{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "toc": true
   },
   "source": [
    "<h1>Table of Contents<span class=\"tocSkip\"></span></h1>\n",
    "<div class=\"toc\"><ul class=\"toc-item\"><li><span><a href=\"#Python-Setup\" data-toc-modified-id=\"Python-Setup-1\"><span class=\"toc-item-num\">1&nbsp;&nbsp;</span>Python Setup</a></span><ul class=\"toc-item\"><li><span><a href=\"#Change-the-width-of-the-page\" data-toc-modified-id=\"Change-the-width-of-the-page-1.1\"><span class=\"toc-item-num\">1.1&nbsp;&nbsp;</span>Change the width of the page</a></span></li><li><span><a href=\"#Import-packages\" data-toc-modified-id=\"Import-packages-1.2\"><span class=\"toc-item-num\">1.2&nbsp;&nbsp;</span>Import packages</a></span></li></ul></li><li><span><a href=\"#Genetic-Code\" data-toc-modified-id=\"Genetic-Code-2\"><span class=\"toc-item-num\">2&nbsp;&nbsp;</span>Genetic Code</a></span></li><li><span><a href=\"#Mutation-rates\" data-toc-modified-id=\"Mutation-rates-3\"><span class=\"toc-item-num\">3&nbsp;&nbsp;</span>Mutation rates</a></span></li><li><span><a href=\"#Functions-for-calculating-mutation-rates-in-DNMT3A\" data-toc-modified-id=\"Functions-for-calculating-mutation-rates-in-DNMT3A-4\"><span class=\"toc-item-num\">4&nbsp;&nbsp;</span>Functions for calculating mutation rates in DNMT3A</a></span><ul class=\"toc-item\"><li><span><a href=\"#Functions-for-calculating-mutation-rate-from-a-csv-file-containing-list-of-variants\" data-toc-modified-id=\"Functions-for-calculating-mutation-rate-from-a-csv-file-containing-list-of-variants-4.1\"><span class=\"toc-item-num\">4.1&nbsp;&nbsp;</span>Functions for calculating mutation rate from a csv file containing list of variants</a></span><ul class=\"toc-item\"><li><span><a href=\"#DNMT3A-transcript-NP072046-=-912-aa--(hg19-or-GRCh38)\" data-toc-modified-id=\"DNMT3A-transcript-NP072046-=-912-aa--(hg19-or-GRCh38)-4.1.1\"><span class=\"toc-item-num\">4.1.1&nbsp;&nbsp;</span>DNMT3A transcript NP072046 = 912 aa  (hg19 or GRCh38)</a></span></li></ul></li><li><span><a href=\"#Functions-for-calculating-mutation-rate-from-a-given-range-of-amino-acid-positions\" data-toc-modified-id=\"Functions-for-calculating-mutation-rate-from-a-given-range-of-amino-acid-positions-4.2\"><span class=\"toc-item-num\">4.2&nbsp;&nbsp;</span>Functions for calculating mutation rate from a given range of amino acid positions</a></span><ul class=\"toc-item\"><li><span><a href=\"#DNMT3A-transcript-NP072046-=-912-aa-(hg19-or-GRCh38)\" data-toc-modified-id=\"DNMT3A-transcript-NP072046-=-912-aa-(hg19-or-GRCh38)-4.2.1\"><span class=\"toc-item-num\">4.2.1&nbsp;&nbsp;</span>DNMT3A transcript NP072046 = 912 aa (hg19 or GRCh38)</a></span></li></ul></li></ul></li><li><span><a href=\"#Calculation-of-mutation-rates-for-each-study\" data-toc-modified-id=\"Calculation-of-mutation-rates-for-each-study-5\"><span class=\"toc-item-num\">5&nbsp;&nbsp;</span>Calculation of mutation rates for each study</a></span><ul class=\"toc-item\"><li><span><a href=\"#Jaiswal-2014\" data-toc-modified-id=\"Jaiswal-2014-5.1\"><span class=\"toc-item-num\">5.1&nbsp;&nbsp;</span>Jaiswal 2014</a></span></li><li><span><a href=\"#Genovese-2014\" data-toc-modified-id=\"Genovese-2014-5.2\"><span class=\"toc-item-num\">5.2&nbsp;&nbsp;</span>Genovese 2014</a></span><ul class=\"toc-item\"><li><ul class=\"toc-item\"><li><span><a href=\"#Complete-codons\" data-toc-modified-id=\"Complete-codons-5.2.0.1\"><span class=\"toc-item-num\">5.2.0.1&nbsp;&nbsp;</span>Complete codons</a></span></li><li><span><a href=\"#Partial-codons\" data-toc-modified-id=\"Partial-codons-5.2.0.2\"><span class=\"toc-item-num\">5.2.0.2&nbsp;&nbsp;</span>Partial codons</a></span></li><li><span><a href=\"#Complete-and-partial-codons-summed\" data-toc-modified-id=\"Complete-and-partial-codons-summed-5.2.0.3\"><span class=\"toc-item-num\">5.2.0.3&nbsp;&nbsp;</span>Complete and partial codons summed</a></span></li></ul></li></ul></li><li><span><a href=\"#McKerrel-2015\" data-toc-modified-id=\"McKerrel-2015-5.3\"><span class=\"toc-item-num\">5.3&nbsp;&nbsp;</span>McKerrel 2015</a></span></li><li><span><a href=\"#Zink-2017\" data-toc-modified-id=\"Zink-2017-5.4\"><span class=\"toc-item-num\">5.4&nbsp;&nbsp;</span>Zink 2017</a></span></li><li><span><a href=\"#Coombs-2017\" data-toc-modified-id=\"Coombs-2017-5.5\"><span class=\"toc-item-num\">5.5&nbsp;&nbsp;</span>Coombs 2017</a></span></li><li><span><a href=\"#Young-2016-&amp;-2019\" data-toc-modified-id=\"Young-2016-&amp;-2019-5.6\"><span class=\"toc-item-num\">5.6&nbsp;&nbsp;</span>Young 2016 &amp; 2019</a></span></li><li><span><a href=\"#Desai-2018\" data-toc-modified-id=\"Desai-2018-5.7\"><span class=\"toc-item-num\">5.7&nbsp;&nbsp;</span>Desai 2018</a></span></li><li><span><a href=\"#Acuna-Hidalgo-2017\" data-toc-modified-id=\"Acuna-Hidalgo-2017-5.8\"><span class=\"toc-item-num\">5.8&nbsp;&nbsp;</span>Acuna-Hidalgo 2017</a></span></li></ul></li></ul></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Python Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Change the width of the page"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>.container { width:80% !important; }</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.core.display import display, HTML\n",
    "display(HTML(\"<style>.container { width:80% !important; }</style>\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/anaconda3/lib/python3.7/site-packages/Bio/codonalign/__init__.py:27: BiopythonExperimentalWarning: Bio.codonalign is an experimental module which may undergo significant changes prior to its future official release.\n",
      "  BiopythonExperimentalWarning)\n"
     ]
    }
   ],
   "source": [
    "# imported packages\n",
    "%matplotlib inline\n",
    "import random\n",
    "import numpy as np\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import math\n",
    "import matplotlib.ticker as plticker\n",
    "import collections\n",
    "from matplotlib.ticker import LinearLocator, FormatStrFormatter\n",
    "from matplotlib.patches import Polygon\n",
    "from matplotlib.patches import Patch\n",
    "from matplotlib.lines import Line2D\n",
    "from matplotlib import cm\n",
    "from mpl_toolkits.mplot3d import Axes3D\n",
    "import scipy.special\n",
    "import scipy.integrate as it\n",
    "from scipy import integrate\n",
    "import copy\n",
    "import glob, os\n",
    "import re\n",
    "from sklearn import datasets, linear_model\n",
    "import pandas as pd\n",
    "from decimal import *\n",
    "from operator import itemgetter    \n",
    "from collections import OrderedDict\n",
    "from matplotlib.ticker import (MultipleLocator, FormatStrFormatter, AutoMinorLocator)\n",
    "\n",
    "from Bio.Alphabet import IUPAC\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.Align import MultipleSeqAlignment\n",
    "from Bio.Alphabet import generic_dna, _ungap\n",
    "from Bio.codonalign.codonalphabet import CodonAlphabet, default_codon_alphabet, default_codon_table \n",
    "from itertools import permutations \n",
    "from math import log \n",
    "from Bio import Entrez"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Genetic Code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "# Amino Acid Code:\n",
    "\n",
    "Amino_acid_abbreviations = {'Ala': 'A', 'Arg': 'R', 'Asn': 'N', 'Asp': 'D',\\\n",
    "                           'Cys': 'C', 'Glu': 'E', 'Gln': 'Q', 'Gly': 'G',\\\n",
    "                           'His': 'H', 'Ile': 'I', 'Leu': 'L', 'Lys': 'K',\\\n",
    "                           'Met': 'M', 'Phe': 'F', 'Pro': 'P', 'Ser': 'S',\\\n",
    "                           'Thr': 'T', 'Trp': 'W', 'Tyr': 'Y', 'Val': 'V', '*': '*'}\n",
    "\n",
    "Genetic_Code = {'TTT': 'Phe', 'TTC': 'Phe', 'TTA': 'Leu', 'TTG': 'Leu',\\\n",
    "               'TCT': 'Ser', 'TCC': 'Ser', 'TCA': 'Ser', 'TCG': 'Ser',\\\n",
    "               'TAT': 'Tyr', 'TAC': 'Tyr', 'TAA': '*', 'TAG': '*',\\\n",
    "               'TGT': 'Cys', 'TGC': 'Cys', 'TGA': '*', 'TGG': 'Trp',\\\n",
    "               'CTT': 'Leu', 'CTC': 'Leu', 'CTA': 'Leu', 'CTG': 'Leu',\\\n",
    "               'CCT': 'Pro', 'CCC': 'Pro', 'CCA': 'Pro', 'CCG': 'Pro',\\\n",
    "               'CAT': 'His', 'CAC': 'His', 'CAA': 'Gln', 'CAG': 'Gln',\\\n",
    "               'CGT': 'Arg', 'CGC': 'Arg', 'CGA': 'Arg', 'CGG': 'Arg',\\\n",
    "               'ATT': 'Ile', 'ATC': 'Ile', 'ATA': 'Ile', 'ATG': 'Met',\\\n",
    "               'ACT': 'Thr', 'ACC': 'Thr', 'ACA': 'Thr', 'ACG': 'Thr',\\\n",
    "               'AAT': 'Asn', 'AAC': 'Asn', 'AAA': 'Lys', 'AAG': 'Lys',\\\n",
    "               'AGT': 'Ser', 'AGC': 'Ser', 'AGA': 'Arg', 'AGG': 'Arg',\\\n",
    "               'GTT': 'Val', 'GTC': 'Val', 'GTA': 'Val', 'GTG': 'Val',\\\n",
    "               'GCT': 'Ala', 'GCC': 'Ala', 'GCA': 'Ala', 'GCG': 'Ala',\\\n",
    "               'GAT': 'Asp', 'GAC': 'Asp', 'GAA': 'Glu', 'GAG': 'Glu',\\\n",
    "               'GGT': 'Gly', 'GGC': 'Gly', 'GGA': 'Gly', 'GGG': 'Gly'}\n",
    "\n",
    "base_alternative = {'A': ['T', 'C', 'G'], 'T': ['A', 'C', 'G'], 'G':['A', 'C', 'T'], 'C':['A', 'G', 'T']}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "Possible_AA_Changes = {}\n",
    "Possible_Base_Changes = {}\n",
    "Possible_Codon_Changes = {}\n",
    "for k, v in Genetic_Code.items():\n",
    "    codon = k\n",
    "    base_1 = str(codon[0])\n",
    "    base_2 = str(codon[1])\n",
    "    base_3 = str(codon[2])\n",
    "    new_codons = []\n",
    "    amino_acid_changes = []\n",
    "    base_changes = []\n",
    "    if base_1 in base_alternative.keys():\n",
    "        alternate_base = base_alternative[base_1]\n",
    "        for i in alternate_base:\n",
    "            base_changes.append(i)\n",
    "            new_codon = i+base_2+base_3\n",
    "            new_codons.append(new_codon)\n",
    "    if base_2 in base_alternative.keys():\n",
    "        alternate_base = base_alternative[base_2]\n",
    "        for i in alternate_base:\n",
    "            base_changes.append(i)\n",
    "            new_codon = base_1+i+base_3\n",
    "            new_codons.append(new_codon)\n",
    "    if base_3 in base_alternative.keys():\n",
    "        alternate_base = base_alternative[base_3]\n",
    "        for i in alternate_base:\n",
    "            base_changes.append(i)\n",
    "            new_codon = base_1+base_2+i\n",
    "            new_codons.append(new_codon)\n",
    "    Possible_Codon_Changes[k]=(new_codons[0],new_codons[1], new_codons[2], new_codons[3], new_codons[4],\\\n",
    "                               new_codons[5], new_codons[6], new_codons[7], new_codons[8]) #i.e. {'TTT': ['ATT', 'CTT', 'GTT'...etc]}\n",
    "    Possible_Base_Changes[k]=(base_changes[0], base_changes[1], base_changes[2], base_changes[3], base_changes[4],\\\n",
    "                              base_changes[5], base_changes[6], base_changes[7], base_changes[8]) #i.e. {'TTT': ['A', 'C', 'G'... etc]}\n",
    "    \n",
    "Possible_Amino_Acid_Changes = {}\n",
    "for k, v in Possible_Codon_Changes.items():\n",
    "    codon = k\n",
    "    codon_changes = v\n",
    "    amino_acid_changes = []\n",
    "    for i in codon_changes:\n",
    "        amino_acid_change = Genetic_Code[i]\n",
    "        amino_acid_changes.append(amino_acid_change)\n",
    "    Possible_Amino_Acid_Changes[k]=(amino_acid_changes[0], amino_acid_changes[1], amino_acid_changes[2],\\\n",
    "                                    amino_acid_changes[3], amino_acid_changes[4], amino_acid_changes[5],\\\n",
    "                                    amino_acid_changes[6], amino_acid_changes[7], amino_acid_changes[8]) #i.e. {'TTT': ['Ile', 'Leu', 'Val'...etc]}\n",
    "    \n",
    "Possible_Amino_Acid_Changes_letters = {}\n",
    "for k, v in Possible_Amino_Acid_Changes.items():\n",
    "    codon = k\n",
    "    codon_changes = v\n",
    "    amino_acid_changes_letters = []\n",
    "    for i in codon_changes:\n",
    "        amino_acid_change_letter = Amino_acid_abbreviations[i]\n",
    "        amino_acid_changes_letters.append(amino_acid_change_letter)\n",
    "    Possible_Amino_Acid_Changes_letters[k]=(amino_acid_changes_letters[0], amino_acid_changes_letters[1],\\\n",
    "                                            amino_acid_changes_letters[2], amino_acid_changes_letters[3],\\\n",
    "                                            amino_acid_changes_letters[4], amino_acid_changes_letters[5],\\\n",
    "                                            amino_acid_changes_letters[6], amino_acid_changes_letters[7],\\\n",
    "                                            amino_acid_changes_letters[8]) #i.e. {'TTT': ['I', 'L', 'V'...etc]}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'TTT': ('ATT', 'CTT', 'GTT', 'TAT', 'TCT', 'TGT', 'TTA', 'TTC', 'TTG'),\n",
       " 'TTC': ('ATC', 'CTC', 'GTC', 'TAC', 'TCC', 'TGC', 'TTA', 'TTG', 'TTT'),\n",
       " 'TTA': ('ATA', 'CTA', 'GTA', 'TAA', 'TCA', 'TGA', 'TTT', 'TTC', 'TTG'),\n",
       " 'TTG': ('ATG', 'CTG', 'GTG', 'TAG', 'TCG', 'TGG', 'TTA', 'TTC', 'TTT'),\n",
       " 'TCT': ('ACT', 'CCT', 'GCT', 'TAT', 'TGT', 'TTT', 'TCA', 'TCC', 'TCG'),\n",
       " 'TCC': ('ACC', 'CCC', 'GCC', 'TAC', 'TGC', 'TTC', 'TCA', 'TCG', 'TCT'),\n",
       " 'TCA': ('ACA', 'CCA', 'GCA', 'TAA', 'TGA', 'TTA', 'TCT', 'TCC', 'TCG'),\n",
       " 'TCG': ('ACG', 'CCG', 'GCG', 'TAG', 'TGG', 'TTG', 'TCA', 'TCC', 'TCT'),\n",
       " 'TAT': ('AAT', 'CAT', 'GAT', 'TTT', 'TCT', 'TGT', 'TAA', 'TAC', 'TAG'),\n",
       " 'TAC': ('AAC', 'CAC', 'GAC', 'TTC', 'TCC', 'TGC', 'TAA', 'TAG', 'TAT'),\n",
       " 'TAA': ('AAA', 'CAA', 'GAA', 'TTA', 'TCA', 'TGA', 'TAT', 'TAC', 'TAG'),\n",
       " 'TAG': ('AAG', 'CAG', 'GAG', 'TTG', 'TCG', 'TGG', 'TAA', 'TAC', 'TAT'),\n",
       " 'TGT': ('AGT', 'CGT', 'GGT', 'TAT', 'TCT', 'TTT', 'TGA', 'TGC', 'TGG'),\n",
       " 'TGC': ('AGC', 'CGC', 'GGC', 'TAC', 'TCC', 'TTC', 'TGA', 'TGG', 'TGT'),\n",
       " 'TGA': ('AGA', 'CGA', 'GGA', 'TAA', 'TCA', 'TTA', 'TGT', 'TGC', 'TGG'),\n",
       " 'TGG': ('AGG', 'CGG', 'GGG', 'TAG', 'TCG', 'TTG', 'TGA', 'TGC', 'TGT'),\n",
       " 'CTT': ('ATT', 'GTT', 'TTT', 'CAT', 'CCT', 'CGT', 'CTA', 'CTC', 'CTG'),\n",
       " 'CTC': ('ATC', 'GTC', 'TTC', 'CAC', 'CCC', 'CGC', 'CTA', 'CTG', 'CTT'),\n",
       " 'CTA': ('ATA', 'GTA', 'TTA', 'CAA', 'CCA', 'CGA', 'CTT', 'CTC', 'CTG'),\n",
       " 'CTG': ('ATG', 'GTG', 'TTG', 'CAG', 'CCG', 'CGG', 'CTA', 'CTC', 'CTT'),\n",
       " 'CCT': ('ACT', 'GCT', 'TCT', 'CAT', 'CGT', 'CTT', 'CCA', 'CCC', 'CCG'),\n",
       " 'CCC': ('ACC', 'GCC', 'TCC', 'CAC', 'CGC', 'CTC', 'CCA', 'CCG', 'CCT'),\n",
       " 'CCA': ('ACA', 'GCA', 'TCA', 'CAA', 'CGA', 'CTA', 'CCT', 'CCC', 'CCG'),\n",
       " 'CCG': ('ACG', 'GCG', 'TCG', 'CAG', 'CGG', 'CTG', 'CCA', 'CCC', 'CCT'),\n",
       " 'CAT': ('AAT', 'GAT', 'TAT', 'CTT', 'CCT', 'CGT', 'CAA', 'CAC', 'CAG'),\n",
       " 'CAC': ('AAC', 'GAC', 'TAC', 'CTC', 'CCC', 'CGC', 'CAA', 'CAG', 'CAT'),\n",
       " 'CAA': ('AAA', 'GAA', 'TAA', 'CTA', 'CCA', 'CGA', 'CAT', 'CAC', 'CAG'),\n",
       " 'CAG': ('AAG', 'GAG', 'TAG', 'CTG', 'CCG', 'CGG', 'CAA', 'CAC', 'CAT'),\n",
       " 'CGT': ('AGT', 'GGT', 'TGT', 'CAT', 'CCT', 'CTT', 'CGA', 'CGC', 'CGG'),\n",
       " 'CGC': ('AGC', 'GGC', 'TGC', 'CAC', 'CCC', 'CTC', 'CGA', 'CGG', 'CGT'),\n",
       " 'CGA': ('AGA', 'GGA', 'TGA', 'CAA', 'CCA', 'CTA', 'CGT', 'CGC', 'CGG'),\n",
       " 'CGG': ('AGG', 'GGG', 'TGG', 'CAG', 'CCG', 'CTG', 'CGA', 'CGC', 'CGT'),\n",
       " 'ATT': ('TTT', 'CTT', 'GTT', 'AAT', 'ACT', 'AGT', 'ATA', 'ATC', 'ATG'),\n",
       " 'ATC': ('TTC', 'CTC', 'GTC', 'AAC', 'ACC', 'AGC', 'ATA', 'ATG', 'ATT'),\n",
       " 'ATA': ('TTA', 'CTA', 'GTA', 'AAA', 'ACA', 'AGA', 'ATT', 'ATC', 'ATG'),\n",
       " 'ATG': ('TTG', 'CTG', 'GTG', 'AAG', 'ACG', 'AGG', 'ATA', 'ATC', 'ATT'),\n",
       " 'ACT': ('TCT', 'CCT', 'GCT', 'AAT', 'AGT', 'ATT', 'ACA', 'ACC', 'ACG'),\n",
       " 'ACC': ('TCC', 'CCC', 'GCC', 'AAC', 'AGC', 'ATC', 'ACA', 'ACG', 'ACT'),\n",
       " 'ACA': ('TCA', 'CCA', 'GCA', 'AAA', 'AGA', 'ATA', 'ACT', 'ACC', 'ACG'),\n",
       " 'ACG': ('TCG', 'CCG', 'GCG', 'AAG', 'AGG', 'ATG', 'ACA', 'ACC', 'ACT'),\n",
       " 'AAT': ('TAT', 'CAT', 'GAT', 'ATT', 'ACT', 'AGT', 'AAA', 'AAC', 'AAG'),\n",
       " 'AAC': ('TAC', 'CAC', 'GAC', 'ATC', 'ACC', 'AGC', 'AAA', 'AAG', 'AAT'),\n",
       " 'AAA': ('TAA', 'CAA', 'GAA', 'ATA', 'ACA', 'AGA', 'AAT', 'AAC', 'AAG'),\n",
       " 'AAG': ('TAG', 'CAG', 'GAG', 'ATG', 'ACG', 'AGG', 'AAA', 'AAC', 'AAT'),\n",
       " 'AGT': ('TGT', 'CGT', 'GGT', 'AAT', 'ACT', 'ATT', 'AGA', 'AGC', 'AGG'),\n",
       " 'AGC': ('TGC', 'CGC', 'GGC', 'AAC', 'ACC', 'ATC', 'AGA', 'AGG', 'AGT'),\n",
       " 'AGA': ('TGA', 'CGA', 'GGA', 'AAA', 'ACA', 'ATA', 'AGT', 'AGC', 'AGG'),\n",
       " 'AGG': ('TGG', 'CGG', 'GGG', 'AAG', 'ACG', 'ATG', 'AGA', 'AGC', 'AGT'),\n",
       " 'GTT': ('ATT', 'CTT', 'TTT', 'GAT', 'GCT', 'GGT', 'GTA', 'GTC', 'GTG'),\n",
       " 'GTC': ('ATC', 'CTC', 'TTC', 'GAC', 'GCC', 'GGC', 'GTA', 'GTG', 'GTT'),\n",
       " 'GTA': ('ATA', 'CTA', 'TTA', 'GAA', 'GCA', 'GGA', 'GTT', 'GTC', 'GTG'),\n",
       " 'GTG': ('ATG', 'CTG', 'TTG', 'GAG', 'GCG', 'GGG', 'GTA', 'GTC', 'GTT'),\n",
       " 'GCT': ('ACT', 'CCT', 'TCT', 'GAT', 'GGT', 'GTT', 'GCA', 'GCC', 'GCG'),\n",
       " 'GCC': ('ACC', 'CCC', 'TCC', 'GAC', 'GGC', 'GTC', 'GCA', 'GCG', 'GCT'),\n",
       " 'GCA': ('ACA', 'CCA', 'TCA', 'GAA', 'GGA', 'GTA', 'GCT', 'GCC', 'GCG'),\n",
       " 'GCG': ('ACG', 'CCG', 'TCG', 'GAG', 'GGG', 'GTG', 'GCA', 'GCC', 'GCT'),\n",
       " 'GAT': ('AAT', 'CAT', 'TAT', 'GTT', 'GCT', 'GGT', 'GAA', 'GAC', 'GAG'),\n",
       " 'GAC': ('AAC', 'CAC', 'TAC', 'GTC', 'GCC', 'GGC', 'GAA', 'GAG', 'GAT'),\n",
       " 'GAA': ('AAA', 'CAA', 'TAA', 'GTA', 'GCA', 'GGA', 'GAT', 'GAC', 'GAG'),\n",
       " 'GAG': ('AAG', 'CAG', 'TAG', 'GTG', 'GCG', 'GGG', 'GAA', 'GAC', 'GAT'),\n",
       " 'GGT': ('AGT', 'CGT', 'TGT', 'GAT', 'GCT', 'GTT', 'GGA', 'GGC', 'GGG'),\n",
       " 'GGC': ('AGC', 'CGC', 'TGC', 'GAC', 'GCC', 'GTC', 'GGA', 'GGG', 'GGT'),\n",
       " 'GGA': ('AGA', 'CGA', 'TGA', 'GAA', 'GCA', 'GTA', 'GGT', 'GGC', 'GGG'),\n",
       " 'GGG': ('AGG', 'CGG', 'TGG', 'GAG', 'GCG', 'GTG', 'GGA', 'GGC', 'GGT')}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Possible_Codon_Changes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'TTT': ('A', 'C', 'G', 'A', 'C', 'G', 'A', 'C', 'G'),\n",
       " 'TTC': ('A', 'C', 'G', 'A', 'C', 'G', 'A', 'G', 'T'),\n",
       " 'TTA': ('A', 'C', 'G', 'A', 'C', 'G', 'T', 'C', 'G'),\n",
       " 'TTG': ('A', 'C', 'G', 'A', 'C', 'G', 'A', 'C', 'T'),\n",
       " 'TCT': ('A', 'C', 'G', 'A', 'G', 'T', 'A', 'C', 'G'),\n",
       " 'TCC': ('A', 'C', 'G', 'A', 'G', 'T', 'A', 'G', 'T'),\n",
       " 'TCA': ('A', 'C', 'G', 'A', 'G', 'T', 'T', 'C', 'G'),\n",
       " 'TCG': ('A', 'C', 'G', 'A', 'G', 'T', 'A', 'C', 'T'),\n",
       " 'TAT': ('A', 'C', 'G', 'T', 'C', 'G', 'A', 'C', 'G'),\n",
       " 'TAC': ('A', 'C', 'G', 'T', 'C', 'G', 'A', 'G', 'T'),\n",
       " 'TAA': ('A', 'C', 'G', 'T', 'C', 'G', 'T', 'C', 'G'),\n",
       " 'TAG': ('A', 'C', 'G', 'T', 'C', 'G', 'A', 'C', 'T'),\n",
       " 'TGT': ('A', 'C', 'G', 'A', 'C', 'T', 'A', 'C', 'G'),\n",
       " 'TGC': ('A', 'C', 'G', 'A', 'C', 'T', 'A', 'G', 'T'),\n",
       " 'TGA': ('A', 'C', 'G', 'A', 'C', 'T', 'T', 'C', 'G'),\n",
       " 'TGG': ('A', 'C', 'G', 'A', 'C', 'T', 'A', 'C', 'T'),\n",
       " 'CTT': ('A', 'G', 'T', 'A', 'C', 'G', 'A', 'C', 'G'),\n",
       " 'CTC': ('A', 'G', 'T', 'A', 'C', 'G', 'A', 'G', 'T'),\n",
       " 'CTA': ('A', 'G', 'T', 'A', 'C', 'G', 'T', 'C', 'G'),\n",
       " 'CTG': ('A', 'G', 'T', 'A', 'C', 'G', 'A', 'C', 'T'),\n",
       " 'CCT': ('A', 'G', 'T', 'A', 'G', 'T', 'A', 'C', 'G'),\n",
       " 'CCC': ('A', 'G', 'T', 'A', 'G', 'T', 'A', 'G', 'T'),\n",
       " 'CCA': ('A', 'G', 'T', 'A', 'G', 'T', 'T', 'C', 'G'),\n",
       " 'CCG': ('A', 'G', 'T', 'A', 'G', 'T', 'A', 'C', 'T'),\n",
       " 'CAT': ('A', 'G', 'T', 'T', 'C', 'G', 'A', 'C', 'G'),\n",
       " 'CAC': ('A', 'G', 'T', 'T', 'C', 'G', 'A', 'G', 'T'),\n",
       " 'CAA': ('A', 'G', 'T', 'T', 'C', 'G', 'T', 'C', 'G'),\n",
       " 'CAG': ('A', 'G', 'T', 'T', 'C', 'G', 'A', 'C', 'T'),\n",
       " 'CGT': ('A', 'G', 'T', 'A', 'C', 'T', 'A', 'C', 'G'),\n",
       " 'CGC': ('A', 'G', 'T', 'A', 'C', 'T', 'A', 'G', 'T'),\n",
       " 'CGA': ('A', 'G', 'T', 'A', 'C', 'T', 'T', 'C', 'G'),\n",
       " 'CGG': ('A', 'G', 'T', 'A', 'C', 'T', 'A', 'C', 'T'),\n",
       " 'ATT': ('T', 'C', 'G', 'A', 'C', 'G', 'A', 'C', 'G'),\n",
       " 'ATC': ('T', 'C', 'G', 'A', 'C', 'G', 'A', 'G', 'T'),\n",
       " 'ATA': ('T', 'C', 'G', 'A', 'C', 'G', 'T', 'C', 'G'),\n",
       " 'ATG': ('T', 'C', 'G', 'A', 'C', 'G', 'A', 'C', 'T'),\n",
       " 'ACT': ('T', 'C', 'G', 'A', 'G', 'T', 'A', 'C', 'G'),\n",
       " 'ACC': ('T', 'C', 'G', 'A', 'G', 'T', 'A', 'G', 'T'),\n",
       " 'ACA': ('T', 'C', 'G', 'A', 'G', 'T', 'T', 'C', 'G'),\n",
       " 'ACG': ('T', 'C', 'G', 'A', 'G', 'T', 'A', 'C', 'T'),\n",
       " 'AAT': ('T', 'C', 'G', 'T', 'C', 'G', 'A', 'C', 'G'),\n",
       " 'AAC': ('T', 'C', 'G', 'T', 'C', 'G', 'A', 'G', 'T'),\n",
       " 'AAA': ('T', 'C', 'G', 'T', 'C', 'G', 'T', 'C', 'G'),\n",
       " 'AAG': ('T', 'C', 'G', 'T', 'C', 'G', 'A', 'C', 'T'),\n",
       " 'AGT': ('T', 'C', 'G', 'A', 'C', 'T', 'A', 'C', 'G'),\n",
       " 'AGC': ('T', 'C', 'G', 'A', 'C', 'T', 'A', 'G', 'T'),\n",
       " 'AGA': ('T', 'C', 'G', 'A', 'C', 'T', 'T', 'C', 'G'),\n",
       " 'AGG': ('T', 'C', 'G', 'A', 'C', 'T', 'A', 'C', 'T'),\n",
       " 'GTT': ('A', 'C', 'T', 'A', 'C', 'G', 'A', 'C', 'G'),\n",
       " 'GTC': ('A', 'C', 'T', 'A', 'C', 'G', 'A', 'G', 'T'),\n",
       " 'GTA': ('A', 'C', 'T', 'A', 'C', 'G', 'T', 'C', 'G'),\n",
       " 'GTG': ('A', 'C', 'T', 'A', 'C', 'G', 'A', 'C', 'T'),\n",
       " 'GCT': ('A', 'C', 'T', 'A', 'G', 'T', 'A', 'C', 'G'),\n",
       " 'GCC': ('A', 'C', 'T', 'A', 'G', 'T', 'A', 'G', 'T'),\n",
       " 'GCA': ('A', 'C', 'T', 'A', 'G', 'T', 'T', 'C', 'G'),\n",
       " 'GCG': ('A', 'C', 'T', 'A', 'G', 'T', 'A', 'C', 'T'),\n",
       " 'GAT': ('A', 'C', 'T', 'T', 'C', 'G', 'A', 'C', 'G'),\n",
       " 'GAC': ('A', 'C', 'T', 'T', 'C', 'G', 'A', 'G', 'T'),\n",
       " 'GAA': ('A', 'C', 'T', 'T', 'C', 'G', 'T', 'C', 'G'),\n",
       " 'GAG': ('A', 'C', 'T', 'T', 'C', 'G', 'A', 'C', 'T'),\n",
       " 'GGT': ('A', 'C', 'T', 'A', 'C', 'T', 'A', 'C', 'G'),\n",
       " 'GGC': ('A', 'C', 'T', 'A', 'C', 'T', 'A', 'G', 'T'),\n",
       " 'GGA': ('A', 'C', 'T', 'A', 'C', 'T', 'T', 'C', 'G'),\n",
       " 'GGG': ('A', 'C', 'T', 'A', 'C', 'T', 'A', 'C', 'T')}"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Possible_Base_Changes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'TTT': ('Ile', 'Leu', 'Val', 'Tyr', 'Ser', 'Cys', 'Leu', 'Phe', 'Leu'),\n",
       " 'TTC': ('Ile', 'Leu', 'Val', 'Tyr', 'Ser', 'Cys', 'Leu', 'Leu', 'Phe'),\n",
       " 'TTA': ('Ile', 'Leu', 'Val', '*', 'Ser', '*', 'Phe', 'Phe', 'Leu'),\n",
       " 'TTG': ('Met', 'Leu', 'Val', '*', 'Ser', 'Trp', 'Leu', 'Phe', 'Phe'),\n",
       " 'TCT': ('Thr', 'Pro', 'Ala', 'Tyr', 'Cys', 'Phe', 'Ser', 'Ser', 'Ser'),\n",
       " 'TCC': ('Thr', 'Pro', 'Ala', 'Tyr', 'Cys', 'Phe', 'Ser', 'Ser', 'Ser'),\n",
       " 'TCA': ('Thr', 'Pro', 'Ala', '*', '*', 'Leu', 'Ser', 'Ser', 'Ser'),\n",
       " 'TCG': ('Thr', 'Pro', 'Ala', '*', 'Trp', 'Leu', 'Ser', 'Ser', 'Ser'),\n",
       " 'TAT': ('Asn', 'His', 'Asp', 'Phe', 'Ser', 'Cys', '*', 'Tyr', '*'),\n",
       " 'TAC': ('Asn', 'His', 'Asp', 'Phe', 'Ser', 'Cys', '*', '*', 'Tyr'),\n",
       " 'TAA': ('Lys', 'Gln', 'Glu', 'Leu', 'Ser', '*', 'Tyr', 'Tyr', '*'),\n",
       " 'TAG': ('Lys', 'Gln', 'Glu', 'Leu', 'Ser', 'Trp', '*', 'Tyr', 'Tyr'),\n",
       " 'TGT': ('Ser', 'Arg', 'Gly', 'Tyr', 'Ser', 'Phe', '*', 'Cys', 'Trp'),\n",
       " 'TGC': ('Ser', 'Arg', 'Gly', 'Tyr', 'Ser', 'Phe', '*', 'Trp', 'Cys'),\n",
       " 'TGA': ('Arg', 'Arg', 'Gly', '*', 'Ser', 'Leu', 'Cys', 'Cys', 'Trp'),\n",
       " 'TGG': ('Arg', 'Arg', 'Gly', '*', 'Ser', 'Leu', '*', 'Cys', 'Cys'),\n",
       " 'CTT': ('Ile', 'Val', 'Phe', 'His', 'Pro', 'Arg', 'Leu', 'Leu', 'Leu'),\n",
       " 'CTC': ('Ile', 'Val', 'Phe', 'His', 'Pro', 'Arg', 'Leu', 'Leu', 'Leu'),\n",
       " 'CTA': ('Ile', 'Val', 'Leu', 'Gln', 'Pro', 'Arg', 'Leu', 'Leu', 'Leu'),\n",
       " 'CTG': ('Met', 'Val', 'Leu', 'Gln', 'Pro', 'Arg', 'Leu', 'Leu', 'Leu'),\n",
       " 'CCT': ('Thr', 'Ala', 'Ser', 'His', 'Arg', 'Leu', 'Pro', 'Pro', 'Pro'),\n",
       " 'CCC': ('Thr', 'Ala', 'Ser', 'His', 'Arg', 'Leu', 'Pro', 'Pro', 'Pro'),\n",
       " 'CCA': ('Thr', 'Ala', 'Ser', 'Gln', 'Arg', 'Leu', 'Pro', 'Pro', 'Pro'),\n",
       " 'CCG': ('Thr', 'Ala', 'Ser', 'Gln', 'Arg', 'Leu', 'Pro', 'Pro', 'Pro'),\n",
       " 'CAT': ('Asn', 'Asp', 'Tyr', 'Leu', 'Pro', 'Arg', 'Gln', 'His', 'Gln'),\n",
       " 'CAC': ('Asn', 'Asp', 'Tyr', 'Leu', 'Pro', 'Arg', 'Gln', 'Gln', 'His'),\n",
       " 'CAA': ('Lys', 'Glu', '*', 'Leu', 'Pro', 'Arg', 'His', 'His', 'Gln'),\n",
       " 'CAG': ('Lys', 'Glu', '*', 'Leu', 'Pro', 'Arg', 'Gln', 'His', 'His'),\n",
       " 'CGT': ('Ser', 'Gly', 'Cys', 'His', 'Pro', 'Leu', 'Arg', 'Arg', 'Arg'),\n",
       " 'CGC': ('Ser', 'Gly', 'Cys', 'His', 'Pro', 'Leu', 'Arg', 'Arg', 'Arg'),\n",
       " 'CGA': ('Arg', 'Gly', '*', 'Gln', 'Pro', 'Leu', 'Arg', 'Arg', 'Arg'),\n",
       " 'CGG': ('Arg', 'Gly', 'Trp', 'Gln', 'Pro', 'Leu', 'Arg', 'Arg', 'Arg'),\n",
       " 'ATT': ('Phe', 'Leu', 'Val', 'Asn', 'Thr', 'Ser', 'Ile', 'Ile', 'Met'),\n",
       " 'ATC': ('Phe', 'Leu', 'Val', 'Asn', 'Thr', 'Ser', 'Ile', 'Met', 'Ile'),\n",
       " 'ATA': ('Leu', 'Leu', 'Val', 'Lys', 'Thr', 'Arg', 'Ile', 'Ile', 'Met'),\n",
       " 'ATG': ('Leu', 'Leu', 'Val', 'Lys', 'Thr', 'Arg', 'Ile', 'Ile', 'Ile'),\n",
       " 'ACT': ('Ser', 'Pro', 'Ala', 'Asn', 'Ser', 'Ile', 'Thr', 'Thr', 'Thr'),\n",
       " 'ACC': ('Ser', 'Pro', 'Ala', 'Asn', 'Ser', 'Ile', 'Thr', 'Thr', 'Thr'),\n",
       " 'ACA': ('Ser', 'Pro', 'Ala', 'Lys', 'Arg', 'Ile', 'Thr', 'Thr', 'Thr'),\n",
       " 'ACG': ('Ser', 'Pro', 'Ala', 'Lys', 'Arg', 'Met', 'Thr', 'Thr', 'Thr'),\n",
       " 'AAT': ('Tyr', 'His', 'Asp', 'Ile', 'Thr', 'Ser', 'Lys', 'Asn', 'Lys'),\n",
       " 'AAC': ('Tyr', 'His', 'Asp', 'Ile', 'Thr', 'Ser', 'Lys', 'Lys', 'Asn'),\n",
       " 'AAA': ('*', 'Gln', 'Glu', 'Ile', 'Thr', 'Arg', 'Asn', 'Asn', 'Lys'),\n",
       " 'AAG': ('*', 'Gln', 'Glu', 'Met', 'Thr', 'Arg', 'Lys', 'Asn', 'Asn'),\n",
       " 'AGT': ('Cys', 'Arg', 'Gly', 'Asn', 'Thr', 'Ile', 'Arg', 'Ser', 'Arg'),\n",
       " 'AGC': ('Cys', 'Arg', 'Gly', 'Asn', 'Thr', 'Ile', 'Arg', 'Arg', 'Ser'),\n",
       " 'AGA': ('*', 'Arg', 'Gly', 'Lys', 'Thr', 'Ile', 'Ser', 'Ser', 'Arg'),\n",
       " 'AGG': ('Trp', 'Arg', 'Gly', 'Lys', 'Thr', 'Met', 'Arg', 'Ser', 'Ser'),\n",
       " 'GTT': ('Ile', 'Leu', 'Phe', 'Asp', 'Ala', 'Gly', 'Val', 'Val', 'Val'),\n",
       " 'GTC': ('Ile', 'Leu', 'Phe', 'Asp', 'Ala', 'Gly', 'Val', 'Val', 'Val'),\n",
       " 'GTA': ('Ile', 'Leu', 'Leu', 'Glu', 'Ala', 'Gly', 'Val', 'Val', 'Val'),\n",
       " 'GTG': ('Met', 'Leu', 'Leu', 'Glu', 'Ala', 'Gly', 'Val', 'Val', 'Val'),\n",
       " 'GCT': ('Thr', 'Pro', 'Ser', 'Asp', 'Gly', 'Val', 'Ala', 'Ala', 'Ala'),\n",
       " 'GCC': ('Thr', 'Pro', 'Ser', 'Asp', 'Gly', 'Val', 'Ala', 'Ala', 'Ala'),\n",
       " 'GCA': ('Thr', 'Pro', 'Ser', 'Glu', 'Gly', 'Val', 'Ala', 'Ala', 'Ala'),\n",
       " 'GCG': ('Thr', 'Pro', 'Ser', 'Glu', 'Gly', 'Val', 'Ala', 'Ala', 'Ala'),\n",
       " 'GAT': ('Asn', 'His', 'Tyr', 'Val', 'Ala', 'Gly', 'Glu', 'Asp', 'Glu'),\n",
       " 'GAC': ('Asn', 'His', 'Tyr', 'Val', 'Ala', 'Gly', 'Glu', 'Glu', 'Asp'),\n",
       " 'GAA': ('Lys', 'Gln', '*', 'Val', 'Ala', 'Gly', 'Asp', 'Asp', 'Glu'),\n",
       " 'GAG': ('Lys', 'Gln', '*', 'Val', 'Ala', 'Gly', 'Glu', 'Asp', 'Asp'),\n",
       " 'GGT': ('Ser', 'Arg', 'Cys', 'Asp', 'Ala', 'Val', 'Gly', 'Gly', 'Gly'),\n",
       " 'GGC': ('Ser', 'Arg', 'Cys', 'Asp', 'Ala', 'Val', 'Gly', 'Gly', 'Gly'),\n",
       " 'GGA': ('Arg', 'Arg', '*', 'Glu', 'Ala', 'Val', 'Gly', 'Gly', 'Gly'),\n",
       " 'GGG': ('Arg', 'Arg', 'Trp', 'Glu', 'Ala', 'Val', 'Gly', 'Gly', 'Gly')}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Possible_Amino_Acid_Changes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'TTT': ('I', 'L', 'V', 'Y', 'S', 'C', 'L', 'F', 'L'),\n",
       " 'TTC': ('I', 'L', 'V', 'Y', 'S', 'C', 'L', 'L', 'F'),\n",
       " 'TTA': ('I', 'L', 'V', '*', 'S', '*', 'F', 'F', 'L'),\n",
       " 'TTG': ('M', 'L', 'V', '*', 'S', 'W', 'L', 'F', 'F'),\n",
       " 'TCT': ('T', 'P', 'A', 'Y', 'C', 'F', 'S', 'S', 'S'),\n",
       " 'TCC': ('T', 'P', 'A', 'Y', 'C', 'F', 'S', 'S', 'S'),\n",
       " 'TCA': ('T', 'P', 'A', '*', '*', 'L', 'S', 'S', 'S'),\n",
       " 'TCG': ('T', 'P', 'A', '*', 'W', 'L', 'S', 'S', 'S'),\n",
       " 'TAT': ('N', 'H', 'D', 'F', 'S', 'C', '*', 'Y', '*'),\n",
       " 'TAC': ('N', 'H', 'D', 'F', 'S', 'C', '*', '*', 'Y'),\n",
       " 'TAA': ('K', 'Q', 'E', 'L', 'S', '*', 'Y', 'Y', '*'),\n",
       " 'TAG': ('K', 'Q', 'E', 'L', 'S', 'W', '*', 'Y', 'Y'),\n",
       " 'TGT': ('S', 'R', 'G', 'Y', 'S', 'F', '*', 'C', 'W'),\n",
       " 'TGC': ('S', 'R', 'G', 'Y', 'S', 'F', '*', 'W', 'C'),\n",
       " 'TGA': ('R', 'R', 'G', '*', 'S', 'L', 'C', 'C', 'W'),\n",
       " 'TGG': ('R', 'R', 'G', '*', 'S', 'L', '*', 'C', 'C'),\n",
       " 'CTT': ('I', 'V', 'F', 'H', 'P', 'R', 'L', 'L', 'L'),\n",
       " 'CTC': ('I', 'V', 'F', 'H', 'P', 'R', 'L', 'L', 'L'),\n",
       " 'CTA': ('I', 'V', 'L', 'Q', 'P', 'R', 'L', 'L', 'L'),\n",
       " 'CTG': ('M', 'V', 'L', 'Q', 'P', 'R', 'L', 'L', 'L'),\n",
       " 'CCT': ('T', 'A', 'S', 'H', 'R', 'L', 'P', 'P', 'P'),\n",
       " 'CCC': ('T', 'A', 'S', 'H', 'R', 'L', 'P', 'P', 'P'),\n",
       " 'CCA': ('T', 'A', 'S', 'Q', 'R', 'L', 'P', 'P', 'P'),\n",
       " 'CCG': ('T', 'A', 'S', 'Q', 'R', 'L', 'P', 'P', 'P'),\n",
       " 'CAT': ('N', 'D', 'Y', 'L', 'P', 'R', 'Q', 'H', 'Q'),\n",
       " 'CAC': ('N', 'D', 'Y', 'L', 'P', 'R', 'Q', 'Q', 'H'),\n",
       " 'CAA': ('K', 'E', '*', 'L', 'P', 'R', 'H', 'H', 'Q'),\n",
       " 'CAG': ('K', 'E', '*', 'L', 'P', 'R', 'Q', 'H', 'H'),\n",
       " 'CGT': ('S', 'G', 'C', 'H', 'P', 'L', 'R', 'R', 'R'),\n",
       " 'CGC': ('S', 'G', 'C', 'H', 'P', 'L', 'R', 'R', 'R'),\n",
       " 'CGA': ('R', 'G', '*', 'Q', 'P', 'L', 'R', 'R', 'R'),\n",
       " 'CGG': ('R', 'G', 'W', 'Q', 'P', 'L', 'R', 'R', 'R'),\n",
       " 'ATT': ('F', 'L', 'V', 'N', 'T', 'S', 'I', 'I', 'M'),\n",
       " 'ATC': ('F', 'L', 'V', 'N', 'T', 'S', 'I', 'M', 'I'),\n",
       " 'ATA': ('L', 'L', 'V', 'K', 'T', 'R', 'I', 'I', 'M'),\n",
       " 'ATG': ('L', 'L', 'V', 'K', 'T', 'R', 'I', 'I', 'I'),\n",
       " 'ACT': ('S', 'P', 'A', 'N', 'S', 'I', 'T', 'T', 'T'),\n",
       " 'ACC': ('S', 'P', 'A', 'N', 'S', 'I', 'T', 'T', 'T'),\n",
       " 'ACA': ('S', 'P', 'A', 'K', 'R', 'I', 'T', 'T', 'T'),\n",
       " 'ACG': ('S', 'P', 'A', 'K', 'R', 'M', 'T', 'T', 'T'),\n",
       " 'AAT': ('Y', 'H', 'D', 'I', 'T', 'S', 'K', 'N', 'K'),\n",
       " 'AAC': ('Y', 'H', 'D', 'I', 'T', 'S', 'K', 'K', 'N'),\n",
       " 'AAA': ('*', 'Q', 'E', 'I', 'T', 'R', 'N', 'N', 'K'),\n",
       " 'AAG': ('*', 'Q', 'E', 'M', 'T', 'R', 'K', 'N', 'N'),\n",
       " 'AGT': ('C', 'R', 'G', 'N', 'T', 'I', 'R', 'S', 'R'),\n",
       " 'AGC': ('C', 'R', 'G', 'N', 'T', 'I', 'R', 'R', 'S'),\n",
       " 'AGA': ('*', 'R', 'G', 'K', 'T', 'I', 'S', 'S', 'R'),\n",
       " 'AGG': ('W', 'R', 'G', 'K', 'T', 'M', 'R', 'S', 'S'),\n",
       " 'GTT': ('I', 'L', 'F', 'D', 'A', 'G', 'V', 'V', 'V'),\n",
       " 'GTC': ('I', 'L', 'F', 'D', 'A', 'G', 'V', 'V', 'V'),\n",
       " 'GTA': ('I', 'L', 'L', 'E', 'A', 'G', 'V', 'V', 'V'),\n",
       " 'GTG': ('M', 'L', 'L', 'E', 'A', 'G', 'V', 'V', 'V'),\n",
       " 'GCT': ('T', 'P', 'S', 'D', 'G', 'V', 'A', 'A', 'A'),\n",
       " 'GCC': ('T', 'P', 'S', 'D', 'G', 'V', 'A', 'A', 'A'),\n",
       " 'GCA': ('T', 'P', 'S', 'E', 'G', 'V', 'A', 'A', 'A'),\n",
       " 'GCG': ('T', 'P', 'S', 'E', 'G', 'V', 'A', 'A', 'A'),\n",
       " 'GAT': ('N', 'H', 'Y', 'V', 'A', 'G', 'E', 'D', 'E'),\n",
       " 'GAC': ('N', 'H', 'Y', 'V', 'A', 'G', 'E', 'E', 'D'),\n",
       " 'GAA': ('K', 'Q', '*', 'V', 'A', 'G', 'D', 'D', 'E'),\n",
       " 'GAG': ('K', 'Q', '*', 'V', 'A', 'G', 'E', 'D', 'D'),\n",
       " 'GGT': ('S', 'R', 'C', 'D', 'A', 'V', 'G', 'G', 'G'),\n",
       " 'GGC': ('S', 'R', 'C', 'D', 'A', 'V', 'G', 'G', 'G'),\n",
       " 'GGA': ('R', 'R', '*', 'E', 'A', 'V', 'G', 'G', 'G'),\n",
       " 'GGG': ('R', 'R', 'W', 'E', 'A', 'V', 'G', 'G', 'G')}"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Possible_Amino_Acid_Changes_letters"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mutation rates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "code_folding": [],
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Import mutation rates\n",
    "import csv\n",
    "filename=\"Data files/Trinucleotide_context_mutation_rates.csv\" #csv file containing trinucleotide-context specific mutation rates (calculated from Lee-Six et al 2018 data)\n",
    "\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    Mutation_rates = {} #create a dictionary where key = trincucleotide site, value = mutation rate\n",
    "\n",
    "    for row in read_reader:\n",
    "#         print(row)\n",
    "        if row_count>0:\n",
    "            site = row[0] #trinucleotide site\n",
    "            comp_site = row[2] #complementary site\n",
    "            if site != '':\n",
    "                mutation_rate = float(row[7])\n",
    "                Mutation_rates[site] = mutation_rate\n",
    "                Mutation_rates[comp_site] = mutation_rate\n",
    "        row_count = row_count+1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functions for calculating mutation rates in DNMT3A"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Functions for calculating mutation rate from a csv file containing list of variants"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- csv file should take the format of column 1 = aa number (e.g. 882), column 2 = variant (e.g. p.R882H)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DNMT3A transcript NP072046 = 912 aa  (hg19 or GRCh38)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "def mutation_rates_list_nonsense_variants_DNMT3A_NP072046(csv_file_variants):\n",
    "    \n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base_DNMT3A_NP072046 = [1279, 1429, 1474, 1554, 1852, 2083, 2082, 2173, 2322, 2478]\n",
    "   \n",
    "    def intron_exon_check_DNMT3A_NP072046(position_in_gene_number):\n",
    "        if position_in_gene_number == 1279:\n",
    "            gene_pos = 1279\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1429:\n",
    "            gene_pos = 1429\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1474:\n",
    "            gene_pos = 1474\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1554:\n",
    "            gene_pos = 1554\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1852:\n",
    "            gene_pos = 1852\n",
    "            first_base = 'G'\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2083:\n",
    "            gene_pos = 2083\n",
    "            first_base = 'G'\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2082:\n",
    "            gene_pos = 2082\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2173:\n",
    "            gene_pos = 2173\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2322:\n",
    "            gene_pos = 2322\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2478:\n",
    "            gene_pos = 2478\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #DNMT3A transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    DNMT3A_exons_NP_072046 = 'ATGCCCGCCATGCCCTCCAGCGGCCCCGGGGACACCAGCAGCTCTGCTGCGGAGCGGGAGGAGGACCGAAAGGACGGAGAGGAGCAGGAGGAGCCGCGTGGCAAGGAGGAGCGCCAAGAGCCCAGCACCACGGCACGGAAGGTGGGGCGGCCTGGGAGGAAGCGCAAGCACCCCCCGGTGGAAAGCGGTGACACGCCAAAGGACCCTGCGGTGATCTCCAAGTCCCCATCCATGGCCCAGGACTCAGGCGCCTCAGAGCTATTACCCAATGGGGACTTGGAGAAGCGGAGTGAGCCCCAGCCAGAGGAGGGGAGCCCTGCTGGGGGGCAGAAGGGCGGGGCCCCAGCAGAGGGAGAGGGTGCAGCTGAGACCCTGCCTGAAGCCTCAAGAGCAGTGGAAAATGGCTGCTGCACCCCCAAGGAGGGCCGAGGAGCCCCTGCAGAAGCGGGCAAAGAACAGAAGGAGACCAACATCGAATCCATGAAAATGGAGGGCTCCCGGGGCCGGCTGCGGGGTGGCTTGGGCTGGGAGTCCAGCCTCCGTCAGCGGCCCATGCCGAGGCTCACCTTCCAGGCGGGGGACCCCTACTACATCAGCAAGCGCAAGCGGGACGAGTGGCTGGCACGCTGGAAAAGGGAGGCTGAGAAGAAAGCCAAGGTCATTGCAGGAATGAATGCTGTGGAAGAAAACCAGGGGCCCGGGGAGTCTCAGAAGGTGGAGGAGGCCAGCCCTCCTGCTGTGCAGCAGCCCACTGACCCCGCATCCCCCACTGTGGCTACCACGCCTGAGCCCGTGGGGTCCGATGCTGGGGACAAGAATGCCACCAAAGCAGGCGATGACGAGCCAGAGTACGAGGACGGCCGGGGCTTTGGCATTGGGGAGCTGGTGTGGGGGAAACTGCGGGGCTTCTCCTGGTGGCCAGGCCGCATTGTGTCTTGGTGGATGACGGGCCGGAGCCGAGCAGCTGAAGGCACCCGCTGGGTCATGTGGTTCGGAGACGGCAAATTCTCAGTGGTGTGTGTTGAGAAGCTGATGCCGCTGAGCTCGTTTTGCAGTGCGTTCCACCAGGCCACGTACAACAAGCAGCCCATGTACCGCAAAGCCATCTACGAGGTCCTGCAGGTGGCCAGCAGCCGCGCGGGGAAGCTGTTCCCGGTGTGCCACGACAGCGATGAGAGTGACACTGCCAAGGCCGTGGAGGTGCAGAACAAGCCCATGATTGAATGGGCCCTGGGGGGCTTCCAGCCTTCTGGCCCTAAGGGCCTGGAGCCACCAGAAGAAGAGAAGAATCCCTACAAAGAAGTGTACACGGACATGTGGGTGGAACCTGAGGCAGCTGCCTACGCACCACCTCCACCAGCCAAAAAGCCCCGGAAGAGCACAGCGGAGAAGCCCAAGGTCAAGGAGATTATTGATGAGCGCACAAGAGAGCGGCTGGTGTACGAGGTGCGGCAGAAGTGCCGGAACATTGAGGACATCTGCATCTCCTGTGGGAGCCTCAATGTTACCCTGGAACACCCCCTCTTCGTTGGAGGAATGTGCCAAAACTGCAAGAACTGCTTTCTGGAGTGTGCGTACCAGTACGACGACGACGGCTACCAGTCCTACTGCACCATCTGCTGTGGGGGCCGTGAGGTGCTCATGTGCGGAAACAACAACTGCTGCAGGTGCTTTTGCGTGGAGTGTGTGGACCTCTTGGTGGGGCCGGGGGCTGCCCAGGCAGCCATTAAGGAAGACCCCTGGAACTGCTACATGTGCGGGCACAAGGGTACCTACGGGCTGCTGCGGCGGCGAGAGGACTGGCCCTCCCGGCTCCAGATGTTCTTCGCTAATAACCACGACCAGGAATTTGACCCTCCAAAGGTTTACCCACCTGTCCCAGCTGAGAAGAGGAAGCCCATCCGGGTGCTGTCTCTCTTTGATGGAATCGCTACAGGGCTCCTGGTGCTGAAGGACTTGGGCATTCAGGTGGACCGCTACATTGCCTCGGAGGTGTGTGAGGACTCCATCACGGTGGGCATGGTGCGGCACCAGGGGAAGATCATGTACGTCGGGGACGTCCGCAGCGTCACACAGAAGCATATCCAGGAGTGGGGCCCATTCGATCTGGTGATTGGGGGCAGTCCCTGCAATGACCTCTCCATCGTCAACCCTGCTCGCAAGGGCCTCTACGAGGGCACTGGCCGGCTCTTCTTTGAGTTCTACCGCCTCCTGCATGATGCGCGGCCCAAGGAGGGAGATGATCGCCCCTTCTTCTGGCTCTTTGAGAATGTGGTGGCCATGGGCGTTAGTGACAAGAGGGACATCTCGCGATTTCTCGAGTCCAACCCTGTGATGATTGATGCCAAAGAAGTGTCAGCTGCACACAGGGCCCGCTACTTCTGGGGTAACCTTCCCGGTATGAACAGGCCGTTGGCATCCACTGTGAATGATAAGCTGGAGCTGCAGGAGTGTCTGGAGCATGGCAGGATAGCCAAGTTCAGCAAAGTGAGGACCATTACTACGAGGTCAAACTCCATAAAGCAGGGCAAAGACCAGCATTTTCCTGTCTTCATGAATGAGAAAGAGGACATCTTATGGTGCACTGAAATGGAAAGGGTATTTGGTTTCCCAGTCCACTATACTGACGTCTCCAACATGAGCCGCTTGGCGAGGCAGAGACTGCTGGGCCGGTCATGGAGCGTGCCAGTCATCCGCCACCTCTTCGCTCCGCTGAAGGAGTATTTTGCGTGTGTGTAA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    DNMT3A_codons_NP_072046 = [DNMT3A_exons_NP_072046[i:i+3] for i in range(0, len(DNMT3A_exons_NP_072046), 3)]\n",
    "\n",
    "    #DNMT3A transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    DNMT3A_exons_extra_bases_NP_072046 = 'GATGCCCGCCATGCCCTCCAGCGGCCCCGGGGACACCAGCAGCTCTGCTGCGGAGCGGGAGGAGGACCGAAAGGACGGAGAGGAGCAGGAGGAGCCGCGTGGCAAGGAGGAGCGCCAAGAGCCCAGCACCACGGCACGGAAGGTGGGGCGGCCTGGGAGGAAGCGCAAGCACCCCCCGGTGGAAAGCGGTGACACGCCAAAGGACCCTGCGGTGATCTCCAAGTCCCCATCCATGGCCCAGGACTCAGGCGCCTCAGAGCTATTACCCAATGGGGACTTGGAGAAGCGGAGTGAGCCCCAGCCAGAGGAGGGGAGCCCTGCTGGGGGGCAGAAGGGCGGGGCCCCAGCAGAGGGAGAGGGTGCAGCTGAGACCCTGCCTGAAGCCTCAAGAGCAGTGGAAAATGGCTGCTGCACCCCCAAGGAGGGCCGAGGAGCCCCTGCAGAAGCGGGCAAAGAACAGAAGGAGACCAACATCGAATCCATGAAAATGGAGGGCTCCCGGGGCCGGCTGCGGGGTGGCTTGGGCTGGGAGTCCAGCCTCCGTCAGCGGCCCATGCCGAGGCTCACCTTCCAGGCGGGGGACCCCTACTACATCAGCAAGCGCAAGCGGGACGAGTGGCTGGCACGCTGGAAAAGGGAGGCTGAGAAGAAAGCCAAGGTCATTGCAGGAATGAATGCTGTGGAAGAAAACCAGGGGCCCGGGGAGTCTCAGAAGGTGGAGGAGGCCAGCCCTCCTGCTGTGCAGCAGCCCACTGACCCCGCATCCCCCACTGTGGCTACCACGCCTGAGCCCGTGGGGTCCGATGCTGGGGACAAGAATGCCACCAAAGCAGGCGATGACGAGCCAGAGTACGAGGACGGCCGGGGCTTTGGCATTGGGGAGCTGGTGTGGGGGAAACTGCGGGGCTTCTCCTGGTGGCCAGGCCGCATTGTGTCTTGGTGGATGACGGGCCGGAGCCGAGCAGCTGAAGGCACCCGCTGGGTCATGTGGTTCGGAGACGGCAAATTCTCAGTGGTGTGTGTTGAGAAGCTGATGCCGCTGAGCTCGTTTTGCAGTGCGTTCCACCAGGCCACGTACAACAAGCAGCCCATGTACCGCAAAGCCATCTACGAGGTCCTGCAGGTGGCCAGCAGCCGCGCGGGGAAGCTGTTCCCGGTGTGCCACGACAGCGATGAGAGTGACACTGCCAAGGCCGTGGAGGTGCAGAACAAGCCCATGATTGAATGGGCCCTGGGGGGCTTCCAGCCTTCTGGCCCTAAGGGCCTGGAGCCACCAGAAGAAGAGAAGAATCCCTACAAAGAAGTGTACACGGACATGTGGGTGGAACCTGAGGCAGCTGCCTACGCACCACCTCCACCAGCCAAAAAGCCCCGGAAGAGCACAGCGGAGAAGCCCAAGGTCAAGGAGATTATTGATGAGCGCACAAGAGAGCGGCTGGTGTACGAGGTGCGGCAGAAGTGCCGGAACATTGAGGACATCTGCATCTCCTGTGGGAGCCTCAATGTTACCCTGGAACACCCCCTCTTCGTTGGAGGAATGTGCCAAAACTGCAAGAACTGCTTTCTGGAGTGTGCGTACCAGTACGACGACGACGGCTACCAGTCCTACTGCACCATCTGCTGTGGGGGCCGTGAGGTGCTCATGTGCGGAAACAACAACTGCTGCAGGTGCTTTTGCGTGGAGTGTGTGGACCTCTTGGTGGGGCCGGGGGCTGCCCAGGCAGCCATTAAGGAAGACCCCTGGAACTGCTACATGTGCGGGCACAAGGGTACCTACGGGCTGCTGCGGCGGCGAGAGGACTGGCCCTCCCGGCTCCAGATGTTCTTCGCTAATAACCACGACCAGGAATTTGACCCTCCAAAGGTTTACCCACCTGTCCCAGCTGAGAAGAGGAAGCCCATCCGGGTGCTGTCTCTCTTTGATGGAATCGCTACAGGGCTCCTGGTGCTGAAGGACTTGGGCATTCAGGTGGACCGCTACATTGCCTCGGAGGTGTGTGAGGACTCCATCACGGTGGGCATGGTGCGGCACCAGGGGAAGATCATGTACGTCGGGGACGTCCGCAGCGTCACACAGAAGCATATCCAGGAGTGGGGCCCATTCGATCTGGTGATTGGGGGCAGTCCCTGCAATGACCTCTCCATCGTCAACCCTGCTCGCAAGGGCCTCTACGAGGGCACTGGCCGGCTCTTCTTTGAGTTCTACCGCCTCCTGCATGATGCGCGGCCCAAGGAGGGAGATGATCGCCCCTTCTTCTGGCTCTTTGAGAATGTGGTGGCCATGGGCGTTAGTGACAAGAGGGACATCTCGCGATTTCTCGAGTCCAACCCTGTGATGATTGATGCCAAAGAAGTGTCAGCTGCACACAGGGCCCGCTACTTCTGGGGTAACCTTCCCGGTATGAACAGGCCGTTGGCATCCACTGTGAATGATAAGCTGGAGCTGCAGGAGTGTCTGGAGCATGGCAGGATAGCCAAGTTCAGCAAAGTGAGGACCATTACTACGAGGTCAAACTCCATAAAGCAGGGCAAAGACCAGCATTTTCCTGTCTTCATGAATGAGAAAGAGGACATCTTATGGTGCACTGAAATGGAAAGGGTATTTGGTTTCCCAGTCCACTATACTGACGTCTCCAACATGAGCCGCTTGGCGAGGCAGAGACTGCTGGGCCGGTCATGGAGCGTGCCAGTCATCCGCCACCTCTTCGCTCCGCTGAAGGAGTATTTTGCGTGTGTGTAAG'\n",
    "\n",
    "    #Create a list of the codons they called nonsense variants at:\n",
    "    import csv\n",
    "    filename = csv_file_variants\n",
    "    with open(filename, 'r') as csvfile:\n",
    "        read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "        row_count=0\n",
    "\n",
    "        DNMT3A_nonsense_variant_list_NP_072046 = []\n",
    "        DNMT3A_nonsense_codons_NP_072046 =[]\n",
    "        DNMT3A_nonsense_variant_DICT_NP_072046 = {}\n",
    "\n",
    "        for row in read_reader: #each row contains aa number (e.g. 11) and variant (e.g. p.D11A)\n",
    "            if row_count>0: \n",
    "                codon0_NP_072046 = int(row[0])-1\n",
    "                codon1_NP_072046 = int(row[0])\n",
    "                variant_NP_072046 = row[1]\n",
    "                DNMT3A_nonsense_codons_NP_072046.append(codon0_NP_072046) #List of codon numbers that they called variants at\n",
    "                DNMT3A_nonsense_variant_DICT_NP_072046[row_count]=(codon1_NP_072046, variant_NP_072046) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76*)\n",
    "\n",
    "            row_count=row_count+1\n",
    "            \n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    DNMT3A_nonsense_codons_NP_072046_DICT = {}\n",
    "    calls_DNMT3A_nonsense_NP_072046 = {}\n",
    "\n",
    "    for i in DNMT3A_nonsense_codons_NP_072046: #Iterates over the list of codon numbers\n",
    "        DNMT3A_nonsense_codons_NP_072046_DICT[i+1]= (DNMT3A_codons_NP_072046[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "        \n",
    "    for k,v in DNMT3A_nonsense_codons_NP_072046_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_DNMT3A_nonsense_NP_072046[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "        \n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_DNMT3A_nonsense_NP_072046_extra = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_nonsense_NP_072046.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_DNMT3A_nonsense_NP_072046_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_DNMT3A_nonsense_NP_072046_extra2 = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_nonsense_NP_072046_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_DNMT3A_nonsense_NP_072046_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "            \n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    DNMT3A_position_in_list_dict_nonsense_NP_072046={}\n",
    "    for k, v in DNMT3A_nonsense_variant_DICT_NP_072046.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_DNMT3A_nonsense_NP_072046[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_DNMT3A_nonsense_NP_072046[v[0]]\n",
    "        codon_seq=calls_DNMT3A_nonsense_NP_072046[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_DNMT3A_nonsense_NP_072046_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                DNMT3A_position_in_list_dict_nonsense_NP_072046[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "            \n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in DNMT3A_position_in_list_dict_nonsense_NP_072046.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "#     print('positions at which nonsense variants called = ', aa_positions_variants_called)\n",
    "                \n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    DNMT3A_bases_nonsense_NP_072046 = {}\n",
    "    for k, v in DNMT3A_position_in_list_dict_nonsense_NP_072046.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        DNMT3A_bases_nonsense_NP_072046[k]=(v, base_changes)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    DNMT3A_bases_nonsense2_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_nonsense_NP_072046.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        DNMT3A_bases_nonsense2_NP_072046[k] = (v, codon_positions)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    DNMT3A_bases_nonsense3_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_nonsense2_NP_072046.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][1][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        DNMT3A_bases_nonsense3_NP_072046[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    DNMT3A_bases_nonsense4_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_nonsense3_NP_072046.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_nonsense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_DNMT3A_NP072046:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check_DNMT3A_NP072046(i)\n",
    "                sites_trio_nonsense.append([adapted_trio])\n",
    "                DNMT3A_bases_nonsense4_NP_072046[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_nonsense = [DNMT3A_exons_extra_bases_NP_072046[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_nonsense.append(site_trio_nonsense)\n",
    "                    DNMT3A_bases_nonsense4_NP_072046[k] = (v, sites_trio_nonsense)\n",
    "                    \n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    DNMT3A_sites_nonsense5_NP_072046 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in DNMT3A_bases_nonsense4_NP_072046.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        DNMT3A_sites_nonsense5_NP_072046[a]=info\n",
    "        a=a+1\n",
    "        \n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    DNMT3A_sites_nonsense6_NP_072046={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in DNMT3A_sites_nonsense5_NP_072046.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "    \n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        DNMT3A_sites_nonsense6_NP_072046[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "        \n",
    "        \n",
    "    #Assign the mutation rates to each site...\n",
    "    DNMT3A_mutation_rates_nonsense_NP_072046 = {}\n",
    "    for k, v in DNMT3A_sites_nonsense6_NP_072046.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            DNMT3A_mutation_rates_nonsense_NP_072046[k] = (v, mutation_rate)\n",
    "\n",
    "    DNMT3A_mutation_rate_nonsense_NP_072046_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in DNMT3A_mutation_rates_nonsense_NP_072046.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        DNMT3A_mutation_rate_nonsense_NP_072046_sum.append(mutation_rate_value)\n",
    "        \n",
    "#     #Get the average of all the mutation rates\n",
    "#     DNMT3A_mutation_rate_nonsense_NP_072046_sum = []\n",
    "#     for k, v in DNMT3A_mutation_rates_nonsense_NP_072046.items():\n",
    "#         mut_rate = v[1]\n",
    "#         DNMT3A_mutation_rate_nonsense_NP_072046_sum.append(mut_rate)\n",
    "\n",
    "    DNMT3A_Average_mutation_rate_nonsense_NP_072046 = np.mean(DNMT3A_mutation_rate_nonsense_NP_072046_sum)\n",
    "    DNMT3A_Summed_mutation_rate_nonsense_NP_072046_added = np.sum(DNMT3A_mutation_rate_nonsense_NP_072046_sum)\n",
    "\n",
    "#     print('Average mutation rate across DNMT3A nonsense_NP_072046 for this study = ', DNMT3A_Average_mutation_rate_nonsense_NP_072046)\n",
    "    print('Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study = ', DNMT3A_Summed_mutation_rate_nonsense_NP_072046_added)\n",
    "    \n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all #in format: column 1 = 882, column 2 = p.R882*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "def mutation_rates_list_missense_variants_DNMT3A_NP072046(csv_file_variants):\n",
    "    \n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base_DNMT3A_NP072046 = [1279, 1429, 1474, 1554, 1852, 2083, 2082, 2173, 2322, 2478]\n",
    "\n",
    "    def intron_exon_check_DNMT3A_NP072046(position_in_gene_number):\n",
    "        if position_in_gene_number == 1279:\n",
    "            gene_pos = 1279\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1429:\n",
    "            gene_pos = 1429\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1474:\n",
    "            gene_pos = 1474\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1554:\n",
    "            gene_pos = 1554\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1852:\n",
    "            gene_pos = 1852\n",
    "            first_base = 'G'\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2083:\n",
    "            gene_pos = 2083\n",
    "            first_base = 'G'\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2082:\n",
    "            gene_pos = 2082\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2173:\n",
    "            gene_pos = 2173\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2322:\n",
    "            gene_pos = 2322\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2478:\n",
    "            gene_pos = 2478\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #DNMT3A transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    DNMT3A_exons_NP_072046 = 'ATGCCCGCCATGCCCTCCAGCGGCCCCGGGGACACCAGCAGCTCTGCTGCGGAGCGGGAGGAGGACCGAAAGGACGGAGAGGAGCAGGAGGAGCCGCGTGGCAAGGAGGAGCGCCAAGAGCCCAGCACCACGGCACGGAAGGTGGGGCGGCCTGGGAGGAAGCGCAAGCACCCCCCGGTGGAAAGCGGTGACACGCCAAAGGACCCTGCGGTGATCTCCAAGTCCCCATCCATGGCCCAGGACTCAGGCGCCTCAGAGCTATTACCCAATGGGGACTTGGAGAAGCGGAGTGAGCCCCAGCCAGAGGAGGGGAGCCCTGCTGGGGGGCAGAAGGGCGGGGCCCCAGCAGAGGGAGAGGGTGCAGCTGAGACCCTGCCTGAAGCCTCAAGAGCAGTGGAAAATGGCTGCTGCACCCCCAAGGAGGGCCGAGGAGCCCCTGCAGAAGCGGGCAAAGAACAGAAGGAGACCAACATCGAATCCATGAAAATGGAGGGCTCCCGGGGCCGGCTGCGGGGTGGCTTGGGCTGGGAGTCCAGCCTCCGTCAGCGGCCCATGCCGAGGCTCACCTTCCAGGCGGGGGACCCCTACTACATCAGCAAGCGCAAGCGGGACGAGTGGCTGGCACGCTGGAAAAGGGAGGCTGAGAAGAAAGCCAAGGTCATTGCAGGAATGAATGCTGTGGAAGAAAACCAGGGGCCCGGGGAGTCTCAGAAGGTGGAGGAGGCCAGCCCTCCTGCTGTGCAGCAGCCCACTGACCCCGCATCCCCCACTGTGGCTACCACGCCTGAGCCCGTGGGGTCCGATGCTGGGGACAAGAATGCCACCAAAGCAGGCGATGACGAGCCAGAGTACGAGGACGGCCGGGGCTTTGGCATTGGGGAGCTGGTGTGGGGGAAACTGCGGGGCTTCTCCTGGTGGCCAGGCCGCATTGTGTCTTGGTGGATGACGGGCCGGAGCCGAGCAGCTGAAGGCACCCGCTGGGTCATGTGGTTCGGAGACGGCAAATTCTCAGTGGTGTGTGTTGAGAAGCTGATGCCGCTGAGCTCGTTTTGCAGTGCGTTCCACCAGGCCACGTACAACAAGCAGCCCATGTACCGCAAAGCCATCTACGAGGTCCTGCAGGTGGCCAGCAGCCGCGCGGGGAAGCTGTTCCCGGTGTGCCACGACAGCGATGAGAGTGACACTGCCAAGGCCGTGGAGGTGCAGAACAAGCCCATGATTGAATGGGCCCTGGGGGGCTTCCAGCCTTCTGGCCCTAAGGGCCTGGAGCCACCAGAAGAAGAGAAGAATCCCTACAAAGAAGTGTACACGGACATGTGGGTGGAACCTGAGGCAGCTGCCTACGCACCACCTCCACCAGCCAAAAAGCCCCGGAAGAGCACAGCGGAGAAGCCCAAGGTCAAGGAGATTATTGATGAGCGCACAAGAGAGCGGCTGGTGTACGAGGTGCGGCAGAAGTGCCGGAACATTGAGGACATCTGCATCTCCTGTGGGAGCCTCAATGTTACCCTGGAACACCCCCTCTTCGTTGGAGGAATGTGCCAAAACTGCAAGAACTGCTTTCTGGAGTGTGCGTACCAGTACGACGACGACGGCTACCAGTCCTACTGCACCATCTGCTGTGGGGGCCGTGAGGTGCTCATGTGCGGAAACAACAACTGCTGCAGGTGCTTTTGCGTGGAGTGTGTGGACCTCTTGGTGGGGCCGGGGGCTGCCCAGGCAGCCATTAAGGAAGACCCCTGGAACTGCTACATGTGCGGGCACAAGGGTACCTACGGGCTGCTGCGGCGGCGAGAGGACTGGCCCTCCCGGCTCCAGATGTTCTTCGCTAATAACCACGACCAGGAATTTGACCCTCCAAAGGTTTACCCACCTGTCCCAGCTGAGAAGAGGAAGCCCATCCGGGTGCTGTCTCTCTTTGATGGAATCGCTACAGGGCTCCTGGTGCTGAAGGACTTGGGCATTCAGGTGGACCGCTACATTGCCTCGGAGGTGTGTGAGGACTCCATCACGGTGGGCATGGTGCGGCACCAGGGGAAGATCATGTACGTCGGGGACGTCCGCAGCGTCACACAGAAGCATATCCAGGAGTGGGGCCCATTCGATCTGGTGATTGGGGGCAGTCCCTGCAATGACCTCTCCATCGTCAACCCTGCTCGCAAGGGCCTCTACGAGGGCACTGGCCGGCTCTTCTTTGAGTTCTACCGCCTCCTGCATGATGCGCGGCCCAAGGAGGGAGATGATCGCCCCTTCTTCTGGCTCTTTGAGAATGTGGTGGCCATGGGCGTTAGTGACAAGAGGGACATCTCGCGATTTCTCGAGTCCAACCCTGTGATGATTGATGCCAAAGAAGTGTCAGCTGCACACAGGGCCCGCTACTTCTGGGGTAACCTTCCCGGTATGAACAGGCCGTTGGCATCCACTGTGAATGATAAGCTGGAGCTGCAGGAGTGTCTGGAGCATGGCAGGATAGCCAAGTTCAGCAAAGTGAGGACCATTACTACGAGGTCAAACTCCATAAAGCAGGGCAAAGACCAGCATTTTCCTGTCTTCATGAATGAGAAAGAGGACATCTTATGGTGCACTGAAATGGAAAGGGTATTTGGTTTCCCAGTCCACTATACTGACGTCTCCAACATGAGCCGCTTGGCGAGGCAGAGACTGCTGGGCCGGTCATGGAGCGTGCCAGTCATCCGCCACCTCTTCGCTCCGCTGAAGGAGTATTTTGCGTGTGTGTAA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    DNMT3A_codons_NP_072046 = [DNMT3A_exons_NP_072046[i:i+3] for i in range(0, len(DNMT3A_exons_NP_072046), 3)]\n",
    "\n",
    "    #DNMT3A transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    DNMT3A_exons_extra_bases_NP_072046 = 'GATGCCCGCCATGCCCTCCAGCGGCCCCGGGGACACCAGCAGCTCTGCTGCGGAGCGGGAGGAGGACCGAAAGGACGGAGAGGAGCAGGAGGAGCCGCGTGGCAAGGAGGAGCGCCAAGAGCCCAGCACCACGGCACGGAAGGTGGGGCGGCCTGGGAGGAAGCGCAAGCACCCCCCGGTGGAAAGCGGTGACACGCCAAAGGACCCTGCGGTGATCTCCAAGTCCCCATCCATGGCCCAGGACTCAGGCGCCTCAGAGCTATTACCCAATGGGGACTTGGAGAAGCGGAGTGAGCCCCAGCCAGAGGAGGGGAGCCCTGCTGGGGGGCAGAAGGGCGGGGCCCCAGCAGAGGGAGAGGGTGCAGCTGAGACCCTGCCTGAAGCCTCAAGAGCAGTGGAAAATGGCTGCTGCACCCCCAAGGAGGGCCGAGGAGCCCCTGCAGAAGCGGGCAAAGAACAGAAGGAGACCAACATCGAATCCATGAAAATGGAGGGCTCCCGGGGCCGGCTGCGGGGTGGCTTGGGCTGGGAGTCCAGCCTCCGTCAGCGGCCCATGCCGAGGCTCACCTTCCAGGCGGGGGACCCCTACTACATCAGCAAGCGCAAGCGGGACGAGTGGCTGGCACGCTGGAAAAGGGAGGCTGAGAAGAAAGCCAAGGTCATTGCAGGAATGAATGCTGTGGAAGAAAACCAGGGGCCCGGGGAGTCTCAGAAGGTGGAGGAGGCCAGCCCTCCTGCTGTGCAGCAGCCCACTGACCCCGCATCCCCCACTGTGGCTACCACGCCTGAGCCCGTGGGGTCCGATGCTGGGGACAAGAATGCCACCAAAGCAGGCGATGACGAGCCAGAGTACGAGGACGGCCGGGGCTTTGGCATTGGGGAGCTGGTGTGGGGGAAACTGCGGGGCTTCTCCTGGTGGCCAGGCCGCATTGTGTCTTGGTGGATGACGGGCCGGAGCCGAGCAGCTGAAGGCACCCGCTGGGTCATGTGGTTCGGAGACGGCAAATTCTCAGTGGTGTGTGTTGAGAAGCTGATGCCGCTGAGCTCGTTTTGCAGTGCGTTCCACCAGGCCACGTACAACAAGCAGCCCATGTACCGCAAAGCCATCTACGAGGTCCTGCAGGTGGCCAGCAGCCGCGCGGGGAAGCTGTTCCCGGTGTGCCACGACAGCGATGAGAGTGACACTGCCAAGGCCGTGGAGGTGCAGAACAAGCCCATGATTGAATGGGCCCTGGGGGGCTTCCAGCCTTCTGGCCCTAAGGGCCTGGAGCCACCAGAAGAAGAGAAGAATCCCTACAAAGAAGTGTACACGGACATGTGGGTGGAACCTGAGGCAGCTGCCTACGCACCACCTCCACCAGCCAAAAAGCCCCGGAAGAGCACAGCGGAGAAGCCCAAGGTCAAGGAGATTATTGATGAGCGCACAAGAGAGCGGCTGGTGTACGAGGTGCGGCAGAAGTGCCGGAACATTGAGGACATCTGCATCTCCTGTGGGAGCCTCAATGTTACCCTGGAACACCCCCTCTTCGTTGGAGGAATGTGCCAAAACTGCAAGAACTGCTTTCTGGAGTGTGCGTACCAGTACGACGACGACGGCTACCAGTCCTACTGCACCATCTGCTGTGGGGGCCGTGAGGTGCTCATGTGCGGAAACAACAACTGCTGCAGGTGCTTTTGCGTGGAGTGTGTGGACCTCTTGGTGGGGCCGGGGGCTGCCCAGGCAGCCATTAAGGAAGACCCCTGGAACTGCTACATGTGCGGGCACAAGGGTACCTACGGGCTGCTGCGGCGGCGAGAGGACTGGCCCTCCCGGCTCCAGATGTTCTTCGCTAATAACCACGACCAGGAATTTGACCCTCCAAAGGTTTACCCACCTGTCCCAGCTGAGAAGAGGAAGCCCATCCGGGTGCTGTCTCTCTTTGATGGAATCGCTACAGGGCTCCTGGTGCTGAAGGACTTGGGCATTCAGGTGGACCGCTACATTGCCTCGGAGGTGTGTGAGGACTCCATCACGGTGGGCATGGTGCGGCACCAGGGGAAGATCATGTACGTCGGGGACGTCCGCAGCGTCACACAGAAGCATATCCAGGAGTGGGGCCCATTCGATCTGGTGATTGGGGGCAGTCCCTGCAATGACCTCTCCATCGTCAACCCTGCTCGCAAGGGCCTCTACGAGGGCACTGGCCGGCTCTTCTTTGAGTTCTACCGCCTCCTGCATGATGCGCGGCCCAAGGAGGGAGATGATCGCCCCTTCTTCTGGCTCTTTGAGAATGTGGTGGCCATGGGCGTTAGTGACAAGAGGGACATCTCGCGATTTCTCGAGTCCAACCCTGTGATGATTGATGCCAAAGAAGTGTCAGCTGCACACAGGGCCCGCTACTTCTGGGGTAACCTTCCCGGTATGAACAGGCCGTTGGCATCCACTGTGAATGATAAGCTGGAGCTGCAGGAGTGTCTGGAGCATGGCAGGATAGCCAAGTTCAGCAAAGTGAGGACCATTACTACGAGGTCAAACTCCATAAAGCAGGGCAAAGACCAGCATTTTCCTGTCTTCATGAATGAGAAAGAGGACATCTTATGGTGCACTGAAATGGAAAGGGTATTTGGTTTCCCAGTCCACTATACTGACGTCTCCAACATGAGCCGCTTGGCGAGGCAGAGACTGCTGGGCCGGTCATGGAGCGTGCCAGTCATCCGCCACCTCTTCGCTCCGCTGAAGGAGTATTTTGCGTGTGTGTAAG'\n",
    "\n",
    "    #Create a list of the codons they called missense variants at:\n",
    "    import csv\n",
    "    filename = csv_file_variants\n",
    "    with open(filename, 'r') as csvfile:\n",
    "        read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "        row_count=0\n",
    "\n",
    "        DNMT3A_missense_variant_list_NP_072046 = []\n",
    "        DNMT3A_missense_codons_NP_072046 =[]\n",
    "        DNMT3A_missense_variant_DICT_NP_072046 = {}\n",
    "\n",
    "        for row in read_reader: #each row contains aa number (e.g. 11) and variant (e.g. p.D11A)\n",
    "    #         print(row)\n",
    "            if row_count>0: \n",
    "                codon0_NP_072046 = int(row[0])-1\n",
    "                codon1_NP_072046 = int(row[0])\n",
    "                variant_NP_072046 = row[1]\n",
    "                DNMT3A_missense_codons_NP_072046.append(codon0_NP_072046) #List of codon numbers that they called variants at\n",
    "                DNMT3A_missense_variant_DICT_NP_072046[row_count]=(codon1_NP_072046, variant_NP_072046) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "            row_count=row_count+1\n",
    "            \n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    DNMT3A_missense_codons_NP_072046_DICT = {}\n",
    "    calls_DNMT3A_missense_NP_072046 = {}\n",
    "\n",
    "    for i in DNMT3A_missense_codons_NP_072046: #Iterates over the list of codon numbers\n",
    "        DNMT3A_missense_codons_NP_072046_DICT[i+1]= (DNMT3A_codons_NP_072046[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "        \n",
    "    for k,v in DNMT3A_missense_codons_NP_072046_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_DNMT3A_missense_NP_072046[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "        \n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_DNMT3A_missense_NP_072046_extra = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_missense_NP_072046.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_DNMT3A_missense_NP_072046_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_DNMT3A_missense_NP_072046_extra2 = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_missense_NP_072046_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_DNMT3A_missense_NP_072046_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "            \n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    DNMT3A_position_in_list_dict_missense_NP_072046={}\n",
    "    for k, v in DNMT3A_missense_variant_DICT_NP_072046.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_DNMT3A_missense_NP_072046[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_DNMT3A_missense_NP_072046[v[0]]\n",
    "        codon_seq=calls_DNMT3A_missense_NP_072046[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_DNMT3A_missense_NP_072046_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                DNMT3A_position_in_list_dict_missense_NP_072046[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "            \n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in DNMT3A_position_in_list_dict_missense_NP_072046.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "#     print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "                \n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    DNMT3A_bases_missense_NP_072046 = {}\n",
    "    for k, v in DNMT3A_position_in_list_dict_missense_NP_072046.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        DNMT3A_bases_missense_NP_072046[k]=(v, base_changes)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    DNMT3A_bases_missense2_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_missense_NP_072046.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        DNMT3A_bases_missense2_NP_072046[k] = (v, codon_positions)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    DNMT3A_bases_missense3_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_missense2_NP_072046.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        DNMT3A_bases_missense3_NP_072046[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    DNMT3A_bases_missense4_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_missense3_NP_072046.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_DNMT3A_NP072046:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check_DNMT3A_NP072046(i)\n",
    "                sites_trio_missense.append([adapted_trio])\n",
    "                DNMT3A_bases_missense4_NP_072046[k] = (v, sites_trio_missense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_missense = [DNMT3A_exons_extra_bases_NP_072046[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_missense.append(site_trio_missense)\n",
    "                    DNMT3A_bases_missense4_NP_072046[k] = (v, sites_trio_missense)\n",
    "                    \n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    DNMT3A_sites_missense5_NP_072046 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in DNMT3A_bases_missense4_NP_072046.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        DNMT3A_sites_missense5_NP_072046[a]=info\n",
    "        a=a+1\n",
    "        \n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    DNMT3A_sites_missense6_NP_072046={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in DNMT3A_sites_missense5_NP_072046.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "    \n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        DNMT3A_sites_missense6_NP_072046[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "        \n",
    "        \n",
    "    #Assign the mutation rates to each site...\n",
    "    DNMT3A_mutation_rates_missense_NP_072046 = {}\n",
    "    for k, v in DNMT3A_sites_missense6_NP_072046.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            DNMT3A_mutation_rates_missense_NP_072046[k] = (v, mutation_rate)\n",
    "\n",
    "    DNMT3A_mutation_rate_missense_NP_072046_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in DNMT3A_mutation_rates_missense_NP_072046.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        DNMT3A_mutation_rate_missense_NP_072046_sum.append(mutation_rate_value)\n",
    "        \n",
    "#     #Get the average of all the mutation rates\n",
    "#     DNMT3A_mutation_rate_missense_NP_072046_sum = []\n",
    "#     for k, v in DNMT3A_mutation_rates_missense_NP_072046.items():\n",
    "#         mut_rate = v[1]\n",
    "#         DNMT3A_mutation_rate_missense_NP_072046_sum.append(mut_rate)\n",
    "\n",
    "    DNMT3A_Average_mutation_rate_missense_NP_072046 = np.mean(DNMT3A_mutation_rate_missense_NP_072046_sum)\n",
    "    DNMT3A_Summed_mutation_rate_missense_NP_072046_added = np.sum(DNMT3A_mutation_rate_missense_NP_072046_sum)\n",
    "\n",
    "#     print('Average mutation rate across DNMT3A missense_NP_072046 for this study = ', DNMT3A_Average_mutation_rate_missense_NP_072046)\n",
    "    print('Total summed mutation rate across DNMT3A missense_NP_072046 for this study = ', DNMT3A_Summed_mutation_rate_missense_NP_072046_added)\n",
    "    \n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all #in format: column 1 = 882, column 2 = p.R882H"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "def mutation_rates_list_synonymous_variants_DNMT3A_NP072046(csv_file_variants):\n",
    "    \n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base_DNMT3A_NP072046 = [1279, 1429, 1474, 1554, 1852, 2083, 2082, 2173, 2322, 2478]\n",
    "\n",
    "    def intron_exon_check_DNMT3A_NP072046(position_in_gene_number):\n",
    "        if position_in_gene_number == 1279:\n",
    "            gene_pos = 1279\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1429:\n",
    "            gene_pos = 1429\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1474:\n",
    "            gene_pos = 1474\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1554:\n",
    "            gene_pos = 1554\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1852:\n",
    "            gene_pos = 1852\n",
    "            first_base = 'G'\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2083:\n",
    "            gene_pos = 2083\n",
    "            first_base = 'G'\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2082:\n",
    "            gene_pos = 2082\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2173:\n",
    "            gene_pos = 2173\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2322:\n",
    "            gene_pos = 2322\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2478:\n",
    "            gene_pos = 2478\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #DNMT3A transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    DNMT3A_exons_NP_072046 = 'ATGCCCGCCATGCCCTCCAGCGGCCCCGGGGACACCAGCAGCTCTGCTGCGGAGCGGGAGGAGGACCGAAAGGACGGAGAGGAGCAGGAGGAGCCGCGTGGCAAGGAGGAGCGCCAAGAGCCCAGCACCACGGCACGGAAGGTGGGGCGGCCTGGGAGGAAGCGCAAGCACCCCCCGGTGGAAAGCGGTGACACGCCAAAGGACCCTGCGGTGATCTCCAAGTCCCCATCCATGGCCCAGGACTCAGGCGCCTCAGAGCTATTACCCAATGGGGACTTGGAGAAGCGGAGTGAGCCCCAGCCAGAGGAGGGGAGCCCTGCTGGGGGGCAGAAGGGCGGGGCCCCAGCAGAGGGAGAGGGTGCAGCTGAGACCCTGCCTGAAGCCTCAAGAGCAGTGGAAAATGGCTGCTGCACCCCCAAGGAGGGCCGAGGAGCCCCTGCAGAAGCGGGCAAAGAACAGAAGGAGACCAACATCGAATCCATGAAAATGGAGGGCTCCCGGGGCCGGCTGCGGGGTGGCTTGGGCTGGGAGTCCAGCCTCCGTCAGCGGCCCATGCCGAGGCTCACCTTCCAGGCGGGGGACCCCTACTACATCAGCAAGCGCAAGCGGGACGAGTGGCTGGCACGCTGGAAAAGGGAGGCTGAGAAGAAAGCCAAGGTCATTGCAGGAATGAATGCTGTGGAAGAAAACCAGGGGCCCGGGGAGTCTCAGAAGGTGGAGGAGGCCAGCCCTCCTGCTGTGCAGCAGCCCACTGACCCCGCATCCCCCACTGTGGCTACCACGCCTGAGCCCGTGGGGTCCGATGCTGGGGACAAGAATGCCACCAAAGCAGGCGATGACGAGCCAGAGTACGAGGACGGCCGGGGCTTTGGCATTGGGGAGCTGGTGTGGGGGAAACTGCGGGGCTTCTCCTGGTGGCCAGGCCGCATTGTGTCTTGGTGGATGACGGGCCGGAGCCGAGCAGCTGAAGGCACCCGCTGGGTCATGTGGTTCGGAGACGGCAAATTCTCAGTGGTGTGTGTTGAGAAGCTGATGCCGCTGAGCTCGTTTTGCAGTGCGTTCCACCAGGCCACGTACAACAAGCAGCCCATGTACCGCAAAGCCATCTACGAGGTCCTGCAGGTGGCCAGCAGCCGCGCGGGGAAGCTGTTCCCGGTGTGCCACGACAGCGATGAGAGTGACACTGCCAAGGCCGTGGAGGTGCAGAACAAGCCCATGATTGAATGGGCCCTGGGGGGCTTCCAGCCTTCTGGCCCTAAGGGCCTGGAGCCACCAGAAGAAGAGAAGAATCCCTACAAAGAAGTGTACACGGACATGTGGGTGGAACCTGAGGCAGCTGCCTACGCACCACCTCCACCAGCCAAAAAGCCCCGGAAGAGCACAGCGGAGAAGCCCAAGGTCAAGGAGATTATTGATGAGCGCACAAGAGAGCGGCTGGTGTACGAGGTGCGGCAGAAGTGCCGGAACATTGAGGACATCTGCATCTCCTGTGGGAGCCTCAATGTTACCCTGGAACACCCCCTCTTCGTTGGAGGAATGTGCCAAAACTGCAAGAACTGCTTTCTGGAGTGTGCGTACCAGTACGACGACGACGGCTACCAGTCCTACTGCACCATCTGCTGTGGGGGCCGTGAGGTGCTCATGTGCGGAAACAACAACTGCTGCAGGTGCTTTTGCGTGGAGTGTGTGGACCTCTTGGTGGGGCCGGGGGCTGCCCAGGCAGCCATTAAGGAAGACCCCTGGAACTGCTACATGTGCGGGCACAAGGGTACCTACGGGCTGCTGCGGCGGCGAGAGGACTGGCCCTCCCGGCTCCAGATGTTCTTCGCTAATAACCACGACCAGGAATTTGACCCTCCAAAGGTTTACCCACCTGTCCCAGCTGAGAAGAGGAAGCCCATCCGGGTGCTGTCTCTCTTTGATGGAATCGCTACAGGGCTCCTGGTGCTGAAGGACTTGGGCATTCAGGTGGACCGCTACATTGCCTCGGAGGTGTGTGAGGACTCCATCACGGTGGGCATGGTGCGGCACCAGGGGAAGATCATGTACGTCGGGGACGTCCGCAGCGTCACACAGAAGCATATCCAGGAGTGGGGCCCATTCGATCTGGTGATTGGGGGCAGTCCCTGCAATGACCTCTCCATCGTCAACCCTGCTCGCAAGGGCCTCTACGAGGGCACTGGCCGGCTCTTCTTTGAGTTCTACCGCCTCCTGCATGATGCGCGGCCCAAGGAGGGAGATGATCGCCCCTTCTTCTGGCTCTTTGAGAATGTGGTGGCCATGGGCGTTAGTGACAAGAGGGACATCTCGCGATTTCTCGAGTCCAACCCTGTGATGATTGATGCCAAAGAAGTGTCAGCTGCACACAGGGCCCGCTACTTCTGGGGTAACCTTCCCGGTATGAACAGGCCGTTGGCATCCACTGTGAATGATAAGCTGGAGCTGCAGGAGTGTCTGGAGCATGGCAGGATAGCCAAGTTCAGCAAAGTGAGGACCATTACTACGAGGTCAAACTCCATAAAGCAGGGCAAAGACCAGCATTTTCCTGTCTTCATGAATGAGAAAGAGGACATCTTATGGTGCACTGAAATGGAAAGGGTATTTGGTTTCCCAGTCCACTATACTGACGTCTCCAACATGAGCCGCTTGGCGAGGCAGAGACTGCTGGGCCGGTCATGGAGCGTGCCAGTCATCCGCCACCTCTTCGCTCCGCTGAAGGAGTATTTTGCGTGTGTGTAA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    DNMT3A_codons_NP_072046 = [DNMT3A_exons_NP_072046[i:i+3] for i in range(0, len(DNMT3A_exons_NP_072046), 3)]\n",
    "\n",
    "    #DNMT3A transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    DNMT3A_exons_extra_bases_NP_072046 = 'GATGCCCGCCATGCCCTCCAGCGGCCCCGGGGACACCAGCAGCTCTGCTGCGGAGCGGGAGGAGGACCGAAAGGACGGAGAGGAGCAGGAGGAGCCGCGTGGCAAGGAGGAGCGCCAAGAGCCCAGCACCACGGCACGGAAGGTGGGGCGGCCTGGGAGGAAGCGCAAGCACCCCCCGGTGGAAAGCGGTGACACGCCAAAGGACCCTGCGGTGATCTCCAAGTCCCCATCCATGGCCCAGGACTCAGGCGCCTCAGAGCTATTACCCAATGGGGACTTGGAGAAGCGGAGTGAGCCCCAGCCAGAGGAGGGGAGCCCTGCTGGGGGGCAGAAGGGCGGGGCCCCAGCAGAGGGAGAGGGTGCAGCTGAGACCCTGCCTGAAGCCTCAAGAGCAGTGGAAAATGGCTGCTGCACCCCCAAGGAGGGCCGAGGAGCCCCTGCAGAAGCGGGCAAAGAACAGAAGGAGACCAACATCGAATCCATGAAAATGGAGGGCTCCCGGGGCCGGCTGCGGGGTGGCTTGGGCTGGGAGTCCAGCCTCCGTCAGCGGCCCATGCCGAGGCTCACCTTCCAGGCGGGGGACCCCTACTACATCAGCAAGCGCAAGCGGGACGAGTGGCTGGCACGCTGGAAAAGGGAGGCTGAGAAGAAAGCCAAGGTCATTGCAGGAATGAATGCTGTGGAAGAAAACCAGGGGCCCGGGGAGTCTCAGAAGGTGGAGGAGGCCAGCCCTCCTGCTGTGCAGCAGCCCACTGACCCCGCATCCCCCACTGTGGCTACCACGCCTGAGCCCGTGGGGTCCGATGCTGGGGACAAGAATGCCACCAAAGCAGGCGATGACGAGCCAGAGTACGAGGACGGCCGGGGCTTTGGCATTGGGGAGCTGGTGTGGGGGAAACTGCGGGGCTTCTCCTGGTGGCCAGGCCGCATTGTGTCTTGGTGGATGACGGGCCGGAGCCGAGCAGCTGAAGGCACCCGCTGGGTCATGTGGTTCGGAGACGGCAAATTCTCAGTGGTGTGTGTTGAGAAGCTGATGCCGCTGAGCTCGTTTTGCAGTGCGTTCCACCAGGCCACGTACAACAAGCAGCCCATGTACCGCAAAGCCATCTACGAGGTCCTGCAGGTGGCCAGCAGCCGCGCGGGGAAGCTGTTCCCGGTGTGCCACGACAGCGATGAGAGTGACACTGCCAAGGCCGTGGAGGTGCAGAACAAGCCCATGATTGAATGGGCCCTGGGGGGCTTCCAGCCTTCTGGCCCTAAGGGCCTGGAGCCACCAGAAGAAGAGAAGAATCCCTACAAAGAAGTGTACACGGACATGTGGGTGGAACCTGAGGCAGCTGCCTACGCACCACCTCCACCAGCCAAAAAGCCCCGGAAGAGCACAGCGGAGAAGCCCAAGGTCAAGGAGATTATTGATGAGCGCACAAGAGAGCGGCTGGTGTACGAGGTGCGGCAGAAGTGCCGGAACATTGAGGACATCTGCATCTCCTGTGGGAGCCTCAATGTTACCCTGGAACACCCCCTCTTCGTTGGAGGAATGTGCCAAAACTGCAAGAACTGCTTTCTGGAGTGTGCGTACCAGTACGACGACGACGGCTACCAGTCCTACTGCACCATCTGCTGTGGGGGCCGTGAGGTGCTCATGTGCGGAAACAACAACTGCTGCAGGTGCTTTTGCGTGGAGTGTGTGGACCTCTTGGTGGGGCCGGGGGCTGCCCAGGCAGCCATTAAGGAAGACCCCTGGAACTGCTACATGTGCGGGCACAAGGGTACCTACGGGCTGCTGCGGCGGCGAGAGGACTGGCCCTCCCGGCTCCAGATGTTCTTCGCTAATAACCACGACCAGGAATTTGACCCTCCAAAGGTTTACCCACCTGTCCCAGCTGAGAAGAGGAAGCCCATCCGGGTGCTGTCTCTCTTTGATGGAATCGCTACAGGGCTCCTGGTGCTGAAGGACTTGGGCATTCAGGTGGACCGCTACATTGCCTCGGAGGTGTGTGAGGACTCCATCACGGTGGGCATGGTGCGGCACCAGGGGAAGATCATGTACGTCGGGGACGTCCGCAGCGTCACACAGAAGCATATCCAGGAGTGGGGCCCATTCGATCTGGTGATTGGGGGCAGTCCCTGCAATGACCTCTCCATCGTCAACCCTGCTCGCAAGGGCCTCTACGAGGGCACTGGCCGGCTCTTCTTTGAGTTCTACCGCCTCCTGCATGATGCGCGGCCCAAGGAGGGAGATGATCGCCCCTTCTTCTGGCTCTTTGAGAATGTGGTGGCCATGGGCGTTAGTGACAAGAGGGACATCTCGCGATTTCTCGAGTCCAACCCTGTGATGATTGATGCCAAAGAAGTGTCAGCTGCACACAGGGCCCGCTACTTCTGGGGTAACCTTCCCGGTATGAACAGGCCGTTGGCATCCACTGTGAATGATAAGCTGGAGCTGCAGGAGTGTCTGGAGCATGGCAGGATAGCCAAGTTCAGCAAAGTGAGGACCATTACTACGAGGTCAAACTCCATAAAGCAGGGCAAAGACCAGCATTTTCCTGTCTTCATGAATGAGAAAGAGGACATCTTATGGTGCACTGAAATGGAAAGGGTATTTGGTTTCCCAGTCCACTATACTGACGTCTCCAACATGAGCCGCTTGGCGAGGCAGAGACTGCTGGGCCGGTCATGGAGCGTGCCAGTCATCCGCCACCTCTTCGCTCCGCTGAAGGAGTATTTTGCGTGTGTGTAAG'\n",
    "\n",
    "    #Create a list of the codons they called synonymous variants at:\n",
    "    import csv\n",
    "    filename = csv_file_variants\n",
    "    with open(filename, 'r') as csvfile:\n",
    "        read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "        row_count=0\n",
    "\n",
    "        DNMT3A_synonymous_variant_list_NP_072046 = []\n",
    "        DNMT3A_synonymous_codons_NP_072046 =[]\n",
    "        DNMT3A_synonymous_variant_DICT_NP_072046 = {}\n",
    "\n",
    "        for row in read_reader: #each row contains aa number (e.g. 11) and variant (e.g. p.D11A)\n",
    "    #         print(row)\n",
    "            if row_count>0: \n",
    "                codon0_NP_072046 = int(row[0])-1\n",
    "                codon1_NP_072046 = int(row[0])\n",
    "                variant_NP_072046 = row[1]\n",
    "                DNMT3A_synonymous_codons_NP_072046.append(codon0_NP_072046) #List of codon numbers that they called variants at\n",
    "                DNMT3A_synonymous_variant_DICT_NP_072046[row_count]=(codon1_NP_072046, variant_NP_072046) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76V)\n",
    "\n",
    "            row_count=row_count+1\n",
    "            \n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    DNMT3A_synonymous_codons_NP_072046_DICT = {}\n",
    "    calls_DNMT3A_synonymous_NP_072046 = {}\n",
    "\n",
    "    for i in DNMT3A_synonymous_codons_NP_072046: #Iterates over the list of codon numbers\n",
    "        DNMT3A_synonymous_codons_NP_072046_DICT[i+1]= (DNMT3A_codons_NP_072046[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "        \n",
    "    for k,v in DNMT3A_synonymous_codons_NP_072046_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_DNMT3A_synonymous_NP_072046[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "        \n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_DNMT3A_synonymous_NP_072046_extra = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_synonymous_NP_072046.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_DNMT3A_synonymous_NP_072046_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_DNMT3A_synonymous_NP_072046_extra2 = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_synonymous_NP_072046_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_DNMT3A_synonymous_NP_072046_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "            \n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    DNMT3A_position_in_list_dict_synonymous_NP_072046={}\n",
    "    for k, v in DNMT3A_synonymous_variant_DICT_NP_072046.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_DNMT3A_synonymous_NP_072046[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_DNMT3A_synonymous_NP_072046[v[0]]\n",
    "        codon_seq=calls_DNMT3A_synonymous_NP_072046[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_DNMT3A_synonymous_NP_072046_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                DNMT3A_position_in_list_dict_synonymous_NP_072046[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "            \n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in DNMT3A_position_in_list_dict_synonymous_NP_072046.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "#     print('positions at which synonymous variants called = ', aa_positions_variants_called)\n",
    "                \n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    DNMT3A_bases_synonymous_NP_072046 = {}\n",
    "    for k, v in DNMT3A_position_in_list_dict_synonymous_NP_072046.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        DNMT3A_bases_synonymous_NP_072046[k]=(v, base_changes)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    DNMT3A_bases_synonymous2_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_synonymous_NP_072046.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        DNMT3A_bases_synonymous2_NP_072046[k] = (v, codon_positions)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    DNMT3A_bases_synonymous3_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_synonymous2_NP_072046.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        DNMT3A_bases_synonymous3_NP_072046[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    DNMT3A_bases_synonymous4_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_synonymous3_NP_072046.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_synonymous=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_DNMT3A_NP072046:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check_DNMT3A_NP072046(i)\n",
    "                sites_trio_synonymous.append([adapted_trio])\n",
    "                DNMT3A_bases_synonymous4_NP_072046[k] = (v, sites_trio_synonymous)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_synonymous = [DNMT3A_exons_extra_bases_NP_072046[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_synonymous.append(site_trio_synonymous)\n",
    "                    DNMT3A_bases_synonymous4_NP_072046[k] = (v, sites_trio_synonymous)\n",
    "                    \n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    DNMT3A_sites_synonymous5_NP_072046 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in DNMT3A_bases_synonymous4_NP_072046.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        DNMT3A_sites_synonymous5_NP_072046[a]=info\n",
    "        a=a+1\n",
    "        \n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    DNMT3A_sites_synonymous6_NP_072046={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in DNMT3A_sites_synonymous5_NP_072046.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "    \n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        DNMT3A_sites_synonymous6_NP_072046[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "        \n",
    "        \n",
    "    #Assign the mutation rates to each site...\n",
    "    DNMT3A_mutation_rates_synonymous_NP_072046 = {}\n",
    "    for k, v in DNMT3A_sites_synonymous6_NP_072046.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            DNMT3A_mutation_rates_synonymous_NP_072046[k] = (v, mutation_rate)\n",
    "\n",
    "    DNMT3A_mutation_rate_synonymous_NP_072046_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in DNMT3A_mutation_rates_synonymous_NP_072046.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        DNMT3A_mutation_rate_synonymous_NP_072046_sum.append(mutation_rate_value)\n",
    "        \n",
    "#     #Get the average of all the mutation rates\n",
    "#     DNMT3A_mutation_rate_synonymous_NP_072046_sum = []\n",
    "#     for k, v in DNMT3A_mutation_rates_synonymous_NP_072046.items():\n",
    "#         mut_rate = v[1]\n",
    "#         DNMT3A_mutation_rate_synonymous_NP_072046_sum.append(mut_rate)\n",
    "\n",
    "    DNMT3A_Average_mutation_rate_synonymous_NP_072046 = np.mean(DNMT3A_mutation_rate_synonymous_NP_072046_sum)\n",
    "    DNMT3A_Summed_mutation_rate_synonymous_NP_072046_added = np.sum(DNMT3A_mutation_rate_synonymous_NP_072046_sum)\n",
    "\n",
    "#     print('Average mutation rate across DNMT3A synonymous_NP_072046 for this study = ', DNMT3A_Average_mutation_rate_synonymous_NP_072046)\n",
    "    print('Total summed mutation rate across DNMT3A synonymous_NP_072046 for this study = ', DNMT3A_Summed_mutation_rate_synonymous_NP_072046_added)\n",
    "    \n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all #in format: column 1 = 882, column 2 = p.R882H"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Functions for calculating mutation rate from a given range of amino acid positions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DNMT3A transcript NP072046 = 912 aa (hg19 or GRCh38)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "def mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(lower_range, upper_range):\n",
    "    \n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base_DNMT3A_NP072046 = [1279, 1431, 1476, 1554, 1852, 2083, 2085, 2175, 2322, 2478]\n",
    "\n",
    "    def intron_exon_check_DNMT3A_NP072046(position_in_gene_number):\n",
    "        if position_in_gene_number == 1279:\n",
    "            gene_pos = 1279\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1429:\n",
    "            gene_pos = 1429\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1474:\n",
    "            gene_pos = 1474\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1554:\n",
    "            gene_pos = 1554\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1852:\n",
    "            gene_pos = 1852\n",
    "            first_base = 'G'\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2083:\n",
    "            gene_pos = 2083\n",
    "            first_base = 'G'\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2082:\n",
    "            gene_pos = 2082\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2173:\n",
    "            gene_pos = 2173\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2322:\n",
    "            gene_pos = 2322\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2478:\n",
    "            gene_pos = 2478\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "\n",
    "    #DNMT3A transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    DNMT3A_exons_NP_072046 = 'ATGCCCGCCATGCCCTCCAGCGGCCCCGGGGACACCAGCAGCTCTGCTGCGGAGCGGGAGGAGGACCGAAAGGACGGAGAGGAGCAGGAGGAGCCGCGTGGCAAGGAGGAGCGCCAAGAGCCCAGCACCACGGCACGGAAGGTGGGGCGGCCTGGGAGGAAGCGCAAGCACCCCCCGGTGGAAAGCGGTGACACGCCAAAGGACCCTGCGGTGATCTCCAAGTCCCCATCCATGGCCCAGGACTCAGGCGCCTCAGAGCTATTACCCAATGGGGACTTGGAGAAGCGGAGTGAGCCCCAGCCAGAGGAGGGGAGCCCTGCTGGGGGGCAGAAGGGCGGGGCCCCAGCAGAGGGAGAGGGTGCAGCTGAGACCCTGCCTGAAGCCTCAAGAGCAGTGGAAAATGGCTGCTGCACCCCCAAGGAGGGCCGAGGAGCCCCTGCAGAAGCGGGCAAAGAACAGAAGGAGACCAACATCGAATCCATGAAAATGGAGGGCTCCCGGGGCCGGCTGCGGGGTGGCTTGGGCTGGGAGTCCAGCCTCCGTCAGCGGCCCATGCCGAGGCTCACCTTCCAGGCGGGGGACCCCTACTACATCAGCAAGCGCAAGCGGGACGAGTGGCTGGCACGCTGGAAAAGGGAGGCTGAGAAGAAAGCCAAGGTCATTGCAGGAATGAATGCTGTGGAAGAAAACCAGGGGCCCGGGGAGTCTCAGAAGGTGGAGGAGGCCAGCCCTCCTGCTGTGCAGCAGCCCACTGACCCCGCATCCCCCACTGTGGCTACCACGCCTGAGCCCGTGGGGTCCGATGCTGGGGACAAGAATGCCACCAAAGCAGGCGATGACGAGCCAGAGTACGAGGACGGCCGGGGCTTTGGCATTGGGGAGCTGGTGTGGGGGAAACTGCGGGGCTTCTCCTGGTGGCCAGGCCGCATTGTGTCTTGGTGGATGACGGGCCGGAGCCGAGCAGCTGAAGGCACCCGCTGGGTCATGTGGTTCGGAGACGGCAAATTCTCAGTGGTGTGTGTTGAGAAGCTGATGCCGCTGAGCTCGTTTTGCAGTGCGTTCCACCAGGCCACGTACAACAAGCAGCCCATGTACCGCAAAGCCATCTACGAGGTCCTGCAGGTGGCCAGCAGCCGCGCGGGGAAGCTGTTCCCGGTGTGCCACGACAGCGATGAGAGTGACACTGCCAAGGCCGTGGAGGTGCAGAACAAGCCCATGATTGAATGGGCCCTGGGGGGCTTCCAGCCTTCTGGCCCTAAGGGCCTGGAGCCACCAGAAGAAGAGAAGAATCCCTACAAAGAAGTGTACACGGACATGTGGGTGGAACCTGAGGCAGCTGCCTACGCACCACCTCCACCAGCCAAAAAGCCCCGGAAGAGCACAGCGGAGAAGCCCAAGGTCAAGGAGATTATTGATGAGCGCACAAGAGAGCGGCTGGTGTACGAGGTGCGGCAGAAGTGCCGGAACATTGAGGACATCTGCATCTCCTGTGGGAGCCTCAATGTTACCCTGGAACACCCCCTCTTCGTTGGAGGAATGTGCCAAAACTGCAAGAACTGCTTTCTGGAGTGTGCGTACCAGTACGACGACGACGGCTACCAGTCCTACTGCACCATCTGCTGTGGGGGCCGTGAGGTGCTCATGTGCGGAAACAACAACTGCTGCAGGTGCTTTTGCGTGGAGTGTGTGGACCTCTTGGTGGGGCCGGGGGCTGCCCAGGCAGCCATTAAGGAAGACCCCTGGAACTGCTACATGTGCGGGCACAAGGGTACCTACGGGCTGCTGCGGCGGCGAGAGGACTGGCCCTCCCGGCTCCAGATGTTCTTCGCTAATAACCACGACCAGGAATTTGACCCTCCAAAGGTTTACCCACCTGTCCCAGCTGAGAAGAGGAAGCCCATCCGGGTGCTGTCTCTCTTTGATGGAATCGCTACAGGGCTCCTGGTGCTGAAGGACTTGGGCATTCAGGTGGACCGCTACATTGCCTCGGAGGTGTGTGAGGACTCCATCACGGTGGGCATGGTGCGGCACCAGGGGAAGATCATGTACGTCGGGGACGTCCGCAGCGTCACACAGAAGCATATCCAGGAGTGGGGCCCATTCGATCTGGTGATTGGGGGCAGTCCCTGCAATGACCTCTCCATCGTCAACCCTGCTCGCAAGGGCCTCTACGAGGGCACTGGCCGGCTCTTCTTTGAGTTCTACCGCCTCCTGCATGATGCGCGGCCCAAGGAGGGAGATGATCGCCCCTTCTTCTGGCTCTTTGAGAATGTGGTGGCCATGGGCGTTAGTGACAAGAGGGACATCTCGCGATTTCTCGAGTCCAACCCTGTGATGATTGATGCCAAAGAAGTGTCAGCTGCACACAGGGCCCGCTACTTCTGGGGTAACCTTCCCGGTATGAACAGGCCGTTGGCATCCACTGTGAATGATAAGCTGGAGCTGCAGGAGTGTCTGGAGCATGGCAGGATAGCCAAGTTCAGCAAAGTGAGGACCATTACTACGAGGTCAAACTCCATAAAGCAGGGCAAAGACCAGCATTTTCCTGTCTTCATGAATGAGAAAGAGGACATCTTATGGTGCACTGAAATGGAAAGGGTATTTGGTTTCCCAGTCCACTATACTGACGTCTCCAACATGAGCCGCTTGGCGAGGCAGAGACTGCTGGGCCGGTCATGGAGCGTGCCAGTCATCCGCCACCTCTTCGCTCCGCTGAAGGAGTATTTTGCGTGTGTGTAA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    DNMT3A_codons_NP_072046 = [DNMT3A_exons_NP_072046[i:i+3] for i in range(0, len(DNMT3A_exons_NP_072046), 3)]\n",
    "\n",
    "    #DNMT3A transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    DNMT3A_exons_extra_bases_NP_072046 = 'GATGCCCGCCATGCCCTCCAGCGGCCCCGGGGACACCAGCAGCTCTGCTGCGGAGCGGGAGGAGGACCGAAAGGACGGAGAGGAGCAGGAGGAGCCGCGTGGCAAGGAGGAGCGCCAAGAGCCCAGCACCACGGCACGGAAGGTGGGGCGGCCTGGGAGGAAGCGCAAGCACCCCCCGGTGGAAAGCGGTGACACGCCAAAGGACCCTGCGGTGATCTCCAAGTCCCCATCCATGGCCCAGGACTCAGGCGCCTCAGAGCTATTACCCAATGGGGACTTGGAGAAGCGGAGTGAGCCCCAGCCAGAGGAGGGGAGCCCTGCTGGGGGGCAGAAGGGCGGGGCCCCAGCAGAGGGAGAGGGTGCAGCTGAGACCCTGCCTGAAGCCTCAAGAGCAGTGGAAAATGGCTGCTGCACCCCCAAGGAGGGCCGAGGAGCCCCTGCAGAAGCGGGCAAAGAACAGAAGGAGACCAACATCGAATCCATGAAAATGGAGGGCTCCCGGGGCCGGCTGCGGGGTGGCTTGGGCTGGGAGTCCAGCCTCCGTCAGCGGCCCATGCCGAGGCTCACCTTCCAGGCGGGGGACCCCTACTACATCAGCAAGCGCAAGCGGGACGAGTGGCTGGCACGCTGGAAAAGGGAGGCTGAGAAGAAAGCCAAGGTCATTGCAGGAATGAATGCTGTGGAAGAAAACCAGGGGCCCGGGGAGTCTCAGAAGGTGGAGGAGGCCAGCCCTCCTGCTGTGCAGCAGCCCACTGACCCCGCATCCCCCACTGTGGCTACCACGCCTGAGCCCGTGGGGTCCGATGCTGGGGACAAGAATGCCACCAAAGCAGGCGATGACGAGCCAGAGTACGAGGACGGCCGGGGCTTTGGCATTGGGGAGCTGGTGTGGGGGAAACTGCGGGGCTTCTCCTGGTGGCCAGGCCGCATTGTGTCTTGGTGGATGACGGGCCGGAGCCGAGCAGCTGAAGGCACCCGCTGGGTCATGTGGTTCGGAGACGGCAAATTCTCAGTGGTGTGTGTTGAGAAGCTGATGCCGCTGAGCTCGTTTTGCAGTGCGTTCCACCAGGCCACGTACAACAAGCAGCCCATGTACCGCAAAGCCATCTACGAGGTCCTGCAGGTGGCCAGCAGCCGCGCGGGGAAGCTGTTCCCGGTGTGCCACGACAGCGATGAGAGTGACACTGCCAAGGCCGTGGAGGTGCAGAACAAGCCCATGATTGAATGGGCCCTGGGGGGCTTCCAGCCTTCTGGCCCTAAGGGCCTGGAGCCACCAGAAGAAGAGAAGAATCCCTACAAAGAAGTGTACACGGACATGTGGGTGGAACCTGAGGCAGCTGCCTACGCACCACCTCCACCAGCCAAAAAGCCCCGGAAGAGCACAGCGGAGAAGCCCAAGGTCAAGGAGATTATTGATGAGCGCACAAGAGAGCGGCTGGTGTACGAGGTGCGGCAGAAGTGCCGGAACATTGAGGACATCTGCATCTCCTGTGGGAGCCTCAATGTTACCCTGGAACACCCCCTCTTCGTTGGAGGAATGTGCCAAAACTGCAAGAACTGCTTTCTGGAGTGTGCGTACCAGTACGACGACGACGGCTACCAGTCCTACTGCACCATCTGCTGTGGGGGCCGTGAGGTGCTCATGTGCGGAAACAACAACTGCTGCAGGTGCTTTTGCGTGGAGTGTGTGGACCTCTTGGTGGGGCCGGGGGCTGCCCAGGCAGCCATTAAGGAAGACCCCTGGAACTGCTACATGTGCGGGCACAAGGGTACCTACGGGCTGCTGCGGCGGCGAGAGGACTGGCCCTCCCGGCTCCAGATGTTCTTCGCTAATAACCACGACCAGGAATTTGACCCTCCAAAGGTTTACCCACCTGTCCCAGCTGAGAAGAGGAAGCCCATCCGGGTGCTGTCTCTCTTTGATGGAATCGCTACAGGGCTCCTGGTGCTGAAGGACTTGGGCATTCAGGTGGACCGCTACATTGCCTCGGAGGTGTGTGAGGACTCCATCACGGTGGGCATGGTGCGGCACCAGGGGAAGATCATGTACGTCGGGGACGTCCGCAGCGTCACACAGAAGCATATCCAGGAGTGGGGCCCATTCGATCTGGTGATTGGGGGCAGTCCCTGCAATGACCTCTCCATCGTCAACCCTGCTCGCAAGGGCCTCTACGAGGGCACTGGCCGGCTCTTCTTTGAGTTCTACCGCCTCCTGCATGATGCGCGGCCCAAGGAGGGAGATGATCGCCCCTTCTTCTGGCTCTTTGAGAATGTGGTGGCCATGGGCGTTAGTGACAAGAGGGACATCTCGCGATTTCTCGAGTCCAACCCTGTGATGATTGATGCCAAAGAAGTGTCAGCTGCACACAGGGCCCGCTACTTCTGGGGTAACCTTCCCGGTATGAACAGGCCGTTGGCATCCACTGTGAATGATAAGCTGGAGCTGCAGGAGTGTCTGGAGCATGGCAGGATAGCCAAGTTCAGCAAAGTGAGGACCATTACTACGAGGTCAAACTCCATAAAGCAGGGCAAAGACCAGCATTTTCCTGTCTTCATGAATGAGAAAGAGGACATCTTATGGTGCACTGAAATGGAAAGGGTATTTGGTTTCCCAGTCCACTATACTGACGTCTCCAACATGAGCCGCTTGGCGAGGCAGAGACTGCTGGGCCGGTCATGGAGCGTGCCAGTCATCCGCCACCTCTTCGCTCCGCTGAAGGAGTATTTTGCGTGTGTGTAAG'\n",
    "\n",
    "    #List of all amino acid positions (across whole coding region) covered (amino acid position minus 1 (as python starts at 0))\n",
    "    DNMT3A_nonsense_codons_DICT_NP_072046={}\n",
    "    for i in range(lower_range, upper_range):\n",
    "        DNMT3A_nonsense_codons_DICT_NP_072046[i+1]= (DNMT3A_codons_NP_072046[i])\n",
    "\n",
    "    calls_DNMT3A_nonsense_NP_072046 = {}\n",
    "    for k,v in DNMT3A_nonsense_codons_DICT_NP_072046.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_DNMT3A_nonsense_NP_072046[k]=(v, changes)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_DNMT3A_nonsense_NP_072046_extra = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_nonsense_NP_072046.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_DNMT3A_nonsense_NP_072046_extra[k] = (Amino_acid_change, v[0], v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_DNMT3A_nonsense_NP_072046_extra2 = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_nonsense_NP_072046_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_DNMT3A_nonsense_NP_072046_extra2[k] = (Amino_acid_abbrev, v[1], v[2])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary that contains details of the positions of nonsense variants at that codon\n",
    "    DNMT3A_position_in_list_dict_nonsense_NP_072046 = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_nonsense_NP_072046_extra2.items():\n",
    "        original_AA = v[0]\n",
    "        possible_AA_changes = v[2]\n",
    "        possible_AA_changes_string = ''.join(v[2])\n",
    "        positions_in_list = []\n",
    "        for i in possible_AA_changes:\n",
    "            if i == '*': #nonsense\n",
    "                position_in_list=-1\n",
    "                nonsense_change = i    \n",
    "                while True:\n",
    "                    position_in_list = possible_AA_changes_string.find(nonsense_change, position_in_list+1)\n",
    "    #               print(position_in_list)\n",
    "                    if position_in_list == -1: break\n",
    "                    positions_in_list.append(position_in_list)\n",
    "                    positions_in_list = set(positions_in_list) #convert into a 'set' to remove duplicate positions\n",
    "                    positions_in_list = list(positions_in_list) #convert back to a list so can be added to dictionary\n",
    "                DNMT3A_position_in_list_dict_nonsense_NP_072046[k]=(positions_in_list, v)\n",
    "\n",
    "        aa_positions_variants_called = []\n",
    "        for k, v in DNMT3A_position_in_list_dict_nonsense_NP_072046.items():\n",
    "            aa_positions_variants_called.append(k)\n",
    "    # print('positions at which nonsense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    DNMT3A_bases_nonsense_NP_072046 = {}\n",
    "    for k, v in DNMT3A_position_in_list_dict_nonsense_NP_072046.items():\n",
    "        original_a_a = v[1][0]\n",
    "        codon_seq = v[1][1]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        DNMT3A_bases_nonsense_NP_072046[k]=(v, base_changes, original_a_a)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    DNMT3A_bases_nonsense2_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_nonsense_NP_072046.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        DNMT3A_bases_nonsense2_NP_072046[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    DNMT3A_bases_nonsense3_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_nonsense2_NP_072046.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_a_a = v[0][2]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = k\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(k)-1)*3)+codon_pos #codon number minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][1][1][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        DNMT3A_bases_nonsense3_NP_072046[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, original_a_a)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    DNMT3A_bases_nonsense4_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_nonsense3_NP_072046.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_nonsense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_DNMT3A_NP072046:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check_DNMT3A_NP072046(i)\n",
    "                sites_trio_nonsense.append([adapted_trio])\n",
    "                DNMT3A_bases_nonsense4_NP_072046[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_nonsense = [DNMT3A_exons_extra_bases_NP_072046[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_nonsense.append(site_trio_nonsense)\n",
    "                    DNMT3A_bases_nonsense4_NP_072046[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    DNMT3A_sites_nonsense5_NP_072046 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in DNMT3A_bases_nonsense4_NP_072046.items():\n",
    "        original_a_a = v[0][0][0]\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, original_a_a])\n",
    "            change = change+1\n",
    "        DNMT3A_sites_nonsense5_NP_072046[a]=info\n",
    "        a=a+1\n",
    "    #     print(DNMT3A_sites_nonsense5_NP_072046)\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    DNMT3A_sites_nonsense6_NP_072046={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in DNMT3A_sites_nonsense5_NP_072046.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        original_a_a = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        DNMT3A_sites_nonsense6_NP_072046[k] = (v[0],changes3, codon_n_912, original_a_a) #site_trio = the affected base and its neighbours\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    DNMT3A_mutation_rates_nonsense_NP_072046 = {}\n",
    "    for k, v in DNMT3A_sites_nonsense6_NP_072046.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            original_a_a = v[3]\n",
    "            aa_positioning = v[2]\n",
    "            variant_type = '*'\n",
    "            variant_id = str(original_a_a)+str(aa_positioning)+variant_type\n",
    "            DNMT3A_mutation_rates_nonsense_NP_072046[k] = (v[0],v[1], mutation_rate, variant_id)\n",
    "\n",
    "    DNMT3A_mutation_rate_nonsense_NP_072046_sum = []\n",
    "\n",
    "    mutation_rates_list = []\n",
    "    for k, v in DNMT3A_mutation_rates_nonsense_NP_072046.items():\n",
    "        codon_position_in_gene=(v[0])\n",
    "        variant = v[3]\n",
    "        site = v[1]\n",
    "        mutation_rate_value = v[2] \n",
    "        mutation_rates_list.append((v[0], variant, site, mutation_rate_value))\n",
    "        DNMT3A_mutation_rate_nonsense_NP_072046_sum.append(mutation_rate_value)\n",
    "\n",
    "    #Get the average of all the mutation rates\n",
    "    DNMT3A_mutation_rate_nonsense_NP_072046_sum = []\n",
    "    for k, v in DNMT3A_mutation_rates_nonsense_NP_072046.items():\n",
    "        mut_rate = v[2]\n",
    "        DNMT3A_mutation_rate_nonsense_NP_072046_sum.append(mut_rate)\n",
    "\n",
    "    DNMT3A_Average_mutation_rate_nonsense_NP_072046 = np.mean(DNMT3A_mutation_rate_nonsense_NP_072046_sum)\n",
    "    DNMT3A_Summed_mutation_rate_nonsense_NP_072046_added = np.sum(DNMT3A_mutation_rate_nonsense_NP_072046_sum)\n",
    "\n",
    "#     print('Average mutation rate across DNMT3A nonsense_NP_072046 for this study = ', DNMT3A_Average_mutation_rate_nonsense_NP_072046)\n",
    "    print('Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study = ', DNMT3A_Summed_mutation_rate_nonsense_NP_072046_added)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all # e.g. lower= 0, upper = 913 for whole gene\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "def mutation_rates_missense_range_of_variants_DNMT3A_NP072046(lower_range, upper_range):\n",
    "    \n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base_DNMT3A_NP072046 = [1279, 1429, 1474, 1554, 1852, 2083, 2082, 2173, 2322, 2478]\n",
    "\n",
    "    def intron_exon_check_DNMT3A_NP072046(position_in_gene_number):\n",
    "        if position_in_gene_number == 1279:\n",
    "            gene_pos = 1279\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1429:\n",
    "            gene_pos = 1429\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1474:\n",
    "            gene_pos = 1474\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1554:\n",
    "            gene_pos = 1554\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1852:\n",
    "            gene_pos = 1852\n",
    "            first_base = 'G'\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2083:\n",
    "            gene_pos = 2083\n",
    "            first_base = 'G'\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2082:\n",
    "            gene_pos = 2082\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2173:\n",
    "            gene_pos = 2173\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2322:\n",
    "            gene_pos = 2322\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2478:\n",
    "            gene_pos = 2478\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "\n",
    "    #DNMT3A transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    DNMT3A_exons_NP_072046 = 'ATGCCCGCCATGCCCTCCAGCGGCCCCGGGGACACCAGCAGCTCTGCTGCGGAGCGGGAGGAGGACCGAAAGGACGGAGAGGAGCAGGAGGAGCCGCGTGGCAAGGAGGAGCGCCAAGAGCCCAGCACCACGGCACGGAAGGTGGGGCGGCCTGGGAGGAAGCGCAAGCACCCCCCGGTGGAAAGCGGTGACACGCCAAAGGACCCTGCGGTGATCTCCAAGTCCCCATCCATGGCCCAGGACTCAGGCGCCTCAGAGCTATTACCCAATGGGGACTTGGAGAAGCGGAGTGAGCCCCAGCCAGAGGAGGGGAGCCCTGCTGGGGGGCAGAAGGGCGGGGCCCCAGCAGAGGGAGAGGGTGCAGCTGAGACCCTGCCTGAAGCCTCAAGAGCAGTGGAAAATGGCTGCTGCACCCCCAAGGAGGGCCGAGGAGCCCCTGCAGAAGCGGGCAAAGAACAGAAGGAGACCAACATCGAATCCATGAAAATGGAGGGCTCCCGGGGCCGGCTGCGGGGTGGCTTGGGCTGGGAGTCCAGCCTCCGTCAGCGGCCCATGCCGAGGCTCACCTTCCAGGCGGGGGACCCCTACTACATCAGCAAGCGCAAGCGGGACGAGTGGCTGGCACGCTGGAAAAGGGAGGCTGAGAAGAAAGCCAAGGTCATTGCAGGAATGAATGCTGTGGAAGAAAACCAGGGGCCCGGGGAGTCTCAGAAGGTGGAGGAGGCCAGCCCTCCTGCTGTGCAGCAGCCCACTGACCCCGCATCCCCCACTGTGGCTACCACGCCTGAGCCCGTGGGGTCCGATGCTGGGGACAAGAATGCCACCAAAGCAGGCGATGACGAGCCAGAGTACGAGGACGGCCGGGGCTTTGGCATTGGGGAGCTGGTGTGGGGGAAACTGCGGGGCTTCTCCTGGTGGCCAGGCCGCATTGTGTCTTGGTGGATGACGGGCCGGAGCCGAGCAGCTGAAGGCACCCGCTGGGTCATGTGGTTCGGAGACGGCAAATTCTCAGTGGTGTGTGTTGAGAAGCTGATGCCGCTGAGCTCGTTTTGCAGTGCGTTCCACCAGGCCACGTACAACAAGCAGCCCATGTACCGCAAAGCCATCTACGAGGTCCTGCAGGTGGCCAGCAGCCGCGCGGGGAAGCTGTTCCCGGTGTGCCACGACAGCGATGAGAGTGACACTGCCAAGGCCGTGGAGGTGCAGAACAAGCCCATGATTGAATGGGCCCTGGGGGGCTTCCAGCCTTCTGGCCCTAAGGGCCTGGAGCCACCAGAAGAAGAGAAGAATCCCTACAAAGAAGTGTACACGGACATGTGGGTGGAACCTGAGGCAGCTGCCTACGCACCACCTCCACCAGCCAAAAAGCCCCGGAAGAGCACAGCGGAGAAGCCCAAGGTCAAGGAGATTATTGATGAGCGCACAAGAGAGCGGCTGGTGTACGAGGTGCGGCAGAAGTGCCGGAACATTGAGGACATCTGCATCTCCTGTGGGAGCCTCAATGTTACCCTGGAACACCCCCTCTTCGTTGGAGGAATGTGCCAAAACTGCAAGAACTGCTTTCTGGAGTGTGCGTACCAGTACGACGACGACGGCTACCAGTCCTACTGCACCATCTGCTGTGGGGGCCGTGAGGTGCTCATGTGCGGAAACAACAACTGCTGCAGGTGCTTTTGCGTGGAGTGTGTGGACCTCTTGGTGGGGCCGGGGGCTGCCCAGGCAGCCATTAAGGAAGACCCCTGGAACTGCTACATGTGCGGGCACAAGGGTACCTACGGGCTGCTGCGGCGGCGAGAGGACTGGCCCTCCCGGCTCCAGATGTTCTTCGCTAATAACCACGACCAGGAATTTGACCCTCCAAAGGTTTACCCACCTGTCCCAGCTGAGAAGAGGAAGCCCATCCGGGTGCTGTCTCTCTTTGATGGAATCGCTACAGGGCTCCTGGTGCTGAAGGACTTGGGCATTCAGGTGGACCGCTACATTGCCTCGGAGGTGTGTGAGGACTCCATCACGGTGGGCATGGTGCGGCACCAGGGGAAGATCATGTACGTCGGGGACGTCCGCAGCGTCACACAGAAGCATATCCAGGAGTGGGGCCCATTCGATCTGGTGATTGGGGGCAGTCCCTGCAATGACCTCTCCATCGTCAACCCTGCTCGCAAGGGCCTCTACGAGGGCACTGGCCGGCTCTTCTTTGAGTTCTACCGCCTCCTGCATGATGCGCGGCCCAAGGAGGGAGATGATCGCCCCTTCTTCTGGCTCTTTGAGAATGTGGTGGCCATGGGCGTTAGTGACAAGAGGGACATCTCGCGATTTCTCGAGTCCAACCCTGTGATGATTGATGCCAAAGAAGTGTCAGCTGCACACAGGGCCCGCTACTTCTGGGGTAACCTTCCCGGTATGAACAGGCCGTTGGCATCCACTGTGAATGATAAGCTGGAGCTGCAGGAGTGTCTGGAGCATGGCAGGATAGCCAAGTTCAGCAAAGTGAGGACCATTACTACGAGGTCAAACTCCATAAAGCAGGGCAAAGACCAGCATTTTCCTGTCTTCATGAATGAGAAAGAGGACATCTTATGGTGCACTGAAATGGAAAGGGTATTTGGTTTCCCAGTCCACTATACTGACGTCTCCAACATGAGCCGCTTGGCGAGGCAGAGACTGCTGGGCCGGTCATGGAGCGTGCCAGTCATCCGCCACCTCTTCGCTCCGCTGAAGGAGTATTTTGCGTGTGTGTAA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    DNMT3A_codons_NP_072046 = [DNMT3A_exons_NP_072046[i:i+3] for i in range(0, len(DNMT3A_exons_NP_072046), 3)]\n",
    "\n",
    "    #DNMT3A transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    DNMT3A_exons_extra_bases_NP_072046 = 'GATGCCCGCCATGCCCTCCAGCGGCCCCGGGGACACCAGCAGCTCTGCTGCGGAGCGGGAGGAGGACCGAAAGGACGGAGAGGAGCAGGAGGAGCCGCGTGGCAAGGAGGAGCGCCAAGAGCCCAGCACCACGGCACGGAAGGTGGGGCGGCCTGGGAGGAAGCGCAAGCACCCCCCGGTGGAAAGCGGTGACACGCCAAAGGACCCTGCGGTGATCTCCAAGTCCCCATCCATGGCCCAGGACTCAGGCGCCTCAGAGCTATTACCCAATGGGGACTTGGAGAAGCGGAGTGAGCCCCAGCCAGAGGAGGGGAGCCCTGCTGGGGGGCAGAAGGGCGGGGCCCCAGCAGAGGGAGAGGGTGCAGCTGAGACCCTGCCTGAAGCCTCAAGAGCAGTGGAAAATGGCTGCTGCACCCCCAAGGAGGGCCGAGGAGCCCCTGCAGAAGCGGGCAAAGAACAGAAGGAGACCAACATCGAATCCATGAAAATGGAGGGCTCCCGGGGCCGGCTGCGGGGTGGCTTGGGCTGGGAGTCCAGCCTCCGTCAGCGGCCCATGCCGAGGCTCACCTTCCAGGCGGGGGACCCCTACTACATCAGCAAGCGCAAGCGGGACGAGTGGCTGGCACGCTGGAAAAGGGAGGCTGAGAAGAAAGCCAAGGTCATTGCAGGAATGAATGCTGTGGAAGAAAACCAGGGGCCCGGGGAGTCTCAGAAGGTGGAGGAGGCCAGCCCTCCTGCTGTGCAGCAGCCCACTGACCCCGCATCCCCCACTGTGGCTACCACGCCTGAGCCCGTGGGGTCCGATGCTGGGGACAAGAATGCCACCAAAGCAGGCGATGACGAGCCAGAGTACGAGGACGGCCGGGGCTTTGGCATTGGGGAGCTGGTGTGGGGGAAACTGCGGGGCTTCTCCTGGTGGCCAGGCCGCATTGTGTCTTGGTGGATGACGGGCCGGAGCCGAGCAGCTGAAGGCACCCGCTGGGTCATGTGGTTCGGAGACGGCAAATTCTCAGTGGTGTGTGTTGAGAAGCTGATGCCGCTGAGCTCGTTTTGCAGTGCGTTCCACCAGGCCACGTACAACAAGCAGCCCATGTACCGCAAAGCCATCTACGAGGTCCTGCAGGTGGCCAGCAGCCGCGCGGGGAAGCTGTTCCCGGTGTGCCACGACAGCGATGAGAGTGACACTGCCAAGGCCGTGGAGGTGCAGAACAAGCCCATGATTGAATGGGCCCTGGGGGGCTTCCAGCCTTCTGGCCCTAAGGGCCTGGAGCCACCAGAAGAAGAGAAGAATCCCTACAAAGAAGTGTACACGGACATGTGGGTGGAACCTGAGGCAGCTGCCTACGCACCACCTCCACCAGCCAAAAAGCCCCGGAAGAGCACAGCGGAGAAGCCCAAGGTCAAGGAGATTATTGATGAGCGCACAAGAGAGCGGCTGGTGTACGAGGTGCGGCAGAAGTGCCGGAACATTGAGGACATCTGCATCTCCTGTGGGAGCCTCAATGTTACCCTGGAACACCCCCTCTTCGTTGGAGGAATGTGCCAAAACTGCAAGAACTGCTTTCTGGAGTGTGCGTACCAGTACGACGACGACGGCTACCAGTCCTACTGCACCATCTGCTGTGGGGGCCGTGAGGTGCTCATGTGCGGAAACAACAACTGCTGCAGGTGCTTTTGCGTGGAGTGTGTGGACCTCTTGGTGGGGCCGGGGGCTGCCCAGGCAGCCATTAAGGAAGACCCCTGGAACTGCTACATGTGCGGGCACAAGGGTACCTACGGGCTGCTGCGGCGGCGAGAGGACTGGCCCTCCCGGCTCCAGATGTTCTTCGCTAATAACCACGACCAGGAATTTGACCCTCCAAAGGTTTACCCACCTGTCCCAGCTGAGAAGAGGAAGCCCATCCGGGTGCTGTCTCTCTTTGATGGAATCGCTACAGGGCTCCTGGTGCTGAAGGACTTGGGCATTCAGGTGGACCGCTACATTGCCTCGGAGGTGTGTGAGGACTCCATCACGGTGGGCATGGTGCGGCACCAGGGGAAGATCATGTACGTCGGGGACGTCCGCAGCGTCACACAGAAGCATATCCAGGAGTGGGGCCCATTCGATCTGGTGATTGGGGGCAGTCCCTGCAATGACCTCTCCATCGTCAACCCTGCTCGCAAGGGCCTCTACGAGGGCACTGGCCGGCTCTTCTTTGAGTTCTACCGCCTCCTGCATGATGCGCGGCCCAAGGAGGGAGATGATCGCCCCTTCTTCTGGCTCTTTGAGAATGTGGTGGCCATGGGCGTTAGTGACAAGAGGGACATCTCGCGATTTCTCGAGTCCAACCCTGTGATGATTGATGCCAAAGAAGTGTCAGCTGCACACAGGGCCCGCTACTTCTGGGGTAACCTTCCCGGTATGAACAGGCCGTTGGCATCCACTGTGAATGATAAGCTGGAGCTGCAGGAGTGTCTGGAGCATGGCAGGATAGCCAAGTTCAGCAAAGTGAGGACCATTACTACGAGGTCAAACTCCATAAAGCAGGGCAAAGACCAGCATTTTCCTGTCTTCATGAATGAGAAAGAGGACATCTTATGGTGCACTGAAATGGAAAGGGTATTTGGTTTCCCAGTCCACTATACTGACGTCTCCAACATGAGCCGCTTGGCGAGGCAGAGACTGCTGGGCCGGTCATGGAGCGTGCCAGTCATCCGCCACCTCTTCGCTCCGCTGAAGGAGTATTTTGCGTGTGTGTAAG'\n",
    "\n",
    "    #List of all amino acid positions (across whole coding region) covered (amino acid position minus 1 (as python starts at 0))\n",
    "    DNMT3A_missense_codons_DICT_NP_072046={}\n",
    "    for i in range(lower_range, upper_range):\n",
    "        DNMT3A_missense_codons_DICT_NP_072046[i+1]= (DNMT3A_codons_NP_072046[i])\n",
    "\n",
    "    calls_DNMT3A_missense_NP_072046 = {}\n",
    "    for k,v in DNMT3A_missense_codons_DICT_NP_072046.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_DNMT3A_missense_NP_072046[k]=(v, changes)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_DNMT3A_missense_NP_072046_extra = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_missense_NP_072046.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_DNMT3A_missense_NP_072046_extra[k] = (Amino_acid_change, v[0], v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_DNMT3A_missense_NP_072046_extra2 = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_missense_NP_072046_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_DNMT3A_missense_NP_072046_extra2[k] = (Amino_acid_abbrev, v[1], v[2])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary that contains details of the positions of missense variants at that codon\n",
    "    DNMT3A_position_in_list_dict_missense_NP_072046 = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_missense_NP_072046_extra2.items():\n",
    "        original_AA = v[0]\n",
    "        possible_AA_changes = v[2]\n",
    "        possible_AA_changes_string = ''.join(v[2])\n",
    "        positions_in_list = []\n",
    "        for i in possible_AA_changes:\n",
    "            if i != original_AA: #not synonymous\n",
    "                if i != '*': #not missense\n",
    "                    position_in_list=-1\n",
    "                    missense_change = i\n",
    "                    new_aa_list = []\n",
    "                    while True:\n",
    "                        position_in_list = possible_AA_changes_string.find(missense_change, position_in_list+1)\n",
    "        #               print(position_in_list)\n",
    "                        if position_in_list == -1: break\n",
    "                        positions_in_list.append(position_in_list)\n",
    "                        positions_in_list = set(positions_in_list) #convert into a 'set' to remove duplicate positions\n",
    "                        positions_in_list = list(positions_in_list) #convert back to a list so can be added to dictionary\n",
    "                    for pos in positions_in_list:\n",
    "                        new_aa = possible_AA_changes_string[pos]\n",
    "                        new_aa_list.append(new_aa)\n",
    "                    DNMT3A_position_in_list_dict_missense_NP_072046[k]=(positions_in_list, v, new_aa_list)\n",
    "\n",
    "        aa_positions_variants_called = []\n",
    "        for k, v in DNMT3A_position_in_list_dict_missense_NP_072046.items():\n",
    "            aa_positions_variants_called.append(k)\n",
    "    # print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    DNMT3A_bases_missense_NP_072046 = {}\n",
    "    for k, v in DNMT3A_position_in_list_dict_missense_NP_072046.items():\n",
    "        original_a_a = v[1][0]\n",
    "        codon_seq = v[1][1]\n",
    "        base_changes=[]\n",
    "        new_a_a_list = v[2]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        DNMT3A_bases_missense_NP_072046[k]=(v, base_changes, original_a_a, new_a_a_list)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    DNMT3A_bases_missense2_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_missense_NP_072046.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        new_a_a_list = v[3]\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        DNMT3A_bases_missense2_NP_072046[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    DNMT3A_bases_missense3_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_missense2_NP_072046.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_a_a = v[0][2]\n",
    "        new_aa_list = v[0][0][2]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = k\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(k)-1)*3)+codon_pos #codon number minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][1][1][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        DNMT3A_bases_missense3_NP_072046[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, original_a_a, new_aa_list)\n",
    "\n",
    "    # Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    DNMT3A_bases_missense4_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_missense3_NP_072046.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "        new_a_a_list = v[6]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_DNMT3A_NP072046:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check_DNMT3A_NP072046(i)\n",
    "                sites_trio_missense.append([adapted_trio])\n",
    "                DNMT3A_bases_missense4_NP_072046[k] = (v, sites_trio_missense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_missense = [DNMT3A_exons_extra_bases_NP_072046[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_missense.append(site_trio_missense)\n",
    "                    DNMT3A_bases_missense4_NP_072046[k] = (v, sites_trio_missense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    DNMT3A_sites_missense5_NP_072046 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in DNMT3A_bases_missense4_NP_072046.items():\n",
    "        original_a_a = v[0][0][0]\n",
    "        new_a_a = v[0][6]\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            new_amino_acid = new_a_a[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, original_a_a, new_amino_acid])\n",
    "            change = change+1\n",
    "        DNMT3A_sites_missense5_NP_072046[a]=info\n",
    "        a=a+1\n",
    "    #     print(DNMT3A_sites_missense5_NP_072046)\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    DNMT3A_sites_missense6_NP_072046={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in DNMT3A_sites_missense5_NP_072046.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        original_a_a = v[5]\n",
    "        new_a_a = v[6]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        DNMT3A_sites_missense6_NP_072046[k] = (v[0],changes3, codon_n_912, original_a_a, new_a_a) #site_trio = the affected base and its neighbours\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    DNMT3A_mutation_rates_missense_NP_072046 = {}\n",
    "    for k, v in DNMT3A_sites_missense6_NP_072046.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            original_a_a = v[3]\n",
    "            new_a_a = v[4]\n",
    "            aa_positioning = v[2]\n",
    "            variant_id = str(original_a_a)+str(aa_positioning)+str(new_a_a)\n",
    "            DNMT3A_mutation_rates_missense_NP_072046[k] = (v[0],v[1], mutation_rate, variant_id)\n",
    "\n",
    "    DNMT3A_mutation_rate_missense_NP_072046_sum = []\n",
    "\n",
    "    mutation_rates_list = []\n",
    "    for k, v in DNMT3A_mutation_rates_missense_NP_072046.items():\n",
    "        codon_position_in_gene=(v[0])\n",
    "        variant = v[3]\n",
    "        site = v[1]\n",
    "        mutation_rate_value = v[2] \n",
    "        mutation_rates_list.append((v[0], variant, site, mutation_rate_value))\n",
    "        DNMT3A_mutation_rate_missense_NP_072046_sum.append(mutation_rate_value)\n",
    "\n",
    "    #Get the average of all the mutation rates\n",
    "    DNMT3A_mutation_rate_missense_NP_072046_sum = []\n",
    "    for k, v in DNMT3A_mutation_rates_missense_NP_072046.items():\n",
    "        mut_rate = v[2]\n",
    "        DNMT3A_mutation_rate_missense_NP_072046_sum.append(mut_rate)\n",
    "\n",
    "    DNMT3A_Average_mutation_rate_missense_NP_072046 = np.mean(DNMT3A_mutation_rate_missense_NP_072046_sum)\n",
    "    DNMT3A_Summed_mutation_rate_missense_NP_072046_added = np.sum(DNMT3A_mutation_rate_missense_NP_072046_sum)\n",
    "\n",
    "#     print('Average mutation rate across DNMT3A missense_NP_072046 for this study = ', DNMT3A_Average_mutation_rate_missense_NP_072046)\n",
    "    print('Total summed mutation rate across DNMT3A missense_NP_072046 for this study = ', DNMT3A_Summed_mutation_rate_missense_NP_072046_added)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all # e.g. lower= 0, upper = 913 for whole gene\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "def mutation_rates_synonymous_range_of_variants_DNMT3A_NP072046(lower_range, upper_range):\n",
    "    intron_exon_boundary_caution_base_DNMT3A_NP072046 = [1279, 1429, 1474, 1554, 1852, 2083, 2082, 2173, 2322, 2478]\n",
    "\n",
    "    def intron_exon_check_DNMT3A_NP072046(position_in_gene_number):\n",
    "        if position_in_gene_number == 1279:\n",
    "            gene_pos = 1279\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1429:\n",
    "            gene_pos = 1429\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1474:\n",
    "            gene_pos = 1474\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1554:\n",
    "            gene_pos = 1554\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1852:\n",
    "            gene_pos = 1852\n",
    "            first_base = 'G'\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2083:\n",
    "            gene_pos = 2083\n",
    "            first_base = 'G'\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2082:\n",
    "            gene_pos = 2082\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2173:\n",
    "            gene_pos = 2173\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2322:\n",
    "            gene_pos = 2322\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2478:\n",
    "            gene_pos = 2478\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "\n",
    "    #DNMT3A transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    DNMT3A_exons_NP_072046 = 'ATGCCCGCCATGCCCTCCAGCGGCCCCGGGGACACCAGCAGCTCTGCTGCGGAGCGGGAGGAGGACCGAAAGGACGGAGAGGAGCAGGAGGAGCCGCGTGGCAAGGAGGAGCGCCAAGAGCCCAGCACCACGGCACGGAAGGTGGGGCGGCCTGGGAGGAAGCGCAAGCACCCCCCGGTGGAAAGCGGTGACACGCCAAAGGACCCTGCGGTGATCTCCAAGTCCCCATCCATGGCCCAGGACTCAGGCGCCTCAGAGCTATTACCCAATGGGGACTTGGAGAAGCGGAGTGAGCCCCAGCCAGAGGAGGGGAGCCCTGCTGGGGGGCAGAAGGGCGGGGCCCCAGCAGAGGGAGAGGGTGCAGCTGAGACCCTGCCTGAAGCCTCAAGAGCAGTGGAAAATGGCTGCTGCACCCCCAAGGAGGGCCGAGGAGCCCCTGCAGAAGCGGGCAAAGAACAGAAGGAGACCAACATCGAATCCATGAAAATGGAGGGCTCCCGGGGCCGGCTGCGGGGTGGCTTGGGCTGGGAGTCCAGCCTCCGTCAGCGGCCCATGCCGAGGCTCACCTTCCAGGCGGGGGACCCCTACTACATCAGCAAGCGCAAGCGGGACGAGTGGCTGGCACGCTGGAAAAGGGAGGCTGAGAAGAAAGCCAAGGTCATTGCAGGAATGAATGCTGTGGAAGAAAACCAGGGGCCCGGGGAGTCTCAGAAGGTGGAGGAGGCCAGCCCTCCTGCTGTGCAGCAGCCCACTGACCCCGCATCCCCCACTGTGGCTACCACGCCTGAGCCCGTGGGGTCCGATGCTGGGGACAAGAATGCCACCAAAGCAGGCGATGACGAGCCAGAGTACGAGGACGGCCGGGGCTTTGGCATTGGGGAGCTGGTGTGGGGGAAACTGCGGGGCTTCTCCTGGTGGCCAGGCCGCATTGTGTCTTGGTGGATGACGGGCCGGAGCCGAGCAGCTGAAGGCACCCGCTGGGTCATGTGGTTCGGAGACGGCAAATTCTCAGTGGTGTGTGTTGAGAAGCTGATGCCGCTGAGCTCGTTTTGCAGTGCGTTCCACCAGGCCACGTACAACAAGCAGCCCATGTACCGCAAAGCCATCTACGAGGTCCTGCAGGTGGCCAGCAGCCGCGCGGGGAAGCTGTTCCCGGTGTGCCACGACAGCGATGAGAGTGACACTGCCAAGGCCGTGGAGGTGCAGAACAAGCCCATGATTGAATGGGCCCTGGGGGGCTTCCAGCCTTCTGGCCCTAAGGGCCTGGAGCCACCAGAAGAAGAGAAGAATCCCTACAAAGAAGTGTACACGGACATGTGGGTGGAACCTGAGGCAGCTGCCTACGCACCACCTCCACCAGCCAAAAAGCCCCGGAAGAGCACAGCGGAGAAGCCCAAGGTCAAGGAGATTATTGATGAGCGCACAAGAGAGCGGCTGGTGTACGAGGTGCGGCAGAAGTGCCGGAACATTGAGGACATCTGCATCTCCTGTGGGAGCCTCAATGTTACCCTGGAACACCCCCTCTTCGTTGGAGGAATGTGCCAAAACTGCAAGAACTGCTTTCTGGAGTGTGCGTACCAGTACGACGACGACGGCTACCAGTCCTACTGCACCATCTGCTGTGGGGGCCGTGAGGTGCTCATGTGCGGAAACAACAACTGCTGCAGGTGCTTTTGCGTGGAGTGTGTGGACCTCTTGGTGGGGCCGGGGGCTGCCCAGGCAGCCATTAAGGAAGACCCCTGGAACTGCTACATGTGCGGGCACAAGGGTACCTACGGGCTGCTGCGGCGGCGAGAGGACTGGCCCTCCCGGCTCCAGATGTTCTTCGCTAATAACCACGACCAGGAATTTGACCCTCCAAAGGTTTACCCACCTGTCCCAGCTGAGAAGAGGAAGCCCATCCGGGTGCTGTCTCTCTTTGATGGAATCGCTACAGGGCTCCTGGTGCTGAAGGACTTGGGCATTCAGGTGGACCGCTACATTGCCTCGGAGGTGTGTGAGGACTCCATCACGGTGGGCATGGTGCGGCACCAGGGGAAGATCATGTACGTCGGGGACGTCCGCAGCGTCACACAGAAGCATATCCAGGAGTGGGGCCCATTCGATCTGGTGATTGGGGGCAGTCCCTGCAATGACCTCTCCATCGTCAACCCTGCTCGCAAGGGCCTCTACGAGGGCACTGGCCGGCTCTTCTTTGAGTTCTACCGCCTCCTGCATGATGCGCGGCCCAAGGAGGGAGATGATCGCCCCTTCTTCTGGCTCTTTGAGAATGTGGTGGCCATGGGCGTTAGTGACAAGAGGGACATCTCGCGATTTCTCGAGTCCAACCCTGTGATGATTGATGCCAAAGAAGTGTCAGCTGCACACAGGGCCCGCTACTTCTGGGGTAACCTTCCCGGTATGAACAGGCCGTTGGCATCCACTGTGAATGATAAGCTGGAGCTGCAGGAGTGTCTGGAGCATGGCAGGATAGCCAAGTTCAGCAAAGTGAGGACCATTACTACGAGGTCAAACTCCATAAAGCAGGGCAAAGACCAGCATTTTCCTGTCTTCATGAATGAGAAAGAGGACATCTTATGGTGCACTGAAATGGAAAGGGTATTTGGTTTCCCAGTCCACTATACTGACGTCTCCAACATGAGCCGCTTGGCGAGGCAGAGACTGCTGGGCCGGTCATGGAGCGTGCCAGTCATCCGCCACCTCTTCGCTCCGCTGAAGGAGTATTTTGCGTGTGTGTAA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    DNMT3A_codons_NP_072046 = [DNMT3A_exons_NP_072046[i:i+3] for i in range(0, len(DNMT3A_exons_NP_072046), 3)]\n",
    "\n",
    "    #DNMT3A transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    DNMT3A_exons_extra_bases_NP_072046 = 'GATGCCCGCCATGCCCTCCAGCGGCCCCGGGGACACCAGCAGCTCTGCTGCGGAGCGGGAGGAGGACCGAAAGGACGGAGAGGAGCAGGAGGAGCCGCGTGGCAAGGAGGAGCGCCAAGAGCCCAGCACCACGGCACGGAAGGTGGGGCGGCCTGGGAGGAAGCGCAAGCACCCCCCGGTGGAAAGCGGTGACACGCCAAAGGACCCTGCGGTGATCTCCAAGTCCCCATCCATGGCCCAGGACTCAGGCGCCTCAGAGCTATTACCCAATGGGGACTTGGAGAAGCGGAGTGAGCCCCAGCCAGAGGAGGGGAGCCCTGCTGGGGGGCAGAAGGGCGGGGCCCCAGCAGAGGGAGAGGGTGCAGCTGAGACCCTGCCTGAAGCCTCAAGAGCAGTGGAAAATGGCTGCTGCACCCCCAAGGAGGGCCGAGGAGCCCCTGCAGAAGCGGGCAAAGAACAGAAGGAGACCAACATCGAATCCATGAAAATGGAGGGCTCCCGGGGCCGGCTGCGGGGTGGCTTGGGCTGGGAGTCCAGCCTCCGTCAGCGGCCCATGCCGAGGCTCACCTTCCAGGCGGGGGACCCCTACTACATCAGCAAGCGCAAGCGGGACGAGTGGCTGGCACGCTGGAAAAGGGAGGCTGAGAAGAAAGCCAAGGTCATTGCAGGAATGAATGCTGTGGAAGAAAACCAGGGGCCCGGGGAGTCTCAGAAGGTGGAGGAGGCCAGCCCTCCTGCTGTGCAGCAGCCCACTGACCCCGCATCCCCCACTGTGGCTACCACGCCTGAGCCCGTGGGGTCCGATGCTGGGGACAAGAATGCCACCAAAGCAGGCGATGACGAGCCAGAGTACGAGGACGGCCGGGGCTTTGGCATTGGGGAGCTGGTGTGGGGGAAACTGCGGGGCTTCTCCTGGTGGCCAGGCCGCATTGTGTCTTGGTGGATGACGGGCCGGAGCCGAGCAGCTGAAGGCACCCGCTGGGTCATGTGGTTCGGAGACGGCAAATTCTCAGTGGTGTGTGTTGAGAAGCTGATGCCGCTGAGCTCGTTTTGCAGTGCGTTCCACCAGGCCACGTACAACAAGCAGCCCATGTACCGCAAAGCCATCTACGAGGTCCTGCAGGTGGCCAGCAGCCGCGCGGGGAAGCTGTTCCCGGTGTGCCACGACAGCGATGAGAGTGACACTGCCAAGGCCGTGGAGGTGCAGAACAAGCCCATGATTGAATGGGCCCTGGGGGGCTTCCAGCCTTCTGGCCCTAAGGGCCTGGAGCCACCAGAAGAAGAGAAGAATCCCTACAAAGAAGTGTACACGGACATGTGGGTGGAACCTGAGGCAGCTGCCTACGCACCACCTCCACCAGCCAAAAAGCCCCGGAAGAGCACAGCGGAGAAGCCCAAGGTCAAGGAGATTATTGATGAGCGCACAAGAGAGCGGCTGGTGTACGAGGTGCGGCAGAAGTGCCGGAACATTGAGGACATCTGCATCTCCTGTGGGAGCCTCAATGTTACCCTGGAACACCCCCTCTTCGTTGGAGGAATGTGCCAAAACTGCAAGAACTGCTTTCTGGAGTGTGCGTACCAGTACGACGACGACGGCTACCAGTCCTACTGCACCATCTGCTGTGGGGGCCGTGAGGTGCTCATGTGCGGAAACAACAACTGCTGCAGGTGCTTTTGCGTGGAGTGTGTGGACCTCTTGGTGGGGCCGGGGGCTGCCCAGGCAGCCATTAAGGAAGACCCCTGGAACTGCTACATGTGCGGGCACAAGGGTACCTACGGGCTGCTGCGGCGGCGAGAGGACTGGCCCTCCCGGCTCCAGATGTTCTTCGCTAATAACCACGACCAGGAATTTGACCCTCCAAAGGTTTACCCACCTGTCCCAGCTGAGAAGAGGAAGCCCATCCGGGTGCTGTCTCTCTTTGATGGAATCGCTACAGGGCTCCTGGTGCTGAAGGACTTGGGCATTCAGGTGGACCGCTACATTGCCTCGGAGGTGTGTGAGGACTCCATCACGGTGGGCATGGTGCGGCACCAGGGGAAGATCATGTACGTCGGGGACGTCCGCAGCGTCACACAGAAGCATATCCAGGAGTGGGGCCCATTCGATCTGGTGATTGGGGGCAGTCCCTGCAATGACCTCTCCATCGTCAACCCTGCTCGCAAGGGCCTCTACGAGGGCACTGGCCGGCTCTTCTTTGAGTTCTACCGCCTCCTGCATGATGCGCGGCCCAAGGAGGGAGATGATCGCCCCTTCTTCTGGCTCTTTGAGAATGTGGTGGCCATGGGCGTTAGTGACAAGAGGGACATCTCGCGATTTCTCGAGTCCAACCCTGTGATGATTGATGCCAAAGAAGTGTCAGCTGCACACAGGGCCCGCTACTTCTGGGGTAACCTTCCCGGTATGAACAGGCCGTTGGCATCCACTGTGAATGATAAGCTGGAGCTGCAGGAGTGTCTGGAGCATGGCAGGATAGCCAAGTTCAGCAAAGTGAGGACCATTACTACGAGGTCAAACTCCATAAAGCAGGGCAAAGACCAGCATTTTCCTGTCTTCATGAATGAGAAAGAGGACATCTTATGGTGCACTGAAATGGAAAGGGTATTTGGTTTCCCAGTCCACTATACTGACGTCTCCAACATGAGCCGCTTGGCGAGGCAGAGACTGCTGGGCCGGTCATGGAGCGTGCCAGTCATCCGCCACCTCTTCGCTCCGCTGAAGGAGTATTTTGCGTGTGTGTAAG'\n",
    "\n",
    "    #List of all amino acid positions (across whole coding region) covered (amino acid position minus 1 (as python starts at 0))\n",
    "    DNMT3A_synonymous_codons_DICT_NP_072046={}\n",
    "    for i in range(lower_range, upper_range):\n",
    "        DNMT3A_synonymous_codons_DICT_NP_072046[i+1]= (DNMT3A_codons_NP_072046[i])\n",
    "\n",
    "    calls_DNMT3A_synonymous_NP_072046 = {}\n",
    "    for k,v in DNMT3A_synonymous_codons_DICT_NP_072046.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_DNMT3A_synonymous_NP_072046[k]=(v, changes)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_DNMT3A_synonymous_NP_072046_extra = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_synonymous_NP_072046.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_DNMT3A_synonymous_NP_072046_extra[k] = (Amino_acid_change, v[0], v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_DNMT3A_synonymous_NP_072046_extra2 = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_synonymous_NP_072046_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_DNMT3A_synonymous_NP_072046_extra2[k] = (Amino_acid_abbrev, v[1], v[2])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary that contains details of the positions of synonymous variants at that codon\n",
    "    DNMT3A_position_in_list_dict_synonymous_NP_072046 = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_synonymous_NP_072046_extra2.items():\n",
    "        original_AA = v[0]\n",
    "        possible_AA_changes = v[2]\n",
    "        possible_AA_changes_string = ''.join(v[2])\n",
    "        positions_in_list = []\n",
    "        for i in possible_AA_changes:\n",
    "            if i == original_AA: #synonymous\n",
    "                position_in_list=-1\n",
    "                synonymous_change = i\n",
    "                new_aa_list = []\n",
    "                while True:\n",
    "                    position_in_list = possible_AA_changes_string.find(synonymous_change, position_in_list+1)\n",
    "    #               print(position_in_list)\n",
    "                    if position_in_list == -1: break\n",
    "                    positions_in_list.append(position_in_list)\n",
    "                    positions_in_list = set(positions_in_list) #convert into a 'set' to remove duplicate positions\n",
    "                    positions_in_list = list(positions_in_list) #convert back to a list so can be added to dictionary\n",
    "                for pos in positions_in_list:\n",
    "                    new_aa = possible_AA_changes_string[pos]\n",
    "                    new_aa_list.append(new_aa)\n",
    "                DNMT3A_position_in_list_dict_synonymous_NP_072046[k]=(positions_in_list, v, new_aa_list)\n",
    "\n",
    "        aa_positions_variants_called = []\n",
    "        for k, v in DNMT3A_position_in_list_dict_synonymous_NP_072046.items():\n",
    "            aa_positions_variants_called.append(k)\n",
    "    # print('positions at which synonymous variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    DNMT3A_bases_synonymous_NP_072046 = {}\n",
    "    for k, v in DNMT3A_position_in_list_dict_synonymous_NP_072046.items():\n",
    "        original_a_a = v[1][0]\n",
    "        codon_seq = v[1][1]\n",
    "        base_changes=[]\n",
    "        new_a_a_list = v[2]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        DNMT3A_bases_synonymous_NP_072046[k]=(v, base_changes, original_a_a, new_a_a_list)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    DNMT3A_bases_synonymous2_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_synonymous_NP_072046.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        new_a_a_list = v[3]\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        DNMT3A_bases_synonymous2_NP_072046[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    DNMT3A_bases_synonymous3_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_synonymous2_NP_072046.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_a_a = v[0][2]\n",
    "        new_aa_list = v[0][0][2]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = k\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(k)-1)*3)+codon_pos #codon number minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][1][1][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        DNMT3A_bases_synonymous3_NP_072046[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, original_a_a, new_aa_list)\n",
    "\n",
    "    # Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    DNMT3A_bases_synonymous4_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_synonymous3_NP_072046.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_synonymous=[]\n",
    "        new_a_a_list = v[6]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_DNMT3A_NP072046:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check_DNMT3A_NP072046(i)\n",
    "                sites_trio_synonymous.append([adapted_trio])\n",
    "                DNMT3A_bases_synonymous4_NP_072046[k] = (v, sites_trio_synonymous)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_synonymous = [DNMT3A_exons_extra_bases_NP_072046[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_synonymous.append(site_trio_synonymous)\n",
    "                    DNMT3A_bases_synonymous4_NP_072046[k] = (v, sites_trio_synonymous)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    DNMT3A_sites_synonymous5_NP_072046 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in DNMT3A_bases_synonymous4_NP_072046.items():\n",
    "        original_a_a = v[0][0][0]\n",
    "        new_a_a = v[0][6]\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            new_amino_acid = new_a_a[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, original_a_a, new_amino_acid])\n",
    "            change = change+1\n",
    "        DNMT3A_sites_synonymous5_NP_072046[a]=info\n",
    "        a=a+1\n",
    "    #     print(DNMT3A_sites_synonymous5_NP_072046)\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    DNMT3A_sites_synonymous6_NP_072046={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in DNMT3A_sites_synonymous5_NP_072046.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        original_a_a = v[5]\n",
    "        new_a_a = v[6]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        DNMT3A_sites_synonymous6_NP_072046[k] = (v[0],changes3, codon_n_912, original_a_a, new_a_a) #site_trio = the affected base and its neighbours\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    DNMT3A_mutation_rates_synonymous_NP_072046 = {}\n",
    "    for k, v in DNMT3A_sites_synonymous6_NP_072046.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            original_a_a = v[3]\n",
    "            new_a_a = v[4]\n",
    "            aa_positioning = v[2]\n",
    "            variant_id = str(original_a_a)+str(aa_positioning)+str(new_a_a)\n",
    "            DNMT3A_mutation_rates_synonymous_NP_072046[k] = (v[0],v[1], mutation_rate, variant_id)\n",
    "\n",
    "    DNMT3A_mutation_rate_synonymous_NP_072046_sum = []\n",
    "\n",
    "    mutation_rates_list = []\n",
    "    for k, v in DNMT3A_mutation_rates_synonymous_NP_072046.items():\n",
    "        codon_position_in_gene=(v[0])\n",
    "        variant = v[3]\n",
    "        site = v[1]\n",
    "        mutation_rate_value = v[2] \n",
    "        mutation_rates_list.append((v[0], variant, site, mutation_rate_value))\n",
    "        DNMT3A_mutation_rate_synonymous_NP_072046_sum.append(mutation_rate_value)\n",
    "\n",
    "    #Get the average of all the mutation rates\n",
    "    DNMT3A_mutation_rate_synonymous_NP_072046_sum = []\n",
    "    for k, v in DNMT3A_mutation_rates_synonymous_NP_072046.items():\n",
    "        mut_rate = v[2]\n",
    "        DNMT3A_mutation_rate_synonymous_NP_072046_sum.append(mut_rate)\n",
    "\n",
    "    DNMT3A_Average_mutation_rate_synonymous_NP_072046 = np.mean(DNMT3A_mutation_rate_synonymous_NP_072046_sum)\n",
    "    DNMT3A_Summed_mutation_rate_synonymous_NP_072046_added = np.sum(DNMT3A_mutation_rate_synonymous_NP_072046_sum)\n",
    "\n",
    "#     print('Average mutation rate across DNMT3A synonymous_NP_072046 for this study = ', DNMT3A_Average_mutation_rate_synonymous_NP_072046)\n",
    "    print('Total summed mutation rate across DNMT3A synonymous_NP_072046 for this study = ', DNMT3A_Summed_mutation_rate_synonymous_NP_072046_added)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all # e.g. lower= 0, upper = 913 for whole gene\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  4.7450332e-08\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[(2641, 'S881C', 'G[A>T]G', 3.78353e-10),\n",
       " (2641, 'S881R', 'G[A>C]G', 1.86315e-10),\n",
       " (2641, 'S881G', 'G[A>G]G', 8.12663e-10),\n",
       " (2642, 'S881N', 'A[G>A]C', 4.08717e-09),\n",
       " (2642, 'S881T', 'A[G>C]C', 4.40263e-10),\n",
       " (2642, 'S881I', 'A[G>T]C', 7.65675e-10),\n",
       " (2643, 'S881R', 'G[C>A]C', 6.41452e-10),\n",
       " (2643, 'S881R', 'G[C>G]C', 4.04844e-10),\n",
       " (2644, 'R882S', 'C[C>A]G', 1.16311e-09),\n",
       " (2644, 'R882G', 'C[C>G]G', 6.5522e-10),\n",
       " (2644, 'R882C', 'C[C>T]G', 1.41512e-08),\n",
       " (2645, 'R882H', 'C[G>A]C', 1.88229e-08),\n",
       " (2645, 'R882P', 'C[G>C]C', 6.29677e-10),\n",
       " (2645, 'R882L', 'C[G>T]C', 8.45567e-10),\n",
       " (2647, 'L883M', 'C[T>A]T', 3.83597e-10),\n",
       " (2647, 'L883V', 'C[T>G]T', 2.52155e-10),\n",
       " (2648, 'L883S', 'T[T>C]G', 5.44182e-10),\n",
       " (2648, 'L883W', 'T[T>G]G', 2.83676e-10),\n",
       " (2649, 'L883F', 'T[G>C]G', 4.64303e-10),\n",
       " (2649, 'L883F', 'T[G>T]G', 1.53801e-09)]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutation_rates_missense_range_of_variants_DNMT3A_NP072046(880, 883)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Calculation of mutation rates for each study"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Jaiswal 2014"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Provide a list (in their paper) of the DNMT3A missense variants they called = 'Jaiswal DNMT3A missense variants called.csv'.\n",
    "- Included all nonsense variants in DNMT3A. \n",
    "- Used NM02252 (NP_072046)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  3.85424736e-07\n"
     ]
    }
   ],
   "source": [
    "Jaiswal_nonsense_DNMT3A = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(0, 912)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  5.08517098e-07\n"
     ]
    }
   ],
   "source": [
    "Jaiswal_missense_DNMT3A = mutation_rates_list_missense_variants_DNMT3A_NP072046('Jaiswal DNMT3A missense variants called.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Mutation rate for DNMT3A missense and nonsense in Jaiswal = 3.85424736e-07 + 5.08517098e-07 = 8.93941834e-07**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Genovese 2014"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Included all DNMT3A missense and nonsense variants in exons 7-23 and all variants found in 7 or more individuals hematopoietic and lymphoid cancers in COSMIC.\n",
    "- Used NM02252 (NP_072046).\n",
    "- Excluded regions described in Genovese 2014 supplement page 3:\n",
    "    - 1) Low complexity regions and sites harboring markers failing Hardy Weinberg equilibrium tests in the 1000 Genomes Project phase 15 (https://github.com/lh3/varcmp/blob/master/scripts/LCR-hs37d5.bed.gz and https://github.com/lh3/varcmp/blob/master/scripts/1000g.hwe-bad.bed)\n",
    "    - 2) Sites with excess coverage within the 1000 Genomes Project phase 1 [6]\n",
    "    - 3) Segmental duplications of the human genome [7,8] (http://hgdownload.cse.ucsc.edu/goldenPath/hg19/database/genomicSuperDups.txt.gz)\n",
    "    - 4) Regions harboring common large insertions in 1000 Genomes Project Phase1 samples (data unpublished)\n",
    "    - 5) Regions excluded from the strict mask of the 1000 Genomes Project phase 1 [9] (ftp://ftp.1000genomes.ebi.ac.uk/vol1/ftp/phase1/analysis_results/supporting/accessible_genome_mask s/20120824_strict_mask.bed)\n",
    "    - (= excludes DNMT3A amino acids 849-866, last base of amino acid 848, 1st base of amino acid 847, amino aicds 842-846.\n",
    "1st 2 bases of amino acids 47, amino acids 29-46, 1st base of amino acid 26, 1st base of amino acid 1).\n",
    "- i.e. they included:\n",
    "    - codons 214-841, 867-912\n",
    "    - first 2 bases of codon 848\n",
    "    - last 2 bases of codon 847"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Complete codons"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  2.7073501399999997e-07\n"
     ]
    }
   ],
   "source": [
    "Genovese_nonsense_DNMT3A_1 = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(213, 841)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  9.564487000000002e-09\n"
     ]
    }
   ],
   "source": [
    "Genovese_nonsense_DNMT3A_2 = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(866, 912)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  5.383253473e-06\n"
     ]
    }
   ],
   "source": [
    "Genovese_missense_DNMT3A_1 = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(213, 841)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  4.6552877900000003e-07\n"
     ]
    }
   ],
   "source": [
    "Genovese_missense_DNMT3A_2 = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(866, 912)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Partial codons"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- No nonsense variants at the partial codons\n",
    "- first 2 bases of codon 848 = 2542, 2543\n",
    "- last 2 bases of codon 847 = 2540, 2541"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  9.167888e-09\n",
      "\n",
      "Total summed mutation rate for partial missense codons in Genovese= 4.7568869999999996e-09\n"
     ]
    }
   ],
   "source": [
    "Genovese_missense_DNMT3A_partial_codons = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(846, 848)\n",
    "\n",
    "base_position_list = [2540, 2541, 2542, 2543]\n",
    "\n",
    "Genovese_partial_missense = []\n",
    "Genovese_partial_missense_sum_values = []\n",
    "for i in Genovese_missense_DNMT3A_partial_codons:\n",
    "    if i[0] in base_position_list:\n",
    "        Genovese_partial_missense.append(i)\n",
    "        Genovese_partial_missense_sum_values.append(i[3])\n",
    "        \n",
    "Genovese_partial_missense_added = np.sum(Genovese_partial_missense_sum_values)\n",
    "print(\"\")\n",
    "# print(Genovese_partial_missense)\n",
    "print('Total summed mutation rate for partial missense codons in Genovese=',Genovese_partial_missense_added)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Complete and partial codons summed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total missense mutation rate in Genovese DNMT3A =  5.8535391390000006e-06\n",
      "Total nonsense mutation rate in Genovese DNMT3A =  2.80299501e-07\n"
     ]
    }
   ],
   "source": [
    "print('Total missense mutation rate in Genovese DNMT3A = ', 5.383253473e-06+4.6552877900000003e-07+4.7568869999999996e-09)\n",
    "print('Total nonsense mutation rate in Genovese DNMT3A = ', 2.7073501399999997e-07+9.564487000000002e-09)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Mutation rate for DNMT3A missense and nonsense in Genovese = 5.8535391390000006e-06 + 2.80299501e-07 = 6.133838640000001e-06**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## McKerrel 2015"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Assumed to have used NM_02252 (NP_072046)\n",
    "- Only included 882 codon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  3.626767399999999e-08\n"
     ]
    }
   ],
   "source": [
    "McKerrel_missense_DNMT3A = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(881, 882)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Mutation rate for DNMT3A in McKerrel = 3.626767399999999e-08**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Zink 2017"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Included all DNMT3A nonsense variants.\n",
    "- Included all DNMT3A missense variants seen in COSMIC = 'Zink DNMT3A missense variants in COSMIC.csv'.\n",
    "- Used both NP_072046 and NP_715640 (and also NP_783328, which is ~NP_072046)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  3.85424736e-07\n"
     ]
    }
   ],
   "source": [
    "Zink_nonsense_DNMT3A_072046 = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(0, 912)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  1.211073846e-06\n"
     ]
    }
   ],
   "source": [
    "Zink_missense_DNMT3A_072046 = mutation_rates_list_missense_variants_DNMT3A_NP072046('Zink DNMT3A missense variants in COSMIC.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Mutation rate for DNMT3A missense and nonsense in Zink = 3.85424736e-07+1.211073846e-06 = 1.596498582e-06**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Coombs 2017"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Used MSK-IMPACT panel which encompassed the full protein-coding sequence of DNMT3A.\n",
    "- Assumed they used NM_02252(NP_072046)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  3.85424736e-07\n"
     ]
    }
   ],
   "source": [
    "Coombs_nonsense_DNMT3A = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(0, 912)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  8.080913063999998e-06\n"
     ]
    }
   ],
   "source": [
    "Coombs_missense_DNMT3A = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(0, 912)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Mutation rate for DNMT3A missense and nonsense in Coombs = 3.85424736e-07 + 8.080913063999998e-06 = 8.466337799999998e-06**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Young 2016 & 2019"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Used Illumina Myeloid Panel, which includes all exons of DNMT3A.\n",
    "- Excluded TET2 germline variants identified by the 1000 Genomes project with MAF >0.01 = none in DNMT3A.\n",
    "- Assumed they used NM_02252 (NP_072046)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  3.85424736e-07\n"
     ]
    }
   ],
   "source": [
    "Young_nonsense_DNMT3A = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(0, 912)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  8.080913063999998e-06\n"
     ]
    }
   ],
   "source": [
    "Young_missense_DNMT3A = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(0, 912)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Mutation rate for DNMT3A missense and nonsense in Young 2016 and Young 2019 = 3.85424736e-07 + 8.080913063999998e-06 = 8.466337799999998e-06**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Desai 2018"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Included all missense except E30A.\n",
    "- Included all nonsense.\n",
    "- Excluded variants with dbSNP v142 MAF >0.25% (0.0025) or ExAC adjusted population >0.25% (0.0025) = E30A is the only one.\n",
    "- Used NP_783328 (~NM_02252 (NP_072046))."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  3.85424736e-07\n"
     ]
    }
   ],
   "source": [
    "Desai_nonsense_DNMT3A = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(0, 912)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  8.080913063999998e-06\n"
     ]
    }
   ],
   "source": [
    "Desai_missense_DNMT3A = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(0, 912)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  4.793991e-09\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[(88, 'E30K', 'G[G>A]A', 1.56027e-09),\n",
       " (88, 'E30Q', 'G[G>C]A', 4.36877e-10),\n",
       " (89, 'E30V', 'G[A>T]G', 3.78353e-10),\n",
       " (89, 'E30A', 'G[A>C]G', 1.86315e-10),\n",
       " (89, 'E30G', 'G[A>G]G', 8.12663e-10),\n",
       " (90, 'E30D', 'A[G>C]G', 4.37005e-10),\n",
       " (90, 'E30D', 'A[G>T]G', 9.82508e-10)]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutation_rates_missense_range_of_variants_DNMT3A_NP072046(29, 30)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Mutation rate for missense variants = 8.080913063999998e-06 - 1.86315e-10 = 8.080726748999998e-06"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Mutation rate for DNMT3A missense and nonsense in Desai 2018 = 8.080726748999998e-06 + 3.85424736e-07 = 8.466151484999998e-06**"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Acuna-Hidalgo 2017"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- See separate python notebook"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
