{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "toc": true
   },
   "source": [
    "<h1>Table of Contents<span class=\"tocSkip\"></span></h1>\n",
    "<div class=\"toc\"><ul class=\"toc-item\"><li><span><a href=\"#Python-Setup\" data-toc-modified-id=\"Python-Setup-1\"><span class=\"toc-item-num\">1&nbsp;&nbsp;</span>Python Setup</a></span><ul class=\"toc-item\"><li><span><a href=\"#Change-the-width-of-the-page\" data-toc-modified-id=\"Change-the-width-of-the-page-1.1\"><span class=\"toc-item-num\">1.1&nbsp;&nbsp;</span>Change the width of the page</a></span></li><li><span><a href=\"#Import-packages\" data-toc-modified-id=\"Import-packages-1.2\"><span class=\"toc-item-num\">1.2&nbsp;&nbsp;</span>Import packages</a></span></li></ul></li><li><span><a href=\"#Colours\" data-toc-modified-id=\"Colours-2\"><span class=\"toc-item-num\">2&nbsp;&nbsp;</span>Colours</a></span><ul class=\"toc-item\"><li><span><a href=\"#Colour-line-graph\" data-toc-modified-id=\"Colour-line-graph-2.1\"><span class=\"toc-item-num\">2.1&nbsp;&nbsp;</span>Colour line graph</a></span></li></ul></li><li><span><a href=\"#Totals-for-studies\" data-toc-modified-id=\"Totals-for-studies-3\"><span class=\"toc-item-num\">3&nbsp;&nbsp;</span>Totals for studies</a></span></li><li><span><a href=\"#Functions-for-calculating-trinucleotide-context-specific-mutation-rates\" data-toc-modified-id=\"Functions-for-calculating-trinucleotide-context-specific-mutation-rates-4\"><span class=\"toc-item-num\">4&nbsp;&nbsp;</span>Functions for calculating trinucleotide-context specific mutation rates</a></span><ul class=\"toc-item\"><li><span><a href=\"#Calculating-mutation-rates-for-individual-variants\" data-toc-modified-id=\"Calculating-mutation-rates-for-individual-variants-4.1\"><span class=\"toc-item-num\">4.1&nbsp;&nbsp;</span>Calculating mutation rates for individual variants</a></span><ul class=\"toc-item\"><li><span><a href=\"#TP53\" data-toc-modified-id=\"TP53-4.1.1\"><span class=\"toc-item-num\">4.1.1&nbsp;&nbsp;</span>TP53</a></span></li></ul></li><li><span><a href=\"#Calculating-mutation-rates-from-a-.csv-file-of-variants\" data-toc-modified-id=\"Calculating-mutation-rates-from-a-.csv-file-of-variants-4.2\"><span class=\"toc-item-num\">4.2&nbsp;&nbsp;</span>Calculating mutation rates from a .csv file of variants</a></span><ul class=\"toc-item\"><li><span><a href=\"#TP53\" data-toc-modified-id=\"TP53-4.2.1\"><span class=\"toc-item-num\">4.2.1&nbsp;&nbsp;</span>TP53</a></span></li></ul></li><li><span><a href=\"#Calculating-mutation-rates-from-a-list-of-variants\" data-toc-modified-id=\"Calculating-mutation-rates-from-a-list-of-variants-4.3\"><span class=\"toc-item-num\">4.3&nbsp;&nbsp;</span>Calculating mutation rates from a list of variants</a></span><ul class=\"toc-item\"><li><span><a href=\"#TP53\" data-toc-modified-id=\"TP53-4.3.1\"><span class=\"toc-item-num\">4.3.1&nbsp;&nbsp;</span>TP53</a></span></li></ul></li></ul></li><li><span><a href=\"#Mutation-rate-for-DBD-and-non-DBD-for-each-study\" data-toc-modified-id=\"Mutation-rate-for-DBD-and-non-DBD-for-each-study-5\"><span class=\"toc-item-num\">5&nbsp;&nbsp;</span>Mutation rate for DBD and non-DBD for each study</a></span><ul class=\"toc-item\"><li><span><a href=\"#Missense-DBD\" data-toc-modified-id=\"Missense-DBD-5.1\"><span class=\"toc-item-num\">5.1&nbsp;&nbsp;</span>Missense DBD</a></span></li><li><span><a href=\"#Missense-non-DBD\" data-toc-modified-id=\"Missense-non-DBD-5.2\"><span class=\"toc-item-num\">5.2&nbsp;&nbsp;</span>Missense non-DBD</a></span></li><li><span><a href=\"#Nonsense-DBD\" data-toc-modified-id=\"Nonsense-DBD-5.3\"><span class=\"toc-item-num\">5.3&nbsp;&nbsp;</span>Nonsense DBD</a></span></li><li><span><a href=\"#Nonsense-non-DBD\" data-toc-modified-id=\"Nonsense-non-DBD-5.4\"><span class=\"toc-item-num\">5.4&nbsp;&nbsp;</span>Nonsense non-DBD</a></span></li></ul></li><li><span><a href=\"#Actual-number-of-observations-of-TP53-variants-(inside-vs-outside-DNA-binding-domain)\" data-toc-modified-id=\"Actual-number-of-observations-of-TP53-variants-(inside-vs-outside-DNA-binding-domain)-6\"><span class=\"toc-item-num\">6&nbsp;&nbsp;</span>Actual number of observations of TP53 variants (inside vs outside DNA binding domain)</a></span></li><li><span><a href=\"#Plotting-the-relative-frequencies-of-mutations-(controlled-for-mutation-rates)\" data-toc-modified-id=\"Plotting-the-relative-frequencies-of-mutations-(controlled-for-mutation-rates)-7\"><span class=\"toc-item-num\">7&nbsp;&nbsp;</span>Plotting the relative frequencies of mutations (controlled for mutation rates)</a></span><ul class=\"toc-item\"><li><span><a href=\"#Overall-p-value\" data-toc-modified-id=\"Overall-p-value-7.1\"><span class=\"toc-item-num\">7.1&nbsp;&nbsp;</span>Overall p-value</a></span></li></ul></li></ul></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Python Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Change the width of the page"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>.container { width:80% !important; }</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.core.display import display, HTML\n",
    "display(HTML(\"<style>.container { width:80% !important; }</style>\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "# imported packages\n",
    "import random\n",
    "import numpy as np\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import math\n",
    "import matplotlib.ticker as plticker\n",
    "from matplotlib.ticker import LinearLocator, FormatStrFormatter\n",
    "from matplotlib.patches import Polygon\n",
    "from matplotlib.patches import Patch\n",
    "from matplotlib.lines import Line2D\n",
    "from matplotlib import cm\n",
    "import scipy.special\n",
    "import scipy.integrate as it\n",
    "from scipy import integrate\n",
    "from scipy.integrate import quad\n",
    "from scipy.interpolate import interp1d\n",
    "from scipy.stats import kde\n",
    "from scipy.stats import poisson\n",
    "import copy\n",
    "import glob, os\n",
    "import re\n",
    "from sklearn import datasets, linear_model\n",
    "import pandas as pd\n",
    "from decimal import *\n",
    "from operator import itemgetter    \n",
    "from collections import OrderedDict\n",
    "import timeit"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Colours"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "# Lists of colors for plots\n",
    "c0 = (0.76, 0.76, 0.76)\n",
    "c1 = (1.00, 0.18, 0.33);\n",
    "c2 = (1.00, 0.23, 0.19);\n",
    "c3 = (1.00, 0.58, 0.00);\n",
    "c4 = (1.00, 0.80, 0.00);\n",
    "c5 = (0.30, 0.85, 0.39);\n",
    "c6 = (0.35, 0.78, 0.98);\n",
    "c7 = (0.20, 0.67, 0.86);\n",
    "c8 = (0.00, 0.48, 1.00);\n",
    "c9 = (0.35, 0.34, 0.84);\n",
    "c10 = (0.00, 0.31, 0.57);\n",
    "c11 = (0.12, 0.29, 0.69);\n",
    "c12 = (0.17, 0.17, 0.42);\n",
    "c13 = (1.00, 1.00, 1.00);\n",
    "c14 = (0.77, 0.04, 0.00);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#define the colors from colorbrewer2\n",
    "orange1 = '#feedde'\n",
    "orange2 = '#fdbe85'\n",
    "orange3 = '#fd8d3c'\n",
    "orange4 = '#e6550d'\n",
    "orange5 = '#a63603'\n",
    "blue1 = '#eff3ff'\n",
    "blue2 = '#bdd7e7'\n",
    "blue3 = '#6baed6'\n",
    "blue4 = '#3182bd'\n",
    "blue5 = '#08519c'\n",
    "green1 = '#edf8e9'\n",
    "green2 = '#bae4b3'\n",
    "green3 = '#74c476'\n",
    "green4 = '#31a354'\n",
    "green5 = '#006d2c'\n",
    "grey1 = '#f7f7f7'\n",
    "grey2 = '#cccccc'\n",
    "grey3 = '#969696'\n",
    "grey4 = '#636363'\n",
    "grey5 = '#252525'\n",
    "purple1 = '#f2f0f7'\n",
    "purple2 = '#cbc9e2'\n",
    "purple3 = '#9e9ac8'\n",
    "purple4 = '#756bb1'\n",
    "purple5 = '#54278f'\n",
    "red1 = '#fee5d9'\n",
    "red2 = '#fcae91'\n",
    "red3 = '#fb6a4a'\n",
    "red4 = '#de2d26'\n",
    "red5 = '#a50f15'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Colour line graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAlkAAAExCAYAAABPt7ftAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAF0xJREFUeJzt3X2QnWd53/Hvj2KcENtgrTAOkW11hpeCiIDhOIEqGEcwQNrBocUW5EWyOmZ2mBiHiSs6oetpAq5CQsDQDs6MRcuwQMybGqfEaUiYJVZCB5geWfbSJVGdpn6R7WLWshkgDl7JV//Y48x22dUenec82j2r72dmZ3Tu+36uc9l//ea+n/M8qSokSZI0XE9Z7QYkSZLWI0OWJElSCwxZkiRJLTBkSZIktcCQJUmS1AJDliRJUgsMWZJGVpJLkxxZ7T4kaSmGLElrQpJfTNJN8r0kDyb5kyQ/s9p9SdKgDFmSVl2Sa4EPA78FPBu4EPg94OdXqZ+nrsb3SlpfDFmSVlWSZwDvBa6uqj+oqu9X1VxV/VFVvSvJmUk+nOSB3t+Hk5y5TK0XJrktyaNJZpJctmDutiRvW/B5d5KvLPhcSa5OchdwV+Z9KMlDSb6TZDrJi1v8XyFpnTFkSVptrwR+BLhlmfkJ4BXAS4GXAD8FXLd4UZIzgD8C/gw4D7gG+P0kLziJXt4E/DTwIuB1wCXA84FnAm8BHj6JWpJOc4YsSattDJitqmPLzP8S8N6qeqiqvg28B9i5xLpXAGcBv11Vj1fVl4FbgV84iV7eV1VHq+oxYA44G/gnQKrqr6rqwZOoJek0Z8iStNoeBjae4D6o5wD3LPh8T29sqXX3VdUTi9b+xEn0ct+T/+iFtI8ANwLfSrIvyTknUUvSac6QJWm1fRX4e+aP6pbyAHDRgs8X9saWWndBkqcsWnt/79/fB56+YO78JWrU//eh6j9W1cuBLcwfG75rmR4l6YcYsiStqqr6DvDvgBuTvCnJ05OckeTnkrwf+DRwXZJnJdnYW/upJUp9nfkg9W96118KvBH4TG/+DuBf9uo/F7jqRH0luTjJT/fu9fo+80HwePP/YkmnC0OWpFVXVTcA1zJ/Q/u3mT+2ewfwh8C/B7rANPAN4Pbe2OIajwOXAT8HzDL/CIhdVfXXvSUfAh4HvgVMAr+/QlvnAB8FHmH+2PFh4AOD/jdKOv2kqlZeJUmSpJPiTpYkSVILDFmSJEktMGRJkiS1wJAlSZLUAkOWJElSC1b9TfMbN26szZs3r3YbkiRJKzp48OBsVT2rn7WrHrI2b95Mt9td7TYkSZJWlOSelVfN87hQkiSpBYYsSZKkFhiyJEmSWmDIkiRJaoEhS5IkqQWGLEmSpBYYsiRJklpgyJIkSWqBIUuSJKkFhixJkqQWGLIkSZJaYMiSJElqwUAhK8klSW5PcizJ5UvMn5Pk/iQfad6iJEnS6Bl0J+teYDdw8zLz1wMHBqwtSZI08p7az6Iku4A9QAHTVbWzN/7EEmtfDjwb+CLQWan2Y3cd5htvuPQkWpYkSVr7VgxZSbYAE8C2qppNsuEEa58CfBDYCbzmBOvGgXGA5/3omSfbsyRJ0prXz07WdmB/Vc0CVNXRE6z9FeC/VdV9SZZdVFX7gH0AnU6nfvKLt/XdsCRJ0qo5Qb5ZrJ+QFeaPCfvxSuBVSX4FOAt4WpLvVdWv992RJEnSOtDPje9TwI4kYwAnOi6sql+qqgurajPz93B9woAlSZJORyuGrKqaAfYCB5LcCdyQ5OIkR4ArgJuSzLTcpyRJ0kjp69eFVTUJTC4a3rTCNR8HPj5QV5IkSSPOJ75LkiS1wJAlSZLUAkOWJElSCwxZkiRJLTBkSZIktcCQJUmS1AJDliRJUgsMWZIkSS0YKGQluSTJ7UmOJbl8wfhFSQ4muSPJTJK3D69VSZKk0dHXE9+XcC+wm/n3Ey70IPBPq+oHSc4C/meSL1TVAw16lCRJGjl9hawku5gPVAVMV9XO3vgTC9dV1eMLPp6Jx5GSJOk0tWLISrIFmAC2VdVskg0rrL8A+GPgucC73MWSJEmno352mrYD+6tqFqCqjp5ocVXdV1VbmQ9ZVyZ59uI1ScaTdJN05+bmBulbkiRpTesnZIX5Y8KT0tvBmgFetcTcvqrqVFXnjDPOONnSkiRJa14/IWsK2JFkDOBEx4VJNiX50d6/zwW2AYeH0agkSdIoWfGerKqaSbIXOJDkOHAoyY3ALcC5wBuTvKeqtgAvBD6YpJjfAftAVX2jxf4lSZLWpL5+XVhVk8DkouFNS6z7ErB1CH1JkiSNNB+xIEmS1AJDliRJUgsMWZIkSS0wZEmSJLXAkCVJktQCQ5YkSVILDFmSJEktMGRJkiS1wJAlSZLUgoFCVpJLktye5FiSyxeMvzTJV5PMJJlO8pbhtSpJkjQ6+nqtzhLuBXYDexaN/x2wq6ruSvIc4GCSP62qRxv0KEmSNHL6CllJdjEfqAqYrqqdvfEnFq6rqv+14N8PJHkIeBawbMi6++5vsXv3hwdoXZIkae1aMWQl2QJMANuqajbJhn4KJ/kp4GnA/15ibhwYBzjnnGefVMOSJEmjIFV14gXJNcD5VTWxxNzHgVurav+i8R8HbgOurKqvnah+p9Opbrd7km1LkiSdekkOVlWnn7X93Pge5o8J+/3yc4A/Bq5bKWBJkiStV/2ErClgR5IxgBMdFyZ5GnAL8Imq+vxwWpQkSRo9K4asqpoB9gIHktwJ3JDk4iRHgCuAm5LM9JbvAC4Bdie5o/f30raalyRJWqv6+nVhVU0Ck4uGNy2x7lPAp4bQlyRJ0kjzie+SJEktMGRJkiS1wJAlSZLUAkOWJElSCwxZkiRJLTBkSZIktcCQJUmS1AJDliRJUgsGCllJLklye5JjSS5fNPfFJI8muXU4LUqSJI2eQXey7gV2AzcvMfe7wM5BG5IkSVoP+gpZSXYlmU5yZ5JPVtXdVTUNPLF4bVVNAd8ddqOSJEmjZMV3FybZAkwA26pqNsmGYTbwt/d/j7f8+leGWVKSJGnV9bOTtR3YX1WzAFV1tOmXJhlP0k3SnXt8rmk5SZKkNWfFnSwgQA3zS6tqH7APoNPp1Gd/+2eGWV6SJKkVn/ud/tf2s5M1BexIMgYw7ONCSZKk9WjFkFVVM8Be4ECSO4Ebklyc5AhwBXBTkpkn1yf5S+DzwGuSHEny+pZ6lyRJWrP6OS6kqiaByUXDm5ZZ+6qmTUmSJI06n/guSZLUAkOWJElSCwxZkiRJLTBkSZIktcCQJUmS1AJDliRJUgsMWZIkSS0wZEmSJLXAkCVJktSCgUJWkkuS3J7kWJLLF81dmeSu3t+Vw2lTkiRptPT1Wp0l3AvsBvYsHOy9PPo3gA5QwMEkX6iqR5o0KUmSNGr6CllJdjEfqAqYrqqdvfEnFi19PfClqjram/8S8Abg08vVPnz/I1w68V8GaF2SJGntWjFkJdkCTADbqmq2t1u1nJ8A7lvw+UhvbHHNcWAc4MyxC06qYUmSpFHQz07WdmB/Vc0CPLlLtYwsMVY/NFC1D9gH0Ol06ra9b+6jDUmSpNWV3+p/bT83voclgtIyjgALt6Y2AQ/0344kSdL60E/ImgJ2JBmDf7i5fTl/CrwuyblJzgVe1xuTJEk6rax4XFhVM0n2AgeSHAcOJbkRuAU4F3hjkvdU1ZaqOprkeuB/9C5/7wrHi5IkSetSX78urKpJYHLR8KZl1n4M+FjDviRJkkaaT3yXJElqgSFLkiSpBYYsSZKkFhiyJEmSWmDIkiRJaoEhS5IkqQWGLEmSpBYYsiRJklow1JCV5KIkU0mmk9yWZMkHlkqSJK13w97J+gDwiaraCrwXeN+Q60uSJI2ERiErya7ertWdST4JvIj5F0oD/Dnw800blCRJGkV9vbtwKUm2ABPAtqqaTbIB+AjwZuA/AP8CODvJWFU9vFydBx94nOvfc8+gbUiSJK1JTXaytgP7q2oWoKqOAnuAVyc5BLwauB84tvjCJONJukm6c3NzDVqQJElam1JVg12Y/CpwXlVdt8z8WcBfV9UJb37vdDrV7XYH6kGSJOlUSnKwqjr9rG2ykzUF7Egy1vvSDUk2Jnmy5ruBjzWoL0mSNLIGvierqmaS7AUOJDkOHAJuBd6XpIC/AK4eTpuSJEmjZeCQBVBVk8DkouH9TWpKkiStBz7xXZIkqQWGLEmSpBYYsiRJklpgyJIkSWqBIUuSJKkFhixJkqQWGLIkSZJaYMiSJElqgSFLkiSpBUMNWUkuTPLnSQ4lmU7yz4ZZX5IkaVQMeyfrOuBzVfUy4K3A7w25viRJ0kho9O7CJLuAPUAB08DfAef0pp8BPLBSjcOzcOl/atKFJEnS2jNwyEqyBZgAtlXVbJINwJnAnyW5Bvgx4LXLXDsOjAOc+Zytg7YgSZK0ZqWqBrtwPkidX1UTC8au7dX8YJJXAv8ZeHFVPbFcnU6nU91ud6AeJEmSTqUkB6uq08/aJvdkhfljwoWuAj4HUFVfBX4E2NjgOyRJkkZSk5A1BexIMgbQOy68F3hN7/MLmQ9Z327apCRJ0qgZ+J6sqppJshc4kOQ4cAj418BHk/wa87tcu2vQ80hJkqQR1ujXhVU1CUwuGt7WpKYkSdJ64BPfJUmSWmDIkiRJaoEhS5IkqQWGLEmSpBYYsiRJklpgyJIkSWqBIUuSJKkFhixJkqQWNHoY6WJJPgT8bO/j04HzquqZw/wOSZKkUTDUkFVVv/bkv5NcA7xsmPUlSZJGRaPjwiS7kkwnuTPJJxdN/wLw6Sb1JUmSRtXAO1lJtgATwLaqmk2yYcHcRcA/Br68Up17vvs447cdGbQNSZKkNanJTtZ2YH9VzQJU1dEFc2/tzR1f6sIk40m6Sbpzc3MNWpAkSVqbmtyTFaCWmXsrcPVyF1bVPmAfQKfTqX2XbmrQhiRJ0qnx0ZNY22QnawrYkWQM4MnjwiQvAM4FvtqgtiRJ0kgbeCerqmaS7AUOJDkOHAJ2M3/D+2eqarldLkmSpHWv0SMcqmoSmFw09ptNakqSJK0HPvFdkiSpBYYsSZKkFhiyJEmSWmDIkiRJaoEhS5IkqQWGLEmSpBYYsiRJklpgyJIkSWqBIUuSJKkFQw9ZSXYk+WaSmSQ3D7u+JEnSKGj0Wp3FkjwPeDewraoeSXLeMOtLkiSNikYhK8kuYA9QwDTwIHBjVT0CUFUPrVTjwceK62eONWlDkiRpzRn4uDDJFmAC2F5VLwHeCTwfeH6S/57ka0nesMy140m6Sbpzc3ODtiBJkrRmpaoGuzC5Bji/qiYWjN0KzAE7gE3AXwIvrqpHl6vT6XSq2+0O1IMkSdKplORgVXX6Wdvkxvcwf0y40BHgv1bVXFX9H+Aw8LwG3yFJkjSSmoSsKWBHkjGAJBuAPwR+tvd5I/PHh3/btElJkqRRM/CN71U1k2QvcCDJceAQ8K+A1yX5JnAceFdVPTycViVJkkZHo18XVtUkMLlo+NrenyRJ0mnLJ75LkiS1wJAlSZLUAkOWJElSCwxZkiRJLTBkSZIktcCQJUmS1AJDliRJUgsMWZIkSS0wZEmSJLVgqCErye4k305yR+/vbcOsL0mSNCoavVZnGZ+tqne0UFeSJGlkNApZSXYBe4ACpoGpk61x/+P/l3973/ubtCFJkrTmDHxcmGQLMAFsr6qXAO/sTb05yXSS/UkuWOba8STdJN25ublBW5AkSVqzUlWDXZhcA5xfVRMLxsaA71XVD5K8HdhRVdtPVKfT6VS32x2oB0mSpFMpycGq6vSztsmN72H+mPAfVNXDVfWD3sePAi9vUF+SJGlkNQlZU8CO3u4VSTYk+fEF85cBf9WkOUmSpFE18I3vVTWTZC9wIMlx4BDwYJLLgGPAUWD3ULqUJEkaMY1+XVhVk8DkouF3N6kpSZK0HvjEd0mSpBYYsiRJklpgyJIkSWqBIUuSJKkFhixJkqQWGLIkSZJaYMiSJElqgSFLkiSpBa2ErCSXJ6kkfb1AUZIkab0ZeshKcjbwq8DXh11bkiRpVDR6rU6SXcAeoIDpqtoJXA+8vze+sr8/DIcvbdKGJEnSmjPwTlaSLcAEsL2qXgK8M8nLgAuq6tYVrh1P0k3SnZubG7QFSZKkNavJTtZ2YH9VzfY+Pwr8AbB7pQurah+wD6DT6RQvuK1BG5IkSadK+l7Z5J6sMH9M+KSzgRcDtyW5G3gF8AVvfpckSaejJiFrCtiRZKz3+R9V1caq2lxVm4GvAZdVVbdpk5IkSaNm4OPCqppJshc4kOQ4cIg+jgolSZJOB41+XVhVk8DkMnOXNqktSZI0ynziuyRJUgsMWZIkSS0wZEmSJLXAkCVJktQCQ5YkSVILDFmSJEktMGRJkiS1wJAlSZLUAkOWJElSC4YaspK8Pck3ktyR5CtJXjTM+pIkSaNi2DtZN1fVT1bVS4H3AzcMub4kSdJIaPTuwiS7gD1AAdNVtXPB9I/1xk/s0cNwy6VN2pAkSVpzBg5ZSbYAE8C2qppNsqE3fjVwLfA0YPsy144D4wBbLzxz0BYkSZLWrFStvNm05IXJNcD5VTWxzPwvAq+vqitPVKfT6VS32x2oB0mSpFMpycGq6vSztsk9WeHEx4GfAd7UoL4kSdLIahKypoAdScYAkmxI8rwF8/8cuKtJc5IkSaNq4HuyqmomyV7gQJLjwCHgO0leC8wBjwAnPCqUJElarxr9urCqJoHJIfUiSZK0bvjEd0mSpBYYsiRJklpgyJIkSWqBIUuSJKkFhixJkqQWGLIkSZJaYMiSJElqgSFLkiSpBUMNWUmuTfLNJNNJppJcNMz6kiRJo2LYO1mHgE5VbQX2A+8fcn1JkqSR0Oi1Okl2AXuAAqaraueC6a8Bv7xikXvvgXe8rUkbkiRJa87AISvJFmAC2FZVs0k2LFpyFfAny1w7DowDbH3mOYO2IEmStGalqga7MLkGOL+qJpaY+2XgHcCrq+oHJ6rT6XSq2+0O1IMkSdKplORgVXX6WdvkuDDMHxMu/vLXMr/DtWLAkiRJWq+a3Pg+BexIMgaQZEOSlwE3AZdV1UPDaFCSJGkUDbyTVVUzSfYCB5IcZ/6XhZuAs4DPJwG4t6ouG0qnkiRJI6TRrwurahKYHFIvkiRJ64ZPfJckSWqBIUuSJKkFhixJkqQWGLIkSZJaYMiSJElqgSFLkiSpBYYsSZKkFhiyJEmSWmDIkiRJasFQQ1aSS5LcnuRYksuHWVuSJGmUDHsn615gN3DzkOtKkiSNlEbvLkyyC9gDFDBdVTt740/0XeTuB+Cq32jShiRJ0pozcMhKsgWYALZV1WySDSdx7TgwDrD1nI2DtiBJkrRmpaoGuzC5Bji/qiaWmPs4cGtV7V+pTqfTqW63O1APkiRJp1KSg1XV6Wdtk3uywvwxoSRJkhZpErKmgB1JxgBO5rhQkiRpvRs4ZFXVDLAXOJDkTuCGJBcnOQJcAdyUZGZIfUqSJI2URr8urKpJYHLR8KYmNSVJktYDn/guSZLUAkOWJElSCwxZkiRJLTBkSZIktcCQJUmS1AJDliRJUgsMWZIkSS0wZEmSJLVgqCEryZlJPpvkb5J8PcnmYdaXJEkaFcPeyboKeKSqngt8CPidIdeXJEkaCY1eq5NkF7AHKGAaOA/4zd70fuAjSVJV1eR7JEmSRs3AISvJFmAC2FZVs0k2AH8B3AdQVceSfAcYA2aXq/PYY48xPT09aBuSJElrUpPjwu3A/qqaBaiqo0CWWPdDu1hJxpN0k3Tn5uYatCBJkrQ2NTkuDD8coI4AFwBHkjwVeAZwdPGFVbUP2AfQ6XRq69atDdqQJElae5rsZE0BO5KMAfSOC78AXNmbvxz4svdjSZKk09HAO1lVNZNkL3AgyXHgEPB24JNJ/ob5Hay3DqdNSZKk0dLo14VVNQlMLhq+oklNSZKk9cAnvkuSJLXAkCVJktQCQ5YkSVILsto//kvyXeDwqjYhSZLUnxdU1dn9LGx04/uQHK6qzmo3IUmStJIk3X7XelwoSZLUAkOWJElSC9ZCyNq32g1IkiT1qe/csuo3vkuSJK1Ha2EnS5Ikad0xZEmSJLXAkCVJktQCQ5YkSVILDFmSJEkt+H8LopM0sc1fuQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 720x360 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Colour graph\n",
    "import matplotlib.pyplot as plt\n",
    "plt.close('all')\n",
    "f, (col) = plt.subplots(1, 1, sharey=True, figsize=(10, 5))\n",
    "col.set_xlim(1, 5)\n",
    "col.set_ylim(-0.5,15)\n",
    "col.set_title('Colours')\n",
    "\n",
    "x_major_ticks = [1, 5]\n",
    "x_major_tick_labels = [\"\", \"\"]\n",
    "col.set_xticks(x_major_ticks)\n",
    "col.set_xticklabels(x_major_tick_labels)\n",
    "\n",
    "y_major_ticks = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14]\n",
    "y_major_tick_labels = [\"c0\", \"c1\", \"c2\", \"c3\", \"c4\", \"c5\", \"c6\", \"c7\", \"c8\", \"c9\", \"c10\", \"c11\", \"c12\", \"c13\", \"c14\"]\n",
    "col.set_yticks(y_major_ticks)\n",
    "col.set_yticklabels(y_major_tick_labels)\n",
    " \n",
    "# c0 points\n",
    "x0 = [1,5]\n",
    "y0 = [0,0]\n",
    "# plotting the c1 points \n",
    "plt.plot(x0, y0, label = \"c0\", color=c0)\n",
    " \n",
    "# c1 points\n",
    "x1 = [1,5]\n",
    "y1 = [1,1]\n",
    "# plotting the c1 points \n",
    "plt.plot(x1, y1, label = \"c1\", color=c1)\n",
    "\n",
    "# c2 points\n",
    "x2 = [1,5]\n",
    "y2 = [2,2]\n",
    "# plotting the c2 points \n",
    "plt.plot(x2, y2, label = \"c2\", color=c2)\n",
    "\n",
    "# c3 points\n",
    "x3 = [1,5]\n",
    "y3 = [3,3]\n",
    "# plotting the c3 points \n",
    "plt.plot(x3, y3, label = \"c3\", color=c3)\n",
    "\n",
    "# c4 points\n",
    "x4 = [1,5]\n",
    "y4 = [4, 4]\n",
    "# plotting the c4 points \n",
    "plt.plot(x4, y4, label = \"c4\", color=c4)\n",
    "\n",
    "# c5 points\n",
    "x5 = [1,5]\n",
    "y5 = [5, 5]\n",
    "# plotting the c5 points \n",
    "plt.plot(x5, y5, label = \"c5\", color=c5)\n",
    "\n",
    "# c6 points\n",
    "x6 = [1,5]\n",
    "y6 = [6, 6]\n",
    "# plotting the c6 points \n",
    "plt.plot(x6, y6, label = \"c6\", color=c6)\n",
    "\n",
    "# c7 points\n",
    "x7 = [1,5]\n",
    "y7 = [7, 7]\n",
    "# plotting the c7 points \n",
    "plt.plot(x7, y7, label = \"c7\", color=c7)\n",
    "\n",
    "# c8 points\n",
    "x8 = [1,5]\n",
    "y8 = [8, 8]\n",
    "# plotting the c8 points \n",
    "plt.plot(x8, y8, label = \"c8\", color=c8)\n",
    "\n",
    "# c9 points\n",
    "x9 = [1,5]\n",
    "y9 = [9, 9]\n",
    "# plotting the c9 points \n",
    "plt.plot(x9, y9, label = \"c9\", color=c9)\n",
    "\n",
    "# c10 points\n",
    "x10 = [1,5]\n",
    "y10 = [10, 10]\n",
    "# plotting the c4 points \n",
    "plt.plot(x10, y10, label = \"c10\", color=c10)\n",
    "\n",
    "# c11 points\n",
    "x11 = [1,5]\n",
    "y11 = [11, 11]\n",
    "# plotting the c11 points \n",
    "plt.plot(x11, y11, label = \"c11\", color=c11)\n",
    "\n",
    "# c12 points\n",
    "x12 = [1,5]\n",
    "y12 = [12, 12]\n",
    "# plotting the c12 points \n",
    "plt.plot(x12, y12, label = \"c12\", color=c12)\n",
    "\n",
    "# c13 points\n",
    "x13 = [1,5]\n",
    "y13 = [13, 13]\n",
    "# plotting the c13 points \n",
    "plt.plot(x13, y13, label = \"c13\", color=c13)\n",
    "\n",
    "# c14 points\n",
    "x14 = [1,5]\n",
    "y14 = [14, 14]\n",
    "# plotting the c14 points \n",
    "plt.plot(x14, y14, label = \"c14\", color=c14)\n",
    "\n",
    "# # show a legend on the plot\n",
    "# plt.legend()\n",
    " \n",
    "# function to show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "missense = c6\n",
    "nonsense = c1\n",
    "synonymous = c4"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Totals for studies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Number of individuals in each study\n",
    "Jaiswaltotal = 17182\n",
    "Genovesetotal = 12380\n",
    "McKerreltotal = 4219 #McKerrel did not include TP53\n",
    "Youngtotal = 20\n",
    "Young2019totalcontrols = 69 #controls only\n",
    "ZinkWGStotal = 11262\n",
    "Acunatotal = 2006\n",
    "Coombstotal = 1591 #chemotherapy and radiotherapy naive individuals\n",
    "Desaitotal = 181 #controls only\n",
    "\n",
    "TP53_study_total = Jaiswaltotal+Genovesetotal+Youngtotal+Young2019totalcontrols+ZinkWGStotal+Acunatotal+Coombstotal+Desaitotal"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functions for calculating trinucleotide-context specific mutation rates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "# Amino Acid Code:\n",
    "\n",
    "Amino_acid_abbreviations = {'Ala': 'A', 'Arg': 'R', 'Asn': 'N', 'Asp': 'D',\\\n",
    "                           'Cys': 'C', 'Glu': 'E', 'Gln': 'Q', 'Gly': 'G',\\\n",
    "                           'His': 'H', 'Ile': 'I', 'Leu': 'L', 'Lys': 'K',\\\n",
    "                           'Met': 'M', 'Phe': 'F', 'Pro': 'P', 'Ser': 'S',\\\n",
    "                           'Thr': 'T', 'Trp': 'W', 'Tyr': 'Y', 'Val': 'V', '*': '*'}\n",
    "\n",
    "Genetic_Code = {'TTT': 'Phe', 'TTC': 'Phe', 'TTA': 'Leu', 'TTG': 'Leu',\\\n",
    "               'TCT': 'Ser', 'TCC': 'Ser', 'TCA': 'Ser', 'TCG': 'Ser',\\\n",
    "               'TAT': 'Tyr', 'TAC': 'Tyr', 'TAA': '*', 'TAG': '*',\\\n",
    "               'TGT': 'Cys', 'TGC': 'Cys', 'TGA': '*', 'TGG': 'Trp',\\\n",
    "               'CTT': 'Leu', 'CTC': 'Leu', 'CTA': 'Leu', 'CTG': 'Leu',\\\n",
    "               'CCT': 'Pro', 'CCC': 'Pro', 'CCA': 'Pro', 'CCG': 'Pro',\\\n",
    "               'CAT': 'His', 'CAC': 'His', 'CAA': 'Gln', 'CAG': 'Gln',\\\n",
    "               'CGT': 'Arg', 'CGC': 'Arg', 'CGA': 'Arg', 'CGG': 'Arg',\\\n",
    "               'ATT': 'Ile', 'ATC': 'Ile', 'ATA': 'Ile', 'ATG': 'Met',\\\n",
    "               'ACT': 'Thr', 'ACC': 'Thr', 'ACA': 'Thr', 'ACG': 'Thr',\\\n",
    "               'AAT': 'Asn', 'AAC': 'Asn', 'AAA': 'Lys', 'AAG': 'Lys',\\\n",
    "               'AGT': 'Ser', 'AGC': 'Ser', 'AGA': 'Arg', 'AGG': 'Arg',\\\n",
    "               'GTT': 'Val', 'GTC': 'Val', 'GTA': 'Val', 'GTG': 'Val',\\\n",
    "               'GCT': 'Ala', 'GCC': 'Ala', 'GCA': 'Ala', 'GCG': 'Ala',\\\n",
    "               'GAT': 'Asp', 'GAC': 'Asp', 'GAA': 'Glu', 'GAG': 'Glu',\\\n",
    "               'GGT': 'Gly', 'GGC': 'Gly', 'GGA': 'Gly', 'GGG': 'Gly'}\n",
    "\n",
    "base_alternative = {'A': ['T', 'C', 'G'], 'T': ['A', 'C', 'G'], 'G':['A', 'C', 'T'], 'C':['A', 'G', 'T']}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "Possible_AA_Changes = {}\n",
    "Possible_Base_Changes = {}\n",
    "Possible_Codon_Changes = {}\n",
    "for k, v in Genetic_Code.items():\n",
    "    codon = k\n",
    "    base_1 = str(codon[0])\n",
    "    base_2 = str(codon[1])\n",
    "    base_3 = str(codon[2])\n",
    "    new_codons = []\n",
    "    amino_acid_changes = []\n",
    "    base_changes = []\n",
    "    if base_1 in base_alternative.keys():\n",
    "        alternate_base = base_alternative[base_1]\n",
    "        for i in alternate_base:\n",
    "            base_changes.append(i)\n",
    "            new_codon = i+base_2+base_3\n",
    "            new_codons.append(new_codon)\n",
    "    if base_2 in base_alternative.keys():\n",
    "        alternate_base = base_alternative[base_2]\n",
    "        for i in alternate_base:\n",
    "            base_changes.append(i)\n",
    "            new_codon = base_1+i+base_3\n",
    "            new_codons.append(new_codon)\n",
    "    if base_3 in base_alternative.keys():\n",
    "        alternate_base = base_alternative[base_3]\n",
    "        for i in alternate_base:\n",
    "            base_changes.append(i)\n",
    "            new_codon = base_1+base_2+i\n",
    "            new_codons.append(new_codon)\n",
    "    Possible_Codon_Changes[k]=(new_codons[0],new_codons[1], new_codons[2], new_codons[3], new_codons[4],\\\n",
    "                               new_codons[5], new_codons[6], new_codons[7], new_codons[8]) #i.e. {'TTT': ['ATT', 'CTT', 'GTT'...etc]}\n",
    "    Possible_Base_Changes[k]=(base_changes[0], base_changes[1], base_changes[2], base_changes[3], base_changes[4],\\\n",
    "                              base_changes[5], base_changes[6], base_changes[7], base_changes[8]) #i.e. {'TTT': ['A', 'C', 'G'... etc]}\n",
    "    \n",
    "Possible_Amino_Acid_Changes = {}\n",
    "for k, v in Possible_Codon_Changes.items():\n",
    "    codon = k\n",
    "    codon_changes = v\n",
    "    amino_acid_changes = []\n",
    "    for i in codon_changes:\n",
    "        amino_acid_change = Genetic_Code[i]\n",
    "        amino_acid_changes.append(amino_acid_change)\n",
    "    Possible_Amino_Acid_Changes[k]=(amino_acid_changes[0], amino_acid_changes[1], amino_acid_changes[2],\\\n",
    "                                    amino_acid_changes[3], amino_acid_changes[4], amino_acid_changes[5],\\\n",
    "                                    amino_acid_changes[6], amino_acid_changes[7], amino_acid_changes[8]) #i.e. {'TTT': ['Ile', 'Leu', 'Val'...etc]}\n",
    "    \n",
    "Possible_Amino_Acid_Changes_letters = {}\n",
    "for k, v in Possible_Amino_Acid_Changes.items():\n",
    "    codon = k\n",
    "    codon_changes = v\n",
    "    amino_acid_changes_letters = []\n",
    "    for i in codon_changes:\n",
    "        amino_acid_change_letter = Amino_acid_abbreviations[i]\n",
    "        amino_acid_changes_letters.append(amino_acid_change_letter)\n",
    "    Possible_Amino_Acid_Changes_letters[k]=(amino_acid_changes_letters[0], amino_acid_changes_letters[1],\\\n",
    "                                            amino_acid_changes_letters[2], amino_acid_changes_letters[3],\\\n",
    "                                            amino_acid_changes_letters[4], amino_acid_changes_letters[5],\\\n",
    "                                            amino_acid_changes_letters[6], amino_acid_changes_letters[7],\\\n",
    "                                            amino_acid_changes_letters[8]) #i.e. {'TTT': ['I', 'L', 'V'...etc]}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'TTT': ('ATT', 'CTT', 'GTT', 'TAT', 'TCT', 'TGT', 'TTA', 'TTC', 'TTG'),\n",
       " 'TTC': ('ATC', 'CTC', 'GTC', 'TAC', 'TCC', 'TGC', 'TTA', 'TTG', 'TTT'),\n",
       " 'TTA': ('ATA', 'CTA', 'GTA', 'TAA', 'TCA', 'TGA', 'TTT', 'TTC', 'TTG'),\n",
       " 'TTG': ('ATG', 'CTG', 'GTG', 'TAG', 'TCG', 'TGG', 'TTA', 'TTC', 'TTT'),\n",
       " 'TCT': ('ACT', 'CCT', 'GCT', 'TAT', 'TGT', 'TTT', 'TCA', 'TCC', 'TCG'),\n",
       " 'TCC': ('ACC', 'CCC', 'GCC', 'TAC', 'TGC', 'TTC', 'TCA', 'TCG', 'TCT'),\n",
       " 'TCA': ('ACA', 'CCA', 'GCA', 'TAA', 'TGA', 'TTA', 'TCT', 'TCC', 'TCG'),\n",
       " 'TCG': ('ACG', 'CCG', 'GCG', 'TAG', 'TGG', 'TTG', 'TCA', 'TCC', 'TCT'),\n",
       " 'TAT': ('AAT', 'CAT', 'GAT', 'TTT', 'TCT', 'TGT', 'TAA', 'TAC', 'TAG'),\n",
       " 'TAC': ('AAC', 'CAC', 'GAC', 'TTC', 'TCC', 'TGC', 'TAA', 'TAG', 'TAT'),\n",
       " 'TAA': ('AAA', 'CAA', 'GAA', 'TTA', 'TCA', 'TGA', 'TAT', 'TAC', 'TAG'),\n",
       " 'TAG': ('AAG', 'CAG', 'GAG', 'TTG', 'TCG', 'TGG', 'TAA', 'TAC', 'TAT'),\n",
       " 'TGT': ('AGT', 'CGT', 'GGT', 'TAT', 'TCT', 'TTT', 'TGA', 'TGC', 'TGG'),\n",
       " 'TGC': ('AGC', 'CGC', 'GGC', 'TAC', 'TCC', 'TTC', 'TGA', 'TGG', 'TGT'),\n",
       " 'TGA': ('AGA', 'CGA', 'GGA', 'TAA', 'TCA', 'TTA', 'TGT', 'TGC', 'TGG'),\n",
       " 'TGG': ('AGG', 'CGG', 'GGG', 'TAG', 'TCG', 'TTG', 'TGA', 'TGC', 'TGT'),\n",
       " 'CTT': ('ATT', 'GTT', 'TTT', 'CAT', 'CCT', 'CGT', 'CTA', 'CTC', 'CTG'),\n",
       " 'CTC': ('ATC', 'GTC', 'TTC', 'CAC', 'CCC', 'CGC', 'CTA', 'CTG', 'CTT'),\n",
       " 'CTA': ('ATA', 'GTA', 'TTA', 'CAA', 'CCA', 'CGA', 'CTT', 'CTC', 'CTG'),\n",
       " 'CTG': ('ATG', 'GTG', 'TTG', 'CAG', 'CCG', 'CGG', 'CTA', 'CTC', 'CTT'),\n",
       " 'CCT': ('ACT', 'GCT', 'TCT', 'CAT', 'CGT', 'CTT', 'CCA', 'CCC', 'CCG'),\n",
       " 'CCC': ('ACC', 'GCC', 'TCC', 'CAC', 'CGC', 'CTC', 'CCA', 'CCG', 'CCT'),\n",
       " 'CCA': ('ACA', 'GCA', 'TCA', 'CAA', 'CGA', 'CTA', 'CCT', 'CCC', 'CCG'),\n",
       " 'CCG': ('ACG', 'GCG', 'TCG', 'CAG', 'CGG', 'CTG', 'CCA', 'CCC', 'CCT'),\n",
       " 'CAT': ('AAT', 'GAT', 'TAT', 'CTT', 'CCT', 'CGT', 'CAA', 'CAC', 'CAG'),\n",
       " 'CAC': ('AAC', 'GAC', 'TAC', 'CTC', 'CCC', 'CGC', 'CAA', 'CAG', 'CAT'),\n",
       " 'CAA': ('AAA', 'GAA', 'TAA', 'CTA', 'CCA', 'CGA', 'CAT', 'CAC', 'CAG'),\n",
       " 'CAG': ('AAG', 'GAG', 'TAG', 'CTG', 'CCG', 'CGG', 'CAA', 'CAC', 'CAT'),\n",
       " 'CGT': ('AGT', 'GGT', 'TGT', 'CAT', 'CCT', 'CTT', 'CGA', 'CGC', 'CGG'),\n",
       " 'CGC': ('AGC', 'GGC', 'TGC', 'CAC', 'CCC', 'CTC', 'CGA', 'CGG', 'CGT'),\n",
       " 'CGA': ('AGA', 'GGA', 'TGA', 'CAA', 'CCA', 'CTA', 'CGT', 'CGC', 'CGG'),\n",
       " 'CGG': ('AGG', 'GGG', 'TGG', 'CAG', 'CCG', 'CTG', 'CGA', 'CGC', 'CGT'),\n",
       " 'ATT': ('TTT', 'CTT', 'GTT', 'AAT', 'ACT', 'AGT', 'ATA', 'ATC', 'ATG'),\n",
       " 'ATC': ('TTC', 'CTC', 'GTC', 'AAC', 'ACC', 'AGC', 'ATA', 'ATG', 'ATT'),\n",
       " 'ATA': ('TTA', 'CTA', 'GTA', 'AAA', 'ACA', 'AGA', 'ATT', 'ATC', 'ATG'),\n",
       " 'ATG': ('TTG', 'CTG', 'GTG', 'AAG', 'ACG', 'AGG', 'ATA', 'ATC', 'ATT'),\n",
       " 'ACT': ('TCT', 'CCT', 'GCT', 'AAT', 'AGT', 'ATT', 'ACA', 'ACC', 'ACG'),\n",
       " 'ACC': ('TCC', 'CCC', 'GCC', 'AAC', 'AGC', 'ATC', 'ACA', 'ACG', 'ACT'),\n",
       " 'ACA': ('TCA', 'CCA', 'GCA', 'AAA', 'AGA', 'ATA', 'ACT', 'ACC', 'ACG'),\n",
       " 'ACG': ('TCG', 'CCG', 'GCG', 'AAG', 'AGG', 'ATG', 'ACA', 'ACC', 'ACT'),\n",
       " 'AAT': ('TAT', 'CAT', 'GAT', 'ATT', 'ACT', 'AGT', 'AAA', 'AAC', 'AAG'),\n",
       " 'AAC': ('TAC', 'CAC', 'GAC', 'ATC', 'ACC', 'AGC', 'AAA', 'AAG', 'AAT'),\n",
       " 'AAA': ('TAA', 'CAA', 'GAA', 'ATA', 'ACA', 'AGA', 'AAT', 'AAC', 'AAG'),\n",
       " 'AAG': ('TAG', 'CAG', 'GAG', 'ATG', 'ACG', 'AGG', 'AAA', 'AAC', 'AAT'),\n",
       " 'AGT': ('TGT', 'CGT', 'GGT', 'AAT', 'ACT', 'ATT', 'AGA', 'AGC', 'AGG'),\n",
       " 'AGC': ('TGC', 'CGC', 'GGC', 'AAC', 'ACC', 'ATC', 'AGA', 'AGG', 'AGT'),\n",
       " 'AGA': ('TGA', 'CGA', 'GGA', 'AAA', 'ACA', 'ATA', 'AGT', 'AGC', 'AGG'),\n",
       " 'AGG': ('TGG', 'CGG', 'GGG', 'AAG', 'ACG', 'ATG', 'AGA', 'AGC', 'AGT'),\n",
       " 'GTT': ('ATT', 'CTT', 'TTT', 'GAT', 'GCT', 'GGT', 'GTA', 'GTC', 'GTG'),\n",
       " 'GTC': ('ATC', 'CTC', 'TTC', 'GAC', 'GCC', 'GGC', 'GTA', 'GTG', 'GTT'),\n",
       " 'GTA': ('ATA', 'CTA', 'TTA', 'GAA', 'GCA', 'GGA', 'GTT', 'GTC', 'GTG'),\n",
       " 'GTG': ('ATG', 'CTG', 'TTG', 'GAG', 'GCG', 'GGG', 'GTA', 'GTC', 'GTT'),\n",
       " 'GCT': ('ACT', 'CCT', 'TCT', 'GAT', 'GGT', 'GTT', 'GCA', 'GCC', 'GCG'),\n",
       " 'GCC': ('ACC', 'CCC', 'TCC', 'GAC', 'GGC', 'GTC', 'GCA', 'GCG', 'GCT'),\n",
       " 'GCA': ('ACA', 'CCA', 'TCA', 'GAA', 'GGA', 'GTA', 'GCT', 'GCC', 'GCG'),\n",
       " 'GCG': ('ACG', 'CCG', 'TCG', 'GAG', 'GGG', 'GTG', 'GCA', 'GCC', 'GCT'),\n",
       " 'GAT': ('AAT', 'CAT', 'TAT', 'GTT', 'GCT', 'GGT', 'GAA', 'GAC', 'GAG'),\n",
       " 'GAC': ('AAC', 'CAC', 'TAC', 'GTC', 'GCC', 'GGC', 'GAA', 'GAG', 'GAT'),\n",
       " 'GAA': ('AAA', 'CAA', 'TAA', 'GTA', 'GCA', 'GGA', 'GAT', 'GAC', 'GAG'),\n",
       " 'GAG': ('AAG', 'CAG', 'TAG', 'GTG', 'GCG', 'GGG', 'GAA', 'GAC', 'GAT'),\n",
       " 'GGT': ('AGT', 'CGT', 'TGT', 'GAT', 'GCT', 'GTT', 'GGA', 'GGC', 'GGG'),\n",
       " 'GGC': ('AGC', 'CGC', 'TGC', 'GAC', 'GCC', 'GTC', 'GGA', 'GGG', 'GGT'),\n",
       " 'GGA': ('AGA', 'CGA', 'TGA', 'GAA', 'GCA', 'GTA', 'GGT', 'GGC', 'GGG'),\n",
       " 'GGG': ('AGG', 'CGG', 'TGG', 'GAG', 'GCG', 'GTG', 'GGA', 'GGC', 'GGT')}"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Possible_Codon_Changes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'TTT': ('A', 'C', 'G', 'A', 'C', 'G', 'A', 'C', 'G'),\n",
       " 'TTC': ('A', 'C', 'G', 'A', 'C', 'G', 'A', 'G', 'T'),\n",
       " 'TTA': ('A', 'C', 'G', 'A', 'C', 'G', 'T', 'C', 'G'),\n",
       " 'TTG': ('A', 'C', 'G', 'A', 'C', 'G', 'A', 'C', 'T'),\n",
       " 'TCT': ('A', 'C', 'G', 'A', 'G', 'T', 'A', 'C', 'G'),\n",
       " 'TCC': ('A', 'C', 'G', 'A', 'G', 'T', 'A', 'G', 'T'),\n",
       " 'TCA': ('A', 'C', 'G', 'A', 'G', 'T', 'T', 'C', 'G'),\n",
       " 'TCG': ('A', 'C', 'G', 'A', 'G', 'T', 'A', 'C', 'T'),\n",
       " 'TAT': ('A', 'C', 'G', 'T', 'C', 'G', 'A', 'C', 'G'),\n",
       " 'TAC': ('A', 'C', 'G', 'T', 'C', 'G', 'A', 'G', 'T'),\n",
       " 'TAA': ('A', 'C', 'G', 'T', 'C', 'G', 'T', 'C', 'G'),\n",
       " 'TAG': ('A', 'C', 'G', 'T', 'C', 'G', 'A', 'C', 'T'),\n",
       " 'TGT': ('A', 'C', 'G', 'A', 'C', 'T', 'A', 'C', 'G'),\n",
       " 'TGC': ('A', 'C', 'G', 'A', 'C', 'T', 'A', 'G', 'T'),\n",
       " 'TGA': ('A', 'C', 'G', 'A', 'C', 'T', 'T', 'C', 'G'),\n",
       " 'TGG': ('A', 'C', 'G', 'A', 'C', 'T', 'A', 'C', 'T'),\n",
       " 'CTT': ('A', 'G', 'T', 'A', 'C', 'G', 'A', 'C', 'G'),\n",
       " 'CTC': ('A', 'G', 'T', 'A', 'C', 'G', 'A', 'G', 'T'),\n",
       " 'CTA': ('A', 'G', 'T', 'A', 'C', 'G', 'T', 'C', 'G'),\n",
       " 'CTG': ('A', 'G', 'T', 'A', 'C', 'G', 'A', 'C', 'T'),\n",
       " 'CCT': ('A', 'G', 'T', 'A', 'G', 'T', 'A', 'C', 'G'),\n",
       " 'CCC': ('A', 'G', 'T', 'A', 'G', 'T', 'A', 'G', 'T'),\n",
       " 'CCA': ('A', 'G', 'T', 'A', 'G', 'T', 'T', 'C', 'G'),\n",
       " 'CCG': ('A', 'G', 'T', 'A', 'G', 'T', 'A', 'C', 'T'),\n",
       " 'CAT': ('A', 'G', 'T', 'T', 'C', 'G', 'A', 'C', 'G'),\n",
       " 'CAC': ('A', 'G', 'T', 'T', 'C', 'G', 'A', 'G', 'T'),\n",
       " 'CAA': ('A', 'G', 'T', 'T', 'C', 'G', 'T', 'C', 'G'),\n",
       " 'CAG': ('A', 'G', 'T', 'T', 'C', 'G', 'A', 'C', 'T'),\n",
       " 'CGT': ('A', 'G', 'T', 'A', 'C', 'T', 'A', 'C', 'G'),\n",
       " 'CGC': ('A', 'G', 'T', 'A', 'C', 'T', 'A', 'G', 'T'),\n",
       " 'CGA': ('A', 'G', 'T', 'A', 'C', 'T', 'T', 'C', 'G'),\n",
       " 'CGG': ('A', 'G', 'T', 'A', 'C', 'T', 'A', 'C', 'T'),\n",
       " 'ATT': ('T', 'C', 'G', 'A', 'C', 'G', 'A', 'C', 'G'),\n",
       " 'ATC': ('T', 'C', 'G', 'A', 'C', 'G', 'A', 'G', 'T'),\n",
       " 'ATA': ('T', 'C', 'G', 'A', 'C', 'G', 'T', 'C', 'G'),\n",
       " 'ATG': ('T', 'C', 'G', 'A', 'C', 'G', 'A', 'C', 'T'),\n",
       " 'ACT': ('T', 'C', 'G', 'A', 'G', 'T', 'A', 'C', 'G'),\n",
       " 'ACC': ('T', 'C', 'G', 'A', 'G', 'T', 'A', 'G', 'T'),\n",
       " 'ACA': ('T', 'C', 'G', 'A', 'G', 'T', 'T', 'C', 'G'),\n",
       " 'ACG': ('T', 'C', 'G', 'A', 'G', 'T', 'A', 'C', 'T'),\n",
       " 'AAT': ('T', 'C', 'G', 'T', 'C', 'G', 'A', 'C', 'G'),\n",
       " 'AAC': ('T', 'C', 'G', 'T', 'C', 'G', 'A', 'G', 'T'),\n",
       " 'AAA': ('T', 'C', 'G', 'T', 'C', 'G', 'T', 'C', 'G'),\n",
       " 'AAG': ('T', 'C', 'G', 'T', 'C', 'G', 'A', 'C', 'T'),\n",
       " 'AGT': ('T', 'C', 'G', 'A', 'C', 'T', 'A', 'C', 'G'),\n",
       " 'AGC': ('T', 'C', 'G', 'A', 'C', 'T', 'A', 'G', 'T'),\n",
       " 'AGA': ('T', 'C', 'G', 'A', 'C', 'T', 'T', 'C', 'G'),\n",
       " 'AGG': ('T', 'C', 'G', 'A', 'C', 'T', 'A', 'C', 'T'),\n",
       " 'GTT': ('A', 'C', 'T', 'A', 'C', 'G', 'A', 'C', 'G'),\n",
       " 'GTC': ('A', 'C', 'T', 'A', 'C', 'G', 'A', 'G', 'T'),\n",
       " 'GTA': ('A', 'C', 'T', 'A', 'C', 'G', 'T', 'C', 'G'),\n",
       " 'GTG': ('A', 'C', 'T', 'A', 'C', 'G', 'A', 'C', 'T'),\n",
       " 'GCT': ('A', 'C', 'T', 'A', 'G', 'T', 'A', 'C', 'G'),\n",
       " 'GCC': ('A', 'C', 'T', 'A', 'G', 'T', 'A', 'G', 'T'),\n",
       " 'GCA': ('A', 'C', 'T', 'A', 'G', 'T', 'T', 'C', 'G'),\n",
       " 'GCG': ('A', 'C', 'T', 'A', 'G', 'T', 'A', 'C', 'T'),\n",
       " 'GAT': ('A', 'C', 'T', 'T', 'C', 'G', 'A', 'C', 'G'),\n",
       " 'GAC': ('A', 'C', 'T', 'T', 'C', 'G', 'A', 'G', 'T'),\n",
       " 'GAA': ('A', 'C', 'T', 'T', 'C', 'G', 'T', 'C', 'G'),\n",
       " 'GAG': ('A', 'C', 'T', 'T', 'C', 'G', 'A', 'C', 'T'),\n",
       " 'GGT': ('A', 'C', 'T', 'A', 'C', 'T', 'A', 'C', 'G'),\n",
       " 'GGC': ('A', 'C', 'T', 'A', 'C', 'T', 'A', 'G', 'T'),\n",
       " 'GGA': ('A', 'C', 'T', 'A', 'C', 'T', 'T', 'C', 'G'),\n",
       " 'GGG': ('A', 'C', 'T', 'A', 'C', 'T', 'A', 'C', 'T')}"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Possible_Base_Changes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'TTT': ('Ile', 'Leu', 'Val', 'Tyr', 'Ser', 'Cys', 'Leu', 'Phe', 'Leu'),\n",
       " 'TTC': ('Ile', 'Leu', 'Val', 'Tyr', 'Ser', 'Cys', 'Leu', 'Leu', 'Phe'),\n",
       " 'TTA': ('Ile', 'Leu', 'Val', '*', 'Ser', '*', 'Phe', 'Phe', 'Leu'),\n",
       " 'TTG': ('Met', 'Leu', 'Val', '*', 'Ser', 'Trp', 'Leu', 'Phe', 'Phe'),\n",
       " 'TCT': ('Thr', 'Pro', 'Ala', 'Tyr', 'Cys', 'Phe', 'Ser', 'Ser', 'Ser'),\n",
       " 'TCC': ('Thr', 'Pro', 'Ala', 'Tyr', 'Cys', 'Phe', 'Ser', 'Ser', 'Ser'),\n",
       " 'TCA': ('Thr', 'Pro', 'Ala', '*', '*', 'Leu', 'Ser', 'Ser', 'Ser'),\n",
       " 'TCG': ('Thr', 'Pro', 'Ala', '*', 'Trp', 'Leu', 'Ser', 'Ser', 'Ser'),\n",
       " 'TAT': ('Asn', 'His', 'Asp', 'Phe', 'Ser', 'Cys', '*', 'Tyr', '*'),\n",
       " 'TAC': ('Asn', 'His', 'Asp', 'Phe', 'Ser', 'Cys', '*', '*', 'Tyr'),\n",
       " 'TAA': ('Lys', 'Gln', 'Glu', 'Leu', 'Ser', '*', 'Tyr', 'Tyr', '*'),\n",
       " 'TAG': ('Lys', 'Gln', 'Glu', 'Leu', 'Ser', 'Trp', '*', 'Tyr', 'Tyr'),\n",
       " 'TGT': ('Ser', 'Arg', 'Gly', 'Tyr', 'Ser', 'Phe', '*', 'Cys', 'Trp'),\n",
       " 'TGC': ('Ser', 'Arg', 'Gly', 'Tyr', 'Ser', 'Phe', '*', 'Trp', 'Cys'),\n",
       " 'TGA': ('Arg', 'Arg', 'Gly', '*', 'Ser', 'Leu', 'Cys', 'Cys', 'Trp'),\n",
       " 'TGG': ('Arg', 'Arg', 'Gly', '*', 'Ser', 'Leu', '*', 'Cys', 'Cys'),\n",
       " 'CTT': ('Ile', 'Val', 'Phe', 'His', 'Pro', 'Arg', 'Leu', 'Leu', 'Leu'),\n",
       " 'CTC': ('Ile', 'Val', 'Phe', 'His', 'Pro', 'Arg', 'Leu', 'Leu', 'Leu'),\n",
       " 'CTA': ('Ile', 'Val', 'Leu', 'Gln', 'Pro', 'Arg', 'Leu', 'Leu', 'Leu'),\n",
       " 'CTG': ('Met', 'Val', 'Leu', 'Gln', 'Pro', 'Arg', 'Leu', 'Leu', 'Leu'),\n",
       " 'CCT': ('Thr', 'Ala', 'Ser', 'His', 'Arg', 'Leu', 'Pro', 'Pro', 'Pro'),\n",
       " 'CCC': ('Thr', 'Ala', 'Ser', 'His', 'Arg', 'Leu', 'Pro', 'Pro', 'Pro'),\n",
       " 'CCA': ('Thr', 'Ala', 'Ser', 'Gln', 'Arg', 'Leu', 'Pro', 'Pro', 'Pro'),\n",
       " 'CCG': ('Thr', 'Ala', 'Ser', 'Gln', 'Arg', 'Leu', 'Pro', 'Pro', 'Pro'),\n",
       " 'CAT': ('Asn', 'Asp', 'Tyr', 'Leu', 'Pro', 'Arg', 'Gln', 'His', 'Gln'),\n",
       " 'CAC': ('Asn', 'Asp', 'Tyr', 'Leu', 'Pro', 'Arg', 'Gln', 'Gln', 'His'),\n",
       " 'CAA': ('Lys', 'Glu', '*', 'Leu', 'Pro', 'Arg', 'His', 'His', 'Gln'),\n",
       " 'CAG': ('Lys', 'Glu', '*', 'Leu', 'Pro', 'Arg', 'Gln', 'His', 'His'),\n",
       " 'CGT': ('Ser', 'Gly', 'Cys', 'His', 'Pro', 'Leu', 'Arg', 'Arg', 'Arg'),\n",
       " 'CGC': ('Ser', 'Gly', 'Cys', 'His', 'Pro', 'Leu', 'Arg', 'Arg', 'Arg'),\n",
       " 'CGA': ('Arg', 'Gly', '*', 'Gln', 'Pro', 'Leu', 'Arg', 'Arg', 'Arg'),\n",
       " 'CGG': ('Arg', 'Gly', 'Trp', 'Gln', 'Pro', 'Leu', 'Arg', 'Arg', 'Arg'),\n",
       " 'ATT': ('Phe', 'Leu', 'Val', 'Asn', 'Thr', 'Ser', 'Ile', 'Ile', 'Met'),\n",
       " 'ATC': ('Phe', 'Leu', 'Val', 'Asn', 'Thr', 'Ser', 'Ile', 'Met', 'Ile'),\n",
       " 'ATA': ('Leu', 'Leu', 'Val', 'Lys', 'Thr', 'Arg', 'Ile', 'Ile', 'Met'),\n",
       " 'ATG': ('Leu', 'Leu', 'Val', 'Lys', 'Thr', 'Arg', 'Ile', 'Ile', 'Ile'),\n",
       " 'ACT': ('Ser', 'Pro', 'Ala', 'Asn', 'Ser', 'Ile', 'Thr', 'Thr', 'Thr'),\n",
       " 'ACC': ('Ser', 'Pro', 'Ala', 'Asn', 'Ser', 'Ile', 'Thr', 'Thr', 'Thr'),\n",
       " 'ACA': ('Ser', 'Pro', 'Ala', 'Lys', 'Arg', 'Ile', 'Thr', 'Thr', 'Thr'),\n",
       " 'ACG': ('Ser', 'Pro', 'Ala', 'Lys', 'Arg', 'Met', 'Thr', 'Thr', 'Thr'),\n",
       " 'AAT': ('Tyr', 'His', 'Asp', 'Ile', 'Thr', 'Ser', 'Lys', 'Asn', 'Lys'),\n",
       " 'AAC': ('Tyr', 'His', 'Asp', 'Ile', 'Thr', 'Ser', 'Lys', 'Lys', 'Asn'),\n",
       " 'AAA': ('*', 'Gln', 'Glu', 'Ile', 'Thr', 'Arg', 'Asn', 'Asn', 'Lys'),\n",
       " 'AAG': ('*', 'Gln', 'Glu', 'Met', 'Thr', 'Arg', 'Lys', 'Asn', 'Asn'),\n",
       " 'AGT': ('Cys', 'Arg', 'Gly', 'Asn', 'Thr', 'Ile', 'Arg', 'Ser', 'Arg'),\n",
       " 'AGC': ('Cys', 'Arg', 'Gly', 'Asn', 'Thr', 'Ile', 'Arg', 'Arg', 'Ser'),\n",
       " 'AGA': ('*', 'Arg', 'Gly', 'Lys', 'Thr', 'Ile', 'Ser', 'Ser', 'Arg'),\n",
       " 'AGG': ('Trp', 'Arg', 'Gly', 'Lys', 'Thr', 'Met', 'Arg', 'Ser', 'Ser'),\n",
       " 'GTT': ('Ile', 'Leu', 'Phe', 'Asp', 'Ala', 'Gly', 'Val', 'Val', 'Val'),\n",
       " 'GTC': ('Ile', 'Leu', 'Phe', 'Asp', 'Ala', 'Gly', 'Val', 'Val', 'Val'),\n",
       " 'GTA': ('Ile', 'Leu', 'Leu', 'Glu', 'Ala', 'Gly', 'Val', 'Val', 'Val'),\n",
       " 'GTG': ('Met', 'Leu', 'Leu', 'Glu', 'Ala', 'Gly', 'Val', 'Val', 'Val'),\n",
       " 'GCT': ('Thr', 'Pro', 'Ser', 'Asp', 'Gly', 'Val', 'Ala', 'Ala', 'Ala'),\n",
       " 'GCC': ('Thr', 'Pro', 'Ser', 'Asp', 'Gly', 'Val', 'Ala', 'Ala', 'Ala'),\n",
       " 'GCA': ('Thr', 'Pro', 'Ser', 'Glu', 'Gly', 'Val', 'Ala', 'Ala', 'Ala'),\n",
       " 'GCG': ('Thr', 'Pro', 'Ser', 'Glu', 'Gly', 'Val', 'Ala', 'Ala', 'Ala'),\n",
       " 'GAT': ('Asn', 'His', 'Tyr', 'Val', 'Ala', 'Gly', 'Glu', 'Asp', 'Glu'),\n",
       " 'GAC': ('Asn', 'His', 'Tyr', 'Val', 'Ala', 'Gly', 'Glu', 'Glu', 'Asp'),\n",
       " 'GAA': ('Lys', 'Gln', '*', 'Val', 'Ala', 'Gly', 'Asp', 'Asp', 'Glu'),\n",
       " 'GAG': ('Lys', 'Gln', '*', 'Val', 'Ala', 'Gly', 'Glu', 'Asp', 'Asp'),\n",
       " 'GGT': ('Ser', 'Arg', 'Cys', 'Asp', 'Ala', 'Val', 'Gly', 'Gly', 'Gly'),\n",
       " 'GGC': ('Ser', 'Arg', 'Cys', 'Asp', 'Ala', 'Val', 'Gly', 'Gly', 'Gly'),\n",
       " 'GGA': ('Arg', 'Arg', '*', 'Glu', 'Ala', 'Val', 'Gly', 'Gly', 'Gly'),\n",
       " 'GGG': ('Arg', 'Arg', 'Trp', 'Glu', 'Ala', 'Val', 'Gly', 'Gly', 'Gly')}"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Possible_Amino_Acid_Changes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'TTT': ('I', 'L', 'V', 'Y', 'S', 'C', 'L', 'F', 'L'),\n",
       " 'TTC': ('I', 'L', 'V', 'Y', 'S', 'C', 'L', 'L', 'F'),\n",
       " 'TTA': ('I', 'L', 'V', '*', 'S', '*', 'F', 'F', 'L'),\n",
       " 'TTG': ('M', 'L', 'V', '*', 'S', 'W', 'L', 'F', 'F'),\n",
       " 'TCT': ('T', 'P', 'A', 'Y', 'C', 'F', 'S', 'S', 'S'),\n",
       " 'TCC': ('T', 'P', 'A', 'Y', 'C', 'F', 'S', 'S', 'S'),\n",
       " 'TCA': ('T', 'P', 'A', '*', '*', 'L', 'S', 'S', 'S'),\n",
       " 'TCG': ('T', 'P', 'A', '*', 'W', 'L', 'S', 'S', 'S'),\n",
       " 'TAT': ('N', 'H', 'D', 'F', 'S', 'C', '*', 'Y', '*'),\n",
       " 'TAC': ('N', 'H', 'D', 'F', 'S', 'C', '*', '*', 'Y'),\n",
       " 'TAA': ('K', 'Q', 'E', 'L', 'S', '*', 'Y', 'Y', '*'),\n",
       " 'TAG': ('K', 'Q', 'E', 'L', 'S', 'W', '*', 'Y', 'Y'),\n",
       " 'TGT': ('S', 'R', 'G', 'Y', 'S', 'F', '*', 'C', 'W'),\n",
       " 'TGC': ('S', 'R', 'G', 'Y', 'S', 'F', '*', 'W', 'C'),\n",
       " 'TGA': ('R', 'R', 'G', '*', 'S', 'L', 'C', 'C', 'W'),\n",
       " 'TGG': ('R', 'R', 'G', '*', 'S', 'L', '*', 'C', 'C'),\n",
       " 'CTT': ('I', 'V', 'F', 'H', 'P', 'R', 'L', 'L', 'L'),\n",
       " 'CTC': ('I', 'V', 'F', 'H', 'P', 'R', 'L', 'L', 'L'),\n",
       " 'CTA': ('I', 'V', 'L', 'Q', 'P', 'R', 'L', 'L', 'L'),\n",
       " 'CTG': ('M', 'V', 'L', 'Q', 'P', 'R', 'L', 'L', 'L'),\n",
       " 'CCT': ('T', 'A', 'S', 'H', 'R', 'L', 'P', 'P', 'P'),\n",
       " 'CCC': ('T', 'A', 'S', 'H', 'R', 'L', 'P', 'P', 'P'),\n",
       " 'CCA': ('T', 'A', 'S', 'Q', 'R', 'L', 'P', 'P', 'P'),\n",
       " 'CCG': ('T', 'A', 'S', 'Q', 'R', 'L', 'P', 'P', 'P'),\n",
       " 'CAT': ('N', 'D', 'Y', 'L', 'P', 'R', 'Q', 'H', 'Q'),\n",
       " 'CAC': ('N', 'D', 'Y', 'L', 'P', 'R', 'Q', 'Q', 'H'),\n",
       " 'CAA': ('K', 'E', '*', 'L', 'P', 'R', 'H', 'H', 'Q'),\n",
       " 'CAG': ('K', 'E', '*', 'L', 'P', 'R', 'Q', 'H', 'H'),\n",
       " 'CGT': ('S', 'G', 'C', 'H', 'P', 'L', 'R', 'R', 'R'),\n",
       " 'CGC': ('S', 'G', 'C', 'H', 'P', 'L', 'R', 'R', 'R'),\n",
       " 'CGA': ('R', 'G', '*', 'Q', 'P', 'L', 'R', 'R', 'R'),\n",
       " 'CGG': ('R', 'G', 'W', 'Q', 'P', 'L', 'R', 'R', 'R'),\n",
       " 'ATT': ('F', 'L', 'V', 'N', 'T', 'S', 'I', 'I', 'M'),\n",
       " 'ATC': ('F', 'L', 'V', 'N', 'T', 'S', 'I', 'M', 'I'),\n",
       " 'ATA': ('L', 'L', 'V', 'K', 'T', 'R', 'I', 'I', 'M'),\n",
       " 'ATG': ('L', 'L', 'V', 'K', 'T', 'R', 'I', 'I', 'I'),\n",
       " 'ACT': ('S', 'P', 'A', 'N', 'S', 'I', 'T', 'T', 'T'),\n",
       " 'ACC': ('S', 'P', 'A', 'N', 'S', 'I', 'T', 'T', 'T'),\n",
       " 'ACA': ('S', 'P', 'A', 'K', 'R', 'I', 'T', 'T', 'T'),\n",
       " 'ACG': ('S', 'P', 'A', 'K', 'R', 'M', 'T', 'T', 'T'),\n",
       " 'AAT': ('Y', 'H', 'D', 'I', 'T', 'S', 'K', 'N', 'K'),\n",
       " 'AAC': ('Y', 'H', 'D', 'I', 'T', 'S', 'K', 'K', 'N'),\n",
       " 'AAA': ('*', 'Q', 'E', 'I', 'T', 'R', 'N', 'N', 'K'),\n",
       " 'AAG': ('*', 'Q', 'E', 'M', 'T', 'R', 'K', 'N', 'N'),\n",
       " 'AGT': ('C', 'R', 'G', 'N', 'T', 'I', 'R', 'S', 'R'),\n",
       " 'AGC': ('C', 'R', 'G', 'N', 'T', 'I', 'R', 'R', 'S'),\n",
       " 'AGA': ('*', 'R', 'G', 'K', 'T', 'I', 'S', 'S', 'R'),\n",
       " 'AGG': ('W', 'R', 'G', 'K', 'T', 'M', 'R', 'S', 'S'),\n",
       " 'GTT': ('I', 'L', 'F', 'D', 'A', 'G', 'V', 'V', 'V'),\n",
       " 'GTC': ('I', 'L', 'F', 'D', 'A', 'G', 'V', 'V', 'V'),\n",
       " 'GTA': ('I', 'L', 'L', 'E', 'A', 'G', 'V', 'V', 'V'),\n",
       " 'GTG': ('M', 'L', 'L', 'E', 'A', 'G', 'V', 'V', 'V'),\n",
       " 'GCT': ('T', 'P', 'S', 'D', 'G', 'V', 'A', 'A', 'A'),\n",
       " 'GCC': ('T', 'P', 'S', 'D', 'G', 'V', 'A', 'A', 'A'),\n",
       " 'GCA': ('T', 'P', 'S', 'E', 'G', 'V', 'A', 'A', 'A'),\n",
       " 'GCG': ('T', 'P', 'S', 'E', 'G', 'V', 'A', 'A', 'A'),\n",
       " 'GAT': ('N', 'H', 'Y', 'V', 'A', 'G', 'E', 'D', 'E'),\n",
       " 'GAC': ('N', 'H', 'Y', 'V', 'A', 'G', 'E', 'E', 'D'),\n",
       " 'GAA': ('K', 'Q', '*', 'V', 'A', 'G', 'D', 'D', 'E'),\n",
       " 'GAG': ('K', 'Q', '*', 'V', 'A', 'G', 'E', 'D', 'D'),\n",
       " 'GGT': ('S', 'R', 'C', 'D', 'A', 'V', 'G', 'G', 'G'),\n",
       " 'GGC': ('S', 'R', 'C', 'D', 'A', 'V', 'G', 'G', 'G'),\n",
       " 'GGA': ('R', 'R', '*', 'E', 'A', 'V', 'G', 'G', 'G'),\n",
       " 'GGG': ('R', 'R', 'W', 'E', 'A', 'V', 'G', 'G', 'G')}"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Possible_Amino_Acid_Changes_letters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#Import mutation rates\n",
    "import csv\n",
    "filename=\"Data files/Trinucleotide_context_mutation_rates.csv\" #csv file containing trinucleotide-context specific mutation rates (calculated from Lee-Six et al 2018 data)\n",
    "\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    Mutation_rates = {} #create a dictionary where key = trincucleotide site, value = mutation rate\n",
    "\n",
    "    for row in read_reader:\n",
    "        if row_count>0:\n",
    "            site = row[0] #trinucleotide site\n",
    "            comp_site = row[2] #complementary site\n",
    "            if site != '':\n",
    "                mutation_rate = float(row[7])\n",
    "                Mutation_rates[site] = mutation_rate\n",
    "                Mutation_rates[comp_site] = mutation_rate\n",
    "        row_count = row_count+1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculating mutation rates for individual variants"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TP53"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_variants_TP53(variant):\n",
    "    intron_exon_boundary_caution_base_TP53 = [74, 75, 96, 375, 672, 782, 919, 993, 1100]\n",
    "\n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    def intron_exon_check(position_in_gene_number):\n",
    "        if position_in_gene_number == 74:\n",
    "            gene_pos = 74\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 75:\n",
    "            gene_pos = 75\n",
    "            first_base = 'G'\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = TP53_exons_extra_bases[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 96:\n",
    "            gene_pos = 96\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 375:\n",
    "            gene_pos = 375\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 672:\n",
    "            gene_pos = 672\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 782:\n",
    "            gene_pos = 782\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 919:\n",
    "            gene_pos = 919\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 993:\n",
    "            gene_pos = 993\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1100:\n",
    "            gene_pos = 1100\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #TP53 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    TP53_exons = 'ATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGACATAGTGTGGTGGTGCCCTATGAGCCGCCTGAGGTTGGCTCTGACTGTACCACCATCCACTACAACTACATGTGTAACAGTTCCTGCATGGGCGGCATGAACCGGAGGCCCATCCTCACCATCATCACACTGGAAGACTCCAGTGGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGAGGAAGAGAATCTCCGCAAGAAAGGGGAGCCTCACCACGAGCTGCCCCCAGGGAGCACTAAGCGAGCACTGCCCAACAACACCAGCTCCTCTCCCCAGCCAAAGAAGAAACCACTGGATGGAGAATATTTCACCCTTCAGATCCGTGGGCGTGAGCGCTTCGAGATGTTCCGAGAGCTGAATGAGGCCTTGGAACTCAAGGATGCCCAGGCTGGGAAGGAGCCAGGGGGGAGCAGGGCTCACTCCAGCCACCTGAAGTCCAAAAAGGGTCAGTCTACCTCCCGCCATAAAAAACTCATGTTCAAGACAGAAGGGCCTGACTCAGACTGA'\n",
    "    \n",
    "    #Split the sequence into codons \n",
    "    TP53_codons = [TP53_exons[i:i+3] for i in range(0, len(TP53_exons), 3)]\n",
    "\n",
    "    #TP53 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    TP53_exons_extra_bases = 'CATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGACATAGTGTGGTGGTGCCCTATGAGCCGCCTGAGGTTGGCTCTGACTGTACCACCATCCACTACAACTACATGTGTAACAGTTCCTGCATGGGCGGCATGAACCGGAGGCCCATCCTCACCATCATCACACTGGAAGACTCCAGTGGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGAGGAAGAGAATCTCCGCAAGAAAGGGGAGCCTCACCACGAGCTGCCCCCAGGGAGCACTAAGCGAGCACTGCCCAACAACACCAGCTCCTCTCCCCAGCCAAAGAAGAAACCACTGGATGGAGAATATTTCACCCTTCAGATCCGTGGGCGTGAGCGCTTCGAGATGTTCCGAGAGCTGAATGAGGCCTTGGAACTCAAGGATGCCCAGGCTGGGAAGGAGCCAGGGGGGAGCAGGGCTCACTCCAGCCACCTGAAGTCCAAAAAGGGTCAGTCTACCTCCCGCCATAAAAAACTCATGTTCAAGACAGAAGGGCCTGACTCAGACTGAC'\n",
    "    \n",
    "    #Create a list of the codons they called missense variants at (from COSMIC)\n",
    "    TP53_variant_list = []\n",
    "    TP53_missense_codons =[]\n",
    "    TP53_missense_variant_DICT = {}\n",
    "\n",
    "    variant_codon = int(((re.findall('\\d+', variant)))[0])\n",
    "\n",
    "    codon0 = variant_codon-1\n",
    "    codon1 = variant_codon\n",
    "    variant = variant\n",
    "    TP53_missense_codons.append(codon0) #List of codon numbers that they called variants at\n",
    "    TP53_missense_variant_DICT[row_count]=(codon1, variant) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    TP53_missense_codons_DICT = {}\n",
    "    calls_TP53_missense = {}\n",
    "\n",
    "    for i in TP53_missense_codons: #Iterates over the list of codon numbers\n",
    "        TP53_missense_codons_DICT[i+1]= (TP53_codons[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "\n",
    "    for k,v in TP53_missense_codons_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_TP53_missense[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_TP53_missense_extra = {}\n",
    "\n",
    "    for k, v in calls_TP53_missense.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_TP53_missense_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_TP53_missense_extra2 = {}\n",
    "\n",
    "    for k, v in calls_TP53_missense_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_TP53_missense_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    TP53_position_in_list_dict_missense={}\n",
    "    for k, v in TP53_missense_variant_DICT.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_TP53_missense[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_TP53_missense[v[0]]\n",
    "        codon_seq=calls_TP53_missense[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_TP53_missense_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                TP53_position_in_list_dict_missense[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "\n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in TP53_position_in_list_dict_missense.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "    #     print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TP53_bases_missense = {}\n",
    "    for k, v in TP53_position_in_list_dict_missense.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TP53_bases_missense[k]=(v, base_changes)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TP53_bases_missense2 = {}\n",
    "\n",
    "    for k, v in TP53_bases_missense.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TP53_bases_missense2[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TP53_bases_missense3 = {}\n",
    "\n",
    "    for k, v in TP53_bases_missense2.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TP53_bases_missense3[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TP53_bases_missense4 = {}\n",
    "\n",
    "    for k, v in TP53_bases_missense3.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_TP53:\n",
    "    #                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check(i)\n",
    "                sites_trio_missense.append([adapted_trio])\n",
    "                TP53_bases_missense4[k] = (v, sites_trio_missense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_missense = [TP53_exons_extra_bases[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_missense.append(site_trio_missense)\n",
    "                    TP53_bases_missense4[k] = (v, sites_trio_missense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TP53_sites_missense5 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TP53_bases_missense4.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        TP53_sites_missense5[a]=info\n",
    "        a=a+1\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TP53_sites_missense6={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TP53_sites_missense5.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (original_base,new_base)\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TP53_sites_missense6[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    TP53_mutation_rates_missense = {}\n",
    "    for k, v in TP53_sites_missense6.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            TP53_mutation_rates_missense[k] = (v, mutation_rate)\n",
    "\n",
    "    TP53_mutation_rate_missense_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TP53_mutation_rates_missense.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        TP53_mutation_rate_missense_sum.append(mutation_rate_value)\n",
    "\n",
    "    TP53_Summed_mutation_rate_missense_added = np.sum(TP53_mutation_rate_missense_sum)\n",
    "    print('mutation rate for specific variant = ', TP53_Summed_mutation_rate_missense_added)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return TP53_Summed_mutation_rate_missense_added"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculating mutation rates from a .csv file of variants"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TP53"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_list_nonsense_variants_TP53(csv_file_variants):\n",
    "\n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base = [74, 75, 96, 375, 672, 782, 919, 993, 1100]\n",
    "   \n",
    "    def intron_exon_check(position_in_gene_number):\n",
    "        if position_in_gene_number == 74:\n",
    "            gene_pos = 74\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 75:\n",
    "            gene_pos = 75\n",
    "            first_base = 'G'\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = TP53_exons_extra_bases[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 96:\n",
    "            gene_pos = 96\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 375:\n",
    "            gene_pos = 375\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 672:\n",
    "            gene_pos = 672\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 782:\n",
    "            gene_pos = 782\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 919:\n",
    "            gene_pos = 919\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 993:\n",
    "            gene_pos = 993\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1100:\n",
    "            gene_pos = 1100\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #TP53 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    TP53_exons = 'ATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGACATAGTGTGGTGGTGCCCTATGAGCCGCCTGAGGTTGGCTCTGACTGTACCACCATCCACTACAACTACATGTGTAACAGTTCCTGCATGGGCGGCATGAACCGGAGGCCCATCCTCACCATCATCACACTGGAAGACTCCAGTGGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGAGGAAGAGAATCTCCGCAAGAAAGGGGAGCCTCACCACGAGCTGCCCCCAGGGAGCACTAAGCGAGCACTGCCCAACAACACCAGCTCCTCTCCCCAGCCAAAGAAGAAACCACTGGATGGAGAATATTTCACCCTTCAGATCCGTGGGCGTGAGCGCTTCGAGATGTTCCGAGAGCTGAATGAGGCCTTGGAACTCAAGGATGCCCAGGCTGGGAAGGAGCCAGGGGGGAGCAGGGCTCACTCCAGCCACCTGAAGTCCAAAAAGGGTCAGTCTACCTCCCGCCATAAAAAACTCATGTTCAAGACAGAAGGGCCTGACTCAGACTGA'\n",
    "    \n",
    "    #Split the sequence into codons \n",
    "    TP53_codons = [TP53_exons[i:i+3] for i in range(0, len(TP53_exons), 3)]\n",
    "\n",
    "    #TP53 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    TP53_exons_extra_bases = 'CATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGACATAGTGTGGTGGTGCCCTATGAGCCGCCTGAGGTTGGCTCTGACTGTACCACCATCCACTACAACTACATGTGTAACAGTTCCTGCATGGGCGGCATGAACCGGAGGCCCATCCTCACCATCATCACACTGGAAGACTCCAGTGGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGAGGAAGAGAATCTCCGCAAGAAAGGGGAGCCTCACCACGAGCTGCCCCCAGGGAGCACTAAGCGAGCACTGCCCAACAACACCAGCTCCTCTCCCCAGCCAAAGAAGAAACCACTGGATGGAGAATATTTCACCCTTCAGATCCGTGGGCGTGAGCGCTTCGAGATGTTCCGAGAGCTGAATGAGGCCTTGGAACTCAAGGATGCCCAGGCTGGGAAGGAGCCAGGGGGGAGCAGGGCTCACTCCAGCCACCTGAAGTCCAAAAAGGGTCAGTCTACCTCCCGCCATAAAAAACTCATGTTCAAGACAGAAGGGCCTGACTCAGACTGAC'\n",
    "    \n",
    "    #Create a list of the codons they called nonsense variants at\n",
    "    import csv\n",
    "    filename = csv_file_variants\n",
    "    with open(filename, 'r') as csvfile:\n",
    "        read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "        row_count=0\n",
    "\n",
    "        TP53_nonsense_variant_list = []\n",
    "        TP53_nonsense_codons =[]\n",
    "        TP53_nonsense_variant_DICT = {}\n",
    "\n",
    "        for row in read_reader: #each row contains aa number (e.g. 11) and variant (e.g. p.D11A)\n",
    "    #         print(row)\n",
    "            if row_count>0: \n",
    "                codon0 = int(row[0])-1\n",
    "                codon1 = int(row[0])\n",
    "                variant = row[1]\n",
    "                TP53_nonsense_codons.append(codon0) #List of codon numbers that they called variants at\n",
    "                TP53_nonsense_variant_DICT[row_count]=(codon1, variant) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "            row_count=row_count+1\n",
    "            \n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    TP53_nonsense_codons_DICT = {}\n",
    "    calls_nonsense = {}\n",
    "\n",
    "    for i in TP53_nonsense_codons: #Iterates over the list of codon numbers\n",
    "        TP53_nonsense_codons_DICT[i+1]= (TP53_codons[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "        \n",
    "    for k,v in TP53_nonsense_codons_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_nonsense[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "        \n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_nonsense_extra = {}\n",
    "\n",
    "    for k, v in calls_nonsense.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_nonsense_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_nonsense_extra2 = {}\n",
    "\n",
    "    for k, v in calls_nonsense_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_nonsense_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "            \n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    TP53_position_in_list_dict_nonsense={}\n",
    "    for k, v in TP53_nonsense_variant_DICT.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_nonsense[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_nonsense[v[0]]\n",
    "        codon_seq=calls_nonsense[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_nonsense_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                TP53_position_in_list_dict_nonsense[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "            \n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in TP53_position_in_list_dict_nonsense.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "#     print('positions at which nonsense variants called = ', aa_positions_variants_called)\n",
    "                \n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TP53_bases_nonsense = {}\n",
    "    for k, v in TP53_position_in_list_dict_nonsense.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TP53_bases_nonsense[k]=(v, base_changes)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TP53_bases_nonsense2 = {}\n",
    "\n",
    "    for k, v in TP53_bases_nonsense.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TP53_bases_nonsense2[k] = (v, codon_positions)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TP53_bases_nonsense3 = {}\n",
    "\n",
    "    for k, v in TP53_bases_nonsense2.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TP53_bases_nonsense3[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TP53_bases_nonsense4 = {}\n",
    "\n",
    "    for k, v in TP53_bases_nonsense3.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_nonsense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check(i)\n",
    "                sites_trio_nonsense.append([adapted_trio])\n",
    "                TP53_bases_nonsense4[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_nonsense = [TP53_exons_extra_bases[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_nonsense.append(site_trio_nonsense)\n",
    "                    TP53_bases_nonsense4[k] = (v, sites_trio_nonsense)\n",
    "                    \n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TP53_sites_nonsense5 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TP53_bases_nonsense4.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        TP53_sites_nonsense5[a]=info\n",
    "        a=a+1\n",
    "#     print(TP53_sites_nonsense5)\n",
    "        \n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TP53_sites_nonsense6={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TP53_sites_nonsense5.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "    \n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TP53_sites_nonsense6[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "        \n",
    "        \n",
    "    #Assign the mutation rates to each site...\n",
    "    TP53_mutation_rates_nonsense = {}\n",
    "    for k, v in TP53_sites_nonsense6.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            TP53_mutation_rates_nonsense[k] = (v, mutation_rate)\n",
    "\n",
    "    TP53_mutation_rate_nonsense_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TP53_mutation_rates_nonsense.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        TP53_mutation_rate_nonsense_sum.append(mutation_rate_value)\n",
    "    \n",
    "    TP53_Summed_mutation_rate_nonsense_added = np.sum(TP53_mutation_rate_nonsense_sum)\n",
    "\n",
    "    print('Total summed mutation rate across TP53 nonsense for this study = ', TP53_Summed_mutation_rate_nonsense_added)\n",
    "    \n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return TP53_Summed_mutation_rate_nonsense_added"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_list_missense_variants_TP53(csv_file_variants):\n",
    "    \n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base_TP53 = [74, 75, 96, 375, 672, 782, 919, 993, 1100]\n",
    "   \n",
    "    def intron_exon_check(position_in_gene_number):\n",
    "        if position_in_gene_number == 74:\n",
    "            gene_pos = 74\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 75:\n",
    "            gene_pos = 75\n",
    "            first_base = 'G'\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = TP53_exons_extra_bases[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 96:\n",
    "            gene_pos = 96\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 375:\n",
    "            gene_pos = 375\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 672:\n",
    "            gene_pos = 672\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 782:\n",
    "            gene_pos = 782\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 919:\n",
    "            gene_pos = 919\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 993:\n",
    "            gene_pos = 993\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1100:\n",
    "            gene_pos = 1100\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #TP53 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    TP53_exons = 'ATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGACATAGTGTGGTGGTGCCCTATGAGCCGCCTGAGGTTGGCTCTGACTGTACCACCATCCACTACAACTACATGTGTAACAGTTCCTGCATGGGCGGCATGAACCGGAGGCCCATCCTCACCATCATCACACTGGAAGACTCCAGTGGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGAGGAAGAGAATCTCCGCAAGAAAGGGGAGCCTCACCACGAGCTGCCCCCAGGGAGCACTAAGCGAGCACTGCCCAACAACACCAGCTCCTCTCCCCAGCCAAAGAAGAAACCACTGGATGGAGAATATTTCACCCTTCAGATCCGTGGGCGTGAGCGCTTCGAGATGTTCCGAGAGCTGAATGAGGCCTTGGAACTCAAGGATGCCCAGGCTGGGAAGGAGCCAGGGGGGAGCAGGGCTCACTCCAGCCACCTGAAGTCCAAAAAGGGTCAGTCTACCTCCCGCCATAAAAAACTCATGTTCAAGACAGAAGGGCCTGACTCAGACTGA'\n",
    "    \n",
    "    #Split the sequence into codons \n",
    "    TP53_codons = [TP53_exons[i:i+3] for i in range(0, len(TP53_exons), 3)]\n",
    "\n",
    "    #TP53 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    TP53_exons_extra_bases = 'CATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGACATAGTGTGGTGGTGCCCTATGAGCCGCCTGAGGTTGGCTCTGACTGTACCACCATCCACTACAACTACATGTGTAACAGTTCCTGCATGGGCGGCATGAACCGGAGGCCCATCCTCACCATCATCACACTGGAAGACTCCAGTGGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGAGGAAGAGAATCTCCGCAAGAAAGGGGAGCCTCACCACGAGCTGCCCCCAGGGAGCACTAAGCGAGCACTGCCCAACAACACCAGCTCCTCTCCCCAGCCAAAGAAGAAACCACTGGATGGAGAATATTTCACCCTTCAGATCCGTGGGCGTGAGCGCTTCGAGATGTTCCGAGAGCTGAATGAGGCCTTGGAACTCAAGGATGCCCAGGCTGGGAAGGAGCCAGGGGGGAGCAGGGCTCACTCCAGCCACCTGAAGTCCAAAAAGGGTCAGTCTACCTCCCGCCATAAAAAACTCATGTTCAAGACAGAAGGGCCTGACTCAGACTGAC'\n",
    "    \n",
    "    #Create a list of the codons they called missense variants at\n",
    "    import csv\n",
    "    filename = csv_file_variants\n",
    "    with open(filename, 'r') as csvfile:\n",
    "        read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "        row_count=0\n",
    "\n",
    "        TP53_missense_variant_list = []\n",
    "        TP53_missense_codons =[]\n",
    "        TP53_missense_variant_DICT = {}\n",
    "\n",
    "        for row in read_reader: #each row contains aa number (e.g. 11) and variant (e.g. p.D11A)\n",
    "    #         print(row)\n",
    "            if row_count>0: \n",
    "                codon0 = int(row[0])-1\n",
    "                codon1 = int(row[0])\n",
    "                variant = row[1]\n",
    "                TP53_missense_codons.append(codon0) #List of codon numbers that they called variants at\n",
    "                TP53_missense_variant_DICT[row_count]=(codon1, variant) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "            row_count=row_count+1\n",
    "            \n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    TP53_missense_codons_DICT = {}\n",
    "    calls_TP53_missense = {}\n",
    "\n",
    "    for i in TP53_missense_codons: #Iterates over the list of codon numbers\n",
    "        TP53_missense_codons_DICT[i+1]= (TP53_codons[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "        \n",
    "    for k,v in TP53_missense_codons_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_TP53_missense[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "        \n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_TP53_missense_extra = {}\n",
    "\n",
    "    for k, v in calls_TP53_missense.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_TP53_missense_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_TP53_missense_extra2 = {}\n",
    "\n",
    "    for k, v in calls_TP53_missense_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_TP53_missense_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "            \n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    TP53_position_in_list_dict_missense={}\n",
    "    for k, v in TP53_missense_variant_DICT.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_TP53_missense[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_TP53_missense[v[0]]\n",
    "        codon_seq=calls_TP53_missense[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_TP53_missense_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                TP53_position_in_list_dict_missense[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "            \n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in TP53_position_in_list_dict_missense.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "#     print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "                \n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TP53_bases_missense = {}\n",
    "    for k, v in TP53_position_in_list_dict_missense.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TP53_bases_missense[k]=(v, base_changes)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TP53_bases_missense2 = {}\n",
    "\n",
    "    for k, v in TP53_bases_missense.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TP53_bases_missense2[k] = (v, codon_positions)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TP53_bases_missense3 = {}\n",
    "\n",
    "    for k, v in TP53_bases_missense2.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TP53_bases_missense3[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TP53_bases_missense4 = {}\n",
    "\n",
    "    for k, v in TP53_bases_missense3.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_TP53:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check(i)\n",
    "                sites_trio_missense.append([adapted_trio])\n",
    "                TP53_bases_missense4[k] = (v, sites_trio_missense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_missense = [TP53_exons_extra_bases[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_missense.append(site_trio_missense)\n",
    "                    TP53_bases_missense4[k] = (v, sites_trio_missense)\n",
    "                    \n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TP53_sites_missense5 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TP53_bases_missense4.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        TP53_sites_missense5[a]=info\n",
    "        a=a+1\n",
    "        \n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TP53_sites_missense6={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TP53_sites_missense5.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "    \n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TP53_sites_missense6[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "        \n",
    "        \n",
    "    #Assign the mutation rates to each site...\n",
    "    TP53_mutation_rates_missense = {}\n",
    "    for k, v in TP53_sites_missense6.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            TP53_mutation_rates_missense[k] = (v, mutation_rate)\n",
    "\n",
    "    TP53_mutation_rate_missense_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TP53_mutation_rates_missense.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        TP53_mutation_rate_missense_sum.append(mutation_rate_value)\n",
    "        \n",
    "    TP53_Summed_mutation_rate_missense_added = np.sum(TP53_mutation_rate_missense_sum)\n",
    "\n",
    "    print('Total summed mutation rate across TP53 missense for this study = ', TP53_Summed_mutation_rate_missense_added)\n",
    "    \n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return TP53_Summed_mutation_rate_missense_added"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculating mutation rates from a list of variants"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TP53"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_nonsense_range_of_variants_TP53(lower_range, upper_range):\n",
    "    \n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base_TP53 = [74, 75, 96, 375, 672, 782, 919, 993, 1100]\n",
    "   \n",
    "    def intron_exon_check(position_in_gene_number):\n",
    "        if position_in_gene_number == 74:\n",
    "            gene_pos = 74\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 75:\n",
    "            gene_pos = 75\n",
    "            first_base = 'G'\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = TP53_exons_extra_bases[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 96:\n",
    "            gene_pos = 96\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 375:\n",
    "            gene_pos = 375\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 672:\n",
    "            gene_pos = 672\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 782:\n",
    "            gene_pos = 782\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 919:\n",
    "            gene_pos = 919\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 993:\n",
    "            gene_pos = 993\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1100:\n",
    "            gene_pos = 1100\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #TP53 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    TP53_exons = 'ATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGACATAGTGTGGTGGTGCCCTATGAGCCGCCTGAGGTTGGCTCTGACTGTACCACCATCCACTACAACTACATGTGTAACAGTTCCTGCATGGGCGGCATGAACCGGAGGCCCATCCTCACCATCATCACACTGGAAGACTCCAGTGGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGAGGAAGAGAATCTCCGCAAGAAAGGGGAGCCTCACCACGAGCTGCCCCCAGGGAGCACTAAGCGAGCACTGCCCAACAACACCAGCTCCTCTCCCCAGCCAAAGAAGAAACCACTGGATGGAGAATATTTCACCCTTCAGATCCGTGGGCGTGAGCGCTTCGAGATGTTCCGAGAGCTGAATGAGGCCTTGGAACTCAAGGATGCCCAGGCTGGGAAGGAGCCAGGGGGGAGCAGGGCTCACTCCAGCCACCTGAAGTCCAAAAAGGGTCAGTCTACCTCCCGCCATAAAAAACTCATGTTCAAGACAGAAGGGCCTGACTCAGACTGA'\n",
    "    \n",
    "    #Split the sequence into codons \n",
    "    TP53_codons = [TP53_exons[i:i+3] for i in range(0, len(TP53_exons), 3)]\n",
    "\n",
    "    #TP53 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    TP53_exons_extra_bases = 'CATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGACATAGTGTGGTGGTGCCCTATGAGCCGCCTGAGGTTGGCTCTGACTGTACCACCATCCACTACAACTACATGTGTAACAGTTCCTGCATGGGCGGCATGAACCGGAGGCCCATCCTCACCATCATCACACTGGAAGACTCCAGTGGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGAGGAAGAGAATCTCCGCAAGAAAGGGGAGCCTCACCACGAGCTGCCCCCAGGGAGCACTAAGCGAGCACTGCCCAACAACACCAGCTCCTCTCCCCAGCCAAAGAAGAAACCACTGGATGGAGAATATTTCACCCTTCAGATCCGTGGGCGTGAGCGCTTCGAGATGTTCCGAGAGCTGAATGAGGCCTTGGAACTCAAGGATGCCCAGGCTGGGAAGGAGCCAGGGGGGAGCAGGGCTCACTCCAGCCACCTGAAGTCCAAAAAGGGTCAGTCTACCTCCCGCCATAAAAAACTCATGTTCAAGACAGAAGGGCCTGACTCAGACTGAC'\n",
    "    \n",
    "    #List of all amino acid positions (across whole coding region) covered (amino acid position minus 1 (as python starts at 0))\n",
    "    TP53_nonsense_codons_DICT={}\n",
    "    for i in range(lower_range, upper_range):\n",
    "        TP53_nonsense_codons_DICT[i+1]= (TP53_codons[i])\n",
    "\n",
    "    calls_TP53_nonsense = {}\n",
    "    for k,v in TP53_nonsense_codons_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_TP53_nonsense[k]=(v, changes)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_TP53_nonsense_extra = {}\n",
    "\n",
    "    for k, v in calls_TP53_nonsense.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_TP53_nonsense_extra[k] = (Amino_acid_change, v[0], v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_TP53_nonsense_extra2 = {}\n",
    "\n",
    "    for k, v in calls_TP53_nonsense_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_TP53_nonsense_extra2[k] = (Amino_acid_abbrev, v[1], v[2])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary that contains details of the positions of nonsense variants at that codon\n",
    "    TP53_position_in_list_dict_nonsense = {}\n",
    "\n",
    "    for k, v in calls_TP53_nonsense_extra2.items():\n",
    "        original_AA = v[0]\n",
    "        possible_AA_changes = v[2]\n",
    "        possible_AA_changes_string = ''.join(v[2])\n",
    "        positions_in_list = []\n",
    "        for i in possible_AA_changes:\n",
    "            if i == '*': #nonsense\n",
    "                position_in_list=-1\n",
    "                nonsense_change = i    \n",
    "                while True:\n",
    "                    position_in_list = possible_AA_changes_string.find(nonsense_change, position_in_list+1)\n",
    "    #               print(position_in_list)\n",
    "                    if position_in_list == -1: break\n",
    "                    positions_in_list.append(position_in_list)\n",
    "                    positions_in_list = set(positions_in_list) #convert into a 'set' to remove duplicate positions\n",
    "                    positions_in_list = list(positions_in_list) #convert back to a list so can be added to dictionary\n",
    "                TP53_position_in_list_dict_nonsense[k]=(positions_in_list, v)\n",
    "\n",
    "        aa_positions_variants_called = []\n",
    "        for k, v in TP53_position_in_list_dict_nonsense.items():\n",
    "            aa_positions_variants_called.append(k)\n",
    "    # print('positions at which nonsense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TP53_bases_nonsense = {}\n",
    "    for k, v in TP53_position_in_list_dict_nonsense.items():\n",
    "        original_a_a = v[1][0]\n",
    "        codon_seq = v[1][1]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TP53_bases_nonsense[k]=(v, base_changes, original_a_a)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TP53_bases_nonsense2 = {}\n",
    "\n",
    "    for k, v in TP53_bases_nonsense.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TP53_bases_nonsense2[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TP53_bases_nonsense3 = {}\n",
    "\n",
    "    for k, v in TP53_bases_nonsense2.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_a_a = v[0][2]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = k\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(k)-1)*3)+codon_pos #codon number minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][1][1][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TP53_bases_nonsense3[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, original_a_a)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TP53_bases_nonsense4 = {}\n",
    "\n",
    "    for k, v in TP53_bases_nonsense3.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_nonsense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_TP53:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check_TP53(i)\n",
    "                sites_trio_nonsense.append([adapted_trio])\n",
    "                TP53_bases_nonsense4[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_nonsense = [TP53_exons_extra_bases[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_nonsense.append(site_trio_nonsense)\n",
    "                    TP53_bases_nonsense4[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TP53_sites_nonsense5 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TP53_bases_nonsense4.items():\n",
    "        original_a_a = v[0][0][0]\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, original_a_a])\n",
    "            change = change+1\n",
    "        TP53_sites_nonsense5[a]=info\n",
    "        a=a+1\n",
    "    #     print(TP53_sites_nonsense5)\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TP53_sites_nonsense6={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TP53_sites_nonsense5.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        original_a_a = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TP53_sites_nonsense6[k] = (v[0],changes3, codon_n_912, original_a_a) #site_trio = the affected base and its neighbours\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    TP53_mutation_rates_nonsense = {}\n",
    "    for k, v in TP53_sites_nonsense6.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            original_a_a = v[3]\n",
    "            aa_positioning = v[2]\n",
    "            variant_type = '*'\n",
    "            variant_id = str(original_a_a)+str(aa_positioning)+variant_type\n",
    "            TP53_mutation_rates_nonsense[k] = (v[0],v[1], mutation_rate, variant_id)\n",
    "\n",
    "    TP53_mutation_rate_nonsense_sum = []\n",
    "\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TP53_mutation_rates_nonsense.items():\n",
    "        codon_position_in_gene=(v[0])\n",
    "        variant = v[3]\n",
    "        site = v[1]\n",
    "        mutation_rate_value = v[2] \n",
    "        mutation_rates_list.append((v[0], variant, site, mutation_rate_value))\n",
    "        TP53_mutation_rate_nonsense_sum.append(mutation_rate_value)\n",
    "\n",
    "    #Get the average of all the mutation rates\n",
    "    TP53_mutation_rate_nonsense_sum = []\n",
    "    for k, v in TP53_mutation_rates_nonsense.items():\n",
    "        mut_rate = v[2]\n",
    "        TP53_mutation_rate_nonsense_sum.append(mut_rate)\n",
    "\n",
    "    TP53_Summed_mutation_rate_nonsense_added = np.sum(TP53_mutation_rate_nonsense_sum)\n",
    "\n",
    "    print('Total summed mutation rate across TP53 nonsense for this study = ', TP53_Summed_mutation_rate_nonsense_added)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return TP53_Summed_mutation_rate_nonsense_added\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_missense_range_of_variants_TP53(lower_range, upper_range):\n",
    "\n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base_TP53 = [74, 75, 96, 375, 672, 782, 919, 993, 1100]\n",
    "   \n",
    "    def intron_exon_check(position_in_gene_number):\n",
    "        if position_in_gene_number == 74:\n",
    "            gene_pos = 74\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 75:\n",
    "            gene_pos = 75\n",
    "            first_base = 'G'\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = TP53_exons_extra_bases[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 96:\n",
    "            gene_pos = 96\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 375:\n",
    "            gene_pos = 375\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 672:\n",
    "            gene_pos = 672\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 782:\n",
    "            gene_pos = 782\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 919:\n",
    "            gene_pos = 919\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 993:\n",
    "            gene_pos = 993\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1100:\n",
    "            gene_pos = 1100\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #TP53 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    TP53_exons = 'ATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGACATAGTGTGGTGGTGCCCTATGAGCCGCCTGAGGTTGGCTCTGACTGTACCACCATCCACTACAACTACATGTGTAACAGTTCCTGCATGGGCGGCATGAACCGGAGGCCCATCCTCACCATCATCACACTGGAAGACTCCAGTGGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGAGGAAGAGAATCTCCGCAAGAAAGGGGAGCCTCACCACGAGCTGCCCCCAGGGAGCACTAAGCGAGCACTGCCCAACAACACCAGCTCCTCTCCCCAGCCAAAGAAGAAACCACTGGATGGAGAATATTTCACCCTTCAGATCCGTGGGCGTGAGCGCTTCGAGATGTTCCGAGAGCTGAATGAGGCCTTGGAACTCAAGGATGCCCAGGCTGGGAAGGAGCCAGGGGGGAGCAGGGCTCACTCCAGCCACCTGAAGTCCAAAAAGGGTCAGTCTACCTCCCGCCATAAAAAACTCATGTTCAAGACAGAAGGGCCTGACTCAGACTGA'\n",
    "    \n",
    "    #Split the sequence into codons \n",
    "    TP53_codons = [TP53_exons[i:i+3] for i in range(0, len(TP53_exons), 3)]\n",
    "\n",
    "    #TP53 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    TP53_exons_extra_bases = 'CATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGACATAGTGTGGTGGTGCCCTATGAGCCGCCTGAGGTTGGCTCTGACTGTACCACCATCCACTACAACTACATGTGTAACAGTTCCTGCATGGGCGGCATGAACCGGAGGCCCATCCTCACCATCATCACACTGGAAGACTCCAGTGGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGAGGAAGAGAATCTCCGCAAGAAAGGGGAGCCTCACCACGAGCTGCCCCCAGGGAGCACTAAGCGAGCACTGCCCAACAACACCAGCTCCTCTCCCCAGCCAAAGAAGAAACCACTGGATGGAGAATATTTCACCCTTCAGATCCGTGGGCGTGAGCGCTTCGAGATGTTCCGAGAGCTGAATGAGGCCTTGGAACTCAAGGATGCCCAGGCTGGGAAGGAGCCAGGGGGGAGCAGGGCTCACTCCAGCCACCTGAAGTCCAAAAAGGGTCAGTCTACCTCCCGCCATAAAAAACTCATGTTCAAGACAGAAGGGCCTGACTCAGACTGAC'\n",
    "    \n",
    "    #List of all amino acid positions (across whole coding region) covered (amino acid position minus 1 (as python starts at 0))\n",
    "    TP53_missense_codons_DICT={}\n",
    "    for i in range(lower_range, upper_range):\n",
    "        TP53_missense_codons_DICT[i+1]= (TP53_codons[i])\n",
    "\n",
    "    calls_TP53_missense = {}\n",
    "    for k,v in TP53_missense_codons_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_TP53_missense[k]=(v, changes)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_TP53_missense_extra = {}\n",
    "\n",
    "    for k, v in calls_TP53_missense.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_TP53_missense_extra[k] = (Amino_acid_change, v[0], v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_TP53_missense_extra2 = {}\n",
    "\n",
    "    for k, v in calls_TP53_missense_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_TP53_missense_extra2[k] = (Amino_acid_abbrev, v[1], v[2])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary that contains details of the positions of missense variants at that codon\n",
    "    TP53_position_in_list_dict_missense = {}\n",
    "\n",
    "    for k, v in calls_TP53_missense_extra2.items():\n",
    "        original_AA = v[0]\n",
    "        possible_AA_changes = v[2]\n",
    "        possible_AA_changes_string = ''.join(v[2])\n",
    "        positions_in_list = []\n",
    "        for i in possible_AA_changes:\n",
    "            if i != original_AA: #not synonymous\n",
    "                if i != '*': #not missense\n",
    "                    position_in_list=-1\n",
    "                    missense_change = i\n",
    "                    new_aa_list = []\n",
    "                    while True:\n",
    "                        position_in_list = possible_AA_changes_string.find(missense_change, position_in_list+1)\n",
    "        #               print(position_in_list)\n",
    "                        if position_in_list == -1: break\n",
    "                        positions_in_list.append(position_in_list)\n",
    "                        positions_in_list = set(positions_in_list) #convert into a 'set' to remove duplicate positions\n",
    "                        positions_in_list = list(positions_in_list) #convert back to a list so can be added to dictionary\n",
    "                    for pos in positions_in_list:\n",
    "                        new_aa = possible_AA_changes_string[pos]\n",
    "                        new_aa_list.append(new_aa)\n",
    "                    TP53_position_in_list_dict_missense[k]=(positions_in_list, v, new_aa_list)\n",
    "\n",
    "        aa_positions_variants_called = []\n",
    "        for k, v in TP53_position_in_list_dict_missense.items():\n",
    "            aa_positions_variants_called.append(k)\n",
    "    # print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TP53_bases_missense = {}\n",
    "    for k, v in TP53_position_in_list_dict_missense.items():\n",
    "        original_a_a = v[1][0]\n",
    "        codon_seq = v[1][1]\n",
    "        base_changes=[]\n",
    "        new_a_a_list = v[2]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TP53_bases_missense[k]=(v, base_changes, original_a_a, new_a_a_list)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TP53_bases_missense2 = {}\n",
    "\n",
    "    for k, v in TP53_bases_missense.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        new_a_a_list = v[3]\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TP53_bases_missense2[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TP53_bases_missense3 = {}\n",
    "\n",
    "    for k, v in TP53_bases_missense2.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_a_a = v[0][2]\n",
    "        new_aa_list = v[0][0][2]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = k\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(k)-1)*3)+codon_pos #codon number minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][1][1][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TP53_bases_missense3[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, original_a_a, new_aa_list)\n",
    "\n",
    "    # Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TP53_bases_missense4 = {}\n",
    "\n",
    "    for k, v in TP53_bases_missense3.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "        new_a_a_list = v[6]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_TP53:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check(i)\n",
    "                sites_trio_missense.append([adapted_trio])\n",
    "                TP53_bases_missense4[k] = (v, sites_trio_missense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_missense = [TP53_exons_extra_bases[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_missense.append(site_trio_missense)\n",
    "                    TP53_bases_missense4[k] = (v, sites_trio_missense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TP53_sites_missense5 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TP53_bases_missense4.items():\n",
    "        original_a_a = v[0][0][0]\n",
    "        new_a_a = v[0][6]\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            new_amino_acid = new_a_a[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, original_a_a, new_amino_acid])\n",
    "            change = change+1\n",
    "        TP53_sites_missense5[a]=info\n",
    "        a=a+1\n",
    "    #     print(TP53_sites_missense5)\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TP53_sites_missense6={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TP53_sites_missense5.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        original_a_a = v[5]\n",
    "        new_a_a = v[6]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TP53_sites_missense6[k] = (v[0],changes3, codon_n_912, original_a_a, new_a_a) #site_trio = the affected base and its neighbours\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    TP53_mutation_rates_missense = {}\n",
    "    for k, v in TP53_sites_missense6.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            original_a_a = v[3]\n",
    "            new_a_a = v[4]\n",
    "            aa_positioning = v[2]\n",
    "            variant_id = str(original_a_a)+str(aa_positioning)+str(new_a_a)\n",
    "            TP53_mutation_rates_missense[k] = (v[0],v[1], mutation_rate, variant_id)\n",
    "\n",
    "    TP53_mutation_rate_missense_sum = []\n",
    "\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TP53_mutation_rates_missense.items():\n",
    "        codon_position_in_gene=(v[0])\n",
    "        variant = v[3]\n",
    "        site = v[1]\n",
    "        mutation_rate_value = v[2] \n",
    "        mutation_rates_list.append((v[0], variant, site, mutation_rate_value))\n",
    "        TP53_mutation_rate_missense_sum.append(mutation_rate_value)\n",
    "\n",
    "    #Get the average of all the mutation rates\n",
    "    TP53_mutation_rate_missense_sum = []\n",
    "    for k, v in TP53_mutation_rates_missense.items():\n",
    "        mut_rate = v[2]\n",
    "        TP53_mutation_rate_missense_sum.append(mut_rate)\n",
    "\n",
    "    TP53_Summed_mutation_rate_missense_added = np.sum(TP53_mutation_rate_missense_sum)\n",
    "\n",
    "    print('Total summed mutation rate across TP53 missense for this study = ', TP53_Summed_mutation_rate_missense_added)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return TP53_Summed_mutation_rate_missense_added\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Mutation rate for DBD and non-DBD for each study"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- In order to control for differences in panel coverage/ mutation rate inside and outside the DNA binding domains, we need to determine which sites in TP53 each study included (and what the mutation rate across those sites was).\n",
    "- This can be calculated by using the mutation rate functions (above) as they will also produce a list of the variants within e.g. range of amino acid positions of interest or from a CSV file of variants."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Missense DBD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 missense for this study =  5.975149040000001e-07\n"
     ]
    }
   ],
   "source": [
    "Jaiswal_missense_DBD_mutation_rate = mutation_rates_list_missense_variants_TP53('Data files/Jaiswal TP53 missense variants called DBD.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 missense for this study =  3.9562103600000004e-07\n"
     ]
    }
   ],
   "source": [
    "Genovese_missense_DBD_mutation_rate = mutation_rates_list_missense_variants_TP53('Data files/Genovese TP53 missense variants called DBD.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 missense for this study =  1.242931322e-06\n"
     ]
    }
   ],
   "source": [
    "Zink_missense_DBD_mutation_rate = mutation_rates_list_missense_variants_TP53('Data files/Zink TP53 missense variants called DBD.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 missense for this study =  1.790603677e-06\n"
     ]
    }
   ],
   "source": [
    "Coombs_missense_DBD_mutation_rate = mutation_rates_missense_range_of_variants_TP53(99, 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 missense for this study =  1.790603677e-06\n"
     ]
    }
   ],
   "source": [
    "Young_missense_DBD_mutation_rate = mutation_rates_missense_range_of_variants_TP53(99, 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 missense for this study =  1.790603677e-06\n"
     ]
    }
   ],
   "source": [
    "Desai_missense_DBD_mutation_rate = mutation_rates_missense_range_of_variants_TP53(99, 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 missense for this study =  2.75197872e-07\n",
      "Total summed mutation rate across TP53 missense for this study =  2.0262653100000002e-07\n",
      "Total summed mutation rate across TP53 missense for this study =  1.9660014099999998e-07\n"
     ]
    }
   ],
   "source": [
    "Acuna_missense_DBD_mutation_rate = mutation_rates_missense_range_of_variants_TP53(185, 221) + \\\n",
    "mutation_rates_missense_range_of_variants_TP53(168, 185) + mutation_rates_missense_range_of_variants_TP53(117, 146)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Missense non-DBD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 missense for this study =  2.8897357000000005e-08\n"
     ]
    }
   ],
   "source": [
    "Jaiswal_missense_nonDBD_mutation_rate = mutation_rates_list_missense_variants_TP53('Data files/Jaiswal TP53 missense variants called nonDBD.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 missense for this study =  1.9851174e-08\n"
     ]
    }
   ],
   "source": [
    "Genovese_missense_nonDBD_mutation_rate = mutation_rates_list_missense_variants_TP53('Data files/Genovese TP53 missense variants called nonDBD.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 missense for this study =  2.2131825300000001e-07\n"
     ]
    }
   ],
   "source": [
    "Zink_missense_nonDBD_mutation_rate = mutation_rates_list_missense_variants_TP53('Data files/Zink TP53 missense variants called nonDBD.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 missense for this study =  7.756362789999999e-07\n",
      "Total summed mutation rate across TP53 missense for this study =  7.07481501e-07\n"
     ]
    }
   ],
   "source": [
    "Coombs_missense_nonDBD_mutation_rate = mutation_rates_missense_range_of_variants_TP53(0, 99)+mutation_rates_missense_range_of_variants_TP53(300, 394)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 missense for this study =  7.756362789999999e-07\n",
      "Total summed mutation rate across TP53 missense for this study =  7.07481501e-07\n",
      "Total summed mutation rate across TP53 missense for this study =  1.27013e-09\n"
     ]
    }
   ],
   "source": [
    "Young_missense_nonDBD_mu = mutation_rates_missense_range_of_variants_TP53(0, 99)+mutation_rates_missense_range_of_variants_TP53(300, 394)\n",
    "Young_missense_nonDBD_excluded_regions = Young_TP53_excluded = mutation_rates_list_missense_variants_TP53('Data files/excluded_variants_Young_TP53.csv')\n",
    "Young_missense_nonDBD_mutation_rate = Young_missense_nonDBD_mu - Young_missense_nonDBD_excluded_regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 missense for this study =  7.756362789999999e-07\n",
      "Total summed mutation rate across TP53 missense for this study =  7.07481501e-07\n",
      "Total summed mutation rate across TP53 missense for this study =  4.730426e-09\n"
     ]
    }
   ],
   "source": [
    "Desai_missense_nonDBD_mu = mutation_rates_missense_range_of_variants_TP53(0, 99)+mutation_rates_missense_range_of_variants_TP53(300, 394)\n",
    "Desai_missense_nonDBD_excluded = mutation_rates_list_missense_variants_TP53('Data files/excluded_variants_Desai_2018_TP53.csv')\n",
    "Desai_missense_nonDBD_mutation_rate = Desai_missense_nonDBD_mu - Desai_missense_nonDBD_excluded"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 missense for this study =  1.9621064499999997e-07\n",
      "mutation rate for specific variant =  4.55944e-10\n"
     ]
    }
   ],
   "source": [
    "Acuna_missense_nonDBD_mu = mutation_rates_missense_range_of_variants_TP53(59, 81)\n",
    "Acuna_missense_nonDBD_excluded = mutation_rates_variants_TP53('p.P72R')\n",
    "Acuna_missense_nonDBD_mutation_rate = Acuna_missense_nonDBD_mu - Acuna_missense_nonDBD_excluded"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Nonsense DBD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 nonsense for this study =  7.948346000000001e-08\n"
     ]
    }
   ],
   "source": [
    "Jaiswal_nonsense_DBD_mutation_rate = mutation_rates_nonsense_range_of_variants_TP53(99, 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 nonsense for this study =  3.5593168e-08\n"
     ]
    }
   ],
   "source": [
    "Genovese_nonsense_DBD_mutation_rate = mutation_rates_list_nonsense_variants_TP53('Data files/Genovese TP53 nonsense variants called DBD.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 nonsense for this study =  7.948346000000001e-08\n"
     ]
    }
   ],
   "source": [
    "Zink_nonsense_DBD_mutation_rate = mutation_rates_nonsense_range_of_variants_TP53(99, 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 nonsense for this study =  7.948346000000001e-08\n"
     ]
    }
   ],
   "source": [
    "Coombs_nonsense_DBD_mutation_rate = mutation_rates_nonsense_range_of_variants_TP53(99, 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 nonsense for this study =  7.948346000000001e-08\n"
     ]
    }
   ],
   "source": [
    "Young_nonsense_DBD_mutation_rate = mutation_rates_nonsense_range_of_variants_TP53(99, 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 nonsense for this study =  7.948346000000001e-08\n"
     ]
    }
   ],
   "source": [
    "Desai_nonsense_DBD_mutation_rate = mutation_rates_nonsense_range_of_variants_TP53(99, 300)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 nonsense for this study =  3.2044719e-08\n",
      "Total summed mutation rate across TP53 nonsense for this study =  3.3097040000000003e-09\n",
      "Total summed mutation rate across TP53 nonsense for this study =  1.3440674e-08\n"
     ]
    }
   ],
   "source": [
    "Acuna_nonsense_DBD_mutation_rate = mutation_rates_nonsense_range_of_variants_TP53(185, 221) + \\\n",
    "mutation_rates_nonsense_range_of_variants_TP53(168, 185) + \\\n",
    "mutation_rates_nonsense_range_of_variants_TP53(117, 146)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Nonsense non-DBD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 nonsense for this study =  3.1374494e-08\n",
      "Total summed mutation rate across TP53 nonsense for this study =  5.0620516e-08\n"
     ]
    }
   ],
   "source": [
    "Jaiswal_nonsense_nonDBD_mutation_rate = mutation_rates_nonsense_range_of_variants_TP53(0, 99)+mutation_rates_nonsense_range_of_variants_TP53(300, 394)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 nonsense for this study =  4.356006999999999e-08\n"
     ]
    }
   ],
   "source": [
    "Genovese_nonsense_nonDBD_mutation_rate = mutation_rates_list_nonsense_variants_TP53('Data files/Genovese TP53 nonsense variants called nonDBD.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 nonsense for this study =  3.1374494e-08\n",
      "Total summed mutation rate across TP53 nonsense for this study =  5.0620516e-08\n"
     ]
    }
   ],
   "source": [
    "Zink_nonsense_nonDBD_mutation_rate = mutation_rates_nonsense_range_of_variants_TP53(0, 99)+mutation_rates_nonsense_range_of_variants_TP53(300, 394)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 nonsense for this study =  3.1374494e-08\n",
      "Total summed mutation rate across TP53 nonsense for this study =  5.0620516e-08\n"
     ]
    }
   ],
   "source": [
    "Coombs_nonsense_nonDBD_mutation_rate = mutation_rates_nonsense_range_of_variants_TP53(0, 99)+mutation_rates_nonsense_range_of_variants_TP53(300, 394)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 nonsense for this study =  3.1374494e-08\n",
      "Total summed mutation rate across TP53 nonsense for this study =  5.0620516e-08\n"
     ]
    }
   ],
   "source": [
    "Young_nonsense_nonDBD_mutation_rate = mutation_rates_nonsense_range_of_variants_TP53(0, 99)+mutation_rates_nonsense_range_of_variants_TP53(300, 394)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 nonsense for this study =  3.1374494e-08\n",
      "Total summed mutation rate across TP53 nonsense for this study =  5.0620516e-08\n"
     ]
    }
   ],
   "source": [
    "Desai_nonsense_nonDBD_mutation_rate = mutation_rates_nonsense_range_of_variants_TP53(0, 99)+mutation_rates_nonsense_range_of_variants_TP53(300, 394)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 nonsense for this study =  1.485146e-09\n"
     ]
    }
   ],
   "source": [
    "Acuna_nonsense_nonDBD_mutation_rate = mutation_rates_nonsense_range_of_variants_TP53(59, 81)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Actual number of observations of TP53 variants (inside vs outside DNA binding domain)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Look at each variant seen across the studies and count how many times it was observed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def TP53_variants_by_study(study_name):\n",
    "\n",
    "    filename = \"Data files/all_studies_trimmed_all_genes.csv\"\n",
    "    with open(filename, 'r') as csvfile:\n",
    "        read_reader = csv.reader(csvfile)\n",
    "        row_count = 0\n",
    "        missense_TP53_DBD = []\n",
    "        missense_TP53_nonDBD = []\n",
    "        nonsense_TP53_DBD = []\n",
    "        nonsense_TP53_nonDBD = []\n",
    "\n",
    "        for row in read_reader:\n",
    "#             print(row)\n",
    "            if row_count > 0:\n",
    "                VAF = float(row[0])\n",
    "                age = row[1]\n",
    "                variant = row[2]\n",
    "                codon_number = int(re.search(r'\\d+', variant).group(0)) #extract the codon number\n",
    "                gene = row[3]\n",
    "                study = row[4]\n",
    "                if study == study_name:\n",
    "                    if gene == 'TP53':\n",
    "                        if 100 <= codon_number <= 300:\n",
    "                            if variant[-1]=='*': #If the final part of the variant is * (i.e. nonsense)\n",
    "                                nonsense_TP53_DBD.append(variant)\n",
    "                            else:\n",
    "                                missense_TP53_DBD.append(variant)\n",
    "                        else:\n",
    "                            if variant[-1]=='*': #If the final part of the variant is * (i.e. nonsense)\n",
    "                                nonsense_TP53_nonDBD.append(variant)\n",
    "                            else:\n",
    "                                missense_TP53_nonDBD.append(variant)\n",
    "            row_count = row_count+1\n",
    "            \n",
    "    return missense_TP53_DBD, nonsense_TP53_DBD, missense_TP53_nonDBD, nonsense_TP53_nonDBD"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "Jaiswal_miss_DBD, Jaiswal_nonsense_DBD, Jaiswal_miss_nonDBD, Jaiswal_nonsense_nonDBD = TP53_variants_by_study('Jaiswal2014')\n",
    "Genovese_miss_DBD, Genovese_nonsense_DBD, Genovese_miss_nonDBD, Genovese_nonsense_nonDBD = TP53_variants_by_study('Genovese2014')\n",
    "Zink_miss_DBD, Zink_nonsense_DBD, Zink_miss_nonDBD, Zink_nonsense_nonDBD = TP53_variants_by_study('ZinkWGS')\n",
    "Acuna_miss_DBD, Acuna_nonsense_DBD, Acuna_miss_nonDBD, Acuna_nonsense_nonDBD = TP53_variants_by_study('Acuna2017')\n",
    "Coombs_miss_DBD, Coombs_nonsense_DBD, Coombs_miss_nonDBD, Coombs_nonsense_nonDBD = TP53_variants_by_study('Coombs2017')\n",
    "#Desai_miss_DBD, Desai_nonsense_DBD, Desai_miss_nonDBD, Desai_nonsense_nonDBD = TP53_variants_by_study('Desai2018')\n",
    "Young2016_miss_DBD, Young2016_nonsense_DBD, Young2016_miss_nonDBD, Young2016_nonsense_nonDBD = TP53_variants_by_study('Young2016')\n",
    "Young2019_miss_DBD, Young2019_nonsense_DBD, Young2019_miss_nonDBD, Young2019_nonsense_nonDBD = TP53_variants_by_study('Young2019')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plotting the relative frequencies of mutations (controlled for mutation rates)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- If the number of missense variants observed in the DBD is Poisson distributed with mean 5 and the number of missense variants outside the DBD is poisson distributed with proportional mean of 5 x (missense non DBD mutation rate/ missense DBD mutation rate) (i.e. null hypothesis is that there is no difference in the number of observations between the DBD and non-DBD), what is the probability that we only observe 1 variant outside the DBD?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def error_bars(observed_variants, mutation_rate):\n",
    "    \n",
    "    errors={}\n",
    "    n=0\n",
    "    for i, j in zip(observed_variants, mutation_rate):\n",
    "        normalised_hist = i/j\n",
    "        sqrt_hist = math.sqrt(i)\n",
    "        upper_error = sqrt_hist\n",
    "        lower_error = sqrt_hist\n",
    "        normalised_upper_error = upper_error/j\n",
    "        normalised_lower_error = lower_error/j\n",
    "        errors[n]=(normalised_hist, normalised_upper_error, normalised_lower_error)\n",
    "        n = n+1\n",
    "        \n",
    "    errors_corrected ={}\n",
    "    for k, v in errors.items():\n",
    "        binheight = v[0]\n",
    "        upper_error = v[1]\n",
    "        lower_error = v[2]\n",
    "        corrected_upper_error = (upper_error+binheight)-binheight\n",
    "        corrected_lower_error = binheight-(binheight-lower_error)\n",
    "        errors_corrected[k] = (binheight, upper_error, lower_error)\n",
    "\n",
    "    lower_err=[]\n",
    "    upper_err=[]\n",
    "    for k, v in errors_corrected.items():\n",
    "        lower_error = v[2]\n",
    "        upper_error = v[1]\n",
    "        lower_err.append(lower_error)\n",
    "        upper_err.append(upper_error)\n",
    "\n",
    "    err = [tuple(lower_err),tuple(upper_err)]\n",
    "    \n",
    "    return err"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "def plotting_frequencies(study_name, observations, mutation_rates, ax):\n",
    "\n",
    "    #colours\n",
    "    missense = c6\n",
    "    nonsense = c10\n",
    "    DBD_missense = missense\n",
    "    nonDBD_missense = missense\n",
    "    DBD_nonsense = nonsense\n",
    "    nonDBD_nonsense = nonsense\n",
    "\n",
    "    axisfont=20\n",
    "    domainfont = 12\n",
    "    titlefont=12\n",
    "    axislabelfont=12\n",
    "\n",
    "    scale = 1\n",
    "    m_size = 18*scale\n",
    "    m_width = 1.5*scale\n",
    "    c_size = 5\n",
    "    c_thick = 1.5\n",
    "    e_width = 1.5\n",
    "    \n",
    "    missense_DBD_number = float(observations[0])\n",
    "    missense_nonDBD_number = float(observations[1])\n",
    "    nonsense_DBD_number = float(observations[2])\n",
    "    nonsense_nonDBD_number = float(observations[3])\n",
    "    \n",
    "    missense_DBD_mu = float(mutation_rates[0])\n",
    "    missense_nonDBD_mu = float(mutation_rates[1])\n",
    "    nonsense_DBD_mu = float(mutation_rates[2])\n",
    "    nonsense_nonDBD_mu = float(mutation_rates[3])\n",
    "    \n",
    "    missense_DBD_freq = missense_DBD_number/missense_DBD_mu\n",
    "    missense_nonDBD_freq = missense_nonDBD_number/missense_nonDBD_mu\n",
    "    nonsense_DBD_freq = nonsense_DBD_number/nonsense_DBD_mu\n",
    "    nonsense_nonDBD_freq = nonsense_nonDBD_number/nonsense_nonDBD_mu\n",
    "\n",
    "    colors_dict = {'missense_DBD': DBD_missense, 'missense_nonDBD': nonDBD_missense, 'nonsense_DBD': DBD_nonsense, 'nonsense_nonDBD': nonDBD_nonsense}\n",
    "\n",
    "    x = ['missense_DBD', 'missense_nonDBD', 'nonsense_DBD', 'nonsense_nonDBD']\n",
    "    y = [missense_DBD_freq, missense_nonDBD_freq, nonsense_DBD_freq, nonsense_nonDBD_freq]\n",
    "\n",
    "    ax.bar(x, y, color = [DBD_missense, nonDBD_missense, DBD_nonsense, nonDBD_nonsense])\n",
    "\n",
    "    errors = error_bars([missense_DBD_number, missense_nonDBD_number, nonsense_DBD_number, nonsense_nonDBD_number], \\\n",
    "                        [missense_DBD_mu, missense_nonDBD_mu, nonsense_DBD_mu, nonsense_nonDBD_mu])\n",
    "\n",
    "    ax.errorbar(x, y, yerr= errors, ecolor = grey3, fmt = 'none', \\\n",
    "             elinewidth = e_width, capsize = c_size, capthick = c_thick, zorder=3)\n",
    "\n",
    "    #Only show the required axis lines\n",
    "    ax.spines['top'].set_visible(False)\n",
    "    ax.spines['right'].set_visible(False)\n",
    "\n",
    "    for axis in ['bottom','left']:\n",
    "        ax.spines[axis].set_linewidth(2)\n",
    "\n",
    "    for axis in ['bottom','left']:\n",
    "        ax.spines[axis].set_color(grey3)\n",
    "\n",
    "    ax.set_ylabel('number of observations / $\\u03BC$', fontsize = axisfont, labelpad = 10)    \n",
    "    ax.set_xlabel('Variant type and TP53 domain', fontsize = axisfont, labelpad = 10)\n",
    "    ax.set_title(study_name, fontsize = axisfont, fontweight = 'bold')\n",
    "\n",
    "    # #x-axis ticks\n",
    "    x1_major_ticks = ['missense_DBD', 'missense_nonDBD', 'nonsense_DBD', 'nonsense_nonDBD']\n",
    "    x1_major_tick_labels = ['Miss DBD', 'Miss non-DBD', 'Nons DBD', 'Nons non-DBD']\n",
    "    ax.set_xticks(x1_major_ticks)\n",
    "    ax.set_xticklabels(x1_major_tick_labels, fontsize = axisfont-5)\n",
    "    \n",
    "    #labels\n",
    "    labels = [\"n = %d\" % i for i in observations]\n",
    "\n",
    "    rects = ax.patches\n",
    "    \n",
    "    for rect, label in zip(rects, labels):\n",
    "        height = rect.get_height()\n",
    "        ax.text(rect.get_x() + rect.get_width() / 2, height*1.1, label,\n",
    "                ha='center', va='bottom', fontsize = axisfont-5)\n",
    "    \n",
    "    print()\n",
    "    \n",
    "    #seeing if difference between missense DBD and missense non-DBD is significant:\n",
    "    nonDBD_mean_null = missense_DBD_number*(missense_nonDBD_mu/missense_DBD_mu)\n",
    "    p_value = poisson.cdf(missense_nonDBD_number, nonDBD_mean_null)\n",
    "    print('Poisson exact test p-value missense DBD vs missense non-DBD '+str(study_name)+'= ', p_value)\n",
    "    print(\"\")    \n",
    "    \n",
    "    return p_value, ax"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "Jaiswal_observations = [len(Jaiswal_miss_DBD), len(Jaiswal_miss_nonDBD), len(Jaiswal_nonsense_DBD), len(Jaiswal_nonsense_nonDBD)]\n",
    "Jaiswal_mutation_rates = [Jaiswal_missense_DBD_mutation_rate, Jaiswal_missense_nonDBD_mutation_rate, \\\n",
    "                          Jaiswal_nonsense_DBD_mutation_rate, Jaiswal_nonsense_nonDBD_mutation_rate]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "Genovese_observations = [len(Genovese_miss_DBD), len(Genovese_miss_nonDBD), len(Genovese_nonsense_DBD), len(Genovese_nonsense_nonDBD)]\n",
    "Genovese_mutation_rates = [Genovese_missense_DBD_mutation_rate, Genovese_missense_nonDBD_mutation_rate, \\\n",
    "                          Genovese_nonsense_DBD_mutation_rate, Genovese_nonsense_nonDBD_mutation_rate]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "Zink_observations = [len(Zink_miss_DBD), len(Zink_miss_nonDBD), len(Zink_nonsense_DBD), len(Zink_nonsense_nonDBD)]\n",
    "Zink_mutation_rates = [Zink_missense_DBD_mutation_rate, Zink_missense_nonDBD_mutation_rate, \\\n",
    "                          Zink_nonsense_DBD_mutation_rate, Zink_nonsense_nonDBD_mutation_rate]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "Acuna_observations = [len(Acuna_miss_DBD), len(Acuna_miss_nonDBD), len(Acuna_nonsense_DBD), len(Acuna_nonsense_nonDBD)]\n",
    "Acuna_mutation_rates = [Acuna_missense_DBD_mutation_rate, Acuna_missense_nonDBD_mutation_rate, \\\n",
    "                          Acuna_nonsense_DBD_mutation_rate, Acuna_nonsense_nonDBD_mutation_rate]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "Coombs_observations = [len(Coombs_miss_DBD), len(Coombs_miss_nonDBD), len(Coombs_nonsense_DBD), len(Coombs_nonsense_nonDBD)]\n",
    "Coombs_mutation_rates = [Coombs_missense_DBD_mutation_rate, Coombs_missense_nonDBD_mutation_rate, \\\n",
    "                          Coombs_nonsense_DBD_mutation_rate, Coombs_nonsense_nonDBD_mutation_rate]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [],
   "source": [
    "Young_observations = [len(Young2016_miss_DBD)+len(Young2019_miss_DBD), len(Young2016_miss_nonDBD)+len(Young2019_miss_nonDBD),\\\n",
    "                      len(Young2016_nonsense_DBD)+len(Young2019_nonsense_DBD), len(Young2016_nonsense_nonDBD)+len(Young2019_nonsense_nonDBD)]\n",
    "Young_mutation_rates = [Young_missense_DBD_mutation_rate, Young_missense_nonDBD_mutation_rate, \\\n",
    "                          Young_nonsense_DBD_mutation_rate, Young_nonsense_nonDBD_mutation_rate]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Poisson exact test p-value missense DBD vs missense non-DBD Jaiswal 2014=  0.313266283136582\n",
      "\n",
      "\n",
      "Poisson exact test p-value missense DBD vs missense non-DBD Genovese 2014=  0.8181504921777598\n",
      "\n",
      "\n",
      "Poisson exact test p-value missense DBD vs missense non-DBD Zink 2017=  0.5241063861117612\n",
      "\n",
      "\n",
      "Poisson exact test p-value missense DBD vs missense non-DBD Coombs 2017=  0.0002528331369878653\n",
      "\n",
      "\n",
      "Poisson exact test p-value missense DBD vs missense non-DBD Acuna-Hidalgo 2017=  0.1752526756219285\n",
      "\n",
      "\n",
      "Poisson exact test p-value missense DBD vs missense non-DBD Young 2016 & 2019=  0.015957222049514833\n",
      "\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABwAAAANYCAYAAADDhfmWAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAIABJREFUeJzs3XmcHWWV+P/PISRsjgnDIo4gjRhAdjAuo0EQFERAwHFAUIdW+OKCCy7AKPwUcaLgMiIgjOBAFDTguIGgIiqLomiIsqpBwRCDQIysKoEs5/dHVSfVN92d7tzuW/fe/rxfr3rlVtVTVed2on14zlPPE5mJJEmSJEmSJEmSpO6wVt0BSJIkSZIkSZIkSRo9FgAlSZIkSZIkSZKkLmIBUJIkSZIkSZIkSeoiFgAlSZIkSZIkSZKkLmIBUJIkSZIkSZIkSeoiFgAlSZIkSZIkSZKkLmIBUFJbiIjeiMjK1lN3TGsqImZWvse8uuORJEmSJEnS6ImIeZW+n5ntek9J45sFQElrJCJ6Ggp2M+uOqdNFxL9GxIcj4pqIuCsiHo2IJyPiTxHxjYjYbzXX7x8RV0TEAxHxVPnnFRGx/yDtp0TEiRFxafm85ZW/z+tGEPfWEfG3hn8PvSP79pIkdbaIeHpEHFf+7r23/N34RNmR84uI+HxEHBgR69cdq4YWEftGxMcj4rqIuLv8u1wcEX+MiEsi4sVDXBsRcWSZzy0qc7n7IuKywa6LiM0j4pSI+Gb5b2eNcuyIeElELGu4fq+R/wQkSRofGgpuw9l66455JMr+mveUfUp3RMRfImJJmaNcGxHHRsTaQ1z/vIi4qPw5LY6IhyPipxHxjoiYOMg1x0TEFyPiV2UetOLnN8RzRvL30NP8T0YaPwb9H7gktdhs4ITK/kN1BVKjC4AdBji+ebm9NiLOzMz3NjaIiLOAdzUcfgZwEHBQRJydme9uON8DnNFMwBGxFjAT2KCZ+0iS1Mki4vXAucCGA5zestxeCLyDIt/5dOui0xr4JgPnNj3ldmREvC8zz6yeLDvCLgVe23DdvwCHAYdFxAcy8zMN56cBH2sm4LKw/CUc5CtJ0liqFrFmAJPLz3fUEMtwvBc4boDjGwF7ldt/RMS+mfmPaoMyv/0yUC30rQO8tNyOjIhXZ+ZjDff+NCt/LmNh0EKipFVZAJTUFjLzTuDOuuNoE3OA64FHgBdQFPH6HB8RV2TmtX0HIuJo+hf/flhevyfwivLYuyLilsy8sOFZT1Ikqr8q2241wljfD0wf4TWSJHWNiHg7RfGv6tfAtcBfKApJzwP2ADZtbXRqQgI/BX4GPMHKTjKAAD4dEVdl5u8r13yY/sW/bwC3AK+hyOkAPhURt2bmDxue93fgNoqc7EgGLiYP5VPAc0d4jSRJ4121iNdoX+CVlf17gCv7djLzgjGMa7TdC1wNzKfo93kDsG557qXAScBH+hpHxPYUg737in9zga9SDGrrpRhw9FLgHOA/Gp61DLiLIqd5JkXf1OoM9vewFsUgqUnl/h3ld5A0XJnp5ubmNuKNYvRzVraZlXOHUIwSuhV4gKLI9A/gbuAS4EUD3K+34X49lXPrUiQjv6Aoii2leENwLvB/wImVtsdV7vF3YGLl3P9Uzp1VOT6xbNt37tjKuROAb5XPWgQsAR6n6KD5b2DzAb7LzMq95o3gZ/pJYLcBjn+44Wfz6cq5tSgSub5zNwFRnotyf0UswFoN37v687mu0va6YcS7A7C4bP+Nhhh76/436ubm5ubmNtYbMBV4qvL770ng8EHargXsA+w9wLkNgPdRFJweKu/5IHA58MoB2u/V8Ht3L4o3zH5e5jSPAt8Bthsklk0pOlPmlG2fAv5cPu+ghrZHV56zDHjWAPe7rdLmGw3nngecB/yujO2JMq/67CD3GnbeV7kmyu9/JXB/+X0eAW4AjgXWHuHf63nAcwc4fmHDz/2dlXOTgb9Vzl3W8J2q+dpPGu67TkOONq/SduYw4n0FsHyQnGyvuv934ubm5ubm1mkbsEvD7/XHgR0b2gz4+5pV+8t6KXLAHwGPlfnQtcCLB3juYPecTP/+nUeBlw3je7wDOLiaZ5THX9YQ480N5y+pnPsbsEnl3Ccr55YDUxuuXb/y+dTqc9bg7+HfG+J8Y93/NtzcOm1zehBJY+GNwJuAnSmmoZwErAc8h2KU0c8i4g0juN8VwOkUU2dNBiZQjIreBngd/aex/FHl8/rA8yv7L6t8ro5AekHZdqB7nERR0NyGYoqEtYGnATtRTKVwWzkyqmmZeWJm/nqAU99s2F+n8vn5wLMr+/+XmVneL4GvV85tSeXnkZlLMnPJmsRaTnH15TKW3wAnr8l9JEnqcO+h/7RIH8nMywZqmJnLM/NHmfnj6vGIeA7FG4OfoRhJvWF5z00p3hz7QUSsbsru04DLgBdT5DRPBw4EboiIjRue9yKK0dOnALuXbSdSjNB+DXBFRHy5nOYbiiktHy8/r0Xxdlr1fjtS5EV9vlg5dzTFG3BvA7YtY1uXIq86Hrg9Il7a8F1GkvcREetQFDsvAw4ANiu/z2SKty6/AFwzkrUXM/PtmfmHAU4NlZPtS/9pQ79Wud/iMsY+0yNik8r5JzNz+XDjq4qIyRSFyaCYBeKcNbmPJEkqlLnT5az8vZ7AmzJzTaf5PBq4Btgb+CeKfGgv4McRsd0w4plM8fbei8pDfwX2ycwbVndtZp6bmZc35hnltX+tHFqR05Q54Gsq567NzL9U9r9W+RwUfWbVe/ebSrRJJ1U+30uRl0oaAacAlTQWHqHogPgN8DDFW2IbU3TKbEfRefS5iPhG2SEyqDIZqk658C3gZoqkaXPgX4Gt+05m5u8i4j7gWeWhlwE3lQlcNbHaKSL+OTMfon8x8N7MvLuyv4Dizbh7y++S5XMPA/6ZokPqkxSdbGPleQ37v6h83rXh3N2r2d+FYr3FZvV1Gi6hKPYO+fcoSVKXquYoSbGe77CVHSzfoniTEIpR4V+heBvvhaycBvzEctrIrw5yqz0ofr9fDbycopAIsAlFp9MZ5fOeTlFg6ys+LaMY0LOAoqNnl/L4myje2Pt4Zv49Ii4DjinPvZFiusk+1UFdfypj6Cs0ns/KNelup+hMC+D1FPnbhsC3ImJqZj460ryv9BmKHBOKUehfL5+1Zfk91qHoZDuT4m3AZox2TtY4DeiaOAvYgiL/fjNOAypJ0horBzt/gyKP6HNqZn67idtOp8irvkmRL7y6PL4exWCytw8RT2Px736K2SGaWkInIp4JTKkcquY0W1PkXn2Gk9OMuoh4Bf0H9X86M5eOxbOkbmYBUNKoy8xjImJtio6rbShGYD8AfJeVRbiNKN68+8lqbrdu5fNjwOsz86lqg4iY2v8SfkzR4QNFAfCT5Z9BMdXC2hSdQXtQdERVC4DVt//IzF0j4p8oOpx6KEaA3V3GfXDZ7BURMXFN36YbSkRsTtGx1edO+o+22qjhksbFlxv3N6ZJETEN+FC5e1pm/ioiepq9ryRJHWjzyue/ZGZ1JDURsYhVf1eTmVF+fDXFjAl9XpmZv6xcfxnFoCOAEynWXhnIL4Hpmbmk7LhawMr1Bl9YaXcU/dchfFtmfrF81gyKqTy3Kc99ICLOyMxlFG/19RUAd46InTLz9ogI4IjK/S6sjDA/gZXFv1uBF/blcBHx38B9FHneJhSFqzMZYd4XERsCb62c/mBmfrJy/lesXJ/xLRHxocxcxBqIiJ0oBkD1uSYzf1rZryMnO4SV6+68KzMXRIQFQEmS1tzZ9J896hsU06Y3408UedDjsCI/2a0898JBryr60qrFv3nAKxoGrY9YREwCLqKYZQGKfrLqDAstz2kG8Z+Vz38B/neMniN1NQuAkkZdRLwe+Bz9O5gGsvlqzgP8luIX/SYUU1TNi4ibKYpwvwVuyMzfNVzzI1YWAF9ajq7vK/L9jKJzaQ9gz4i4ipWj5Puu7fseawH/RbEmT3WKp0brUCQ89w/j+wxbROxCMV3UFuWh+cABDR1h0XjZavazyZjWpXhTYG2KEWKfaOZ+kiR1kTX5HbtHw/4vipragHaJiH/q6zxq8MW+gUhlEfCPrMzDNqy0m1753Pf2H+V1T0bEVynWaum77nnAHZn5i4i4nZVTfb6RYkqm6awcIb+cYirKgb7bLsCTQ3y36RQFwJHmfS+m/3/TnjHEdKkTyvZXDhbEYCJib4o3C59eHroFOLyx2Qj3m83JNqaY3hSKdRcvaeZ+kiSNdxHxDvoPLLoNOKpvmZUmXNyQv93FygLghgO071OdWvN3FAPFFjQTSET8M0VRc6/y0JPA6zLz99VmjZetZr/Zn88qIuL5FOsm9jkrM58Y7edI44FrAEoaVRGxG8XUVasr/sHQRTWg6IwC/g24pzz0TIrpsI6n6PT4bUT8ICLWq1xWfYtvCkVnVd8IrhvKDYqi4O4Ua/r1qa7L807gg8OJc5hthi0iDgJ+ysri328oRvbf29C0cRT701ez/1eaczxFZ+A/gP8o3wqQJGm8qnbCbFp2qlR9mOJNuJsHub6x/eoMNsJ6XsP+k5XP1f/mqz7v4ca36yhmbKiqtq+Ouj6yHChVnf7zB5k5f5BrV2cTWKO8b6Q/v01W36S/iDiWYvR9XwfdT4C9M/Phhqatzsn+iyLffpBijUVJkrSGImIvioHsfRYBB2fm30fh9vMa9gfL04byV4pladZYRGwD3MTK4t+jwKsz8/sNTVud0wyk+vbf34DPj8EzpHHBNwBbLCIupFgrbGFm7riatp+lWMMDigViN83MKUNcIrWDf2dlApMUI8S/k5mPR8T2FFNYjkhm/qSczmhnihHkz6Uo6h1EMZr7lRSda6eV7RdExF2snMLqIFZOr3UDRbHuZIq51/um8QS4MzOrHV+vr3z+M0WH1K/LEfLvYIwSkIh4P8W0pX0/x6uBwzPz0QGa39Kw3zjtU+M6Obc2Gd5m5Z/rA3OHGMl/UURcBHw0M09t8pmStApzKrWJa1iZbwTQC/x338nMPBcgInYEpg1w/UOVz0kxxfZQa5sM1vHTOA35YCOxq8/bMCImNRQBNxui/cUU00OtQzGLwyso8r4+Xxzg2r4BYb9m8OlLoVJIHWHe91DDfS6gGFU/mMEKsasoC5yfBt5bOfwl4NgBCqcwcE7268r+WOVkzwD+MkROdm157s2ZObPJZ0rqMuZTEkTEVhRv+vf1ky+leCtu3ig9Yrh5WqPfA1uVcb0UuCIiDsjMxSMNoDKbQd+ApnuAgzLzNwM0vxt4nJXrAI51P1M/ZR742sqhLwww8ErSMFkAbL2ZwDlUptsZTGau+I/NiHgXK18Pl9pZdWT6o8CllbVgXj9A+yGVc5Nvk5l3UCQVt1bOXUHRGQTFeoJVP2Jlh9x7KIppT1JMW7k2RUK3NnBcwzWDfZc5mXlT+dy16N/hNSrKNXvOA46uHD4beO8Qb9rNoZga9Nnl/usi4tOZmeW6PNU455ftJakbzMScSvU7CzgWmFju/1dE/C4zvzvM639KsbYfFAXEBzPzosZGZcfUNpn5SJPx3sjKNQUnUKwf17cG4DrAkZW2D1NMuwlAZj4UEd9iZT53HivfwFsIXNHwrJ+ysvPmX4BLGgZa9eVU+wB/KPdHmvfdxMqcDmCdzPx045eOiCnA/pl5e+O5gZTrP88CDigPLQc+lJmDTS8K8AOKGRLWL/cPA/6vvN+6ldgBbszMvwwnFkkaYzMxn9I4FhFPAy6n/7p378nM62sKqepnFOsPfokiT9wb+FZEHDzIYKQBlbMZfJ6V+dINwGsb167uk5nLI+I7rMwL94qITTNzYbl/WLU58O3hxjJMJ7JyQPxTVAbXSRo5C4Atlpk3RERP9VhEbE3xf8SbUPxH4/8bYE2zI4CPtCJGqUlzK5+nAN+LiJ8Az6f//OXD9XTg9oj4PUXycz/FgsNTgVdX2jWOAP8R8Pbyc18h75fl1FJPRsQcioWUJzdc0/hdppafD4iIC4D7KDqDBhrF36xv0L9z6BaKot17G0Z1/ykzL4MVidlprBx1/0Lgmoi4lmJah2ph9LRKMZaI2JDiTcg+1VFcW0dEtQNtRjni6pYyzkbrA/tX9m8G7qWYulSSRp05ldpBZt5Vvrl/VnloPeCqiPgpxVSRj1G8BbfPILf4LnAH0PfWxQURcQjFm2NLKaYCfxHFm3BfopgVoBlfAk5h5VSY50XESyjewHsNKwdPAfz3AAOQvsjKAuBzqvftW4Ow4tMUud9aFG+p3R4RX6fIbdYHtqOYjn0TijdK/sgI877MfLjMz/pyvv+IiOcBP6QYub4pRQf1Syhmc5g11A+n4kZWrncIcD2wLCI+0NDujr5pszLz0Yj4b4qfLxSDsr5O8Xd5CCundYeV6ywCK/6/6+2VQ9X1gKZVc7LM7IvhZwz8tugmrJz6HopOvr+w6vRjkmQ+pXGtHDR9Mf1/5/8JWH+A3/l9Vvzub4XMvLjsu+mbnvRVwNci4nWZOdSsEQBExPuAz1QOPQR8H3hz4+wBDYOoPk4xC9Y6FHnbTyLiKxRrP/dW2n2lYf1AIuJDrBwk9pKGc9VnXJqZNzec34xigFqfizPzz0N8RUmrk5luLd6AHopfGH37PwKmlp9fBPy4of2WFP/xO6Hu2N3c+jaKaQiysl1UHt+QImHKAbYLG/Z7K/frbTjXUx7feJB7Vbe/A7s3xPfPwLKGdv9VOf/JhnNLgckN93gJxWijxuctoUgSV4m3vG5m5fi8EfxM5w3juyZw3QDXnr2aa84e4JqeYT6v3/cbJPbGe/UO93u7ubm5remGOZVbm2xlHvPYMH+n/rXh2udQTFu5uutmVq7Zq+HcXg33vG6wvAH4V4q1XYZ61lcH+t8Jxejzuwdov+0gP5ejKWZgWN1326tsP+K8D1gXuHIY180bwd/ncPOjmQ3XTQS+uZprPjDA8xr/PgfdhhH7kP823Nzc3Bo3zKfcxulW/lse7u/8gfKxeYMc72m4prfhuTMHy0+GuOdHG+552XD+N9jwrBHlGBSF/iVDXHMj8PQBrps3zGf2DnDtGZXzyyhmhqj934qbWydvw11oVGOkfNX8JcD/RcQtFIvbP7Oh2euBr+fgUwBKddiiYf8xgCzeEpsOfA14BFhMMX3TWyjX6BuhR4F3UBTcbqeYYmopxUjEucD5wPMz81fVizLzIVZdi+WGyufG6RzmZMMae5n5M4r1bX5Sfo/HgR9TjFZvfFuwVpn5Loo3E6+iGOW9tPzzKuCA8rwkdS1zKtUpi7XVtgQ+QDEV5P0Uha+nKHKXn1NMsXYwDf8uM/MeinWJ30NRuPsrxe/xRcBtFDnQ68vzoxHrz4EdKEZ23wL8rXzeA8B3gEMz88iB/neSmQn8b8Phn2Tm3Ma2Zfv/pVjL72yKdaD/TvFzmU/xBtvpFGva9OVoI877MnNxZh5IMUr9corZGp6iyE1/R1EcPA7YY1g/oCZk8Rbkv1Gsgf1jilH2Syj+Pfwf8JIcYIpSSWoX5lMaZwZdQLfdZOZHKPKpPocBF5bTqY/VM2dR5KgzKXK3pyhytZ8B76QYZPTYaD0vIiYDb6sc+lZmDrW2s6RhiOK/4dRK5fQKV2bmjhHxdGBuZjYmVNX2vwaOK4sRUu0iYgOKxYNfVTl8bGZeUFNIkqRxyJxKkiSpOeZTkiR1L98ArFk5UuKPEfHvUMw/HRG79J2PiG0pplT8eU0hSitExNPKUYAP0L/49zjFdEeSJNXCnEqSJKk55lOSJHUXC4AtFhGzKBKlbSNiQUQcDbwBODoibqWYGufgyiVHUCyK6quaagdrA7sAT6scexI4KjP/Wk9IkqTxyJxKkiSpOeZTkiR1N6cAlTRsETEFeJhi/ZZ7KdapOWuwdV8kSZIkSZIkSVLrWQBskfPOOy8B3v72t3fMArOSJEntxpxKkiSpOeZTkiSND2vXHcA4ZMVVkqTOZAdJezGnkiSpM5lTtQ/zKUmSOtOw8inXAJQkSZIkSZIkSZK6iAVASZIkSZIkqQ1FxIURsTAi7hjkfETEWRHxh4i4LSJ2b3WMkiSpPVkAbFJEzIuI2yPiloi4ue54JEmSJEmS1DVmAq8a4vz+wNRyOxY4rwUxSZKkDuAagKPj5Zm5qO4gJEmSJEmS1D0y84aI6BmiycHAlzMzgZsiYkpEPDMz729JgJIkqW35BqAkSZIkSZLUmZ4F/Kmyv6A8toqIONbZqyRJGj8sADYvgR9ExJyIOLbxpMmVJEmSJEmSxkgMcCwHapiZ52fmtDGOR5IktQkLgM17aWbuTjHn+nER8bLqSZMrSZIkSZIkjZEFwBaV/c2BP9cUiyRJaiMWAJuUmX8u/1wIfAt4Yb0RSZIkSZIkaZy4AviPKLwYeNT1/yRJEsDadQfQySJiA2CtzHy8/LwvcFrNYUmSJEmSJKkLRMQsYC9g44hYAHwEmAiQmf8DfBd4NfAH4B/Am+uJVJIktRsLgM15BvCtiIDiZ/nVzPx+vSFJkiRJkiSpG2TmEas5n8BxLQpHkiR1EAuATcjMe4Bd6o5DkiRJkiRJkiRJ6uMagJIkSZIkSZIkSVIXsQAoSZIkSZIkSZIkdRELgJIkSZIkSZIkSVIXsQAoSZIkSZIkSZIkdZG16w5A48vll18+rHYHH3zwGEciSZLUucypJEmSmmM+JUnqdr4BqFotWrSIRYsW1R2GJElSRzOnkiRJao75lCSp2/gGoFqqcdRU32grR1NJkiQNnzmVJElSc8ynJEndzjcAJUmSJEmSJEmSpC5iAVCSJEmSJEmSJEnqIhYAJUmSJEmSJEmSpC5iAVCSJEmSJEmSJEnqIhYAJUmSJEmSJEmSpC5iAVCSJEmSJEmSJEnqIhYAJUmSJEmSJEmSpC5iAVCSJEmSJEmSJEnqIhYAJUmSJEmSJEmSpC5iAVCSJEmSJEmSJEnqIhYAJUmSJEmSJEmSpC5iAVCSJEmSJEmSJEnqIhYA1daWLVvGGWecwR577MFGG23ERhttxL777svs2bOHvO6QQw4hIjjnnHNaFKkkSZIkSZIkSVJ7sACotvbEE09w+umn84IXvICLL76YSy65hIkTJzJ9+nTmzJkz4DU/+MEPuOmmm1ocqSRJkiRJkiRJUntYu+4ApKGst9563HPPPWy44YYrju2zzz5ss802nHPOOVx00UX92i9ZsoT3vOc9zJgxg2OOOabV4UqSJEmSJEmSJNXONwC1Wr29vUybNo1rrrmGnXfemQ022IDp06dz5513jvmzJ0yY0K/4BzBp0iR22GEHFi5cuEr7z33uc6y33nq8+c1vHvPYJEmSJEmSJEmS2pEFQA3L/PnzOeGEEzj55JOZNWsWCxcu5LDDDiMzh7xu6dKlQ27Lli1b7T0aPfnkk8yZM4ftt9++3/EHHniAj33sY5x55pmstZb/tCVJkiRJkiRJ0vjkFKAaloceeogbb7yRqVOnArB8+XIOPfRQ5s6dy3bbbTfodRMnTlztvd/1rndxyCGHDDuWGTNm8PDDD68yxeeJJ57Ifvvtx8te9rJh30uSJEmSJEmSJKnbWADUsPT09Kwo/gEr3r5bsGDBkAXA2bNnD3nf66+/nk033XTYcVx11VXMmDGDz3zmM2y77bYrjv/85z/n61//Or/97W+HfS9JkiRJkiRJ3eXyyy8fVruDDz54jCORpHpZANSwTJkypd/+pEmTAFi8ePGQ1+26665Dnp8/f/6wp+ucPXs2hx9+OG9961s5/vjj+507/vjjeetb38rkyZN55JFHVhx/4oknePTRR5k8efKwniFJkiRJkiSpeyxatAiAjTfeuOZIJKm1uqYAGBE/BG4Gfg3Mycw/DNDmhcD8zHyg1fGNV6M1Behdd93FAQccwD777MPZZ5+9yvm5c+fyy1/+kjPPPLPf8RNPPJEPfvCDLF26dGSBS5LUZerIlSLiQuBAYGFm7jhEuxcANwGHZ+bXR+PZkiRJo82+J6kzNL7Z1/dGoG/8SRpvuqYACOxdbgkQEY9TJGS/AuYAtwMnAlsAL6opxnFnNKYAvf/++9lvv/3YeuutmTVrFhMmTFilzZVXXrlKke/lL3857373u3nta1878sAlSeo+deRKM4FzgC8P1iAiJgBnAFeP0jMlSZLGin1PkiSpY3RTAXAKsDvw/HKbBrwM2JMyMQMC+Hst0Y1T06ZNG/L8fffdN+T5J554gv3335+HH36Yc845h9tuu23FuXXWWYfddtsNgOnTpw94/dSpU9lzzz1HGLUkSV2p5blSZt4QET2rafYu4BvAC0bruZIkSWPEvidJktQxuqYAmJmPAdeVGwAR8XTgpcARwBuAB4BP1RCe1tCDDz7IrbfeCsCBBx7Y79yWW27JvHnzaohKkqTO0465UkQ8CziUYiT9kAXAiDgWOPbcc89tRWiSJEmraMd8SpIkaTBdUwAcSJmYfQ/4XkRcCVwAXF9vVJ1n5syZqxzr6ekhM1dtPMqaeU4r4pMkqZO1Qa50JnBSZi6LiCEbZub5wPnnnXeev+AlSVLbaIN8SpIkaUBr1R1Aq2Tm14BbKNaYkSRJUkVNudI04NKImAe8Djg3Ig5p4fMlSZJGjX1PkiSpnYybAmDpVuDFdQchSZLUplqaK2XmVpnZk5k9wNeBd2Tmt1v1fEmSpDFg35MkSWoLXVMAjIjeiNgpIob6ThsBT7QqJkmSpHZRR64UEbOAnwPbRsSCiDg6It4WEW8brWdIkiS1in1PkiSpk3TTGoAXAgksjojbgDmV7T5gD+C1wNm1RShJklSfludKmXnECNr2jtZzJUmSxoh9T5IkqWN0UwHweGA3YHfg+cCLKJKyqgeBeyJiGnB7Zj7Z2hAlSZJqY64kSZLUHPMpSZLUMbqmAJiZZ/V9joh1gJ0pErK+bUdgM+DzZbNlEfE74NeZeVSLw5UkSWopcyVJkqTmmE9JkqRO0jUFwKpydNXscgMgItamSMSqidnOwA6ASZgkSRo3zJUkSZKaYz4lSZLaXVcWAAeSmUuBW8rtQoBy0ebt6oxLkiSpHZgrSZIkNcd8SpIktZNxUwAcSGYuB35TdxyS1O4uv/zyYbU7+OCDxzgSSa1kriRJktRb6xEQAAAgAElEQVQc8ylJklSXteoOQJLUeRYtWsSiRYvqDkOSJEmSJEmSNIBx/QagJGl4Gt/s63sj0Df+JEmSJEmSJKn9+AagJEmSJEmSJEmS1EUsAEqSJEmSJEmSJEldpKMLgBFxZkS8LCKi7lgkSZLajbmSJElSc8ynJElSp+roAiDwTuBa4MGI+GJEvDoiJtUdlCRJUpswV5IkSWqO+ZQkSepInV4A/Bfg7cAc4I3Ad4BFEXFpRBweEf9Ua3SSJEn1MleSJElqjvmUJEnqSB1dAMzMhZl5fmbuD2wKvAm4Gng1MAtYGBFXRsTREbFJnbFKkiS1mrmSJElSc8yntDqZyYwZM3j2s5/Nuuuuy+67787VV19dd1iSJHV2AbAqMx/LzK9m5r8DmwCHAJcCLwIuAP4cEddHxHsiYss6Y5UkSWo1cyVJkqTmmE9pIKeffjqnnXYaxx13HJdffjk77LADBx10ELNnz647NEnSONc1BcCqzHwyM6/IzDcDzwD2Af4H2Ar4LHBPRMypM0ZJkqS6mCtJkiQ1x3xKAE899RSf+MQnOOmkkzjppJPYb7/9uPjii9lpp5346Ec/Wnd4kqRxrisLgFWZuTwzr83Md2Xms4EXAmcA69UcmiRJUu3MlSRJkppjPlWv3t5epk2bxjXXXMPOO+/MBhtswPTp07nzzjvH/Nl33303jz/+OK94xSv6HX/lK1/JNddcw1NPPTXmMUiSNJiuLwA2ysybM/NDmbl93bFIkiS1G3MlSZKk5phPtd78+fM54YQTOPnkk5k1axYLFy7ksMMOIzOHvG7p0qUrtmXLlrFs2bJ+x5YuXTrkPRYvXgzApEmT+h1fZ511eOqpp7jnnnua/3KSJK2htesOQJIkSZIkSZLW1EMPPcSNN97I1KlTAVi+fDmHHnooc+fOZbvtthv0uokTJ6723hdddBG9vb0DnnvOc55DRDB79mxe/OIXrzj+y1/+ckVckiTVxQKgJEmSJEmSpI7V09OzovgHsP32xcuXCxYsGLIAOHv27BWfr7/+egD23HPPfm222mqrQa+fPHkyRxxxBDNmzGDHHXdkl1124Stf+Qo//OEPAZgwYcLIv4wkSaPEAqAkSZIkSZKkjjVlypR++31TcvZN0TmYXXfddcXn+fPnr3IMVl/EO/PMMzn88MPZe++9Adhiiy045ZRTOPXUU3nGM54xvC8gSdIYsAAoSZIkSZIkadxpdgpQgE022YQf//jHLFiwgEcffZRtt92WM888k80224yenp7RC1aSpBGyAChJkiRJkiRp3Gl2CtCqzTffnM0335zFixdz4YUX8pa3vGX0ApUkaQ1YAGxSREwAbgbuy8wD645HkiRJkiRJ3SMiXgV8DpgAfDEzT284/2zgS8CUss1/ZuZ3Wx5oB5o2bdqKz/fdd98qx4bj4osvZsmSJTznOc9h/vz5fPazn2XChAl88IMfHNVYJUkaqa4vAEbExsAewD+AH2bmslF+xHuA3wJPH+X7SpIkjbkW5EqSJEldbSzzqXLg+eeBVwILgNkRcUVm/qbS7BTga5l5XkRsD3wX6BmtGDS05cuXc8YZZ3DvvfcyefJkDjnkED7+8Y/ztKc9re7QJEnjXNcUACPi7UAvsH9mPlQeez7wfeCfy2Y3R8Temfn3UXrm5sABwAzgfaNxzzV15E1L6nz8GtvpseUAXNah8QN89cWrny9ekqS61ZErSZIkdZOa8qkXAn/IzHvK510KHAxUC4DJyoHpk4E/j9KzO8LMmTNXOdbT00NmtuT5Rx11FEcddVRLniVJ0kisVXcAo+hwIPsSsNKngA2BiyhGP70AeNsoPvNM4ERg+WANIuLYiLh5FJ8pSZK0JurIlSRJkrpJHfnUs4A/VfYXlMeqTgXeGBELyhjeNdCN7KOSJGl86aYC4FTgtr6dcvqFPYH/zcxjMvMgYDZw5Gg8LCIOBBZm5pyh2mXm+Zk5ssnDJUmSRl9LcyVJkqQuVEc+FQMca3y17QhgZmZuDrwauDgiVunzs49KkqTxpZsKgBsBCyv7Ly3//Fbl2E+ALUfpeS8FXhMR84BLgb0j4pJRurckSdJoa3WuJEmS1G3qyKcWAFtU9jdn1Sk+jwa+BpCZPwfWBTYexRgkSVIH6qYC4EP0T272pJia82eVY0mRBDUtMz+YmZtnZg/weuDHmfnG0bi3JEnSGGhpriRJktSF6sinZgNTI2KriJhE0Qd1RUOb+cA+ABHxvPL5fxnFGCRJUgfqpgLgb4GDImKjiJhCMS/77Mx8rNKmB3igjuAkSZJqZq4kSZLUnJbnU5m5FHgncHX5/K9l5p0RcVpEvKZs9n7g/0XErcAsoDczG6cJlSRJ48zadQcwij4HfJtiaoSlwPrASX0nI2ICMJ3+o7JGRWZeB1w32veVJEkaRbXlSpIkSV2ilnwqM78LfLfh2Icrn3/DyulIJUmSgC4qAGbmFRHxNuDY8tBXMrO6Jt8rKKZAuLrlwUmSJNXMXEmSJKk55lOSJKmTdE0BECAzzwfOH+Tc1cCGrY1IkiSpfZgrSZIkNcd8SpIkdYpuWgNQkiRJkiRJkiRJGvcsAEqSJGlMRMSFEbEwIu4Y5PwbIuK2cvtZROzS6hglSZIkSZK6UVcVACNiz4i4suxoWhIRywbYltYdpyRJUh1qyJVmAq8a4vwfgT0zc2fgYwwynZYkSVK7sO9JkiR1iq5ZAzAiDgC+DUwA5gNzARMuSZIk6smVMvOGiOgZ4vzPKrs3AZuPZTySJEnNsO9JkiR1kq4pAAKnAkuAAzLzBzXHIkmS1G5Opb1zpaOB7w12MiKOBY4999xzWxeRJElSf6fS3vmUJEnSCt00BeiOwGUmYJIkSQNq21wpIl5OUQA8abA2mXl+Zk5rXVSSJEmraNt8SpIkqVE3vQH4N+ChuoOQJElqU22ZK0XEzsAXgf0z8691xyNJkjSEtsynJEmSBtJNbwD+CPjXuoOQJElqU22XK0XEs4FvAm/KzLvqjkeSJGk12i6fkiRJGkw3FQBPAraOiFMiIuoORpIkqc20PFeKiFnAz4FtI2JBRBwdEW+LiLeVTT4MbAScGxG3RMTNrYhLkiRpDdn3JEmSOkY3TQH6EeBO4KPAWyLiFuCRAdplZh7d0sgkSZLq1/JcKTOPWM35Y4BjRuNZas63v/1tDj30UJ7//Odz883WYSVJGoR9T5IkqWN0UwGwt/K5p9wGkoBJmCRJGm96K597MFdSafHixbzvfe/jGc94Rt2hSJLU7norn3swn5IkSW2smwqAW9UdgCRJUhszV9KAPvWpT/GsZz2LrbfemjvuuKPucCRJamfmU5IkqWN0TQEwM++tOwZJkqR2Za7Ufnp7e7njjjv4xCc+wfvf/37uvvtudtttN77whS+www47tCSG+fPn88lPfpLrr7+es846qyXPlCSpU5lPSZKkTrJW3QGMlYh4ekRsERFPrzsWSZKkdmOu1B7mz5/PCSecwMknn8ysWbNYuHAhhx12GJk55HVLly7tty1btoxly5b1O7a6ewC8//3v57DDDmP33Xcfra8kSdK4YT4lSZLaWde8AQgQEROAE4BjqEzLEBF/BL4IfDozl9YUniRJUq3MldrPQw89xI033sjUqVMBWL58OYceeihz585lu+22G/S6iRMnrvbeF110Eb29vYOev/baa7n66qu56667Rhy3JEnjlfmUJEnqFF1TAIyIScD3gT0pFlv+E3A/8EyKRZlnAK+KiH0z86m64pQkSaqDuVJ76unpWVH8A9h+++0BWLBgwZAFwNmzZ/fbv/766wHYc889VxzbaqvBlylaunQp7373uznllFPYbLPN1ih2SZLGG/MpSZLUSbqmAAi8D9gLuBJ4f2b+vu9ERGwNfAY4qGx3eh0BSpIk1chcqQ1NmTKl3/6kSZMAWLx48ZDX7brrrv3258+fv8rxCRMmDHr9BRdcwCOPPMJRRx3FI488AsBTTz3FsmXLeOSRR9hggw2G9ZahJEnjjPmUJEnqGN1UADwSuAM4JDOXV09k5t0R8VrgFuANmIRJkqTxx1ypizQ7BejcuXNZsGDBgG//bbjhhlx88cW88Y1vbDZMSZK6jfmUJEnqGN1UAHwucHZjAtYnM5dHxPeAd7U2LEmSpLZgrtRFmp0C9J3vfCeHHHJIv2Onn346f/zjH/nCF77A8573vFGMVpKkrmE+NQxH3rSk7hDWyE6PFX+tl3Vo/ABffbEzOEiSVuqmAuBTwNNW02YDoHN/i0uSJK05c6UuMm3atH77991334DHB/Pc5z6X5z73uf2OzZw5k0WLFrHXXnuNSoySJHUh8ylJktQx1qo7gFF0G/C6iNhkoJMRsTHwOuDWlkYlSZLUHsyVJEmSmmM+JUmSOkY3vQF4DnAp8MuI+C/gWuB+YDOKBZpPATYB3l1XgJIkSTUyV2ozM2fOXOVYT08Pmdn6YBg4HkmS1I/5lCRJ6hhdUwDMzK9FxK7AfwLnD9AkgE9m5tdaG5kkSVL9zJUkSZKaYz4lSZI6SdcUAAEy80MRcQVwNLAbMBl4FPg1cGFm/rzO+CRJkupkriRJktQc8ylJktQpuqoACJCZNwE31R2HJElSOzJXkiRJao75lCRJ6gRr1R2AJEmSJEmSJEmSpNFjAVCSJEmSJEmSJEnqIh07BWhEXAgk8KHMfLDcH47MzKPHMDRJkqTamStJkiQ1x3xKkiR1so4tAAK9FEnYGcCD5f5wJMVCzZIkSd2sF3MlSZKkZvRiPiVJkjpUJxcAtyr/vK9hX5IkSeZKkiRJzTKfkiRJHatjC4CZeW/DoT2BWzLztsGuiYgdgd2BL49lbJIkSXUzV5IkSWqO+ZQkSepka9UdwCiaCRyymjYHAxeNfSiSJEltZybmSpIkSc2YifmUJEnqEN1UAByOCRTzsEuSJGlV5kqSJEnNMZ+SJEltYbwVALcBHq47CEmSpDZlriRJktQc8ylJktQWOnYNQICIuLDh0CER0TNA0wnAs4E9gKvGOCxJkqS2YK4kSZLUHPMpSZLUqTq6AAj0Vj4nsGu5DSSBXwDvHeOYJEmS2kVv5bO5kiRJ0sj1Vj6bT0mSpI7R6QXArco/A7gHOBP43ADtlgEPZ+bfWxWYJElSGzBXkiRJao75lCRJ6kgdXQDMzHv7PkfER4Frq8ckSZLGM3MlSZKk5phPSZKkTtXRBcCqzPxo3TFIkiS1K3MlSZKk5phPSZKkTrJW3QFIkiRJkiRJkiRJGj1d8wZgn4h4JrAP8CxgnQGaZGZ+rLVRSZIktQdzJUmSpOaYT0mSpE7QVQXAci72/6T/9wogGz6bhEmSpHHHXEmSJKk55lOSJKlTdM0UoBHxBuD/A34CvI4i4foScCRwAbAcuBTYu64YJUmS6mKuJEmS1BzzKUmS1Em66Q3AtwMLgFdl5tKIAJiXmZcCl0bEt4CrgFk1xihJklQXcyVJkqTmmE9JkqSO0TVvAAI7Ad/NzKWVYxP6PmTm1cDVwAmtDkySJKkNtDxXiogLI2JhRNwxyPmIiLMi4g8RcVtE7D5az5YkSRoD9j1JkqSO0U1vAE4E/lrZfwKY3NDmDuBtLYtIkiSpfdSRK80EzgG+PMj5/YGp5fYi4Lzyz1ocedOSuh7dtJ0eWw7AZR36Hb764ol1hyBJ0nDY9yRJkjpGNxUA7weeWdmfD+zc0OZZwFJUm51+f2W//af9468DHr996oEti0mSpHGi5blSZt4QET1DNDkY+HJmJnBTREyJiGdm5v2jFYMkSdIosu9JkiR1jG6aAvTXFFMx9PkxsEdEvCkiNoiIA4B/K9upTfxt/Y342/ob1R2GJEnjQTvmSs8C/lTZX1AeW0VEHBsRN7ckKkmSpIG1Yz4lSZI0oG56A/BK4NyI2Coz/wicDhxOMfXUzLLNEuCUWqIT4Jt9kiTVqB1zpRjgWA7UMDPPB84/77zzBjwvSZLUAu2YT0mSJA2oawqAmTmTlckWmfmniHgB8H5ga2AecG5m3l5HfJIkSXVq01xpAbBFZX9z4M8tfL4kSdKwtWk+JUmSNKCuKQAOpByN9c6645AkSWpHbZArXQG8MyIuBV4EPOr6f5IkqZO0QT4lSZI0oK4pAEbElMx8pO44JEmS2lEduVJEzAL2AjaOiAXAR4CJAJn5P8B3gVcDfwD+Aby5lfFJkiSNhH1PkiSpk3RNARC4PyKuAL4EfD8zl9cdkCRJUhtpea6UmUes5nwCx411HJIkSaPEvidJktQx1qo7gFE0D/h34DvAfRHxqYjYqd6QJEmS2sY8zJUkSZKaMQ/zKUmS1CG6pgCYmc+jWDvmfyimlno/cEtEzImId0fExrUGKEmSVCNzJUmSpOaYT0mSpE7SNQVAgMycnZnHAc+kGJF1FbATcCbFyKxvR8QhdcYoSZJUF3MlSZKk5phPSZKkTtFVBcA+mbkkM7+Rma8B/gV4H3AH8Brg67UGJ0mSVDNzJUmSpOaYT0mSpHbXlQXABn8F7gR+CywBYrRuHBHrRsQvI+LWiLgzIj46WveWJElqkTHLlSRJksaJMc2nIuJVETE3Iv4QEf85SJvDIuI3Zf/UV0fz+ZIkqTOtXXcAYyUitgOOAt5IMRIrgD8AXxrFxzwJ7J2Zf4uIicBPI+J7mXnTKD5DkiRp1LUoV5IkSeparcinImIC8HnglcACYHZEXJGZv6m0mQp8EHhpZj4cEZuO1vMlSVLn6qoCYERsCBxBkXxNo0i8HgP+F5iZmT8bzedlZgJ/K3cnlluO5jMkSZJGS6tzJUmSpG5TQz71QuAPmXlP+fxLgYOB31Ta/D/g85n5MEBmLhzlGCRJUgfqmilAI+LrwP3A2cDzgR8CbwA2y8xjx6pDKyImRMQtwELgmsz8RcP5YyPi5rF4tiRJ0nDVlStJkiR1i5ryqWcBf6rsLyiPVW0DbBMRN0bETRHxqoFuZB+VJEnjS9cUAIHXAn8ETga2zMz9MnNWZi4ey4dm5rLM3BXYHHhhROzYcP78zJw2ljFIkiQNQy25kiRJUhepI58aaD3Bxtmn1gamAntRvJ34xYiYsspF9lFJkjSudFMB8EPAEZl5embe1+qHZ+YjwHXAgKOsJGk8W7ZsGWeccQZ77LEHG220ERtttBH77rsvs2fPrjs0aTypNVeSJEnqAnXkUwuALSr7mwN/HqDN5Zm5JDP/CMylKAhKkqRxrJsKgDMo5jxvmYjYpG9EVUSsB7wC+F0rY5CkTvDEE09w+umn84IXvICLL76YSy65hIkTJzJ9+nTmzJlTd3jSeNHyXEmSJKnL1JFPzQamRsRWETEJeD1wRUObbwMvB4iIjSmmBL2npVFKkqS2s3bdAYyiRUCrp7B6JvCliJhAUUz9WmZe2eIYJKntrbfeetxzzz1suOGGK47ts88+bLPNNpxzzjlcdNFFNUYnjRt15EqSJEndpOX5VGYujYh3AlcDE4ALM/POiDgNuDkzryjP7RsRvwGWASdk5l9bGackSWo/3fQG4HXAS1r5wMy8LTN3y8ydM3PHzDytlc+XpJHo7e1l2rRpXHPNNey8885ssMEGTJ8+nTvvvHPMnz1hwoR+xT+ASZMmscMOO7Bw4cIxf74koIZcSZIkqctcRw35VGZ+NzO3ycytM3NGeezDZfGPLLwvM7fPzJ0y89JWxyhJktpPNxUATwG2jYiPRcTEuoORpHY0f/58TjjhBE4++WRmzZrFwoULOeyww8hsXEO+v6VLl/bbli1bxrJly/odW909Gj355JPMmTOH7bffvpmvJGn4zJUkSZKaYz4lSZI6RjdNAfpB4A6KBZmPjohbgQeAxh7pzMyjWx2cJLWDhx56iBtvvJGpU4v14JcvX86hhx7K3Llz2W677Qa9buLE1f+37UUXXURvb++wY5kxYwYPP/wwxxxzzLCvkdQUcyVJkqTmmE9JkqSO0U0FwN7K583KbSAJmIRJGpd6enpWFP+AFW/fLViwYMgC4OzZs/vtX3/99QDsueeeK45ttdVWw47jqquuYsaMGXzmM59h2223HfZ1kprSW/lsriRJkjRyvZXP5lOSJKmtdVMBcPg9z5I0Tk2ZMqXf/qRJkwBYvHjodex33XXXfvvz589f5fiECROGFcPs2bM5/PDDeetb38rxxx8/rGskjQpzJUmSpOaYT0mSpI7RNQXAzLy37hgkqVuN1hSgd911FwcccAD77LMPZ5999ihFJ2k4zJUkSZKaYz4lSZI6SdcUAKsiYgNgG+BpmfmTuuORpE43GlOA3n///ey3335svfXWzJo1a9hvDEoafeZKkiRJzTGfkiRJ7a6rCoARsTnwOeAgYALFnOtrl+emA+cD78jM6+qKUZI60bRp0/rt33fffQMeH8wTTzzB/vvvz8MPP8w555zDbbfdtuLcOuusw2677TZ6wUoalLmSJElSc8ynJElSp+iaAmBEPBP4BfAM4ApgU+BfK01+UR47HLiu1fFJ0nj24IMPcuuttwJw4IEH9ju35ZZbMm/evBqiksYXcyVJkqTmmE9JkqRO0jUFQOAjFEnWKzLzuoj4CJUkLDOXRMRPgJfWFaAk1WnmzJmrHOvp6SEzx/zZrXqOpCGZK0mSJDXHfEqSJHWMteoOYBS9GrhiNVMszAf+pTXhSJIktRVzJUmSpOaYT0mSpI7RTQXAZwC/X02bJcAGLYhFkiSp3ZgrSZIkNcd8SpIkdYxuKgA+BGyxmjbbAA+0IBZJkqR2Y64kSZLUHPMpSZLUMbqpAHgj8JqI2GygkxExFXgVcG1Lo5IkSWoP5kqSJEnNMZ+SJEkdo5sKgJ8C1gWuj4j9gfUBImKDcv87wHLgM/WFKEmSVBtzJUmSpOaYT0mSpI7RNQXAzPwFcCzQA1wJfKA89Vi5vxVwdGbeWUuAkiRJNTJXkiRJao75lDS+XHPNNRxxxBH09PQQEZx66ql1hyRJI9I1BUCAzLwI2BE4C/glcDfwK+BcYOfM/EqN4UmSJNXKXEmSJKk55lPS+PH973+f2267jX322Yf111+/7nAkacTWrjuA0Zb/P3t3Hy9HXd/9//VJCChBQUHRApeJGgV+wgU2Um+wgIKAWgELCKi9jlKxWGu9KS2KImptLdTbaqkUAb1UbttqpAGMlahQwHCJoolyI0IIiKjcaZWbJJ/fHzMnbjZ7ck7O7tm52dfz8ZhHzs58d+Zzzndm55397sxm3gS8reo6JEmS6sisJEmS1B/zlDQaTjvtND784eKOvl/+8pcrrkaSNl2rrgCUJEmSJEmSJDXf2NgYCxcuZMmSJey+++7MnTuXvffem+XLh3OX3VmzfOtcUrO1/lUsIraLiMMi4sCImF11PZIkSXViVpIkSeqPeUqaOStXruSEE07gpJNO4txzz+Xuu+/myCOPJDM3+rzVq1evm9asWcOaNWvWm7d69epJ1yFJTdeaW4BGxPHAGHBwZt5Tzvt94FLg8WWzayPiRZn5P9VUKUmSVA2zkiRJUn/MU9Lw3XPPPVx55ZUsWLAAgLVr13LYYYdxww03sPPOO0/4vDlz5ky67rPPPpuxsbFBlSpJtdOaAUDgVUCOB7DSacDjgLOB7YGXAX8GfHj45UmSJFXKrCRJktQf85Q0ZPPmzVs3+Aew6667ArBq1aqNDgAuW7Zs3c/f+MY3ANhnn33WazN//vxBlipJtdOmAcAFwH+OP4iI7YB9gDMz843lvGuAYzCESZKk0TP0rBQRBwEfB2aX2/lQ1/L/BXwW2KZsc2JmLh7EtiVJkmaA7z1JQ7bNNtus93jzzTcH4MEHH9zo8/bYY491P69cuXKDeQCzZ3vHXknt1qYBwG2Buzsev6D89z865n2L4lYNkiRJo2aoWan8/ptPAQcAq4BlEbEoM1d0NHs3cEFmnh4RuwKLgXmD2L4kSdIM8L0nqSG8BagktWsA8B5gu47H+wBrgf/umJfAo4ZZlCRJUk0MOyvtBdycmbcARMR5wCFA5wBgAo8tf94auHNA25YkSZoJvvckNYS3AJWkdg0A/hD4o4g4CVhDcV/2ZZn5QEebecBdFdQmSZJUtWFnpR2A2zserwL+oKvNKcBXI+IvgLnA/hOtLCKOA47753/+5wGVJ0mStMl870lqiIULF677+Y477thg3lTcdttt6wYSH374YVasWMFFF13E3LlzOfjggwdXrCTNkFlVFzBAHweeTPHm0u3Ak4B17xCVt6HaG/heJdVJkiRVa9hZKXrMy67HRwPnZOaOwEuB/xsRPfNpZp6RmZv2P3ZJkqTB8r0naYRcfvnlHHHEERxxxBH86le/4sILL+SII47g+OOPr7o0SZqS1lwBmJmLIuLPgOPKWV/IzM93NNmf4hYMlw29OEmSpIpVkJVWATt1PN6RDW/xeSxwUFnfVRHxKIrbat2NJElSzfjekzRc55xzzgbz5s2bR2b35wpnxtjYmN8RKKnRWjMACMUnw4EzJlh2GfC44VYkSROLQz9RdQnT9tbnzAbg0LOb+Tvkl95SdQlSJYaclZYBCyJiPnAHcBRwTFeblcCLgXMiYheKN8x+PsAaJEmSBsr3niRJUlO0agAQICK2Ag4D9gS2Bu4HrgP+IzN/XWVtkiRJVRtWVsrM1RHxZopPwM8GzsrM5RHxfuDazFwEvAP414h4G8XtQcdyWB/nlSRJmibfe5IkSU3QqgHAiDgC+BdgG9b/3pkEPhYRb8zMiyopTpIkqWLDzkqZuRhY3DXv5I6fVwAvGNT2JEmSZprvPUmSpKZozQBgRBwAnAusBT4HLAXuovhC5v0objl1bkTcl5lfq6pOSZKkKpiVJEmS+mOekiRJTdKaAUDgZOAh4IWZ+Z2uZZ+NiE8C3yzbGcIkSQK+/OUvT6ndIYccMsOVaAjMSpIkSf0xT0mSpMZo0wDgnsD5PQIYAJl5bURcABw+3LIkSWqOX/ziFwBst912FVeiGWBWkiRJ6k8kslwAACAASURBVI95qkV2u+ni9R5v9Ztf9pz//QUvH1pNkiQNUpsGAB8CfjpJmzvLdpIkiQ2v7Bu/ItAr/lrJrCRJktQf81SL/XrLbasuQZKkgWrTAOC3gL0nafMCilsxSJIkjRqzkiRJUn/MUy3ilX2SpLabVXUBA/Q3wG4R8aGImNu5ICLmRsSpwLOAEyupTpIkqVpmJUmSpP6YpyRJUmM09grAiDirx+zrgROA4yLiO8DPgO2BZwNbU3wC66+BY4dVpyRJUhXMSpIkSf0xT0mSpCZr7AAgMLaRZdsAL+oxfx/gDzGESZKk9hvbyDKzkiRJ0uTGNrLMPCVJkmqtyQOA86suQJIkqcbMSpIkSf0xT0mSpMZq7ABgZt5WdQ2SJEl1ZVaSJEnqj3lKkiQ1WWMHACcTEY+luPf6/Zn5QNX1SJIk1YlZSZIkqT/mKUmSVGezqi5gkCJidkScGBE3A/cCtwL3RsTN5fzWDnhKkiRNxqwkSZLUH/OUJElqitaEkojYHLiU4suWE7gd+CnwZGAe8EHgoIh4SWY+XFWdkiRJVTArSZIk9cc8JUmSmqRNVwC+HdgX+E9gl8ycl5nPy8x5wDOBrwAvLNtJkiSNGrOSJElSf8xTkiSpMVpzBSBwDPAD4NDMXNu5IDN/HBGvBL4LvBr4UAX1SZIkVcmsJEmS1B/zlEbaMVc/UnUJ07LbA8Xhen5D6wf44nPnVF2CpAZq0xWATwcu6Q5g48r5lwBPG2pVkiRJ9WBWkiRJ6o95SpIkNUabBgAfBraapM1coLkf9ZAkSZo+s5IkSVJ/zFOSJKkx2jQAeD1weEQ8odfCiNgOOBz43lCrkiRJqgezkiRJUn/MU5IkqTHaNAD4SeAJwLcj4tiIeGpEPDoi5kfE64BryuWfrLRKSZKkapiVJEmS+mOekiRJjbFZ1QUMSmZeEBF7ACcCZ/RoEsCpmXnBcCuTJEmqnllJkiSpP+YpSZLUJK0ZAATIzHdFxCLgWGBPYGvgfuA64KzMvKrK+iRJkqpkVpIkSeqPeUqSJDVFqwYAATLzauDqquuQJEmqI7OSJElSf8xTkiSpCdr0HYCSJEmSJEmSJEnSyHMAUJIkSZIkSZIkSWoRBwAlSZIkSZIkSZKkFnEAUJIkSZIkSZIkSWoRBwAlSZIkSZIkSZKkFnEAUJIkSZIkSZIkSWqRxg4ARsRbImKvquuQJEmqI7OSJElSf+qSpyLioIi4ISJujogTN9Lu8IjIiFg4zPokSVI9NXYAEPgYcND4g4hYExHvGdbGI2KniLg8In4YEcsj4i+HtW1JkqQpqDQrSZIktUDleSoiZgOfAg4GdgWOjohde7R7DPAW4Jph1idJkuqryQOADwJbdDyOchqW1cA7MnMX4LnAn/cKYJIkSRWpOitJkiQ1XR3y1F7AzZl5S2Y+DJwHHNKj3QeAUylqliRJavQA4E+AAyNi+455OayNZ+ZPM/M75c+/An4I7DCs7UuSJE2i0qwkSZLUAnXIUzsAt3c8XkXX+08RsSewU2ZePMzCJElSvTV5APDTwLOBOyNiTTnvlPJ2DBubVg+6kIiYB+xJj9ssRMRxEXHtoLcpSZI0idpkJUmSpIaqQ57qdcXhukHIiJgFfBR4x6Qr8j0qSZJGymZVFzBdmfmJiLgbeBnwe8B+wErg1mHWERFbAf8GvDUzH+hR5xnAGaeffrqfuJckSUNTl6wkSZLUVDXJU6uAnToe7wjc2fH4McCzgKURAfAkYFFEvCIz1xvs8z0qSZJGS2MHAAEy8zyKe58TEWuBszPz/cPafkTMoRj8+0Jm/vuwtitJkjQVNchKBwEfB2YDZ2bmh3q0ORI4heKT7N/LzGOGVZ8kSdJkqs5TwDJgQUTMB+4AjgLW5aXMvB/YbvxxRCwF/qp78E+SJI2eRg8AdnkfsHRYG4viY1WfAX6YmR8Z1nYlSZKmadhZaTbwKeAAik+uL4uIRZm5oqPNAuCdwAsy896IeOKw6pMkSZqGoeYpgMxcHRFvBi6j+FDVWZm5PCLeD1ybmYuGWY8kSWqO1gwAZub7Oh9HxGOBrYH7e92acwBeALwW+H5EfLec967MXDwD25IkSepLBVlpL+DmzLyl3N55wCHAio42bwA+lZn3ljXePQN1SJIkDUQFeWp8u4uBxV3zTp6g7b4zVYckSWqWWVUXMEgRMTsiToyIm4F7Ke7Jfm9E3FzOH9iAZ2ZekZmRmbtn5h7l5OCfJEmqrWFmJWAH4PaOx6vKeZ2eATwjIq6MiKvLW4ZKkiTV1pDzlCRJ0rS1ZgAwIjYHlgAfBOZRvOH07fLfeeX8r5XtJEmSRkoFWSl6zMuux5sBC4B9gaOBMyNim54rizguIvwuG0mSVBnfe5IkSU3SmgFA4O0Ubx79J7BLZs7LzOdl5jzgmcBXgBeW7SRJkkbNsLPSKmCnjsc7Anf2aPPlzHwkM38C3EAxILiBzDwjMxcOqDZJkqTp8L0nSZLUGG0aADwG+AFwaGbe1LkgM38MvBJYDry6gtokSZKqNuystAxYEBHzy0/BHwUs6mrzJWA/gIjYjuKWoLcMaPuSJEmD5ntPkiSpMdo0APh04JLMXNtrYTn/EuBpQ61KkiSpHoaalTJzNfBm4DLgh8AFmbk8It4fEa8om10G/DIiVgCXAydk5i8HsX1JkqQZ4HtPkiSpMdr0xcQPA1tN0mYu8MgQapEkSaqboWelzFwMLO6ad3LHz0lxiyxvkyVJkprA954kSVJjtOkKwOuBwyPiCb0WlreVOhz43lCrkiRJqgezkiRJUn/MU5IkqTHaNAD4SeAJwLcj4tiIeGpEPLr83pnXAdeUyz9ZaZWSJEnVMCtJkiT1xzwlSZIaozW3AM3MCyJiD+BE4IweTQI4NTMvGG5lkiRJ1TMrSZIk9cc8JUmSmqQ1A4AAmfmuiFgEHAvsCWwN3A9cB5yVmVdVWZ8kSVKVzEqSJEn9MU9JkqSmaNUAIEBmXg1cXXUdkiRJdWRWkiRJ6o95SpIkNUGbvgNQkiRJkiRJkiRJGnkOAEqSJEmSJEmSJEkt4gCgJEmSJEmSJEmS1CIOAEqSJEmSJEmSJEkt4gCgJEmSJEmSJEmS1CIOAEqSJEmSJEmSJEkt0poBwIj4w4jYo+o6JEmS6sisJEmS1B/zlCRJapLWDAAClwPHVV2EJElSTZmVJEmS+mOekiRJjdGmAcBfAL+tughJkqSaMitJkiT1xzwlSZIaY7OqCxigpcDzqy5CkiSpppZiVmqN3W66eL3HW/3mlz3nf3/By4dWkyRJI2Ap5ilJktQQbboC8N3AMyPiAxExp+piJEmSasas1GK/3nJbfr3ltlWXIUlS25mnJElSY7TpCsB3Aj8A3gUcGxHfA+4CsqtdZuaxwy5OkiSpYmalFvHKPkmSKmGekiRJjdGmAcCxjp+fVE69JGAIkyRJo2as42ezkiRJ0qYb6/jZPCVJkmqtTQOA86suQJIkqcbMSpIkSf0xT0mSpMZozQBgZt5WdQ2SJI2iJUuWcNZZZ3HVVVdx22238d73vpdTTjml6rLUxawkSZLUH/OUJElqkllVFzATImJuROwZES+suhZJktru0ksv5frrr+fFL34xW265ZdXlaArMSpIkSf0xT0mSpLpr1QBgROwYEf8G3AtcC1zesWzviFgREftWVZ8kSW102mmnsXz5cj7zmc/w6Ec/uupytBFmJUmSpP6YpyRJUlO0ZgAwIp4MXAMcAlwMXAVER5NrgCcCrxp+dZIkzZyxsTEWLlzIkiVL2H333Zk7dy577703y5cvH8r2Z81qTZxoNbOSJElSf8xTkiSpSdr0jt17KULW/pn5SmBJ58LMfAT4FvCCCmqTJGlGrVy5khNOOIGTTjqJc889l7vvvpsjjzySzNzo81avXr3etGbNGtasWbPevMnWocYwK0mSJPXHPCVJkhpjs6oLGKCXAosyc+lG2qwEvDe7JKl17rnnHq688koWLFgAwNq1aznssMO44YYb2HnnnSd83pw5cyZd99lnn83Y2NigSlV1zEqSJEn9MU9JkqTGaNMA4PbATZO0eQSYO4RaJEkaqnnz5q0b/APYddddAVi1atVGBwCXLVu23uNvfOMbAOyzzz7r5s2fP3+Qpao6ZiVJkqT+mKckSVJjtGkA8B5gp0naPAO4awi1SJI0VNtss816jzfffHMAHnzwwY0+b4899ljv8cqVKzeYP3v27EGUqOqZlSRJkvpjnpIkSY3RpgHAK4FXRMSTMnODoBURC4CDgM8PvTJJkmrKW4COFLOSJElSf8xTkiSpMdo0AHgacAjwjYh4K7AlQETMBf4Q+CiwFvhwZRVKklQz3gJ0pJiVJEmS+mOekiRJjdGaAcDMvCYijgP+Bbi4Y9ED5b+rgddn5vKhFydJUk0tXLhwvcd33HFHz/kbc9ttt60bSHz44YdZsWIFF110EXPnzuXggw8eXLHqi1lJkiSpP+YpSZLUJK0ZAATIzLMj4grgTcBzgW2B+4GrgU9m5g1V1idJUhtdfvnlvO51r1v3+MILL+TCCy/kKU95Crfeemt1hWkDZiVJkqT+mKckSVJTtGoAECAzbwLeVnUdkiQNyznnnLPBvHnz5pGZQ9n+2NiY3xHYIGYlSZKk/pinJElSE8yqugBJkiRJkiRJkiRJg9O6KwAjYivgMGBPYGuK2zBcB/xHZv66ytokSZKqZlaSJEnqj3lKkiQ1QasGACPiCIovYt4GiI5FCXwsIt6YmRdVUpwkSVLFzEqSJEn9MU9JkqSmaM0AYEQcAJwLrAU+BywF7gKeBOwHHAOcGxH3ZebXqqpTkiSpClVkpYg4CPg4MBs4MzM/NEG7w4ELgedk5rWD2LYkSdKg+d6TJElqktYMAAInAw8BL8zM73Qt+2xEfBL4ZtnOECZJkkbNULNSRMwGPgUcAKwClkXEosxc0dXuMcBbgGv63aYkSdIM870nSZLUGLOqLmCA9gTO7xHAACg/TX4B8OyhViVJklQPw85KewE3Z+YtmfkwcB5wSI92HwBOBR4c0HYlSZJmiu89SZKkxmjTAOBDwE8naXNn2U6SJGnUDDsr7QDc3vF4VTlvnYjYE9gpMy+ebGURcVxEeHtQSZJUJd97kiRJjdGmAcBvAXtP0uYFFLdikCRJGjXDzkrRY16uWxgxC/go8I6prCwzz8jMhQOqTZIkaTp870mSJDVGmwYA/wbYLSI+FBFzOxdExNyIOBV4FnBiJdVJkiRVa9hZaRWwU8fjHSk+ET/uMeX2lkbErcBzgUUR4SCfJEmqK997kiRJjbFZ1QVMV0Sc1WP29cAJwHER8R3gZ8D2FPde35riE1h/DRw7rDolSZKqUIOstAxYEBHzgTuAo4Bjxhdm5v3Adh31LgX+qvzuHEmSpMrVIE9JkiRNW2MHAIGxjSzbBnhRj/n7AH+IIUySNIPi0E9UXcK0vfU5swE49Ozm/g75pbdUXUJdjG1k2YxnpcxcHRFvBi4DZgNnZebyiHg/cG1mLup3G5IkSTNsbCPLfO9JkiTVWpMHAOdXXYAkSVKNVZ6VMnMxsLhr3skTtN13GDVJkiRtgsrzlCRJ0nQ1dgAwM2+rugZJkqS6MitJkiT1xzwlSZKabFbVBUiSJEmSJEmSJEkanMZeAdhLROwIvA3YA9gRmNOjWWbm04ZamCRJUg2YlSRJkvpjnpIkSU3RmgHAiNiX4jtmHgWsBn5W/rtB0yGWJUmSVAtmJUmSpP6YpyRJUpO0ZgAQOBWYDfwJ8MXMXFtxPZIkSXViVpIkSeqPeUqSJDVGmwYAdwPOzczPV12IJElSDZmVJEmS+mOekiRJjTGr6gIG6F7gnqqLkCRJqimzkiRJUn/MU5IkqTHaNAB4MbBP1UVIkiTVlFlJkiSpP+YpSZLUGG0aAHwXsHVEfCoi5lZdjCRJUs2YlSRJkvpjnpIkSY3Rmu8AzMxfRMRBwDXAn0TEjcD9vZvmiwexzYg4C3g5cHdmPmsQ65QkSZoJVWQlSZKkNqkqT5Xb/DgwGzgzMz/UtfztwJ8Cq4GfA6/PzNsGtX1JktRMrRkAjIj/D7gceFw5a88JmuYAN3sO8EngcwNcpyRJ0sBVlJUkSZJao4o8FRGzgU8BBwCrgGURsSgzV3Q0uw5YmJm/iYjjgVOBVw2qBkmS1ExtugXoR4BtgZOBpwBzMnNWj2n2oDaYmd/EL3+WJEnNMPSsJEmS1DJV5Km9gJsz85bMfBg4Dziks0FmXp6ZvykfXg3sOMDtS5KkhmrTAODzgH/PzL/NzNszc03VBUmSJNWIWUmSJKk/VeSpHYDbOx6vKudN5Fjgkl4LIuK4iLh2gLVJkqQaa9MA4MPArVUX0c1wJUmSaqKWWUmSJKlBqshT0WNez1uMRsRrgIXAab2WZ+YZmblwgLVJkqQaa9MA4FKK2yLUiuFKkiTVxFJqmJUkSZIaZCnDz1OrgJ06Hu8I3NndKCL2B04CXpGZDw2pNkmSVGNtGgD8a2DXiDgxInp9OkqSJGmUmZUkSZL6U0WeWgYsiIj5EbE5cBSwqLNBROwJfJpi8O/uIdUlSZJqbrOqCxigdwM/AD4IvCEivgvc36NdZuaxg9hgRJwL7AtsFxGrgPdm5mcGsW5JkqQBG3pWkiRJapmh56nMXB0RbwYuA2YDZ2Xm8oh4P3BtZi6iuOXnVsCF5bjkysx8xSC2L0mSmqtNA4BjHT/PL6dekuILkfuWmUcPYj2SJElDMNbx81CykiRJUsuMdfw8tDyVmYuBxV3zTu74ef9BbUuSJLVHmwYAJwpdkiRJMitJkiT1yzwlSZIaozUDgJl5W9U1SJIk1ZVZSZIkqT/mKUmS1CSzqi5AkiRJkiRJkiRJ0uA4AChJkiRJkiRJkiS1SGtuARoRt0yxaWbm02a0GEmSpJoxK0mSJPXHPCVJkpqkNQOAFFczZo/5WwPblD/fCTwytIokSZLqw6wkSZLUH/OUJElqjNYMAGbmvImWRcTTgU8Ac4EDh1WTJElSXZiVJEmS+mOekiRJTTIS3wGYmTcDrwR2AN5bcTmSJEm1YlaSJEnqj3lKkiTVzUgMAAJk5oPAEuDoqmuRJEmqG7OSJElSf8xTkiSpTkZmALC0GnhS1UVIkiTVlFlJkiSpP+YpSZJUCyMzABgR2wGHAbdXXYskSVLdmJUkSZL6Y56SJEl1slnVBQxKRJw8waLNgJ2AQ4CtgXcOrShJaom3Pmf2eo93fEz0nP+xZWuGVpOkTWNWkiRJ6o95SpIkNUlrBgCBUyZZ/gDwt5l56hBqkaRWW/WrrLoESZvulEmWm5UkSZI27pRJlpunJElSbbRpAHC/CeavBe4FfpSZq4dYjyS1hlf2Sa1gVpIkSeqPeUqSJDVGawYAM/MbVdcgSZJUV2YlSZKk/pinJElSk8yqugBJkiRJkiRJkiRJg9OqAcCI2CciLo6IuyPikYhY02PyVgySJGkkDTsrRcRBEXFDRNwcESf2WP72iFgREddHxH9FxFMGtW1JkqSZ4HtPkiSpKVpzC9CIeBnwJWA2sBK4ATBwSZIkMfysFBGzgU8BBwCrgGURsSgzV3Q0uw5YmJm/iYjjgVOBV81UTZIkSf3wvSdJktQkrRkABE4BHgFelplfrbgWSZKkujmF4WalvYCbM/MWgIg4DzgEWDcAmJmXd7S/GnjNEOqSJEmarlPwvSdJktQQbboF6LOA8w1gkiRJPQ07K+0A3N7xeFU5byLHApdMtDAijouIawdUmyRJ0nT43pMkSWqMNg0A/hq4p+oiJEmSamrYWSl6zMueDSNeAywETptoZZl5RmYuHFBtkiRJ0+F7T5IkqTHadAvQ/wKeV3URkiRJNTXsrLQK2Knj8Y7And2NImJ/4CRgn8x8aEi1SZIkTYfvPUmSpMZo0xWAfwM8LSLeHRG9PnEuSZI0yoadlZYBCyJifkRsDhwFLOpsEBF7Ap8GXpGZdw+hJkmSpH743pMkSWqMNl0B+F5gOfA+4PUR8V3gvh7tMjOPHWplkiRJ1RtqVsrM1RHxZuAyYDZwVmYuj4j3A9dm5iKKW35uBVxYvoe2MjNf0e+2JUmSZojvPUmSpMZo0wDgWMfP88qplwQMYZIkadSMdfw8jyFkpcxcDCzumndyx8/7D2I7kiRJQzLW8fM8fO9JkiTVWJsGAOdXXYAkSVKNmZUkSZL6Y56SJEmN0ZoBwMy8reoaJEmS6sqsJEmS1B/zlCRJapJZVRcgSZIkSZIkSZIkaXAcAJQkSZIkSZIkSZJaxAFASZIkSZIkSZIkqUUcAJQkSZIkSZIkSZJaxAFASZIkSZIkSZIkqUUcAJQkSZIkSZIkSZJaxAFASZIkSZIkSZIkqUUcAJQkSZIkSZIkSZJaxAFASZIkSZIkSZIkqUUcAJQkSZIkSZIkSZJaxAFASZIkSZIkSZIkqUUcAJQkSZIkSZIkSZJaxAFASZIkSZIkSZIkqUUcAJQkSZIkSZIkSZJaxAFASZIkSZIkSZIkqUUcAJQkSZIkSZIkSZJaxAFASZIkSZIkSZIkqUUcAJQkSZIkSZIkSZJaxAFASZIkSZIkSZIkqUUcAJQkSZIkSZIkSZJaxAFASZIkSZIkSZIkqUUcAJRUO1/+8pfZbbfdeNSjHsWuu+7K+eefX3VJkiRJjWKeajf7t73sW0mShsNzbnvZt7/jAKCkWrniiiv44z/+Y/bbbz8uueQSXvayl3H00Ufz1a9+terSJEmSGsE81W72b3vZt5IkDYfn3Payb9cXmVl1DSPh9NNPT4Djjz9+RtZ/zNWPzMh6NbkvPndO1SW0yoEHHsgjjzzC17/+9XXzXvrSl/LAAw9wxRVXVFjZ4MWhn6i6hJGVX3rLjK6/yX371ufMBuBjy9ZUXMn0zWD/xkytWFM3k5nKPFUd89RgjVKeGkX2b3uNUN+aqSrme1TtNdOZqql9u9tNFwPw/QUvr7iS6TMvD9YInXNHzgj17ZTylFcASlrP2NgYCxcuZMmSJey+++7MnTuXvffem+XLl8/4th966CEuv/xyjjzyyPXmH3XUUVx11VXcf//9M16DJElSv8xT7Wb/tpd9K0nScHjObS/7tl4cAJS0gZUrV3LCCSdw0kknce6553L33Xdz5JFHMtkVw6tXr5502tg6fvzjH/PII4+w8847rzd/l112Ye3atdx4440D+f0kSZJmmnmq3ezf9rJvJUkaDs+57WXf1sdmVRcgqX7uuecerrzyShYsWADA2rVrOeyww7jhhhs2eAHtNGfO5LcjOPvssxkbG+u57N577wVgm222WW/+4x73uPWWS5Ik1Z15qt3s3/aybyVJGg7Pue1l39aHA4CSNjBv3rx1L9AAu+66KwCrVq3a6Iv0smXLJl33/PnzJ20Tsf4tjMc/2dE9X5Ikqa7MU+1m/7aXfStJ0nB4zm0v+7Y+HACUtIHuT0lsvvnmADz44IMbfd4ee+wx6bpnz5494bLxT2Pcd999680ff9xdlyRJUl2Zp9rN/m0v+1aSpOHwnNte9m19+B2AkgZmzpw5k06f/exnJ3z+0572NObMmcOPfvSj9eb/6Ec/YtasWTzjGc+Y6V9BkiSpUuapdrN/28u+lSRpODzntpd9O3heAShpYPq9THuLLbZgv/3248ILL+SNb3zjuvnnn38+z3ve89h6660HUqckSVJdmafazf5tL/tWkqTh8JzbXvbt4DkAKGlgFi5c2Pc63vOe97Dvvvvy1re+lUMPPZTFixezePFiLr300gFUKEmSVG/mqXazf9vLvpUkaTg857aXfTt43gK0DxFxUETcEBE3R8SJVdcjtcHee+/NRRddxNe+9jUOPPBAFi1axBe/+EVe8pKXVF2aJGkTTZaVImKLiDi/XH5NRMwbfpVS+5in2s3+bS/7VhMxU0nSYHnObS/7dn2RmVXX0EgRMRu4ETgAWAUsA47OzBW92p9++ukJcPzxx89IPcdc/ciMrFeT++Jz51RdghoqDv1E1SWMrPzSW2Z0/U3u27c+p/gy5Y8tW1NxJdM3g/0bM7XiNppKVoqINwG7Z+afRcRRwGGZ+aqNrXcmM5V5qjrmKUkaKWaqTTATmcr3qNprpjNVU/t2t5suBuD7C15ecSXTZ16W1GVKecoBwGmKiOcBp2TmgeXjdwJk5t/3aj8eriRJUnMdf/zxvmE1RVPJShFxWdnmqojYDLgLeEJuJKCaqSRJaj4z1dTNRKYyT0mS1HxTyVPeAnT6dgBu73i8qpy3nog4LiKuHVpVkiRJ9TCVrLSuTWauBu4Htu21MjOVJEkaUQPLVOYpSZJGy2ZVF9BgvUZXN/gEVWaeAZwx8+U0V0Rcm5n9f8Onasn+bS/7tr3sWw3IVLLSlPIUmKmmwmO3vezb9rJv28u+1QANLFOZpybnsdte9m272b/tZd/2xysAp28VsFPH4x2BOyuqRZIkqW6mkpXWtSlvV7U1cM9QqpMkSWoGM5UkSZoWBwCnbxmwICLmR8TmwFHAooprkiRJqoupZKVFwP8pfz4c+PrGvv9PkiRpBJmpJEnStHgL0GnKzNUR8WbgMmA2cFZmLq+4rKby9hPtZv+2l33bXvat+jZRVoqI9wPXZuYi4DPA/42Imyk+pX5UdRW3gsdue9m37WXftpd9q4EwUw2dx2572bftZv+2l33bh/ADQZIkSZIkSZIkSVJ7eAtQSZIkSZIkSZIkqUUcAJQkSZIkSZIkSZJaxAHAERYRp0RERsRNEyy/uVx+Sse8cyLi2hmua2m53YyIRyLi7oj4r4j484jYoqvtvh1tx9vfEhGnRsTcrra3drR7KCLujIjFEfHaiKjNsVDXfqmLUd0/OvaLy3osuygillZQVk+j1EdN6pe6GKX9Q6OhruftUT/W6tovdTGq+0eTztuj1EdN6pe6GKX9Q6OjrufuUT/e6tovdTGq+0eTzt2j1EdN6pe6GIX9vX4WTgAAIABJREFUY7Nhbky19CAwPyIWZua6k3NEPAd4Srm80weARw+hrsuBd1EMUj8B2Bf4B+DYiHhRZt7X1f7VwC3AHODZwAeBbYDjutp9Efgnii/OfjJwIMWXZb86Il6RmQ/PyG+z6eraL3UxyvvHSyLiOZm5rOpCJjFqfdSUfqmLUds/1H51PW+P+rFW136pi1HeP5py3h61PmpKv9TFqO0fGg11PXeP+vFW136pi1HeP5py7h61PmpKv9RFu/ePzHQa0Qk4BfgF8FXgH7uWfRi4rFx+ypDrWgpc1GP+7sCvgbM75u0LJPCsrranAnd3zbu1+/cs5x8ArAbeW3Wf1Llf6jKN6v5R7he/BL4HfKlr2UXA0qr7ZhT7qEn9UpdplPYPp9GY6nreHvVjra79UpdpVPePJp23R6mPmtQvdZlGaf9wGp2prufuUT/e6tovdZlGdf9o0rl7lPqoSf1Sl2kU9o/KLxlWLZwHHBkRAVD+e2Q5fz3RdRl/RGwTEWeWl7E+GBErI+JfO5bvGBEXlJfP/jYifhwRH5hOkZl5PfBJilHyx07S/FcUo/BTWe8SihfB46dT1wyqVb+Uly7/Y0S8LSJWRcS9EXFeRGzT1W5+RHwpIh6IiF9FxFci4uldbTIi/jIi/i4ifl7W8anuS6s3xYjsHwn8HfCKiNhtYw0jYo8oLln/TdlXX4iI7TuWzyv74ciI+HRE3F/26/ui41J0j+EpaUS/eAzX4hhWu9XqvD2RETzWatUvvhbXYv9oxHl7wuLb20eN6BeP4Vocw2q/Wp27JzKCx1ut+sXX41rsH404d09YfHv7qBH94jE8vP3DAUAB/DuwPbB3+fiFFJe7/scUnvuR8nlvo7iM9V0ULzTjPgfsRHEJ7MEUl8RO++ADlvC7y2s7zY6IzSLi0RHxAuCNU6y/c73bR8S8PmobtDr2y5HAi8vn/Q3wcoqTCgDlC+t/AbsAbwDGgPnANyLi8V3regfwe8BrgNMo+uwvp1DDxozC/nEhcCNw0kQNIuIJFJ9g2RI4BvgLYB9gSURs3tX8VIpPtBwOfB44ufx5nMfw1DSlXzyGpZlTx/P2REbpWKtjv/haXL2mnLcn0tY+akq/eAxLM6uO5+6JjNLxVsd+8fW4ek05d0+krX3UlH7xGB4CvwNQZOZ9EXEpcBTwrfLfS8v5kz19L+BTmXl+x7zPdy0/OjO/Uj5e2me5q8p/t++a/92ux1ewaQd553pv3fSyBq+m/fIIcGhmrgaIiF3Lut5ULn8d8L+AZ2TmLWWbayjui/xG4O871nVrZo6VP19WvmC+kuKkMl2t3z8yc21EfAj4TEScnJk39mj2jvLfAzPzAYCIuBG4Bvhj4NyOtt/MzPH2SyLiIIp+uKCc5zE8BQ3qF4/hmrzGq31qet6eyMgcazXtF1+LzVNL+/wVWtlHDeoXj+GavMarnWp67p7IyBxvNe0XX4/NVEv7/BVa2UcN6heP4SHsH14BqHHnAYeXI+uH0+MS/gl8FzghIt4UEc+YYPnfR8RYRPyvAdQ5Uao4CngO8AfA0cB2wJei43Lkaa63anXrl8vHX5RLK4AndnwyZC/gO+MvygCZuQq4kt99SmzcV7serwB2HH9QfpJi3TTF+kZl//g8sBJ45wTL9wK+On4CB8jMb1OcVDapH/AY3hRN6BePYWlm1e28PZFRO9bq1i++FtdDE87bE2lzHzWhXzyGpZlXt3P3REbteKtbv/h6XA9NOHdPpM191IR+8RgeAgcANW4RsBXFJbpzga9svPk6bwa+RHHp7w0RcVNEHNWx/FXAtcBHgdsi4rsR8eI+6tyh/PdnXfOXZ+a1mfntzDyP4nLfFwEv7XO9Vatbv9zX9fhhihet8RfmJ9P7b/gzoPvS7F7relTH40e6pqkYif2jPDmeCrwmIp7So8kg+8FjeIoa0i8ew9LMqtt5eyKjdqzVrV98La6Bhpy3J9LaPmpIv3gMSzOvbufuiYza8Va3fvH1uAYacu6eSGv7qCH94jE8BA4ACoDM/B/gYor7cX+lfDyV592XmW/JzCcB/5viMuEvRHHJLpl5R3n57bbA84C7gEURse00S30JxUH6/yZpt6L8d5dNWO9dmXnrNOuaEQ3ql3E/BZ7YY/72wD2buK7ndE1TMUr7x1nA3RT3yO42sH7wGN5kTe2XgdeIx7BGUIPO2yN1rDWoX8b5Wjw8TT1vt72PmtovA68Rj2GNqAadu0fqeGtQv4zz9Xh4mnrubnsfNbVfBl4jI3wMOwCoTqdTfHrnX6bz5My8HjiBYr/auWvZ2sy8GngfxZeL9vrkwUZFxO7AnwOfz8xfTdL8WeW/t09hvQdQ3Lrg9E2taUhq3S9drgF+PyLmj8+IiB2A51PcD3lT6r62c5qs/ajtH5n5EPCPwOspPhHT6RrgwIh4zPiMiHgOMI9N7IeO7XkMT0ET+6VHjR7DUn9qfd4e4WOt1v3SxdfiIWnieXsU+qiJ/dKjRo9hqX+1PneP8PFW637p4uvxkDTx3D0KfdTEfulRo8dwn6Z6v1ONgMxcyiZ+eWpEXAH8B/ADIIE3AP8DfDsitgYuAz4H3AhsQfEFo3cBP5xk1Y+PiOdSBIJtgf3Kdd8IvL1H+90jYquy/VOB91Dc5/iSrnZPLtc7G3gScCAwBixh/S8OrY2a9ctkzqH4VMklEXEysAY4BfgF8Ok+193J/aPwaeBdFCe+b3TM/whwPMWX3v4DxS06PgR8H/i3qa7cY3ja6twvkzkHj2GpLzU7b3uslWrWL5M5B1+Lh6nO5+1R7qM698tkzsFjWOpbzc7dHm+lmvXLZM7B1+NhqvO5e5T7qM79Mplz8BjumwOA6tdVFDvuPIqD8Drg4MxcFcWXAn8f+EtgJ+A3wNXASzLzt5Osd79y3asp7uH7feBE4Mzy0wvdvlD+m8CdwLeAd2fm/V3tjimnR4BfUnxJ6bHAFzJz7dR+5UaYqX7ZqMx8KCL2pziJfIbivs1LgVdm5qZemr0x7h9AZv4mIj5Kcf/9zvk/j4j9gA8D51Lc93ox8LbMfHgTNvEgHsObrOb9MlntHsNSNcxT9WSeWl8r94+an7dHto9q3i+T1e4xLFXHTFVPZqr1tXL/qPm5e2T7qOb9MlntHsMDEJk5rG1JkiRJkiRJkiRJmmF+B6AkSZIkSZIkSZLUIg4ASpIkSZIkSZIkSS3iAKAkSZIkSZIkSZLUIg4ASpIkSZIkSZIkSS3iAKAkSZIkSZIkSZLUIg4ASpIkSZIkSZIkSS3iAKAkSZIkSZIkSZLUIg4ASpIkSZIkSZIkSS3iAKAkSZIkSZIkSZLUIg4ASpIkSZIkSZIkSS3iAKAkSZIkSZIkSZLUIg4ASpIkSZIkSZIkSS3iAKBUkYi4IiJWV12HqhER+0dERsS7q66lCSJiVUTcXHUd/YiIp5d9fmbVtUhSW5inRpt5atOYpyRJEzFTjTYz1aYxU6lJHADUyIiIL5YvbMdPoe2Ssu2hw6htmPp5gW/DCa5pOkLYpkw7ls9d1TV/bUTcFxFXRsSfRcTsrm1dMcl6N+tqv01EfKx83p0R8WBE3B0R346It0TElsP8W0mSZp55qmCeahbzlCSpbsxUBTNVs5ippObZbPImUmucARwNvAE4faJGETEPeDHwU+DiGaznGODRM7h+tcMtwPu65j0e+AvgXuATPZ7zQNfjj5bzZgNPBV4JPB94EXBkV9sE3j9BLWu7Hm9HcTwtA/4T+DmwNbA/8HHgTyPi+Zn56wnWN2puA3YB7qu6EEnqg3lKTWSeag/zlKS2MFOpicxU7WGmGhEOAGpkZObSiLgR2DMinp2Z35mg6bFAAGdn5ozd/iAzV87UutUemXkLcErnvIh4OkW4uiczT+nxtG4fycxVHc/fHbgGOCIiXpCZV3a0XTvFdQL8BNi6+ziJiAC+CBwFHAd8ZIrra7XMfAT4UdV1SFI/zFNqIvNUe5inJLWFmUpNZKZqDzPV6PAWoBo1/1r++4ZeC8vLzV9H8QmTMzvm7xAR742I/46IuyLi4Yi4IyK+EBE791jPulsYRMQzI+LCiPh5eXn73mWbDe6vHhFbRMRfRMQlEXFbRDwUEfeUt3s4cIKaV0XEzRExNyI+HBEry+fdFBF/VZ7oxtv+LXBT+fDYrkvnXzPRH238En9gB+BpXc87MyK2jYjfRsQNndvrWselZfv/3eNvtGtELCp/1/+JiG9FxP4bqefVEbG0vFXAgxGxIiLeFRGbT/ScHut4ZkT8Q0RcW/bNQxFxa0R8OiJ2mOhvEBHvjog9yz66v6x3aUT8wQTbeXJEnF3eduC3EXFdRLx2qnXOhMy8HvhW+XCvPtazptd/QDIzgYvKhwumur4ovKXsz4fKY+wTEfHYjTznUWXf/6D8+z4QEd+MiMN7tO3c5xZExL+X+9wD5f65a9nuiWWbn5b717cjYp8e65v260LX/M+X83eKiDeVv8uD5Tr/ZWO/vyRVxDxlnhpfh3mqYJ4yT0nSdJipzFTj6zBTFcxUZioNmFcAatR8FvggcExEvCMzf9O1/GCKALEkM3/SMX8/4K+By4HrgP+hOGkcCfxRFJeQ/6DH9p4BfBtYAXwe2BL41UbqewLwMeC/gSUUl6s/GXgFcElEvD4zz+nxvM2BrwFPBBYDa4DDgNOALcrfGeDrwGMpPplzHbCoYx3Xb6Su8Uv83w6sZv1L+r+Tmb+MiAuB1wL7Uvyd1omIpwAHANdk5ve61v104Crgu8CnKf7+RwKXRcSRmflvXev6LPAnwEqKE/j9FLcK+CDwoog4MDPXbOR3GXcExSd/LgeuBB4BdqMI3i+PiIWZ+dMez9sLOAm4giKsPwX4Y+DrEbFHZo6HVyLiiRR9OQ/4ZvnzDuXzLptCjTNpPATnBgsiXkVxG4aHKPbdyzPzoU1c/x+V/25sv+r2SeBNwJ0U+8Jq4FCKv/kc4MGuOregOE72Luv8JDCXom8vjIgPZObJPbbzVIpPl/0AOBuYT3G8LI2I55frvBc4H9iW4lNil0bEgs5PqTH914WJfJjiOLmYYv94MfBG4GnlfEmqC/OUeWqceapgnjJPSdJ0mKnMVOPMVAUzlZlKg5aZTk4jNVG8YCYw1mPZl8tlh3fN3x7Yqkf7PSleUL/SNf/p5XoSeP8EdVwBrO6a9yhghx5ttwF+SBG2tuhatqrczleAR3fMfxLFPbXvAWb3qO3MafztVgE3T7DsueV6z+ux7G/LZa+b4G/0oa72f0BxYv1l598d+NOy/QXAo7qe84Fy2Z9P8XfZsftvWc4/mOI+4v/UNX//jnpf07Xsz8v5n+iaf1Y5/7QJfr8E3j2Nfhj/2/Xsix77xo5d83cHflsue17XPpk9pruAwzaync0pbgFxCkXw/l75vCXA5lP8nf6wfM6NwOM65j+a4j8oG/y+wHvK+YuAzbr2/dvLfvyDCfa5v+la1/vK+fcAnwJmdSwb/8Rldz9O93XhzK75ny/n/6SzrygC5X+Xy569qfuJk5OT00xOmKfMU2mewjxlnnJycnLqc8JMZaZKMxVmKjOV04xNlRfg5DTsieITCwlc0TX/yRSfsLkLmLMJ61sM/IbeAeaOiU4u9AhXk2znr8t1Pr9r/vgJdF6P53yhXLZzj9oGGq7K5d+h+ETOEzrmbUbxaZn7gC171PFLYG6PdY2fcF7dMe/75fof26P9ZhSfiPnvAewjK4Abu+aNh6ulPdpvQfGJtqu75v22/L0fs5Hfbxjh6iMUwecD5XbHg9UFXe3fAbwU+D2KoP9M4EPAwxRh8IAJtrMVGways3v160ZqPbt83mt7LBv/23eHq5+Uf/cFPZ7zxvI5Z/T6u9ERnsplTy2XPdBdN0XIWUPxqctBvC5MFK7GeqznDeWyP+t3v3ZycnIa5IR5yjw1+e9pnjJPdS4zTzk5OTn1mDBTmakm/z3NVGaqzmVmKqdNmrwFqEbR14EfAy+IiF0y84fl/NdRnKDPyeKLUNcTEa+geMH+fYpLrruPn8dTfPqp03cz8+FNKS4idgNOoLhk/PcoTtKdNrj3N/DLzLy1x/zby38ftyk19OF04AyKv+Wp5bw/ogiu/5Qb3s4C4P9l5v/0mL8UeDXFJ1W+EBGPAZ4F/Ax4e/S+jfuDwC5TKTSKFbwW+D8UnzZ6HDC7o0mvWgGu7Z6RmQ9FxM9Z/++8K0VAuSoze91SYynF7zcMbyv/TYrbe1xHcUL/dGejzPxw1/NuAE6MiLuAjwJ/R/GJKbqe92vK26NT7LMvAf4eWBYRB+XUvkz82eW/3+ix7JsUn5RaJyIeR3Hbituy45YWHb5e/rtnj2XXZebarnl3lv/e0L0/ZuYjZf/u2L2iab4uTGSDfYvhH8OSNFXmqZljnvod85R5CsxTktrNTDVzzFS/Y6YyU4GZaiQ5AKiRk5njX3D69xSX67+jPCm8nvKTD93PiYi3U9z7+B6K+5jfxu8+ofJKivtyd4cgKD6pNWUR8YJy/bOA/6K43cOvKE4sz6YIKr22c98Eqxz/8tvZEywftC9Q3NP9uIg4LTOT4sQDRejq5WcTzB//221d/vv48t/tgfdupIYNvvB3Ap8A3kxxUr2U4pNw4/fvfj1FSOhlY3/rzr/zeN2T/X7DsFOuf1/wTXUGxf7/+xGx5QQhmbK/7wDOjoibKL7E+RMU90ifzIR/r8x8OCLunaB9r3vgd87fpsey+3vMW72RZePL53TO6ON1YSK99q1hH8OSNCXmqRllnvod85R5yjwlqdXMVDPKTPU7ZiozlZlqRDkAqFF1NvB+4E8i4p3ACym+xPTrmXlzZ8OImENxafqdFPc4/lnX8hduZDu5iXW9h+ITOS/MzCu6tvMefveltbWUmb+JiM9RfIHzi8sT7AHAlTnxF81uP8H8J5X/3t/177LM3KufOiPiyRT3RP//2bvzMLuKMvHj3zesAYUEWeQHjkGJxIAKggEUAdnBBRRlcYEgDsqo4ziC4OgMuO+iuKDI7sI6CqgsBgVGFCQo+yaLCGELEFZZE97fH3U6ffrk3l5vutPd38/znKfvqVN1Tt3bN33eVNWpuhrYohodVD/+vqGcv9JV377e3xKv+r3+E3ghZZHwdiPP6mUuiYjHKQtu90f98+oxGisilqWMLnqkRf52n+OajXwdNcS/C5I0VhhPLQbGUz0YTxlPSdJ4YEy1GBhT9WBMZUylcWrCSFdAGgnVH8KzgVUpIz8+UB1qNQJoDcpN5ZIWf0BXovXj24O1LjC3GVhVturQNRZUPwczWmNBP8r9oPr5Qcrc0BNoPMbfsHFErNgifevq55UAmfkI5XH/V0VEqxEzA/FyIIDzWwRWL6U8tj9UN1BGa722mhqiaesOXGNYRMT6lH8Dj1JGEvWnzCTKvOv9He321+pnq+/5ljTuV5n5MGU0079ExMtalHlT47ydNpx/FyRpiWQ8BRhPGU/1k/FUS8ZTkoQxVfXTmMqYql+MqVoyplJbdgBqPPtx9fMTwNuBB4Fftsh3L+Um+bp6EFCN+PgunZ33+A5gtepmtlBEfJCyMHQndN0c/2UQZR8CVo+Ito+MZ+ZNlLnDdwUOqK53ei/nXIUyqmyhiNgU2IuyYPJZtUPfoow+OzYiVqYhIlaJiP7c1O6ofr4xIhYGi1UQdDQd+NuYmc8AJ1OmAfifRj273t8SIyJeXo06a6avDhxX7f68Pi95RLy2VaBbfT++Rwlgf9PPKhxf/fzv+jkjYiJlXvdWjqP8rr4eERNqZVYHPl3LszgM598FSVqSGU8NnPFUPxlPGU8tputK0pLImGrgjKn6yZjKmGoxXVejgFOAajz7LfB3oOtR/e+1Wgw5MxdExPeAg4BrI+JsypzJ21BunBfTuZFPR1CCqD9FxGnAY1X9Ngf+F9h9qBfIzEcj4grgTRHxU+BvlPnbz+xlCoQuv6OMGjkvIv4APEtZrLZ5A/0BZfTQasARmfk07V0MHBgRmwN/pCwgvSflxnxAffRTZh4dERtTgratIuK3lEfxVwFeRpkm48eUedN7+wzmRMQZwDuBv0bEBZTf5Q7AE8C1lAWSh+pQyvfkoIiY0Xh/vwHe1oFrdMqbgB9GxMXArZTA9qXALsBKwOWU91P3fuD9EXEhZaTTY3QvsLwGcCPwyf5cPDP/LyKOAg4Erq9+P/Mpox8fAOa2KPZVYCfKXOZXR8S5wIrAuyjfvS9l5mX9evcDNMx/FyRpSWY8ZTxlPNXNeGoAjKckqQdjKmMqY6puxlQDYEyl3tgBqHGrWmj5WOALVdKPe8n+Kcof9/dTpg14BJhFGcHx5Q7W6TcRsWt13r0oN5fLKYHKNDoQXFXeQxmptAvwbkogcwfQV3D1WcqN9i2UQGYp4FgWHUHzS8rNeTLtF1bucitlrvMvU26sywGzgc9m5gXNzJn5wYj4DeX3sD1lAd2HKDf3rwE/7eN6XWZW135Xdf25lCk3/hv4VT/P0avMnFsFjV+ifGYzgJso007cy5IVXF1BWSB7Y0oAvRIlWLoaOBU4OjOfa5Q5lTLf+mbA6ynTKTxGmVria8BRmfnUAOrwEcoUGh+qtgeBX1D+PdzQzJyZz0TEtpQRknsD/w48B1wFfDQzTx3AtQdjWP4uSNKSzHjKeArjqTrjqYEznpIkjKkwppqJMVWdMdXAGVOppcgc6PqvktS7iJhKuUlenJlvapNnXeAW4NjM/ECrPJIkSeOV8ZQkSdLQGVNJGs9cA1DS4nAwZcTW90a6IpIkSaOU8ZQkSdLQGVNJGrecAlRSR0TESymPuK8H7Av8ldYLVkuSJKkF4ylJkqShM6aSpMIOQEmdMpUyp/Q/gfOBAzPz+ZGtkiRJ0qhiPCVJkjR0xlSShGsASpIkSZIkSZIkSWOKawBKkiRJkiRJkiRJY4gdgJIkSZIkSZIkSdIYYgegJEmSJEmSJEmSNIbYAShJkiRJkiRJkiSNIXYASpIkSZIkSZIkSWOIHYCSJEmSJEmSJEnSGGIHoCRJkiRJkiRJkjSG2AEoSZIkSZIkSZIkjSF2AEqSJEmSJEmSJEljiB2AkiRJkiRJkiRJ0hhiB6AktRERh0dEdm0jXR9JkqQlWT1uiojDB1Duolq5izpQD2M4SZIkSeOeHYCS+i0iVomIgyLi/Ii4JyKejoh5EXFtRPw8It4REcuPdD3HkohYISL2jIgfRMTlEXFXRDwTEU9Un/u3IuIlvZRfMSI+FRFXRMRjEfFURPwtIr7brlxEbBoRX42IWRHxUKMxb2abMoc38vW2Hd6ZT0eSpL5FxNkt7kUbjHS9ljQt7uVbt8iztff0zomIV0fEJyPi1xFxYxVXPxsR90XEORGxZx/lN4uIUyPi7io+fLCK394dEdEi/zIR8R8RcVJEXBcR82u/yzt6uU5/Yzw7WyVpjImIZSPi6sbf++1b5JsQEX9o5Ov1PjbWRcRaEfGhqr3smoi4PyKei4iHI+KPEfGJiJjYS/mXRsT3I+KWqi3n0YiYHRGHRMQKbcrsUbUfXVaVqf8+1u5HnWdExAkRcVtEPFm1Pf09Is6MiPcM8nN4TUQcFxG3Vud8JCJujohTIuIdgzlndd4dI+LLEXFxVd8nqvd8exXrzOilbETEeyPigip+eiYi5lR1alkuIv4lIv47In4ZEXc2Pttj+qjrKyLiqCjtcU9GxD+rWOyrEbHaYD8DqT+WHukKSBodImJf4LvACxuHlgMmAxsAewP7AScMa+XGtunAKS3Sl6V85hsAH4iIXTLzknqGiFgLuACY1ig7tdr2jYi3ZeZFjeN7Ax/rQN3bsXFIkjQsIuLFwM4tDu0PfHyYqzMeHFx7/acRq8Xo8SXgzS3S16B8b3eOiHcBe2Tm8/UMEfEJ4BuNci8Ctqu23SNir8x8rnZ8ReCITlVekjT2ZeazEbEPcDmlHQLgmIjYIDMfr2X9d2CL2v6pmXnqcNVzCfU+4Mst0icBr6+290fE1pn5QD1DRGwLnAm8oJa8PLBJte0TEdtl5r2Nc/8PsP5AKxoRE4CvAZ9ocXhFYAqwKvCzAZ53O+AcYJla8kRgZeAVlL6JXwy0vpWzKG2STetU23sj4t8z83uNOi0LnAbs2ii3FrAnsEdE/GdmfrtxfAbwuYFWMiLeCfyE8vurW7/aZkbEjpl51UDPLfWHHYCS+hQRHwWObCRfBFwCPAWsDWwDrDe8NRtXngJ+C1xJCZbeQenEg9Ip+5OIeHlX41BEBHAy3Z1/zwBHA49SOmnXqsqdERHTM3Nu43oPVNf6O/DBftTvt8ATbY7tDmxW2/9NP84nSVIn7Evr//O8NyIOycxnh7tCY1lmNjuk1D83AbOAuZTBX3sAS1XHdqd8j4/vylw1ptU/6ysojWCvAd5Zpb0D+G9KQ2DdfOBG4K9V/g37Ub+D26S/sHF+YzxJGoMy8+ooT/1/qUr6F8p96IMAEfFy4Iu1IvcC/zacdVzC3Q+cC9wKrEnpGFypOjYd+AplcBoAEbEmcDrdnX9zKHHAZOAASkfsdEpn3DaNayWlHeevlA6nVgONWjmMnp1/l1Pa/B6idFi+HHiuRbm+fIzuzr8FwI8p72d1YONBnrMugT9QBp49A7wJ2LI6FsAREXFOZt5eK/NZenb+nQ5cA+xW1SmAb0XEVS0GzP8TuJry+e5D9++xpYj4F+Akujv/7qf7oYn9KJ/D6sAvIuJVmfnPvt+yNECZ6ebm5tZ2o4zIeY5yU03gSWCnNnm3B7ZspC0LfAi4EHiwOtdDwMXAR4Dl2pxrdeDzwF8onVbPAvdQGjfe2iL/1rU6JuWm/xFKA8fTwN+Aj1R5lwIOpQRfz1Q/PwlE45yHN865TFXu5uqcc4BvAyu3qM8OVV3vrq7xJHBn9Tl8DXhFPz//6VU9JjfSlwduaNRvg9rx7RrHDqwdWx94vnbs841zr1B7PaVxnpkD/P5MpHQmdpW/YKS/025ubm6UoPu1AAAgAElEQVRu42er7tld96CbG/e0d/ZSbrkqfrmA0inzbBXH/AX4JrBsla/X+yTlP/hdx+5oHLujduwEYF1KQ87cKna4Dti3Rd02BH4AXArcRWmIeIYSJ50D7DnIz6oZ92zdIs/WjTyHN473dmwC8FHgekocdQ9wFGU0+UW1chc1yr2fMhvC9bXfxROUGO9HwCv7ei8tjq9LGfk9r/r8LqN0tDXfX6vPYBfKSPW7q7o8BlxFaRRdY4Cf+aeBrVqk79OoxxmN45fUjv0DWL527LTasSeAlWrHopG37fezn/U/uFHPLUb637ybm5ub2+LZKO0of2r83d+2urdc3EjfpVF2JUpbyqXAI9X9837gfOA9LNoW02zP2KJxvH4fvKCWvnSj3GcoT8v9CniYMrD6z7Rv03oFcEYtPrgUeHtf9enlM3s38F5gmUb6KymxW9f57msc/0rt2PPAerVj/96oy1aNsvX2nA808q7dpp5TKAOEuvJ9oIPfm3qscU2L4xOGcO4fAS9vkX5S431/qHZsMqV9ruvYz2rHJlLa+bqOXdg473L1+jbyHtOmjv/VqMtmtWNvbBw7cCDv382tv9uIV8DNzW3J3oDvN25IBw2g7KqUhrLsZbsSWK1RblNKA09v5U5q3Hi3bhyf3abc4ZRGm1bHDmvU4/DG8V+1KXc18IJauff0UfdkgB1pbT7fbzTOuXEvv7cXNcpeVzt2fS/XmDKUelMa+urltxvp77Sbm5ub2/jYWvynek/Kk1Zd++e2Kbd24z7ZaptU5e31Pkn/OwCvpgx4anWtfRvlPtKPOOOHg/i8mnHP1i3ybN3Ic3jjeG/HjmtT11spnXtd+xc1yl3Rx3t9ulnX5ntpHNuA0rDX6lxntvsMKB2Yx/dRlweAzTvw3X1B47y/qh1brXHsu42y72wcf1cv12n7/exHHZeldIJ2lb9kpP/Nu7m5ubkt3o0ygOaf9XsHi3Zw/KhR5hXA7X3cP8+j5wCVTnUAXkrpbGxebwGLDl5/NaWTsFX9zuqtPoP8LK+qne/hxrH6oLUrG8de3KjLd3q5Rn87AD9by3ML5UnAKykDieZROmp3HOT7fBU9B6B/eBi+p7s13vfHasf2bhzbtVH2h43jk3u5Tn86AI9tnK/+PX9h41jL/5u4uQ11cwpQSX3ZtvY6KY03/fUT4LW1/fMpI6xfRxk9DWUU+88oT8wRESsBZ1MaN6AEZidRbqxvo0xVBGXKhJvonoKiaRNKEDmbEvSsWaUfVv08h9I5+aHatf4zIr6UPddKqXsz8HPgNuCtdE+Z9GrKPOD/We1/tFbmZsp0As9SGhRfCWze5vwD9cra68coI+G71KdzejQzH2qUvY3ueeGnRcRymflMh+oFQEQsTc8pJP6SmRd08hqSJPVi/9rrxynxxTRK5xDADhGxdmbO6cpUrX9yFj3XTrmRMm3TU5TOo13ovK5GpyMoo4//le4pIA8FTqzlfYYyLdOVlKcSn6B0GL2B0kEH8MGIOCYzrxhCnfaMiE0aaS8fzIkiYlfKNEdd7qfEd8tV6c01puseAH5N6Sh8mDKbxIspI/JfUp3je5TfTX+cQBn93eU8yue5LYuuBVN3MDCztn8d5buyBmWKzmUog9/OioipmfloP+vTyisb+3+uvW5O2XlbH/uvocSinbYP8P9q+19ZDNeQJC1BMvPWiPgk5b4L8FJ6Tv15O7U2gKpN4CzKemxdTqO05WxHWQMPYEfgW3R+2tDNKDMxnVzVda8qfQLlvv5/tbwnUaa67HIOZRDS9pS2qI6JiBWq+nT5c+PY1NqxHvf1zLwvIp4EVqiSXsPQ1dduXJee04yvSGmv2yEiDs3Mr/b3pBGxHiUGiVrydyNiYtamjY+IN9L9u7g8Mzcd6Bto6HQcddEQ6vJIY39DSrtoq7q8agjXkdqyA1BSX15Sez03M+f1p1BEvArYqZb0s8x8b+34iZSGA4DtI2KjzLyS0oCyeq3chzLzmKrMFynzcr+iOnZQRHw1Mxe0qMIsyrQTGRFzKFMDdDkvM99cnfNeyjRaUKalmAZc2+ZtHZaZn6/V5Vq6A7MPRMQnM3M+PRf2PTwzT6mfpOrkbC7+OyARsS89GyC/mplP1vZfVHv9WItT1NMmUBrC7htKnVrYi55BrQ1DkqRhEREvBN5VSzozM5+KiJ/T3QE4gdKh84Vavp3pOXjpbMpUoQsHB1VreXR6fY4Etq1iISLiaeA/qmPTIuKFmfk4QGb+GPhxRKxP6ThcldIp9htgBt0NQjtTGq4G60NDKNvbueYDb8zMWwAi4n8pU4e1lJk7R8TylEa8l1M6C++iTM/a1am4fkS8JDPv6q0SETGDsrZLl1Myc+/q2GeB3wNbtSg3ATiolnQb8LrMfLo6/ie6B8mtRpm29Ije6tJLHVemrNvc5T7KVKldXtSzxCJxXnN/1cHUozfV51FfF/A6XP9PksaLH1A6xHZopD9PmQnhiVraWyltLF0+m5mHw8L77kWUGRsA/jUiPp2ZD3ewro8Dm2bmfdU1XwC8pTo2oytTRLyenh1pP83M91XHPkeJU97QiQpFRFDu612djQsoA8q7rELPDrN27Tld8V4n7vNrNfafAY6hfH770z1o/ssRcUFm/qWvE0bEapSHALrahG4E1qPE31+PiFUz89DqWL3D8+bBvYWF190Q+FQt6dzMvKy2P9xx1Fl0PywA8MuIOIHyO96vkXcy0mIwYaQrIGnM2qKxf3xjv/kkYVcwVS/X9fQfANUTaj+vHZ/MoiN7uvw8M7N6fUfj2Mm117c0jvV2w104+r6qS71j74V0d0zWR5GdEBEXR8SxEXFIRGwLPJWZc3u5Tq8i4hP0/PxOAL7czNbmdbu0bJFnqD5Ze/03ytSrkiQNh73pbhiBMtsAVadTvVNsv6ohpsuWjfP8T3NmgMy8s5fZAgbr0q7Ov0qz8WNhfBIRG0bE1ZROl58DR1JGan+dnu957VqZnSLioBbb6xker6u9vrSr8w8gM/8P+Hu7ghHxMcoTgxdSGqOOoLzXZqPJ2vTtdY39Y2v1eJ4SU7WyHj0bgE7u6vyr/ITSsdllUI2EVefyH+geEf4w8ObGTA7NGK6v/cUR472d7rgXykC0xXEdSdISpvp7vz+LPtn0zcz8QyOtbbtQdd+tz3CwNGU5mE76ZVfnX6UeX9XbfnqLDxbQPj4YkIiYSHkCsmsw/PPA/pn5p3q2ZrFWp6q97sT9d9nG/kGZ+ZHM/BQ9n34Mymxc/fGfdHf+/Z3S4frh2vFDIuLHEbEUJW7v8vv+V7uniNie0qncNbPEXxvnhmGOo6o499u1pBdTZvc4hJ4PP0DpeJU6zg5ASX2ZU3u9ekSs0s9yzXzNp8ua+6s0fkKZB/3ZfpZrurv2unmOe2qv5zeO9fZ38f4+9rsCyP+ie4745SiNie+nPAF3AXBnRAy4YSgilomIYyiNfF31PJISMDaDkgdrr1dqcbp62vMsGrwPSUS8mZ7TF3y9CvAlSRoO76+9nku5/3apDyZ6Gd3TZsKicUXbjqk2mo0Gy/Wz3B2N/WYDwARY2HD0G8qTf32pX3svSqdZc2uO3q97U2ZGfQPe1I/rtlKfUqsZP7VLIyLeRmk0aRXLNPXns57U2L+3sd9uNoRe49pqBoh67NXfeHmhiNiMMhVpV/x0N2UNwr82sj7Y2G9+Ns395jTwnXBI7fU/6DkoTpI0xlXTp5/dSD6qRdbBtgs1LY74qt72M9j4oN8iYk3KYPF31uqyd2ae2Mj6ED07nVrFQPWp0ztxn28+dbmwE656eu6p2rF1+3nO7WuvT8/MJzLzh8D/1NI/QJkOc7tq/3FKW9qARcSBlGlbV66SLgK2aTEl+7DHUZn5ccrv/XeUtrenKetfH0bP6UXvWaSw1AF2AErqS73BLOi59klvmlOFvriP/XmNnwCTI6I5EqlduabeRuY3O/36a40+9h8BqAKb3SjrDu5GmSLpaLqDyBdTRor3W9Xx+lu61zOaD3w0Mz/WpmPtqtrrlarpF+rqa/jc1On1/+jZMHQPtSc5JakTIuK4iJgbEdf1I+8REXFVtf0tIjo66EFLlmpqzPoI8tWB+RGREZGUNWbq6msFNuOKdehd8x48sbE/lf5pxi3tRhu/kZ7rrh1BiUcmVB10D/TzesOt/m+uGT+1S4PutXqgTLu6M7BC9V7fPMR6wKIjr5txZpde49pqjaP6E4L9mjK/Vn5vyhOOXZ/DXyhTll3TIvtVjf1mQ1xzncarB1KXvkTENvR8UuIbVQeoJGl8acYqrWKXwbYLtY2vqifGXtafCtL/+Gqw8UG/RMRGwGyga23luZTOqdOaeaulXeozVfW4z1cdifUZHzpxn28ug9Pb7/Yp+qfeqbowRqqW1flu7Vh9vemvD3QK2IhYKiKOpExN27XU2XHAjm3WYx5oHNUqFhuwzPzfzNwuMydn5sTM3AD4Jj0H9f2xE9eSmuwAlNSXI+nZYfb5iGg5WjwitqsW74VFb1zNaZre39j/Y+MnwFJ0T41ARCwHvLt2/GHKPOLDZd9GXeqNUo9TTScRERtExHKZeX9mnpWZ38jMDwIfqeVfJyKac4+3FBGvoIyK2rpKeoSyvuH32haCMxv7e9TOtwEwvXbsl/2pR39FxOZ0z+MPcESLJzklaahOoOdas21l5sczc8PM3JDyH06nJB7b9u87Sw/vqNZdg57TeAMcXnXuLBQRa0XEMtVus8Fos1q+Hem53lwnNNch+Wlmzq3WPN6G7jVaesjMmc2n+art8A7Xr53ZtdebR8TCjtEqdmzX0Vp/v7dn5nmZ2dXwtFerAn24vLH/nlo9utaEbOVmeo4Y36tal7DL++hudIJ+NuBE8VnKU6ld5zsD2DIz725VJjMfAOpThb21UZc9aq+fpAwg66RDa68foDZNmiRJDW3bhar77r61Y/OBP1ev28ZXlHWFB/ykfR96iw+Wov8D4RcREbtRpvfuWmfvGmBGY9rPpnp7zqsior70zR6NvJ1oz/lVY3/hjA/V+sn1DsfZ9E+9rW7viKivsfgxFp3q83bgS/08d1fdVqLU/aNV0vPAwZm5fy9tUOdRnsDrUm8rW4HuNSIBLs7MAQ3qalPPRdYRrP4v8V16fpePGeq1pFaW7juLpPEsM2+OiIMpI8yh3PjPj4jfU4K5pyhrrmxLWR9lP+APmXlNRPyW7qml3lPd9C6jjPCpj9r+XW3dmxOBz9DdgHVUtT7NHMrc4/X1Rr5Vzcc+XD4bEdOA2yiLWddH9R9XG/38FeCN1Wd0J2Vaq5XoOff4M5RGmV5FxMspn1l9fvozgNc0AigoixtfX72+gBJkdnXEfTMi1qME0vWG0YfoOfqKqoO36/fWnP5gz6oDEeC2zGw1zUe9YegR4Eet3pskDUVm/l9ETKmnVX8zv0+5hzwJ/Gtm3tQoujdluhWNQdXMAfW1SeZSnqxqWg3Ypno9kTLA6CjgXOBKYKPq2G7A1RFxDuU7NQ3YlTIS/JHMfCwibqrSAd4XEWtR4qPeptccrObagD+LiJMpsw7MXAzX65Sj6e6wXxr4Q0ScRFlzpjkorO5muqeQelVEnEpZ+3Brun9//ZaZsyPiL3R3zO5fzZJwJWX6qZZTtGfm8xHxLbobptYFZkfEmZTvQr0B80EWXfu6ne/Q3WgFZTrNy4F/67k0JY9m5o9r+4cBs6rXL6F8nmdS1g58Zy3ft5qj3yPiG7Xd+qj7yY1jR2XmbY2yG9FzSq8jax2ykiQ1nU25l69X7R9WtancxKL33WNrT3/dADwBvKBWbhNgRQZx/+9LZl5ara/c1cbywYhYg/J03fbAoNZMjoi9KIN8um7qT1I67N7VuM8D/DAzn6hefxv4V0o7UACzIuK4av+AWpkLM/OixjU/TPfAqvqyLACfiYiua5ybmb+rXp9Fea9d7/+bVafjE/QczP8o/Y9xvkd3Z9pE4LKI+CllivONWfT3+DLKe2vVxrSIKB/gn4D1a8m/r44d1Mh+TWb+FiAzH46Ib9PdbrVX1cl7NWWN466ZNhL4bOOaU4EP1pJWrr2eUYujFmRmfVasH1bf+/+jzJC1OrAjPds3j8nM/nauSgOTmW5ubm59bpROoycoN8Hetpm1Ml0NKr3lvwZYo3GtzSmNJ72V+zmwVK3M1o3jW3fg2OGNY79vU5drgZVq5X7dj8/py/383Jv169dnX5VdmxJYt8v/eP399vK+220XtSg7nTLqqivPF0b6u+vm5jZ2N2AKcF1t/3fA1Or1psDvG/lfSpmOeanhqqPbsH8n3tm4V32qTb4VKINUuvLNrh17CaWTqbd74KRa/plt8syljGTv2r+jUYc7asdOaBxrnnNK7dhv2lzvt8Bd7c7Zj8+uef/fukWeZlxyeON4b8dObFPvOcDfWsUXlGmYHm1T7rh29W2+l0Y9NqBMMdbqnM0YbstauQmUKc17+148BLxhAJ/5RX2cr+V3pyp7UB9lfgEs06Jcf67X7vd/au34Y8Dkkf437+bm5uY2MhtlNo6WsUoj3zTKAJfe7jmzgImNcl9ok/dvlE7Frv0LamWWbuT9TC/nnN849hp6xoa9xQev7+dn1O49tNrWbpTdlt7b4K4H/l+La17Sz+s1P5up9IxNm9tjwHYD/I58EljQRz3+WXv9HLBDP8/d/F33th3TKLsspdOzXf7ngY+3uOZ2/bxe87t1Rh/5fwIsN9L/pt3G7uYUoJL6JTOPpTScHkJ5uuw+4FlKgHQ9cDKl0e3UWpkHKFM1fAS4mNLYMp8ydecllMf+Z2Tm/Y1rXUoZxfMlyvzcT1Tl7qM83v/2zHx3Du/Tf1DWnTkMuJXy3u+hPD23ZWY+Vsv3Tcr6QpdQngB8ihLI3Et5smCvzPzU4q5sloW5NwY+TemIfYLy5OGtlCdk1s/GaLEO+CTdo9ueooxsl6TFLiJeQBmde3pEXEV5+njNRra9gDNG4P6h4VN/yn0+bUYpZ1lf5ee1pE0i4tXVsbsoT0b9G2Xwz4PVuR6hjA7+NrWn+DPzBEqH3XWU+OABSkfRxiyeqcrfCXyDMoL6OUpjzZcpMyUsyd/t/YCPUz6TZykzJBxHWU/unlYFsjyB9kZK/PRPyud+GeW9Dmp94cy8jjJA4AzK7/Qp4ArK34efNbI/XCv3fGbuU137LEpc91xVr2uArwIbZOawrN+Smd+g+ptXq8s8ynf2vcDumdnbmtgDUj1hvXst6egc4Do9kqTxJ8tsHK+mtEtcTulImk+Jr2ZRnqLfMRd9ovy/KU9p3Ua5x91NaV/YlMWw5nFmXl2d+3/pjg9mU6aIPLWRfbHf/7I8nbc+ZW272yhtOY8DfwU+BbwuM1vGT4O83i2UmQQ+R4lr/knP9qNXZ+YFAzzn1yhtgj+ltI09S5l+8xbKwLDtKIPIu36fSwOnRcT0Rc/WOVmmB307ZcmhCynx03OUePRUYPPMPKL9GQbsZ5QnP/9Od7vcncApwPaZ+b7MfKaD15N6iMwc6TpIkiRJA1ZNAfrrzNygWgPi5sxsdvrV818JfDh7X3ND0hhWrbmS2T11e/3YmZQpXqE0UK7ayU40SZK0ZKqmkV/QaqBgRPya7mVsHgFWaxVHSNKSyDUAJUmSNOplWYvt7xHxrsw8vVoX4tXVaF6qdVAnA5eOaEUljbSXAxdGxM8p07g/QHlaeE/KSPQu37PzT5KkceMVwG+r+OA6uuODvem5Xt2Rdv5JGk18AlCSJEmjTkScTFmPbFXKVIKHUaa+O4ryn/VlgFMy83NV/sOB5TPz0FbnkzQ+RMQ0+p6a9RTgfTbwSZI0PkTEBpSBQb35GTDT+EDSaGIHoCRJkiRpXIiIlYHPAFsC6wCTKGsR3Qv8GTgxM88fuRpKkqThFhGTKesUbglMoTs+uIcSH5yQmbNGrIKSNEh2AA6To446KgEOPPDAGOm6SJIkjVbGVJIkSUNjPCVJ0vjgGoDDzx5XSZJGJxtIlizGVJIkjU7GVEsO4ylJkkanfsVTExZ3LSRJkiRJkiRJkiQNHzsAJUmSJEmSJEmSpDHEDkBJkiRJkiRJkiRpDLEDUJIkSZIkSZIkSRpD7ACUJEmSJEmSJEmSxhA7ACVJkiRJkiRJkqQxxA5ASZIkSZIkSZIkaQyxA1CSJEmSJEmSJEkaQ+wAlCRJkiRJkiRJksYQOwAlSZIkSZIkSZKkMcQOQEmSJEmSJEmSJGkMsQNQkiRJkiRJkiRJGkPsAJQkSdKIiYj1IuKq2vZYRPzHSNdLkiRJkiRpNFt6pCsgSZKk8SszbwY2BIiIpYC7gV+OaKUkSZIkSZJGOTsANazOOuusfuXbddddF3NNJEnSEmhb4LbM/MdIV0SSOs3/C0nSksW/y5Kksc4pQDWiHnzwQR588MGRroYkSVoy7AWc3OpARBwQEVcMc30kabHx/0KStGTx77IkaazxCUANq+aoqa7RVo6mkiRpfIuIZYG3AZ9qdTwzjwaOPuqoo3JYKyZJHeL/haSxpZq6/Arg7sx8S0SsA5wCrAL8FXhfZj4bEcsBJwEbAw8Be2bmHdU5PgXsDywA/j0zz6/SdwK+AywFHJOZX6nSB3wNteffZUnSWOcTgJIkSVoS7Az8NTPvH+mKSJIk9cPHgBtr+18FjsjMqcDDlI49qp8PZ+a6wBFVPiJiOmX2g/WBnYAfRMRSVcfi9ymx0XRg7yrvgK8hSZLGNzsAJUmStCTYmzbTf0qSJC1JImJt4M3AMdV+ANsAZ1RZTgR2q17vWu1THd+2yr8rcEpmPpOZfwduBWZU262ZeXtmPkt54m/XQV5DkiSNY3YASpIkaURFxArA9sAvRroukiRJ/fBt4JPA89X+i4BHMnN+tT8HWKt6vRZwF0B1/NEq/8L0Rpl26YO5Rg+uqSxJ0vhiB6AkSZJGVGY+mZkvysxHR7oukiRJvYmItwBzM/Mv9eQWWbOPY51K7+v63QmZR2fmJi3ySpKkMWjpka6AJEmSJEmSNEq8AXhbROwCLA+sRHkicFJELF09gbc2cE+Vfw7wEmBORCwNrAzMq6V3qZdplf7gIK4hSZLGMZ8AlCRJkiRJkvohMz+VmWtn5hRgL+D3mfke4ELgnVW2fYGzqtdnV/tUx3+fmVml7xURy0XEOsBU4HJgNjA1ItaJiGWra5xdlRnoNSRJ0jjmE4CSJEmSJEnS0BwCnBIRXwCuBI6t0o8FfhIRt1KeytsLIDOvj4jTgBuA+cCHM3MBQER8BDgfWAo4LjOvH8w1JEnS+GYHoCRJkiRJkjRAmXkRcFH1+nZgRos8TwPvalP+i8AXW6SfA5zTIn3A15AkSeOXU4BKkiRJkiRJkiRJY4gdgJIkSZIkSZIkSdIYYgegJEmSJEmSJEmSNIaMig7AiFgvIq6qbY9FxH9ExCoRMSsibql+Tq7yR0QcGRG3RsQ1EfHa2rn2rfLfEhH71tI3johrqzJHRkRU6QO+hiRJkiRJkiRJkjRSRkUHYGbenJkbZuaGwMbAk8AvgUOB32XmVOB31T7AzsDUajsAOApKZx5wGLApZdHkw7o69Ko8B9TK7VSlD+gakiRJkiRJkiRJ0kgaFR2ADdsCt2XmP4BdgROr9BOB3arXuwInZXEZMCki1gR2BGZl5rzMfBiYBexUHVspMy/NzAROapxrINeQJEmSJEmSJEmSRsxo7ADcCzi5er1GZt4LUP1cvUpfC7irVmZOldZb+pwW6YO5Rg8RcUBEXDGA9ydJkiRJkiRJkiQN2qjqAIyIZYG3Aaf3lbVFWg4ifTDX6JmQeXRmbtLHuSRJkiRJkiRJkqSOGFUdgJR19/6amfdX+/d3TbtZ/Zxbpc8BXlIrtzZwTx/pa7dIH8w1JEmSJEmSJEmSpBEz2joA96Z7+k+As4F9q9f7AmfV0veJYjPg0Wr6zvOBHSJickRMBnYAzq+OPR4Rm0VEAPs0zjWQa0iSJEmSJEmSJEkjZumRrkB/RcQKwPbAB2vJXwFOi4j9gTuBd1Xp5wC7ALcCTwL7AWTmvIj4PDC7yve5zJxXvT4QOAGYCJxbbQO+hiRJkiRJkiRJkjSSRk0HYGY+CbyokfYQsG2LvAl8uM15jgOOa5F+BbBBi/QBX0OSJEmSJEmSJEkaKaNtClBJkiRJkiRJkiRJvRjwE4ARcQFwBXAl8JfMvLVFnhnAnZl539CrKEmSpOFgnCdJkkYTYxdJkqT2BjMF6DbVlgAR8Tgl0Por8BfgWuCTwEuATTtTTUmSJA0D4zxJkjSaGLtIkiS1MZgOwEnAa4GNq20TYEtgK6qACwjgn52ooCRJkoaNcZ4kSRpNjF0kSZLaGHAHYGY+BlxUbQBExErAG4C9gfcA9wFf70gNJUmSNCyM8yRJ0mhi7CJJktTehE6cJDMfy8xzM3MfSoC1InBxJ84tSZKkkWOcJ0mSRhNjF0mSpKIjHYB1mXkacBXw1U6fW5IkSSPHOE+SJI0mxi6SJGk863gHYOVqYLPFdG5JkiSNHOM8SZI0mhi7SJKkcWnAHYARMTMiXhURvZV9EfDU4KslSZKk4WacJ0mSRhNjF0mSpPaWHkSZ44AEno6Ia4C/1La7gTcC7wC+26lKSpIkaVgY50mSpNHE2EWSJKmNwXQA/gewEfBaYGNgU0qwVXc/cHtEbAJcm5nPDKmWkiRJGg7GeZIkaTQxdpEkSWpjwB2AmXlk1+uIWA54NSXQ6to2AF4MfL/KtiAibgKuzMx9h1xjSZIkLRbGeZIkaTQxdpEkSWpvME8ALlSNmppdbQBExNKUAKsecL0aWB8wuJIkSRoFjPMkSdJoYuwiSZLU05A6AFvJzPnAVdV2HEC1GPO0Tl9LkiRJw8c4T5IkjSbGLpIkaTzreAdgK5n5PHDDcFxLkiRJw8c4T5IkjSbGLpIkabyYMNIVkCRJkiRJkiRJktQ5dgBKkiRJkiRJkiRJY4gdgJIkSZIkSah1w6QAACAASURBVJIkSdIYYgeglnizZs1i7733ZsqUKUQEhx9+eMt8jz76KPvttx+TJ09m5ZVX5j3veQ8PPfTQ8FZWkiRJkiRJkiRphPWrAzAivh0RW0ZELO4KSU3nnXce11xzDdtuuy0rrLBC23x77rknF110EccccwwnnHACs2fPZrfddhvGmkqSNPoY50mSpNFkpGOXiFg+Ii6PiKsj4vqI+GyVfkJE/D0irqq2Dav0iIgjI+LWiLgmIl5bO9e+EXFLte1bS984Iq6tyhzZ9V4jYpWImFXlnxURk/u6hiRJGr/6+wTgR4ALgfsj4piI2CUill2M9ZIW+vrXv87111/Psccey8SJE1vmufTSSzn//PM58cQT2X333Xn729/OT3/6Uy655BIuuOCCYa6xJEmjinGeJEkaTUY6dnkG2CYzXwNsCOwUEZtVxw7OzA2r7aoqbWdgarUdABwFpTMPOAzYFJgBHNbVoVflOaBWbqcq/VDgd5k5Ffhdtd/2GpIkaXzrbwfg/wMOBP4CvBf4FfBgRJwSEXtGxAsXVwU18mbOnMkmm2zCrFmzePWrX82KK67IFltswfXXXz8s158woe+v6bnnnssaa6zBlltuuTBtxowZrLPOOpx77rmLs3qSJI12xnmSJGk0GdHYJYsnqt1lqi17KbIrcFJV7jJgUkSsCewIzMrMeZn5MDCL0pm4JrBSZl6amQmcBOxWO9eJ1esTG+mtriFJksaxfnUAZubczDw6M3cGVgfeB5wP7AKcDMyNiF9HxP4Rsdriq65Gyp133snBBx/Mpz/9aU4++WTmzp3LHnvsQYlF25s/f36v24IFC/o8R3/cdNNNTJs2bZH0V77yldx0001DPr8kSWPVkhDnRcSkiDgjIm6KiBsjYvPFcR1JkjT6LSGxy1IRcRUwl9KJ9+fq0BerKTiPiIjlqrS1gLtqxedUab2lz2mRDrBGZt4LUP1cvY9rNOt9QERcMaA3K0mSRq3+PgG4UGY+lpk/z8x3AatRRhudQpmy4MfAPRFxcUR8LCJe2tnqaqTMmzeP008/nT333JO3ve1tfO1rX+OGG27g5ptv7rXcMsss0+u2++678/vf/37I9Xv44YeZNGnSIumTJ0/m4YcfHvL5JUkaD0YwzvsOcF5mTgNeA9zYwXNLkqQxaqRil8xckJkbAmsDMyJiA+BTwDTgdcAqwCFV9lZrFeYg0nvTrzJVx+kmfZxLkiSNEUsPpXBmPgOcDZwdEROArYB3UKYeOAL4VkRclZkbD7mmGlFTpkxh6tSpC/enT58OwJw5c1o+eddl9uzZvZ734osvZvXVV+81T3+1Wv87M1umS5Kk3g1XnBcRKwFbAjOr6z4LPDuUc0qSpPFnJNqoMvORiLgI2Ckzv1ElPxMRxwMHVftzgJfUiq0N3FOlb91Iv6hKX7tFfijrHq6ZmfdWU3zO7eMakiRpHBvwE4DtZObzmXlhZn40M/+FsoDxV4GJnbqGRk7z6bplly3raz/99NO9lttwww173dZZZx1e+MKhT88/efJkHnnkkUXSH3nkkZZPBkqSpP5bzHHey4AHgOMj4sqIOCYiVmxmcsoqSZLUX4szdomI1SJiUvV6IrAdcFPXmntRRiHvBlxXFTkb2CeKzYBHq+k7zwd2iIjJETEZ2AE4vzr2eERsVp1rH+Cs2rn2rV7v20hvdQ1JkjSOdawDsCkzr8jM/8rM6YvrGlryDdcUoNOmTWu51l+7tQElSdLgdTjOWxp4LXBUZm4E/BM4tMU1nbJKkiQNSodjlzWBCyPiGmA2ZQ3AXwM/i4hrgWuBVYEvVPnPAW4HbqVMS/pvVZ3mAZ+vzjEb+FyVBnAgcExV5jbg3Cr9K8D2EXELsH213/YakiRpfBvSFKBSX4ZrCtCdd96Zz3/+81xyySVsscUWAFxxxRXcfvvt7LzzzkM+vyRJWmzmAHMy88/V/hm06ACUJElaEmTmNcBGLdK3aZM/gQ+3OXYccFyL9CuADVqkPwRsO5BrSJKk8csOQC1Wm2zS+0D9u+++u89z/OMf/1jYkfjss89yww03cMYZZ7Diiisu7NzbfPPN2XHHHdlnn334xje+wYQJEzjkkEPYYost2G677Yb+RiRJ0mKRmfdFxF0RsV5m3kxp1LphpOslSZIkSZI0mtkBqCXehRdeyH777bdw//TTT+f000/npS99KXfcccfC9FNOOYWPf/zjvP/97+f555/nLW95C0ceeeQI1FiSJA3QRynTZi1Lmb5qvz7yS5IkSZIkqRd2AKpPJ5xwwiJpU6ZMocwwsfjNnDmTmTNn9plv0qRJHH/88Rx//PGLv1KSJKljMvMqwPX9JEmSJEmSOmTCSFdAkiRJkiRJkiRJUufYAShJkiRJkiRJkiSNIR2bAjQiVgXeCDwJXJCZCzp1bkmSJI0c4zxJkjSaGLtIkiQN4gnAiDgwIv4cEavU0jYGbgTOAM4B/hQRK3aumpIkSVrcjPMkSdJoYuwiSZLU3mCmAN0TyMycV0v7OjAZOJ4SXL0O+NDQq9ctIiZFxBkRcVNE3BgRm0fEKhExKyJuqX5OrvJGRBwZEbdGxDUR8draefat8t8SEfvW0jeOiGurMkdGRFTpA76GJEnSKDUicZ4kSdIgGbtIkiS1MZgOwKnANV071bQKWwHHZuYHMvOtwGzg3Z2p4kLfAc7LzGnAayijuQ4FfpeZU4HfVfsAO1f1nAocABxV1XUV4DBgU2AGcFhXh16V54BauZ2q9AFdQ5IkaRQbqThPkiRpMIxdJEmS2hhMB+CLgLm1/TdUP39ZS/sD8NLBVqopIlYCtgSOBcjMZzPzEWBX4MQq24nAbtXrXYGTsrgMmBQRawI7ArMyc15mPgzMAnaqjq2UmZdmZgInNc41kGtIkiSNVsMe50mSJA2BsYskSVIbg+kAnAesWtvfCnge+FMtLYHlh1CvppcBDwDHR8SVEXFMNX/7Gpl5L0D1c/Uq/1rAXbXyc6q03tLntEhnENeQJEkarUYizpMkSRosYxdJkqQ2BtMBeCPw1oh4UURMosy3PjszH6vlmQLc14H6dVkaeC1wVGZuBPyT7qk4W4kWaTmI9N70q0xEHBARV/RxLkmSpCXBSMR5kiRJg2XsIkmS1MZgOgC/A6xJeeLtLuDFwA+6DkbEUsAWwNWdqGBlDjAnM/9c7Z9B6RC8v2vazern3Fr+l9TKrw3c00f62i3SGcQ1esjMozNzk36/U0mSpJEzEnGeJEnSYBm7SJIktTHgDsDMPBv4EHA9cDNwUGb+tJZlO8rUCud3pIblmvcBd0XEelXStsANwNnAvlXavsBZ1euzgX2i2Ax4tJq+83xgh4iYHBGTgR2A86tjj0fEZhERwD6Ncw3kGpIkSaPSSMR5kiRJg2XsIkmS1N7SgymUmUcDR7c5dj4weSiVauOjwM8iYlngdmA/SgfmaRGxP3An8K4q7znALsCtwJNVXjJzXkR8Hphd5ftcZs6rXh8InABMBM6tNoCvDOQakiRJo9kIxXmSJEmDYuwiSZLU2qA6AEdCZl4FtJpKc9sWeRP4cJvzHAcc1yL9CmCDFukPDfQakiRJkiRJksa+mTNncuKJJy6SfuONNzJt2rQRqJEkScWo6QCUJEmSJEmSpCXNtGnTOP7443ukTZkyZWQqI0lSZVAdgBGxFXAwMIMylUKrtQQzM+1glCRJGkWM8yRJ0mhi7KIlwYorrshmm2020tWQJKmHAQc/EfFm4ExgKcqaeDcD8ztcL0mSJA0z4zxJkjSaGLsIyhSc1113HV/+8pf5xCc+wW233cZGG23Ej370I9Zff/2Rrp4kSSNmMKOfDgeeA96cmb/tbHUkSZI0gg7HOE+SJI0eh2PsIuDOO+/k4IMP5tOf/jQTJ07koIMOYo899uC6664jItqWmz+/u794wYIFi6QBLLXUUr2eA+CGG25gpZVW4plnnuF1r3sdX/ziF9lqq62G8I4kSRq6wXQAbgCcYmAlSZI05hjnSZKk0cTYRQDMmzePP/7xj0ydOhWA559/nre//e3cfPPNTJs2rW25ZZZZps9zH3/88cycObPt8Y022ohNN92U6dOn88ADD/DNb36T7bffnksuuYQZM2YM+L1IktQpg+kAfAKY1+mKSJIkacQZ50mSpNHE2EUATJkyZWHnH8D06dMBmDNnTq8dgLNnz174+uKLLwZY5Mm9ddZZp9drf+xjH+ux/+Y3v5np06fzpS99iTPPPLN/b0CSpMVgMB2AvwM273RFJEmSNOKM8yRJ0mhi7CIAJk2a1GN/2WWXBeDpp5/utdyGG2648PWdd965SBqUKUAHYuLEieyyyy786le/GlA5SZI6bcIgyhwCvDwiPhN9TYAtSZKk0cQ4T5IkjSbGLhqSZZZZZuG2++67s/vuu/dIW2aZZTjxxBMHdW6/kpKkkTaYJwAPA64HPgu8PyKuAh5pkS8zc/+hVE6SJEnDyjhPkiSNJsYuGpJOTAHa9NRTT3Huueey8cYbD72CkiQNwWA6AGfWXk+ptlYSMLiSJEkaPWbWXk/BOE+SJC3ZZtZeT8HYRQO0ySabLHx99913L5LWl0cffZS3vOUtvPe972XdddflwQcf5IgjjuDuu+/mtNNO63h9JUkaiMF0AA5s2IskSZJGC+M8SZI0mhi7aEQtt9xyrLbaanzhC19g7ty5LL/88my++eZcfPHFA+pIlCRpcRhwB2Bm/mNxVESSJEkjyzhPkiSNJsYuAjjhhBMWSZsyZQqZudivvfzyy/OLX/xisV9HkqTBGMwTgD1ExErAysCjmfnY0KskSZKkJYFxniRJGk2MXSRJkroNqgMwIpYCDgY+QG26hYj4O3AM8I3MnN+RGqpf3n3ZcyNdhUF51WPPA3DqKK0/wM83W2akqyBJUscY50mSpNHE2EWSJKm1AXcARsSywHnAVpRFlO8C7gXWpCy2/EVgp4jYITOf7VxVJUmStDgZ50mSpNHE2EWSJKm9CYMo85/A1sBvgFdm5pTM3DwzpwDrAb8C3ljlkyRJ0uhhnCdJkkYTYxdJkqQ2BtMB+G7gOmC3zLylfiAzbwPeAVwPvGfo1ZMkSdIwMs6TJEmjybDHLhGxfERcHhFXR8T1EfHZKn2diPhzRNwSEadWTycSEctV+7dWx6fUzvWpKv3miNixlr5TlXZrRBxaSx/wNSRJ0vg1mA7AdYFzM/P5Vger9HOBlw+lYpIkSRp2xnmSJGk0GYnY5Rlgm8x8DbAhZYrRzYCvAkdk5lTgYWD/Kv/+wMOZuS5wRJWPiJgO7AWsD+wE/CAilqrWNPw+sDMwHdi7ystAryFJksa3wXQAPgu8oI88KwLPDeLckiRJGjnGeZIkaTQZ9tgliyeq3WWqLYFtgDOq9BOB3arXu1b7VMe3jYio0k/JzGcy8+/ArcCMars1M2+v1i08Bdi1KjPQa0iSpHFsMB2A1wDvjIjVWh2MiFWBdwJXD6VikiRJGnbGeZIkaTQZkdilelLvKmAuMAu4DXgkM+dXWeYAa1Wv1wLuAqiOPwq8qJ7eKNMu/UWDuEaz3gdExBWDe9eSJGm0GUwH4PeA1YDLI2L/iHhZREys5iHfD/hzdfx7nayoJEmSFjvjPEmSNJqMSOySmQsyc0NgbcoTe69sla362epJvOxgem/X6JmQeXRmbtIiryRJGoOWHmiBzDwtIjYEDgWObpElgK9l5mlDrZwkSZKGj3GeJEkaTUY6dsnMRyLiImAzYFJELF09gbc2cE+VbQ7wEmBORCwNrAzMq6V3qZdplf7gIK4hSZLGscE8AUhm/hfweuA44Erg9urnccAbMvPQjtVQkiRJw8Y4T5IkjSbDHbtExGoRMal6PRHYDrgRuJAy3SjAvsBZ1euzq32q47/PzKzS94qI5SJiHWAqcDkwG5haPcW4LLAXcHZVZqDXkCRJ49iAnwDskpmXAZd1sC6SJElaAhjnSZKk0WSYY5c1gRMjYinKwPrTMvPXEXEDcEpEfIHSAXlslf9Y4CcRcSvlqby9qjpfHxGnATcA84EPZ+YCgIj4CHA+sBRwXGZeX53rkIFcQ5IkjW+D7gCUJEmSJEmSxpPMvAbYqEX67ZT1AJvpTwPvanOuLwJfbJF+DnBOJ64hSZLGLzsAJUmSNKIi4g7gcWABMD8zNxnZGkmSJEnS/2fv3qPtqst7/78fQoJcKuEiiIRjcjCoXFpQBFpQbhUCegz4A0V7CrH0xPqDX6Uy1IBW8IKiPbYqNNFYQoIXAsNLiYjQiESthZCAglxKCRjDlkh+kAh4lEuS5/yxvsGVnbX2ZWbttfZM3q8x1lhzPfM75/dZGYMxPqzvnnNKUr0NugAYEXOABC7MzMfK56HIzDx7s7qTJEnSiBllOe/YzHy8w+eUJElbkFGWXSRJkka1oVwBOI1GuPo08Fj5PBQJGK4kSZJGr2mY8yRJUn1Mw+wiSZI0JENZAJxU3n/V77MkSZLqbbTkvAT+LSIS+FJmzu4/ICKmA9NnzpzZ9eYkSdKoMVqyiyRJ0qi3zWADMvOX5bW2lI4Gdm6qb/IC/qiMkyRJ0ig1inLekZn5GuAk4JyIeEOLXmf7bEBJkrZuoyi7SJIkjXqDLgC2MBc4ZZAxU4ErK5xbkiRJvTOXHuS8zHy0vK8Cvg0c1snzS5KkLdZc/I1KkiSppSoLgEMxhsatnCRJkrRl6WjOi4gdI+KPNmwDJwD3dOr8kiRpq+dvVJIkaas0lGcAVrEfsGaEzi1JkqTe6XTO2xP4dkRAI5t+PTNv7OD5JUnS1s3fqCRJ0lZpSAuAETGnX+mUiJjYYugY4L8Brwe+u1mdSZIkacT1Oudl5sPAn3TqfJIkacvW6+wiSZJUF0O9AnBa03YCB5dXKwksBv6ueluSJEnqkmlN2+Y8SZI02k1r2ja7SJIktTHUBcBJ5T2Ah4HPAZ9vMW4dsCYz/08HepMkSdLIM+dJkqQ6MbtIkiQNwZAWADPzlxu2I+KjwC3NtW6IiOXA0zQC3NrMPDQidgWuASYCy4G3ZeaaaDxE5vPAycDvgGmZeWc5z1nAh8tpP5GZ80r9tcBcYHvgBuC9mZlV5pAkSaqL0ZDzJEmShsrsMvq987bne91CJQc9tR6Aa2raP8DXjxjb6xYkSaPINsM9IDM/mpk/GolmhuDYzDw4Mw8tn2cAN2fmZODm8hngJGByeU0HZgGUxbyLgMOBw4CLImKXcsysMnbDcVOqzCFJklRXPc55kqQBrFu3jk9/+tO8/vWvZ7fddmO33XbjhBNOYMmSJb1uTeoZs4skSVJ7w14AHGWmAvPK9jzglKb6VdlwGzA+IvYCTgQWZubqzFwDLASmlH0vzsxbMzOBq/qdazhzSJIkSZLUUb///e+59NJLed3rXsdXvvIVvvrVrzJ27FiOOuoo7rjjjl63J0mSJGmUGeozADdRFruOB/YGtmsxJDPz41XP3+p8wL9FRAJfyszZwJ6ZubJMtjIi9ihj9wYeaTq2r9QGqve1qFNhjpXNTUfEdGD6zJkzh/+NJUmSeqAHOU+SNIjtt9+ehx9+mF122eWF2vHHH89+++3H5ZdfzpVXXtnD7qTeMrtIkiRtqtICYLnH+ox+xweNRbrm7U6GqyMz89GyALcwIv5zoBZb1LJCfSBDOqYsVM6eNWvWYOeTJEnquR7lPEmqhWnTpnHPPffwqU99ivPPP5+HHnqIQw45hC996UsccMABIzr3mDFjNlr8Axg3bhwHHHAAq1atGtG5pdHM7CJJktTasG8BGhF/Afw98GPgNBpBah7wTuDLwHpgPnBc59qEzHy0vK8Cvk3jGX6PbbjtZnnf8H89fcA+TYdPAB4dpD6hRZ0Kc0iSJNVSr3KeJNXJihUreP/738+HPvQhrr76alatWsXb3vY2Gk+TaG/t2rUbvdatW8e6des2qg12jv6effZZ7rjjDvbff//N+UpSbZldJEmS2qtyBeB7aCx+TcnMtREBsDwz5wPzI+LbwHeBqzvVZETsCGyTmU+X7ROAjwELgLOAS8v7deWQBcC5ETEfOBx4sty+8ybgkxGx4c8mTwAuyMzVEfF0RBwBLAbOBC5rOteQ5+jUd5YkSeqBruc8Saqb1atX85Of/ITJkycDsH79ek499VQeeOABXvWqV7U9buzYsYOe+8orr2TatGlD7uWSSy5hzZo1/PVf//WQj5G2MGYXSZKkNqosAB4EXJ2Za5tqYzZsZOZNZaHt/cB3NrO/DfYEvl2C3LbA1zPzxohYAlwbEWcDK4DTy/gbgJOBZcDvgHeV3lZHxMeBJWXcxzJzddl+DzAX2B74XnlBY+FvyHNIkiTVWC9yniTVysSJE19Y/ANeuPqur69vwAXAJUuWbPT5hz/8IQBHH330C7VJkyYNuY/vfve7XHLJJXz2s5/lla985ZCPk7YwZhdJkqQ2qiwAjgWeaPr8e2DnfmPuAf6malP9ZebDwJ+0qD9B4yHP/esJnNPmXHOAOS3qS4EDOzGHJElSTXU950lS3YwfP36jz+PGjQPgmWeeGfC4gw8+eKPPK1as2KQ+ZswYhmLJkiW8/e1v593vfjfnnXfekI6RtlBmF0mSpDaqLACuBPZq+rwC+ON+Y/YG1iJJkqQ6MedJ0gjp1C1A/+u//os3velNHH/88Vx22WUDjpW2AmYXSZKkNqosAP6Uxi0WNvgBMD0i/hL4FnAM8P8AP9ns7iRJktRN5jxJGiGduAXoypUrOfHEE9l33325+uqrh3zFoLQFM7tIkiS1UWUB8HpgZkRMysxf0HhG3ttpPD9vbhnzPPDhTjQoSZKkrjHnSdIIOfTQQzf6/Ktf/aplvZ3f//73nHTSSaxZs4bLL7+cu++++4V92223HYccckjnmpXqw+wiSZLUxrAXADNzLn8IUWTmIxHxOuB8YF9gOTAzM3/emRYlSZLUDeY8SRq9HnvsMe666y4A3vzmN2+07+UvfznLly/vQVdSb5ldJEmS2qtyBeAmyl9ZnduJc0mSJGn0MOdJ0h/MnTt3k9rEiRPJzBGfu1vzSHVndpEkSWrYZrgHRMT4kWhEkiRJvWXOkyRJdWJ2kSRJam/YC4DAyoi4JiJOjogqx0uSJGl0MudJkqQ6MbtIkiS1USUcLQdOB74D/Coi/iEiDupoV5IkSeqF5ZjzJElSfSzH7CJJktTSsBcAM/PVwOHAF4GxNB6s/LOIuCMi/jYidu9wj5IkSeoCc54kSaoTs4skSVJ7lW6PkJlLMvMcYC8af2n1XeAg4HM0/uLqXyPilM61KUmSpG4w50mSpDoxu0iSJLW2WfdHz8znM/ObmfkW4GXA+4B7gLcA3+hAf5IkSeoBc54kSaoTs4skSdLGOvmA5CeAe4H7geeB6OC5JUmS1DvmPEmSVCcjll0iYp+IuCUi7o+IeyPivaV+cUT8KiJ+Vl4nNx1zQUQsi4gHIuLEpvqUUlsWETOa6pMiYnFEPBgR10TEuFLfrnxeVvZPHGwOSZK09dp2c08QEa8CzgL+J42/sApgGTBvc88tSZKk3jHnSZKkOulSdlkLnJ+Zd0bEHwF3RMTCsu+fMvN/9+tpf+AM4IDS0/cjYr+y+5+BNwJ9wJKIWJCZ9wGfLueaHxFfBM4GZpX3NZn5iog4o4x7e7s5MnNdB7+3JEmqmUoLgBGxC/AOGqHqUBqB6ingCmBuZv5HxzqUJElS15jzJElSnXQ7u2TmSmBl2X46Iu4H9h7gkKnA/Mx8FvhFRCwDDiv7lmXmw+V7zAemlvMdB7yzjJkHXExjAXBq2YbGbU0vj4gYYI5bN/8bS5Kkuhr2LUAj4hs0gs5lwGuB7wN/Abw0M6f7o5AkSVI9mfMkSVKd9Dq7lFtwHgIsLqVzI+LuiJhTFiahsTj4SNNhfaXWrr4b8JvMXNuvvtG5yv4ny/h25+rf7/SIWDrsLypJkmqpyjMA3wr8AvgQ8PLMPDEzr87MZzrbmiRJkrrMnCdJkuqkZ9klInYCvgmcl5lP0bhCb1/gYBqLkp/dMLTF4VmhXuVcGxcyZ2fmoS3GSpKkLVCVW4BeCNyYmT/rdDOSJEnqKXOeJEmqk55kl4gYS2Px72uZ+S2AzHysaf+XgevLxz5gn6bDJwCPlu1W9ceB8RGxbbnKr3n8hnP1RcS2wM7A6kHmkCRJW6kqVwBeAvyvTjciSZKknjPnSZKkOul6dinP3LsCuD8z/7GpvlfTsFOBe8r2AuCMiNguIiYBk4HbgSXA5IiYFBHjgDOABZmZwC3AaeX4s4Drms51Vtk+DfhBGd9uDkmStBWrcgXg44C3gZIkSdrymPMkSVKd9CK7HAn8JfDziNhw5eGFwDsi4mAat95cDrwbIDPvjYhrgfuAtcA5mbkOICLOBW4CxgBzMvPecr4PAvMj4hPAT2ksOFLevxIRy2hc+XfGYHNIkqStV5UFwEXAn3W4D0mSJPXeIsx5kiSpPhbR5eySmf9O62fu3TDAMZfQuFqxf/2GVsdl5sPAYS3qzwCnD2cOSZK09apyC9APA6+MiI+Xe55LkiRpy2DOkyRJdWJ2kSRJaqPKFYAX0LiP+YXA2RFxF/BrGrc4aJaZefZm9idJkqTuMedJkqQ6MbtIkiS1UWUBcFrT9kvLq5UEDFeSJEn1Ma1p25wnSZJGu2lN22YXSZKkJlUWACd1vAtJkiSNBuY8SZJUJ2YXSZKkNoa9AJiZvxyJRiRJktRb5jxJklQnZhdJkqT2qlwB+IKI2BHYD9gpM3/cmZYkSZLUa+Y8SXURp3yh1y1Uct7rxgBwypX17B8g//Vve92C9AKziyRJ0sa2qXJQREyIiG8Ca4ClwC1N+46KiPsi4pjOtChJkqRu6VXOi4gxEfHTiLi+0+eWJElbLn+jkiRJam3YC4ARsRewGJgKXA/cCkTTkMXAHsDbO9GgJEmSuqPHOe+9wP0jcF5JkrSF8jcqSZKk9qpcAXgRjfD0LiFfKwAAIABJREFU55n5VmBh887MfB74MXDk5rcnSZKkLupJzouICcCbgH/p5HklSdIWz9+oJEmS2qiyAHgysCAzFw0wZgXwskodSZIkqVd6lfM+B3wAWN/h80qSpC2bv1FJkiS1UWUBcE/gwUHGPA/sWOHckiRJ6p2u57yIeDOwKjPvGGTc9IhY2ql5JUnSFsHfqCRJktqosgC4GthnkDH7Ab+ucG5JkiT1Ti9y3pHAWyJiOTAfOC4ivtp/UGbOzsxDOzivJEmqP3+jkiRJaqPKAuBPaPxI89JWOyNiMjAFuGVzGpMkSVLXdT3nZeYFmTkhMycCZwA/yMz/2anzS5KkLZq/UUmSJLVRZQHwH4AXAT+MiJOAHQAiYsfy+Ts0nt/y2Y51KUmSpG4w50mSpDoxu0iSJLWx7XAPyMzFETEd+CJwfdOup8r7WuCvMvPeDvQnSZKkLul1zsvMRcCikTi3JEna8vQ6u0iSJI1mVa4AJDOvBA4EvgDcDjwE3AnMBP44M7/WsQ6bRMSYiPhpRFxfPk+KiMUR8WBEXBMR40p9u/J5Wdk/sekcF5T6AxFxYlN9Sqkti4gZTfVhzyFJklRXvcp5kiRJVZhdJEmSWhv2FYAbZOaDwN91sJeheC9wP/Di8vnTwD9l5vyI+CJwNjCrvK/JzFdExBll3NsjYn8az5Y5AHgZ8P2I2K+c65+BNwJ9wJKIWJCZ9w13jpH+B5AkSRppPcp5kiRJlZhdJEmSNlXpCsBeiIgJwJuAfymfAzgO+EYZMg84pWxPLZ8p+48v46cC8zPz2cz8BbAMOKy8lmXmw5n5HDAfmFpxDkmSJEmSJEmSJKlnOrYAGBG7R8SpEXFiRIzp1HmbfA74AI2HNwPsBvwmM9eWz33A3mV7b+ARgLL/yTL+hXq/Y9rVq8whSZK0RelCzpMkSeoYs4skSVKFBcCIeE955t2uTbXX0rg15zeAG4D/iIgdO9VkRLwZWJWZdzSXWwzNQfZ1qj7Y/C+IiOkRsbTFWEmSpFGlFzlPkiSpKrOLJElSe1WuAHw7kJm5uqn2D8AuwJU0wtXrgL/Z/PZecCTwlohYTuP2nMfRuCJwfERseI7hBODRst0H7ANQ9u8MrG6u9zumXf3xCnNsJDNnZ+ahVb60JElSl/Ui50mSJFVldpEkSWqjygLgZODuDR8iYnfgaOCKzPzrzPwfwBLgnZ1pETLzgsyckJkTgTOAH2TmXwC3AKeVYWcB15XtBeUzZf8PMjNL/YyI2C4iJpXvcnvpd3JETIqIcWWOBeWY4c4hSZJUV13PeZIkSZvB7CJJktRGlQXA3YBVTZ+PLO/fbqr9GHh51aaG4YPA+yJiWenrilK/Atit1N8HzADIzHuBa4H7gBuBczJzXXmG37nATTRuE3FtGTvsOSRJkmpsNOU8SZKkwZhdJEmS2th28CGbWA3s3vT5aGA98B9NtQRetBl9tZWZi4BFZfth4LAWY54BTm9z/CXAJS3qN9C4NUT/+rDnkCRJqqme5jxJkqRhMrtIkiS1UeUKwPuB/xERu0XEeBr3W1+SmU81jZkI/LoD/UmSJKl7zHmSJKlOzC6SJEltVFkA/DywF9AHPAK8FJi5YWdEjAGOAu7qRIOSJEnqGnOeJEmqE7OLJElSG8O+BWhmLoiIvwGml9LXMvOrTUP+nMatFW7qQH/awhz04PUbfd7pd0+0rP988pu71pMkSWow50mSpDoxu0iSJLVX5RmAZOZsYHabfTcBu2xOU9p6/HaH3XrdgiRJamLOkyRJdWJ2kSRJaq3SAiBAROwEnAocAuwMPAn8FPh2Zv62M+1pS+OVfZIkjX7mPEmSVCdmF0mSpE1VWgCMiNOBLwLjgWjalcDnIuLdmfmNDvQnSZKkLjLnSZKkOjG7SJIktTbsBcCIeCNwNbAeuApYBPyaxoOWjwXeCVwdEb/JzO93rlVJkiSNJHOeJEmqE7OLJElSe1WuAPwI8Czw+sy8s9++eRFxOfCjMs5wJUmSVB/mPEmSVCdmF0mSpDa2qXDMIcA1LYIVAJm5FLgWeM3mNCZJkqSuM+dJkqQ66Xp2iYh9IuKWiLg/Iu6NiPeW+q4RsTAiHizvu5R6RMQXImJZRNwdEa9pOtdZZfyDEXFWU/21EfHzcswXIiKqziFJkrZeVRYAnwVWDjLm0TJOkiRJ9WHOkyRJddKL7LIWOD8zXw0cAZwTEfsDM4CbM3MycHP5DHASMLm8pgOzoLGYB1wEHA4cBly0YUGvjJnedNyUUh/WHJIkaetWZQHwx8BRg4w5ksYtFiRJklQf5jxJklQnXc8umblywxWHmfk0cD+wNzAVmFeGzQNOKdtTgauy4TZgfETsBZwILMzM1Zm5BlgITCn7XpyZt2Zm0ni2YfO5hjOHJEnailVZAPwgcFBEXBoROzbviIgdI+IzwIH84a+QJEmSVA/mPEmSVCc9zS4RMZHGbUgXA3tm5kpoLBICe5RhewOPNB3WV2oD1fta1KkwR/9+p0fE0uF8R0mSVF/bDjYgIua0KN8NvB+YHhF3Ao8Be9K4p/rONP6y6gPA2Z1rVZIkSZ1kzpMkSXUymrJLROwEfBM4LzOfKo/pazm0RS0r1AdsZyjHZOZsYPasWbMGO58kSdoCDLoACEwbYN944LgW9aOBN+APQ5IkSaPZtAH2mfMkSdJoM22AfV3LLhExlsbi39cy81ul/FhE7JWZK8vtN1eVeh+wT9PhE2g8l7APOKZffVGpT2gxvsockiRpKzaUBcBJI96FJEmSesGcJ0mS6qTn2SUal/pdAdyfmf/YtGsBcBZwaXm/rql+bkTMBw4HniwLeDcBn4yIXcq4E4ALMnN1RDwdEUfQuLXomcBlVebo9HeXJEn1MugCYGb+shuNSJIkqbvMeZIkqU5GSXY5EvhL4OcR8bNSu5DGoty1EXE2sAI4vey7ATgZWAb8DngXQFno+ziwpIz7WGauLtvvAeYC2wPfKy+GO4ckSdq6DeUKwAFFxItp3FP9ycx8avNbkiRJ0mhgzpMkSXXSjeySmf9O62fuARzfYnwC57Q51xxgk+caZuZS4MAW9SeGO4ckSdp6bVPloIgYExEzImIZsAZYDqyJiGWlvtkLi5IkSeo+c54kSaoTs4skSVJrww5BETEOuJHGQ5QTeARYCewFTAQuAaZExAmZ+VznWpUkSdJIMudJkqQ6MbtIkiS1V+UKwPcBxwDfBV6dmRMz808zcyLwSuA7wOvLOEmSJNWHOU+SJNWJ2UWSJKmNKguA7wTuAU7JzAebd2TmQ8BbgXuBv9j89iRJktRF5jxJklQnZhdJkqQ2qiwAvgL4Xmaub7Wz1L8H7Ls5jUmSJKnrzHmSJKlOzC6SJEltVFkAfA7YaZAxOwLPVzi3JEmSesecJ0mS6sTsIkmS1EaVBcC7gdMi4iWtdkbE7sBpwF2b05gkSZK6zpwnSZLqxOwiSZLURpUFwMuBlwC3R8TZEfHfI2L7iJgUEe8CFpf9l3eyUUmSJI04c54kSaoTs4skSVIb2w73gMy8NiIOBmYAs1sMCeAzmXnt5jYnSZKk7jHnSZKkOjG7SJIktTfsBUCAzLwwIhYAZwOHADsDTwI/BeZk5q2da1GSJEndYs6TJEl1YnaRJElqrdICIEBm3gbc1sFeJEmSNAqY8yRJUp2YXSRJkjZV5RmAkiRJkiRJkiRJkkYpFwAlSZLUMxHxooi4PSLuioh7I+Kjve5JkiRJkiSp7irfAlSSJEnqgGeB4zLztxExFvj3iPheuZWXJEmSJEmSKnABUJIkST2TmQn8tnwcW17Zu44kSZIkSZLqz1uASpIkqaciYkxE/AxYBSzMzMUtxkyPiKXd706SJEmSJKl+XACUJElST2Xmusw8GJgAHBYRB7YYMzszD+1+d5IkSZIkSfUz6AJgRPxtRBzWjWYkSZLUPaMt52Xmb4BFwJQetyJJkkah0ZZdJEmSRrOhXAH4OZp+hImIdRHx9yPXkiRJkrqk5zkvIl4SEePL9vbAnwP/2c0eJElSbfQ8u0iSJNXFUBYAnwG2a/oc5dU1EfGiiLg9Iu6KiHsj4qOlPikiFkfEgxFxTUSMK/XtyudlZf/EpnNdUOoPRMSJTfUppbYsImY01Yc9hyRJUk30POcBewG3RMTdwBIazwC8vss9SJKkehgN2UWSJKkWhrIA+AvgxIjYs6mWI9RPO88Cx2XmnwAHA1Mi4gjg08A/ZeZkYA1wdhl/NrAmM18B/FMZR0TsD5wBHEDjL8ZmRsSYiBgD/DNwErA/8I4yluHOIUmSVCM9z3mZeXdmHpKZf5yZB2bmx7o5vyRJqpWeZxdJkqS6GMoC4JeA1wCPRsS6Uru43GZhoNfaTjWZDb8tH8eWVwLHAd8o9XnAKWV7avlM2X98RESpz8/MZzPzF8Ay4LDyWpaZD2fmc8B8YGo5ZrhzSJIk1UXPc54kSdIwmF0kSZKGaNvBBmTmFyJiFfAm4GXAscAKYPnItraxcpXeHcAraFyt9xDwm8zcEOL6gL3L9t7AIwCZuTYingR2K/Xbmk7bfMwj/eqHl2OGO8fj/fqeDkyfOXNmtS8uSZI0QkZLzpMkSRoKs4skSdLQDboACJCZ82lcFUdErAeu7PbtmTJzHXBwRIwHvg28utWw8t7qSrwcoN7qSsiBxg80x8aFzNnA7FmzZnlLCkmSNOqMhpwnSZI0VGYXSZKkoRnKLUD7+yiwqMN9DFlm/qbMfwQwPiI2LGJOAB4t233APgBl/87A6uZ6v2Pa1R+vMIckSVJd9TTnSZIkDZPZRZIkqY1hLwBm5kcz80cbPkfEiyNin4h4cWdb+4OIeEm58o+I2B74c+B+4BbgtDLsLOC6sr2gfKbs/0FmZqmfERHbRcQkYDJwO7AEmBwRkyJiHHAGsKAcM9w5JEmSaqkXOU+SJKkqs4skSVJ7Va4AJCLGRMSMiFgGrKFxr/U1EbGs1Id0a9Fh2Au4JSLuprFYtzAzrwc+CLyv9LEbcEUZfwWwW6m/D5gBkJn3AtcC9wE3Audk5rryjL9zgZtoLCxeW8Yy3DkkSZLqrAc5T5IkqTKziyRJUmvDDkHlCrkbgaNpPPPuEWAljUW6icAlwJSIOCEzn+tEk5l5N3BIi/rDwGEt6s8Ap7c51yWlx/71G4AbOjGHJElSHfUi50mSJFVldpEkSWqvyhWA7wOOAb4LvDozJ2bmn2bmROCVwHeA15dxkiRJqg9zniRJqhOziyRJUhtVFgDfCdwDnJKZDzbvyMyHgLcC9wJ/sfntSZIkqYvMeZIkqU7MLpIkSW1UuQ/6K4DLMnN9q52ZuT4ivgf8f5vVmSRJkrrNnCdJkurE7KLKDnrw+o0+7/S7J1rWfz75zV3rSZKkTqpyBeBzwE6DjNkReL7CuSVJktQ75jxJklQnXc8uETEnIlZFxD1NtYsj4lcR8bPyOrlp3wURsSwiHoiIE5vqU0ptWUTMaKpPiojFEfFgRFxTnnNIRGxXPi8r+ycONoeG57c77MZvd9it121IktQxVa4AvBs4LSIuzsz/v//OiNgdOA24a3ObkyRJUleZ8yRJUp30IrvMBS4HrupX/6fM/N/95t8fOAM4AHgZ8P2I2K/s/mfgjUAfsCQiFmTmfcCny7nmR8QXgbOBWeV9TWa+IiLOKOPe3m6OzFzXwe+8RfLKPknSlq7KFYCXAy8Bbo+IsyPiv0fE9uUvlN4FLC77L+9ko5IkSRpx5jxJklQnXc8umfkjYPUQh08F5mfms5n5C2AZcFh5LcvMhzPzOWA+MDUiAjgO+EY5fh5wStO55pXtbwDHl/Ht5pAkSVu5YV8BmJnXRsTBwAxgdoshAXwmM6/d3OYkSZLUPeY8SZJUJ6Msu5wbEWcCS4HzM3MNsDdwW9OYvlIDeKRf/XBgN+A3mbm2xfi9NxyTmWsj4skyfqA5NhIR04HpM2fOrPQFJUlSvVS5ApDMvBD4M2AO8FPg4fI+BzgyM2cMcLgkSZJGKXOeJEmqk1GSXWYB+wIHAyuBz5Z6tBibFepVzrVpMXN2Zh7aap8kSdryVHkGIACZeRsb/4WRJEmStgDmPEmSVCe9zi6Z+diG7Yj4MnB9+dgH7NM0dALwaNluVX8cGB8R25arAJvHbzhXX0RsC+xM41akA80hSZK2YpWuAJQkSZIkSZIEEbFX08dTgXvK9gLgjIjYLiImAZOB24ElwOTyrMJxwBnAgsxM4BbgtHL8WcB1Tec6q2yfBvygjG83hyRJ2spVvgJQkiRJkiRJ2ppExNXAMcDuEdEHXAQcU55FmMBy4N0AmXlvRFwL3AesBc7JzHXlPOcCNwFjgDmZeW+Z4oPA/Ij4BI1bmV5R6lcAX4mIZTSu/DtjsDkkSdLWzQVASZIkSZIkaQgy8x0tyle0qG0YfwlwSYv6DcANLeoPA4e1qD8DnD6cOSRJ0tbNW4BKkiRJkiRJkiRJWxAXACVJkiRJkiRJkqQtiAuAkiRJkiRJkiRJ0hZk2AuAEfGG8mBjSZIkbUHMeZIkqU7MLpIkSe1VuQLwFmB6pxuRJElSz5nzJElSnZhdJEmS2qiyAPg48PtONyJJkqSeM+dJkqQ6MbtIkiS1UWUBcBHwZx3uQ5IkSb23CHOeJEmqj0WYXSRJklqqsgD4YeCVEfHxiBjb6YYkSZLUM+Y8SZJUJ2YXSZKkNratcMwFwD3AhcDZEXEX8Gsg+43LzDx7M/uTJElS95jzJElSnZhdJEmS2qiyADitaful5dVKAoYrSZKk+pjWtG3OkyRJo920pm2ziyRJUpMqC4CTOt6FJEmSRgNzniRJqhOziyRJUhvDXgDMzF+ORCOSJEnqLXOeJEmqE7OLJElSe1WuAHxBROwI7AfslJk/7kxLkiRJ6jVzniRJqhOziyRJ0sa2qXJQREyIiG8Ca4ClwC1N+46KiPsi4pjOtChJkqRuMedJkqQ6MbtIkiS1NuwFwIjYC1gMTAWuB24FomnIYmAP4O2daFCSJEndYc6TJEl1YnaRJElqr8oVgBfRCE9/nplvBRY278zM54EfA0dufnuSJEnqInOeJEmqE7OLJElSG1UWAE8GFmTmogHGrABeVqkjSZIk9Yo5T5Ik1YnZRZIkqY0qC4B7Ag8OMuZ5YMcK55YkSVLvmPMkSVKdmF0kSZLaqLIAuBrYZ5Ax+wG/rnBuSZIk9Y45T5Ik1YnZRZIkqY0qC4A/Ad4SES9ttTMiJgNTgFs2pzFJkiR1XddzXkTsExG3RMT9EXFvRLy3U+eWJElbPH+jkiRJaqPKAuA/AC8CfhgRJwE7AETEjuXzd4D1wGc71qUkSZK6oRc5by1wfma+GjgCOCci9u/g+SVJ0pbL36gkSZLa2Ha4B2Tm4oiYDnwRuL5p11PlfS3wV5l5bwf6kyRJUpf0Iudl5kpgZdl+OiLuB/YG7uvUHJIkacvkb1SSJEntVbkCkMy8EjgQ+AJwO/AQcCcwE/jjzPxaxzqUJElS1/Qy50XEROAQYHGLfdMjYulIzS1JkurJ36gkSZJaG/YVgBtk5oPA33WwF0mSJI0Cvch5EbET8E3gvMx8qv/+zJwNzJ41a1Z2sy9JkjT6+RuVJEnSpipdAdhtEbFPRNwSEfdHxL0R8d5S3zUiFkbEg+V9l1KPiPhCRCyLiLsj4jVN5zqrjH8wIs5qqr82In5ejvlCRETVOSRJkjR0ETGWxuLf1zLzW73uR5IkSZIkqe4qXwFY/kr7VBq3adoZeBL4KfDtzPxtZ9p7wVrg/My8MyL+CLgjIhYC04CbM/PSiJgBzAA+CJwETC6vw4FZwOERsStwEXAokOU8CzJzTRkzHbgNuAGYAnyvnHPIc3T4e0uSJHVdN3Ne+aOrK4D7M/MfO3luSZK0dejyb1SSJEm1UGkBMCJOp/GA5fFANO1K4HMR8e7M/EYH+mucNHMlsLJsPx0R9wN7A1OBY8qwecAiGotzU4GrMjOB2yJifETsVcYuzMzV5XssBKZExCLgxZl5a6lfBZxCYwFwWHOUXiVJkmqp2zkPOBL4S+DnEfGzUrswM2/o4BySJGkL1YPsIkmSVAvDXgCMiDcCVwPrgatoLIj9GngpcCzwTuDqiPhNZn6/c62+MP9EGn/RtRjYc8OCW2aujIg9yrC9gUeaDusrtYHqfS3qVJhjowXAiJgOTJ85c+Zwv6okSVJX9SLnZea/s/GPdZIkSUPS69+oJEmSRrMqVwB+BHgWeH1m3tlv37yIuBz4URnX0XBVbunwTeC8zHyqPKav5dAWtaxQH7CdoRyTmbOB2bNmzRrsfJIkSb3Ws5wnSZJUgdlFkiSpjW0qHHMIcE2LYAVAZi4FrgVeszmN9RcRY2ks/n0tM79Vyo+VW3tS3leVeh+wT9PhE4BHB6lPaFGvMockSVJd9STnSZIkVWR2kSRJaqPKAuCz9LvNZQuPlnEdEY1L/a4A7s/Mf2zatQA4q2yfBVzXVD8zGo4Aniy38bwJOCEidomIXYATgJvKvqcj4ogy15n9zjWcOSRJkuqq6zlPkiRpM5hdJEmS2qhyC9AfA0cNMuZIGrdY6JQjgb8Efh4RPyu1C4FLgWsj4mxgBXB62XcDcDKwDPgd8C6AzFwdER8HlpRxH8vM1WX7PcBcYHvge+XFcOeQJEmqsV7kPEmSpKrMLpIkSW1UWQD8IHBrRFwKfDwz/8+GHRGxI3ARcCDwZ51pETLz32n9zD2A41uMT+CcNueaA8xpUV9Ko+/+9SeGO4ckSVJNdT3nSZIkbYauZ5eImAO8GViVmQeW2q7ANcBEYDnwtsxcU+4y9Xkaf0D+O2DahtuVRsRZwIfLaT+RmfNK/bX84Q/UbwDem5lZZQ5JkrR1G3QBsASb/u4G3g9Mj4g7gceAPWncU31nGn9Z9QHg7M61KkmSpE4y50mSpDoZJdllLnA5cFVTbQZwc2ZeGhEzyucPAicBk8vrcGAWcHhZzLsIOBRI4I6IWJCZa8qY6cBtNBYAp9C4S9Ww5ujQd5UkSTU2lCsApw2wbzxwXIv60cAb8IchSZKk0WzaAPvMeZIkabSZNsC+rmSXzPxRREzsV54KHFO25wGLaCzOTQWuKneRui0ixkfEXmXswg2PpYmIhcCUiFgEvDgzby31q4BTaCwADmuOzBzs2YiSJGkLN5QFwEkj3oUkSZJ6wZwnSZLqZLRmlz03LLhl5sqI2KPU9wYeaRrXV2oD1fta1KvMsckCYERMB6bPnDlz2F9QkiTVz6ALgJn5y240IkmSpO4y50mSpDqpYXaJFrWsUK8yx6bFzNnA7FmzZg12TkmStAXYptcNSJIkSZIkSTX2WLm1J+V9Van3Afs0jZsAPDpIfUKLepU5JEnSVm4otwDdRERMAP4OOJhGsBjbYlhm5r6b0ZskSZK6zJwnSZLqZJRklwXAWcCl5f26pvq5ETEfOBx4sty+8ybgkxGxSxl3AnBBZq6OiKcj4ghgMXAmcFmVOUbwu0qSpJoY9gJgRBwD3AC8CFgLPFbeNxm6WZ1JkiSpq8x5kiSpTnqRXSLiauAYYPeI6AMuorEod21EnA2sAE4vw28ATgaWAb8D3gVQFvo+Diwp4z6WmavL9nuAucD2wPfKi+HOIUmSVOUKwM8AY2j8FdLXM3N9Z1uSJElSj5jzJElSnXQ9u2TmO9rsOr7F2ATOaXOeOcCcFvWlwIEt6k8Mdw5JkrR1q7IAeBBwdWZ+tdPNSJIkqafMeZIkqU7MLpIkSW1sU+GYNcDqQUdJkiSpbsx5kiSpTswukiRJbVRZALweOLrTjUiSJKnnzHmSJKlOzC6SJEltVFkAvBDYOSL+OSJ27HRDkiRJ6hlzniRJqhOziyRJUhvDfgZgZj4eEVOAxcCZEfFfwJOth+YmDyeWJEnS6GTOkyRJdWJ2kSRJam/YC4ARcQBwC7BLKR3SZmhWbUqSJEndZ86TJEl1YnaRJElqr8otQP8R2A34CPByYGxmbtPiNaajnUqSJGmkmfMkSVKdmF0kSZLaGPYVgMCfAt/KzE90uhlJkiT1lDlPkiTVidlFkiSpjSpXAD4HLO9wH5IkSeo9c54kSaoTs4skSVIbVRYAFwGHdbgPSZIk9d4izHmSJKk+FmF2kSRJaqnKAuAHgP0jYkZERKcbkiRJUs+Y8yRJUp2YXSRJktqo8gzADwP3AJcA/ysifgY82WJcZubZm9OcJEmSusqcJ0mS6sTsIkmS1EaVBcBpTduTyquVBAxXkiRJ9TGtaducJ0mSRrtpTdtmF0mSpCZVFgDbhSlJkiTVmzlPkiTVidlFkiSpjWEvAGbmL0eiEUmSJPWWOU+SJNWJ2UWSJKm9KlcASpIkSeqh6667bkjjpk6dOuD+J554ggsvvJDrrruOJ598kpe//OVceOGFnHnmmZ1oU5IkSZIk9YgLgJIkSVLNPf744wDsvvvuQz7mqaee4g1veAM77bQTl112Gbvvvjv33Xcfzz333Ei1KUmSJEmSumTYC4AR8fAQh2Zm7jvc80uSJKk3zHn10f/Kvg1XBA52xV+zT37ykzz77LMsXbqU7bffHoBjjz22c01KkjTCzC6SJEntVbkCcBsgW9R3BsaX7UeB56s2JUmSpJ4w53XRtGnTuOeee/jUpz7F+eefz0MPPcQhhxzCl770JQ444IARn//KK6/kvPPOe2HxT9LIOu91Yzb6POGPomX9c0vWda0naQtgdpEkSWpjm+EekJkTM3NSi9euwH7AjcBDwKs73awkSZJGjjmv+1asWMH73/9+PvShD3H11VezatUq3va2t5HZ6rfMP1i7du1Gr3Xr1rFu3bqNagOd4xe/+AWrVq1i/PjxnHzyyYwbN46XvOQlvO997/MWoFKX9D2d9D098H/rkgZmdpEkSWqvo88AzMxlEfFW4B7gIuCCTp5fkiRJvWHOGxmrV6/mJz/5CZMnTwZg/fr1nHrqqTzwwAO86lWvanvc2LFjBz12UgPHAAAgAElEQVT3lVdeybRp01ru+/Wvfw3ABz7wAc444wxuvPFG7rrrLi688EK23XZbPvOZzwz/y0gakFf2Sd1ldpEkSVu7ji4AAmTmMxGxEHgHhitJkqQthjmv8yZOnPjC4h/A/vvvD0BfX9+AC4BLlizZ6PMPf/hDAI4++ugXapMmTWp7/Pr16wE44IAD+PKXvwzAcccdx9NPP80nP/lJLr74YnbYYYdhfhtJkkYXs4skSdqadXwBsFgLvHSEzi1JkqTeMed10Pjx4zf6PG7cOACeeeaZAY87+OCDN/q8YsWKTepjxmz8XLFmu+66KwDHHnvsRvXjjjuOiy66iIceeoiDDjpokO4lSaoFs4skSdoqdXwBMCJ2B04FHun0uSVJktQ75rzRY3NvAbrvvvu+sNjYbMNzA7fZZtiPCpckadQxu0iSpK3ZsBcAI+IjA5xrH2AqsDPeWkGSJKlWzHn1sbm3AB03bhxvfOMb+cEPfrBR/eabb2aHHXbgFa94RQe7lSRpZJhdJEmS2qtyBeDFg+x/CvhEZn6mwrklSZLUOxcPsr/jOS8i5gBvBlZl5oGdOu+W7tBDD93o869+9auW9YF85CMf4aijjuJd73oX73jHO7j77ru59NJL+fu//3u22267jvYrSdIIuXiQ/f5GJUmStlpVFgCPbVNfD6wB/jMz11ZvSZIkST3Si5w3F7gcuKrD59UgDjvsML7zne9wwQUX8PWvf5099tiDD33oQ1xwgRdJSJJqw9+oJEmS2hj2AmBm/nAkGhlIq78Mj4hdgWuAicBy4G2ZuSYiAvg8cDLwO2BaZt5ZjjkL+HA57Scyc16pv5bGj0/bAzcA783MrDKHJElSXfUi52XmjyJiYrfnHQ3mzp27SW3ixIkvPIevG0488UROPPHErs0nSVIn9SK7SJIk1cU2vW5giOYCU/rVZgA3Z+Zk4ObyGeAkYHJ5TQdmwQsLhhcBhwOHARdFxC7lmFll7IbjplSZQ5IkSSMjIqZHxNJe9yFJkiRJklQHlRYAI+LoiLg+IlZFxPMRsa7Fq2O3WMjMHwGr+5WnAvPK9jzglKb6VdlwGzA+IvYCTgQWZubqzFwDLASmlH0vzsxbs/Hn1lf1O9dw5pAkSaq1bue8ocrM2Zk59AfcSZKkrcJozS6SJEm9NuxbgEbEm4B/BcYAK4AHgF4EqT0zcyVAZq6MiD1KfW/gkaZxfaU2UL2vRb3KHCv7NxkR04HpM2fOHPYXlCRJ6qZRlPMkSZIGZXaRJElqb9gLgMDFwPPAmzLz3zrbTkdEi1pWqFeZY9Ni5mxg9qxZs7r3MBdJkqRqLmZ05zxJkqRmFzOKsktELAeeBtYBazPz0PJImmuAicBy4G2ZuSYiAvg8cDLwO2BaZt5ZznMW8OFy2k9k5rxSfy2Nx+RsD9wAvDczs90cI/x1JUnSKFflFqAHAteMgmD12Ibbbpb3VaXeB+zTNG4C8Ogg9Qkt6lXmkCRJqrOu57yIuBq4FXhlRPRFxNndmluSJNXeaPmNqtmxmXlw063LZwA3Z+Zk4ObyGeAkYHJ5TQdmAZTFvIuAw4HDgIsiYpdyzKwydsNxUwaZQ5IkbcWqXAH4WzZ9Hl8vLADOAi4t79c11c+NiPk0wtKT5fadNwGfbApNJwAXZObqiHg6Io4AFgNnApdVmWMEv6skSVI3dD3nZeY7ujnfQN552/O9bqGyg55aD8A1Nf0OXz9ibK9bkCTV02j5jWogU4FjyvY8YBHwwVK/KjMTuC0ixpc/Pj8GWJiZqwEiYiEwJSIWAS/OzFtL/SrgFOB7A8whSZK2YlWuALwZ+NNONzKQNn8Zfinwxoh4EHhj+QyNWyA8DCwDvgz8vwAlOH0cWFJeH9sQpoD3AP9SjnmIRnhiuHNIkiTVXNdzniRJ0mYYbdklgX+LiDsiYnqp7bnhj8bL+x6lvjfwSNOxfaU2UL2vRX2gOTYSEdMjYmnF7yZJkmqmyhWAHwRuj4gPA5eUv1QaUQP8ZfjxLcYmcE6b88wB5rSoL6Vx24j+9SeGO4ckSVKNdT3nSZIkbYbRll2OzMxHI2IPYGFE/OcAY6NFLSvUhywzZwOzZ82a1et/J0mS1AVVFgAvAu4FPgr8VUT8DPhNi3GZ+X/bu/NoOapyYePPSwiDKJMKIihBQIFPUBFREK9EhCgOF6cIOEVUMM4TXicURxQV7od4UWTSxSioGJQpCkGDjBdQFCGgIkbEIBBAhpCQff/Y1dCpdJ+xz+nqque3Vq3O2TXt7rd27ReqalfyHS6SJEmDwzxPkiQNkkrlLimlW4vPRRHxE/I7/P4ZERsVr6fZCFhULL4QeErb6psAtxblu5bK5xXlm3RYniH2IUmSGmwsFwBntf17WjF1kgD/x5AkSdLgmNX272mY50mSpGqb1fbvafQxd4mItYBVUkr3Fv/eA/gCMAd4G/m1Mm8DflqsMgd4X0ScCjwfuLu4gHce8JWIWK9Ybg/gkymlOyPi3oh4AXAZ8FbgW23b6rQPSZLUYGO5ALhZz2shSZKkKjDPkyRJg6RKucuGwE8iAvL/bzs5pXRuRFwB/DAi3gHcAryhWP5sYE/gJuB+4O0AxYW+LwJXFMt9IaV0Z/Hv2cAJwJrAOcUE+cJfp31IkqQGG/UFwJTSXyeiIpIkSeov8zxJkjRIqpS7pJT+DDyrQ/kdwG4dyhPw3i7bOg44rkP5lcAzR7oPSZLUbKv0uwKSJEmSJEmSJEmSescLgJIkSZIkSZIkSVKNeAFQkiRJkiRJkiRJqhEvAEqSJEmSJEmSJEk14gVASZIkSZIkSZIkqUa8AChJkiRJkiRJkiTViBcAJUmSJEmSJEmSpBrxAqAkSZIkSZIkSZJUI14AlCRJkiRJkiRJkmpk1X5XQJIkSdLobHvjz1b4+7H339Gx/NotXzlpdZIkSZIkSdXhBUBJkiRpwP37MY/vdxUkSZIkSVKFeAFQkiRJGjA+2SdJkiRJkobiOwAlSZIkSZIkSZKkGvECoCRJkiRJkiRJklQjXgCUJEmSJEmSJEmSasQLgJIkSZIkSZIkSVKNeAFQkiRJkiRJkiRJqhEvAEqSJEmSJEmSJEk14gVASZIkSZIkSZIkqUa8AChJkiRJkiRJkiTViBcAJUmSJEmSJEmSpBrxAqCkyvnpT3/KtttuyxprrME222zDaaed1u8qqUeMbX0ZW0mqFs/L0mCy7UpStXheri9jW1/G9lFeAJRUKfPnz+d1r3sd06dP55xzzuEVr3gF++yzD+eff36/q6ZxMrb1ZWwlqVo8L0uDybYrSdXiebm+jG19GdsVRUqp33VohKOOOioBzJ49e0K2v++lSydkuxreyS+Y2u8q1MqMGTNYunQpF1xwwSNle+65J/fccw/z58/vY800Xsa2vhoU2+h3BTSxOZX5VP+YT/VWg87LAMReR/S7Co2VzvxAv6tQKw1qu+ZUfeb/o6ovc6reatB5uXGMbX01KLYjyqd8AlDSCmbNmsUOO+zA3Llz2W677VhrrbXYZZdd+MMf/jDh+16yZAkXXnghM2fOXKF877335pJLLuHuu++e8DrUmbGtL2MrSdXieVkaTLZdSaoWz8v1ZWzry9hWixcAJa3klltu4cADD+TTn/40p5xyCosWLWLmzJkM98TwsmXLhp2G2saf/vQnli5dylZbbbVC+dZbb83y5ctZsGBBT75fkxnb+jK2klQtnpelwWTblaRq8bxcX8a2voxtdaza7wpIqp4777yTiy++mC233BKA5cuX85rXvIYbbrhhpRNou6lThx9q4vjjj2fWrFkd5911110ArLvuuiuUr7feeivM19gZ2/oytpJULZ6XpcFk25WkavG8XF/Gtr6MbXV4AVDSSqZNm/bICRpgm222AWDhwoVDnqSvuOKKYbe92WabDbtMxIpDGLfu7CiXa/SMbX0ZW0mqFs/L0mCy7UpStXheri9jW1/Gtjq8AChpJeW7JFZbbTUAHnzwwSHXe/aznz3stqdMmdJ1XutujMWLF69Q3vq7XC+NnrGtL2MrSdXieVkaTLZdSaoWz8v1ZWzry9hWh+8AlNQzU6dOHXb6/ve/33X9zTffnKlTp3L99devUH799dezyiqr8PSnP32iv4K6MLb1ZWwlqVo8L0uDybYrSdXiebm+jG19Gdve8wlAST0z3se0V199daZPn87pp5/OAQcc8Ej5aaedxk477cQ666zTk3pq9IxtfRlbSaoWz8vSYLLtSlK1eF6uL2NbX8a297wAKKlndthhh3Fv46CDDmLXXXflQx/6EHvttRdnn302Z599Nueee24PaqixMrb1ZWwlqVo8L0uDybYrSdXiebm+jG19GdvecwjQcYiIl0XEDRFxU0R8ot/1kepgl1124YwzzuAXv/gFM2bMYM6cOZx88snsscce/a6axsnY1pex1XiZU0m95XlZGky2XY2H+ZTUe56X68vY1pexXVGklPpdh4EUEVOABcDuwELgCmCflNJ1nZY/6qijEsDs2bMnpD77Xrp0Qrar4Z38gqn9roIkaXJEvytQR1XKqcyn+sd8SuMRex3R7yo0VjrzA/2uggaTOVWPVSmfAnOqfjKnkqTGGFE+5QXAMYqInYCDU0ozir8/CZBSOqTT8q3kSpIkDa7Zs2f7P6x6zJxKkqTmMafqLfMpSZKaZyT5lEOAjt3GwN/a/l5YlK0gIvaPiCsnrVaSJEmDxZxKkiRpfMynJEnSSlbtdwUGWKerqyvdQZVSOho4euKrM7gi4sqU0vjf8KlKMr71ZWzry9hqkplT9Yhtt76MbX0Z2/oytppk5lM9YtutL2Nbb8a3vozt+PgE4NgtBJ7S9vcmwK19qoskSdKgMqeSJEkaH/MpSZK0Ei8Ajt0VwJYRsVlErAbsDczpc50kSZIGjTmVJEnS+JhPSZKklTgE6BillJZFxPuA84ApwHEppT/0uVqDyuEn6s341pexrS9jq0ljTtVTtt36Mrb1ZWzry9hq0phP9ZRtt76Mbb0Z3/oytuMQKa00JLgkSZIkSZIkSZKkAeUQoJIkSZIkSZIkSVKNeAFQkiRJkiRJkiRJqhEvADZYRBwcESkibuwy/6Zi/sFtZSdExJUTXK95xX5TRCyNiEUR8cuIeG9ErF5adte2ZVvL/zkiDo2ItUrL3ty23JKIuDUizo6It0REZdpCVeNSFU09PtqOi/M6zDsjIub1oVodNSlGgxSXqmjS8aFmqGq/3fS2VtW4VEVTj49B6rebFKNBiktVNOn4UHNUte9uenuralyqoqnHxyD13U2K0SDFpSqacHysOpk7UyU9CGwWETuklB7pnCPiecCmxfx2XwTWnIR6XQh8inyR+onArsDXgHdExEtSSotLy78J+DMwFdge+DKwLrB/abmTgW+RX4q9ETADOBZ4U0S8OqX00IR8m9GralyqosnHxx4R8byU0hX9rsgwmhajQYlLVTTt+FD9VbXfbnpbq2pcqqLJx8eg9NtNi9GgxKUqmnZ8qBmq2nc3vb1VNS5V0eTjY1D67qbFaFDiUhX1Pj5SSk4NnYCDgX8B5wPfKM37JnBeMf/gSa7XPOCMDuXbAf8Gjm8r2xVIwDNLyx4KLCqV3Vz+nkX57sAy4HP9jkmV41KVqanHR3Fc3AH8FjizNO8MYF6/Y9PEGA1SXKoyNen4cGrGVNV+u+ltrapxqcrU1ONjkPrtJsVokOJSlalJx4dTc6aq9t1Nb29VjUtVpqYeH4PUdzcpRoMUl6pMTTg++v7IsCrhVGBmRARA8TmzKF9BlB7jj4h1I+KY4jHWByPiloj4Xtv8TSLih8Xjsw9ExJ8i4otjqWRK6XfAkeSr5GsPs/i95KvwI9nuXPJJcPZY6jWBKhWX4tHlb0TEhyNiYUTcFRGnRsS6peU2i4gzI+KeiLg3Is6KiC1Ky6SI+GBEfCUibi/q8e3yo9Wj0ZDjIwFfAV4dEdsOtWBEPDvyI+v3F7E6KSI2bJs/rYjDzIj4bkTcXcT189H2KLpteEQGIi624Uq0YdVbpfrtbhrY1ioVF8/FlTg+BqLf7lr5+sZoIOJiG65EG1b9Varv7qaB7a1ScfF8XInjYyD67q6Vr2+MBiIutuHJOz68ACiAHwMbArsUf7+I/LjrT0aw7mHFeh8mP8b6KfKJpuUHwFPIj8C+nPxI7JgbHzCXRx+vbTclIlaNiDUj4oXAASOsf/t2N4yIaeOoW69VMS4zgd2K9f4LeCW5UwGgOLH+EtgaeBcwC9gMuCgi1i9t66PAk4E3A18nx+yDI6jDUJpwfJwOLAA+3W2BiHgi+Q6WxwD7Au8HXgzMjYjVSosfSr6j5fXAicBni3+32IZHZlDiYhuWJk4V++1umtTWqhgXz8X9Nyj9djd1jdGgxMU2LE2sKvbd3TSpvVUxLp6P+29Q+u5u6hqjQYmLbXgS+A5AkVJaHBHnAnsDvy4+zy3Kh1t9R+DbKaXT2spOLM3fJ6V0VvH3vHFWd2HxuWGp/JrS3/MZXSNv3+7No69W71U0LkuBvVJKywAiYpuiXu8p5r8deCrw9JTSn4tlLiOPi3wAcEjbtm5OKc0q/n1eccJ8LblTGavaHx8ppeUR8VXg2Ij4bEppQYfFPlp8zkgp3QMQEQuAy4DXAae0LfurlFJr+bkR8TJyHH5YlNmGR2CA4mIbrsg5XvVT0X67m8a0tYrGxXOx+dS8cX6FWsZogOJiG67IOV71VNG+u5vGtLeKxsXzsTnVvHF+hVrGaIDiYhuehOPDJwDVcirw+uLK+uvp8Ah/F9cAB0bEeyLi6V3mHxIRsyLiqT2oZ7esYm/gecDzgX2AJwBnRtvjyGPcbr9VLS4Xtk7KheuADdruDNkRuKp1UgZIKS0ELubRu8Razi/9fR2wSeuP4k6KR6YR1q8px8eJwC3AJ7vM3xE4v9WBA6SULid3KqOKA7bh0RiEuNiGpYlVtX67m6a1tarFxXNxNQxCv91NnWM0CHGxDUsTr2p9dzdNa29Vi4vn42oYhL67mzrHaBDiYhueBF4AVMsc4LHkR3TXAs4aevFHvA84k/zo7w0RcWNE7N02/43AlcDhwF8j4pqI2G0c9dy4+PxnqfwPKaUrU0qXp5ROJT/u+xJgz3Fut9+qFpfFpb8fIp+0Wifmjej8G/4TKD+a3Wlba7T9vbQ0jUQjjo+iczwUeHNEbNphkV7GwTY8QgMSF9uwNLGq1m9307S2VrW4eC6ugAHpt7upbYwGJC62YWniVa3v7qZp7a1qcfF8XAED0nd3U9sYDUhcbMOTwAuAAiCldB/wM/J43GcVf49kvcUppQ+klJ4EPIv8mPBJkR/ZJaX09+Lx28cDOwG3AXMi4vFjrOoe5Eb6v8Msd13xufUotntbSunmMdZrQgxQXFr+AWzQoXxD4M5Rbut5pWkkmnR8HAcsIo+RXdazONiGR21Q49LzOmIbVgMNUL/dqLY2QHFp8Vw8eQa13657jAY1Lj2vI7ZhNdQA9d2Nam8DFJcWz8eTZ1D77rrHaFDj0vM60uA27AVAtTuKfPfOd8ayckrpd8CB5ONqq9K85SmlS4HPk18u2unOgyFFxHbAe4ETU0r3DrP4M4vPv41gu7uThy44arR1miSVjkvJZcBzI2KzVkFEbAzsTB4PeTT1vrJ9Gm75ph0fKaUlwDeA/ch3xLS7DJgREY9rFUTE84BpjDIObfuzDY/AIMalQx1tw9L4VLrfbnBbq3RcSjwXT5JB7LebEKNBjEuHOtqGpfGrdN/d4PZW6biUeD6eJIPYdzchRoMYlw51tA2P00jHO1UDpJTmMcqXp0bEfOAnwO+BBLwLuA+4PCLWAc4DfgAsAFYnv2D0NuCPw2x6/Yh4ATkheDwwvdj2AuAjHZbfLiIeWyz/NOAg8jjH55SW26jY7hTgScAMYBYwlxVfHFoZFYvLcE4g31VyTkR8FngYOBj4F/DdcW67ncdH9l3gU+SO76K28sOA2eSX3n6NPETHV4FrgR+NdOO24TGrclyGcwK2YWlcKtZv29YKFYvLcE7Ac/FkqnK/3eQYVTkuwzkB27A0bhXru21vhYrFZTgn4Pl4MlW5725yjKocl+GcgG143LwAqPG6hHzgTiM3wquBl6eUFkZ+KfC1wAeBpwD3A5cCe6SUHhhmu9OLbS8jj+F7LfAJ4Jji7oWyk4rPBNwK/Br4TErp7tJy+xbTUuAO8ktK3wGclFJaPrKvPBAmKi5DSiktiYiXkjuRY8njNs8DXptSGu2j2UPx+ABSSvdHxOHk8ffby2+PiOnAN4FTyONenw18OKX00Ch28SC24VGreFyGq7ttWOoP86lqMp9aUS2Pj4r3242NUcXjMlzdbcNS/5hTVZM51YpqeXxUvO9ubIwqHpfh6m4b7oFIKU3WviRJkiRJkiRJkiRNMN8BKEmSJEmSJEmSJNWIFwAlSZIkSZIkSZKkGvECoCRJkiRJkiRJklQjXgCUJEmSJEmSJEmSasQLgJIkSZIkSZIkSVKNeAFQkiRJkiRJkiRJqhEvAEqSJEmSJEmSJEk14gVASZIkSZIkSZIkqUa8AChJkiRJkiRJkiTViBcAJUmSJEmSJEmSpBrxAqAkSZIkSZIkSZJUI14AlCRJkiRJkiRJkmrEC4BSn0TE/IhY1u96qD8i4qURkSLiM/2uyyCIiIURcVO/6zEeEbFFEfNj+l0XSaoL86lmM58aHfMpSVI35lTNZk41OuZUGiReAFRjRMTJxYlt9giWnVssu9dk1G0yjecEX4cObtC0JWGjmTYp1l1YKl8eEYsj4uKIeHdETCnta/4w2121tPy6EfHfxXq3RsSDEbEoIi6PiA9ExGMm87eSJE0886nMfGqwmE9JkqrGnCozpxos5lTS4Fl1+EWk2jga2Ad4F3BUt4UiYhqwG/AP4GcTWJ99gTUncPuqhz8Dny+VrQ+8H7gLOKLDOveU/j68KJsCPA14LbAz8BJgZmnZBHyhS12Wl/5+Ark9XQH8HLgdWAd4KfD/gXdGxM4ppX932V7T/BXYGljc74pI0jiYT2kQmU/Vh/mUpLowp9IgMqeqD3OqhvACoBojpTQvIhYAz4mI7VNKV3VZ9B1AAMenlCZs+IOU0i0TtW3VR0rpz8DB7WURsQU5ubozpXRwh9XKDkspLWxbfzvgMuANEfHClNLFbcsuH+E2Af4CrFNuJxERwMnA3sD+wGEj3F6tpZSWAtf3ux6SNB7mUxpE5lP1YT4lqS7MqTSIzKnqw5yqORwCVE3zveLzXZ1mFo+bv518h8kxbeUbR8TnIuI3EXFbRDwUEX+PiJMiYqsO23lkCIOIeEZEnB4RtxePt+9SLLPS+OoRsXpEvD8izomIv0bEkoi4sxjuYUaXOi+MiJsiYq2I+GZE3FKsd2NEfKzo6FrLfgm4sfjzHaVH59/c7UdrPeIPbAxsXlrvmIh4fEQ8EBE3tO+vtI1zi+Wf1eE32iYi5hTf9b6I+HVEvHSI+rwpIuYVQwU8GBHXRcSnImK1but02MYzIuJrEXFlEZslEXFzRHw3Ijbu9htExGci4jlFjO4u6jsvIp7fZT8bRcTxxbADD0TE1RHxlpHWcyKklH4H/Lr4c8dxbOfhTv8BklJKwBnFn1uOdHuRfaCI55KijR0REWsPsc4aRex/X/y+90TEryLi9R2WbT/mtoyIHxfH3D3F8blNsdwGxTL/KI6vyyPixR22N+bzQqn8xKL8KRHxnuK7PFhs8ztDfX9J6hPzKfOp1jbMpzLzKfMpSRoLcypzqtY2zKkycypzKvWYTwCqab4PfBnYNyI+mlK6vzT/5eQEYm5K6S9t5dOBjwMXAlcD95E7jZnAqyI/Qv77Dvt7OnA5cB1wIvAY4N4h6vdE4L+B3wBzyY+rbwS8GjgnIvZLKZ3QYb3VgF8AGwBnAw8DrwG+DqxefGeAC4C1yXfmXA3MadvG74aoV+sR/48Ay1jxkf6rUkp3RMTpwFuAXcm/0yMiYlNgd+CylNJvS9veArgEuAb4Lvn3nwmcFxEzU0o/Km3r+8BbgVvIHfjd5KECvgy8JCJmpJQeHuK7tLyBfOfPhcDFwFJgW3Li/cqI2CGl9I8O6+0IfBqYT07WNwVeB1wQEc9OKbWSVyJiA3IspwG/Kv69cbHeeSOo40RqJcFppRkRbyQPw7CEfOxemFJaMsrtv6r4HOq4KjsSeA9wK/lYWAbsRf7NpwIPluq5Ormd7FLU80hgLXJsT4+IL6aUPtthP08j3132e+B4YDNye5kXETsX27wLOA14PPkusXMjYsv2u9QY+3mhm2+S28nPyMfHbsABwOZFuSRVhfmU+VSL+VRmPmU+JUljYU5lTtViTpWZU5lTqddSSk5OjZrIJ8wEzOow76fFvNeXyjcEHtth+eeQT6hnlcq3KLaTgC90qcd8YFmpbA1g4w7Lrgv8kZxsrV6at7DYz1nAmm3lTyKPqX0nMKVD3Y4Zw2+3ELipy7wXFNs9tcO8LxXz3t7lN/pqafnnkzvWO9p/d+CdxfI/BNYorfPFYt57R/hdNin/lkX5y8njiH+rVP7Stvq+uTTvvUX5EaXy44ryr3f5fgn4zBji0PrtOsaiw7GxSal8O+CBYt5OpWMydZhuA14zxH5WIw8BcTA58f5tsd5cYLURfqf/KNZZAKzXVr4m+T9QVvq+wEFF+Rxg1dKx/7cijs/vcsz9V2lbny/K7wS+DazSNq91x2U5jmM9LxxTKj+xKP9Le6zICeVvinnbj/Y4cXJycprICfMp86lkPoX5lPmUk5OT0zgnzKnMqZI5FeZU5lROEzb1vQJOTpM9ke9YSMD8UvlG5DtsbgOmjmJ7ZwP30zmB+Xu3zoUOydUw+/l4sc2dS+WtDnRah3VOKuZt1aFuPU2uivlXke/IeWJb2arku2UWA4/pUI87gLU6bKvV4byprezaYvtrd1h+VfIdMb/pwTFyHbCgVNZKrlO+BGQAAAkNSURBVOZ1WH518h1tl5bKHii+9+OG+H6TkVwdRk58vljst5VY/bC0/EeBPYEnkxP9ZwBfBR4iJ4O7d9nPY1k5ITu+U1yHqOvxxXpv6TCv9duXk6u/FL/7lh3WOaBY5+hOvxttyVMx72nFvHvK9SYnOQ+T77rsxXmhW3I1q8N23lXMe/d4j2snJyenXk6YT5lPDf89zafMp9rnmU85OTk5dZgwpzKnGv57mlOZU7XPM6dyGtXkEKBqoguAPwEvjIitU0p/LMrfTu6gT0j5RagriIhXk0/YzyU/cl1uP+uT735qd01K6aHRVC4itgUOJD8y/mRyJ91upbG/gTtSSjd3KP9b8bneaOowDkcBR5N/y0OLsleRE9dvpZWHswD435TSfR3K5wFvIt+pclJEPA54JvBP4CPReRj3B4GtR1LRyBt4C/A28t1G6wFT2hbpVFeAK8sFKaUlEXE7K/7O25ATlEtSSp2G1JhH/n6T4cPFZyIP73E1uUP/bvtCKaVvlta7AfhERNwGHA58hXzHFKX1/k0xPDr5mN0DOAS4IiJelkb2MvHti8+LOsz7FflOqUdExHrkYSv+mtqGtGhzQfH5nA7zrk4pLS+V3Vp83lA+HlNKS4v4blLe0BjPC92sdGwx+W1YkkbKfGrimE89ynzKfArMpyTVmznVxDGnepQ5lTkVmFM1khcA1TgppdYLTg8hP67/0aJT2I/izofyOhHxEfLYx3eSxzH/K4/eofJa8rjc5SQI8p1aIxYRLyy2vwrwS/JwD/eSO5btyYlKp/0s7rLJ1stvp3SZ32snkcd03z8ivp5SSuSOB3LS1ck/u5S3frt1is/1i88Ngc8NUYeVXvjbxRHA+8id6rnkO+Fa43fvR04SOhnqt27/nVv1Hu77TYanpBXHBR+to8nH/3Mj4jFdkmSKeP8dOD4ibiS/xPkI8hjpw+n6e6WUHoqIu7os32kM/PbydTvMu7tD2bIh5rXmT20vGMd5oZtOx9Zkt2FJGhHzqQllPvUo8ynzKfMpSbVmTjWhzKkeZU5lTmVO1VBeAFRTHQ98AXhrRHwSeBH5JaYXpJRual8wIqaSH02/lTzG8T9L8180xH7SKOt1EPmOnBellOaX9nMQj760tpJSSvdHxA/IL3DerehgdwcuTt1fNLthl/InFZ93lz6vSCntOJ56RsRG5DHRfwvsUtwd1D7/LePZfqFV3+G+X+UVcb0PeBz5JeHd7jxrX2d+RNxLfuH2SLT/XivcjRURq5HvLlrcYfluv+NGpeV6apznBUmqC/OpCWA+tQLzKfMpSWoCc6oJYE61AnMqcyo11Cr9roDUD8WJcA7wBPKdH+8sZnW6A2hDcqcyv8MJdG06P749VlsAi8qJVeHFPdrHw8XnWO7WeHgE6/1P8XkAeWzoVSg9xl/y3IhYq0P5rsXn1QAppcXkx/23jYhOd8yMxuZAAOd1SKw2JT+2P17Xke/W2r4YGqJs1x7sY1JExP8jt4G7yXcSjWSddcnjro/0breris9Ox/l/UOqvUkp3ke9mempEPK3DOtNL2+21yTwvSFIlmU8B5lPmUyNkPtWR+ZQkYU5VfJpTmVONiDlVR+ZU6soLgGqy7xWfHwVeA/wL+EmH5f5B7iSf154EFHd8fIvejnt8M/DEojN7REQcQH4xdC+0OsenjmHdO4ANIqLrI+MppevJY4f/J7B/sb/Th9jm+uS7yh4REc8H9ia/MPmnbbMOI999dmxErENJRKwfESPp1G4uPl8UEY8ki0USdDQ9ODemlJYAp5CHAfhsqZ6t71cZEbF5cddZuXwD4Ljiz5PbxyWPiO07JbrF8XEkOYH9+QircHzxeVD7NiNiTfK47p0cR47V1yNilbZ1NgA+3bbMRJjM84IkVZn51OiZT42Q+ZT51ATtV5KqyJxq9MypRsicypxqgvarAeAQoGqy84G/AK1H9Y/s9DLklNLDEXEk8DHg2oiYQx4z+SXkjvMienfn0+HkJOo3EfFD4J6ifjsBPwJeN94dpJTujogrgekRcSKwgDx++5lDDIHQ8kvyXSPnRsSvgYfIL6std6D/Q7576InA4SmlB+nuImB2ROwEXEx+gfQbyR3z/u13P6WUjo6I55KTthdHxPnkR/HXB55GHibje+Rx04f6DRZGxBnA64GrIuIX5FjuAfwbuJb8guTx+gT5OPlYROxY+n4/B17dg330ynTgOxFxEXATObHdFNgTWBu4nPx92u0H7BcRF5LvdLqHR1+wvCHwR+DjI9l5SulXEXEUMBv4QxGfZeS7H28HFnVY7WvAy8hjmf82Is4B1gLeQD72vpJSunRE336UJvm8IElVZj5lPmU+9SjzqVEwn5KkFZhTmVOZUz3KnGoUzKk0FC8AqrGKFy0fC3ypKPreEIt/knxy3488bMBiYC75Do5Delinn0fEfxbb3ZvcuVxOTlS2ogfJVeFN5DuV9gT2JScyNwPDJVefJ3e0ryQnMlOAY1n5DpqfkDvn9ej+YuWWm8hjnR9C7lhXB64APp9S+kV54ZTSARHxc3Icdie/QPcOcud+KHDiMPtrmVXs+w3F/heRh9w4CDhrhNsYUkppUZE0foX8m+0IXE8eduIfVCu5upL8guznkhPotcnJ0m+B04CjU0pLS+ucRh5v/QXAzuThFO4hDy1xKHBUSumBUdThfeQhNN5dTP8CfkxuD9eVF04pLYmI3ch3SO4DfABYClwDvD+ldNoo9j0Wk3JekKQqM58yn8J8qp351OiZT0kS5lSYU83CnKqdOdXomVOpo0hptO9/laShRcSW5E7yopTS9C7LbAHcCBybUnpnp2UkSZKaynxKkiRp/MypJDWZ7wCUNBEOJN+xdWS/KyJJkjSgzKckSZLGz5xKUmM5BKiknoiITcmPuD8DeBtwFZ1fWC1JkqQOzKckSZLGz5xKkjIvAErqlS3JY0rfB5wHzE4pLe9vlSRJkgaK+ZQkSdL4mVNJEr4DUJIkSZIkSZIkSaoV3wEoSZIkSZIkSZIk1YgXACVJkiRJkiRJkqQa8QKgJEmSJEmSJEmSVCNeAJQkSZIkSZIkSZJqxAuAkiRJkiRJkiRJUo38HzufvdWLj5a+AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 1800x864 with 6 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.close('all')\n",
    "fig = plt.figure(12, figsize=(25, 12))\n",
    "\n",
    "ax1 = plt.subplot2grid((2, 3), (0, 0)) #Jaiswal\n",
    "ax2 = plt.subplot2grid((2, 3), (0, 1)) #Genovese\n",
    "ax3 = plt.subplot2grid((2, 3), (0, 2)) #Zink WGS\n",
    "ax4 = plt.subplot2grid((2, 3), (1, 0)) #Acuna\n",
    "ax5 = plt.subplot2grid((2, 3), (1, 1)) #Coombs\n",
    "ax6 = plt.subplot2grid((2, 3), (1, 2)) #Young\n",
    "\n",
    "Jaiswal_p, Jaiswal = plotting_frequencies('Jaiswal 2014', Jaiswal_observations, Jaiswal_mutation_rates, ax1)\n",
    "Genovese_p, Genovese = plotting_frequencies('Genovese 2014', Genovese_observations, Genovese_mutation_rates, ax2)\n",
    "Zink_p, Zink = plotting_frequencies('Zink 2017', Zink_observations, Zink_mutation_rates, ax3)\n",
    "Coombs_p, Coombs = plotting_frequencies('Coombs 2017', Coombs_observations, Coombs_mutation_rates, ax4)\n",
    "Acuna_p, Acuna = plotting_frequencies('Acuna-Hidalgo 2017', Acuna_observations, Acuna_mutation_rates, ax5)\n",
    "Young_p, Young = plotting_frequencies('Young 2016 & 2019', Young_observations, Young_mutation_rates, ax6)\n",
    "\n",
    "plt.tight_layout()\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "if missense DBD is Poisson with mean 58\n",
    "if missense nonDBD is Poisson with 58*(missense_DBD_mut_rate/missense_nonDBD_mut_rate)\n",
    "what is probability of 1?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Overall p-value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 365,
   "metadata": {},
   "outputs": [],
   "source": [
    "l_value = -np.log(Jaiswal_p)-np.log(Genovese_p)-np.log(Zink_p)-np.log(Coombs_p)-np.log(Acuna_p)-np.log(Young_p)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- What is the area above l_value in l_value^5 x e^-l_value? (5 because number of values - 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 282,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "16.169622346455107"
      ]
     },
     "execution_count": 282,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "l_value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 317,
   "metadata": {},
   "outputs": [],
   "source": [
    "def integrand(l):\n",
    "    return ((l**5)*(np.exp(-l)))/(math.factorial(6))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 319,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0.0002043933140211318, 3.987876604314408e-10)"
      ]
     },
     "execution_count": 319,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "integrate.quad(lambda l: integrand(l), l_value, np.inf)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
