{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "toc": true
   },
   "source": [
    "<h1>Table of Contents<span class=\"tocSkip\"></span></h1>\n",
    "<div class=\"toc\"><ul class=\"toc-item\"><li><span><a href=\"#Python-Setup\" data-toc-modified-id=\"Python-Setup-1\"><span class=\"toc-item-num\">1&nbsp;&nbsp;</span>Python Setup</a></span><ul class=\"toc-item\"><li><span><a href=\"#Change-the-width-of-the-page\" data-toc-modified-id=\"Change-the-width-of-the-page-1.1\"><span class=\"toc-item-num\">1.1&nbsp;&nbsp;</span>Change the width of the page</a></span></li><li><span><a href=\"#Import-packages\" data-toc-modified-id=\"Import-packages-1.2\"><span class=\"toc-item-num\">1.2&nbsp;&nbsp;</span>Import packages</a></span></li></ul></li><li><span><a href=\"#Colours\" data-toc-modified-id=\"Colours-2\"><span class=\"toc-item-num\">2&nbsp;&nbsp;</span>Colours</a></span><ul class=\"toc-item\"><li><span><a href=\"#Colour-line-graph\" data-toc-modified-id=\"Colour-line-graph-2.1\"><span class=\"toc-item-num\">2.1&nbsp;&nbsp;</span>Colour line graph</a></span></li></ul></li><li><span><a href=\"#Totals-for-studies\" data-toc-modified-id=\"Totals-for-studies-3\"><span class=\"toc-item-num\">3&nbsp;&nbsp;</span>Totals for studies</a></span></li><li><span><a href=\"#Functions-for-calculating-trinucleotide-context-specific-mutation-rates\" data-toc-modified-id=\"Functions-for-calculating-trinucleotide-context-specific-mutation-rates-4\"><span class=\"toc-item-num\">4&nbsp;&nbsp;</span>Functions for calculating trinucleotide-context specific mutation rates</a></span><ul class=\"toc-item\"><li><span><a href=\"#Calculating-mutation-rates-for-individual-variants\" data-toc-modified-id=\"Calculating-mutation-rates-for-individual-variants-4.1\"><span class=\"toc-item-num\">4.1&nbsp;&nbsp;</span>Calculating mutation rates for individual variants</a></span><ul class=\"toc-item\"><li><span><a href=\"#DNMT3A\" data-toc-modified-id=\"DNMT3A-4.1.1\"><span class=\"toc-item-num\">4.1.1&nbsp;&nbsp;</span>DNMT3A</a></span></li><li><span><a href=\"#TET2\" data-toc-modified-id=\"TET2-4.1.2\"><span class=\"toc-item-num\">4.1.2&nbsp;&nbsp;</span>TET2</a></span></li><li><span><a href=\"#ASXL1\" data-toc-modified-id=\"ASXL1-4.1.3\"><span class=\"toc-item-num\">4.1.3&nbsp;&nbsp;</span>ASXL1</a></span></li><li><span><a href=\"#TP53\" data-toc-modified-id=\"TP53-4.1.4\"><span class=\"toc-item-num\">4.1.4&nbsp;&nbsp;</span>TP53</a></span></li></ul></li><li><span><a href=\"#Calculating-mutation-rates-from-a-.csv-file-of-variants\" data-toc-modified-id=\"Calculating-mutation-rates-from-a-.csv-file-of-variants-4.2\"><span class=\"toc-item-num\">4.2&nbsp;&nbsp;</span>Calculating mutation rates from a .csv file of variants</a></span><ul class=\"toc-item\"><li><span><a href=\"#DNMT3A\" data-toc-modified-id=\"DNMT3A-4.2.1\"><span class=\"toc-item-num\">4.2.1&nbsp;&nbsp;</span>DNMT3A</a></span></li><li><span><a href=\"#TET2\" data-toc-modified-id=\"TET2-4.2.2\"><span class=\"toc-item-num\">4.2.2&nbsp;&nbsp;</span>TET2</a></span></li><li><span><a href=\"#ASXL1\" data-toc-modified-id=\"ASXL1-4.2.3\"><span class=\"toc-item-num\">4.2.3&nbsp;&nbsp;</span>ASXL1</a></span></li><li><span><a href=\"#TP53\" data-toc-modified-id=\"TP53-4.2.4\"><span class=\"toc-item-num\">4.2.4&nbsp;&nbsp;</span>TP53</a></span></li></ul></li><li><span><a href=\"#Calculating-mutation-rates-from-a-list-of-variants\" data-toc-modified-id=\"Calculating-mutation-rates-from-a-list-of-variants-4.3\"><span class=\"toc-item-num\">4.3&nbsp;&nbsp;</span>Calculating mutation rates from a list of variants</a></span><ul class=\"toc-item\"><li><span><a href=\"#DNMT3A\" data-toc-modified-id=\"DNMT3A-4.3.1\"><span class=\"toc-item-num\">4.3.1&nbsp;&nbsp;</span>DNMT3A</a></span></li><li><span><a href=\"#TET2\" data-toc-modified-id=\"TET2-4.3.2\"><span class=\"toc-item-num\">4.3.2&nbsp;&nbsp;</span>TET2</a></span></li><li><span><a href=\"#ASXL1\" data-toc-modified-id=\"ASXL1-4.3.3\"><span class=\"toc-item-num\">4.3.3&nbsp;&nbsp;</span>ASXL1</a></span></li><li><span><a href=\"#TP53\" data-toc-modified-id=\"TP53-4.3.4\"><span class=\"toc-item-num\">4.3.4&nbsp;&nbsp;</span>TP53</a></span></li></ul></li></ul></li><li><span><a href=\"#Lists-of-variants-targeted-by-each-study\" data-toc-modified-id=\"Lists-of-variants-targeted-by-each-study-5\"><span class=\"toc-item-num\">5&nbsp;&nbsp;</span>Lists of variants targeted by each study</a></span><ul class=\"toc-item\"><li><span><a href=\"#Jaiswal-2014\" data-toc-modified-id=\"Jaiswal-2014-5.1\"><span class=\"toc-item-num\">5.1&nbsp;&nbsp;</span>Jaiswal 2014</a></span></li><li><span><a href=\"#Genovese-2014\" data-toc-modified-id=\"Genovese-2014-5.2\"><span class=\"toc-item-num\">5.2&nbsp;&nbsp;</span>Genovese 2014</a></span></li><li><span><a href=\"#McKerrel-2015\" data-toc-modified-id=\"McKerrel-2015-5.3\"><span class=\"toc-item-num\">5.3&nbsp;&nbsp;</span>McKerrel 2015</a></span></li><li><span><a href=\"#Zink-2017\" data-toc-modified-id=\"Zink-2017-5.4\"><span class=\"toc-item-num\">5.4&nbsp;&nbsp;</span>Zink 2017</a></span></li><li><span><a href=\"#Coombs-2017\" data-toc-modified-id=\"Coombs-2017-5.5\"><span class=\"toc-item-num\">5.5&nbsp;&nbsp;</span>Coombs 2017</a></span></li><li><span><a href=\"#Young-2016-&amp;-2019\" data-toc-modified-id=\"Young-2016-&amp;-2019-5.6\"><span class=\"toc-item-num\">5.6&nbsp;&nbsp;</span>Young 2016 &amp; 2019</a></span></li><li><span><a href=\"#Desai-2018\" data-toc-modified-id=\"Desai-2018-5.7\"><span class=\"toc-item-num\">5.7&nbsp;&nbsp;</span>Desai 2018</a></span></li><li><span><a href=\"#Acuna-Hidalgo-2017\" data-toc-modified-id=\"Acuna-Hidalgo-2017-5.8\"><span class=\"toc-item-num\">5.8&nbsp;&nbsp;</span>Acuna-Hidalgo 2017</a></span></li></ul></li><li><span><a href=\"#Lists-of-all-possible-variants-in-DNMT3A,-TET2,-ASXL1,-TP53\" data-toc-modified-id=\"Lists-of-all-possible-variants-in-DNMT3A,-TET2,-ASXL1,-TP53-6\"><span class=\"toc-item-num\">6&nbsp;&nbsp;</span>Lists of all possible variants in DNMT3A, TET2, ASXL1, TP53</a></span><ul class=\"toc-item\"><li><span><a href=\"#DNMT3A\" data-toc-modified-id=\"DNMT3A-6.1\"><span class=\"toc-item-num\">6.1&nbsp;&nbsp;</span>DNMT3A</a></span></li><li><span><a href=\"#TET2\" data-toc-modified-id=\"TET2-6.2\"><span class=\"toc-item-num\">6.2&nbsp;&nbsp;</span>TET2</a></span></li><li><span><a href=\"#ASXL1\" data-toc-modified-id=\"ASXL1-6.3\"><span class=\"toc-item-num\">6.3&nbsp;&nbsp;</span>ASXL1</a></span></li><li><span><a href=\"#TP53\" data-toc-modified-id=\"TP53-6.4\"><span class=\"toc-item-num\">6.4&nbsp;&nbsp;</span>TP53</a></span></li></ul></li><li><span><a href=\"#Actual-number-of-observations-of-each-variant\" data-toc-modified-id=\"Actual-number-of-observations-of-each-variant-7\"><span class=\"toc-item-num\">7&nbsp;&nbsp;</span>Actual number of observations of each variant</a></span><ul class=\"toc-item\"><li><ul class=\"toc-item\"><li><span><a href=\"#DNMT3A\" data-toc-modified-id=\"DNMT3A-7.0.1\"><span class=\"toc-item-num\">7.0.1&nbsp;&nbsp;</span>DNMT3A</a></span></li><li><span><a href=\"#TET2\" data-toc-modified-id=\"TET2-7.0.2\"><span class=\"toc-item-num\">7.0.2&nbsp;&nbsp;</span>TET2</a></span></li><li><span><a href=\"#ASXL1\" data-toc-modified-id=\"ASXL1-7.0.3\"><span class=\"toc-item-num\">7.0.3&nbsp;&nbsp;</span>ASXL1</a></span></li><li><span><a href=\"#TP53\" data-toc-modified-id=\"TP53-7.0.4\"><span class=\"toc-item-num\">7.0.4&nbsp;&nbsp;</span>TP53</a></span></li></ul></li></ul></li><li><span><a href=\"#Functions-for-calculating-the-expected-number-of-observations-of-a-variant\" data-toc-modified-id=\"Functions-for-calculating-the-expected-number-of-observations-of-a-variant-8\"><span class=\"toc-item-num\">8&nbsp;&nbsp;</span>Functions for calculating the expected number of observations of a variant</a></span></li><li><span><a href=\"#Maximum-Likelihood-Estimation-for-s\" data-toc-modified-id=\"Maximum-Likelihood-Estimation-for-s-9\"><span class=\"toc-item-num\">9&nbsp;&nbsp;</span>Maximum Likelihood Estimation for s</a></span><ul class=\"toc-item\"><li><span><a href=\"#DNMT3A-variants\" data-toc-modified-id=\"DNMT3A-variants-9.1\"><span class=\"toc-item-num\">9.1&nbsp;&nbsp;</span>DNMT3A variants</a></span></li><li><span><a href=\"#TET2-variants\" data-toc-modified-id=\"TET2-variants-9.2\"><span class=\"toc-item-num\">9.2&nbsp;&nbsp;</span>TET2 variants</a></span></li><li><span><a href=\"#ASXL1-variants\" data-toc-modified-id=\"ASXL1-variants-9.3\"><span class=\"toc-item-num\">9.3&nbsp;&nbsp;</span>ASXL1 variants</a></span></li><li><span><a href=\"#TP53-variants\" data-toc-modified-id=\"TP53-variants-9.4\"><span class=\"toc-item-num\">9.4&nbsp;&nbsp;</span>TP53 variants</a></span></li></ul></li></ul></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Python Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Change the width of the page"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>.container { width:80% !important; }</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.core.display import display, HTML\n",
    "display(HTML(\"<style>.container { width:80% !important; }</style>\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "# imported packages\n",
    "import random\n",
    "import numpy as np\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import math\n",
    "import matplotlib.ticker as plticker\n",
    "from matplotlib.ticker import LinearLocator, FormatStrFormatter\n",
    "from matplotlib.patches import Polygon\n",
    "from matplotlib.patches import Patch\n",
    "from matplotlib.lines import Line2D\n",
    "from matplotlib import cm\n",
    "import scipy.special\n",
    "import scipy.integrate as it\n",
    "from scipy import integrate\n",
    "from scipy.interpolate import interp1d\n",
    "from scipy.stats import kde\n",
    "import copy\n",
    "import glob, os\n",
    "import re\n",
    "from sklearn import datasets, linear_model\n",
    "import pandas as pd\n",
    "from decimal import *\n",
    "from operator import itemgetter    \n",
    "from collections import OrderedDict\n",
    "import timeit"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Colours"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "# Lists of colors for plots\n",
    "c0 = (0.76, 0.76, 0.76)\n",
    "c1 = (1.00, 0.18, 0.33);\n",
    "c2 = (1.00, 0.23, 0.19);\n",
    "c3 = (1.00, 0.58, 0.00);\n",
    "c4 = (1.00, 0.80, 0.00);\n",
    "c5 = (0.30, 0.85, 0.39);\n",
    "c6 = (0.35, 0.78, 0.98);\n",
    "c7 = (0.20, 0.67, 0.86);\n",
    "c8 = (0.00, 0.48, 1.00);\n",
    "c9 = (0.35, 0.34, 0.84);\n",
    "c10 = (0.00, 0.31, 0.57);\n",
    "c11 = (0.12, 0.29, 0.69);\n",
    "c12 = (0.17, 0.17, 0.42);\n",
    "c13 = (1.00, 1.00, 1.00);\n",
    "c14 = (0.77, 0.04, 0.00);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#define the colors from colorbrewer2\n",
    "orange1 = '#feedde'\n",
    "orange2 = '#fdbe85'\n",
    "orange3 = '#fd8d3c'\n",
    "orange4 = '#e6550d'\n",
    "orange5 = '#a63603'\n",
    "blue1 = '#eff3ff'\n",
    "blue2 = '#bdd7e7'\n",
    "blue3 = '#6baed6'\n",
    "blue4 = '#3182bd'\n",
    "blue5 = '#08519c'\n",
    "green1 = '#edf8e9'\n",
    "green2 = '#bae4b3'\n",
    "green3 = '#74c476'\n",
    "green4 = '#31a354'\n",
    "green5 = '#006d2c'\n",
    "grey1 = '#f7f7f7'\n",
    "grey2 = '#cccccc'\n",
    "grey3 = '#969696'\n",
    "grey4 = '#636363'\n",
    "grey5 = '#252525'\n",
    "purple1 = '#f2f0f7'\n",
    "purple2 = '#cbc9e2'\n",
    "purple3 = '#9e9ac8'\n",
    "purple4 = '#756bb1'\n",
    "purple5 = '#54278f'\n",
    "red1 = '#fee5d9'\n",
    "red2 = '#fcae91'\n",
    "red3 = '#fb6a4a'\n",
    "red4 = '#de2d26'\n",
    "red5 = '#a50f15'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Colour line graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAlkAAAExCAYAAABPt7ftAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAF0xJREFUeJzt3X2QnWd53/Hvj2KcENtgrTAOkW11hpeCiIDhOIEqGEcwQNrBocUW5EWyOmZ2mBiHiSs6oetpAq5CQsDQDs6MRcuwQMybGqfEaUiYJVZCB5geWfbSJVGdpn6R7WLWshkgDl7JV//Y48x22dUenec82j2r72dmZ3Tu+36uc9l//ea+n/M8qSokSZI0XE9Z7QYkSZLWI0OWJElSCwxZkiRJLTBkSZIktcCQJUmS1AJDliRJUgsMWZJGVpJLkxxZ7T4kaSmGLElrQpJfTNJN8r0kDyb5kyQ/s9p9SdKgDFmSVl2Sa4EPA78FPBu4EPg94OdXqZ+nrsb3SlpfDFmSVlWSZwDvBa6uqj+oqu9X1VxV/VFVvSvJmUk+nOSB3t+Hk5y5TK0XJrktyaNJZpJctmDutiRvW/B5d5KvLPhcSa5OchdwV+Z9KMlDSb6TZDrJi1v8XyFpnTFkSVptrwR+BLhlmfkJ4BXAS4GXAD8FXLd4UZIzgD8C/gw4D7gG+P0kLziJXt4E/DTwIuB1wCXA84FnAm8BHj6JWpJOc4YsSattDJitqmPLzP8S8N6qeqiqvg28B9i5xLpXAGcBv11Vj1fVl4FbgV84iV7eV1VHq+oxYA44G/gnQKrqr6rqwZOoJek0Z8iStNoeBjae4D6o5wD3LPh8T29sqXX3VdUTi9b+xEn0ct+T/+iFtI8ANwLfSrIvyTknUUvSac6QJWm1fRX4e+aP6pbyAHDRgs8X9saWWndBkqcsWnt/79/fB56+YO78JWrU//eh6j9W1cuBLcwfG75rmR4l6YcYsiStqqr6DvDvgBuTvCnJ05OckeTnkrwf+DRwXZJnJdnYW/upJUp9nfkg9W96118KvBH4TG/+DuBf9uo/F7jqRH0luTjJT/fu9fo+80HwePP/YkmnC0OWpFVXVTcA1zJ/Q/u3mT+2ewfwh8C/B7rANPAN4Pbe2OIajwOXAT8HzDL/CIhdVfXXvSUfAh4HvgVMAr+/QlvnAB8FHmH+2PFh4AOD/jdKOv2kqlZeJUmSpJPiTpYkSVILDFmSJEktMGRJkiS1wJAlSZLUAkOWJElSC1b9TfMbN26szZs3r3YbkiRJKzp48OBsVT2rn7WrHrI2b95Mt9td7TYkSZJWlOSelVfN87hQkiSpBYYsSZKkFhiyJEmSWmDIkiRJaoEhS5IkqQWGLEmSpBYYsiRJklpgyJIkSWqBIUuSJKkFhixJkqQWGLIkSZJaYMiSJElqwUAhK8klSW5PcizJ5UvMn5Pk/iQfad6iJEnS6Bl0J+teYDdw8zLz1wMHBqwtSZI08p7az6Iku4A9QAHTVbWzN/7EEmtfDjwb+CLQWan2Y3cd5htvuPQkWpYkSVr7VgxZSbYAE8C2qppNsuEEa58CfBDYCbzmBOvGgXGA5/3omSfbsyRJ0prXz07WdmB/Vc0CVNXRE6z9FeC/VdV9SZZdVFX7gH0AnU6nfvKLt/XdsCRJ0qo5Qb5ZrJ+QFeaPCfvxSuBVSX4FOAt4WpLvVdWv992RJEnSOtDPje9TwI4kYwAnOi6sql+qqgurajPz93B9woAlSZJORyuGrKqaAfYCB5LcCdyQ5OIkR4ArgJuSzLTcpyRJ0kjp69eFVTUJTC4a3rTCNR8HPj5QV5IkSSPOJ75LkiS1wJAlSZLUAkOWJElSCwxZkiRJLTBkSZIktcCQJUmS1AJDliRJUgsMWZIkSS0YKGQluSTJ7UmOJbl8wfhFSQ4muSPJTJK3D69VSZKk0dHXE9+XcC+wm/n3Ey70IPBPq+oHSc4C/meSL1TVAw16lCRJGjl9hawku5gPVAVMV9XO3vgTC9dV1eMLPp6Jx5GSJOk0tWLISrIFmAC2VdVskg0rrL8A+GPgucC73MWSJEmno352mrYD+6tqFqCqjp5ocVXdV1VbmQ9ZVyZ59uI1ScaTdJN05+bmBulbkiRpTesnZIX5Y8KT0tvBmgFetcTcvqrqVFXnjDPOONnSkiRJa14/IWsK2JFkDOBEx4VJNiX50d6/zwW2AYeH0agkSdIoWfGerKqaSbIXOJDkOHAoyY3ALcC5wBuTvKeqtgAvBD6YpJjfAftAVX2jxf4lSZLWpL5+XVhVk8DkouFNS6z7ErB1CH1JkiSNNB+xIEmS1AJDliRJUgsMWZIkSS0wZEmSJLXAkCVJktQCQ5YkSVILDFmSJEktMGRJkiS1wJAlSZLUgoFCVpJLktye5FiSyxeMvzTJV5PMJJlO8pbhtSpJkjQ6+nqtzhLuBXYDexaN/x2wq6ruSvIc4GCSP62qRxv0KEmSNHL6CllJdjEfqAqYrqqdvfEnFq6rqv+14N8PJHkIeBawbMi6++5vsXv3hwdoXZIkae1aMWQl2QJMANuqajbJhn4KJ/kp4GnA/15ibhwYBzjnnGefVMOSJEmjIFV14gXJNcD5VTWxxNzHgVurav+i8R8HbgOurKqvnah+p9Opbrd7km1LkiSdekkOVlWnn7X93Pge5o8J+/3yc4A/Bq5bKWBJkiStV/2ErClgR5IxgBMdFyZ5GnAL8Imq+vxwWpQkSRo9K4asqpoB9gIHktwJ3JDk4iRHgCuAm5LM9JbvAC4Bdie5o/f30raalyRJWqv6+nVhVU0Ck4uGNy2x7lPAp4bQlyRJ0kjzie+SJEktMGRJkiS1wJAlSZLUAkOWJElSCwxZkiRJLTBkSZIktcCQJUmS1AJDliRJUgsGCllJLklye5JjSS5fNPfFJI8muXU4LUqSJI2eQXey7gV2AzcvMfe7wM5BG5IkSVoP+gpZSXYlmU5yZ5JPVtXdVTUNPLF4bVVNAd8ddqOSJEmjZMV3FybZAkwA26pqNsmGYTbwt/d/j7f8+leGWVKSJGnV9bOTtR3YX1WzAFV1tOmXJhlP0k3SnXt8rmk5SZKkNWfFnSwgQA3zS6tqH7APoNPp1Gd/+2eGWV6SJKkVn/ud/tf2s5M1BexIMgYw7ONCSZKk9WjFkFVVM8Be4ECSO4Ebklyc5AhwBXBTkpkn1yf5S+DzwGuSHEny+pZ6lyRJWrP6OS6kqiaByUXDm5ZZ+6qmTUmSJI06n/guSZLUAkOWJElSCwxZkiRJLTBkSZIktcCQJUmS1AJDliRJUgsMWZIkSS0wZEmSJLXAkCVJktSCgUJWkkuS3J7kWJLLF81dmeSu3t+Vw2lTkiRptPT1Wp0l3AvsBvYsHOy9PPo3gA5QwMEkX6iqR5o0KUmSNGr6CllJdjEfqAqYrqqdvfEnFi19PfClqjram/8S8Abg08vVPnz/I1w68V8GaF2SJGntWjFkJdkCTADbqmq2t1u1nJ8A7lvw+UhvbHHNcWAc4MyxC06qYUmSpFHQz07WdmB/Vc0CPLlLtYwsMVY/NFC1D9gH0Ol06ra9b+6jDUmSpNWV3+p/bT83voclgtIyjgALt6Y2AQ/0344kSdL60E/ImgJ2JBmDf7i5fTl/CrwuyblJzgVe1xuTJEk6rax4XFhVM0n2AgeSHAcOJbkRuAU4F3hjkvdU1ZaqOprkeuB/9C5/7wrHi5IkSetSX78urKpJYHLR8KZl1n4M+FjDviRJkkaaT3yXJElqgSFLkiSpBYYsSZKkFhiyJEmSWmDIkiRJaoEhS5IkqQWGLEmSpBYYsiRJklow1JCV5KIkU0mmk9yWZMkHlkqSJK13w97J+gDwiaraCrwXeN+Q60uSJI2ERiErya7ertWdST4JvIj5F0oD/Dnw800blCRJGkV9vbtwKUm2ABPAtqqaTbIB+AjwZuA/AP8CODvJWFU9vFydBx94nOvfc8+gbUiSJK1JTXaytgP7q2oWoKqOAnuAVyc5BLwauB84tvjCJONJukm6c3NzDVqQJElam1JVg12Y/CpwXlVdt8z8WcBfV9UJb37vdDrV7XYH6kGSJOlUSnKwqjr9rG2ykzUF7Egy1vvSDUk2Jnmy5ruBjzWoL0mSNLIGvierqmaS7AUOJDkOHAJuBd6XpIC/AK4eTpuSJEmjZeCQBVBVk8DkouH9TWpKkiStBz7xXZIkqQWGLEmSpBYYsiRJklpgyJIkSWqBIUuSJKkFhixJkqQWGLIkSZJaYMiSJElqgSFLkiSpBUMNWUkuTPLnSQ4lmU7yz4ZZX5IkaVQMeyfrOuBzVfUy4K3A7w25viRJ0kho9O7CJLuAPUAB08DfAef0pp8BPLBSjcOzcOl/atKFJEnS2jNwyEqyBZgAtlXVbJINwJnAnyW5Bvgx4LXLXDsOjAOc+Zytg7YgSZK0ZqWqBrtwPkidX1UTC8au7dX8YJJXAv8ZeHFVPbFcnU6nU91ud6AeJEmSTqUkB6uq08/aJvdkhfljwoWuAj4HUFVfBX4E2NjgOyRJkkZSk5A1BexIMgbQOy68F3hN7/MLmQ9Z327apCRJ0qgZ+J6sqppJshc4kOQ4cAj418BHk/wa87tcu2vQ80hJkqQR1ujXhVU1CUwuGt7WpKYkSdJ64BPfJUmSWmDIkiRJaoEhS5IkqQWGLEmSpBYYsiRJklpgyJIkSWqBIUuSJKkFhixJkqQWNHoY6WJJPgT8bO/j04HzquqZw/wOSZKkUTDUkFVVv/bkv5NcA7xsmPUlSZJGRaPjwiS7kkwnuTPJJxdN/wLw6Sb1JUmSRtXAO1lJtgATwLaqmk2yYcHcRcA/Br68Up17vvs447cdGbQNSZKkNanJTtZ2YH9VzQJU1dEFc2/tzR1f6sIk40m6Sbpzc3MNWpAkSVqbmtyTFaCWmXsrcPVyF1bVPmAfQKfTqX2XbmrQhiRJ0qnx0ZNY22QnawrYkWQM4MnjwiQvAM4FvtqgtiRJ0kgbeCerqmaS7AUOJDkOHAJ2M3/D+2eqarldLkmSpHWv0SMcqmoSmFw09ptNakqSJK0HPvFdkiSpBYYsSZKkFhiyJEmSWmDIkiRJaoEhS5IkqQWGLEmSpBYYsiRJklpgyJIkSWqBIUuSJKkFQw9ZSXYk+WaSmSQ3D7u+JEnSKGj0Wp3FkjwPeDewraoeSXLeMOtLkiSNikYhK8kuYA9QwDTwIHBjVT0CUFUPrVTjwceK62eONWlDkiRpzRn4uDDJFmAC2F5VLwHeCTwfeH6S/57ka0nesMy140m6Sbpzc3ODtiBJkrRmpaoGuzC5Bji/qiYWjN0KzAE7gE3AXwIvrqpHl6vT6XSq2+0O1IMkSdKplORgVXX6Wdvkxvcwf0y40BHgv1bVXFX9H+Aw8LwG3yFJkjSSmoSsKWBHkjGAJBuAPwR+tvd5I/PHh3/btElJkqRRM/CN71U1k2QvcCDJceAQ8K+A1yX5JnAceFdVPTycViVJkkZHo18XVtUkMLlo+NrenyRJ0mnLJ75LkiS1wJAlSZLUAkOWJElSCwxZkiRJLTBkSZIktcCQJUmS1AJDliRJUgsMWZIkSS0wZEmSJLVgqCErye4k305yR+/vbcOsL0mSNCoavVZnGZ+tqne0UFeSJGlkNApZSXYBe4ACpoGpk61x/+P/l3973/ubtCFJkrTmDHxcmGQLMAFsr6qXAO/sTb05yXSS/UkuWOba8STdJN25ublBW5AkSVqzUlWDXZhcA5xfVRMLxsaA71XVD5K8HdhRVdtPVKfT6VS32x2oB0mSpFMpycGq6vSztsmN72H+mPAfVNXDVfWD3sePAi9vUF+SJGlkNQlZU8CO3u4VSTYk+fEF85cBf9WkOUmSpFE18I3vVTWTZC9wIMlx4BDwYJLLgGPAUWD3ULqUJEkaMY1+XVhVk8DkouF3N6kpSZK0HvjEd0mSpBYYsiRJklpgyJIkSWqBIUuSJKkFhixJkqQWGLIkSZJaYMiSJElqgSFLkiSpBa2ErCSXJ6kkfb1AUZIkab0ZeshKcjbwq8DXh11bkiRpVDR6rU6SXcAeoIDpqtoJXA+8vze+sr8/DIcvbdKGJEnSmjPwTlaSLcAEsL2qXgK8M8nLgAuq6tYVrh1P0k3SnZubG7QFSZKkNavJTtZ2YH9VzfY+Pwr8AbB7pQurah+wD6DT6RQvuK1BG5IkSadK+l7Z5J6sMH9M+KSzgRcDtyW5G3gF8AVvfpckSaejJiFrCtiRZKz3+R9V1caq2lxVm4GvAZdVVbdpk5IkSaNm4OPCqppJshc4kOQ4cIg+jgolSZJOB41+XVhVk8DkMnOXNqktSZI0ynziuyRJUgsMWZIkSS0wZEmSJLXAkCVJktQCQ5YkSVILDFmSJEktMGRJkiS1wJAlSZLUAkOWJElSC4YaspK8Pck3ktyR5CtJXjTM+pIkSaNi2DtZN1fVT1bVS4H3AzcMub4kSdJIaPTuwiS7gD1AAdNVtXPB9I/1xk/s0cNwy6VN2pAkSVpzBg5ZSbYAE8C2qppNsqE3fjVwLfA0YPsy144D4wBbLzxz0BYkSZLWrFStvNm05IXJNcD5VTWxzPwvAq+vqitPVKfT6VS32x2oB0mSpFMpycGq6vSztsk9WeHEx4GfAd7UoL4kSdLIahKypoAdScYAkmxI8rwF8/8cuKtJc5IkSaNq4HuyqmomyV7gQJLjwCHgO0leC8wBjwAnPCqUJElarxr9urCqJoHJIfUiSZK0bvjEd0mSpBYYsiRJklpgyJIkSWqBIUuSJKkFhixJkqQWGLIkSZJaYMiSJElqgSFLkiSpBUMNWUmuTfLNJNNJppJcNMz6kiRJo2LYO1mHgE5VbQX2A+8fcn1JkqSR0Oi1Okl2AXuAAqaraueC6a8Bv7xikXvvgXe8rUkbkiRJa87AISvJFmAC2FZVs0k2LFpyFfAny1w7DowDbH3mOYO2IEmStGalqga7MLkGOL+qJpaY+2XgHcCrq+oHJ6rT6XSq2+0O1IMkSdKplORgVXX6WdvkuDDMHxMu/vLXMr/DtWLAkiRJWq+a3Pg+BexIMgaQZEOSlwE3AZdV1UPDaFCSJGkUDbyTVVUzSfYCB5IcZ/6XhZuAs4DPJwG4t6ouG0qnkiRJI6TRrwurahKYHFIvkiRJ64ZPfJckSWqBIUuSJKkFhixJkqQWGLIkSZJaYMiSJElqgSFLkiSpBYYsSZKkFhiyJEmSWmDIkiRJasFQQ1aSS5LcnuRYksuHWVuSJGmUDHsn615gN3DzkOtKkiSNlEbvLkyyC9gDFDBdVTt740/0XeTuB+Cq32jShiRJ0pozcMhKsgWYALZV1WySDSdx7TgwDrD1nI2DtiBJkrRmpaoGuzC5Bji/qiaWmPs4cGtV7V+pTqfTqW63O1APkiRJp1KSg1XV6Wdtk3uywvwxoSRJkhZpErKmgB1JxgBO5rhQkiRpvRs4ZFXVDLAXOJDkTuCGJBcnOQJcAdyUZGZIfUqSJI2URr8urKpJYHLR8KYmNSVJktYDn/guSZLUAkOWJElSCwxZkiRJLTBkSZIktcCQJUmS1AJDliRJUgsMWZIkSS0wZEmSJLVgqCEryZlJPpvkb5J8PcnmYdaXJEkaFcPeyboKeKSqngt8CPidIdeXJEkaCY1eq5NkF7AHKGAaOA/4zd70fuAjSVJV1eR7JEmSRs3AISvJFmAC2FZVs0k2AH8B3AdQVceSfAcYA2aXq/PYY48xPT09aBuSJElrUpPjwu3A/qqaBaiqo0CWWPdDu1hJxpN0k3Tn5uYatCBJkrQ2NTkuDD8coI4AFwBHkjwVeAZwdPGFVbUP2AfQ6XRq69atDdqQJElae5rsZE0BO5KMAfSOC78AXNmbvxz4svdjSZKk09HAO1lVNZNkL3AgyXHgEPB24JNJ/ob5Hay3DqdNSZKk0dLo14VVNQlMLhq+oklNSZKk9cAnvkuSJLXAkCVJktQCQ5YkSVILsto//kvyXeDwqjYhSZLUnxdU1dn9LGx04/uQHK6qzmo3IUmStJIk3X7XelwoSZLUAkOWJElSC9ZCyNq32g1IkiT1qe/csuo3vkuSJK1Ha2EnS5Ikad0xZEmSJLXAkCVJktQCQ5YkSVILDFmSJEkt+H8LopM0sc1fuQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 720x360 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Colour graph\n",
    "import matplotlib.pyplot as plt\n",
    "plt.close('all')\n",
    "f, (col) = plt.subplots(1, 1, sharey=True, figsize=(10, 5))\n",
    "col.set_xlim(1, 5)\n",
    "col.set_ylim(-0.5,15)\n",
    "col.set_title('Colours')\n",
    "\n",
    "x_major_ticks = [1, 5]\n",
    "x_major_tick_labels = [\"\", \"\"]\n",
    "col.set_xticks(x_major_ticks)\n",
    "col.set_xticklabels(x_major_tick_labels)\n",
    "\n",
    "y_major_ticks = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14]\n",
    "y_major_tick_labels = [\"c0\", \"c1\", \"c2\", \"c3\", \"c4\", \"c5\", \"c6\", \"c7\", \"c8\", \"c9\", \"c10\", \"c11\", \"c12\", \"c13\", \"c14\"]\n",
    "col.set_yticks(y_major_ticks)\n",
    "col.set_yticklabels(y_major_tick_labels)\n",
    " \n",
    "# c0 points\n",
    "x0 = [1,5]\n",
    "y0 = [0,0]\n",
    "# plotting the c1 points \n",
    "plt.plot(x0, y0, label = \"c0\", color=c0)\n",
    " \n",
    "# c1 points\n",
    "x1 = [1,5]\n",
    "y1 = [1,1]\n",
    "# plotting the c1 points \n",
    "plt.plot(x1, y1, label = \"c1\", color=c1)\n",
    "\n",
    "# c2 points\n",
    "x2 = [1,5]\n",
    "y2 = [2,2]\n",
    "# plotting the c2 points \n",
    "plt.plot(x2, y2, label = \"c2\", color=c2)\n",
    "\n",
    "# c3 points\n",
    "x3 = [1,5]\n",
    "y3 = [3,3]\n",
    "# plotting the c3 points \n",
    "plt.plot(x3, y3, label = \"c3\", color=c3)\n",
    "\n",
    "# c4 points\n",
    "x4 = [1,5]\n",
    "y4 = [4, 4]\n",
    "# plotting the c4 points \n",
    "plt.plot(x4, y4, label = \"c4\", color=c4)\n",
    "\n",
    "# c5 points\n",
    "x5 = [1,5]\n",
    "y5 = [5, 5]\n",
    "# plotting the c5 points \n",
    "plt.plot(x5, y5, label = \"c5\", color=c5)\n",
    "\n",
    "# c6 points\n",
    "x6 = [1,5]\n",
    "y6 = [6, 6]\n",
    "# plotting the c6 points \n",
    "plt.plot(x6, y6, label = \"c6\", color=c6)\n",
    "\n",
    "# c7 points\n",
    "x7 = [1,5]\n",
    "y7 = [7, 7]\n",
    "# plotting the c7 points \n",
    "plt.plot(x7, y7, label = \"c7\", color=c7)\n",
    "\n",
    "# c8 points\n",
    "x8 = [1,5]\n",
    "y8 = [8, 8]\n",
    "# plotting the c8 points \n",
    "plt.plot(x8, y8, label = \"c8\", color=c8)\n",
    "\n",
    "# c9 points\n",
    "x9 = [1,5]\n",
    "y9 = [9, 9]\n",
    "# plotting the c9 points \n",
    "plt.plot(x9, y9, label = \"c9\", color=c9)\n",
    "\n",
    "# c10 points\n",
    "x10 = [1,5]\n",
    "y10 = [10, 10]\n",
    "# plotting the c4 points \n",
    "plt.plot(x10, y10, label = \"c10\", color=c10)\n",
    "\n",
    "# c11 points\n",
    "x11 = [1,5]\n",
    "y11 = [11, 11]\n",
    "# plotting the c11 points \n",
    "plt.plot(x11, y11, label = \"c11\", color=c11)\n",
    "\n",
    "# c12 points\n",
    "x12 = [1,5]\n",
    "y12 = [12, 12]\n",
    "# plotting the c12 points \n",
    "plt.plot(x12, y12, label = \"c12\", color=c12)\n",
    "\n",
    "# c13 points\n",
    "x13 = [1,5]\n",
    "y13 = [13, 13]\n",
    "# plotting the c13 points \n",
    "plt.plot(x13, y13, label = \"c13\", color=c13)\n",
    "\n",
    "# c14 points\n",
    "x14 = [1,5]\n",
    "y14 = [14, 14]\n",
    "# plotting the c14 points \n",
    "plt.plot(x14, y14, label = \"c14\", color=c14)\n",
    "\n",
    "# # show a legend on the plot\n",
    "# plt.legend()\n",
    " \n",
    "# function to show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "missense = c6\n",
    "nonsense = c1\n",
    "synonymous = c4"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Totals for studies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Number of individuals in each study\n",
    "Jaiswaltotal = 17182\n",
    "Genovesetotal = 12380\n",
    "McKerreltotal = 4219\n",
    "Youngtotal = 20\n",
    "Young2019totalcontrols = 69 #controls only\n",
    "ZinkWGStotal = 11262\n",
    "Acunatotal = 2006\n",
    "Coombstotal = 1591 #chemotherapy and radiotherapy naive individuals\n",
    "Desaitotal = 181 #controls only"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functions for calculating trinucleotide-context specific mutation rates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "# Amino Acid Code:\n",
    "\n",
    "Amino_acid_abbreviations = {'Ala': 'A', 'Arg': 'R', 'Asn': 'N', 'Asp': 'D',\\\n",
    "                           'Cys': 'C', 'Glu': 'E', 'Gln': 'Q', 'Gly': 'G',\\\n",
    "                           'His': 'H', 'Ile': 'I', 'Leu': 'L', 'Lys': 'K',\\\n",
    "                           'Met': 'M', 'Phe': 'F', 'Pro': 'P', 'Ser': 'S',\\\n",
    "                           'Thr': 'T', 'Trp': 'W', 'Tyr': 'Y', 'Val': 'V', '*': '*'}\n",
    "\n",
    "Genetic_Code = {'TTT': 'Phe', 'TTC': 'Phe', 'TTA': 'Leu', 'TTG': 'Leu',\\\n",
    "               'TCT': 'Ser', 'TCC': 'Ser', 'TCA': 'Ser', 'TCG': 'Ser',\\\n",
    "               'TAT': 'Tyr', 'TAC': 'Tyr', 'TAA': '*', 'TAG': '*',\\\n",
    "               'TGT': 'Cys', 'TGC': 'Cys', 'TGA': '*', 'TGG': 'Trp',\\\n",
    "               'CTT': 'Leu', 'CTC': 'Leu', 'CTA': 'Leu', 'CTG': 'Leu',\\\n",
    "               'CCT': 'Pro', 'CCC': 'Pro', 'CCA': 'Pro', 'CCG': 'Pro',\\\n",
    "               'CAT': 'His', 'CAC': 'His', 'CAA': 'Gln', 'CAG': 'Gln',\\\n",
    "               'CGT': 'Arg', 'CGC': 'Arg', 'CGA': 'Arg', 'CGG': 'Arg',\\\n",
    "               'ATT': 'Ile', 'ATC': 'Ile', 'ATA': 'Ile', 'ATG': 'Met',\\\n",
    "               'ACT': 'Thr', 'ACC': 'Thr', 'ACA': 'Thr', 'ACG': 'Thr',\\\n",
    "               'AAT': 'Asn', 'AAC': 'Asn', 'AAA': 'Lys', 'AAG': 'Lys',\\\n",
    "               'AGT': 'Ser', 'AGC': 'Ser', 'AGA': 'Arg', 'AGG': 'Arg',\\\n",
    "               'GTT': 'Val', 'GTC': 'Val', 'GTA': 'Val', 'GTG': 'Val',\\\n",
    "               'GCT': 'Ala', 'GCC': 'Ala', 'GCA': 'Ala', 'GCG': 'Ala',\\\n",
    "               'GAT': 'Asp', 'GAC': 'Asp', 'GAA': 'Glu', 'GAG': 'Glu',\\\n",
    "               'GGT': 'Gly', 'GGC': 'Gly', 'GGA': 'Gly', 'GGG': 'Gly'}\n",
    "\n",
    "base_alternative = {'A': ['T', 'C', 'G'], 'T': ['A', 'C', 'G'], 'G':['A', 'C', 'T'], 'C':['A', 'G', 'T']}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "Possible_AA_Changes = {}\n",
    "Possible_Base_Changes = {}\n",
    "Possible_Codon_Changes = {}\n",
    "for k, v in Genetic_Code.items():\n",
    "    codon = k\n",
    "    base_1 = str(codon[0])\n",
    "    base_2 = str(codon[1])\n",
    "    base_3 = str(codon[2])\n",
    "    new_codons = []\n",
    "    amino_acid_changes = []\n",
    "    base_changes = []\n",
    "    if base_1 in base_alternative.keys():\n",
    "        alternate_base = base_alternative[base_1]\n",
    "        for i in alternate_base:\n",
    "            base_changes.append(i)\n",
    "            new_codon = i+base_2+base_3\n",
    "            new_codons.append(new_codon)\n",
    "    if base_2 in base_alternative.keys():\n",
    "        alternate_base = base_alternative[base_2]\n",
    "        for i in alternate_base:\n",
    "            base_changes.append(i)\n",
    "            new_codon = base_1+i+base_3\n",
    "            new_codons.append(new_codon)\n",
    "    if base_3 in base_alternative.keys():\n",
    "        alternate_base = base_alternative[base_3]\n",
    "        for i in alternate_base:\n",
    "            base_changes.append(i)\n",
    "            new_codon = base_1+base_2+i\n",
    "            new_codons.append(new_codon)\n",
    "    Possible_Codon_Changes[k]=(new_codons[0],new_codons[1], new_codons[2], new_codons[3], new_codons[4],\\\n",
    "                               new_codons[5], new_codons[6], new_codons[7], new_codons[8]) #i.e. {'TTT': ['ATT', 'CTT', 'GTT'...etc]}\n",
    "    Possible_Base_Changes[k]=(base_changes[0], base_changes[1], base_changes[2], base_changes[3], base_changes[4],\\\n",
    "                              base_changes[5], base_changes[6], base_changes[7], base_changes[8]) #i.e. {'TTT': ['A', 'C', 'G'... etc]}\n",
    "    \n",
    "Possible_Amino_Acid_Changes = {}\n",
    "for k, v in Possible_Codon_Changes.items():\n",
    "    codon = k\n",
    "    codon_changes = v\n",
    "    amino_acid_changes = []\n",
    "    for i in codon_changes:\n",
    "        amino_acid_change = Genetic_Code[i]\n",
    "        amino_acid_changes.append(amino_acid_change)\n",
    "    Possible_Amino_Acid_Changes[k]=(amino_acid_changes[0], amino_acid_changes[1], amino_acid_changes[2],\\\n",
    "                                    amino_acid_changes[3], amino_acid_changes[4], amino_acid_changes[5],\\\n",
    "                                    amino_acid_changes[6], amino_acid_changes[7], amino_acid_changes[8]) #i.e. {'TTT': ['Ile', 'Leu', 'Val'...etc]}\n",
    "    \n",
    "Possible_Amino_Acid_Changes_letters = {}\n",
    "for k, v in Possible_Amino_Acid_Changes.items():\n",
    "    codon = k\n",
    "    codon_changes = v\n",
    "    amino_acid_changes_letters = []\n",
    "    for i in codon_changes:\n",
    "        amino_acid_change_letter = Amino_acid_abbreviations[i]\n",
    "        amino_acid_changes_letters.append(amino_acid_change_letter)\n",
    "    Possible_Amino_Acid_Changes_letters[k]=(amino_acid_changes_letters[0], amino_acid_changes_letters[1],\\\n",
    "                                            amino_acid_changes_letters[2], amino_acid_changes_letters[3],\\\n",
    "                                            amino_acid_changes_letters[4], amino_acid_changes_letters[5],\\\n",
    "                                            amino_acid_changes_letters[6], amino_acid_changes_letters[7],\\\n",
    "                                            amino_acid_changes_letters[8]) #i.e. {'TTT': ['I', 'L', 'V'...etc]}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'TTT': ('ATT', 'CTT', 'GTT', 'TAT', 'TCT', 'TGT', 'TTA', 'TTC', 'TTG'),\n",
       " 'TTC': ('ATC', 'CTC', 'GTC', 'TAC', 'TCC', 'TGC', 'TTA', 'TTG', 'TTT'),\n",
       " 'TTA': ('ATA', 'CTA', 'GTA', 'TAA', 'TCA', 'TGA', 'TTT', 'TTC', 'TTG'),\n",
       " 'TTG': ('ATG', 'CTG', 'GTG', 'TAG', 'TCG', 'TGG', 'TTA', 'TTC', 'TTT'),\n",
       " 'TCT': ('ACT', 'CCT', 'GCT', 'TAT', 'TGT', 'TTT', 'TCA', 'TCC', 'TCG'),\n",
       " 'TCC': ('ACC', 'CCC', 'GCC', 'TAC', 'TGC', 'TTC', 'TCA', 'TCG', 'TCT'),\n",
       " 'TCA': ('ACA', 'CCA', 'GCA', 'TAA', 'TGA', 'TTA', 'TCT', 'TCC', 'TCG'),\n",
       " 'TCG': ('ACG', 'CCG', 'GCG', 'TAG', 'TGG', 'TTG', 'TCA', 'TCC', 'TCT'),\n",
       " 'TAT': ('AAT', 'CAT', 'GAT', 'TTT', 'TCT', 'TGT', 'TAA', 'TAC', 'TAG'),\n",
       " 'TAC': ('AAC', 'CAC', 'GAC', 'TTC', 'TCC', 'TGC', 'TAA', 'TAG', 'TAT'),\n",
       " 'TAA': ('AAA', 'CAA', 'GAA', 'TTA', 'TCA', 'TGA', 'TAT', 'TAC', 'TAG'),\n",
       " 'TAG': ('AAG', 'CAG', 'GAG', 'TTG', 'TCG', 'TGG', 'TAA', 'TAC', 'TAT'),\n",
       " 'TGT': ('AGT', 'CGT', 'GGT', 'TAT', 'TCT', 'TTT', 'TGA', 'TGC', 'TGG'),\n",
       " 'TGC': ('AGC', 'CGC', 'GGC', 'TAC', 'TCC', 'TTC', 'TGA', 'TGG', 'TGT'),\n",
       " 'TGA': ('AGA', 'CGA', 'GGA', 'TAA', 'TCA', 'TTA', 'TGT', 'TGC', 'TGG'),\n",
       " 'TGG': ('AGG', 'CGG', 'GGG', 'TAG', 'TCG', 'TTG', 'TGA', 'TGC', 'TGT'),\n",
       " 'CTT': ('ATT', 'GTT', 'TTT', 'CAT', 'CCT', 'CGT', 'CTA', 'CTC', 'CTG'),\n",
       " 'CTC': ('ATC', 'GTC', 'TTC', 'CAC', 'CCC', 'CGC', 'CTA', 'CTG', 'CTT'),\n",
       " 'CTA': ('ATA', 'GTA', 'TTA', 'CAA', 'CCA', 'CGA', 'CTT', 'CTC', 'CTG'),\n",
       " 'CTG': ('ATG', 'GTG', 'TTG', 'CAG', 'CCG', 'CGG', 'CTA', 'CTC', 'CTT'),\n",
       " 'CCT': ('ACT', 'GCT', 'TCT', 'CAT', 'CGT', 'CTT', 'CCA', 'CCC', 'CCG'),\n",
       " 'CCC': ('ACC', 'GCC', 'TCC', 'CAC', 'CGC', 'CTC', 'CCA', 'CCG', 'CCT'),\n",
       " 'CCA': ('ACA', 'GCA', 'TCA', 'CAA', 'CGA', 'CTA', 'CCT', 'CCC', 'CCG'),\n",
       " 'CCG': ('ACG', 'GCG', 'TCG', 'CAG', 'CGG', 'CTG', 'CCA', 'CCC', 'CCT'),\n",
       " 'CAT': ('AAT', 'GAT', 'TAT', 'CTT', 'CCT', 'CGT', 'CAA', 'CAC', 'CAG'),\n",
       " 'CAC': ('AAC', 'GAC', 'TAC', 'CTC', 'CCC', 'CGC', 'CAA', 'CAG', 'CAT'),\n",
       " 'CAA': ('AAA', 'GAA', 'TAA', 'CTA', 'CCA', 'CGA', 'CAT', 'CAC', 'CAG'),\n",
       " 'CAG': ('AAG', 'GAG', 'TAG', 'CTG', 'CCG', 'CGG', 'CAA', 'CAC', 'CAT'),\n",
       " 'CGT': ('AGT', 'GGT', 'TGT', 'CAT', 'CCT', 'CTT', 'CGA', 'CGC', 'CGG'),\n",
       " 'CGC': ('AGC', 'GGC', 'TGC', 'CAC', 'CCC', 'CTC', 'CGA', 'CGG', 'CGT'),\n",
       " 'CGA': ('AGA', 'GGA', 'TGA', 'CAA', 'CCA', 'CTA', 'CGT', 'CGC', 'CGG'),\n",
       " 'CGG': ('AGG', 'GGG', 'TGG', 'CAG', 'CCG', 'CTG', 'CGA', 'CGC', 'CGT'),\n",
       " 'ATT': ('TTT', 'CTT', 'GTT', 'AAT', 'ACT', 'AGT', 'ATA', 'ATC', 'ATG'),\n",
       " 'ATC': ('TTC', 'CTC', 'GTC', 'AAC', 'ACC', 'AGC', 'ATA', 'ATG', 'ATT'),\n",
       " 'ATA': ('TTA', 'CTA', 'GTA', 'AAA', 'ACA', 'AGA', 'ATT', 'ATC', 'ATG'),\n",
       " 'ATG': ('TTG', 'CTG', 'GTG', 'AAG', 'ACG', 'AGG', 'ATA', 'ATC', 'ATT'),\n",
       " 'ACT': ('TCT', 'CCT', 'GCT', 'AAT', 'AGT', 'ATT', 'ACA', 'ACC', 'ACG'),\n",
       " 'ACC': ('TCC', 'CCC', 'GCC', 'AAC', 'AGC', 'ATC', 'ACA', 'ACG', 'ACT'),\n",
       " 'ACA': ('TCA', 'CCA', 'GCA', 'AAA', 'AGA', 'ATA', 'ACT', 'ACC', 'ACG'),\n",
       " 'ACG': ('TCG', 'CCG', 'GCG', 'AAG', 'AGG', 'ATG', 'ACA', 'ACC', 'ACT'),\n",
       " 'AAT': ('TAT', 'CAT', 'GAT', 'ATT', 'ACT', 'AGT', 'AAA', 'AAC', 'AAG'),\n",
       " 'AAC': ('TAC', 'CAC', 'GAC', 'ATC', 'ACC', 'AGC', 'AAA', 'AAG', 'AAT'),\n",
       " 'AAA': ('TAA', 'CAA', 'GAA', 'ATA', 'ACA', 'AGA', 'AAT', 'AAC', 'AAG'),\n",
       " 'AAG': ('TAG', 'CAG', 'GAG', 'ATG', 'ACG', 'AGG', 'AAA', 'AAC', 'AAT'),\n",
       " 'AGT': ('TGT', 'CGT', 'GGT', 'AAT', 'ACT', 'ATT', 'AGA', 'AGC', 'AGG'),\n",
       " 'AGC': ('TGC', 'CGC', 'GGC', 'AAC', 'ACC', 'ATC', 'AGA', 'AGG', 'AGT'),\n",
       " 'AGA': ('TGA', 'CGA', 'GGA', 'AAA', 'ACA', 'ATA', 'AGT', 'AGC', 'AGG'),\n",
       " 'AGG': ('TGG', 'CGG', 'GGG', 'AAG', 'ACG', 'ATG', 'AGA', 'AGC', 'AGT'),\n",
       " 'GTT': ('ATT', 'CTT', 'TTT', 'GAT', 'GCT', 'GGT', 'GTA', 'GTC', 'GTG'),\n",
       " 'GTC': ('ATC', 'CTC', 'TTC', 'GAC', 'GCC', 'GGC', 'GTA', 'GTG', 'GTT'),\n",
       " 'GTA': ('ATA', 'CTA', 'TTA', 'GAA', 'GCA', 'GGA', 'GTT', 'GTC', 'GTG'),\n",
       " 'GTG': ('ATG', 'CTG', 'TTG', 'GAG', 'GCG', 'GGG', 'GTA', 'GTC', 'GTT'),\n",
       " 'GCT': ('ACT', 'CCT', 'TCT', 'GAT', 'GGT', 'GTT', 'GCA', 'GCC', 'GCG'),\n",
       " 'GCC': ('ACC', 'CCC', 'TCC', 'GAC', 'GGC', 'GTC', 'GCA', 'GCG', 'GCT'),\n",
       " 'GCA': ('ACA', 'CCA', 'TCA', 'GAA', 'GGA', 'GTA', 'GCT', 'GCC', 'GCG'),\n",
       " 'GCG': ('ACG', 'CCG', 'TCG', 'GAG', 'GGG', 'GTG', 'GCA', 'GCC', 'GCT'),\n",
       " 'GAT': ('AAT', 'CAT', 'TAT', 'GTT', 'GCT', 'GGT', 'GAA', 'GAC', 'GAG'),\n",
       " 'GAC': ('AAC', 'CAC', 'TAC', 'GTC', 'GCC', 'GGC', 'GAA', 'GAG', 'GAT'),\n",
       " 'GAA': ('AAA', 'CAA', 'TAA', 'GTA', 'GCA', 'GGA', 'GAT', 'GAC', 'GAG'),\n",
       " 'GAG': ('AAG', 'CAG', 'TAG', 'GTG', 'GCG', 'GGG', 'GAA', 'GAC', 'GAT'),\n",
       " 'GGT': ('AGT', 'CGT', 'TGT', 'GAT', 'GCT', 'GTT', 'GGA', 'GGC', 'GGG'),\n",
       " 'GGC': ('AGC', 'CGC', 'TGC', 'GAC', 'GCC', 'GTC', 'GGA', 'GGG', 'GGT'),\n",
       " 'GGA': ('AGA', 'CGA', 'TGA', 'GAA', 'GCA', 'GTA', 'GGT', 'GGC', 'GGG'),\n",
       " 'GGG': ('AGG', 'CGG', 'TGG', 'GAG', 'GCG', 'GTG', 'GGA', 'GGC', 'GGT')}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Possible_Codon_Changes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'TTT': ('A', 'C', 'G', 'A', 'C', 'G', 'A', 'C', 'G'),\n",
       " 'TTC': ('A', 'C', 'G', 'A', 'C', 'G', 'A', 'G', 'T'),\n",
       " 'TTA': ('A', 'C', 'G', 'A', 'C', 'G', 'T', 'C', 'G'),\n",
       " 'TTG': ('A', 'C', 'G', 'A', 'C', 'G', 'A', 'C', 'T'),\n",
       " 'TCT': ('A', 'C', 'G', 'A', 'G', 'T', 'A', 'C', 'G'),\n",
       " 'TCC': ('A', 'C', 'G', 'A', 'G', 'T', 'A', 'G', 'T'),\n",
       " 'TCA': ('A', 'C', 'G', 'A', 'G', 'T', 'T', 'C', 'G'),\n",
       " 'TCG': ('A', 'C', 'G', 'A', 'G', 'T', 'A', 'C', 'T'),\n",
       " 'TAT': ('A', 'C', 'G', 'T', 'C', 'G', 'A', 'C', 'G'),\n",
       " 'TAC': ('A', 'C', 'G', 'T', 'C', 'G', 'A', 'G', 'T'),\n",
       " 'TAA': ('A', 'C', 'G', 'T', 'C', 'G', 'T', 'C', 'G'),\n",
       " 'TAG': ('A', 'C', 'G', 'T', 'C', 'G', 'A', 'C', 'T'),\n",
       " 'TGT': ('A', 'C', 'G', 'A', 'C', 'T', 'A', 'C', 'G'),\n",
       " 'TGC': ('A', 'C', 'G', 'A', 'C', 'T', 'A', 'G', 'T'),\n",
       " 'TGA': ('A', 'C', 'G', 'A', 'C', 'T', 'T', 'C', 'G'),\n",
       " 'TGG': ('A', 'C', 'G', 'A', 'C', 'T', 'A', 'C', 'T'),\n",
       " 'CTT': ('A', 'G', 'T', 'A', 'C', 'G', 'A', 'C', 'G'),\n",
       " 'CTC': ('A', 'G', 'T', 'A', 'C', 'G', 'A', 'G', 'T'),\n",
       " 'CTA': ('A', 'G', 'T', 'A', 'C', 'G', 'T', 'C', 'G'),\n",
       " 'CTG': ('A', 'G', 'T', 'A', 'C', 'G', 'A', 'C', 'T'),\n",
       " 'CCT': ('A', 'G', 'T', 'A', 'G', 'T', 'A', 'C', 'G'),\n",
       " 'CCC': ('A', 'G', 'T', 'A', 'G', 'T', 'A', 'G', 'T'),\n",
       " 'CCA': ('A', 'G', 'T', 'A', 'G', 'T', 'T', 'C', 'G'),\n",
       " 'CCG': ('A', 'G', 'T', 'A', 'G', 'T', 'A', 'C', 'T'),\n",
       " 'CAT': ('A', 'G', 'T', 'T', 'C', 'G', 'A', 'C', 'G'),\n",
       " 'CAC': ('A', 'G', 'T', 'T', 'C', 'G', 'A', 'G', 'T'),\n",
       " 'CAA': ('A', 'G', 'T', 'T', 'C', 'G', 'T', 'C', 'G'),\n",
       " 'CAG': ('A', 'G', 'T', 'T', 'C', 'G', 'A', 'C', 'T'),\n",
       " 'CGT': ('A', 'G', 'T', 'A', 'C', 'T', 'A', 'C', 'G'),\n",
       " 'CGC': ('A', 'G', 'T', 'A', 'C', 'T', 'A', 'G', 'T'),\n",
       " 'CGA': ('A', 'G', 'T', 'A', 'C', 'T', 'T', 'C', 'G'),\n",
       " 'CGG': ('A', 'G', 'T', 'A', 'C', 'T', 'A', 'C', 'T'),\n",
       " 'ATT': ('T', 'C', 'G', 'A', 'C', 'G', 'A', 'C', 'G'),\n",
       " 'ATC': ('T', 'C', 'G', 'A', 'C', 'G', 'A', 'G', 'T'),\n",
       " 'ATA': ('T', 'C', 'G', 'A', 'C', 'G', 'T', 'C', 'G'),\n",
       " 'ATG': ('T', 'C', 'G', 'A', 'C', 'G', 'A', 'C', 'T'),\n",
       " 'ACT': ('T', 'C', 'G', 'A', 'G', 'T', 'A', 'C', 'G'),\n",
       " 'ACC': ('T', 'C', 'G', 'A', 'G', 'T', 'A', 'G', 'T'),\n",
       " 'ACA': ('T', 'C', 'G', 'A', 'G', 'T', 'T', 'C', 'G'),\n",
       " 'ACG': ('T', 'C', 'G', 'A', 'G', 'T', 'A', 'C', 'T'),\n",
       " 'AAT': ('T', 'C', 'G', 'T', 'C', 'G', 'A', 'C', 'G'),\n",
       " 'AAC': ('T', 'C', 'G', 'T', 'C', 'G', 'A', 'G', 'T'),\n",
       " 'AAA': ('T', 'C', 'G', 'T', 'C', 'G', 'T', 'C', 'G'),\n",
       " 'AAG': ('T', 'C', 'G', 'T', 'C', 'G', 'A', 'C', 'T'),\n",
       " 'AGT': ('T', 'C', 'G', 'A', 'C', 'T', 'A', 'C', 'G'),\n",
       " 'AGC': ('T', 'C', 'G', 'A', 'C', 'T', 'A', 'G', 'T'),\n",
       " 'AGA': ('T', 'C', 'G', 'A', 'C', 'T', 'T', 'C', 'G'),\n",
       " 'AGG': ('T', 'C', 'G', 'A', 'C', 'T', 'A', 'C', 'T'),\n",
       " 'GTT': ('A', 'C', 'T', 'A', 'C', 'G', 'A', 'C', 'G'),\n",
       " 'GTC': ('A', 'C', 'T', 'A', 'C', 'G', 'A', 'G', 'T'),\n",
       " 'GTA': ('A', 'C', 'T', 'A', 'C', 'G', 'T', 'C', 'G'),\n",
       " 'GTG': ('A', 'C', 'T', 'A', 'C', 'G', 'A', 'C', 'T'),\n",
       " 'GCT': ('A', 'C', 'T', 'A', 'G', 'T', 'A', 'C', 'G'),\n",
       " 'GCC': ('A', 'C', 'T', 'A', 'G', 'T', 'A', 'G', 'T'),\n",
       " 'GCA': ('A', 'C', 'T', 'A', 'G', 'T', 'T', 'C', 'G'),\n",
       " 'GCG': ('A', 'C', 'T', 'A', 'G', 'T', 'A', 'C', 'T'),\n",
       " 'GAT': ('A', 'C', 'T', 'T', 'C', 'G', 'A', 'C', 'G'),\n",
       " 'GAC': ('A', 'C', 'T', 'T', 'C', 'G', 'A', 'G', 'T'),\n",
       " 'GAA': ('A', 'C', 'T', 'T', 'C', 'G', 'T', 'C', 'G'),\n",
       " 'GAG': ('A', 'C', 'T', 'T', 'C', 'G', 'A', 'C', 'T'),\n",
       " 'GGT': ('A', 'C', 'T', 'A', 'C', 'T', 'A', 'C', 'G'),\n",
       " 'GGC': ('A', 'C', 'T', 'A', 'C', 'T', 'A', 'G', 'T'),\n",
       " 'GGA': ('A', 'C', 'T', 'A', 'C', 'T', 'T', 'C', 'G'),\n",
       " 'GGG': ('A', 'C', 'T', 'A', 'C', 'T', 'A', 'C', 'T')}"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Possible_Base_Changes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'TTT': ('Ile', 'Leu', 'Val', 'Tyr', 'Ser', 'Cys', 'Leu', 'Phe', 'Leu'),\n",
       " 'TTC': ('Ile', 'Leu', 'Val', 'Tyr', 'Ser', 'Cys', 'Leu', 'Leu', 'Phe'),\n",
       " 'TTA': ('Ile', 'Leu', 'Val', '*', 'Ser', '*', 'Phe', 'Phe', 'Leu'),\n",
       " 'TTG': ('Met', 'Leu', 'Val', '*', 'Ser', 'Trp', 'Leu', 'Phe', 'Phe'),\n",
       " 'TCT': ('Thr', 'Pro', 'Ala', 'Tyr', 'Cys', 'Phe', 'Ser', 'Ser', 'Ser'),\n",
       " 'TCC': ('Thr', 'Pro', 'Ala', 'Tyr', 'Cys', 'Phe', 'Ser', 'Ser', 'Ser'),\n",
       " 'TCA': ('Thr', 'Pro', 'Ala', '*', '*', 'Leu', 'Ser', 'Ser', 'Ser'),\n",
       " 'TCG': ('Thr', 'Pro', 'Ala', '*', 'Trp', 'Leu', 'Ser', 'Ser', 'Ser'),\n",
       " 'TAT': ('Asn', 'His', 'Asp', 'Phe', 'Ser', 'Cys', '*', 'Tyr', '*'),\n",
       " 'TAC': ('Asn', 'His', 'Asp', 'Phe', 'Ser', 'Cys', '*', '*', 'Tyr'),\n",
       " 'TAA': ('Lys', 'Gln', 'Glu', 'Leu', 'Ser', '*', 'Tyr', 'Tyr', '*'),\n",
       " 'TAG': ('Lys', 'Gln', 'Glu', 'Leu', 'Ser', 'Trp', '*', 'Tyr', 'Tyr'),\n",
       " 'TGT': ('Ser', 'Arg', 'Gly', 'Tyr', 'Ser', 'Phe', '*', 'Cys', 'Trp'),\n",
       " 'TGC': ('Ser', 'Arg', 'Gly', 'Tyr', 'Ser', 'Phe', '*', 'Trp', 'Cys'),\n",
       " 'TGA': ('Arg', 'Arg', 'Gly', '*', 'Ser', 'Leu', 'Cys', 'Cys', 'Trp'),\n",
       " 'TGG': ('Arg', 'Arg', 'Gly', '*', 'Ser', 'Leu', '*', 'Cys', 'Cys'),\n",
       " 'CTT': ('Ile', 'Val', 'Phe', 'His', 'Pro', 'Arg', 'Leu', 'Leu', 'Leu'),\n",
       " 'CTC': ('Ile', 'Val', 'Phe', 'His', 'Pro', 'Arg', 'Leu', 'Leu', 'Leu'),\n",
       " 'CTA': ('Ile', 'Val', 'Leu', 'Gln', 'Pro', 'Arg', 'Leu', 'Leu', 'Leu'),\n",
       " 'CTG': ('Met', 'Val', 'Leu', 'Gln', 'Pro', 'Arg', 'Leu', 'Leu', 'Leu'),\n",
       " 'CCT': ('Thr', 'Ala', 'Ser', 'His', 'Arg', 'Leu', 'Pro', 'Pro', 'Pro'),\n",
       " 'CCC': ('Thr', 'Ala', 'Ser', 'His', 'Arg', 'Leu', 'Pro', 'Pro', 'Pro'),\n",
       " 'CCA': ('Thr', 'Ala', 'Ser', 'Gln', 'Arg', 'Leu', 'Pro', 'Pro', 'Pro'),\n",
       " 'CCG': ('Thr', 'Ala', 'Ser', 'Gln', 'Arg', 'Leu', 'Pro', 'Pro', 'Pro'),\n",
       " 'CAT': ('Asn', 'Asp', 'Tyr', 'Leu', 'Pro', 'Arg', 'Gln', 'His', 'Gln'),\n",
       " 'CAC': ('Asn', 'Asp', 'Tyr', 'Leu', 'Pro', 'Arg', 'Gln', 'Gln', 'His'),\n",
       " 'CAA': ('Lys', 'Glu', '*', 'Leu', 'Pro', 'Arg', 'His', 'His', 'Gln'),\n",
       " 'CAG': ('Lys', 'Glu', '*', 'Leu', 'Pro', 'Arg', 'Gln', 'His', 'His'),\n",
       " 'CGT': ('Ser', 'Gly', 'Cys', 'His', 'Pro', 'Leu', 'Arg', 'Arg', 'Arg'),\n",
       " 'CGC': ('Ser', 'Gly', 'Cys', 'His', 'Pro', 'Leu', 'Arg', 'Arg', 'Arg'),\n",
       " 'CGA': ('Arg', 'Gly', '*', 'Gln', 'Pro', 'Leu', 'Arg', 'Arg', 'Arg'),\n",
       " 'CGG': ('Arg', 'Gly', 'Trp', 'Gln', 'Pro', 'Leu', 'Arg', 'Arg', 'Arg'),\n",
       " 'ATT': ('Phe', 'Leu', 'Val', 'Asn', 'Thr', 'Ser', 'Ile', 'Ile', 'Met'),\n",
       " 'ATC': ('Phe', 'Leu', 'Val', 'Asn', 'Thr', 'Ser', 'Ile', 'Met', 'Ile'),\n",
       " 'ATA': ('Leu', 'Leu', 'Val', 'Lys', 'Thr', 'Arg', 'Ile', 'Ile', 'Met'),\n",
       " 'ATG': ('Leu', 'Leu', 'Val', 'Lys', 'Thr', 'Arg', 'Ile', 'Ile', 'Ile'),\n",
       " 'ACT': ('Ser', 'Pro', 'Ala', 'Asn', 'Ser', 'Ile', 'Thr', 'Thr', 'Thr'),\n",
       " 'ACC': ('Ser', 'Pro', 'Ala', 'Asn', 'Ser', 'Ile', 'Thr', 'Thr', 'Thr'),\n",
       " 'ACA': ('Ser', 'Pro', 'Ala', 'Lys', 'Arg', 'Ile', 'Thr', 'Thr', 'Thr'),\n",
       " 'ACG': ('Ser', 'Pro', 'Ala', 'Lys', 'Arg', 'Met', 'Thr', 'Thr', 'Thr'),\n",
       " 'AAT': ('Tyr', 'His', 'Asp', 'Ile', 'Thr', 'Ser', 'Lys', 'Asn', 'Lys'),\n",
       " 'AAC': ('Tyr', 'His', 'Asp', 'Ile', 'Thr', 'Ser', 'Lys', 'Lys', 'Asn'),\n",
       " 'AAA': ('*', 'Gln', 'Glu', 'Ile', 'Thr', 'Arg', 'Asn', 'Asn', 'Lys'),\n",
       " 'AAG': ('*', 'Gln', 'Glu', 'Met', 'Thr', 'Arg', 'Lys', 'Asn', 'Asn'),\n",
       " 'AGT': ('Cys', 'Arg', 'Gly', 'Asn', 'Thr', 'Ile', 'Arg', 'Ser', 'Arg'),\n",
       " 'AGC': ('Cys', 'Arg', 'Gly', 'Asn', 'Thr', 'Ile', 'Arg', 'Arg', 'Ser'),\n",
       " 'AGA': ('*', 'Arg', 'Gly', 'Lys', 'Thr', 'Ile', 'Ser', 'Ser', 'Arg'),\n",
       " 'AGG': ('Trp', 'Arg', 'Gly', 'Lys', 'Thr', 'Met', 'Arg', 'Ser', 'Ser'),\n",
       " 'GTT': ('Ile', 'Leu', 'Phe', 'Asp', 'Ala', 'Gly', 'Val', 'Val', 'Val'),\n",
       " 'GTC': ('Ile', 'Leu', 'Phe', 'Asp', 'Ala', 'Gly', 'Val', 'Val', 'Val'),\n",
       " 'GTA': ('Ile', 'Leu', 'Leu', 'Glu', 'Ala', 'Gly', 'Val', 'Val', 'Val'),\n",
       " 'GTG': ('Met', 'Leu', 'Leu', 'Glu', 'Ala', 'Gly', 'Val', 'Val', 'Val'),\n",
       " 'GCT': ('Thr', 'Pro', 'Ser', 'Asp', 'Gly', 'Val', 'Ala', 'Ala', 'Ala'),\n",
       " 'GCC': ('Thr', 'Pro', 'Ser', 'Asp', 'Gly', 'Val', 'Ala', 'Ala', 'Ala'),\n",
       " 'GCA': ('Thr', 'Pro', 'Ser', 'Glu', 'Gly', 'Val', 'Ala', 'Ala', 'Ala'),\n",
       " 'GCG': ('Thr', 'Pro', 'Ser', 'Glu', 'Gly', 'Val', 'Ala', 'Ala', 'Ala'),\n",
       " 'GAT': ('Asn', 'His', 'Tyr', 'Val', 'Ala', 'Gly', 'Glu', 'Asp', 'Glu'),\n",
       " 'GAC': ('Asn', 'His', 'Tyr', 'Val', 'Ala', 'Gly', 'Glu', 'Glu', 'Asp'),\n",
       " 'GAA': ('Lys', 'Gln', '*', 'Val', 'Ala', 'Gly', 'Asp', 'Asp', 'Glu'),\n",
       " 'GAG': ('Lys', 'Gln', '*', 'Val', 'Ala', 'Gly', 'Glu', 'Asp', 'Asp'),\n",
       " 'GGT': ('Ser', 'Arg', 'Cys', 'Asp', 'Ala', 'Val', 'Gly', 'Gly', 'Gly'),\n",
       " 'GGC': ('Ser', 'Arg', 'Cys', 'Asp', 'Ala', 'Val', 'Gly', 'Gly', 'Gly'),\n",
       " 'GGA': ('Arg', 'Arg', '*', 'Glu', 'Ala', 'Val', 'Gly', 'Gly', 'Gly'),\n",
       " 'GGG': ('Arg', 'Arg', 'Trp', 'Glu', 'Ala', 'Val', 'Gly', 'Gly', 'Gly')}"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Possible_Amino_Acid_Changes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'TTT': ('I', 'L', 'V', 'Y', 'S', 'C', 'L', 'F', 'L'),\n",
       " 'TTC': ('I', 'L', 'V', 'Y', 'S', 'C', 'L', 'L', 'F'),\n",
       " 'TTA': ('I', 'L', 'V', '*', 'S', '*', 'F', 'F', 'L'),\n",
       " 'TTG': ('M', 'L', 'V', '*', 'S', 'W', 'L', 'F', 'F'),\n",
       " 'TCT': ('T', 'P', 'A', 'Y', 'C', 'F', 'S', 'S', 'S'),\n",
       " 'TCC': ('T', 'P', 'A', 'Y', 'C', 'F', 'S', 'S', 'S'),\n",
       " 'TCA': ('T', 'P', 'A', '*', '*', 'L', 'S', 'S', 'S'),\n",
       " 'TCG': ('T', 'P', 'A', '*', 'W', 'L', 'S', 'S', 'S'),\n",
       " 'TAT': ('N', 'H', 'D', 'F', 'S', 'C', '*', 'Y', '*'),\n",
       " 'TAC': ('N', 'H', 'D', 'F', 'S', 'C', '*', '*', 'Y'),\n",
       " 'TAA': ('K', 'Q', 'E', 'L', 'S', '*', 'Y', 'Y', '*'),\n",
       " 'TAG': ('K', 'Q', 'E', 'L', 'S', 'W', '*', 'Y', 'Y'),\n",
       " 'TGT': ('S', 'R', 'G', 'Y', 'S', 'F', '*', 'C', 'W'),\n",
       " 'TGC': ('S', 'R', 'G', 'Y', 'S', 'F', '*', 'W', 'C'),\n",
       " 'TGA': ('R', 'R', 'G', '*', 'S', 'L', 'C', 'C', 'W'),\n",
       " 'TGG': ('R', 'R', 'G', '*', 'S', 'L', '*', 'C', 'C'),\n",
       " 'CTT': ('I', 'V', 'F', 'H', 'P', 'R', 'L', 'L', 'L'),\n",
       " 'CTC': ('I', 'V', 'F', 'H', 'P', 'R', 'L', 'L', 'L'),\n",
       " 'CTA': ('I', 'V', 'L', 'Q', 'P', 'R', 'L', 'L', 'L'),\n",
       " 'CTG': ('M', 'V', 'L', 'Q', 'P', 'R', 'L', 'L', 'L'),\n",
       " 'CCT': ('T', 'A', 'S', 'H', 'R', 'L', 'P', 'P', 'P'),\n",
       " 'CCC': ('T', 'A', 'S', 'H', 'R', 'L', 'P', 'P', 'P'),\n",
       " 'CCA': ('T', 'A', 'S', 'Q', 'R', 'L', 'P', 'P', 'P'),\n",
       " 'CCG': ('T', 'A', 'S', 'Q', 'R', 'L', 'P', 'P', 'P'),\n",
       " 'CAT': ('N', 'D', 'Y', 'L', 'P', 'R', 'Q', 'H', 'Q'),\n",
       " 'CAC': ('N', 'D', 'Y', 'L', 'P', 'R', 'Q', 'Q', 'H'),\n",
       " 'CAA': ('K', 'E', '*', 'L', 'P', 'R', 'H', 'H', 'Q'),\n",
       " 'CAG': ('K', 'E', '*', 'L', 'P', 'R', 'Q', 'H', 'H'),\n",
       " 'CGT': ('S', 'G', 'C', 'H', 'P', 'L', 'R', 'R', 'R'),\n",
       " 'CGC': ('S', 'G', 'C', 'H', 'P', 'L', 'R', 'R', 'R'),\n",
       " 'CGA': ('R', 'G', '*', 'Q', 'P', 'L', 'R', 'R', 'R'),\n",
       " 'CGG': ('R', 'G', 'W', 'Q', 'P', 'L', 'R', 'R', 'R'),\n",
       " 'ATT': ('F', 'L', 'V', 'N', 'T', 'S', 'I', 'I', 'M'),\n",
       " 'ATC': ('F', 'L', 'V', 'N', 'T', 'S', 'I', 'M', 'I'),\n",
       " 'ATA': ('L', 'L', 'V', 'K', 'T', 'R', 'I', 'I', 'M'),\n",
       " 'ATG': ('L', 'L', 'V', 'K', 'T', 'R', 'I', 'I', 'I'),\n",
       " 'ACT': ('S', 'P', 'A', 'N', 'S', 'I', 'T', 'T', 'T'),\n",
       " 'ACC': ('S', 'P', 'A', 'N', 'S', 'I', 'T', 'T', 'T'),\n",
       " 'ACA': ('S', 'P', 'A', 'K', 'R', 'I', 'T', 'T', 'T'),\n",
       " 'ACG': ('S', 'P', 'A', 'K', 'R', 'M', 'T', 'T', 'T'),\n",
       " 'AAT': ('Y', 'H', 'D', 'I', 'T', 'S', 'K', 'N', 'K'),\n",
       " 'AAC': ('Y', 'H', 'D', 'I', 'T', 'S', 'K', 'K', 'N'),\n",
       " 'AAA': ('*', 'Q', 'E', 'I', 'T', 'R', 'N', 'N', 'K'),\n",
       " 'AAG': ('*', 'Q', 'E', 'M', 'T', 'R', 'K', 'N', 'N'),\n",
       " 'AGT': ('C', 'R', 'G', 'N', 'T', 'I', 'R', 'S', 'R'),\n",
       " 'AGC': ('C', 'R', 'G', 'N', 'T', 'I', 'R', 'R', 'S'),\n",
       " 'AGA': ('*', 'R', 'G', 'K', 'T', 'I', 'S', 'S', 'R'),\n",
       " 'AGG': ('W', 'R', 'G', 'K', 'T', 'M', 'R', 'S', 'S'),\n",
       " 'GTT': ('I', 'L', 'F', 'D', 'A', 'G', 'V', 'V', 'V'),\n",
       " 'GTC': ('I', 'L', 'F', 'D', 'A', 'G', 'V', 'V', 'V'),\n",
       " 'GTA': ('I', 'L', 'L', 'E', 'A', 'G', 'V', 'V', 'V'),\n",
       " 'GTG': ('M', 'L', 'L', 'E', 'A', 'G', 'V', 'V', 'V'),\n",
       " 'GCT': ('T', 'P', 'S', 'D', 'G', 'V', 'A', 'A', 'A'),\n",
       " 'GCC': ('T', 'P', 'S', 'D', 'G', 'V', 'A', 'A', 'A'),\n",
       " 'GCA': ('T', 'P', 'S', 'E', 'G', 'V', 'A', 'A', 'A'),\n",
       " 'GCG': ('T', 'P', 'S', 'E', 'G', 'V', 'A', 'A', 'A'),\n",
       " 'GAT': ('N', 'H', 'Y', 'V', 'A', 'G', 'E', 'D', 'E'),\n",
       " 'GAC': ('N', 'H', 'Y', 'V', 'A', 'G', 'E', 'E', 'D'),\n",
       " 'GAA': ('K', 'Q', '*', 'V', 'A', 'G', 'D', 'D', 'E'),\n",
       " 'GAG': ('K', 'Q', '*', 'V', 'A', 'G', 'E', 'D', 'D'),\n",
       " 'GGT': ('S', 'R', 'C', 'D', 'A', 'V', 'G', 'G', 'G'),\n",
       " 'GGC': ('S', 'R', 'C', 'D', 'A', 'V', 'G', 'G', 'G'),\n",
       " 'GGA': ('R', 'R', '*', 'E', 'A', 'V', 'G', 'G', 'G'),\n",
       " 'GGG': ('R', 'R', 'W', 'E', 'A', 'V', 'G', 'G', 'G')}"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Possible_Amino_Acid_Changes_letters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "#Import mutation rates\n",
    "import csv\n",
    "filename=\"Data files/Trinucleotide_context_mutation_rates.csv\" #csv file containing trinucleotide-context specific mutation rates (calculated from Lee-Six et al 2018 data)\n",
    "\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    Mutation_rates = {} #create a dictionary where key = trincucleotide site, value = mutation rate\n",
    "\n",
    "    for row in read_reader:\n",
    "        if row_count>0:\n",
    "            site = row[0] #trinucleotide site\n",
    "            comp_site = row[2] #complementary site\n",
    "            if site != '':\n",
    "                mutation_rate = float(row[7])\n",
    "                Mutation_rates[site] = mutation_rate\n",
    "                Mutation_rates[comp_site] = mutation_rate\n",
    "        row_count = row_count+1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculating mutation rates for individual variants"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DNMT3A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_variants_DNMT3A(variant):\n",
    "    intron_exon_boundary_caution_base_DNMT3A_NP072046 = [1279, 1429, 1474, 1554, 1852, 2083, 2082, 2173, 2322, 2478]\n",
    "\n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    def intron_exon_check_DNMT3A_NP072046(position_in_gene_number):\n",
    "        if position_in_gene_number == 1279:\n",
    "            gene_pos = 1279\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "    #             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1429:\n",
    "            gene_pos = 1429\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "    #             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1474:\n",
    "            gene_pos = 1474\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "    #             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1554:\n",
    "            gene_pos = 1554\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "    #             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1852:\n",
    "            gene_pos = 1852\n",
    "            first_base = 'G'\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "    #             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2083:\n",
    "            gene_pos = 2083\n",
    "            first_base = 'G'\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "    #             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2082:\n",
    "            gene_pos = 2082\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "    #             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2173:\n",
    "            gene_pos = 2173\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "    #             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2322:\n",
    "            gene_pos = 2322\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "    #             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2478:\n",
    "            gene_pos = 2478\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "    #             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "\n",
    "    #DNMT3A transcript used:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000119772;r=2:25455845-25564754;t=ENST00000321117\n",
    "    DNMT3A_exons_NP_072046 = 'ATGCCCGCCATGCCCTCCAGCGGCCCCGGGGACACCAGCAGCTCTGCTGCGGAGCGGGAGGAGGACCGAAAGGACGGAGAGGAGCAGGAGGAGCCGCGTGGCAAGGAGGAGCGCCAAGAGCCCAGCACCACGGCACGGAAGGTGGGGCGGCCTGGGAGGAAGCGCAAGCACCCCCCGGTGGAAAGCGGTGACACGCCAAAGGACCCTGCGGTGATCTCCAAGTCCCCATCCATGGCCCAGGACTCAGGCGCCTCAGAGCTATTACCCAATGGGGACTTGGAGAAGCGGAGTGAGCCCCAGCCAGAGGAGGGGAGCCCTGCTGGGGGGCAGAAGGGCGGGGCCCCAGCAGAGGGAGAGGGTGCAGCTGAGACCCTGCCTGAAGCCTCAAGAGCAGTGGAAAATGGCTGCTGCACCCCCAAGGAGGGCCGAGGAGCCCCTGCAGAAGCGGGCAAAGAACAGAAGGAGACCAACATCGAATCCATGAAAATGGAGGGCTCCCGGGGCCGGCTGCGGGGTGGCTTGGGCTGGGAGTCCAGCCTCCGTCAGCGGCCCATGCCGAGGCTCACCTTCCAGGCGGGGGACCCCTACTACATCAGCAAGCGCAAGCGGGACGAGTGGCTGGCACGCTGGAAAAGGGAGGCTGAGAAGAAAGCCAAGGTCATTGCAGGAATGAATGCTGTGGAAGAAAACCAGGGGCCCGGGGAGTCTCAGAAGGTGGAGGAGGCCAGCCCTCCTGCTGTGCAGCAGCCCACTGACCCCGCATCCCCCACTGTGGCTACCACGCCTGAGCCCGTGGGGTCCGATGCTGGGGACAAGAATGCCACCAAAGCAGGCGATGACGAGCCAGAGTACGAGGACGGCCGGGGCTTTGGCATTGGGGAGCTGGTGTGGGGGAAACTGCGGGGCTTCTCCTGGTGGCCAGGCCGCATTGTGTCTTGGTGGATGACGGGCCGGAGCCGAGCAGCTGAAGGCACCCGCTGGGTCATGTGGTTCGGAGACGGCAAATTCTCAGTGGTGTGTGTTGAGAAGCTGATGCCGCTGAGCTCGTTTTGCAGTGCGTTCCACCAGGCCACGTACAACAAGCAGCCCATGTACCGCAAAGCCATCTACGAGGTCCTGCAGGTGGCCAGCAGCCGCGCGGGGAAGCTGTTCCCGGTGTGCCACGACAGCGATGAGAGTGACACTGCCAAGGCCGTGGAGGTGCAGAACAAGCCCATGATTGAATGGGCCCTGGGGGGCTTCCAGCCTTCTGGCCCTAAGGGCCTGGAGCCACCAGAAGAAGAGAAGAATCCCTACAAAGAAGTGTACACGGACATGTGGGTGGAACCTGAGGCAGCTGCCTACGCACCACCTCCACCAGCCAAAAAGCCCCGGAAGAGCACAGCGGAGAAGCCCAAGGTCAAGGAGATTATTGATGAGCGCACAAGAGAGCGGCTGGTGTACGAGGTGCGGCAGAAGTGCCGGAACATTGAGGACATCTGCATCTCCTGTGGGAGCCTCAATGTTACCCTGGAACACCCCCTCTTCGTTGGAGGAATGTGCCAAAACTGCAAGAACTGCTTTCTGGAGTGTGCGTACCAGTACGACGACGACGGCTACCAGTCCTACTGCACCATCTGCTGTGGGGGCCGTGAGGTGCTCATGTGCGGAAACAACAACTGCTGCAGGTGCTTTTGCGTGGAGTGTGTGGACCTCTTGGTGGGGCCGGGGGCTGCCCAGGCAGCCATTAAGGAAGACCCCTGGAACTGCTACATGTGCGGGCACAAGGGTACCTACGGGCTGCTGCGGCGGCGAGAGGACTGGCCCTCCCGGCTCCAGATGTTCTTCGCTAATAACCACGACCAGGAATTTGACCCTCCAAAGGTTTACCCACCTGTCCCAGCTGAGAAGAGGAAGCCCATCCGGGTGCTGTCTCTCTTTGATGGAATCGCTACAGGGCTCCTGGTGCTGAAGGACTTGGGCATTCAGGTGGACCGCTACATTGCCTCGGAGGTGTGTGAGGACTCCATCACGGTGGGCATGGTGCGGCACCAGGGGAAGATCATGTACGTCGGGGACGTCCGCAGCGTCACACAGAAGCATATCCAGGAGTGGGGCCCATTCGATCTGGTGATTGGGGGCAGTCCCTGCAATGACCTCTCCATCGTCAACCCTGCTCGCAAGGGCCTCTACGAGGGCACTGGCCGGCTCTTCTTTGAGTTCTACCGCCTCCTGCATGATGCGCGGCCCAAGGAGGGAGATGATCGCCCCTTCTTCTGGCTCTTTGAGAATGTGGTGGCCATGGGCGTTAGTGACAAGAGGGACATCTCGCGATTTCTCGAGTCCAACCCTGTGATGATTGATGCCAAAGAAGTGTCAGCTGCACACAGGGCCCGCTACTTCTGGGGTAACCTTCCCGGTATGAACAGGCCGTTGGCATCCACTGTGAATGATAAGCTGGAGCTGCAGGAGTGTCTGGAGCATGGCAGGATAGCCAAGTTCAGCAAAGTGAGGACCATTACTACGAGGTCAAACTCCATAAAGCAGGGCAAAGACCAGCATTTTCCTGTCTTCATGAATGAGAAAGAGGACATCTTATGGTGCACTGAAATGGAAAGGGTATTTGGTTTCCCAGTCCACTATACTGACGTCTCCAACATGAGCCGCTTGGCGAGGCAGAGACTGCTGGGCCGGTCATGGAGCGTGCCAGTCATCCGCCACCTCTTCGCTCCGCTGAAGGAGTATTTTGCGTGTGTGTAA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    DNMT3A_codons_NP_072046 = [DNMT3A_exons_NP_072046[i:i+3] for i in range(0, len(DNMT3A_exons_NP_072046), 3)]\n",
    "\n",
    "    #DNMT3A transcript used:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000119772;r=2:25455845-25564754;t=ENST00000321117\n",
    "    DNMT3A_exons_extra_bases_NP_072046 = 'GATGCCCGCCATGCCCTCCAGCGGCCCCGGGGACACCAGCAGCTCTGCTGCGGAGCGGGAGGAGGACCGAAAGGACGGAGAGGAGCAGGAGGAGCCGCGTGGCAAGGAGGAGCGCCAAGAGCCCAGCACCACGGCACGGAAGGTGGGGCGGCCTGGGAGGAAGCGCAAGCACCCCCCGGTGGAAAGCGGTGACACGCCAAAGGACCCTGCGGTGATCTCCAAGTCCCCATCCATGGCCCAGGACTCAGGCGCCTCAGAGCTATTACCCAATGGGGACTTGGAGAAGCGGAGTGAGCCCCAGCCAGAGGAGGGGAGCCCTGCTGGGGGGCAGAAGGGCGGGGCCCCAGCAGAGGGAGAGGGTGCAGCTGAGACCCTGCCTGAAGCCTCAAGAGCAGTGGAAAATGGCTGCTGCACCCCCAAGGAGGGCCGAGGAGCCCCTGCAGAAGCGGGCAAAGAACAGAAGGAGACCAACATCGAATCCATGAAAATGGAGGGCTCCCGGGGCCGGCTGCGGGGTGGCTTGGGCTGGGAGTCCAGCCTCCGTCAGCGGCCCATGCCGAGGCTCACCTTCCAGGCGGGGGACCCCTACTACATCAGCAAGCGCAAGCGGGACGAGTGGCTGGCACGCTGGAAAAGGGAGGCTGAGAAGAAAGCCAAGGTCATTGCAGGAATGAATGCTGTGGAAGAAAACCAGGGGCCCGGGGAGTCTCAGAAGGTGGAGGAGGCCAGCCCTCCTGCTGTGCAGCAGCCCACTGACCCCGCATCCCCCACTGTGGCTACCACGCCTGAGCCCGTGGGGTCCGATGCTGGGGACAAGAATGCCACCAAAGCAGGCGATGACGAGCCAGAGTACGAGGACGGCCGGGGCTTTGGCATTGGGGAGCTGGTGTGGGGGAAACTGCGGGGCTTCTCCTGGTGGCCAGGCCGCATTGTGTCTTGGTGGATGACGGGCCGGAGCCGAGCAGCTGAAGGCACCCGCTGGGTCATGTGGTTCGGAGACGGCAAATTCTCAGTGGTGTGTGTTGAGAAGCTGATGCCGCTGAGCTCGTTTTGCAGTGCGTTCCACCAGGCCACGTACAACAAGCAGCCCATGTACCGCAAAGCCATCTACGAGGTCCTGCAGGTGGCCAGCAGCCGCGCGGGGAAGCTGTTCCCGGTGTGCCACGACAGCGATGAGAGTGACACTGCCAAGGCCGTGGAGGTGCAGAACAAGCCCATGATTGAATGGGCCCTGGGGGGCTTCCAGCCTTCTGGCCCTAAGGGCCTGGAGCCACCAGAAGAAGAGAAGAATCCCTACAAAGAAGTGTACACGGACATGTGGGTGGAACCTGAGGCAGCTGCCTACGCACCACCTCCACCAGCCAAAAAGCCCCGGAAGAGCACAGCGGAGAAGCCCAAGGTCAAGGAGATTATTGATGAGCGCACAAGAGAGCGGCTGGTGTACGAGGTGCGGCAGAAGTGCCGGAACATTGAGGACATCTGCATCTCCTGTGGGAGCCTCAATGTTACCCTGGAACACCCCCTCTTCGTTGGAGGAATGTGCCAAAACTGCAAGAACTGCTTTCTGGAGTGTGCGTACCAGTACGACGACGACGGCTACCAGTCCTACTGCACCATCTGCTGTGGGGGCCGTGAGGTGCTCATGTGCGGAAACAACAACTGCTGCAGGTGCTTTTGCGTGGAGTGTGTGGACCTCTTGGTGGGGCCGGGGGCTGCCCAGGCAGCCATTAAGGAAGACCCCTGGAACTGCTACATGTGCGGGCACAAGGGTACCTACGGGCTGCTGCGGCGGCGAGAGGACTGGCCCTCCCGGCTCCAGATGTTCTTCGCTAATAACCACGACCAGGAATTTGACCCTCCAAAGGTTTACCCACCTGTCCCAGCTGAGAAGAGGAAGCCCATCCGGGTGCTGTCTCTCTTTGATGGAATCGCTACAGGGCTCCTGGTGCTGAAGGACTTGGGCATTCAGGTGGACCGCTACATTGCCTCGGAGGTGTGTGAGGACTCCATCACGGTGGGCATGGTGCGGCACCAGGGGAAGATCATGTACGTCGGGGACGTCCGCAGCGTCACACAGAAGCATATCCAGGAGTGGGGCCCATTCGATCTGGTGATTGGGGGCAGTCCCTGCAATGACCTCTCCATCGTCAACCCTGCTCGCAAGGGCCTCTACGAGGGCACTGGCCGGCTCTTCTTTGAGTTCTACCGCCTCCTGCATGATGCGCGGCCCAAGGAGGGAGATGATCGCCCCTTCTTCTGGCTCTTTGAGAATGTGGTGGCCATGGGCGTTAGTGACAAGAGGGACATCTCGCGATTTCTCGAGTCCAACCCTGTGATGATTGATGCCAAAGAAGTGTCAGCTGCACACAGGGCCCGCTACTTCTGGGGTAACCTTCCCGGTATGAACAGGCCGTTGGCATCCACTGTGAATGATAAGCTGGAGCTGCAGGAGTGTCTGGAGCATGGCAGGATAGCCAAGTTCAGCAAAGTGAGGACCATTACTACGAGGTCAAACTCCATAAAGCAGGGCAAAGACCAGCATTTTCCTGTCTTCATGAATGAGAAAGAGGACATCTTATGGTGCACTGAAATGGAAAGGGTATTTGGTTTCCCAGTCCACTATACTGACGTCTCCAACATGAGCCGCTTGGCGAGGCAGAGACTGCTGGGCCGGTCATGGAGCGTGCCAGTCATCCGCCACCTCTTCGCTCCGCTGAAGGAGTATTTTGCGTGTGTGTAAG'\n",
    "\n",
    "    #Create a list of the codons they called missense variants at\n",
    "    DNMT3A_variant_list_NP_072046 = []\n",
    "    DNMT3A_missense_codons_NP_072046 =[]\n",
    "    DNMT3A_missense_variant_DICT_NP_072046 = {}\n",
    "\n",
    "    variant_codon = int(((re.findall('\\d+', variant)))[0])\n",
    "\n",
    "    codon0_NP_072046 = variant_codon-1\n",
    "    codon1_NP_072046 = variant_codon\n",
    "    variant_NP_072046 = variant\n",
    "    DNMT3A_missense_codons_NP_072046.append(codon0_NP_072046) #List of codon numbers that they called variants at\n",
    "    DNMT3A_missense_variant_DICT_NP_072046[row_count]=(codon1_NP_072046, variant_NP_072046) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    DNMT3A_missense_codons_NP_072046_DICT = {}\n",
    "    calls_DNMT3A_missense_NP_072046 = {}\n",
    "\n",
    "    for i in DNMT3A_missense_codons_NP_072046: #Iterates over the list of codon numbers\n",
    "        DNMT3A_missense_codons_NP_072046_DICT[i+1]= (DNMT3A_codons_NP_072046[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "\n",
    "    for k,v in DNMT3A_missense_codons_NP_072046_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_DNMT3A_missense_NP_072046[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_DNMT3A_missense_NP_072046_extra = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_missense_NP_072046.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_DNMT3A_missense_NP_072046_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_DNMT3A_missense_NP_072046_extra2 = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_missense_NP_072046_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_DNMT3A_missense_NP_072046_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    DNMT3A_position_in_list_dict_missense_NP_072046={}\n",
    "    for k, v in DNMT3A_missense_variant_DICT_NP_072046.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_DNMT3A_missense_NP_072046[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_DNMT3A_missense_NP_072046[v[0]]\n",
    "        codon_seq=calls_DNMT3A_missense_NP_072046[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_DNMT3A_missense_NP_072046_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                DNMT3A_position_in_list_dict_missense_NP_072046[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "\n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in DNMT3A_position_in_list_dict_missense_NP_072046.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "    #     print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    DNMT3A_bases_missense_NP_072046 = {}\n",
    "    for k, v in DNMT3A_position_in_list_dict_missense_NP_072046.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        DNMT3A_bases_missense_NP_072046[k]=(v, base_changes)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    DNMT3A_bases_missense2_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_missense_NP_072046.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        DNMT3A_bases_missense2_NP_072046[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    DNMT3A_bases_missense3_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_missense2_NP_072046.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        DNMT3A_bases_missense3_NP_072046[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    DNMT3A_bases_missense4_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_missense3_NP_072046.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_DNMT3A_NP072046:\n",
    "                adapted_trio = intron_exon_check_DNMT3A_NP072046(i)\n",
    "                sites_trio_missense.append([adapted_trio])\n",
    "                DNMT3A_bases_missense4_NP_072046[k] = (v, sites_trio_missense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_missense = [DNMT3A_exons_extra_bases_NP_072046[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_missense.append(site_trio_missense)\n",
    "                    DNMT3A_bases_missense4_NP_072046[k] = (v, sites_trio_missense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    DNMT3A_sites_missense5_NP_072046 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in DNMT3A_bases_missense4_NP_072046.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        DNMT3A_sites_missense5_NP_072046[a]=info\n",
    "        a=a+1\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    DNMT3A_sites_missense6_NP_072046={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in DNMT3A_sites_missense5_NP_072046.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (original_base,new_base)\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        DNMT3A_sites_missense6_NP_072046[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    DNMT3A_mutation_rates_missense_NP_072046 = {}\n",
    "    for k, v in DNMT3A_sites_missense6_NP_072046.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            DNMT3A_mutation_rates_missense_NP_072046[k] = (v, mutation_rate)\n",
    "\n",
    "    DNMT3A_mutation_rate_missense_NP_072046_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in DNMT3A_mutation_rates_missense_NP_072046.items():\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        DNMT3A_mutation_rate_missense_NP_072046_sum.append(mutation_rate_value)\n",
    "\n",
    "    DNMT3A_Summed_mutation_rate_missense_NP_072046_added = np.sum(DNMT3A_mutation_rate_missense_NP_072046_sum)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return DNMT3A_Summed_mutation_rate_missense_NP_072046_added"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TET2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_variants_TET2(variant):\n",
    "    intron_exon_boundary_caution_base_TET2 = [3409, 3804, 4044, 4537]\n",
    "\n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "\n",
    "    def intron_exon_check_TET2(position_in_gene_number):\n",
    "        if position_in_gene_number == 3409:\n",
    "            gene_pos = 3409\n",
    "            first_base = TET2_exons_extra_bases_NP[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 3804:\n",
    "            gene_pos = 3804\n",
    "            first_base = 'G'\n",
    "            middle_base = TET2_exons_extra_bases_NP[gene_pos]\n",
    "            third_base = TET2_exons_extra_bases_NP[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 4044:\n",
    "            gene_pos = 4044\n",
    "            first_base = TET2_exons_extra_bases_NP[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 4537:\n",
    "            gene_pos = 4537\n",
    "            first_base = TET2_exons_extra_bases_NP[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #TET2 transcript (NP):\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000168769;r=4:106067032-106163928;t=ENST00000380013\n",
    "    TET2_exons_NP = 'ATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGAGCAAATTATTGAAAAAGATGAAGGTCCTTTTTATACCCATCTAGGAGCAGGTCCTAATGTGGCAGCTATTAGAGAAATCATGGAAGAAAGGTTTGGACAGAAGGGTAAAGCTATTAGGATTGAAAGAGTCATCTATACTGGTAAAGAAGGCAAAAGTTCTCAGGGATGTCCTATTGCTAAGTGGGTGGTTCGCAGAAGCAGCAGTGAAGAGAAGCTACTGTGTTTGGTGCGGGAGCGAGCTGGCCACACCTGTGAGGCTGCAGTGATTGTGATTCTCATCCTGGTGTGGGAAGGAATCCCGCTGTCTCTGGCTGACAAACTCTACTCGGAGCTTACCGAGACGCTGAGGAAATACGGCACGCTCACCAATCGCCGGTGTGCCTTGAATGAAGAGAGAACTTGCGCCTGTCAGGGGCTGGATCCAGAAACCTGTGGTGCCTCCTTCTCTTTTGGTTGTTCATGGAGCATGTACTACAATGGATGTAAGTTTGCCAGAAGCAAGATCCCAAGGAAGTTTAAGCTGCTTGGGGATGACCCAAAAGAGGAAGAGAAACTGGAGTCTCATTTGCAAAACCTGTCCACTCTTATGGCACCAACATATAAGAAACTTGCACCTGATGCATATAATAATCAGATTGAATATGAACACAGAGCACCAGAGTGCCGTCTGGGTCTGAAGGAAGGCCGTCCATTCTCAGGGGTCACTGCATGTTTGGACTTCTGTGCTCATGCCCACAGAGACTTGCACAACATGCAGAATGGCAGCACATTGGTATGCACTCTCACTAGAGAAGACAATCGAGAATTTGGAGGAAAACCTGAGGATGAGCAGCTTCACGTTCTGCCTTTATACAAAGTCTCTGACGTGGATGAGTTTGGGAGTGTGGAAGCTCAGGAGGAGAAAAAACGGAGTGGTGCCATTCAGGTACTGAGTTCTTTTCGGCGAAAAGTCAGGATGTTAGCAGAGCCAGTCAAGACTTGCCGACAAAGGAAACTAGAAGCCAAGAAAGCTGCAGCTGAAAAGCTTTCCTCCCTGGAGAACAGCTCAAATAAAAATGAAAAGGAAAAGTCAGCCCCATCACGTACAAAACAAACTGAAAACGCAAGCCAGGCTAAACAGTTGGCAGAACTTTTGCGACTTTCAGGACCAGTCATGCAGCAGTCCCAGCAGCCCCAGCCTCTACAGAAGCAGCCACCACAGCCCCAGCAGCAGCAGAGACCCCAGCAGCAGCAGCCACATCACCCTCAGACAGAGTCTGTCAACTCTTATTCTGCTTCTGGATCCACCAATCCATACATGAGACGGCCCAATCCAGTTAGTCCTTATCCAAACTCTTCACACACTTCAGATATCTATGGAAGCACCAGCCCTATGAACTTCTATTCCACCTCATCTCAAGCTGCAGGTTCATATTTGAATTCTTCTAATCCCATGAACCCTTACCCTGGGCTTTTGAATCAGAATACCCAATATCCATCATATCAATGCAATGGAAACCTATCAGTGGACAACTGCTCCCCATATCTGGGTTCCTATTCTCCCCAGTCTCAGCCGATGGATCTGTATAGGTATCCAAGCCAAGACCCTCTGTCTAAGCTCAGTCTACCACCCATCCATACACTTTACCAGCCAAGGTTTGGAAATAGCCAGAGTTTTACATCTAAATACTTAGGTTATGGAAACCAAAATATGCAGGGAGATGGTTTCAGCAGTTGTACCATTAGACCAAATGTACATCATGTAGGGAAATTGCCTCCTTATCCCACTCATGAGATGGATGGCCACTTCATGGGAGCCACCTCTAGATTACCACCCAATCTGAGCAATCCAAACATGGACTATAAAAATGGTGAACATCATTCACCTTCTCACATAATCCATAACTACAGTGCAGCTCCGGGCATGTTCAACAGCTCTCTTCATGCCCTGCATCTCCAAAACAAGGAGAATGACATGCTTTCCCACACAGCTAATGGGTTATCAAAGATGCTTCCAGCTCTTAACCATGATAGAACTGCTTGTGTCCAAGGAGGCTTACACAAATTAAGTGATGCTAATGGTCAGGAAAAGCAGCCATTGGCACTAGTCCAGGGTGTGGCTTCTGGTGCAGAGGACAACGATGAGGTCTGGTCAGACAGCGAGCAGAGCTTTCTGGATCCTGACATTGGGGGAGTGGCCGTGGCTCCAACTCATGGGTCAATTCTCATTGAGTGTGCAAAGCGTGAGCTGCATGCCACAACCCCTTTAAAGAATCCCAATAGGAATCACCCCACCAGGATCTCCCTCGTCTTTTACCAGCATAAGAGCATGAATGAGCCAAAACATGGCTTGGCTCTTTGGGAAGCCAAAATGGCTGAAAAAGCCCGTGAGAAAGAGGAAGAGTGTGAAAAGTATGGCCCAGACTATGTGCCTCAGAAATCCCATGGCAAAAAAGTGAAACGGGAGCCTGCTGAGCCACATGAAACTTCAGAGCCCACTTACCTGCGTTTCATCAAGTCTCTTGCCGAAAGGACCATGTCCGTGACCACAGACTCCACAGTAACTACATCTCCATATGCCTTCACTCGGGTCACAGGGCCTTACAACAGATATATATGA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    TET2_codons_NP = [TET2_exons_NP[i:i+3] for i in range(0, len(TET2_exons_NP), 3)]\n",
    "\n",
    "    #TET2 transcript (transcript variant 2 (TET2-001), NM_017628, NP_001120680):\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000168769;r=4:106067032-106163928;t=ENST00000380013\n",
    "    TET2_exons_extra_bases_NP = 'GATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGAGCAAATTATTGAAAAAGATGAAGGTCCTTTTTATACCCATCTAGGAGCAGGTCCTAATGTGGCAGCTATTAGAGAAATCATGGAAGAAAGGTTTGGACAGAAGGGTAAAGCTATTAGGATTGAAAGAGTCATCTATACTGGTAAAGAAGGCAAAAGTTCTCAGGGATGTCCTATTGCTAAGTGGGTGGTTCGCAGAAGCAGCAGTGAAGAGAAGCTACTGTGTTTGGTGCGGGAGCGAGCTGGCCACACCTGTGAGGCTGCAGTGATTGTGATTCTCATCCTGGTGTGGGAAGGAATCCCGCTGTCTCTGGCTGACAAACTCTACTCGGAGCTTACCGAGACGCTGAGGAAATACGGCACGCTCACCAATCGCCGGTGTGCCTTGAATGAAGAGAGAACTTGCGCCTGTCAGGGGCTGGATCCAGAAACCTGTGGTGCCTCCTTCTCTTTTGGTTGTTCATGGAGCATGTACTACAATGGATGTAAGTTTGCCAGAAGCAAGATCCCAAGGAAGTTTAAGCTGCTTGGGGATGACCCAAAAGAGGAAGAGAAACTGGAGTCTCATTTGCAAAACCTGTCCACTCTTATGGCACCAACATATAAGAAACTTGCACCTGATGCATATAATAATCAGATTGAATATGAACACAGAGCACCAGAGTGCCGTCTGGGTCTGAAGGAAGGCCGTCCATTCTCAGGGGTCACTGCATGTTTGGACTTCTGTGCTCATGCCCACAGAGACTTGCACAACATGCAGAATGGCAGCACATTGGTATGCACTCTCACTAGAGAAGACAATCGAGAATTTGGAGGAAAACCTGAGGATGAGCAGCTTCACGTTCTGCCTTTATACAAAGTCTCTGACGTGGATGAGTTTGGGAGTGTGGAAGCTCAGGAGGAGAAAAAACGGAGTGGTGCCATTCAGGTACTGAGTTCTTTTCGGCGAAAAGTCAGGATGTTAGCAGAGCCAGTCAAGACTTGCCGACAAAGGAAACTAGAAGCCAAGAAAGCTGCAGCTGAAAAGCTTTCCTCCCTGGAGAACAGCTCAAATAAAAATGAAAAGGAAAAGTCAGCCCCATCACGTACAAAACAAACTGAAAACGCAAGCCAGGCTAAACAGTTGGCAGAACTTTTGCGACTTTCAGGACCAGTCATGCAGCAGTCCCAGCAGCCCCAGCCTCTACAGAAGCAGCCACCACAGCCCCAGCAGCAGCAGAGACCCCAGCAGCAGCAGCCACATCACCCTCAGACAGAGTCTGTCAACTCTTATTCTGCTTCTGGATCCACCAATCCATACATGAGACGGCCCAATCCAGTTAGTCCTTATCCAAACTCTTCACACACTTCAGATATCTATGGAAGCACCAGCCCTATGAACTTCTATTCCACCTCATCTCAAGCTGCAGGTTCATATTTGAATTCTTCTAATCCCATGAACCCTTACCCTGGGCTTTTGAATCAGAATACCCAATATCCATCATATCAATGCAATGGAAACCTATCAGTGGACAACTGCTCCCCATATCTGGGTTCCTATTCTCCCCAGTCTCAGCCGATGGATCTGTATAGGTATCCAAGCCAAGACCCTCTGTCTAAGCTCAGTCTACCACCCATCCATACACTTTACCAGCCAAGGTTTGGAAATAGCCAGAGTTTTACATCTAAATACTTAGGTTATGGAAACCAAAATATGCAGGGAGATGGTTTCAGCAGTTGTACCATTAGACCAAATGTACATCATGTAGGGAAATTGCCTCCTTATCCCACTCATGAGATGGATGGCCACTTCATGGGAGCCACCTCTAGATTACCACCCAATCTGAGCAATCCAAACATGGACTATAAAAATGGTGAACATCATTCACCTTCTCACATAATCCATAACTACAGTGCAGCTCCGGGCATGTTCAACAGCTCTCTTCATGCCCTGCATCTCCAAAACAAGGAGAATGACATGCTTTCCCACACAGCTAATGGGTTATCAAAGATGCTTCCAGCTCTTAACCATGATAGAACTGCTTGTGTCCAAGGAGGCTTACACAAATTAAGTGATGCTAATGGTCAGGAAAAGCAGCCATTGGCACTAGTCCAGGGTGTGGCTTCTGGTGCAGAGGACAACGATGAGGTCTGGTCAGACAGCGAGCAGAGCTTTCTGGATCCTGACATTGGGGGAGTGGCCGTGGCTCCAACTCATGGGTCAATTCTCATTGAGTGTGCAAAGCGTGAGCTGCATGCCACAACCCCTTTAAAGAATCCCAATAGGAATCACCCCACCAGGATCTCCCTCGTCTTTTACCAGCATAAGAGCATGAATGAGCCAAAACATGGCTTGGCTCTTTGGGAAGCCAAAATGGCTGAAAAAGCCCGTGAGAAAGAGGAAGAGTGTGAAAAGTATGGCCCAGACTATGTGCCTCAGAAATCCCATGGCAAAAAAGTGAAACGGGAGCCTGCTGAGCCACATGAAACTTCAGAGCCCACTTACCTGCGTTTCATCAAGTCTCTTGCCGAAAGGACCATGTCCGTGACCACAGACTCCACAGTAACTACATCTCCATATGCCTTCACTCGGGTCACAGGGCCTTACAACAGATATATATGAT'\n",
    "\n",
    "    #Create a list of the codons they called missense variants at\n",
    "    TET2_variant_list_NP = []\n",
    "    TET2_missense_codons_NP =[]\n",
    "    TET2_missense_variant_DICT_NP = {}\n",
    "\n",
    "    variant_codon = int(((re.findall('\\d+', variant)))[0])\n",
    "\n",
    "    codon0_NP = variant_codon-1\n",
    "    codon1_NP = variant_codon\n",
    "    variant_NP = variant\n",
    "    TET2_missense_codons_NP.append(codon0_NP) #List of codon numbers that they called variants at\n",
    "    TET2_missense_variant_DICT_NP[row_count]=(codon1_NP, variant_NP) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    TET2_missense_codons_NP_DICT = {}\n",
    "    calls_TET2_missense_NP = {}\n",
    "\n",
    "    for i in TET2_missense_codons_NP: #Iterates over the list of codon numbers\n",
    "        TET2_missense_codons_NP_DICT[i+1]= (TET2_codons_NP[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "\n",
    "    for k,v in TET2_missense_codons_NP_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_TET2_missense_NP[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_TET2_missense_NP_extra = {}\n",
    "\n",
    "    for k, v in calls_TET2_missense_NP.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_TET2_missense_NP_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_TET2_missense_NP_extra2 = {}\n",
    "\n",
    "    for k, v in calls_TET2_missense_NP_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_TET2_missense_NP_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    TET2_position_in_list_dict_missense_NP={}\n",
    "    for k, v in TET2_missense_variant_DICT_NP.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_TET2_missense_NP[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_TET2_missense_NP[v[0]]\n",
    "        codon_seq=calls_TET2_missense_NP[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_TET2_missense_NP_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                TET2_position_in_list_dict_missense_NP[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "\n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in TET2_position_in_list_dict_missense_NP.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "    #     print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TET2_bases_missense_NP = {}\n",
    "    for k, v in TET2_position_in_list_dict_missense_NP.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TET2_bases_missense_NP[k]=(v, base_changes)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TET2_bases_missense2_NP = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense_NP.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TET2_bases_missense2_NP[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TET2_bases_missense3_NP = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense2_NP.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TET2_bases_missense3_NP[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TET2_bases_missense4_NP = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense3_NP.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_TET2:\n",
    "    #                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check_TET2(i)\n",
    "                sites_trio_missense.append([adapted_trio])\n",
    "                TET2_bases_missense4_NP[k] = (v, sites_trio_missense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_missense = [TET2_exons_extra_bases_NP[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_missense.append(site_trio_missense)\n",
    "                    TET2_bases_missense4_NP[k] = (v, sites_trio_missense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TET2_sites_missense5_NP = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TET2_bases_missense4_NP.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        TET2_sites_missense5_NP[a]=info\n",
    "        a=a+1\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TET2_sites_missense6_NP={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TET2_sites_missense5_NP.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (original_base,new_base)\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TET2_sites_missense6_NP[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    TET2_mutation_rates_missense_NP = {}\n",
    "    for k, v in TET2_sites_missense6_NP.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            TET2_mutation_rates_missense_NP[k] = (v, mutation_rate)\n",
    "\n",
    "    TET2_mutation_rate_missense_NP_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TET2_mutation_rates_missense_NP.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        TET2_mutation_rate_missense_NP_sum.append(mutation_rate_value)\n",
    "\n",
    "    TET2_Summed_mutation_rate_missense_NP_added = np.sum(TET2_mutation_rate_missense_NP_sum)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return TET2_Summed_mutation_rate_missense_NP_added"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ASXL1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_variants_ASXL1(variant):\n",
    "    intron_exon_boundary_caution_base_ASXL1 = [140, 143, 252, 373, 1086, 1085]\n",
    "   \n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "\n",
    "    def intron_exon_check(position_in_gene_number):\n",
    "        if position_in_gene_number == 140:\n",
    "            gene_pos = 140\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 143:\n",
    "            gene_pos = 143\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 252:\n",
    "            gene_pos = 252\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 373:\n",
    "            gene_pos = 373\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1086:\n",
    "            gene_pos = 1086\n",
    "            first_base = 'G'\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = ASXL1_exons_extra_bases[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1085:\n",
    "            gene_pos = 1085\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #ASXL1 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    ASXL1_exons = 'ATGAAGGACAAACAGAAGAAGAAGAAGGAGCGCACGTGGGCCGAGGCCGCGCGCCTGGTATTAGAAAACTACTCGGATGCTCCAATGACACCAAAACAGATTCTGCAGGTCATAGAGGCAGAAGGACTAAAGGAAATGAGAAGTGGGACTTCCCCTCTCGCATGCCTCAATGCTATGCTACATTCCAATTCAAGAGGAGGAGAGGGGTTGTTTTATAAACTGCCTGGCCGAATCAGCCTTTTCACGCTCAAGAAGGATGCCCTGCAGTGGTCTCGCCATCCAGCTACAGTGGAGGGAGAGGAGCCAGAGGACACGGCTGATGTGGAGAGCTGTGGGTCTAATGAAGCCAGCACTGTGAGTGGTGAAAACGATGTATCTCTTGATGAAACATCTTCGAACGCATCCTGTTCTACAGAATCTCAGAGTCGACCTCTTTCCAATCCCAGGGACAGCTACAGAGCTTCCTCACAGGCGAACAAACAAAAGAAAAAGACTGGGGTGATGCTGCCTCGAGTTGTCCTGACTCCTCTGAAGGTAAACGGGGCCCACGTGGAATCTGCATCAGGGTTCTCGGGCTGCCACGCCGATGGCGAGAGCGGCAGCCCGTCCAGCAGCAGCAGCGGCTCTCTGGCCCTGGGCAGCGCTGCTATTCGTGGCCAGGCCGAGGTCACCCAGGACCCTGCCCCGCTCCTGAGAGGCTTCCGGAAGCCAGCCACAGGTCAAATGAAGCGCAACAGAGGGGAAGAAATAGATTTTGAGACACCTGGGTCCATTCTTGTCAACACCAACCTCCGTGCCCTGATCAACTCTCGGACCTTCCATGCCTTACCATCACACTTCCAGCAGCAGCTCCTCTTCCTCCTGCCTGAAGTAGACAGACAGGTGGGGACGGATGGCCTGTTGCGTCTCAGCAGCAGTGCACTAAATAACGAGTTTTTTACCCATGCGGCTCAGAGCTGGCGGGAGCGCCTGGCTGATGGTGAATTTACTCATGAGATGCAAGTCAGGATACGACAGGAAATGGAGAAGGAAAAGAAGGTGGAACAATGGAAAGAAAAGTTCTTTGAAGACTACTATGGACAGAAGCTGGGTTTGACCAAAGAAGAGTCATTGCAGCAGAACGTGGGCCAGGAGGAGGCTGAAATCAAAAGTGGCTTGTGTGTCCCAGGAGAATCAGTGCGTATACAGCGTGGTCCAGCCACCCGACAGCGAGATGGGCATTTTAAGAAACGCTCTCGGCCAGATCTCCGAACCAGAGCCAGAAGGAATCTGTACAAAAAACAGGAGTCAGAACAAGCAGGGGTTGCTAAGGATGCAAAATCTGTGGCCTCAGATGTTCCCCTCTACAAGGATGGGGAGGCTAAGACTGACCCAGCAGGGCTGAGCAGTCCCCATCTGCCAGGCACATCCTCTGCAGCACCCGACCTGGAGGGTCCCGAATTCCCAGTTGAGTCTGTGGCTTCTCGGATCCAGGCTGAGCCAGACAACTTGGCACGTGCCTCTGCATCTCCAGACAGAATTCCTAGCCTGCCTCAGGAAACTGTGGATCAGGAACCCAAGGATCAGAAGAGGAAATCCTTTGAGCAGGCGGCCTCTGCATCCTTTCCCGAAAAGAAGCCCCGGCTTGAAGATCGTCAGTCCTTTCGTAACACAATTGAAAGTGTTCACACCGAAAAGCCACAGCCCACTAAAGAGGAGCCCAAAGTCCCGCCCATCCGGATTCAACTTTCACGTATCAAACCACCCTGGGTGGTTAAAGGTCAGCCCACTTACCAGATATGCCCCCGGATCATCCCCACCACGGAGTCCTCCTGCCGGGGTTGGACTGGCGCCAGGACCCTCGCAGACATTAAAGCCCGTGCTCTGCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCCACCACTGCCATCGGAGGGGGGGGTGGCCCGGGTGGAGGTGGCGGCGGGGCCACCGATGAGGGAGGTGGCAGAGGCAGCAGCAGTGGTGATGGTGGTGAGGCCTGTGGCCACCCTGAGCCCAGGGGAGGCCCGAGCACCCCTGGAAAGTGTACGTCAGATCTACAGCGAACACAACTACTGCCGCCTTATCCTCTAAATGGGGAGCATACCCAGGCCGGAACTGCCATGTCCAGAGCTAGGAGAGAGGACCTGCCTTCTCTGAGAAAGGAGGAAAGCTGCCTACTACAGAGGGCTACAGTTGGACTCACAGATGGGCTAGGAGATGCCTCCCAACTCCCCGTTGCTCCCACTGGGGACCAGCCATGCCAGGCCTTGCCCCTACTGTCCTCCCAAACCTCAGTAGCTGAGAGATTAGTGGAGCAGCCTCAGTTGCATCCGGATGTTAGAACTGAATGTGAGTCTGGCACCACTTCCTGGGAAAGTGATGATGAGGAGCAAGGACCCACCGTTCCTGCAGACAATGGTCCCATTCTGTCTCTAGTGGGAGATGATACATTAGAGAAAGGAACTGGCCAAGCTCTTGACAGTCATCCCACTATGAAGGATCCTGTAAATGTGACCCCCAGTTCCACACCTGAATCCTCACCGACTGATTGCCTGCAGAACAGAGCATTTGATGACGAATTAGGGCTTGGTGGCTCATGCCCTCCTATGAGGGAAAGTGATACTAGACAAGAAAACTTGAAAACCAAGGCTCTCGTTTCTAACAGTTCTTTGCATTGGATACCCATCCCATCGAATGATGAGGTAGTGAAACAGCCCAAACCAGAATCCAGAGAACACATACCATCTGTTGAGCCCCAGGTTGGAGAGGAGTGGGAGAAAGCTGCTCCCACCCCTCCTGCATTGCCTGGGGATTTGACAGCTGAGGAGGGTCTAGATCCTCTTGACAGCCTTACTTCACTCTGGACTGTGCCATCTCGAGGAGGCAGTGACAGCAATGGCAGTTACTGTCAACAGGTGGACATTGAAAAGCTGAAAATCAACGGAGACTCTGAAGCACTGAGTCCTCACGGTGAGTCCACGGATACAGCCTCTGACTTTGAAGGTCACCTCACGGAGGACAGCAGTGAGGCTGACACTAGAGAAGCTGCAGTGACAAAGGGATCTTCGGTGGACAAGGATGAGAAACCCAATTGGAACCAATCTGCCCCACTGTCCAAGGTGAATGGTGACATGCGTCTGGTTACAAGGACAGATGGGATGGTTGCTCCTCAGAGCTGGGTGTCTCGAGTATGTGCGGTCCGCCAAAAGATCCCAGATTCCCTACTGCTGGCCAGTACTGAGTACCAGCCAAGAGCCGTGTGCCTGTCCATGCCTGGGTCCTCAGTGGAGGCCACTAACCCACTTGTGATGCAGTTGCTGCAGGGTAGCTTGCCCCTAGAGAAGGTTCTTCCACCAGCCCACGATGACAGCATGTCAGAATCCCCACAAGTACCACTTACAAAAGACCAGAGCCATGGCTCGCTACGCATGGGATCTTTACATGGTCTTGGAAAAAACAGTGGCATGGTTGATGGAAGCAGCCCCAGTTCTTTAAGGGCTTTGAAGGAGCCTCTTCTGCCAGATAGCTGTGAAACAGGCACTGGTCTTGCCAGGATTGAGGCCACCCAGGCTCCTGGAGCACCCCAAAAGAATTGCAAGGCAGTCCCAAGTTTTGACTCCCTCCATCCAGTGACAAATCCCATTACATCCTCTAGGAAACTGGAAGAAATGGATTCCAAAGAGCAGTTCTCTTCCTTTAGTTGTGAAGATCAGAAGGAAGTCCGTGCTATGTCACAGGACAGTAATTCAAATGCTGCTCCAGGAAAGAGCCCAGGAGATCTTACTACCTCGAGAACACCTCGTTTCTCATCTCCAAATGTGATCTCCTTTGGTCCAGAGCAGACAGGTCGGGCCCTGGGTGATCAGAGCAATGTTACAGGCCAAGGGAAGAAGCTTTTTGGCTCTGGGAATGTGGCTGCAACCCTTCAGCGCCCCAGGCCTGCGGACCCGATGCCTCTTCCTGCTGAGATCCCTCCAGTTTTTCCCAGTGGGAAGTTGGGACCAAGCACAAACTCCATGTCTGGTGGGGTACAGACTCCAAGGGAAGACTGGGCTCCAAAGCCACATGCCTTTGTTGGCAGCGTCAAGAATGAGAAGACTTTTGTGGGGGGTCCTCTTAAGGCAAATGCCGAGAACAGGAAAGCTACTGGGCATAGTCCCCTGGAACTGGTGGGTCACTTGGAAGGGATGCCCTTTGTCATGGACTTGCCCTTCTGGAAATTACCCCGAGAGCCAGGGAAGGGGCTCAGTGAGCCTCTGGAGCCTTCTTCTCTCCCCTCCCAACTCAGCATCAAGCAGGCATTTTATGGGAAGCTTTCTAAACTCCAACTGAGTTCCACCAGCTTTAATTATTCCTCTAGCTCTCCCACCTTTCCCAAAGGCCTTGCTGGAAGTGTGGTGCAGCTGAGCCACAAAGCAAACTTTGGTGCGAGCCACAGTGCATCACTTTCCTTGCAAATGTTCACTGACAGCAGCACGGTGGAAAGCATCTCGCTCCAGTGTGCGTGCAGCCTGAAAGCCATGATCATGTGCCAAGGCTGCGGTGCGTTCTGTCACGATGACTGTATTGGACCCTCAAAGCTCTGTGTATTGTGCCTTGTGGTGAGATAA'\n",
    "    \n",
    "    #Split the sequence into codons \n",
    "    ASXL1_codons = [ASXL1_exons[i:i+3] for i in range(0, len(ASXL1_exons), 3)]\n",
    "\n",
    "    #ASXL1 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    ASXL1_exons_extra_bases = 'GATGAAGGACAAACAGAAGAAGAAGAAGGAGCGCACGTGGGCCGAGGCCGCGCGCCTGGTATTAGAAAACTACTCGGATGCTCCAATGACACCAAAACAGATTCTGCAGGTCATAGAGGCAGAAGGACTAAAGGAAATGAGAAGTGGGACTTCCCCTCTCGCATGCCTCAATGCTATGCTACATTCCAATTCAAGAGGAGGAGAGGGGTTGTTTTATAAACTGCCTGGCCGAATCAGCCTTTTCACGCTCAAGAAGGATGCCCTGCAGTGGTCTCGCCATCCAGCTACAGTGGAGGGAGAGGAGCCAGAGGACACGGCTGATGTGGAGAGCTGTGGGTCTAATGAAGCCAGCACTGTGAGTGGTGAAAACGATGTATCTCTTGATGAAACATCTTCGAACGCATCCTGTTCTACAGAATCTCAGAGTCGACCTCTTTCCAATCCCAGGGACAGCTACAGAGCTTCCTCACAGGCGAACAAACAAAAGAAAAAGACTGGGGTGATGCTGCCTCGAGTTGTCCTGACTCCTCTGAAGGTAAACGGGGCCCACGTGGAATCTGCATCAGGGTTCTCGGGCTGCCACGCCGATGGCGAGAGCGGCAGCCCGTCCAGCAGCAGCAGCGGCTCTCTGGCCCTGGGCAGCGCTGCTATTCGTGGCCAGGCCGAGGTCACCCAGGACCCTGCCCCGCTCCTGAGAGGCTTCCGGAAGCCAGCCACAGGTCAAATGAAGCGCAACAGAGGGGAAGAAATAGATTTTGAGACACCTGGGTCCATTCTTGTCAACACCAACCTCCGTGCCCTGATCAACTCTCGGACCTTCCATGCCTTACCATCACACTTCCAGCAGCAGCTCCTCTTCCTCCTGCCTGAAGTAGACAGACAGGTGGGGACGGATGGCCTGTTGCGTCTCAGCAGCAGTGCACTAAATAACGAGTTTTTTACCCATGCGGCTCAGAGCTGGCGGGAGCGCCTGGCTGATGGTGAATTTACTCATGAGATGCAAGTCAGGATACGACAGGAAATGGAGAAGGAAAAGAAGGTGGAACAATGGAAAGAAAAGTTCTTTGAAGACTACTATGGACAGAAGCTGGGTTTGACCAAAGAAGAGTCATTGCAGCAGAACGTGGGCCAGGAGGAGGCTGAAATCAAAAGTGGCTTGTGTGTCCCAGGAGAATCAGTGCGTATACAGCGTGGTCCAGCCACCCGACAGCGAGATGGGCATTTTAAGAAACGCTCTCGGCCAGATCTCCGAACCAGAGCCAGAAGGAATCTGTACAAAAAACAGGAGTCAGAACAAGCAGGGGTTGCTAAGGATGCAAAATCTGTGGCCTCAGATGTTCCCCTCTACAAGGATGGGGAGGCTAAGACTGACCCAGCAGGGCTGAGCAGTCCCCATCTGCCAGGCACATCCTCTGCAGCACCCGACCTGGAGGGTCCCGAATTCCCAGTTGAGTCTGTGGCTTCTCGGATCCAGGCTGAGCCAGACAACTTGGCACGTGCCTCTGCATCTCCAGACAGAATTCCTAGCCTGCCTCAGGAAACTGTGGATCAGGAACCCAAGGATCAGAAGAGGAAATCCTTTGAGCAGGCGGCCTCTGCATCCTTTCCCGAAAAGAAGCCCCGGCTTGAAGATCGTCAGTCCTTTCGTAACACAATTGAAAGTGTTCACACCGAAAAGCCACAGCCCACTAAAGAGGAGCCCAAAGTCCCGCCCATCCGGATTCAACTTTCACGTATCAAACCACCCTGGGTGGTTAAAGGTCAGCCCACTTACCAGATATGCCCCCGGATCATCCCCACCACGGAGTCCTCCTGCCGGGGTTGGACTGGCGCCAGGACCCTCGCAGACATTAAAGCCCGTGCTCTGCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCCACCACTGCCATCGGAGGGGGGGGTGGCCCGGGTGGAGGTGGCGGCGGGGCCACCGATGAGGGAGGTGGCAGAGGCAGCAGCAGTGGTGATGGTGGTGAGGCCTGTGGCCACCCTGAGCCCAGGGGAGGCCCGAGCACCCCTGGAAAGTGTACGTCAGATCTACAGCGAACACAACTACTGCCGCCTTATCCTCTAAATGGGGAGCATACCCAGGCCGGAACTGCCATGTCCAGAGCTAGGAGAGAGGACCTGCCTTCTCTGAGAAAGGAGGAAAGCTGCCTACTACAGAGGGCTACAGTTGGACTCACAGATGGGCTAGGAGATGCCTCCCAACTCCCCGTTGCTCCCACTGGGGACCAGCCATGCCAGGCCTTGCCCCTACTGTCCTCCCAAACCTCAGTAGCTGAGAGATTAGTGGAGCAGCCTCAGTTGCATCCGGATGTTAGAACTGAATGTGAGTCTGGCACCACTTCCTGGGAAAGTGATGATGAGGAGCAAGGACCCACCGTTCCTGCAGACAATGGTCCCATTCTGTCTCTAGTGGGAGATGATACATTAGAGAAAGGAACTGGCCAAGCTCTTGACAGTCATCCCACTATGAAGGATCCTGTAAATGTGACCCCCAGTTCCACACCTGAATCCTCACCGACTGATTGCCTGCAGAACAGAGCATTTGATGACGAATTAGGGCTTGGTGGCTCATGCCCTCCTATGAGGGAAAGTGATACTAGACAAGAAAACTTGAAAACCAAGGCTCTCGTTTCTAACAGTTCTTTGCATTGGATACCCATCCCATCGAATGATGAGGTAGTGAAACAGCCCAAACCAGAATCCAGAGAACACATACCATCTGTTGAGCCCCAGGTTGGAGAGGAGTGGGAGAAAGCTGCTCCCACCCCTCCTGCATTGCCTGGGGATTTGACAGCTGAGGAGGGTCTAGATCCTCTTGACAGCCTTACTTCACTCTGGACTGTGCCATCTCGAGGAGGCAGTGACAGCAATGGCAGTTACTGTCAACAGGTGGACATTGAAAAGCTGAAAATCAACGGAGACTCTGAAGCACTGAGTCCTCACGGTGAGTCCACGGATACAGCCTCTGACTTTGAAGGTCACCTCACGGAGGACAGCAGTGAGGCTGACACTAGAGAAGCTGCAGTGACAAAGGGATCTTCGGTGGACAAGGATGAGAAACCCAATTGGAACCAATCTGCCCCACTGTCCAAGGTGAATGGTGACATGCGTCTGGTTACAAGGACAGATGGGATGGTTGCTCCTCAGAGCTGGGTGTCTCGAGTATGTGCGGTCCGCCAAAAGATCCCAGATTCCCTACTGCTGGCCAGTACTGAGTACCAGCCAAGAGCCGTGTGCCTGTCCATGCCTGGGTCCTCAGTGGAGGCCACTAACCCACTTGTGATGCAGTTGCTGCAGGGTAGCTTGCCCCTAGAGAAGGTTCTTCCACCAGCCCACGATGACAGCATGTCAGAATCCCCACAAGTACCACTTACAAAAGACCAGAGCCATGGCTCGCTACGCATGGGATCTTTACATGGTCTTGGAAAAAACAGTGGCATGGTTGATGGAAGCAGCCCCAGTTCTTTAAGGGCTTTGAAGGAGCCTCTTCTGCCAGATAGCTGTGAAACAGGCACTGGTCTTGCCAGGATTGAGGCCACCCAGGCTCCTGGAGCACCCCAAAAGAATTGCAAGGCAGTCCCAAGTTTTGACTCCCTCCATCCAGTGACAAATCCCATTACATCCTCTAGGAAACTGGAAGAAATGGATTCCAAAGAGCAGTTCTCTTCCTTTAGTTGTGAAGATCAGAAGGAAGTCCGTGCTATGTCACAGGACAGTAATTCAAATGCTGCTCCAGGAAAGAGCCCAGGAGATCTTACTACCTCGAGAACACCTCGTTTCTCATCTCCAAATGTGATCTCCTTTGGTCCAGAGCAGACAGGTCGGGCCCTGGGTGATCAGAGCAATGTTACAGGCCAAGGGAAGAAGCTTTTTGGCTCTGGGAATGTGGCTGCAACCCTTCAGCGCCCCAGGCCTGCGGACCCGATGCCTCTTCCTGCTGAGATCCCTCCAGTTTTTCCCAGTGGGAAGTTGGGACCAAGCACAAACTCCATGTCTGGTGGGGTACAGACTCCAAGGGAAGACTGGGCTCCAAAGCCACATGCCTTTGTTGGCAGCGTCAAGAATGAGAAGACTTTTGTGGGGGGTCCTCTTAAGGCAAATGCCGAGAACAGGAAAGCTACTGGGCATAGTCCCCTGGAACTGGTGGGTCACTTGGAAGGGATGCCCTTTGTCATGGACTTGCCCTTCTGGAAATTACCCCGAGAGCCAGGGAAGGGGCTCAGTGAGCCTCTGGAGCCTTCTTCTCTCCCCTCCCAACTCAGCATCAAGCAGGCATTTTATGGGAAGCTTTCTAAACTCCAACTGAGTTCCACCAGCTTTAATTATTCCTCTAGCTCTCCCACCTTTCCCAAAGGCCTTGCTGGAAGTGTGGTGCAGCTGAGCCACAAAGCAAACTTTGGTGCGAGCCACAGTGCATCACTTTCCTTGCAAATGTTCACTGACAGCAGCACGGTGGAAAGCATCTCGCTCCAGTGTGCGTGCAGCCTGAAAGCCATGATCATGTGCCAAGGCTGCGGTGCGTTCTGTCACGATGACTGTATTGGACCCTCAAAGCTCTGTGTATTGTGCCTTGTGGTGAGATAAT'\n",
    "  \n",
    "    #Create a list of the codons they called missense variants at\n",
    "    ASXL1_variant_list = []\n",
    "    ASXL1_missense_codons =[]\n",
    "    ASXL1_missense_variant_DICT = {}\n",
    "\n",
    "    variant_codon = int(((re.findall('\\d+', variant)))[0])\n",
    "\n",
    "    codon0 = variant_codon-1\n",
    "    codon1 = variant_codon\n",
    "    variant = variant\n",
    "    ASXL1_missense_codons.append(codon0) #List of codon numbers that they called variants at\n",
    "    ASXL1_missense_variant_DICT[row_count]=(codon1, variant) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    ASXL1_missense_codons_DICT = {}\n",
    "    calls_ASXL1_missense = {}\n",
    "\n",
    "    for i in ASXL1_missense_codons: #Iterates over the list of codon numbers\n",
    "        ASXL1_missense_codons_DICT[i+1]= (ASXL1_codons[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "\n",
    "    for k,v in ASXL1_missense_codons_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_ASXL1_missense[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_ASXL1_missense_extra = {}\n",
    "\n",
    "    for k, v in calls_ASXL1_missense.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_ASXL1_missense_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_ASXL1_missense_extra2 = {}\n",
    "\n",
    "    for k, v in calls_ASXL1_missense_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_ASXL1_missense_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    ASXL1_position_in_list_dict_missense={}\n",
    "    for k, v in ASXL1_missense_variant_DICT.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_ASXL1_missense[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_ASXL1_missense[v[0]]\n",
    "        codon_seq=calls_ASXL1_missense[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_ASXL1_missense_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                ASXL1_position_in_list_dict_missense[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "\n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in ASXL1_position_in_list_dict_missense.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "    #     print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    ASXL1_bases_missense = {}\n",
    "    for k, v in ASXL1_position_in_list_dict_missense.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        ASXL1_bases_missense[k]=(v, base_changes)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    ASXL1_bases_missense2 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_missense.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        ASXL1_bases_missense2[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    ASXL1_bases_missense3 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_missense2.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        ASXL1_bases_missense3[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    ASXL1_bases_missense4 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_missense3.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_ASXL1:\n",
    "    #                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check_ASXL1(i)\n",
    "                sites_trio_missense.append([adapted_trio])\n",
    "                ASXL1_bases_missense4[k] = (v, sites_trio_missense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_missense = [ASXL1_exons_extra_bases[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_missense.append(site_trio_missense)\n",
    "                    ASXL1_bases_missense4[k] = (v, sites_trio_missense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    ASXL1_sites_missense5 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in ASXL1_bases_missense4.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        ASXL1_sites_missense5[a]=info\n",
    "        a=a+1\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    ASXL1_sites_missense6={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in ASXL1_sites_missense5.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (original_base,new_base)\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        ASXL1_sites_missense6[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    ASXL1_mutation_rates_missense = {}\n",
    "    for k, v in ASXL1_sites_missense6.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            ASXL1_mutation_rates_missense[k] = (v, mutation_rate)\n",
    "\n",
    "    ASXL1_mutation_rate_missense_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in ASXL1_mutation_rates_missense.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        ASXL1_mutation_rate_missense_sum.append(mutation_rate_value)\n",
    "\n",
    "    ASXL1_Summed_mutation_rate_missense_added = np.sum(ASXL1_mutation_rate_missense_sum)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return ASXL1_Summed_mutation_rate_missense_added"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TP53"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_variants_TP53(variant):\n",
    "    intron_exon_boundary_caution_base_TP53 = [74, 75, 96, 375, 672, 782, 919, 993, 1100]\n",
    "\n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    def intron_exon_check(position_in_gene_number):\n",
    "        if position_in_gene_number == 74:\n",
    "            gene_pos = 74\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 75:\n",
    "            gene_pos = 75\n",
    "            first_base = 'G'\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = TP53_exons_extra_bases[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 96:\n",
    "            gene_pos = 96\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 375:\n",
    "            gene_pos = 375\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 672:\n",
    "            gene_pos = 672\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 782:\n",
    "            gene_pos = 782\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 919:\n",
    "            gene_pos = 919\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 993:\n",
    "            gene_pos = 993\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1100:\n",
    "            gene_pos = 1100\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #TP53 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    TP53_exons = 'ATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGACATAGTGTGGTGGTGCCCTATGAGCCGCCTGAGGTTGGCTCTGACTGTACCACCATCCACTACAACTACATGTGTAACAGTTCCTGCATGGGCGGCATGAACCGGAGGCCCATCCTCACCATCATCACACTGGAAGACTCCAGTGGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGAGGAAGAGAATCTCCGCAAGAAAGGGGAGCCTCACCACGAGCTGCCCCCAGGGAGCACTAAGCGAGCACTGCCCAACAACACCAGCTCCTCTCCCCAGCCAAAGAAGAAACCACTGGATGGAGAATATTTCACCCTTCAGATCCGTGGGCGTGAGCGCTTCGAGATGTTCCGAGAGCTGAATGAGGCCTTGGAACTCAAGGATGCCCAGGCTGGGAAGGAGCCAGGGGGGAGCAGGGCTCACTCCAGCCACCTGAAGTCCAAAAAGGGTCAGTCTACCTCCCGCCATAAAAAACTCATGTTCAAGACAGAAGGGCCTGACTCAGACTGA'\n",
    "    \n",
    "    #Split the sequence into codons \n",
    "    TP53_codons = [TP53_exons[i:i+3] for i in range(0, len(TP53_exons), 3)]\n",
    "\n",
    "    #TP53 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    TP53_exons_extra_bases = 'CATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGACATAGTGTGGTGGTGCCCTATGAGCCGCCTGAGGTTGGCTCTGACTGTACCACCATCCACTACAACTACATGTGTAACAGTTCCTGCATGGGCGGCATGAACCGGAGGCCCATCCTCACCATCATCACACTGGAAGACTCCAGTGGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGAGGAAGAGAATCTCCGCAAGAAAGGGGAGCCTCACCACGAGCTGCCCCCAGGGAGCACTAAGCGAGCACTGCCCAACAACACCAGCTCCTCTCCCCAGCCAAAGAAGAAACCACTGGATGGAGAATATTTCACCCTTCAGATCCGTGGGCGTGAGCGCTTCGAGATGTTCCGAGAGCTGAATGAGGCCTTGGAACTCAAGGATGCCCAGGCTGGGAAGGAGCCAGGGGGGAGCAGGGCTCACTCCAGCCACCTGAAGTCCAAAAAGGGTCAGTCTACCTCCCGCCATAAAAAACTCATGTTCAAGACAGAAGGGCCTGACTCAGACTGAC'\n",
    "    \n",
    "    #Create a list of the codons they called missense variants at (from COSMIC)\n",
    "    TP53_variant_list = []\n",
    "    TP53_missense_codons =[]\n",
    "    TP53_missense_variant_DICT = {}\n",
    "\n",
    "    variant_codon = int(((re.findall('\\d+', variant)))[0])\n",
    "\n",
    "    codon0 = variant_codon-1\n",
    "    codon1 = variant_codon\n",
    "    variant = variant\n",
    "    TP53_missense_codons.append(codon0) #List of codon numbers that they called variants at\n",
    "    TP53_missense_variant_DICT[row_count]=(codon1, variant) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    TP53_missense_codons_DICT = {}\n",
    "    calls_TP53_missense = {}\n",
    "\n",
    "    for i in TP53_missense_codons: #Iterates over the list of codon numbers\n",
    "        TP53_missense_codons_DICT[i+1]= (TP53_codons[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "\n",
    "    for k,v in TP53_missense_codons_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_TP53_missense[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_TP53_missense_extra = {}\n",
    "\n",
    "    for k, v in calls_TP53_missense.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_TP53_missense_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_TP53_missense_extra2 = {}\n",
    "\n",
    "    for k, v in calls_TP53_missense_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_TP53_missense_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    TP53_position_in_list_dict_missense={}\n",
    "    for k, v in TP53_missense_variant_DICT.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_TP53_missense[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_TP53_missense[v[0]]\n",
    "        codon_seq=calls_TP53_missense[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_TP53_missense_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                TP53_position_in_list_dict_missense[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "\n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in TP53_position_in_list_dict_missense.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "    #     print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TP53_bases_missense = {}\n",
    "    for k, v in TP53_position_in_list_dict_missense.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TP53_bases_missense[k]=(v, base_changes)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TP53_bases_missense2 = {}\n",
    "\n",
    "    for k, v in TP53_bases_missense.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TP53_bases_missense2[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TP53_bases_missense3 = {}\n",
    "\n",
    "    for k, v in TP53_bases_missense2.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TP53_bases_missense3[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TP53_bases_missense4 = {}\n",
    "\n",
    "    for k, v in TP53_bases_missense3.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_TP53:\n",
    "    #                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check(i)\n",
    "                sites_trio_missense.append([adapted_trio])\n",
    "                TP53_bases_missense4[k] = (v, sites_trio_missense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_missense = [TP53_exons_extra_bases[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_missense.append(site_trio_missense)\n",
    "                    TP53_bases_missense4[k] = (v, sites_trio_missense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TP53_sites_missense5 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TP53_bases_missense4.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        TP53_sites_missense5[a]=info\n",
    "        a=a+1\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TP53_sites_missense6={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TP53_sites_missense5.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (original_base,new_base)\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TP53_sites_missense6[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    TP53_mutation_rates_missense = {}\n",
    "    for k, v in TP53_sites_missense6.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            TP53_mutation_rates_missense[k] = (v, mutation_rate)\n",
    "\n",
    "    TP53_mutation_rate_missense_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TP53_mutation_rates_missense.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        TP53_mutation_rate_missense_sum.append(mutation_rate_value)\n",
    "\n",
    "    TP53_Summed_mutation_rate_missense_added = np.sum(TP53_mutation_rate_missense_sum)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return TP53_Summed_mutation_rate_missense_added"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculating mutation rates from a .csv file of variants"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DNMT3A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_list_nonsense_variants_DNMT3A_NP072046(csv_file_variants):\n",
    "   \n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "\n",
    "    intron_exon_boundary_caution_base_DNMT3A_NP072046 = [1279, 1429, 1474, 1554, 1852, 2083, 2082, 2173, 2322, 2478]\n",
    "   \n",
    "    def intron_exon_check_DNMT3A_NP072046(position_in_gene_number):\n",
    "        if position_in_gene_number == 1279:\n",
    "            gene_pos = 1279\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1429:\n",
    "            gene_pos = 1429\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1474:\n",
    "            gene_pos = 1474\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1554:\n",
    "            gene_pos = 1554\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1852:\n",
    "            gene_pos = 1852\n",
    "            first_base = 'G'\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2083:\n",
    "            gene_pos = 2083\n",
    "            first_base = 'G'\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2082:\n",
    "            gene_pos = 2082\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2173:\n",
    "            gene_pos = 2173\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2322:\n",
    "            gene_pos = 2322\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2478:\n",
    "            gene_pos = 2478\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #DNMT3A transcript used:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000119772;r=2:25455845-25564754;t=ENST00000321117\n",
    "    DNMT3A_exons_NP_072046 = 'ATGCCCGCCATGCCCTCCAGCGGCCCCGGGGACACCAGCAGCTCTGCTGCGGAGCGGGAGGAGGACCGAAAGGACGGAGAGGAGCAGGAGGAGCCGCGTGGCAAGGAGGAGCGCCAAGAGCCCAGCACCACGGCACGGAAGGTGGGGCGGCCTGGGAGGAAGCGCAAGCACCCCCCGGTGGAAAGCGGTGACACGCCAAAGGACCCTGCGGTGATCTCCAAGTCCCCATCCATGGCCCAGGACTCAGGCGCCTCAGAGCTATTACCCAATGGGGACTTGGAGAAGCGGAGTGAGCCCCAGCCAGAGGAGGGGAGCCCTGCTGGGGGGCAGAAGGGCGGGGCCCCAGCAGAGGGAGAGGGTGCAGCTGAGACCCTGCCTGAAGCCTCAAGAGCAGTGGAAAATGGCTGCTGCACCCCCAAGGAGGGCCGAGGAGCCCCTGCAGAAGCGGGCAAAGAACAGAAGGAGACCAACATCGAATCCATGAAAATGGAGGGCTCCCGGGGCCGGCTGCGGGGTGGCTTGGGCTGGGAGTCCAGCCTCCGTCAGCGGCCCATGCCGAGGCTCACCTTCCAGGCGGGGGACCCCTACTACATCAGCAAGCGCAAGCGGGACGAGTGGCTGGCACGCTGGAAAAGGGAGGCTGAGAAGAAAGCCAAGGTCATTGCAGGAATGAATGCTGTGGAAGAAAACCAGGGGCCCGGGGAGTCTCAGAAGGTGGAGGAGGCCAGCCCTCCTGCTGTGCAGCAGCCCACTGACCCCGCATCCCCCACTGTGGCTACCACGCCTGAGCCCGTGGGGTCCGATGCTGGGGACAAGAATGCCACCAAAGCAGGCGATGACGAGCCAGAGTACGAGGACGGCCGGGGCTTTGGCATTGGGGAGCTGGTGTGGGGGAAACTGCGGGGCTTCTCCTGGTGGCCAGGCCGCATTGTGTCTTGGTGGATGACGGGCCGGAGCCGAGCAGCTGAAGGCACCCGCTGGGTCATGTGGTTCGGAGACGGCAAATTCTCAGTGGTGTGTGTTGAGAAGCTGATGCCGCTGAGCTCGTTTTGCAGTGCGTTCCACCAGGCCACGTACAACAAGCAGCCCATGTACCGCAAAGCCATCTACGAGGTCCTGCAGGTGGCCAGCAGCCGCGCGGGGAAGCTGTTCCCGGTGTGCCACGACAGCGATGAGAGTGACACTGCCAAGGCCGTGGAGGTGCAGAACAAGCCCATGATTGAATGGGCCCTGGGGGGCTTCCAGCCTTCTGGCCCTAAGGGCCTGGAGCCACCAGAAGAAGAGAAGAATCCCTACAAAGAAGTGTACACGGACATGTGGGTGGAACCTGAGGCAGCTGCCTACGCACCACCTCCACCAGCCAAAAAGCCCCGGAAGAGCACAGCGGAGAAGCCCAAGGTCAAGGAGATTATTGATGAGCGCACAAGAGAGCGGCTGGTGTACGAGGTGCGGCAGAAGTGCCGGAACATTGAGGACATCTGCATCTCCTGTGGGAGCCTCAATGTTACCCTGGAACACCCCCTCTTCGTTGGAGGAATGTGCCAAAACTGCAAGAACTGCTTTCTGGAGTGTGCGTACCAGTACGACGACGACGGCTACCAGTCCTACTGCACCATCTGCTGTGGGGGCCGTGAGGTGCTCATGTGCGGAAACAACAACTGCTGCAGGTGCTTTTGCGTGGAGTGTGTGGACCTCTTGGTGGGGCCGGGGGCTGCCCAGGCAGCCATTAAGGAAGACCCCTGGAACTGCTACATGTGCGGGCACAAGGGTACCTACGGGCTGCTGCGGCGGCGAGAGGACTGGCCCTCCCGGCTCCAGATGTTCTTCGCTAATAACCACGACCAGGAATTTGACCCTCCAAAGGTTTACCCACCTGTCCCAGCTGAGAAGAGGAAGCCCATCCGGGTGCTGTCTCTCTTTGATGGAATCGCTACAGGGCTCCTGGTGCTGAAGGACTTGGGCATTCAGGTGGACCGCTACATTGCCTCGGAGGTGTGTGAGGACTCCATCACGGTGGGCATGGTGCGGCACCAGGGGAAGATCATGTACGTCGGGGACGTCCGCAGCGTCACACAGAAGCATATCCAGGAGTGGGGCCCATTCGATCTGGTGATTGGGGGCAGTCCCTGCAATGACCTCTCCATCGTCAACCCTGCTCGCAAGGGCCTCTACGAGGGCACTGGCCGGCTCTTCTTTGAGTTCTACCGCCTCCTGCATGATGCGCGGCCCAAGGAGGGAGATGATCGCCCCTTCTTCTGGCTCTTTGAGAATGTGGTGGCCATGGGCGTTAGTGACAAGAGGGACATCTCGCGATTTCTCGAGTCCAACCCTGTGATGATTGATGCCAAAGAAGTGTCAGCTGCACACAGGGCCCGCTACTTCTGGGGTAACCTTCCCGGTATGAACAGGCCGTTGGCATCCACTGTGAATGATAAGCTGGAGCTGCAGGAGTGTCTGGAGCATGGCAGGATAGCCAAGTTCAGCAAAGTGAGGACCATTACTACGAGGTCAAACTCCATAAAGCAGGGCAAAGACCAGCATTTTCCTGTCTTCATGAATGAGAAAGAGGACATCTTATGGTGCACTGAAATGGAAAGGGTATTTGGTTTCCCAGTCCACTATACTGACGTCTCCAACATGAGCCGCTTGGCGAGGCAGAGACTGCTGGGCCGGTCATGGAGCGTGCCAGTCATCCGCCACCTCTTCGCTCCGCTGAAGGAGTATTTTGCGTGTGTGTAA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    DNMT3A_codons_NP_072046 = [DNMT3A_exons_NP_072046[i:i+3] for i in range(0, len(DNMT3A_exons_NP_072046), 3)]\n",
    "\n",
    "    #DNMT3A transcript used:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000119772;r=2:25455845-25564754;t=ENST00000321117\n",
    "    DNMT3A_exons_extra_bases_NP_072046 = 'GATGCCCGCCATGCCCTCCAGCGGCCCCGGGGACACCAGCAGCTCTGCTGCGGAGCGGGAGGAGGACCGAAAGGACGGAGAGGAGCAGGAGGAGCCGCGTGGCAAGGAGGAGCGCCAAGAGCCCAGCACCACGGCACGGAAGGTGGGGCGGCCTGGGAGGAAGCGCAAGCACCCCCCGGTGGAAAGCGGTGACACGCCAAAGGACCCTGCGGTGATCTCCAAGTCCCCATCCATGGCCCAGGACTCAGGCGCCTCAGAGCTATTACCCAATGGGGACTTGGAGAAGCGGAGTGAGCCCCAGCCAGAGGAGGGGAGCCCTGCTGGGGGGCAGAAGGGCGGGGCCCCAGCAGAGGGAGAGGGTGCAGCTGAGACCCTGCCTGAAGCCTCAAGAGCAGTGGAAAATGGCTGCTGCACCCCCAAGGAGGGCCGAGGAGCCCCTGCAGAAGCGGGCAAAGAACAGAAGGAGACCAACATCGAATCCATGAAAATGGAGGGCTCCCGGGGCCGGCTGCGGGGTGGCTTGGGCTGGGAGTCCAGCCTCCGTCAGCGGCCCATGCCGAGGCTCACCTTCCAGGCGGGGGACCCCTACTACATCAGCAAGCGCAAGCGGGACGAGTGGCTGGCACGCTGGAAAAGGGAGGCTGAGAAGAAAGCCAAGGTCATTGCAGGAATGAATGCTGTGGAAGAAAACCAGGGGCCCGGGGAGTCTCAGAAGGTGGAGGAGGCCAGCCCTCCTGCTGTGCAGCAGCCCACTGACCCCGCATCCCCCACTGTGGCTACCACGCCTGAGCCCGTGGGGTCCGATGCTGGGGACAAGAATGCCACCAAAGCAGGCGATGACGAGCCAGAGTACGAGGACGGCCGGGGCTTTGGCATTGGGGAGCTGGTGTGGGGGAAACTGCGGGGCTTCTCCTGGTGGCCAGGCCGCATTGTGTCTTGGTGGATGACGGGCCGGAGCCGAGCAGCTGAAGGCACCCGCTGGGTCATGTGGTTCGGAGACGGCAAATTCTCAGTGGTGTGTGTTGAGAAGCTGATGCCGCTGAGCTCGTTTTGCAGTGCGTTCCACCAGGCCACGTACAACAAGCAGCCCATGTACCGCAAAGCCATCTACGAGGTCCTGCAGGTGGCCAGCAGCCGCGCGGGGAAGCTGTTCCCGGTGTGCCACGACAGCGATGAGAGTGACACTGCCAAGGCCGTGGAGGTGCAGAACAAGCCCATGATTGAATGGGCCCTGGGGGGCTTCCAGCCTTCTGGCCCTAAGGGCCTGGAGCCACCAGAAGAAGAGAAGAATCCCTACAAAGAAGTGTACACGGACATGTGGGTGGAACCTGAGGCAGCTGCCTACGCACCACCTCCACCAGCCAAAAAGCCCCGGAAGAGCACAGCGGAGAAGCCCAAGGTCAAGGAGATTATTGATGAGCGCACAAGAGAGCGGCTGGTGTACGAGGTGCGGCAGAAGTGCCGGAACATTGAGGACATCTGCATCTCCTGTGGGAGCCTCAATGTTACCCTGGAACACCCCCTCTTCGTTGGAGGAATGTGCCAAAACTGCAAGAACTGCTTTCTGGAGTGTGCGTACCAGTACGACGACGACGGCTACCAGTCCTACTGCACCATCTGCTGTGGGGGCCGTGAGGTGCTCATGTGCGGAAACAACAACTGCTGCAGGTGCTTTTGCGTGGAGTGTGTGGACCTCTTGGTGGGGCCGGGGGCTGCCCAGGCAGCCATTAAGGAAGACCCCTGGAACTGCTACATGTGCGGGCACAAGGGTACCTACGGGCTGCTGCGGCGGCGAGAGGACTGGCCCTCCCGGCTCCAGATGTTCTTCGCTAATAACCACGACCAGGAATTTGACCCTCCAAAGGTTTACCCACCTGTCCCAGCTGAGAAGAGGAAGCCCATCCGGGTGCTGTCTCTCTTTGATGGAATCGCTACAGGGCTCCTGGTGCTGAAGGACTTGGGCATTCAGGTGGACCGCTACATTGCCTCGGAGGTGTGTGAGGACTCCATCACGGTGGGCATGGTGCGGCACCAGGGGAAGATCATGTACGTCGGGGACGTCCGCAGCGTCACACAGAAGCATATCCAGGAGTGGGGCCCATTCGATCTGGTGATTGGGGGCAGTCCCTGCAATGACCTCTCCATCGTCAACCCTGCTCGCAAGGGCCTCTACGAGGGCACTGGCCGGCTCTTCTTTGAGTTCTACCGCCTCCTGCATGATGCGCGGCCCAAGGAGGGAGATGATCGCCCCTTCTTCTGGCTCTTTGAGAATGTGGTGGCCATGGGCGTTAGTGACAAGAGGGACATCTCGCGATTTCTCGAGTCCAACCCTGTGATGATTGATGCCAAAGAAGTGTCAGCTGCACACAGGGCCCGCTACTTCTGGGGTAACCTTCCCGGTATGAACAGGCCGTTGGCATCCACTGTGAATGATAAGCTGGAGCTGCAGGAGTGTCTGGAGCATGGCAGGATAGCCAAGTTCAGCAAAGTGAGGACCATTACTACGAGGTCAAACTCCATAAAGCAGGGCAAAGACCAGCATTTTCCTGTCTTCATGAATGAGAAAGAGGACATCTTATGGTGCACTGAAATGGAAAGGGTATTTGGTTTCCCAGTCCACTATACTGACGTCTCCAACATGAGCCGCTTGGCGAGGCAGAGACTGCTGGGCCGGTCATGGAGCGTGCCAGTCATCCGCCACCTCTTCGCTCCGCTGAAGGAGTATTTTGCGTGTGTGTAAG'\n",
    "\n",
    "    #Create a list of the codons they called nonsense variants at\n",
    "    import csv\n",
    "    filename = csv_file_variants\n",
    "    with open(filename, 'r') as csvfile:\n",
    "        read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "        row_count=0\n",
    "\n",
    "        DNMT3A_nonsense_variant_list_NP_072046 = []\n",
    "        DNMT3A_nonsense_codons_NP_072046 =[]\n",
    "        DNMT3A_nonsense_variant_DICT_NP_072046 = {}\n",
    "\n",
    "        for row in read_reader: #each row contains aa number (e.g. 11) and variant (e.g. p.D11A)\n",
    "    #         print(row)\n",
    "            if row_count>0: \n",
    "                codon0_NP_072046 = int(row[0])-1\n",
    "                codon1_NP_072046 = int(row[0])\n",
    "                variant_NP_072046 = row[1]\n",
    "                DNMT3A_nonsense_codons_NP_072046.append(codon0_NP_072046) #List of codon numbers that they called variants at\n",
    "                DNMT3A_nonsense_variant_DICT_NP_072046[row_count]=(codon1_NP_072046, variant_NP_072046) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "            row_count=row_count+1\n",
    "            \n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    DNMT3A_nonsense_codons_NP_072046_DICT = {}\n",
    "    calls_DNMT3A_nonsense_NP_072046 = {}\n",
    "\n",
    "    for i in DNMT3A_nonsense_codons_NP_072046: #Iterates over the list of codon numbers\n",
    "        DNMT3A_nonsense_codons_NP_072046_DICT[i+1]= (DNMT3A_codons_NP_072046[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "        \n",
    "    for k,v in DNMT3A_nonsense_codons_NP_072046_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_DNMT3A_nonsense_NP_072046[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "        \n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_DNMT3A_nonsense_NP_072046_extra = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_nonsense_NP_072046.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_DNMT3A_nonsense_NP_072046_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_DNMT3A_nonsense_NP_072046_extra2 = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_nonsense_NP_072046_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_DNMT3A_nonsense_NP_072046_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "            \n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    DNMT3A_position_in_list_dict_nonsense_NP_072046={}\n",
    "    for k, v in DNMT3A_nonsense_variant_DICT_NP_072046.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_DNMT3A_nonsense_NP_072046[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_DNMT3A_nonsense_NP_072046[v[0]]\n",
    "        codon_seq=calls_DNMT3A_nonsense_NP_072046[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_DNMT3A_nonsense_NP_072046_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                DNMT3A_position_in_list_dict_nonsense_NP_072046[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "            \n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in DNMT3A_position_in_list_dict_nonsense_NP_072046.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "#     print('positions at which nonsense variants called = ', aa_positions_variants_called)\n",
    "                \n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    DNMT3A_bases_nonsense_NP_072046 = {}\n",
    "    for k, v in DNMT3A_position_in_list_dict_nonsense_NP_072046.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        DNMT3A_bases_nonsense_NP_072046[k]=(v, base_changes)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    DNMT3A_bases_nonsense2_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_nonsense_NP_072046.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        DNMT3A_bases_nonsense2_NP_072046[k] = (v, codon_positions)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    DNMT3A_bases_nonsense3_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_nonsense2_NP_072046.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        DNMT3A_bases_nonsense3_NP_072046[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    DNMT3A_bases_nonsense4_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_nonsense3_NP_072046.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_nonsense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_DNMT3A_NP072046:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check_DNMT3A_NP072046(i)\n",
    "                sites_trio_nonsense.append([adapted_trio])\n",
    "                DNMT3A_bases_nonsense4_NP_072046[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_nonsense = [DNMT3A_exons_extra_bases_NP_072046[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_nonsense.append(site_trio_nonsense)\n",
    "                    DNMT3A_bases_nonsense4_NP_072046[k] = (v, sites_trio_nonsense)\n",
    "                    \n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    DNMT3A_sites_nonsense5_NP_072046 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in DNMT3A_bases_nonsense4_NP_072046.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        DNMT3A_sites_nonsense5_NP_072046[a]=info\n",
    "        a=a+1\n",
    "#     print(DNMT3A_sites_nonsense5_NP_072046)\n",
    "        \n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    DNMT3A_sites_nonsense6_NP_072046={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in DNMT3A_sites_nonsense5_NP_072046.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "    \n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        DNMT3A_sites_nonsense6_NP_072046[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "        \n",
    "        \n",
    "    #Assign the mutation rates to each site...\n",
    "    DNMT3A_mutation_rates_nonsense_NP_072046 = {}\n",
    "    for k, v in DNMT3A_sites_nonsense6_NP_072046.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            DNMT3A_mutation_rates_nonsense_NP_072046[k] = (v, mutation_rate)\n",
    "\n",
    "    DNMT3A_mutation_rate_nonsense_NP_072046_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in DNMT3A_mutation_rates_nonsense_NP_072046.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        DNMT3A_mutation_rate_nonsense_NP_072046_sum.append(mutation_rate_value)\n",
    "\n",
    "    DNMT3A_Summed_mutation_rate_nonsense_NP_072046_added = np.sum(DNMT3A_mutation_rate_nonsense_NP_072046_sum)\n",
    "\n",
    "    print('Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study = ', DNMT3A_Summed_mutation_rate_nonsense_NP_072046_added)\n",
    "    \n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all #in format: column 1 = 882, column 2 = p.R882*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_list_missense_variants_DNMT3A_NP072046(csv_file_variants):\n",
    "    \n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base_DNMT3A_NP072046 = [1279, 1429, 1474, 1554, 1852, 2083, 2082, 2173, 2322, 2478]\n",
    "\n",
    "    def intron_exon_check_DNMT3A_NP072046(position_in_gene_number):\n",
    "        if position_in_gene_number == 1279:\n",
    "            gene_pos = 1279\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1429:\n",
    "            gene_pos = 1429\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1474:\n",
    "            gene_pos = 1474\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1554:\n",
    "            gene_pos = 1554\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1852:\n",
    "            gene_pos = 1852\n",
    "            first_base = 'G'\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2083:\n",
    "            gene_pos = 2083\n",
    "            first_base = 'G'\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2082:\n",
    "            gene_pos = 2082\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2173:\n",
    "            gene_pos = 2173\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2322:\n",
    "            gene_pos = 2322\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2478:\n",
    "            gene_pos = 2478\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #DNMT3A transcript used:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000119772;r=2:25455845-25564754;t=ENST00000321117\n",
    "    DNMT3A_exons_NP_072046 = 'ATGCCCGCCATGCCCTCCAGCGGCCCCGGGGACACCAGCAGCTCTGCTGCGGAGCGGGAGGAGGACCGAAAGGACGGAGAGGAGCAGGAGGAGCCGCGTGGCAAGGAGGAGCGCCAAGAGCCCAGCACCACGGCACGGAAGGTGGGGCGGCCTGGGAGGAAGCGCAAGCACCCCCCGGTGGAAAGCGGTGACACGCCAAAGGACCCTGCGGTGATCTCCAAGTCCCCATCCATGGCCCAGGACTCAGGCGCCTCAGAGCTATTACCCAATGGGGACTTGGAGAAGCGGAGTGAGCCCCAGCCAGAGGAGGGGAGCCCTGCTGGGGGGCAGAAGGGCGGGGCCCCAGCAGAGGGAGAGGGTGCAGCTGAGACCCTGCCTGAAGCCTCAAGAGCAGTGGAAAATGGCTGCTGCACCCCCAAGGAGGGCCGAGGAGCCCCTGCAGAAGCGGGCAAAGAACAGAAGGAGACCAACATCGAATCCATGAAAATGGAGGGCTCCCGGGGCCGGCTGCGGGGTGGCTTGGGCTGGGAGTCCAGCCTCCGTCAGCGGCCCATGCCGAGGCTCACCTTCCAGGCGGGGGACCCCTACTACATCAGCAAGCGCAAGCGGGACGAGTGGCTGGCACGCTGGAAAAGGGAGGCTGAGAAGAAAGCCAAGGTCATTGCAGGAATGAATGCTGTGGAAGAAAACCAGGGGCCCGGGGAGTCTCAGAAGGTGGAGGAGGCCAGCCCTCCTGCTGTGCAGCAGCCCACTGACCCCGCATCCCCCACTGTGGCTACCACGCCTGAGCCCGTGGGGTCCGATGCTGGGGACAAGAATGCCACCAAAGCAGGCGATGACGAGCCAGAGTACGAGGACGGCCGGGGCTTTGGCATTGGGGAGCTGGTGTGGGGGAAACTGCGGGGCTTCTCCTGGTGGCCAGGCCGCATTGTGTCTTGGTGGATGACGGGCCGGAGCCGAGCAGCTGAAGGCACCCGCTGGGTCATGTGGTTCGGAGACGGCAAATTCTCAGTGGTGTGTGTTGAGAAGCTGATGCCGCTGAGCTCGTTTTGCAGTGCGTTCCACCAGGCCACGTACAACAAGCAGCCCATGTACCGCAAAGCCATCTACGAGGTCCTGCAGGTGGCCAGCAGCCGCGCGGGGAAGCTGTTCCCGGTGTGCCACGACAGCGATGAGAGTGACACTGCCAAGGCCGTGGAGGTGCAGAACAAGCCCATGATTGAATGGGCCCTGGGGGGCTTCCAGCCTTCTGGCCCTAAGGGCCTGGAGCCACCAGAAGAAGAGAAGAATCCCTACAAAGAAGTGTACACGGACATGTGGGTGGAACCTGAGGCAGCTGCCTACGCACCACCTCCACCAGCCAAAAAGCCCCGGAAGAGCACAGCGGAGAAGCCCAAGGTCAAGGAGATTATTGATGAGCGCACAAGAGAGCGGCTGGTGTACGAGGTGCGGCAGAAGTGCCGGAACATTGAGGACATCTGCATCTCCTGTGGGAGCCTCAATGTTACCCTGGAACACCCCCTCTTCGTTGGAGGAATGTGCCAAAACTGCAAGAACTGCTTTCTGGAGTGTGCGTACCAGTACGACGACGACGGCTACCAGTCCTACTGCACCATCTGCTGTGGGGGCCGTGAGGTGCTCATGTGCGGAAACAACAACTGCTGCAGGTGCTTTTGCGTGGAGTGTGTGGACCTCTTGGTGGGGCCGGGGGCTGCCCAGGCAGCCATTAAGGAAGACCCCTGGAACTGCTACATGTGCGGGCACAAGGGTACCTACGGGCTGCTGCGGCGGCGAGAGGACTGGCCCTCCCGGCTCCAGATGTTCTTCGCTAATAACCACGACCAGGAATTTGACCCTCCAAAGGTTTACCCACCTGTCCCAGCTGAGAAGAGGAAGCCCATCCGGGTGCTGTCTCTCTTTGATGGAATCGCTACAGGGCTCCTGGTGCTGAAGGACTTGGGCATTCAGGTGGACCGCTACATTGCCTCGGAGGTGTGTGAGGACTCCATCACGGTGGGCATGGTGCGGCACCAGGGGAAGATCATGTACGTCGGGGACGTCCGCAGCGTCACACAGAAGCATATCCAGGAGTGGGGCCCATTCGATCTGGTGATTGGGGGCAGTCCCTGCAATGACCTCTCCATCGTCAACCCTGCTCGCAAGGGCCTCTACGAGGGCACTGGCCGGCTCTTCTTTGAGTTCTACCGCCTCCTGCATGATGCGCGGCCCAAGGAGGGAGATGATCGCCCCTTCTTCTGGCTCTTTGAGAATGTGGTGGCCATGGGCGTTAGTGACAAGAGGGACATCTCGCGATTTCTCGAGTCCAACCCTGTGATGATTGATGCCAAAGAAGTGTCAGCTGCACACAGGGCCCGCTACTTCTGGGGTAACCTTCCCGGTATGAACAGGCCGTTGGCATCCACTGTGAATGATAAGCTGGAGCTGCAGGAGTGTCTGGAGCATGGCAGGATAGCCAAGTTCAGCAAAGTGAGGACCATTACTACGAGGTCAAACTCCATAAAGCAGGGCAAAGACCAGCATTTTCCTGTCTTCATGAATGAGAAAGAGGACATCTTATGGTGCACTGAAATGGAAAGGGTATTTGGTTTCCCAGTCCACTATACTGACGTCTCCAACATGAGCCGCTTGGCGAGGCAGAGACTGCTGGGCCGGTCATGGAGCGTGCCAGTCATCCGCCACCTCTTCGCTCCGCTGAAGGAGTATTTTGCGTGTGTGTAA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    DNMT3A_codons_NP_072046 = [DNMT3A_exons_NP_072046[i:i+3] for i in range(0, len(DNMT3A_exons_NP_072046), 3)]\n",
    "\n",
    "    #DNMT3A transcript used:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000119772;r=2:25455845-25564754;t=ENST00000321117\n",
    "    DNMT3A_exons_extra_bases_NP_072046 = 'GATGCCCGCCATGCCCTCCAGCGGCCCCGGGGACACCAGCAGCTCTGCTGCGGAGCGGGAGGAGGACCGAAAGGACGGAGAGGAGCAGGAGGAGCCGCGTGGCAAGGAGGAGCGCCAAGAGCCCAGCACCACGGCACGGAAGGTGGGGCGGCCTGGGAGGAAGCGCAAGCACCCCCCGGTGGAAAGCGGTGACACGCCAAAGGACCCTGCGGTGATCTCCAAGTCCCCATCCATGGCCCAGGACTCAGGCGCCTCAGAGCTATTACCCAATGGGGACTTGGAGAAGCGGAGTGAGCCCCAGCCAGAGGAGGGGAGCCCTGCTGGGGGGCAGAAGGGCGGGGCCCCAGCAGAGGGAGAGGGTGCAGCTGAGACCCTGCCTGAAGCCTCAAGAGCAGTGGAAAATGGCTGCTGCACCCCCAAGGAGGGCCGAGGAGCCCCTGCAGAAGCGGGCAAAGAACAGAAGGAGACCAACATCGAATCCATGAAAATGGAGGGCTCCCGGGGCCGGCTGCGGGGTGGCTTGGGCTGGGAGTCCAGCCTCCGTCAGCGGCCCATGCCGAGGCTCACCTTCCAGGCGGGGGACCCCTACTACATCAGCAAGCGCAAGCGGGACGAGTGGCTGGCACGCTGGAAAAGGGAGGCTGAGAAGAAAGCCAAGGTCATTGCAGGAATGAATGCTGTGGAAGAAAACCAGGGGCCCGGGGAGTCTCAGAAGGTGGAGGAGGCCAGCCCTCCTGCTGTGCAGCAGCCCACTGACCCCGCATCCCCCACTGTGGCTACCACGCCTGAGCCCGTGGGGTCCGATGCTGGGGACAAGAATGCCACCAAAGCAGGCGATGACGAGCCAGAGTACGAGGACGGCCGGGGCTTTGGCATTGGGGAGCTGGTGTGGGGGAAACTGCGGGGCTTCTCCTGGTGGCCAGGCCGCATTGTGTCTTGGTGGATGACGGGCCGGAGCCGAGCAGCTGAAGGCACCCGCTGGGTCATGTGGTTCGGAGACGGCAAATTCTCAGTGGTGTGTGTTGAGAAGCTGATGCCGCTGAGCTCGTTTTGCAGTGCGTTCCACCAGGCCACGTACAACAAGCAGCCCATGTACCGCAAAGCCATCTACGAGGTCCTGCAGGTGGCCAGCAGCCGCGCGGGGAAGCTGTTCCCGGTGTGCCACGACAGCGATGAGAGTGACACTGCCAAGGCCGTGGAGGTGCAGAACAAGCCCATGATTGAATGGGCCCTGGGGGGCTTCCAGCCTTCTGGCCCTAAGGGCCTGGAGCCACCAGAAGAAGAGAAGAATCCCTACAAAGAAGTGTACACGGACATGTGGGTGGAACCTGAGGCAGCTGCCTACGCACCACCTCCACCAGCCAAAAAGCCCCGGAAGAGCACAGCGGAGAAGCCCAAGGTCAAGGAGATTATTGATGAGCGCACAAGAGAGCGGCTGGTGTACGAGGTGCGGCAGAAGTGCCGGAACATTGAGGACATCTGCATCTCCTGTGGGAGCCTCAATGTTACCCTGGAACACCCCCTCTTCGTTGGAGGAATGTGCCAAAACTGCAAGAACTGCTTTCTGGAGTGTGCGTACCAGTACGACGACGACGGCTACCAGTCCTACTGCACCATCTGCTGTGGGGGCCGTGAGGTGCTCATGTGCGGAAACAACAACTGCTGCAGGTGCTTTTGCGTGGAGTGTGTGGACCTCTTGGTGGGGCCGGGGGCTGCCCAGGCAGCCATTAAGGAAGACCCCTGGAACTGCTACATGTGCGGGCACAAGGGTACCTACGGGCTGCTGCGGCGGCGAGAGGACTGGCCCTCCCGGCTCCAGATGTTCTTCGCTAATAACCACGACCAGGAATTTGACCCTCCAAAGGTTTACCCACCTGTCCCAGCTGAGAAGAGGAAGCCCATCCGGGTGCTGTCTCTCTTTGATGGAATCGCTACAGGGCTCCTGGTGCTGAAGGACTTGGGCATTCAGGTGGACCGCTACATTGCCTCGGAGGTGTGTGAGGACTCCATCACGGTGGGCATGGTGCGGCACCAGGGGAAGATCATGTACGTCGGGGACGTCCGCAGCGTCACACAGAAGCATATCCAGGAGTGGGGCCCATTCGATCTGGTGATTGGGGGCAGTCCCTGCAATGACCTCTCCATCGTCAACCCTGCTCGCAAGGGCCTCTACGAGGGCACTGGCCGGCTCTTCTTTGAGTTCTACCGCCTCCTGCATGATGCGCGGCCCAAGGAGGGAGATGATCGCCCCTTCTTCTGGCTCTTTGAGAATGTGGTGGCCATGGGCGTTAGTGACAAGAGGGACATCTCGCGATTTCTCGAGTCCAACCCTGTGATGATTGATGCCAAAGAAGTGTCAGCTGCACACAGGGCCCGCTACTTCTGGGGTAACCTTCCCGGTATGAACAGGCCGTTGGCATCCACTGTGAATGATAAGCTGGAGCTGCAGGAGTGTCTGGAGCATGGCAGGATAGCCAAGTTCAGCAAAGTGAGGACCATTACTACGAGGTCAAACTCCATAAAGCAGGGCAAAGACCAGCATTTTCCTGTCTTCATGAATGAGAAAGAGGACATCTTATGGTGCACTGAAATGGAAAGGGTATTTGGTTTCCCAGTCCACTATACTGACGTCTCCAACATGAGCCGCTTGGCGAGGCAGAGACTGCTGGGCCGGTCATGGAGCGTGCCAGTCATCCGCCACCTCTTCGCTCCGCTGAAGGAGTATTTTGCGTGTGTGTAAG'\n",
    "\n",
    "    #Create a list of the codons they called missense variants at\n",
    "    import csv\n",
    "    filename = csv_file_variants\n",
    "    with open(filename, 'r') as csvfile:\n",
    "        read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "        row_count=0\n",
    "\n",
    "        DNMT3A_missense_variant_list_NP_072046 = []\n",
    "        DNMT3A_missense_codons_NP_072046 =[]\n",
    "        DNMT3A_missense_variant_DICT_NP_072046 = {}\n",
    "\n",
    "        for row in read_reader: #each row contains aa number (e.g. 11) and variant (e.g. p.D11A)\n",
    "    #         print(row)\n",
    "            if row_count>0: \n",
    "                codon0_NP_072046 = int(row[0])-1\n",
    "                codon1_NP_072046 = int(row[0])\n",
    "                variant_NP_072046 = row[1]\n",
    "                DNMT3A_missense_codons_NP_072046.append(codon0_NP_072046) #List of codon numbers that they called variants at\n",
    "                DNMT3A_missense_variant_DICT_NP_072046[row_count]=(codon1_NP_072046, variant_NP_072046) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "            row_count=row_count+1\n",
    "            \n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    DNMT3A_missense_codons_NP_072046_DICT = {}\n",
    "    calls_DNMT3A_missense_NP_072046 = {}\n",
    "\n",
    "    for i in DNMT3A_missense_codons_NP_072046: #Iterates over the list of codon numbers\n",
    "        DNMT3A_missense_codons_NP_072046_DICT[i+1]= (DNMT3A_codons_NP_072046[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "        \n",
    "    for k,v in DNMT3A_missense_codons_NP_072046_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_DNMT3A_missense_NP_072046[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "        \n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_DNMT3A_missense_NP_072046_extra = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_missense_NP_072046.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_DNMT3A_missense_NP_072046_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_DNMT3A_missense_NP_072046_extra2 = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_missense_NP_072046_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_DNMT3A_missense_NP_072046_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "            \n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    DNMT3A_position_in_list_dict_missense_NP_072046={}\n",
    "    for k, v in DNMT3A_missense_variant_DICT_NP_072046.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_DNMT3A_missense_NP_072046[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_DNMT3A_missense_NP_072046[v[0]]\n",
    "        codon_seq=calls_DNMT3A_missense_NP_072046[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_DNMT3A_missense_NP_072046_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                DNMT3A_position_in_list_dict_missense_NP_072046[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "            \n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in DNMT3A_position_in_list_dict_missense_NP_072046.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "#     print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "                \n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    DNMT3A_bases_missense_NP_072046 = {}\n",
    "    for k, v in DNMT3A_position_in_list_dict_missense_NP_072046.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        DNMT3A_bases_missense_NP_072046[k]=(v, base_changes)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    DNMT3A_bases_missense2_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_missense_NP_072046.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        DNMT3A_bases_missense2_NP_072046[k] = (v, codon_positions)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    DNMT3A_bases_missense3_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_missense2_NP_072046.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        DNMT3A_bases_missense3_NP_072046[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    DNMT3A_bases_missense4_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_missense3_NP_072046.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_DNMT3A_NP072046:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check_DNMT3A_NP072046(i)\n",
    "                sites_trio_missense.append([adapted_trio])\n",
    "                DNMT3A_bases_missense4_NP_072046[k] = (v, sites_trio_missense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_missense = [DNMT3A_exons_extra_bases_NP_072046[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_missense.append(site_trio_missense)\n",
    "                    DNMT3A_bases_missense4_NP_072046[k] = (v, sites_trio_missense)\n",
    "                    \n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    DNMT3A_sites_missense5_NP_072046 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in DNMT3A_bases_missense4_NP_072046.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        DNMT3A_sites_missense5_NP_072046[a]=info\n",
    "        a=a+1\n",
    "        \n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    DNMT3A_sites_missense6_NP_072046={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in DNMT3A_sites_missense5_NP_072046.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "    \n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        DNMT3A_sites_missense6_NP_072046[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "        \n",
    "        \n",
    "    #Assign the mutation rates to each site...\n",
    "    DNMT3A_mutation_rates_missense_NP_072046 = {}\n",
    "    for k, v in DNMT3A_sites_missense6_NP_072046.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            DNMT3A_mutation_rates_missense_NP_072046[k] = (v, mutation_rate)\n",
    "\n",
    "    DNMT3A_mutation_rate_missense_NP_072046_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in DNMT3A_mutation_rates_missense_NP_072046.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        DNMT3A_mutation_rate_missense_NP_072046_sum.append(mutation_rate_value)\n",
    "        \n",
    "    DNMT3A_Summed_mutation_rate_missense_NP_072046_added = np.sum(DNMT3A_mutation_rate_missense_NP_072046_sum)\n",
    "\n",
    "    print('Total summed mutation rate across DNMT3A missense_NP_072046 for this study = ', DNMT3A_Summed_mutation_rate_missense_NP_072046_added)\n",
    "    \n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all #in format: column 1 = 882, column 2 = p.R882H"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TET2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_list_nonsense_variants_TET2(csv_file_variants):\n",
    "\n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base_TET2_NP01120680 = [3409, 3804, 4044, 4537]\n",
    "   \n",
    "    def intron_exon_check_TET2_NP01120680(position_in_gene_number):\n",
    "        if position_in_gene_number == 3409:\n",
    "            gene_pos = 3409\n",
    "            first_base = TET2_exons_extra_bases_NP_01120680[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 3804:\n",
    "            gene_pos = 3804\n",
    "            first_base = 'G'\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = TET2_exons_extra_bases_NP_01120680[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 4044:\n",
    "            gene_pos = 4044\n",
    "            first_base = TET2_exons_extra_bases_NP_01120680[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 4537:\n",
    "            gene_pos = 4537\n",
    "            first_base = TET2_exons_extra_bases_NP_01120680[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #TET2 transcript (NP_01120680):\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000168769;r=4:106067032-106163928;t=ENST00000380013\n",
    "    TET2_exons_NP_01120680 = 'ATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGAGCAAATTATTGAAAAAGATGAAGGTCCTTTTTATACCCATCTAGGAGCAGGTCCTAATGTGGCAGCTATTAGAGAAATCATGGAAGAAAGGTTTGGACAGAAGGGTAAAGCTATTAGGATTGAAAGAGTCATCTATACTGGTAAAGAAGGCAAAAGTTCTCAGGGATGTCCTATTGCTAAGTGGGTGGTTCGCAGAAGCAGCAGTGAAGAGAAGCTACTGTGTTTGGTGCGGGAGCGAGCTGGCCACACCTGTGAGGCTGCAGTGATTGTGATTCTCATCCTGGTGTGGGAAGGAATCCCGCTGTCTCTGGCTGACAAACTCTACTCGGAGCTTACCGAGACGCTGAGGAAATACGGCACGCTCACCAATCGCCGGTGTGCCTTGAATGAAGAGAGAACTTGCGCCTGTCAGGGGCTGGATCCAGAAACCTGTGGTGCCTCCTTCTCTTTTGGTTGTTCATGGAGCATGTACTACAATGGATGTAAGTTTGCCAGAAGCAAGATCCCAAGGAAGTTTAAGCTGCTTGGGGATGACCCAAAAGAGGAAGAGAAACTGGAGTCTCATTTGCAAAACCTGTCCACTCTTATGGCACCAACATATAAGAAACTTGCACCTGATGCATATAATAATCAGATTGAATATGAACACAGAGCACCAGAGTGCCGTCTGGGTCTGAAGGAAGGCCGTCCATTCTCAGGGGTCACTGCATGTTTGGACTTCTGTGCTCATGCCCACAGAGACTTGCACAACATGCAGAATGGCAGCACATTGGTATGCACTCTCACTAGAGAAGACAATCGAGAATTTGGAGGAAAACCTGAGGATGAGCAGCTTCACGTTCTGCCTTTATACAAAGTCTCTGACGTGGATGAGTTTGGGAGTGTGGAAGCTCAGGAGGAGAAAAAACGGAGTGGTGCCATTCAGGTACTGAGTTCTTTTCGGCGAAAAGTCAGGATGTTAGCAGAGCCAGTCAAGACTTGCCGACAAAGGAAACTAGAAGCCAAGAAAGCTGCAGCTGAAAAGCTTTCCTCCCTGGAGAACAGCTCAAATAAAAATGAAAAGGAAAAGTCAGCCCCATCACGTACAAAACAAACTGAAAACGCAAGCCAGGCTAAACAGTTGGCAGAACTTTTGCGACTTTCAGGACCAGTCATGCAGCAGTCCCAGCAGCCCCAGCCTCTACAGAAGCAGCCACCACAGCCCCAGCAGCAGCAGAGACCCCAGCAGCAGCAGCCACATCACCCTCAGACAGAGTCTGTCAACTCTTATTCTGCTTCTGGATCCACCAATCCATACATGAGACGGCCCAATCCAGTTAGTCCTTATCCAAACTCTTCACACACTTCAGATATCTATGGAAGCACCAGCCCTATGAACTTCTATTCCACCTCATCTCAAGCTGCAGGTTCATATTTGAATTCTTCTAATCCCATGAACCCTTACCCTGGGCTTTTGAATCAGAATACCCAATATCCATCATATCAATGCAATGGAAACCTATCAGTGGACAACTGCTCCCCATATCTGGGTTCCTATTCTCCCCAGTCTCAGCCGATGGATCTGTATAGGTATCCAAGCCAAGACCCTCTGTCTAAGCTCAGTCTACCACCCATCCATACACTTTACCAGCCAAGGTTTGGAAATAGCCAGAGTTTTACATCTAAATACTTAGGTTATGGAAACCAAAATATGCAGGGAGATGGTTTCAGCAGTTGTACCATTAGACCAAATGTACATCATGTAGGGAAATTGCCTCCTTATCCCACTCATGAGATGGATGGCCACTTCATGGGAGCCACCTCTAGATTACCACCCAATCTGAGCAATCCAAACATGGACTATAAAAATGGTGAACATCATTCACCTTCTCACATAATCCATAACTACAGTGCAGCTCCGGGCATGTTCAACAGCTCTCTTCATGCCCTGCATCTCCAAAACAAGGAGAATGACATGCTTTCCCACACAGCTAATGGGTTATCAAAGATGCTTCCAGCTCTTAACCATGATAGAACTGCTTGTGTCCAAGGAGGCTTACACAAATTAAGTGATGCTAATGGTCAGGAAAAGCAGCCATTGGCACTAGTCCAGGGTGTGGCTTCTGGTGCAGAGGACAACGATGAGGTCTGGTCAGACAGCGAGCAGAGCTTTCTGGATCCTGACATTGGGGGAGTGGCCGTGGCTCCAACTCATGGGTCAATTCTCATTGAGTGTGCAAAGCGTGAGCTGCATGCCACAACCCCTTTAAAGAATCCCAATAGGAATCACCCCACCAGGATCTCCCTCGTCTTTTACCAGCATAAGAGCATGAATGAGCCAAAACATGGCTTGGCTCTTTGGGAAGCCAAAATGGCTGAAAAAGCCCGTGAGAAAGAGGAAGAGTGTGAAAAGTATGGCCCAGACTATGTGCCTCAGAAATCCCATGGCAAAAAAGTGAAACGGGAGCCTGCTGAGCCACATGAAACTTCAGAGCCCACTTACCTGCGTTTCATCAAGTCTCTTGCCGAAAGGACCATGTCCGTGACCACAGACTCCACAGTAACTACATCTCCATATGCCTTCACTCGGGTCACAGGGCCTTACAACAGATATATATGA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    TET2_codons_NP_01120680 = [TET2_exons_NP_01120680[i:i+3] for i in range(0, len(TET2_exons_NP_01120680), 3)]\n",
    "\n",
    "    #TET2 transcript (transcript variant 2 (TET2-001), NM_017628, NP_001120680):\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000168769;r=4:106067032-106163928;t=ENST00000380013\n",
    "    TET2_exons_extra_bases_NP_01120680 = 'GATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGAGCAAATTATTGAAAAAGATGAAGGTCCTTTTTATACCCATCTAGGAGCAGGTCCTAATGTGGCAGCTATTAGAGAAATCATGGAAGAAAGGTTTGGACAGAAGGGTAAAGCTATTAGGATTGAAAGAGTCATCTATACTGGTAAAGAAGGCAAAAGTTCTCAGGGATGTCCTATTGCTAAGTGGGTGGTTCGCAGAAGCAGCAGTGAAGAGAAGCTACTGTGTTTGGTGCGGGAGCGAGCTGGCCACACCTGTGAGGCTGCAGTGATTGTGATTCTCATCCTGGTGTGGGAAGGAATCCCGCTGTCTCTGGCTGACAAACTCTACTCGGAGCTTACCGAGACGCTGAGGAAATACGGCACGCTCACCAATCGCCGGTGTGCCTTGAATGAAGAGAGAACTTGCGCCTGTCAGGGGCTGGATCCAGAAACCTGTGGTGCCTCCTTCTCTTTTGGTTGTTCATGGAGCATGTACTACAATGGATGTAAGTTTGCCAGAAGCAAGATCCCAAGGAAGTTTAAGCTGCTTGGGGATGACCCAAAAGAGGAAGAGAAACTGGAGTCTCATTTGCAAAACCTGTCCACTCTTATGGCACCAACATATAAGAAACTTGCACCTGATGCATATAATAATCAGATTGAATATGAACACAGAGCACCAGAGTGCCGTCTGGGTCTGAAGGAAGGCCGTCCATTCTCAGGGGTCACTGCATGTTTGGACTTCTGTGCTCATGCCCACAGAGACTTGCACAACATGCAGAATGGCAGCACATTGGTATGCACTCTCACTAGAGAAGACAATCGAGAATTTGGAGGAAAACCTGAGGATGAGCAGCTTCACGTTCTGCCTTTATACAAAGTCTCTGACGTGGATGAGTTTGGGAGTGTGGAAGCTCAGGAGGAGAAAAAACGGAGTGGTGCCATTCAGGTACTGAGTTCTTTTCGGCGAAAAGTCAGGATGTTAGCAGAGCCAGTCAAGACTTGCCGACAAAGGAAACTAGAAGCCAAGAAAGCTGCAGCTGAAAAGCTTTCCTCCCTGGAGAACAGCTCAAATAAAAATGAAAAGGAAAAGTCAGCCCCATCACGTACAAAACAAACTGAAAACGCAAGCCAGGCTAAACAGTTGGCAGAACTTTTGCGACTTTCAGGACCAGTCATGCAGCAGTCCCAGCAGCCCCAGCCTCTACAGAAGCAGCCACCACAGCCCCAGCAGCAGCAGAGACCCCAGCAGCAGCAGCCACATCACCCTCAGACAGAGTCTGTCAACTCTTATTCTGCTTCTGGATCCACCAATCCATACATGAGACGGCCCAATCCAGTTAGTCCTTATCCAAACTCTTCACACACTTCAGATATCTATGGAAGCACCAGCCCTATGAACTTCTATTCCACCTCATCTCAAGCTGCAGGTTCATATTTGAATTCTTCTAATCCCATGAACCCTTACCCTGGGCTTTTGAATCAGAATACCCAATATCCATCATATCAATGCAATGGAAACCTATCAGTGGACAACTGCTCCCCATATCTGGGTTCCTATTCTCCCCAGTCTCAGCCGATGGATCTGTATAGGTATCCAAGCCAAGACCCTCTGTCTAAGCTCAGTCTACCACCCATCCATACACTTTACCAGCCAAGGTTTGGAAATAGCCAGAGTTTTACATCTAAATACTTAGGTTATGGAAACCAAAATATGCAGGGAGATGGTTTCAGCAGTTGTACCATTAGACCAAATGTACATCATGTAGGGAAATTGCCTCCTTATCCCACTCATGAGATGGATGGCCACTTCATGGGAGCCACCTCTAGATTACCACCCAATCTGAGCAATCCAAACATGGACTATAAAAATGGTGAACATCATTCACCTTCTCACATAATCCATAACTACAGTGCAGCTCCGGGCATGTTCAACAGCTCTCTTCATGCCCTGCATCTCCAAAACAAGGAGAATGACATGCTTTCCCACACAGCTAATGGGTTATCAAAGATGCTTCCAGCTCTTAACCATGATAGAACTGCTTGTGTCCAAGGAGGCTTACACAAATTAAGTGATGCTAATGGTCAGGAAAAGCAGCCATTGGCACTAGTCCAGGGTGTGGCTTCTGGTGCAGAGGACAACGATGAGGTCTGGTCAGACAGCGAGCAGAGCTTTCTGGATCCTGACATTGGGGGAGTGGCCGTGGCTCCAACTCATGGGTCAATTCTCATTGAGTGTGCAAAGCGTGAGCTGCATGCCACAACCCCTTTAAAGAATCCCAATAGGAATCACCCCACCAGGATCTCCCTCGTCTTTTACCAGCATAAGAGCATGAATGAGCCAAAACATGGCTTGGCTCTTTGGGAAGCCAAAATGGCTGAAAAAGCCCGTGAGAAAGAGGAAGAGTGTGAAAAGTATGGCCCAGACTATGTGCCTCAGAAATCCCATGGCAAAAAAGTGAAACGGGAGCCTGCTGAGCCACATGAAACTTCAGAGCCCACTTACCTGCGTTTCATCAAGTCTCTTGCCGAAAGGACCATGTCCGTGACCACAGACTCCACAGTAACTACATCTCCATATGCCTTCACTCGGGTCACAGGGCCTTACAACAGATATATATGAT'\n",
    "\n",
    "    #Create a list of the codons they called nonsense variants at\n",
    "    import csv\n",
    "    filename = csv_file_variants\n",
    "    with open(filename, 'r') as csvfile:\n",
    "        read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "        row_count=0\n",
    "\n",
    "        TET2_nonsense_variant_list_NP_01120680 = []\n",
    "        TET2_nonsense_codons_NP_01120680 =[]\n",
    "        TET2_nonsense_variant_DICT_NP_01120680 = {}\n",
    "\n",
    "        for row in read_reader: #each row contains aa number (e.g. 11) and variant (e.g. p.D11A)\n",
    "    #         print(row)\n",
    "            if row_count>0: \n",
    "                codon0_NP_01120680 = int(row[0])-1\n",
    "                codon1_NP_01120680 = int(row[0])\n",
    "                variant_NP_01120680 = row[1]\n",
    "                TET2_nonsense_codons_NP_01120680.append(codon0_NP_01120680) #List of codon numbers that they called variants at\n",
    "                TET2_nonsense_variant_DICT_NP_01120680[row_count]=(codon1_NP_01120680, variant_NP_01120680) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "            row_count=row_count+1\n",
    "            \n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    TET2_nonsense_codons_NP_01120680_DICT = {}\n",
    "    calls_TET2_nonsense_NP_01120680 = {}\n",
    "\n",
    "    for i in TET2_nonsense_codons_NP_01120680: #Iterates over the list of codon numbers\n",
    "        TET2_nonsense_codons_NP_01120680_DICT[i+1]= (TET2_codons_NP_01120680[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "        \n",
    "    for k,v in TET2_nonsense_codons_NP_01120680_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_TET2_nonsense_NP_01120680[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "        \n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_TET2_nonsense_NP_01120680_extra = {}\n",
    "\n",
    "    for k, v in calls_TET2_nonsense_NP_01120680.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_TET2_nonsense_NP_01120680_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_TET2_nonsense_NP_01120680_extra2 = {}\n",
    "\n",
    "    for k, v in calls_TET2_nonsense_NP_01120680_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_TET2_nonsense_NP_01120680_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "            \n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    TET2_position_in_list_dict_nonsense_NP_01120680={}\n",
    "    for k, v in TET2_nonsense_variant_DICT_NP_01120680.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_TET2_nonsense_NP_01120680[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_TET2_nonsense_NP_01120680[v[0]]\n",
    "        codon_seq=calls_TET2_nonsense_NP_01120680[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_TET2_nonsense_NP_01120680_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                TET2_position_in_list_dict_nonsense_NP_01120680[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "            \n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in TET2_position_in_list_dict_nonsense_NP_01120680.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "#     print('positions at which nonsense variants called = ', aa_positions_variants_called)\n",
    "                \n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TET2_bases_nonsense_NP_01120680 = {}\n",
    "    for k, v in TET2_position_in_list_dict_nonsense_NP_01120680.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TET2_bases_nonsense_NP_01120680[k]=(v, base_changes)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TET2_bases_nonsense2_NP_01120680 = {}\n",
    "\n",
    "    for k, v in TET2_bases_nonsense_NP_01120680.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TET2_bases_nonsense2_NP_01120680[k] = (v, codon_positions)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TET2_bases_nonsense3_NP_01120680 = {}\n",
    "\n",
    "    for k, v in TET2_bases_nonsense2_NP_01120680.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TET2_bases_nonsense3_NP_01120680[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TET2_bases_nonsense4_NP_01120680 = {}\n",
    "\n",
    "    for k, v in TET2_bases_nonsense3_NP_01120680.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_nonsense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_TET2_NP01120680:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check_TET2_NP01120680(i)\n",
    "                sites_trio_nonsense.append([adapted_trio])\n",
    "                TET2_bases_nonsense4_NP_01120680[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_nonsense = [TET2_exons_extra_bases_NP_01120680[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_nonsense.append(site_trio_nonsense)\n",
    "                    TET2_bases_nonsense4_NP_01120680[k] = (v, sites_trio_nonsense)\n",
    "                    \n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TET2_sites_nonsense5_NP_01120680 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TET2_bases_nonsense4_NP_01120680.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        TET2_sites_nonsense5_NP_01120680[a]=info\n",
    "        a=a+1\n",
    "#     print(TET2_sites_nonsense5_NP_01120680)\n",
    "        \n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TET2_sites_nonsense6_NP_01120680={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TET2_sites_nonsense5_NP_01120680.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "    \n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TET2_sites_nonsense6_NP_01120680[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "        \n",
    "        \n",
    "    #Assign the mutation rates to each site...\n",
    "    TET2_mutation_rates_nonsense_NP_01120680 = {}\n",
    "    for k, v in TET2_sites_nonsense6_NP_01120680.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            TET2_mutation_rates_nonsense_NP_01120680[k] = (v, mutation_rate)\n",
    "\n",
    "    TET2_mutation_rate_nonsense_NP_01120680_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TET2_mutation_rates_nonsense_NP_01120680.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        TET2_mutation_rate_nonsense_NP_01120680_sum.append(mutation_rate_value)\n",
    "        \n",
    "    TET2_Summed_mutation_rate_nonsense_NP_01120680_added = np.sum(TET2_mutation_rate_nonsense_NP_01120680_sum)\n",
    "\n",
    "    print('Total summed mutation rate across TET2 nonsense_NP_01120680 for this study = ', TET2_Summed_mutation_rate_nonsense_NP_01120680_added)\n",
    "    \n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all #in format: column 1 = 882, column 2 = p.R882* #2002 aa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_list_missense_variants_TET2(csv_file_variants):\n",
    "    \n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base_TET2_NP01120680 = [3409, 3804, 4044, 4537]\n",
    "   \n",
    "    def intron_exon_check_TET2_NP01120680(position_in_gene_number):\n",
    "        if position_in_gene_number == 3409:\n",
    "            gene_pos = 3409\n",
    "            first_base = TET2_exons_extra_bases_NP_01120680[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 3804:\n",
    "            gene_pos = 3804\n",
    "            first_base = 'G'\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = TET2_exons_extra_bases_NP_01120680[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 4044:\n",
    "            gene_pos = 4044\n",
    "            first_base = TET2_exons_extra_bases_NP_01120680[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 4537:\n",
    "            gene_pos = 4537\n",
    "            first_base = TET2_exons_extra_bases_NP_01120680[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #TET2 transcript (NP_01120680):\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000168769;r=4:106067032-106163928;t=ENST00000380013\n",
    "    TET2_exons_NP_01120680 = 'ATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGAGCAAATTATTGAAAAAGATGAAGGTCCTTTTTATACCCATCTAGGAGCAGGTCCTAATGTGGCAGCTATTAGAGAAATCATGGAAGAAAGGTTTGGACAGAAGGGTAAAGCTATTAGGATTGAAAGAGTCATCTATACTGGTAAAGAAGGCAAAAGTTCTCAGGGATGTCCTATTGCTAAGTGGGTGGTTCGCAGAAGCAGCAGTGAAGAGAAGCTACTGTGTTTGGTGCGGGAGCGAGCTGGCCACACCTGTGAGGCTGCAGTGATTGTGATTCTCATCCTGGTGTGGGAAGGAATCCCGCTGTCTCTGGCTGACAAACTCTACTCGGAGCTTACCGAGACGCTGAGGAAATACGGCACGCTCACCAATCGCCGGTGTGCCTTGAATGAAGAGAGAACTTGCGCCTGTCAGGGGCTGGATCCAGAAACCTGTGGTGCCTCCTTCTCTTTTGGTTGTTCATGGAGCATGTACTACAATGGATGTAAGTTTGCCAGAAGCAAGATCCCAAGGAAGTTTAAGCTGCTTGGGGATGACCCAAAAGAGGAAGAGAAACTGGAGTCTCATTTGCAAAACCTGTCCACTCTTATGGCACCAACATATAAGAAACTTGCACCTGATGCATATAATAATCAGATTGAATATGAACACAGAGCACCAGAGTGCCGTCTGGGTCTGAAGGAAGGCCGTCCATTCTCAGGGGTCACTGCATGTTTGGACTTCTGTGCTCATGCCCACAGAGACTTGCACAACATGCAGAATGGCAGCACATTGGTATGCACTCTCACTAGAGAAGACAATCGAGAATTTGGAGGAAAACCTGAGGATGAGCAGCTTCACGTTCTGCCTTTATACAAAGTCTCTGACGTGGATGAGTTTGGGAGTGTGGAAGCTCAGGAGGAGAAAAAACGGAGTGGTGCCATTCAGGTACTGAGTTCTTTTCGGCGAAAAGTCAGGATGTTAGCAGAGCCAGTCAAGACTTGCCGACAAAGGAAACTAGAAGCCAAGAAAGCTGCAGCTGAAAAGCTTTCCTCCCTGGAGAACAGCTCAAATAAAAATGAAAAGGAAAAGTCAGCCCCATCACGTACAAAACAAACTGAAAACGCAAGCCAGGCTAAACAGTTGGCAGAACTTTTGCGACTTTCAGGACCAGTCATGCAGCAGTCCCAGCAGCCCCAGCCTCTACAGAAGCAGCCACCACAGCCCCAGCAGCAGCAGAGACCCCAGCAGCAGCAGCCACATCACCCTCAGACAGAGTCTGTCAACTCTTATTCTGCTTCTGGATCCACCAATCCATACATGAGACGGCCCAATCCAGTTAGTCCTTATCCAAACTCTTCACACACTTCAGATATCTATGGAAGCACCAGCCCTATGAACTTCTATTCCACCTCATCTCAAGCTGCAGGTTCATATTTGAATTCTTCTAATCCCATGAACCCTTACCCTGGGCTTTTGAATCAGAATACCCAATATCCATCATATCAATGCAATGGAAACCTATCAGTGGACAACTGCTCCCCATATCTGGGTTCCTATTCTCCCCAGTCTCAGCCGATGGATCTGTATAGGTATCCAAGCCAAGACCCTCTGTCTAAGCTCAGTCTACCACCCATCCATACACTTTACCAGCCAAGGTTTGGAAATAGCCAGAGTTTTACATCTAAATACTTAGGTTATGGAAACCAAAATATGCAGGGAGATGGTTTCAGCAGTTGTACCATTAGACCAAATGTACATCATGTAGGGAAATTGCCTCCTTATCCCACTCATGAGATGGATGGCCACTTCATGGGAGCCACCTCTAGATTACCACCCAATCTGAGCAATCCAAACATGGACTATAAAAATGGTGAACATCATTCACCTTCTCACATAATCCATAACTACAGTGCAGCTCCGGGCATGTTCAACAGCTCTCTTCATGCCCTGCATCTCCAAAACAAGGAGAATGACATGCTTTCCCACACAGCTAATGGGTTATCAAAGATGCTTCCAGCTCTTAACCATGATAGAACTGCTTGTGTCCAAGGAGGCTTACACAAATTAAGTGATGCTAATGGTCAGGAAAAGCAGCCATTGGCACTAGTCCAGGGTGTGGCTTCTGGTGCAGAGGACAACGATGAGGTCTGGTCAGACAGCGAGCAGAGCTTTCTGGATCCTGACATTGGGGGAGTGGCCGTGGCTCCAACTCATGGGTCAATTCTCATTGAGTGTGCAAAGCGTGAGCTGCATGCCACAACCCCTTTAAAGAATCCCAATAGGAATCACCCCACCAGGATCTCCCTCGTCTTTTACCAGCATAAGAGCATGAATGAGCCAAAACATGGCTTGGCTCTTTGGGAAGCCAAAATGGCTGAAAAAGCCCGTGAGAAAGAGGAAGAGTGTGAAAAGTATGGCCCAGACTATGTGCCTCAGAAATCCCATGGCAAAAAAGTGAAACGGGAGCCTGCTGAGCCACATGAAACTTCAGAGCCCACTTACCTGCGTTTCATCAAGTCTCTTGCCGAAAGGACCATGTCCGTGACCACAGACTCCACAGTAACTACATCTCCATATGCCTTCACTCGGGTCACAGGGCCTTACAACAGATATATATGA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    TET2_codons_NP_01120680 = [TET2_exons_NP_01120680[i:i+3] for i in range(0, len(TET2_exons_NP_01120680), 3)]\n",
    "\n",
    "    #TET2 transcript (transcript variant 2 (TET2-001), NM_017628, NP_001120680):\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000168769;r=4:106067032-106163928;t=ENST00000380013\n",
    "    TET2_exons_extra_bases_NP_01120680 = 'GATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGAGCAAATTATTGAAAAAGATGAAGGTCCTTTTTATACCCATCTAGGAGCAGGTCCTAATGTGGCAGCTATTAGAGAAATCATGGAAGAAAGGTTTGGACAGAAGGGTAAAGCTATTAGGATTGAAAGAGTCATCTATACTGGTAAAGAAGGCAAAAGTTCTCAGGGATGTCCTATTGCTAAGTGGGTGGTTCGCAGAAGCAGCAGTGAAGAGAAGCTACTGTGTTTGGTGCGGGAGCGAGCTGGCCACACCTGTGAGGCTGCAGTGATTGTGATTCTCATCCTGGTGTGGGAAGGAATCCCGCTGTCTCTGGCTGACAAACTCTACTCGGAGCTTACCGAGACGCTGAGGAAATACGGCACGCTCACCAATCGCCGGTGTGCCTTGAATGAAGAGAGAACTTGCGCCTGTCAGGGGCTGGATCCAGAAACCTGTGGTGCCTCCTTCTCTTTTGGTTGTTCATGGAGCATGTACTACAATGGATGTAAGTTTGCCAGAAGCAAGATCCCAAGGAAGTTTAAGCTGCTTGGGGATGACCCAAAAGAGGAAGAGAAACTGGAGTCTCATTTGCAAAACCTGTCCACTCTTATGGCACCAACATATAAGAAACTTGCACCTGATGCATATAATAATCAGATTGAATATGAACACAGAGCACCAGAGTGCCGTCTGGGTCTGAAGGAAGGCCGTCCATTCTCAGGGGTCACTGCATGTTTGGACTTCTGTGCTCATGCCCACAGAGACTTGCACAACATGCAGAATGGCAGCACATTGGTATGCACTCTCACTAGAGAAGACAATCGAGAATTTGGAGGAAAACCTGAGGATGAGCAGCTTCACGTTCTGCCTTTATACAAAGTCTCTGACGTGGATGAGTTTGGGAGTGTGGAAGCTCAGGAGGAGAAAAAACGGAGTGGTGCCATTCAGGTACTGAGTTCTTTTCGGCGAAAAGTCAGGATGTTAGCAGAGCCAGTCAAGACTTGCCGACAAAGGAAACTAGAAGCCAAGAAAGCTGCAGCTGAAAAGCTTTCCTCCCTGGAGAACAGCTCAAATAAAAATGAAAAGGAAAAGTCAGCCCCATCACGTACAAAACAAACTGAAAACGCAAGCCAGGCTAAACAGTTGGCAGAACTTTTGCGACTTTCAGGACCAGTCATGCAGCAGTCCCAGCAGCCCCAGCCTCTACAGAAGCAGCCACCACAGCCCCAGCAGCAGCAGAGACCCCAGCAGCAGCAGCCACATCACCCTCAGACAGAGTCTGTCAACTCTTATTCTGCTTCTGGATCCACCAATCCATACATGAGACGGCCCAATCCAGTTAGTCCTTATCCAAACTCTTCACACACTTCAGATATCTATGGAAGCACCAGCCCTATGAACTTCTATTCCACCTCATCTCAAGCTGCAGGTTCATATTTGAATTCTTCTAATCCCATGAACCCTTACCCTGGGCTTTTGAATCAGAATACCCAATATCCATCATATCAATGCAATGGAAACCTATCAGTGGACAACTGCTCCCCATATCTGGGTTCCTATTCTCCCCAGTCTCAGCCGATGGATCTGTATAGGTATCCAAGCCAAGACCCTCTGTCTAAGCTCAGTCTACCACCCATCCATACACTTTACCAGCCAAGGTTTGGAAATAGCCAGAGTTTTACATCTAAATACTTAGGTTATGGAAACCAAAATATGCAGGGAGATGGTTTCAGCAGTTGTACCATTAGACCAAATGTACATCATGTAGGGAAATTGCCTCCTTATCCCACTCATGAGATGGATGGCCACTTCATGGGAGCCACCTCTAGATTACCACCCAATCTGAGCAATCCAAACATGGACTATAAAAATGGTGAACATCATTCACCTTCTCACATAATCCATAACTACAGTGCAGCTCCGGGCATGTTCAACAGCTCTCTTCATGCCCTGCATCTCCAAAACAAGGAGAATGACATGCTTTCCCACACAGCTAATGGGTTATCAAAGATGCTTCCAGCTCTTAACCATGATAGAACTGCTTGTGTCCAAGGAGGCTTACACAAATTAAGTGATGCTAATGGTCAGGAAAAGCAGCCATTGGCACTAGTCCAGGGTGTGGCTTCTGGTGCAGAGGACAACGATGAGGTCTGGTCAGACAGCGAGCAGAGCTTTCTGGATCCTGACATTGGGGGAGTGGCCGTGGCTCCAACTCATGGGTCAATTCTCATTGAGTGTGCAAAGCGTGAGCTGCATGCCACAACCCCTTTAAAGAATCCCAATAGGAATCACCCCACCAGGATCTCCCTCGTCTTTTACCAGCATAAGAGCATGAATGAGCCAAAACATGGCTTGGCTCTTTGGGAAGCCAAAATGGCTGAAAAAGCCCGTGAGAAAGAGGAAGAGTGTGAAAAGTATGGCCCAGACTATGTGCCTCAGAAATCCCATGGCAAAAAAGTGAAACGGGAGCCTGCTGAGCCACATGAAACTTCAGAGCCCACTTACCTGCGTTTCATCAAGTCTCTTGCCGAAAGGACCATGTCCGTGACCACAGACTCCACAGTAACTACATCTCCATATGCCTTCACTCGGGTCACAGGGCCTTACAACAGATATATATGAT'\n",
    "\n",
    "\n",
    "    #Create a list of the codons they called missense variants at\n",
    "    import csv\n",
    "    filename = csv_file_variants\n",
    "    with open(filename, 'r') as csvfile:\n",
    "        read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "        row_count=0\n",
    "\n",
    "        TET2_missense_variant_list_NP_01120680 = []\n",
    "        TET2_missense_codons_NP_01120680 =[]\n",
    "        TET2_missense_variant_DICT_NP_01120680 = {}\n",
    "\n",
    "        for row in read_reader: #each row contains aa number (e.g. 11) and variant (e.g. p.D11A)\n",
    "    #         print(row)\n",
    "            if row_count>0: \n",
    "                codon0_NP_01120680 = int(row[0])-1\n",
    "                codon1_NP_01120680 = int(row[0])\n",
    "                variant_NP_01120680 = row[1]\n",
    "                TET2_missense_codons_NP_01120680.append(codon0_NP_01120680) #List of codon numbers that they called variants at\n",
    "                TET2_missense_variant_DICT_NP_01120680[row_count]=(codon1_NP_01120680, variant_NP_01120680) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "            row_count=row_count+1\n",
    "            \n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    TET2_missense_codons_NP_01120680_DICT = {}\n",
    "    calls_TET2_missense_NP_01120680 = {}\n",
    "\n",
    "    for i in TET2_missense_codons_NP_01120680: #Iterates over the list of codon numbers\n",
    "        TET2_missense_codons_NP_01120680_DICT[i+1]= (TET2_codons_NP_01120680[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "        \n",
    "    for k,v in TET2_missense_codons_NP_01120680_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_TET2_missense_NP_01120680[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "        \n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_TET2_missense_NP_01120680_extra = {}\n",
    "\n",
    "    for k, v in calls_TET2_missense_NP_01120680.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_TET2_missense_NP_01120680_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_TET2_missense_NP_01120680_extra2 = {}\n",
    "\n",
    "    for k, v in calls_TET2_missense_NP_01120680_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_TET2_missense_NP_01120680_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "            \n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    TET2_position_in_list_dict_missense_NP_01120680={}\n",
    "    for k, v in TET2_missense_variant_DICT_NP_01120680.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_TET2_missense_NP_01120680[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_TET2_missense_NP_01120680[v[0]]\n",
    "        codon_seq=calls_TET2_missense_NP_01120680[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_TET2_missense_NP_01120680_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                TET2_position_in_list_dict_missense_NP_01120680[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "            \n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in TET2_position_in_list_dict_missense_NP_01120680.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "#     print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "                \n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TET2_bases_missense_NP_01120680 = {}\n",
    "    for k, v in TET2_position_in_list_dict_missense_NP_01120680.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TET2_bases_missense_NP_01120680[k]=(v, base_changes)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TET2_bases_missense2_NP_01120680 = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense_NP_01120680.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TET2_bases_missense2_NP_01120680[k] = (v, codon_positions)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TET2_bases_missense3_NP_01120680 = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense2_NP_01120680.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TET2_bases_missense3_NP_01120680[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TET2_bases_missense4_NP_01120680 = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense3_NP_01120680.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_TET2_NP01120680:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check_TET2_NP01120680(i)\n",
    "                sites_trio_missense.append([adapted_trio])\n",
    "                TET2_bases_missense4_NP_01120680[k] = (v, sites_trio_missense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_missense = [TET2_exons_extra_bases_NP_01120680[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_missense.append(site_trio_missense)\n",
    "                    TET2_bases_missense4_NP_01120680[k] = (v, sites_trio_missense)\n",
    "                    \n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TET2_sites_missense5_NP_01120680 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TET2_bases_missense4_NP_01120680.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        TET2_sites_missense5_NP_01120680[a]=info\n",
    "        a=a+1\n",
    "        \n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TET2_sites_missense6_NP_01120680={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TET2_sites_missense5_NP_01120680.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "    \n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TET2_sites_missense6_NP_01120680[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "        \n",
    "        \n",
    "    #Assign the mutation rates to each site...\n",
    "    TET2_mutation_rates_missense_NP_01120680 = {}\n",
    "    for k, v in TET2_sites_missense6_NP_01120680.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            TET2_mutation_rates_missense_NP_01120680[k] = (v, mutation_rate)\n",
    "\n",
    "    TET2_mutation_rate_missense_NP_01120680_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TET2_mutation_rates_missense_NP_01120680.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        TET2_mutation_rate_missense_NP_01120680_sum.append(mutation_rate_value)\n",
    "        \n",
    "    TET2_Summed_mutation_rate_missense_NP_01120680_added = np.sum(TET2_mutation_rate_missense_NP_01120680_sum)\n",
    "\n",
    "    print('Total summed mutation rate across TET2 missense_NP_01120680 for this study = ', TET2_Summed_mutation_rate_missense_NP_01120680_added)\n",
    "    \n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all #in format: column 1 = 882, column 2 = p.R882H  #2002 aa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_list_nonsense_variants_TET2_shorter(csv_file_variants):\n",
    "\n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    #TET2 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Summary?db=core;g=ENSG00000168769;r=4:106067032-106200973;t=ENST00000305737\n",
    "    TET2_exons_NP_060098 = 'ATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGGTAAGTGCCAGAAATGTACTGAGACACATGGCGTTTATCCAGAATTAGCAAATTTATCTTCAGATATGGGATTTTCCTTCTTTTTTTAA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    TET2_codons_NP_060098 = [TET2_exons_NP_060098[i:i+3] for i in range(0, len(TET2_exons_NP_060098), 3)]\n",
    "\n",
    "    #TET2 transcript 1 extra base at each end:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Summary?db=core;g=ENSG00000168769;r=4:106067032-106200973;t=ENST00000305737\n",
    "    TET2_exons_extra_bases_NP_060098 = 'GATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGGTAAGTGCCAGAAATGTACTGAGACACATGGCGTTTATCCAGAATTAGCAAATTTATCTTCAGATATGGGATTTTCCTTCTTTTTTTAAA'\n",
    "\n",
    "    #Create a list of the codons they called nonsense variants at\n",
    "    import csv\n",
    "    filename = csv_file_variants\n",
    "    with open(filename, 'r') as csvfile:\n",
    "        read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "        row_count=0\n",
    "\n",
    "        TET2_nonsense_variant_list_NP_060098 = []\n",
    "        TET2_nonsense_codons_NP_060098 =[]\n",
    "        TET2_nonsense_variant_DICT_NP_060098 = {}\n",
    "\n",
    "        for row in read_reader: #each row contains aa number (e.g. 11) and variant (e.g. p.D11A)\n",
    "    #         print(row)\n",
    "            if row_count>0: \n",
    "                codon0_NP_060098 = int(row[0])-1\n",
    "                codon1_NP_060098 = int(row[0])\n",
    "                variant_NP_060098 = row[1]\n",
    "                if codon1_NP_060098 < 1167:\n",
    "                    TET2_nonsense_codons_NP_060098.append(codon0_NP_060098) #List of codon numbers that they called variants at\n",
    "                    TET2_nonsense_variant_DICT_NP_060098[row_count]=(codon1_NP_060098, variant_NP_060098) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "            row_count=row_count+1\n",
    "            \n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    TET2_nonsense_codons_NP_060098_DICT = {}\n",
    "    calls_TET2_nonsense_NP_060098 = {}\n",
    "\n",
    "    for i in TET2_nonsense_codons_NP_060098: #Iterates over the list of codon numbers\n",
    "        TET2_nonsense_codons_NP_060098_DICT[i+1]= (TET2_codons_NP_060098[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "        \n",
    "    for k,v in TET2_nonsense_codons_NP_060098_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_TET2_nonsense_NP_060098[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "        \n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_TET2_nonsense_NP_060098_extra = {}\n",
    "\n",
    "    for k, v in calls_TET2_nonsense_NP_060098.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_TET2_nonsense_NP_060098_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_TET2_nonsense_NP_060098_extra2 = {}\n",
    "\n",
    "    for k, v in calls_TET2_nonsense_NP_060098_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_TET2_nonsense_NP_060098_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "            \n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    TET2_position_in_list_dict_nonsense_NP_060098={}\n",
    "    for k, v in TET2_nonsense_variant_DICT_NP_060098.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_TET2_nonsense_NP_060098[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_TET2_nonsense_NP_060098[v[0]]\n",
    "        codon_seq=calls_TET2_nonsense_NP_060098[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_TET2_nonsense_NP_060098_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                TET2_position_in_list_dict_nonsense_NP_060098[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "            \n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in TET2_position_in_list_dict_nonsense_NP_060098.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "#     print('positions at which nonsense variants called = ', aa_positions_variants_called)\n",
    "                \n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TET2_bases_nonsense_NP_060098 = {}\n",
    "    for k, v in TET2_position_in_list_dict_nonsense_NP_060098.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TET2_bases_nonsense_NP_060098[k]=(v, base_changes)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TET2_bases_nonsense2_NP_060098 = {}\n",
    "\n",
    "    for k, v in TET2_bases_nonsense_NP_060098.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TET2_bases_nonsense2_NP_060098[k] = (v, codon_positions)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TET2_bases_nonsense3_NP_060098 = {}\n",
    "\n",
    "    for k, v in TET2_bases_nonsense2_NP_060098.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TET2_bases_nonsense3_NP_060098[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TET2_bases_nonsense4_NP_060098 = {}\n",
    "\n",
    "    for k, v in TET2_bases_nonsense3_NP_060098.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_nonsense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "            gene_pos = i\n",
    "            site_trio_nonsense = [TET2_exons_extra_bases_NP_060098[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "            sites_trio_nonsense.append(site_trio_nonsense)\n",
    "            TET2_bases_nonsense4_NP_060098[k] = (v, sites_trio_nonsense)\n",
    "                    \n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TET2_sites_nonsense5_NP_060098 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TET2_bases_nonsense4_NP_060098.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        TET2_sites_nonsense5_NP_060098[a]=info\n",
    "        a=a+1\n",
    "#     print(TET2_sites_nonsense5_NP_060098)\n",
    "        \n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TET2_sites_nonsense6_NP_060098={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TET2_sites_nonsense5_NP_060098.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "    \n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TET2_sites_nonsense6_NP_060098[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "        \n",
    "        \n",
    "    #Assign the mutation rates to each site...\n",
    "    TET2_mutation_rates_nonsense_NP_060098 = {}\n",
    "    for k, v in TET2_sites_nonsense6_NP_060098.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            TET2_mutation_rates_nonsense_NP_060098[k] = (v, mutation_rate)\n",
    "\n",
    "    TET2_mutation_rate_nonsense_NP_060098_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TET2_mutation_rates_nonsense_NP_060098.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        TET2_mutation_rate_nonsense_NP_060098_sum.append(mutation_rate_value)\n",
    "        \n",
    "    TET2_Summed_mutation_rate_nonsense_NP_060098_added = np.sum(TET2_mutation_rate_nonsense_NP_060098_sum)\n",
    "\n",
    "    print('Total summed mutation rate across TET2 nonsense_NP_060098 for this study = ', TET2_Summed_mutation_rate_nonsense_NP_060098_added)\n",
    "    \n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all #in format: column 1 = 882, column 2 = p.R882*  #1165 aa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_list_missense_variants_TET2_shorter(csv_file_variants):\n",
    "    \n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    #TET2 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Summary?db=core;g=ENSG00000168769;r=4:106067032-106200973;t=ENST00000305737\n",
    "    TET2_exons_NP_060098 = 'ATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGGTAAGTGCCAGAAATGTACTGAGACACATGGCGTTTATCCAGAATTAGCAAATTTATCTTCAGATATGGGATTTTCCTTCTTTTTTTAA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    TET2_codons_NP_060098 = [TET2_exons_NP_060098[i:i+3] for i in range(0, len(TET2_exons_NP_060098), 3)]\n",
    "\n",
    "    #TET2 transcript 1 extra base at each end:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Summary?db=core;g=ENSG00000168769;r=4:106067032-106200973;t=ENST00000305737\n",
    "    TET2_exons_extra_bases_NP_060098 = 'GATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGGTAAGTGCCAGAAATGTACTGAGACACATGGCGTTTATCCAGAATTAGCAAATTTATCTTCAGATATGGGATTTTCCTTCTTTTTTTAAA'\n",
    "\n",
    "    #Create a list of the codons they called missense variants at\n",
    "    import csv\n",
    "    filename = csv_file_variants\n",
    "    with open(filename, 'r') as csvfile:\n",
    "        read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "        row_count=0\n",
    "\n",
    "        TET2_missense_variant_list_NP_060098 = []\n",
    "        TET2_missense_codons_NP_060098 =[]\n",
    "        TET2_missense_variant_DICT_NP_060098 = {}\n",
    "\n",
    "        for row in read_reader: #each row contains aa number (e.g. 11) and variant (e.g. p.D11A)\n",
    "    #         print(row)\n",
    "            if row_count>0: \n",
    "                codon0_NP_060098 = int(row[0])-1\n",
    "                codon1_NP_060098 = int(row[0])\n",
    "                variant_NP_060098 = row[1]\n",
    "                if codon1_NP_060098 < 1167:\n",
    "                    TET2_missense_codons_NP_060098.append(codon0_NP_060098) #List of codon numbers that they called variants at\n",
    "                    TET2_missense_variant_DICT_NP_060098[row_count]=(codon1_NP_060098, variant_NP_060098) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "            row_count=row_count+1\n",
    "            \n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    TET2_missense_codons_NP_060098_DICT = {}\n",
    "    calls_TET2_missense_NP_060098 = {}\n",
    "\n",
    "    for i in TET2_missense_codons_NP_060098: #Iterates over the list of codon numbers\n",
    "        TET2_missense_codons_NP_060098_DICT[i+1]= (TET2_codons_NP_060098[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "        \n",
    "    for k,v in TET2_missense_codons_NP_060098_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_TET2_missense_NP_060098[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "        \n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_TET2_missense_NP_060098_extra = {}\n",
    "\n",
    "    for k, v in calls_TET2_missense_NP_060098.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_TET2_missense_NP_060098_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_TET2_missense_NP_060098_extra2 = {}\n",
    "\n",
    "    for k, v in calls_TET2_missense_NP_060098_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_TET2_missense_NP_060098_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "            \n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    TET2_position_in_list_dict_missense_NP_060098={}\n",
    "    for k, v in TET2_missense_variant_DICT_NP_060098.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_TET2_missense_NP_060098[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_TET2_missense_NP_060098[v[0]]\n",
    "        codon_seq=calls_TET2_missense_NP_060098[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_TET2_missense_NP_060098_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                TET2_position_in_list_dict_missense_NP_060098[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "            \n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in TET2_position_in_list_dict_missense_NP_060098.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "#     print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "                \n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TET2_bases_missense_NP_060098 = {}\n",
    "    for k, v in TET2_position_in_list_dict_missense_NP_060098.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TET2_bases_missense_NP_060098[k]=(v, base_changes)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TET2_bases_missense2_NP_060098 = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense_NP_060098.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TET2_bases_missense2_NP_060098[k] = (v, codon_positions)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TET2_bases_missense3_NP_060098 = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense2_NP_060098.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TET2_bases_missense3_NP_060098[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TET2_bases_missense4_NP_060098 = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense3_NP_060098.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "            gene_pos = i\n",
    "            site_trio_missense = [TET2_exons_extra_bases_NP_060098[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "            sites_trio_missense.append(site_trio_missense)\n",
    "            TET2_bases_missense4_NP_060098[k] = (v, sites_trio_missense)\n",
    "                    \n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TET2_sites_missense5_NP_060098 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TET2_bases_missense4_NP_060098.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        TET2_sites_missense5_NP_060098[a]=info\n",
    "        a=a+1\n",
    "        \n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TET2_sites_missense6_NP_060098={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TET2_sites_missense5_NP_060098.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "    \n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TET2_sites_missense6_NP_060098[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "        \n",
    "        \n",
    "    #Assign the mutation rates to each site...\n",
    "    TET2_mutation_rates_missense_NP_060098 = {}\n",
    "    for k, v in TET2_sites_missense6_NP_060098.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            TET2_mutation_rates_missense_NP_060098[k] = (v, mutation_rate)\n",
    "\n",
    "    TET2_mutation_rate_missense_NP_060098_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TET2_mutation_rates_missense_NP_060098.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        TET2_mutation_rate_missense_NP_060098_sum.append(mutation_rate_value)\n",
    "        \n",
    "    TET2_Summed_mutation_rate_missense_NP_060098_added = np.sum(TET2_mutation_rate_missense_NP_060098_sum)\n",
    "\n",
    "    print('Total summed mutation rate across TET2 missense_NP_060098 for this study = ', TET2_Summed_mutation_rate_missense_NP_060098_added)\n",
    "    \n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all #in format: column 1 = 882, column 2 = p.R882H  #1165 aa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ASXL1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_list_nonsense_variants_ASXL1(csv_file_variants):\n",
    "    \n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base = [140, 143, 252, 373, 1086, 1085]\n",
    "   \n",
    "    def intron_exon_check(position_in_gene_number):\n",
    "        if position_in_gene_number == 140:\n",
    "            gene_pos = 140\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 143:\n",
    "            gene_pos = 143\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 252:\n",
    "            gene_pos = 252\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 373:\n",
    "            gene_pos = 373\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1086:\n",
    "            gene_pos = 1086\n",
    "            first_base = 'G'\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = ASXL1_exons_extra_bases[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1085:\n",
    "            gene_pos = 1085\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #ASXL1 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    ASXL1_exons = 'ATGAAGGACAAACAGAAGAAGAAGAAGGAGCGCACGTGGGCCGAGGCCGCGCGCCTGGTATTAGAAAACTACTCGGATGCTCCAATGACACCAAAACAGATTCTGCAGGTCATAGAGGCAGAAGGACTAAAGGAAATGAGAAGTGGGACTTCCCCTCTCGCATGCCTCAATGCTATGCTACATTCCAATTCAAGAGGAGGAGAGGGGTTGTTTTATAAACTGCCTGGCCGAATCAGCCTTTTCACGCTCAAGAAGGATGCCCTGCAGTGGTCTCGCCATCCAGCTACAGTGGAGGGAGAGGAGCCAGAGGACACGGCTGATGTGGAGAGCTGTGGGTCTAATGAAGCCAGCACTGTGAGTGGTGAAAACGATGTATCTCTTGATGAAACATCTTCGAACGCATCCTGTTCTACAGAATCTCAGAGTCGACCTCTTTCCAATCCCAGGGACAGCTACAGAGCTTCCTCACAGGCGAACAAACAAAAGAAAAAGACTGGGGTGATGCTGCCTCGAGTTGTCCTGACTCCTCTGAAGGTAAACGGGGCCCACGTGGAATCTGCATCAGGGTTCTCGGGCTGCCACGCCGATGGCGAGAGCGGCAGCCCGTCCAGCAGCAGCAGCGGCTCTCTGGCCCTGGGCAGCGCTGCTATTCGTGGCCAGGCCGAGGTCACCCAGGACCCTGCCCCGCTCCTGAGAGGCTTCCGGAAGCCAGCCACAGGTCAAATGAAGCGCAACAGAGGGGAAGAAATAGATTTTGAGACACCTGGGTCCATTCTTGTCAACACCAACCTCCGTGCCCTGATCAACTCTCGGACCTTCCATGCCTTACCATCACACTTCCAGCAGCAGCTCCTCTTCCTCCTGCCTGAAGTAGACAGACAGGTGGGGACGGATGGCCTGTTGCGTCTCAGCAGCAGTGCACTAAATAACGAGTTTTTTACCCATGCGGCTCAGAGCTGGCGGGAGCGCCTGGCTGATGGTGAATTTACTCATGAGATGCAAGTCAGGATACGACAGGAAATGGAGAAGGAAAAGAAGGTGGAACAATGGAAAGAAAAGTTCTTTGAAGACTACTATGGACAGAAGCTGGGTTTGACCAAAGAAGAGTCATTGCAGCAGAACGTGGGCCAGGAGGAGGCTGAAATCAAAAGTGGCTTGTGTGTCCCAGGAGAATCAGTGCGTATACAGCGTGGTCCAGCCACCCGACAGCGAGATGGGCATTTTAAGAAACGCTCTCGGCCAGATCTCCGAACCAGAGCCAGAAGGAATCTGTACAAAAAACAGGAGTCAGAACAAGCAGGGGTTGCTAAGGATGCAAAATCTGTGGCCTCAGATGTTCCCCTCTACAAGGATGGGGAGGCTAAGACTGACCCAGCAGGGCTGAGCAGTCCCCATCTGCCAGGCACATCCTCTGCAGCACCCGACCTGGAGGGTCCCGAATTCCCAGTTGAGTCTGTGGCTTCTCGGATCCAGGCTGAGCCAGACAACTTGGCACGTGCCTCTGCATCTCCAGACAGAATTCCTAGCCTGCCTCAGGAAACTGTGGATCAGGAACCCAAGGATCAGAAGAGGAAATCCTTTGAGCAGGCGGCCTCTGCATCCTTTCCCGAAAAGAAGCCCCGGCTTGAAGATCGTCAGTCCTTTCGTAACACAATTGAAAGTGTTCACACCGAAAAGCCACAGCCCACTAAAGAGGAGCCCAAAGTCCCGCCCATCCGGATTCAACTTTCACGTATCAAACCACCCTGGGTGGTTAAAGGTCAGCCCACTTACCAGATATGCCCCCGGATCATCCCCACCACGGAGTCCTCCTGCCGGGGTTGGACTGGCGCCAGGACCCTCGCAGACATTAAAGCCCGTGCTCTGCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCCACCACTGCCATCGGAGGGGGGGGTGGCCCGGGTGGAGGTGGCGGCGGGGCCACCGATGAGGGAGGTGGCAGAGGCAGCAGCAGTGGTGATGGTGGTGAGGCCTGTGGCCACCCTGAGCCCAGGGGAGGCCCGAGCACCCCTGGAAAGTGTACGTCAGATCTACAGCGAACACAACTACTGCCGCCTTATCCTCTAAATGGGGAGCATACCCAGGCCGGAACTGCCATGTCCAGAGCTAGGAGAGAGGACCTGCCTTCTCTGAGAAAGGAGGAAAGCTGCCTACTACAGAGGGCTACAGTTGGACTCACAGATGGGCTAGGAGATGCCTCCCAACTCCCCGTTGCTCCCACTGGGGACCAGCCATGCCAGGCCTTGCCCCTACTGTCCTCCCAAACCTCAGTAGCTGAGAGATTAGTGGAGCAGCCTCAGTTGCATCCGGATGTTAGAACTGAATGTGAGTCTGGCACCACTTCCTGGGAAAGTGATGATGAGGAGCAAGGACCCACCGTTCCTGCAGACAATGGTCCCATTCTGTCTCTAGTGGGAGATGATACATTAGAGAAAGGAACTGGCCAAGCTCTTGACAGTCATCCCACTATGAAGGATCCTGTAAATGTGACCCCCAGTTCCACACCTGAATCCTCACCGACTGATTGCCTGCAGAACAGAGCATTTGATGACGAATTAGGGCTTGGTGGCTCATGCCCTCCTATGAGGGAAAGTGATACTAGACAAGAAAACTTGAAAACCAAGGCTCTCGTTTCTAACAGTTCTTTGCATTGGATACCCATCCCATCGAATGATGAGGTAGTGAAACAGCCCAAACCAGAATCCAGAGAACACATACCATCTGTTGAGCCCCAGGTTGGAGAGGAGTGGGAGAAAGCTGCTCCCACCCCTCCTGCATTGCCTGGGGATTTGACAGCTGAGGAGGGTCTAGATCCTCTTGACAGCCTTACTTCACTCTGGACTGTGCCATCTCGAGGAGGCAGTGACAGCAATGGCAGTTACTGTCAACAGGTGGACATTGAAAAGCTGAAAATCAACGGAGACTCTGAAGCACTGAGTCCTCACGGTGAGTCCACGGATACAGCCTCTGACTTTGAAGGTCACCTCACGGAGGACAGCAGTGAGGCTGACACTAGAGAAGCTGCAGTGACAAAGGGATCTTCGGTGGACAAGGATGAGAAACCCAATTGGAACCAATCTGCCCCACTGTCCAAGGTGAATGGTGACATGCGTCTGGTTACAAGGACAGATGGGATGGTTGCTCCTCAGAGCTGGGTGTCTCGAGTATGTGCGGTCCGCCAAAAGATCCCAGATTCCCTACTGCTGGCCAGTACTGAGTACCAGCCAAGAGCCGTGTGCCTGTCCATGCCTGGGTCCTCAGTGGAGGCCACTAACCCACTTGTGATGCAGTTGCTGCAGGGTAGCTTGCCCCTAGAGAAGGTTCTTCCACCAGCCCACGATGACAGCATGTCAGAATCCCCACAAGTACCACTTACAAAAGACCAGAGCCATGGCTCGCTACGCATGGGATCTTTACATGGTCTTGGAAAAAACAGTGGCATGGTTGATGGAAGCAGCCCCAGTTCTTTAAGGGCTTTGAAGGAGCCTCTTCTGCCAGATAGCTGTGAAACAGGCACTGGTCTTGCCAGGATTGAGGCCACCCAGGCTCCTGGAGCACCCCAAAAGAATTGCAAGGCAGTCCCAAGTTTTGACTCCCTCCATCCAGTGACAAATCCCATTACATCCTCTAGGAAACTGGAAGAAATGGATTCCAAAGAGCAGTTCTCTTCCTTTAGTTGTGAAGATCAGAAGGAAGTCCGTGCTATGTCACAGGACAGTAATTCAAATGCTGCTCCAGGAAAGAGCCCAGGAGATCTTACTACCTCGAGAACACCTCGTTTCTCATCTCCAAATGTGATCTCCTTTGGTCCAGAGCAGACAGGTCGGGCCCTGGGTGATCAGAGCAATGTTACAGGCCAAGGGAAGAAGCTTTTTGGCTCTGGGAATGTGGCTGCAACCCTTCAGCGCCCCAGGCCTGCGGACCCGATGCCTCTTCCTGCTGAGATCCCTCCAGTTTTTCCCAGTGGGAAGTTGGGACCAAGCACAAACTCCATGTCTGGTGGGGTACAGACTCCAAGGGAAGACTGGGCTCCAAAGCCACATGCCTTTGTTGGCAGCGTCAAGAATGAGAAGACTTTTGTGGGGGGTCCTCTTAAGGCAAATGCCGAGAACAGGAAAGCTACTGGGCATAGTCCCCTGGAACTGGTGGGTCACTTGGAAGGGATGCCCTTTGTCATGGACTTGCCCTTCTGGAAATTACCCCGAGAGCCAGGGAAGGGGCTCAGTGAGCCTCTGGAGCCTTCTTCTCTCCCCTCCCAACTCAGCATCAAGCAGGCATTTTATGGGAAGCTTTCTAAACTCCAACTGAGTTCCACCAGCTTTAATTATTCCTCTAGCTCTCCCACCTTTCCCAAAGGCCTTGCTGGAAGTGTGGTGCAGCTGAGCCACAAAGCAAACTTTGGTGCGAGCCACAGTGCATCACTTTCCTTGCAAATGTTCACTGACAGCAGCACGGTGGAAAGCATCTCGCTCCAGTGTGCGTGCAGCCTGAAAGCCATGATCATGTGCCAAGGCTGCGGTGCGTTCTGTCACGATGACTGTATTGGACCCTCAAAGCTCTGTGTATTGTGCCTTGTGGTGAGATAA'\n",
    "    \n",
    "    #Split the sequence into codons \n",
    "    ASXL1_codons = [ASXL1_exons[i:i+3] for i in range(0, len(ASXL1_exons), 3)]\n",
    "\n",
    "    #ASXL1 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    ASXL1_exons_extra_bases = 'GATGAAGGACAAACAGAAGAAGAAGAAGGAGCGCACGTGGGCCGAGGCCGCGCGCCTGGTATTAGAAAACTACTCGGATGCTCCAATGACACCAAAACAGATTCTGCAGGTCATAGAGGCAGAAGGACTAAAGGAAATGAGAAGTGGGACTTCCCCTCTCGCATGCCTCAATGCTATGCTACATTCCAATTCAAGAGGAGGAGAGGGGTTGTTTTATAAACTGCCTGGCCGAATCAGCCTTTTCACGCTCAAGAAGGATGCCCTGCAGTGGTCTCGCCATCCAGCTACAGTGGAGGGAGAGGAGCCAGAGGACACGGCTGATGTGGAGAGCTGTGGGTCTAATGAAGCCAGCACTGTGAGTGGTGAAAACGATGTATCTCTTGATGAAACATCTTCGAACGCATCCTGTTCTACAGAATCTCAGAGTCGACCTCTTTCCAATCCCAGGGACAGCTACAGAGCTTCCTCACAGGCGAACAAACAAAAGAAAAAGACTGGGGTGATGCTGCCTCGAGTTGTCCTGACTCCTCTGAAGGTAAACGGGGCCCACGTGGAATCTGCATCAGGGTTCTCGGGCTGCCACGCCGATGGCGAGAGCGGCAGCCCGTCCAGCAGCAGCAGCGGCTCTCTGGCCCTGGGCAGCGCTGCTATTCGTGGCCAGGCCGAGGTCACCCAGGACCCTGCCCCGCTCCTGAGAGGCTTCCGGAAGCCAGCCACAGGTCAAATGAAGCGCAACAGAGGGGAAGAAATAGATTTTGAGACACCTGGGTCCATTCTTGTCAACACCAACCTCCGTGCCCTGATCAACTCTCGGACCTTCCATGCCTTACCATCACACTTCCAGCAGCAGCTCCTCTTCCTCCTGCCTGAAGTAGACAGACAGGTGGGGACGGATGGCCTGTTGCGTCTCAGCAGCAGTGCACTAAATAACGAGTTTTTTACCCATGCGGCTCAGAGCTGGCGGGAGCGCCTGGCTGATGGTGAATTTACTCATGAGATGCAAGTCAGGATACGACAGGAAATGGAGAAGGAAAAGAAGGTGGAACAATGGAAAGAAAAGTTCTTTGAAGACTACTATGGACAGAAGCTGGGTTTGACCAAAGAAGAGTCATTGCAGCAGAACGTGGGCCAGGAGGAGGCTGAAATCAAAAGTGGCTTGTGTGTCCCAGGAGAATCAGTGCGTATACAGCGTGGTCCAGCCACCCGACAGCGAGATGGGCATTTTAAGAAACGCTCTCGGCCAGATCTCCGAACCAGAGCCAGAAGGAATCTGTACAAAAAACAGGAGTCAGAACAAGCAGGGGTTGCTAAGGATGCAAAATCTGTGGCCTCAGATGTTCCCCTCTACAAGGATGGGGAGGCTAAGACTGACCCAGCAGGGCTGAGCAGTCCCCATCTGCCAGGCACATCCTCTGCAGCACCCGACCTGGAGGGTCCCGAATTCCCAGTTGAGTCTGTGGCTTCTCGGATCCAGGCTGAGCCAGACAACTTGGCACGTGCCTCTGCATCTCCAGACAGAATTCCTAGCCTGCCTCAGGAAACTGTGGATCAGGAACCCAAGGATCAGAAGAGGAAATCCTTTGAGCAGGCGGCCTCTGCATCCTTTCCCGAAAAGAAGCCCCGGCTTGAAGATCGTCAGTCCTTTCGTAACACAATTGAAAGTGTTCACACCGAAAAGCCACAGCCCACTAAAGAGGAGCCCAAAGTCCCGCCCATCCGGATTCAACTTTCACGTATCAAACCACCCTGGGTGGTTAAAGGTCAGCCCACTTACCAGATATGCCCCCGGATCATCCCCACCACGGAGTCCTCCTGCCGGGGTTGGACTGGCGCCAGGACCCTCGCAGACATTAAAGCCCGTGCTCTGCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCCACCACTGCCATCGGAGGGGGGGGTGGCCCGGGTGGAGGTGGCGGCGGGGCCACCGATGAGGGAGGTGGCAGAGGCAGCAGCAGTGGTGATGGTGGTGAGGCCTGTGGCCACCCTGAGCCCAGGGGAGGCCCGAGCACCCCTGGAAAGTGTACGTCAGATCTACAGCGAACACAACTACTGCCGCCTTATCCTCTAAATGGGGAGCATACCCAGGCCGGAACTGCCATGTCCAGAGCTAGGAGAGAGGACCTGCCTTCTCTGAGAAAGGAGGAAAGCTGCCTACTACAGAGGGCTACAGTTGGACTCACAGATGGGCTAGGAGATGCCTCCCAACTCCCCGTTGCTCCCACTGGGGACCAGCCATGCCAGGCCTTGCCCCTACTGTCCTCCCAAACCTCAGTAGCTGAGAGATTAGTGGAGCAGCCTCAGTTGCATCCGGATGTTAGAACTGAATGTGAGTCTGGCACCACTTCCTGGGAAAGTGATGATGAGGAGCAAGGACCCACCGTTCCTGCAGACAATGGTCCCATTCTGTCTCTAGTGGGAGATGATACATTAGAGAAAGGAACTGGCCAAGCTCTTGACAGTCATCCCACTATGAAGGATCCTGTAAATGTGACCCCCAGTTCCACACCTGAATCCTCACCGACTGATTGCCTGCAGAACAGAGCATTTGATGACGAATTAGGGCTTGGTGGCTCATGCCCTCCTATGAGGGAAAGTGATACTAGACAAGAAAACTTGAAAACCAAGGCTCTCGTTTCTAACAGTTCTTTGCATTGGATACCCATCCCATCGAATGATGAGGTAGTGAAACAGCCCAAACCAGAATCCAGAGAACACATACCATCTGTTGAGCCCCAGGTTGGAGAGGAGTGGGAGAAAGCTGCTCCCACCCCTCCTGCATTGCCTGGGGATTTGACAGCTGAGGAGGGTCTAGATCCTCTTGACAGCCTTACTTCACTCTGGACTGTGCCATCTCGAGGAGGCAGTGACAGCAATGGCAGTTACTGTCAACAGGTGGACATTGAAAAGCTGAAAATCAACGGAGACTCTGAAGCACTGAGTCCTCACGGTGAGTCCACGGATACAGCCTCTGACTTTGAAGGTCACCTCACGGAGGACAGCAGTGAGGCTGACACTAGAGAAGCTGCAGTGACAAAGGGATCTTCGGTGGACAAGGATGAGAAACCCAATTGGAACCAATCTGCCCCACTGTCCAAGGTGAATGGTGACATGCGTCTGGTTACAAGGACAGATGGGATGGTTGCTCCTCAGAGCTGGGTGTCTCGAGTATGTGCGGTCCGCCAAAAGATCCCAGATTCCCTACTGCTGGCCAGTACTGAGTACCAGCCAAGAGCCGTGTGCCTGTCCATGCCTGGGTCCTCAGTGGAGGCCACTAACCCACTTGTGATGCAGTTGCTGCAGGGTAGCTTGCCCCTAGAGAAGGTTCTTCCACCAGCCCACGATGACAGCATGTCAGAATCCCCACAAGTACCACTTACAAAAGACCAGAGCCATGGCTCGCTACGCATGGGATCTTTACATGGTCTTGGAAAAAACAGTGGCATGGTTGATGGAAGCAGCCCCAGTTCTTTAAGGGCTTTGAAGGAGCCTCTTCTGCCAGATAGCTGTGAAACAGGCACTGGTCTTGCCAGGATTGAGGCCACCCAGGCTCCTGGAGCACCCCAAAAGAATTGCAAGGCAGTCCCAAGTTTTGACTCCCTCCATCCAGTGACAAATCCCATTACATCCTCTAGGAAACTGGAAGAAATGGATTCCAAAGAGCAGTTCTCTTCCTTTAGTTGTGAAGATCAGAAGGAAGTCCGTGCTATGTCACAGGACAGTAATTCAAATGCTGCTCCAGGAAAGAGCCCAGGAGATCTTACTACCTCGAGAACACCTCGTTTCTCATCTCCAAATGTGATCTCCTTTGGTCCAGAGCAGACAGGTCGGGCCCTGGGTGATCAGAGCAATGTTACAGGCCAAGGGAAGAAGCTTTTTGGCTCTGGGAATGTGGCTGCAACCCTTCAGCGCCCCAGGCCTGCGGACCCGATGCCTCTTCCTGCTGAGATCCCTCCAGTTTTTCCCAGTGGGAAGTTGGGACCAAGCACAAACTCCATGTCTGGTGGGGTACAGACTCCAAGGGAAGACTGGGCTCCAAAGCCACATGCCTTTGTTGGCAGCGTCAAGAATGAGAAGACTTTTGTGGGGGGTCCTCTTAAGGCAAATGCCGAGAACAGGAAAGCTACTGGGCATAGTCCCCTGGAACTGGTGGGTCACTTGGAAGGGATGCCCTTTGTCATGGACTTGCCCTTCTGGAAATTACCCCGAGAGCCAGGGAAGGGGCTCAGTGAGCCTCTGGAGCCTTCTTCTCTCCCCTCCCAACTCAGCATCAAGCAGGCATTTTATGGGAAGCTTTCTAAACTCCAACTGAGTTCCACCAGCTTTAATTATTCCTCTAGCTCTCCCACCTTTCCCAAAGGCCTTGCTGGAAGTGTGGTGCAGCTGAGCCACAAAGCAAACTTTGGTGCGAGCCACAGTGCATCACTTTCCTTGCAAATGTTCACTGACAGCAGCACGGTGGAAAGCATCTCGCTCCAGTGTGCGTGCAGCCTGAAAGCCATGATCATGTGCCAAGGCTGCGGTGCGTTCTGTCACGATGACTGTATTGGACCCTCAAAGCTCTGTGTATTGTGCCTTGTGGTGAGATAAT'\n",
    "    \n",
    "    #Create a list of the codons they called nonsense variants at\n",
    "    import csv\n",
    "    filename = csv_file_variants\n",
    "    with open(filename, 'r') as csvfile:\n",
    "        read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "        row_count=0\n",
    "\n",
    "        ASXL1_nonsense_variant_list = []\n",
    "        ASXL1_nonsense_codons =[]\n",
    "        ASXL1_nonsense_variant_DICT = {}\n",
    "\n",
    "        for row in read_reader: #each row contains aa number (e.g. 11) and variant (e.g. p.D11A)\n",
    "    #         print(row)\n",
    "            if row_count>0: \n",
    "                codon0 = int(row[0])-1\n",
    "                codon1 = int(row[0])\n",
    "                variant = row[1]\n",
    "                ASXL1_nonsense_codons.append(codon0) #List of codon numbers that they called variants at\n",
    "                ASXL1_nonsense_variant_DICT[row_count]=(codon1, variant) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "            row_count=row_count+1\n",
    "            \n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    ASXL1_nonsense_codons_DICT = {}\n",
    "    calls_nonsense = {}\n",
    "\n",
    "    for i in ASXL1_nonsense_codons: #Iterates over the list of codon numbers\n",
    "        ASXL1_nonsense_codons_DICT[i+1]= (ASXL1_codons[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "        \n",
    "    for k,v in ASXL1_nonsense_codons_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_nonsense[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "        \n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_nonsense_extra = {}\n",
    "\n",
    "    for k, v in calls_nonsense.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_nonsense_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_nonsense_extra2 = {}\n",
    "\n",
    "    for k, v in calls_nonsense_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_nonsense_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "            \n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    ASXL1_position_in_list_dict_nonsense={}\n",
    "    for k, v in ASXL1_nonsense_variant_DICT.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_nonsense[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_nonsense[v[0]]\n",
    "        codon_seq=calls_nonsense[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_nonsense_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                ASXL1_position_in_list_dict_nonsense[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "            \n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in ASXL1_position_in_list_dict_nonsense.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "#     print('positions at which nonsense variants called = ', aa_positions_variants_called)\n",
    "                \n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    ASXL1_bases_nonsense = {}\n",
    "    for k, v in ASXL1_position_in_list_dict_nonsense.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        ASXL1_bases_nonsense[k]=(v, base_changes)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    ASXL1_bases_nonsense2 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_nonsense.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        ASXL1_bases_nonsense2[k] = (v, codon_positions)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    ASXL1_bases_nonsense3 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_nonsense2.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        ASXL1_bases_nonsense3[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    ASXL1_bases_nonsense4 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_nonsense3.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_nonsense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check(i)\n",
    "                sites_trio_nonsense.append([adapted_trio])\n",
    "                ASXL1_bases_nonsense4[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_nonsense = [ASXL1_exons_extra_bases[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_nonsense.append(site_trio_nonsense)\n",
    "                    ASXL1_bases_nonsense4[k] = (v, sites_trio_nonsense)\n",
    "                    \n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    ASXL1_sites_nonsense5 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in ASXL1_bases_nonsense4.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        ASXL1_sites_nonsense5[a]=info\n",
    "        a=a+1\n",
    "#     print(ASXL1_sites_nonsense5)\n",
    "        \n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    ASXL1_sites_nonsense6={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in ASXL1_sites_nonsense5.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "    \n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        ASXL1_sites_nonsense6[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "        \n",
    "        \n",
    "    #Assign the mutation rates to each site...\n",
    "    ASXL1_mutation_rates_nonsense = {}\n",
    "    for k, v in ASXL1_sites_nonsense6.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            ASXL1_mutation_rates_nonsense[k] = (v, mutation_rate)\n",
    "\n",
    "    ASXL1_mutation_rate_nonsense_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in ASXL1_mutation_rates_nonsense.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        ASXL1_mutation_rate_nonsense_sum.append(mutation_rate_value)\n",
    "        \n",
    "    ASXL1_Summed_mutation_rate_nonsense_added = np.sum(ASXL1_mutation_rate_nonsense_sum)\n",
    "\n",
    "    print('Total summed mutation rate across ASXL1 nonsense for this study = ', ASXL1_Summed_mutation_rate_nonsense_added)\n",
    "    \n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all #in format: column 1 = 882, column 2 = p.R882*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_list_missense_variants_ASXL1(csv_file_variants):\n",
    "    \n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base = [140, 143, 252, 373, 1086, 1085]\n",
    "   \n",
    "    def intron_exon_check(position_in_gene_number):\n",
    "        if position_in_gene_number == 140:\n",
    "            gene_pos = 140\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 143:\n",
    "            gene_pos = 143\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 252:\n",
    "            gene_pos = 252\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 373:\n",
    "            gene_pos = 373\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1086:\n",
    "            gene_pos = 1086\n",
    "            first_base = 'G'\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = ASXL1_exons_extra_bases[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1085:\n",
    "            gene_pos = 1085\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #ASXL1 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    ASXL1_exons = 'ATGAAGGACAAACAGAAGAAGAAGAAGGAGCGCACGTGGGCCGAGGCCGCGCGCCTGGTATTAGAAAACTACTCGGATGCTCCAATGACACCAAAACAGATTCTGCAGGTCATAGAGGCAGAAGGACTAAAGGAAATGAGAAGTGGGACTTCCCCTCTCGCATGCCTCAATGCTATGCTACATTCCAATTCAAGAGGAGGAGAGGGGTTGTTTTATAAACTGCCTGGCCGAATCAGCCTTTTCACGCTCAAGAAGGATGCCCTGCAGTGGTCTCGCCATCCAGCTACAGTGGAGGGAGAGGAGCCAGAGGACACGGCTGATGTGGAGAGCTGTGGGTCTAATGAAGCCAGCACTGTGAGTGGTGAAAACGATGTATCTCTTGATGAAACATCTTCGAACGCATCCTGTTCTACAGAATCTCAGAGTCGACCTCTTTCCAATCCCAGGGACAGCTACAGAGCTTCCTCACAGGCGAACAAACAAAAGAAAAAGACTGGGGTGATGCTGCCTCGAGTTGTCCTGACTCCTCTGAAGGTAAACGGGGCCCACGTGGAATCTGCATCAGGGTTCTCGGGCTGCCACGCCGATGGCGAGAGCGGCAGCCCGTCCAGCAGCAGCAGCGGCTCTCTGGCCCTGGGCAGCGCTGCTATTCGTGGCCAGGCCGAGGTCACCCAGGACCCTGCCCCGCTCCTGAGAGGCTTCCGGAAGCCAGCCACAGGTCAAATGAAGCGCAACAGAGGGGAAGAAATAGATTTTGAGACACCTGGGTCCATTCTTGTCAACACCAACCTCCGTGCCCTGATCAACTCTCGGACCTTCCATGCCTTACCATCACACTTCCAGCAGCAGCTCCTCTTCCTCCTGCCTGAAGTAGACAGACAGGTGGGGACGGATGGCCTGTTGCGTCTCAGCAGCAGTGCACTAAATAACGAGTTTTTTACCCATGCGGCTCAGAGCTGGCGGGAGCGCCTGGCTGATGGTGAATTTACTCATGAGATGCAAGTCAGGATACGACAGGAAATGGAGAAGGAAAAGAAGGTGGAACAATGGAAAGAAAAGTTCTTTGAAGACTACTATGGACAGAAGCTGGGTTTGACCAAAGAAGAGTCATTGCAGCAGAACGTGGGCCAGGAGGAGGCTGAAATCAAAAGTGGCTTGTGTGTCCCAGGAGAATCAGTGCGTATACAGCGTGGTCCAGCCACCCGACAGCGAGATGGGCATTTTAAGAAACGCTCTCGGCCAGATCTCCGAACCAGAGCCAGAAGGAATCTGTACAAAAAACAGGAGTCAGAACAAGCAGGGGTTGCTAAGGATGCAAAATCTGTGGCCTCAGATGTTCCCCTCTACAAGGATGGGGAGGCTAAGACTGACCCAGCAGGGCTGAGCAGTCCCCATCTGCCAGGCACATCCTCTGCAGCACCCGACCTGGAGGGTCCCGAATTCCCAGTTGAGTCTGTGGCTTCTCGGATCCAGGCTGAGCCAGACAACTTGGCACGTGCCTCTGCATCTCCAGACAGAATTCCTAGCCTGCCTCAGGAAACTGTGGATCAGGAACCCAAGGATCAGAAGAGGAAATCCTTTGAGCAGGCGGCCTCTGCATCCTTTCCCGAAAAGAAGCCCCGGCTTGAAGATCGTCAGTCCTTTCGTAACACAATTGAAAGTGTTCACACCGAAAAGCCACAGCCCACTAAAGAGGAGCCCAAAGTCCCGCCCATCCGGATTCAACTTTCACGTATCAAACCACCCTGGGTGGTTAAAGGTCAGCCCACTTACCAGATATGCCCCCGGATCATCCCCACCACGGAGTCCTCCTGCCGGGGTTGGACTGGCGCCAGGACCCTCGCAGACATTAAAGCCCGTGCTCTGCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCCACCACTGCCATCGGAGGGGGGGGTGGCCCGGGTGGAGGTGGCGGCGGGGCCACCGATGAGGGAGGTGGCAGAGGCAGCAGCAGTGGTGATGGTGGTGAGGCCTGTGGCCACCCTGAGCCCAGGGGAGGCCCGAGCACCCCTGGAAAGTGTACGTCAGATCTACAGCGAACACAACTACTGCCGCCTTATCCTCTAAATGGGGAGCATACCCAGGCCGGAACTGCCATGTCCAGAGCTAGGAGAGAGGACCTGCCTTCTCTGAGAAAGGAGGAAAGCTGCCTACTACAGAGGGCTACAGTTGGACTCACAGATGGGCTAGGAGATGCCTCCCAACTCCCCGTTGCTCCCACTGGGGACCAGCCATGCCAGGCCTTGCCCCTACTGTCCTCCCAAACCTCAGTAGCTGAGAGATTAGTGGAGCAGCCTCAGTTGCATCCGGATGTTAGAACTGAATGTGAGTCTGGCACCACTTCCTGGGAAAGTGATGATGAGGAGCAAGGACCCACCGTTCCTGCAGACAATGGTCCCATTCTGTCTCTAGTGGGAGATGATACATTAGAGAAAGGAACTGGCCAAGCTCTTGACAGTCATCCCACTATGAAGGATCCTGTAAATGTGACCCCCAGTTCCACACCTGAATCCTCACCGACTGATTGCCTGCAGAACAGAGCATTTGATGACGAATTAGGGCTTGGTGGCTCATGCCCTCCTATGAGGGAAAGTGATACTAGACAAGAAAACTTGAAAACCAAGGCTCTCGTTTCTAACAGTTCTTTGCATTGGATACCCATCCCATCGAATGATGAGGTAGTGAAACAGCCCAAACCAGAATCCAGAGAACACATACCATCTGTTGAGCCCCAGGTTGGAGAGGAGTGGGAGAAAGCTGCTCCCACCCCTCCTGCATTGCCTGGGGATTTGACAGCTGAGGAGGGTCTAGATCCTCTTGACAGCCTTACTTCACTCTGGACTGTGCCATCTCGAGGAGGCAGTGACAGCAATGGCAGTTACTGTCAACAGGTGGACATTGAAAAGCTGAAAATCAACGGAGACTCTGAAGCACTGAGTCCTCACGGTGAGTCCACGGATACAGCCTCTGACTTTGAAGGTCACCTCACGGAGGACAGCAGTGAGGCTGACACTAGAGAAGCTGCAGTGACAAAGGGATCTTCGGTGGACAAGGATGAGAAACCCAATTGGAACCAATCTGCCCCACTGTCCAAGGTGAATGGTGACATGCGTCTGGTTACAAGGACAGATGGGATGGTTGCTCCTCAGAGCTGGGTGTCTCGAGTATGTGCGGTCCGCCAAAAGATCCCAGATTCCCTACTGCTGGCCAGTACTGAGTACCAGCCAAGAGCCGTGTGCCTGTCCATGCCTGGGTCCTCAGTGGAGGCCACTAACCCACTTGTGATGCAGTTGCTGCAGGGTAGCTTGCCCCTAGAGAAGGTTCTTCCACCAGCCCACGATGACAGCATGTCAGAATCCCCACAAGTACCACTTACAAAAGACCAGAGCCATGGCTCGCTACGCATGGGATCTTTACATGGTCTTGGAAAAAACAGTGGCATGGTTGATGGAAGCAGCCCCAGTTCTTTAAGGGCTTTGAAGGAGCCTCTTCTGCCAGATAGCTGTGAAACAGGCACTGGTCTTGCCAGGATTGAGGCCACCCAGGCTCCTGGAGCACCCCAAAAGAATTGCAAGGCAGTCCCAAGTTTTGACTCCCTCCATCCAGTGACAAATCCCATTACATCCTCTAGGAAACTGGAAGAAATGGATTCCAAAGAGCAGTTCTCTTCCTTTAGTTGTGAAGATCAGAAGGAAGTCCGTGCTATGTCACAGGACAGTAATTCAAATGCTGCTCCAGGAAAGAGCCCAGGAGATCTTACTACCTCGAGAACACCTCGTTTCTCATCTCCAAATGTGATCTCCTTTGGTCCAGAGCAGACAGGTCGGGCCCTGGGTGATCAGAGCAATGTTACAGGCCAAGGGAAGAAGCTTTTTGGCTCTGGGAATGTGGCTGCAACCCTTCAGCGCCCCAGGCCTGCGGACCCGATGCCTCTTCCTGCTGAGATCCCTCCAGTTTTTCCCAGTGGGAAGTTGGGACCAAGCACAAACTCCATGTCTGGTGGGGTACAGACTCCAAGGGAAGACTGGGCTCCAAAGCCACATGCCTTTGTTGGCAGCGTCAAGAATGAGAAGACTTTTGTGGGGGGTCCTCTTAAGGCAAATGCCGAGAACAGGAAAGCTACTGGGCATAGTCCCCTGGAACTGGTGGGTCACTTGGAAGGGATGCCCTTTGTCATGGACTTGCCCTTCTGGAAATTACCCCGAGAGCCAGGGAAGGGGCTCAGTGAGCCTCTGGAGCCTTCTTCTCTCCCCTCCCAACTCAGCATCAAGCAGGCATTTTATGGGAAGCTTTCTAAACTCCAACTGAGTTCCACCAGCTTTAATTATTCCTCTAGCTCTCCCACCTTTCCCAAAGGCCTTGCTGGAAGTGTGGTGCAGCTGAGCCACAAAGCAAACTTTGGTGCGAGCCACAGTGCATCACTTTCCTTGCAAATGTTCACTGACAGCAGCACGGTGGAAAGCATCTCGCTCCAGTGTGCGTGCAGCCTGAAAGCCATGATCATGTGCCAAGGCTGCGGTGCGTTCTGTCACGATGACTGTATTGGACCCTCAAAGCTCTGTGTATTGTGCCTTGTGGTGAGATAA'\n",
    "    \n",
    "    #Split the sequence into codons \n",
    "    ASXL1_codons = [ASXL1_exons[i:i+3] for i in range(0, len(ASXL1_exons), 3)]\n",
    "\n",
    "    #ASXL1 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    ASXL1_exons_extra_bases = 'GATGAAGGACAAACAGAAGAAGAAGAAGGAGCGCACGTGGGCCGAGGCCGCGCGCCTGGTATTAGAAAACTACTCGGATGCTCCAATGACACCAAAACAGATTCTGCAGGTCATAGAGGCAGAAGGACTAAAGGAAATGAGAAGTGGGACTTCCCCTCTCGCATGCCTCAATGCTATGCTACATTCCAATTCAAGAGGAGGAGAGGGGTTGTTTTATAAACTGCCTGGCCGAATCAGCCTTTTCACGCTCAAGAAGGATGCCCTGCAGTGGTCTCGCCATCCAGCTACAGTGGAGGGAGAGGAGCCAGAGGACACGGCTGATGTGGAGAGCTGTGGGTCTAATGAAGCCAGCACTGTGAGTGGTGAAAACGATGTATCTCTTGATGAAACATCTTCGAACGCATCCTGTTCTACAGAATCTCAGAGTCGACCTCTTTCCAATCCCAGGGACAGCTACAGAGCTTCCTCACAGGCGAACAAACAAAAGAAAAAGACTGGGGTGATGCTGCCTCGAGTTGTCCTGACTCCTCTGAAGGTAAACGGGGCCCACGTGGAATCTGCATCAGGGTTCTCGGGCTGCCACGCCGATGGCGAGAGCGGCAGCCCGTCCAGCAGCAGCAGCGGCTCTCTGGCCCTGGGCAGCGCTGCTATTCGTGGCCAGGCCGAGGTCACCCAGGACCCTGCCCCGCTCCTGAGAGGCTTCCGGAAGCCAGCCACAGGTCAAATGAAGCGCAACAGAGGGGAAGAAATAGATTTTGAGACACCTGGGTCCATTCTTGTCAACACCAACCTCCGTGCCCTGATCAACTCTCGGACCTTCCATGCCTTACCATCACACTTCCAGCAGCAGCTCCTCTTCCTCCTGCCTGAAGTAGACAGACAGGTGGGGACGGATGGCCTGTTGCGTCTCAGCAGCAGTGCACTAAATAACGAGTTTTTTACCCATGCGGCTCAGAGCTGGCGGGAGCGCCTGGCTGATGGTGAATTTACTCATGAGATGCAAGTCAGGATACGACAGGAAATGGAGAAGGAAAAGAAGGTGGAACAATGGAAAGAAAAGTTCTTTGAAGACTACTATGGACAGAAGCTGGGTTTGACCAAAGAAGAGTCATTGCAGCAGAACGTGGGCCAGGAGGAGGCTGAAATCAAAAGTGGCTTGTGTGTCCCAGGAGAATCAGTGCGTATACAGCGTGGTCCAGCCACCCGACAGCGAGATGGGCATTTTAAGAAACGCTCTCGGCCAGATCTCCGAACCAGAGCCAGAAGGAATCTGTACAAAAAACAGGAGTCAGAACAAGCAGGGGTTGCTAAGGATGCAAAATCTGTGGCCTCAGATGTTCCCCTCTACAAGGATGGGGAGGCTAAGACTGACCCAGCAGGGCTGAGCAGTCCCCATCTGCCAGGCACATCCTCTGCAGCACCCGACCTGGAGGGTCCCGAATTCCCAGTTGAGTCTGTGGCTTCTCGGATCCAGGCTGAGCCAGACAACTTGGCACGTGCCTCTGCATCTCCAGACAGAATTCCTAGCCTGCCTCAGGAAACTGTGGATCAGGAACCCAAGGATCAGAAGAGGAAATCCTTTGAGCAGGCGGCCTCTGCATCCTTTCCCGAAAAGAAGCCCCGGCTTGAAGATCGTCAGTCCTTTCGTAACACAATTGAAAGTGTTCACACCGAAAAGCCACAGCCCACTAAAGAGGAGCCCAAAGTCCCGCCCATCCGGATTCAACTTTCACGTATCAAACCACCCTGGGTGGTTAAAGGTCAGCCCACTTACCAGATATGCCCCCGGATCATCCCCACCACGGAGTCCTCCTGCCGGGGTTGGACTGGCGCCAGGACCCTCGCAGACATTAAAGCCCGTGCTCTGCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCCACCACTGCCATCGGAGGGGGGGGTGGCCCGGGTGGAGGTGGCGGCGGGGCCACCGATGAGGGAGGTGGCAGAGGCAGCAGCAGTGGTGATGGTGGTGAGGCCTGTGGCCACCCTGAGCCCAGGGGAGGCCCGAGCACCCCTGGAAAGTGTACGTCAGATCTACAGCGAACACAACTACTGCCGCCTTATCCTCTAAATGGGGAGCATACCCAGGCCGGAACTGCCATGTCCAGAGCTAGGAGAGAGGACCTGCCTTCTCTGAGAAAGGAGGAAAGCTGCCTACTACAGAGGGCTACAGTTGGACTCACAGATGGGCTAGGAGATGCCTCCCAACTCCCCGTTGCTCCCACTGGGGACCAGCCATGCCAGGCCTTGCCCCTACTGTCCTCCCAAACCTCAGTAGCTGAGAGATTAGTGGAGCAGCCTCAGTTGCATCCGGATGTTAGAACTGAATGTGAGTCTGGCACCACTTCCTGGGAAAGTGATGATGAGGAGCAAGGACCCACCGTTCCTGCAGACAATGGTCCCATTCTGTCTCTAGTGGGAGATGATACATTAGAGAAAGGAACTGGCCAAGCTCTTGACAGTCATCCCACTATGAAGGATCCTGTAAATGTGACCCCCAGTTCCACACCTGAATCCTCACCGACTGATTGCCTGCAGAACAGAGCATTTGATGACGAATTAGGGCTTGGTGGCTCATGCCCTCCTATGAGGGAAAGTGATACTAGACAAGAAAACTTGAAAACCAAGGCTCTCGTTTCTAACAGTTCTTTGCATTGGATACCCATCCCATCGAATGATGAGGTAGTGAAACAGCCCAAACCAGAATCCAGAGAACACATACCATCTGTTGAGCCCCAGGTTGGAGAGGAGTGGGAGAAAGCTGCTCCCACCCCTCCTGCATTGCCTGGGGATTTGACAGCTGAGGAGGGTCTAGATCCTCTTGACAGCCTTACTTCACTCTGGACTGTGCCATCTCGAGGAGGCAGTGACAGCAATGGCAGTTACTGTCAACAGGTGGACATTGAAAAGCTGAAAATCAACGGAGACTCTGAAGCACTGAGTCCTCACGGTGAGTCCACGGATACAGCCTCTGACTTTGAAGGTCACCTCACGGAGGACAGCAGTGAGGCTGACACTAGAGAAGCTGCAGTGACAAAGGGATCTTCGGTGGACAAGGATGAGAAACCCAATTGGAACCAATCTGCCCCACTGTCCAAGGTGAATGGTGACATGCGTCTGGTTACAAGGACAGATGGGATGGTTGCTCCTCAGAGCTGGGTGTCTCGAGTATGTGCGGTCCGCCAAAAGATCCCAGATTCCCTACTGCTGGCCAGTACTGAGTACCAGCCAAGAGCCGTGTGCCTGTCCATGCCTGGGTCCTCAGTGGAGGCCACTAACCCACTTGTGATGCAGTTGCTGCAGGGTAGCTTGCCCCTAGAGAAGGTTCTTCCACCAGCCCACGATGACAGCATGTCAGAATCCCCACAAGTACCACTTACAAAAGACCAGAGCCATGGCTCGCTACGCATGGGATCTTTACATGGTCTTGGAAAAAACAGTGGCATGGTTGATGGAAGCAGCCCCAGTTCTTTAAGGGCTTTGAAGGAGCCTCTTCTGCCAGATAGCTGTGAAACAGGCACTGGTCTTGCCAGGATTGAGGCCACCCAGGCTCCTGGAGCACCCCAAAAGAATTGCAAGGCAGTCCCAAGTTTTGACTCCCTCCATCCAGTGACAAATCCCATTACATCCTCTAGGAAACTGGAAGAAATGGATTCCAAAGAGCAGTTCTCTTCCTTTAGTTGTGAAGATCAGAAGGAAGTCCGTGCTATGTCACAGGACAGTAATTCAAATGCTGCTCCAGGAAAGAGCCCAGGAGATCTTACTACCTCGAGAACACCTCGTTTCTCATCTCCAAATGTGATCTCCTTTGGTCCAGAGCAGACAGGTCGGGCCCTGGGTGATCAGAGCAATGTTACAGGCCAAGGGAAGAAGCTTTTTGGCTCTGGGAATGTGGCTGCAACCCTTCAGCGCCCCAGGCCTGCGGACCCGATGCCTCTTCCTGCTGAGATCCCTCCAGTTTTTCCCAGTGGGAAGTTGGGACCAAGCACAAACTCCATGTCTGGTGGGGTACAGACTCCAAGGGAAGACTGGGCTCCAAAGCCACATGCCTTTGTTGGCAGCGTCAAGAATGAGAAGACTTTTGTGGGGGGTCCTCTTAAGGCAAATGCCGAGAACAGGAAAGCTACTGGGCATAGTCCCCTGGAACTGGTGGGTCACTTGGAAGGGATGCCCTTTGTCATGGACTTGCCCTTCTGGAAATTACCCCGAGAGCCAGGGAAGGGGCTCAGTGAGCCTCTGGAGCCTTCTTCTCTCCCCTCCCAACTCAGCATCAAGCAGGCATTTTATGGGAAGCTTTCTAAACTCCAACTGAGTTCCACCAGCTTTAATTATTCCTCTAGCTCTCCCACCTTTCCCAAAGGCCTTGCTGGAAGTGTGGTGCAGCTGAGCCACAAAGCAAACTTTGGTGCGAGCCACAGTGCATCACTTTCCTTGCAAATGTTCACTGACAGCAGCACGGTGGAAAGCATCTCGCTCCAGTGTGCGTGCAGCCTGAAAGCCATGATCATGTGCCAAGGCTGCGGTGCGTTCTGTCACGATGACTGTATTGGACCCTCAAAGCTCTGTGTATTGTGCCTTGTGGTGAGATAAT'\n",
    "  \n",
    "    #Create a list of the codons they called missense variants at\n",
    "    import csv\n",
    "    filename = csv_file_variants\n",
    "    with open(filename, 'r') as csvfile:\n",
    "        read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "        row_count=0\n",
    "\n",
    "        ASXL1_missense_variant_list = []\n",
    "        ASXL1_missense_codons =[]\n",
    "        ASXL1_missense_variant_DICT = {}\n",
    "\n",
    "        for row in read_reader: #each row contains aa number (e.g. 11) and variant (e.g. p.D11A)\n",
    "    #         print(row)\n",
    "            if row_count>0: \n",
    "                codon0 = int(row[0])-1\n",
    "                codon1 = int(row[0])\n",
    "                variant = row[1]\n",
    "                ASXL1_missense_codons.append(codon0) #List of codon numbers that they called variants at\n",
    "                ASXL1_missense_variant_DICT[row_count]=(codon1, variant) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "            row_count=row_count+1\n",
    "            \n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    ASXL1_missense_codons_DICT = {}\n",
    "    calls_ASXL1_missense = {}\n",
    "\n",
    "    for i in ASXL1_missense_codons: #Iterates over the list of codon numbers\n",
    "        ASXL1_missense_codons_DICT[i+1]= (ASXL1_codons[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "        \n",
    "    for k,v in ASXL1_missense_codons_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_ASXL1_missense[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "        \n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_ASXL1_missense_extra = {}\n",
    "\n",
    "    for k, v in calls_ASXL1_missense.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_ASXL1_missense_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_ASXL1_missense_extra2 = {}\n",
    "\n",
    "    for k, v in calls_ASXL1_missense_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_ASXL1_missense_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "            \n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    ASXL1_position_in_list_dict_missense={}\n",
    "    for k, v in ASXL1_missense_variant_DICT.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_ASXL1_missense[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_ASXL1_missense[v[0]]\n",
    "        codon_seq=calls_ASXL1_missense[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_ASXL1_missense_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                ASXL1_position_in_list_dict_missense[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "            \n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in ASXL1_position_in_list_dict_missense.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "#     print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "                \n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    ASXL1_bases_missense = {}\n",
    "    for k, v in ASXL1_position_in_list_dict_missense.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        ASXL1_bases_missense[k]=(v, base_changes)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    ASXL1_bases_missense2 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_missense.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        ASXL1_bases_missense2[k] = (v, codon_positions)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    ASXL1_bases_missense3 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_missense2.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        ASXL1_bases_missense3[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    ASXL1_bases_missense4 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_missense3.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check(i)\n",
    "                sites_trio_missense.append([adapted_trio])\n",
    "                ASXL1_bases_missense4[k] = (v, sites_trio_missense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_missense = [ASXL1_exons_extra_bases[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_missense.append(site_trio_missense)\n",
    "                    ASXL1_bases_missense4[k] = (v, sites_trio_missense)\n",
    "                    \n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    ASXL1_sites_missense5 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in ASXL1_bases_missense4.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        ASXL1_sites_missense5[a]=info\n",
    "        a=a+1\n",
    "        \n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    ASXL1_sites_missense6={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in ASXL1_sites_missense5.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "    \n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        ASXL1_sites_missense6[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "        \n",
    "        \n",
    "    #Assign the mutation rates to each site...\n",
    "    ASXL1_mutation_rates_missense = {}\n",
    "    for k, v in ASXL1_sites_missense6.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            ASXL1_mutation_rates_missense[k] = (v, mutation_rate)\n",
    "\n",
    "    ASXL1_mutation_rate_missense_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in ASXL1_mutation_rates_missense.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        ASXL1_mutation_rate_missense_sum.append(mutation_rate_value)\n",
    "        \n",
    "    ASXL1_Summed_mutation_rate_missense_added = np.sum(ASXL1_mutation_rate_missense_sum)\n",
    "\n",
    "    print('Total summed mutation rate across ASXL1 missense for this study = ', ASXL1_Summed_mutation_rate_missense_added)\n",
    "    \n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all #in format: column 1 = 882, column 2 = p.R882H"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TP53"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_list_nonsense_variants_TP53(csv_file_variants):\n",
    "\n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base = [74, 75, 96, 375, 672, 782, 919, 993, 1100]\n",
    "   \n",
    "    def intron_exon_check(position_in_gene_number):\n",
    "        if position_in_gene_number == 74:\n",
    "            gene_pos = 74\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 75:\n",
    "            gene_pos = 75\n",
    "            first_base = 'G'\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = TP53_exons_extra_bases[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 96:\n",
    "            gene_pos = 96\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 375:\n",
    "            gene_pos = 375\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 672:\n",
    "            gene_pos = 672\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 782:\n",
    "            gene_pos = 782\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 919:\n",
    "            gene_pos = 919\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 993:\n",
    "            gene_pos = 993\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1100:\n",
    "            gene_pos = 1100\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #TP53 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    TP53_exons = 'ATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGACATAGTGTGGTGGTGCCCTATGAGCCGCCTGAGGTTGGCTCTGACTGTACCACCATCCACTACAACTACATGTGTAACAGTTCCTGCATGGGCGGCATGAACCGGAGGCCCATCCTCACCATCATCACACTGGAAGACTCCAGTGGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGAGGAAGAGAATCTCCGCAAGAAAGGGGAGCCTCACCACGAGCTGCCCCCAGGGAGCACTAAGCGAGCACTGCCCAACAACACCAGCTCCTCTCCCCAGCCAAAGAAGAAACCACTGGATGGAGAATATTTCACCCTTCAGATCCGTGGGCGTGAGCGCTTCGAGATGTTCCGAGAGCTGAATGAGGCCTTGGAACTCAAGGATGCCCAGGCTGGGAAGGAGCCAGGGGGGAGCAGGGCTCACTCCAGCCACCTGAAGTCCAAAAAGGGTCAGTCTACCTCCCGCCATAAAAAACTCATGTTCAAGACAGAAGGGCCTGACTCAGACTGA'\n",
    "    \n",
    "    #Split the sequence into codons \n",
    "    TP53_codons = [TP53_exons[i:i+3] for i in range(0, len(TP53_exons), 3)]\n",
    "\n",
    "    #TP53 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    TP53_exons_extra_bases = 'CATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGACATAGTGTGGTGGTGCCCTATGAGCCGCCTGAGGTTGGCTCTGACTGTACCACCATCCACTACAACTACATGTGTAACAGTTCCTGCATGGGCGGCATGAACCGGAGGCCCATCCTCACCATCATCACACTGGAAGACTCCAGTGGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGAGGAAGAGAATCTCCGCAAGAAAGGGGAGCCTCACCACGAGCTGCCCCCAGGGAGCACTAAGCGAGCACTGCCCAACAACACCAGCTCCTCTCCCCAGCCAAAGAAGAAACCACTGGATGGAGAATATTTCACCCTTCAGATCCGTGGGCGTGAGCGCTTCGAGATGTTCCGAGAGCTGAATGAGGCCTTGGAACTCAAGGATGCCCAGGCTGGGAAGGAGCCAGGGGGGAGCAGGGCTCACTCCAGCCACCTGAAGTCCAAAAAGGGTCAGTCTACCTCCCGCCATAAAAAACTCATGTTCAAGACAGAAGGGCCTGACTCAGACTGAC'\n",
    "    \n",
    "    #Create a list of the codons they called nonsense variants at\n",
    "    import csv\n",
    "    filename = csv_file_variants\n",
    "    with open(filename, 'r') as csvfile:\n",
    "        read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "        row_count=0\n",
    "\n",
    "        TP53_nonsense_variant_list = []\n",
    "        TP53_nonsense_codons =[]\n",
    "        TP53_nonsense_variant_DICT = {}\n",
    "\n",
    "        for row in read_reader: #each row contains aa number (e.g. 11) and variant (e.g. p.D11A)\n",
    "    #         print(row)\n",
    "            if row_count>0: \n",
    "                codon0 = int(row[0])-1\n",
    "                codon1 = int(row[0])\n",
    "                variant = row[1]\n",
    "                TP53_nonsense_codons.append(codon0) #List of codon numbers that they called variants at\n",
    "                TP53_nonsense_variant_DICT[row_count]=(codon1, variant) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "            row_count=row_count+1\n",
    "            \n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    TP53_nonsense_codons_DICT = {}\n",
    "    calls_nonsense = {}\n",
    "\n",
    "    for i in TP53_nonsense_codons: #Iterates over the list of codon numbers\n",
    "        TP53_nonsense_codons_DICT[i+1]= (TP53_codons[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "        \n",
    "    for k,v in TP53_nonsense_codons_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_nonsense[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "        \n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_nonsense_extra = {}\n",
    "\n",
    "    for k, v in calls_nonsense.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_nonsense_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_nonsense_extra2 = {}\n",
    "\n",
    "    for k, v in calls_nonsense_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_nonsense_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "            \n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    TP53_position_in_list_dict_nonsense={}\n",
    "    for k, v in TP53_nonsense_variant_DICT.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_nonsense[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_nonsense[v[0]]\n",
    "        codon_seq=calls_nonsense[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_nonsense_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                TP53_position_in_list_dict_nonsense[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "            \n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in TP53_position_in_list_dict_nonsense.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "#     print('positions at which nonsense variants called = ', aa_positions_variants_called)\n",
    "                \n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TP53_bases_nonsense = {}\n",
    "    for k, v in TP53_position_in_list_dict_nonsense.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TP53_bases_nonsense[k]=(v, base_changes)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TP53_bases_nonsense2 = {}\n",
    "\n",
    "    for k, v in TP53_bases_nonsense.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TP53_bases_nonsense2[k] = (v, codon_positions)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TP53_bases_nonsense3 = {}\n",
    "\n",
    "    for k, v in TP53_bases_nonsense2.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TP53_bases_nonsense3[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TP53_bases_nonsense4 = {}\n",
    "\n",
    "    for k, v in TP53_bases_nonsense3.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_nonsense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check(i)\n",
    "                sites_trio_nonsense.append([adapted_trio])\n",
    "                TP53_bases_nonsense4[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_nonsense = [TP53_exons_extra_bases[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_nonsense.append(site_trio_nonsense)\n",
    "                    TP53_bases_nonsense4[k] = (v, sites_trio_nonsense)\n",
    "                    \n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TP53_sites_nonsense5 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TP53_bases_nonsense4.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        TP53_sites_nonsense5[a]=info\n",
    "        a=a+1\n",
    "#     print(TP53_sites_nonsense5)\n",
    "        \n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TP53_sites_nonsense6={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TP53_sites_nonsense5.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "    \n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TP53_sites_nonsense6[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "        \n",
    "        \n",
    "    #Assign the mutation rates to each site...\n",
    "    TP53_mutation_rates_nonsense = {}\n",
    "    for k, v in TP53_sites_nonsense6.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            TP53_mutation_rates_nonsense[k] = (v, mutation_rate)\n",
    "\n",
    "    TP53_mutation_rate_nonsense_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TP53_mutation_rates_nonsense.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        TP53_mutation_rate_nonsense_sum.append(mutation_rate_value)\n",
    "    \n",
    "    TP53_Summed_mutation_rate_nonsense_added = np.sum(TP53_mutation_rate_nonsense_sum)\n",
    "\n",
    "    print('Total summed mutation rate across TP53 nonsense for this study = ', TP53_Summed_mutation_rate_nonsense_added)\n",
    "    \n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all #in format: column 1 = 882, column 2 = p.R882*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_list_missense_variants_TP53(csv_file_variants):\n",
    "    \n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base_TP53 = [74, 75, 96, 375, 672, 782, 919, 993, 1100]\n",
    "   \n",
    "    def intron_exon_check(position_in_gene_number):\n",
    "        if position_in_gene_number == 74:\n",
    "            gene_pos = 74\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 75:\n",
    "            gene_pos = 75\n",
    "            first_base = 'G'\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = TP53_exons_extra_bases[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 96:\n",
    "            gene_pos = 96\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 375:\n",
    "            gene_pos = 375\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 672:\n",
    "            gene_pos = 672\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 782:\n",
    "            gene_pos = 782\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 919:\n",
    "            gene_pos = 919\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 993:\n",
    "            gene_pos = 993\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1100:\n",
    "            gene_pos = 1100\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #TP53 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    TP53_exons = 'ATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGACATAGTGTGGTGGTGCCCTATGAGCCGCCTGAGGTTGGCTCTGACTGTACCACCATCCACTACAACTACATGTGTAACAGTTCCTGCATGGGCGGCATGAACCGGAGGCCCATCCTCACCATCATCACACTGGAAGACTCCAGTGGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGAGGAAGAGAATCTCCGCAAGAAAGGGGAGCCTCACCACGAGCTGCCCCCAGGGAGCACTAAGCGAGCACTGCCCAACAACACCAGCTCCTCTCCCCAGCCAAAGAAGAAACCACTGGATGGAGAATATTTCACCCTTCAGATCCGTGGGCGTGAGCGCTTCGAGATGTTCCGAGAGCTGAATGAGGCCTTGGAACTCAAGGATGCCCAGGCTGGGAAGGAGCCAGGGGGGAGCAGGGCTCACTCCAGCCACCTGAAGTCCAAAAAGGGTCAGTCTACCTCCCGCCATAAAAAACTCATGTTCAAGACAGAAGGGCCTGACTCAGACTGA'\n",
    "    \n",
    "    #Split the sequence into codons \n",
    "    TP53_codons = [TP53_exons[i:i+3] for i in range(0, len(TP53_exons), 3)]\n",
    "\n",
    "    #TP53 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    TP53_exons_extra_bases = 'CATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGACATAGTGTGGTGGTGCCCTATGAGCCGCCTGAGGTTGGCTCTGACTGTACCACCATCCACTACAACTACATGTGTAACAGTTCCTGCATGGGCGGCATGAACCGGAGGCCCATCCTCACCATCATCACACTGGAAGACTCCAGTGGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGAGGAAGAGAATCTCCGCAAGAAAGGGGAGCCTCACCACGAGCTGCCCCCAGGGAGCACTAAGCGAGCACTGCCCAACAACACCAGCTCCTCTCCCCAGCCAAAGAAGAAACCACTGGATGGAGAATATTTCACCCTTCAGATCCGTGGGCGTGAGCGCTTCGAGATGTTCCGAGAGCTGAATGAGGCCTTGGAACTCAAGGATGCCCAGGCTGGGAAGGAGCCAGGGGGGAGCAGGGCTCACTCCAGCCACCTGAAGTCCAAAAAGGGTCAGTCTACCTCCCGCCATAAAAAACTCATGTTCAAGACAGAAGGGCCTGACTCAGACTGAC'\n",
    "    \n",
    "    #Create a list of the codons they called missense variants at\n",
    "    import csv\n",
    "    filename = csv_file_variants\n",
    "    with open(filename, 'r') as csvfile:\n",
    "        read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "        row_count=0\n",
    "\n",
    "        TP53_missense_variant_list = []\n",
    "        TP53_missense_codons =[]\n",
    "        TP53_missense_variant_DICT = {}\n",
    "\n",
    "        for row in read_reader: #each row contains aa number (e.g. 11) and variant (e.g. p.D11A)\n",
    "    #         print(row)\n",
    "            if row_count>0: \n",
    "                codon0 = int(row[0])-1\n",
    "                codon1 = int(row[0])\n",
    "                variant = row[1]\n",
    "                TP53_missense_codons.append(codon0) #List of codon numbers that they called variants at\n",
    "                TP53_missense_variant_DICT[row_count]=(codon1, variant) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "            row_count=row_count+1\n",
    "            \n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    TP53_missense_codons_DICT = {}\n",
    "    calls_TP53_missense = {}\n",
    "\n",
    "    for i in TP53_missense_codons: #Iterates over the list of codon numbers\n",
    "        TP53_missense_codons_DICT[i+1]= (TP53_codons[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "        \n",
    "    for k,v in TP53_missense_codons_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_TP53_missense[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "        \n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_TP53_missense_extra = {}\n",
    "\n",
    "    for k, v in calls_TP53_missense.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_TP53_missense_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_TP53_missense_extra2 = {}\n",
    "\n",
    "    for k, v in calls_TP53_missense_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_TP53_missense_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "            \n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    TP53_position_in_list_dict_missense={}\n",
    "    for k, v in TP53_missense_variant_DICT.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_TP53_missense[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_TP53_missense[v[0]]\n",
    "        codon_seq=calls_TP53_missense[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_TP53_missense_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                TP53_position_in_list_dict_missense[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "            \n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in TP53_position_in_list_dict_missense.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "#     print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "                \n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TP53_bases_missense = {}\n",
    "    for k, v in TP53_position_in_list_dict_missense.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TP53_bases_missense[k]=(v, base_changes)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TP53_bases_missense2 = {}\n",
    "\n",
    "    for k, v in TP53_bases_missense.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TP53_bases_missense2[k] = (v, codon_positions)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TP53_bases_missense3 = {}\n",
    "\n",
    "    for k, v in TP53_bases_missense2.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TP53_bases_missense3[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TP53_bases_missense4 = {}\n",
    "\n",
    "    for k, v in TP53_bases_missense3.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_TP53:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check(i)\n",
    "                sites_trio_missense.append([adapted_trio])\n",
    "                TP53_bases_missense4[k] = (v, sites_trio_missense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_missense = [TP53_exons_extra_bases[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_missense.append(site_trio_missense)\n",
    "                    TP53_bases_missense4[k] = (v, sites_trio_missense)\n",
    "                    \n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TP53_sites_missense5 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TP53_bases_missense4.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        TP53_sites_missense5[a]=info\n",
    "        a=a+1\n",
    "        \n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TP53_sites_missense6={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TP53_sites_missense5.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "    \n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TP53_sites_missense6[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "        \n",
    "        \n",
    "    #Assign the mutation rates to each site...\n",
    "    TP53_mutation_rates_missense = {}\n",
    "    for k, v in TP53_sites_missense6.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            TP53_mutation_rates_missense[k] = (v, mutation_rate)\n",
    "\n",
    "    TP53_mutation_rate_missense_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TP53_mutation_rates_missense.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        TP53_mutation_rate_missense_sum.append(mutation_rate_value)\n",
    "        \n",
    "    TP53_Summed_mutation_rate_missense_added = np.sum(TP53_mutation_rate_missense_sum)\n",
    "\n",
    "    print('Total summed mutation rate across TP53 missense for this study = ', TP53_Summed_mutation_rate_missense_added)\n",
    "    \n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all #in format: column 1 = 882, column 2 = p.R882H"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculating mutation rates from a list of variants"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DNMT3A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(lower_range, upper_range):\n",
    "    \n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base_DNMT3A_NP072046 = [1279, 1431, 1476, 1554, 1852, 2083, 2085, 2175, 2322, 2478]\n",
    "\n",
    "    def intron_exon_check_DNMT3A_NP072046(position_in_gene_number):\n",
    "        if position_in_gene_number == 1279:\n",
    "            gene_pos = 1279\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1429:\n",
    "            gene_pos = 1429\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1474:\n",
    "            gene_pos = 1474\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1554:\n",
    "            gene_pos = 1554\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1852:\n",
    "            gene_pos = 1852\n",
    "            first_base = 'G'\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2083:\n",
    "            gene_pos = 2083\n",
    "            first_base = 'G'\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2082:\n",
    "            gene_pos = 2082\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2173:\n",
    "            gene_pos = 2173\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2322:\n",
    "            gene_pos = 2322\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2478:\n",
    "            gene_pos = 2478\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "\n",
    "    #DNMT3A transcript used:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000119772;r=2:25455845-25564754;t=ENST00000321117\n",
    "    DNMT3A_exons_NP_072046 = 'ATGCCCGCCATGCCCTCCAGCGGCCCCGGGGACACCAGCAGCTCTGCTGCGGAGCGGGAGGAGGACCGAAAGGACGGAGAGGAGCAGGAGGAGCCGCGTGGCAAGGAGGAGCGCCAAGAGCCCAGCACCACGGCACGGAAGGTGGGGCGGCCTGGGAGGAAGCGCAAGCACCCCCCGGTGGAAAGCGGTGACACGCCAAAGGACCCTGCGGTGATCTCCAAGTCCCCATCCATGGCCCAGGACTCAGGCGCCTCAGAGCTATTACCCAATGGGGACTTGGAGAAGCGGAGTGAGCCCCAGCCAGAGGAGGGGAGCCCTGCTGGGGGGCAGAAGGGCGGGGCCCCAGCAGAGGGAGAGGGTGCAGCTGAGACCCTGCCTGAAGCCTCAAGAGCAGTGGAAAATGGCTGCTGCACCCCCAAGGAGGGCCGAGGAGCCCCTGCAGAAGCGGGCAAAGAACAGAAGGAGACCAACATCGAATCCATGAAAATGGAGGGCTCCCGGGGCCGGCTGCGGGGTGGCTTGGGCTGGGAGTCCAGCCTCCGTCAGCGGCCCATGCCGAGGCTCACCTTCCAGGCGGGGGACCCCTACTACATCAGCAAGCGCAAGCGGGACGAGTGGCTGGCACGCTGGAAAAGGGAGGCTGAGAAGAAAGCCAAGGTCATTGCAGGAATGAATGCTGTGGAAGAAAACCAGGGGCCCGGGGAGTCTCAGAAGGTGGAGGAGGCCAGCCCTCCTGCTGTGCAGCAGCCCACTGACCCCGCATCCCCCACTGTGGCTACCACGCCTGAGCCCGTGGGGTCCGATGCTGGGGACAAGAATGCCACCAAAGCAGGCGATGACGAGCCAGAGTACGAGGACGGCCGGGGCTTTGGCATTGGGGAGCTGGTGTGGGGGAAACTGCGGGGCTTCTCCTGGTGGCCAGGCCGCATTGTGTCTTGGTGGATGACGGGCCGGAGCCGAGCAGCTGAAGGCACCCGCTGGGTCATGTGGTTCGGAGACGGCAAATTCTCAGTGGTGTGTGTTGAGAAGCTGATGCCGCTGAGCTCGTTTTGCAGTGCGTTCCACCAGGCCACGTACAACAAGCAGCCCATGTACCGCAAAGCCATCTACGAGGTCCTGCAGGTGGCCAGCAGCCGCGCGGGGAAGCTGTTCCCGGTGTGCCACGACAGCGATGAGAGTGACACTGCCAAGGCCGTGGAGGTGCAGAACAAGCCCATGATTGAATGGGCCCTGGGGGGCTTCCAGCCTTCTGGCCCTAAGGGCCTGGAGCCACCAGAAGAAGAGAAGAATCCCTACAAAGAAGTGTACACGGACATGTGGGTGGAACCTGAGGCAGCTGCCTACGCACCACCTCCACCAGCCAAAAAGCCCCGGAAGAGCACAGCGGAGAAGCCCAAGGTCAAGGAGATTATTGATGAGCGCACAAGAGAGCGGCTGGTGTACGAGGTGCGGCAGAAGTGCCGGAACATTGAGGACATCTGCATCTCCTGTGGGAGCCTCAATGTTACCCTGGAACACCCCCTCTTCGTTGGAGGAATGTGCCAAAACTGCAAGAACTGCTTTCTGGAGTGTGCGTACCAGTACGACGACGACGGCTACCAGTCCTACTGCACCATCTGCTGTGGGGGCCGTGAGGTGCTCATGTGCGGAAACAACAACTGCTGCAGGTGCTTTTGCGTGGAGTGTGTGGACCTCTTGGTGGGGCCGGGGGCTGCCCAGGCAGCCATTAAGGAAGACCCCTGGAACTGCTACATGTGCGGGCACAAGGGTACCTACGGGCTGCTGCGGCGGCGAGAGGACTGGCCCTCCCGGCTCCAGATGTTCTTCGCTAATAACCACGACCAGGAATTTGACCCTCCAAAGGTTTACCCACCTGTCCCAGCTGAGAAGAGGAAGCCCATCCGGGTGCTGTCTCTCTTTGATGGAATCGCTACAGGGCTCCTGGTGCTGAAGGACTTGGGCATTCAGGTGGACCGCTACATTGCCTCGGAGGTGTGTGAGGACTCCATCACGGTGGGCATGGTGCGGCACCAGGGGAAGATCATGTACGTCGGGGACGTCCGCAGCGTCACACAGAAGCATATCCAGGAGTGGGGCCCATTCGATCTGGTGATTGGGGGCAGTCCCTGCAATGACCTCTCCATCGTCAACCCTGCTCGCAAGGGCCTCTACGAGGGCACTGGCCGGCTCTTCTTTGAGTTCTACCGCCTCCTGCATGATGCGCGGCCCAAGGAGGGAGATGATCGCCCCTTCTTCTGGCTCTTTGAGAATGTGGTGGCCATGGGCGTTAGTGACAAGAGGGACATCTCGCGATTTCTCGAGTCCAACCCTGTGATGATTGATGCCAAAGAAGTGTCAGCTGCACACAGGGCCCGCTACTTCTGGGGTAACCTTCCCGGTATGAACAGGCCGTTGGCATCCACTGTGAATGATAAGCTGGAGCTGCAGGAGTGTCTGGAGCATGGCAGGATAGCCAAGTTCAGCAAAGTGAGGACCATTACTACGAGGTCAAACTCCATAAAGCAGGGCAAAGACCAGCATTTTCCTGTCTTCATGAATGAGAAAGAGGACATCTTATGGTGCACTGAAATGGAAAGGGTATTTGGTTTCCCAGTCCACTATACTGACGTCTCCAACATGAGCCGCTTGGCGAGGCAGAGACTGCTGGGCCGGTCATGGAGCGTGCCAGTCATCCGCCACCTCTTCGCTCCGCTGAAGGAGTATTTTGCGTGTGTGTAA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    DNMT3A_codons_NP_072046 = [DNMT3A_exons_NP_072046[i:i+3] for i in range(0, len(DNMT3A_exons_NP_072046), 3)]\n",
    "\n",
    "    #DNMT3A transcript used:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000119772;r=2:25455845-25564754;t=ENST00000321117\n",
    "    DNMT3A_exons_extra_bases_NP_072046 = 'GATGCCCGCCATGCCCTCCAGCGGCCCCGGGGACACCAGCAGCTCTGCTGCGGAGCGGGAGGAGGACCGAAAGGACGGAGAGGAGCAGGAGGAGCCGCGTGGCAAGGAGGAGCGCCAAGAGCCCAGCACCACGGCACGGAAGGTGGGGCGGCCTGGGAGGAAGCGCAAGCACCCCCCGGTGGAAAGCGGTGACACGCCAAAGGACCCTGCGGTGATCTCCAAGTCCCCATCCATGGCCCAGGACTCAGGCGCCTCAGAGCTATTACCCAATGGGGACTTGGAGAAGCGGAGTGAGCCCCAGCCAGAGGAGGGGAGCCCTGCTGGGGGGCAGAAGGGCGGGGCCCCAGCAGAGGGAGAGGGTGCAGCTGAGACCCTGCCTGAAGCCTCAAGAGCAGTGGAAAATGGCTGCTGCACCCCCAAGGAGGGCCGAGGAGCCCCTGCAGAAGCGGGCAAAGAACAGAAGGAGACCAACATCGAATCCATGAAAATGGAGGGCTCCCGGGGCCGGCTGCGGGGTGGCTTGGGCTGGGAGTCCAGCCTCCGTCAGCGGCCCATGCCGAGGCTCACCTTCCAGGCGGGGGACCCCTACTACATCAGCAAGCGCAAGCGGGACGAGTGGCTGGCACGCTGGAAAAGGGAGGCTGAGAAGAAAGCCAAGGTCATTGCAGGAATGAATGCTGTGGAAGAAAACCAGGGGCCCGGGGAGTCTCAGAAGGTGGAGGAGGCCAGCCCTCCTGCTGTGCAGCAGCCCACTGACCCCGCATCCCCCACTGTGGCTACCACGCCTGAGCCCGTGGGGTCCGATGCTGGGGACAAGAATGCCACCAAAGCAGGCGATGACGAGCCAGAGTACGAGGACGGCCGGGGCTTTGGCATTGGGGAGCTGGTGTGGGGGAAACTGCGGGGCTTCTCCTGGTGGCCAGGCCGCATTGTGTCTTGGTGGATGACGGGCCGGAGCCGAGCAGCTGAAGGCACCCGCTGGGTCATGTGGTTCGGAGACGGCAAATTCTCAGTGGTGTGTGTTGAGAAGCTGATGCCGCTGAGCTCGTTTTGCAGTGCGTTCCACCAGGCCACGTACAACAAGCAGCCCATGTACCGCAAAGCCATCTACGAGGTCCTGCAGGTGGCCAGCAGCCGCGCGGGGAAGCTGTTCCCGGTGTGCCACGACAGCGATGAGAGTGACACTGCCAAGGCCGTGGAGGTGCAGAACAAGCCCATGATTGAATGGGCCCTGGGGGGCTTCCAGCCTTCTGGCCCTAAGGGCCTGGAGCCACCAGAAGAAGAGAAGAATCCCTACAAAGAAGTGTACACGGACATGTGGGTGGAACCTGAGGCAGCTGCCTACGCACCACCTCCACCAGCCAAAAAGCCCCGGAAGAGCACAGCGGAGAAGCCCAAGGTCAAGGAGATTATTGATGAGCGCACAAGAGAGCGGCTGGTGTACGAGGTGCGGCAGAAGTGCCGGAACATTGAGGACATCTGCATCTCCTGTGGGAGCCTCAATGTTACCCTGGAACACCCCCTCTTCGTTGGAGGAATGTGCCAAAACTGCAAGAACTGCTTTCTGGAGTGTGCGTACCAGTACGACGACGACGGCTACCAGTCCTACTGCACCATCTGCTGTGGGGGCCGTGAGGTGCTCATGTGCGGAAACAACAACTGCTGCAGGTGCTTTTGCGTGGAGTGTGTGGACCTCTTGGTGGGGCCGGGGGCTGCCCAGGCAGCCATTAAGGAAGACCCCTGGAACTGCTACATGTGCGGGCACAAGGGTACCTACGGGCTGCTGCGGCGGCGAGAGGACTGGCCCTCCCGGCTCCAGATGTTCTTCGCTAATAACCACGACCAGGAATTTGACCCTCCAAAGGTTTACCCACCTGTCCCAGCTGAGAAGAGGAAGCCCATCCGGGTGCTGTCTCTCTTTGATGGAATCGCTACAGGGCTCCTGGTGCTGAAGGACTTGGGCATTCAGGTGGACCGCTACATTGCCTCGGAGGTGTGTGAGGACTCCATCACGGTGGGCATGGTGCGGCACCAGGGGAAGATCATGTACGTCGGGGACGTCCGCAGCGTCACACAGAAGCATATCCAGGAGTGGGGCCCATTCGATCTGGTGATTGGGGGCAGTCCCTGCAATGACCTCTCCATCGTCAACCCTGCTCGCAAGGGCCTCTACGAGGGCACTGGCCGGCTCTTCTTTGAGTTCTACCGCCTCCTGCATGATGCGCGGCCCAAGGAGGGAGATGATCGCCCCTTCTTCTGGCTCTTTGAGAATGTGGTGGCCATGGGCGTTAGTGACAAGAGGGACATCTCGCGATTTCTCGAGTCCAACCCTGTGATGATTGATGCCAAAGAAGTGTCAGCTGCACACAGGGCCCGCTACTTCTGGGGTAACCTTCCCGGTATGAACAGGCCGTTGGCATCCACTGTGAATGATAAGCTGGAGCTGCAGGAGTGTCTGGAGCATGGCAGGATAGCCAAGTTCAGCAAAGTGAGGACCATTACTACGAGGTCAAACTCCATAAAGCAGGGCAAAGACCAGCATTTTCCTGTCTTCATGAATGAGAAAGAGGACATCTTATGGTGCACTGAAATGGAAAGGGTATTTGGTTTCCCAGTCCACTATACTGACGTCTCCAACATGAGCCGCTTGGCGAGGCAGAGACTGCTGGGCCGGTCATGGAGCGTGCCAGTCATCCGCCACCTCTTCGCTCCGCTGAAGGAGTATTTTGCGTGTGTGTAAG'\n",
    "\n",
    "    #List of all amino acid positions (across whole coding region) covered (amino acid position minus 1 (as python starts at 0))\n",
    "    DNMT3A_nonsense_codons_DICT_NP_072046={}\n",
    "    for i in range(lower_range, upper_range):\n",
    "        DNMT3A_nonsense_codons_DICT_NP_072046[i+1]= (DNMT3A_codons_NP_072046[i])\n",
    "\n",
    "    calls_DNMT3A_nonsense_NP_072046 = {}\n",
    "    for k,v in DNMT3A_nonsense_codons_DICT_NP_072046.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_DNMT3A_nonsense_NP_072046[k]=(v, changes)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_DNMT3A_nonsense_NP_072046_extra = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_nonsense_NP_072046.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_DNMT3A_nonsense_NP_072046_extra[k] = (Amino_acid_change, v[0], v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_DNMT3A_nonsense_NP_072046_extra2 = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_nonsense_NP_072046_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_DNMT3A_nonsense_NP_072046_extra2[k] = (Amino_acid_abbrev, v[1], v[2])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary that contains details of the positions of nonsense variants at that codon\n",
    "    DNMT3A_position_in_list_dict_nonsense_NP_072046 = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_nonsense_NP_072046_extra2.items():\n",
    "        original_AA = v[0]\n",
    "        possible_AA_changes = v[2]\n",
    "        possible_AA_changes_string = ''.join(v[2])\n",
    "        positions_in_list = []\n",
    "        for i in possible_AA_changes:\n",
    "            if i == '*': #nonsense\n",
    "                position_in_list=-1\n",
    "                nonsense_change = i    \n",
    "                while True:\n",
    "                    position_in_list = possible_AA_changes_string.find(nonsense_change, position_in_list+1)\n",
    "    #               print(position_in_list)\n",
    "                    if position_in_list == -1: break\n",
    "                    positions_in_list.append(position_in_list)\n",
    "                    positions_in_list = set(positions_in_list) #convert into a 'set' to remove duplicate positions\n",
    "                    positions_in_list = list(positions_in_list) #convert back to a list so can be added to dictionary\n",
    "                DNMT3A_position_in_list_dict_nonsense_NP_072046[k]=(positions_in_list, v)\n",
    "\n",
    "        aa_positions_variants_called = []\n",
    "        for k, v in DNMT3A_position_in_list_dict_nonsense_NP_072046.items():\n",
    "            aa_positions_variants_called.append(k)\n",
    "    # print('positions at which nonsense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    DNMT3A_bases_nonsense_NP_072046 = {}\n",
    "    for k, v in DNMT3A_position_in_list_dict_nonsense_NP_072046.items():\n",
    "        original_a_a = v[1][0]\n",
    "        codon_seq = v[1][1]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        DNMT3A_bases_nonsense_NP_072046[k]=(v, base_changes, original_a_a)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    DNMT3A_bases_nonsense2_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_nonsense_NP_072046.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        DNMT3A_bases_nonsense2_NP_072046[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    DNMT3A_bases_nonsense3_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_nonsense2_NP_072046.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_a_a = v[0][2]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = k\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(k)-1)*3)+codon_pos #codon number minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][1][1][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        DNMT3A_bases_nonsense3_NP_072046[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, original_a_a)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    DNMT3A_bases_nonsense4_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_nonsense3_NP_072046.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_nonsense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_DNMT3A_NP072046:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check_DNMT3A_NP072046(i)\n",
    "                sites_trio_nonsense.append([adapted_trio])\n",
    "                DNMT3A_bases_nonsense4_NP_072046[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_nonsense = [DNMT3A_exons_extra_bases_NP_072046[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_nonsense.append(site_trio_nonsense)\n",
    "                    DNMT3A_bases_nonsense4_NP_072046[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    DNMT3A_sites_nonsense5_NP_072046 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in DNMT3A_bases_nonsense4_NP_072046.items():\n",
    "        original_a_a = v[0][0][0]\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, original_a_a])\n",
    "            change = change+1\n",
    "        DNMT3A_sites_nonsense5_NP_072046[a]=info\n",
    "        a=a+1\n",
    "    #     print(DNMT3A_sites_nonsense5_NP_072046)\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    DNMT3A_sites_nonsense6_NP_072046={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in DNMT3A_sites_nonsense5_NP_072046.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        original_a_a = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        DNMT3A_sites_nonsense6_NP_072046[k] = (v[0],changes3, codon_n_912, original_a_a) #site_trio = the affected base and its neighbours\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    DNMT3A_mutation_rates_nonsense_NP_072046 = {}\n",
    "    for k, v in DNMT3A_sites_nonsense6_NP_072046.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            original_a_a = v[3]\n",
    "            aa_positioning = v[2]\n",
    "            variant_type = '*'\n",
    "            variant_id = str(original_a_a)+str(aa_positioning)+variant_type\n",
    "            DNMT3A_mutation_rates_nonsense_NP_072046[k] = (v[0],v[1], mutation_rate, variant_id)\n",
    "\n",
    "    DNMT3A_mutation_rate_nonsense_NP_072046_sum = []\n",
    "\n",
    "    mutation_rates_list = []\n",
    "    for k, v in DNMT3A_mutation_rates_nonsense_NP_072046.items():\n",
    "        codon_position_in_gene=(v[0])\n",
    "        variant = v[3]\n",
    "        site = v[1]\n",
    "        mutation_rate_value = v[2] \n",
    "        mutation_rates_list.append((v[0], variant, site, mutation_rate_value))\n",
    "        DNMT3A_mutation_rate_nonsense_NP_072046_sum.append(mutation_rate_value)\n",
    "\n",
    "    #Get the average of all the mutation rates\n",
    "    DNMT3A_mutation_rate_nonsense_NP_072046_sum = []\n",
    "    for k, v in DNMT3A_mutation_rates_nonsense_NP_072046.items():\n",
    "        mut_rate = v[2]\n",
    "        DNMT3A_mutation_rate_nonsense_NP_072046_sum.append(mut_rate)\n",
    "\n",
    "    DNMT3A_Summed_mutation_rate_nonsense_NP_072046_added = np.sum(DNMT3A_mutation_rate_nonsense_NP_072046_sum)\n",
    "\n",
    "    print('Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study = ', DNMT3A_Summed_mutation_rate_nonsense_NP_072046_added)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all # e.g. lower= 0, upper = 913 for whole gene\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_missense_range_of_variants_DNMT3A_NP072046(lower_range, upper_range):\n",
    "    intron_exon_boundary_caution_base_DNMT3A_NP072046 = [1279, 1429, 1474, 1554, 1852, 2083, 2082, 2173, 2322, 2478]\n",
    "\n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    def intron_exon_check_DNMT3A_NP072046(position_in_gene_number):\n",
    "        if position_in_gene_number == 1279:\n",
    "            gene_pos = 1279\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1429:\n",
    "            gene_pos = 1429\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1474:\n",
    "            gene_pos = 1474\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1554:\n",
    "            gene_pos = 1554\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1852:\n",
    "            gene_pos = 1852\n",
    "            first_base = 'G'\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2083:\n",
    "            gene_pos = 2083\n",
    "            first_base = 'G'\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2082:\n",
    "            gene_pos = 2082\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2173:\n",
    "            gene_pos = 2173\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2322:\n",
    "            gene_pos = 2322\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2478:\n",
    "            gene_pos = 2478\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "\n",
    "    #DNMT3A transcript used (NP_072046):\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000119772;r=2:25455845-25564754;t=ENST00000321117\n",
    "    DNMT3A_exons_NP_072046 = 'ATGCCCGCCATGCCCTCCAGCGGCCCCGGGGACACCAGCAGCTCTGCTGCGGAGCGGGAGGAGGACCGAAAGGACGGAGAGGAGCAGGAGGAGCCGCGTGGCAAGGAGGAGCGCCAAGAGCCCAGCACCACGGCACGGAAGGTGGGGCGGCCTGGGAGGAAGCGCAAGCACCCCCCGGTGGAAAGCGGTGACACGCCAAAGGACCCTGCGGTGATCTCCAAGTCCCCATCCATGGCCCAGGACTCAGGCGCCTCAGAGCTATTACCCAATGGGGACTTGGAGAAGCGGAGTGAGCCCCAGCCAGAGGAGGGGAGCCCTGCTGGGGGGCAGAAGGGCGGGGCCCCAGCAGAGGGAGAGGGTGCAGCTGAGACCCTGCCTGAAGCCTCAAGAGCAGTGGAAAATGGCTGCTGCACCCCCAAGGAGGGCCGAGGAGCCCCTGCAGAAGCGGGCAAAGAACAGAAGGAGACCAACATCGAATCCATGAAAATGGAGGGCTCCCGGGGCCGGCTGCGGGGTGGCTTGGGCTGGGAGTCCAGCCTCCGTCAGCGGCCCATGCCGAGGCTCACCTTCCAGGCGGGGGACCCCTACTACATCAGCAAGCGCAAGCGGGACGAGTGGCTGGCACGCTGGAAAAGGGAGGCTGAGAAGAAAGCCAAGGTCATTGCAGGAATGAATGCTGTGGAAGAAAACCAGGGGCCCGGGGAGTCTCAGAAGGTGGAGGAGGCCAGCCCTCCTGCTGTGCAGCAGCCCACTGACCCCGCATCCCCCACTGTGGCTACCACGCCTGAGCCCGTGGGGTCCGATGCTGGGGACAAGAATGCCACCAAAGCAGGCGATGACGAGCCAGAGTACGAGGACGGCCGGGGCTTTGGCATTGGGGAGCTGGTGTGGGGGAAACTGCGGGGCTTCTCCTGGTGGCCAGGCCGCATTGTGTCTTGGTGGATGACGGGCCGGAGCCGAGCAGCTGAAGGCACCCGCTGGGTCATGTGGTTCGGAGACGGCAAATTCTCAGTGGTGTGTGTTGAGAAGCTGATGCCGCTGAGCTCGTTTTGCAGTGCGTTCCACCAGGCCACGTACAACAAGCAGCCCATGTACCGCAAAGCCATCTACGAGGTCCTGCAGGTGGCCAGCAGCCGCGCGGGGAAGCTGTTCCCGGTGTGCCACGACAGCGATGAGAGTGACACTGCCAAGGCCGTGGAGGTGCAGAACAAGCCCATGATTGAATGGGCCCTGGGGGGCTTCCAGCCTTCTGGCCCTAAGGGCCTGGAGCCACCAGAAGAAGAGAAGAATCCCTACAAAGAAGTGTACACGGACATGTGGGTGGAACCTGAGGCAGCTGCCTACGCACCACCTCCACCAGCCAAAAAGCCCCGGAAGAGCACAGCGGAGAAGCCCAAGGTCAAGGAGATTATTGATGAGCGCACAAGAGAGCGGCTGGTGTACGAGGTGCGGCAGAAGTGCCGGAACATTGAGGACATCTGCATCTCCTGTGGGAGCCTCAATGTTACCCTGGAACACCCCCTCTTCGTTGGAGGAATGTGCCAAAACTGCAAGAACTGCTTTCTGGAGTGTGCGTACCAGTACGACGACGACGGCTACCAGTCCTACTGCACCATCTGCTGTGGGGGCCGTGAGGTGCTCATGTGCGGAAACAACAACTGCTGCAGGTGCTTTTGCGTGGAGTGTGTGGACCTCTTGGTGGGGCCGGGGGCTGCCCAGGCAGCCATTAAGGAAGACCCCTGGAACTGCTACATGTGCGGGCACAAGGGTACCTACGGGCTGCTGCGGCGGCGAGAGGACTGGCCCTCCCGGCTCCAGATGTTCTTCGCTAATAACCACGACCAGGAATTTGACCCTCCAAAGGTTTACCCACCTGTCCCAGCTGAGAAGAGGAAGCCCATCCGGGTGCTGTCTCTCTTTGATGGAATCGCTACAGGGCTCCTGGTGCTGAAGGACTTGGGCATTCAGGTGGACCGCTACATTGCCTCGGAGGTGTGTGAGGACTCCATCACGGTGGGCATGGTGCGGCACCAGGGGAAGATCATGTACGTCGGGGACGTCCGCAGCGTCACACAGAAGCATATCCAGGAGTGGGGCCCATTCGATCTGGTGATTGGGGGCAGTCCCTGCAATGACCTCTCCATCGTCAACCCTGCTCGCAAGGGCCTCTACGAGGGCACTGGCCGGCTCTTCTTTGAGTTCTACCGCCTCCTGCATGATGCGCGGCCCAAGGAGGGAGATGATCGCCCCTTCTTCTGGCTCTTTGAGAATGTGGTGGCCATGGGCGTTAGTGACAAGAGGGACATCTCGCGATTTCTCGAGTCCAACCCTGTGATGATTGATGCCAAAGAAGTGTCAGCTGCACACAGGGCCCGCTACTTCTGGGGTAACCTTCCCGGTATGAACAGGCCGTTGGCATCCACTGTGAATGATAAGCTGGAGCTGCAGGAGTGTCTGGAGCATGGCAGGATAGCCAAGTTCAGCAAAGTGAGGACCATTACTACGAGGTCAAACTCCATAAAGCAGGGCAAAGACCAGCATTTTCCTGTCTTCATGAATGAGAAAGAGGACATCTTATGGTGCACTGAAATGGAAAGGGTATTTGGTTTCCCAGTCCACTATACTGACGTCTCCAACATGAGCCGCTTGGCGAGGCAGAGACTGCTGGGCCGGTCATGGAGCGTGCCAGTCATCCGCCACCTCTTCGCTCCGCTGAAGGAGTATTTTGCGTGTGTGTAA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    DNMT3A_codons_NP_072046 = [DNMT3A_exons_NP_072046[i:i+3] for i in range(0, len(DNMT3A_exons_NP_072046), 3)]\n",
    "\n",
    "    #DNMT3A transcript used (NP_072046):\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000119772;r=2:25455845-25564754;t=ENST00000321117\n",
    "    DNMT3A_exons_extra_bases_NP_072046 = 'GATGCCCGCCATGCCCTCCAGCGGCCCCGGGGACACCAGCAGCTCTGCTGCGGAGCGGGAGGAGGACCGAAAGGACGGAGAGGAGCAGGAGGAGCCGCGTGGCAAGGAGGAGCGCCAAGAGCCCAGCACCACGGCACGGAAGGTGGGGCGGCCTGGGAGGAAGCGCAAGCACCCCCCGGTGGAAAGCGGTGACACGCCAAAGGACCCTGCGGTGATCTCCAAGTCCCCATCCATGGCCCAGGACTCAGGCGCCTCAGAGCTATTACCCAATGGGGACTTGGAGAAGCGGAGTGAGCCCCAGCCAGAGGAGGGGAGCCCTGCTGGGGGGCAGAAGGGCGGGGCCCCAGCAGAGGGAGAGGGTGCAGCTGAGACCCTGCCTGAAGCCTCAAGAGCAGTGGAAAATGGCTGCTGCACCCCCAAGGAGGGCCGAGGAGCCCCTGCAGAAGCGGGCAAAGAACAGAAGGAGACCAACATCGAATCCATGAAAATGGAGGGCTCCCGGGGCCGGCTGCGGGGTGGCTTGGGCTGGGAGTCCAGCCTCCGTCAGCGGCCCATGCCGAGGCTCACCTTCCAGGCGGGGGACCCCTACTACATCAGCAAGCGCAAGCGGGACGAGTGGCTGGCACGCTGGAAAAGGGAGGCTGAGAAGAAAGCCAAGGTCATTGCAGGAATGAATGCTGTGGAAGAAAACCAGGGGCCCGGGGAGTCTCAGAAGGTGGAGGAGGCCAGCCCTCCTGCTGTGCAGCAGCCCACTGACCCCGCATCCCCCACTGTGGCTACCACGCCTGAGCCCGTGGGGTCCGATGCTGGGGACAAGAATGCCACCAAAGCAGGCGATGACGAGCCAGAGTACGAGGACGGCCGGGGCTTTGGCATTGGGGAGCTGGTGTGGGGGAAACTGCGGGGCTTCTCCTGGTGGCCAGGCCGCATTGTGTCTTGGTGGATGACGGGCCGGAGCCGAGCAGCTGAAGGCACCCGCTGGGTCATGTGGTTCGGAGACGGCAAATTCTCAGTGGTGTGTGTTGAGAAGCTGATGCCGCTGAGCTCGTTTTGCAGTGCGTTCCACCAGGCCACGTACAACAAGCAGCCCATGTACCGCAAAGCCATCTACGAGGTCCTGCAGGTGGCCAGCAGCCGCGCGGGGAAGCTGTTCCCGGTGTGCCACGACAGCGATGAGAGTGACACTGCCAAGGCCGTGGAGGTGCAGAACAAGCCCATGATTGAATGGGCCCTGGGGGGCTTCCAGCCTTCTGGCCCTAAGGGCCTGGAGCCACCAGAAGAAGAGAAGAATCCCTACAAAGAAGTGTACACGGACATGTGGGTGGAACCTGAGGCAGCTGCCTACGCACCACCTCCACCAGCCAAAAAGCCCCGGAAGAGCACAGCGGAGAAGCCCAAGGTCAAGGAGATTATTGATGAGCGCACAAGAGAGCGGCTGGTGTACGAGGTGCGGCAGAAGTGCCGGAACATTGAGGACATCTGCATCTCCTGTGGGAGCCTCAATGTTACCCTGGAACACCCCCTCTTCGTTGGAGGAATGTGCCAAAACTGCAAGAACTGCTTTCTGGAGTGTGCGTACCAGTACGACGACGACGGCTACCAGTCCTACTGCACCATCTGCTGTGGGGGCCGTGAGGTGCTCATGTGCGGAAACAACAACTGCTGCAGGTGCTTTTGCGTGGAGTGTGTGGACCTCTTGGTGGGGCCGGGGGCTGCCCAGGCAGCCATTAAGGAAGACCCCTGGAACTGCTACATGTGCGGGCACAAGGGTACCTACGGGCTGCTGCGGCGGCGAGAGGACTGGCCCTCCCGGCTCCAGATGTTCTTCGCTAATAACCACGACCAGGAATTTGACCCTCCAAAGGTTTACCCACCTGTCCCAGCTGAGAAGAGGAAGCCCATCCGGGTGCTGTCTCTCTTTGATGGAATCGCTACAGGGCTCCTGGTGCTGAAGGACTTGGGCATTCAGGTGGACCGCTACATTGCCTCGGAGGTGTGTGAGGACTCCATCACGGTGGGCATGGTGCGGCACCAGGGGAAGATCATGTACGTCGGGGACGTCCGCAGCGTCACACAGAAGCATATCCAGGAGTGGGGCCCATTCGATCTGGTGATTGGGGGCAGTCCCTGCAATGACCTCTCCATCGTCAACCCTGCTCGCAAGGGCCTCTACGAGGGCACTGGCCGGCTCTTCTTTGAGTTCTACCGCCTCCTGCATGATGCGCGGCCCAAGGAGGGAGATGATCGCCCCTTCTTCTGGCTCTTTGAGAATGTGGTGGCCATGGGCGTTAGTGACAAGAGGGACATCTCGCGATTTCTCGAGTCCAACCCTGTGATGATTGATGCCAAAGAAGTGTCAGCTGCACACAGGGCCCGCTACTTCTGGGGTAACCTTCCCGGTATGAACAGGCCGTTGGCATCCACTGTGAATGATAAGCTGGAGCTGCAGGAGTGTCTGGAGCATGGCAGGATAGCCAAGTTCAGCAAAGTGAGGACCATTACTACGAGGTCAAACTCCATAAAGCAGGGCAAAGACCAGCATTTTCCTGTCTTCATGAATGAGAAAGAGGACATCTTATGGTGCACTGAAATGGAAAGGGTATTTGGTTTCCCAGTCCACTATACTGACGTCTCCAACATGAGCCGCTTGGCGAGGCAGAGACTGCTGGGCCGGTCATGGAGCGTGCCAGTCATCCGCCACCTCTTCGCTCCGCTGAAGGAGTATTTTGCGTGTGTGTAAG'\n",
    "\n",
    "    #List of all amino acid positions (across whole coding region) covered (amino acid position minus 1 (as python starts at 0))\n",
    "    DNMT3A_missense_codons_DICT_NP_072046={}\n",
    "    for i in range(lower_range, upper_range):\n",
    "        DNMT3A_missense_codons_DICT_NP_072046[i+1]= (DNMT3A_codons_NP_072046[i])\n",
    "\n",
    "    calls_DNMT3A_missense_NP_072046 = {}\n",
    "    for k,v in DNMT3A_missense_codons_DICT_NP_072046.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_DNMT3A_missense_NP_072046[k]=(v, changes)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_DNMT3A_missense_NP_072046_extra = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_missense_NP_072046.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_DNMT3A_missense_NP_072046_extra[k] = (Amino_acid_change, v[0], v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_DNMT3A_missense_NP_072046_extra2 = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_missense_NP_072046_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_DNMT3A_missense_NP_072046_extra2[k] = (Amino_acid_abbrev, v[1], v[2])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary that contains details of the positions of missense variants at that codon\n",
    "    DNMT3A_position_in_list_dict_missense_NP_072046 = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_missense_NP_072046_extra2.items():\n",
    "        original_AA = v[0]\n",
    "        possible_AA_changes = v[2]\n",
    "        possible_AA_changes_string = ''.join(v[2])\n",
    "        positions_in_list = []\n",
    "        for i in possible_AA_changes:\n",
    "            if i != original_AA: #not synonymous\n",
    "                if i != '*': #not missense\n",
    "                    position_in_list=-1\n",
    "                    missense_change = i\n",
    "                    new_aa_list = []\n",
    "                    while True:\n",
    "                        position_in_list = possible_AA_changes_string.find(missense_change, position_in_list+1)\n",
    "        #               print(position_in_list)\n",
    "                        if position_in_list == -1: break\n",
    "                        positions_in_list.append(position_in_list)\n",
    "                        positions_in_list = set(positions_in_list) #convert into a 'set' to remove duplicate positions\n",
    "                        positions_in_list = list(positions_in_list) #convert back to a list so can be added to dictionary\n",
    "                    for pos in positions_in_list:\n",
    "                        new_aa = possible_AA_changes_string[pos]\n",
    "                        new_aa_list.append(new_aa)\n",
    "                    DNMT3A_position_in_list_dict_missense_NP_072046[k]=(positions_in_list, v, new_aa_list)\n",
    "\n",
    "        aa_positions_variants_called = []\n",
    "        for k, v in DNMT3A_position_in_list_dict_missense_NP_072046.items():\n",
    "            aa_positions_variants_called.append(k)\n",
    "    # print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    DNMT3A_bases_missense_NP_072046 = {}\n",
    "    for k, v in DNMT3A_position_in_list_dict_missense_NP_072046.items():\n",
    "        original_a_a = v[1][0]\n",
    "        codon_seq = v[1][1]\n",
    "        base_changes=[]\n",
    "        new_a_a_list = v[2]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        DNMT3A_bases_missense_NP_072046[k]=(v, base_changes, original_a_a, new_a_a_list)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    DNMT3A_bases_missense2_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_missense_NP_072046.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        new_a_a_list = v[3]\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        DNMT3A_bases_missense2_NP_072046[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    DNMT3A_bases_missense3_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_missense2_NP_072046.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_a_a = v[0][2]\n",
    "        new_aa_list = v[0][0][2]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = k\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(k)-1)*3)+codon_pos #codon number minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][1][1][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        DNMT3A_bases_missense3_NP_072046[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, original_a_a, new_aa_list)\n",
    "\n",
    "    # Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    DNMT3A_bases_missense4_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_missense3_NP_072046.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "        new_a_a_list = v[6]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_DNMT3A_NP072046:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check_DNMT3A_NP072046(i)\n",
    "                sites_trio_missense.append([adapted_trio])\n",
    "                DNMT3A_bases_missense4_NP_072046[k] = (v, sites_trio_missense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_missense = [DNMT3A_exons_extra_bases_NP_072046[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_missense.append(site_trio_missense)\n",
    "                    DNMT3A_bases_missense4_NP_072046[k] = (v, sites_trio_missense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    DNMT3A_sites_missense5_NP_072046 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in DNMT3A_bases_missense4_NP_072046.items():\n",
    "        original_a_a = v[0][0][0]\n",
    "        new_a_a = v[0][6]\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            new_amino_acid = new_a_a[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, original_a_a, new_amino_acid])\n",
    "            change = change+1\n",
    "        DNMT3A_sites_missense5_NP_072046[a]=info\n",
    "        a=a+1\n",
    "    #     print(DNMT3A_sites_missense5_NP_072046)\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    DNMT3A_sites_missense6_NP_072046={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in DNMT3A_sites_missense5_NP_072046.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        original_a_a = v[5]\n",
    "        new_a_a = v[6]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        DNMT3A_sites_missense6_NP_072046[k] = (v[0],changes3, codon_n_912, original_a_a, new_a_a) #site_trio = the affected base and its neighbours\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    DNMT3A_mutation_rates_missense_NP_072046 = {}\n",
    "    for k, v in DNMT3A_sites_missense6_NP_072046.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            original_a_a = v[3]\n",
    "            new_a_a = v[4]\n",
    "            aa_positioning = v[2]\n",
    "            variant_id = str(original_a_a)+str(aa_positioning)+str(new_a_a)\n",
    "            DNMT3A_mutation_rates_missense_NP_072046[k] = (v[0],v[1], mutation_rate, variant_id)\n",
    "\n",
    "    DNMT3A_mutation_rate_missense_NP_072046_sum = []\n",
    "\n",
    "    mutation_rates_list = []\n",
    "    for k, v in DNMT3A_mutation_rates_missense_NP_072046.items():\n",
    "        codon_position_in_gene=(v[0])\n",
    "        variant = v[3]\n",
    "        site = v[1]\n",
    "        mutation_rate_value = v[2] \n",
    "        mutation_rates_list.append((v[0], variant, site, mutation_rate_value))\n",
    "        DNMT3A_mutation_rate_missense_NP_072046_sum.append(mutation_rate_value)\n",
    "\n",
    "    #Get the average of all the mutation rates\n",
    "    DNMT3A_mutation_rate_missense_NP_072046_sum = []\n",
    "    for k, v in DNMT3A_mutation_rates_missense_NP_072046.items():\n",
    "        mut_rate = v[2]\n",
    "        DNMT3A_mutation_rate_missense_NP_072046_sum.append(mut_rate)\n",
    "\n",
    "    DNMT3A_Summed_mutation_rate_missense_NP_072046_added = np.sum(DNMT3A_mutation_rate_missense_NP_072046_sum)\n",
    "\n",
    "    print('Total summed mutation rate across DNMT3A missense_NP_072046 for this study = ', DNMT3A_Summed_mutation_rate_missense_NP_072046_added)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all # e.g. lower= 0, upper = 913 for whole gene\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TET2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_nonsense_range_of_variants_TET2(lower_range, upper_range):\n",
    "    \n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base_TET2_NP01120680 = [3409, 3804, 4044, 4537]\n",
    "   \n",
    "    def intron_exon_check_TET2_NP01120680(position_in_gene_number):\n",
    "        if position_in_gene_number == 3409:\n",
    "            gene_pos = 3409\n",
    "            first_base = TET2_exons_extra_bases_NP_01120680[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 3804:\n",
    "            gene_pos = 3804\n",
    "            first_base = 'G'\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = TET2_exons_extra_bases_NP_01120680[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 4044:\n",
    "            gene_pos = 4044\n",
    "            first_base = TET2_exons_extra_bases_NP_01120680[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 4537:\n",
    "            gene_pos = 4537\n",
    "            first_base = TET2_exons_extra_bases_NP_01120680[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #TET2 transcript (NP_01120680):\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000168769;r=4:106067032-106163928;t=ENST00000380013\n",
    "    TET2_exons_NP_01120680 = 'ATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGAGCAAATTATTGAAAAAGATGAAGGTCCTTTTTATACCCATCTAGGAGCAGGTCCTAATGTGGCAGCTATTAGAGAAATCATGGAAGAAAGGTTTGGACAGAAGGGTAAAGCTATTAGGATTGAAAGAGTCATCTATACTGGTAAAGAAGGCAAAAGTTCTCAGGGATGTCCTATTGCTAAGTGGGTGGTTCGCAGAAGCAGCAGTGAAGAGAAGCTACTGTGTTTGGTGCGGGAGCGAGCTGGCCACACCTGTGAGGCTGCAGTGATTGTGATTCTCATCCTGGTGTGGGAAGGAATCCCGCTGTCTCTGGCTGACAAACTCTACTCGGAGCTTACCGAGACGCTGAGGAAATACGGCACGCTCACCAATCGCCGGTGTGCCTTGAATGAAGAGAGAACTTGCGCCTGTCAGGGGCTGGATCCAGAAACCTGTGGTGCCTCCTTCTCTTTTGGTTGTTCATGGAGCATGTACTACAATGGATGTAAGTTTGCCAGAAGCAAGATCCCAAGGAAGTTTAAGCTGCTTGGGGATGACCCAAAAGAGGAAGAGAAACTGGAGTCTCATTTGCAAAACCTGTCCACTCTTATGGCACCAACATATAAGAAACTTGCACCTGATGCATATAATAATCAGATTGAATATGAACACAGAGCACCAGAGTGCCGTCTGGGTCTGAAGGAAGGCCGTCCATTCTCAGGGGTCACTGCATGTTTGGACTTCTGTGCTCATGCCCACAGAGACTTGCACAACATGCAGAATGGCAGCACATTGGTATGCACTCTCACTAGAGAAGACAATCGAGAATTTGGAGGAAAACCTGAGGATGAGCAGCTTCACGTTCTGCCTTTATACAAAGTCTCTGACGTGGATGAGTTTGGGAGTGTGGAAGCTCAGGAGGAGAAAAAACGGAGTGGTGCCATTCAGGTACTGAGTTCTTTTCGGCGAAAAGTCAGGATGTTAGCAGAGCCAGTCAAGACTTGCCGACAAAGGAAACTAGAAGCCAAGAAAGCTGCAGCTGAAAAGCTTTCCTCCCTGGAGAACAGCTCAAATAAAAATGAAAAGGAAAAGTCAGCCCCATCACGTACAAAACAAACTGAAAACGCAAGCCAGGCTAAACAGTTGGCAGAACTTTTGCGACTTTCAGGACCAGTCATGCAGCAGTCCCAGCAGCCCCAGCCTCTACAGAAGCAGCCACCACAGCCCCAGCAGCAGCAGAGACCCCAGCAGCAGCAGCCACATCACCCTCAGACAGAGTCTGTCAACTCTTATTCTGCTTCTGGATCCACCAATCCATACATGAGACGGCCCAATCCAGTTAGTCCTTATCCAAACTCTTCACACACTTCAGATATCTATGGAAGCACCAGCCCTATGAACTTCTATTCCACCTCATCTCAAGCTGCAGGTTCATATTTGAATTCTTCTAATCCCATGAACCCTTACCCTGGGCTTTTGAATCAGAATACCCAATATCCATCATATCAATGCAATGGAAACCTATCAGTGGACAACTGCTCCCCATATCTGGGTTCCTATTCTCCCCAGTCTCAGCCGATGGATCTGTATAGGTATCCAAGCCAAGACCCTCTGTCTAAGCTCAGTCTACCACCCATCCATACACTTTACCAGCCAAGGTTTGGAAATAGCCAGAGTTTTACATCTAAATACTTAGGTTATGGAAACCAAAATATGCAGGGAGATGGTTTCAGCAGTTGTACCATTAGACCAAATGTACATCATGTAGGGAAATTGCCTCCTTATCCCACTCATGAGATGGATGGCCACTTCATGGGAGCCACCTCTAGATTACCACCCAATCTGAGCAATCCAAACATGGACTATAAAAATGGTGAACATCATTCACCTTCTCACATAATCCATAACTACAGTGCAGCTCCGGGCATGTTCAACAGCTCTCTTCATGCCCTGCATCTCCAAAACAAGGAGAATGACATGCTTTCCCACACAGCTAATGGGTTATCAAAGATGCTTCCAGCTCTTAACCATGATAGAACTGCTTGTGTCCAAGGAGGCTTACACAAATTAAGTGATGCTAATGGTCAGGAAAAGCAGCCATTGGCACTAGTCCAGGGTGTGGCTTCTGGTGCAGAGGACAACGATGAGGTCTGGTCAGACAGCGAGCAGAGCTTTCTGGATCCTGACATTGGGGGAGTGGCCGTGGCTCCAACTCATGGGTCAATTCTCATTGAGTGTGCAAAGCGTGAGCTGCATGCCACAACCCCTTTAAAGAATCCCAATAGGAATCACCCCACCAGGATCTCCCTCGTCTTTTACCAGCATAAGAGCATGAATGAGCCAAAACATGGCTTGGCTCTTTGGGAAGCCAAAATGGCTGAAAAAGCCCGTGAGAAAGAGGAAGAGTGTGAAAAGTATGGCCCAGACTATGTGCCTCAGAAATCCCATGGCAAAAAAGTGAAACGGGAGCCTGCTGAGCCACATGAAACTTCAGAGCCCACTTACCTGCGTTTCATCAAGTCTCTTGCCGAAAGGACCATGTCCGTGACCACAGACTCCACAGTAACTACATCTCCATATGCCTTCACTCGGGTCACAGGGCCTTACAACAGATATATATGA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    TET2_codons_NP_01120680 = [TET2_exons_NP_01120680[i:i+3] for i in range(0, len(TET2_exons_NP_01120680), 3)]\n",
    "\n",
    "    #TET2 transcript (transcript variant 2 (TET2-001), NM_017628, NP_001120680):\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000168769;r=4:106067032-106163928;t=ENST00000380013\n",
    "    TET2_exons_extra_bases_NP_01120680 = 'GATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGAGCAAATTATTGAAAAAGATGAAGGTCCTTTTTATACCCATCTAGGAGCAGGTCCTAATGTGGCAGCTATTAGAGAAATCATGGAAGAAAGGTTTGGACAGAAGGGTAAAGCTATTAGGATTGAAAGAGTCATCTATACTGGTAAAGAAGGCAAAAGTTCTCAGGGATGTCCTATTGCTAAGTGGGTGGTTCGCAGAAGCAGCAGTGAAGAGAAGCTACTGTGTTTGGTGCGGGAGCGAGCTGGCCACACCTGTGAGGCTGCAGTGATTGTGATTCTCATCCTGGTGTGGGAAGGAATCCCGCTGTCTCTGGCTGACAAACTCTACTCGGAGCTTACCGAGACGCTGAGGAAATACGGCACGCTCACCAATCGCCGGTGTGCCTTGAATGAAGAGAGAACTTGCGCCTGTCAGGGGCTGGATCCAGAAACCTGTGGTGCCTCCTTCTCTTTTGGTTGTTCATGGAGCATGTACTACAATGGATGTAAGTTTGCCAGAAGCAAGATCCCAAGGAAGTTTAAGCTGCTTGGGGATGACCCAAAAGAGGAAGAGAAACTGGAGTCTCATTTGCAAAACCTGTCCACTCTTATGGCACCAACATATAAGAAACTTGCACCTGATGCATATAATAATCAGATTGAATATGAACACAGAGCACCAGAGTGCCGTCTGGGTCTGAAGGAAGGCCGTCCATTCTCAGGGGTCACTGCATGTTTGGACTTCTGTGCTCATGCCCACAGAGACTTGCACAACATGCAGAATGGCAGCACATTGGTATGCACTCTCACTAGAGAAGACAATCGAGAATTTGGAGGAAAACCTGAGGATGAGCAGCTTCACGTTCTGCCTTTATACAAAGTCTCTGACGTGGATGAGTTTGGGAGTGTGGAAGCTCAGGAGGAGAAAAAACGGAGTGGTGCCATTCAGGTACTGAGTTCTTTTCGGCGAAAAGTCAGGATGTTAGCAGAGCCAGTCAAGACTTGCCGACAAAGGAAACTAGAAGCCAAGAAAGCTGCAGCTGAAAAGCTTTCCTCCCTGGAGAACAGCTCAAATAAAAATGAAAAGGAAAAGTCAGCCCCATCACGTACAAAACAAACTGAAAACGCAAGCCAGGCTAAACAGTTGGCAGAACTTTTGCGACTTTCAGGACCAGTCATGCAGCAGTCCCAGCAGCCCCAGCCTCTACAGAAGCAGCCACCACAGCCCCAGCAGCAGCAGAGACCCCAGCAGCAGCAGCCACATCACCCTCAGACAGAGTCTGTCAACTCTTATTCTGCTTCTGGATCCACCAATCCATACATGAGACGGCCCAATCCAGTTAGTCCTTATCCAAACTCTTCACACACTTCAGATATCTATGGAAGCACCAGCCCTATGAACTTCTATTCCACCTCATCTCAAGCTGCAGGTTCATATTTGAATTCTTCTAATCCCATGAACCCTTACCCTGGGCTTTTGAATCAGAATACCCAATATCCATCATATCAATGCAATGGAAACCTATCAGTGGACAACTGCTCCCCATATCTGGGTTCCTATTCTCCCCAGTCTCAGCCGATGGATCTGTATAGGTATCCAAGCCAAGACCCTCTGTCTAAGCTCAGTCTACCACCCATCCATACACTTTACCAGCCAAGGTTTGGAAATAGCCAGAGTTTTACATCTAAATACTTAGGTTATGGAAACCAAAATATGCAGGGAGATGGTTTCAGCAGTTGTACCATTAGACCAAATGTACATCATGTAGGGAAATTGCCTCCTTATCCCACTCATGAGATGGATGGCCACTTCATGGGAGCCACCTCTAGATTACCACCCAATCTGAGCAATCCAAACATGGACTATAAAAATGGTGAACATCATTCACCTTCTCACATAATCCATAACTACAGTGCAGCTCCGGGCATGTTCAACAGCTCTCTTCATGCCCTGCATCTCCAAAACAAGGAGAATGACATGCTTTCCCACACAGCTAATGGGTTATCAAAGATGCTTCCAGCTCTTAACCATGATAGAACTGCTTGTGTCCAAGGAGGCTTACACAAATTAAGTGATGCTAATGGTCAGGAAAAGCAGCCATTGGCACTAGTCCAGGGTGTGGCTTCTGGTGCAGAGGACAACGATGAGGTCTGGTCAGACAGCGAGCAGAGCTTTCTGGATCCTGACATTGGGGGAGTGGCCGTGGCTCCAACTCATGGGTCAATTCTCATTGAGTGTGCAAAGCGTGAGCTGCATGCCACAACCCCTTTAAAGAATCCCAATAGGAATCACCCCACCAGGATCTCCCTCGTCTTTTACCAGCATAAGAGCATGAATGAGCCAAAACATGGCTTGGCTCTTTGGGAAGCCAAAATGGCTGAAAAAGCCCGTGAGAAAGAGGAAGAGTGTGAAAAGTATGGCCCAGACTATGTGCCTCAGAAATCCCATGGCAAAAAAGTGAAACGGGAGCCTGCTGAGCCACATGAAACTTCAGAGCCCACTTACCTGCGTTTCATCAAGTCTCTTGCCGAAAGGACCATGTCCGTGACCACAGACTCCACAGTAACTACATCTCCATATGCCTTCACTCGGGTCACAGGGCCTTACAACAGATATATATGAT'\n",
    "\n",
    "    #List of all amino acid positions (across whole coding region) covered by Coombs (amino acid position minus 1 (as python starts at 0))\n",
    "    TET2_nonsense_codons_DICT_NP_01120680={}\n",
    "    for i in range(lower_range, upper_range):\n",
    "        TET2_nonsense_codons_DICT_NP_01120680[i+1]= (TET2_codons_NP_01120680[i])\n",
    "\n",
    "    calls_TET2_nonsense_NP_01120680 = {}\n",
    "    for k,v in TET2_nonsense_codons_DICT_NP_01120680.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_TET2_nonsense_NP_01120680[k]=(v, changes)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_TET2_nonsense_NP_01120680_extra = {}\n",
    "\n",
    "    for k, v in calls_TET2_nonsense_NP_01120680.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_TET2_nonsense_NP_01120680_extra[k] = (Amino_acid_change, v[0], v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_TET2_nonsense_NP_01120680_extra2 = {}\n",
    "\n",
    "    for k, v in calls_TET2_nonsense_NP_01120680_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_TET2_nonsense_NP_01120680_extra2[k] = (Amino_acid_abbrev, v[1], v[2])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary that contains details of the positions of nonsense variants at that codon\n",
    "    TET2_position_in_list_dict_nonsense_NP_01120680 = {}\n",
    "\n",
    "    for k, v in calls_TET2_nonsense_NP_01120680_extra2.items():\n",
    "        original_AA = v[0]\n",
    "        possible_AA_changes = v[2]\n",
    "        possible_AA_changes_string = ''.join(v[2])\n",
    "        positions_in_list = []\n",
    "        for i in possible_AA_changes:\n",
    "            if i == '*': #nonsense\n",
    "                position_in_list=-1\n",
    "                nonsense_change = i    \n",
    "                while True:\n",
    "                    position_in_list = possible_AA_changes_string.find(nonsense_change, position_in_list+1)\n",
    "    #               print(position_in_list)\n",
    "                    if position_in_list == -1: break\n",
    "                    positions_in_list.append(position_in_list)\n",
    "                    positions_in_list = set(positions_in_list) #convert into a 'set' to remove duplicate positions\n",
    "                    positions_in_list = list(positions_in_list) #convert back to a list so can be added to dictionary\n",
    "                TET2_position_in_list_dict_nonsense_NP_01120680[k]=(positions_in_list, v)\n",
    "\n",
    "        aa_positions_variants_called = []\n",
    "        for k, v in TET2_position_in_list_dict_nonsense_NP_01120680.items():\n",
    "            aa_positions_variants_called.append(k)\n",
    "    # print('positions at which nonsense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TET2_bases_nonsense_NP_01120680 = {}\n",
    "    for k, v in TET2_position_in_list_dict_nonsense_NP_01120680.items():\n",
    "        original_a_a = v[1][0]\n",
    "        codon_seq = v[1][1]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TET2_bases_nonsense_NP_01120680[k]=(v, base_changes, original_a_a)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TET2_bases_nonsense2_NP_01120680 = {}\n",
    "\n",
    "    for k, v in TET2_bases_nonsense_NP_01120680.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TET2_bases_nonsense2_NP_01120680[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TET2_bases_nonsense3_NP_01120680 = {}\n",
    "\n",
    "    for k, v in TET2_bases_nonsense2_NP_01120680.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_a_a = v[0][2]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = k\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(k)-1)*3)+codon_pos #codon number minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][1][1][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TET2_bases_nonsense3_NP_01120680[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, original_a_a)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TET2_bases_nonsense4_NP_01120680 = {}\n",
    "\n",
    "    for k, v in TET2_bases_nonsense3_NP_01120680.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_nonsense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_TET2_NP01120680:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check_TET2_NP01120680(i)\n",
    "                sites_trio_nonsense.append([adapted_trio])\n",
    "                TET2_bases_nonsense4_NP_01120680[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_nonsense = [TET2_exons_extra_bases_NP_01120680[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_nonsense.append(site_trio_nonsense)\n",
    "                    TET2_bases_nonsense4_NP_01120680[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TET2_sites_nonsense5_NP_01120680 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TET2_bases_nonsense4_NP_01120680.items():\n",
    "        original_a_a = v[0][0][0]\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, original_a_a])\n",
    "            change = change+1\n",
    "        TET2_sites_nonsense5_NP_01120680[a]=info\n",
    "        a=a+1\n",
    "    #     print(TET2_sites_nonsense5_NP_01120680)\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TET2_sites_nonsense6_NP_01120680={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TET2_sites_nonsense5_NP_01120680.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        original_a_a = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TET2_sites_nonsense6_NP_01120680[k] = (v[0],changes3, codon_n_912, original_a_a) #site_trio = the affected base and its neighbours\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    TET2_mutation_rates_nonsense_NP_01120680 = {}\n",
    "    for k, v in TET2_sites_nonsense6_NP_01120680.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            original_a_a = v[3]\n",
    "            aa_positioning = v[2]\n",
    "            variant_type = '*'\n",
    "            variant_id = str(original_a_a)+str(aa_positioning)+variant_type\n",
    "            TET2_mutation_rates_nonsense_NP_01120680[k] = (v[0],v[1], mutation_rate, variant_id)\n",
    "\n",
    "    TET2_mutation_rate_nonsense_NP_01120680_sum = []\n",
    "\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TET2_mutation_rates_nonsense_NP_01120680.items():\n",
    "        codon_position_in_gene=(v[0])\n",
    "        variant = v[3]\n",
    "        site = v[1]\n",
    "        mutation_rate_value = v[2] \n",
    "        mutation_rates_list.append((v[0], variant, site, mutation_rate_value))\n",
    "        TET2_mutation_rate_nonsense_NP_01120680_sum.append(mutation_rate_value)\n",
    "\n",
    "    #Get the average of all the mutation rates\n",
    "    TET2_mutation_rate_nonsense_NP_01120680_sum = []\n",
    "    for k, v in TET2_mutation_rates_nonsense_NP_01120680.items():\n",
    "        mut_rate = v[2]\n",
    "        TET2_mutation_rate_nonsense_NP_01120680_sum.append(mut_rate)\n",
    "\n",
    "    TET2_Summed_mutation_rate_nonsense_NP_01120680_added = np.sum(TET2_mutation_rate_nonsense_NP_01120680_sum)\n",
    "\n",
    "    print('Total summed mutation rate across TET2 nonsense_NP_01120680 for this study = ', TET2_Summed_mutation_rate_nonsense_NP_01120680_added)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all # e.g. lower= 0, upper = 913 for whole gene\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_missense_range_of_variants_TET2(lower_range, upper_range):\n",
    "\n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base_TET2_NP01120680 = [3409, 3804, 4044, 4537]\n",
    "   \n",
    "    def intron_exon_check_TET2_NP01120680(position_in_gene_number):\n",
    "        if position_in_gene_number == 3409:\n",
    "            gene_pos = 3409\n",
    "            first_base = TET2_exons_extra_bases_NP_01120680[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 3804:\n",
    "            gene_pos = 3804\n",
    "            first_base = 'G'\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = TET2_exons_extra_bases_NP_01120680[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 4044:\n",
    "            gene_pos = 4044\n",
    "            first_base = TET2_exons_extra_bases_NP_01120680[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 4537:\n",
    "            gene_pos = 4537\n",
    "            first_base = TET2_exons_extra_bases_NP_01120680[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #TET2 transcript (NP_01120680):\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000168769;r=4:106067032-106163928;t=ENST00000380013\n",
    "    TET2_exons_NP_01120680 = 'ATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGAGCAAATTATTGAAAAAGATGAAGGTCCTTTTTATACCCATCTAGGAGCAGGTCCTAATGTGGCAGCTATTAGAGAAATCATGGAAGAAAGGTTTGGACAGAAGGGTAAAGCTATTAGGATTGAAAGAGTCATCTATACTGGTAAAGAAGGCAAAAGTTCTCAGGGATGTCCTATTGCTAAGTGGGTGGTTCGCAGAAGCAGCAGTGAAGAGAAGCTACTGTGTTTGGTGCGGGAGCGAGCTGGCCACACCTGTGAGGCTGCAGTGATTGTGATTCTCATCCTGGTGTGGGAAGGAATCCCGCTGTCTCTGGCTGACAAACTCTACTCGGAGCTTACCGAGACGCTGAGGAAATACGGCACGCTCACCAATCGCCGGTGTGCCTTGAATGAAGAGAGAACTTGCGCCTGTCAGGGGCTGGATCCAGAAACCTGTGGTGCCTCCTTCTCTTTTGGTTGTTCATGGAGCATGTACTACAATGGATGTAAGTTTGCCAGAAGCAAGATCCCAAGGAAGTTTAAGCTGCTTGGGGATGACCCAAAAGAGGAAGAGAAACTGGAGTCTCATTTGCAAAACCTGTCCACTCTTATGGCACCAACATATAAGAAACTTGCACCTGATGCATATAATAATCAGATTGAATATGAACACAGAGCACCAGAGTGCCGTCTGGGTCTGAAGGAAGGCCGTCCATTCTCAGGGGTCACTGCATGTTTGGACTTCTGTGCTCATGCCCACAGAGACTTGCACAACATGCAGAATGGCAGCACATTGGTATGCACTCTCACTAGAGAAGACAATCGAGAATTTGGAGGAAAACCTGAGGATGAGCAGCTTCACGTTCTGCCTTTATACAAAGTCTCTGACGTGGATGAGTTTGGGAGTGTGGAAGCTCAGGAGGAGAAAAAACGGAGTGGTGCCATTCAGGTACTGAGTTCTTTTCGGCGAAAAGTCAGGATGTTAGCAGAGCCAGTCAAGACTTGCCGACAAAGGAAACTAGAAGCCAAGAAAGCTGCAGCTGAAAAGCTTTCCTCCCTGGAGAACAGCTCAAATAAAAATGAAAAGGAAAAGTCAGCCCCATCACGTACAAAACAAACTGAAAACGCAAGCCAGGCTAAACAGTTGGCAGAACTTTTGCGACTTTCAGGACCAGTCATGCAGCAGTCCCAGCAGCCCCAGCCTCTACAGAAGCAGCCACCACAGCCCCAGCAGCAGCAGAGACCCCAGCAGCAGCAGCCACATCACCCTCAGACAGAGTCTGTCAACTCTTATTCTGCTTCTGGATCCACCAATCCATACATGAGACGGCCCAATCCAGTTAGTCCTTATCCAAACTCTTCACACACTTCAGATATCTATGGAAGCACCAGCCCTATGAACTTCTATTCCACCTCATCTCAAGCTGCAGGTTCATATTTGAATTCTTCTAATCCCATGAACCCTTACCCTGGGCTTTTGAATCAGAATACCCAATATCCATCATATCAATGCAATGGAAACCTATCAGTGGACAACTGCTCCCCATATCTGGGTTCCTATTCTCCCCAGTCTCAGCCGATGGATCTGTATAGGTATCCAAGCCAAGACCCTCTGTCTAAGCTCAGTCTACCACCCATCCATACACTTTACCAGCCAAGGTTTGGAAATAGCCAGAGTTTTACATCTAAATACTTAGGTTATGGAAACCAAAATATGCAGGGAGATGGTTTCAGCAGTTGTACCATTAGACCAAATGTACATCATGTAGGGAAATTGCCTCCTTATCCCACTCATGAGATGGATGGCCACTTCATGGGAGCCACCTCTAGATTACCACCCAATCTGAGCAATCCAAACATGGACTATAAAAATGGTGAACATCATTCACCTTCTCACATAATCCATAACTACAGTGCAGCTCCGGGCATGTTCAACAGCTCTCTTCATGCCCTGCATCTCCAAAACAAGGAGAATGACATGCTTTCCCACACAGCTAATGGGTTATCAAAGATGCTTCCAGCTCTTAACCATGATAGAACTGCTTGTGTCCAAGGAGGCTTACACAAATTAAGTGATGCTAATGGTCAGGAAAAGCAGCCATTGGCACTAGTCCAGGGTGTGGCTTCTGGTGCAGAGGACAACGATGAGGTCTGGTCAGACAGCGAGCAGAGCTTTCTGGATCCTGACATTGGGGGAGTGGCCGTGGCTCCAACTCATGGGTCAATTCTCATTGAGTGTGCAAAGCGTGAGCTGCATGCCACAACCCCTTTAAAGAATCCCAATAGGAATCACCCCACCAGGATCTCCCTCGTCTTTTACCAGCATAAGAGCATGAATGAGCCAAAACATGGCTTGGCTCTTTGGGAAGCCAAAATGGCTGAAAAAGCCCGTGAGAAAGAGGAAGAGTGTGAAAAGTATGGCCCAGACTATGTGCCTCAGAAATCCCATGGCAAAAAAGTGAAACGGGAGCCTGCTGAGCCACATGAAACTTCAGAGCCCACTTACCTGCGTTTCATCAAGTCTCTTGCCGAAAGGACCATGTCCGTGACCACAGACTCCACAGTAACTACATCTCCATATGCCTTCACTCGGGTCACAGGGCCTTACAACAGATATATATGA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    TET2_codons_NP_01120680 = [TET2_exons_NP_01120680[i:i+3] for i in range(0, len(TET2_exons_NP_01120680), 3)]\n",
    "\n",
    "    #TET2 transcript (transcript variant 2 (TET2-001), NM_017628, NP_001120680):\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000168769;r=4:106067032-106163928;t=ENST00000380013\n",
    "    TET2_exons_extra_bases_NP_01120680 = 'GATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGAGCAAATTATTGAAAAAGATGAAGGTCCTTTTTATACCCATCTAGGAGCAGGTCCTAATGTGGCAGCTATTAGAGAAATCATGGAAGAAAGGTTTGGACAGAAGGGTAAAGCTATTAGGATTGAAAGAGTCATCTATACTGGTAAAGAAGGCAAAAGTTCTCAGGGATGTCCTATTGCTAAGTGGGTGGTTCGCAGAAGCAGCAGTGAAGAGAAGCTACTGTGTTTGGTGCGGGAGCGAGCTGGCCACACCTGTGAGGCTGCAGTGATTGTGATTCTCATCCTGGTGTGGGAAGGAATCCCGCTGTCTCTGGCTGACAAACTCTACTCGGAGCTTACCGAGACGCTGAGGAAATACGGCACGCTCACCAATCGCCGGTGTGCCTTGAATGAAGAGAGAACTTGCGCCTGTCAGGGGCTGGATCCAGAAACCTGTGGTGCCTCCTTCTCTTTTGGTTGTTCATGGAGCATGTACTACAATGGATGTAAGTTTGCCAGAAGCAAGATCCCAAGGAAGTTTAAGCTGCTTGGGGATGACCCAAAAGAGGAAGAGAAACTGGAGTCTCATTTGCAAAACCTGTCCACTCTTATGGCACCAACATATAAGAAACTTGCACCTGATGCATATAATAATCAGATTGAATATGAACACAGAGCACCAGAGTGCCGTCTGGGTCTGAAGGAAGGCCGTCCATTCTCAGGGGTCACTGCATGTTTGGACTTCTGTGCTCATGCCCACAGAGACTTGCACAACATGCAGAATGGCAGCACATTGGTATGCACTCTCACTAGAGAAGACAATCGAGAATTTGGAGGAAAACCTGAGGATGAGCAGCTTCACGTTCTGCCTTTATACAAAGTCTCTGACGTGGATGAGTTTGGGAGTGTGGAAGCTCAGGAGGAGAAAAAACGGAGTGGTGCCATTCAGGTACTGAGTTCTTTTCGGCGAAAAGTCAGGATGTTAGCAGAGCCAGTCAAGACTTGCCGACAAAGGAAACTAGAAGCCAAGAAAGCTGCAGCTGAAAAGCTTTCCTCCCTGGAGAACAGCTCAAATAAAAATGAAAAGGAAAAGTCAGCCCCATCACGTACAAAACAAACTGAAAACGCAAGCCAGGCTAAACAGTTGGCAGAACTTTTGCGACTTTCAGGACCAGTCATGCAGCAGTCCCAGCAGCCCCAGCCTCTACAGAAGCAGCCACCACAGCCCCAGCAGCAGCAGAGACCCCAGCAGCAGCAGCCACATCACCCTCAGACAGAGTCTGTCAACTCTTATTCTGCTTCTGGATCCACCAATCCATACATGAGACGGCCCAATCCAGTTAGTCCTTATCCAAACTCTTCACACACTTCAGATATCTATGGAAGCACCAGCCCTATGAACTTCTATTCCACCTCATCTCAAGCTGCAGGTTCATATTTGAATTCTTCTAATCCCATGAACCCTTACCCTGGGCTTTTGAATCAGAATACCCAATATCCATCATATCAATGCAATGGAAACCTATCAGTGGACAACTGCTCCCCATATCTGGGTTCCTATTCTCCCCAGTCTCAGCCGATGGATCTGTATAGGTATCCAAGCCAAGACCCTCTGTCTAAGCTCAGTCTACCACCCATCCATACACTTTACCAGCCAAGGTTTGGAAATAGCCAGAGTTTTACATCTAAATACTTAGGTTATGGAAACCAAAATATGCAGGGAGATGGTTTCAGCAGTTGTACCATTAGACCAAATGTACATCATGTAGGGAAATTGCCTCCTTATCCCACTCATGAGATGGATGGCCACTTCATGGGAGCCACCTCTAGATTACCACCCAATCTGAGCAATCCAAACATGGACTATAAAAATGGTGAACATCATTCACCTTCTCACATAATCCATAACTACAGTGCAGCTCCGGGCATGTTCAACAGCTCTCTTCATGCCCTGCATCTCCAAAACAAGGAGAATGACATGCTTTCCCACACAGCTAATGGGTTATCAAAGATGCTTCCAGCTCTTAACCATGATAGAACTGCTTGTGTCCAAGGAGGCTTACACAAATTAAGTGATGCTAATGGTCAGGAAAAGCAGCCATTGGCACTAGTCCAGGGTGTGGCTTCTGGTGCAGAGGACAACGATGAGGTCTGGTCAGACAGCGAGCAGAGCTTTCTGGATCCTGACATTGGGGGAGTGGCCGTGGCTCCAACTCATGGGTCAATTCTCATTGAGTGTGCAAAGCGTGAGCTGCATGCCACAACCCCTTTAAAGAATCCCAATAGGAATCACCCCACCAGGATCTCCCTCGTCTTTTACCAGCATAAGAGCATGAATGAGCCAAAACATGGCTTGGCTCTTTGGGAAGCCAAAATGGCTGAAAAAGCCCGTGAGAAAGAGGAAGAGTGTGAAAAGTATGGCCCAGACTATGTGCCTCAGAAATCCCATGGCAAAAAAGTGAAACGGGAGCCTGCTGAGCCACATGAAACTTCAGAGCCCACTTACCTGCGTTTCATCAAGTCTCTTGCCGAAAGGACCATGTCCGTGACCACAGACTCCACAGTAACTACATCTCCATATGCCTTCACTCGGGTCACAGGGCCTTACAACAGATATATATGAT'\n",
    "\n",
    "    #List of all amino acid positions (across whole coding region) covered (amino acid position minus 1 (as python starts at 0))\n",
    "    TET2_missense_codons_DICT_NP_01120680={}\n",
    "    for i in range(lower_range, upper_range):\n",
    "        TET2_missense_codons_DICT_NP_01120680[i+1]= (TET2_codons_NP_01120680[i])\n",
    "\n",
    "    calls_TET2_missense_NP_01120680 = {}\n",
    "    for k,v in TET2_missense_codons_DICT_NP_01120680.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_TET2_missense_NP_01120680[k]=(v, changes)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_TET2_missense_NP_01120680_extra = {}\n",
    "\n",
    "    for k, v in calls_TET2_missense_NP_01120680.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_TET2_missense_NP_01120680_extra[k] = (Amino_acid_change, v[0], v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_TET2_missense_NP_01120680_extra2 = {}\n",
    "\n",
    "    for k, v in calls_TET2_missense_NP_01120680_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_TET2_missense_NP_01120680_extra2[k] = (Amino_acid_abbrev, v[1], v[2])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary that contains details of the positions of missense variants at that codon\n",
    "    TET2_position_in_list_dict_missense_NP_01120680 = {}\n",
    "\n",
    "    for k, v in calls_TET2_missense_NP_01120680_extra2.items():\n",
    "        original_AA = v[0]\n",
    "        possible_AA_changes = v[2]\n",
    "        possible_AA_changes_string = ''.join(v[2])\n",
    "        positions_in_list = []\n",
    "        for i in possible_AA_changes:\n",
    "            if i != original_AA: #not synonymous\n",
    "                if i != '*': #not missense\n",
    "                    position_in_list=-1\n",
    "                    missense_change = i\n",
    "                    new_aa_list = []\n",
    "                    while True:\n",
    "                        position_in_list = possible_AA_changes_string.find(missense_change, position_in_list+1)\n",
    "        #               print(position_in_list)\n",
    "                        if position_in_list == -1: break\n",
    "                        positions_in_list.append(position_in_list)\n",
    "                        positions_in_list = set(positions_in_list) #convert into a 'set' to remove duplicate positions\n",
    "                        positions_in_list = list(positions_in_list) #convert back to a list so can be added to dictionary\n",
    "                    for pos in positions_in_list:\n",
    "                        new_aa = possible_AA_changes_string[pos]\n",
    "                        new_aa_list.append(new_aa)\n",
    "                    TET2_position_in_list_dict_missense_NP_01120680[k]=(positions_in_list, v, new_aa_list)\n",
    "\n",
    "        aa_positions_variants_called = []\n",
    "        for k, v in TET2_position_in_list_dict_missense_NP_01120680.items():\n",
    "            aa_positions_variants_called.append(k)\n",
    "    # print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TET2_bases_missense_NP_01120680 = {}\n",
    "    for k, v in TET2_position_in_list_dict_missense_NP_01120680.items():\n",
    "        original_a_a = v[1][0]\n",
    "        codon_seq = v[1][1]\n",
    "        base_changes=[]\n",
    "        new_a_a_list = v[2]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TET2_bases_missense_NP_01120680[k]=(v, base_changes, original_a_a, new_a_a_list)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TET2_bases_missense2_NP_01120680 = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense_NP_01120680.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        new_a_a_list = v[3]\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TET2_bases_missense2_NP_01120680[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TET2_bases_missense3_NP_01120680 = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense2_NP_01120680.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_a_a = v[0][2]\n",
    "        new_aa_list = v[0][0][2]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = k\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(k)-1)*3)+codon_pos #codon number minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][1][1][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TET2_bases_missense3_NP_01120680[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, original_a_a, new_aa_list)\n",
    "\n",
    "    # Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TET2_bases_missense4_NP_01120680 = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense3_NP_01120680.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "        new_a_a_list = v[6]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_TET2_NP01120680:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check_TET2_NP01120680(i)\n",
    "                sites_trio_missense.append([adapted_trio])\n",
    "                TET2_bases_missense4_NP_01120680[k] = (v, sites_trio_missense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_missense = [TET2_exons_extra_bases_NP_01120680[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_missense.append(site_trio_missense)\n",
    "                    TET2_bases_missense4_NP_01120680[k] = (v, sites_trio_missense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TET2_sites_missense5_NP_01120680 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TET2_bases_missense4_NP_01120680.items():\n",
    "        original_a_a = v[0][0][0]\n",
    "        new_a_a = v[0][6]\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            new_amino_acid = new_a_a[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, original_a_a, new_amino_acid])\n",
    "            change = change+1\n",
    "        TET2_sites_missense5_NP_01120680[a]=info\n",
    "        a=a+1\n",
    "    #     print(TET2_sites_missense5_NP_01120680)\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TET2_sites_missense6_NP_01120680={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TET2_sites_missense5_NP_01120680.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        original_a_a = v[5]\n",
    "        new_a_a = v[6]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TET2_sites_missense6_NP_01120680[k] = (v[0],changes3, codon_n_912, original_a_a, new_a_a) #site_trio = the affected base and its neighbours\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    TET2_mutation_rates_missense_NP_01120680 = {}\n",
    "    for k, v in TET2_sites_missense6_NP_01120680.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            original_a_a = v[3]\n",
    "            new_a_a = v[4]\n",
    "            aa_positioning = v[2]\n",
    "            variant_id = str(original_a_a)+str(aa_positioning)+str(new_a_a)\n",
    "            TET2_mutation_rates_missense_NP_01120680[k] = (v[0],v[1], mutation_rate, variant_id)\n",
    "\n",
    "    TET2_mutation_rate_missense_NP_01120680_sum = []\n",
    "\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TET2_mutation_rates_missense_NP_01120680.items():\n",
    "        codon_position_in_gene=(v[0])\n",
    "        variant = v[3]\n",
    "        site = v[1]\n",
    "        mutation_rate_value = v[2] \n",
    "        mutation_rates_list.append((v[0], variant, site, mutation_rate_value))\n",
    "        TET2_mutation_rate_missense_NP_01120680_sum.append(mutation_rate_value)\n",
    "\n",
    "    #Get the average of all the mutation rates\n",
    "    TET2_mutation_rate_missense_NP_01120680_sum = []\n",
    "    for k, v in TET2_mutation_rates_missense_NP_01120680.items():\n",
    "        mut_rate = v[2]\n",
    "        TET2_mutation_rate_missense_NP_01120680_sum.append(mut_rate)\n",
    "\n",
    "    TET2_Summed_mutation_rate_missense_NP_01120680_added = np.sum(TET2_mutation_rate_missense_NP_01120680_sum)\n",
    "\n",
    "    print('Total summed mutation rate across TET2 missense_NP_01120680 for this study = ', TET2_Summed_mutation_rate_missense_NP_01120680_added)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all # e.g. lower= 0, upper = 913 for whole gene\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_nonsense_range_of_variants_TET2_shorter(lower_range, upper_range):\n",
    "    \n",
    "    #TET2 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Summary?db=core;g=ENSG00000168769;r=4:106067032-106200973;t=ENST00000305737\n",
    "    TET2_exons_NP_060098 = 'ATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGGTAAGTGCCAGAAATGTACTGAGACACATGGCGTTTATCCAGAATTAGCAAATTTATCTTCAGATATGGGATTTTCCTTCTTTTTTTAA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    TET2_codons_NP_060098 = [TET2_exons_NP_060098[i:i+3] for i in range(0, len(TET2_exons_NP_060098), 3)]\n",
    "\n",
    "    #TET2 transcript 1 extra base at each end:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Summary?db=core;g=ENSG00000168769;r=4:106067032-106200973;t=ENST00000305737\n",
    "    TET2_exons_extra_bases_NP_060098 = 'GATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGGTAAGTGCCAGAAATGTACTGAGACACATGGCGTTTATCCAGAATTAGCAAATTTATCTTCAGATATGGGATTTTCCTTCTTTTTTTAAA'\n",
    "\n",
    "    #List of all amino acid positions (across whole coding region) covered (amino acid position minus 1 (as python starts at 0))\n",
    "    TET2_nonsense_codons_DICT_NP_060098={}\n",
    "    for i in range(lower_range, upper_range):\n",
    "        TET2_nonsense_codons_DICT_NP_060098[i+1]= (TET2_codons_NP_060098[i])\n",
    "\n",
    "    calls_TET2_nonsense_NP_060098 = {}\n",
    "    for k,v in TET2_nonsense_codons_DICT_NP_060098.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_TET2_nonsense_NP_060098[k]=(v, changes)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_TET2_nonsense_NP_060098_extra = {}\n",
    "\n",
    "    for k, v in calls_TET2_nonsense_NP_060098.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_TET2_nonsense_NP_060098_extra[k] = (Amino_acid_change, v[0], v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_TET2_nonsense_NP_060098_extra2 = {}\n",
    "\n",
    "    for k, v in calls_TET2_nonsense_NP_060098_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_TET2_nonsense_NP_060098_extra2[k] = (Amino_acid_abbrev, v[1], v[2])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary that contains details of the positions of nonsense variants at that codon\n",
    "    TET2_position_in_list_dict_nonsense_NP_060098 = {}\n",
    "\n",
    "    for k, v in calls_TET2_nonsense_NP_060098_extra2.items():\n",
    "        original_AA = v[0]\n",
    "        possible_AA_changes = v[2]\n",
    "        possible_AA_changes_string = ''.join(v[2])\n",
    "        positions_in_list = []\n",
    "        for i in possible_AA_changes:\n",
    "            if i == '*': #nonsense\n",
    "                position_in_list=-1\n",
    "                nonsense_change = i    \n",
    "                while True:\n",
    "                    position_in_list = possible_AA_changes_string.find(nonsense_change, position_in_list+1)\n",
    "    #               print(position_in_list)\n",
    "                    if position_in_list == -1: break\n",
    "                    positions_in_list.append(position_in_list)\n",
    "                    positions_in_list = set(positions_in_list) #convert into a 'set' to remove duplicate positions\n",
    "                    positions_in_list = list(positions_in_list) #convert back to a list so can be added to dictionary\n",
    "                TET2_position_in_list_dict_nonsense_NP_060098[k]=(positions_in_list, v)\n",
    "\n",
    "        aa_positions_variants_called = []\n",
    "        for k, v in TET2_position_in_list_dict_nonsense_NP_060098.items():\n",
    "            aa_positions_variants_called.append(k)\n",
    "    # print('positions at which nonsense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TET2_bases_nonsense_NP_060098 = {}\n",
    "    for k, v in TET2_position_in_list_dict_nonsense_NP_060098.items():\n",
    "        original_a_a = v[1][0]\n",
    "        codon_seq = v[1][1]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TET2_bases_nonsense_NP_060098[k]=(v, base_changes, original_a_a)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TET2_bases_nonsense2_NP_060098 = {}\n",
    "\n",
    "    for k, v in TET2_bases_nonsense_NP_060098.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TET2_bases_nonsense2_NP_060098[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TET2_bases_nonsense3_NP_060098 = {}\n",
    "\n",
    "    for k, v in TET2_bases_nonsense2_NP_060098.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_a_a = v[0][2]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = k\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(k)-1)*3)+codon_pos #codon number minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][1][1][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TET2_bases_nonsense3_NP_060098[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, original_a_a)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TET2_bases_nonsense4_NP_060098 = {}\n",
    "\n",
    "    for k, v in TET2_bases_nonsense3_NP_060098.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_nonsense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "            gene_pos = i\n",
    "            site_trio_nonsense = [TET2_exons_extra_bases_NP_060098[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "            sites_trio_nonsense.append(site_trio_nonsense)\n",
    "            TET2_bases_nonsense4_NP_060098[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TET2_sites_nonsense5_NP_060098 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TET2_bases_nonsense4_NP_060098.items():\n",
    "        original_a_a = v[0][0][0]\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, original_a_a])\n",
    "            change = change+1\n",
    "        TET2_sites_nonsense5_NP_060098[a]=info\n",
    "        a=a+1\n",
    "    #     print(TET2_sites_nonsense5_NP_060098)\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TET2_sites_nonsense6_NP_060098={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TET2_sites_nonsense5_NP_060098.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        original_a_a = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TET2_sites_nonsense6_NP_060098[k] = (v[0],changes3, codon_n_912, original_a_a) #site_trio = the affected base and its neighbours\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    TET2_mutation_rates_nonsense_NP_060098 = {}\n",
    "    for k, v in TET2_sites_nonsense6_NP_060098.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            original_a_a = v[3]\n",
    "            aa_positioning = v[2]\n",
    "            variant_type = '*'\n",
    "            variant_id = str(original_a_a)+str(aa_positioning)+variant_type\n",
    "            TET2_mutation_rates_nonsense_NP_060098[k] = (v[0],v[1], mutation_rate, variant_id)\n",
    "\n",
    "    TET2_mutation_rate_nonsense_NP_060098_sum = []\n",
    "\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TET2_mutation_rates_nonsense_NP_060098.items():\n",
    "        codon_position_in_gene=(v[0])\n",
    "        variant = v[3]\n",
    "        site = v[1]\n",
    "        mutation_rate_value = v[2] \n",
    "        mutation_rates_list.append((v[0], variant, site, mutation_rate_value))\n",
    "        TET2_mutation_rate_nonsense_NP_060098_sum.append(mutation_rate_value)\n",
    "\n",
    "    #Get the average of all the mutation rates\n",
    "    TET2_mutation_rate_nonsense_NP_060098_sum = []\n",
    "    for k, v in TET2_mutation_rates_nonsense_NP_060098.items():\n",
    "        mut_rate = v[2]\n",
    "        TET2_mutation_rate_nonsense_NP_060098_sum.append(mut_rate)\n",
    "\n",
    "    TET2_Summed_mutation_rate_nonsense_NP_060098_added = np.sum(TET2_mutation_rate_nonsense_NP_060098_sum)\n",
    "\n",
    "    print('Total summed mutation rate across TET2 nonsense_NP_060098 for this study = ', TET2_Summed_mutation_rate_nonsense_NP_060098_added)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all # e.g. lower= 0, upper = 913 for whole gene\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_missense_range_of_variants_TET2_shorter(lower_range, upper_range):\n",
    "    \n",
    "    #TET2 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Summary?db=core;g=ENSG00000168769;r=4:106067032-106200973;t=ENST00000305737\n",
    "    TET2_exons_NP_060098 = 'ATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGGTAAGTGCCAGAAATGTACTGAGACACATGGCGTTTATCCAGAATTAGCAAATTTATCTTCAGATATGGGATTTTCCTTCTTTTTTTAA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    TET2_codons_NP_060098 = [TET2_exons_NP_060098[i:i+3] for i in range(0, len(TET2_exons_NP_060098), 3)]\n",
    "\n",
    "    #TET2 transcript 1 extra base at each end:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Summary?db=core;g=ENSG00000168769;r=4:106067032-106200973;t=ENST00000305737\n",
    "    TET2_exons_extra_bases_NP_060098 = 'GATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGGTAAGTGCCAGAAATGTACTGAGACACATGGCGTTTATCCAGAATTAGCAAATTTATCTTCAGATATGGGATTTTCCTTCTTTTTTTAAA'\n",
    "\n",
    "    #List of all amino acid positions (across whole coding region) covered (amino acid position minus 1 (as python starts at 0))\n",
    "    TET2_missense_codons_DICT_NP_072046={}\n",
    "    for i in range(lower_range, upper_range):\n",
    "        TET2_missense_codons_DICT_NP_072046[i+1]= (TET2_codons_NP_072046[i])\n",
    "\n",
    "    calls_TET2_missense_NP_072046 = {}\n",
    "    for k,v in TET2_missense_codons_DICT_NP_072046.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_TET2_missense_NP_072046[k]=(v, changes)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_TET2_missense_NP_072046_extra = {}\n",
    "\n",
    "    for k, v in calls_TET2_missense_NP_072046.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_TET2_missense_NP_072046_extra[k] = (Amino_acid_change, v[0], v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_TET2_missense_NP_072046_extra2 = {}\n",
    "\n",
    "    for k, v in calls_TET2_missense_NP_072046_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_TET2_missense_NP_072046_extra2[k] = (Amino_acid_abbrev, v[1], v[2])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary that contains details of the positions of missense variants at that codon\n",
    "    TET2_position_in_list_dict_missense_NP_072046 = {}\n",
    "\n",
    "    for k, v in calls_TET2_missense_NP_072046_extra2.items():\n",
    "        original_AA = v[0]\n",
    "        possible_AA_changes = v[2]\n",
    "        possible_AA_changes_string = ''.join(v[2])\n",
    "        positions_in_list = []\n",
    "        for i in possible_AA_changes:\n",
    "            if i != original_AA: #not synonymous\n",
    "                if i != '*': #not missense\n",
    "                    position_in_list=-1\n",
    "                    missense_change = i\n",
    "                    new_aa_list = []\n",
    "                    while True:\n",
    "                        position_in_list = possible_AA_changes_string.find(missense_change, position_in_list+1)\n",
    "        #               print(position_in_list)\n",
    "                        if position_in_list == -1: break\n",
    "                        positions_in_list.append(position_in_list)\n",
    "                        positions_in_list = set(positions_in_list) #convert into a 'set' to remove duplicate positions\n",
    "                        positions_in_list = list(positions_in_list) #convert back to a list so can be added to dictionary\n",
    "                    for pos in positions_in_list:\n",
    "                        new_aa = possible_AA_changes_string[pos]\n",
    "                        new_aa_list.append(new_aa)\n",
    "                    TET2_position_in_list_dict_missense_NP_072046[k]=(positions_in_list, v, new_aa_list)\n",
    "\n",
    "        aa_positions_variants_called = []\n",
    "        for k, v in TET2_position_in_list_dict_missense_NP_072046.items():\n",
    "            aa_positions_variants_called.append(k)\n",
    "    # print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TET2_bases_missense_NP_072046 = {}\n",
    "    for k, v in TET2_position_in_list_dict_missense_NP_072046.items():\n",
    "        original_a_a = v[1][0]\n",
    "        codon_seq = v[1][1]\n",
    "        base_changes=[]\n",
    "        new_a_a_list = v[2]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TET2_bases_missense_NP_072046[k]=(v, base_changes, original_a_a, new_a_a_list)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TET2_bases_missense2_NP_072046 = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense_NP_072046.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        new_a_a_list = v[3]\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TET2_bases_missense2_NP_072046[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TET2_bases_missense3_NP_072046 = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense2_NP_072046.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_a_a = v[0][2]\n",
    "        new_aa_list = v[0][0][2]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = k\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(k)-1)*3)+codon_pos #codon number minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][1][1][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TET2_bases_missense3_NP_072046[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, original_a_a, new_aa_list)\n",
    "\n",
    "    # Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TET2_bases_missense4_NP_072046 = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense3_NP_072046.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "        new_a_a_list = v[6]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "            gene_pos = i\n",
    "            site_trio_missense = [TET2_exons_extra_bases_NP_072046[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "            sites_trio_missense.append(site_trio_missense)\n",
    "            TET2_bases_missense4_NP_072046[k] = (v, sites_trio_missense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TET2_sites_missense5_NP_072046 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TET2_bases_missense4_NP_072046.items():\n",
    "        original_a_a = v[0][0][0]\n",
    "        new_a_a = v[0][6]\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            new_amino_acid = new_a_a[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, original_a_a, new_amino_acid])\n",
    "            change = change+1\n",
    "        TET2_sites_missense5_NP_072046[a]=info\n",
    "        a=a+1\n",
    "    #     print(TET2_sites_missense5_NP_072046)\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TET2_sites_missense6_NP_072046={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TET2_sites_missense5_NP_072046.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        original_a_a = v[5]\n",
    "        new_a_a = v[6]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TET2_sites_missense6_NP_072046[k] = (v[0],changes3, codon_n_912, original_a_a, new_a_a) #site_trio = the affected base and its neighbours\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    TET2_mutation_rates_missense_NP_072046 = {}\n",
    "    for k, v in TET2_sites_missense6_NP_072046.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            original_a_a = v[3]\n",
    "            new_a_a = v[4]\n",
    "            aa_positioning = v[2]\n",
    "            variant_id = str(original_a_a)+str(aa_positioning)+str(new_a_a)\n",
    "            TET2_mutation_rates_missense_NP_072046[k] = (v[0],v[1], mutation_rate, variant_id)\n",
    "\n",
    "    TET2_mutation_rate_missense_NP_072046_sum = []\n",
    "\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TET2_mutation_rates_missense_NP_072046.items():\n",
    "        codon_position_in_gene=(v[0])\n",
    "        variant = v[3]\n",
    "        site = v[1]\n",
    "        mutation_rate_value = v[2] \n",
    "        mutation_rates_list.append((v[0], variant, site, mutation_rate_value))\n",
    "        TET2_mutation_rate_missense_NP_072046_sum.append(mutation_rate_value)\n",
    "\n",
    "    #Get the average of all the mutation rates\n",
    "    TET2_mutation_rate_missense_NP_072046_sum = []\n",
    "    for k, v in TET2_mutation_rates_missense_NP_072046.items():\n",
    "        mut_rate = v[2]\n",
    "        TET2_mutation_rate_missense_NP_072046_sum.append(mut_rate)\n",
    "\n",
    "    TET2_Summed_mutation_rate_missense_NP_072046_added = np.sum(TET2_mutation_rate_missense_NP_072046_sum)\n",
    "\n",
    "    print('Total summed mutation rate across TET2 missense_NP_072046 for this study = ', TET2_Summed_mutation_rate_missense_NP_072046_added)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all # e.g. lower= 0, upper = 913 for whole gene\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ASXL1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_nonsense_range_of_variants_ASXL1(lower_range, upper_range):\n",
    "    \n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base = [140, 143, 252, 373, 1086, 1085]\n",
    "   \n",
    "    def intron_exon_check(position_in_gene_number):\n",
    "        if position_in_gene_number == 140:\n",
    "            gene_pos = 140\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 143:\n",
    "            gene_pos = 143\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 252:\n",
    "            gene_pos = 252\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 373:\n",
    "            gene_pos = 373\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1086:\n",
    "            gene_pos = 1086\n",
    "            first_base = 'G'\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = ASXL1_exons_extra_bases[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1085:\n",
    "            gene_pos = 1085\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #ASXL1 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    ASXL1_exons = 'ATGAAGGACAAACAGAAGAAGAAGAAGGAGCGCACGTGGGCCGAGGCCGCGCGCCTGGTATTAGAAAACTACTCGGATGCTCCAATGACACCAAAACAGATTCTGCAGGTCATAGAGGCAGAAGGACTAAAGGAAATGAGAAGTGGGACTTCCCCTCTCGCATGCCTCAATGCTATGCTACATTCCAATTCAAGAGGAGGAGAGGGGTTGTTTTATAAACTGCCTGGCCGAATCAGCCTTTTCACGCTCAAGAAGGATGCCCTGCAGTGGTCTCGCCATCCAGCTACAGTGGAGGGAGAGGAGCCAGAGGACACGGCTGATGTGGAGAGCTGTGGGTCTAATGAAGCCAGCACTGTGAGTGGTGAAAACGATGTATCTCTTGATGAAACATCTTCGAACGCATCCTGTTCTACAGAATCTCAGAGTCGACCTCTTTCCAATCCCAGGGACAGCTACAGAGCTTCCTCACAGGCGAACAAACAAAAGAAAAAGACTGGGGTGATGCTGCCTCGAGTTGTCCTGACTCCTCTGAAGGTAAACGGGGCCCACGTGGAATCTGCATCAGGGTTCTCGGGCTGCCACGCCGATGGCGAGAGCGGCAGCCCGTCCAGCAGCAGCAGCGGCTCTCTGGCCCTGGGCAGCGCTGCTATTCGTGGCCAGGCCGAGGTCACCCAGGACCCTGCCCCGCTCCTGAGAGGCTTCCGGAAGCCAGCCACAGGTCAAATGAAGCGCAACAGAGGGGAAGAAATAGATTTTGAGACACCTGGGTCCATTCTTGTCAACACCAACCTCCGTGCCCTGATCAACTCTCGGACCTTCCATGCCTTACCATCACACTTCCAGCAGCAGCTCCTCTTCCTCCTGCCTGAAGTAGACAGACAGGTGGGGACGGATGGCCTGTTGCGTCTCAGCAGCAGTGCACTAAATAACGAGTTTTTTACCCATGCGGCTCAGAGCTGGCGGGAGCGCCTGGCTGATGGTGAATTTACTCATGAGATGCAAGTCAGGATACGACAGGAAATGGAGAAGGAAAAGAAGGTGGAACAATGGAAAGAAAAGTTCTTTGAAGACTACTATGGACAGAAGCTGGGTTTGACCAAAGAAGAGTCATTGCAGCAGAACGTGGGCCAGGAGGAGGCTGAAATCAAAAGTGGCTTGTGTGTCCCAGGAGAATCAGTGCGTATACAGCGTGGTCCAGCCACCCGACAGCGAGATGGGCATTTTAAGAAACGCTCTCGGCCAGATCTCCGAACCAGAGCCAGAAGGAATCTGTACAAAAAACAGGAGTCAGAACAAGCAGGGGTTGCTAAGGATGCAAAATCTGTGGCCTCAGATGTTCCCCTCTACAAGGATGGGGAGGCTAAGACTGACCCAGCAGGGCTGAGCAGTCCCCATCTGCCAGGCACATCCTCTGCAGCACCCGACCTGGAGGGTCCCGAATTCCCAGTTGAGTCTGTGGCTTCTCGGATCCAGGCTGAGCCAGACAACTTGGCACGTGCCTCTGCATCTCCAGACAGAATTCCTAGCCTGCCTCAGGAAACTGTGGATCAGGAACCCAAGGATCAGAAGAGGAAATCCTTTGAGCAGGCGGCCTCTGCATCCTTTCCCGAAAAGAAGCCCCGGCTTGAAGATCGTCAGTCCTTTCGTAACACAATTGAAAGTGTTCACACCGAAAAGCCACAGCCCACTAAAGAGGAGCCCAAAGTCCCGCCCATCCGGATTCAACTTTCACGTATCAAACCACCCTGGGTGGTTAAAGGTCAGCCCACTTACCAGATATGCCCCCGGATCATCCCCACCACGGAGTCCTCCTGCCGGGGTTGGACTGGCGCCAGGACCCTCGCAGACATTAAAGCCCGTGCTCTGCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCCACCACTGCCATCGGAGGGGGGGGTGGCCCGGGTGGAGGTGGCGGCGGGGCCACCGATGAGGGAGGTGGCAGAGGCAGCAGCAGTGGTGATGGTGGTGAGGCCTGTGGCCACCCTGAGCCCAGGGGAGGCCCGAGCACCCCTGGAAAGTGTACGTCAGATCTACAGCGAACACAACTACTGCCGCCTTATCCTCTAAATGGGGAGCATACCCAGGCCGGAACTGCCATGTCCAGAGCTAGGAGAGAGGACCTGCCTTCTCTGAGAAAGGAGGAAAGCTGCCTACTACAGAGGGCTACAGTTGGACTCACAGATGGGCTAGGAGATGCCTCCCAACTCCCCGTTGCTCCCACTGGGGACCAGCCATGCCAGGCCTTGCCCCTACTGTCCTCCCAAACCTCAGTAGCTGAGAGATTAGTGGAGCAGCCTCAGTTGCATCCGGATGTTAGAACTGAATGTGAGTCTGGCACCACTTCCTGGGAAAGTGATGATGAGGAGCAAGGACCCACCGTTCCTGCAGACAATGGTCCCATTCTGTCTCTAGTGGGAGATGATACATTAGAGAAAGGAACTGGCCAAGCTCTTGACAGTCATCCCACTATGAAGGATCCTGTAAATGTGACCCCCAGTTCCACACCTGAATCCTCACCGACTGATTGCCTGCAGAACAGAGCATTTGATGACGAATTAGGGCTTGGTGGCTCATGCCCTCCTATGAGGGAAAGTGATACTAGACAAGAAAACTTGAAAACCAAGGCTCTCGTTTCTAACAGTTCTTTGCATTGGATACCCATCCCATCGAATGATGAGGTAGTGAAACAGCCCAAACCAGAATCCAGAGAACACATACCATCTGTTGAGCCCCAGGTTGGAGAGGAGTGGGAGAAAGCTGCTCCCACCCCTCCTGCATTGCCTGGGGATTTGACAGCTGAGGAGGGTCTAGATCCTCTTGACAGCCTTACTTCACTCTGGACTGTGCCATCTCGAGGAGGCAGTGACAGCAATGGCAGTTACTGTCAACAGGTGGACATTGAAAAGCTGAAAATCAACGGAGACTCTGAAGCACTGAGTCCTCACGGTGAGTCCACGGATACAGCCTCTGACTTTGAAGGTCACCTCACGGAGGACAGCAGTGAGGCTGACACTAGAGAAGCTGCAGTGACAAAGGGATCTTCGGTGGACAAGGATGAGAAACCCAATTGGAACCAATCTGCCCCACTGTCCAAGGTGAATGGTGACATGCGTCTGGTTACAAGGACAGATGGGATGGTTGCTCCTCAGAGCTGGGTGTCTCGAGTATGTGCGGTCCGCCAAAAGATCCCAGATTCCCTACTGCTGGCCAGTACTGAGTACCAGCCAAGAGCCGTGTGCCTGTCCATGCCTGGGTCCTCAGTGGAGGCCACTAACCCACTTGTGATGCAGTTGCTGCAGGGTAGCTTGCCCCTAGAGAAGGTTCTTCCACCAGCCCACGATGACAGCATGTCAGAATCCCCACAAGTACCACTTACAAAAGACCAGAGCCATGGCTCGCTACGCATGGGATCTTTACATGGTCTTGGAAAAAACAGTGGCATGGTTGATGGAAGCAGCCCCAGTTCTTTAAGGGCTTTGAAGGAGCCTCTTCTGCCAGATAGCTGTGAAACAGGCACTGGTCTTGCCAGGATTGAGGCCACCCAGGCTCCTGGAGCACCCCAAAAGAATTGCAAGGCAGTCCCAAGTTTTGACTCCCTCCATCCAGTGACAAATCCCATTACATCCTCTAGGAAACTGGAAGAAATGGATTCCAAAGAGCAGTTCTCTTCCTTTAGTTGTGAAGATCAGAAGGAAGTCCGTGCTATGTCACAGGACAGTAATTCAAATGCTGCTCCAGGAAAGAGCCCAGGAGATCTTACTACCTCGAGAACACCTCGTTTCTCATCTCCAAATGTGATCTCCTTTGGTCCAGAGCAGACAGGTCGGGCCCTGGGTGATCAGAGCAATGTTACAGGCCAAGGGAAGAAGCTTTTTGGCTCTGGGAATGTGGCTGCAACCCTTCAGCGCCCCAGGCCTGCGGACCCGATGCCTCTTCCTGCTGAGATCCCTCCAGTTTTTCCCAGTGGGAAGTTGGGACCAAGCACAAACTCCATGTCTGGTGGGGTACAGACTCCAAGGGAAGACTGGGCTCCAAAGCCACATGCCTTTGTTGGCAGCGTCAAGAATGAGAAGACTTTTGTGGGGGGTCCTCTTAAGGCAAATGCCGAGAACAGGAAAGCTACTGGGCATAGTCCCCTGGAACTGGTGGGTCACTTGGAAGGGATGCCCTTTGTCATGGACTTGCCCTTCTGGAAATTACCCCGAGAGCCAGGGAAGGGGCTCAGTGAGCCTCTGGAGCCTTCTTCTCTCCCCTCCCAACTCAGCATCAAGCAGGCATTTTATGGGAAGCTTTCTAAACTCCAACTGAGTTCCACCAGCTTTAATTATTCCTCTAGCTCTCCCACCTTTCCCAAAGGCCTTGCTGGAAGTGTGGTGCAGCTGAGCCACAAAGCAAACTTTGGTGCGAGCCACAGTGCATCACTTTCCTTGCAAATGTTCACTGACAGCAGCACGGTGGAAAGCATCTCGCTCCAGTGTGCGTGCAGCCTGAAAGCCATGATCATGTGCCAAGGCTGCGGTGCGTTCTGTCACGATGACTGTATTGGACCCTCAAAGCTCTGTGTATTGTGCCTTGTGGTGAGATAA'\n",
    "    \n",
    "    #Split the sequence into codons \n",
    "    ASXL1_codons = [ASXL1_exons[i:i+3] for i in range(0, len(ASXL1_exons), 3)]\n",
    "\n",
    "    #ASXL1 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    ASXL1_exons_extra_bases = 'GATGAAGGACAAACAGAAGAAGAAGAAGGAGCGCACGTGGGCCGAGGCCGCGCGCCTGGTATTAGAAAACTACTCGGATGCTCCAATGACACCAAAACAGATTCTGCAGGTCATAGAGGCAGAAGGACTAAAGGAAATGAGAAGTGGGACTTCCCCTCTCGCATGCCTCAATGCTATGCTACATTCCAATTCAAGAGGAGGAGAGGGGTTGTTTTATAAACTGCCTGGCCGAATCAGCCTTTTCACGCTCAAGAAGGATGCCCTGCAGTGGTCTCGCCATCCAGCTACAGTGGAGGGAGAGGAGCCAGAGGACACGGCTGATGTGGAGAGCTGTGGGTCTAATGAAGCCAGCACTGTGAGTGGTGAAAACGATGTATCTCTTGATGAAACATCTTCGAACGCATCCTGTTCTACAGAATCTCAGAGTCGACCTCTTTCCAATCCCAGGGACAGCTACAGAGCTTCCTCACAGGCGAACAAACAAAAGAAAAAGACTGGGGTGATGCTGCCTCGAGTTGTCCTGACTCCTCTGAAGGTAAACGGGGCCCACGTGGAATCTGCATCAGGGTTCTCGGGCTGCCACGCCGATGGCGAGAGCGGCAGCCCGTCCAGCAGCAGCAGCGGCTCTCTGGCCCTGGGCAGCGCTGCTATTCGTGGCCAGGCCGAGGTCACCCAGGACCCTGCCCCGCTCCTGAGAGGCTTCCGGAAGCCAGCCACAGGTCAAATGAAGCGCAACAGAGGGGAAGAAATAGATTTTGAGACACCTGGGTCCATTCTTGTCAACACCAACCTCCGTGCCCTGATCAACTCTCGGACCTTCCATGCCTTACCATCACACTTCCAGCAGCAGCTCCTCTTCCTCCTGCCTGAAGTAGACAGACAGGTGGGGACGGATGGCCTGTTGCGTCTCAGCAGCAGTGCACTAAATAACGAGTTTTTTACCCATGCGGCTCAGAGCTGGCGGGAGCGCCTGGCTGATGGTGAATTTACTCATGAGATGCAAGTCAGGATACGACAGGAAATGGAGAAGGAAAAGAAGGTGGAACAATGGAAAGAAAAGTTCTTTGAAGACTACTATGGACAGAAGCTGGGTTTGACCAAAGAAGAGTCATTGCAGCAGAACGTGGGCCAGGAGGAGGCTGAAATCAAAAGTGGCTTGTGTGTCCCAGGAGAATCAGTGCGTATACAGCGTGGTCCAGCCACCCGACAGCGAGATGGGCATTTTAAGAAACGCTCTCGGCCAGATCTCCGAACCAGAGCCAGAAGGAATCTGTACAAAAAACAGGAGTCAGAACAAGCAGGGGTTGCTAAGGATGCAAAATCTGTGGCCTCAGATGTTCCCCTCTACAAGGATGGGGAGGCTAAGACTGACCCAGCAGGGCTGAGCAGTCCCCATCTGCCAGGCACATCCTCTGCAGCACCCGACCTGGAGGGTCCCGAATTCCCAGTTGAGTCTGTGGCTTCTCGGATCCAGGCTGAGCCAGACAACTTGGCACGTGCCTCTGCATCTCCAGACAGAATTCCTAGCCTGCCTCAGGAAACTGTGGATCAGGAACCCAAGGATCAGAAGAGGAAATCCTTTGAGCAGGCGGCCTCTGCATCCTTTCCCGAAAAGAAGCCCCGGCTTGAAGATCGTCAGTCCTTTCGTAACACAATTGAAAGTGTTCACACCGAAAAGCCACAGCCCACTAAAGAGGAGCCCAAAGTCCCGCCCATCCGGATTCAACTTTCACGTATCAAACCACCCTGGGTGGTTAAAGGTCAGCCCACTTACCAGATATGCCCCCGGATCATCCCCACCACGGAGTCCTCCTGCCGGGGTTGGACTGGCGCCAGGACCCTCGCAGACATTAAAGCCCGTGCTCTGCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCCACCACTGCCATCGGAGGGGGGGGTGGCCCGGGTGGAGGTGGCGGCGGGGCCACCGATGAGGGAGGTGGCAGAGGCAGCAGCAGTGGTGATGGTGGTGAGGCCTGTGGCCACCCTGAGCCCAGGGGAGGCCCGAGCACCCCTGGAAAGTGTACGTCAGATCTACAGCGAACACAACTACTGCCGCCTTATCCTCTAAATGGGGAGCATACCCAGGCCGGAACTGCCATGTCCAGAGCTAGGAGAGAGGACCTGCCTTCTCTGAGAAAGGAGGAAAGCTGCCTACTACAGAGGGCTACAGTTGGACTCACAGATGGGCTAGGAGATGCCTCCCAACTCCCCGTTGCTCCCACTGGGGACCAGCCATGCCAGGCCTTGCCCCTACTGTCCTCCCAAACCTCAGTAGCTGAGAGATTAGTGGAGCAGCCTCAGTTGCATCCGGATGTTAGAACTGAATGTGAGTCTGGCACCACTTCCTGGGAAAGTGATGATGAGGAGCAAGGACCCACCGTTCCTGCAGACAATGGTCCCATTCTGTCTCTAGTGGGAGATGATACATTAGAGAAAGGAACTGGCCAAGCTCTTGACAGTCATCCCACTATGAAGGATCCTGTAAATGTGACCCCCAGTTCCACACCTGAATCCTCACCGACTGATTGCCTGCAGAACAGAGCATTTGATGACGAATTAGGGCTTGGTGGCTCATGCCCTCCTATGAGGGAAAGTGATACTAGACAAGAAAACTTGAAAACCAAGGCTCTCGTTTCTAACAGTTCTTTGCATTGGATACCCATCCCATCGAATGATGAGGTAGTGAAACAGCCCAAACCAGAATCCAGAGAACACATACCATCTGTTGAGCCCCAGGTTGGAGAGGAGTGGGAGAAAGCTGCTCCCACCCCTCCTGCATTGCCTGGGGATTTGACAGCTGAGGAGGGTCTAGATCCTCTTGACAGCCTTACTTCACTCTGGACTGTGCCATCTCGAGGAGGCAGTGACAGCAATGGCAGTTACTGTCAACAGGTGGACATTGAAAAGCTGAAAATCAACGGAGACTCTGAAGCACTGAGTCCTCACGGTGAGTCCACGGATACAGCCTCTGACTTTGAAGGTCACCTCACGGAGGACAGCAGTGAGGCTGACACTAGAGAAGCTGCAGTGACAAAGGGATCTTCGGTGGACAAGGATGAGAAACCCAATTGGAACCAATCTGCCCCACTGTCCAAGGTGAATGGTGACATGCGTCTGGTTACAAGGACAGATGGGATGGTTGCTCCTCAGAGCTGGGTGTCTCGAGTATGTGCGGTCCGCCAAAAGATCCCAGATTCCCTACTGCTGGCCAGTACTGAGTACCAGCCAAGAGCCGTGTGCCTGTCCATGCCTGGGTCCTCAGTGGAGGCCACTAACCCACTTGTGATGCAGTTGCTGCAGGGTAGCTTGCCCCTAGAGAAGGTTCTTCCACCAGCCCACGATGACAGCATGTCAGAATCCCCACAAGTACCACTTACAAAAGACCAGAGCCATGGCTCGCTACGCATGGGATCTTTACATGGTCTTGGAAAAAACAGTGGCATGGTTGATGGAAGCAGCCCCAGTTCTTTAAGGGCTTTGAAGGAGCCTCTTCTGCCAGATAGCTGTGAAACAGGCACTGGTCTTGCCAGGATTGAGGCCACCCAGGCTCCTGGAGCACCCCAAAAGAATTGCAAGGCAGTCCCAAGTTTTGACTCCCTCCATCCAGTGACAAATCCCATTACATCCTCTAGGAAACTGGAAGAAATGGATTCCAAAGAGCAGTTCTCTTCCTTTAGTTGTGAAGATCAGAAGGAAGTCCGTGCTATGTCACAGGACAGTAATTCAAATGCTGCTCCAGGAAAGAGCCCAGGAGATCTTACTACCTCGAGAACACCTCGTTTCTCATCTCCAAATGTGATCTCCTTTGGTCCAGAGCAGACAGGTCGGGCCCTGGGTGATCAGAGCAATGTTACAGGCCAAGGGAAGAAGCTTTTTGGCTCTGGGAATGTGGCTGCAACCCTTCAGCGCCCCAGGCCTGCGGACCCGATGCCTCTTCCTGCTGAGATCCCTCCAGTTTTTCCCAGTGGGAAGTTGGGACCAAGCACAAACTCCATGTCTGGTGGGGTACAGACTCCAAGGGAAGACTGGGCTCCAAAGCCACATGCCTTTGTTGGCAGCGTCAAGAATGAGAAGACTTTTGTGGGGGGTCCTCTTAAGGCAAATGCCGAGAACAGGAAAGCTACTGGGCATAGTCCCCTGGAACTGGTGGGTCACTTGGAAGGGATGCCCTTTGTCATGGACTTGCCCTTCTGGAAATTACCCCGAGAGCCAGGGAAGGGGCTCAGTGAGCCTCTGGAGCCTTCTTCTCTCCCCTCCCAACTCAGCATCAAGCAGGCATTTTATGGGAAGCTTTCTAAACTCCAACTGAGTTCCACCAGCTTTAATTATTCCTCTAGCTCTCCCACCTTTCCCAAAGGCCTTGCTGGAAGTGTGGTGCAGCTGAGCCACAAAGCAAACTTTGGTGCGAGCCACAGTGCATCACTTTCCTTGCAAATGTTCACTGACAGCAGCACGGTGGAAAGCATCTCGCTCCAGTGTGCGTGCAGCCTGAAAGCCATGATCATGTGCCAAGGCTGCGGTGCGTTCTGTCACGATGACTGTATTGGACCCTCAAAGCTCTGTGTATTGTGCCTTGTGGTGAGATAAT'\n",
    "  \n",
    "    #List of all amino acid positions (across whole coding region) covered by (amino acid position minus 1 (as python starts at 0))\n",
    "    ASXL1_nonsense_codons_DICT={}\n",
    "    for i in range(lower_range, upper_range):\n",
    "        ASXL1_nonsense_codons_DICT[i+1]= (ASXL1_codons[i])\n",
    "\n",
    "    calls_ASXL1_nonsense = {}\n",
    "    for k,v in ASXL1_nonsense_codons_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_ASXL1_nonsense[k]=(v, changes)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_ASXL1_nonsense_extra = {}\n",
    "\n",
    "    for k, v in calls_ASXL1_nonsense.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_ASXL1_nonsense_extra[k] = (Amino_acid_change, v[0], v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_ASXL1_nonsense_extra2 = {}\n",
    "\n",
    "    for k, v in calls_ASXL1_nonsense_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_ASXL1_nonsense_extra2[k] = (Amino_acid_abbrev, v[1], v[2])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary that contains details of the positions of nonsense variants at that codon\n",
    "    ASXL1_position_in_list_dict_nonsense = {}\n",
    "\n",
    "    for k, v in calls_ASXL1_nonsense_extra2.items():\n",
    "        original_AA = v[0]\n",
    "        possible_AA_changes = v[2]\n",
    "        possible_AA_changes_string = ''.join(v[2])\n",
    "        positions_in_list = []\n",
    "        for i in possible_AA_changes:\n",
    "            if i == '*': #nonsense\n",
    "                position_in_list=-1\n",
    "                nonsense_change = i    \n",
    "                while True:\n",
    "                    position_in_list = possible_AA_changes_string.find(nonsense_change, position_in_list+1)\n",
    "    #               print(position_in_list)\n",
    "                    if position_in_list == -1: break\n",
    "                    positions_in_list.append(position_in_list)\n",
    "                    positions_in_list = set(positions_in_list) #convert into a 'set' to remove duplicate positions\n",
    "                    positions_in_list = list(positions_in_list) #convert back to a list so can be added to dictionary\n",
    "                ASXL1_position_in_list_dict_nonsense[k]=(positions_in_list, v)\n",
    "\n",
    "        aa_positions_variants_called = []\n",
    "        for k, v in ASXL1_position_in_list_dict_nonsense.items():\n",
    "            aa_positions_variants_called.append(k)\n",
    "    # print('positions at which nonsense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    ASXL1_bases_nonsense = {}\n",
    "    for k, v in ASXL1_position_in_list_dict_nonsense.items():\n",
    "        original_a_a = v[1][0]\n",
    "        codon_seq = v[1][1]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        ASXL1_bases_nonsense[k]=(v, base_changes, original_a_a)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    ASXL1_bases_nonsense2 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_nonsense.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        ASXL1_bases_nonsense2[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    ASXL1_bases_nonsense3 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_nonsense2.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_a_a = v[0][2]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = k\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(k)-1)*3)+codon_pos #codon number minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][1][1][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        ASXL1_bases_nonsense3[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, original_a_a)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    ASXL1_bases_nonsense4 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_nonsense3.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_nonsense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check(i)\n",
    "                sites_trio_nonsense.append([adapted_trio])\n",
    "                ASXL1_bases_nonsense4[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_nonsense = [ASXL1_exons_extra_bases[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_nonsense.append(site_trio_nonsense)\n",
    "                    ASXL1_bases_nonsense4[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    ASXL1_sites_nonsense5 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in ASXL1_bases_nonsense4.items():\n",
    "        original_a_a = v[0][0][0]\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, original_a_a])\n",
    "            change = change+1\n",
    "        ASXL1_sites_nonsense5[a]=info\n",
    "        a=a+1\n",
    "    #     print(ASXL1_sites_nonsense5)\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    ASXL1_sites_nonsense6={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in ASXL1_sites_nonsense5.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        original_a_a = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        ASXL1_sites_nonsense6[k] = (v[0],changes3, codon_n_912, original_a_a) #site_trio = the affected base and its neighbours\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    ASXL1_mutation_rates_nonsense = {}\n",
    "    for k, v in ASXL1_sites_nonsense6.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            original_a_a = v[3]\n",
    "            aa_positioning = v[2]\n",
    "            variant_type = '*'\n",
    "            variant_id = str(original_a_a)+str(aa_positioning)+variant_type\n",
    "            ASXL1_mutation_rates_nonsense[k] = (v[0],v[1], mutation_rate, variant_id)\n",
    "\n",
    "    ASXL1_mutation_rate_nonsense_sum = []\n",
    "\n",
    "    mutation_rates_list = []\n",
    "    for k, v in ASXL1_mutation_rates_nonsense.items():\n",
    "        codon_position_in_gene=(v[0])\n",
    "        variant = v[3]\n",
    "        site = v[1]\n",
    "        mutation_rate_value = v[2] \n",
    "        mutation_rates_list.append((v[0], variant, site, mutation_rate_value))\n",
    "        ASXL1_mutation_rate_nonsense_sum.append(mutation_rate_value)\n",
    "\n",
    "    #Get the average of all the mutation rates\n",
    "    ASXL1_mutation_rate_nonsense_sum = []\n",
    "    for k, v in ASXL1_mutation_rates_nonsense.items():\n",
    "        mut_rate = v[2]\n",
    "        ASXL1_mutation_rate_nonsense_sum.append(mut_rate)\n",
    "\n",
    "    ASXL1_Summed_mutation_rate_nonsense_added = np.sum(ASXL1_mutation_rate_nonsense_sum)\n",
    "\n",
    "    print('Total summed mutation rate across ASXL1 nonsense for this study = ', ASXL1_Summed_mutation_rate_nonsense_added)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all # e.g. lower= 0, upper = 913 for whole gene\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_missense_range_of_variants_ASXL1(lower_range, upper_range):\n",
    "\n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base = [140, 143, 252, 373, 1086, 1085]\n",
    "   \n",
    "    def intron_exon_check(position_in_gene_number):\n",
    "        if position_in_gene_number == 140:\n",
    "            gene_pos = 140\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 143:\n",
    "            gene_pos = 143\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 252:\n",
    "            gene_pos = 252\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 373:\n",
    "            gene_pos = 373\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1086:\n",
    "            gene_pos = 1086\n",
    "            first_base = 'G'\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = ASXL1_exons_extra_bases[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1085:\n",
    "            gene_pos = 1085\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #ASXL1 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    ASXL1_exons = 'ATGAAGGACAAACAGAAGAAGAAGAAGGAGCGCACGTGGGCCGAGGCCGCGCGCCTGGTATTAGAAAACTACTCGGATGCTCCAATGACACCAAAACAGATTCTGCAGGTCATAGAGGCAGAAGGACTAAAGGAAATGAGAAGTGGGACTTCCCCTCTCGCATGCCTCAATGCTATGCTACATTCCAATTCAAGAGGAGGAGAGGGGTTGTTTTATAAACTGCCTGGCCGAATCAGCCTTTTCACGCTCAAGAAGGATGCCCTGCAGTGGTCTCGCCATCCAGCTACAGTGGAGGGAGAGGAGCCAGAGGACACGGCTGATGTGGAGAGCTGTGGGTCTAATGAAGCCAGCACTGTGAGTGGTGAAAACGATGTATCTCTTGATGAAACATCTTCGAACGCATCCTGTTCTACAGAATCTCAGAGTCGACCTCTTTCCAATCCCAGGGACAGCTACAGAGCTTCCTCACAGGCGAACAAACAAAAGAAAAAGACTGGGGTGATGCTGCCTCGAGTTGTCCTGACTCCTCTGAAGGTAAACGGGGCCCACGTGGAATCTGCATCAGGGTTCTCGGGCTGCCACGCCGATGGCGAGAGCGGCAGCCCGTCCAGCAGCAGCAGCGGCTCTCTGGCCCTGGGCAGCGCTGCTATTCGTGGCCAGGCCGAGGTCACCCAGGACCCTGCCCCGCTCCTGAGAGGCTTCCGGAAGCCAGCCACAGGTCAAATGAAGCGCAACAGAGGGGAAGAAATAGATTTTGAGACACCTGGGTCCATTCTTGTCAACACCAACCTCCGTGCCCTGATCAACTCTCGGACCTTCCATGCCTTACCATCACACTTCCAGCAGCAGCTCCTCTTCCTCCTGCCTGAAGTAGACAGACAGGTGGGGACGGATGGCCTGTTGCGTCTCAGCAGCAGTGCACTAAATAACGAGTTTTTTACCCATGCGGCTCAGAGCTGGCGGGAGCGCCTGGCTGATGGTGAATTTACTCATGAGATGCAAGTCAGGATACGACAGGAAATGGAGAAGGAAAAGAAGGTGGAACAATGGAAAGAAAAGTTCTTTGAAGACTACTATGGACAGAAGCTGGGTTTGACCAAAGAAGAGTCATTGCAGCAGAACGTGGGCCAGGAGGAGGCTGAAATCAAAAGTGGCTTGTGTGTCCCAGGAGAATCAGTGCGTATACAGCGTGGTCCAGCCACCCGACAGCGAGATGGGCATTTTAAGAAACGCTCTCGGCCAGATCTCCGAACCAGAGCCAGAAGGAATCTGTACAAAAAACAGGAGTCAGAACAAGCAGGGGTTGCTAAGGATGCAAAATCTGTGGCCTCAGATGTTCCCCTCTACAAGGATGGGGAGGCTAAGACTGACCCAGCAGGGCTGAGCAGTCCCCATCTGCCAGGCACATCCTCTGCAGCACCCGACCTGGAGGGTCCCGAATTCCCAGTTGAGTCTGTGGCTTCTCGGATCCAGGCTGAGCCAGACAACTTGGCACGTGCCTCTGCATCTCCAGACAGAATTCCTAGCCTGCCTCAGGAAACTGTGGATCAGGAACCCAAGGATCAGAAGAGGAAATCCTTTGAGCAGGCGGCCTCTGCATCCTTTCCCGAAAAGAAGCCCCGGCTTGAAGATCGTCAGTCCTTTCGTAACACAATTGAAAGTGTTCACACCGAAAAGCCACAGCCCACTAAAGAGGAGCCCAAAGTCCCGCCCATCCGGATTCAACTTTCACGTATCAAACCACCCTGGGTGGTTAAAGGTCAGCCCACTTACCAGATATGCCCCCGGATCATCCCCACCACGGAGTCCTCCTGCCGGGGTTGGACTGGCGCCAGGACCCTCGCAGACATTAAAGCCCGTGCTCTGCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCCACCACTGCCATCGGAGGGGGGGGTGGCCCGGGTGGAGGTGGCGGCGGGGCCACCGATGAGGGAGGTGGCAGAGGCAGCAGCAGTGGTGATGGTGGTGAGGCCTGTGGCCACCCTGAGCCCAGGGGAGGCCCGAGCACCCCTGGAAAGTGTACGTCAGATCTACAGCGAACACAACTACTGCCGCCTTATCCTCTAAATGGGGAGCATACCCAGGCCGGAACTGCCATGTCCAGAGCTAGGAGAGAGGACCTGCCTTCTCTGAGAAAGGAGGAAAGCTGCCTACTACAGAGGGCTACAGTTGGACTCACAGATGGGCTAGGAGATGCCTCCCAACTCCCCGTTGCTCCCACTGGGGACCAGCCATGCCAGGCCTTGCCCCTACTGTCCTCCCAAACCTCAGTAGCTGAGAGATTAGTGGAGCAGCCTCAGTTGCATCCGGATGTTAGAACTGAATGTGAGTCTGGCACCACTTCCTGGGAAAGTGATGATGAGGAGCAAGGACCCACCGTTCCTGCAGACAATGGTCCCATTCTGTCTCTAGTGGGAGATGATACATTAGAGAAAGGAACTGGCCAAGCTCTTGACAGTCATCCCACTATGAAGGATCCTGTAAATGTGACCCCCAGTTCCACACCTGAATCCTCACCGACTGATTGCCTGCAGAACAGAGCATTTGATGACGAATTAGGGCTTGGTGGCTCATGCCCTCCTATGAGGGAAAGTGATACTAGACAAGAAAACTTGAAAACCAAGGCTCTCGTTTCTAACAGTTCTTTGCATTGGATACCCATCCCATCGAATGATGAGGTAGTGAAACAGCCCAAACCAGAATCCAGAGAACACATACCATCTGTTGAGCCCCAGGTTGGAGAGGAGTGGGAGAAAGCTGCTCCCACCCCTCCTGCATTGCCTGGGGATTTGACAGCTGAGGAGGGTCTAGATCCTCTTGACAGCCTTACTTCACTCTGGACTGTGCCATCTCGAGGAGGCAGTGACAGCAATGGCAGTTACTGTCAACAGGTGGACATTGAAAAGCTGAAAATCAACGGAGACTCTGAAGCACTGAGTCCTCACGGTGAGTCCACGGATACAGCCTCTGACTTTGAAGGTCACCTCACGGAGGACAGCAGTGAGGCTGACACTAGAGAAGCTGCAGTGACAAAGGGATCTTCGGTGGACAAGGATGAGAAACCCAATTGGAACCAATCTGCCCCACTGTCCAAGGTGAATGGTGACATGCGTCTGGTTACAAGGACAGATGGGATGGTTGCTCCTCAGAGCTGGGTGTCTCGAGTATGTGCGGTCCGCCAAAAGATCCCAGATTCCCTACTGCTGGCCAGTACTGAGTACCAGCCAAGAGCCGTGTGCCTGTCCATGCCTGGGTCCTCAGTGGAGGCCACTAACCCACTTGTGATGCAGTTGCTGCAGGGTAGCTTGCCCCTAGAGAAGGTTCTTCCACCAGCCCACGATGACAGCATGTCAGAATCCCCACAAGTACCACTTACAAAAGACCAGAGCCATGGCTCGCTACGCATGGGATCTTTACATGGTCTTGGAAAAAACAGTGGCATGGTTGATGGAAGCAGCCCCAGTTCTTTAAGGGCTTTGAAGGAGCCTCTTCTGCCAGATAGCTGTGAAACAGGCACTGGTCTTGCCAGGATTGAGGCCACCCAGGCTCCTGGAGCACCCCAAAAGAATTGCAAGGCAGTCCCAAGTTTTGACTCCCTCCATCCAGTGACAAATCCCATTACATCCTCTAGGAAACTGGAAGAAATGGATTCCAAAGAGCAGTTCTCTTCCTTTAGTTGTGAAGATCAGAAGGAAGTCCGTGCTATGTCACAGGACAGTAATTCAAATGCTGCTCCAGGAAAGAGCCCAGGAGATCTTACTACCTCGAGAACACCTCGTTTCTCATCTCCAAATGTGATCTCCTTTGGTCCAGAGCAGACAGGTCGGGCCCTGGGTGATCAGAGCAATGTTACAGGCCAAGGGAAGAAGCTTTTTGGCTCTGGGAATGTGGCTGCAACCCTTCAGCGCCCCAGGCCTGCGGACCCGATGCCTCTTCCTGCTGAGATCCCTCCAGTTTTTCCCAGTGGGAAGTTGGGACCAAGCACAAACTCCATGTCTGGTGGGGTACAGACTCCAAGGGAAGACTGGGCTCCAAAGCCACATGCCTTTGTTGGCAGCGTCAAGAATGAGAAGACTTTTGTGGGGGGTCCTCTTAAGGCAAATGCCGAGAACAGGAAAGCTACTGGGCATAGTCCCCTGGAACTGGTGGGTCACTTGGAAGGGATGCCCTTTGTCATGGACTTGCCCTTCTGGAAATTACCCCGAGAGCCAGGGAAGGGGCTCAGTGAGCCTCTGGAGCCTTCTTCTCTCCCCTCCCAACTCAGCATCAAGCAGGCATTTTATGGGAAGCTTTCTAAACTCCAACTGAGTTCCACCAGCTTTAATTATTCCTCTAGCTCTCCCACCTTTCCCAAAGGCCTTGCTGGAAGTGTGGTGCAGCTGAGCCACAAAGCAAACTTTGGTGCGAGCCACAGTGCATCACTTTCCTTGCAAATGTTCACTGACAGCAGCACGGTGGAAAGCATCTCGCTCCAGTGTGCGTGCAGCCTGAAAGCCATGATCATGTGCCAAGGCTGCGGTGCGTTCTGTCACGATGACTGTATTGGACCCTCAAAGCTCTGTGTATTGTGCCTTGTGGTGAGATAA'\n",
    "    \n",
    "    #Split the sequence into codons \n",
    "    ASXL1_codons = [ASXL1_exons[i:i+3] for i in range(0, len(ASXL1_exons), 3)]\n",
    "\n",
    "    #ASXL1 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    ASXL1_exons_extra_bases = 'GATGAAGGACAAACAGAAGAAGAAGAAGGAGCGCACGTGGGCCGAGGCCGCGCGCCTGGTATTAGAAAACTACTCGGATGCTCCAATGACACCAAAACAGATTCTGCAGGTCATAGAGGCAGAAGGACTAAAGGAAATGAGAAGTGGGACTTCCCCTCTCGCATGCCTCAATGCTATGCTACATTCCAATTCAAGAGGAGGAGAGGGGTTGTTTTATAAACTGCCTGGCCGAATCAGCCTTTTCACGCTCAAGAAGGATGCCCTGCAGTGGTCTCGCCATCCAGCTACAGTGGAGGGAGAGGAGCCAGAGGACACGGCTGATGTGGAGAGCTGTGGGTCTAATGAAGCCAGCACTGTGAGTGGTGAAAACGATGTATCTCTTGATGAAACATCTTCGAACGCATCCTGTTCTACAGAATCTCAGAGTCGACCTCTTTCCAATCCCAGGGACAGCTACAGAGCTTCCTCACAGGCGAACAAACAAAAGAAAAAGACTGGGGTGATGCTGCCTCGAGTTGTCCTGACTCCTCTGAAGGTAAACGGGGCCCACGTGGAATCTGCATCAGGGTTCTCGGGCTGCCACGCCGATGGCGAGAGCGGCAGCCCGTCCAGCAGCAGCAGCGGCTCTCTGGCCCTGGGCAGCGCTGCTATTCGTGGCCAGGCCGAGGTCACCCAGGACCCTGCCCCGCTCCTGAGAGGCTTCCGGAAGCCAGCCACAGGTCAAATGAAGCGCAACAGAGGGGAAGAAATAGATTTTGAGACACCTGGGTCCATTCTTGTCAACACCAACCTCCGTGCCCTGATCAACTCTCGGACCTTCCATGCCTTACCATCACACTTCCAGCAGCAGCTCCTCTTCCTCCTGCCTGAAGTAGACAGACAGGTGGGGACGGATGGCCTGTTGCGTCTCAGCAGCAGTGCACTAAATAACGAGTTTTTTACCCATGCGGCTCAGAGCTGGCGGGAGCGCCTGGCTGATGGTGAATTTACTCATGAGATGCAAGTCAGGATACGACAGGAAATGGAGAAGGAAAAGAAGGTGGAACAATGGAAAGAAAAGTTCTTTGAAGACTACTATGGACAGAAGCTGGGTTTGACCAAAGAAGAGTCATTGCAGCAGAACGTGGGCCAGGAGGAGGCTGAAATCAAAAGTGGCTTGTGTGTCCCAGGAGAATCAGTGCGTATACAGCGTGGTCCAGCCACCCGACAGCGAGATGGGCATTTTAAGAAACGCTCTCGGCCAGATCTCCGAACCAGAGCCAGAAGGAATCTGTACAAAAAACAGGAGTCAGAACAAGCAGGGGTTGCTAAGGATGCAAAATCTGTGGCCTCAGATGTTCCCCTCTACAAGGATGGGGAGGCTAAGACTGACCCAGCAGGGCTGAGCAGTCCCCATCTGCCAGGCACATCCTCTGCAGCACCCGACCTGGAGGGTCCCGAATTCCCAGTTGAGTCTGTGGCTTCTCGGATCCAGGCTGAGCCAGACAACTTGGCACGTGCCTCTGCATCTCCAGACAGAATTCCTAGCCTGCCTCAGGAAACTGTGGATCAGGAACCCAAGGATCAGAAGAGGAAATCCTTTGAGCAGGCGGCCTCTGCATCCTTTCCCGAAAAGAAGCCCCGGCTTGAAGATCGTCAGTCCTTTCGTAACACAATTGAAAGTGTTCACACCGAAAAGCCACAGCCCACTAAAGAGGAGCCCAAAGTCCCGCCCATCCGGATTCAACTTTCACGTATCAAACCACCCTGGGTGGTTAAAGGTCAGCCCACTTACCAGATATGCCCCCGGATCATCCCCACCACGGAGTCCTCCTGCCGGGGTTGGACTGGCGCCAGGACCCTCGCAGACATTAAAGCCCGTGCTCTGCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCCACCACTGCCATCGGAGGGGGGGGTGGCCCGGGTGGAGGTGGCGGCGGGGCCACCGATGAGGGAGGTGGCAGAGGCAGCAGCAGTGGTGATGGTGGTGAGGCCTGTGGCCACCCTGAGCCCAGGGGAGGCCCGAGCACCCCTGGAAAGTGTACGTCAGATCTACAGCGAACACAACTACTGCCGCCTTATCCTCTAAATGGGGAGCATACCCAGGCCGGAACTGCCATGTCCAGAGCTAGGAGAGAGGACCTGCCTTCTCTGAGAAAGGAGGAAAGCTGCCTACTACAGAGGGCTACAGTTGGACTCACAGATGGGCTAGGAGATGCCTCCCAACTCCCCGTTGCTCCCACTGGGGACCAGCCATGCCAGGCCTTGCCCCTACTGTCCTCCCAAACCTCAGTAGCTGAGAGATTAGTGGAGCAGCCTCAGTTGCATCCGGATGTTAGAACTGAATGTGAGTCTGGCACCACTTCCTGGGAAAGTGATGATGAGGAGCAAGGACCCACCGTTCCTGCAGACAATGGTCCCATTCTGTCTCTAGTGGGAGATGATACATTAGAGAAAGGAACTGGCCAAGCTCTTGACAGTCATCCCACTATGAAGGATCCTGTAAATGTGACCCCCAGTTCCACACCTGAATCCTCACCGACTGATTGCCTGCAGAACAGAGCATTTGATGACGAATTAGGGCTTGGTGGCTCATGCCCTCCTATGAGGGAAAGTGATACTAGACAAGAAAACTTGAAAACCAAGGCTCTCGTTTCTAACAGTTCTTTGCATTGGATACCCATCCCATCGAATGATGAGGTAGTGAAACAGCCCAAACCAGAATCCAGAGAACACATACCATCTGTTGAGCCCCAGGTTGGAGAGGAGTGGGAGAAAGCTGCTCCCACCCCTCCTGCATTGCCTGGGGATTTGACAGCTGAGGAGGGTCTAGATCCTCTTGACAGCCTTACTTCACTCTGGACTGTGCCATCTCGAGGAGGCAGTGACAGCAATGGCAGTTACTGTCAACAGGTGGACATTGAAAAGCTGAAAATCAACGGAGACTCTGAAGCACTGAGTCCTCACGGTGAGTCCACGGATACAGCCTCTGACTTTGAAGGTCACCTCACGGAGGACAGCAGTGAGGCTGACACTAGAGAAGCTGCAGTGACAAAGGGATCTTCGGTGGACAAGGATGAGAAACCCAATTGGAACCAATCTGCCCCACTGTCCAAGGTGAATGGTGACATGCGTCTGGTTACAAGGACAGATGGGATGGTTGCTCCTCAGAGCTGGGTGTCTCGAGTATGTGCGGTCCGCCAAAAGATCCCAGATTCCCTACTGCTGGCCAGTACTGAGTACCAGCCAAGAGCCGTGTGCCTGTCCATGCCTGGGTCCTCAGTGGAGGCCACTAACCCACTTGTGATGCAGTTGCTGCAGGGTAGCTTGCCCCTAGAGAAGGTTCTTCCACCAGCCCACGATGACAGCATGTCAGAATCCCCACAAGTACCACTTACAAAAGACCAGAGCCATGGCTCGCTACGCATGGGATCTTTACATGGTCTTGGAAAAAACAGTGGCATGGTTGATGGAAGCAGCCCCAGTTCTTTAAGGGCTTTGAAGGAGCCTCTTCTGCCAGATAGCTGTGAAACAGGCACTGGTCTTGCCAGGATTGAGGCCACCCAGGCTCCTGGAGCACCCCAAAAGAATTGCAAGGCAGTCCCAAGTTTTGACTCCCTCCATCCAGTGACAAATCCCATTACATCCTCTAGGAAACTGGAAGAAATGGATTCCAAAGAGCAGTTCTCTTCCTTTAGTTGTGAAGATCAGAAGGAAGTCCGTGCTATGTCACAGGACAGTAATTCAAATGCTGCTCCAGGAAAGAGCCCAGGAGATCTTACTACCTCGAGAACACCTCGTTTCTCATCTCCAAATGTGATCTCCTTTGGTCCAGAGCAGACAGGTCGGGCCCTGGGTGATCAGAGCAATGTTACAGGCCAAGGGAAGAAGCTTTTTGGCTCTGGGAATGTGGCTGCAACCCTTCAGCGCCCCAGGCCTGCGGACCCGATGCCTCTTCCTGCTGAGATCCCTCCAGTTTTTCCCAGTGGGAAGTTGGGACCAAGCACAAACTCCATGTCTGGTGGGGTACAGACTCCAAGGGAAGACTGGGCTCCAAAGCCACATGCCTTTGTTGGCAGCGTCAAGAATGAGAAGACTTTTGTGGGGGGTCCTCTTAAGGCAAATGCCGAGAACAGGAAAGCTACTGGGCATAGTCCCCTGGAACTGGTGGGTCACTTGGAAGGGATGCCCTTTGTCATGGACTTGCCCTTCTGGAAATTACCCCGAGAGCCAGGGAAGGGGCTCAGTGAGCCTCTGGAGCCTTCTTCTCTCCCCTCCCAACTCAGCATCAAGCAGGCATTTTATGGGAAGCTTTCTAAACTCCAACTGAGTTCCACCAGCTTTAATTATTCCTCTAGCTCTCCCACCTTTCCCAAAGGCCTTGCTGGAAGTGTGGTGCAGCTGAGCCACAAAGCAAACTTTGGTGCGAGCCACAGTGCATCACTTTCCTTGCAAATGTTCACTGACAGCAGCACGGTGGAAAGCATCTCGCTCCAGTGTGCGTGCAGCCTGAAAGCCATGATCATGTGCCAAGGCTGCGGTGCGTTCTGTCACGATGACTGTATTGGACCCTCAAAGCTCTGTGTATTGTGCCTTGTGGTGAGATAAT'\n",
    "  \n",
    "    #List of all amino acid positions (across whole coding region) covered (amino acid position minus 1 (as python starts at 0))\n",
    "    ASXL1_missense_codons_DICT={}\n",
    "    for i in range(lower_range, upper_range):\n",
    "        ASXL1_missense_codons_DICT[i+1]= (ASXL1_codons[i])\n",
    "\n",
    "    calls_ASXL1_missense = {}\n",
    "    for k,v in ASXL1_missense_codons_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_ASXL1_missense[k]=(v, changes)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_ASXL1_missense_extra = {}\n",
    "\n",
    "    for k, v in calls_ASXL1_missense.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_ASXL1_missense_extra[k] = (Amino_acid_change, v[0], v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_ASXL1_missense_extra2 = {}\n",
    "\n",
    "    for k, v in calls_ASXL1_missense_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_ASXL1_missense_extra2[k] = (Amino_acid_abbrev, v[1], v[2])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary that contains details of the positions of missense variants at that codon\n",
    "    ASXL1_position_in_list_dict_missense = {}\n",
    "\n",
    "    for k, v in calls_ASXL1_missense_extra2.items():\n",
    "        original_AA = v[0]\n",
    "        possible_AA_changes = v[2]\n",
    "        possible_AA_changes_string = ''.join(v[2])\n",
    "        positions_in_list = []\n",
    "        for i in possible_AA_changes:\n",
    "            if i != original_AA: #not synonymous\n",
    "                if i != '*': #not missense\n",
    "                    position_in_list=-1\n",
    "                    missense_change = i\n",
    "                    new_aa_list = []\n",
    "                    while True:\n",
    "                        position_in_list = possible_AA_changes_string.find(missense_change, position_in_list+1)\n",
    "        #               print(position_in_list)\n",
    "                        if position_in_list == -1: break\n",
    "                        positions_in_list.append(position_in_list)\n",
    "                        positions_in_list = set(positions_in_list) #convert into a 'set' to remove duplicate positions\n",
    "                        positions_in_list = list(positions_in_list) #convert back to a list so can be added to dictionary\n",
    "                    for pos in positions_in_list:\n",
    "                        new_aa = possible_AA_changes_string[pos]\n",
    "                        new_aa_list.append(new_aa)\n",
    "                    ASXL1_position_in_list_dict_missense[k]=(positions_in_list, v, new_aa_list)\n",
    "\n",
    "        aa_positions_variants_called = []\n",
    "        for k, v in ASXL1_position_in_list_dict_missense.items():\n",
    "            aa_positions_variants_called.append(k)\n",
    "    # print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    ASXL1_bases_missense = {}\n",
    "    for k, v in ASXL1_position_in_list_dict_missense.items():\n",
    "        original_a_a = v[1][0]\n",
    "        codon_seq = v[1][1]\n",
    "        base_changes=[]\n",
    "        new_a_a_list = v[2]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        ASXL1_bases_missense[k]=(v, base_changes, original_a_a, new_a_a_list)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    ASXL1_bases_missense2 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_missense.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        new_a_a_list = v[3]\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        ASXL1_bases_missense2[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    ASXL1_bases_missense3 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_missense2.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_a_a = v[0][2]\n",
    "        new_aa_list = v[0][0][2]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = k\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(k)-1)*3)+codon_pos #codon number minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][1][1][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        ASXL1_bases_missense3[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, original_a_a, new_aa_list)\n",
    "\n",
    "    # Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    ASXL1_bases_missense4 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_missense3.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "        new_a_a_list = v[6]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check(i)\n",
    "                sites_trio_missense.append([adapted_trio])\n",
    "                ASXL1_bases_missense4[k] = (v, sites_trio_missense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_missense = [ASXL1_exons_extra_bases[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_missense.append(site_trio_missense)\n",
    "                    ASXL1_bases_missense4[k] = (v, sites_trio_missense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    ASXL1_sites_missense5 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in ASXL1_bases_missense4.items():\n",
    "        original_a_a = v[0][0][0]\n",
    "        new_a_a = v[0][6]\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            new_amino_acid = new_a_a[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, original_a_a, new_amino_acid])\n",
    "            change = change+1\n",
    "        ASXL1_sites_missense5[a]=info\n",
    "        a=a+1\n",
    "    #     print(ASXL1_sites_missense5)\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    ASXL1_sites_missense6={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in ASXL1_sites_missense5.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        original_a_a = v[5]\n",
    "        new_a_a = v[6]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        ASXL1_sites_missense6[k] = (v[0],changes3, codon_n_912, original_a_a, new_a_a) #site_trio = the affected base and its neighbours\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    ASXL1_mutation_rates_missense = {}\n",
    "    for k, v in ASXL1_sites_missense6.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            original_a_a = v[3]\n",
    "            new_a_a = v[4]\n",
    "            aa_positioning = v[2]\n",
    "            variant_id = str(original_a_a)+str(aa_positioning)+str(new_a_a)\n",
    "            ASXL1_mutation_rates_missense[k] = (v[0],v[1], mutation_rate, variant_id)\n",
    "\n",
    "    ASXL1_mutation_rate_missense_sum = []\n",
    "\n",
    "    mutation_rates_list = []\n",
    "    for k, v in ASXL1_mutation_rates_missense.items():\n",
    "        codon_position_in_gene=(v[0])\n",
    "        variant = v[3]\n",
    "        site = v[1]\n",
    "        mutation_rate_value = v[2] \n",
    "        mutation_rates_list.append((v[0], variant, site, mutation_rate_value))\n",
    "        ASXL1_mutation_rate_missense_sum.append(mutation_rate_value)\n",
    "\n",
    "    #Get the average of all the mutation rates\n",
    "    ASXL1_mutation_rate_missense_sum = []\n",
    "    for k, v in ASXL1_mutation_rates_missense.items():\n",
    "        mut_rate = v[2]\n",
    "        ASXL1_mutation_rate_missense_sum.append(mut_rate)\n",
    "\n",
    "    ASXL1_Summed_mutation_rate_missense_added = np.sum(ASXL1_mutation_rate_missense_sum)\n",
    "\n",
    "    print('Total summed mutation rate across ASXL1 missense for this study = ', ASXL1_Summed_mutation_rate_missense_added)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all # e.g. lower= 0, upper = 913 for whole gene\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TP53"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_nonsense_range_of_variants_TP53(lower_range, upper_range):\n",
    "    \n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base_TP53 = [74, 75, 96, 375, 672, 782, 919, 993, 1100]\n",
    "   \n",
    "    def intron_exon_check(position_in_gene_number):\n",
    "        if position_in_gene_number == 74:\n",
    "            gene_pos = 74\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 75:\n",
    "            gene_pos = 75\n",
    "            first_base = 'G'\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = TP53_exons_extra_bases[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 96:\n",
    "            gene_pos = 96\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 375:\n",
    "            gene_pos = 375\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 672:\n",
    "            gene_pos = 672\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 782:\n",
    "            gene_pos = 782\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 919:\n",
    "            gene_pos = 919\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 993:\n",
    "            gene_pos = 993\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1100:\n",
    "            gene_pos = 1100\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #TP53 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    TP53_exons = 'ATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGACATAGTGTGGTGGTGCCCTATGAGCCGCCTGAGGTTGGCTCTGACTGTACCACCATCCACTACAACTACATGTGTAACAGTTCCTGCATGGGCGGCATGAACCGGAGGCCCATCCTCACCATCATCACACTGGAAGACTCCAGTGGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGAGGAAGAGAATCTCCGCAAGAAAGGGGAGCCTCACCACGAGCTGCCCCCAGGGAGCACTAAGCGAGCACTGCCCAACAACACCAGCTCCTCTCCCCAGCCAAAGAAGAAACCACTGGATGGAGAATATTTCACCCTTCAGATCCGTGGGCGTGAGCGCTTCGAGATGTTCCGAGAGCTGAATGAGGCCTTGGAACTCAAGGATGCCCAGGCTGGGAAGGAGCCAGGGGGGAGCAGGGCTCACTCCAGCCACCTGAAGTCCAAAAAGGGTCAGTCTACCTCCCGCCATAAAAAACTCATGTTCAAGACAGAAGGGCCTGACTCAGACTGA'\n",
    "    \n",
    "    #Split the sequence into codons \n",
    "    TP53_codons = [TP53_exons[i:i+3] for i in range(0, len(TP53_exons), 3)]\n",
    "\n",
    "    #TP53 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    TP53_exons_extra_bases = 'CATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGACATAGTGTGGTGGTGCCCTATGAGCCGCCTGAGGTTGGCTCTGACTGTACCACCATCCACTACAACTACATGTGTAACAGTTCCTGCATGGGCGGCATGAACCGGAGGCCCATCCTCACCATCATCACACTGGAAGACTCCAGTGGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGAGGAAGAGAATCTCCGCAAGAAAGGGGAGCCTCACCACGAGCTGCCCCCAGGGAGCACTAAGCGAGCACTGCCCAACAACACCAGCTCCTCTCCCCAGCCAAAGAAGAAACCACTGGATGGAGAATATTTCACCCTTCAGATCCGTGGGCGTGAGCGCTTCGAGATGTTCCGAGAGCTGAATGAGGCCTTGGAACTCAAGGATGCCCAGGCTGGGAAGGAGCCAGGGGGGAGCAGGGCTCACTCCAGCCACCTGAAGTCCAAAAAGGGTCAGTCTACCTCCCGCCATAAAAAACTCATGTTCAAGACAGAAGGGCCTGACTCAGACTGAC'\n",
    "    \n",
    "    #List of all amino acid positions (across whole coding region) covered (amino acid position minus 1 (as python starts at 0))\n",
    "    TP53_nonsense_codons_DICT={}\n",
    "    for i in range(lower_range, upper_range):\n",
    "        TP53_nonsense_codons_DICT[i+1]= (TP53_codons[i])\n",
    "\n",
    "    calls_TP53_nonsense = {}\n",
    "    for k,v in TP53_nonsense_codons_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_TP53_nonsense[k]=(v, changes)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_TP53_nonsense_extra = {}\n",
    "\n",
    "    for k, v in calls_TP53_nonsense.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_TP53_nonsense_extra[k] = (Amino_acid_change, v[0], v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_TP53_nonsense_extra2 = {}\n",
    "\n",
    "    for k, v in calls_TP53_nonsense_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_TP53_nonsense_extra2[k] = (Amino_acid_abbrev, v[1], v[2])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary that contains details of the positions of nonsense variants at that codon\n",
    "    TP53_position_in_list_dict_nonsense = {}\n",
    "\n",
    "    for k, v in calls_TP53_nonsense_extra2.items():\n",
    "        original_AA = v[0]\n",
    "        possible_AA_changes = v[2]\n",
    "        possible_AA_changes_string = ''.join(v[2])\n",
    "        positions_in_list = []\n",
    "        for i in possible_AA_changes:\n",
    "            if i == '*': #nonsense\n",
    "                position_in_list=-1\n",
    "                nonsense_change = i    \n",
    "                while True:\n",
    "                    position_in_list = possible_AA_changes_string.find(nonsense_change, position_in_list+1)\n",
    "    #               print(position_in_list)\n",
    "                    if position_in_list == -1: break\n",
    "                    positions_in_list.append(position_in_list)\n",
    "                    positions_in_list = set(positions_in_list) #convert into a 'set' to remove duplicate positions\n",
    "                    positions_in_list = list(positions_in_list) #convert back to a list so can be added to dictionary\n",
    "                TP53_position_in_list_dict_nonsense[k]=(positions_in_list, v)\n",
    "\n",
    "        aa_positions_variants_called = []\n",
    "        for k, v in TP53_position_in_list_dict_nonsense.items():\n",
    "            aa_positions_variants_called.append(k)\n",
    "    # print('positions at which nonsense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TP53_bases_nonsense = {}\n",
    "    for k, v in TP53_position_in_list_dict_nonsense.items():\n",
    "        original_a_a = v[1][0]\n",
    "        codon_seq = v[1][1]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TP53_bases_nonsense[k]=(v, base_changes, original_a_a)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TP53_bases_nonsense2 = {}\n",
    "\n",
    "    for k, v in TP53_bases_nonsense.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TP53_bases_nonsense2[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TP53_bases_nonsense3 = {}\n",
    "\n",
    "    for k, v in TP53_bases_nonsense2.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_a_a = v[0][2]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = k\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(k)-1)*3)+codon_pos #codon number minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][1][1][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TP53_bases_nonsense3[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, original_a_a)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TP53_bases_nonsense4 = {}\n",
    "\n",
    "    for k, v in TP53_bases_nonsense3.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_nonsense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_TP53:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check_TP53(i)\n",
    "                sites_trio_nonsense.append([adapted_trio])\n",
    "                TP53_bases_nonsense4[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_nonsense = [TP53_exons_extra_bases[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_nonsense.append(site_trio_nonsense)\n",
    "                    TP53_bases_nonsense4[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TP53_sites_nonsense5 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TP53_bases_nonsense4.items():\n",
    "        original_a_a = v[0][0][0]\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, original_a_a])\n",
    "            change = change+1\n",
    "        TP53_sites_nonsense5[a]=info\n",
    "        a=a+1\n",
    "    #     print(TP53_sites_nonsense5)\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TP53_sites_nonsense6={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TP53_sites_nonsense5.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        original_a_a = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TP53_sites_nonsense6[k] = (v[0],changes3, codon_n_912, original_a_a) #site_trio = the affected base and its neighbours\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    TP53_mutation_rates_nonsense = {}\n",
    "    for k, v in TP53_sites_nonsense6.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            original_a_a = v[3]\n",
    "            aa_positioning = v[2]\n",
    "            variant_type = '*'\n",
    "            variant_id = str(original_a_a)+str(aa_positioning)+variant_type\n",
    "            TP53_mutation_rates_nonsense[k] = (v[0],v[1], mutation_rate, variant_id)\n",
    "\n",
    "    TP53_mutation_rate_nonsense_sum = []\n",
    "\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TP53_mutation_rates_nonsense.items():\n",
    "        codon_position_in_gene=(v[0])\n",
    "        variant = v[3]\n",
    "        site = v[1]\n",
    "        mutation_rate_value = v[2] \n",
    "        mutation_rates_list.append((v[0], variant, site, mutation_rate_value))\n",
    "        TP53_mutation_rate_nonsense_sum.append(mutation_rate_value)\n",
    "\n",
    "    #Get the average of all the mutation rates\n",
    "    TP53_mutation_rate_nonsense_sum = []\n",
    "    for k, v in TP53_mutation_rates_nonsense.items():\n",
    "        mut_rate = v[2]\n",
    "        TP53_mutation_rate_nonsense_sum.append(mut_rate)\n",
    "\n",
    "    TP53_Summed_mutation_rate_nonsense_added = np.sum(TP53_mutation_rate_nonsense_sum)\n",
    "\n",
    "    print('Total summed mutation rate across TP53 nonsense for this study = ', TP53_Summed_mutation_rate_nonsense_added)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all # e.g. lower= 0, upper = 913 for whole gene\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_missense_range_of_variants_TP53(lower_range, upper_range):\n",
    "\n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base_TP53 = [74, 75, 96, 375, 672, 782, 919, 993, 1100]\n",
    "   \n",
    "    def intron_exon_check(position_in_gene_number):\n",
    "        if position_in_gene_number == 74:\n",
    "            gene_pos = 74\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 75:\n",
    "            gene_pos = 75\n",
    "            first_base = 'G'\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = TP53_exons_extra_bases[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 96:\n",
    "            gene_pos = 96\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 375:\n",
    "            gene_pos = 375\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 672:\n",
    "            gene_pos = 672\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 782:\n",
    "            gene_pos = 782\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 919:\n",
    "            gene_pos = 919\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 993:\n",
    "            gene_pos = 993\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1100:\n",
    "            gene_pos = 1100\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #TP53 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    TP53_exons = 'ATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGACATAGTGTGGTGGTGCCCTATGAGCCGCCTGAGGTTGGCTCTGACTGTACCACCATCCACTACAACTACATGTGTAACAGTTCCTGCATGGGCGGCATGAACCGGAGGCCCATCCTCACCATCATCACACTGGAAGACTCCAGTGGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGAGGAAGAGAATCTCCGCAAGAAAGGGGAGCCTCACCACGAGCTGCCCCCAGGGAGCACTAAGCGAGCACTGCCCAACAACACCAGCTCCTCTCCCCAGCCAAAGAAGAAACCACTGGATGGAGAATATTTCACCCTTCAGATCCGTGGGCGTGAGCGCTTCGAGATGTTCCGAGAGCTGAATGAGGCCTTGGAACTCAAGGATGCCCAGGCTGGGAAGGAGCCAGGGGGGAGCAGGGCTCACTCCAGCCACCTGAAGTCCAAAAAGGGTCAGTCTACCTCCCGCCATAAAAAACTCATGTTCAAGACAGAAGGGCCTGACTCAGACTGA'\n",
    "    \n",
    "    #Split the sequence into codons \n",
    "    TP53_codons = [TP53_exons[i:i+3] for i in range(0, len(TP53_exons), 3)]\n",
    "\n",
    "    #TP53 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    TP53_exons_extra_bases = 'CATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGACATAGTGTGGTGGTGCCCTATGAGCCGCCTGAGGTTGGCTCTGACTGTACCACCATCCACTACAACTACATGTGTAACAGTTCCTGCATGGGCGGCATGAACCGGAGGCCCATCCTCACCATCATCACACTGGAAGACTCCAGTGGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGAGGAAGAGAATCTCCGCAAGAAAGGGGAGCCTCACCACGAGCTGCCCCCAGGGAGCACTAAGCGAGCACTGCCCAACAACACCAGCTCCTCTCCCCAGCCAAAGAAGAAACCACTGGATGGAGAATATTTCACCCTTCAGATCCGTGGGCGTGAGCGCTTCGAGATGTTCCGAGAGCTGAATGAGGCCTTGGAACTCAAGGATGCCCAGGCTGGGAAGGAGCCAGGGGGGAGCAGGGCTCACTCCAGCCACCTGAAGTCCAAAAAGGGTCAGTCTACCTCCCGCCATAAAAAACTCATGTTCAAGACAGAAGGGCCTGACTCAGACTGAC'\n",
    "    \n",
    "    #List of all amino acid positions (across whole coding region) covered (amino acid position minus 1 (as python starts at 0))\n",
    "    TP53_missense_codons_DICT={}\n",
    "    for i in range(lower_range, upper_range):\n",
    "        TP53_missense_codons_DICT[i+1]= (TP53_codons[i])\n",
    "\n",
    "    calls_TP53_missense = {}\n",
    "    for k,v in TP53_missense_codons_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_TP53_missense[k]=(v, changes)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_TP53_missense_extra = {}\n",
    "\n",
    "    for k, v in calls_TP53_missense.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_TP53_missense_extra[k] = (Amino_acid_change, v[0], v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_TP53_missense_extra2 = {}\n",
    "\n",
    "    for k, v in calls_TP53_missense_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_TP53_missense_extra2[k] = (Amino_acid_abbrev, v[1], v[2])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary that contains details of the positions of missense variants at that codon\n",
    "    TP53_position_in_list_dict_missense = {}\n",
    "\n",
    "    for k, v in calls_TP53_missense_extra2.items():\n",
    "        original_AA = v[0]\n",
    "        possible_AA_changes = v[2]\n",
    "        possible_AA_changes_string = ''.join(v[2])\n",
    "        positions_in_list = []\n",
    "        for i in possible_AA_changes:\n",
    "            if i != original_AA: #not synonymous\n",
    "                if i != '*': #not missense\n",
    "                    position_in_list=-1\n",
    "                    missense_change = i\n",
    "                    new_aa_list = []\n",
    "                    while True:\n",
    "                        position_in_list = possible_AA_changes_string.find(missense_change, position_in_list+1)\n",
    "        #               print(position_in_list)\n",
    "                        if position_in_list == -1: break\n",
    "                        positions_in_list.append(position_in_list)\n",
    "                        positions_in_list = set(positions_in_list) #convert into a 'set' to remove duplicate positions\n",
    "                        positions_in_list = list(positions_in_list) #convert back to a list so can be added to dictionary\n",
    "                    for pos in positions_in_list:\n",
    "                        new_aa = possible_AA_changes_string[pos]\n",
    "                        new_aa_list.append(new_aa)\n",
    "                    TP53_position_in_list_dict_missense[k]=(positions_in_list, v, new_aa_list)\n",
    "\n",
    "        aa_positions_variants_called = []\n",
    "        for k, v in TP53_position_in_list_dict_missense.items():\n",
    "            aa_positions_variants_called.append(k)\n",
    "    # print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TP53_bases_missense = {}\n",
    "    for k, v in TP53_position_in_list_dict_missense.items():\n",
    "        original_a_a = v[1][0]\n",
    "        codon_seq = v[1][1]\n",
    "        base_changes=[]\n",
    "        new_a_a_list = v[2]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TP53_bases_missense[k]=(v, base_changes, original_a_a, new_a_a_list)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TP53_bases_missense2 = {}\n",
    "\n",
    "    for k, v in TP53_bases_missense.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        new_a_a_list = v[3]\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TP53_bases_missense2[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TP53_bases_missense3 = {}\n",
    "\n",
    "    for k, v in TP53_bases_missense2.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_a_a = v[0][2]\n",
    "        new_aa_list = v[0][0][2]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = k\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(k)-1)*3)+codon_pos #codon number minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][1][1][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TP53_bases_missense3[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, original_a_a, new_aa_list)\n",
    "\n",
    "    # Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TP53_bases_missense4 = {}\n",
    "\n",
    "    for k, v in TP53_bases_missense3.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "        new_a_a_list = v[6]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_TP53:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check(i)\n",
    "                sites_trio_missense.append([adapted_trio])\n",
    "                TP53_bases_missense4[k] = (v, sites_trio_missense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_missense = [TP53_exons_extra_bases[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_missense.append(site_trio_missense)\n",
    "                    TP53_bases_missense4[k] = (v, sites_trio_missense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TP53_sites_missense5 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TP53_bases_missense4.items():\n",
    "        original_a_a = v[0][0][0]\n",
    "        new_a_a = v[0][6]\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            new_amino_acid = new_a_a[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, original_a_a, new_amino_acid])\n",
    "            change = change+1\n",
    "        TP53_sites_missense5[a]=info\n",
    "        a=a+1\n",
    "    #     print(TP53_sites_missense5)\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TP53_sites_missense6={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TP53_sites_missense5.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        original_a_a = v[5]\n",
    "        new_a_a = v[6]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TP53_sites_missense6[k] = (v[0],changes3, codon_n_912, original_a_a, new_a_a) #site_trio = the affected base and its neighbours\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    TP53_mutation_rates_missense = {}\n",
    "    for k, v in TP53_sites_missense6.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            original_a_a = v[3]\n",
    "            new_a_a = v[4]\n",
    "            aa_positioning = v[2]\n",
    "            variant_id = str(original_a_a)+str(aa_positioning)+str(new_a_a)\n",
    "            TP53_mutation_rates_missense[k] = (v[0],v[1], mutation_rate, variant_id)\n",
    "\n",
    "    TP53_mutation_rate_missense_sum = []\n",
    "\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TP53_mutation_rates_missense.items():\n",
    "        codon_position_in_gene=(v[0])\n",
    "        variant = v[3]\n",
    "        site = v[1]\n",
    "        mutation_rate_value = v[2] \n",
    "        mutation_rates_list.append((v[0], variant, site, mutation_rate_value))\n",
    "        TP53_mutation_rate_missense_sum.append(mutation_rate_value)\n",
    "\n",
    "    #Get the average of all the mutation rates\n",
    "    TP53_mutation_rate_missense_sum = []\n",
    "    for k, v in TP53_mutation_rates_missense.items():\n",
    "        mut_rate = v[2]\n",
    "        TP53_mutation_rate_missense_sum.append(mut_rate)\n",
    "\n",
    "    TP53_Summed_mutation_rate_missense_added = np.sum(TP53_mutation_rate_missense_sum)\n",
    "\n",
    "    print('Total summed mutation rate across TP53 missense for this study = ', TP53_Summed_mutation_rate_missense_added)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all # e.g. lower= 0, upper = 913 for whole gene\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Lists of variants targeted by each study"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- In order to calculate the number of times you expect to see a variant across all studies, you need to know which variants were call-able by which studies.\n",
    "- This can be calculated by using the mutation rate functions (above) as they will also produce a list of the variants within e.g. range of amino acid positions of interest or from a CSV file of variants."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Jaiswal 2014"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "code_folding": [],
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  3.8678204400000003e-07\n",
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  5.08517098e-07\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Jaiswal's DNMT3A variants\n",
    "Jaiswal_non_DNMT3A = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(0, 913)\n",
    "Jaiswal_miss_DNMT3A = mutation_rates_list_missense_variants_DNMT3A_NP072046('Data files/Jaiswal DNMT3A missense variants called.csv')\n",
    "\n",
    "#Lists of DNMT3A variants targetable in Jaiswal 2014 (i.e. included in their panel)\n",
    "Jaiswal_nonsense_DNMT3A = []\n",
    "for i in Jaiswal_non_DNMT3A:\n",
    "    Jaiswal_nonsense_DNMT3A.append(str('p.')+i[1])\n",
    "    \n",
    "Jaiswal_missense_DNMT3A = []\n",
    "for i in Jaiswal_miss_DNMT3A:\n",
    "    Jaiswal_missense_DNMT3A.append(i[3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "code_folding": [],
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TET2 nonsense_NP_01120680 for this study =  5.10993418e-07\n",
      "Total summed mutation rate across TET2 missense_NP_01120680 for this study =  1.1931589999999999e-08\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Jaiswal's TET2 variants\n",
    "Jaiswal_non_TET2 = mutation_rates_nonsense_range_of_variants_TET2(0, 1137)\n",
    "Jaiswal_miss_TET2 = mutation_rates_list_missense_variants_TET2('Data files/Jaiswal TET2 missense variants called.csv')\n",
    "\n",
    "#Lists of TET2 variants targetable in Jaiswal 2014 (i.e. included in their panel)\n",
    "Jaiswal_nonsense_TET2 = []\n",
    "for i in Jaiswal_non_TET2:\n",
    "    Jaiswal_nonsense_TET2.append(str('p.')+i[1])\n",
    "    \n",
    "Jaiswal_missense_TET2 = []\n",
    "for i in Jaiswal_miss_TET2:\n",
    "    Jaiswal_missense_TET2.append(i[3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "code_folding": [],
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across ASXL1 nonsense for this study =  6.254075530000001e-07\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Jaiswal's ASXL1 variants\n",
    "Jaiswal_non_ASXL1 = mutation_rates_nonsense_range_of_variants_ASXL1(0, 1542)\n",
    "\n",
    "#Lists of ASXL1 variants targetable in Jaiswal 2014 (i.e. included in their panel)\n",
    "Jaiswal_nonsense_ASXL1 = []\n",
    "for i in Jaiswal_non_ASXL1:\n",
    "    Jaiswal_nonsense_ASXL1.append(str('p.')+i[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "code_folding": [],
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 nonsense for this study =  1.6147847e-07\n",
      "Total summed mutation rate across TP53 missense for this study =  6.26412261e-07\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Jaiswal's TP53 variants\n",
    "Jaiswal_non_TP53 = mutation_rates_nonsense_range_of_variants_TP53(0, 394)\n",
    "Jaiswal_miss_TP53 = mutation_rates_list_missense_variants_TP53('Data files/Jaiswal TP53 missense variants called.csv')\n",
    "\n",
    "#Lists of TP53 variants targetable in Jaiswal 2014 (i.e. included in their panel)\n",
    "Jaiswal_nonsense_TP53 = []\n",
    "for i in Jaiswal_non_TP53:\n",
    "    Jaiswal_nonsense_TP53.append(str('p.')+i[1])\n",
    "    \n",
    "Jaiswal_missense_TP53 = []\n",
    "for i in Jaiswal_miss_TP53:\n",
    "    Jaiswal_missense_TP53.append(i[3])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Genovese 2014"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  2.7073501399999997e-07\n",
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  1.0921795000000001e-08\n",
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  5.383253473e-06\n",
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  4.68086602e-07\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Genovese's DNMT3A variants\n",
    "Genovese_nonsense_DNMT3A_1 = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(213, 841)\n",
    "Genovese_nonsense_DNMT3A_2 = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(866, 913)\n",
    "Genovese_missense_DNMT3A_1 = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(213, 841)\n",
    "Genovese_missense_DNMT3A_2 = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(866, 913)\n",
    "\n",
    "#Lists of DNMT3A variants targetable in Genovese 2014 (i.e. included in their panel)\n",
    "Genovese_nonsense_DNMT3A = []\n",
    "Genovese_missense_DNMT3A = []\n",
    "for i in Genovese_nonsense_DNMT3A_1:\n",
    "    Genovese_nonsense_DNMT3A.append(str('p.')+i[1])\n",
    "for i in Genovese_nonsense_DNMT3A_2:\n",
    "    Genovese_nonsense_DNMT3A.append(str('p.')+i[1])\n",
    "for i in Genovese_missense_DNMT3A_1:\n",
    "    Genovese_missense_DNMT3A.append(str('p.')+i[1])\n",
    "for i in Genovese_missense_DNMT3A_1:\n",
    "    Genovese_missense_DNMT3A.append(str('p.')+i[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "code_folding": [
     0,
     6
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  9.167888e-09\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Genovese's DNMT3A variants\n",
    "Genovese_missense_DNMT3A_partial_codons = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(846, 848)\n",
    "base_position_list = [2540, 2541, 2542, 2543]\n",
    "Genovese_partial_missense = []\n",
    "Genovese_partial_missense_sum_values = []\n",
    "for i in Genovese_missense_DNMT3A_partial_codons:\n",
    "    if i[0] in base_position_list:\n",
    "        Genovese_partial_missense.append(i)\n",
    "        Genovese_partial_missense_sum_values.append(i[3])\n",
    "        \n",
    "Genovese_partial_missense_added = np.sum(Genovese_partial_missense_sum_values)\n",
    "\n",
    "#Lists of DNMT3A variants targetable in Genovese 2014 (i.e. included in their panel)\n",
    "for i in Genovese_partial_missense:\n",
    "    Genovese_missense_DNMT3A.append(str('p.'+i[1]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "code_folding": [],
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TET2 nonsense_NP_060098 for this study =  5.192273050000001e-07\n",
      "Total summed mutation rate across TET2 missense_NP_060098 for this study =  3.8538210999999995e-08\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Genovese's TET2 variants\n",
    "Genovese_non_TET2 = mutation_rates_nonsense_range_of_variants_TET2_shorter(0, 1166)\n",
    "Genovese_miss_TET2 = mutation_rates_list_missense_variants_TET2_shorter('Data files/Genovese TET2 missense variants called.csv')\n",
    "\n",
    "#Lists of TET2 variants targetable in Genovese 2014 (i.e. included in their panel)\n",
    "Genovese_nonsense_TET2 = []\n",
    "for i in Genovese_non_TET2:\n",
    "    Genovese_nonsense_TET2.append(str('p.')+i[1])\n",
    "    \n",
    "Genovese_missense_TET2 = []\n",
    "for i in Genovese_miss_TET2:\n",
    "    Genovese_missense_TET2.append(i[3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across ASXL1 missense for this study =  4.871308e-09\n",
      "Total summed mutation rate across ASXL1 nonsense for this study =  4.4622556000000006e-08\n",
      "Total summed mutation rate across ASXL1 nonsense for this study =  2.6285938e-08\n",
      "Total summed mutation rate across ASXL1 nonsense for this study =  2.2726005200000001e-07\n",
      "Total summed mutation rate across ASXL1 nonsense for this study =  2.93543906e-07\n",
      "Total summed mutation rate across ASXL1 nonsense for this study =  1.20044e-08\n",
      "Total summed mutation rate across ASXL1 nonsense for this study =  3.78353e-10\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Genovese's ASXL1 variants\n",
    "Genovese_miss_ASXL1 = mutation_rates_list_missense_variants_ASXL1('Data files/Genovese ASXL1 missense variants called.csv')\n",
    "Genovese_non_ASXL1_1 = mutation_rates_nonsense_range_of_variants_ASXL1(19, 128)\n",
    "Genovese_non_ASXL1_2 = mutation_rates_nonsense_range_of_variants_ASXL1(143, 196)\n",
    "Genovese_non_ASXL1_3 = mutation_rates_nonsense_range_of_variants_ASXL1(200, 627)\n",
    "Genovese_non_ASXL1_4 = mutation_rates_nonsense_range_of_variants_ASXL1(655, 1541)\n",
    "Genovese_non_ASXL1_5 = mutation_rates_nonsense_range_of_variants_ASXL1(141, 143)\n",
    "Genovese_non_ASXL1_6 = mutation_rates_nonsense_range_of_variants_ASXL1(627, 629)\n",
    "\n",
    "#Lists of ASXL1 variants targetable in Genovese 2014 (i.e. included in their panel)\n",
    "Genovese_nonsense_ASXL1 = []\n",
    "for i in Genovese_non_ASXL1_1:\n",
    "    Genovese_nonsense_ASXL1.append(str('p.')+i[1])\n",
    "for i in Genovese_non_ASXL1_2:\n",
    "    Genovese_nonsense_ASXL1.append(str('p.')+i[1])\n",
    "for i in Genovese_non_ASXL1_3:\n",
    "    Genovese_nonsense_ASXL1.append(str('p.')+i[1])\n",
    "for i in Genovese_non_ASXL1_4:\n",
    "    Genovese_nonsense_ASXL1.append(str('p.')+i[1])\n",
    "for i in Genovese_non_ASXL1_5:\n",
    "    Genovese_nonsense_ASXL1.append(str('p.')+i[1])\n",
    "for i in Genovese_non_ASXL1_6:\n",
    "    Genovese_nonsense_ASXL1.append(str('p.')+i[1])\n",
    "    \n",
    "Genovese_missense_ASXL1 = []\n",
    "for i in Genovese_miss_ASXL1:\n",
    "    Genovese_missense_ASXL1.append(i[3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 nonsense for this study =  7.915323800000001e-08\n",
      "Total summed mutation rate across TP53 missense for this study =  4.1547220999999997e-07\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Genovese's TP53 variants\n",
    "Genovese_non_TP53 = mutation_rates_list_nonsense_variants_TP53('Data files/Genovese TP53 nonsense variants called.csv')\n",
    "Genovese_miss_TP53 = mutation_rates_list_missense_variants_TP53('Data files/Genovese TP53 missense variants called.csv')\n",
    "\n",
    "#Lists of TP53 variants targetable in Genovese 2014 (i.e. included in their panel)\n",
    "Genovese_nonsense_TP53 = []\n",
    "for i in Genovese_non_TP53:\n",
    "    Genovese_nonsense_TP53.append(i[3])\n",
    "    \n",
    "Genovese_missense_TP53 = []\n",
    "for i in Genovese_miss_TP53:\n",
    "    Genovese_missense_TP53.append(i[3])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## McKerrel 2015"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "#List of missense variants in DNMT3A called by McKerrel\n",
    "McKerrel_missense_DNMT3A = ['p.R882H', 'p.R882C', 'p.R882P', 'p.R882S', 'p.R882G', 'p.R882L']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Zink 2017"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "code_folding": [],
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  3.8678204400000003e-07\n",
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  1.211073846e-06\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Zink's DNMT3A variants\n",
    "Zink_non_DNMT3A = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(0, 913)\n",
    "Zink_miss_DNMT3A = mutation_rates_list_missense_variants_DNMT3A_NP072046('Data files/Zink DNMT3A missense variants in COSMIC.csv')\n",
    "\n",
    "#Lists of DNMT3A variants targetable in Zink 2017 (i.e. included in their panel)\n",
    "Zink_nonsense_DNMT3A = []\n",
    "for i in Zink_non_DNMT3A:\n",
    "    Zink_nonsense_DNMT3A.append(str('p.')+i[1])\n",
    "\n",
    "Zink_missense_DNMT3A = []\n",
    "for i in Zink_miss_DNMT3A:\n",
    "    Zink_missense_DNMT3A.append(i[3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "code_folding": [],
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TET2 nonsense_NP_01120680 for this study =  8.664021850000001e-07\n",
      "Total summed mutation rate across TET2 missense_NP_01120680 for this study =  1.090214402e-06\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Zink's TET2 variants\n",
    "Zink_non_TET2 = mutation_rates_nonsense_range_of_variants_TET2(0, 2003)\n",
    "Zink_miss_TET2 = mutation_rates_list_missense_variants_TET2('Data files/Zink TET2 missense variants in COSMIC.csv')\n",
    "\n",
    "#Lists of TET2 variants targetable in Zink 2017 (i.e. included in their panel)\n",
    "Zink_nonsense_TET2 = []\n",
    "for i in Zink_non_TET2:\n",
    "    Zink_nonsense_TET2.append(str('p.')+i[1])\n",
    "    \n",
    "Zink_missense_TET2 = []\n",
    "for i in Zink_miss_TET2:\n",
    "    Zink_missense_TET2.append(i[3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "code_folding": [],
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across ASXL1 nonsense for this study =  6.254075530000001e-07\n",
      "Total summed mutation rate across ASXL1 missense for this study =  5.189833720000001e-07\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Zink's ASXL1 variants\n",
    "Zink_non_ASXL1 = mutation_rates_nonsense_range_of_variants_ASXL1(0, 1542)\n",
    "Zink_miss_ASXL1 = mutation_rates_list_missense_variants_ASXL1('Data files/Zink ASXL1 missense variants called.csv')\n",
    "\n",
    "#Lists of ASXL1 variants targetable in Zink 2017 (i.e. included in their panel)\n",
    "Zink_nonsense_ASXL1 = []\n",
    "for i in Zink_non_ASXL1:\n",
    "    Zink_nonsense_ASXL1.append(str('p.')+i[1])\n",
    "    \n",
    "Zink_missense_ASXL1 = []\n",
    "for i in Zink_miss_ASXL1:\n",
    "    Zink_missense_ASXL1.append(i[3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "code_folding": [],
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 nonsense for this study =  1.6147847e-07\n",
      "Total summed mutation rate across TP53 missense for this study =  1.464249575e-06\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Zink's TP53 variants\n",
    "Zink_non_TP53 = mutation_rates_nonsense_range_of_variants_TP53(0, 394)\n",
    "Zink_miss_TP53 = mutation_rates_list_missense_variants_TP53('Data files/Zink TP53 missense variants called.csv')\n",
    "\n",
    "#Lists of TP53 variants targetable in Zink 2017 (i.e. included in their panel)\n",
    "Zink_nonsense_TP53 = []\n",
    "for i in Zink_non_TP53:\n",
    "    Zink_nonsense_TP53.append(str('p.')+i[1])\n",
    "    \n",
    "Zink_missense_TP53 = []\n",
    "for i in Zink_miss_TP53:\n",
    "    Zink_missense_TP53.append(i[3])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Coombs 2017"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  3.8678204400000003e-07\n",
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  8.083470887e-06\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Coombs' DNMT3A variants\n",
    "Coombs_non_DNMT3A = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(0, 913)\n",
    "Coombs_miss_DNMT3A = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(0, 913)\n",
    "\n",
    "#Lists of DNMT3A variants targetable in Coombs 2017 (i.e. included in their panel)\n",
    "Coombs_nonsense_DNMT3A = []\n",
    "for i in Coombs_non_DNMT3A:\n",
    "    Coombs_nonsense_DNMT3A.append(str('p.')+i[1])\n",
    "    \n",
    "Coombs_missense_DNMT3A = []\n",
    "for i in Coombs_miss_DNMT3A:\n",
    "    Coombs_missense_DNMT3A.append(str('p.')+i[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TET2 nonsense_NP_01120680 for this study =  8.664021850000001e-07\n",
      "Total summed mutation rate across TET2 missense_NP_01120680 for this study =  1.2216578715e-05\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Coombs' TET2 variants\n",
    "Coombs_non_TET2 = mutation_rates_nonsense_range_of_variants_TET2(0, 2003)\n",
    "Coombs_miss_TET2 = mutation_rates_missense_range_of_variants_TET2(0, 2003)\n",
    "\n",
    "#Lists of TET2 variants targetable in Coombs 2017 (i.e. included in their panel)\n",
    "Coombs_nonsense_TET2 = []\n",
    "for i in Coombs_non_TET2:\n",
    "    Coombs_nonsense_TET2.append(str('p.')+i[1])\n",
    "    \n",
    "Coombs_missense_TET2 = []\n",
    "for i in Coombs_miss_TET2:\n",
    "    Coombs_missense_TET2.append(str('p.')+i[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across ASXL1 nonsense for this study =  6.254075530000001e-07\n",
      "Total summed mutation rate across ASXL1 missense for this study =  1.1360431228e-05\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Coombs' ASXL1 variants\n",
    "Coombs_non_ASXL1 = mutation_rates_nonsense_range_of_variants_ASXL1(0, 1542)\n",
    "Coombs_miss_ASXL1 = mutation_rates_missense_range_of_variants_ASXL1(0, 1542)\n",
    "\n",
    "#Lists of ASXL1 variants targetable in Coombs 2017 (i.e. included in their panel)\n",
    "Coombs_nonsense_ASXL1 = []\n",
    "for i in Coombs_non_ASXL1:\n",
    "    Coombs_nonsense_ASXL1.append(str('p.')+i[1])\n",
    "    \n",
    "Coombs_missense_ASXL1 = []\n",
    "for i in Coombs_miss_ASXL1:\n",
    "    Coombs_missense_ASXL1.append(str('p.')+i[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 nonsense for this study =  1.6147847e-07\n",
      "Total summed mutation rate across TP53 missense for this study =  3.273721457e-06\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Coombs' TP53 variants\n",
    "Coombs_non_TP53 = mutation_rates_nonsense_range_of_variants_TP53(0, 394)\n",
    "Coombs_miss_TP53 = mutation_rates_missense_range_of_variants_TP53(0, 394)\n",
    "\n",
    "#Lists of TP53 variants targetable in Coombs 2017 (i.e. included in their panel)\n",
    "Coombs_nonsense_TP53 = []\n",
    "for i in Coombs_non_TP53:\n",
    "    Coombs_nonsense_TP53.append(str('p.')+i[1])\n",
    "    \n",
    "Coombs_missense_TP53 = []\n",
    "for i in Coombs_miss_TP53:\n",
    "    Coombs_missense_TP53.append(str('p.')+i[1])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Young 2016 & 2019"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  3.8678204400000003e-07\n",
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  8.083470887e-06\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Young DNMT3A variants\n",
    "Young_non_DNMT3A = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(0, 913)\n",
    "Young_miss_DNMT3A = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(0, 913)\n",
    "\n",
    "#Lists of DNMT3A variants targetable in Young (i.e. included in their panel)\n",
    "Young_nonsense_DNMT3A = []\n",
    "for i in Young_non_DNMT3A:\n",
    "    Young_nonsense_DNMT3A.append(str('p.')+i[1])\n",
    "    \n",
    "Young_missense_DNMT3A = []\n",
    "for i in Young_miss_DNMT3A:\n",
    "    Young_missense_DNMT3A.append(str('p.')+i[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TET2 nonsense_NP_01120680 for this study =  8.664021850000001e-07\n",
      "Total summed mutation rate across TET2 missense_NP_01120680 for this study =  1.2216578715e-05\n",
      "Total summed mutation rate across TET2 missense_NP_01120680 for this study =  5.206542600000001e-08\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Young TET2 variants\n",
    "Young_non_TET2 = mutation_rates_nonsense_range_of_variants_TET2(0, 2003)\n",
    "Young_miss_TET2 = mutation_rates_missense_range_of_variants_TET2(0, 2003)\n",
    "\n",
    "#Lists of TET2 variants targetable in Young (i.e. included in their panel)\n",
    "Young_nonsense_TET2 = []\n",
    "for i in Young_non_TET2:\n",
    "    Young_nonsense_TET2.append(str('p.')+i[1])\n",
    "    \n",
    "Young_missense_TET2_all = []\n",
    "for i in Young_miss_TET2:\n",
    "    Young_missense_TET2_all.append(str('p.')+i[1])\n",
    "    \n",
    "#Excluded regions:\n",
    "excluded_TET2_Young=[]\n",
    "Young_TET2_excluded = mutation_rates_list_missense_variants_TET2('Data files/excluded_variants_Young_TET2.csv')\n",
    "for i in Young_TET2_excluded:\n",
    "    excluded_TET2_Young.append(i[3])\n",
    "\n",
    "Young_missense_TET2=[]\n",
    "for i in Young_missense_TET2_all:\n",
    "    if i not in Young_TET2_excluded:\n",
    "        Young_missense_TET2.append(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across ASXL1 nonsense for this study =  3.3358587600000003e-07\n",
      "Total summed mutation rate across ASXL1 missense for this study =  6.820950358e-06\n",
      "Total summed mutation rate across ASXL1 missense for this study =  5.2494673999999997e-08\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Young ASXL1 variants\n",
    "Young_non_ASXL1 = mutation_rates_nonsense_range_of_variants_ASXL1(573, 1542)\n",
    "Young_miss_ASXL1 = mutation_rates_missense_range_of_variants_ASXL1(573, 1542)\n",
    "\n",
    "#Lists of ASXL1 variants targetable in Young (i.e. included in their panel)\n",
    "Young_nonsense_ASXL1 = []\n",
    "for i in Young_non_ASXL1:\n",
    "    Young_nonsense_ASXL1.append(str('p.')+i[1])\n",
    "    \n",
    "Young_missense_ASXL1_all = []\n",
    "for i in Young_miss_ASXL1:\n",
    "    Young_missense_ASXL1_all.append(str('p.')+i[1])\n",
    "    \n",
    "#Excluded regions:\n",
    "excluded_ASXL1_Young=[]\n",
    "Young_ASXL1_excluded = mutation_rates_list_missense_variants_ASXL1('Data files/excluded_variants_Young_ASXL1.csv')\n",
    "for i in Young_ASXL1_excluded:\n",
    "    excluded_ASXL1_Young.append(i[3])\n",
    "\n",
    "Young_missense_ASXL1=[]\n",
    "for i in Young_missense_ASXL1_all:\n",
    "    if i not in Young_ASXL1_excluded:\n",
    "        Young_missense_ASXL1.append(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 nonsense for this study =  1.6147847e-07\n",
      "Total summed mutation rate across TP53 missense for this study =  3.273721457e-06\n",
      "Total summed mutation rate across TP53 missense for this study =  1.27013e-09\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Young TP53 variants\n",
    "Young_non_TP53 = mutation_rates_nonsense_range_of_variants_TP53(0, 394)\n",
    "Young_miss_TP53 = mutation_rates_missense_range_of_variants_TP53(0, 394)\n",
    "\n",
    "#Lists of TP53 variants targetable in Young (i.e. included in their panel)\n",
    "Young_nonsense_TP53 = []\n",
    "for i in Young_non_TP53:\n",
    "    Young_nonsense_TP53.append(str('p.')+i[1])\n",
    "    \n",
    "Young_missense_TP53_all = []\n",
    "for i in Young_miss_TP53:\n",
    "    Young_missense_TP53_all.append(str('p.')+i[1])\n",
    "    \n",
    "#Excluded regions:\n",
    "excluded_TP53_Young=[]\n",
    "Young_TP53_excluded = mutation_rates_list_missense_variants_TP53('Data files/excluded_variants_Young_TP53.csv')\n",
    "for i in Young_TP53_excluded:\n",
    "    excluded_TP53_Young.append(i[3])\n",
    "\n",
    "Young_missense_TP53=[]\n",
    "for i in Young_missense_TP53_all:\n",
    "    if i not in Young_TP53_excluded:\n",
    "        Young_missense_TP53.append(i)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Desai 2018"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  3.8678204400000003e-07\n",
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  8.083470887e-06\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Desai DNMT3A variants\n",
    "Desai_non_DNMT3A = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(0, 913)\n",
    "Desai_miss_DNMT3A = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(0, 913)\n",
    "\n",
    "#Lists of DNMT3A variants targetable in Desai (i.e. included in their panel)\n",
    "Desai_nonsense_DNMT3A = []\n",
    "for i in Desai_non_DNMT3A:\n",
    "    Desai_nonsense_DNMT3A.append(str('p.')+i[1])\n",
    "    \n",
    "Desai_missense_DNMT3A = []\n",
    "for i in Desai_miss_DNMT3A:\n",
    "    Desai_missense_DNMT3A.append(str('p.')+i[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TET2 nonsense_NP_01120680 for this study =  8.664021850000001e-07\n",
      "Total summed mutation rate across TET2 missense_NP_01120680 for this study =  1.2216578715e-05\n",
      "Total summed mutation rate across TET2 missense_NP_01120680 for this study =  8.3116988e-08\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Desai TET2 variants\n",
    "Desai_non_TET2 = mutation_rates_nonsense_range_of_variants_TET2(0, 2003)\n",
    "Desai_miss_TET2 = mutation_rates_missense_range_of_variants_TET2(0, 2003)\n",
    "\n",
    "#Lists of TET2 variants targetable in Desai (i.e. included in their panel)\n",
    "Desai_nonsense_TET2 = []\n",
    "for i in Desai_non_TET2:\n",
    "    Desai_nonsense_TET2.append(str('p.')+i[1])\n",
    "    \n",
    "Desai_missense_TET2_all = []\n",
    "for i in Desai_miss_TET2:\n",
    "    Desai_missense_TET2_all.append(str('p.')+i[1])\n",
    "    \n",
    "#Excluded regions:\n",
    "excluded_TET2_Desai=[]\n",
    "Desai_TET2_excluded = mutation_rates_list_missense_variants_TET2('Data files/excluded_variants_Desai_2018_TET2.csv')\n",
    "for i in Desai_TET2_excluded:\n",
    "    excluded_TET2_Desai.append(i[3])\n",
    "\n",
    "Desai_missense_TET2=[]\n",
    "for i in Desai_missense_TET2_all:\n",
    "    if i not in Desai_TET2_excluded:\n",
    "        Desai_missense_TET2.append(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across ASXL1 nonsense for this study =  6.254075530000001e-07\n",
      "Total summed mutation rate across ASXL1 missense for this study =  1.1360431228e-05\n",
      "Total summed mutation rate across ASXL1 missense for this study =  5.8171873e-08\n",
      "Total summed mutation rate across ASXL1 nonsense for this study =  5.39508e-10\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Desai ASXL1 variants\n",
    "Desai_non_ASXL1 = mutation_rates_nonsense_range_of_variants_ASXL1(0, 1542)\n",
    "Desai_miss_ASXL1 = mutation_rates_missense_range_of_variants_ASXL1(0, 1542)\n",
    "\n",
    "#Lists of ASXL1 variants targetable in Desai (i.e. included in their panel)\n",
    "Desai_nonsense_ASXL1_all = []\n",
    "for i in Desai_non_ASXL1:\n",
    "    Desai_nonsense_ASXL1_all.append(str('p.')+i[1])\n",
    "    \n",
    "Desai_missense_ASXL1_all = []\n",
    "for i in Desai_miss_ASXL1:\n",
    "    Desai_missense_ASXL1_all.append(str('p.')+i[1])\n",
    "    \n",
    "#Excluded regions:\n",
    "excluded_ASXL1_Desai_missense=[]\n",
    "Desai_ASXL1_excluded_missense = mutation_rates_list_missense_variants_ASXL1('Data files/excluded_variants_Desai_2018_ASXL1_missense.csv')\n",
    "for i in Desai_ASXL1_excluded_missense:\n",
    "    excluded_ASXL1_Desai_missense.append(i[3])\n",
    "\n",
    "Desai_missense_ASXL1=[]\n",
    "for i in Desai_missense_ASXL1_all:\n",
    "    if i not in Desai_ASXL1_excluded_missense:\n",
    "        Desai_missense_ASXL1.append(i)\n",
    "        \n",
    "excluded_ASXL1_Desai_nonsense=[]\n",
    "Desai_ASXL1_excluded_nonsense = mutation_rates_list_nonsense_variants_ASXL1('Data files/excluded_variants_Desai_2018_ASXL1_nonsense.csv')\n",
    "for i in Desai_ASXL1_excluded_nonsense:\n",
    "    excluded_ASXL1_Desai_nonsense.append(i[3])\n",
    "\n",
    "Desai_nonsense_ASXL1=[]\n",
    "for i in Desai_nonsense_ASXL1_all:\n",
    "    if i not in Desai_ASXL1_excluded_nonsense:\n",
    "        Desai_nonsense_ASXL1.append(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 nonsense for this study =  1.6147847e-07\n",
      "Total summed mutation rate across TP53 missense for this study =  3.273721457e-06\n",
      "Total summed mutation rate across TP53 missense for this study =  4.730426e-09\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Desai TP53 variants\n",
    "Desai_non_TP53 = mutation_rates_nonsense_range_of_variants_TP53(0, 394)\n",
    "Desai_miss_TP53 = mutation_rates_missense_range_of_variants_TP53(0, 394)\n",
    "\n",
    "#Lists of TP53 variants targetable in Desai (i.e. included in their panel)\n",
    "Desai_nonsense_TP53 = []\n",
    "for i in Desai_non_TP53:\n",
    "    Desai_nonsense_TP53.append(str('p.')+i[1])\n",
    "    \n",
    "Desai_missense_TP53_all = []\n",
    "for i in Desai_miss_TP53:\n",
    "    Desai_missense_TP53_all.append(str('p.')+i[1])\n",
    "    \n",
    "#Excluded regions:\n",
    "excluded_TP53_Desai=[]\n",
    "Desai_TP53_excluded = mutation_rates_list_missense_variants_TP53('Data files/excluded_variants_Desai_2018_TP53.csv')\n",
    "for i in Desai_TP53_excluded:\n",
    "    excluded_TP53_Desai.append(i[3])\n",
    "\n",
    "Desai_missense_TP53=[]\n",
    "for i in Desai_missense_TP53_all:\n",
    "    if i not in Desai_TP53_excluded:\n",
    "        Desai_missense_TP53.append(i)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Acuna-Hidalgo 2017"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  3.741791e-09\n",
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  9.093981e-09\n",
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  1.3048360000000002e-09\n",
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  2.0864156e-08\n",
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  1.4313799999999999e-09\n",
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  2.5605291000000003e-08\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Acuna-Hidalgo DNMT3A variants\n",
    "Acuna_complete_nonsense_DNMT3A_1 = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(874,892)\n",
    "Acuna_complete_nonsense_DNMT3A_2 = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(851,866)\n",
    "Acuna_complete_nonsense_DNMT3A_3 = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(849,857)\n",
    "Acuna_complete_nonsense_DNMT3A_4 = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(758,774)\n",
    "Acuna_complete_nonsense_DNMT3A_5 = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(733,740)\n",
    "Acuna_complete_nonsense_DNMT3A_6 = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(317,334)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  1.987322e-07\n",
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  7.3492762e-08\n",
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  3.745106700000001e-08\n",
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  1.535384e-07\n",
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  6.461601599999999e-08\n",
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  2.0140387e-07\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Acuna-Hidalgo DNMT3A variants\n",
    "Acuna_complete_missense_DNMT3A_1 = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(874,892)\n",
    "Acuna_complete_missense_DNMT3A_2 = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(851,866)\n",
    "Acuna_complete_missense_DNMT3A_3 = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(849,857)\n",
    "Acuna_complete_missense_DNMT3A_4 = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(758,774)\n",
    "Acuna_complete_missense_DNMT3A_5 = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(733,740)\n",
    "Acuna_complete_missense_DNMT3A_6 = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(317,334)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "#Lists of DNMT3A variants targetable in Acunda-Hidalgo (i.e. included in their panel)\n",
    "Acuna_nonsense_DNMT3A = []\n",
    "for i in Acuna_complete_nonsense_DNMT3A_1:\n",
    "    Acuna_nonsense_DNMT3A.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_DNMT3A_2:\n",
    "    Acuna_nonsense_DNMT3A.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_DNMT3A_3:\n",
    "    Acuna_nonsense_DNMT3A.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_DNMT3A_4:\n",
    "    Acuna_nonsense_DNMT3A.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_DNMT3A_5:\n",
    "    Acuna_nonsense_DNMT3A.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_DNMT3A_6:\n",
    "    Acuna_nonsense_DNMT3A.append(str('p.')+i[1])\n",
    "    \n",
    "Acuna_missense_DNMT3A = []\n",
    "for i in Acuna_complete_missense_DNMT3A_1:\n",
    "    Acuna_missense_DNMT3A.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_DNMT3A_2:\n",
    "    Acuna_missense_DNMT3A.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_DNMT3A_3:\n",
    "    Acuna_missense_DNMT3A.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_DNMT3A_4:\n",
    "    Acuna_missense_DNMT3A.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_DNMT3A_5:\n",
    "    Acuna_missense_DNMT3A.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_DNMT3A_6:\n",
    "    Acuna_missense_DNMT3A.append(str('p.')+i[1])\n",
    "    \n",
    "for i in ['p.Y874S', 'p.Y874F', 'p.Y874C', 'p.F851Y', 'p.F851S', 'p.F851C', 'p.F851L', 'p.I858L', 'p.I858F', 'p.I858V', 'p.I858N', 'p.I858T', 'p.I858S',\\\n",
    "'p.A741T', 'p.A741P', 'p.A741S', 'p.A741E', 'p.A741V', 'p.A741G', 'p.E733A', 'p.E733V', 'p.E733G', 'p.E733D']:\n",
    "    Acuna_missense_DNMT3A.append(i)\n",
    "    \n",
    "for i in ['p.Y874*', 'p.E733*']:\n",
    "    Acuna_nonsense_DNMT3A.append(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TET2 nonsense_NP_01120680 for this study =  5.5247350000000006e-09\n",
      "Total summed mutation rate across TET2 nonsense_NP_01120680 for this study =  1.249072e-08\n",
      "Total summed mutation rate across TET2 nonsense_NP_01120680 for this study =  3.501569e-08\n",
      "Total summed mutation rate across TET2 nonsense_NP_01120680 for this study =  9.645627e-09\n",
      "Total summed mutation rate across TET2 nonsense_NP_01120680 for this study =  3.998035e-09\n",
      "Total summed mutation rate across TET2 nonsense_NP_01120680 for this study =  7.165286e-09\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Acuna-Hidalgo TET2 variants\n",
    "Acuna_complete_nonsense_TET2_1 = mutation_rates_nonsense_range_of_variants_TET2(527, 534)\n",
    "Acuna_complete_nonsense_TET2_2 = mutation_rates_nonsense_range_of_variants_TET2(541, 546)\n",
    "Acuna_complete_nonsense_TET2_3 = mutation_rates_nonsense_range_of_variants_TET2(547, 552)\n",
    "Acuna_complete_nonsense_TET2_4 = mutation_rates_nonsense_range_of_variants_TET2(754, 773)\n",
    "Acuna_complete_nonsense_TET2_5 = mutation_rates_nonsense_range_of_variants_TET2(844, 849)\n",
    "Acuna_complete_nonsense_TET2_6 = mutation_rates_nonsense_range_of_variants_TET2(885, 890)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TET2 missense_NP_01120680 for this study =  3.135999e-08\n",
      "Total summed mutation rate across TET2 missense_NP_01120680 for this study =  3.5936637999999997e-08\n",
      "Total summed mutation rate across TET2 missense_NP_01120680 for this study =  3.3724941999999994e-08\n",
      "Total summed mutation rate across TET2 missense_NP_01120680 for this study =  9.4240322e-08\n",
      "Total summed mutation rate across TET2 missense_NP_01120680 for this study =  2.3836145e-08\n",
      "Total summed mutation rate across TET2 missense_NP_01120680 for this study =  2.016166e-08\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Acuna-Hidalgo TET2 variants\n",
    "Acuna_complete_missense_TET2_1 = mutation_rates_missense_range_of_variants_TET2(527, 534)\n",
    "Acuna_complete_missense_TET2_2 = mutation_rates_missense_range_of_variants_TET2(541, 546)\n",
    "Acuna_complete_missense_TET2_3 = mutation_rates_missense_range_of_variants_TET2(547, 552)\n",
    "Acuna_complete_missense_TET2_4 = mutation_rates_missense_range_of_variants_TET2(754, 773)\n",
    "Acuna_complete_missense_TET2_5 = mutation_rates_missense_range_of_variants_TET2(844, 849)\n",
    "Acuna_complete_missense_TET2_6 = mutation_rates_missense_range_of_variants_TET2(885, 890)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "#Lists of TET2 variants targetable in Acunda-Hidalgo (i.e. included in their panel)\n",
    "Acuna_nonsense_TET2 = []\n",
    "for i in Acuna_complete_nonsense_TET2_1:\n",
    "    Acuna_nonsense_TET2.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_TET2_2:\n",
    "    Acuna_nonsense_TET2.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_TET2_3:\n",
    "    Acuna_nonsense_TET2.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_TET2_4:\n",
    "    Acuna_nonsense_TET2.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_TET2_5:\n",
    "    Acuna_nonsense_TET2.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_TET2_6:\n",
    "    Acuna_nonsense_TET2.append(str('p.')+i[1])\n",
    "    \n",
    "Acuna_missense_TET2 = []\n",
    "for i in Acuna_complete_missense_TET2_1:\n",
    "    Acuna_missense_TET2.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_TET2_2:\n",
    "    Acuna_missense_TET2.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_TET2_3:\n",
    "    Acuna_missense_TET2.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_TET2_4:\n",
    "    Acuna_missense_TET2.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_TET2_5:\n",
    "    Acuna_missense_TET2.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_TET2_6:\n",
    "    Acuna_missense_TET2.append(str('p.')+i[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across ASXL1 nonsense for this study =  3.618524e-08\n",
      "Total summed mutation rate across ASXL1 nonsense for this study =  5.3296666999999994e-08\n",
      "Total summed mutation rate across ASXL1 nonsense for this study =  1.1963488e-08\n",
      "Total summed mutation rate across ASXL1 nonsense for this study =  2.2857128e-08\n",
      "Total summed mutation rate across ASXL1 nonsense for this study =  4.384787e-09\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Acuna-Hidalgo ASXL1 variants\n",
    "Acuna_complete_nonsense_ASXL1_1 = mutation_rates_nonsense_range_of_variants_ASXL1(400, 407)\n",
    "Acuna_complete_nonsense_ASXL1_2 = mutation_rates_nonsense_range_of_variants_ASXL1(413, 573)\n",
    "Acuna_complete_nonsense_ASXL1_3 = mutation_rates_nonsense_range_of_variants_ASXL1(572, 596)\n",
    "Acuna_complete_nonsense_ASXL1_4 = mutation_rates_nonsense_range_of_variants_ASXL1(688, 694)\n",
    "Acuna_complete_nonsense_ASXL1_5 = mutation_rates_nonsense_range_of_variants_ASXL1(799, 804)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across ASXL1 missense for this study =  5.904737100000001e-08\n",
      "Total summed mutation rate across ASXL1 missense for this study =  1.1949677830000001e-06\n",
      "Total summed mutation rate across ASXL1 missense for this study =  2.43229732e-07\n",
      "Total summed mutation rate across ASXL1 missense for this study =  3.6595521e-08\n",
      "Total summed mutation rate across ASXL1 missense for this study =  2.3812267999999998e-08\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Acuna-Hidalgo ASXL1 variants\n",
    "Acuna_complete_missense_ASXL1_1 = mutation_rates_missense_range_of_variants_ASXL1(400, 407)\n",
    "Acuna_complete_missense_ASXL1_2 = mutation_rates_missense_range_of_variants_ASXL1(413, 573)\n",
    "Acuna_complete_missense_ASXL1_3 = mutation_rates_missense_range_of_variants_ASXL1(572, 596)\n",
    "Acuna_complete_missense_ASXL1_4 = mutation_rates_missense_range_of_variants_ASXL1(688, 694)\n",
    "Acuna_complete_missense_ASXL1_5 = mutation_rates_missense_range_of_variants_ASXL1(799, 804)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#Lists of ASXL1 variants targetable in Acunda-Hidalgo (i.e. included in their panel)\n",
    "Acuna_nonsense_ASXL1 = []\n",
    "for i in Acuna_complete_nonsense_ASXL1_1:\n",
    "    Acuna_nonsense_ASXL1.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_ASXL1_2:\n",
    "    Acuna_nonsense_ASXL1.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_ASXL1_3:\n",
    "    Acuna_nonsense_ASXL1.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_ASXL1_4:\n",
    "    Acuna_nonsense_ASXL1.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_ASXL1_5:\n",
    "    Acuna_nonsense_ASXL1.append(str('p.')+i[1])\n",
    "    \n",
    "Acuna_missense_ASXL1 = []\n",
    "for i in Acuna_complete_missense_ASXL1_1:\n",
    "    Acuna_missense_ASXL1.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_ASXL1_2:\n",
    "    Acuna_missense_ASXL1.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_ASXL1_3:\n",
    "    Acuna_missense_ASXL1.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_ASXL1_4:\n",
    "    Acuna_missense_ASXL1.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_ASXL1_5:\n",
    "    Acuna_missense_ASXL1.append(str('p.')+i[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 nonsense for this study =  3.2044719e-08\n",
      "Total summed mutation rate across TP53 nonsense for this study =  3.3097040000000003e-09\n",
      "Total summed mutation rate across TP53 nonsense for this study =  1.3440674e-08\n",
      "Total summed mutation rate across TP53 nonsense for this study =  1.485146e-09\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Acuna-Hidalgo TP53 variants\n",
    "Acuna_complete_nonsense_TP53_1 = mutation_rates_nonsense_range_of_variants_TP53(185, 221)\n",
    "Acuna_complete_nonsense_TP53_2 = mutation_rates_nonsense_range_of_variants_TP53(168, 185)\n",
    "Acuna_complete_nonsense_TP53_3 = mutation_rates_nonsense_range_of_variants_TP53(117, 146)\n",
    "Acuna_complete_nonsense_TP53_4 = mutation_rates_nonsense_range_of_variants_TP53(59, 81)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 missense for this study =  2.75197872e-07\n",
      "Total summed mutation rate across TP53 missense for this study =  2.0262653100000002e-07\n",
      "Total summed mutation rate across TP53 missense for this study =  1.9660014099999998e-07\n",
      "Total summed mutation rate across TP53 missense for this study =  1.9621064499999997e-07\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Acuna-Hidalgo TP53 variants\n",
    "Acuna_complete_missense_TP53_1 = mutation_rates_missense_range_of_variants_TP53(185, 221)\n",
    "Acuna_complete_missense_TP53_2 = mutation_rates_missense_range_of_variants_TP53(168, 185)\n",
    "Acuna_complete_missense_TP53_3 = mutation_rates_missense_range_of_variants_TP53(117, 146)\n",
    "Acuna_complete_missense_TP53_4 = mutation_rates_missense_range_of_variants_TP53(59, 81)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#Lists of TP53 variants targetable in Acunda-Hidalgo (i.e. included in their panel)\n",
    "Acuna_nonsense_TP53 = []\n",
    "for i in Acuna_complete_nonsense_TP53_1:\n",
    "    Acuna_nonsense_TP53.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_TP53_2:\n",
    "    Acuna_nonsense_TP53.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_TP53_3:\n",
    "    Acuna_nonsense_TP53.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_TP53_4:\n",
    "    Acuna_nonsense_TP53.append(str('p.')+i[1])\n",
    "    \n",
    "Acuna_missense_TP53_all = []\n",
    "for i in Acuna_complete_missense_TP53_1:\n",
    "    Acuna_missense_TP53_all.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_TP53_2:\n",
    "    Acuna_missense_TP53_all.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_TP53_3:\n",
    "    Acuna_missense_TP53_all.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_TP53_4:\n",
    "    Acuna_missense_TP53_all.append(str('p.')+i[1])\n",
    "    \n",
    "#Excluded regions:\n",
    "Acuna_TP53_excluded=['p.P72R']\n",
    "\n",
    "Acuna_missense_TP53=[]\n",
    "for i in Acuna_missense_TP53_all:\n",
    "    if i not in Acuna_TP53_excluded:\n",
    "        Acuna_missense_TP53.append(i)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Lists of all possible variants in DNMT3A, TET2, ASXL1, TP53"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Next, create lists of ALL possible variants in DNMT3A, TET2, ASXL1, TP53 so can look at each of them in turn and see how often they are observed and then use this to calculate what fitness effect would explain this number of observations.\n",
    "- The mutation rate calculator functions also produce a list of all the possible variants so use this function to generate a list of all the possible variants within each gene."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DNMT3A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  3.8678204400000003e-07\n",
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  8.083470887e-06\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across DNMT3A\n",
    "all_nonsense_DNMT3A = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(0, 913)\n",
    "all_missense_DNMT3A = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(0, 913)\n",
    "\n",
    "#List of possible variants\n",
    "all_nonsense_variants_DNMT3A = []\n",
    "for i in all_nonsense_DNMT3A:\n",
    "    all_nonsense_variants_DNMT3A.append(str('p.')+i[1])\n",
    "    \n",
    "all_missense_variants_DNMT3A = []\n",
    "for i in all_missense_DNMT3A:\n",
    "    all_missense_variants_DNMT3A.append(str('p.')+i[1])\n",
    "\n",
    "all_missense_variants_DNMT3A = list(dict.fromkeys(all_missense_variants_DNMT3A))\n",
    "all_nonsense_variants_DNMT3A = list(dict.fromkeys(all_nonsense_variants_DNMT3A))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TET2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TET2 nonsense_NP_01120680 for this study =  8.664021850000001e-07\n",
      "Total summed mutation rate across TET2 missense_NP_01120680 for this study =  1.2216578715e-05\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across TET2\n",
    "all_nonsense_TET2 = mutation_rates_nonsense_range_of_variants_TET2(0, 2003)\n",
    "all_missense_TET2 = mutation_rates_missense_range_of_variants_TET2(0, 2003)\n",
    "\n",
    "#List of possible variants\n",
    "all_nonsense_variants_TET2 = []\n",
    "for i in all_nonsense_TET2:\n",
    "    all_nonsense_variants_TET2.append(str('p.')+i[1])\n",
    "    \n",
    "all_missense_variants_TET2 = []\n",
    "for i in all_missense_TET2:\n",
    "    all_missense_variants_TET2.append(str('p.')+i[1])\n",
    "\n",
    "all_missense_variants_TET2 = list(dict.fromkeys(all_missense_variants_TET2))\n",
    "all_nonsense_variants_TET2 = list(dict.fromkeys(all_nonsense_variants_TET2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ASXL1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across ASXL1 nonsense for this study =  6.254075530000001e-07\n",
      "Total summed mutation rate across ASXL1 missense for this study =  1.1360431228e-05\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across TET2\n",
    "all_nonsense_ASXL1 = mutation_rates_nonsense_range_of_variants_ASXL1(0, 1542)\n",
    "all_missense_ASXL1 = mutation_rates_missense_range_of_variants_ASXL1(0, 1542)\n",
    "\n",
    "#List of possible variants\n",
    "all_nonsense_variants_ASXL1 = []\n",
    "for i in all_nonsense_ASXL1:\n",
    "    all_nonsense_variants_ASXL1.append(str('p.')+i[1])\n",
    "    \n",
    "all_missense_variants_ASXL1 = []\n",
    "for i in all_missense_ASXL1:\n",
    "    all_missense_variants_ASXL1.append(str('p.')+i[1])\n",
    "\n",
    "all_missense_variants_ASXL1 = list(dict.fromkeys(all_missense_variants_ASXL1))\n",
    "all_nonsense_variants_ASXL1 = list(dict.fromkeys(all_nonsense_variants_ASXL1))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TP53"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 nonsense for this study =  1.6147847e-07\n",
      "Total summed mutation rate across TP53 missense for this study =  3.273721457e-06\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across TP53\n",
    "all_nonsense_TP53 = mutation_rates_nonsense_range_of_variants_TP53(0, 394)\n",
    "all_missense_TP53 = mutation_rates_missense_range_of_variants_TP53(0, 394)\n",
    "\n",
    "#List of possible variants\n",
    "all_nonsense_variants_TP53 = []\n",
    "for i in all_nonsense_TP53:\n",
    "    all_nonsense_variants_TP53.append(str('p.')+i[1])\n",
    "    \n",
    "all_missense_variants_TP53 = []\n",
    "for i in all_missense_TP53:\n",
    "    all_missense_variants_TP53.append(str('p.')+i[1])\n",
    "\n",
    "all_missense_variants_TP53 = list(dict.fromkeys(all_missense_variants_TP53))\n",
    "all_nonsense_variants_TP53 = list(dict.fromkeys(all_nonsense_variants_TP53))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Actual number of observations of each variant"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Look at each variant seen across the studies and count how many times it was observed."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DNMT3A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#Create lists of variants by studies and counts of each variant\n",
    "import csv\n",
    "filename=\"Data files/all_studies_trimmed_all_genes.csv\"\n",
    "\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    DNMT3A_results={} #dictionary of DNMTA results (key = ID (row[13]), value = VAF (row[10]))\n",
    "\n",
    "    for row in read_reader:\n",
    "        if row_count>0:\n",
    "#             print(row)\n",
    "            if row[3] == 'DNMT3A':\n",
    "#                 print(row)\n",
    "                VAF = row[0]\n",
    "                age = row[1]\n",
    "                variant = str('p.')+row[2]\n",
    "                study = row[4]\n",
    "                DNMT3A_results[row_count]=(variant, study)\n",
    "                \n",
    "        row_count=row_count+1\n",
    "        \n",
    "all_DNMT3A_results_list = []\n",
    "for k, v in DNMT3A_results.items():\n",
    "    all_DNMT3A_results_list.append(v[0])\n",
    "        \n",
    "all_DNMT3A_df = pd.DataFrame({'variant':all_DNMT3A_results_list})\n",
    "all_grouped_df=all_DNMT3A_df.groupby(['variant']).size().reset_index(name='counts')\n",
    "all_grouped_df.to_csv('Data files/all_DNMT3A_variant_counts.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#All_DNMT3A_counts\n",
    "filename=\"Data files/all_DNMT3A_variant_counts.csv\"\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    all_studies_DNMT3A_counts={} #dictionary of DNMTA results (key = ID (row[13]), value = VAF (row[10]))\n",
    "\n",
    "    for row in read_reader:\n",
    "#         print(row)\n",
    "        if row_count>0:\n",
    "            variant = row[1]\n",
    "            counts = row[2]\n",
    "            all_studies_DNMT3A_counts[variant]=counts\n",
    "                \n",
    "        row_count=row_count+1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TET2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#Create lists of variants by studies and counts of each variant\n",
    "import csv\n",
    "filename=\"Data files/all_studies_trimmed_all_genes.csv\"\n",
    "\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    TET2_results={} #dictionary of DNMTA results (key = ID (row[13]), value = VAF (row[10]))\n",
    "\n",
    "    for row in read_reader:\n",
    "        if row_count>0:\n",
    "#             print(row)\n",
    "            if row[3] == 'TET2':\n",
    "#                 print(row)\n",
    "                VAF = row[0]\n",
    "                age = row[1]\n",
    "                variant = str('p.')+row[2]\n",
    "                study = row[4]\n",
    "                TET2_results[row_count]=(variant, study)\n",
    "                \n",
    "        row_count=row_count+1\n",
    "        \n",
    "all_TET2_results_list = []\n",
    "for k, v in TET2_results.items():\n",
    "    all_TET2_results_list.append(v[0])\n",
    "\n",
    "all_TET2_df = pd.DataFrame({'variant':all_TET2_results_list})\n",
    "all_grouped_df=all_TET2_df.groupby(['variant']).size().reset_index(name='counts')\n",
    "all_grouped_df.to_csv('Data files/all_TET2_variant_counts.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#All_TET2_counts\n",
    "filename=\"Data files/all_TET2_variant_counts.csv\"\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    all_studies_TET2_counts={} #dictionary of DNMTA results (key = ID (row[13]), value = VAF (row[10]))\n",
    "\n",
    "    for row in read_reader:\n",
    "#         print(row)\n",
    "        if row_count>0:\n",
    "            variant = row[1]\n",
    "            counts = row[2]\n",
    "            all_studies_TET2_counts[variant]=counts\n",
    "                \n",
    "        row_count=row_count+1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ASXL1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#Create lists of variants by studies and counts of each variant\n",
    "import csv\n",
    "filename=\"Data files/all_studies_trimmed_all_genes.csv\"\n",
    "\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    ASXL1_results={} #dictionary of DNMTA results (key = ID (row[13]), value = VAF (row[10]))\n",
    "\n",
    "    for row in read_reader:\n",
    "        if row_count>0:\n",
    "#             print(row)\n",
    "            if row[3] == 'ASXL1':\n",
    "#                 print(row)\n",
    "                VAF = row[0]\n",
    "                age = row[1]\n",
    "                variant = str('p.')+row[2]\n",
    "                study = row[4]\n",
    "                ASXL1_results[row_count]=(variant, study)\n",
    "                \n",
    "        row_count=row_count+1\n",
    "\n",
    "all_ASXL1_results_list = []\n",
    "\n",
    "for k, v in ASXL1_results.items():\n",
    "    all_ASXL1_results_list.append(v[0])\n",
    "        \n",
    "all_ASXL1_df = pd.DataFrame({'variant':all_ASXL1_results_list})\n",
    "all_grouped_df=all_ASXL1_df.groupby(['variant']).size().reset_index(name='counts')\n",
    "all_grouped_df.to_csv('Data files/all_ASXL1_variant_counts.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#All_ASXL1_counts\n",
    "filename=\"Data files/all_ASXL1_variant_counts.csv\"\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    all_studies_ASXL1_counts={} #dictionary of DNMTA results (key = ID (row[13]), value = VAF (row[10]))\n",
    "\n",
    "    for row in read_reader:\n",
    "#         print(row)\n",
    "        if row_count>0:\n",
    "            variant = row[1]\n",
    "            counts = row[2]\n",
    "            all_studies_ASXL1_counts[variant]=counts\n",
    "                \n",
    "        row_count=row_count+1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TP53"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#Create lists of variants by studies and counts of each variant\n",
    "import csv\n",
    "filename=\"Data files/all_studies_trimmed_all_genes.csv\"\n",
    "\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    TP53_results={} #dictionary of DNMTA results (key = ID (row[13]), value = VAF (row[10]))\n",
    "\n",
    "    for row in read_reader:\n",
    "        if row_count>0:\n",
    "#             print(row)\n",
    "            if row[3] == 'TP53':\n",
    "#                 print(row)\n",
    "                VAF = row[0]\n",
    "                age = row[1]\n",
    "                variant = str('p.')+row[2]\n",
    "                study = row[4]\n",
    "                TP53_results[row_count]=(variant, study)\n",
    "                \n",
    "        row_count=row_count+1\n",
    "\n",
    "all_TP53_results_list = []\n",
    "for k, v in TP53_results.items():\n",
    "    all_TP53_results_list.append(v[0])\n",
    "        \n",
    "all_TP53_df = pd.DataFrame({'variant':all_TP53_results_list})\n",
    "all_grouped_df=all_TP53_df.groupby(['variant']).size().reset_index(name='counts')\n",
    "all_grouped_df.to_csv('Data files/all_TP53_variant_counts.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#All_TP53_counts\n",
    "filename=\"Data files/all_TP53_variant_counts.csv\"\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    all_studies_TP53_counts={} #dictionary of DNMTA results (key = ID (row[13]), value = VAF (row[10]))\n",
    "\n",
    "    for row in read_reader:\n",
    "#         print(row)\n",
    "        if row_count>0:\n",
    "            variant = row[1]\n",
    "            counts = row[2]\n",
    "            all_studies_TP53_counts[variant]=counts\n",
    "                \n",
    "        row_count=row_count+1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functions for calculating the expected number of observations of a variant"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- The number of times you expect to see a variant depends on it's fitness effect (s), the size of the study (i.e. number of participants), the ages of the individuals in the study and the study's sequencing limit of detection.\n",
    "- The distribution of ages in the studys is presumed to be ~normally distributed, except for Acuna-Hidalgo, which has a ~uniform distribution of ages."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "def density_of_clones(f, s, t, N, u): \n",
    "    return ((2*N*u)/(f*(1-2*f)))*np.exp(-f/((1-2*f)*((np.exp(s*t)-1)/(2*N*s))))\n",
    "\n",
    "def prevalence_single_s(fmin, s, t, N, u): #expected prevalence of variant of a particular fitness effect\n",
    "    expected_number_clones=integrate.quad(lambda f: density_of_clones(f, s, t, N, u), fmin, 0.4999)[0]\n",
    "    return 1.0-np.exp(-expected_number_clones)\n",
    "\n",
    "def prevalence_averaged_over_ages(fmin, s, N, u, mean, sigma, study_size): #expected prevalence across range of ages\n",
    "    expected_prev_clones=integrate.quad(lambda t: (prevalence_single_s(fmin, s, t, N, u))*\\\n",
    "                                 ((1/((2*np.pi*(sigma)**2)**0.5))*(np.exp(-(t-mean)**2/(2*(sigma**2))))), mean-2*sigma, mean+2*sigma)\n",
    "    prev = expected_prev_clones[0]\n",
    "    expect_number_clones = prev*study_size\n",
    "    return expect_number_clones\n",
    "\n",
    "def prevalence_averaged_over_uniform_ages(fmin, s, N, u, mean, study_size): #expected prevalence across uniform distribution of ages\n",
    "    expected_prevalence_clones=integrate.quad(lambda t: (prevalence_single_s(fmin, s, t, N, u))*(1/49), 20, 69)\n",
    "    prevalence = expected_prevalence_clones[0]\n",
    "    expected_number_clones = prevalence*study_size\n",
    "    return expected_number_clones"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Maximum Likelihood Estimation for s"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- This involves calculating the most likely fitness effect (s) that would explain the number of observations of the variant."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " **Calculating the expected number of observations of each variant**\n",
    "- To calculate the expected number of observations of each variant, need to look at each variant in turn and for each variant, need to then look at each study in turn and see if that variant is targeted by that study.  If it is, need to work out how many times you expect to see that variant in that study, taking in to account the study's limit of detection, study size and ages of the individuals in the study. Then, once looked at all studies, sum up the number of observations across all the studies to get the overall expected number of observations for that variant.\n",
    "- The expected number also depends on the fitness effect of the variant - the value of s that means the observed number of variants matches as closely as possible the expected number of variants is calculated."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DNMT3A variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {
    "code_folding": [
     0,
     96,
     107,
     119
    ]
   },
   "outputs": [],
   "source": [
    "def expected_number_variants_missense_DNMT3A(data, params):\n",
    "    s_value = params[0]\n",
    "    \n",
    "    a = 0\n",
    "    expected_number_by_variant_missense = {}\n",
    "    \n",
    "    Jaiswal_fmin = 0.056\n",
    "    Jaiswal_mean = 58\n",
    "    Jaiswal_sigma = 11.5\n",
    "    Jaiswal_total = 17182\n",
    "    \n",
    "    Genovese_fmin = 0.1194\n",
    "    Genovese_mean = 55.1\n",
    "    Genovese_total = 12380\n",
    "    Genovese_sigma = 11.95\n",
    "    \n",
    "    McKerrel_fmin = 0.008\n",
    "    McKerrel_mean = 51.88\n",
    "    McKerrel_total = 4219\n",
    "    McKerrel_sigma = 18.36\n",
    "    \n",
    "    Zink_fmin = 0.2363\n",
    "    Zink_mean = 60\n",
    "    Zink_total = 11262\n",
    "    Zink_sigma = 1.18907720e+01\n",
    "    \n",
    "    Coombs_fmin = 0.019\n",
    "    Coombs_mean = 59.82\n",
    "    Coombs_total = 5649\n",
    "    Coombs_sigma = 15.14\n",
    "    \n",
    "    Young_fmin = 0.0009\n",
    "    Young_mean = 63\n",
    "    Young_total = 89\n",
    "    Young_sigma = 6.8\n",
    "    \n",
    "    Desai_fmin = 0.0169\n",
    "    Desai_mean = 66.1\n",
    "    Desai_total = 424\n",
    "    Desai_sigma = 6 #estimated\n",
    "    \n",
    "    Acuna_fmin = 0.0015\n",
    "    Acuna_mean = 4.5\n",
    "    Acuna_total = 2014\n",
    "    \n",
    "    N = 9.40166610e+04\n",
    "\n",
    "    for i in data:\n",
    "        variant_mutation_rate = mutation_rates_variants_DNMT3A(i)\n",
    "\n",
    "        if i in Jaiswal_missense_DNMT3A:\n",
    "            Jaiswal_number = prevalence_averaged_over_ages(Jaiswal_fmin, s_value, N, variant_mutation_rate, Jaiswal_mean, Jaiswal_sigma, Jaiswal_total)\n",
    "        else:\n",
    "            Jaiswal_number = 0\n",
    "\n",
    "        if i in Genovese_missense_DNMT3A:\n",
    "            Genovese_number = prevalence_averaged_over_ages(Genovese_fmin, s_value, N, variant_mutation_rate, Genovese_mean, Genovese_sigma, Genovese_total)\n",
    "        else:\n",
    "            Genovese_number = 0 \n",
    "\n",
    "        if i in Zink_missense_DNMT3A:\n",
    "            Zink_number = prevalence_averaged_over_ages(Zink_fmin, s_value, N, variant_mutation_rate, Zink_mean, Zink_sigma, Zink_total)\n",
    "        else:\n",
    "            Zink_number = 0\n",
    "            \n",
    "        if i in McKerrel_missense_DNMT3A:\n",
    "            McKerrel_number = prevalence_averaged_over_ages(McKerrel_fmin, s_value, N, variant_mutation_rate, McKerrel_mean, McKerrel_sigma, McKerrel_total)\n",
    "        else:\n",
    "            McKerrel_number = 0\n",
    "\n",
    "        if i in Coombs_missense_DNMT3A:\n",
    "            Coombs_number = prevalence_averaged_over_ages(Coombs_fmin, s_value, N, variant_mutation_rate, Coombs_mean, Coombs_sigma, Coombs_total)\n",
    "        else:\n",
    "            Coombs_number = 0\n",
    "\n",
    "        if i in Young_missense_DNMT3A:\n",
    "            Young_number = prevalence_averaged_over_ages(Young_fmin, s_value, N, variant_mutation_rate, Young_mean, Young_sigma, Young_total)\n",
    "        else:\n",
    "            Young_number = 0\n",
    "\n",
    "        if i in Desai_missense_DNMT3A:\n",
    "            Desai_number = prevalence_averaged_over_ages(Desai_fmin, s_value, N, variant_mutation_rate, Desai_mean, Desai_sigma, Desai_total)\n",
    "        else:\n",
    "            Desai_number = 0\n",
    "\n",
    "        if i in Acuna_missense_DNMT3A:\n",
    "            Acuna_number = prevalence_averaged_over_uniform_ages(Acuna_fmin, s_value, N, variant_mutation_rate, Acuna_mean, Acuna_total)\n",
    "        else:\n",
    "            Acuna_number = 0\n",
    "\n",
    "        expected_total = Jaiswal_number + Genovese_number + Zink_number + McKerrel_number + Coombs_number + Young_number + Desai_number + Acuna_number\n",
    "\n",
    "        expected_number_by_variant_missense[i] = expected_total\n",
    "    \n",
    "    return expected_number_by_variant_missense\n",
    "\n",
    "def prob_observed_given_expected_missense_DNMT3A(params, data): #data = list of variant (e.g. p.R882H) (or all_missense_variants_DNMT3A)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_DNMT3A_counts[variant])\n",
    "        expected_number = float(expected_number_variants_missense_DNMT3A(data, params)[variant])\n",
    "        square_distance = ((observed_number - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   \n",
    "\n",
    "def prob_observed_given_expected_missense_DNMT3A_upper_error(params, data): #data = list of variant (e.g. p.R882H) (or all_missense_variants_DNMT3A)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_DNMT3A_counts[variant])\n",
    "        observed_number_upper_error = observed_number+np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_missense_DNMT3A(data, params)[variant])\n",
    "        square_distance = ((observed_number_upper_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   \n",
    "\n",
    "def prob_observed_given_expected_missense_DNMT3A_lower_error(params, data): #data = list of variant (e.g. p.R882H) (or all_missense_variants_DNMT3A)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_DNMT3A_counts[variant])\n",
    "        observed_number_lower_error = observed_number-np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_missense_DNMT3A(data, params)[variant])\n",
    "        square_distance = ((observed_number_lower_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {
    "code_folding": [
     0,
     91,
     102,
     114
    ]
   },
   "outputs": [],
   "source": [
    "def expected_number_variants_nonsense_DNMT3A(data, params):\n",
    "    s_value = params[0]\n",
    "    \n",
    "    a = 0\n",
    "    expected_number_by_variant_nonsense = {}\n",
    "    \n",
    "    Jaiswal_fmin = 0.056\n",
    "    Jaiswal_mean = 58\n",
    "    Jaiswal_sigma = 11.5\n",
    "    Jaiswal_total = 17182\n",
    "    \n",
    "    Genovese_fmin = 0.1194\n",
    "    Genovese_mean = 55.1\n",
    "    Genovese_total = 12380\n",
    "    Genovese_sigma = 11.95\n",
    "    \n",
    "    McKerrel_fmin = 0.008\n",
    "    McKerrel_mean = 51.88\n",
    "    McKerrel_total = 4219\n",
    "    McKerrel_sigma = 18.36\n",
    "    \n",
    "    Zink_fmin = 0.2363\n",
    "    Zink_mean = 60\n",
    "    Zink_total = 11262\n",
    "    Zink_sigma = 1.18907720e+01\n",
    "    \n",
    "    Coombs_fmin = 0.019\n",
    "    Coombs_mean = 59.82\n",
    "    Coombs_total = 5649\n",
    "    Coombs_sigma = 15.14\n",
    "    \n",
    "    Young_fmin = 0.0009\n",
    "    Young_mean = 63\n",
    "    Young_total = 89\n",
    "    Young_sigma = 6.8\n",
    "    \n",
    "    Desai_fmin = 0.0169\n",
    "    Desai_mean = 66.1\n",
    "    Desai_total = 424\n",
    "    Desai_sigma = 6 #estimated\n",
    "    \n",
    "    Acuna_fmin = 0.0015\n",
    "    Acuna_mean = 4.5\n",
    "    Acuna_total = 2014\n",
    "    \n",
    "    N = 9.40166610e+04\n",
    "\n",
    "    for i in data:\n",
    "        variant_mutation_rate = mutation_rates_variants_DNMT3A(i)\n",
    "\n",
    "        if i in Jaiswal_nonsense_DNMT3A:\n",
    "            Jaiswal_number = prevalence_averaged_over_ages(Jaiswal_fmin, s_value, N, variant_mutation_rate, Jaiswal_mean, Jaiswal_sigma, Jaiswal_total)\n",
    "        else:\n",
    "            Jaiswal_number = 0\n",
    "\n",
    "        if i in Genovese_nonsense_DNMT3A:\n",
    "            Genovese_number = prevalence_averaged_over_ages(Genovese_fmin, s_value, N, variant_mutation_rate, Genovese_mean, Genovese_sigma, Genovese_total)\n",
    "        else:\n",
    "            Genovese_number = 0 \n",
    "\n",
    "        if i in Zink_nonsense_DNMT3A:\n",
    "            Zink_number = prevalence_averaged_over_ages(Zink_fmin, s_value, N, variant_mutation_rate, Zink_mean, Zink_sigma, Zink_total)\n",
    "        else:\n",
    "            Zink_number = 0\n",
    "\n",
    "        if i in Coombs_nonsense_DNMT3A:\n",
    "            Coombs_number = prevalence_averaged_over_ages(Coombs_fmin, s_value, N, variant_mutation_rate, Coombs_mean, Coombs_sigma, Coombs_total)\n",
    "        else:\n",
    "            Coombs_number = 0\n",
    "\n",
    "        if i in Young_nonsense_DNMT3A:\n",
    "            Young_number = prevalence_averaged_over_ages(Young_fmin, s_value, N, variant_mutation_rate, Young_mean, Young_sigma, Young_total)\n",
    "        else:\n",
    "            Young_number = 0\n",
    "\n",
    "        if i in Desai_nonsense_DNMT3A:\n",
    "            Desai_number = prevalence_averaged_over_ages(Desai_fmin, s_value, N, variant_mutation_rate, Desai_mean, Desai_sigma, Desai_total)\n",
    "        else:\n",
    "            Desai_number = 0\n",
    "\n",
    "        if i in Acuna_nonsense_DNMT3A:\n",
    "            Acuna_number = prevalence_averaged_over_uniform_ages(Acuna_fmin, s_value, N, variant_mutation_rate, Acuna_mean, Acuna_total)\n",
    "        else:\n",
    "            Acuna_number = 0\n",
    "\n",
    "        expected_total = Jaiswal_number + Genovese_number + Zink_number + Coombs_number + Young_number + Desai_number + Acuna_number\n",
    "\n",
    "        expected_number_by_variant_nonsense[i] = expected_total\n",
    "    \n",
    "    return expected_number_by_variant_nonsense\n",
    "\n",
    "def prob_observed_given_expected_nonsense_DNMT3A(params, data): #data = list of variant (e.g. p.R882H) (or all_nonsense_variants_DNMT3A)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_DNMT3A_counts[variant])\n",
    "        expected_number = float(expected_number_variants_nonsense_DNMT3A(data, params)[variant])\n",
    "        square_distance = ((observed_number - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance\n",
    "\n",
    "def prob_observed_given_expected_nonsense_DNMT3A_upper_error(params, data): #data = list of variant (e.g. p.R882H) (or all_nonsense_variants_DNMT3A)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_DNMT3A_counts[variant])\n",
    "        observed_number_upper_error = observed_number+np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_nonsense_DNMT3A(data, params)[variant])\n",
    "        square_distance = ((observed_number_upper_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   \n",
    "\n",
    "def prob_observed_given_expected_nonsense_DNMT3A_lower_error(params, data): #data = list of variant (e.g. p.R882H) (or all_nonsense_variants_DNMT3A)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_DNMT3A_counts[variant])\n",
    "        observed_number_lower_error = observed_number-np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_nonsense_DNMT3A(data, params)[variant])\n",
    "        square_distance = ((observed_number_lower_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#CALCULATING S FOR MISSENSE VARIANTS\n",
    "DNMT3A_missense_s = {}\n",
    "a = 0\n",
    "for var in all_missense_variants_DNMT3A: #iterate over all the possible missense variants in DNMT3A\n",
    "    initial_guess=[0.05] #initial guess for s = 0.05 (i.e. 5%)\n",
    "    if var in all_studies_DNMT3A_counts.keys(): #look to see if the variant is seen across any of the studies\n",
    "        observed_number = float(all_studies_DNMT3A_counts[var]) #if it is, retrieve how many times it was observed\n",
    "#         print(observed_number)\n",
    "        if observed_number >1: #only calculate s for variants seen more than 1 time\n",
    "            variant_mutation_rate = mutation_rates_variants_DNMT3A(var) #trinucleotide-context dependent mutation rate for the variant\n",
    "            \n",
    "            MLE = scipy.optimize.minimize(prob_observed_given_expected_missense_DNMT3A, initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value = MLE.x[0]\n",
    "\n",
    "            upper_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_missense_DNMT3A_upper_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_upper = upper_error_MLE.x[0]\n",
    "\n",
    "            lower_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_missense_DNMT3A_lower_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_lower = lower_error_MLE.x[0]\n",
    "\n",
    "#             print('a = ', a)\n",
    "            #Create a dictionary where key = variant\n",
    "            DNMT3A_missense_s[var]=(s_value, observed_number, variant_mutation_rate, s_value_upper, s_value_lower) \n",
    "    a = a+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#CALCULATING S FOR NONSENSE VARIANTS\n",
    "DNMT3A_nonsense_s = {}\n",
    "a = 0\n",
    "for var in all_nonsense_variants_DNMT3A: #iterate over all the possible nonsense variants in DNMT3A\n",
    "    initial_guess=[0.05] #initial guess for s = 0.05 (i.e. 5%)\n",
    "    if var in all_studies_DNMT3A_counts.keys(): #look to see if the variant is seen across any of the studies\n",
    "        observed_number = float(all_studies_DNMT3A_counts[var]) #if it is, retrieve how many times it was observed\n",
    "#         print(observed_number)\n",
    "        if observed_number >1: #only calculate s for variants seen more than 1 time\n",
    "            variant_mutation_rate = mutation_rates_variants_DNMT3A(var) #trinucleotide-context dependent mutation rate for the variant\n",
    "            \n",
    "            MLE = scipy.optimize.minimize(prob_observed_given_expected_nonsense_DNMT3A, initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value = MLE.x[0]\n",
    "\n",
    "            upper_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_nonsense_DNMT3A_upper_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_upper = upper_error_MLE.x[0]\n",
    "\n",
    "            lower_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_nonsense_DNMT3A_lower_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_lower = lower_error_MLE.x[0]\n",
    "\n",
    "#             print('a = ', a)\n",
    "            DNMT3A_nonsense_s[var]=(s_value, observed_number, variant_mutation_rate, s_value_upper, s_value_lower) \n",
    "    a = a+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#Exac - retrieve the frequencies in Exac for each of the variants\n",
    "import csv\n",
    "filename=\"Data files/DNMT3A_Exac.csv\"\n",
    "\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    Exac_DNMT3A_results={} \n",
    "    \n",
    "    for row in read_reader:\n",
    "        if row_count>1:\n",
    "#             print(row)\n",
    "            variant = row[6]\n",
    "            allele_freq = float(row[14])\n",
    "            african_freq = float(row[15])/float(row[16])\n",
    "            east_asian_freq = float(row[18])/float(row[19])\n",
    "            european_freq = float(row[21])/float(row[22])\n",
    "            finnish_freq = float(row[24])/float(row[25])\n",
    "            latino_freq = float(row[27])/float(row[28])\n",
    "            south_asian_freq = float(row[33])/float(row[34])\n",
    "            Exac_DNMT3A_results[variant]=(allele_freq, african_freq, east_asian_freq, european_freq, finnish_freq, latino_freq, south_asian_freq)\n",
    "\n",
    "        row_count=row_count+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#COSMIC - retrieve the frequencies in COSMIC for each of the variants\n",
    "import csv\n",
    "filename=\"Data files/DNMT3A_COSMIC_variant_counts.csv\"\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    DNMT3A_COSMIC_results={} #dictionary of DNMTA results (key = ID (row[13]), value = VAF (row[10]))\n",
    "\n",
    "    for row in read_reader:\n",
    "        if row_count>1:\n",
    "            variant = row[1]\n",
    "            counts = int(row[2])\n",
    "            DNMT3A_COSMIC_results[variant]=counts\n",
    "\n",
    "        row_count=row_count+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Create dictionary of results and save to .csv file\n",
    "DNMT3A_s_dictionary={}\n",
    "for k, v in DNMT3A_missense_s.items():\n",
    "#     print(k)\n",
    "    if k in Exac_DNMT3A_results.keys():\n",
    "        exac_freq = round((Exac_DNMT3A_results[k][0])*100, 4)\n",
    "        african_freq = round((Exac_DNMT3A_results[k][1])*100, 4)\n",
    "        east_asian_freq = round((Exac_DNMT3A_results[k][2])*100, 4)\n",
    "        european_freq = round((Exac_DNMT3A_results[k][3])*100, 4)\n",
    "        finnish_freq = round((Exac_DNMT3A_results[k][4])*100, 4)\n",
    "        latino_freq = round((Exac_DNMT3A_results[k][5])*100, 4)\n",
    "        south_asian_freq = round((Exac_DNMT3A_results[k][6])*100, 4)\n",
    "    if k not in Exac_DNMT3A_results.keys():\n",
    "        exac_freq = '-'\n",
    "        african_freq = '-'\n",
    "        east_asian_freq = '-'\n",
    "        european_freq = '-'\n",
    "        finnish_freq = '-'\n",
    "        latino_freq = '-'\n",
    "        south_asian_freq = '-'\n",
    "    if k in DNMT3A_COSMIC_results.keys():\n",
    "        COSMIC_number = DNMT3A_COSMIC_results[k]\n",
    "    if k not in DNMT3A_COSMIC_results.keys():\n",
    "        COSMIC_number = '-'\n",
    "    DNMT3A_s_dictionary[k]=(round(v[0]*100, 2),round(v[4]*100, 2),round(v[3]*100, 2), v[1], v[2], COSMIC_number, exac_freq, \\\n",
    "                            african_freq, east_asian_freq, european_freq, finnish_freq, latino_freq, south_asian_freq)\n",
    "\n",
    "for k, v in DNMT3A_nonsense_s.items():\n",
    "    if k in Exac_DNMT3A_results.keys():\n",
    "        exac_freq = round((Exac_DNMT3A_results[k][0])*100, 4)\n",
    "        african_freq = round((Exac_DNMT3A_results[k][1])*100, 4)\n",
    "        east_asian_freq = round((Exac_DNMT3A_results[k][2])*100, 4)\n",
    "        european_freq = round((Exac_DNMT3A_results[k][3])*100, 4)\n",
    "        finnish_freq = round((Exac_DNMT3A_results[k][4])*100, 4)\n",
    "        latino_freq = round((Exac_DNMT3A_results[k][5])*100, 4)\n",
    "        south_asian_freq = round((Exac_DNMT3A_results[k][6])*100, 4)\n",
    "    if k not in Exac_DNMT3A_results.keys():\n",
    "        exac_freq = '-'\n",
    "        african_freq = '-'\n",
    "        east_asian_freq = '-'\n",
    "        european_freq = '-'\n",
    "        finnish_freq = '-'\n",
    "        latino_freq = '-'\n",
    "        south_asian_freq = '-'\n",
    "    if k in DNMT3A_COSMIC_results.keys():\n",
    "        COSMIC_number = DNMT3A_COSMIC_results[k]\n",
    "    if k not in DNMT3A_COSMIC_results.keys():\n",
    "        COSMIC_number = '-'\n",
    "    DNMT3A_s_dictionary[k]=(round(v[0]*100, 2),round(v[4]*100, 2),round(v[3]*100, 2), v[1], v[2], COSMIC_number, exac_freq, \\\n",
    "                            african_freq, east_asian_freq, european_freq, finnish_freq, latino_freq, south_asian_freq)\n",
    "\n",
    "DNMT3A_df = pd.DataFrame.from_dict(DNMT3A_s_dictionary, orient='index', columns=['s (%)', 's lower value (%)', 's upper value (%)','observed number', 'site-specific mutation rate', 'Number of times in COSMIC',\\\n",
    "                                                                                'Exac Overall freq (%)', 'Exac African freq (%)', 'Exac East Asian freq (%)',\\\n",
    "                                                                                'Exac European freq (%)', 'Exac Finnish freq (%)', 'Exac Latino freq (%)', 'Exac South Asian freq (%)'])\n",
    "DNMT3A_df_sorted = DNMT3A_df.sort_values(by = ['s (%)'], ascending = False)\n",
    "\n",
    "DNMT3A_df_sorted.to_csv('Data files/DNMT3A_s_across_whole_gene.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s (%)</th>\n",
       "      <th>s lower value (%)</th>\n",
       "      <th>s upper value (%)</th>\n",
       "      <th>observed number</th>\n",
       "      <th>site-specific mutation rate</th>\n",
       "      <th>Number of times in COSMIC</th>\n",
       "      <th>Exac Overall freq (%)</th>\n",
       "      <th>Exac African freq (%)</th>\n",
       "      <th>Exac East Asian freq (%)</th>\n",
       "      <th>Exac European freq (%)</th>\n",
       "      <th>Exac Finnish freq (%)</th>\n",
       "      <th>Exac Latino freq (%)</th>\n",
       "      <th>Exac South Asian freq (%)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>p.E30A</th>\n",
       "      <td>28.05</td>\n",
       "      <td>20.70</td>\n",
       "      <td>196.75</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.863150e-10</td>\n",
       "      <td>3</td>\n",
       "      <td>0.2959</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0</td>\n",
       "      <td>0.3033</td>\n",
       "      <td>2.1424</td>\n",
       "      <td>0.1413</td>\n",
       "      <td>0.0247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.F751C</th>\n",
       "      <td>21.62</td>\n",
       "      <td>14.91</td>\n",
       "      <td>27.48</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.574550e-10</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.Y735C</th>\n",
       "      <td>19.93</td>\n",
       "      <td>18.54</td>\n",
       "      <td>21.24</td>\n",
       "      <td>23.0</td>\n",
       "      <td>8.813490e-10</td>\n",
       "      <td>12</td>\n",
       "      <td>0.0118</td>\n",
       "      <td>0.0097</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0182</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0087</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.F732C</th>\n",
       "      <td>18.41</td>\n",
       "      <td>13.86</td>\n",
       "      <td>21.62</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.857950e-10</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.S714C</th>\n",
       "      <td>17.63</td>\n",
       "      <td>15.68</td>\n",
       "      <td>19.31</td>\n",
       "      <td>7.0</td>\n",
       "      <td>4.368770e-10</td>\n",
       "      <td>25</td>\n",
       "      <td>0.0041</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0075</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.F414C</th>\n",
       "      <td>17.23</td>\n",
       "      <td>12.95</td>\n",
       "      <td>20.37</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.574550e-10</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.L737R</th>\n",
       "      <td>15.58</td>\n",
       "      <td>11.98</td>\n",
       "      <td>17.91</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.863150e-10</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.P507R</th>\n",
       "      <td>15.56</td>\n",
       "      <td>11.66</td>\n",
       "      <td>18.09</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.559440e-10</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.W795S</th>\n",
       "      <td>15.48</td>\n",
       "      <td>11.61</td>\n",
       "      <td>18.00</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.643030e-10</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.W860R</th>\n",
       "      <td>15.45</td>\n",
       "      <td>14.46</td>\n",
       "      <td>16.34</td>\n",
       "      <td>18.0</td>\n",
       "      <td>1.990571e-09</td>\n",
       "      <td>7</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.Y908C</th>\n",
       "      <td>15.04</td>\n",
       "      <td>12.16</td>\n",
       "      <td>17.53</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.472060e-09</td>\n",
       "      <td>-</td>\n",
       "      <td>0.0033</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0045</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0086</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.F755S</th>\n",
       "      <td>14.96</td>\n",
       "      <td>13.02</td>\n",
       "      <td>16.41</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.958240e-10</td>\n",
       "      <td>3</td>\n",
       "      <td>0.0041</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0086</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R882P</th>\n",
       "      <td>14.92</td>\n",
       "      <td>13.09</td>\n",
       "      <td>16.40</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.296770e-10</td>\n",
       "      <td>36</td>\n",
       "      <td>0.0058</td>\n",
       "      <td>0.029</td>\n",
       "      <td>0.0116</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0087</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.N838D</th>\n",
       "      <td>14.59</td>\n",
       "      <td>11.38</td>\n",
       "      <td>16.45</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.958240e-10</td>\n",
       "      <td>5</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.E599D</th>\n",
       "      <td>13.93</td>\n",
       "      <td>11.89</td>\n",
       "      <td>15.41</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.419513e-09</td>\n",
       "      <td>-</td>\n",
       "      <td>0.0034</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0061</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.L547H</th>\n",
       "      <td>13.72</td>\n",
       "      <td>11.03</td>\n",
       "      <td>15.34</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.783530e-10</td>\n",
       "      <td>5</td>\n",
       "      <td>0.0017</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0087</td>\n",
       "      <td>0.0061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.E733G</th>\n",
       "      <td>13.62</td>\n",
       "      <td>11.10</td>\n",
       "      <td>15.17</td>\n",
       "      <td>3.0</td>\n",
       "      <td>8.126630e-10</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.G543A</th>\n",
       "      <td>13.54</td>\n",
       "      <td>10.91</td>\n",
       "      <td>15.11</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.048440e-10</td>\n",
       "      <td>5</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R729G</th>\n",
       "      <td>13.51</td>\n",
       "      <td>11.48</td>\n",
       "      <td>14.92</td>\n",
       "      <td>3.0</td>\n",
       "      <td>6.552200e-10</td>\n",
       "      <td>-</td>\n",
       "      <td>0.0009</td>\n",
       "      <td>0.0097</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.S770W</th>\n",
       "      <td>13.28</td>\n",
       "      <td>11.19</td>\n",
       "      <td>14.66</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.723300e-10</td>\n",
       "      <td>5</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.P307R</th>\n",
       "      <td>13.19</td>\n",
       "      <td>10.66</td>\n",
       "      <td>14.66</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.643030e-10</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R882H</th>\n",
       "      <td>13.07</td>\n",
       "      <td>12.75</td>\n",
       "      <td>13.36</td>\n",
       "      <td>105.0</td>\n",
       "      <td>1.882290e-08</td>\n",
       "      <td>1069</td>\n",
       "      <td>0.0545</td>\n",
       "      <td>0.0869</td>\n",
       "      <td>0.0694</td>\n",
       "      <td>0.0571</td>\n",
       "      <td>0.0758</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.F732S</th>\n",
       "      <td>13.02</td>\n",
       "      <td>10.55</td>\n",
       "      <td>14.47</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.958240e-10</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.E756D</th>\n",
       "      <td>12.96</td>\n",
       "      <td>9.97</td>\n",
       "      <td>14.73</td>\n",
       "      <td>2.0</td>\n",
       "      <td>9.655700e-10</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.I780T</th>\n",
       "      <td>12.91</td>\n",
       "      <td>10.88</td>\n",
       "      <td>14.18</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.346250e-09</td>\n",
       "      <td>7</td>\n",
       "      <td>0.0051</td>\n",
       "      <td>0.0201</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0061</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.G543C</th>\n",
       "      <td>12.81</td>\n",
       "      <td>11.04</td>\n",
       "      <td>13.98</td>\n",
       "      <td>3.0</td>\n",
       "      <td>8.141860e-10</td>\n",
       "      <td>25</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.M880V</th>\n",
       "      <td>12.79</td>\n",
       "      <td>11.45</td>\n",
       "      <td>13.81</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.486370e-09</td>\n",
       "      <td>7</td>\n",
       "      <td>0.0017</td>\n",
       "      <td>0.0097</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.W753C</th>\n",
       "      <td>12.73</td>\n",
       "      <td>9.81</td>\n",
       "      <td>14.44</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.046296e-09</td>\n",
       "      <td>-</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.E733*</th>\n",
       "      <td>12.38</td>\n",
       "      <td>9.73</td>\n",
       "      <td>13.82</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.399810e-10</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.I705T</th>\n",
       "      <td>12.32</td>\n",
       "      <td>10.95</td>\n",
       "      <td>13.27</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.346250e-09</td>\n",
       "      <td>5</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.D768E</th>\n",
       "      <td>9.49</td>\n",
       "      <td>7.03</td>\n",
       "      <td>10.86</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.420300e-09</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R598*</th>\n",
       "      <td>9.45</td>\n",
       "      <td>8.94</td>\n",
       "      <td>9.84</td>\n",
       "      <td>11.0</td>\n",
       "      <td>1.882290e-08</td>\n",
       "      <td>9</td>\n",
       "      <td>0.0059</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0061</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R326C</th>\n",
       "      <td>9.42</td>\n",
       "      <td>8.81</td>\n",
       "      <td>9.89</td>\n",
       "      <td>13.0</td>\n",
       "      <td>1.415120e-08</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0058</td>\n",
       "      <td>0.0096</td>\n",
       "      <td>0.0116</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.0151</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R736C</th>\n",
       "      <td>9.42</td>\n",
       "      <td>8.81</td>\n",
       "      <td>9.89</td>\n",
       "      <td>13.0</td>\n",
       "      <td>1.415120e-08</td>\n",
       "      <td>21</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.W893*</th>\n",
       "      <td>9.41</td>\n",
       "      <td>7.80</td>\n",
       "      <td>10.22</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.524850e-09</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0016</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.W581*</th>\n",
       "      <td>9.41</td>\n",
       "      <td>7.80</td>\n",
       "      <td>10.22</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.524850e-09</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0027</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0016</td>\n",
       "      <td>0.0177</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0063</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.W601*</th>\n",
       "      <td>9.37</td>\n",
       "      <td>8.22</td>\n",
       "      <td>10.06</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.410250e-09</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.G890D</th>\n",
       "      <td>9.32</td>\n",
       "      <td>7.75</td>\n",
       "      <td>10.23</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.445670e-09</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R771*</th>\n",
       "      <td>9.27</td>\n",
       "      <td>8.74</td>\n",
       "      <td>9.70</td>\n",
       "      <td>16.0</td>\n",
       "      <td>1.882290e-08</td>\n",
       "      <td>12</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R749C</th>\n",
       "      <td>9.22</td>\n",
       "      <td>8.51</td>\n",
       "      <td>9.72</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.200440e-08</td>\n",
       "      <td>10</td>\n",
       "      <td>0.0025</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0045</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.L547F</th>\n",
       "      <td>9.19</td>\n",
       "      <td>7.63</td>\n",
       "      <td>9.98</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.087170e-09</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.W440*</th>\n",
       "      <td>9.02</td>\n",
       "      <td>7.49</td>\n",
       "      <td>9.80</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.610690e-09</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R635W</th>\n",
       "      <td>8.98</td>\n",
       "      <td>8.09</td>\n",
       "      <td>9.64</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.415120e-08</td>\n",
       "      <td>18</td>\n",
       "      <td>0.0095</td>\n",
       "      <td>0.0353</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.W313*</th>\n",
       "      <td>8.88</td>\n",
       "      <td>7.38</td>\n",
       "      <td>9.63</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.140310e-09</td>\n",
       "      <td>3</td>\n",
       "      <td>0.0049</td>\n",
       "      <td>0.0096</td>\n",
       "      <td>0.0116</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.W306*</th>\n",
       "      <td>8.80</td>\n",
       "      <td>7.31</td>\n",
       "      <td>9.56</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.410250e-09</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.G550R</th>\n",
       "      <td>8.69</td>\n",
       "      <td>7.95</td>\n",
       "      <td>9.19</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.480642e-08</td>\n",
       "      <td>8</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.P385L</th>\n",
       "      <td>8.21</td>\n",
       "      <td>7.13</td>\n",
       "      <td>8.90</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.415120e-08</td>\n",
       "      <td>-</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0092</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.S770L</th>\n",
       "      <td>8.04</td>\n",
       "      <td>6.95</td>\n",
       "      <td>8.79</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.200440e-08</td>\n",
       "      <td>6</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.0096</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0086</td>\n",
       "      <td>0.0182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.W327*</th>\n",
       "      <td>8.04</td>\n",
       "      <td>5.93</td>\n",
       "      <td>9.01</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.610690e-09</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.W860*</th>\n",
       "      <td>7.85</td>\n",
       "      <td>5.75</td>\n",
       "      <td>8.82</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.140310e-09</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0.0096</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.D702N</th>\n",
       "      <td>7.80</td>\n",
       "      <td>6.46</td>\n",
       "      <td>8.46</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.200440e-08</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R326H</th>\n",
       "      <td>7.77</td>\n",
       "      <td>6.95</td>\n",
       "      <td>8.34</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.882290e-08</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0025</td>\n",
       "      <td>0.0096</td>\n",
       "      <td>0</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R899C</th>\n",
       "      <td>7.67</td>\n",
       "      <td>6.30</td>\n",
       "      <td>8.39</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.415120e-08</td>\n",
       "      <td>-</td>\n",
       "      <td>0.0016</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.G685R</th>\n",
       "      <td>7.56</td>\n",
       "      <td>6.24</td>\n",
       "      <td>8.20</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.480642e-08</td>\n",
       "      <td>4</td>\n",
       "      <td>0.0017</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R899H</th>\n",
       "      <td>7.33</td>\n",
       "      <td>5.99</td>\n",
       "      <td>7.99</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.882290e-08</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R688H</th>\n",
       "      <td>7.33</td>\n",
       "      <td>5.99</td>\n",
       "      <td>8.00</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.882290e-08</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0033</td>\n",
       "      <td>0.0096</td>\n",
       "      <td>0</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R659H</th>\n",
       "      <td>7.30</td>\n",
       "      <td>5.99</td>\n",
       "      <td>7.91</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.882290e-08</td>\n",
       "      <td>-</td>\n",
       "      <td>0.0017</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R366H</th>\n",
       "      <td>7.30</td>\n",
       "      <td>5.99</td>\n",
       "      <td>7.91</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.882290e-08</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.G332R</th>\n",
       "      <td>7.23</td>\n",
       "      <td>6.03</td>\n",
       "      <td>7.98</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.480642e-08</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0041</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0075</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R771Q</th>\n",
       "      <td>6.39</td>\n",
       "      <td>4.38</td>\n",
       "      <td>7.30</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.200440e-08</td>\n",
       "      <td>4</td>\n",
       "      <td>0.0025</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>102 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         s (%)  s lower value (%)  s upper value (%)  observed number  \\\n",
       "p.E30A   28.05              20.70             196.75              4.0   \n",
       "p.F751C  21.62              14.91              27.48              2.0   \n",
       "p.Y735C  19.93              18.54              21.24             23.0   \n",
       "p.F732C  18.41              13.86              21.62              2.0   \n",
       "p.S714C  17.63              15.68              19.31              7.0   \n",
       "p.F414C  17.23              12.95              20.37              2.0   \n",
       "p.L737R  15.58              11.98              17.91              2.0   \n",
       "p.P507R  15.56              11.66              18.09              2.0   \n",
       "p.W795S  15.48              11.61              18.00              2.0   \n",
       "p.W860R  15.45              14.46              16.34             18.0   \n",
       "p.Y908C  15.04              12.16              17.53              4.0   \n",
       "p.F755S  14.96              13.02              16.41              4.0   \n",
       "p.R882P  14.92              13.09              16.40              6.0   \n",
       "p.N838D  14.59              11.38              16.45              2.0   \n",
       "p.E599D  13.93              11.89              15.41              4.0   \n",
       "p.L547H  13.72              11.03              15.34              2.0   \n",
       "p.E733G  13.62              11.10              15.17              3.0   \n",
       "p.G543A  13.54              10.91              15.11              2.0   \n",
       "p.R729G  13.51              11.48              14.92              3.0   \n",
       "p.S770W  13.28              11.19              14.66              3.0   \n",
       "p.P307R  13.19              10.66              14.66              2.0   \n",
       "p.R882H  13.07              12.75              13.36            105.0   \n",
       "p.F732S  13.02              10.55              14.47              2.0   \n",
       "p.E756D  12.96               9.97              14.73              2.0   \n",
       "p.I780T  12.91              10.88              14.18              3.0   \n",
       "p.G543C  12.81              11.04              13.98              3.0   \n",
       "p.M880V  12.79              11.45              13.81              6.0   \n",
       "p.W753C  12.73               9.81              14.44              2.0   \n",
       "p.E733*  12.38               9.73              13.82              2.0   \n",
       "p.I705T  12.32              10.95              13.27              4.0   \n",
       "...        ...                ...                ...              ...   \n",
       "p.D768E   9.49               7.03              10.86              2.0   \n",
       "p.R598*   9.45               8.94               9.84             11.0   \n",
       "p.R326C   9.42               8.81               9.89             13.0   \n",
       "p.R736C   9.42               8.81               9.89             13.0   \n",
       "p.W893*   9.41               7.80              10.22              2.0   \n",
       "p.W581*   9.41               7.80              10.22              2.0   \n",
       "p.W601*   9.37               8.22              10.06              3.0   \n",
       "p.G890D   9.32               7.75              10.23              3.0   \n",
       "p.R771*   9.27               8.74               9.70             16.0   \n",
       "p.R749C   9.22               8.51               9.72              6.0   \n",
       "p.L547F   9.19               7.63               9.98              2.0   \n",
       "p.W440*   9.02               7.49               9.80              2.0   \n",
       "p.R635W   8.98               8.09               9.64              5.0   \n",
       "p.W313*   8.88               7.38               9.63              2.0   \n",
       "p.W306*   8.80               7.31               9.56              2.0   \n",
       "p.G550R   8.69               7.95               9.19              5.0   \n",
       "p.P385L   8.21               7.13               8.90              3.0   \n",
       "p.S770L   8.04               6.95               8.79              5.0   \n",
       "p.W327*   8.04               5.93               9.01              2.0   \n",
       "p.W860*   7.85               5.75               8.82              2.0   \n",
       "p.D702N   7.80               6.46               8.46              2.0   \n",
       "p.R326H   7.77               6.95               8.34              7.0   \n",
       "p.R899C   7.67               6.30               8.39              2.0   \n",
       "p.G685R   7.56               6.24               8.20              2.0   \n",
       "p.R899H   7.33               5.99               7.99              2.0   \n",
       "p.R688H   7.33               5.99               8.00              2.0   \n",
       "p.R659H   7.30               5.99               7.91              2.0   \n",
       "p.R366H   7.30               5.99               7.91              2.0   \n",
       "p.G332R   7.23               6.03               7.98              4.0   \n",
       "p.R771Q   6.39               4.38               7.30              2.0   \n",
       "\n",
       "         site-specific mutation rate Number of times in COSMIC  \\\n",
       "p.E30A                  1.863150e-10                         3   \n",
       "p.F751C                 1.574550e-10                         -   \n",
       "p.Y735C                 8.813490e-10                        12   \n",
       "p.F732C                 1.857950e-10                         1   \n",
       "p.S714C                 4.368770e-10                        25   \n",
       "p.F414C                 1.574550e-10                         -   \n",
       "p.L737R                 1.863150e-10                         1   \n",
       "p.P507R                 4.559440e-10                         -   \n",
       "p.W795S                 4.643030e-10                         -   \n",
       "p.W860R                 1.990571e-09                         7   \n",
       "p.Y908C                 1.472060e-09                         -   \n",
       "p.F755S                 4.958240e-10                         3   \n",
       "p.R882P                 6.296770e-10                        36   \n",
       "p.N838D                 4.958240e-10                         5   \n",
       "p.E599D                 1.419513e-09                         -   \n",
       "p.L547H                 3.783530e-10                         5   \n",
       "p.E733G                 8.126630e-10                         1   \n",
       "p.G543A                 4.048440e-10                         5   \n",
       "p.R729G                 6.552200e-10                         -   \n",
       "p.S770W                 5.723300e-10                         5   \n",
       "p.P307R                 4.643030e-10                         2   \n",
       "p.R882H                 1.882290e-08                      1069   \n",
       "p.F732S                 4.958240e-10                         3   \n",
       "p.E756D                 9.655700e-10                         -   \n",
       "p.I780T                 1.346250e-09                         7   \n",
       "p.G543C                 8.141860e-10                        25   \n",
       "p.M880V                 1.486370e-09                         7   \n",
       "p.W753C                 1.046296e-09                         -   \n",
       "p.E733*                 5.399810e-10                         -   \n",
       "p.I705T                 1.346250e-09                         5   \n",
       "...                              ...                       ...   \n",
       "p.D768E                 2.420300e-09                         -   \n",
       "p.R598*                 1.882290e-08                         9   \n",
       "p.R326C                 1.415120e-08                         2   \n",
       "p.R736C                 1.415120e-08                        21   \n",
       "p.W893*                 3.524850e-09                         2   \n",
       "p.W581*                 3.524850e-09                         2   \n",
       "p.W601*                 5.410250e-09                         1   \n",
       "p.G890D                 3.445670e-09                         1   \n",
       "p.R771*                 1.882290e-08                        12   \n",
       "p.R749C                 1.200440e-08                        10   \n",
       "p.L547F                 4.087170e-09                         1   \n",
       "p.W440*                 4.610690e-09                         2   \n",
       "p.R635W                 1.415120e-08                        18   \n",
       "p.W313*                 5.140310e-09                         3   \n",
       "p.W306*                 5.410250e-09                         1   \n",
       "p.G550R                 1.480642e-08                         8   \n",
       "p.P385L                 1.415120e-08                         -   \n",
       "p.S770L                 1.200440e-08                         6   \n",
       "p.W327*                 4.610690e-09                         2   \n",
       "p.W860*                 5.140310e-09                         1   \n",
       "p.D702N                 1.200440e-08                         -   \n",
       "p.R326H                 1.882290e-08                         2   \n",
       "p.R899C                 1.415120e-08                         -   \n",
       "p.G685R                 1.480642e-08                         4   \n",
       "p.R899H                 1.882290e-08                         3   \n",
       "p.R688H                 1.882290e-08                         1   \n",
       "p.R659H                 1.882290e-08                         -   \n",
       "p.R366H                 1.882290e-08                         -   \n",
       "p.G332R                 1.480642e-08                         2   \n",
       "p.R771Q                 1.200440e-08                         4   \n",
       "\n",
       "        Exac Overall freq (%) Exac African freq (%) Exac East Asian freq (%)  \\\n",
       "p.E30A                 0.2959                  0.02                        0   \n",
       "p.F751C                     -                     -                        -   \n",
       "p.Y735C                0.0118                0.0097                        0   \n",
       "p.F732C                     -                     -                        -   \n",
       "p.S714C                0.0041                     0                        0   \n",
       "p.F414C                     -                     -                        -   \n",
       "p.L737R                     -                     -                        -   \n",
       "p.P507R                     -                     -                        -   \n",
       "p.W795S                     -                     -                        -   \n",
       "p.W860R                     -                     -                        -   \n",
       "p.Y908C                0.0033                     0                        0   \n",
       "p.F755S                0.0041                     0                        0   \n",
       "p.R882P                0.0058                 0.029                   0.0116   \n",
       "p.N838D                     -                     -                        -   \n",
       "p.E599D                0.0034                     0                        0   \n",
       "p.L547H                0.0017                     0                        0   \n",
       "p.E733G                     -                     -                        -   \n",
       "p.G543A                     -                     -                        -   \n",
       "p.R729G                0.0009                0.0097                        0   \n",
       "p.S770W                     -                     -                        -   \n",
       "p.P307R                     -                     -                        -   \n",
       "p.R882H                0.0545                0.0869                   0.0694   \n",
       "p.F732S                     -                     -                        -   \n",
       "p.E756D                     -                     -                        -   \n",
       "p.I780T                0.0051                0.0201                        0   \n",
       "p.G543C                     -                     -                        -   \n",
       "p.M880V                0.0017                0.0097                        0   \n",
       "p.W753C                0.0008                     0                        0   \n",
       "p.E733*                     -                     -                        -   \n",
       "p.I705T                     -                     -                        -   \n",
       "...                       ...                   ...                      ...   \n",
       "p.D768E                     -                     -                        -   \n",
       "p.R598*                0.0059                  0.01                        0   \n",
       "p.R326C                0.0058                0.0096                   0.0116   \n",
       "p.R736C                     -                     -                        -   \n",
       "p.W893*                0.0016                     0                        0   \n",
       "p.W581*                0.0027                     0                        0   \n",
       "p.W601*                     -                     -                        -   \n",
       "p.G890D                0.0008                     0                        0   \n",
       "p.R771*                     -                     -                        -   \n",
       "p.R749C                0.0025                     0                        0   \n",
       "p.L547F                     -                     -                        -   \n",
       "p.W440*                0.0008                     0                        0   \n",
       "p.R635W                0.0095                0.0353                        0   \n",
       "p.W313*                0.0049                0.0096                   0.0116   \n",
       "p.W306*                0.0008                     0                        0   \n",
       "p.G550R                     -                     -                        -   \n",
       "p.P385L                 0.005                     0                        0   \n",
       "p.S770L                 0.005                0.0096                        0   \n",
       "p.W327*                     -                     -                        -   \n",
       "p.W860*                0.0008                0.0096                        0   \n",
       "p.D702N                     -                     -                        -   \n",
       "p.R326H                0.0025                0.0096                        0   \n",
       "p.R899C                0.0016                     0                        0   \n",
       "p.G685R                0.0017                     0                        0   \n",
       "p.R899H                     -                     -                        -   \n",
       "p.R688H                0.0033                0.0096                        0   \n",
       "p.R659H                0.0017                     0                        0   \n",
       "p.R366H                     -                     -                        -   \n",
       "p.G332R                0.0041                     0                        0   \n",
       "p.R771Q                0.0025                     0                        0   \n",
       "\n",
       "        Exac European freq (%) Exac Finnish freq (%) Exac Latino freq (%)  \\\n",
       "p.E30A                  0.3033                2.1424               0.1413   \n",
       "p.F751C                      -                     -                    -   \n",
       "p.Y735C                 0.0182                     0               0.0087   \n",
       "p.F732C                      -                     -                    -   \n",
       "p.S714C                 0.0075                     0                    0   \n",
       "p.F414C                      -                     -                    -   \n",
       "p.L737R                      -                     -                    -   \n",
       "p.P507R                      -                     -                    -   \n",
       "p.W795S                      -                     -                    -   \n",
       "p.W860R                      -                     -                    -   \n",
       "p.Y908C                 0.0045                     0               0.0086   \n",
       "p.F755S                  0.006                     0               0.0086   \n",
       "p.R882P                  0.003                     0               0.0087   \n",
       "p.N838D                      -                     -                    -   \n",
       "p.E599D                 0.0061                     0                    0   \n",
       "p.L547H                      0                     0               0.0087   \n",
       "p.E733G                      -                     -                    -   \n",
       "p.G543A                      -                     -                    -   \n",
       "p.R729G                      0                     0                    0   \n",
       "p.S770W                      -                     -                    -   \n",
       "p.P307R                      -                     -                    -   \n",
       "p.R882H                 0.0571                0.0758                    0   \n",
       "p.F732S                      -                     -                    -   \n",
       "p.E756D                      -                     -                    -   \n",
       "p.I780T                 0.0061                     0                    0   \n",
       "p.G543C                      -                     -                    -   \n",
       "p.M880V                 0.0015                     0                    0   \n",
       "p.W753C                 0.0015                     0                    0   \n",
       "p.E733*                      -                     -                    -   \n",
       "p.I705T                      -                     -                    -   \n",
       "...                        ...                   ...                  ...   \n",
       "p.D768E                      -                     -                    -   \n",
       "p.R598*                 0.0061                     0                    0   \n",
       "p.R326C                  0.006                0.0151                    0   \n",
       "p.R736C                      -                     -                    -   \n",
       "p.W893*                  0.003                     0                    0   \n",
       "p.W581*                 0.0016                0.0177                    0   \n",
       "p.W601*                      -                     -                    -   \n",
       "p.G890D                 0.0015                     0                    0   \n",
       "p.R771*                      -                     -                    -   \n",
       "p.R749C                 0.0045                     0                    0   \n",
       "p.L547F                      -                     -                    -   \n",
       "p.W440*                 0.0015                     0                    0   \n",
       "p.R635W                      0                     0                    0   \n",
       "p.W313*                  0.006                     0                    0   \n",
       "p.W306*                 0.0015                     0                    0   \n",
       "p.G550R                      -                     -                    -   \n",
       "p.P385L                 0.0092                     0                    0   \n",
       "p.S770L                 0.0015                     0               0.0086   \n",
       "p.W327*                      -                     -                    -   \n",
       "p.W860*                      0                     0                    0   \n",
       "p.D702N                      -                     -                    -   \n",
       "p.R326H                  0.003                     0                    0   \n",
       "p.R899C                  0.003                     0                    0   \n",
       "p.G685R                  0.003                     0                    0   \n",
       "p.R899H                      -                     -                    -   \n",
       "p.R688H                  0.003                     0                    0   \n",
       "p.R659H                 0.0015                     0                    0   \n",
       "p.R366H                      -                     -                    -   \n",
       "p.G332R                 0.0075                     0                    0   \n",
       "p.R771Q                  0.003                     0                    0   \n",
       "\n",
       "        Exac South Asian freq (%)  \n",
       "p.E30A                     0.0247  \n",
       "p.F751C                         -  \n",
       "p.Y735C                         0  \n",
       "p.F732C                         -  \n",
       "p.S714C                         0  \n",
       "p.F414C                         -  \n",
       "p.L737R                         -  \n",
       "p.P507R                         -  \n",
       "p.W795S                         -  \n",
       "p.W860R                         -  \n",
       "p.Y908C                         0  \n",
       "p.F755S                         0  \n",
       "p.R882P                         0  \n",
       "p.N838D                         -  \n",
       "p.E599D                         0  \n",
       "p.L547H                    0.0061  \n",
       "p.E733G                         -  \n",
       "p.G543A                         -  \n",
       "p.R729G                         0  \n",
       "p.S770W                         -  \n",
       "p.P307R                         -  \n",
       "p.R882H                    0.0485  \n",
       "p.F732S                         -  \n",
       "p.E756D                         -  \n",
       "p.I780T                         0  \n",
       "p.G543C                         -  \n",
       "p.M880V                         0  \n",
       "p.W753C                         0  \n",
       "p.E733*                         -  \n",
       "p.I705T                         -  \n",
       "...                           ...  \n",
       "p.D768E                         -  \n",
       "p.R598*                    0.0121  \n",
       "p.R326C                         0  \n",
       "p.R736C                         -  \n",
       "p.W893*                         0  \n",
       "p.W581*                    0.0063  \n",
       "p.W601*                         -  \n",
       "p.G890D                         0  \n",
       "p.R771*                         -  \n",
       "p.R749C                         0  \n",
       "p.L547F                         -  \n",
       "p.W440*                         0  \n",
       "p.R635W                    0.0128  \n",
       "p.W313*                         0  \n",
       "p.W306*                         0  \n",
       "p.G550R                         -  \n",
       "p.P385L                         0  \n",
       "p.S770L                    0.0182  \n",
       "p.W327*                         -  \n",
       "p.W860*                         0  \n",
       "p.D702N                         -  \n",
       "p.R326H                         0  \n",
       "p.R899C                         0  \n",
       "p.G685R                         0  \n",
       "p.R899H                         -  \n",
       "p.R688H                    0.0061  \n",
       "p.R659H                    0.0062  \n",
       "p.R366H                         -  \n",
       "p.G332R                         0  \n",
       "p.R771Q                         0  \n",
       "\n",
       "[102 rows x 13 columns]"
      ]
     },
     "execution_count": 97,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DNMT3A_df_sorted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABZoAAAHvCAYAAADdMGqTAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAIABJREFUeJzs3Xd4VFX+x/H3CYFAgCS0EKkRFVBUEBRYwAVpCkgVRWEFxLIrCrIqrrpqAupPXCyACLooiooVaaIQUECKgPQaOqGGEkJoIZByfn/cmdmZzCQEQgn6eT3PPJOce+6533vP3Hn0m8P3GmstIiIiIiIiIiIiIiLnK+hyByAiIiIiIiIiIiIiVzYlmkVEREREREREREQkX5RoFhEREREREREREZF8UaJZRERERERERERERPJFiWYRERERERERERERyRclmkVEREREREREREQkX5RoFhEREblMjDHBxhhrjPnocsdSEBlj+hljNhljzriuUyVXewtjzBJjzAlX+98ud6wiIiIiIn92SjSLiIiIXGDGmGauBGhOr4a57BtkjIk1xnS4lDEXNMaYVsAIYB3wGPAgkGyMKQtMBIoCT7vaF16kGOq65qLKxRj/z8oYc222+yHTGJPi+qPC18aYe4wxwQH2a+m1T+8A291/uJmcrX2Bq/2MMaZcDjG97zV2E6+x8vqqZIyJMsZ8YoxZa4w5Yow5ZYzZYowZY4ypdpZr0t81zhFjTLFzvKQiIiIiBYLff8CJiIiIyAXzFfBTgPatANbaDFdSKcNrWxAQA3wMTL3oERZcrVzvD1lrj7kbjTHNgDDgZWvtxb4+dXHm4mdg10U+1p9RHPCF6+cSwLVAO6AbsMwY09lauyeHfQcbY7621qbl8VjpOPfW34B3vTcYY4oCDwBpOH/AAMjE+SOGt2bAw8Bo4Lds25KBq4HrgBk4n5dTQA2gD3CfMaa+tXZTDvH1AbYB1wBdgPF5PC8RERGRAkOJZhEREZGLZ4W19ovcOpxDouzPJgrI9E4ye7WDk9iTXBhjDFDcWnvicseSg43Z7w9jzEDgGWAoMM0YU89am5ltv2XArUA/V7+8SMVJDj9EtkQz0BkoBXwJdAew1lr+lwR3x1YUJ9H8Ww739XqgSfZGY8xE17GfAPoH2F4PqO069vM4SWclmkVEROSKo9IZIiIiIpdJ9hrNxphrcVZeAjzs9c/yM7L3d/3z/vnGmFRjTJIx5r/GmOIBjlHRGPOBMWa3q3TAXtfvZbP1K2OMGW6M2W6MSTPGHDbGLDPGPJ2tX29jzFJXqYOTxphtxpgvjDFl8njOdxpjZhljjrqOs9oY85jX9muNMRZnNWkhr2vwszFmD85Kb4D53tfGtW9RY8xLxpgNrrGPGGOmGmNqB4jDGGP+boz53Ti1no8bY9YYY2Jc218DxmQ7lvdcFTPGDDZOuYdU17HWGmOG5OEaBLvinG+MOeCal52u8g2lc9jnXmPMr67rlmqM2eiar2DXdndZiQeNU9s6HjgNDPAao5nrOh5zjbHcBC5BcZMx5nvXZ+W0MSbRGDPbGNPGq895n39urOMt4Buc5Ou9Abp9CawGXjDGlDqH4T8BbnIldr09BCwH1p5HyHmx0/WeU6wPA8eAycA44A5jzNUXKRYRERGRi0YrmkVEREQuntDsCV3gtLX2eA799wO9cJJNc/lfUjUrW796OKswP8ZZddkceBSnBEdfdydXsuo3oJCr73acf9r/OE4y6zavFcMTgb8AHwBrgOLA9TjlAt5xjfcQMBb4FXgFpzRAFaAtUBY4nNvFMMY8DrzviulVnFWmdwIfGmOutta+4LoGD7pibOi6HgCJOCUz2uEk5l4FNruvjTGmCDATaAB8hlPfuZTruvxmjGlirV3p6mtwypp0AxYBrwMprvO9BxgEfIezetr7WOAqe+K6Tj2BT11jFHZd2+a5XQOXojirdr/HSS6eBOrj1KJu7JoX9x8cMMa8CTyHs2L2beAATomFe4B/A94rlp9xnffHrmu50zVGJ2CC6zoOdR3zAeAT17V3J9jLAbNxSkd8gFMCoixwmyvG6Rfg/PPiI5z5aQd8nW2bxVn5Ox14Aefa5MUUnM9oH5zEMsapv90CZ3V0WL6jxvNZDAOK4JQDGeTa5FdGx/yvbMe31tpTxpgvgDdxkt+vXIh4RERERC4Za61eeumll1566aWXXhfwhZOctTm8vvbqF+xq+yi3tgDbMoFbs22Lw1nBWsyr7UecZGOFbH0buMZ4yfV7ade4I85yXlNxSlYUOo9rUskV32cBtr2PkySv6tX2BZARoO8jrlibZGsfiJN0bpmtPQLYA/zs1dbdNcanQFC2/kFnO5Zr2zFg6nl+PoK858mr/e+u43XxamvkapsFhGTrb7x+bunqlwSUzdavsOsaJANRXu0hwGLXZ6Gaq61L9hhyOIf8nP+1rmMMy6VPpKvPkgDnOMD1+y84f+yolO3+mJxtrAVAiuvn4a7rEOL63f0Hkwic5HXA+c72efjbWc6vE773fCLwVA59e2Q/JvADToI/KLfj6KWXXnrppZdeehW0l0pniIiIiFw8/8V5qJ3367ULMO4Ca+2ybG2zcVZQVgVwlWBog7Ni9owxpqz7hfPQsR1Aa9e+qTglOxoaY6rmctyjQEmgrWtV8Lm41xXfWO9YXPH8gLPqusU5juntbzgrfldlGzsYJyHZ1BgT4urrTu49a631WS2e/fdcHMUpw1DrXAO11mZZa08BGGMKGWMiXLHOdnVp4NW9h+v9eWvt6Wzj2ADDf2qtTcrWdhtQEeePF/u99j8NvIWT+O7gdV7gzHHJXE7jvM8/j9wr7XNbZfwvnGT54HMYdyzOiu9Ors9wL2CStTblvKIMbCHOvd4BZ8X1fqC0u8xJNg8D26y1C7zaPgUq878HYoqIiIhcEVQ6Q0REROTi2WKt/fkijLs9QJu7bIW7VnJNwOCskv17DuNkgvNAQuPUYn4HSDDGrMdJek6y1s7x6v8azsPOpgJJxphfccoXfGPP/sC5613vc3LpU/4sY+SmJk4i+1AufUrjrC69DtgTICF7Lp7CKXGyzhizDee8pgLTckgA+zDG3I9T5qI2zopjb961fK/Dmac1eYxrc4A2d73f9QG2rXO9V3O9z8apgfww0NMYsxSnJMk31tqNXvvl6/zzwJ1gzv4wSA9r7TJjzHdAL2PM28Cmsw1qrV1tjFmJU5riIM55/+MCxOt9jEOA+77/wVUOYw1OCZIn3P2MMdVw/vXDaOPUZ3eLxylt8jDOv1QQERERuSIo0SwiIiJy5cnMZZvJ9j4OpwxFIKnuH6y1I40xk3Bq4v4VuA/oZ4wZb639m6vPJmNMTZwSBi2Apji1dGONMX+11u7IQ1w9cBJ8gWzNoT0vgoBVOCU0cpLsFUu+kqHW2omuRHtbnOvQCqe0wlxjTGvrVWM5O2PMfTg1ohcD/XHKWqThJMp/xPeB3ee6cjw1QFuex3AliXu4HurXBrgdpwbyy8aYftba0a5+533+eXSz6/1syeN/49QrH+J6z4uxOCU0CuOUqPjlfALMK2vtHmPMbOBRY8wAr2vTB2du+uJVW91LR2NMGWttrrXPRURERAoKJZpFRERECpYLsRoUnKStBQrndVW1tXYvTrmP/7r+mf94nKTj29b1ID1XuYUfXS+MMR1wHrI2AGeVa062uN4PXaRV3luAcsAveVhRuwmnNETZs6xqznUcVwLwc+BzVxmGoTirlO8GJuWy64M4CeE7rLVp7kZjzI05xNoSuAlYkVs8udjmeg9U5uIG17vPKnlr7VpgLfAfVxmW33GSuaO9+pzv+efFI673H3PrZK3daoz5L85K4dvzOPZ4nJIhzYHB51AuJT+K4SS2SwBHjDFBOGU7luNc1+wqAsNwSsIMvwTxiYiIiOSbajSLiIiIFCDW2kyc1a2l8znOAZySB/caY27Lvt04yrl+DjXGFMu2fwZOohF3LK46wtmt8O6Ti2+AM8BgY0zRAPFEGGOKnGWM3HyGk5wLmOw2xniX5RiPs5L0P9lrTWf73V0OpHS2PsHGmHDvNldye1Wg/gFk4jy40PPf4q7jvhSg75eu9zeyX59zqJO9FNgLPGyMifTavwjwrCuWqa620tnHtdYmAwlACWNMkQtw/rkyxjwDdANWAhPysMtgnLl6My/jW2uP4JTLGASMOc8w/WT7jHm334RTImOz69gAd+I8IPMza+2EAK/hwG6cVc8iIiIiVwStaBYREREpeBYDdxpjnsNJNmVaa789j3H+DiwAFhhjPsNJ3AXj1OztBHyMU3f5BuBnV+mM9cARV9vjOKthF7rGm22MOQTMd8VVGqfWbRY5l+cAwFq70xjzJPABsMFVt3YXzirkm4GOQHWcMhLn4x2clb/vGmNaAnOB40AVnDIfx/nfw9W+Brq4Yq9ujPkB5+F21V19a7v6/Y6zqvklV1L+JP97kOIuY8xUnOSqu9bv4zjlOXJdhYuTPO2Icz0/x3mgXWfALwFvrf3NVX/4GcBdk3g/zhzea4y55Wz1sa21GcaYfsB3wFJjzBjXudwP1MdZ1ete0dwHeNL1WdgKZAB3uK7Ll9baM64/OOTn/N1qGmP+5vq5OHAN0B6n3vbvQGfXH15yZa096LpGMXk8LtbaT/Pa9xy8bIxpCvyEk5gvhLMS/W+un5/w6vswzmcrt5XfE4GnjDG3Bnj4p4iIiEiBo0SziIiISMHzD2AkzgrXkjgrYM850exK7tYFngc6AD1xSjbsBibzv9WiO3FqOTfDSXiG4CR8PwTe9Crv8D5wryu+0jgPIFwB9LXWzs1DPGOMMfE4q2gfB8KBJGAjTq3d3B7kd7axzxhj7gKexEnsDXJt2oeTuB/n1dcaY7rhJP764CQoM3HKR3zn1W+HMeYRnBrFo3FKH3zs2m8ETvK1FU6SNBEnafiGtXb/WWL9whhTAmf19ds4ydmpwMsEqF9trX3W9QC7J4B/4azG3g1Mw1n9npfrM8kY0wrnM/UvnP8PiAf6WGs/8eo6Gyfx3x64CifRvAMn0T3S1ed4fs7fy52ul8VZkbwf5/P0Ms6DKM+aZPbyFs7nMj8PlMyvH4AKOKuxI3GSy3tw/rDxlrU2Hjz/MqA9sNRauzuX8b7H+Yz0AZRoFhERkQLPXJiHQl8cxpg3gVtxVpeUBU7h/I/QZGBkoAdjGGMa4fwHdEOcVSFbcR748d45/seqiIiIiIiIiIiIiORBQU80n8FZ1bABZ3VHcZwE8q04q1Maeq8CMMZ0xPnLfxpOHcBknNUCNYAJ1tp7L+kJiIiIiIiIiIiIiPwJFPREc1HvJ3F7tb8OvAiMttb2dbWF4axeDgcau+uYuR42Mxv4C/CAtfbrSxW/iIiIiIiIiIiIyJ9B0Nm7XD6Bkswu7hqF13m1dcV5mMzX3g/LcI3hfoL34xc8SBEREREREREREZE/uQKdaM5Fe9f7Gq+25q73GQH6z8N58E0jY0zIxQxMRERERERERERE5M8m+HIHkBfGmGeBEjhlMW4FmuAkmYd4davhet+cfX9rbYYxZgdQC6iG84TtgEaPHm0BHn/8cXNBghcRERERERERERH5g7tSVjQ/C8QAA3CSzDOA1tbaQ159wl3vR3MYw90eEWijMeYxY4yn5MbRo0ftypUrbXp6ul22bJkF7LJly+zp06ft6tWr7ZEjR+zWrVvtnj17bGJiot20aZM9duyYXblypc3IyPDbZ+3atTY5Odlu27bN7t692+7fv99u3LjRHj9+3C5fvtxmZmb67HPq1Cm7bt06e/jwYbt9+3a7a9cue+DAARsfH29PnDhhly1bZrOysuyw6b/boRsz7LDpv9vU1FS7fv16e+jQIbtjxw67c+dOe/DgQbthwwZ78uRJz/hLly71HOfkyZN2w4YN9uDBg3bnzp12x44d9tChQ3b9+vU2NTXVJ6asrCy7bNkye+LECRsfH28PHDhgd+3aZbdv324PHz5s161bZ0+dOuWzT2Zmpl2+fLk9fvy43bhxo92/f7/dvXu33bZtm01OTrZr1661p0+f9tknIyPDrly50h47dsxu2rTJJiYm2j179titW7faI0eO2NWrV/vtk56ebleuXGmPHj1qN2/ebPft22f37t1rt2zZYlNSUuyqVatsenq6z7mfOXPGM5dbtmyxe/futfv27bObN2+23vPv3mfp0qX29OnTds2aNZ65zD7/K1as8JvLtLQ0u3btWp+5zD7/7mubff6TkpI8c3ngwAG7YcMGz/xba3328Z7/hIQEm5CQYA8dOuQ3/97vJ06c8Jv/pKQku27dOpuamupzvbKysuzy5ctznf+0tDSfz1lmZqZdsWKFPXbsWMD5X7NmTY7zf/ToUbtp0ya7b98+v/k/c+aM3/yvWrXKpqSk2M2bN9u9e/favXv32s2bN/vMv/c+3vPvvpf37dtnN23a5Jn/QPey9/x738ve8+993dLS0nK9l7PP/9KlS21qaqrf/LvvZff8Z59L972cff4D3cvuz457/g8cOOA3/9nvZff8e9/L3vO/du1an/l33//u+fe+l7PPv/c9ltu9HGj+3fu45z+3e7kgf5fndC/ru1zf5fou13e5vsv1Xa7vcn2X67tc3+X6Ltd3ub7L9V2e3+9yLpEC/TDA7Iwx5YFGOCuZSwJ3W2tXuLZtxqnZfJ21dmuAfX/DeSDgX6y1i3M6hteK5gt/AhfB0I0Znp8H1rwiFqiLiIiIiIiIiIjIpXNJKjdcKSuaAbDWHrDWTgJaA2WAz7w2u1csh/vt6AjL1k9ERERERERERERELoArKtHsZq3dCWwAahljyrqaN7neq2fvb4wJBq4GMoDtlyRIERERERERERERkT+JKzLR7FLB9Z7pep/ter8rQN+/AqHAb9ba0xc7MBEREREREREREZE/kwKbaDbG1DTGRAVoDzLGvA5E4iSOj7g2TQCSgPuNMbd69S8KvOb6dfRFDltERERERERERETkT6cgPz3uLmCoMWYesA04DJQHmgLVgP3Ao+7O1tpjxphHcRLOc40xXwPJQAeghqv9m0t6BiIiIiIiIiIiIiJ/AgU50fwz8F+gMVAbiABOApuBz4ER1tpk7x2stZONMU2BfwP3AEWBrcDTrv720oUvIiIiIiIiIiIi8udQYBPN1tp1wBPnsd9CoO2Fj0hEREREREREREREAimwNZpFRERERERERERE5MqgRLOIiIiIiIiIiIiI5IsSzSIiIiIiIiIiIiKSL0o0i4iIiIiIiIiIiEi+KNEsIiIiIiIiIiIiIvmiRLOIiIiIiIiIiIiI5IsSzSIiIiIiIiIiIiKSL0o0i4iIiIiIiIiIiEi+KNEsIiIiIiIiIiIiIvmiRPMVbE+q5eBpy/40531Pqr3cIeVbbGwsxhjmzp17uUMRERERERERERGRPAq+3AHIuVt3NIuJe7LYeNw3sfzcmgxqljR0qRTEjeH6G4KIiIiIiIiIiIhcGspGXmHmHMzijfhMvySz28bjljfiM5l7MOsSR3ZhPPnkk8THx1O/fv3LHYqIiIiIiIiIiIjkkVY0X0HWHc3io+2ZnK1AhgXGbM+kbAhX3MrmsmXLUrZs2csdhoiIiIiIiIiIiJyDKysL+Sc3cU/WWZPMbhaYtOfSrGpOSEjAGEPv3r3Ztm0bXbt2pUyZMpQsWZLWrVuzbt06AA4dOsRjjz3GVVddRdGiRbntttuYM2eOz1g51WieP38+7du3p1KlSoSEhBAVFUXDhg0ZNGiQT78DBw7w7LPPUqNGDYoXL05ERAQ1atSgd+/ebN++3S/2uLg42rZtS9myZQkJCeGaa65h4MCBpKSk+PWNjo4mOjqa1NRUBg4cSJUqVQgJCeHaa6/lzTffxFr/2Zk6dSotWrTgqquuIiQkhAoVKtC0aVNGjRrl1zc5OZkXXniB66+/nmLFihEeHk6LFi2YOXPmWedARERERERERETkctKK5ivEnlSbY7mMnMQfdx4QWCnUXKSofCUkJNCgQQOuv/56evfuTUJCApMmTaJZs2YsWrSIu+66i7CwMLp160ZycjJff/01bdq0YfPmzVSpUiXHcWfMmEG7du0ICwujQ4cOVKxYkeTkZOLj4xk1ahQxMTEApKam0rhxY7Zt20arVq1o37491lp27tzJlClT6Nq1K9WqVfOMO3jwYGJiYihdujR33303kZGRrFmzhrfeeouffvqJRYsWERYW5hNLeno6rVu3Zt++fbRp04bg4GAmT57M888/T1pamicWgP/+97/8/e9/Jyoqivbt21O2bFkOHjzImjVr+OSTT+jbt6+n786dO2nWrBkJCQncfvvt3HXXXZw8eZJp06Zx11138eGHH/Loo49eqKkSERERERERERG5oJRovkKsO3p+q5PXHc2iUmihCxxNYL/++iuvvfYa//73vz1tr776Kq+88goNGjTgvvvuY9SoUQQFOQvpW7VqRc+ePXn33Xd59913cxx3zJgxZGVlMXfuXGrXru2zLSkpyfPzL7/8wrZt2xgwYIDfeGfOnOH06dOe3+fMmUNMTAx/+ctf+Omnn4iIiPBs+/TTT3nooYeIiYnxG2ffvn3Url2bWbNmUaxYMQBiYmKoXr067777Li+++CKFCxcG4MMPP6RIkSKsXr2ayMjIHOMG6NWrFzt37uSrr77i/vvv97SnpKTQrFkz+vfvT4cOHShfvnyO10lERERERERERORyUaL5Ium+OP1yhwDAZzuz+Gxn7knqLxsWviDHio6O5vnnn/dp69WrF6+88gqnT59m6NChniQzQPfu3enTpw+rVq3K0/juxK63QPWcA/UrUqQIRYoU8fw+YsQIwElieyeZAXr37s3w4cMZP358wAT4iBEjfI4RGRlJx44d+eyzz9i0aRM33nijZ1twcLAn8ZxT3KtXr+bXX3+la9euPklmgIiICAYNGkSnTp34/vvvfVZBi4iIiIiIiIiIFBRKNMsFU6dOHQoV8l09XaFCBQCqV69OyZIlfbYVKlSI8uXLs2fPnlzH7dGjBxMnTqRBgwZ069aNO+64g8aNG1OpUiWffk2bNqVixYoMGTKEFStW0LZtWxo3bhwwrkWLFlG4cGG+++47vvvuO79jnjlzhkOHDnH48GHKlCnjaQ8PD+faa6/161+5cmUAjhw54hP3M888Q61atejWrRtNmzalcePGlCtXzi8WgKNHjxIbG+s39qFDhwCIj4/P7TKJiIiIiIiIiIhcNko0ywUTHh7u1xYcHJzjNvf29PTcV3936dKFadOm8fbbbzN27Fg+/PBDAOrVq8cbb7xBq1atAAgLC2Px4sXExMQwdepU4uLiAGf1cN++fXnppZc8q4sPHz5MRkaG38MEsztx4oRPojn76ufs55mZmelpe/rppylbtiyjRo1ixIgRDBs2DGMMTZs2ZejQodx6662eWABmzZrFrFmzco1FRERERERERESkIFKi+SK5UOUovA1en3FODwS8vqTh5Vp/jClu164d7dq14+TJkyxZsoRp06YxevRo7r77blauXMkNN9wAQKVKlfj444+x1rJhwwZmz57N+++/z+DBg8nKyuLVV18FnMR3VlYWycnJFzXunj170rNnT1JSUvjtt9+YNGkSY8eO5c477yQ+Pp7IyEhPEn748OH079//osYjIiIiIiIiIiJyMQSdvYsUFF0qBWHy2NcAnSv98aa3ePHiNG/enHfeeYcXX3yRM2fOMH36dL9+xhhq1apFv379PKuEJ0+e7NnesGFDjhw5wvr16y9J3BEREbRt25YxY8bQu3dvkpOTmT9/vicWwPO7iIiIiIiIiIjIleaPl4n8A7sxPIhHqhU6a7LZAI9WK8SN4X+M6f3ll184deqUX/uBAwcACA0NBWDdunUkJCSctR/AP//5TwAeffRR9u3b57fPyZMnWbx4cb7injFjBhkZGX7tBw8e9Inn1ltv5fbbb2fixImMHTs24Fhr16717CciIiIiIiIiIlLQ/DHqKvyJ3BEZRLkQmLQni/gAZTSuL2noXCnoD5NkBnjmmWdISEigWbNmREdHU6RIEZYvX87s2bOpWrUq999/PwA///wzTz/9NI0aNaJmzZpERkayZ88epkyZQlBQEAMHDvSM2aJFC4YMGcILL7zAddddR9u2bbn66qs5ceIEO3fu5Ndff6VJkybMmDHjvOO+//77KVq0KE2aNCE6OhprLfPnz2fp0qXUq1ePli1bevp++eWXNG/enIcffpgRI0bQoEEDIiIi2LNnD2vWrGHdunUsWrSIyMjI87+QIiIiIiIiIudp6Eb/hVTeBtZUiknkz07fAlegG8OdRPKeVMvwLRlkWQgy8NR1wVQKzWtxjSvHiy++yKRJk1i2bBk///wzQUFBVKlShRdffJEBAwZQqlQpAO68804GDBjAvHnzmDJlCseOHeOqq66iVatWngS0t3/96180btyYESNGsGDBAqZMmUJ4eDgVK1bkscceo3v37vmKe8iQIcTFxbFixQp++uknihYtStWqVXnzzTd5/PHHPQ8mBKe29PLly3nvvff4/vvvGT9+PJmZmURFRXHDDTfQr18/brrppnzFIyIiIiIiInIhHDnzv4VvpYr88fIQInJ+jLV5f7jcn8Ho0aMtwOOPP365Q8kT778o6q+HIiIiIiIiInKxKRchcsW5JH8R+uPUVxARERERERERERGRy0KJZhERERERERERERHJFyWaRURERERERERERCRflGgWERERERERERERkXxRollERERERERERERE8kWJZhERERERERERERHJFyWaRURERERERERERCRflGgWERERERERERERkXxRollERERERERERERE8kWJZhERERERERERERHJFyWaRURERERERERERCRflGgWERERERERERERkXxRollERERERERERERE8kWJZhERERERERERERHJFyWaRURERERERERERCRflGgWERERERERERERkXxRollERERERERERERE8kWJZsm3hIQEjDH07t2bhIQE7r//fsqWLUvRokW59dZbmTZtmt8+p0+fZsiQIdx8882EhoYSFhbG7bffzrfffntBxj9z5gwjRoxZLfXTAAAgAElEQVSgbt26lCpVitDQUKKjo+nYsSM///yzX/+NGzfSu3dvKleuTEhICOXLl6d79+5s2rTJr2/v3r0xxpCQkMCHH37ITTfdRNGiRSlfvjyPPfYYR48e9dtnzZo1PPDAA0RHRxMSEkK5cuWoW7cuAwYMID093advRkYGo0aNomHDhoSFhREaGsott9zCyJEjycrKynUuRERERERERERELofgyx2A/HHs3LmT+vXrU61aNR588EGSk5P55ptvPMndO+64A3CSwHfeeSe//vorNWvW5IknniA1NZUJEybQrVs3Vq1axf/93/+d9/jgJIO/+uorbrzxRnr27EmxYsXYt28fCxYsYMaMGbRs2dLTd8aMGXTp0oX09HTat2/Ptddey549e5g4cSI//vgjc+bMoW7dun7xPPfcc8TFxdG+fXtat27NnDlzGDNmDFu3bmX27NmefmvWrKFBgwYYY+jQoQNXX301x44dY+vWrYwaNYrXXnuNwoULA3hiiIuLo0aNGnTv3p2iRYsyZ84c+vXrx5IlS/j8888v2JyJiIiIiIiIiIhcCEo0ywUzd+5cYmNjiYmJ8bR1796du+66i6FDh3oSwW+//Ta//vorbdq0YerUqQQHOx/DmJgY6tevzxtvvMHdd99No0aNzmv8o0eP8vXXX1OvXj2WLFlCoUKFfMY5fPiw5+cjR47wwAMPEBoayrx587jhhhs829avX0+DBg145JFHWLFihd/5Ll68mLVr11KlShXAWYncvHlz5syZw++//079+vUBGDduHGlpaUyePJmOHTv6jHHkyBFCQ0M9v7/++uvExcXx5JNPMmzYME/smZmZPPbYY4wdO5auXbv6jSMiIiIiIiIiInI5KdF8kZhOIy75MZ87z/3s5P4X5PhVq1blpZde8mm78847qVKlCr///runbezYsRhjeOeddzxJZoDIyEhefvllHnnkET766CO/RHNexzfGYK0lJCSEoCD/6jBlypTx/PzZZ5+RkpLCyJEjfZLMALVq1eLRRx9l2LBhbNiwwW/7K6+84kkyAwQHB/PQQw8xf/58n0SzW7FixfxiKVWqlOfnrKwsRo4cSVRUFO+++65PgrxQoUK8/fbbfPLJJ4wfP16JZhERERERERERKVCUaJYLpk6dOn6rhwEqV67MokWLADh+/Dhbt26lYsWK1KxZ069v8+bNAVi5cuV5jQ8QFhZG+/bt+eGHH6hTpw733HMPt99+Ow0aNPBZPQx49lu9ejWxsbF+Y2/evBmA+Ph4v0TzrbfeGjAWcFYqu3Xr1o3hw4fTqVMnunbtSsuWLWncuDHXXHON37EOHz7Mddddx2uvveY3NjjJ6vj4+IDbRERERERERERELhclmuWCiYiICNgeHBzseYid+0F5V111VcC+7vaUlJTzGt/tm2++4c033+TLL7/0lNooWrQoXbt25a233qJ8+fLA/8pojBkzJtdzO3HiRJ7ica/QzszM9LTVr1+f+fPn8/rrrzNhwgRPjeUaNWoQExPDAw884BPLli1bGDRo0DnFIiIiIiIiIiIicjkV2ESzMaYM0BloB9wEVATOAGuBT4BPrLVZXv2jgR25DPmNtfb+ixVvdheqHMXZDN2Y4fl5YM0CO50e4eHhAOzfvz/g9sTERJ9+56tYsWLExsYSGxvL7t27mTdvHp9++ilffPEFCQkJzJ8/3+c4q1ev5uabb87XMXPzl7/8hWnTpnH69GmWL1/OjBkzeO+99+jevTvlypWjZcuWnlg6d+7MxIkTL1osIiIiIiIiIiIiF5p/AduC415gDNAAWAIMA74HbgQ+Ar41xpgA+60GBgV4TbgEMctZlCxZkmuuuYa9e/eyZcsWv+1z5swBoG7duhfsmJUrV6ZHjx7ExcVx3XXXsWDBAs/q4YYNGwJ4Es8XW0hICI0aNWLw4MGMGOHU8Z4yZQoANWvWJCIigsWLF5Oenn5J4hEREREREREREbkQCnKieTPQAahkre1hrX3BWtsHqAnsBu4BugTYb5W1NjbAS4nmAqJPnz5Yaxk4cKBPiYmkpCReffVVT5/zdejQIZYsWeLXfvLkSY4fP05wcDBFihQB4KGHHiIiIoJBgwb5PFDQLSsri7lz5553LOAksd0lQ7wdOHAAwFM3Ojg4mH79+pGYmEj//v05deqU3z6JiYls2LAhX/GIiIiIiIiIiIhcaAW21oK1dnYO7fuNMR8ArwPNcFY5yxXk2WefZfr06UyZMoXatWvTtm1bUlNT+e677zh48CDPPfccTZo0Oe/x9+7dS8OGDbn++uupW7culStX5tixY0ybNo39+/fTv39/SpYsCUCZMmWYMGECnTt3pmHDhrRo0YJatWoRFBTErl27WLRoEYcPHyYtLe2843n77beZOXMmzZo1o1q1apQoUYL169czffp0SpUqxWOPPebp+/LLL7N69Wo++OADfvjhB5o3b07FihU5ePAgW7ZsYeHChbz++ut+DyYUERERERERERG5nApsovks3HUFMgJsq2CM+TtQBjgMLLLWrrlkkclZFSlShFmzZvHOO+/w5Zdf8t577xEcHEzt2rUZNmyY5+F45ys6OppBgwYxd+5c5syZQ1JSEqVLl6ZGjRoMGTKE++/3LdXdokUL1qxZw1tvvUVcXBzz58+nSJEiVKhQgebNm3PPPffkK56+fftSqlQplixZwsKFC8nIyKBSpUr07duXZ555hqpVq3r6Fi5cmMmTJ/PFF1/w6aefMm3aNE6cOEG5cuW4+uqrefXVV+nRo0e+4hEREREREREREbnQjLX2csdwTowxwcBKnFrNd1lr41zt0eT8MMC5QC9r7a6zjT969GgL8Pjjj1+AaC++K+1hgCIiIiIiIiJyZVMuQuSKE+g5dxdcQa7RnJMhOEnmn9xJZpdU4FWgHlDK9WoKzMEpsfGLMab4pQ1VRERERERERERE5I+vUGxs7OWOIc+MMf2BwcBG4O7Y2FjP09JiY2NPxsbGzomNjU2MjY1Nc712Dho0aDxwJ3AzcDA2Ntb/KXHO2I8NGjTow3bt2lUAqF69OvHx8ZQtW5aVK1dSoUIFli1bRpkyZVi/fj0hISHs3buX1NRUTp48yb59+yhSpAgbNmygXLlyrFixwmef+Ph4ChcuzL59+zh58iSnTp1iz549hISEsG7dOsqXL++zT6lSpdi0aROFChVi//79HD9+nLS0NHbv3k2xYsVYs2YNUVFRLFu2nKxSV1F4xwpurliKzZs3Y4zh4MGDHDt2jDNnzrBz505CQ0NZs2YNFSpUYOnSpVSsWJFly5YRHh7O1q1bsdaSlJTEkSNHyMjIICEhgRIlSrB69WpPTFFRUSxfvpywsDC2bdtGVlYWycnJHD58mKysLLZv307JkiV99ilfvjwrV66kZMmS7Nixg4yMDI4cOUJSUhIAW7duJSIiglWrVnn2iYyMZPXq1RQvXpyEhATS09M5evQoBw8exBjD5s2bKVWqlM8+5cqVY82aNYSGhrJr1y5Onz7NsWPHOHDgAIUKFWLjxo2UKVOGFStWeM69bNmyrFu3jpCQEPbs2UNaWhonTpwgMTGR4OBgz/y791m6dClly5Zlw4YNnrnMPv/r168nMjLSZy5Lly5NfHw8wcHBnrnMPv/ua5t9/oOCgjhw4ADHjh3j9OnT7Nq1yzP/V111lc8+ERERnvk/dOgQKSkppKen+83/smXLPO9hYWF+85+Zmcn27dspUaIEq1at8lyvqKgoVqxYkev8h4WFeeZl6dKlREVFsWrVKooXLx5w/rds2ZLj/IeGhrJz507OnDnjN/+lS5f2uS/LlSvH2rVrKVq0KLt27SItLY3jx4+zf/9+n/n33sd7/t338okTJ9i3bx+FCxcmPj4+4L3sPf/e97L3/C9fvtxz3UqXLs3GjRtzvJfXrl3rM/9Lly4NOP/ue9k9/9nn0n0vAz7zH+hedn923POflZXF4cOHfeY/+73snn/ve9l7/rdt20Z4eLjPXJYvX94z/9738qFDh3zmf+XKlZ57LDIyMsd7OSgoyG/+3fu453/37t053ssF+bs8p3tZ3+X6Ltd3ub7L9V2u73J9l+u7XN/l+i4viN/la+M3kVW8FIV3rqbhdRX1Xa7vcn2XF/Dv8quuumrQ+eZjz8UVUzrDGPMEMBLYALSw1u4/h30fAcYAE621uRbcVekMEREREREREZGcKRchcsVR6Qw3Y8wAnCTzOuCOc0kyuxxyvat0hoiIiIiIiIiIiMgFVuATzcaYfwHvAqtwkswHz2OYhq737RcsMBEREREREREREREBCnii2RjzMs7D/5bjlMtIyqVvA2NMkQDtzYF/un794qIEKiIiIiIiIiIiIvInVmAL6RhjeuE8+C8TmA/0N8avnEiCtfZT189vArWMMXOBPa62m4Hmrp9fttb+djFjFhEREREREREREfkzKrCJZuBq13shYEAOfX4FPnX9/DnQGbgNaAMUBg4A3wIjrbXzL1qkIiIiIiIiIiIiIn9iBTbRbK2NBWLPof/HwMcXKx4RERERERERERERCaxA12gWERERERERERERkYJPiWYRERERERERERERyRclmkVEREREREREREQkX5RoFhEREREREREREZF8UaJZpADp3bs3xhgSEhIudygF3ty5czHGEBsbe7lDybNmzZphjLncYYiIiIiIiIiIXHBKNEu+JCQkYIyhd+/elzsUP+np6UyaNImHH36YG2+8kbCwMEJDQ7npppt45ZVXOH78eI777tmzhz59+lChQgVCQkKIjo5mwIABHDly5BKegciFUZDvUxERERERERH5Ywi+3AGIXCzbtm2jS5cuFC9enDvuuIN27dpx4sQJ4uLiePXVV/nmm29YuHAhZcuW9duvUaNGHDx4kI4dO1KzZk1+//13hg8fzowZM1i4cCFlypS5TGclV7LPPvuM1NTUyx2GiIiIiIiIiMgFp0Sz/GGVLFmS999/n169elG8eHFP+5kzZ+jSpQs//vgjgwYN4r333vPZr2/fvhw8eJARI0bQr18/T/vTTz/Nu+++y7///W8++OCDS3Ye8sdRpUqVyx2CiIiIiIiIiMhFodIZct5iY2O5+uqrARg3bhzGGM/r008/9fTLysrigw8+4LbbbqNEiRIUL16c2267jdGjR5OVleU3rjGGZs2asW/fPh588EEiIyMpVqwY9erV48svv8xzfBUrVqRv374+SWaAIkWK8OKLLwJOnV9v27dvZ+bMmURHR/PEE0/4bBs0aBDFixfn888/5+TJk3mOY/ny5dxzzz1ERkYSEhJC1apV6du3L4mJiTnuk5WVxTvvvEPNmjUpWrQolSpV4p///CfHjh3z67tmzRoeeOABoqOjCQkJoVy5ctStW5cBAwaQnp7u0zcjI4NRo0bRsGFDTymRW265hZEjR/rNhXe5hc2bN9OtWzciIyMJCgpi7ty51KxZkyJFipCUlBTwHIYMGYIxhvfff9+nfc+ePTz55JNUq1aNkJAQypQpQ4cOHVi6dGnAcQ4cOMDDDz9M+fLlKVasGHXq1GHcuHE5XrtA3njjDYwxjBgxIuD2ffv2UahQIW677TaftsGDB9O4cWOioqIoUqQIFSpUoHv37sTHx/uNcbbrBYFrNJ85c4aRI0fStm1bqlatSkhICKVLl6Zly5ZMnz49YLzR0dFER0eTmprKwIEDqVKlCiEhIVx77bW8+eabWGs9ffN6n4qIiIiIiIiI5IdWNMt5a9asGSkpKQwfPpzatWvTqVMnz7Y6dep4fn7wwQf58ssvqVy5Mo888gjGGCZNmkTfvn1ZsGAB48eP9xv7yJEjNGrUiIiICB566CFSUlL49ttv6dGjB3v37mXgwIH5ir1w4cIABAf73gKzZ88GoHXr1gQF+f4dpmTJkjRu3JiZM2eyePFiWrRocdbjTJs2jXvuuQdrLV27dqVq1aosX76c0aNHM2XKFBYuXEh0dLTffv/85z+ZN28e9913Hx07diQuLo5hw4Yxf/58FixYQNGiRQEnydygQQOMMXTo0IGrr76aY8eOsXXrVkaNGsVrr73mOdf09HTat29PXFwcNWrUoHv37hQtWpQ5c+bQr18/lixZwueff+4Xy7Zt22jQoAHVq1enR48enDp1irCwMHr16sWLL77IV1995bPy2+2zzz6jSJEi3H///Z62FStW0Lp1a5KTk7nzzjvp0qULSUlJTJ48mSZNmjBp0iTatm3r6X/48GEaNWrE9u3badKkCU2aNCExMZF//OMftG7d+qzX361nz5689NJLjBs3jv79+/tt/+KLL8jKyqJXr16etnnz5jFkyBDuuOMO7rnnHkqUKMGWLVuYMGECU6dOZeHChdSuXTvP1ysnycnJPPXUUzRq1IhWrVpRrlw5EhMT+eGHH2jbti1jxozhkUce8dsvPT2d1q1bs2/fPtq0aUNwcDCTJ0/m+eefJy0tjZiYGCDv96mIiIiIiIiISL5Ya/Xyeo0aNcqOGjXKXin+E5/ueV0OO3bssIDt1atXwO1ffvmlBewtt9xijx8/7mk/ceKErVevngXs+PHjffYBLGDvvfdem5mZ6Wnfvn27LVWqlC1cuLDdtm1bvuL+xz/+YQH7/PPP+7Q/++yzFrBvvfVWwP2eeOIJC+TpM3L8+HFbpkwZGxQUZOfNm+ezbciQIRawrVq18mnv1auXBWyZMmVsQkKCpz0zM9N26dLFAnbw4MGe9qefftoCdvLkyX7HT05O9rl+MTExFrBPPvmkzcjI8LRnZGTYPn36+I3jnlvAvvDCC37j79mzxwYFBdl69er5bfv9998tYLt06eJpS09Pt9dcc40NCQmxc+fO9em/d+9eW6FCBRsVFWXT0tI87Y8++qgF7IABA3z6L1261AYHB1vAxsTE+B0/kNatW1vArl271m/bDTfcYAsXLmyTkpI8bQcOHLDHjh3z67tq1SpbvHhxe9ddd/m0n+16WWtt06ZNrfO1+z9paWl29+7dfn1TUlJsrVq1bKlSpWxqaqrPtqpVq1rAtmnTxmfbgQMHbHh4uA0PD7dnzpzxiy2n+1RERERERORcXO5chIics0uSV9WK5oul9r2X5DD5W9frsvq7CzFKQGPHjgWcMgolSpTwtBcvXpw333yTli1b8tFHH9G9e3ef/QoVKsSbb77ps6r46quvpn///gwaNIjPP//cs2LzXE2dOpUPP/yQSpUq8dxzz/lsO3r0KADh4eEB93W3p6SknPU4U6ZM4fDhwzzwwAPcfvvtPtueeeYZPvjgA2bNmsWuXbv8avc+9dRTVK1a1fN7UFAQQ4cOZfLkyYwdO5aXX37Zp3+xYsX8jl+qVCnPz1lZWYwcOZKoqCjeffddChUq5NlWqFAh3n77bT755BPGjx9Px44dfcYpX758wGtdsWJFWrRowaxZs1i/fj21atXybHOXtvBeIfzjjz+ybds2nn32WZo2beozVoUKFXjuuecYMGAAv/zyC23btiU9PZ3x48dTsmRJYmNjffrfeuut9OjR45xKaPTq1YuZM2cybtw4hg4d6mlftmwZGzZsoHPnzj4PeYyMjAw4Tu3atWnevDkzZ84kPT3ds2LcLafrlZOQkBAqVark1x4eHk6fPn145plnWLp0KX/961/9+owYMcJn7iMjI+nYsSOfffYZmzZt4sYbb8xzHCIiIiIiIiIi+aFEs1xUK1asICgoiGbNmvlta9q0KYUKFWLlypV+26pUqeKpK+utWbNmDBo0KOA+efHbb7/RvXt3ihcvzvfff++TjM0L66p9m73ObiArVqwAoHnz5n7bgoOD+etf/0pCQgIrV670SzRnT8QCVKtWjcqVK5OQkEBKSgoRERF069aN4cOH06lTJ7p27UrLli1p3Lgx11xzjc++mzdv5vDhw1x33XW89tprAeMtVqxYwNrDtWvXJiQkJOA+vXv3ZtasWYwbN47//Oc/gFNz+Ouvv6ZcuXI+ZTAWLVoEwM6dO/0SxwBbtmwBID4+nrZt27Jx40ZSU1O5/fbbAyb+mzVrdk6J5s6dOxMeHs4XX3zBkCFDPMl29xi9e/f22+fHH3/kgw8+YNmyZSQlJZGRkeGzPSkpiauuusqnLbfrlZP169czdOhQ5s2bR2JiImlpaT7b9+7d67dPeHg41157rV975cqVAaf8jIiIiIiIiIjIpaJEs1xUR48epXTp0hQpUsRvW3BwMGXLluXgwYN+28qXLx9wvKioKM+452rRokW0adOGoKAgpk+fTv369f36uBOaOY3vfhhfTiuevbnHyJ6IdHO3B1odndv579y5k6NHjxIREUH9+vWZP38+r7/+OhMmTPDUWK5RowYxMTE88MADgFPrGJxk7qBBg3KM+cSJEwGPmZPOnTsTFhbGF198wRtvvEGhQoWYNm0ahw8fZsCAAT41sN0xfPdd7ivo3TG4r9/ZPgt5VaxYMe677z7GjBnDzJkzadOmDenp6Xz11VeUK1eONm3a+PQfMWIETz31FKVKlaJVq1ZUqVKF0NBQjDFMnjyZ1atXc/r06XzHtXjxYpo3b05GRgYtWrSgQ4cOhIWFERQUxKpVq5gyZUrA40RERAQcz33NMzMzzykOEREREREREZH8UKL5YrmI5Si8Dd34vxWWA2sWvOkMDw8nOTk5YImBjIwMkpKSAj4o7cCBAwHH279/v2fcczF//nzatWtHUFAQcXFxNGzYMGC/GjVqAM4K4EDcq26rV69+1mO6Y3THnF1iYqJPP28HDhzwxOIt0Pn/5S9/Ydq0aZw+fZrly5czY8YM3nvvPbp37065cuVo2bKlp3/nzp2ZOHHiWWP3ltvqbXfy9qOPPmLWrFncddddActmeMc8ZcoUOnTocNbjuvuf7bNwLnr16sWYMWMYN24cbdq08STFn3rqKZ/PZ0ZGBjExMURFRbFixQq/Pxa4V2cHkpfV7t5ee+01Tp06xZw5c/xW/r/xxhtMmTLlnMYTEREREREREbkcgs7eRSRn7vIDOa2evOWWW8jKymLevHl+2+bNm0dmZiZ169b127Zr1y4SEhL82ufOnesZN69mz55NmzZtCA4OZtasWTkmmQHuuOMOAGbOnElWVpbPtuPHj7Nw4UKKFSuW6xhu7hjdMXvLyMhgwYIFAAHP/9dff/Vr2759O7t37yY6OjrgataQkBAaNWrE4MGDGTFiBIAnSVmzZk0iIiJYvHgx6enpZ439XLhLTowbN46kpCSmT5/OzTffTJ06dXz6ua/Z/Pnz8zRuzZo1CQ0NZdWqVQFXmAe6rmfTuHFjrrvuOqZMmcLRo0dzTIonJSWRkpJCo0aN/JLMJ06c8JRFuRC2bt1K6dKlA5aXCfQ5OB9nu09FRERERERERPJLiWbJl1KlSmGMYdeuXQG39+nTB4AXXniB1NRUT3tqairPP/88AA8//LDffpmZmfzrX//ySfbu2LGDESNGEBwczN/+9rc8xTdz5kzuvvtuihYtyi+//MJtt92Wa/9rrrmG1q1bk5CQwPvvv++zLSYmhpMnT9KzZ0+KFy9+1mN36tSJ0qVL89VXX7F48WKfbcOGDWP79u20bNnSrz4zwPDhw9m5c6fn96ysLAYOHEhWVhYPPfSQp33+/PkBk7DuVcChoaGAU06hX79+JCYm0r9/f06dOuW3T2JiIhs2bDjreWXnnbwdPXo06enpAesdd+zYkWuuuYb333+fn376KeBYixYt8nxOChcuTI8ePTh+/LhfTedly5Yxfvz4c44VnKRyWloao0aN4qeffuLmm2/2+8NFZGQkoaGhLF++3KecSHp6Ok899RRJSUnndexAoqOjSU5OZs2aNT7tH3/8MXFxcRfkGGe7T0VERERERERE8qvg1VqQK0qJEiVo0KAB8+fPp0ePHlSvXp1ChQrRoUMHbr75Zrp3786UKVP49ttvqVWrFp06dfLUuN2xYwf33XcfPXr08Bv35ptvZsmSJdSrV4/WrVtz9OhRvvnmG1JSUvjPf/7j97C7QDZt2kTHjh1JS0ujbdu2TJkyJWAZguxJzFGjRtGoUSP69+/PL7/8wvXXX8+SJUuYM2cO1atX5/XXX8/ztRk7diz33nsvTZs25d5776VKlSosX76cmTNnEhUVxYcffhhw38aNG1OnTh26detGeHg4cXFxrF69mnr16vHcc895+r399tvMnDmTZs2aUa1aNUqUKMH69euZPn06pUqV4rHHHvP0ffnll1m9ejUffPABP/zwA82bN6dixYocPHiQLVu2sHDhQl5//XVuuOGGPJ2ft549e/Lyyy/z6quvEhwcTPfu3f36FC5cmIkTJ3LnnXfSrl07GjVqRJ06dQgNDWX37t0sXbqU7du3k5iY6EmQ/9///R+//PILw4YNY9myZTRp0oTExES++eYb2rZty9SpU88r1ldeeYWYmBjS09P9VjMDBAUF0b9/f4YMGcJNN91Ex44dOXPmDHPmzCE5OZk77riDOXPmnPOxAxkwYABxcXE0adKE++67j/DwcJYtW8aCBQvo2rUrEyZMyPcxznafioiIiIiIiIjkm7VWL6/XqFGj7KhRo+yV4j/x6Z7X5bJlyxZ7991329KlS1tjjAXsJ5984tmemZlp33//fVuvXj1brFgxW6xYMVu3bl07cuRIm5mZ6TceYJs2bWr37t1re/ToYcuVK2dDQkLsLbfcYsePH5/nuObMmWOBs74C2bVrl+3du7eNioqyhQsXtlWqVLH9+/e3hw8fPufr8/vvv9tOnTrZsmXL2sKFC9vKlSvbf/zjH3bv3r1+fXv16mUBu23bNvvWW2/ZGjVq2JCQEFuhQgX71FNP2aNHj/r0j4uLs71797bXX3+9DQsLs6H/z96dx8dxl3ke//xakmXdtmzJii07ieMEO3YITLI4B4NzAAMECDEJyywMx5BzGQiwkwALAdsBDGQHNjBMyAUMTAYSEuD0GbsAACAASURBVCfADGGYDU64huDblizHtyXZluXYjm7JUvdv/2h1uyX1VdXV9/f9eukldXdV908q1a+rn3rqeSor7QUXXGA/9rGP2YMHD055/kAgYH/4wx/aa665xs6cOdOWlZXZuXPn2iuvvNJ++ctftu3t7eFlDxw4YAH7wQ9+MOHveOjQIevz+Sxg3/72t8dd9tixY/bTn/60Xbp0qa2oqLBVVVV20aJF9t3vfrf90Y9+ZEdHJ/4vHz161H74wx+2s2fPttOnT7cXX3yx/f73vx/evl/84hcTjm+ya6+91gK2tLTUdnV1RV1mdHTU/sM//INdsmSJnT59up0zZ459//vfbw8ePBjeTgcOHAgvn8zfa8WKFVH/537xi1/Y5cuX2+rqaltXV2ff9KY32RdeeMF+//vfn7I/WWvt2Wefbc8+++yor/HFL37RAnb9+vUT7k+0n4qIiIiIiCQrF2IRIuJIRuKqxlqbybh2znvggQcswB133JHtoSQl15sBumGMYcWKFa5q8IqIiIiIiIhIehViLEKkwJlMvIhqNIuIiIiIiIiIiIhIShRoFhEREREREREREZGUKNAsIiIiIiIiIiIiIilRIR3JOaobLiIiIiIiIiIikl+U0SwiIiIiIiIiIiIiKVGgWURERERERERERERSokCziIiIiIiIiIiIiKREgWYRERERERERERERSYmaAYqIiIiIiIiISFI6By3dI5aABZ8J3m6uNNkelojkAAWaRUREREREREQkrpaeAOs6A+zqsxPuv3v7GItrDCubfSyr04XzIsVMM4CIiIiIiIiIiMS0vjvA2jb/lCBzyK4+y9o2P893BzI8MhHJJQo0i4iIiIiIiIhIVC09AR7Z7yd6iPkMCzy8309Lj4LNIsVKgWYREREREREREYlqXWcgYZA5xAJPdyrQLFKsFGgWEREREREREZEpOgdtzHIZsbT1WToHna0jIoVBzQBFRERERERERPLMfbvG4j5+1+LUQz5uy2C09ARorixJ+fVFJL8o0CwiIiIiIiIiksdOnT6TQTxzmvHseYf8mV1PRPKbAs0iIiIiIiIiInkmMmM5MrvZi0zmkAqXSclu1xOR/KYazSIiIiIiIiIiMsWyOndhI7friUh+054vIiIiIiIiIiJTNFcaFtc4K8WxpMbQXOld+Q4RyR8qnSEiIiIiIiIikudu+Np9wR+qYwR5v/0ZV8+7stnH2jY/NvGiGOCGZuU0ihQrV3u/MabKGHOBMeYKY8xrjTHzvB6YiIiIiIiIiIi40H3izFeKltX5uHlhCYlylA1wy8ISlc0QKWJJZzQbY84D/hZ4E/BaJgWpjTEngPXAU8A6a+3YlCcRERERERERERHPPf3pu4DxZoAf++qZB1xmMke6utFHQzk83RmgrW9qbvOSGsMNzT4FmUWKXMJAszHmUuBLwBsJBpdHgR1AF3ASqABmAa8CbgJuBI4bY+4HvmGtHUnP0EVEREREREREJBOW1QUDyZ2Dlvv3jBGw4DNw5/mlqsksIkCCQLMx5p+B9wE9wEPAT4A/W2uHYyx/DvBXwAeBLwO3G2P+xlr7Ww/HLCIiIiIiIiIiWdBcaWgsNxNui4hA4hrNbwY+Bcy11v5Pa+1vYwWZAay1B621D1prrwBeA2wGVng3XBERERERERERERHJNYlKZyy01g65eWJr7XbgBmNMhZv1Jbb7dp0pf72v3065/67FSZfeFhEREREREREREUlZ3Iik2yCz188hsdVPy/YIREREREREREREpNh5kvpqjCkDlgGD1tqXvHhOiU0ZyyIiIiIiIiIiIpJLEtVonsAY8x5jzBPGmPqI+84DWoGNwE5jzDpjjCKhIiIiIiIiIiIiIkXCUaAZ+FtgsbX2ZMR9/wAsAtYD24HrgQ+nOjBjzCxjzM3GmKeNMXuNMUPGmB5jzO+NMR8xxkQduzHmCmPML40xJ40xg8aY7caYTxhjSlIdk4iIiIiIiIhITtvbAYe74dDR4Pe9HdkekYgUCaeZxxcC/xm6YYypBd4GPGGtfe94CY2tBAPND6c4tpuAB4CjBIPY7cAcYCXwCPBWY8xN1tpwNzxjzPXAU8Aw8DhwEngH8E3gyvHnFBEREZEsi2xuHI1KhYmIiDizYEcLVzz1FLTtmvjAuz8FlyyB226C5RdlZ3AiUhScZjQ3EAz8hlxOMFj9EwBr7SjBQPR5HoxtN/BOoNla+z5r7WettX8LLAY6gHcTDDoD4aD3w4AfuMpa+xFr7V3Aa4D/Am40xrzXg3GJiIiIiIdOnbbhLxEREXHuot+s56Yvf4X5k4PMIZva4PZ74enfZHZgIlJUnKaK9AF1EbdXABb4fcR9w0BNiuPCWht19rPWdhljvgt8GbiKYAYzwI0EA+E/tNZujFh+2BjzeeA54A7Gg+IiIiIikj2RGcuR2c3KZBYREXHoxR28+aGH8dkEJ2wDFtZ8F+Y2KLNZRNLCaUbzHoIlK8qNMdMIlqLYbq19OWKZs4FurwYYw+j498hrLq8Z//6rKMv/FhgErjDGlKdzYCIiIiIiIiIiGfPgTxMHmUMCFh56Mr3jEZGi5TTQ/BCwkGDAuW385+9NWmY50Jr60KIzxpQCHxi/GRlUftX4992T17HWjgEHCGZwL0zX2EREREREREREMmZvB2xqw1HxqY071SBQRNKiZNWqVUkvvGrVqm2rV68uBy4DpgOPAmtCz2GMuQb4OPDwqlWr/uj1YAFWr179deBdwC+ttWsi7r8LmAXct2rVqpNR1nsfMB94bNWqVZ2THzfG3Lp69eoHr7vuurkAF1xwAW1tbcyePZstW7Ywd+5cNm7cyKxZs2htbaW8vJzDhw8zODjIwMAAR44cYdq0aezcuZOGhgY2b948YZ22tjbKyso4cuQIAwMDDA0N0dnZSXl5OS0tLcyZM2fCOjNnzuSll16ipKSErq4u+vr6GB4epqOjg4qKCrZv305TUxObNm0KrzNjxgx2796NMYbu7m56e3s5ffo0hw4dorKyku3btzN37lw2bNjAvHnz2LhxI3V1dezduxdrLS+//DKnTp1ibGyMgwcPUl1dzbZt28LPH3q92tpa9u3bRyAQ4OTJk5w4cYJAIMD+/fupqamZsM6cOXPYsmULNTU1HDhwgLGxMU6dOsXLLweT4Pfu3cuMGTPYunVreJ3Gxka2bdtGVVUVBw8eZHR0lJ6eHrq7uzHGsHv3bmbOnDlhnYaGBrZv305lZSXt7e2MjIzQ29vLsWPHKCkpYdeuXcyaNYvNmzeHf/fZs2fT0tJCeXk5nZ2dDA8P09/fz9GjRyktLQ1v/9A6GzZsYPbs2ezcuTO8LSdv/9bWVhobGydsy/r6etra2igtLQ1vy8nbf/K2DG1/n8/HsWPH6O3tZWRkhPb29vD2P+uss2Ju/+PHj/PKK68wOjo6Zftv3Lgx/L22tnbK9vf7/ezfv5/q6mq2bt0a/ns1NTWxefPmuNu/trY2vF02bNhAU1MTW7dupaqqKur237NnT8ztX1lZyaFDhzh9+vSU7V9fXz9hv2xoaGDHjh1Mnz6d9vZ2hoeH6evro6ura8L2j1wncvuH9uX+/n6OHDlCWVkZbW1tUfflyO0fuS9Hbv9NmzaF/2719fXs2rUr5r68Y8eOCdt/w4YNUbd/aF8Obf/J2zK0LwMTtn+0fTn0vxPa/oFAgBMnTkzY/pP35dD2j9yXI7f/vn37qKurm7At58yZE97+kfvy8ePHJ2z/LVu2hPexxsbGmPuyz+ebsv1D64S2f0dHR8x9WXO55nLN5bk7l2/YsJFA7WzKOltZPKtCc7nmcs3lmss1l+fhXK7j8uzM5dN+s4HyjW0Yh7GVV2ZWcri+0vVcvqPtJQJVMyk7tI3Lzp+nuVxzuebyHJ/LzzrrrNUOpwlXjE328opknixYTqMCGBjPIvaUMebjwP3ALuBKa+3JiMd2A+cD51tr90ZZ948Emxdebq39U6zXeOCBByzAHXfc4fHoRURERCQa1WgWERFx6aGn4DsuWlF99L1w67tdv6zeu0XyjtPzUa54OhtYa08Dp718zhBjzEcJBpl3AtdGBpnH9Yx/ryO62knLiYiIiIiIiIjkr6qKzK4nIhJH3BrNxpiUZx6PnuMTwD8CLcDV1tquKIu9NP79gijrlwLnEmweuD/V8YiIiIiIiIiIZN3yiwCc1WiOWE9ExEuJmgEeMMbcaYwpd/rExpiLjTE/A/7e3dDCz/Np4JvAVoJB5u4Yi/5m/Ptbojz2BqAS+KO1diSV8YiIiIiIiIiI5IRF8+GSJc6uib/0wuB6IiIeSxRo/jXwDeCoMeYBY8zV8TKUjTELjTF3GGP+C9gMXAysdzs4Y8w9wFeBTQTLZbwcZ/EngZeB9xpjLo14junAl8ZvPuB2LCIiIiIiIiIiOee2mwiYJEPNPgO33pje8YhI0Ypbo9la+wFjzLeArwC3jn/5jTFtwFHgFDAdmAW8CphNsLj0MeBzwDfdZhAbYz4IrAH8wO+Aj5upE+dBa+0Pxsfaa4y5hWDA+XljzE+Ak8A7x8f2JPC4m7GIiIiIiIiIiOSk5Rfx61tv4c0PPYzPximi4TPwhdtVNkNE0iZhM0Br7UbgzcaY84GPANcCrwEmz0zHgXXAU8BT1trRFMd27vj3EuATMZZ5AfhBxFifMcasIBjkfjfBIPhe4FPAt6yNN+OKiIiIiIiIiOSfHddcTU9DA5evW8eCnW1TF7j0wmAms4LMIpJGCQPNIdbaPcBnAIwxlcA8gpnMQ0C3tfaolwOz1q4CVrlY7w/A27wci4iIiIiIiIhIIvftGov7+F2Lkw7DONZ+0TLaL1rGXaVH4e5vwpgfSkvg659UTWYRyQhXM5y1dhDYM/4lIiIiIiIiIiIRTp0+c1H1zGmO2vWlZtF8mNc48baISAak71SaiIiIiIiIiEgRicxYjsxuTmcms4hIrvBlewAiIiIiIiIiIiIikt8UaBYRERERERERERGRlCjQLCIiIiIiIiIiIiIpUZEgEREREREREZE81Tlo6R6xBCz4TPB2c7YHJSJFSYFmEREREREREZE809ITYF1ngF19dsL9d28fY1WfpakCaktNlkYnIsXIUaDZGLMAeMVa2xtnmRpgprW2PdXBiYiIiIiIiIjIROu7Azyy34+N8Xj/mGVvHyyogtkZHZmIFDOnNZoPAHcmWObj48uJiIiIiIiIiIiHWnriB5kjtQ9YWnoCaR+TiAg4L51hxr9ERERERERERCTD1nUGogaZ//4f7wv/fN7BfeGfj935Nagx8O3PZGB0IlLM0lGjeQ4wkIbnFREREREREREpWp2DdkpN5mhOzqgP/9w/ZhnyQ0U6ByYiQhKBZmPMBybd9Zoo9wGUAAuAvwF2eDA2EREREREREREZF68Mxv/5u7tiPvaBs328JR0DEhGJkExG8w8gfFWGBa4f/5osVFJjEFid8shERERERERERCRsyJ/Z9UREnEgm0Pzh8e8G+B7wDPCzKMv5gRPAf1lrX/FmeCIiIiIiIiIiAlBRktn1REScSBhottb+c+hnY8wHgWestT9M66hERERERERERGSCZXU+IHb5jPjriYikl6NmgNbaq9M1EBEREREREYD7do3FffyuxenoaS4i4p3OQUv3iCVgwWeCt5srTeIVE2iuNCyuMUk1BAxZUmM8eW0RkUQcHaEZY/4CeDvwoLX2WJTHm4BbgZ9ba7d6M0QRERERESlWp06fCabMnKZAiYjktpaeAOs6A1MCwXdvH2NxjWFlsy/l7OKVzT7WtvlJJtRsgBualc3sBZ0EFUnM6V7w98DrgXtjPH4M+AiwCPhACuMSEREREZEiFflhPfKDvT7Ei0guW98d4JH9sQPAu/osa9v83LIQrmp0H/xdVufj5oXEfa2Q+RXw7NEAzx6dWG5D82lqdBJUJDqnM8vlwHprbdS5zFprjTG/Ad6Q8shERERERERERPJAS0/8IHOIBR7e72d2eWp1k69u9NFQDk93BmiLUkajugSapkNNmVFQ1CM6CSqSmNO9oQnoTLDMEeAsd8MREREREREREckv6zoDSZWygGCw+enOQMolNJbVBctwdA5a7t8zFq4Hfef5pRNqMisoKiKZ4nSGGQQaEizTAIy4G46IiIiIFJN0NUsSERHJlM5B66g5H0Bbn/W0QWBjuZlwW0QkG5wGmrcC1xtjPmWt7Z/8oDGmFrh+fDkRERERkagy0SxJREQkE1p6AokXirFec2WJx6MREckep4Hmh4AfA/9pjLnNWrs99IAx5mLgQWD2+HIiIiIiIlNkqlmSiIhIJgz5M7ueiEhIZGmcaDJdLsfRq1lrHzfGvBX4ALDFGHMMOAzMA+YABvhna+2PPR+piIiIiOS9TDdLEhERSbcKl0nJbtcTEYkmFxp/Og5rW2s/ZIz5I/AxYCnBBoEALcC3rLWPeDg+ERERESkg2WiWJCIikk7B9ynn5TP0/iYiqYrMWM6Fxp+uZjVr7UPW2ouAaqAZqLbWvlpBZhERERGJJZVmSSIiIrmqudKwuMZZ9uCSGqOmfSJScFI6fWatHbTWHrHWDno1IBEREREpTKk0SxIREcllK5t9JBs2NsANzcpmFpHC42pmM8Y0GGNuN8bcb4x5ZNL9rzPGVHg3RBEREREpBGqWJCIihWpZnY+bF5YkDDYb4JaFJSqbISIFyXHBDmPMR4BvAdMJzpEWuHn84TnAfwG3Ao96NEYRERERKQBqliQiIoXs6kYfDeXB/gJtUUpFLakx3NDsU5BZRAqWo0CzMeZNwEPAduCLwF8Bt4cet9a2GGNagXehQLOIiIiIRFCzJBERKXTL6oKB5M5By/17xghY8Bm48/xS1WQWkYLnNKP508BRYIW1ttcY89ooy2wHLk95ZCIiIiJSUELNkpw0BFSzJBERyUfNlYbGcjPhtohIoXOaHnIp8G/W2t44y3QCTe6HJCIiIiKFSs2SREREREQKk9Mj92nAQIJlZgBq2SIiIiIiU6hZkoiIiIhIYXJaOuMgcEmCZZYDL7kajYiIiIgUPDVLEhEREREpPE4DzT8D7jbG3GSt/enkB40xHwZeDXzOi8GJiIiISGFSsyQRERERkcLiNND8deC9wI+NMTcCdQDGmL8D/hJYCewBvu3lIEVERESkMKlZkoiIiIhIYXAUaLbWnjLGrAB+CNwU8dC3xr//Dvgf1tpEdZxFRERERERERCSP7eu33LdrLObjdy12mt8ouSjeNobi2M76GyTH8V/BWtsOXGWMeTVwOTAL6AH+ZK3d5PH4REREREREREQkx506fabvwsxpukKpUGVtO3/sq/Ef//ZnMjMO9L8eT9xAszFmHfATa+0T47ffABy01rZba7cD2zMwRhERERERERERyTHnVZtwJmdkxqeyOwtL5PbMie3cfeLMz42zMvKSOfc3yFGJ/hrvArZG3F4PrAbWpG1EIiIiIiIiIiIiIiGRGcuR2c0ZzGSWxHwJHn8FqI24rXxwEREREREREREREZkgUUZzG/DXxpgNwNHx+84ZL6ERl7X2t6kOTkREREREilfnoKV7xBKw4DPB282Vyn0REUmW5lERyaREgebVwDPAv0bc98Hxr0RK3A5KRERERESKV0tPgHWdAXb12Qn33719jMU1hpXNPpbVJbo4U0SkeGkeTR8F77Nsbwcc7oYxP5SWBG8vmp/tUcm4uIFma+2vjTFLgDcC84BVwAvjXyIiIiIiIp5a3x3gkf1+bIzHd/VZ1rb5uWUhXNWoIImIyGSaR9NDwfsse3EHPPhT2NQ28f53fwouWQK33QTLL8rO2CQsYWtEa+0h4FEAY8wq4HlrrZoBioiIiIiIp1p64gdHQizw8H4/s8vRh3oRkQh9o1bzaBooeJ9l656Dex+EQIwtsKkNbr8XvnA73HBNZscmE8QNNBtjvgH8ylr76/G7PgxsSfuoRERERESk6KzrDCQMjoRY4OnOgAIkIpJT7ts1Fv55X7+dcv9dixPm+6WkaxjNox7TSdAse3FH/CBzSMDCmu/C3AZlNmdRov/8TwCXRdz+HvCu9A1HRERERESKUeegnXI5ciJtfZbOQWfriIhkSv20M1+ZMOS39PudraN5NDE3J0HFQw/+NHGQOSRg4aEn0zseiSvRqbR+oDLitqqbi4iIiIiI51p63H0wb+kJ0FypPuQikhvSnbEcT99Y4mWi0TwaWyonQdUg0AN7O6bWZE5k4041CMyiRDPgXmClMeZp4Oj4fTOMMQsSPbG1tj3VwYmIiIiISHEYcpiFl+p6IiKFJtmkz8k0j8amk6BZ9uIO9+sVWaC5c9DSPWIJWPAZsnayI1Gg+T7gX4A/Rtx35/hXPDaJ507IGHMjsAJ4DXAxUAM8Zq19f5RlzwEOxHm6x6217011TCIiIiIi4r0Kl5/H3a4n+Smy/m002cwmFck2n8uYkhfzaK4Eubymk6BZNjCU2fWSkGv/6y09AdZ1BqZk3t+9fYzFNYaVzb6M1gyP+y5srf2xMeYAcB0wD/gQsB3Ymv6hAfB5ggHmfqATWJzEOtuAZ6Lc3+LhuERERERExEPBD0HOM8fyveGSAqfunTp95kP1zGn5H9ASSVWNy+kilXm0b9SypnUsZ4JcXtNJ0CyrqsjsenHkWkAXYH13/EaVu/osa9v83LIQrmrMzNgSTkPW2j8BfwIwxnwIeNpauybN4wr5JMEA816Cmc3rk1hnq7V2VToHJSIiIiIi3mquNCyuMY5qYS6pMQWRMReiwGlikYH3yCC9AvIiUFFiqC5x1hAwlXn0xIilfQiIEebKRpDLa8V6EjRnLL8os+vFkIsB3Zae+GMKscDD+/08tN9/7b9eVvZcusfl9N34w8CWdAwkGmttOLBsjA60REREREQK2cpmH2vbEn9ogmCX8hua8/+DvAKnIpKK0Lyxrz84czqp05zKPNo3GgoyxxcKcs0uz8/gq06CZtmi+XDJEmcNAS+90NP6zE4Dupn6X1/XGUjqeAnCp4LuAXIr0Gyt/ed0DcRDc40xtwGzgBPAf1lrt2d5TCIiIiIiksCyOh83LyThBzoD3LKwJK+CFiqRISLpNjqeeDtrGpw4HX/ZVOfRruHkl7XA052BvJqzIxXjSdBYslKf+Lab4PZ7kzuL4jNw642evrzTgG4m/tc7B62jkx/jVqRjLJM5PpoxxviAjwLvA5YAVdba0vHHXgvcAvxfa+1uLwfqwJvGv8KMMc8DH7TWtmdlRCIiIiIikpSrG300lAc/qLVF+RC1pMZwQ57X/FSJDBHxSugk1X27xsJZzQsqDbed50vbPNo56Kw8B0Bbn8160zS3CvkkaLKyWp94+UVwz21w74Pxg80+A1+43dOyGW4Cupn4X2/pcV7OJVNKVq1alfTCxphpwH8AtwMVwCBQs2rVqjUAq1evHgYeA3pXrVr1vJcDXb169TkEmxHuWLVq1booj08DhoFPAZ8BvgY8C5wDXA28Y/Xq1d9btWrVaLTnN8bcunr16gevu+66uQAXXHABbW1tzJ49my1btjB37lw2btzIrFmzaG1tpby8nMOHDzM4OMjAwABHjhxh2rRp7Ny5k4aGBjZv3jxhnba2NsrKyjhy5AgDAwMMDQ3R2dlJeXk5LS0tzJkzZ8I6M2fO5KWXXqKkpISuri76+voYHh6mo6ODiooKtm/fTlNTE5s2bQqvM2PGDHbv3o0xhu7ubnp7ezl9+jSHDh2isrKS7du3M3fuXDZs2MC8efPYuHEjdXV17N27F2stL7/8MqdOnWJsbIyDBw9SXV3Ntm3bws8fer3a2lr27dtHIBDg5MmTnDhxgkAgwP79+6mpqZmwzpw5c9iyZQs1NTUcOHCAsbExTp06xcsvvwzA3r17mTFjBlu3bg2v09jYyLZt26iqquLgwYOMjo7S09NDd3c3xhh2797NzJkzJ6zT0NDA9u3bqayspL29nZGREXp7ezl27BglJSXs2rWLWbNmsXnz5vDvPnv2bFpaWigvL6ezs5Ph4WH6+/s5evQopaWl4e0fWmfDhg3Mnj2bnTt3hrfl5O3f2tpKY2PjhG1ZX19PW1sbpaWl4W05eftP3pah7e/z+Th27Bi9vb2MjIzQ3t4e3v5nnXVWzO1//PhxXnnlFUZHR6ds/40bN4a/19bWTtn+fr+f/fv3U11dzdatW8N/r6amJjZv3hx3+9fW1oa3y4YNG2hqamLr1q1UVVVF3f579uyJuf0rKys5dOgQp0+fnrL96+vrJ+yXDQ0N7Nixg+nTp9Pe3s7w8DB9fX10dXVN2P6R60Ru/9C+3N/fz5EjRygrK6OtrS3qvhy5/SP35cjtv2nTpvDfrb6+nl27dsXcl3fs2DFh+2/YsCHq9g/ty6HtP3lbhvZlYML2j7Yvh/53Qts/EAhw4sSJCdt/8r4c2v6R+3Lk9t+3bx91dXUTtuWcOXPC2z9yXz5+/PiE7b9ly5bwPtbY2BhzX/b5fFO2f2id0Pbv6OiIuS9rLtdcrrk8d+fyDRs2EqidTVlnK4tnVWgu11xOrR3hLyoGWezvpuu0j7kndjN9xizeM7adD/3FfNpbNuXdXH5p7RjNAx28aUE1L7VsZ9rsudS2b+Z/Xb6A8oObqaurY8uuvWAtvr4TNPlf0VyeYC7fvGMntrSMkleOMr9kSHO5jssLei5/rLeRDRs28rvRBjZs2szv+6t4sW0ff+we5U+HTuAbOMnAaIAZJ/czvaaW+d07+OvXNDOzYzPHpjfQ3L2DqqoqPlDVyTWz/JQP9zqeyx9uOcWf9nfxwglLVfceBqfPZP6xbfTUnMWCo5vpq5zN3OOtjJVOZ2bfYcrGhikf7ae2/xgBXwllR1/iVXPz87i8cuA455f0MjJyGt/L7Zwuq2Bed0v4d587dy5vHtzKlQtmFtxx+a9aDvHj/UP09/VSO9iNNSXMObGbgYp65h/biF6YIQAAIABJREFUxr5pZ3FoxyZM7Wz6DrSm57h86BRzr7uGvt37KT/ROyWmd/rVizh88zvxvfkKT+fy57a+RMcQ1Awep2Kkl5LAKPU97YyUVdLcvSO8/UPfh8prmX1qP+UmQMXgybTN5V1UcbLzIL7AGJXDr1A9eAKAhlP7GCqvo/nY9jP7ZVUD87p30FvdxLubS1bHirt6xVibfGTeGPM54F5gFfAl4AvAPdbakohlfg3UWmsv83SgxlxFsBngY9ba9ztYrxT4PbAc+IS19v54yz/wwAMW4I477nA/WBERERFJmurSSjyF+P8R63cqxN81nfT3kmIS+f8e66qIUEbzedUmLXNL6Hm6hi1HHZTOCLmp2ccNzSWJF8xxnYOW+/eMhctH3Hl+aV5maiejpSfgqGzIZ5dkIKN7bwfc/U0Y80NpCXz9k57WZI70dKefn3Y6zx5O9//6r476+eEh5+P618vK0v6P6nSGeR/wB2vtGgBjTLT/tQPAO1IdmFestWPGmEcIBprfAMQNNIuIiIiIiIiI5JJEgeNEdeC9HIPbIFdF/seYgWCDwMZyM+F2ocrF+sQsmg/zGifeThO3/7Pp/l8P/o1zs3yG00DzucC/J1jmJFDvbjhpc3z8e1VWRyEiIpKj1KRKRERERJLhNshViLWLC1mu1ifOpFz9X2+uNCyuMU63zwvpGk8kp58ah4AZCZZZALzibjhpEyrjsT+roxAREckDalIlIiIiIrG4CXItqTGOg49KhMgutw3nWnoCNFcWRvp6pv7X3VjZ7HNU1sQGSyGnndMQ+1bgzeNNAacwxtQBfwX8OdWBOWWMWR5tXMaYa4BPjt/8l8yOSkREJD/ctbg0/DVzmgl/he4TEREREQlZ2ewj2VCaAW5oTi3D89RpG/6SzBjyZ3a9XJXp//VkLavzcfPCkoRjM8AtC0v418vKnsvEuJx+cnwYeAx4zBjzkcgHjDEzgO8DM4HvejE4Y8y7gHeN32wa/365MeYH4z+/bK39+/GfvwYsNcY8D3SO3/dq4Jrxn++x1v7Ri3GJiIiIiIiIiGTTvn4bzvrd128Z9E+8P53JAs8eDTC/AtqH4i8XCnK5KSWgZqnZlXP1iT/21TM/t+ydev+3P5OWlw0GdOGR/fGzh1P5X3fr6kYfDeXB2thtUbKul9QYbmj2ZXRMjvZOa+2PjTFvBD4MvBM4BWCM2QgsBcqB71hrf+nR+F4DfHDSfQvHvwAOAaFA84+AG4D/BrwVKAOOAU8A/2it/Z1HYxIREREREZEs6hy0dI9YAhZ8hoKqCSrixmgAAhlO9p1Vbpjms3QMwUiUKgvVJfDxCzIbeBPv5Gp9YgAaZ6b/NSLkYkA3ZFld8HU7By337xkLvy/eeX5pVt4XHZ8GstZ+xBjzO+BOghnDBvgLoBX4hrX2+14Nzlq7CliV5LKPAo969doiIiIiIiKSW1p6AqzrDEypl3n39jEW1xhWZumDvkg2nFdtwtm9t28cZdROvT9dJmcbD/ktBwagfhqcPA3nVkFFidH+mMdyrj5xmjKWk5VrAd3JmisNjeVmwu1scLXHW2t/YK19LVANNAM11tqLvAwyi4iIiIiIFItQhm7XcPB756DqkE62vjvA2jZ/zKDHrj7L2jY/z3e7a2AlIu5VlBjKfdA0Pfi9oiT7gTdJXa7WJ86mUEC3aXrwey4EmXNJSqe4rLVDQIKKPCIiIiIiIhKNMnST09ITSFgfE8ACD+/3M7s8Q5dvi0jOiaznHI1qPCcvl+sTS27Sf4CIiIiIiEgWnBixytBN0rrOQMIgc4glWEdTpJj4bbBGc8CiqyIinDptw1/iztWNPj67pIQlNdEzd5fUGD67pISrGhVilBQzmkVERERERCSxUIZdqI5pwBKupxqPMnSDZUWc1AgFaOuzahAoBWlyI8z1x/z87mXLgP/MMoeHJl4VUWwm14+Odr84k+v1iSV3aC8TERERERFJs75RS9cw9PsTLztZKEO3WAPNLT3uspNbegI0V5Z4PBqR7IhVZufhA7H3j9BVEfMrYFa5goGSulxpOCe5S4FmERERERGRNFrfHWDfAEmXfoimmDN0h1wE51NZTyTXrO+OXaP8hZ/+Mu66K256G+1DMM1nqSkrvvlDJBWx6n3v6w/ujedVa5+aTIFmERERERGRNEm2iV2yz1WMGboVLn9lt+uJ5BInc8hQ/0D454rqqgmPdQ1DTZnHg8uQyeVCivWkm2SX6nwnR4FmERERCdOBvIiIt5w0sUukWDN0gyVDnJfPKNZSI1JYEs0hK256W/jnyOzmyPshWLZnyJ9fgbJY5UIi609rP5d0ilXve+a0bIwmP3gSaDbGzAb+EhgE/p+1tkgPgURERPKTDuRFRLznpoldPMWaodtcaVhcYxz9LZfUGJ0olbzn9RzSF70KQE6KVy4EztSfvmUhXNWoY1SRXOEo0GyMuQP4EPBWa+3J8fsuAX4F1I8vttEYc421diD6s4iIiEgu0YG8iEh6uG1iF0sxn/Bb2exjbVty5QMMcENz8f6tpHB4PYcE8iShOdlyIRZ4eL+f2eXFPT9KZs3q6ODsllamDQ5yurISSi+GRfOzPayc4TSj+b8DNhRkHncfMBP4PjAHuA64HfgHT0YoIiIiaaMDeRGR9PGy1EWxZ+guq/Nx80ISvmcZ4JaFJQXxXhWrCVVI5CXdUpi8Lpfjy8AUMuS3jASga9i6LsPmpOSQBZ7uDBTEPi+5bcGOFq546inmt+2a+MAPgEuWwG03wfKLsjG0nOJ0Tzwf2B66MV4yYwXwqLX2ZmvtO4ANwP/wbogiIiKSLm4O5EVEJDlelbpQhm7Q1Y0+PrukhCU10YNWS2oMn11SUpBX35w6bcNfUjy8LpdTk8ZzE32jloEx2NUHIwE4OgyHh4Jl2Na0jiWdne2mXEhbn6VzcOo6od4jXcPB79GWEUnKuud4z5e+zPy2XdE/O21qg9vvhad/k+mR5Ryn08wsoDvi9pXj35+OuO93BMtriIiIiMe8zG5K5UC+mLPqpDApczCziuXv7baJXaRCytD1wrK6YM+AzkHL/XvGws1r7zy/tODem2I1oSqU/SMfZHuu8mIOCakugYqS9OwjJ0Ys7UOxH3dShs1tuZCWngDNlSXhn9V7RDzz4g7svQ8S2nti7kUBC2u+C3Mbijqz2emedRKYHXF7BcFZ748R91lgeorjEhERkQRSzW5K5UBepJApczCzCvnvHWpi51YhZ+imqrnS0FhuaJoe/F5oQWbJPdmYq1KdQyI1pSlKs/6YP26QOSRUhi3RcaTbciGh9dZ3B1jb5o+ZTBEKej/freNZSaylJ8ChbzyBSbbAecDCQ0+md1A5zunptzbgHcaYzwF+gjWbN1hreyOWOQfo8mZ4IiIiEsnL7KZUD+RFCokyBzOrmP7eTprYAdSXQUVpYWboSmHIdpZvJuXCXOV0Doml1OPzVbGyhuNJpp6y23IhFSXqPSLeWt8d4NnfH+Lru4LlMpJ+R964E/Z2FG2DQKd71P3AWUAn0AE0Af8UetAYUwK8Htjm1QBFREQkPVI5kBcRkeQ5CRDVT0MZupI3CvmKhFwRbIRZknyQK4a++OcHHEmUNRxPrHrKIW4Dv8vqfOo9Ip4JnbRYuqsFcBBkDnlxh+djyheOTsNZa39ujLkduHX8rsestf8SscgbCZbN+A+PxiciIiJp4rbunzI/xAuRmWH7+u2U+wspI06K2zqHgYyuYagpS9NgRDyQC1m+xebqRh8N5cHAaJuL4C4Er+j3QrJZw4meI1RPebJQuRAnQexQg1D1HhGvhE5aVAwlURcmmgGX6xUAx+8E1tqHgIdiPPYfwMxUByUiIiLp5/ZAXgfj4rX6adkegUh6uGm62u+HIb+yQ0VkolAjzJ+0+/n5EeeJAj6PDt+cZA3HkqgMm5NyIQa4odnnSRNBEZj43j1UUeHuSapcrlcAdMpRRESkiLk5kBfxgjLfpBi4DXx4eYm7iKSuc9DSPWIJ2GDANptZsK+f7YsaaO45fpL+V3qxNoAxPnqOn6SuoT78eI0Hb7tuTp5Fk6gMW7BcCAkzpw1wy8ISltX52NPnromIeo/IZJHv3S2LlwE4q9EMsPwiT8eUTxxNNcaY1wKXEyyZ0TN+XxXBOs3XA4PA16y193s9UBEREfGemwN5ERFJjtsAhleXuItIamI1vLt7+xiLawwrm30ZPzZqrjRUlwSvfgDoOthJyx82crzj6ITlfvno4zTMP4tlV17KovOaqShJPTDu9uTZZMn8zRKVC1lSY7gh4u+fyd4j9+0ai1r2K0Qn0/Nb5Hv34bnNtJ2/mCV7diX/BJdeWLSNAMF5RvOngb+01v5TxH1rgb8B+oFZwDeMMW3W2l97NEYRERFJI6cH8iIikhy3gQ+vLnEXEffWd8evRbyrz/KVNj8LKvzMKp+606Yz2Ng0HfYOwL5tbfz5Vy+AjT7K4x1HWf/4v1F9/QrOv+zClF/Xi+zf6hL4cfvEJ4r1twqVC+kctNy/ZyycUX7n+aVTMsqz1XtkNEC4GebMaZq8C8Hk9+51163ks/evxRdjP5vAZ+DWG9MzsDzhdOa7FHg+dMMYUwZ8EPgzcBVQD2wBPg4o0CwiIpInnBzIi4hIctwGPry4xF1E3HPS8K59CEYCNhycykSwsabMEDjaETfIHGYt//azF2iaXcv5i5pTel23J88iVZU6D8w2VxoaI4L50Y5NM9l75K7FpeEs5n39Nvx7KJO5MEx+725dsozfLX89b/jT7zBMLaMRum0B8/YVRV02A8DpqZtGoCPi9qVADfCgtXbYWnsE+Bnwao/GJyIiIhkUOpBvmh78riCziIh7ocCHE9UleHKJu4i457Th3cBYMGg6c5rhrsWlGQk4/vm3mxIHmcdZa3nuuY0pv2Yq2b8GuHVhCXMrTNr+ViubfUnX0VXvEYll8nv30rYW/vLF34f/tyb/j024/99egBd3pH+QOczpHm0nrfP68fteiLjvONCQ4rhEREQkjlxqSiPpM7nm32TKnMlt2n6JFcNc5qTpKgQviReJlIn9RPPVGW4a3vX7YchvM3aSqOvYSQ4cOOJonf0HjtB17CQsbnT9um6yhmFiGbYNJ5O/yuMdX/p5+OfImsi/rQ7+nX/x+XdOWF69R8Qrke/dK/99XXJlMyDYZOGhJ4s6q9npu0U7cFnE7euBTmvt/oj75gKnUh2YiIiITJWLTWkkM0KXmYJqAOYjbb+JCm0uiwyGRPOpW9+eVOBjfkXwkngRyN5+UuzzlduGd31j3pSWSMbevZ3u11vhPtAMzk+e3bLQx9WNqf9henoHztyoro65nHqPiBdCJy2e/f0hluzZNaVcRlwbd8LejqJtCOg00PwEsNoY8yQwDFwO/N9JyywD9nkwNhEREYmQTFOatW1+blkIVzU6O3iOzGSK1kW7mDKZcknk3z1yG2l7pI+XmYPaftGlcy7LBYdPnAmGzJtVBSQf+Hj2qLsAlxSeTO8nbuarQr0iwW3Du4CzJN+UjAyfzuh6kZLNGgZYUEFKQebIjOULP/OzqPfHGqN6j0iqrm708apTOwEHQeaQF3co0JykbwJvAVaO394KrAk9aIy5ELgE+IonoxMREREg+aY0Fnh4v5/Z5e7r6I3mQJxDl/BKphVahm2ucjOX5YPIoEdkdnPk/ZGBj3tazsxx9y47E/hQoFkgs+/5kyUTPC70+dJtVrIvg/HL8unTMrpepNAx2nlV0DUcLBsyWXVJsAxQtq/QSKaJoEg8c/3D7lYcGPJ2IHHkWsKQo1ez1vYDVxpjlo3ftdNaG3k0NAjcAKReZV5ERETCnDSlsQSz5px8yIs8ALl942jU+7Ol2C/hlfQr9AzbXOJmLpueoUvRM6W50lDum3hbJFK63/OjSTZ4XAzzZfBv6fykT00GD5kWLWrO6HrR1JQZxqylNgDHR84E2u9dVsqP26OnhRdaFvyQ3zISgK5hWxC/j0RRVZHZ9VJUn/q5pJS5mgqttS0x7j8IHExhPCIiIjKJm6Y0bX2WNa2jVJSYnAgWO6WSA5Ip2cwcLDZu57LFNWSswZZItrndT1IJcCUTPP5Km5/GaX66k6i8kO/zpZuGd9MM9IxC31hq2yJZTXPqOffcuY4aAi48dy5Nc+pTfu1ox2j9Y5bzxhv0RfvdCy0LvqUnwJ4+G87mPjqe9Jqvv08+CV01FJm5G+nDH7oO8PAzg9umfhlsBphrn49c/ecbY8qMMW8xxnzSGHNPxP3TjTGNxhjtUSIiIh5JpSmNiMTnJnNQ3NFcJpKY2/0klfWSOdkGJBVkDsn3+XJls89RTdbTNhhsPDwUDDauaR1zvU2S9cZrL8WY5EZpjGHJFZewr99y366xhCXKvLS+O8DaNn/MwH0oC/757vz4fwn9PtFKhkD+/T75rKd3gJ7eAU72DDAaSFP5v0Xz4ZIlzta59MKirc8MLjKajTFvAR4FmgjWw7bAveMPvwb4A/B+4McejVFERKSouW1Kc8UsHzc05+Y156rBLLkgG5mDxSwfGmxJ7su1WpRec7ufuF3Pyck2p/J5vnTS8C6aTJQQOX9RM+9euYJ1614gYGOP0hjDW69fwatf5V3ZjGT1jdqCumpIV0FlX7SeCPv6rfeZzJFuuwluvze5AxKfgVtv9H4MecTRFjDGXAo8A7wMfBJ4HfDXocettX8yxhwgWKdZgWYREREPuG1K43a9TFMNZsmWVDIAmyvzZAfLIfnQYMuNyOZ/fUOn2bTvONZajDFc9bknef7Lxf2BM51yoRal1zL5nu/mZJtTuTxfJnPSu6E8mJnd5uLvlIlg4+v+24Xc/uqZ3PvEn3mh9fCUx1csncc973kd116cnezKrmEyXm88nbJRP11ywPKL4J7bCKx5EJ+1WIh+xYPPwBduz2jZjFzkNNR/D8GGf5daa7uMMV+MsswG4C9SHpmIiIgA7pvS5PKBrWowSy7IdOZgscuHBltuvdw7xO4jr3Cyf2TC/S+0HmHF557kC+9ZnrVAT6Ep9PeJTL7np7u0A+TPfBnrpPeyumCt3c5By/17xghYOHEaTif5p8tEsPHai+dz7cXzec+zx/j1U78mEAhQWerjuc+/jaULZqXtdSMb+40Ego3xIuvpD/ltzPISseRyFryugipyK6/lp/5ZXL5uHQt2tk19/NILg5nMRR5kBueB5iuBZ6y1XXGW6QCucz8kERERieSmKU11SfRmLCJyRqFfLZBr3M5lR4YglBOXi+UR3rX8PG79p9/EvHT9t61HePOqZ3j4o9fwt29cmuHRSb5xs58sqTGu3vMzEQTO5fnSyUnv5kpDY7lhyG/Djd+SlalgY31DPdUzagGoLSVtQeZYjf129UF1iQ2fwHBbXz9aFnxr+wlOnuwlEAjg8/lobT+R1iB6rHG5XS9Xs/rFmfaLltF+0TJmdXRwdksr0wYHOV1ZybXvuLioazJP5vQIrZpg2Yx4KnHZZFBERESiW9nsY21b8nUCm6andTiSYZFZQz6DsmM8kqnMQW2/M9zMZWM5XKP5uW0dcYPMIQFrueU7v+HshlplNktCTvYTA9zQ7O7jdyaCwLl8dZUbXgZPvdQ5aBkJEK5T4U/TvLm+O36N4n4/rG3zM7/CfX39yBMgz23rYM0TL/Lb1iMTlln28cd4w9K5Gb1aRFdBSciJ+fM5Mf/M/921i3Ln5HcucPrXOAwkOg3/GmC/u+GIiIhINE6a0iyogJoyd0Gs8AeViNvFGhDLBbGyhu7ePsbiGsPKZl/BfYjPpHRnDmr7TZXsXGaAWxaWpK2JllfWPPFiwiBzSMBa7n3izwo0S0JO9xO384jbk23JcptpnWlOTgZ6ETx1KzLrekePDY/nb/40SujpQ8PrH4M1rWOevs84aYTXPgSzXdZQD50AefQ/W3PqahFdBVWcYjWgDTmvOvfnuExzGmh+FrjdGPN6a+3vJz9ojHkrcAXwVS8GJyIiImdc3eiL25RmYSX0jsGohe4RZ5dpKiCWexJlDWWio30xSFfmoLZfbInmsiU1hhvyYM5pbT8xJcsukRdaD9PafgKoTc+gpGBkYj9xc7ItWalkWmeKm2Mft81JvQ42lpng1R7xThN4/T7jpBEewIDL7O9ldb6cvFqkEHumOBEr4JqLJa3SJbIB7cnT2RtHrnP6n7AWeC/wa2PMt4FzAIwx1wFvAD4KHAW+4eEYRUREZFy0pjSnA1BbBvsHJi6bbJBYAbHc4yRrKN0d7QtdOjIHtf0SizaX+QzceX5pXmRAAjy3vcP9ek2q1SyJZWI/cXqyrWEadCcIsKSaaZ0Jbo993DYn9eJvERnIa+kJsLYtcZq0V+8zbhrhDQWCAXYn2dyhLPj35eDVIpmsn57r6l1mq+ejWAH0yMC7TORomrTWHjbGvBl4Argr4qGfE3w/2QestNYmquMsIiIiKQg1pTkxYjl6OtgBPZpdfZavtPlZUOFnVrmZcrCkgFhy3vGln8d9/Beff6enr+ckaygTHe0LndeZg9p+yQvNZZG380XvoLt0JrfrSfFK537i9GTbhpMBasssXcPBeryT5cMVCakc+1SUGKpLbNTfPZZ0BBsz/T7jthFeZQkM+3F01VAqV4uku0Fgpuqn56Jcy1hubT/BgWO9jPkDvDJm6Dp2kqY59dkeluA8oxlr7WZjzKuA64DLgVlAD/An4GfWWoX1RUREMqBv1NI+lNyy7UMwzTf1sDjbAbF8bJJ2+MSZ1PF5s6o8f343WUOZ6mhfyLzKHNT2Kx61le5SutyuJ5IuyZ5sm1Fm+Pejwffsumlw6/wSHu/w590VCake+zRNh30DzoKnXsrG+4zbGtPTfHDzwhJHVw1963furxZJd6A5U/XTJbZYDSK/8c2f0Hz2Wcx/33Le+7oFWRqdgItAM4C11k8wizl+eo+IiIikTdews+XLJhUWzGZALNm6iLkSiI7MWI7MbvY6kxncZw2lu6N9sUg1c1Dbr3hc+2p3l2lf++r57Oz2eDAiKYp3su2VURv1Pfsbu/1UlwQDrzVl+VEiIJVjn5CaMsPNC31ZCzZm433GbY1pn3F+1VCuXy1SKH0G8lGiBpGdh47y12t/xk/edRVfetdSbYMsya3cdxEREUlK56CzyzZhapA4WwGxZOoifqXNz1nT/RydFEwvhuaEbrOGvOhoL6nT9iseXacGHa+z8Ny5/HKwjkH/mRnw9o2jQLBzfbE2WJLcMflk255+G/c9u98PewdgQYX3DQXTIZVjn0jZDDZm433GbSO8UE1rJ1cN5cPVIoXQZyDfJNsgEmv52TPP01dezT3XnJ2R/jKZLrGX61wdtRhjLgVeB8wEon3StNbae1MZmIiIiMTmRZA4Gx9Ukq2LCEwJMocUcnPCd3zp53SPMCFzKtKKm94Wc12vO9qLO263g7Zf/lnzxIuO11lyxSXs6w9epREyGDGnFlODJcl9faPxg8yR2oeC7/G5fhLY7THMfx4LMBBRJDR0IuiepaUZDzZm433GTSO86pJgTevJz5PoqqFUrhbJtHzuM5Bv1jhoEIm1tPxhEw+f05zx/jLpLrGXDxwFmo0xtcA64GqCV4LEYgEFmkVERNLEiyBxNj6oOKmLGE8hNyesHT86G+o/c6BaUZ34QLXQ/g75ym3Wl7ZffnHTrArgVY1VNFUbTp0+MxPOnBb8WKXMZcmmUOAUCGfWD4wlV4c4JB8am6ZSAiLWiaBMBxsHXB4DprptnDTCg2BJFTd+OVjHuefO5cCB5OfYFUvnTajPHO3/OfJ+zbf5xc17bnfHEV45fpKna2Z5Pi9NLu33wKfeEd7v011iLx843bvuA64Bfgd8H+gA1PxPREQkw7wIEmc6IOamLmI86WhOmEhkh+vSEp/nHc5DB6RrWsd48Hv/Hr4/XiYzpKejvbjjJutL2y//PLfdXbOqs/uO8PEVjR6PRsRb9dPgdAB6HX7Sz4fGpm6PfXKlJEKox4VTXrzPOGmEN78iWMvarTdeeymPPPoLbBIZrD5juOc9r4v5uK4UyX9u33O7DnXS1lDv2byUTI8ZcR5ovh7YDFxtrXV3za6IiIikzIsgcaYDYm7LfcSTqQ+1sTpcL/v4Y7xh6Vy+8J7lXHuxd5dsrmz28WCSy6ajo70XirlenZOsr1zdfhJfrjerEnFqcobnr476+eEh5+/b6WxsGpmlGk0yWar5fjLQzZVhXr7PJKpNXV0CH7+ghGePuj/mu2txKSw+h+Xl13DLP/0mbrDZZwwPf/SaKcdgylguLG7fO0dHgj0QvJiXkukxs7bNz4nTMKvIT2443fvqgB8pyCwiIpJdzZWG6hIcNQSM9kEpkwGxdDU7S+eHWkjc4fq3rUd486pnePij1/C3b1zqyWsuq/OxoMrQPhB/y6Sro73Xiq1enZOsr3zYfjJVPjSrEklFrjc2jVZ+Jln5ejLQ7ZVhN3rclDBaI7yTp+HcqmBN5mV1vpQCzSEfedNSXhyp4rnnNrI/ShmNFUvncc97XufpiX7JTW7fO8vKy4DU56Vke8xY4NCAZZrPhBtheupjXw1+HxiCV/pgzA+lJTCjBqoq4NufScOLOuf0V98DzEnHQERERMSZpunBTu/JmvxBKZRx+vJpogY0I8s1eBEQS1ezs3R+qE22w3XAWm75zm84u6HWkw88re0n6DvVy7TRAAMBQ8/xk9Q11E9YJp0d7b0QylhubT/Bf7/v2XC5ka9+4EpPy43kqkRZX7m+/SS+fGpWJeJGLjY2jcxSjcxudpq9mq8nA91eGZaubRJZm3pgzE5p/OeF8xc1c/6iZrqOneQ/nvx1+Fji8bveWhTHEhLk9r2z6exmIPV9INkrCXqOn6T/lV5aegI0VZR4XmKPU73QfhR6+qc+VlcNL+6A5Rd593ouOQ00fwf4qjFmnrX2cDoGJCIiIsmpKTMsqLC0DyW3fKim3+QPTLMMV4HLAAAgAElEQVSnwSs9A/SchuHA1MZzXgXE3Jb7SCSdH2qddLgOWMu9T/w5pUBzrBIdv3z0cZrPPovLVlzKOec150ytyHgyXW4kF0XL+vKZ3Kn1Ke4tXTCLNyyd66g50eRmVSK5rNAbm17d6KN/zPKzwwEGo5ywriyB6+f5uKoxd36fi7/wdf5+yDJ9eIia/j5KAgH8Ph991TUMT6/g//zdXVHXy1SWeTo1zann3Dm14duaS4uLm/fcxvlzw0kaqcxLyVxJ0HWwk5Y/bOR4x9HwfSfw+Jh33XPQuhcCMcbS0w+33wtfuB1uuCa110qR00DzswSbAf7BGLMa2AS8Em1Ba217KgMzxtwIrABeA1wM1ACPWWvfH2edK4DPA5cB04G9wPeAb1trC2B6FRERmWhWuWGaz1LmM1GzJiOFaofdshCuavRNyTitNQFqSn1c8ZbLqG+o9zwg5qYuYjLS9aHWTYfrF1oPu85eSFSio/PQUZ760b9x48qraL44+9kK8WSj3Egui8z6Ct0uFl7UVM1VX3jPct686pmkTkYlalYlkmvyvZZxIuu7A/ykPXaW4qAfftIeoKbU5EyweXpvL+d3dFEzMDWbsa+qmqVtLbwQZb10npAXyRQn77kYw9IrLwFSn5cSXUmwb1sbf/7VC5DOY94Xd8C9D8YOMocELKz5LsxtyGpms9Mju4MEy44Y4JE4y1kXzz3Z5wkGmPuBTmBxvIWNMdcDTwHDwOPASeAdwDeBK4GbUhyPiIhITqopM7z1LB/fXfOLuMutuOltWODh/X5ml8Oxg4ejZpzu/s7jnHvuXN547aU0X3yO+4Ht7QgeGA0MBeuGLb+Ilc3zkq6LmIx0fqh12+H6ue0djgPNyZbosNby5Lrnee7VM3I2Gzhb5UYk96VSUzUXXXvxfB76n9ck/H+P1axKJNc5qWUMmatl3Dlo6R6x4atEnDYFdlJvNXTMlPVM7XXPMWP3PkzAhgMyIRaoGejns/ev5cTi5Tw+b9GEVbM+dhEPJPueizG87i0raDqn2ZMa6/GuCOg62Bk3yByS8jHvgz9NHGQ+82Lw0JN5FWj+IXj22TCRTxIMMO8lmNm8PtaCxpha4GHAD1xlrd04fv89wG+AG40x77XW/iTtoxYREcmCUFmMkKH+M8WbJ5fCsMDnn2nhF8+8EPNA7cCBIzzy6C+4bLqLs+/v/98x64ctq6vmrv/9Ce6bceGUA4oXfvrLuE8bWTMa0t+gx22HazfrOSnRYT0o0ZFOmS43IrnNq5qqueojb1rKOY213PvEn3mhdWplwWJqVhWq+x9L6CoayR/J1jIGWFCR/oBmS0+AdZ2BKVnWd28fY3GNYWWSZb6SrbcKwWOmpzsD2Q3WjmczmvFA0+SQeui2z1rua/sThyuq+H39WUB+ZZmLJBJ6z/1fj73Itt1TrzpsnD+XpVdeEg4ye1FjPd4VAS1/2JgwyBwSsJZ3fvWXnH12U/i+86on7ptR3yf3dsCmtqReI2zjzuB6i7Jz7OHoCM9a+6E0jSPaa4UDy8YknBhvBBqAH4aCzOPPMWyM+TzwHHAHoECziIgUnCG/ZVffxEBsZNB2coC262An659JfPbdujn7vu452LEn9uM9/bzms1/ma3fdxvcvXhG13Mf0EjjVEztQDplp0OO2w7XT9TJdouP/s3ff8W3V9/7HX0fy3vG2Y2eSkJDFKrMQIGWV0gJt6aWM0jIKdHDbX+m6DAcovffSSSlcKJSUcaG0UEbZDRAuM3s5NtmxncQjduK9JJ3fH7IUb0uyxpH8fj4eiawjHekjHZ31Od/v5xtKsfRZRHy1ZFEpSxaVessQ7a5vwTAMjpuZx7u/+HKkw4uIvY2Ht+GTc4ZuwyV6jDWwaZrdPThxenxok5nv1I/eCnlwebCR+FJvdbCKVtPvVtNB5UdrRjtw885NvJ9dFPIL8v7of6FxR5s5ZHqsXHyU0FuyqJT1i0p5obyB63//Ol0OF4Zh47MXneOtyRzMAZdHqlff3NA0oCazLzo6u2nr7CEtOYHmlnb2dru3KaPuJz/Z5Nd7DJgvGhLNFuapdP36MI+9B3QApxiGkWiaZnf4whIREQm91kElUD0jHpum+8CruaHJe+AF/l9997nFqad+2NgvSsm9D3Hb/+RTs3C+d5C0r1/1eW9N6DPKXmJ/p0mbY2iiPJgHj6MJdIRrf+cLZ4mOUIulzyLir3lTcphekEFDi3uE1vTkwC5WRav+LbH6t24OZUtmtaIOj9EGNn26KvRDIQWz1MVY9VZHi6EkJQLFjv1szWgCpxyqZ3bboZBfkA9U9sTaNEqIXDQvj0enZNLlhJ3tJnNKgz++DIxcr752T01Ar3fMwpm8eM2xvu8n230c9T1Y8wWBX4lmwzCcQJlpmneN8pz/AJaaphnOJPaRfbdbBz9gmqbDMIxdwDxgBuBnm3MRERFr8zRyGW7EY4BXH/0reaVFzD/1eJJTU/y++u5zi9MA6oeVPLpgyCBp5VWNHGhswXC6SDNsJLY2hWRwwrEEMsL14nmT/U6YhrNER6jF0meR4BpvTdVYFssDJoaTWlGHXqQGNg1mqYvR6q2OJtD5xs3P1oyeJXJxR627Zfcw42WEqpWjp6XyvZWOoLRaHrzf6HK6e72J9Jdkh0QbFCa5f/2h2C4NV6++t7s3oNfq7vLzmDc1OaD3CXi+ILCXlZX5/OSlS5eWASvKysqGG8zU85zTgbPKysruHHd0h19zGnA1sKmsrOz5YR6/BcgB7i0rK2sa5vHLgVLgqbKysmEvOxiGcf3SpUsfuuCCC4oBZs+eTUVFBbm5uaxbt47i4mJWr15NTk4O5eXlJCYmsnfvXjo6Omhvb2ffvn0kJCSwZcsW8vLyWLt27YB5KioqiI+PZ9++fbS3t9PZ2UlNTQ2JiYls3ryZgoKCAfNMmjSJTz/9FLvdTm1tLa2trXR1dVFdXU1ycjIbN26ksLCQNWvWeOfJyspi69atGIZBfX09LS0t9PT0sGfPHlJSUti4cSPFxcWsWrWKyZMns3r1ajIzM9m+fTumaXLgwAEOHjyIw+Fg9+7dpKWlsWHDBu/re94vIyODHTt24HK5aGpqorGxEZfLxc6dO0lPTx8wT0FBAevWrSM9PZ1du3bhcDg4ePAgBw4cAGD79u1kZWWxfv167zz5+fls2LCB1NRUdu/eTW9vL83NzdTX12MYBlu3bmXSpEkD5snLy2Pjxo2kpKRQVVVFd3c3LS0t1NXVYbfbqaysJCcnh7Vr13o/e25uLps3byYxMZGamhq6urpoa2tj//79xMXFeZe/Z55Vq1aRm5vLli1bvMty8PIvLy8nPz9/wLLMzs6moqKCuLg477IcvPwHL0vP8rfZbNTV1dHS0kJ3dzdVVVXe5V9UVDTi8m9oaODQoUP09vYOWf6rV6/23mZkZAxZ/k6nk507d5KWlsb69eu931dhYSFr164ddflnZGR4l8uqVasoLCxk/fr1pKamDrv8t23bNuLyT0lJYc+ePfT09AxZ/tnZ2QPWy7y8PDZt2kRSUhJVVVV0dXXR2tpKbW3tgOXff57+y9+zLre1tbFv3z7i4+OpqKgYdl3uv/z7r8v9l/+aNWu831t2djaVlZUjrsubNm0asPxXrVo17PL3rMue5T94WXrWZWDA8h9uXfb8djzL3+Vy0djYOGD5D16XPcu//7rcf/nv2LGDzMzMAcuyoKDAu/z7r8sNDQ0Dlv+6deu861h+fv6I67LNZhuy/D3zeJZ/dXX1iOtyLG7LdycWEb9rLZ1JGex4/33efu1D2oepiwzQ0dLGnvKtdHd20XzgoO874T4prjZOnlM84rZ8UmMb9j/4XqHKBIx9DbS9/n/Mf305x730InNff5Otj7xE9TPLuc+VTmNrF42tnaxfVU7dziqmmI185ZjSsG7LT1p4JE+uqPTpBNdmGNxydgmfmTfDr235y++t46Mdh/xeJhedOAN7S42ltuX13fG8vnaP35/l2KJ4zjx6hmW25bvsucTvXodrUjGJu9cGZVv+aX0zhqMHe2M1C/KG3y/H4rb8w8o9PLOjk3/uPATN9bQ4bSTXb+PF5kx2bl5PRn4RezatxpWeR3xNOXNykqN6W75ieyu765uxGQaZCXDB0YU+b8vNlAx69u2g02HibDlAUlczhtNB5qE9UXVcXt9l54//XEdjWzeNzW3ML0omwewJ6nH5phdeI2dlBZ87WM3lhcmceXQBb9W0kZJoJ8Hm4m8/OodFOY6YOy53TSomftdaTphZFLHj8i21h7C3NGAaNlIbtrGDLOL3bMA1qYjE3WuDflze6Ijjk01baEvJpbR2Pc3pRUzZv5b25GwKD1TitMWT2VZLQm878b2dZLXuo9aRQGbdFqYUFQw5Lq+oqKSq205Gex1JPa3EO7qZ1FpDb1wSxQ3ltKQWMKV2nfd9OpKyKGjaxrQ0G4ltDSMel69atRrXJPe6vKA4K2jb8ta3PiBty26/96tHN9Rie/jvGM+8Dh+sh1Wb3bfPvoH5p+doKs2jKTlu3MflnuV/ILWIgr1r6UzNJb++HFtCElmte0kzu0jsbcfeXMdRk+J92paXN7Tx1NZW3qrcyyFXPGl1leyLz2XzJ2s5aCbiaG/hqjPmWOa4fHdCIfG71niPGZRjCV+O5bXNDXR3d9Pe4+TYI3KwH9zL/NzkoG/LC5JsZHY2sK+xGburl+zmKvY197Jv99CxGcaysDCO+aV53PfSag52OKhrauaU2fm0NOwbdvlnFhcQ99xyn1/fM1jo5nOPJX/OrAHLv6ioaKnfAQfAMH3sOgtgGIYLd4vmEZPIhmH8CrjRNM2gXUY2DOMM3IMBPmWa5hXDPL4VmAXMMk1z+zCPfwicDJxsmubHo73Xgw8+aALceOONQYhcREQkdDwtRMq3VvOXx/7pczmMQNz19ZO49dITRn7CU6/Cfz8WlPdqtMVxaf4c3k6eNGC6YRg88p0ABiccp0ffKh9zhGubYfCnAGMrr2pk/vef8nu+zfddbrlyE7HwWS68+6UBLbF8GqjFB7E4EN5YxqqpCu6TodJkyOlrJRnt382Fd7/Eym11AJwwq8Dv30s0/06Wb6jmzmc/GbYXyOnzirn90hPHPyjiKIPNHoqL59+PPI7VOQX89ZbzLbNNCRar/DYGxxHquF7f7+TxPf6Xu7hqqo3zioY2f63pMPnxxtF7EQznvxeO3qMqZN9DgMdXnmTTiAwD7rgBLj5rtGcNy/NZd7SZdPS19HaaYO97wwWZgW/PR9tv9B9/5KmfXDhqLe5wssq6ORF5yk/saDP55tUXAKFdBpubXd569c0NTbz66F/9fo3Jk/PYu7dhyPRR95Pfut2/AQGPPwoeHTanHJZuKGMuAcMwTh80adow08Bdd34KcDnwaRBi80dz323mCI9nDHqeiIhIzHjvnTUhTTKDDwPcBakOmAnkuBy8WbuZ63Jn8Vj64ZGZAxqcMAg8I1zf9exKVpQPbbmweN5kbrv0hIBjCleJjnCIpc8C+D5QiwzhT03Vqk5IsJkhH0xMQmesC3Lvle/jnLIXAr4gB4w62KwJZDl6ebT8Y67LncX87zcFL7ktERXsUhcj1Vsdzdx0I3Klfk5c4NfTPQnmMaM1TVj6P1CcN+Z7DC7v47kY2+F0V0JLtLnfL26cX9FY+43ZbYf4bFMt6Y5eqh6xse3zC5l19NTxvamIHwbWq89mw9Qi9u7xryThcElmGGM/+e2vwg13+Vai0GbA9V/xK6Zg8yXV/y5413UT+Ebfv+EYuIdj/H/jjsw/nwLHA7OBNQMCMow4YDrgAHaGOS4REZGQqq1rYtcu35N6gRpzgLsg1QHznKPYgT8d2MaeuMQBLZv9GpwwiJYsKmXJolLKqxr52r2v4XC6iLPbgtZq7vZLT+ScshdGbTXtYRgGt43WujzC/PksNgt+lpdv/aL3pPqxZa8wua9FswYV848/NVUBarsgPT5k4USFaK1jvXxD9Zi9PsC9/Q74YuEYg80Ou+8oZ/zJ7TBRne6RJQdYk3e0+YartzoSA7i4JIItZ48ohePm+tya0a8thmnCr/4Cf/uVT0/vdJrUd7m3TzbgiDRIthtB+32OtN9Y+p+3U1hfS3p7v54MOzfAX4HMNLj3h34n5EXGw1Ov/vNnf4ZHHn0ZfypFjGbE/eSJC+C2b7svDo3GZsDtN0R8ffBli3Anhy+M3Y478TxcjWYn0Ai8Y5pmZbAC9NHbuFtSnwc8Peix04EU4D3TNLvDHJeIiEhIbd8e2IjH/lg8bzI/ffyDUZ/z8r8dE/T3tQO3HaoeUkLD58EJQ2DelBymF2QMuB8MSxaV8vBNZ42ZrDEMg69ccoalW+j5+lk85Uas/FkkMDUdpl+tBQHanO4kxkS0udnF8zWuId/Zjzc6mJNucEmJbcSBzazgzmc/8enCEozjYqEfg83233eMK7kdIQd7Dn/OSQnWv9AQau7fvv+lM0ZbZ+Zn2rh2Bj6V9rluhj3y658/rRn9tXWPe8DAUQYIPL/I1reNGji9qRfmpLu3YeP9jkbab5zx/jscsWs7BkPLgZiA0dzm/m5uD6wMiMh4zDqihC9fspjnn1/h835wLCPuJy9Z4u7ZM0L5KCtddBkz0WyaZpnnb8MwvgG8YJrmfaEMKgB/B/4L+DfDMP5gmuZqAMMwkoC7+57zYKSCExERCRW/Ry7u4zlgH4unxenvXl7nnba3sd37t7ecgJ8tbnxhAmd0NXNUTztbEgaWLVi+sdqy5RYCNVaJjhnTi1my5HhmHVESgej8E+pyI+FQW9dEU1MLrhaTOLstYhc3otHmZv+TQgCt/pdNjXpj1bGubDX5ZYWT62ZgmXqk/ZVXNfpVKgcCuFi4vdqvfcvgfUekesL4o3+LUNV7HShUpS7OzLeRl4i33upwr3GxVS7yjNGa0ZOAdeFuaey3V96Dmy8f9qFwbaOG22/Mq9jMtU8+4k0uD16i3vsuE+70rQxIsPRfT/uP6+CZrnV34jjhM0dxw8JJIx7zZhfm0VQ7fLmMkYy4n3zyHvft9mp3T5/2Tnev0hMXjHqxKNz8+vWbpjk9VIEMZhjGRcBFfXc9BRpPNgxjWd/fB0zT/FFfXC2GYVyHO+H8rmEYzwBNwBeBI/um+1+lW0RExOISk8aonTyCr502i2ff3z5m61lPi9P+J+iegTdgUDmBILe48ZxALOk8NCTR3NIRWILd6kYr0fFqx0hDUVhTqMuNhMryDdU8uOwTb0maur7p87//lGq++ijQmqqhaKxnZf7Usf7TTie5iaO30oyE5RurA57P5+3AJ5v8eu3h9h2R7Anjj2gtnxJqoSp1MbDeqsP7vd88a/SB/zzCmmwcpTWjAXSmpbE/r4AZu3b4/9pb9ww7OZzbqOH2G5e88jw2X1uJukx4+O8Rac2ZHdihuMSQwce8tZ1Oek0bn73oHGr31PidaIYx9pNHlFoqsTyYlS+zHM3QWtAz+v4B7AF+5HnANM0XDMNYDPwH8GUgCdgO/BC4zwxW0RQREZEIGnxSk1g8OaDXufWrJ3Dt5+aP2Xr2W5+bNmB6eVUju+pavEnDASfunhY3dz0U1IxRhmvo2ceYgxNGueFKdLw6Rg1PqwpVuZFQCMuAZhNAoDVVm3omVmswf+pYm7hbXlot0RzoRT+/5gtwsNnB+w4r94TxpXzKRBbqUheeeqv97/srLMnGEVoz/jl/Lo2lpZx3938F+MLDf95wbqMG7zcm76th7rbKIeUyRrV6y5hlQIJlIuyjxH/zpuRQlJvBwRaTRCAzL5vqrbsCeq1oblQz6tphGMbb9A0AaJpmTd99X5imaS4ZT2B9JTvK/JznA+Dz43lfERGRaJGdANmTsymdWkS1HyMeL543mXlTcpg3Jcd79f1L97xMdUMbGJCekcZFF51OYUG2d57lG6q589lPhnSRHtLKc6z6YQFosQ3NWo05OKGIn8IyoNkEEWhN1Zwovn7k6emxclsdh9q7vX97pg8eTDKQOtYVrablWrgGetHPr/kCHGx28L7DqiftvpQmuKfCSZINEvpyeBPpgoyHFUtdROz7H9SasbHv91BTOoUFm9b7/3qzpwyZFO5t1OD9xvzKzYCfgxuCOwFv4ZaeEvtaBrULiU8MbKTjaG5UM9aW8Yy+25RB98ei1sMiIiIhMNxJzQE/Rjz21Fz28CSQd9S2eKc1NrXwm98+w/TpxRx79Ynsrm/xr5XnpAz3v/ZOqK4FlwtsNsjNgi07ff6snlYsy5OzBkz3JMpFgum6ZWEY0GyCCLSm6m3zYiNplhTvTnBOzk4b8TmB1rH+/TYH9y4K7KQ1FAK96OfXfH52hR9p32HFk3ZfSxMAdLkgMz7wHgPB5CkTcW+lI6z1acdb6iLWrTzps5z/6ktjP3GwC04fMinQbdTmZhclKf7/SAfvN5I7A+vJEGgPCJFgGdyps3BqYGOrRHOjmlH3AKZp2ka7LyIiIpHnGfH4uedXjJpstvWruQxjlwnYtWsfZ9/xDzBgrPzbgFaef/jpyE/81u0+D+pkAO8mZQ6ozzw4US4SDOVVjd6azL6KlpqvkRKqmqpW5WmxPGIN+0FipY71vCk5nD6v2K8BAf2+WOjnYLPD7TvAmift/pQmAMhOsN4FmUjUpw1GqYtY0T/Z31FYQnVJKaU1ftROnz1l2BbAgW6jAp0PBu43OpMD68kQaA8IkWCxDdocZeZlk1daREO1/71Po9WoR3SGYXzfMAydzYmIiFjYLXPi+NuVC3mr7CIWzxu+ZvPieZN5s+wib11ZX8sEmIydZPbwtPIc1be/OvQIbARO4K6swyc/xqBEuVjbhXe/5P23clud959nmpWMZ0AzGZ67pqp9zG7PgdZUjXaBtkp1mYG3NAyV2y89EZvh23Y94IuF49h3gDVP2sdTmiDSZqYZzEwzuGVOHPcsjPf+u2VO3IQq52E18Qa8ecWVvhcuMoAfXT3sQ4Fuo8bT4r7/fmPznPlAAF3lIzAYoEh/GcNsAuefejyEej9pIWMd0f0OOM9zxzAMp2EYt4U2JBEREQnEkkWlvPuLL7P5vsuZV5pNXm4mBfmT+OEP/o13f/HlAQnaO5/1vUyAPzytPEfkGTCwL2EwOALPfSdwXe4s3k6eBLgHJ7z2mgs1AFuUmpyd5v1nRWEZ0GwCOjPfxs/m2pmbPvzJ1dx0g5/NtXNG/sRKMgMBJ9bruuGXFU7erbdOsnnJolIevumsMZPNg3vV+MWz7+jj677D875WPGkfT2kCEY9b5sR5E/9zMgyuuewYbHfcMPaMBnDHjSMmZgPdRo33oqFnv5FxZCkVs+b4V6P5+KNUn3kC8bTm39Fm0u7E+/e9lY4Bg6eHW5Id0gYlmwunlXDCeYvHTDaPaz9pIWNdbuwCEvvdNwigHruIiIiEz7wpOUwvyMDV15Wy/6B+4C4T4E83Z38t31g9esuxS5bA5Hw2/cdDLGioG/CQAazPzuPWtMlsLy1lnt3GuV85Z8hnkOGNdWAdzpZmo5UNsJqwDGg2QU2Emqr9W+iv3FY3ZPpw60Igdaw9TOBPO53kJo4/qRMs15w9j2n5Gdz17EpWlO8d8vjieZO57dITxnfy3G+wWWPQYLMGcMAWx9fy5wxJMlv1pD0SpQlkghhrYObMNLj3h6O2/g201n4wtuue/UbD976K6+a7sfnSMMJmwPVfGfd7S3RKMCJTxmckRckG2watOzMXzSU1M53yD9ZQXz30PCwo+0mLGOtsYxdwrmEY95mm6TlqinxfHREREQlYqLv7+9TK88lXWDC3BKbk0L15Bw6HuybfrknZHH1UKUbBUczqe6qSzIE52HP4kG1SQmwk9EIlLAOaTXATpaaqP632/aljPZgJ/KPGZZlEM7hbNi9ZVEp5VSNfu/c1HE4XcXYbf73l/OCVrXjyHvft9mr4ZBM7dtTySmUdDx1iSE1mq5+0R6I0gQzU/+JsOAc1DIv+AzNX7HTXQTMMmDvDXcfYhxITka61n7d4IY9/41qu+Msj2EzTO9Cnh/e+zYDbb1DZjAlgcOm1xv313r83vfSaZRo5pMfBlFQDg4EJ1MJpJRROK6G5oYn3X3gT03SRk2Tnnz8N4n7SAsbacj6Eu3zGPuNwE+8ywzDKxpjPNE0zirfKIiIisSvU3f39auWZmsyhoiIam5oB6IyPD1FU0clz4tt/sB/PfRh6Etz/fv8T6Kg+WQ6DUA1oFtNJDPEK9MTWXY8UHtkZWLLZU6/Xakl7T6+a/veD7ohSOKKUmcD3gSWhTG6HiPsigf9lMKx0cSGWWKk1pC/G3L94JqQmw5TCwzP6MVier9uoUNbav21fJ/PSs5nR0cokx8DjVwN8ap0tsWlvYztdPe4uHkkJ1rsCl5sAP5tr5x81LioGtW7OzMumMDeDomSD9LgQ7ScjaNSjW9M07zMMox64ACgGzgSqgN2hD01ERERCIdTd/X1q5fmHn3r/fLzSwWPLXgHcAwy9fOsXwWKDxVmBPyfBNR0m9d2mt0yBFZNRVnP7pSdyTtkLPtUuD6Tma7QlMSQ8zsy3kZcIj+50Utft//ybm12UpFjvBDvcwpLcDrJIliYQt1i52Dfs/qXfcdZ4eLZRwyXLwP2bvLjEFrILIKl2aElMYn1iEondXUzrasdmmsTF2difms6Co0qVZJ5A+l/YvfDul9jb5C4Nc8KsAsu0Zu5vtPJhN/4rdrflY25ZTdN8BngGwDAMF/CYaZp3hjowERERCY1Qdvf3pZWn+G5mmvsg1NeT4c3NLp6vcQ1JXPx4o4M56QaXhPBkMNp5BjS7/oG3R002+1PzNVaSGBJa8zNtnJ5n8rca/1u3ql5vdIt0aQKJXuHcv0Sy1v43r76A1Vdf4L3/Qd/tLXPiyA3pO4sEz0QpH+bh79ZpKfBuCOIQERGRMAmkTIAvAmnlCVBb10RTUwsulwtbu53yqgPmKFUAACAASURBVMagxhUMgQz2FW7v1LtG7d5a2Wryywon182AM/KVrBhOWAY0ExmG6vVOTFYoTSAylv5lOlp7D09/usp9pSscSe9Op0mrA/XUEokCfm0RTNNcGqpAREREZHwGJ0M7+s4LHlv2Cu95SlL08adMgNH332hP9aeVp8fyDdXc+ewnAxLeDcD87z9FvN1GalIcCXF2dvSV1fhmvxYtkeTPYF/hsrl59CSzhwn8aaeT3ETV+RxJWAY0ExlE9XonrkiXJvDFRKo33/+zDieWPmsgwl0GqrXXpLYL2gb13lBPLbGSYDZIiYVtkPUjFBEREb9Nzk6jrnvktKO/ZQKm5gW3ledXn9jIc8+vwBzhvXudLg619zAl1xpJ3XC1WA70QPX5GpfPg4mZuBMaOjEbXTTWfJXopXq9E1skSxP4ayLVmz/Yc3h9nJRgreUQTpFKbL1T72J7+8iPq6fWxHTh3S/R2tnDmh0NdPa4k7IfVu63TC/DYDZIidZtkBLNIiIiMWLwgVX/K+LDnST4WyYgWK08l2+oHjXJ3F9NYzvXXHQWs44o8es9YoGvB6o1HaZfySmAilZT3U5FLOaSEhv3VPhWdFn1emOTVet4RkMLumDxfNbBSf/LptgtszwmAk9PrbGop9bEsnxDNR9W7qepbeDouU1t3XxYuZ/ZxVkRiSuYye3+29uxzuWsKnoiFRERkaDzt0xAMFp53vnsJz4lmQFcpsny5asnTKI5kAPVzc3+d7f3zFeSogKvIlYxP9PGlGQnVZ2jP0/1eg+Lhvr5EjzhKOGhQXWtQT21ZLBH3yoftSdmU1s3K7fV8+d/lfOtz80Lc3TSnxLNIiIiErYyAeVVjX4PQrhz1z5q65ooLMgOSUzRrtO3BpBBm09EQicn0SDBNnxNUrBGvV6rsmL9fF9NpBrI49H/u2lzDD99PDSorjWop5YMtnxD9Zjl/sDdQOW6P77N1LyMqB+4uabDpL7bjMoBMLXHEhERiSFWP1l98pOqgObbvr1GieYRJAfYKDnQ+UQktNLjDdLj3V31rV6vN9JiscXyRKqB7K+Zae7ff6fTZEcbOPoOc5wm2I2hg2j5c8yjQXWtQz21ZLA7n/3EpwHMwZ1svuvZlVGbaL59U++IA2Cm2eH7s63fo0mJZhERkRhlxZPVlXXdYz9pGPtbetjRZlomYW4l7oNN/0/KrH6QKrHB6he/rMjzPT1d5aS19/D0p6vcZ536zmKPlqlvzi+y9ZW1GDi93QmJNtjXaZJsD2zQLJVqsA711JL+AukNuaJ8L+VVjVE3kPNYA2C2OYmKXhXao4mIiMQQK5+s1nSYHCI+oHld8fG0OYLXPTZcBreuGiwYy6skxWBOuuFXN9O56YZaRkrYWfHil9XpO5NICMe+y19jlbXodkFdN0xJ9j8+lWqwFvXUkv6Wb6wOeL5oSjTH0gCY1j0bFRERkZiyudlF4dTABvWbMr2ERNvhbrPRptNpUt/lbndsA/KTINkevM9ySYmNX1aM3eUX3AOJXVxizQNTiT1WvvhlVZ7tXLR9dxqYL/aEet/lK1/LWgBUdcI79U7OzPc966hSDdainlrSX0tHT1jnC6XHlr0CwHtpxpD95NZWk+Mu/rxPr2P1XhXRdfQiIiIiUavTCZl52eSVFtFQvd/n+YqmFDN9cjY2w12zNJpaD43UzbepF+aku09Sg3GQOD/TxrUzGPNE3ACum2H92m4iEt2ieWA+Cd++y1f+lLUA+NNOF//XYHKJjwNnqlSDtainlvSXkRJY155A5xuNv+XABj+/3Xn477jUVAoS3b/ZLufAAU59YeVeFUo0i4iISFh4ujTOP/V43vnrP8GXQT0MgzmnHMf+LvfdH290MCfd8PnkMZLCPXr9mfk28hLdLRwqhjk5m5tucHEUfG8iE02s1LFWi+XYEO5911gCKWsB/sWpUg3Wo55a4rFkYWCD+gU6n6/8LW2VnQCXX+luseypI+/Zv7++38nje/xvxW/VXhXRcdQiIiIiUc/TFbJwWgknnLeYla+vGD3ZbBiccN5iCqcNLLcR7pPcQERq9Pr5me5Eck2Hye+3OXCZYDPg5llxlmzxICIDqSazRFKk9l1jxRQoX+NUqQbr8fTU+tMYNWvVUyv2zZuSw+nziv0aEHDxvMkhqc/s74VfX58fa70qlGgWERGRsOjfFXLmormkZqZT/sEa6quHHjjmlxYz79TjhiSZPaw+EEakR68vSTHITzQG3BffqM6shFu0tFiW2BfpfddwxptI8SVOlWqwpjPzbbxT56S2C9qG+R2op9bEcfulJ3JO2Qu4fOgNaTMMbrv0hDBEFTyx1qtCRzUiIiIypmAl3/p3hSycVkLhtBKaG5p4/4U3MU0XhmHjsxedQ2Ze9pivZdWBMDR6fexQnVkRmSisuu8KRiLFlzhVqsGa0uMN0uPdA1O2OlBPrQlqyaJSHr7pLK5/4O1Rk802w+BP3zmLJYtCWzYj2HITA/stBzpfqCnRLCIiMoEFkkAeT/JtuEHrMvOyScvK8D7HlySzhxUTtBq9PrqpxbKITERW3XcFWtZisLHi1KC61tW/dr3H01WHmzirV8jEcM3Z85iWn8Fdz65kRfneIY8vnjeZ2y49IeqSzAAHuv2vQz+e+UJNa6SIiIgAoyeQg5l8G2vQOn9ZLUEba3XWREQk9ll13xVIWYvh+BKnBtW1vt6+aw4He0zvgGoycSxZVMqSRaWUVzXytXtfw+F0EWe38ddbzg9JTeZwser2N1BKNIuIiExgkWq9OXjQuiQ7OExwBNBoyWoHWZGss3ZvpcP7d/8WQJ7pavUjIiLDsXKNUH/KWozE1zg1qK71zEw7/L17jm0mJRg6ppnA5k3JYXpBxoD70czK299A6FKciIiIRIxn0LpEG8QFeP5mtYOsQFs6BbuFVHbC4X8iIiKjscq+a6T3uHaGnfGkef2N03N8UpjkvlWSWURCxcrb30DoEpCIiIhYQqCJZqsdZEVy9Hq17hERkUBEct/li/GU3QpnnDJ+I/XOAuhwgoWqpYkEhdW3v/6y1pmZiIiITAj3Vjq8/3a0mbQ7ocvlf7LZqgdZl5TYfG55pdHrRUTECqy+75qfaeO2eXFcN93399U+Nrr1ug7/63CC03Tf7mgzBySkRaKd1be//rBuZCIiIjIhZCdAguH+l+hHKxUrH2T52s1Xo9eLiIhVRMu+68wCO1OSx35epOOUwNwyJ877b06G4f2Xn3i4QUK3Czqd4x9QWsQqomX76wv1rxQREZGwG1ziYVPG4cOq62bYeWTn6IP+RMNBlkavFxGRaBMt+66cRIMEm0ltF7QNMyiwVeKU8bllThybm108X+NiX9fh6Q4TKlvhznIHl2g5S4yIlu3vWJRoFhEREUuJlYMs0Oj1IiISfaJl35Ueb5AeD5dNsVs6TgncO/WuURsfVLaa/LLCyXUz4Ix86x8Xioyl//b3ts2Hy8PcNT96tmtKNIuIiIjlRMtJrq88o9f3vy8iImJl0bLvipY4xT+bm0dPMnuYwJ92OslNtN4A0SKBKkkxSLQNvB8tlGgWERERy9LJo4iIiMjE83yNa8wks4eJuyecEs0ikadEs4iIiIiIiIhMePdWHu6qvqPNHDJ98BgTEho1HSaVw5RPG01Fq0lNh6lGCSIRpq2kiIiIiIiIiEg/2QmRjmDi2tzsCni+khR7kKMREX8o0SwiIiIiIiIiE55aLFtDpzO884lYRf9eFR3OodOjYRtl/QhFREREREREZEIbrazFjjaTmWkqmRArkgNslBzofCJWFB+lmzQlmkVEREREREQkaqisRWxzD+rnf/kMDQYo0a5/i+X+F9eioSWzR/REKiIiIhOGBuMRERGR/kbb9/c/bpDoV5JiMCfd8GtAwLnphgYCFLEAnaWJiIiIpUVzqyUlzEVEJNpEy74rWuKUwFxSYuOXFU58STUbwMUlas0s0S8WtmvWj1BERERi0oV3v+T9e+W2uiHTX771i2GPKZSiOWEuIiITU7Tsu6IlTvHd/Ewb186AR3aOnmw2gOtm2FU2Q2JOtG7XlGgWERGRiJucnRbpEEIiGlodiIiI9Bct+65oiVMCd2a+jbxE+EeNi4phymjMTTe4uMSmJLPEjFjYrkX/JxAREZGoFGstlkVEREQkuOZnuhPJNR0mv9/mwGWCzYCbZ8WpJvMEM9F6Q0YrJZpFRERERERERMSySlIM8hONAfdl4orV3pCxQIlmERERERERERERsSy1WI4OKmQjIiIiIiIiIiIiIuMSc4lmwzB2G4ZhjvCvNtLxiYiIiIiIiIiIiMSaWC2d0Qz8bpjpbeEORERERERERERERCTWxWqi+ZBpmmWRDkJERERERERERERkIoi50hkiIiIiIiIiIiIiEl6x2qI50TCMK4ApQDuwEXjPNE1nZMMSERERERERERERiT2xmmguBJ4YNG2XYRjfNE1zRSQCEhEREREREREREYlVsVg64zFgCe5kcyqwAHgImAa8ZhjGosiFJiIiIiIiIiIiIhJ77GVlZZGOIajKyspWlJWV7SorK2svKyvrLSsrqy8rK/vn0qVL04HTgcKysrK/Dp7PMIzrly5d+tAFF1xQDDB79mwqKirIzc1l3bp1FBcXs3r1anJycigvLycxMZG9e/fS0dFBe3s7+/btIyEhgS1btpCXl8fatWsHzFNRUUF8fDz79u2jvb2dzs5OampqSExMZPPmzRQUFAyYZ9KkSXz66afY7XZqa2tpbW2lq6uL6upqkpOT2bhxI4WFhaxZs8Y7T1ZWFlu3bsUwDOrr62lpaaGnp4c9e/aQkpLCxo0bKS4uZtWqVUyePJnVq1eTmZnJ9u3bMU2TAwcOcPDgQRwOB7t37yYtLY0NGzZ4X9/zfhkZGezYsQOXy0VTUxONjY24XC527txJenr6gHkKCgpYt24d6enp7Nq1C4fDwcGDBzlw4AAA27dvJysri/Xr13vnyc/PZ8OGDaSmprJ79256e3tpbm6mvr4ewzDYunUrkyZNGjBPXl4eGzduJCUlhaqqKrq7u2lpaaGurg673U5lZSU5OTmsXbvW+9lzc3PZvHkziYmJ1NTU0NXVRVtbG/v37ycuLs67/D3zrFq1itzcXLZs2eJdloOXf3l5Ofn5+QOWZXZ2NhUVFcTFxXmX5eDlP3hZepa/zWajrq6OlpYWuru7qaqq8i7/oqKiEZd/Q0MDhw4dore3d8jyX716tfc2IyNjyPJ3Op3s3LmTtLQ01q9f7/2+CgsLWbt27ajLPyMjw7tcVq1aRWFhIevXryc1NXXY5b9t27YRl39KSgp79uyhp6dnyPLPzs4esF7m5eWxadMmkpKSqKqqoquri9bWVmprawcs//7z9F/+nnW5ra2Nffv2ER8fT0VFxbDrcv/l339d7r/816xZ4/3esrOzqaysHHFd3rRp04Dlv2rVqmGXv2dd9iz/wcvSsy4DA5b/cOuy57fjWf4ul4vGxsYBy3/wuuxZ/v3X5f7Lf8eOHWRmZg5YlgUFBd7l339dbmhoGLD8161b513H8vPzR1yXbTbbkOXvmcez/Kurq0dcl7Ut17Zc23Jty7Ut17Zc23Jty7Ut17Zc2/Lo3ZbvTigkftdaXGnZZNR/qm25tuXalvuxLS8qKloanMzr6AzTNMPxPhFnGMYRwDagyTTNnJGe9+CDD5oAN954Y7hCExERERERERGRUdxb6fD+fcucWK0EKxIyRjjeJBZLZ4ykvu82NaJRiIiIiIiIiIiIiMSYiZRoPrnvdmdEoxARERERERERERGJMTGVaDYMY55hGNnDTJ8K3N9398nwRiUiIiIiIiIiIiIS22KtqM1XgZ8ahvEOsAtoBWYCFwBJwKvAryIXnoiIiIiIiIiIiEjsibVE8zvAkcAxuEtlpAKHgPeBJ4AnzIky+qGIiIiIiIiIiIhImMRUotk0zRXAikjHISIiIiIiIiIi43NvpcP79442c8j0W+bEVFpLJOppjRQREREREREREUvLToh0BCIyFiWaRURERERERETEctRiWSS62CIdgIiIiIiIiIiIiIhENyWaRURERERERERERGRclGgWERERERERERERkXFRollERERERERERERExkWJZhEREREREREREREZFyWaRURERERERERERGRclGgWERERERERERERkXFRollERERERERERERExkWJZhEREREREREREREZFyWaRURERERERERERGRclGgWERERERERERERkXFRollERERERERERERExkWJZhEREREREREREREZFyWaRURERERERERERGRclGgWERERERERERERkXFRollERERERERERERExkWJZhEREREREREREREZFyWaRURERERERERERGRclGgWERERERERERERkXFRollERERERERERERExkWJZhEREREREREREREZFyWaRURERERERERERGRclGgWERERERERERERkXFRollERERERERERERExkWJZhEREREREREREREZFyWaRURERERERERERGRclGgWERERERERERERkXFRollERERERERERERExkWJZhEREREREREREREZFyWaRURERERERERERGRclGgWERERERERERERkXFRollERERERERERERExkWJZhEREREREREREREZFyWaRURERERERERERGRclGgWERERERERERERkXFRollERERERERERERExkWJZhEREREREREREREZFyWaRURERERERERERGRclGgWERERERERERERkXFRollERERERERERERExkWJZhEREREREREREREZFyWaRURERERERERERGRclGgWERERERERERERkXFRollERERERERERERExkWJZhEREREREREREREZFyWaRURERERERERERGRclGgWERERERERERERkXGJuUSzYRglhmH82TCMfYZhdBuGsdswjN8ZhjEp0rGJiIiIiIiIiIiIxKK4SAcQTIZhzAQ+BPKBF4FK4ATgZuA8wzBONU2zMYIhioiIiIiIiIiIiMScWGvR/ADuJPP3TdO8yDTNn5qmeRbwW+BI4BcRjU5EREREREREREQkBsVMotkwjBnAOcBu4I+DHr4DaAeuNAwjNcyhiYiIiIiIiIiIiMS0mEk0A2f13b5pmqar/wOmabYCHwApwEnhDkxEREREREREREQklsVSovnIvtutIzy+re92dhhiEREREREREREREZkwYmkwwMy+2+YRHvdMzxruQcMwrgeuf+CBBwB48MEHgxqciIiIiIiIiIiISASYN954oxHqN4mlFs1j8XyZ5nAPmqb5sGmax4cxHhEREREREREREZGYEEstmj0tljNHeDxj0POGFY7svoiIiIiIiIiIiEgsiaUWzZ/23Y5Ug3lW3+1INZxFREREREREREREJACGaQ5bSSLqGIYxE9gO7AZmmqbp6vdYOrAfd2I9zzTN9ogEKSIiIiIiIiIiIhKDYqZFs2maO4A3gWnAdwY9vBRIBR5XkllEREREREREREQkuGKmRTN4WzV/COQDLwIVwInAmbhLZpximmZj5CIUERERERERERERiT0xlWgGMAyjFLgTOA/IwV0y4wVgqWmaTZGMTURERERERERERCQWxVyiORjsdpvpclnjezEMg1hZRjYDLPK1ioXYANeYz/JPLK03oWAYNvqVsRcrMgzQb9h/hg1i7bet30KADEDf22hCsf8F6x7vxfqxgVW/91gTqvUm2GL99z4eUXkcrGOB0UXT8Z9lluXEOE6Klm12GB0wTTMv1G8SF+o3iEYul4m5OtJRuBnHm7z33nuRDiMoTj/9dMz7Ix2FWI3xXTAXfiW4r7nx71z46TVBfc1Y8vKRj3LZRz2RDkNG8fTJCRz3t22RDiPqrPnqLPh5jB0032Nw2de2RDqKqPP0X4+CL/0+0mFYmuvFmzFPPTror2t8sN4yx9H9xdIx9XB0nB0eoThuDQUdC48sGo+DdVw4uqg6/rPIcd1EOU4K1bFOtDI+WL8nHO8TM4MBioiIiIiIiIiIiEhkKNEsIiIiIiIiIiIiIuOi0hlB1t0DN/0n/GslNLXAESVwz3fg/FMjHdnw7rrrLtasWUNXVxfZ2dl8/etf5wtf+EKkw4p52+phwT3wlWPgyW9EOhprO2PHu3zc0UScYQAwOS6ZT+ecF+GofLf3lR1svX8dnfvbScxN5uj/PJ2c4wsjHVZM+NtZkwbcd3Z3csQlN3D8//tdhCKKbd31NVQ9Ukb71nUYcQlMOuk8Sr/5Hxj2CXYocaAC3vgO1K6BlDw461448uJIRxVSW7c9xa5dL3CoeStTp1zASSfe432stu4jVq+5m46O/eTkLOSkE35BaurkCEYbYTvfg6qV0LoPJh8Hx17unu5ywOrH4VA1dDbBqd+F3FmRjXUY9+9vYFldE5s6urgsL4tls6YC8HFrO7ft2c+a9k7swBmZadw3o4SihPjgvO9fYdk/YdN2uOxcWFbmnr5lJ1x1B+yocd8/bi7c9yM4akZQ3jYonnvuOV5//XV27tzJkiVL+PnPfz7kOY899hiPPfYYv/nNbzj++OMjEGVk3b8Cln0Mm/bDZcfBsivd03c3wvQ7IDXh8HN/cjbcdn5k4gy1+w9sZ9nBPWzqauayrFKWlX7G+1iHy8GP9m/k2UM19JouFiVn8d7MMyIXbAB2PbmF6ue30bq1ieIvzOSY/zwdgJqXtrPxjg+8zzNdJq4uJ6c99yWy5udGKtyAbf3bA+x69XEO7djM1LO/xkm3Pep9rOpff2PTI3fS0bCXlPwSFt1wFyWLvxTBaMen/rUnaHz3eTqrPiX71C8w7bv/7X3swPJnqf3HQ/QeOkDanOOYetMvScguiGC0AXJ0wxs3wa5/QVcTTDoCzrgHZvZtiHYtdx/3tVRB8Ylw4TLInBrysJzOHlavuZPauo/o6WkmLW0Kixb+O8VFp+N09vDRxz+mqWkz7R37OOvMZRTknxDymHzidMDGZ6FhK/R0QFouzP0CFBwFTbuh8hU4VOOuVZ07CxZcAkmZkY46aLpdLm7aUcO/mltp6nVyRHIi90wt4vxJGWzp6OKqrXvY0eUu3XNcWjL3zSjhqJSkCEcdfBPs7DD0HE4oLYAVD8OUQnj1A7j0Z7DpGZhWHOnohrriiiv4yU9+QkJCAnv27OHmm29m1qxZHHnkkZEOLaZ951n4TOj3TzHj/uJjuDZneqTD8FvDB3up+NUqjvvtWWQtzKOroSPSIcWUr7590Pu3o7Odf1xQwpSzvhzBiGJb1SNlxGdks/DhD3G2t7D1rqtpeOMp8j8/ga6WuRzw9y/BMTfAZW9B1Qr424XwrXWQMzvS0YVMcnI+8476NvtrP8Dp7PZO7+4+yPsf3MwJn7mTycVnsnHTfXzw4f/jnLOfiWC0EZaUCUeeA/WV4Owd+FjODJi5GFYti0hovihOiOfW0gLeONRKp+vw8DkHHU6uL8zl3Kx04gyD7+6s4Zvbqnh93szgvG8e3HoNvPERdHYPnP73/4KpReBywR//Bv/2c9hooZ9Ybm4uV111FStXrqS7u3vI43v37uXdd98lJycnAtFZQ3Em3HoevFEBnb1DHz90L8TZwx9XuBXHJ3Nr/hzeaK2j03QOeOz6mrU4TBcVR55Ltj2B9Z2HIhRl4JLyU5h10yIa/m8vzu7Dn6/ki0dQ8sUjvPern9/K1gfWkzkvOteJ5Lwi5l39M/Z/8hbO7k7v9I76vXy09GpO++/nKDrpXPZ9+Bof/MdlfPH5bSRl50cw4sDFZ+dT9OWbaF7/f5g9Xd7preWfsPd/f83ssidJLJxK9WN3s+t3P+DIO/83gtEGyOWA9FK4YgVkToHtr8I/LoVrN0FCGjx/CXz+EZh1Iay4Df7xNbj645CHZZoOUlIKWXLW46SmFLFv/3t88OEPOf+8F0lOyiM391iOnH0l73/4g5DH4hfTCcmT4LPfc9/WbXEf95z1E+jtgKmnwGfmugdw3PR3WPe/cPKNkY46aBymSWliPCvmH8GUxARePdjCpZW72XTMkRQnxPH3OdOYmpiAC/jj/gP826e72XjMnEiHHXQqnRFkqclQ9m13Utlmgy+cBtOLYU1FpCMb3vTp00lIcDcjMPpajO7duzeSIcW8Z1ZDVjIsid2chPT59A9rmX3TMUw6Oh/DZpBckEpyQWqkw4pJVW8/R+KkfPKO/mykQ4lZPfXVTDrl89gSEomflEfm0afRWT3BBqY5UOluqXrCD8Bmh2lnQcmpsPmJSEcWUqUlZ1NS8jkSE7MGTK+ueYvMjCOYUnoednsiC+Z/h0PNn9LSsjNCkVpA8SIoWggJg7b1tjiYeQbkzHS34rGoS3KyuCgni5y4gW1Rzp+UwVdzs8iIs5Nit/Hdolw+aGkP3vueBRedATmDGjVlpbuPqQ0DTBPsNtheHbS3DYrFixdz2mmnkZGRMezjv/vd77jhhhuIjw9O6+9odMnRcNEiyJngh0CXZE7moszJ5MQlDJj+aVcrL7Xs4+GS48iLS8RuGByXMmmEV7GuonOmUfS5aSRkJY76vOp/bKfkoiO8557RpvSMiylZ/CUSM7MHTO9o2Et8ehbFJ5+HYRhMPvXzxCWn0rY3eveJk048l6wTziYufeD+v3nNO0w6+XySS2dhi0+g6Cvfoa1iFd21YRlnLLgSUuH0Msia5k5+zvoCZE5391z79HnInQdzvwpxSXBaGdRvcB8PhlhcXAoL5n+XtNTJGIaNycVnkJpawsGmcuz2BOYceRV5ecdhGBa7SheXCHPOh5Qc9/dZOB9Ss909ugqOgsnHQHwSxCXA9NOgcVekIw6qVLudsilFTEtKxGYYfCE7k+lJCaxp6yQrLo5pSYkYhoEJ2A3Y3jX0AnUsUKI5xOoaYWsVBKnBR0j85je/4eyzz+aKK64gNOW52gAAIABJREFUJyeHk046KdIhxayWTrj9Ffj1JZGOJLr8rHYTueUvcer2d3i3rT7S4fjEdLo4tPkA3Qe7WH72s7x1+tNsuvNDnF2OSIcWk3a99iTTz788ak9aokH+579B0wev4OrupKexlub175Fx9OmRDivMhhnR3DShYXP4Q7GA5ubtZGUd7gEVF5dCWmopzc3bIxiVhMN7LW3MC2NXz6wzIOlU+N698PNvhu1tx+2dd94hLi6Ok08+OdKhWNrU26HkVvjmE3CgLdLRhN8nnU1MTUjhjrpycstfYsHWN3muuSbSYYVEx95WGlfXUvol65UOGq/sOceRMXUONf/3Mi6nk5oVL2KLTyTriAWRDi3oTNN0H/8cngAQGw0Q2uqgaas7wdxQDvmLDj+WkApZM+FAedjD6uw6QGvrbjIzjxj7yVbS1QJtDZBeNPSxxh2QEdslJet6etna2T3gmCnr440kfbiB7+3cy89LorDcjA9UOiOEeh1w+W3wjQtgzrRIRzOyH/7wh9x8882Ul5ezbt06bwtnCb7bXoFrToHS6GukEDH/VbSAoxIzSDBsPHOomgt3f8j6WZ9jZmJapEMbVfeBTsxeF/tf382pT30BI87GqpveYuuD65n7g4lXnzGU2muraFj3Hif+/KFIhxLT0o46gQP/epZ1Vx0DLic5iy8m64SzIx1WeOXMgdR8+Phed6vmPe+4y2dMPTPSkUWEw9FBYuLAFl3x8en0OoLX0lWsZ2N7J3dW1/HinPCVtDr0LrR3wl/+6S6jEQ06Ojp4+OGH+fWvfx3pUCwrNw1W3QJHl0Bju7u03OXL4I3vRjqy8Krp7WBzVwtfzpjMvrlf4KOORi7Y/T5HJWYwN2n4lvLRquaF7eQcX0BKaXqkQwk6m93O9POv4KM7rsLZ04UtLoFTf/E0ccmx15Q/89jF7Pztv5N7zmUkFU5j/9/vB8PA1a+USFRy9sJLl8OCb0DuHOhtc4/H0V9SJvS0hjUsl6uXjz7+MdOnfYmMDAsNUjAWlxPWPAGlJ0D6oIRq81749A048drIxBYGvS6Ty7fu4Rv52czpl2g+dNJC2p1O/lJ/kKmJsdnbSS2aQ8Tlgitvg4Q4uP8nkY5mbHa7nYULF9LQ0MALL7wQ6XBi0voa+Fcl/GBi5iMCdmJKDun2eBJtdr6RPY1TU3J4tbU20mGNyZbkvo43/cqjSMpPITE7iRnfnE/9Cov1+Y0Bu157ktyFp5JWHH11vKOF6XKx7e5vkXXiORzz5AYW/XkljvYW9j7532PPHEvs8fCVF2D7K3BfIXzya5h7KWSURDqyiIiLS6G3d2Dzw15HG/FxsXdSLW7bO7s5f8tOfj99MqdlhveCb2oy3PBl9+CA9U1hfeuA/PnPf+bcc8+luNiCg7RYRFoiHD/VXZ+5IAPuvxTerHT3AJxIkg078YbBrQVzSbDZWJyWx5mp+bzZVhfp0IKu5sXtlFwUe62ZAWpXLmf9H3/GWX/8F197r50lDyxn5S9v4ODW9ZEOLegyFpxC8aXfZ+evvsummxaTkF+CLSmV+Jwobp1quuClK8GeAOfe754WnwbdLQOf190CCeG7UGKaLj76+KfYbPEcf9ytYXvfcTNdsPYJd6m5hV8Z+FhbA3z8kHsgwBwLd/0fB5dpcuW2PSTYDO6fMfQ8IdVu54bCHK7aVkV9zzCDF0Q5tWgOAdOEa+6CuiZ49fcQH0XfstPpVI3mEHl3G+xugim3ue+3dYPThC37Ye1PIxtbNDEMMIfrvm4xCZmJJBWmgio5hNzu155k7pU/jnQYMc3Zdojexv3kn3cltvhEbPGJ5J75ZfY+81tKroyCq6nBlL8Qrlxx+P5fTnG3fJmAMjOPYNeuF733HY4O2tqqo69bp/hkT1cPnyvfzm0lBVyZnz32DCHgckFHF+ythwiF4LO1a9cOaMBx6NAh7rjjDr7+9a9z+eWXRzg6a/IcMln/KC+4FiZljv2kGNC0po6u+g6Kz50W6VBC4uC2DeQf/Vly5h4HQM5Rx5Mz7zPUrnqbSbOPjnB0wZd/3hXkn3cFAF37dlH73AMkl0bpIESmCa9cA+118LVX3Q0LAPLmwaa/HH5eTzsc3OEuqxGWsEw+WXkrXd2NLD7tf7DZoqT1q2nCuqehuxVO+rY72ezR0QQfPgCzz4HSz0QuxhAyTZNrtldT1+Pg1aNmEG8bPiHgAjpcLvb29JKfECXL1kdq0RwCN/4SKnbBy7+F5PCVr/PbwYMHWb58OR0dHTidTlauXMny5cs59thjIx1aTLr+VNhRBut/5v53w2fhgnnwxnciHZl1HXL28EZrLV0uJw7TxVMHq3iv7QDnpkXH1fLSS2ax64ktdDd20tPcza6/lFNwxpRIhxVTGjZ+REfDPqac9eVIhxLT4jKyScgvoeHN/8V0OnC0t9D47j9ImRp7oySPqX4jOLrcI2d//Cto2w8Lr450VCHlcjlwOrtxmU5M0+n+2+WgZPLnaG7ZRnX1mzid3Wwuf5CszNnR1a0z2FxOd9db0+X+5+x1TwNwOtz3BzzPWik1h2nS5XLhNE2cJnS5XDhMk73dPZxVvp3vFOZxQ1Fu8N/XAV3d4HSB0+n+2+GAtz6GdZXuaS1t8MPfwqR0mGuhDiwOh4Pu7m5cLhcul4vu7m4cDge//e1vWbZsGY8++iiPPvooOTk5/OhHP+Liiy+OdMhh53BCV2/f8nW5/3Y44ZPd8Gmd+wJCYxt8/+9wxizITI50xKHhMF10uZx965fpPb49PS2PKfEp/LK+Eofp4oP2A7zb3sC5adFVu9PlcOHsdmC6TEyn+2+Xw+V9vPqFbRSdM424tOgu0+hyOHB2d+FyOjFdTvffDgc5c4+nfsMH3hbMTZ+uo2H9B1Fdo9l0OnD1dIPLhely4erp9k7rrNqKaZr0NOxjz0O3kv/5bxCXFqUXTV6/EQ5UwKUvQ3y/DdDsi93jcFQ+5z72e/9Od4OD3PAc/65es5SWlp2c/tk/Ehc3MLHkdPbgdLoHknO5enE6u921s61g47PuWtcnXu9uIe7ReQg+uB+mf9b9L0bduKOGio4uXj5qOsn2wynXtw61sq6tA6dp0uJw8sNde5lktzM3jGNehEsUtbWNDnv2w0PPQ2ICFJ57ePpDP4fLz49cXMMxDIMXXniBX//617hcLgoKCvje977HaaedFunQYlJKgvufR1oiJMVDXuyVKAuaXtPk1tpyKrtbsWMwJymdF6adwpFJ0fGlzb7pGHoOdvH2uX/Hnmin+PzpzLpx0dgzis92vfoEpWdcRHxqdPwmotnMH/2R6mW/oPbFh8FmJ33eiZRc/R+RDiv8Nj0BGx5xJwlLT4PL3nKPsB3Dyrf8D5vLH/De373nZebPu4kF87/LZ0/5HavX/oKPPvkJOdkLOeWUCV6Pduub8Onrh+/XrIYjz3OPwL78F9DZV/Phowfdt2ff7h6Z3SLurq5lafXh7vpPNhzkjtICDAx2dvWwtLqWpdWHy1e1nbwwOO/7KCz90+H7T74Gd1znHkz7e/dCTT0kJ8JnjoLX/wBJFlrlHn/8cZYtW+a9/+abb3L11VfzrW99a8Dz7HY76enppKSkhDnCyLv7dVj62uH7T66CO86HIwvg5y9DfStkJMHZc+DpKBrs0V9311WwtL7Ce//JQ1XckT+XssJ5vDjtFK6tWcN/1n/K1IQUHi/9DHOirD7ztgfXs/X+dd77e1/awezvHsOR3zsWZ7eDfa/t4vg/LIlghMFRvuweNj96t/f+7tf/l/nX3MqCa29nwTW38f7PL6PrYB2JWXkc9Y2fUHRi9I5n8f/bu/MwK6ozj+Pft4Fu6IV9xygIBBeMC7jvEzUawXEBNQ4uIWQ0M6OYxESHZGY0M5jRDHEjTjRqHMcZiWKMYVGzKCq4REBERR0QGqFpoFl7gd7f+eNUN5fm3l64t/s29O/zPOep5546p+pU19tVdc+tOlX4/MMUPvdQ/edtb77IoIk30//iG1jzwPeo2PQFGV1z6HvuFQy+6tY0tjQJO9fC+49Apyx4IOZmposegdF/A5c/HwaO//0kGHwyXDqrTZpVVlbAqs+fJSMjk9/9fs/Lt08ccydDh45n3vyvU7ZrAwALXv82AOPH/ZHcnCFt0r6Edm2D/LcgozO8HDPUx7FXQVkR7NoarpNir5XG/azt29lK1pZX8simrWSZMfAve14a+cjwQ8jMyODm1etZX1FFtwzjxLxsXj56OF0zDr77f63d/OrRjpiZ++J0tyKwsfDGG2+kuxkpcdZZZ+Ez090KaW/sH8AbjtuU7DKXz2b8Z99K6TIPJnNGPc433q5MdzOkEc+cmsmY5w6CN3e3sSUTR8K0g+y65m7jG1etSHcrDjjP/OYo+OsH0t2M9u3FqfjpqX+c2xYto71cR8c6mK6p49F1dttojevW1qBr4cQOxOtgXRc27oC6/msn13Ud5jqpla51DlS2aNkSdx/b2us5+LrORURERERERERERKRNqaNZRERERERERERERJKijmYRERERERERERERSYo6mkVEREREREREREQkKepoFhEREREREREREZGkmPsB8nbONmRmRcDadLdDREREREREREREJEmHuXu/1l6JOppFREREREREREREJCkaOkNEREREREREREREkqKOZhERERERERERERFJijqaRURERERERERERCQp6mgWERERERERERERkaSoo1lEREREREREREREktI53Q1ojzp37uwZGeqDFxGR9qu2ugbTuapVOI6uA6Sjqq2uxczS3QyRtHNr23NBdXUNGRn63xMRkdZRU1Ozxd37tfZ61NEcR0ZGBsXFxeluhoiISEJ9cnryxpH3prsZB6Vz8n9EUVFRupshkha9svvw8IAl6W6GSNrdUnwiRVva7lyQk9ODY465o83WJyIiHcuyZT9Z2xbr0e06IiIiIiIiIiIiIpIUdTSLiIiIiIiIiIiISFI0dIaIiIgkNPrlG+jSJxuvraV2dzU738yn5N31DJwylhWXPl1fbuSjl9Klf+4+ecXvrsMwcscMZtXf/b5+3tFzr6Piix375G2Y+TbbX17JmA9voWZXFeDUlFSy/ZWVrJ+xEGq9TbZbRKQ1XDj7UHoflcWs41ZRWxmOZ2fcN4jDL+tObWUtAKXrq1n3x1KWz9xKVUnIG3FlD06fMZCa8lCnfGs1G9/exfKHtlK8uio9GyMish/mzbuVfv3yuOCCGezYsas+f9asmxg1ahCLFq3k+OMPBaBLl86AU1VVA8D8+cuZPn1uOpotIs2kjmYRERFp1Kqb51Dyzjq69M9h5C8vxatr6Ta8N517daN6+27oZHT7cl9qy6v3yss5diAbfvEOmDHwW2Mgw6DW6dwnG+ucQfaR/fbK63pYT0qXbKhf7ycT/peKdTvJGtaLUU9cTnn+drY891Ea/xIiIvsv95AuDDi5G1XFtRx6QS75c0vq53308FaW3ruFTllGryOyGPvjflz84mHMvTif6t2hc7loyW7mX/oFlgG5X+rC0Tf25pKXhzJ3/Fp2fFaZrs0SEWmxgoIdXHjhMcya9S4AI0b0JyurCwDTp8+lsHAHAHfddSmbNhXz8MOvpq2tItIyGjpDREREmqVqcxk7F+bTpV8OFet2kjtmMADZR/an/PNtlC4u2CvPzNj18WZ2fbQpdCwf0ReAvLFDKHlvPeX5O/bKK/9iB1VFZfust2LNdkqXbqDbyD5ttKUiIqk3fGJ3ipbuZuWzOxkxsUfcMjUVzpYPyvnTDQVk9erEyKv3Lee1ULK2inembWLj27s5/vt9W7vpIiIpNW/eB4wbd2z95/Hjj2Pu3A+aVbdXr2xmzpzEq6/+kAULbue++75B37659fMvu2wM8+bdysKF05gzZyrnnXdU/bwrrhjLCy/czIIFt/Pgg39Dv355qdsoEQHU0SwiIiLN1GVALj3OHMruT4soWVJA7pghAOSNGUzJ0gJK39+wV17p8o14dS1eXUvZh5vq5+WOGUzpkg2ULt3QIK8g7nq7Ht6b3BMGs+uTojbYShGR1jFiQg9W/7aY1b/dyZBzcujat1PCstVltWx4o4wBJ2c3usy1L5U0WUZEpL358MP15ORkMWxYXzIyjAsuGM38+c3raDYzZs9+j4su+jnjxt0PwG23XQhAXl5Xvvvd87nxxv/ijDPuZvLkJ/j8880AfO1ro7nmmlOYOvV/Oe+8n/Hpp4VMn35F62ygSAemjmYRERFp1PAHLubYRTcy6qkJlCwuoPBX71G6uIC8+k7iIZQu3UDJ0g175y3e03FcsnhPx3TuCYMpXVqwd0fzCYMpWbx3R/ORz36DYxf9LSNmjmfLbz9m6+9WtMXmioikXP+TupF7SBfWzClh64cVFK+t5PDLujdaZ9emajJ7Ju6MBti1sZqsJsqIiLRH4a7m4zjllOHk529h8+aSpisB27aVsWDBZ1RUVFNaWsETT7zJCScM3avMiBH9yczsTFFRCWvWbAHC3cyPPfY6X3yxlerqWh59dAHHHXcovXvnpHrTRDo0jdEsIiIijfp86jxK3lm3V17JkgIOu+urdOqeRc5XBlJ2+yvU7q6iS79sOnXPIvf4QWz672X15UuXFNDv6q/QqXsWnXt1o+KLnVRt3c3Q6efTqXsW3Ub02Wt8ZoBPrnyGinU722QbRURa04iJPSh4vYyKbeGFVqtfKGbElT1Y8avtCetkD+xM5Y6aRpebPagzFU2UERFpj+bN+4DHH5/MkCE9mTt3WdMVItnZmfzgBxdx8smHk5fXFaB+fOeSknKmTXueSZNO4yc/uYylS9cyY8YrrFu3jUGDevCjH43nH//x4vpl1dTUMmBAd7Zt23foNhHZP+poFhERkRarXF9MVVEZfSeMprKwhNrdVQCUfrCRvhNGk5GdSdnywvrypR8U0ikvk74TRlO2LOTXllVStTkso6qojMqC4rRsi4hIa+rU1Rg2Pg/rZFy1bETIyzSyenai11FZcet0zjYGn5nD8ge3NLrswy7MY9O7u1LeZhGR1lZYuJOCgu2cfvpI7rzzxWbX++Y3z2TAgO5MmvQo27aVccwxh/DEE5Pr5y9cuJKFC1fStWsXpk49n2nTxvGd7zzFpk3F3H//H/jznz9pjc0RkYg6mkVERGS/lC7dwIDrjmfbvM/2yhs4eQy7Pt6EV+y5y84ratj18WYGXHc8hY++t6f8+2EZxe980aZtFxFpK4demIvXwO++upraSq/PP+eRIYyYsPfL/jIyjV6jMhn74/5U7qxh5ax9n+qwDMgZ0oWjb+zNwNOymTd+batvg4hIa7jrrhfp3r0b5eVVdOrUvJFdc3IyKS+voqSknJ49s5ky5az6ef365TFq1EAWL86nsrKa3bsrqampBWD27PeYMuVsVq3azNq1W8nL68pJJw1Tx7NIiqmjWURERPZLyeICen99FKVL9wx5Ubp0A11+cCZbXth3POWSxQXkHjeI0vf3Lt//mmP3GTZDRORgMWJiD1b+ZidlBdV75X/y6+2c/K8DKHyzjNF/14ejpvSCDKN0XRXr/1TKaw9tpXr3no7pfmO6MWnll8GgYlsNG9/axZyL8tm5qrKtN0lEJCXWr98OJB5CKJ6nnnqLu+++ggULbmfTpmKeeeZdTjstelqkUwaTJ5/JT386AXdnxYpC7r57DgAvv/wRXbtmMmPG1QwY0J2SknIWLVqpjmaRFDN3b7pUB5OZmenFxXp8V0RE2q8+OT1546h7092Mg9I5+T+iqKgo3c0QSYte2X14eOCSdDdDJO1uKT6Roi1tdy7IyenBV75yR5utT0REOpZly36yxN3HtvZ6mvdsgoiIiIiIiIiIiIhIAupoFhEREREREREREZGkqKNZRERERERERERERJKijmYRERERERERERERSYo6mkVEREREREREREQkKebu6W5Du2NmJcBn6W6HHBT6AlvS3Qg5KCiWJBUUR5IqiiVJFcWSpIpiSVJFsSSpoliS9mSUu+e19ko6t/YKDlCfufvYdDdCDnxmtlixJKmgWJJUUBxJqiiWJFUUS5IqiiVJFcWSpIpiSdoTM1vcFuvR0BkiIiIiIiIiIiIikhR1NIuIiIiIiIiIiIhIUtTRHN+j6W6AHDQUS5IqiiVJBcWRpIpiSVJFsSSpoliSVFEsSaoolqQ9aZN41MsARURERERERERERCQpuqNZRERERERERERERJKijmYRERERERERERERSYo6miNmdoiZPWFmG8yswszyzex+M+uV7rZJepjZBDN7yMzeNLNiM3Mze7qJOqeZ2Xwz22Zmu8xsuZndamadGqkzzswWmNlOMys1s3fN7PrUb5Gkg5n1MbMpZvaCma0ys93Rvl5oZt8ys7jHYcWSxGNm95jZn81sXRRL28zsfTP7FzPrk6COYkmaZGbXRuc5N7MpCcq0OC7M7Hoz+0tUfmdUf1zrbIW0teh62ROkjQnq6JgkCZnZmWb2vJkVRt/JCs3sD2b29ThlFUuyFzO7oZFjUl2qiVNPsSRxmdnF0TFofXTtvdrMnjOzUxOUVyxJq7Bgspm9Y2YlUXy9b2a3NIwvM+tiZlPN7NdmtszMKhu7xo/qnG5m95rZe2ZWFJ2D15jZY2Y2okVt1RjNYGbDgbeA/sCLwKfAScC5wGfA6e6+NX0tlHQws2XAsUApsB44Avgfd5+UoPxfA88D5cBvgG3AeGAUMNvdJ8ap8w/AQ8DWqE4lMAE4BJjh7releLOkjZnZTcB/AoXAa8AXwADgcqAHIWYmeszBWLEkiZhZJbAUWAFsBnKAU4CxwAbgFHdfF1NesSRNMrMvAR8CnYBc4Nvu/liDMi2OCzP7D+D7hHPobCATuBroDdzs7jNba5ukbZhZPtATuD/O7FJ3/48G5XVMkoTM7MfAvwJbgLmEa6e+wPHAa+7+w5iyiiXZh5kdB1yaYPaZwF8B89x9XEwdxZLEZWb3AD8k7OffEY5NI4BLgM7Ade7+dEx5xZK0GjN7CriW8B1wDlAGnAccRYM+BTPrCWyPqm4ixNWXiHONH7P8jUA/Qt/oEqAaOBU4LVrX+e7+drMa6+4dPgGvAE740hOb//Mo/5fpbqNSWuLiXGAkYMA5USw8naBs9+gfvgIYG5PfNfpHdeDqBnWGEk5CW4GhMfm9gFVRnVPT/XdQSjqO/opwgZHRIH8godPZgSsUS0rNjKeuCfKnR/v5YcWSUktSdI77E/A58LNoH09JNi4IF6Ueze/VYFlbo+UNba3tUmqz+MkH8ptZVsckpcbiY2K0P/8I5MWZ30WxpJRMAt6O9vMliiWlZsTLQKAG2Aj0bzDv3Gg/r1YsKbVFIvyA5sBqoG9MfhfghWjeDTH5mcBFwKDo853EucZvsI7bgcFx8qdFdT9sbns7/NAZZnY4cAHhQvkXDWb/C6Hn/lozy2njpkmauftr7r7So/+uJkwg/Pozy90XxyyjHPhx9PE7DepMBrKAme6eH1NnO3B39PGm/Wy+tBPu/qq7z3H32gb5G4FfRh/PiZmlWJKEojiI59loOjImT7EkzXEL4QexbxKueeLZn7io+zw9KldXJ59wvZUVrVM6Dh2TJC4Lw4jdA+wCrnH3koZl3L0q5qNiSVrEzEYTngArAObFzFIsSSKHEYaafdfdN8fOcPfXgBJC7NRRLElrujyaznD3LXWZ0bnxn6KPN8fkV7r7S+5e2NwVuPs97r4hzqx7gN3AaEswVGNDHb6jmfDlCuAPcTqCSoBFQDbhxCSSSF0cvRxn3huEC+fTzCyrmXVealBGDk51X5qqY/IUS7I/xkfT5TF5iiVplJkdCfw78IC7v9FI0f2JC8VSx5FlZpPMbFo0HuC5Ccai1DFJEjkNGAbMB7ZHY6LeHsVTvHFQFUvSUjdG08fdPXaMZsWSJLKSMNzASWbWN3aGmZ0F5BGeCKujWJLWNDCaro4zry7vhGjIjFRz9vRX7DPGfTzqaA7j5QD8X4L5K6Ppl9ugLXLgShhH7l4NrCGM43R4M+sUEu4sO8TMslPbVGkPzKwzcF30MfbiQrEkTTKz28zsTjO7z8zeJIxpuZzQaVhHsSQJRceg/yYM4TOtieItiovoKbAhhDF6491JoWurg8tAQixNJ4zV/Cqw0szOblBOxyRJ5MRouonwHoK5hPPZ/cBbZva6mcXeOahYkmYzs27AJKAWaDg2qWJJ4nL3bYShBAYAK8zsUTP7qZk9C/yBMMzPjTFVFEvSmuruYh4WZ15sTB3RCuueSPhh5R1339GcCupoDi/jAtiZYH5dfmv8MiAHj/2Jo+bW6ZFgvhzY/h0YDcx391di8hVL0hy3EYZ3uhU4g/BjxQXuXhRTRrEkjflnwgu2bnD33U2UbWlc6Nqq4/g18FVCZ3MOcAzwCGHcyZfM7NiYsjomSSL9o+lNQDfCy43yCNdJrwBnAc/FlFcsSUtcSYiFlzzmhckRxZIk5O73E4Ys6Ax8G7iD0Om2DniywZAaiiVpTXOj6ffMrHddZnTjyF0x5XqlcqVmNozwsspqwgu+m0UdzU2zaNqccXpFEtmfOFLsHaTM7BbCgfpTwptjW1Q9miqWOjB3H+juRujcuZzwS/b7ZnZCCxajWOqgzOwkwl3MM7y5b49uYpHRtKVxoTg6wLn7XdG7CDa5+y53/8jdbyK8ULsb4eUzzaVjUsdVN9SKARPc/c/uXuruHwOXAeuBsxMMoxGPYkli/W00fWQ/6iqWOjAz+yEwG3gSGE74QXUMYaiC/zGze1uyuGiqWJL9MYswnMpw9txhfz+wDPg6e54WbNbQFs1hZv2jdfYDprr7W82tq47mpn8l6t6gnEg8+xNHza1TnES7pJ0xs78HHgBWAOdGj2XFUixJs0WdOy8QXmrbB3gqZrZiSfYRM2TG/7Hn5SFNaWlcNFW+qTt45MBX97Lbs2LydEySROpeGLra3T/nv/ChAAAFuUlEQVSInRE9cVH35NdJ0VSxJM1iZkcRxgBfTxgDvCHFksRlZucQXoL2e3f/nruvjn5QXUr4AawA+L6Z1Q1boFiSVhO9T+4SwlOtGwk3q00mHNvOALZGRTfHXUALRZ3MrxKGd5nq7g+3pL46muGzaJponMCR0TTRGM4i0EgcRV/qhxEeN1jdzDqDCL+Yrnf3XaltqqSLmd0KzAQ+InQyb4xTTLEkLebuawk/Xhwd88ISxZLEk0vYv0cC5WbmdYkwHAvAr6K8+6PPLYoLdy8jfAHLjeY3pGurg1/dF52cmDwdkySRuv2caOzHuo7obg3KK5akKYleAlhHsSSJjIumrzWcEe3bvxD6046PshVL0qrcvdrdZ7j7ce7ezd27u/uFhO+AxwG7gY+TXU8UewuAo4C/d/cHW7oMdTTvOXBcYGZ7/T3MLA84nbDD3mnrhskB5dVoemGceWcB2cBb7l7RzDoXNSgjBzgzux24j/B4y7kNxvSKpViS/TU4mtZ9kVIsSTwVwOMJ0vtRmYXR57phNfYnLhRLHVvdEAexX6h1TJJE3iB0wIw0s8w480dH0/xoqliSJplZV8Jdf7WEc1o8iiVJJCua9kswvy6/MpoqliRdrgW6As+6e1UyCzKzQ4DXCS8VvKmldzLXc/cOnwiPYzlwc4P8n0f5v0x3G5XSm4Bzolh4OsH87kAR4Qv82Jj8rsBbUd2rG9QZBpQTHnMYGpPfC1gV1Tk13duulJL4+adofy4GejdRVrGklCg2jgAGxsnPAKZH+3mRYklpfxNhPF0HpiQbF4RHlT2a3ysmf2i0nPLYZSkdeAk4Ot45DTiMMFagA9Ni8nVMUmosnp6O9ue/Ncg/n9BRuAPoqVhSakFMXRvt1zmNlFEsKSWKjSujfbkRGNJg3kXRcWk30EexpNQWCegeJ+9EYBtQAhzeSN07iXON36DMocDnhJuWvplMWy1aYIdmZsMJ//z9gReBT4CTgXMJj3We5u5bEy9BDkZmdilwafRxIPA1wp05b0Z5W9z9tgblZxNOFrMI//CXEMa1mQ1c6Q3+4czsZuBBwsnlN4RfRCcAhxBe0nQbckAzs+sJL5CoIbyxNd6YpPnu/mRMHcWS7CMaeuVnhDu/Pifs6wHA2YSXAW4EvuruK2LqKJak2czsTsLwGd9298cazGtxXJjZDOB7hPHjZgOZwFWE8cRvdveZrbYx0uqieLmD8HTgGsKXnOHAxYQv1vOBy9y9MqaOjkkSVzQe5CJgBOFa+y+EHy0uI3w5vsbdn4spr1iSRpnZm4SxSy9x9zmNlFMsyT6ip91fAc4jnN9eIFxrH0kYVsOAW939gZg6iiVpNWb2LuHHjY8IMXk04UWAFcDl7v5Kg/J3EG5UgjC0xrGEfs+6FwcujL3eN7M1hBtClgBzEzTjSXfPb7Kx6e6Vby8J+BLwa6CQ8M+9lvDCrkbvPlQ6eBN7fvVJlPLj1Dmd8MVqe3QQ+BD4LtCpkfWMJzyeUAKUAe8B16d7+5XaLI4cWKBYUmpGLI0GfkEYfmUL4THjndF+vjPR+UqxpNTcRBN3O+xPXADXR+XKonqvA+PSva1KKYmXs4FngE8Jd5tWEe7m+iNwHYQbWuLU0zFJKdF+7k14onQN4fvYVsJNQKcolpRaGEtHRuezdY3FQ0x5xZJSvH3cBbiVMIxqcXTtvZnQCXeBYkmpLRPwA0In8A5C5/IawsuXhyYov4DG+yCebFC+qT4LB85pTlt1R7OIiIiIiIiIiIiIJEUvAxQRERERERERERGRpKijWURERERERERERESSoo5mEREREREREREREUmKOppFREREREREREREJCnqaBYRERERERERERGRpKijWURERERERERERESSoo5mEREREREREREREUmKOppFREREREREREREJCnqaBYRERERERERERGRpKijWURERERERERERESS8v+IWtZ+7qnT7QAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 1728x576 with 3 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Plotting the predicted fitness effect for the variants seen more than once\n",
    "plt.close('all')\n",
    "f, (ax1, ax2) = plt.subplots(2, 1, sharey=True, sharex = True, figsize=(24, 8))\n",
    "gs = matplotlib.gridspec.GridSpec(3, 1, width_ratios=[1], height_ratios=[10,0.5, 0.5])\n",
    "ax1 = plt.subplot(gs[0])\n",
    "ax2 = plt.subplot(gs[1])\n",
    "ax3 = plt.subplot(gs[2])\n",
    "gs.update(hspace=0.05)\n",
    "\n",
    "nonsense_list = []\n",
    "y_list = []\n",
    "x_list=[]\n",
    "y_list_top20 = []\n",
    "x_list_top20=[]\n",
    "y_list_nonsense = []\n",
    "x_list_nonsense = []\n",
    "x=[]\n",
    "y=[]\n",
    "OBS_EXP_DICT={}\n",
    "top20_lower_err = []\n",
    "top20_upper_err = []\n",
    "missense_lower_err = []\n",
    "missense_upper_err = []\n",
    "nonsense_lower_err = []\n",
    "nonsense_upper_err = []\n",
    "\n",
    "top20 = c1\n",
    "missense = c6\n",
    "nonsense = c10\n",
    "synonymous = c3\n",
    "\n",
    "axisfont=20\n",
    "domainfont = 12\n",
    "titlefont=12\n",
    "axislabelfont=12\n",
    "\n",
    "top_20_DNMT3A = ['p.R882C', 'p.R729W', 'p.R326C', 'p.R320*', 'p.R882H', 'p.R736H', 'p.Y735C',\\\n",
    "                'p.R736C', 'p.W860R', 'p.R771*', 'p.R598*', 'p.P904L']\n",
    "\n",
    "PWWP_col = 'mediumvioletred'\n",
    "ADD_col = 'darkviolet'\n",
    "MTase_col = 'midnightblue'\n",
    "\n",
    "for k, v in DNMT3A_missense_s.items():\n",
    "    if k in top_20_DNMT3A:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_top20.append(codon_number)\n",
    "        y_list_top20.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        top20_lower_err.append(error_lower)\n",
    "        top20_upper_err.append(error_upper)\n",
    "\n",
    "    if k not in top_20_DNMT3A:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list.append(codon_number)\n",
    "        y_list.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        missense_lower_err.append(error_lower)\n",
    "        missense_upper_err.append(error_upper)\n",
    "\n",
    "    missense_err = [tuple(missense_lower_err), tuple(missense_upper_err)]   \n",
    "    \n",
    "for k, v in DNMT3A_nonsense_s.items():\n",
    "    if k in top_20_DNMT3A:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_top20.append(codon_number)\n",
    "        y_list_top20.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        top20_lower_err.append(error_lower)\n",
    "        top20_upper_err.append(error_upper)\n",
    "    \n",
    "    top20_err = [tuple(top20_lower_err), tuple(top20_upper_err)]\n",
    "        \n",
    "    if k not in top_20_DNMT3A:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_nonsense.append(codon_number)\n",
    "        y_list_nonsense.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        nonsense_lower_err.append(error_lower)\n",
    "        nonsense_upper_err.append(error_upper)\n",
    "\n",
    "    nonsense_err = [tuple(nonsense_lower_err), tuple(nonsense_upper_err)]        \n",
    "    \n",
    "msize = 12\n",
    "medgewidth = 2\n",
    "linew = 3\n",
    "csize = 4\n",
    "cthick = 3\n",
    "\n",
    "markers1, caps1, bars1 = ax1.errorbar(x_list, y_list, yerr = missense_err, fmt = 'o', ecolor = missense, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = missense, markerfacecolor = missense, zorder = 1)    \n",
    "\n",
    "markers2, caps2, bars2 = ax1.errorbar(x_list_nonsense, y_list_nonsense, yerr = nonsense_err, fmt = 'o', ecolor = nonsense, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = nonsense, markerfacecolor = nonsense, zorder = 2)    \n",
    "\n",
    "markers3, caps3, bars3 = ax1.errorbar(x_list_top20, y_list_top20, yerr = top20_err, fmt = 'o', ecolor = top20, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = top20, markerfacecolor = top20, zorder = 3)    \n",
    "\n",
    "# loop through bars and caps and set the alpha value\n",
    "a = 0.7\n",
    "[bar.set_alpha(a) for bar in bars1]\n",
    "[cap.set_alpha(a) for cap in caps1]\n",
    "[bar.set_alpha(a) for bar in bars2]\n",
    "[cap.set_alpha(a) for cap in caps2]\n",
    "[bar.set_alpha(a) for bar in bars3]\n",
    "[cap.set_alpha(a) for cap in caps3]\n",
    "\n",
    "#Only show the required axis lines\n",
    "ax1.spines['top'].set_visible(False)\n",
    "ax1.spines['right'].set_visible(False)\n",
    "\n",
    "for axis in ['bottom','left']:\n",
    "    ax1.spines[axis].set_linewidth(2)\n",
    "\n",
    "for axis in ['bottom','left']:\n",
    "    ax1.spines[axis].set_color(grey3)\n",
    "    \n",
    "ax1.set_ylabel('fitness effect (s)', fontsize = axisfont, labelpad = 10)\n",
    "ax1.set_title('Fitness effects across DNMT3A', fontsize = 18)\n",
    "\n",
    "# ax1.set_yscale('log')\n",
    "\n",
    "# STACKED BAR CHART\n",
    "# define the exons + create data\n",
    "exons = [{'2':(24-1)+1, '3': (59-25)+1, '4': (150-60)+1, '5':(164-151)+1, '6':(213-165)+1, '7': (285-214)+1, '8': (338-286)+1,\\\n",
    "          '9': (374-339)+1, '10': (427-375)+1, '11': (477-428)+1, '12': (492-478)+1, '13': (518-493)+1, '14':(556-519)+1, \\\n",
    "          '15': (617-557)+1, '16': (646-618)+1, '17': (694-647)+1, '18': (725-695)+1, '19': (774-726)+1, '20': (803-775)+1, \\\n",
    "          '21': (826-804)+1, '22': (866-827)+1, '23': (912-867)+1}]\n",
    "exon_orders = [['2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23']]\n",
    "colors2 = [c4, c0, c3, c1, c5, c6,c7, c8, c9, c10, c2,c4, c0, c3, c1, c5, c6,c7, c8, c9, c10, c2]\n",
    "names = exons[0].keys()\n",
    "values = np.array([[data[name] for name in order] for data,order in zip(exons, exon_orders)])\n",
    "lefts = np.insert(np.cumsum(values, axis=1),0,0, axis=1)[:, :-1]\n",
    "orders = np.array(exon_orders)\n",
    "\n",
    "#plot the bar chart\n",
    "for name, color in zip(names, colors2):\n",
    "    idx = np.where(orders == name)\n",
    "    value = values[idx]\n",
    "    left = lefts[idx]\n",
    "    ax2.bar(x=left, height=-10, width=value, bottom=-5, color=color, orientation = 'horizontal', label = name, edgecolor='k', linewidth=1, tick_label = name)\n",
    "    ax2.text(left + (value/2.), (-0.26), name, ha='center', va='center', fontsize = axislabelfont)\n",
    "    \n",
    "\n",
    "# STACKED BAR CHART of domains    \n",
    "domains = [{' ': (291-1)+1, 'PWWP' :(350-292)+1, '  ':(481-351)+1, 'ADD': (614-482)+1, '   ': (633-615)+1, 'MTase': (912-634)+1}]\n",
    "domain_orders = [[' ','PWWP','  ','ADD','   ','MTase']]\n",
    "# blank_col = c13\n",
    "blank_col = grey1\n",
    "colors3 = [blank_col, PWWP_col, blank_col, ADD_col, blank_col, MTase_col]\n",
    "names_domains = domains[0].keys()\n",
    "values = np.array([[data[name] for name in order] for data,order in zip(domains, domain_orders)])\n",
    "lefts = np.insert(np.cumsum(values, axis=1),0,0, axis=1)[:, :-1]\n",
    "orders = np.array(domain_orders)\n",
    "\n",
    "#plot the bar chart\n",
    "for name, color in zip(names_domains, colors3):\n",
    "    idx = np.where(orders == name)\n",
    "    value = values[idx]\n",
    "    left = lefts[idx]\n",
    "    ax3.bar(x=left, height=-10, width=value, bottom=-5, color=color, orientation = 'horizontal', alpha = 0.9, label = name, edgecolor='k', linewidth=1)\n",
    "    ax3.text(left + (value/2.), (-0.25), name, ha='center', va='center', fontsize = domainfont, color = 'white', fontweight = 530)\n",
    "\n",
    "    \n",
    "\n",
    "# CONFIGURING THE GRAPH APPEARANCE\n",
    "#Set the x and y axis limits\n",
    "ax1.set_xlim(0, 912)\n",
    "ax1.set_ylim(-0.01, 30)\n",
    "ax2.set_xlim(0, 912)\n",
    "ax2.set_ylim(-0.5, 0)\n",
    "ax3.set_xlim(0, 912)\n",
    "ax3.set_ylim(-0.5, 0)\n",
    "\n",
    "#x-axis ticks\n",
    "x1_major_ticks = []\n",
    "x1_major_tick_labels = []\n",
    "x2_major_ticks = [0, 100, 200, 300, 400, 500, 600, 700, 800, 912]\n",
    "x2_major_tick_labels = [\"0\", \"100\", \"200\", \"300\", \"400\", \"500\", \"600\", \"700\", \"800\", \"912\"]\n",
    "ax3.set_xticks(x2_major_ticks)\n",
    "ax3.set_xticklabels(x2_major_tick_labels, fontsize = axisfont)\n",
    "ax1.set_yticklabels([])\n",
    "ax1.set_yticks([])\n",
    "ax1.set_xticklabels([])\n",
    "ax1.set_xticks([])\n",
    "ax2.set_yticklabels([])\n",
    "ax2.set_yticks([])\n",
    "ax2.set_xticklabels([])\n",
    "ax2.set_xticks([])\n",
    "ax3.set_yticklabels([])\n",
    "ax3.set_yticks([])\n",
    "\n",
    "#y-axis ticks\n",
    "y1_major_ticks = [0, 5, 10, 15, 20, 25, 30]\n",
    "y1_major_tick_labels = [\"0\", \"5\", \"10\", \"15\", \"20\", \"25\", \"30\"]\n",
    "ax1.set_yticks(y1_major_ticks)\n",
    "ax1.set_yticklabels(y1_major_tick_labels, fontsize = axisfont)\n",
    "\n",
    "#set panel legends\n",
    "legend_marker_size = 2\n",
    "legend_elements = [Line2D([0], [0], marker = '_', color = missense, alpha = 1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='missense'),\n",
    "                   Line2D([0], [0], marker = '_', color=nonsense, alpha=1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='nonsense'),\n",
    "                  Line2D([0], [0], marker = '_', color=top20, alpha=1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='top 20 observed variant')]\n",
    "    \n",
    "ax1.legend(ncol=1, handles=legend_elements, loc='upper left', frameon=0, fontsize = 20)\n",
    "\n",
    "ax1.grid(which='major', linestyle=':', linewidth='0.7', color=grey3, zorder = 0)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABacAAAHvCAYAAABAPWMbAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAIABJREFUeJzs3Xd4VFX+x/H3CYFAKAktdIiggKKCoICAC9IUkCKiKKyAWHZFQVbFVVdNgrriIiqIgD8URcWyIk2UpoAUAek1dEINJYTQQiDl/P64M7MzmUkILQn6eT3PfWbm3HPOPfeeufPoN4fvNdZaRERERERERERERERyU1BeD0BERERERERERERE/nwUnBYRERERERERERGRXKfgtIiIiIiIiIiIiIjkOgWnRURERERERERERCTXKTgtIiIiIiIiIiIiIrlOwWkRERERERERERERyXUKTouIiIhcRYwxwcYYa4z5OK/Hkh8ZY/obY7YYY865rlNlV3krY8wyY8wpV/lf83qsIiIiIiJ/dgpOi4iIiOQDxpgWrqBpVlvjbNoGGWOijTGdcnPM+Y0xpg0wAtgAPAE8DCQaY8oAk4DCwLOu8sVXaAz1XXNR9Ur0/2dljLk20/2QboxJcv0h4htjzH3GmOAA7Vp7tekTYL/7jz1TMpUvcpWfM8aUzWJMH3r13cyrr5xulY0x5Y0xnxpj1htjjhljzhhjthljxhpjqp/nmgxw9XPMGFPkAi+piIiISL7g9x9wIiIiIpKnvgZ+ClC+HcBam+YKRKV57QsCooBPgGlXfIT5VxvX6yPW2hPuQmNMC6AE8Kq19kpfn/o4c/EzsOcKH+vPaBbwpet9MeBaoAPQHVhhjLnXWrsvi7aDjTHfWGtTcnisVJx766/Ae947jDGFgYeAFJw/egCk4/zhw1sL4FFgNPBbpn2JwDXAdcBMnO/LGaAW0Bd4wBjT0Fq7JYvx9QV2ADWArsCEHJ6XiIiISL6h4LSIiIhI/rLKWvtldhUuILj2Z1MeSPcOTHuVgxMMlGwYYwxQ1Fp7Kq/HkoXNme8PY8wg4DlgKDDdGNPAWpueqd0K4Fagv6teTiTjBJQfIVNwGrgXKAl8BfQAsNZa/hc4d4+tME5w+rcs7uuNQLPMhcaYSa5jPwUMCLC/AVDXdewXcQLVCk6LiIjIVUdpPURERESuIplzThtjrsVZ4QnwqFfKgLTM9V2pBxYaY5KNMQnGmP8zxhQNcIxKxpgxxpi9rrQG+12fy2SqV9oYM9wYs9MYk2KMOWqMWWGMeTZTvT7GmOWuNAynjTE7jDFfGmNK5/Cc7zLGzDHGHHcdZ60x5gmv/dcaYyzOqtUCXtfgZ2PMPpwV5QALva+Nq21hY8wrxphNrr6PGWOmGWPqBhiHMcb8zRjzu3FyV580xqwzxkS59r8BjM10LO+5KmKMGWycVBTJrmOtN8YMycE1CHaNc6Ex5pBrXna7UkuUyqLN/caYX13XLdkYs9k1X8Gu/e6UFw8bJ1d3LHAWGOjVRwvXdTzh6mOlCZwe4yZjzPeu78pZY0y8MWauMaadV52LPv/sWMc7wLc4Adv7A1T7ClgLvGSMKXkB3X8K3OQKBnt7BFgJrL+IIefEbtdrVmN9FDgBTAHGA3caY665QmMRERERuWK0clpEREQkfwnNHAQGzlprT2ZR/yDQGydANZ//BWIzMtVrgLPa8xOc1Z0tgcdx0oP0c1dyBbh+Awq46u7ESTvwJE4A7DavlcmTgNuBMcA6oChwPU4qg3dd/T0CjAN+BV7DSVtQFWgPlAGOZncxjDFPAh+6xvQ6zmrWu4CPjDHXWGtfcl2Dh11jbOy6HgDxOOk8OuAE814HtrqvjTGmEDAbaAR8jpOvuqTruvxmjGlmrV3tqmtwUq50B5YAbwJJrvO9D4gBvsNZpe19LHClZHFdp17AZ64+CrqubcvsroFLYZzVwd/jBCRPAw1xcms3dc2L+48UGGPeBl7AWZk7DDiEk/7hPuBfgPfK6Odc5/2J61rudvXRBZjouo5DXcd8CPjUde3dQfmywFyctBZjcNJTlAFuc41xxmU4/5z4GGd+OgDfZNpncVYYzwBewrk2OTEV5zvaFycYjXHyibfCWYVd4pJHjee7WAIohJOqJMa1yy/Fj/lfSpH/WmvPGGO+BN7GCZi/djnGIyIiIpJrrLXatGnTpk2bNm3a8njDCejaLLZvvOoFu8o+zq4swL504NZM+2bhrJQt4lX2I06AsmKmuo1cfbzi+lzK1e+I85zXNJx0GgUu4ppUdo3v8wD7PsQJrFfzKvsSSAtQ9zHXWJtlKh+EE6hunak8HNgH/OxV1sPVx2dAUKb6Qec7lmvfCWDaRX4/grznyav8b67jdfUqa+IqmwOEZKpvvN63dtVLAMpkqlfQdQ0SgfJe5SHAUtd3obqrrGvmMWRxDpdy/te6jvF+NnUiXHWWBTjHga7Pv+D8gaRypvtjSqa+FgFJrvfDXdchxPXZ/UeWcJyAd8D5zvR9+Ot5zq8Lvvd8PPBMFnV7Zj4m8APOHwWCsjuONm3atGnTpk1bftuU1kNEREQkf/k/nAf7eW9vXIZ+F1lrV2Qqm4uzUrMagCs9RDuclbnnjDFl3BvOg9d2AW1dbZNx0ok0NsZUy+a4x4HiQHvX6uMLcb9rfOO8x+Iazw84q7tbXWCf3v6Ks7J4Taa+g3GCmM2NMSGuuu6A4PPWWp9V6Zk/Z+M4ToqIOhc6UGtthrX2DIAxpoAxJtw11rmuKo28qvd0vb5orT2bqR8boPvPrLUJmcpuAyrh/MHjoFf7s8A7OMHyTl7nBc4cF8/mNC76/HPIvaI/u9XM/8QJsA++gH7H4aws7+L6DvcGJltrky5qlIEtxrnXO+Gs7D4IlHKnYMnkUWCHtXaRV9lnQBX+91BQERERkauC0nqIiIiI5C/brLU/X4F+dwYoc6fUcOd+rg0YnNW4f8uin3RwHsponNzS7wJxxpiNOIHSydbaeV7138B54Ns0IMEY8ytOaoVv7fkfune963VeNnXKnaeP7NTGCX4fyaZOKZxVrNcB+wIEcS/EMzjpVzYYY3bgnNc0YHoWQWMfxpgHcVJw1MVZ2ezNOzfxdTjztC6H49oaoMydv3hjgH0bXK/VXa9zcXI6Pwr0MsYsx0mX8q21drNXu0s6/xxwB6UzPxDTw1q7whjzHdDbGDMM2HK+Tq21a40xq3HSZhzGOe+/X4bxeh/jCOC+739wpepYh5Me5Sl3PWNMdZx/ZTHaOPnm3WJx0q48ivMvIkRERESuCgpOi4iIiPw5pGezz2R6HY+TIiOQZPcba+1IY8xknBy/fwEeAPobYyZYa//qqrPFGFMbJ71CK6A5Tm7gaGPMX6y1u3Iwrp44QcFAtmdRnhNBwBqc9B5ZSfQayyUFUK21k1zB+fY416ENTtqH+caYttYrZ3RmxpgHcHJeLwUG4KTcSMEJrv+I74POL3SFenKAshz34Qos93Q92LAdcAdOTudXjTH9rbWjXfUu+vxz6GbX6/kCzv/Cyb8+xPWaE+Nw0nsUxEmf8cvFDDCnrLX7jDFzgceNMQO9rk1fnLnph1eueC+djTGlrbXZ5nIXERERyS8UnBYRERG5+l2OVafgBHotUDCnq7ettftxUpH8nysFwQScQOUw63qYoCsVxI+uDWNMJ5wHzQ3EWU2blW2u1yNXaDX5NqAs8EsOVu5uwUlbUeY8q6ez7ccVNPwC+MKVImIozmroe4DJ2TR9GCeIfKe1NsVdaIy5MYuxtgZuAlZlN55s7HC9BkrBcYPr1Wc1vrV2PbAe+I8rRczvOAHg0V51Lvb8c+Ix1+uP2VWy1m43xvwfzorkO3LY9wScdCYtgcEXkMrlUhTBCYYXA44ZY4JwUoqsxLmumVUC3sdJVzM8F8YnIiIicsmUc1pERETkKmetTcdZRVvqEvs5hJOO4X5jzG2Z9xtHWdf7UGNMkUzt03CCk7jH4sqLnNkq7zrZ+BY4Bww2xhQOMJ5wY0yh8/SRnc9xAnoBA+TGGO+UIRNwVqz+J3Pu7Eyf3alKSmWqE2yMCfMucwXE1wSqH0A6zsMbPf/97jruKwHqfuV6fSvz9bmAvN/Lgf3Ao8aYCK/2hYDnXWOZ5iorlblfa20iEAcUM8YUugznny1jzHNAd2A1MDEHTQbjzNXbOenfWnsMJ5VHDDD2IofpJ9N3zLv8Jpz0HVtdxwa4C+choZ9baycG2IYDe3FWV4uIiIhcFbRyWkREROSPYSlwlzHmBZwAVbq19r8X0c/fgEXAImPM5zjBvmCcHMRdgE9w8kjfAPzsSuuxETjmKnsSZ9XtYld/c40xR4CFrnGVwsndm0HWqUMAsNbuNsY8DYwBNrny8O7BWe18M9AZqImT4uJivIuzwvg9Y0xrYD5wEqiKk4LkJP97wNw3QFfX2GsaY37AecBfTVfduq56v+Osnn7FFcg/zf8eJrnHGDMNJyDrzl38JE7qkGxX++IEXDvjXM8vcB7qdy/gF7S31v7myqf8HODOsXwQZw7vN8bccr5839baNGNMf+A7YLkxZqzrXB4EGuKsHnavnO4LPO36LmwH0oA7XdflK2vtOdcfKS7l/N1qG2P+6npfFKgBdMTJH/47cK/rjzXZstYedl2jqBweF2vtZzmtewFeNcY0B37CCeYXwFnx/lfX+6e86j6K893KboX5JOAZY8ytAR6AKiIiIpLvKDgtIiIi8sfwd2Akzkra4jgrbS84OO0KCNcHXgQ6Ab1w0knsBabwv1Wpu3FyU7fACZKG4ASJPwLe9ko98SFwv2t8pXAewrgK6GetnZ+D8Yw1xsTirNZ9EggDEoDNOLmDs3uY4fn6PmeMuRt4GicYGOPadQAn2D/eq641xnTHCRb2xQlqpuOktvjOq94uY8xjODmXR+OkZfjE1W4ETsC2DU5gNR4n0PiWtfbgecb6pTGmGM4q72E4Ad1pwKsEyMdtrX3e9RC/p4B/4qz63gtMx1lln5PrM9kY0wbnO/VPnP93iAX6Wms/9ao6F+ePBR2BCjjB6V04wfGRrjonL+X8vdzl2izOyueDON+nV3EexnnewLSXd3C+l5fyUM1L9QNQEWfVdwROQHofzh9D3rHWxoLnXyB0BJZba/dm09/3ON+RvoCC0yIiIpLvmcvzYOzLzxhTGud/dDrgrB6ohPPPOtcDnwKfBsr1ZoxpgvMf0I1xVpJsx3mAyQdZ/ceqMeYenP/huQXnPwg3AqOsteMD1RcRERERERERERGRS5Ofg9N/x1ltEg/Mw/knnOVw/jllGM6qgPu9H15jjOnsKk/ByVGYiLPCoBYw0Vp7f4DjPA18gLOKx53XsBtOPrdh1trnr9ApioiIiIiIiIiIiPxp5efgdEucf+73o/cKaWNMeZx8clWAbtba713lJXBWSYcBTd051lwPz5kL3A48ZK39xquvSJx/EnoaaGCtjXOVl8R5CEwNoIm1dsmVPFcRERERERERERGRP5ug81fJG9baudbaHzKn7nDloxvj+tjCa1c3nIfjfOP98A9XvkP3U8yfzHSYvjj5EUe6A9OuNseAf7s+/v3SzkREREREREREREREMsu3wenzSHW9pnmVtXS9zgxQfwHOg3yaGGNCcthmRqY6IiIiIiIiIiIiInKZXHXBaWNMMM5T48E3qFzL9bo1cxtrrfuJ4cFA9Ry2icdJ91HZGBOa3ZhGjx5tR48enT/zo4iIiIiIiIiIiIjkQ1ddcBoYAtwI/GStneVVHuZ6PZ5FO3d5+EW0CQu00xjzhDHGk0Lk+PHjdvXq1TY1NdWuWLHCAnbFihX27Nmzdu3atfbYsWN2+/btdt++fTY+Pt5u2bLFnjhxwq5evdqmpaX5tVm/fr1NTEy0O3bssHv37rUHDx60mzdvtidPnrQrV6606enpPm3OnDljN2zYYI8ePWp37txp9+zZYw8dOmRjY2PtqVOn7IoVK2xGRoZPm+TkZLtx40Z75MgRu2vXLrt79257+PBhu2nTJnv69GlP3eXLl3vanD592m7atMkePnzY7t692+7atcseOXLEbty40SYnJ/v07z7eqVOnbGxsrD106JDds2eP3blzpz169KjdsGGDPXPmjE+b9PR0u3LlSnvy5Em7efNme/DgQbt37167Y8cOm5iYaNevX2/Pnj3r0yYtLc2uXr3anjhxwm7ZssXGx8fbffv22e3bt9tjx47ZtWvX+rVJTU21q1evtsePH7dbt261Bw4csPv377fbtm2zSUlJds2aNTY1NdXn3M+dO+eZy23bttn9+/fbAwcO2K1bt1rv+Xe3Wb58uT179qxdt26dZy4zz/+qVav85jIlJcWuX7/eZy4zz3/muXTPf0JCgmcuDx06ZDdt2uSZf2ttlvMfFxdn4+Li7JEjR/zm3/v11KlTfvOfkJBgN2zYYJOTk32uV0ZGhl25cmW285+SkuLzPUtPT7erVq2yJ06cCDj/69aty3L+jx8/brds2WIPHDjgN//nzp3zm/81a9bYpKQku3XrVrt//367f/9+u3XrVp/5927jPf/ue/nAgQN2y5YtnvkPdC97z7/3vew9/97XLSUlJdt7OfP8L1++3CYnJ/vNv/teds9/5rl038uZ5z/Qvez+7rjn/9ChQ37zn/leds+/973sPf/r16/3mX/3/e+ef+97OfP8e99j2d3Lgebf3cY9/9ndy/ot12+5fsv1W67fcv2W67dcv+X6LddvuX7L9Vuu33L9lv8Zf8vJJfn2gYiBGGMGAMNxHmLY1Fqb6LVvK3AdcJ21dnuAtr/hPBTxdmvtUlfZOaAgUNC1ujpzmwNABaCCK9d1QO5V008+mTmltYiIiIiIiIiIiMhVx+TGQa6aldPGmKdwAtObgDu9A9Mu2a5yBkpkqnchbU5cwFBFRERERERERERE5DyuiuC0MWYgMBLYgBOYDrSKeYvrtWaA9sHANTgPUNyZwzYVgKLAPmtt8sWPXkREREREREREREQyy/fBaWPMP4H3gDU4genDWVSd63q9O8C+vwChwG/W2rM5bNMuUx0RERERERERERERuUzydXDaGPMqzgMQVwKtrLUJ2VSfCCQADxpjbvXqozDwhuvj6ExtPgXOAk8bYyK92pQEXnZ9HHMJpyAiIiIiIiIiIiIiAQTn9QCyYozpDQwG0oGFwABj/PJwx1lrPwOw1p4wxjyOE6Seb4z5BkgEOgG1XOXfeje21u4yxgwCRgArjDHfAueAbkBlYJi1dsmVOUMRERERERERERGRP698G5zGyRENUAAYmEWdX4HP3B+stVOMMc2BfwH3AYWB7cCzwAhrrc3cgbX2A2NMHPA80AtnNfkm4BVr7fjLciYiIiIiIiIiIiIi4iPfBqettdFA9EW0Wwy0v8A2PwA/XOixREREREREREREROTi5Ouc0yIiIiIiIiIiIiLyx6TgtIiIiIiIiIiIiIjkOgWnRURERERERERERCTXKTgtIiIiIiIiIiIiIrlOwWkRERERERERERERyXUKTouIiIiIiIiIiIhIrlNwWkRERERERERERERynYLTIiIiIiIiIiIiIpLrFJwWERERERERERERkVyn4LRc9aKjozHGMH/+/LweioiIiIiIiIiIiOSQgtMiIiIiIiIiIiIikusUnJar3tNPP01sbCwNGzbM66GIiIiIiIiIiIhIDgXn9QBELlWZMmUoU6ZMXg9DRERERERERERELoBWTkueiIuLwxhDnz592LFjB926daN06dIUL16ctm3bsmHDBgCOHDnCE088QYUKFShcuDC33XYb8+bN8+krq5zTCxcupGPHjlSuXJmQkBDKly9P48aNiYmJ8al36NAhnn/+eWrVqkXRokUJDw+nVq1a9OnTh507d/qNfdasWbRv354yZcoQEhJCjRo1GDRoEElJSX51IyMjiYyMJDk5mUGDBlG1alVCQkK49tprefvtt7HW+rWZNm0arVq1okKFCoSEhFCxYkWaN2/OqFGj/OomJiby0ksvcf3111OkSBHCwsJo1aoVs2fPPu8ciIiIiIiIiIiI5CWtnJY8FRcXR6NGjbj++uvp06cPcXFxTJ48mRYtWrBkyRLuvvtuSpQoQffu3UlMTOSbb76hXbt2bN26lapVq2bZ78yZM+nQoQMlSpSgU6dOVKpUicTERGJjYxk1ahRRUVEAJCcn07RpU3bs2EGbNm3o2LEj1lp2797N1KlT6datG9WrV/f0O3jwYKKioihVqhT33HMPERERrFu3jnfeeYeffvqJJUuWUKJECZ+xpKam0rZtWw4cOEC7du0IDg5mypQpvPjii6SkpHjGAvB///d//O1vf6N8+fJ07NiRMmXKcPjwYdatW8enn35Kv379PHV3795NixYtiIuL44477uDuu+/m9OnTTJ8+nbvvvpuPPvqIxx9//HJNlYiIiIiIiIiIyGWl4LTkqV9//ZU33niDf/3rX56y119/nddee41GjRrxwAMPMGrUKIKCnEX+bdq0oVevXrz33nu89957WfY7duxYMjIymD9/PnXr1vXZl5CQ4Hn/yy+/sGPHDgYOHOjX37lz5zh79qzn87x584iKiuL222/np59+Ijw83LPvs88+45FHHiEqKsqvnwMHDlC3bl3mzJlDkSJFAIiKiqJmzZq89957vPzyyxQsWBCAjz76iEKFCrF27VoiIiKyHDdA79692b17N19//TUPPvigpzwpKYkWLVowYMAAOnXqRLly5bK8TiIiIiIiIiIiInlFwel8pMfS1LweQo591bjgZeknMjKSF1980aesd+/evPbaa5w9e5ahQ4d6AtMAPXr0oG/fvqxZsyZH/buDwd4C5acOVK9QoUIUKlTI83nEiBGAE/j2DkwD9OnTh+HDhzNhwoSAQfMRI0b4HCMiIoLOnTvz+eefs2XLFm688UbPvuDgYE+wOqtxr127ll9//ZVu3br5BKYBwsPDiYmJoUuXLnz//fc+q61FRERERERERETyCwWnJU/Vq1ePAgUK+JRVrFgRgJo1a1K8eHGffQUKFKBcuXLs27cv23579uzJpEmTaNSoEd27d+fOO++kadOmVK5c2ade8+bNqVSpEkOGDGHVqlW0b9+epk2bBhzXkiVLKFiwIN999x3fffed3zHPnTvHkSNHOHr0KKVLl/aUh4WFce211/rVr1KlCgDHjh3zGfdzzz1HnTp16N69O82bN6dp06aULVvWbywAx48fJzo62q/vI0eOABAbG5vdZRIREREREREREckzCk5LngoLC/MrCw4OznKfe39qavarzLt27cr06dMZNmwY48aN46OPPgKgQYMGvPXWW7Rp0waAEiVKsHTpUqKiopg2bRqzZs0CnFXK/fr145VXXvGsYj569ChpaWl+D1TM7NSpUz7B6cyrrDOfZ3p6uqfs2WefpUyZMowaNYoRI0bw/vvvY4yhefPmDB06lFtvvdUzFoA5c+YwZ86cbMciIiIiIiIiIiKSHyk4nY9crlQZ4ujQoQMdOnTg9OnTLFu2jOnTpzN69GjuueceVq9ezQ033ABA5cqV+eSTT7DWsmnTJubOncuHH37I4MGDycjI4PXXXwecYHlGRgaJiYlXdNy9evWiV69eJCUl8dtvvzF58mTGjRvHXXfdRWxsLBEREZ7A/fDhwxkwYMAVHY+IiIiIiIiIiMiVEHT+KiJXt6JFi9KyZUveffddXn75Zc6dO8eMGTP86hljqFOnDv379/esRp4yZYpnf+PGjTl27BgbN27MlXGHh4fTvn17xo4dS58+fUhMTGThwoWesQCezyIiIiIiIiIiIlcbBaflD+mXX37hzJkzfuWHDh0CIDQ0FIANGzYQFxd33noA//jHPwB4/PHHOXDggF+b06dPs3Tp0ksa98yZM0lLS/MrP3z4sM94br31Vu644w4mTZrEuHHjAva1fv16TzsREREREREREZH8Rmk95A/pueeeIy4ujhYtWhAZGUmhQoVYuXIlc+fOpVq1ajz44IMA/Pzzzzz77LM0adKE2rVrExERwb59+5g6dSpBQUEMGjTI02erVq0YMmQIL730Etdddx3t27fnmmuu4dSpU+zevZtff/2VZs2aMXPmzIse94MPPkjhwoVp1qwZkZGRWGtZuHAhy5cvp0GDBrRu3dpT96uvvqJly5Y8+uijjBgxgkaNGhEeHs6+fftYt24dGzZsYMmSJURERFz8hRQREREREREREblCFJyWP6SXX36ZyZMns2LFCn7++WeCgoKoWrUqL7/8MgMHDqRkyZIA3HXXXQwcOJAFCxYwdepUTpw4QYUKFWjTpo0naO3tn//8J02bNmXEiBEsWrSIqVOnEhYWRqVKlXjiiSfo0aPHJY17yJAhzJo1i1WrVvHTTz9RuHBhqlWrxttvv82TTz7peTgjOLmyV65cyQcffMD333/PhAkTSE9Pp3z58txwww3079+fm2666ZLGIyIiIiIiIiIicqUYa21ej+GqN3r0aAvw5JNP5vVQRERERERERERERC6VyY2DKOe0iIiIiIiIiIiIiOQ6BadFREREREREREREJNcpOC0iIiIiIiIiIiIiuU7BaRERERERERERERHJdQpOi4iIiIiIiIiIiEiuU3BaRERERERERERERHJdcF4PQHLH0M1p2e4fVFtfBREREREREREREck9ikj+CR07Zz3vSxYyeTgSERERERERERER+bNScPpPwntltPcqaq2YFhERERERERERkbygnNMiIiIiIiIiIiIikusUnBYRERERERERERGRXKfgtIiIiIiIiIiIiIjkOgWnRURERERERERERCTXKTgtIiIiIiIiIiIiIrlOwWkRERERERERERERyXUKTouIiIiIiIiIiIhIrlNwWkRERERERERERERynYLTkifi4uIwxtCnTx/i4uJ48MEHKVOmDIULF+bWW29l+vTpfm3Onj3LkCFDuPnmmwkNDaVEiRLccccd/Pe//70s/Z87d44RI0ZQv359SpYsSWhoKJGRkXTu3Jmff/7Zr/7mzZvp06cPVapUISQkhHLlytGjRw+2bNniV7dPnz4YY4iLi+Ojjz7ipptuonDhwpQrV44nnniC48eP+7VZt27jbMaDAAAgAElEQVQdDz30EJGRkYSEhFC2bFnq16/PwIEDSU1N9amblpbGqFGjaNy4MSVKlCA0NJRbbrmFkSNHkpGRke1ciIiIiIiIiIiI5IXgvB6A/Lnt3r2bhg0bUr16dR5++GESExP59ttvPQHhO++8E3ACx3fddRe//vortWvX5qmnniI5OZmJEyfSvXt31qxZw7///e+L7h+cAPLXX3/NjTfeSK9evShSpAgHDhxg0aJFzJw5k9atW3vqzpw5k65du5KamkrHjh259tpr2bdvH5MmTeLHH39k3rx51K9f3288L7zwArNmzaJjx460bduWefPmMXbsWLZv387cuXM99datW0ejRo0wxtCpUyeuueYaTpw4wfbt2xk1ahRvvPEGBQsWBPCMYdasWdSqVYsePXpQuHBh5s2bR//+/Vm2bBlffPHFZZszERERERERERGRy0HBaclT8+fPJzo6mqioKE9Zjx49uPvuuxk6dKgneDxs2DB+/fVX2rVrx7Rp0wgOdr66UVFRNGzYkLfeeot77rmHJk2aXFT/x48f55tvvqFBgwYsW7aMAgUK+PRz9OhRz/tjx47x0EMPERoayoIFC7jhhhs8+zZu3EijRo147LHHWLVqld/5Ll26lPXr11O1alXAWfHcsmVL5s2bx++//07Dhg0BGD9+PCkpKUyZMoXOnTv79HHs2DFCQ0M9n998801mzZrF008/zfvvv+8Ze3p6Ok888QTjxo2jW7dufv2IiIiIiIiIiIjkJQWn8xHTZUSuH/OFi2xnpwy4LMevVq0ar7zyik/ZXXfdRdWqVfn99989ZePGjcMYw7vvvusJTANERETw6quv8thjj/Hxxx/7Badz2r8xBmstISEhBAX5Z7spXbq05/3nn39OUlISI0eO9AlMA9SpU4fHH3+c999/n02bNvntf+211zyBaYDg4GAeeeQRFi5c6BOcditSpIjfWEqWLOl5n5GRwciRIylfvjzvvfeeT1C9QIECDBs2jE8//ZQJEyYoOC0iIiIiIiIiIvmKgtOSp+rVq+e3ShmgSpUqLFmyBICTJ0+yfft2KlWqRO3atf3qtmzZEoDVq1dfVP8AJUqUoGPHjvzwww/Uq1eP++67jzvuuINGjRr5rFIGPO3Wrl1LdHS0X99bt24FIDY21i84feuttwYcCzgrot26d+/O8OHD6dKlC926daN169Y0bdqUGjVq+B3r6NGjXHfddbzxxht+fYMT4I6NjQ24T0REREREREREJK8oOC15Kjw8PGB5cHCw50F+7ocFVqhQIWBdd3lSUtJF9e/27bff8vbbb/PVV1950oAULlyYbt268c4771CuXDngfyk+xo4dm+25nTp1Kkfjca8ET09P95Q1bNiQhQsX8uabbzJx4kRPzuhatWoRFRXFQw895DOWbdu2ERMTc0FjERERERERERERyUv5OjhtjOkGNAfqAXWB4sAEa+1fs6hfDPgn0A24BkgBVgLDrLU/BajfB/g0myE8aa0dcynncCEuV6qM8xm6Oc3zflDtfP0VACAsLAyAgwcPBtwfHx/vU+9iFSlShOjoaKKjo9m7dy8LFizgs88+48svvyQuLo6FCxf6HGft2rXcfPPNl3TM7Nx+++1Mnz6ds2fPsnLlSmbOnMkHH3xAjx49KFu2LK1bt/aM5d5772XSpElXbCwiIiIiIiIiIiKXm39y3fzlFeBpnOD0/uwqGmPCgSWuNunAR8BE4CbgR2NMdpHfqUBMgG3FJY5fLoPixYtTo0YN9u/fz7Zt2/z2z5s3D4D69etftmNWqVKFnj17MmvWLK677joWLVrkWaXcuHFjAE+w+koLCQmhSZMmDB48mBEjnLzkU6dOBaB27dqEh4ezdOlSUlNTc2U8IiIiIiIiIiIil0N+D07/A6gJlACePE/daOBGYBJQz1r7jLX2MaAOEAe8Y4y5Lou2U6y10QE2Bafzib59+2KtZdCgQT7pLxISEnj99dc9dS7WkSNHWLZsmV/56dOnOXnyJMHBwRQqVAiARx55hPDwcGJiYnwequiWkZHB/PnzL3os4AS+3elMvB06dAjAkwc7ODiY/v37Ex8fz4ABAzhz5oxfm/j4eDZt2nRJ4xEREREREREREbnc8nVOB2vtPPd7Y8z5qnd1vb5mrfXkrbDWHjHGDAM+AP4OPHe5xylX3vPPP8+MGTOYOnUqdevWpX379iQnJ/Pdd99x+PBhXnjhBZo1a3bR/e/fv5/GjRtz/fXXU79+fapUqcKJEyeYPn06Bw8eZMCAARQvXhyA0qVLM3HiRO69914aN25Mq1atqFOnDkFBQezZs4clS5Zw9OhRUlJSLno8w4YNY/bs2bRo0YLq1atTrFgxNm7cyIwZMyhZsiRPPPGEp+6rr77K2rVrGTNmDD/88AMtW7akUqVKHD58mG3btrF48WLefPNNv4czioiIiIiIiIiI5KV8HZy+QOVdrzsD7HOXtcqibT1jzECgME76kHnW2n2XeXxyCQoVKsScOXN49913+eqrr/jggw8IDg6mbt26vP/++54HBF6syMhIYmJimD9/PvPmzSMhIYFSpUpRq1YthgwZwoMPPuhTv1WrVqxbt4533nmHWbNmsXDhQgoVKkTFihVp2bIl99133yWNp1+/fpQsWZJly5axePFi0tLSqFy5Mv369eO5556jWrVqnroFCxZkypQpfPnll3z22WdMnz6dU6dOUbZsWa655hpef/11evbseUnjERERERERERERudyMtTavx5AjxpgWwDyyeCCiMeYAUAGoY63dlGnf0zgrp89aawt7lfch8AMR04GPgYHW2vMufx09erQFePLJ82UeyR+utgciioiIiIiIiIiISK46bxqLyyG/55y+ENNdr9HGmALuQmNMaeBZ18cQY0wRrza7gP5ALaAoUBF4ACdH9d+AcVd4zCIiIiIiIiIiIiJ/SgWio6Pzegw5EhMTEwn0AdZHR0dPCrB/OXA/0AToGhMTUysmJqYrrhXTOCk7CgKDo6Oj0wGio6N3R0dH/x4dHX00Ojo6NTo6+lR0dPSmmJiYya5j3RYTEzMlOjr6UKAxGWOeiImJ+ahDhw4VAWrWrElsbCxlypRh9erVVKxYkRUrVlC6dGk2btxISEgI+/fvJzk5mdOnT3PgwAEKFSrEpk2bKFu2LKtWrfJpExsbS8GCBTlw4ACnT5/mzJkz7Nu3j5CQEDZs2EC5cuV82pQsWZItW7ZQoEABDh48yMmTJ0lJSWHv3r0UKVKEdevWUb58eVasWElGyQoU3LWKmyuVZOvWrRhjOHz4MCdOnODcuXPs3r2b0NBQ1q1bR8WKFVm+fDmVKlVixYoVhIWFsX37dqy1JCQkcOzYMdLS0oiLi6NYsWKsXbvWM6by5cuzcuVKSpQowY4dO8jIyCAxMZGjR4+SkZHBzp07KV68uE+bcuXKsXr1aooXL86uXbtIS0vj2LFjJCQkALB9+3bCw8NZs2aNp01ERARr166laNGixMXFkZqayvHjxzl8+DDGGLZu3UrJkiV92pQtW5Z169YRGhrKnj17OHv2LCdOnODQoUMUKFCAzZs3U7p0aVatWuU59zJlyrBhwwZCQkLYt28fKSkpnDp1ivj4eIKDgz3z726zfPlyypQpw6ZNmzxzmXn+N27cSEREhM9clipVitjYWIKDgz1zmXn+3dc28/wHBQVx6NAhTpw4wdmzZ9mzZ49n/itUqODTJjw83DP/R44cISkpidTUVL/5X7Fihee1RIkSfvOfnp7Ozp07KVasGGvWrPFcr/Lly7Nq1aps579EiRKeeVm+fDnly5dnzZo1FC1aNOD8b9u2Lcv5Dw0NZffu3Zw7d85v/kuVKuVzX5YtW5b169dTuHBh9uzZQ0pKCidPnuTgwYM+8+/dxnv+3ffyqVOnOHDgAAULFiQ2Njbgvew9/973svf8r1y50nPdSpUqxebNm7O8l9evX+8z/8uXLw84/+572T3/mefSfS8DPvMf6F52f3fc85+RkcHRo0d95j/zveyef+972Xv+d+zYQVhYmM9clitXzjP/3vfykSNHfOZ/9erVnnssIiIiy3s5KCjIb/7dbdzzv3fv3izv5fz8W57Vvazfcv2W67dcv+X6LddvuX7L9Vuu33L9luu3XL/l+i3Xb/ml/pZXqFAhJicx20v1h0nr4apTFngF6AhUBo7hrKh+HSfv9AlrbXgOj/cl0BMYYK39ILu6SushIiIiIiIiIiIifyC5ktbjDxWZtNYeAZ5xbR7GmDtxLujyC+juiOu16OUZnYiIiIiIiIiIiIi4/ZFyTmfncdfrhAto08j1uvMyj0VERERERERERETkT+8PE5w2xgQZY4oFKH8MeAhYQ6bgtDHmjgD1jTHmJeB2IAGYeWVGLCIiIiIiIiIiIvLnla/TehhjugBdXB/Lu15vN8Z85nqfYK193vU+FDhkjJkDbHeV3QE0BHYA91prUzMdYoExZitOuo/9QBjQFLgRSAZ6WmtPXN6zEhEREREREREREZF8HZwG6gG9M5VVd20AuwF3cPos8A3QDGjjKtsBRAHvWmtPBej/HZzgdUugFJAB7AE+dLVRSg8RERERERERERGRKyBfB6ettdFAdA7rpgKPXmD/gy58VCIiIiIiIiIiIiJyqf4wOadFRERERERERERE5Oqh4LSIiIiIiIiIiIiI5DoFp0VEREREREREREQk1yk4LSIiIiIiIiIiIiK5TsFpkatcnz59MMYQFxeX10PJ9+bPn48xhujo6LweSo61aNECY0xeD0NERERERERE5LJTcFpyXVxcHMYY+vTpk9dD8ZOamsrkyZN59NFHufHGGylRogShoaHcdNNNvPbaa5w8eTLLtvv27aNv375UrFiRkJAQIiMjGThwIMeOHcvFMxC5PPLzfSoiIiIiIiIifwzBeT0Akfxkx44ddO3alaJFi3LnnXfSoUMHTp06xaxZs3j99df59ttvWbx4MWXKlPFr16RJEw4fPkznzp2pXbs2v//+O8OHD2fmzJksXryY0qVL59FZydXs888/Jzk5Oa+HISIiIiIiIiJy2Sk4LeKlePHifPjhh/Tu3ZuiRYt6ys+dO0fXrl358ccfiYmJ4YMPPvBp169fPw4fPsyIESPo37+/p/zZZ5/lvffe41//+hdjxozJtfOQP46qVavm9RBERERERERERK4IpfWQXBUdHc0111wDwPjx4zHGeLbPPvvMUy8jI4MxY8Zw2223UaxYMYoWLcptt93G6NGjycjI8OvXGEOLFi04cOAADz/8MBERERQpUoQGDRrw1Vdf5Xh8lSpVol+/fj6BaYBChQrx8ssvA07eYm87d+5k9uzZREZG8tRTT/nsi4mJoWjRonzxxRecPn06x+NYuXIl9913HxEREYSEhFCtWjX69etHfHx8lm0yMjJ49913qV27NoULF6Zy5cr84x//4MSJE351161bx0MPPURkZCQhISGULVuW+vXrM3DgQFJTU33qpqWlMWrUKBo3buxJc3LLLbcwcuRIv7nwTgWxdetWunfvTkREBEFBQcyfP5/atWtTqFAhEhISAp7DkCFDMMbw4Ycf+pTv27ePp59+murVqxMSEkLp0qXp1KkTy5cvD9jPoUOHePTRRylXrhxFihShXr16jB8/PstrF8hbb72FMYYRI0YE3H/gwAEKFCjAbbfd5lM2ePBgmjZtSvny5SlUqBAVK1akR48exMbG+vVxvusFgXNOnzt3jpEjR9K+fXuqVatGSEgIpUqVonXr1syYMSPgeCMjI4mMjCQ5OZlBgwZRtWpVQkJCuPbaa3n77bex1nrq5vQ+FRERERERERG5FFo5LbmqRYsWJCUlMXz4cOrWrUuXLl08++rVq+d5//DDD/PVV19RpUoVHnvsMYwxTJ48mX79+rFo0SImTJjg1/exY8do0qQJ4eHhPPLIIyQlJfHf//6Xnj17sn//fgYNGnRJYy9YsCAAwcG+t83cuXMBaNu2LUFBvn/vKV68OE2bNmX27NksXbqUVq1anfc406dP57777sNaS7du3ahWrRorV65k9OjRTJ06lcWLFxMZGenX7h//+AcLFizggQceoHPnzsyaNYv333+fhQsXsmjRIgoXLgw4gelGjRphjKFTp05cc801nDhxgu3btzNq1CjeeOMNz7mmpqbSsWNHZs2aRa1atejRoweFCxdm3rx59O/fn2XLlvHFF1/4jWXHjh00atSImjVr0rNnT86cOUOJEiXo3bs3L7/8Ml9//bXPCnO3zz//nEKFCvHggw96ylatWkXbtm1JTEzkrrvuomvXriQkJDBlyhSaNWvG5MmTad++vaf+0aNHadKkCTt37qRZs2Y0a9aM+Ph4/v73v9O2bdvzXn+3Xr168corrzB+/HgGDBjgt//LL78kIyOD3r17e8oWLFjAkCFDuPPOO7nvvvsoVqwY27ZtY+LEiUybNo3FixdTt27dHF+vrCQmJvLMM8/QpEkT2rRpQ9myZYmPj+eHH36gffv2jB07lscee8yvXWpqKm3btuXAgQO0a9eO4OBgpkyZwosvvkhKSgpRUVFAzu9TEREREREREZFLYq3VdonbqFGj7KhRo+zV4j+xqZ4tL+zatcsCtnfv3gH3f/XVVxawt9xyiz158qSn/NSpU7ZBgwYWsBMmTPBpA1jA3n///TY9Pd1TvnPnTluyZElbsGBBu2PHjksa99///ncL2BdffNGn/Pnnn7eAfeeddwK2e+qppyyQo+/IyZMnbenSpW1QUJBdsGCBz74hQ4ZYwLZp08anvHfv3hawpUuXtnFxcZ7y9PR027VrVwvYwYMHe8qfffZZC9gpU6b4HT8xMdHn+kVFRVnAPv300zYtLc1TnpaWZvv27evXj3tuAfvSSy/59b9v3z4bFBRkGzRo4Lfv999/t4Dt2rWrpyw1NdXWqFHDhoSE2Pnz5/vU379/v61YsaItX768TUlJ8ZQ//vjjFrADBw70qb98+XIbHBxsARsVFeV3/EDatm1rAbt+/Xq/fTfccIMtWLCgTUhI8JQdOnTInjhxwq/umjVrbNGiRe3dd9/tU36+62Wttc2bN7fOT/X/pKSk2L179/rVTUpKsnXq1LElS5a0ycnJPvuqVatmAduuXTuffYcOHbJhYWE2LCzMnjt3zm9sWd2nIiIiIiIiIvKHlitxVa2czk/q3p8rh7m09cMua7+7HL0ENG7cOMBJ8VCsWDFPedGiRXn77bdp3bo1H3/8MT169PBpV6BAAd5++22f1cvXXHMNAwYMICYmhi+++MKzMvRCTZs2jY8++ojKlSvzwgsv+Ow7fvw4AGFhYQHbusuTkpLOe5ypU6dy9OhRHnroIe644w6ffc899xxjxoxhzpw57Nmzxy8X8TPPPEO1atU8n4OCghg6dChTpkxh3LhxvPrqqz71ixQp4nf8kiVLet5nZGQwcuRIypcvz3vvvUeBAgU8+woUKMCwYcP49NNPmTBhAp07d/bpp1y5cgGvdaVKlWjVqhVz5sxh48aN1KlTx7PPnXbDeyXyjz/+yI4dO3j++edp3ry5T18VK1bkhRdeYODAgfzyyy+0b9+e1NRUJkyYQPHixYmOjvapf+utt9KzZ88LSu/Ru3dvZs+ezfjx4xk6dKinfMWKFWzatIl7773X50GXERERAfupW7cuLVu2ZPbs2aSmpnpWprtldb2yEhISQuXKlf3Kw8LC6Nu3L8899xzLly/nL3/5i1+dESNG+Mx9REQEnTt35vPPP2fLli3ceOONOR6HiIiIiIiIiMilUHBa8p1Vq1YRFBREixYt/PY1b96cAgUKsHr1ar99VatW9eTJ9daiRQtiYmICtsmJ3377jR49elC0aFG+//57nwBuTlhXLt/MeYMDWbVqFQAtW7b02xccHMxf/vIX4uLiWL16tV9wOnPwFqB69epUqVKFuLg4kpKSCA8Pp3v37gwfPpwuXbrQrVs3WrduTdOmTalRo4ZP261bt3L06FGuu+463njjjYDjLVKkSMBcynXr1iUkJCRgmz59+jBnzhzGjx/Pf/7zH8DJofzNN99QtmxZnxQdS5YsAWD37t1+wWaAbdu2ARAbG0v79u3ZvHkzycnJ3HHHHQH/WNCiRYsLCk7fe++9hIWF8eWXXzJkyBBPgN7dR58+ffza/Pjjj4wZM4YVK1aQkJBAWlqaz/6EhAQqVKjgU5bd9crKxo0bGTp0KAsWLCA+Pp6UlBSf/fv37/drExYWxrXXXutXXqVKFcBJjSMiIiIiIiIiklsUnJZ85/jx45QqVYpChQr57QsODqZMmTIcPnzYb1+5cuUC9le+fHlPvxdqyZIltGvXjqCgIGbMmEHDhg396riDoFn1734gYVYrq725+8gcvHRzlwdahZ3d+e/evZvjx48THh5Ow4YNWbhwIW+++SYTJ0705IyuVasWUVFRPPTQQ4CTuxmcAHBMTEyWYz516lTAY2bl3nvvpUSJEnz55Ze89dZbFChQgOnTp3P06FEGDhzok9PbPYbvvst+pb57DO7rd77vQk4VKVKEBx54gLFjxzJ79mzatWtHamoqX3/9NWXLlqVdu3Y+9UeMGMEzzzxDyZIladOmDVWrViU0NBRjDFOmTGHt2rWcPXv2kse1dOlSWrZsSVpaGq1ataJTp06UKFGCoKAg1qxZw9SpUwMeJzw8PGB/7muenp5+QeMQEREREREREbkUCk7nJ1cwVYa3oZv/t5JzUO389xUICwsjMTExYPqDtLQ0EhISAj4s7tChQwH7O3jwoKffC7Fw4UI6dOhAUFAQs2bNonHjxgHr1apVC3BWGgfiXt1bs2bN8x7TPUb3mDOLj4/3qeft0KFDnrF4C3T+t99+O9OnT+fs2bOsXLmSmTNn8sEHH9CjRw/Kli1L69atPfXvvfdeJk2adN6xe8tulbg74Pvxxx8zZ84c7r777oApPbzHPHXqVDp16nTe47rrn++7cCF69+7N2LFjGT9+PO3atfME0p955hmf72daWhpRUVGUL1+eVatW+f2Bwb0KPJCcrKr39sYbb3DmzBnmzZvn9y8M3nrrLaZOnXpB/YmIiIiIiIiI5IWg81cRubzcqRGyWqV5yy23kJGRwYIFC/z2LViwgPT0dOrXr++3b8+ePcTFxfmVz58/39NvTs2dO5d27doRHBzMnDlzsgxMA9x5550AzJ49m4yMDJ99J0+eZPHixRQpUiTbPtzcY3SP2VtaWhqLFi0CCHj+v/76q1/Zzp072bt3L5GRkQFXzYaEhNCkSRMGDx7MiBEjADyBzdq1axMeHs7SpUtJTU0979gvhDsdxvjx40lISGDGjBncfPPN1KtXz6ee+5otXLgwR/3Wrl2b0NBQ1qxZE3Ale6Drej5NmzbluuuuY+rUqRw/fjzLQHpCQgJJSUk0adLELzB96tQpT8qWy2H79u2UKlUqYOqbQN+Di3G++1RERERERERE5FIpOC25rmTJkhhj2LNnT8D9ffv2BeCll14iOTnZU56cnMyLL74IwKOPPurXLj09nX/+858+AeJdu3YxYsQIgoOD+etf/5qj8c2ePZt77rmHwoUL88svv3DbbbdlW79GjRq0bduWuLg4PvzwQ599UVFRnD59ml69elG0aNHzHrtLly6UKlWKr7/+mqVLl/rse//999m5cyetW7f2yzcNMHz4cHbv3u35nJGRwaBBg8jIyOCRRx7xlC9cuDBg4Na92jg0NBRwUj3079+f+Ph4BgwYwJkzZ/zaxMfHs2nTpvOeV2beAd/Ro0eTmpoaMH9z586dqVGjBh9++CE//fRTwL6WLFni+Z4ULFiQnj17cvLkSb8c1StWrGDChAkXPFZwAtEpKSmMGjWKn376iZtvvtnvjx0RERGEhoaycuVKn1QnqampPPPMMyQkJFzUsQOJjIwkMTGRdevW+ZR/8sknzJo167Ic43z3qYiIiIiIiIjIpcp/OR3kD69YsWI0atSIhQsX0rNnT2rWrEmBAgXo1KkTN998Mz169GDq1Kn897//pU6dOnTp0sWTs3fXrl088MAD9OzZ06/fm2++mWXLltGgQQPatm3L8ePH+fbbb0lKSuI///mP3wP/AtmyZQudO3cmJSWF9u3bM3Xq1IApEjIHPkeNGkWTJk0YMGAAv/zyC9dffz3Lli1j3rx51KxZkzfffDPH12bcuHHcf//9NG/enPvvv5+qVauycuVKZs+eTfny5fnoo48Ctm3atCn16tWje/fuhIWFMWvWLNauXUuDBg144YUXPPWGDRvG7NmzadGiBdWrV6dYsWJs3LiRGTNmULJkSZ544glP3VdffZW1a9cyZswYfvjhB1q2bEmlSpU4fPgw27ZtY/Hixbz55pvccMMNOTo/b7169eLVV1/l9ddfJzg4mB49evjVKViwIJMmTeKuu+6iQ4cONGnShHr16hEaGsrevXtZvnw5O3fuJD4+3hNU//e//80vv/zC+++/z4oVK2jWrBnx8fF8++23tG/fnmnTpl3UWF977TWioqJITU31WzUNEBQUxIABAxgyZAg33XQTnTt35ty5c8ybN4/ExETuvPNO5s2bd8HHDmTgwIHMmjWLZs2a8cADDxAWFsaKFStYtGgR3bp1Y+LEiZd8jPPdpyIiIiIiIiIil8xaq+0St1GjRtlRo0bZq8V/YlM9W17Ztm2bveeee2ypUqWsMcYC9tNPP/XsT09Ptx9++KFt0KCBLVKkiC1SpIitX7++HTlypE1PT/frD7DNmze3+/+fvXuP8yuv68P/eueyySbZ3C+TbOWmLlAtLIoioK6CoHTxByr8bAu6i1oebBXwgtXWCxOU9teCtED5bcFawELlZ5VCRVErLsqvapncs7mQzf22uW52s8lubptP/5jvjDOzk0wmmXxZJ8/n4/F9nDnnfM6Z93feOeePV87jcw4caG94wxvakiVL2kIn0N4AACAASURBVIwZM9oLXvCC9slPfvKK67rvvvtakjE/o9m7d2+7++67W09PT5s+fXp72tOe1t72tre148ePj/vv8+Uvf7m99rWvbYsXL27Tp09vX/M1X9Pe8pa3tAMHDjxp7F133dWStB07drT3vve97dnPfnabMWNGW7FiRXv729/eHnnkkWHj/+RP/qTdfffd7bnPfW6bO3dumzVrVrvtttvaW9/61rZ79+4nnf/ixYvtt3/7t9vLXvaytmDBgjZ9+vS2YsWK9tKXvrS9+93vbnv37h0cu2vXrpak3XXXXWN+xz179rQpU6a0JO3Vr371ZccePny4/cIv/EL7hm/4hnbzzTe32bNnt6/7uq9rP/RDP9T+y3/5L+38+eH/lh988MH2pje9qS1evLjNnDmzPf/5z28f/ehHB/v7zne+c8z6Rnr5y1/ekrRp06a1Q4cOjTrm/Pnz7Td+4zfac5/73DZz5sy2bNmy9sY3vrHt3r17sE+7du0aHH8lf6877rhj1H9zf/AHf9Be9KIXtTlz5rR58+a1V7ziFe0v/uIv2kc/+tEnXU+ttfb0pz+9Pf3pTx/1d7zzne9sSdp99903bPtY1ykAAAAwaXUlV63WWjez8Enp3nvvbUlyzz33fLVLuSJP9RciXo2qyh133HFVcwoDAAAAAMNUN36JOacBAAAAAOg64TQAAAAAAF0nnAYAAAAAoOsmx4TD3PDMnQ4AAAAAf7d4choAAAAAgK4TTgMAAAAA0HXCaQAAAAAAuk44DQAAAABA1wmnAQAAAADoOuE0AAAAAABdJ5wGAAAAAKDrhNMAAAAAAHSdcBoAAAAAgK4TTgMAAAAA0HXCaQAAAAAAuk44DQAAAABA1wmnAQAAAADoOuE0AAAAAABdJ5wGAAAAAKDrhNMAAAAAAHTdtKs5qKpmJ7k1yeIkjyc50lo7MJGFAQAAAAAweV1xOF1VX5vkx5K8IskLMuKp66o6nuS+JL+f5NOttQsTWCcAAAAAAJPImOF0Vb0wya8n+Z70B9Lnk2xMcijJQ0luTrIoybOTvD7J65Icrar3J3lfa+3s9SkdAAAAAIC/qy4bTlfVx5O8IckjST6S5FNJvtxaO3OJ8c9I8r1J7kry7iRvqaofaa395QTWDAAAAADA33FjvRDxlUl+NsmK1to/a6395aWC6SRpre1urX24tfaSJLcnWZPkjokrFwAAAACAyWCsaT2e1Vp7/GpO3FrbkOQHqurmqzkeAAAAAIDJ67JPTl9tMD3R5wAAAAAAYHIZa1qPK1JV06vqBVX17Ik4HwAAAAAAk9u4wumq+r+r6nerauGQbV+bZFOSVUk2V9Wnq2qs6UIAAAAAALiBjffJ6R9L8pzW2kNDtv1Gkq9Lcl+SDUlek+RNE1FcVb2uqj5YVV+qqpNV1arqE5cZP6eqfq2qtlTVmap6uKq+UFX/cIzfc1dVfbmqTlXVI1X1xap69UR8BwAAAAAAnmy84fTfT9I3sFJVc5P8wyS/21r7niTfmmRrJiicTvLLSX4qye1JDlxuYFXNT/LXnWOeSPLhJL+X5B8k+cOqetsljntvko8lWZ7kN5N8onPMH1TVT03ItwAAAAAAYJjxTr+xJMmDQ9Zf3DnHp5KktXa+qv5nkn88MeXlZ5LsT7I9yR3pfzr7UnqTfGOSTyf54dbahSSpqiVJvpzkvVX1+dbaAwMHVNVLkvxckh1JvqW1dqKz/T1JVneO+VxrbfcEfR8AAAAAADL+J6cfTTJvyPodSVqS/3/ItjNJbrnGupIkrbX7WmsPtNbaFQz/wc7yVweC6c45jqZ/6pHpSd4y4piB9XcPBNOdY3Yn+VCSGZm4p8ABAAAAAOgYbzj9QJJXVdWMqropyeuTbGitHRsy5ulJjkxUgePQ01nuHGXfwLaXj9j+ss7yj0c55vMjxgAAAAAAMEHGG05/JMmz0h9Sb+n8/J9HjHlRkk3XXtq4DQTkzxxl37M6y+cMbKiq2UluTXKqtfbgKMcMTP9x24RVCAAAAABAkmRqb2/vFQ/u7e1dv3LlyhlJvi3JzCS/leRdA+eoqpcleVuS3+zt7f2riSx05cqVz0hyd5KNvb29nx5l/7OTfHOSpStXrvx0b29v69S0KMlHkyxIMm3lypX/tre398LKlSuXpH++6aO9vb3vG+V855P8SpLpvb29vz5aTVX15pUrV374zjvvXJEkt912W7Zs2ZLFixdn7dq1WbFiRVatWpVFixZl06ZNmTFjRg4cOJDHHnssp0+fzsGDB3PTTTdl8+bNWbJkSdasWTPsmC1btmT69Ok5ePBgTp8+nccffzz79+/PjBkzcv/992fZsmXDjlmwYEG+8pWvZOrUqTl06FAeffTRnDlzJvv27cvNN9+cDRs2pKenJ6tWrc7FBcszfdeaPO/WBdm2bVuqKkeOHMnJkydz7ty57NmzJ7NmzcqGDRuyYsWK9PX15dZbb82qVasyb968bN++Pa21HDt2LCdOnMiFCxeye/fuzJkzJ+vXrx+sqaenJ6tXr87cuXOzY8eOXLx4MQ899FCOHz+eixcvZufOnbnllluGHbNs2bKsXbs2t9xyS3bt2pULFy7kxIkTOXas//8ftm/fnvnz52fdunWDxyxdujTr16/P7Nmzs3v37pw/fz6PPPJIjhw5kqrKtm3bsmDBgmHHLFmyJBs2bMisWbOyd+/enD17NidPnszhw4czderUbN26NYsWLcqaNWsGv/vixYtz//33Z8aMGdm/f3/OnDmTU6dO5cEHH8y0adMG+z9wTF9fXxYvXpzNmzcP9nJk/zdt2pSlS5cO6+XChQuzZcuWTJs2bbCXI/s/8Lcd2f8pU6bk8OHDOXnyZM6ePZu9e/cO9n/58uXDjpk/f/5g/48ePZqHH34458+ff1L/V61aNbicO3fuk/r/xBNPZOfOnZkzZ07WrVs3+Pfq6enJmjVrLtv/uXPnDvalr68vPT09WbduXWbPnj1q/x944IFL9n/WrFnZs2dPzp0796T+L1y4cNh1uWTJkmzcuDEzZ87M3r17c+bMmTz66KM5dOjQsP4PPWZo/weu5VOnTuXgwYOZPn16tmzZMuq1PLT/Q6/lof1fvXr14N9t4cKF2bp16yWv5Y0bNw7rf19f36j9H7iWB/o/spcD13KSYf0f7Voe+Lcz0P+LFy/m+PHjw/o/8loe6P/Qa3lo/3fs2JF58+YN6+WyZcsG+z/0Wj569Oiw/q9du3bwGlu6dOklr+UpU6Y8qf8Dxwz0f9++fZe8lp/K9/JLXcvu5e7l7uXu5e7l7uXu5e7l7uXu5e7l7uXu5e7l13ovX758+cpL5bQTqa5sOucrPFn/VB83Jzk9dN7nCTr3d6X/hYifbK29cZT9PUn+Jv3Tityf5AtJZiV5Tfrnyl7eWZ/RWjtXVSuSHEhyoLX290Y53/Qk55Kcba3NvFxt9957b0uSe+6556q/Xze9Z+vftubnnzPed2ICAAAAAJNcdeOXjHdaj8tqrZ1rrT0y0cH0Ff7uQ0m+JckHksxO8s/SH0x/Lsn3pD80f6S1dq5zyCOd5byMbt6IcQAAAAAATJDLhtNVdfO1/oKJOMeVaq0dba29vbX2rNbaTa21Za21H0//PNSVpG/I2NPpf3J6TlUtH+V0X99ZbrvuhQMAAAAA3GDGenJ6V1W9vapmjPfEVfX8qvpskndcXWkT6p92lp8csf3PO8vvG+WYV40YAwAAAADABBkrnP7TJO9L8mBV3VtV3325J6Gr6llVdU9V/XWSNUmen/55oq+7qppSVXNG2f4TSf5xknV5cjj9HzvLX6qqBUOOeUaSn0xyNv0vUwQAAAAAYAJd9m14rbUfraoPJPlXSd7c+TxRVVuSPJjkRJKZSRYleXaSxemfPuNwkl9K8u9aa2evtriqem2S13ZWezrLF1fVxzo/H2utDTyZPSvJ4ar6n0m2d7Z9R5JvTbIjyQ+01s6P+H5/VVXvS/KzSTZU1e8luSnJDydZmOStrbXdV1s/AAAAAACju2w4nSSttVVJXllVX5/kx5O8PMntSf7BiKFHk3w6ye8n+f2RQfBVuj3JXSO2PavzSZI9+dtpQ84m+VSSb0/yis62HUnemeR9rbVTo/2C1trPVdWGJD+V/vD9Yvqf+n5Pa+1zE/AdAAAAAAAYYcxwekBr7YEkv5gkVTUrya3pf2L68SRHWmsPTnRxrbXeJL1XOPZ8+sPzq/k9H0/y8as5FgAAAACA8bvicHqo1tpjSR7ofAAAAAAAYFzGeiEiAAAAAABMOOE0AAAAAABdJ5wGAAAAAKDrhNMAAAAAAHSdcBoAAAAAgK4TTgMAAAAA0HXjCqer6mlVNXeMMbdU1dOurSwAAAAAACaz8T45vSvJ28cY87bOOAAAAAAAGNV4w+nqfAAAAAAA4KpdjzmnlyU5fR3OCwAAAADAJDFtrAFV9aMjNt0+yrYkmZrkaUl+JMnGCagNAAAAAIBJasxwOsnHkrTOzy3JazqfkQam+3gsycprrgwAAAAAgEnrSsLpN3WWleQ/J/lMks+OMu6JJMeT/HVr7eGJKQ8AAAAAgMlozHC6tfbxgZ+r6q4kn2mt/fZ1rQoAAAAAgEntSp6cHtRa++7rVQgAAAAAADeOKeMZXFXfVFW/WlXLLrG/p7P/9okpDwAAAACAyWhc4XSSdyT5iSRHLrH/cJIfT/Kz11IUAAAAAACT23jD6Rcnua+11kbb2dn+50leeq2FAQAAAAAweY03nO5Jsn+MMQeTLL+6cgAAAAAAuBGMN5x+LMmSMcYsSXL26soBAAAAAOBGMN5wel2S11TVnNF2VtXcJK/pjAMAAAAAgFGNN5z+SPqfjP6fVfW8oTuq6vlJ/jTJ4s44AAAAAAAY1bTxDG6t/X9V9aokP5pkbVUdTnIgya1JliWpJB9vrf3OhFcKAAAAAMCkMd4np9NauzvJW5JsTv8LEr+5s9yU5M2ttTdNZIEAAAAAAEw+43pyekBr7SNJPlJVs5LMT/Jwa+2xCa0MAAAAAIBJ66rC6QGdQFooDQAAAADAuFxVOF1VS5L8UJLnJpndWvuJIdufmWRja+3xCasSAAAAAIBJZdzhdFX9eJIPJJmZ/hcgtiQ/0dm9LMlfJ3lzkt+aoBoBAAAAAJhkxvVCxKp6RZKPJNmW5AeS3Dt0f2vt/vS/GPG1E1UgAAAAAACTz3ifnP6FJA8muaO1drKqXjDKmA1JXnzNlQEAAAAAMGmN68npJC9M8rnW2snLjNmfpOfqSwIAAAAAYLIbbzh9U5LTY4yZn+SJqysHAAAAAIAbwXjD6d1JvnmMMS9K8pWrqgYAAAAAgBvCeMPpzyb5jqp6/Wg7q+pNSZ6X5PevtTAAAAAAACav8b4Q8d8m+UdJfqeqXpdkXpJU1U8l+Y4kP5jkgSQfnMgiAQAAAACYXMYVTrfWTlTVHUl+O8nQp6c/0Fl+Kck/aa2NNS81AAAAAAA3sPE+OZ3W2t4k31VVz0vy4iSLkjyS5G9aa6snuD4AAAAAACahy4bTVfXpJJ9qrf1uZ/07k+xure1trW1IsqELNQIAAAAAMMmM9ULE1yZ5zpD1+5Lcfd2qAQAAAADghjBWOP1wkrlD1us61gIAAAAAwA1irDmntyT5x1XVl+TBzrZndKb3uKzW2l9ea3EAAAAAAExOY4XTK5N8Jsl/HbLtrs5nLFOvtigAAAAAACa3y4bTrbU/rarnJvmeJLcm6U3yF50PAAAAAABclbGenE5rbU+S30qSqupN8sXW2ruuc10AAAAAAExil30hYlW9r6peOWTTm9I/zQcAAAAAAFy1y4bTSX46ybcNWf/PSV57/coBAAAAAOBGMFY4fSrJrCHrdR1rAQAAAADgBjHWnNPbk/xgVf33JA92ts2vqqeNdeLW2t5rLQ4AAAAAgMlprCen35Pka5P8VZJdnW1v7/x8uc/OiSiuql5XVR+sqi9V1cmqalX1icuMn1FVP1lVX66qY1V1qqq2VNUHqurpo4y/u3POS33eMhHfAwAAAACA4S775HRr7XeqaleSO5PcmuTuJBuSrLv+pSVJfjnJ89M/vcj+JM+51MCqmpbkC0lemmRrkt9JcjbJtyR5a5IfraqXtNY2j3L4ZzP6d1p1TdUDAAAAADCqsab1SGvtb5L8TdL/pHGS/95ae9d1rmvAz6Q/lN6e5I4k911m7A+kP5j+QpJXttYuDuyoqpVJfjXJO5L82CjHfqa19rEJqhkAAAAAgDGMNa3HSG9K8pnrUchoWmv3tdYeaK21Kxj+rM7yD4cG0x2f7SyXTFx1AAAAAABcrTGfnB6qtfbx61XIBNjUWb6qqt4/IqB+dWf5Z5c49vaq+ukkM5McSHJfa23/daoTAAAAAOCGN65wOkmqakqSn0zyhiTPTTK7tTats+8FSf5pkn/fWts2kYVegT9M8ukkP5hkY1X9WZJzSb45ybcn+WCS/3CJY98+Yv2JqvpPSX66tXbmOtULAAAAAHDDmtrb23vFg6vqpiR/kuQtSW5O8liSW3p7e9+VJCtXrjyT5JNJTvb29n5xIgtduXLlM9L/QsaNvb29nx65v7e3NytXrvxvSVr6A+oXJ3lJkmcm+WKS97fW9o0457wkq9M/t/UvJvlQkv+d5IVJXpHka3t7e3//UjVV1ZtXrlz54TvvvHNFktx2223ZsmVLFi9enLVr12bFihVZtWpVFi1alE2bNmXGjBk5cOBAHnvssZw+fToHDx7MTTfdlM2bN2fJkiVZs2bNsGO2bNmS6dOn5+DBgzl9+nQef/zx7N+/PzNmzMj999+fZcuWDTtmwYIF+cpXvpKpU6fm0KFDefTRR3PmzJns27cvN998czZs2JCenp6sWrU6Fxcsz/Rda/K8Wxdk27ZtqaocOXIkJ0+ezLlz57Jnz57MmjUrGzZsyIoVK9LX15dbb701q1atyrx587J9+/a01nLs2LGcOHEiFy5cyO7duzNnzpysX79+sKaenp6sXr06c+fOzY4dO3Lx4sU89NBDOX78eC5evJidO3fmlltuGXbMsmXLsnbt2txyyy3ZtWtXLly4kBMnTuTYsWNJku3bt2f+/PlZt27d4DFLly7N+vXrM3v27OzevTvnz5/PI488kiNHjqSqsm3btixYsGDYMUuWLMmGDRsya9as7N27N2fPns3Jkydz+PDhTJ06NVu3bs2iRYuyZs2awe++ePHi3H///ZkxY0b279+fM2fO5NSpU3nwwQczbdq0wf4PHNPX15fFixdn8+bNg70c2f9NmzZl6dKlw3q5cOHCbNmyJdOmTRvs5cj+D/xtR/Z/ypQpOXz4cE6ePJmzZ89m7969g/1fvnz5sGPmz58/2P+jR4/m4Ycfzvnz55/U/1WrVg0u586d+6T+P/HEE9m5c2fmzJmTdevWDf69enp6smbNmsv2f+7cuYN96evrS09PT9atW5fZs2eP2v8HHnjgkv2fNWtW9uzZk3Pnzj2p/wsXLhx2XS5ZsiQbN27MzJkzs3fv3pw5cyaPPvpoDh06NKz/Q48Z2v+Ba/nUqVM5ePBgpk+fni1btox6LQ/t/9BreWj/V69ePfh3W7hwYbZu3XrJa3njxo3D+t/X1zdq/weu5YH+j+zlwLWcZFj/R7uWB/7tDPT/4sWLOX78+LD+j7yWB/o/9Foe2v8dO3Zk3rx5w3q5bNmywf4PvZaPHj06rP9r164dvMaWLl16yWt5ypQpT+r/wDED/d+3b98lr+Wn8r38Uteye7l7uXu5e7l7uXu5e7l7uXu5e7l7uXu5e7l7+bXey5cvX77yUpnoRKorm865M7jql5L8WpLeJL+e/pcM/kprbeqQMX+aZG5r7dsmtNCq70r/CxE/2Vp74yj7Zyb57SSvSv+LDz+b/vD8pUk+kOTpSV7fWvvsyGNHOdfXJFmfZEGS21tr6y83/t57721Jcs8994zjG331vGfrhcGff/454354HgAAAACY3Kobv2S8L0R8Q5L/1Vp7V2dO59GS7V1JnnbNlY3fLyZ5fZJfaq19uLV2qLV2srX2+SSvSzI9yfuv5ESdJ6z/qLP6ndelWgAAAACAG9h4w+lnJvmbMcY8lGTh1ZVzTQZeenjfyB2dJ58fSvL0qlp0hec72lnOnoDaAAAAAAAYYrzh9ONJ5o8x5mlJHr66cq7JjM5yycgdVTUjydzO6rkrPN+LOsud11gXAAAAAAAjjDecXpfklZ0XIz5JVc1L8r1JvnythV2FL3WW/7ITRg/Vm2Rakr7W2qMDG6vqO0aepPr9i/S/UPFYkj++PuUCAAAAANy4xvs2vN9M8skkn6yqHx+6o6rmJ/lo+l8i+B8noriqem2S13ZWezrLF1fVxzo/H2utvaPz87uTfH+SlyfZWlV/nP4nvV+a5Fs7P799xK/4y6ralqQvyYEk8zrjvzH9L1N8Q2vt5ER8FwAAAAAA/ta4wunW2u9U1fckeVOS/yvJiSSpqlVJviH9U2t8qLX2R5c+y7jcnuSuEdue1fkkyZ4k7+jUdqCqvinJLyS5s1PjlCQPJvlYkn/TWts64lzvTX9w/bL0z5N9McneJB9K8r7Wmik9AAAAAACug/E+OZ3W2o9X1ZfS/xTy85JUkm9Ksin9ge5HJ6q41lpv+qfkuNLxR9MfVr9jrLGd8T9/VYUBAAAAAHBNxh1OJ0lr7WNJPlZVN6d/Go9HWmunJ7IwAAAAAAAmr6sKpwe01h5P/1zOAAAAAABwxaZ8tQsAAAAAAODGI5wGAAAAAKDrhNMAAAAAAHSdcBoAAAAAgK4TTgMAAAAA0HXCaQAAAAAAum7aRJykqhYn+Y4kjyX5s9baExNxXgAAAAAAJqdxPTldVfdU1f+uqoVDtn1zki1Jfi/JHyX5q6qaPbFlAgAAAAAwmYx3Wo8fTtJaaw8N2faeJAuSfDT94fS3JHnLxJQHAAAAAMBkNN5w+uuTbBhY6UzncUeS32qt/URr7fuT9CX5JxNXIgAAAAAAk814w+lFSY4MWX9pZ/nfh2z7UpKnX0tRAAAAAABMbuMNpx9KsnjI+h1JLib5qyHbWpKZ11gXAAAAAACT2HjD6S1Jvr+qFlXV/PTPQd3XWjs5ZMwzkhyaoPoAAAAAAJiExhtOvz/J8iT7k+xL0pPk/x3YWVVTk3x7kvUTVSAAAAAAAJPPtPEMbq39j6p6S5I3dzZ9srX2iSFDvif9U3r8yQTVBwAAAADAJDSucDpJWmsfSfKRS+z7kyQLrrUoAAAAAAAmt/FO6wEAAAAAANdsXOF0Vb2gqv5ZVc0bsm12VX28qh6uqoNV9faJLxMAAAAAgMlkvE9O/0KSX2qtPTJk279O8iOdcy1K8r6qeuUE1QcAAAAAwCQ03nD6hUm+OLBSVdOT3JXky0mWJnlmkmNJ3jZB9QEAAAAAMAmNN5xemmTfkPUXJrklyYdba2daaweTfDbJ8yaoPgAAAAAAJqHxhtMtybQh69/e2fYXQ7YdTbLkGusCAAAAAGASG284vTfJtw1Zf02S/a21nUO2rUhy4loLAwAAAABg8hpvOP27SV5SVb9XVZ9I8uIkvzdizDcm2TERxQEAAAAAMDlNG3vIMP8uyfcl+cHO+rok7xrYWVV/P8k3J/lXE1IdAAAAAACT0rjC6dbaqSQvrapv7Gza3Fq7OGTIY0l+IMmqCaoPAAAAAIBJaLxPTidJWmv3X2L77iS7r6EeAAAAAABuAFcVTlfV9CQvT/LcJHNaa7/W2T4zydwkx0Y8UQ0AAAAAAIPG+0LEVNX3pf/p6D9M8htJeofsvj3Jg0l+eAJqAwAAAABgkhpXOF1VL0zymSQtyc8k+a9D97fW/ibJrvTPOw0AAAAAAKMa75PTv5L+lx6+sLX2gSQPjDKmL8nzr7UwAAAAAAAmr/GG0y9N8pnW2qHLjNmXZPnVlwQAAAAAwGQ33nB6TpJjY4yZdRXnBQAAAADgBjLeEPlAkm8YY8ztSXZeXTkAAAAAANwIxhtOfz7J91bVt4+2s6peleQlST53rYUBAAAAADB5jTec/tdJHk7yp1X1b5L8/SSpqjs76/8tyYNJ3jehVQIAAAAAMKlMG8/g1tqBqnplkt9N8vNDdv2PJJVkR5IfbK2NNS81AAAAAAA3sHGF00nSWltTVc9OcmeSFydZlOSRJH+T5LOttQsTWyIAAAAAAJPNuMPpJGmtPZH+p6X/x8SWAwAAAADAjWC8c04DAAAAAMA1u6onp6vqhUm+NcmCJFNHGdJaa792LYUBAAAAADB5jSucrqq5ST6d5LvT/wLES2lJhNMAAAAAAIxqvE9OvyfJy5J8KclHk+xL4gWIAAAAAACMy3jD6dckWZPku1trF69DPQAAAAAA3ADG+0LEeUnuE0wDAAAAAHAtxhtOP5Bk2fUoBAAAAACAG8d4w+kPJfn+qrr1ehQDAAAAAMCNYbzh9OeT/GmS/1VVb6qq51XV00b7TERxVfW6qvpgVX2pqk5WVauqT1xm/Iyq+smq+nJVHauqU1W1pao+UFVPv8xxd3WOOVVVj1TVF6vq1RPxHQAAAAAAeLLxvhBxd5KWpJL8p8uMa1dx7tH8cpLnJzmVZH+S51xqYFVNS/KFJC9NsjXJ7yQ5m+Rbkrw1yY9W1Utaa5tHHPfeJD/XOf9vJrkpyT9K8gdV9dbW2n+YgO8BAAAAAMAQ4w2Qfzv9wXO3/Ez6Q+PtSe5Ict9lxv5A+oPpLyR55dCXNlbVyiS/muQdSX5syPaXpD+Y3pHkW1prJzrb35NkdZL3VtXnWmu7J/A7AQAAAADc8MYVTrfW7r5OdVzq9w2G0VU11vBndZZ/ODSY7vhs+sPpJSO2v6WzfPdAMN35vbur6kNJfiXJm5K8c5ylAwAAAABwGeOdc/qpbFNn+aqqGvm9BuaP/rMR21/WWf7xKOf7/IgxAAAAAABMkHGF01X1RFX9yhhjfqmqLlxbWVflk1HK0QAAIABJREFUD5N8OskrkmysqvdX1Xuq6s/TP3f1B5MMzh9dVbOT3JrkVGvtwVHO90Bnedv1LRsAAAAA4MYztbe394oHr1y5sjfJX/T29v7FZcZ8Z5KX9fb2vuuaqxt+3mckuTvJxt7e3k+P3N/b25uVK1f+t/TPif2DSV6c5CVJnpnki0ne31rbN+R8S9I/3/TR3t7e943y+86nf1qP6b29vb8+Wk1V9eaVK1d++M4771yRJLfddlu2bNmSxYsXZ+3atVmxYkVWrVqVRYsWZdOmTZkxY0YOHDiQxx57LKdPn87Bgwdz0003ZfPmzVmyZEnWrFkz7JgtW7Zk+vTpOXjwYE6fPp3HH388+/fvz4wZM3L//fdn2bJlw45ZsGBBvvKVr2Tq1Kk5dOhQHn300Zw5cyb79u3LzTffnA0bNqSnpyerVq3OxQXLM33Xmjzv1gXZtm1bqipHjhzJyZMnc+7cuezZsyezZs3Khg0bsmLFivT19eXWW2/NqlWrMm/evGzfvj2ttRw7diwnTpzIhQsXsnv37syZMyfr168frKmnpyerV6/O3Llzs2PHjly8eDEPPfRQjh8/nosXL2bnzp255ZZbhh2zbNmyrF27Nrfcckt27dqVCxcu5MSJEzl27FiSZPv27Zk/f37WrVs3eMzSpUuzfv36zJ49O7t378758+fzyCOP5MiRI6mqbNu2LQsWLBh2zJIlS7Jhw4bMmjUre/fuzdmzZ3Py5MkcPnw4U6dOzdatW7No0aKsWbNm8LsvXrw4999/f2bMmJH9+/fnzJkzOXXqVB588MFMmzZtsP8Dx/T19WXx4sXZvHnzYC9H9n/Tpk1ZunTpsF4uXLgwW7ZsybRp0wZ7ObL/A3/bkf2fMmVKDh8+nJMnT+bs2bPZu3fvYP+XL18+7Jj58+cP9v/o0aN5+OGHc/78+Sf1f9WqVYPLuXPnPqn/TzzxRHbu3Jk5c+Zk3bp1g3+vnp6erFmz5rL9nzt37mBf+vr60tPTk3Xr1mX27Nmj9v+BBx64ZP9nzZqVPXv25Ny5c0/q/8KFC4ddl0uWLMnGjRszc+bM7N27N2fOnMmjjz6aQ4cODev/0GOG9n/gWj516lQOHjyY6dOnZ8uWLaNey0P7P/RaHtr/1atXD/7dFi5cmK1bt17yWt64ceOw/vf19Y3a/4FreaD/I3s5cC0nGdb/0a7lgX87A/2/ePFijh8/Pqz/I6/lgf4PvZaH9n/Hjh2ZN2/esF4uW7ZssP9Dr+WjR48O6//atWsHr7GlS5de8lqeMmXKk/o/cMxA//ft23fJa/mpfC+/1LXsXu5e7l7uXu5e7l7uXu5e7l7uXu5e7l7uXu5efq338uXLl68cLQ+daNXalb/fsKouJultrV0yeK6q9ya5p7U2ewLqG3re70r/CxE/2Vp74yj7Z6b/hY2vSv+LDz+b5LH0vyTxA0menuT1rbXPdsavSHIgyYHW2t8b5XzTk5xLcra1NvNytd17770tSe65556r/Xpd9Z6tf/tg+88/Z7zvxAQAAAAAJrkxXwA4EcZMJqvqO0dsesYo25JkapKnJXlDkq9MQG3j9YtJXp/k7a21Dw/Z/vmqel2SdUnen/7QOkke6SznXeJ880aMAwAAAABgglzJY7NfTP9UGeks7+p8RlNJLqZ/uoxuG3jp4X0jd7TW1lfVQ0meXlWLWmvHW2unq+pAkluravko805/fWe57TrWDAAAAABwQ7qScPpd6Q+lK8mvpj+sHm3O6SeSHE9yX2tt60QVOA4zOsslI3dU1Ywkczur54bs+vMkP5Lk+5J8dMRhrxoyBgAAAACACTRmON1a6x34uaruSvKZ1toHrmdRV+lLSb4xyb+sqv/VWjs7ZF9v+r9rX2vt0SHb/2P6w+lfqqrPtNZOJElVPSPJTyY5myeH1gAAAAAAXKNxvQ2vtfbM61XIaKrqtUle21nt6SxfXFUf6/x8rLX2js7P707y/UlenmRrVf1xksfT/0LEb+38/Pah52+t/VVVvS/JzybZUFW/l+SmJD+cZGGSt7bWdl+HrwYAAAAAcEMbVzj9VXB7njy/9bM6nyTZk+QdSdJaO1BV35TkF5LcmeRNSaYkeTDJx5L8m9GmG2mt/VxVbUjyU0nenP45s9ckeU9r7XMT/YUAAAAAABgjnK6qP0/nJYittf2d9SvRWmsvv9biOlOK9I5j/NH0h9XvGGvsiOM+nuTj4zkGAAAAAICrN9aT09/VWc4asT6WdjXFAAAAAABwY7hsON1am3K5dQAAAAAAuBqXDZur6m1V9a3dKgYAAAAAgBvDWE9C//sk3zewUlVPVNWvXN+SAAAAAACY7MYKp88kmTFkvTofAAAAAAC4amOF07uSfG9VLRuyzcsOAQAAAAC4JmOF0x9O8k1JDlbVE51tvZ3pPS73uXB9ywYAAAAA4O+yaZfb2Vr7QFUdSXJnkhVJvjvJ3iS7r39pAAAAAABMVpcNp5OktfapJJ9Kkqq6mOSjrbV3Xe/CAAAAAACYvMaa1mOklUm+eB3qAAAAAADgBjLmk9NDtdZWXq9CAAAAAAC4cYz3yWkAAAAAALhmwmkAAAAAALpOOA0AAAAAQNcJpwEAAAAA6DrhNAAAAAAAXSecBgAAAACg64TTAAAAAAB0nXAaAAAAAICuE04DAAAAANB1wmkAAAAAALpOOA0AAAAAQNcJpwEAAAAA6DrhNAAAAAAAXSecBgAAAACg64TTAAAAAAB0nXAaAAAAAICuE04DAAAAANB1wmkAAAAAALpOOA0AAAAAQNcJpwEAAAAA6DrhNAAAAAAAXSecBgAAAACg64TTAAAAAAB0nXAaAAAAAICuE04DAAAAANB1wmkAAAAAALpOOA0AAAAAQNcJpwEAAAAA6DrhNAAAAAAAXSecBgAAAACg64TTAAAAAAB0nXAaAAAAAICuE04DAAAAANB1wmkAAAAAALpOOA0AAAAAQNcJpwEAAAAA6DrhNAAAAAAAXSecBgAAAACg657S4XRVva6qPlhVX6qqk1XVquoTlxj7sc7+y32+MOKYu8cY/5bufFMAAAAAgBvLtK92AWP45STPT3Iqyf4kz7nM2M8k2X2JfT+S5FlJPn+J/Z9Nsm6U7auuqEoAAAAAAMblqR5O/0z6Q+ntSe5Ict+lBrbWPpP+gHqYqpqf5J8nOZfkY5c4/DOttUvtAwAAAABggj2lw+nW2mAYXVVXe5ofSXJzkk+11o5NRF0AAAAAAFybp3Q4PUH+aWf5kcuMub2qfjrJzCQHktzXWtt/3SsDAAAAALhBTepwuqpenOQfJNk29CnsUbx9xPoTVfWfkvx0a+3MdSsQAAAAAOAGNbW3t/erXcMVWbly5TOS3J1kY29v76ev8JhfS3J7kv+nt7f3r0bZPy/J6vTPbf2LST6U5H8neWGSVyT52t7e3t+/1Pmr6s0rV6788J133rkiSW677bZs2bIlixcvztq1a7NixYqsWrUqixYtyqZNmzJjxowcOHAgjz32WE6fPp2DBw/mpptuyubNm7NkyZKsWbNm2DFbtmzJ9OnTc/DgwZw+fTqPP/549u/fnxkzZuT+++/PsmXLhh2zYMGCfOUrX8nUqVNz6NChPProozlz5kz27duXm2++ORs2bEhPT09WrVqdiwuWZ/quNXnerQuybdu2VFWOHDmSkydP5ty5c9mzZ09mzZqVDRs2ZMWKFenr68utt96aVatWZd68edm+fXtaazl27FhOnDiRCxcuZPfu3ZkzZ07Wr18/WFNPT09Wr16duXPnZseOHbl48WIeeuihHD9+PBcvXszOnTtzyy23DDtm2bJlWbt2bW655Zbs2rUrFy5cyIkTJ3LsWP+sLNu3b8/8+fOzbt26wWOWLl2a9evXZ/bs2dm9e3fOnz+fRx55JEeOHElVZdu2bVmwYMGwY5YsWZINGzZk1qxZ2bt3b86ePZuTJ0/m8OHDmTp1arZu3ZpFixZlzZo1g9998eLFuf/++zNjxozs378/Z86cyalTp/Lggw9m2rRpg/0fOKavry+LFy/O5s2bB3s5sv+bNm3K0qVLh/Vy4cKF2bJlS6ZNmzbYy5H9H/jbjuz/lClTcvjw4Zw8eTJnz57N3r17B/u/fPnyYcfMnz9/sP9Hjx7Nww8/nPPnzz+p/6tWrRpczp0790n9f+KJJ7Jz587MmTMn69atG/x79fT0ZM2aNZft/9y5cwf70tfXl56enqxbty6zZ88etf8PPPDAJfs/a9as7NmzJ+fOnXtS/xcuXDjsulyyZEk2btyYmTNnZu/evTlz5kweffTRHDp0aFj/hx4ztP8D1/KpU6dy8ODBTJ8+PVu2bBn1Wh7a/6HX8tD+r169evDvtnDhwmzduvWS1/LGjRuH9b+vr2/U/g9cywP9H9nLgWs5ybD+j3YtD/zbGej/xYsXc/z48WH9H3ktD/R/6LU8tP87duzIvHnzhvVy2bJlg/0fei0fPXp0WP/Xrl07eI0tXbr0ktfylClTntT/gWMG+r9v375LXstP5Xv5pa5l93L3cvdy93L3cvdy93L3cvdy93L3cvdy93L38mu9ly9fvnzlleSv16paa934Pdesqr4r/S9E/GRr7Y1XMH5ekoPpfzr81vHMN11VX5NkfZIFSW5vra2/3Ph77723Jck999xzpb/iq+o9Wy8M/vzzz5nUD88DAAAAAON31S8AHI8p3fglXyVvTDIryafH+yLE1tq+JH/UWf3OiS4MAAAAAOBGN5nD6YEXIX74Ko8/2lnOnoBaAAAAAAAYYlKG01X1oiTPT/+LEL94lad5UWe5c0KKAgAAAABg0KQMp5O8ubP8yOUGVdV3jLKtqupfJHlxkmNJ/njiywMAAAAAuLE9pd+GV1WvTfLazmpPZ/niqvpY5+djrbV3jDhmbpIfTnIuycfH+BV/WVXbkvQlOZBkXpKXJvnGJI8leUNr7eS1fg8AAAAAAIZ7SofTSW5PcteIbc/qfJJkT5J3jNj/hvTPE/2pK3gR4nuTfGuSlyVZmORikr1JPpTkfa01U3oAAAAAAFwHT+lwurXWm6R3nMfcm+TeKxz78+OvCgAAAACAazVZ55wGAAAAAOApTDgNAAAAAEDXCacBAAAAAOg64TQAAAAAAF0nnAYAAAAAoOuE0wAAAAAAdJ1wGgAAAACArhNOAwAAAADQdcJpAAAAAAC6TjgNAAAAAEDXCacBAAAAAOg64TQAAAAAAF0nnAYAAAAAoOuE0wAAAAAAdJ1wGgAAAACArhNOAwAAAADQdcJpAAAAAAC6TjgNAAAAAEDXCacBAAAAAOg64TQAAAAAAF0nnAYAAAAAoOuE0wAAAAAAdJ1wGgAAAACArhNOAwAAAADQdcJpAAAAAAC6TjgNAAAAAEDXCacBAAAAAOg64TQAAAAAAF0nnAYAAAAAoOuE0wAAAAAAdJ1wGgAAAACArhNOAwAAAADQdcJpAAAAAAC6TjgNAAAAAEDXCacBAAAAAOg64TQAAAAAAF0nnAYAAAAAoOuE0wAAAAAAdJ1wGgAAAACArhNOAwAAAADQdcJpAAAAAAC6TjgNAAAAAEDXCacBAAAAAOg64TQAAAAAAF0nnAYAAAAAoOuE0wAAAAAAdJ1wGgAAAACArhNOAwAAAADQdU/pcLqqXldVH6yqL1XVyapqVfWJS4z9WGf/5T5fuMSxd1XVl6vqVFU9UlVfrKpXX99vBwAAAABw45r21S5gDL+c5PlJTiXZn+Q5lxn7mSS7L7HvR5I8K8nnR+6oqvcm+bnO+X8zyU1J/lGSP6iqt7bW/sPVFg8AAAAAwOie6uH0z6Q/NN6e5I4k911qYGvtM+kPqIepqvlJ/nmSc/+nvfuOj6rK/z/+OpNkUiEJKbSEIr2DYEUUcS0o1rWsBbu72FbX9be6rq7dLXbXsvrVlVVWXXtX7KJgRVgQlN4SWkKAkJ6ZOb8/7k1IQgIhTOYmw/v5eNzHZc49d+5nMh/unDlz7rnA1AbbDsbpmF4G7Get3eyW3w3MBu4xxrxtrV0ZhtciIiIiIiIiIiIiIq42Pa2HtfZTa+0Sa63dg6eZDCQCr1prCxtsm+Ku76zpmHaPuxJ4BIgHLtiDY4uIiIiIiIiIiIhII9p053SYXOKun2hk2wR3/X4j295rUEdEREREREREREREwiSqO6eNMQcBw4DF1tpPG2xLBroDJdbadY3svsRd92/dKEVERERERERERET2Pm19zuk99Wt3/X+NbEt111ub2LemPK25B3vssceaW9VTKXX+/ViTs3iLiIiIiIiIiIjIXspeeumlprUPErUjp40xqcDpNHIjxN3U5HzXxphfG2O+34PnFhEREREREREREdkrRfPI6XOAJOCFRm6ECNtHRqc2sq1ueVMjq7HWPkHjc1mLiIiIiIiIiIiIyE5E7chptt8I8fHGNlprS4F8IMUY07WRKv3c9eJWiE1ERERERERERERkrxaVndPGmAOAETg3QvxsJ1U/cdfHNLJtYoM6IiIiIiIiIiIiIhImUdk5zfYbIe5qyo1/uus/GWPSawqNMb2Ay4FK4OlwByciIiIiIiIiIiKytzPWNnm/P88ZY04CTnIfdgGOBpYDX7hlhdbaaxvs0xFYC8QB3ZuYb7pu/XuBa4A84GXAD5wBZABXWmsfDs+rEREREREREREREZEabf2GiCOB8xqU7eMuAKuAaxtsPxtIpukbIdZjrf29MWYecAXOiOsQ8ANwt7X27T2IXURERERERERERESa0KZHTrcnMTE+Gwq1jb+lMYZoeV99BtrIn1XaEB/Or0jhFE3/b8LNGB/WhvsvLmFnDCiHd5/xQbTlt3KhhQygv9vOtMbnL7Td9l60tw3a6t892rTW/5twi/Z83xPtsi2stsDOtaf2X5t5L/eOdlJ7OWdHUKG1Nqu1D9LWR063G6GQxX7vdRQOM8YyY8YMr8MIi0MPPRRNrCINmSvADj81vM8572WOX3RRWJ8zWrw14CnO/KrK6zBkF54/yM/ol5Z4HUa7M/u0fnBDlDW07zKcecZCr6Nod57/72A48UGvw2jTQm9chR07MuzPa2bObTPt6LqiqU3dGLWzI6M12q2tQW3hprXHtrDahTvXrtp/baRdt7e0k1qrrdNemZlzV0XiONF6Q0QRERERERERERERacPUOS0iIiIiIiIiIiIiEadpPTxWWQWX/RU++haKiqFvDtx1OUwc63VkTbv99tuZPXs2FRUVdOrUibPOOotJkyZ5HVbUW7IRht0Fp46CaQ1vEyr1jF/2GV+XFRFrDADdYxNZNPAYj6Nqvvx3lrH44TmUryslPjORkX89lIwxXbwOKyq8NCG93uNgZTl9T5nCmN8/4FFE0a1yYx6rn7yF0sVzMLF+0g88htwL/oSJ2YuaH4U/wfTLYf1sSMqCCXfDgJO9jqrVLV7yH1aseJ0tWxfTs8dxHHjAXbXb1m/4iu9n30FZ2ToyMoZz4P53kpzc3cNoPbZ8Bqz+Frathe6jYd+znfJQAL5/BrasgfIiGHsFZPbzNtZGPLyugKkbiphfVsGZWWlM7dcTgK+3lXLTqnXMLi0nBhifmsJD++TQ1R8XnuP+F6a+DfOXwplHw9RbnPKFy+Hcm2FZnvN49CB46FoYvE+TTxVxr7zyCu+//z7Lly/niCOO4IYbbtihztNPP83TTz/Nfffdx5gxYzyI0lsPfw5Tv4b56+DM0TB1slO+chP0vhmS/dvrXnck3DTRmzhb28OFS5m6eRXzK7ZyZlouU3P3q91WFgpw7bp5vLglj2obYkRiGjP6jPcu2BZYMW0ha15dwrbFRXSb1IdRfz0UgLw3lzLv5pm19WzIEqoIMu6VE0kbmulVuC22+KVHWfHuM2xZ9iM9jzyDA296qnbb6o9eYv6Tt1FWkE9Sdg4jptxOzmEnehjtntn43rNs+uxVylcvotPYSfS64u+12wo/fpH1rz1O9ZZCUgaOpudlf8HfqbOH0bZQoBKmXwYrPoKKIkjvC+Pvgj7uiWjFx07br3g1dDsAjp8KqT1bPaxgsIrvZ9/G+g1fUVW1lZSUHowYfjXduh5KMFjFV1//gaKiHyktW8uEw6fSOXv/Vo+pWYIBmPciFCyGqjJIyYRBk6DzYChaCT+/A1vynLm3M/vBsFMgIdXrqMOiMhTismV5fLR1G0XVQfomxnNXz65MTO/IwrIKzl28imUVzpRCo1MSeWifHAYnJXgcdevYi74dtk2BIOR2hs+fgB5d4N2ZcPofYf4L0Kub19E17pxzzuG6667D7/ezatUqrrrqKvr168eAAQO8Di2qXf4i7Nf6n2lR4+Fuo7g4o7fXYey2gpn5/HTPd4y+fwJpw7OoKCjzOqSoctonm2v/HSgv5bXjcugx4ZceRhTdVj95C3EdOzH8iVkES4tZfPv5FEz/D9nH7iW/sIUC8PKJMGoKnPkhrP4cXjoeLpwDGf29jq5VJSZmM2Twb1i3fibBYGVteWXlZr6ceRX773cb3bsdzrz5DzFz1u856sgXPIzWYwmpMOAo2PgzBKvrb8vYB/ocBt9N9SS05ujmj+PG3M5M37KN8tD2WwhtDgT5dZdMjk7rQKwxXLE8jwuWrOb9IX3Cc9wsuPEimP4VlFfWL3/5b9CzK4RC8MhL8KsbYF4bSrHMzEzOPfdcvv32WyorK3fYnp+fz2effUZGRoYH0bUN3VLhxmNg+k9QXr3j9i13Q2xM5OOKtG5xidyYPZDp2zZQboP1tv067wcCNsRPA46mU4yfueVbPIqy5RKyk+h32QgKvsgnWLn99eWc0JecE/rWPl7z6mIWPzqX1CHt8/9EYlZXhpz/R9Z98yHByvLa8rKN+Xx16/mM+/srdD3waNbOeo+ZfzqTE15dQkKnbA8jbrm4Ttl0/eVlbJ37BbaqorZ824JvyH/uXvrfMo34Lj1Z8/QdrHjgdwy47TkPo22hUAA65MI5n0NqD1j6Lrx2Olw8H/wp8OopcOyT0O94+PwmeO0MOP/rVg/L2gBJSV04YsIzJCd1Ze26GcycdQ0Tj3mDxIQsMjP3ZUD/yXw563etHstusUFITIdDrnTWGxY67Z4J10F1GfQ8GPYb5NzEcv7LMOc5OOhSr6MOi4C15MbH8fnQvvSI9/Pu5mJO/3kl80cNoJs/lpcH9qJnvJ8Q8Mi6Qn61aCXzRg30OuxWoWk9PJacCLf8xumI9vlg0jjo3Q1m/+R1ZE3r3bs3fr8zXMG4I1Pz8/O9DCnqvfA9pCXCEdHdlyHAon/8QP/LRpE+MhvjMyR2Tiaxc7LXYUWl1Z+8Qnx6NlkjD/E6lKhVtXEN6Qcfi88fT1x6Fqkjx1G+Zi+6OU/hz85o2P1/B74Y6DUBcsbCj896HVmry805kpycXxAfn1avfE3eh6R27EuP3GOIiYln2NDL2bJ1EcXFyz2KtA3oNgK6Dgd/g3O9Lxb6jIeMPs5ooTbqlIw0TspIIyO2/piXiekdOS0zjY6xMSTF+LiiayYzi0vDd9wJcNJ4yGgweCqtg9OuNgashRgfLF0TtsOGxWGHHca4cePo2LFjo9sfeOABpkyZQlxceEaZt0enjISTRkDGXt4EOiW1Oyeldicj1l+vfFHFNt4sXssTOaPJio0nxhhGJ6U38SxtV9ejetH1F73wp8XvtN6a15aSc1Lf2u+e7U3u+JPJOexE4lM71SsvK8gnrkMa3Q46BmMM3cceS2xiMiX57fczMf2Ao0nb/0hiO9T//N86+1PSD5pIYm4/fHF+up56OSU/fUfl+ojcay28/Mlw6C2Q1svpMO03CVJ7O1fJLXoVMofAoNMgNgHG3QIb/+e0CVtZbGwSw4ZeQUpyd4zx0b3beJKTc9hctICYGD8DB5xLVtZojGljv+zFxsPAiZCU4fw9uwyF5E7OlWOdB0P3URCXALF+6D0ONq3wOuKwSY6J4ZYeXemVEI/PGCZ1SqV3gp/ZJeWkxcbSKyEeYwwWiDGwtGLHH7SjhTqn25gNm2DxagjToJJWc99993HkkUdyzjnnkJGRwYEHHuh1SFGruBz+/A7ce4rXkbQvf1w/n8wFbzJ26ad8VrLR63CaxQZDbPmxkMrNFXx85It8eOjzzL9tFsGKgNehRaUV702j98Sz2+0XnfYg+9jzKJr5DqHKcqo2rWfr3Bl0HHmo12FFUCN3gbcWCn6MfChtxNatS0lL236lVWxsEinJuWzdutTDqCQSZhSXMCSCl6KmjYeEsXDl3XDDBRE77B779NNPiY2N5aCDDvI6lDat558h50a44FkoLPE6msj7pryInv4kbt6wgMwFbzJs8Qe8sjXP67BaRVn+NjZ9v57cE9vetEZ7qtPA0XTsOZC8L94iFAyS9/kb+OLiSes7zOvQws5a67SBthcARMeghZINULTY6ZQuWADZI7Zv8ydDWh8oXBDxsMorCtm2bSWpqX13XbktqSiGkgLo0HXHbZuWQcfone5yQ1U1i8sr67WX0r6eR8Ks/3Hl8nxuyGmH0+A0k6b1aEOqA3D2TXDecTCwl9fR7Nw111zDVVddxYIFC5gzZ07tSGoJv5vegYsOhtz2NxjCM3/rOozB8R3xGx8vbFnD8StnMbffL+gTn+J1aDtVWViOrQ6x7v2VjP3PJEysj+8u+5DFj81l0O/2vvkmW1Pp+tUUzJnBATc87nUoUS1l8P4UfvQic84dBaEgGYedTNr+R3odVuRkDITkbPj6bmf09KpPnak9eh7udWSeCQTKiI+vP3IsLq4D1YHwjaiVtmdeaTm3rdnAGwMjN93Wls+gtBz+/bYzxUd7UFZWxhNPPMG9997rdShtVmYKfPf/YGQObCp1pr07eypMv8LryCIrr7qMHyuK+WXH7qwdNImvyjZx3MovGRzfkUEJjY/Ib6/yXl9KxpjOJOV28DqUsPPFxNB74jl8dfO5BKsq8MX6GXvn88QmRt8lA6n7Hsa+wgg9AAAgAElEQVTy+68m86gzSejSi3UvPwzGEKozzUm7FKyGN8+GYedB5kCoLnHuMVJXQipUbYtoWKFQNV99/Qd69zqRjh3b0E0XdiUUhNnPQu7+0KFBR+zWfFg0HQ642JvYWll1yHL24lWcl92JgXU6p7ccOJzSYJB/b9xMz/jovaJKI6fbiFAIJt8E/lh4+Dqvo2memJgYhg8fTkFBAa+//rrX4USluXnw0c/wu723H6NFDkjKoENMHPG+GM7r1IuxSRm8u22912Htki/B+b2w9+TBJGQnEd8pgX0uGMrGz9vY9chRYMV708gcPpaUbu1vXvL2woZCLLnjQtIOOIpR0/7HiH99S6C0mPxpf9/1ztEiJg5OfR2WvgMPdYFv7oVBp0PHHK8j80xsbBLV1fWHOVYHSoiLjb4v4uJYWl7JxIXLebB3d8alRvZH4uREmPJL5waJG4sieugW+de//sXRRx9Nt25t9MYzbUBKPIzp6cw33bkjPHw6fPCzc6Xh3iTRxBBnDDd2HoTf5+OwlCwOT87mg5INXocWdnlvLCXnpOgbNQ2w/tuPmfvIH5nwyEecMaOUIx79mG//MoXNi+d6HVrYdRx2MN1O/y3L77mC+Zcdhj87B19CMnEZ7XgUrA3Bm5Mhxg9HP+yUxaVAZXH9epXF4I/cjyvWhvjq6+vx+eIYM/rGiB13j9kQ/PCsMxXe8FPrbyspgK8fd26GmNHGpxlogZC1TF6yCr/P8PA+O35PSI6JYUqXDM5dspqNVY3ciCEKaOR0G2AtXHQ7bCiCdx+EuHb2rgSDQc053Uo+WwIri6DHTc7jkkoIWli4Dn643tvY2hNjwDZ2eX0b40+NJ6FLMmiWiVa38r1pDJr8B6/DiGrBki1Ub1pH9jGT8cXF44uLJ/PwX5L/wv3kTG4nv8KGQ/ZwmPz59sf/PtgZXbOXSk3ty4oVb9Q+DgTKKClZ0/4uOZVmWVVRxS8WLOWmnM5Mzu606x1aQSgEZRWQvxE8CqHZfvjhh3qDPrZs2cLNN9/MWWedxdlnn+1xdG1TTZOp7bfywmt4QuquK0WBotkbqNhYRreje3kdSqvYvOR/ZI88hIxBowHIGDyGjCH7sf67T0jvP9Lj6MIv+5hzyD7mHAAq1q5g/SuPkpjbTm+qZC28cxGUboAz3nUGJABkDYH5/95er6oUNi9zpvyISFiWb769kYrKTRw27p/4fO1kpK21MOd5qNwGB/7G6aCuUVYEsx6F/kdB7n7exdhKrLVctHQNG6oCvDt4H+J8jXcGhICyUIj8qmqy/e3kfd0NGjndBlz6F/hpBbx1PyRGbiq+Ftm8eTMff/wxZWVlBINBvv32Wz7++GP23Xdfr0OLSr8eC8tugbl/dJYph8BxQ2D65V5H1nZtCVYxfdt6KkJBAjbEfzavZkZJIUentI9f5XNP6ceKZxdSuamcqq2VrPj3AjqP7+F1WFGlYN5XlBWspceEX3odSlSL7dgJf3YOBR88hw0GCJQWs+mz10jqGZ13mG7SxnkQqHDuNv71PVCyDoaf73VUrS4UChAMVhKyQawNOv8OBcjp/gu2Fi9hzZoPCAYr+XHBY6Sl9m9fl5yGWyjoXBZsQ84SrHbKAIIB53G9em2rGy5gLRWhEEFrCVqoCIUIWEt+ZRUTFizl8i5ZTOmaGf7jBqCiEoIhCAadfwcC8OHXMOdnp6y4BK65H9I7wKA2dKFMIBCgsrKSUChEKBSisrKSQCDA/fffz9SpU3nqqad46qmnyMjI4Nprr+Xkk0/2OuSICwShotp9f0POvwNB+GYlLNrg/OiwqQR++zKM7wepiV5H3DoCNkRFKOj+/7K17dtDU7LoEZfEXzb+TMCGmFlayGelBRyd0r7mIw0FQgQrA9iQxQadf4cCodrta15fQtejehGb0r6nkAwFAgQrKwgFg9hQ0Pl3IEDGoDFs/N/M2pHSRYvmUDB3Zruec9oGA4SqKiEUwoZChKoqa8vKVy/GWktVwVpWPX4j2ceeR2xKO/2h5f1LofAnOP0tiKtzAup/snNvkZ9fcdp/X97mDFTIjEz79/vZt1JcvJxDD3mE2Nj6nUvBYBXBoHNDvVCommCw0pkLvC2Y96Izd/cBv3ZGotco3wIzH4behzhLFLp0WR4/lVXw1uDeJMZs76L9cMs25pSUEbSW4kCQa1bkkx4Tw6AI3r8jktrZGN3os2odPP4qxPuhy9Hbyx+/Ac6e6F1cTTHG8Prrr3PvvfcSCoXo3LkzV155JePGjfM6tKiU5HeWGinxkBAHWdE35VrYVFvLjesX8HPlNmIwDEzowOu9DmZAQvv4o/W/bBRVmyv45OiXiYmPodvE3vS7dMSud5RmW/Hus+SOP4m45PaRE+1Zn2sfYc3UO1n/xhPgi6HDkAPIOf9PXocVWfOfhf896XQq5o6DMz907koe5RYs/Cc/Lni09vHKVW8xdMhlDBt6BYcc/ADf/3AnX31zHRmdhnPwwXv5/LqLP4BF729/nPc9DDjGuXP9x3dCuTsfxVePOesj/+zc0b6NuGPNem5ds30qgWkFm7k5tzMGw/KKKm5ds55b12yfWqvkoOHhOe5TcOv/bX887T24+RLnpuJX3g15GyExHvYbDO//AxLa0H+7Z555hqlTp9Y+/uCDDzj//PO58MIL69WLiYmhQ4cOJCUlRThC793xPtz63vbH076DmyfCgM5ww1uwcRt0TIAjB8Lz7eiGl7vrjg0/cevGn2ofT9uympuzB3FLlyG80etgLs6bzV83LqKnP4lncvdjYDubb3rJY3NZ/PCc2sf5by6j/xWjGHDlvgQrA6x9bwVj/nGEhxGGx4Kpd/HjU3fUPl75/nMMvehGhl38Z4ZddBNf3nAmFZs3EJ+WxeDzrqPrAe33/hzrXnmUdS/9o/Zx0Rdv0PW0K8k+7nxWPHgNlRtW40tIJvPwX9LtjKs9jHQPbF0Fcx6HmHh4sM4AqImPw9Cz4ZRXnInw3zwHuh0AJ70QkbBKS/NZuuxFfD4/r7+5/Qbk+42+hV69juedd4+ltGwtAJ99fgkAx0/6kJTk7hGJr0llRbByFvhi4f0605CMOANKC6Bsk9NOqttWmnR35ONsBasqqnh8wybijaHLt9tvmvl4nxz8Ph9XLs8jr7KaRJ9hvw5JvD+kDwm+6BxjbNrMLyXtnDHG2u+9jsJhxsCMGTO8DiMsDj30UOzDXkchbY25AmzDeaj29Dnnvczxiy4K63NGi7cGPMWZX1V5HYbswvMH+Rn9UhTc8TzCZp/WD26IsrbQXYYzz1jodRTtzvP/HQwnPuh1GG3bG1dhx4b/UnMzcy5tpR1dVzS1qRujdnZktEa7tTWoLdy09tgWVrtw59pV+6+NtOv2mnZSK7V12iszc+5sa+2Y1j5OdHa5i4iIiIiIiIiIiEibps5pEREREREREREREYk4dU6LiIiIiIiIiIiISMSpc1pEREREREREREREIk6d0yIiIiIiIiIiIiISccbadnKH0jbOGFMArPI6DhEREREREREREZE91NNam9XaB1HntIiIiIiIiIiIiIhEnKb1EBEREREREREREZGIU+e0iIiIiIiIiIiIiEScOqdFREREREREREREJOLUOS0iIiIiIiIiIiIiEafOaRERERERERERERGJuFivA4gWsbGx1udTX7+IiLRdoUAQo8+qVmGxqB0ge6tQIIQxxuswRDxnTWQ/CwKBID6f/u+JiEjrCAaDhdbarNY+jjqnw8Tn81FcXOx1GCIiIk3KSE5jxqC/ex1GVBq/8k8UFBR4HYaIJ9KTMni082yvwxDx3G+L96OgMHKfBcnJqQwbdn3EjiciInuXuXNvWxWJ42iIj4iIiIiIiIiIiIhEnDqnRURERERERERERCTiNK2HiIiIhM3xi26lKLANnzEkGj9jOwxmTHJ/phZ+yEv9bqitd9mKRygIbN2hbP+U/lhgTukyHuo1pXbbyYtvJ9eftUPZlOzjODptX8b8eBUJxo8xkOJL5MjUUVzV5URijH6HF5H262+bzmVN9SLu7/wFccYPwFNb/sjX5e8Qa+IAyIzpxoj4wzk25RKSfB0A+LLsNZ7eeiN+kwBAB186A/z7c1zKJXSJ7e3NixERaaEFCx4kENjGkCHXEBubVFv+88+PU1GxgQ4d+lBauhqAUCiIMWBMDADp6cPJzT3Ok7hFpHnUOS0iIiJhdV/PSzggZQAbq7dwxcrHiDUxrKjcwOZACemxKQRskCUVa0nw+euVzS9fyW86H4u1lqkFHxG0IWKMj8LqYgI2xKKKvHpla6oK2Te5T+1xn+/7B3Ljs1hZuYHfrPgHPeKzOLXTIR7+JUREWq4wkM/iqtkkmQ7MrfiE/RKPqd02MeVCTulwNdW2krzqxby07R7+suksbsz4L/E+p+OmT9xIbsj8DyEbpDCYz/TSp7m18FT+lPE8OXH9vXpZIiIt4vens3nzj2Rl7Q9AefkGrA0AkJNzHPHxaQCsWvUGfn8Hunad4FmsIrJ7NJxIREREWkV2XBpjOwymILCV7nEZ/FC6FICfy/PYJ6EL+yb3qVcWspbBCbkMSexBwAZZXJEPwJyyZYxJ7ktPf3a9shx/JllxqTsct1d8Z0Ym9WFZxfoIvVIRkfCbVf46feJGMDbpJGaWv95onTgTT2//MK5Mf5SS0Ba+LH9thzo+E0N2bA8mp97MAP9+vFHySGuHLiISdunpwygqmlf7uKjof6SnD2/WvoFAKcuW/Yf58+9h/vy/s3z5C1RXb6vdvmnTDyxY8CDz5v2VhQsfYsuWhbXbCgtn89NPj7j7PUdVVXH4XpSIAOqcFhERkVayvmozM7ctZEBCDvsm92FO2TLA6VgeldSHkUn71CsbltSLOF8scb5Yhib1rO24/qF0GSOT+zAieZ96ZaOS+jR63OUV65lbtowBCd0j8CpFRFrHrPI3OTBxEgcmHs+CyplsDRY2WTfRl8zg+INZXPX9Tp9zdMKRLKmaHe5QRURaXXJyDqFQJRUVBVgbYsuWhXTqNKxZ+1prycwcw5AhVzF48G8ByM//AIBAoJz8/A/p23cyw4dfT9++55OQkAXA5s0/UlDwDb17/4qhQ39PYmIXVq/e8UdAEdkz6pwWERGRsLp29ZOMX3g9F694kH2T+3BB1pHsm9yXH0rdjujSZYxM3odRyX3qldWdomPfpL7McbfNLVvGqKR9GJXUp17Z6OS+9Y57zrK7OXzh9fxu1ROcmH4QJ6QfEImXKyISdourZrMpuJb9EifSK24IWTG5fFP+9k73SfNlUxrausd1RETaqprR09u2LSc+PoO4uI7N2i8uLoXU1AH4fHHExCTQufMhlJSsqlenoqKAUCiA39+xtnN606bZdO48joSEDIyJoUuXwygpWUN1dUnYX5vI3kxzTouIiEhY3dPjYg5IGVCvbFRyH27Pf57iYBk/lq3kjpxzSYqJZ1OgmOJgGf8rW85ZGeNr6++b3IeXir6gOFjG5kAJPeKz6RTbkVvyplEcLGNZxTpGJdcfOT2tz/8jNz4rEi9RRKRVzSp7nSHxB9PBlw7AgYnHMbP8DY5KOb/JfbaENpDs23Gqo7o2N6OOiEhb1anTcJYs+TdVVVvo1GlEs/cLBivJz5/Otm3LCQYrAAiFnPmqY2MT6dnzZAoKvmbVqtdJSelJ9+5HER/fiaqqreTlvUNe3ru1z2WMj+rqYuLiUsL74kT2YuqcFhERkVaX488kKzaVV4tm0cWfTlJMPADDEnvxatEsykKVDE3qWVt/WFIvSkIVvFo0ixFJ+wCQEpNAZpzzHFmxqXT3Z3jyWkREWlOVreC7ivcJEeLqDeMACNgqymwxq6t/bnSfilApCyu/YlLKlJ0+9w8VH9HPPzrsMYuIRILfn4bfn0Zx8RJyc49v9n4bN86kurqY/v0vJi4uhdLSPJYsebp2e2pqf1JT+xMKVbN27UesWfMOfftOJi6uI926HUla2qDWeDki4lLntIiIiETEyOR9eK7wU45JG1Ov7JmCjxmU2IMEn7+2PMHnZ3BCLs8VfsqF2Udtr5/kPMf+DUZmi4hEix8qPsZHDLdlvUEscbXlj235HbPK36hXt9pWkR9YwkvF95DkS+WQxJN3eL6QDbIpuI4PSqeyqOo7/pTxfKu/BhGR1tKjx/EEgxXExPixNtSsfYLBKoxxpvQIBMrYsOGL2m1VVcWUl6+nQ4feGBODzxeHMc4MuJmZY9iw4QsSErJISMgkECinpGSlOqtFwkyd0yIiIhIR+yb3ZfrWHxjpjoQGGJnUhweCb3BC0oGN1p9XvrJe/VHJfXix6IsdpvQQEYkWs8pfZ2zSyWTEdKtXPiHpbJ4rvovB/oN4r+RffFj6LBZLZkw3RsSP57iUXxPvS6qtv6x6LpeuH43F0sGXxgD//tyU8SLd4nT+FJH2Kz6+027vk519EKtWvcqPP95NXFxHsrL2p7h4qbvVsnHjTFatehUwJCV1JTf3OADS04cSClWzcuVLVFVtJSYmkY4d+6pzWiTMjLXW6xiigt/vt8XFxV6HISIi0qSM5DRmDP6712FEpfEr/0RBQYHXYYh4Ij0pg0e7zPY6DBHP/bZ4PwoKI/dZkJycyvDh10fseCIisneZO/e22dbaMbuuuWd8rX0AEREREREREREREZGG1DktIiIiIiIiIiIiIhGnzmkRERERERERERERiTh1TouIiIiIiIiIiIhIxKlzWkREREREREREREQizlhrvY4hKhhjtgGLvI5DokImUOh1ENLuKY8kXJRLEi7KJQkX5ZKEi3JJwkW5JOGgPJK2ZoC1tkNrHyS2tQ+wF1lkrR3jdRDS/hljvlcuyZ5SHkm4KJckXJRLEi7KJQkX5ZKEi3JJwkF5JG2NMeb7SBxH03qIiIiIiIiIiIiISMSpc1pEREREREREREREIk6d0+HzhNcBSNRQLkk4KI8kXJRLEi7KJQkX5ZKEi3JJwkW5JOGgPJK2JiI5qRsiioiIiIiIiIiIiEjEaeS0iIiIiIiIiIiIiEScOqdFREREREREREREJOLUOb0HjDE5xph/GWPWGmMqjTErjTEPGGPSvY5NIs8Yc6ox5h/GmC+MMcXGGGuMmbaLfQ42xrxrjCkyxpQZY+YZY642xsTsZJ9JxpjPjDFbjTElxphvjDHnhf8ViVeMMRnGmIuNMa8ZY5YaY8rd9/tLY8xFxphGz93KJ2mMMeZvxpiPjTFr3FwqMsbMMcbcbIzJaGIf5ZLskjFmsvtZZ40xFzdRZ7fzwhhznjHmW7f+Vnf/Sa3zKiTS3PaybWJZ38Q+OidJk4wx44wxrxhj1rnfydYZYz4wxhzbSF3lktRjjDl/J+ekmiXYyH7KJWmUMeY49xyU57a9lxtjXjLGHNREfeWStArjuNAY87UxZpubX3OMMb9tmF/GmDhjzFXGmKeNMXONMVU7a+O7+4w1xvzdGPOdMabA/QxeYYx50hjTd7di1ZzTLWOM6QPMArKBN4Cfgf2Bw4FFwFhr7SbvIpRIM8bMBUYAJUAeMBD4j7X2nCbqnwi8AlQA/wWKgOOBAcDL1trTGtnnCuAfwCZ3nyrgVCAHuNdae22YX5Z4wBgzBXgMWAd8CqwGOgOnAKk4eXOarXMCVz5JU4wxVcAPwEJgI5AMHAiMAdYCB1pr19Spr1ySXTLG5ALzgRggBbjEWvtkgzq7nRfGmHuA3+N8jr4M+IFfAZ2AK621D7fWa5LIMMasBNKABxrZXGKtvadBfZ2TpEnGmBuB24FC4G2ctlMmMAr41Fr7hzp1lUuyA2PMSOCkJjaPAyYA71hrJ9XZR7kkjTLG/A34A877/DrOuakvcAIQC5xrrZ1Wp75ySVqNMeYZYDLOd8C3gFLgF8BgGvQpGGPSgM3urhtw8iqXRtr4dZ5/PZCF0zc6GwgABwEHu8c60lr7VbOCtdZqacECTAcszheluuX3ueX/9DpGLRHPicOBfoABxrt5MK2Juh3dE0QlMKZOeYL7H9sCv2qwTy+cD61NQK865enAUnefg7z+O2gJSy5NwGmU+BqUd8HpqLbAL5VPWpqZTwlNlN/pvs+PKpe07M7ifs59BCwD7nbf44v3NC9wGrLW3Z7e4Lk2uc/Xq7Vel5aI5c9KYGUz6+qcpGVn+XGa+35+CHRoZHuccknLnizAV+77fIJySUsz8qULEATWA9kNth3uvs/LlUtaIrHg/OhmgeVAZp3yOOA1d9v5dcr9wESgq/v4Fhpp4zc4xnVAt0bKb3D3nd/ceDWtRwsYY/YBjsJpXD/SYPPNOL8QTDbGJEc4NPGQtfZTa+0S6/5v3IVTcX5hesFa+32d56gAbnQfXtpgnwuBeOBha+3KOvtsBu5yH05pYfjShlhrP7HWvmWtDTUoXw/80304vs4m5ZM0yc2DxrzorvvVKVMuSXP8FudHtAtw2jyNaUle1Dy+061Xs89KnPZWvHtM2XvonCSNMs4UZ38DyoCzrLXbGtax1lbXeahckt1ijBmKc6VZPvBOnU3KJWlKT5ypc7+x1m6su8Fa+ymwDSd3aiiXpDWd4q7vtdYW1hS6n403uQ+vrFNeZa19z1q7rrkHsNb+zVq7tpFNfwPKgaGmiWkkG1LndMtMcNcfNNJ5tA2YCSThfJiJNKYmh95vZNsMnIb2wcaY+Gbu816DOhK9ar5oBeqUKZ+kJY531/PqlCmXZKeMMYOAvwIPWmtn7KRqS/JCubT3iDfGnGOMucGd3/DwJubW1DlJmnIw0Bt4F9jszvF6nZtPjc3rqlyS3fUbd/2UtbbunNPKJWnKEpypEPY3xmTW3WCMORTogHPlWQ3lkrSmLu56eSPbasr2dafzCDfL9v6KHebsb4w6p1tmgLte3MT2Je66fwRikfapyRyy1gaAFThzUu3TzH3W4YxeyzHGJIU3VGkrjDGxwLnuw7oNEuWT7JIx5lpjzC3GmPuNMV/gzNE5D6ejsYZySZrknoOexZle6IZdVN+tvHCvNuuOM+dwYyM21LaKLl1wculOnLmnPwGWGGMOa1BP5yRpyn7uegPOfRXexvk8ewCYZYz53BhTd4SickmazRiTCJwDhICGc60ql6RR1toinGkOOgMLjTFPGGP+Yox5EfgAZwqi39TZRbkkralmtHTvRrbVzamBrXDs03B+jPnaWrulOTuoc7plUt311ia215S3xi8QEh1akkPN3Se1ie3S/v0VGAq8a62dXqdc+STNcS3O1FNXA4fg/MBxlLW2oE4d5ZLszJ9xbjJ2vrW2fBd1dzcv1LbaezwNHIHTQZ0MDAMex5lH8z1jzIg6dXVOkqZku+spQCLODZ464LSTpgOHAi/Vqa9ckt1xOk4uvGfr3DTapVySJllrH8CZTiEWuAS4Hqejbg0wtcF0H8olaU1vu+trjDGdagrdwSa31qmXHs6DGmN649ywM4Bzk/NmUed06zDuujlzD4s0piU5pLyLYsaY3+Kc3H/GuePubu3urpVPezFrbRdrrcHpEDoF5xfzOcaYfXfjaZRLeyljzP44o6Xvtc296/YuntJd725eKI/aOWvtre69FTZYa8ustT9aa6fg3FQ8EecGPM2lc9Leq2YaGAOcaq392FpbYq1dAJwM5AGHNTHFR2OUS1LXr9314y3YV7m0FzPG/AF4GZgK9MH5EXY0zjQK/zHG/H13ns5dK5ekJV7AmeqlD9tH8j8AzAWOZftVic2adqM5jDHZ7jGzgKustbOau686p1tmV79GdWxQT6ShluRQc/cp3oO4pA0yxlwOPAgsBA53LxmrS/kkzeZ2CL2Gc2PfDOCZOpuVS7KDOtN5LGb7DVR2ZXfzYlf1dzVSSNq/mhv+HlqnTOckaUrNTVOXW2v/V3eDe2VHzRVm+7tr5ZI0izFmMM6c5nk4c5o3pFySRhljxuPcCO5Na+011trl7o+wP+D8aJYP/N4YUzOlgnJJWo17f7wTcK6eXY8zwO1CnHPbIcAmt+rGRp9gN7kd05/gTD1zlbX20d3ZX53TLbPIXTc172E/d93UnNQiTeaQ2wnQG+cyiOXN3Kcrzq+yedbasvCGKl4yxlwNPAz8iNMxvb6Rason2W3W2lU4P3gMqXPTFuWSNCYF5/0dBFQYY2zNgjNVDMD/uWUPuI93Ky+staU4X9pS3O0NqW0V/Wq+HCXXKdM5SZpS8z43NZdlTed1YoP6yiXZlaZuhFhDuSRNmeSuP224wX1vv8XpgxvlFiuXpFVZawPW2nuttSOttYnW2o7W2mNwvgOOBMqBBXt6HDf3PgMGA5dbax/a3edQ53TL1JxsjjLG1PsbGmM6AGNx3uSvIx2YtBufuOtjGtl2KJAEzLLWVjZzn4kN6kgUMMZcB9yPc+nN4Q3mKKtL+SQt1c1d13z5Ui5JYyqBp5pY5rh1vnQf10z50ZK8UC7t3WqmX6j7JVznJGnKDJxOm37GGH8j24e665XuWrkku2SMScAZXRjC+UxrjHJJmhLvrrOa2F5TXuWulUvilclAAvCitbZ6T57IGJMDfI5zY8Upuztiupa1VksLFpxLxSxwZYPy+9zyf3odoxZP82O8mwfTmtjeESjA+cI/pk55AjDL3fdXDfbpDVTgXH7Rq055OrDU3ecgr1+7lrDl0E3ue/o90GkXdZVPWprKjYFAl0bKfcCd7vs8U7mkpaULzvzAFrh4T/MC5zJq625Pr1Pey32eirrPpaX9LcCQxj7TgJ44cx9a4IY65TonadlZPk1z3887GpQfidO5uAVIUy5p2Y2cmuy+r2/tpI5ySUtTuXG6+16uB7o32DbRPS+VAxnKJS2RWICOjZTtBxQB24B9drLvLTTSxm9QpwewDGeg0wV7Eqtxn1B2kzGmD84JIxt4A/gJOAA4HOeS04OttZuafgaJNsaYk4CT3IddgKNxRv984ZYVWmuvbVD/ZZwPlxdwThAn4MzR8zJwum3wH9QYc+NLdkoAAAKzSURBVCXwEM6H0X9xfnU9FcjBuUnVtUi7Z4w5D+cmGkGcO902NsfqSmvt1Dr7KJ9kB+60MHfjjDBbhvNedwYOw7kh4nrgCGvtwjr7KJek2Ywxt+BM7XGJtfbJBtt2Oy+MMfcC1+DMh/cy4AfOwJkf/Upr7cOt9mKk1bn5cj3OVYgrcL4Y9QGOw/ky/i5wsrW2qs4+OidJo9z5LWcCfXHa29/i/NBxMs4X6rOstS/Vqa9ckp0yxnyBMxfrCdbat3ZST7kkO3Cvqp8O/ALn8+01nLb2IJwpPwxwtbX2wTr7KJek1RhjvsH5QeRHnJwcgnMzxErgFGvt9Ab1r8cZ3ATOtB8jcPo9a26e+GXd9r4xZgXOIJLZwNtNhDHVWrtyl8F63ZPfnhcgF3gaWIdzQliFc9OynY5y1BKdC9t/WWpqWdnIPmNxvohtdk8a84HfATE7Oc7xOJdNbANKge+A87x+/VoimksW+Ez5pKUZuTQUeARnaphCnEugt7rv8y1NfV4pl7Q0d2EXoypakhfAeW69Une/z4FJXr9WLWHJl8OA54GfcUa1VuOMGvsQOBecgTON7Kdzkpam3udOOFeursD5PrYJZ+DQgcolLbuZS4Pcz7M1O8uHOvWVS1oae4/jgKtxpngtdtveG3E67o5SLmmJ5AL8P5yO4y04HdIrcG5A3auJ+p+x8z6IqQ3q76rPwgLjmxOrRk6LiIiIiIiIiIiISMTphogiIiIiIiIiIiIiEnHqnBYRERERERERERGRiFPntIiIiIiIiIiIiIhEnDqnRURERERERERERCTi1DktIiIiIiIiIiIiIhGnzmkRERERERERERERiTh1TouIiIiIiIiIiIhIxKlzWkREREREREREREQiTp3TIiIiIiIiIiIiIhJx6pwWERERERERERERkYj7/0SgTDjm/YfaAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 1728x576 with 3 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Plotting the predicted fitness effect for the variants seen more than once\n",
    "plt.close('all')\n",
    "f, (ax1, ax2) = plt.subplots(2, 1, sharey=True, sharex = True, figsize=(24, 8))\n",
    "gs = matplotlib.gridspec.GridSpec(3, 1, width_ratios=[1], height_ratios=[10,0.5, 0.5])\n",
    "ax1 = plt.subplot(gs[0])\n",
    "ax2 = plt.subplot(gs[1])\n",
    "ax3 = plt.subplot(gs[2])\n",
    "gs.update(hspace=0.05)\n",
    "\n",
    "nonsense_list = []\n",
    "y_list = []\n",
    "x_list=[]\n",
    "y_list_top20 = []\n",
    "x_list_top20=[]\n",
    "y_list_nonsense = []\n",
    "x_list_nonsense = []\n",
    "x=[]\n",
    "y=[]\n",
    "OBS_EXP_DICT={}\n",
    "top20_lower_err = []\n",
    "top20_upper_err = []\n",
    "missense_lower_err = []\n",
    "missense_upper_err = []\n",
    "nonsense_lower_err = []\n",
    "nonsense_upper_err = []\n",
    "\n",
    "top20 = c1\n",
    "missense = c6\n",
    "nonsense = c10\n",
    "synonymous = c3\n",
    "\n",
    "axisfont=20\n",
    "domainfont = 12\n",
    "titlefont=12\n",
    "axislabelfont=12\n",
    "\n",
    "top_20_DNMT3A = ['p.R882C', 'p.R729W', 'p.R326C', 'p.R320*', 'p.R882H', 'p.R736H', 'p.Y735C',\\\n",
    "                'p.R736C', 'p.W860R', 'p.R771*', 'p.R598*', 'p.P904L']\n",
    "\n",
    "for k, v in DNMT3A_missense_s.items():\n",
    "    if k in top_20_DNMT3A:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_top20.append(codon_number)\n",
    "        y_list_top20.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        top20_lower_err.append(error_lower)\n",
    "        top20_upper_err.append(error_upper)\n",
    "\n",
    "    if k not in top_20_DNMT3A:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list.append(codon_number)\n",
    "        y_list.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        missense_lower_err.append(error_lower)\n",
    "        missense_upper_err.append(error_upper)\n",
    "\n",
    "    missense_err = [tuple(missense_lower_err), tuple(missense_upper_err)]   \n",
    "    \n",
    "for k, v in DNMT3A_nonsense_s.items():\n",
    "    if k in top_20_DNMT3A:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_top20.append(codon_number)\n",
    "        y_list_top20.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        top20_lower_err.append(error_lower)\n",
    "        top20_upper_err.append(error_upper)\n",
    "    \n",
    "    top20_err = [tuple(top20_lower_err), tuple(top20_upper_err)]\n",
    "        \n",
    "    if k not in top_20_DNMT3A:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_nonsense.append(codon_number)\n",
    "        y_list_nonsense.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        nonsense_lower_err.append(error_lower)\n",
    "        nonsense_upper_err.append(error_upper)\n",
    "\n",
    "    nonsense_err = [tuple(nonsense_lower_err), tuple(nonsense_upper_err)]        \n",
    "    \n",
    "msize = 12\n",
    "medgewidth = 2\n",
    "linew = 3\n",
    "csize = 4\n",
    "cthick = 3\n",
    "\n",
    "markers1, caps1, bars1 = ax1.errorbar(x_list, y_list, yerr = missense_err, fmt = 'o', ecolor = missense, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = missense, markerfacecolor = missense, zorder = 1)    \n",
    "\n",
    "markers2, caps2, bars2 = ax1.errorbar(x_list_nonsense, y_list_nonsense, yerr = nonsense_err, fmt = 'o', ecolor = nonsense, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = nonsense, markerfacecolor = nonsense, zorder = 2)    \n",
    "\n",
    "markers3, caps3, bars3 = ax1.errorbar(x_list_top20, y_list_top20, yerr = top20_err, fmt = 'o', ecolor = top20, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = top20, markerfacecolor = top20, zorder = 3)    \n",
    "\n",
    "# loop through bars and caps and set the alpha value\n",
    "a = 0.7\n",
    "[bar.set_alpha(a) for bar in bars1]\n",
    "[cap.set_alpha(a) for cap in caps1]\n",
    "[bar.set_alpha(a) for bar in bars2]\n",
    "[cap.set_alpha(a) for cap in caps2]\n",
    "[bar.set_alpha(a) for bar in bars3]\n",
    "[cap.set_alpha(a) for cap in caps3]\n",
    "\n",
    "#Only show the required axis lines\n",
    "ax1.spines['top'].set_visible(False)\n",
    "ax1.spines['right'].set_visible(False)\n",
    "\n",
    "for axis in ['bottom','left']:\n",
    "    ax1.spines[axis].set_linewidth(2)\n",
    "\n",
    "for axis in ['bottom','left']:\n",
    "    ax1.spines[axis].set_color(grey3)\n",
    "    \n",
    "ax1.set_ylabel('fitness effect (s)', fontsize = axisfont, labelpad = 10)\n",
    "ax1.set_title('Fitness effects across DNMT3A', fontsize = 18)\n",
    "\n",
    "# ax1.set_yscale('log')\n",
    "\n",
    "# STACKED BAR CHART\n",
    "# define the exons + create data\n",
    "exons = [{'2':(24-1)+1, '3': (59-25)+1, '4': (150-60)+1, '5':(164-151)+1, '6':(213-165)+1, '7': (285-214)+1, '8': (338-286)+1,\\\n",
    "          '9': (374-339)+1, '10': (427-375)+1, '11': (477-428)+1, '12': (492-478)+1, '13': (518-493)+1, '14':(556-519)+1, \\\n",
    "          '15': (617-557)+1, '16': (646-618)+1, '17': (694-647)+1, '18': (725-695)+1, '19': (774-726)+1, '20': (803-775)+1, \\\n",
    "          '21': (826-804)+1, '22': (866-827)+1, '23': (912-867)+1}]\n",
    "exon_orders = [['2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23']]\n",
    "colors2 = [c4, c0, c3, c1, c5, c6,c7, c8, c9, c10, c2,c4, c0, c3, c1, c5, c6,c7, c8, c9, c10, c2]\n",
    "names = exons[0].keys()\n",
    "values = np.array([[data[name] for name in order] for data,order in zip(exons, exon_orders)])\n",
    "lefts = np.insert(np.cumsum(values, axis=1),0,0, axis=1)[:, :-1]\n",
    "orders = np.array(exon_orders)\n",
    "\n",
    "#plot the bar chart\n",
    "for name, color in zip(names, colors2):\n",
    "    idx = np.where(orders == name)\n",
    "    value = values[idx]\n",
    "    left = lefts[idx]\n",
    "    ax2.bar(x=left, height=-10, width=value, bottom=-5, color=color, orientation = 'horizontal', label = name, edgecolor='k', linewidth=1, tick_label = name)\n",
    "    ax2.text(left + (value/2.), (-0.26), name, ha='center', va='center', fontsize = axislabelfont)\n",
    "    \n",
    "\n",
    "# STACKED BAR CHART of domains    \n",
    "domains = [{' ': (291-1)+1, 'PWWP' :(350-292)+1, '  ':(481-351)+1, 'ADD': (614-482)+1, '   ': (633-615)+1, 'MTase': (912-634)+1}]\n",
    "domain_orders = [[' ','PWWP','  ','ADD','   ','MTase']]\n",
    "# blank_col = c13\n",
    "blank_col = grey1\n",
    "colors3 = [blank_col, PWWP_col, blank_col, ADD_col, blank_col, MTase_col]\n",
    "names_domains = domains[0].keys()\n",
    "values = np.array([[data[name] for name in order] for data,order in zip(domains, domain_orders)])\n",
    "lefts = np.insert(np.cumsum(values, axis=1),0,0, axis=1)[:, :-1]\n",
    "orders = np.array(domain_orders)\n",
    "\n",
    "#plot the bar chart\n",
    "for name, color in zip(names_domains, colors3):\n",
    "    idx = np.where(orders == name)\n",
    "    value = values[idx]\n",
    "    left = lefts[idx]\n",
    "    ax3.bar(x=left, height=-10, width=value, bottom=-5, color=color, orientation = 'horizontal', alpha = 0.9, label = name, edgecolor='k', linewidth=1)\n",
    "    ax3.text(left + (value/2.), (-0.25), name, ha='center', va='center', fontsize = domainfont, color = 'k', fontweight = 530)\n",
    "\n",
    "    \n",
    "\n",
    "# CONFIGURING THE GRAPH APPEARANCE\n",
    "#Set the x and y axis limits\n",
    "ax1.set_xlim(0, 912)\n",
    "ax1.set_ylim(170, 200)\n",
    "ax2.set_xlim(0, 912)\n",
    "ax2.set_ylim(-0.5, 0)\n",
    "ax3.set_xlim(0, 912)\n",
    "ax3.set_ylim(-0.5, 0)\n",
    "\n",
    "#x-axis ticks\n",
    "x1_major_ticks = []\n",
    "x1_major_tick_labels = []\n",
    "x2_major_ticks = [0, 100, 200, 300, 400, 500, 600, 700, 800, 912]\n",
    "x2_major_tick_labels = [\"0\", \"100\", \"200\", \"300\", \"400\", \"500\", \"600\", \"700\", \"800\", \"912\"]\n",
    "ax3.set_xticks(x2_major_ticks)\n",
    "ax3.set_xticklabels(x2_major_tick_labels, fontsize = axisfont)\n",
    "ax1.set_yticklabels([])\n",
    "ax1.set_yticks([])\n",
    "ax1.set_xticklabels([])\n",
    "ax1.set_xticks([])\n",
    "ax2.set_yticklabels([])\n",
    "ax2.set_yticks([])\n",
    "ax2.set_xticklabels([])\n",
    "ax2.set_xticks([])\n",
    "ax3.set_yticklabels([])\n",
    "ax3.set_yticks([])\n",
    "\n",
    "#y-axis ticks\n",
    "y1_major_ticks = [170, 175, 180, 185, 190, 195, 200]\n",
    "y1_major_tick_labels = [\"170\", \"175\", \"180\", \"185\", \"190\", \"195\", \"200\"]\n",
    "ax1.set_yticks(y1_major_ticks)\n",
    "ax1.set_yticklabels(y1_major_tick_labels, fontsize = axisfont)\n",
    "\n",
    "#set panel legends\n",
    "legend_marker_size = 2\n",
    "legend_elements = [Line2D([0], [0], marker = '_', color = missense, alpha = 1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='missense'),\n",
    "                   Line2D([0], [0], marker = '_', color=nonsense, alpha=1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='nonsense'),\n",
    "                  Line2D([0], [0], marker = '_', color=top20, alpha=1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='top 20 observed variant')]\n",
    "    \n",
    "ax1.legend(ncol=1, handles=legend_elements, loc='upper left', frameon=0, fontsize = 20)\n",
    "\n",
    "ax1.grid(which='major', linestyle=':', linewidth='0.7', color=grey3, zorder = 0)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TET2 variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {
    "code_folding": [
     0,
     91,
     102,
     114
    ]
   },
   "outputs": [],
   "source": [
    "def expected_number_variants_missense_TET2(data, params):\n",
    "    s_value = params[0]\n",
    "    \n",
    "    a = 0\n",
    "    expected_number_by_variant_missense = {}\n",
    "    \n",
    "    Jaiswal_fmin = 0.056\n",
    "    Jaiswal_mean = 58\n",
    "    Jaiswal_sigma = 11.5\n",
    "    Jaiswal_total = 17182\n",
    "    \n",
    "    Genovese_fmin = 0.1194\n",
    "    Genovese_mean = 55.1\n",
    "    Genovese_total = 12380\n",
    "    Genovese_sigma = 11.95\n",
    "    \n",
    "    McKerrel_fmin = 0.008\n",
    "    McKerrel_mean = 51.88\n",
    "    McKerrel_total = 4219\n",
    "    McKerrel_sigma = 18.36\n",
    "    \n",
    "    Zink_fmin = 0.2363\n",
    "    Zink_mean = 60\n",
    "    Zink_total = 11262\n",
    "    Zink_sigma = 1.18907720e+01\n",
    "    \n",
    "    Coombs_fmin = 0.019\n",
    "    Coombs_mean = 59.82\n",
    "    Coombs_total = 5649\n",
    "    Coombs_sigma = 15.14\n",
    "    \n",
    "    Young_fmin = 0.0009\n",
    "    Young_mean = 63\n",
    "    Young_total = 89\n",
    "    Young_sigma = 6.8\n",
    "    \n",
    "    Desai_fmin = 0.0169\n",
    "    Desai_mean = 66.1\n",
    "    Desai_total = 424\n",
    "    Desai_sigma = 6 #estimated\n",
    "    \n",
    "    Acuna_fmin = 0.0015\n",
    "    Acuna_mean = 4.5\n",
    "    Acuna_total = 2014\n",
    "    \n",
    "    N = 9.40166610e+04\n",
    "\n",
    "    for i in data:\n",
    "        variant_mutation_rate = mutation_rates_variants_TET2(i)\n",
    "\n",
    "        if i in Jaiswal_missense_TET2:\n",
    "            Jaiswal_number = prevalence_averaged_over_ages(Jaiswal_fmin, s_value, N, variant_mutation_rate, Jaiswal_mean, Jaiswal_sigma, Jaiswal_total)\n",
    "        else:\n",
    "            Jaiswal_number = 0\n",
    "\n",
    "        if i in Genovese_missense_TET2:\n",
    "            Genovese_number = prevalence_averaged_over_ages(Genovese_fmin, s_value, N, variant_mutation_rate, Genovese_mean, Genovese_sigma, Genovese_total)\n",
    "        else:\n",
    "            Genovese_number = 0 \n",
    "\n",
    "        if i in Zink_missense_TET2:\n",
    "            Zink_number = prevalence_averaged_over_ages(Zink_fmin, s_value, N, variant_mutation_rate, Zink_mean, Zink_sigma, Zink_total)\n",
    "        else:\n",
    "            Zink_number = 0\n",
    "\n",
    "        if i in Coombs_missense_TET2:\n",
    "            Coombs_number = prevalence_averaged_over_ages(Coombs_fmin, s_value, N, variant_mutation_rate, Coombs_mean, Coombs_sigma, Coombs_total)\n",
    "        else:\n",
    "            Coombs_number = 0\n",
    "\n",
    "        if i in Young_missense_TET2:\n",
    "            Young_number = prevalence_averaged_over_ages(Young_fmin, s_value, N, variant_mutation_rate, Young_mean, Young_sigma, Young_total)\n",
    "        else:\n",
    "            Young_number = 0\n",
    "\n",
    "        if i in Desai_missense_TET2:\n",
    "            Desai_number = prevalence_averaged_over_ages(Desai_fmin, s_value, N, variant_mutation_rate, Desai_mean, Desai_sigma, Desai_total)\n",
    "        else:\n",
    "            Desai_number = 0\n",
    "\n",
    "        if i in Acuna_missense_TET2:\n",
    "            Acuna_number = prevalence_averaged_over_uniform_ages(Acuna_fmin, s_value, N, variant_mutation_rate, Acuna_mean, Acuna_total)\n",
    "        else:\n",
    "            Acuna_number = 0\n",
    "\n",
    "        expected_total = Jaiswal_number + Genovese_number + Zink_number + Coombs_number + Young_number + Desai_number + Acuna_number\n",
    "\n",
    "        expected_number_by_variant_missense[i] = expected_total\n",
    "    \n",
    "    return expected_number_by_variant_missense\n",
    "\n",
    "def prob_observed_given_expected_missense_TET2(params, data): #data = list of variant (e.g. p.R882H) (or all_missense_variants_TET2)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_TET2_counts[variant])\n",
    "        expected_number = float(expected_number_variants_missense_TET2(data, params)[variant])\n",
    "        square_distance = ((observed_number - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance\n",
    "\n",
    "def prob_observed_given_expected_missense_TET2_upper_error(params, data): #data = list of variant (e.g. p.R882H) (or all_missense_variants_TET2)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_TET2_counts[variant])\n",
    "        observed_number_upper_error = observed_number+np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_missense_TET2(data, params)[variant])\n",
    "        square_distance = ((observed_number_upper_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   \n",
    "\n",
    "def prob_observed_given_expected_missense_TET2_lower_error(params, data): #data = list of variant (e.g. p.R882H) (or all_missense_variants_TET2)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_TET2_counts[variant])\n",
    "        observed_number_lower_error = observed_number-np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_missense_TET2(data, params)[variant])\n",
    "        square_distance = ((observed_number_lower_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {
    "code_folding": [
     0,
     91,
     102,
     114
    ]
   },
   "outputs": [],
   "source": [
    "def expected_number_variants_nonsense_TET2(data, params):\n",
    "    s_value = params[0]\n",
    "    \n",
    "    a = 0\n",
    "    expected_number_by_variant_nonsense = {}\n",
    "    \n",
    "    Jaiswal_fmin = 0.056\n",
    "    Jaiswal_mean = 58\n",
    "    Jaiswal_sigma = 11.5\n",
    "    Jaiswal_total = 17182\n",
    "    \n",
    "    Genovese_fmin = 0.1194\n",
    "    Genovese_mean = 55.1\n",
    "    Genovese_total = 12380\n",
    "    Genovese_sigma = 11.95\n",
    "    \n",
    "    McKerrel_fmin = 0.008\n",
    "    McKerrel_mean = 51.88\n",
    "    McKerrel_total = 4219\n",
    "    McKerrel_sigma = 18.36\n",
    "    \n",
    "    Zink_fmin = 0.2363\n",
    "    Zink_mean = 60\n",
    "    Zink_total = 11262\n",
    "    Zink_sigma = 1.18907720e+01\n",
    "    \n",
    "    Coombs_fmin = 0.019\n",
    "    Coombs_mean = 59.82\n",
    "    Coombs_total = 5649\n",
    "    Coombs_sigma = 15.14\n",
    "    \n",
    "    Young_fmin = 0.0009\n",
    "    Young_mean = 63\n",
    "    Young_total = 89\n",
    "    Young_sigma = 6.8\n",
    "    \n",
    "    Desai_fmin = 0.0169\n",
    "    Desai_mean = 66.1\n",
    "    Desai_total = 424\n",
    "    Desai_sigma = 6 #estimated\n",
    "    \n",
    "    Acuna_fmin = 0.0015\n",
    "    Acuna_mean = 4.5\n",
    "    Acuna_total = 2014\n",
    "    \n",
    "    N = 9.40166610e+04\n",
    "\n",
    "    for i in data:\n",
    "        variant_mutation_rate = mutation_rates_variants_TET2(i)\n",
    "\n",
    "        if i in Jaiswal_nonsense_TET2:\n",
    "            Jaiswal_number = prevalence_averaged_over_ages(Jaiswal_fmin, s_value, N, variant_mutation_rate, Jaiswal_mean, Jaiswal_sigma, Jaiswal_total)\n",
    "        else:\n",
    "            Jaiswal_number = 0\n",
    "\n",
    "        if i in Genovese_nonsense_TET2:\n",
    "            Genovese_number = prevalence_averaged_over_ages(Genovese_fmin, s_value, N, variant_mutation_rate, Genovese_mean, Genovese_sigma, Genovese_total)\n",
    "        else:\n",
    "            Genovese_number = 0 \n",
    "\n",
    "        if i in Zink_nonsense_TET2:\n",
    "            Zink_number = prevalence_averaged_over_ages(Zink_fmin, s_value, N, variant_mutation_rate, Zink_mean, Zink_sigma, Zink_total)\n",
    "        else:\n",
    "            Zink_number = 0\n",
    "\n",
    "        if i in Coombs_nonsense_TET2:\n",
    "            Coombs_number = prevalence_averaged_over_ages(Coombs_fmin, s_value, N, variant_mutation_rate, Coombs_mean, Coombs_sigma, Coombs_total)\n",
    "        else:\n",
    "            Coombs_number = 0\n",
    "\n",
    "        if i in Young_nonsense_TET2:\n",
    "            Young_number = prevalence_averaged_over_ages(Young_fmin, s_value, N, variant_mutation_rate, Young_mean, Young_sigma, Young_total)\n",
    "        else:\n",
    "            Young_number = 0\n",
    "\n",
    "        if i in Desai_nonsense_TET2:\n",
    "            Desai_number = prevalence_averaged_over_ages(Desai_fmin, s_value, N, variant_mutation_rate, Desai_mean, Desai_sigma, Desai_total)\n",
    "        else:\n",
    "            Desai_number = 0\n",
    "\n",
    "        if i in Acuna_nonsense_TET2:\n",
    "            Acuna_number = prevalence_averaged_over_uniform_ages(Acuna_fmin, s_value, N, variant_mutation_rate, Acuna_mean, Acuna_total)\n",
    "        else:\n",
    "            Acuna_number = 0\n",
    "\n",
    "        expected_total = Jaiswal_number + Genovese_number + Zink_number + Coombs_number + Young_number + Desai_number + Acuna_number\n",
    "\n",
    "        expected_number_by_variant_nonsense[i] = expected_total\n",
    "    \n",
    "    return expected_number_by_variant_nonsense\n",
    "\n",
    "def prob_observed_given_expected_nonsense_TET2(params, data): #data = list of variant (e.g. p.R882H) (or all_nonsense_variants_TET2)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_TET2_counts[variant])\n",
    "        expected_number = float(expected_number_variants_nonsense_TET2(data, params)[variant])\n",
    "        square_distance = ((observed_number - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance\n",
    "\n",
    "def prob_observed_given_expected_nonsense_TET2_upper_error(params, data): #data = list of variant (e.g. p.R882H) (or all_nonsense_variants_TET2)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_TET2_counts[variant])\n",
    "        observed_number_upper_error = observed_number+np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_nonsense_TET2(data, params)[variant])\n",
    "        square_distance = ((observed_number_upper_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   \n",
    "\n",
    "def prob_observed_given_expected_nonsense_TET2_lower_error(params, data): #data = list of variant (e.g. p.R882H) (or all_nonsense_variants_TET2)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_TET2_counts[variant])\n",
    "        observed_number_lower_error = observed_number-np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_nonsense_TET2(data, params)[variant])\n",
    "        square_distance = ((observed_number_lower_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#CALCULATING S FOR MISSENSE VARIANTS\n",
    "TET2_missense_s = {}\n",
    "a = 0\n",
    "for var in all_missense_variants_TET2:\n",
    "    initial_guess=[0.05]\n",
    "    if var in all_studies_TET2_counts.keys():\n",
    "        observed_number = float(all_studies_TET2_counts[var])\n",
    "#         print(observed_number)\n",
    "        if observed_number >1:\n",
    "            variant_mutation_rate = mutation_rates_variants_TET2(var)\n",
    "            \n",
    "            MLE = scipy.optimize.minimize(prob_observed_given_expected_missense_TET2, initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value = MLE.x[0]\n",
    "\n",
    "            upper_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_missense_TET2_upper_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_upper = upper_error_MLE.x[0]\n",
    "\n",
    "            lower_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_missense_TET2_lower_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_lower = lower_error_MLE.x[0]\n",
    "\n",
    "#             print('a = ', a)\n",
    "            TET2_missense_s[var]=(s_value, observed_number, variant_mutation_rate, s_value_upper, s_value_lower) \n",
    "    a = a+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#CALCULATING S FOR NONSENSE VARIANTS\n",
    "TET2_nonsense_s = {}\n",
    "a = 0\n",
    "for var in all_nonsense_variants_TET2:\n",
    "    initial_guess=[0.05]\n",
    "    if var in all_studies_TET2_counts.keys():\n",
    "        observed_number = float(all_studies_TET2_counts[var])\n",
    "#         print(observed_number)\n",
    "        if observed_number >1:\n",
    "            variant_mutation_rate = mutation_rates_variants_TET2(var)\n",
    "            \n",
    "            MLE = scipy.optimize.minimize(prob_observed_given_expected_nonsense_TET2, initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value = MLE.x[0]\n",
    "\n",
    "            upper_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_nonsense_TET2_upper_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_upper = upper_error_MLE.x[0]\n",
    "\n",
    "            lower_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_nonsense_TET2_lower_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_lower = lower_error_MLE.x[0]\n",
    "\n",
    "#             print('a = ', a)\n",
    "            TET2_nonsense_s[var]=(s_value, observed_number, variant_mutation_rate, s_value_upper, s_value_lower) \n",
    "    a = a+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#Exac - retrieve the frequencies in Exac for each of the variants\n",
    "import csv\n",
    "filename=\"Data files/TET2_Exac.csv\"\n",
    "\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    Exac_TET2_results={} \n",
    "    \n",
    "    for row in read_reader:\n",
    "#         print(row)\n",
    "        if row_count>1:\n",
    "            variant = row[6]\n",
    "            allele_freq = float(row[14])\n",
    "            african_freq = float(row[15])/float(row[16])\n",
    "            east_asian_freq = float(row[18])/float(row[19])\n",
    "            european_freq = float(row[21])/float(row[22])\n",
    "            finnish_freq = float(row[24])/float(row[25])\n",
    "            latino_freq = float(row[27])/float(row[28])\n",
    "            south_asian_freq = float(row[33])/float(row[34])\n",
    "            Exac_TET2_results[variant]=(allele_freq, african_freq, east_asian_freq, european_freq, finnish_freq, latino_freq, south_asian_freq)\n",
    "\n",
    "        row_count=row_count+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#COSMIC - retrieve the frequencies in COSMIC for each of the variants\n",
    "import csv\n",
    "filename=\"Data files/TET2_COSMIC_variant_counts.csv\"\n",
    "\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    TET2_COSMIC_results={} \n",
    "    \n",
    "    for row in read_reader:\n",
    "#         print(row)\n",
    "        if row_count>1:\n",
    "            variant = row[1]\n",
    "            counts = int(row[2])\n",
    "            TET2_COSMIC_results[variant]=counts\n",
    "\n",
    "        row_count=row_count+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Create dictionary of results and save to .csv file\n",
    "TET2_s_dictionary={}\n",
    "for k, v in TET2_missense_s.items():\n",
    "#     print(k)\n",
    "    if k in Exac_TET2_results.keys():\n",
    "        exac_freq = round((Exac_TET2_results[k][0])*100, 4)\n",
    "        african_freq = round((Exac_TET2_results[k][1])*100, 4)\n",
    "        east_asian_freq = round((Exac_TET2_results[k][2])*100, 4)\n",
    "        european_freq = round((Exac_TET2_results[k][3])*100, 4)\n",
    "        finnish_freq = round((Exac_TET2_results[k][4])*100, 4)\n",
    "        latino_freq = round((Exac_TET2_results[k][5])*100, 4)\n",
    "        south_asian_freq = round((Exac_TET2_results[k][6])*100, 4)\n",
    "    if k not in Exac_TET2_results.keys():\n",
    "        exac_freq = '-'\n",
    "        african_freq = '-'\n",
    "        east_asian_freq = '-'\n",
    "        european_freq = '-'\n",
    "        finnish_freq = '-'\n",
    "        latino_freq = '-'\n",
    "        south_asian_freq = '-'\n",
    "    if k in TET2_COSMIC_results.keys():\n",
    "        COSMIC_number = TET2_COSMIC_results[k]\n",
    "    if k not in TET2_COSMIC_results.keys():\n",
    "        COSMIC_number = '-'\n",
    "    TET2_s_dictionary[k]=(round(v[0]*100, 2),round(v[4]*100, 2),round(v[3]*100, 2), v[1], v[2], COSMIC_number, exac_freq, \\\n",
    "                            african_freq, east_asian_freq, european_freq, finnish_freq, latino_freq, south_asian_freq)\n",
    "\n",
    "for k, v in TET2_nonsense_s.items():\n",
    "    if k in Exac_TET2_results.keys():\n",
    "        exac_freq = round((Exac_TET2_results[k][0])*100, 4)\n",
    "        african_freq = round((Exac_TET2_results[k][1])*100, 4)\n",
    "        east_asian_freq = round((Exac_TET2_results[k][2])*100, 4)\n",
    "        european_freq = round((Exac_TET2_results[k][3])*100, 4)\n",
    "        finnish_freq = round((Exac_TET2_results[k][4])*100, 4)\n",
    "        latino_freq = round((Exac_TET2_results[k][5])*100, 4)\n",
    "        south_asian_freq = round((Exac_TET2_results[k][6])*100, 4)\n",
    "    if k not in Exac_TET2_results.keys():\n",
    "        exac_freq = '-'\n",
    "        african_freq = '-'\n",
    "        east_asian_freq = '-'\n",
    "        european_freq = '-'\n",
    "        finnish_freq = '-'\n",
    "        latino_freq = '-'\n",
    "        south_asian_freq = '-'\n",
    "    if k in TET2_COSMIC_results.keys():\n",
    "        COSMIC_number = TET2_COSMIC_results[k]\n",
    "    if k not in TET2_COSMIC_results.keys():\n",
    "        COSMIC_number = '-'\n",
    "    TET2_s_dictionary[k]=(round(v[0]*100, 2),round(v[4]*100, 2),round(v[3]*100, 2), v[1], v[2], COSMIC_number, exac_freq, \\\n",
    "                            african_freq, east_asian_freq, european_freq, finnish_freq, latino_freq, south_asian_freq)\n",
    "\n",
    "TET2_df = pd.DataFrame.from_dict(TET2_s_dictionary, orient='index', columns=['s (%)', 's lower value (%)', 's upper value (%)','observed number', 'site-specific mutation rate', 'Number of times in COSMIC',\\\n",
    "                                                                                'Exac Overall freq (%)', 'Exac African freq (%)', 'Exac East Asian freq (%)',\\\n",
    "                                                                                'Exac European freq (%)', 'Exac Finnish freq (%)', 'Exac Latino freq (%)', 'Exac South Asian freq (%)'])\n",
    "TET2_df_sorted = TET2_df.sort_values(by = ['s (%)'], ascending = False)\n",
    "\n",
    "TET2_df_sorted.to_csv('Data files/TET2_s_across_whole_gene.csv')\n",
    "# TET2_s_dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s (%)</th>\n",
       "      <th>s lower value (%)</th>\n",
       "      <th>s upper value (%)</th>\n",
       "      <th>observed number</th>\n",
       "      <th>site-specific mutation rate</th>\n",
       "      <th>Number of times in COSMIC</th>\n",
       "      <th>Exac Overall freq (%)</th>\n",
       "      <th>Exac African freq (%)</th>\n",
       "      <th>Exac East Asian freq (%)</th>\n",
       "      <th>Exac European freq (%)</th>\n",
       "      <th>Exac Finnish freq (%)</th>\n",
       "      <th>Exac Latino freq (%)</th>\n",
       "      <th>Exac South Asian freq (%)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>p.L532*</th>\n",
       "      <td>15.02</td>\n",
       "      <td>11.63</td>\n",
       "      <td>17.16</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.175600e-10</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.Q916*</th>\n",
       "      <td>13.59</td>\n",
       "      <td>12.19</td>\n",
       "      <td>14.64</td>\n",
       "      <td>5.0</td>\n",
       "      <td>9.919690e-10</td>\n",
       "      <td>43</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.S588*</th>\n",
       "      <td>11.90</td>\n",
       "      <td>9.74</td>\n",
       "      <td>13.09</td>\n",
       "      <td>2.0</td>\n",
       "      <td>8.230880e-10</td>\n",
       "      <td>4</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.Q530*</th>\n",
       "      <td>9.61</td>\n",
       "      <td>7.38</td>\n",
       "      <td>10.68</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.964580e-09</td>\n",
       "      <td>4</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.C1135Y</th>\n",
       "      <td>9.55</td>\n",
       "      <td>7.91</td>\n",
       "      <td>10.42</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.211140e-09</td>\n",
       "      <td>3</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0087</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.Q3*</th>\n",
       "      <td>9.54</td>\n",
       "      <td>7.91</td>\n",
       "      <td>10.38</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.211140e-09</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R1516*</th>\n",
       "      <td>7.33</td>\n",
       "      <td>5.99</td>\n",
       "      <td>7.99</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.882290e-08</td>\n",
       "      <td>26</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R550*</th>\n",
       "      <td>6.82</td>\n",
       "      <td>6.01</td>\n",
       "      <td>7.37</td>\n",
       "      <td>7.0</td>\n",
       "      <td>3.269290e-08</td>\n",
       "      <td>61</td>\n",
       "      <td>0.0058</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0116</td>\n",
       "      <td>0.0075</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0087</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R1359H</th>\n",
       "      <td>6.71</td>\n",
       "      <td>5.35</td>\n",
       "      <td>7.31</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.269290e-08</td>\n",
       "      <td>6</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          s (%)  s lower value (%)  s upper value (%)  observed number  \\\n",
       "p.L532*   15.02              11.63              17.16              2.0   \n",
       "p.Q916*   13.59              12.19              14.64              5.0   \n",
       "p.S588*   11.90               9.74              13.09              2.0   \n",
       "p.Q530*    9.61               7.38              10.68              2.0   \n",
       "p.C1135Y   9.55               7.91              10.42              2.0   \n",
       "p.Q3*      9.54               7.91              10.38              2.0   \n",
       "p.R1516*   7.33               5.99               7.99              2.0   \n",
       "p.R550*    6.82               6.01               7.37              7.0   \n",
       "p.R1359H   6.71               5.35               7.31              2.0   \n",
       "\n",
       "          site-specific mutation rate Number of times in COSMIC  \\\n",
       "p.L532*                  2.175600e-10                         -   \n",
       "p.Q916*                  9.919690e-10                        43   \n",
       "p.S588*                  8.230880e-10                         4   \n",
       "p.Q530*                  1.964580e-09                         4   \n",
       "p.C1135Y                 3.211140e-09                         3   \n",
       "p.Q3*                    3.211140e-09                         -   \n",
       "p.R1516*                 1.882290e-08                        26   \n",
       "p.R550*                  3.269290e-08                        61   \n",
       "p.R1359H                 3.269290e-08                         6   \n",
       "\n",
       "         Exac Overall freq (%) Exac African freq (%) Exac East Asian freq (%)  \\\n",
       "p.L532*                      -                     -                        -   \n",
       "p.Q916*                 0.0008                     0                        0   \n",
       "p.S588*                      -                     -                        -   \n",
       "p.Q530*                 0.0008                     0                        0   \n",
       "p.C1135Y                0.0008                     0                        0   \n",
       "p.Q3*                        -                     -                        -   \n",
       "p.R1516*                     -                     -                        -   \n",
       "p.R550*                 0.0058                     0                   0.0116   \n",
       "p.R1359H                     -                     -                        -   \n",
       "\n",
       "         Exac European freq (%) Exac Finnish freq (%) Exac Latino freq (%)  \\\n",
       "p.L532*                       -                     -                    -   \n",
       "p.Q916*                       0                     0                    0   \n",
       "p.S588*                       -                     -                    -   \n",
       "p.Q530*                  0.0015                     0                    0   \n",
       "p.C1135Y                      0                     0               0.0087   \n",
       "p.Q3*                         -                     -                    -   \n",
       "p.R1516*                      -                     -                    -   \n",
       "p.R550*                  0.0075                     0               0.0087   \n",
       "p.R1359H                      -                     -                    -   \n",
       "\n",
       "         Exac South Asian freq (%)  \n",
       "p.L532*                          -  \n",
       "p.Q916*                     0.0061  \n",
       "p.S588*                          -  \n",
       "p.Q530*                          0  \n",
       "p.C1135Y                         0  \n",
       "p.Q3*                            -  \n",
       "p.R1516*                         -  \n",
       "p.R550*                          0  \n",
       "p.R1359H                         -  "
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TET2_df_sorted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABZ8AAAHvCAYAAAA7GaHXAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAIABJREFUeJzs3Xd4VFX+x/H3SSUBUiAkdIIgRRSUIgiydBSkidjAlYiru7CCrF13NSC6wloQpP5QFBXL2kBRKUoVQQkt9B5qCITQQyDl/P6YsjOZSWgBon5ezzPPZM4999zvvSd3Hv3m8L3GWouIiIiIiIiIiIiISFEKuNIBiIiIiIiIiIiIiMjvj5LPIiIiIiIiIiIiIlLklHwWERERERERERERkSKn5LOIiIiIiIiIiIiIFDkln0VERERERERERESkyCn5LCIiIiIiIiIiIiJFTslnERERkd8wY0yQMcYaY96+0rEUR8aYgcaYTcaYM87rVNnZ3s4Y84sx5oSz/b4rHauIiIiIyO+Nks8iIiIixZAxprUzKVrQq1kh+wYYY4YYY7pdzpiLG2NMB2A0sBZ4GPgzkGGMiQG+BEoAjznbF1+iGBo656LqpRj/j8gYs+cs94bn62aPP9AUej8ZY146j3HfdsbSwxjznjFmozHmpDFmnzFmjjGm45W+TiIiIiLFQdCVDkBERERECvUx8J2f9q0A1tocY0wYkOOxLQBIBN4Bvr7kERZfHZzvD1hrj7kajTGtgQjgeWvtpb4+DXHMxQ/Arkt8rD+KgUBJj8/1gGeAz4Hp+fpu8vh5OfBmAWNuBT4DNuZrHw2cAZ7w0x/gbeCQ87ibgTLAg8AsY8wz1toRZzsZERERkd8zJZ9FREREircV1toPC+tgrc26XMH8xpQHcj0Tzx7tABmXOZ7fHGOMAUpaa09c6VhcrLVfeX42xrTHkXxe7e9eMca4/p9nz1nupXRgdb59hwNZhex3l7V2br59xjrHGWqMmWCtPVroCYmIiIj8jqnshoiIiMhvWP6az8aYmkC2c/ODHmUCcvL3d5YkWGSMyTTGpBtj/s8YU9LPMSoZYyYYY3Y7ayfvdX6OydevrDFmlDFmuzEmyxhzyBiTZIx5LF+/BGPMMmPMEWepgm3GmA+NMWXP8ZxvcZY2OOo8zmpjzMMe22saYyyOchqBHtfgB2PMHhwrwgEWeV4b574ljDH/Msasd4592BjztTGmgZ84jDHmr8aYX42jdvRxY0yyMSbRuf0lYFK+Y3nOVZgx5kXjqEmd6TzWGmfC82zXIMgZ5yJjTJpzXnYaY8YaY8oUsM+dxpgFzuuW6SwVMcqVnDXGtHfG92fjqJW9ATgNDPYYo7XzOh5zjrHcGJPg51jXGWO+cP6unDbGpBpj5hpjOnn0ueDzLy7yJ56dbSeBb4FQ4OrLHpSIiIhIMaKVzyIiIiLFW3j+JC9w2lp7vID++4G+wBRgPv9LtObl69cIuN25/UOgLfAQjvIdA1ydjDHVgZ+BQGff7TgSav2BNsaYJh4ri78EbgImAMk4SiPUBVoDbzjHewCYDCwAXgBOAVWBzkAMjhIGBTLG9AfGOmMaBmQCtwATjTHVrbXPOq/Bn50xNnNeD4BUHOU2bsNRGmEYjlIJec6xQ4DZQFPgfRwlF6Kd1+VnY8zN1tqVzr4GR0mUu4ElwMvAEef53gEMxVHGoXy+Y8H/SjZMAO4H3nOOEey8tm0LuwZOJYDHgS+AacBJ4EYcta1bOOfF9UcIjDEjgKeAdcDrQBpQwxnrPwHPlc2PO8/7Hee13OkcoweO0hapwKvOY94LvOu89q6kezlgLpDrPMddOOa2iTPG74vg/C9EiJ97CSDPWlvUq+ArO98PFPG4IiIiIr8pxlp7pWMQERERkXyMoy7xvAI2f2qtvcfZLwjHSud3rLV/KajNY1zXtjygqbU2yWPbLByJ4ihr7Sln27c4EtUNrbX7PPo2xZEATrTWvuRcbXsIeMtaO6iQ8/oauBkoZ63NPcfL4dq3MrDNef7359s2FvgrUMNa60qWfgjcY60Nytf3LzhWJLe01v7k0f4kMALoaK39waM9CsdDCzdaa9s723oDU3Ek+ftZa/M8+ge4Phd0LOe2Y8B8a+15PxjSGBMAhLrmyaP9rziSundYa790tjXH8UDFH4Au1trTHv2Ndf4PgXGUr5iDYx7rWGvTPfoFAzuAcOAaa+1+Z3sojj8kNAGuttZuN8b0xJEUd8dQwDlc8PnnG8cV9/PW2pf8bHf9zhfkqLU2qoCx9+Aou1HzPOJpCPwKLLTWXqpEuoiIiMhvgspuiIiIiBRv/4fjwXmeL58E2wX4yTPx7DQXCAGqATgTyp1wrKw9Y4yJcb1wJIF3AB2d+2biSPA1M8ZUK+S4R4HSQGfn6uHzcaczvsmesTjj+QbH6ux25zmmp/twrAxelW/sIOBHoJUz2QrQB7DAE56JZ4D8nwtxFLjOGFPvfAO11uZ5/IEg0BgT5YzVVQaiqUf3Ps73ZzwTz85x/K1Eec8z8ezUBKgEvO1KPDv3Pw28huP/K1xJZFeN487GmNKFnMYFn/8F+hnfe6kD0L2oDmCMicOReD+JYxW6iIiIyB+aym6IiIiIFG9bPFfhFqHtftpcJS9ctZfrAAbHiuK/FjBOLjgeemgctZ3fAFKMMetwJEK/stZ6ruB+CcfK56+BdGPMAhxlGD49h4fa1XW+F7QiHCDuLGMUpg6O5PbBQvqUwVF24mocD7DLn6Q9H4/iWDm91hizDcd5fQ3MKCAp7MUYcw+OEhkNcJSs8BTt8fPVOOYp+Rzj2uynrbrzfZ2fbWud71c53+cCH+EoN3K/MWYZjnImn1prN3rsd1HnfwEOXqJ7CQBn8v8HHL+Dna21W8+yi4iIiMjvnpLPIiIiIn9MhZW8MPnep+CoC+1PpusHa+0YY8xXOGoq/wm4CxhojJlqrb3P2WeTMaYO0B7HKuVWwNvAEGPMn6y1O84hrj4UXEv3YhJ+AcAq4MlC+rhqAxscK58vmLX2S2fyvTOO69AB+Asw3xjT0bNmc37GmLtw1JxeCgwC9gBZOJLn3+L9LxzPd4V5pp+2cx7DmTju43xwYCegJY56088bYwZaa8c7+13w+Rc3xvGwzB+AmkA3a+38KxuRiIiISPGg5LOIiIjI709RrRrd6hwr+FxXjFpr9+IoFfJ/zlq7U3EkIl93PazPWarhW+cLY0w3YDowGMdq2IJscb5fqhWsW4BywI/nsPJ2E46yEjFnWf1c6DjW2kPAB8AHzjIkr+JYzdwF+KqQXf+MI0ncxlqb5Wo0xlxbQKztgeuAFYXFU4htznd/JTKucb57raa31q4B1gD/cZZw+RUYDoz36HOh519sOBPPP+JYOd/NWjvnCockIiIiUmyo5rOIiIjI74zzQX5ZOEpEXMw4aTjKJdxpjGmSf7txKOf8OdwYE5Zv/xwcyUdcsThLE+S3wrNPIT4FzgAvGmNK+IknyhgTcpYxCvM+jrrGfhPgznq+LlNxrAb+T/7a1fk+u0qJlMnXJ8gYE+nZ5kx4r/LX349cHA+NdP/3vPO4//LT9yPn+yv5r8951N1eBuwFHjTGxHrsHwI84Yzla2dbmfzjWmszgBSglDEmpAjOv1hwJtV/wJF4vt1aO/sKhyQiIiJSrGjls4iIiMjv01LgFmPMU8BuINda+98LGOevwE/AT8aY94GVOP4bsjrQA3gHRx3na4AfnGU31gGHnW39cayaXewcb64x5iCwyBlXGeABHMnLgkp7AGCt3WmMeQSYAKw3xnwI7MKxWrk+jgfH1cJRguJCvIFjhfBIY0x7YD5wHKiKo0TIcRylIQA+AXo6Y69ljPkGxwP0ajn7NnD2+xXH6ud/ORP1J/nfwxp3GWO+xpFwPYCjZnJ/HKU9vj1LrJ87z3euMeYDIBS4HfBJyltrfzbGvI5jRXGSMeYzYD+OObzTGHPD2eptW2tzjDEDgc+AZcaYSc5zuQe4EXjRWuta+dwPeMT5u7AVyAHaOK/LR9baM84/QlzM+V+IysaY+wrY9rNH/OfEmWD/Abgexx8jyvoZf/FZSsmIiIiI/K4p+SwiIiLy+/Q3YAyOlbClcayUPe/kszPh2xB4BugG3I+j3MNuYBqOJCjAThy1oVvjSIKG4kgCTwRGeJSGGAvc6YyvDI6HHK4ABpxLnVxr7SRjzAYcq237A5FAOrAR+CeFPyzwbGOfMcbcCjwC3AcMdW7ahyOZP8WjrzXG3A38HUeyNRHHNd6OI0Hr6rfDGPMXHDWPx+N4MOA7zv1G40jIdgBK4niQ4VfAK9ba/WeJ9UNjTCkcq7Rfx5Gw/Rp4Hj/1sK21TxhjVjqP+zSOVdu7gRk4Vsmfy/X5yhjTAcfv1NM4/l9iA9DPWvuuR9e5OP4Y0BWogCP5vANH8nuMs8/xizn/C9QIR4kPfx7A/0M4CxMI3OD8uY/zld+fcZy7iIiIyB+SuTQPkr54ztppt+N4YM11OP4J5Bkc/3TzXeBda22eR/94Cv8Pu0+ttfdcqnhFRERERERERERE5H+K88rnO3GsDkkF5uH4J5VxOP5549tAJ2PMnX4eBrMaxyqc/NZewlhFRERERERERERExENxXvncFsc/v/s23wrn8jhq51UBellrv3C2x+NY+TzFWptwueMVERERERERERERkf8JOHuXK8NaO9da+41n4tnZvh/HQ2bAUVNQRERERERERERERIqZ4lx2ozDZzvccP9sqGmP+CpTF8QCbJdba5MsWmYiIiIiIiIiIiIj89pLPxpggHE9ZB5jpp0sH58tzn/lAX2vtrrONP378eAvQv39/c3GRioiIiIiIiIiIiPxxFduyG4UYDlwLfGetneXRngkMAxoB0c5XKxwPK2wN/GiMKVnQoMaYh40xSa7PR48etStXrrTZ2dk2KSnJAjYpKcmePn3arl692h4+fNhu3brV7tmzx6amptpNmzbZY8eO2ZUrV9qcnByffdasWWMzMjLstm3b7O7du+3+/fvtxo0b7fHjx+3y5cttbm6u1z6nTp2ya9eutYcOHbLbt2+3u3btsmlpaXbDhg32xIkTNikpyebl5Xntk5mZadetW2cPHjxod+zYYXfu3GkPHDhg169fb0+ePOnuu2zZMvc+J0+etOvXr7cHDhywO3futDt27LAHDx6069ats5mZmV7ju4534sQJu2HDBpuWlmZ37dplt2/fbg8dOmTXrl1rT5065bVPbm6uXb58uT1+/LjduHGj3b9/v929e7fdtm2bzcjIsGvWrLGnT5/22icnJ8euXLnSHjt2zG7atMmmpqbaPXv22K1bt9rDhw/b1atX++yTnZ1tV65caY8ePWo3b95s9+3bZ/fu3Wu3bNlijxw5YletWmWzs7O9zv3MmTPuudyyZYvdu3ev3bdvn928ebP1nH/XPsuWLbOnT5+2ycnJ7rnMP/8rVqzwmcusrCy7Zs0ar7nMP//559I1/+np6e65TEtLs+vXr3fPv7W2wPlPSUmxKSkp9uDBgz7z7/l+4sQJn/lPT0+3a9eutZmZmV7XKy8vzy5fvrzQ+c/KyvL6PcvNzbUrVqywx44d8zv/ycnJBc7/0aNH7aZNm+y+fft85v/MmTM+879q1Sp75MgRu3nzZrt37167d+9eu3nzZq/599zHc/5d9/K+ffvspk2b3PPv7172nH/Pe9lz/j2vW1ZWVqH3cv75X7Zsmc3MzPSZf9e97Jr//HPpupfzz7+/e9n1u+Oa/7S0NJ/5z38vu+bf8172nP81a9Z4zb/r/nfNv+e9nH/+Pe+xwu5lf/Pv2sc1/4Xdy/ou13e5vsv1Xa7vcn2X67tc3+X6Ltd3ub7L9V2u73J9l/8Rv8u5gortAwf9McYMAkYBG4EW1tqMc9gnCPgJaAoMttaOKqy/x8rniw9YRERERERERERE5Mq6YhUefjMrn40xf8eReF4PtDmXxDOAtTYHeNv58U+XKDwRERERERERERER8fCbSD4bYwYDY4C1OBLP+89ziIPO9wLLboiIiIiIiIiIiIhI0Sn2yWdjzNPASGAVjsTzgQsYppnzfXuRBSYiIiIiIiIiIiIiBSrWyWdjzPM4HjC4HGhnrU0vpG9TY0yIn/a2wD+cHz+8JIGKiIiIiIiIiIiIiJegKx1AQYwxfYEXgVxgETDIGJ/a2CnW2vecP48A6hlj5gN7nG31gbbOn5+31v58KWMWEREREREREREREYdim3wGqjvfA4HBBfRZALzn/PkD4HagCdAJCAbSgP8CY6y1iy5ZpCIiIiIiIiIiIiLipdgmn621Q4Ah59H/HeCdSxWPiIiIiIiIiIiIiJy7Yl3zWURERERERERERER+m5R8FhEREREREREREZEip+SziIiIiIiIiIiIiBQ5JZ9FREREREREREREpMgp+SwiIiIiIiIiIiIiRU7JZxEREREREREREREpcko+i4iIiIiIiIiIiEiRU/JZRERERERERERERIqcks8iIiIiIiIiIiIiUuSUfJZiZciQIRhjmD9//pUORURERERERERERC6Cks8iIiIiIiIiIiIiUuSUfJZi5ZFHHmHDhg3ceOONVzoUERERERERERERuQhBVzoAEU8xMTHExMRc6TBERERERERERETkImnls1y0lJQUjDEkJCSwbds2evXqRdmyZSldujQdO3Zk7dq1ABw8eJCHH36YChUqUKJECZo0acK8efO8xiqo5vOiRYvo2rUrlStXJjQ0lPLly9OsWTOGDh3q1S8tLY0nnniC2rVrU7JkSaKioqhduzYJCQls377dJ/ZZs2bRuXNnYmJiCA0NpUaNGjz55JMcOXLEp298fDzx8fFkZmby5JNPUrVqVUJDQ6lZsyYjRozAWuuzz9dff027du2oUKECoaGhVKxYkVatWjFu3DifvhkZGTz77LPUrVuXsLAwIiMjadeuHbNnzz7rHIiIiIiIiIiIiBQ3WvksRSYlJYWmTZtSt25dEhISSElJ4auvvqJ169YsWbKEW2+9lYiICO6++24yMjL45JNP6NSpE5s3b6Zq1aoFjjtz5kxuu+02IiIi6NatG5UqVSIjI4MNGzYwbtw4EhMTAcjMzKRFixZs27aNDh060LVrV6y17Ny5k+nTp9OrVy+uuuoq97gvvvgiiYmJlClThi5duhAbG0tycjKvvfYa3333HUuWLCEiIsIrluzsbDp27Mi+ffvo1KkTQUFBTJs2jWeeeYasrCx3LAD/93//x1//+lfKly9P165diYmJ4cCBAyQnJ/Puu+8yYMAAd9+dO3fSunVrUlJSaNmyJbfeeisnT55kxowZ3HrrrUycOJGHHnqoqKZKRERERERERETkklPyWYrMggULeOmll/jnP//pbhs2bBgvvPACTZs25a677mLcuHEEBDgW3Hfo0IH777+fkSNHMnLkyALHnTRpEnl5ecyfP58GDRp4bUtPT3f//OOPP7Jt2zYGDx7sM96ZM2c4ffq0+/O8efNITEzkpptu4rvvviMqKsq97b333uOBBx4gMTHRZ5x9+/bRoEED5syZQ1hYGACJiYnUqlWLkSNH8txzzxEcHAzAxIkTCQkJYfXq1cTGxhYYN0Dfvn3ZuXMnH3/8Mffcc4+7/ciRI7Ru3ZpBgwbRrVs34uLiCrxOIiIiIiIiIiIixYmSz5dI76XZVzqEc/ZRs+AiGSc+Pp5nnnnGq61v37688MILnD59mldffdWdeAbo3bs3/fr1Y9WqVec0vivZ68lffWh//UJCQggJCXF/Hj16NOBIbHsmngESEhIYNWoUU6dO9ZsUHz16tNcxYmNj6d69O++//z6bNm3i2muvdW8LCgpyJ6MLinv16tUsWLCAXr16eSWeAaKiohg6dCg9evTgiy++8FotLSIiIiIiIiIiUpwp+SxF5vrrrycwMNCrrWLFigDUqlWL0qVLe20LDAwkLi6OPXv2FDpunz59+PLLL2natCl33303bdq0oUWLFlSuXNmrX6tWrahUqRLDhw9nxYoVdO7cmRYtWviNa8mSJQQHB/PZZ5/x2Wef+RzzzJkzHDx4kEOHDlG2bFl3e2RkJDVr1vTpX6VKFQAOHz7sFffjjz9OvXr1uPvuu2nVqhUtWrSgXLlyPrEAHD16lCFDhviMffDgQQA2bNhQ2GUSEREREREREREpVpR8liITGRnp0xYUFFTgNtf27OzCV4n37NmTGTNm8PrrrzN58mQmTpwIQKNGjXjllVfo0KEDABERESxdupTExES+/vprZs2aBThWGQ8YMIB//etf7lXIhw4dIicnx+eBhfmdOHHCK/mcf5V0/vPMzc11tz322GPExMQwbtw4Ro8ezZtvvokxhlatWvHqq6/SuHFjdywAc+bMYc6cOYXGIiIiIiIiIiIi8luh5PMlUlSlLMThtttu47bbbuPkyZP88ssvzJgxg/Hjx9OlSxdWrlzJNddcA0DlypV55513sNayfv165s6dy9ixY3nxxRfJy8tj2LBhgCMZnpeXR0ZGxiWN+/777+f+++/nyJEj/Pzzz3z11VdMnjyZW265hQ0bNhAbG+tOzI8aNYpBgwZd0nhEREREREREREQul4CzdxEpPkqWLEnbtm154403eO655zhz5gzff/+9Tz9jDPXq1WPgwIHu1cTTpk1zb2/WrBmHDx9m3bp1lyXuqKgoOnfuzKRJk0hISCAjI4NFixa5YwHcn0VERERERERERH4PlHyWYu/HH3/k1KlTPu1paWkAhIeHA7B27VpSUlLO2g/gH//4BwAPPfQQ+/bt89nn5MmTLF269KLinjlzJjk5OT7tBw4c8IqncePGtGzZki+//JLJkyf7HWvNmjXu/URERERERERERH4LVHZDir3HH3+clJQUWrduTXx8PCEhISxfvpy5c+dSrVo17rnnHgB++OEHHnvsMZo3b06dOnWIjY1lz549TJ8+nYCAAJ588kn3mO3atWP48OE8++yzXH311XTu3Jnq1atz4sQJdu7cyYIFC7j55puZOXPmBcd9zz33UKJECW6++Wbi4+Ox1rJo0SKWLVtGo0aNaN++vbvvRx99RNu2bXnwwQcZPXo0TZs2JSoqij179pCcnMzatWtZsmQJsbGxF34hRURERERERERELiMln6XYe+655/jqq69ISkrihx9+ICAggKpVq/Lcc88xePBgoqOjAbjlllsYPHgwCxcuZPr06Rw7dowKFSrQoUMHd1La09NPP02LFi0YPXo0P/30E9OnTycyMpJKlSrx8MMP07t374uKe/jw4cyaNYsVK1bw3XffUaJECapVq8aIESPo37+/++GH4KhVvXz5ct566y2++OILpk6dSm5uLuXLl+eaa65h4MCBXHfddRcVj4iIiIiIiIiIyOVkrLVXOoZiZfz48Ragf//+VzoUERERERERERERkYtlrtSBVfNZRERERERERERERIqcks8iIiIiIiIiIiIiUuSUfBYRERERERERERGRIqfks4iIiIiIiIiIiIgUOSWfRURERERERERERKTIKfksIiIiIiIiIiIiIkVOyWcRERERERERERERKXJKPouIiIiIiIiIiIhIkVPyWURERERERERERESKnJLPIiIiIiIiIiIiIlLklHwWERERERERERERkSKn5LOIiIiIiIiIiIiIFDkln0VERERERERERESkyCn5LCIiIiIiIiIiIiJFTslnERERERERERERESlySj6LiIiIiIiIiIiISJFT8llEREREREREREREipySz3LRUlJSMMaQkJBASkoK99xzDzExMZQoUYLGjRszY8YMn31Onz7N8OHDqV+/PuHh4URERNCyZUv++9//Fsn4Z86cYfTo0TRs2JDo6GjCw8OJj4+ne/fu/PDDDz79N27cSEJCAlWqVCE0NJS4uDh69+7Npk2bfPomJCRgjCElJYWJEydy3XXXUaJECeLi4nj44Yc5evSozz7Jycnce++9xMfHExoaSrly5WjYsCGDBw8mOzvbq29OTg7jxo2jWbNmREREEB4ezg033MCYMWPIy8srdC5ERERERERERESKC2OtvdIxFCvjx4+3AP3797/SofxmpKSkUL16dVq3bs26deu46qqruOmmm8jIyODTTz8lOzubH374gTZt2gCOxHDHjh1ZsGABderUoUuXLmRmZvL5559z4MABnn32Wf79739f8PgAvXv35uOPP+baa6+lbdu2hIWFsW/fPn766Sd69uzJa6+95u47c+ZMevbsSXZ2Nl27dqVmzZrs2bOHL7/8ktDQUObNm0fDhg3d/RMSEpgyZQp33nkns2bNomvXrsTFxTFv3jxWrlxJmzZtmDt3rrt/cnIyTZs2xRhDt27dqF69OseOHWPr1q3MmzePjIwMSpUqBeCOYdasWdSuXZvWrVtTokQJ5s2bR3JyMvfddx8ffPDBJZtLERERERERERH53TFX7MjWWr08XuPGjbPjxo2zcu527NhhAQvYIUOGeG2bOXOmBWynTp3cbf/+97/dbdnZ2e72tLQ0W61aNQvYxYsXX/D4R44cscYY26hRI5uTk+MTb3p6uvvnjIwMGxUVZcuWLWvXrVvn1W/t2rW2ZMmS9oYbbvBq79u3rwVslSpV7M6dO93t2dnZtmXLlhawv/zyi7v9scces4CdNm2aTywZGRk2NzfX/TkxMdEC9pFHHvGKPScnx/br16/AcURERERERERERApwxXKtQVck4/0HYHqMvtIhnDM7bVCRjFOtWjX+9a9/ebXdcsstVK1alV9//dXdNnnyZIwxvPHGGwQF/e9XMDY2lueff56//OUvvP322zRv3vyCxjfGYK0lNDSUgADfyjJly5Z1//z+++9z5MgRxowZwzXXXOPVr169ejz00EO8+eabrF+/3mf7Cy+8QNWqVd2fg4KCeOCBB1i0aBG//vorN954o1f/sLAwn1iio6PdP+fl5TFmzBjKly/PyJEjCQwMdG8LDAzk9ddf591332Xq1Kl0797dZywREREREREREZHiRMlnKTLXX3+9V8LUpUqVKixZsgSA48ePs3XrVipVqkSdOnV8+rZt2xaAlStXXtD4ABEREXTt2pVvvvmG66+/njvuuIOWLVtnlJFzAAAgAElEQVTStGlTwsPDvfZ17bd69WqGDBniM/bmzZsB2LBhg0/yuXHjxn5jATh8+LC77e6772bUqFH06NGDXr160b59e1q0aEGNGjV8jnXo0CGuvvpqXnrpJZ+xwZHA3rBhg99tIiIiIiIiIiIixYmSz1JkoqKi/LYHBQW5H5TnehhfhQoV/PZ1tR85cuSCxnf59NNPGTFiBB999BGJiYkAlChRgl69evHaa68RFxcHwKFDhwCYNGlSoed24sSJc4rHtZI7NzfX3XbjjTeyaNEiXn75ZT7//HN3zebatWuTmJjIvffe6xXLli1bGDp06HnFIiIiIiIiIiIiUtwo+XyJFFUpi9+byMhIAPbv3+93e2pqqle/CxUWFsaQIUMYMmQIu3fvZuHChbz33nt8+OGHpKSksGjRIq/jrF69mvr161/UMQtz0003MWPGDE6fPs3y5cuZOXMmb731Fr1796ZcuXK0b9/eHcvtt9/Ol19+ecliERERERERERERuRx8C+KKXEKlS5emRo0a7N27ly1btvhsnzdvHgANGzYssmNWqVKFPn36MGvWLK6++mp++ukn9yrjZs2aAbiT0ZdaaGgozZs358UXX2T0aEdd8OnTpwNQp04doqKiWLp0KdnZ2ZclHhERERERERERkUtFyWe57Pr164e1lieffNKrPEV6ejrDhg1z97lQBw8e5JdffvFpP3nyJMePHycoKIiQkBAAHnjgAaKiohg6dKjXQwtd8vLymD9//gXHAo7EtqvciKe0tDQAdx3qoKAgBg4cSGpqKoMGDeLUqVM++6SmprJ+/fqLikdERERERERERORyUNkNueyeeOIJvv/+e6ZPn06DBg3o3LkzmZmZfPbZZxw4cICnnnqKm2+++YLH37t3L82aNaNu3bo0bNiQKlWqcOzYMWbMmMH+/fsZNGgQpUuXBqBs2bJ8/vnn3H777TRr1ox27dpRr149AgIC2LVrF0uWLOHQoUNkZWVdcDyvv/46s2fPpnXr1lx11VWUKlWKdevW8f333xMdHc3DDz/s7vv888+zevVqJkyYwDfffEPbtm2pVKkSBw4cYMuWLSxevJiXX37Z5+GHIiIiIiIiIiIixY2Sz3LZhYSEMGfOHN544w0++ugj3nrrLYKCgmjQoAFvvvmm+wF8Fyo+Pp6hQ4cyf/585s2bR3p6OmXKlKF27doMHz6ce+65x6t/u3btSE5O5rXXXmPWrFksWrSIkJAQKlasSNu2bbnjjjsuKp4BAwYQHR3NL7/8wuLFi8nJyaFy5coMGDCAxx9/nGrVqrn7BgcHM23aND788EPee+89ZsyYwYkTJyhXrhzVq1dn2LBh9OnT56LiERERERERERERuRyMtfZKx1CsjB8/3gL079//SociIiIiIiIiIiIicrHMlTqwaj6LiIiIiIiIiIiISJFT8llEREREREREREREipySzyIiIiIiIiIiIiJS5JR8FhEREREREREREZEip+SziIiIiIiIiIiIiBQ5JZ9FREREREREREREpMgp+SwiIiIiIiIiIiIiRU7JZxEREREREREREREpcko+i4iIiIiIiIiIiEiRU/JZRERERERERERERIqcks8iIiIiIiIiIiIiUuSKbfLZGFPWGPMXY8xXxpitxphTxpijxpifjDEPGmP8xm6MaW6M+c4Yk2GMyTTGJBtjBhtjAi/3OYiIiIiIiIiIiIj8UQVd6QAKcScwHkgF5gG7gDigJ/A20MkYc6e11rp2MMZ0B74AsoBPgQygKzASaOEcU0REREREREREREQusWK78hnYDHQDKltr+1hrn7XW9gPqALuBO3AkogEwxkQAk4BcoLW19kFr7ZPA9cASoJcx5p7LfRIi5yMhIQFjDCkpKVc6lGJv/vz5GGMYMmTIlQ7lnLVu3RpjzJUOQ0RERERERETksii2yWdr7Vxr7TfW2rx87fuBCc6PrT029QLKAZ9Ya5M8+mcB/3J+7H/pIv5jSklJwRhDQkLClQ7FR3Z2Nl999RUPPvgg1157LREREYSHh3PdddfxwgsvcPz48QL33bNnD/369aNixYqEhoYSHx/P4MGDOXz48GU8A5GiUZzvUxERERERERH5/SrOZTcKk+18z/Foa+t8n+mn/0IgE2hujAm11p6+lMFJ8bBt2zZ69uxJyZIladOmDbfddhsnTpxg1qxZDBs2jE8//ZTFixcTExPjs1/z5s05cOAA3bt3p06dOvz666+MGjWKmTNnsnjxYsqWLXuFzkp+y95//30yMzOvdBgiIiIiIiIiIpfFby75bIwJAu53fvRMNNd2vm/Ov4+1NscYswOoB1wFbLikQUqxULp0acaOHUvfvn0pWbKku/3MmTP07NmTb7/9lqFDh/LWW2957TdgwAAOHDjA6NGjGThwoLv9scceY+TIkfzzn/9kwoQJiJyvqlWrXukQREREREREREQum2JbdqMQw4Frge+stbM82iOd70cL2M/VHnWpAvujGTJkCNWrVwdgypQpGGPcr/fee8/dLy8vjwkTJtCkSRNKlSpFyZIladKkCePHjycvL89nXGMMrVu3Zt++ffz5z38mNjaWsLAwGjVqxEcffXTO8VWqVIkBAwZ4JZ4BQkJCeO655wBH3WBP27dvZ/bs2cTHx/P3v//da9vQoUMpWbIkH3zwASdPnjznOJYvX84dd9xBbGwsoaGhVKtWjQEDBpCamlrgPnl5ebzxxhvUqVOHEiVKULlyZf7xj39w7Ngxn77Jycnce++9xMfHExoaSrly5WjYsCGDBw8mOzvbq29OTg7jxo2jWbNm7jIkN9xwA2PGjPGZC89SDZs3b+buu+8mNjaWgIAA5s+fT506dQgJCSE9Pd3vOQwfPhxjDGPHjvVq37NnD4888ghXXXUVoaGhlC1blm7durFs2TK/46SlpfHggw8SFxdHWFgY119/PVOmTCnw2vnzyiuvYIxh9OjRfrfv27ePwMBAmjRp4tX24osv0qJFC8qXL09ISAgVK1akd+/ebNjg+/ers10v8F/z+cyZM4wZM4bOnTtTrVo1QkNDKVOmDO3bt+f777/3G298fDzx8fFkZmby5JNPUrVqVUJDQ6lZsyYjRozA4zms53yfioiIiIiIiIgUtcDf0sO6jDGDgBeBjUCXIUOGnHJtGzp06ECgLPDWkCFDMvLvO3To0L8AVYC3hwwZssfP2A8PHTp04m233VYRoFatWmzYsIGYmBhWrlxJxYoVSUpKomzZsqxbt47Q0FD27t1LZmYmJ0+eZN++fYSEhLB+/XrKlSvHihUrvPbZsGEDwcHB7Nu3j5MnT3Lq1Cn27NlDaGgoa9euJS4uzmuf6OhoNm3aRGBgIPv37+f48eNkZWWxe/duwsLCSE5Opnz58ixfvty9T1RUFJs3b8YYw4EDBzh27Bhnzpxh586dhIeHk5ycTMWKFVm2bBmVKlUiKSmJyMhItm7dirWW9PR0Dh8+TE5ODikpKZQqVYrVq1e7x3cdLyIigm3bthEYGEipUqVYuXIl1157LXfccQcdO3akTp06dO3alT179hAXF0eXLl0YNWoU1lp69uzJtddey6pVq/jkk09ISkqiV69erFq1yn2cSZMmERISwsSJEzl48CBdunThqquuYunSpXz88cccPXqUNm3aeO1Trlw5kpOTCQ8PZ9euXZw+fZpjx46RlpZGYGAgGzdupGzZsqxYsYJKlSoxd+5cpk2bRnR0NH369GHPnj1kZWXx2WefMWfOHHr06EF8fDwxMTHufZKTk9m8eTObNm2icePGhISE+Mz/unXriI2Ndc/lyJEjuffee9m4cSMdO3bk5ptv5syZM3zxxRd88MEHdOrUid27d7uv7a+//srq1avZtm0bU6ZMoX379jRu3Ji0tDRmzJjBN998w1133cX69eupUKECn3zyCV27dmXDhg20adOGG2+8kVq1anHixAm++uor+vbtS2pqKuHh4axYsYKHH36Y0aNHU6pUKZo1a0bLli3ZtGkTH374IWvWrOHGG28kNzeX7du3k5uby9ixYyldujTDhg0jODiYxo0b06RJEypXrkx0dDQLFiwgLCyM+vXrs337diIiItzzcv/995OZmcngwYMpW7YsO3bsYMWKFbRr145FixbRsGFDmjdvTs2aNZk5cyaTJk0iIiKCxo0bs3r1asLDw1m1ahW33nor8+fPp27duvTs2ZO8vDxGjx7Nli1b2LlzJ7Vq1eLWW29lzZo1lChRgl27dpGVlcXx48fZv38/gYGBWGv56KOP2LFjB4888ghJSUnExMSwdu1aQkNDee2111i8eDEDBgwgLi6O4OBg3nnnHV5++WUqVqxImzZtqF27NjExMUybNo3JkydTv359wsLC3PfyqVOnGD9+PKVKlfK6Xk2bNqV8+fKUL1+e999/n71799K/f3/3vbxw4UL+/Oc/U7JkSW6++Wbq1atHjRo1mD9/PpMnT6ZEiRJUqVLFfS+HhYXx5ptvkpeXxyeffMKqVavc1zI5OZnp06dz9OhR6tSpQ3Z2NgcOHKBSpUosW7aMBg0a0K1bN7p06UKtWrVo164dx48fJy8vj0OHDnH48GH3/JcuXdrn/l+xYgWlS5dmx44d5OTkkJGRwaFDh8jLy2Pbtm1ERkZ63ZdxcXGsWrWKkiVLkpKSQnZ2NkePHuXgwYMAbNmyhaioKFauXOmOMTY2tsB7OSAggM2bN1OmTBn397JrH9f87969m6ysLE6cOEFqaipBQUH6Lj/Ld3leXp7XXPqb/7i4OFauXOk1/4cPH3b/8Wnr1q1ERUV5zX9sbCyrV6/2mf8DBw5gjGHz5s1ER0df1Hd5/nvZ9V3ub/5d+yxbtoyYmBjWr1/vnsuzfZcnJSVRpkwZNmzYQFBQkHsu889//rl0zX9AQABpaWkcO3aM06dPs2vXLvf8V6hQocD5P3jwIEeOHCE7O9tn/pOSktzvERERPvPvupdLlSrFqlWr3NfLdS8XNv+e3+XLli2jfPny7nvZ3/y77mV/8x8eHs7OnTs5c+aMz/x73suu+S/su9w1/577eM6/614+ceIE+/btIzg4mA0bNvi9lz3n3/Ne9pz/5cuXu69bmTJl2LhxY4H38po1a7zmf9myZX7n3/O73N9cuu5lwGv+/d3Lrt8d1/zru1zf5fou13e5vsv1Xa7vcn2X67tc3+UFzX+FChWGXlg2tghYa38TL+DvgAXWAeX9bF/m3N6ogP3XOrfXLew448aNs+PGjbNybnbs2GEB27dvX7/bP/roIwvYG264wR4/ftzdfuLECduoUSML2KlTp3rt45wne+edd9rc3Fx3+/bt2210dLQNDg6227Ztu6i4//a3v1nAPvPMM17tTzzxhAXsa6+95ne/v//97xY4p9+R48eP27Jly9qAgAC7cOFCr23Dhw+3gO3QoYNXe9++fS1gy5Yta1NSUtztubm5tmfPnhawL774orv9scces4CdNm2az/EzMjK8rl9iYqIF7COPPGJzcnLc7Tk5ObZfv34+47jmFrDPPvusz/h79uyxAQEBtlGjRj7bfv31VwvYnj17utuys7NtjRo1bGhoqJ0/f75X/71799qKFSva8uXL26ysLHf7Qw89ZAE7ePBgr/7Lli2zQUFBFrCJiYk+x/enY8eOFrBr1qzx2XbNNdfY4OBgm56e7m5LS0uzx44d8+m7atUqW7JkSXvrrbd6tZ/tellrbatWrazja/d/srKy7O7du336HjlyxNarV89GR0fbzMxMr23VqlWzgO3UqZPXtrS0NBsZGWkjIyPtmTNnfGIr6D4VERERERERkd+1K5bT/U3UfDbGDAZG4kggt7PWHvDTbRPQGKgFLM+3fxBQHccDCrdf2midGtx5WQ5TJFZ/dsmGnjx5MuAowVCqVCl3e8mSJRkxYgTt27fn7bffpnfv3l77BQYGMmLECAIC/lcZpnr16gwaNIihQ4fywQcfkJiYeEExff3110ycOJHKlSvz1FNPeW07etRRnSUyMtLfru72I0eOnPU406dP59ChQ9x77720bNnSa9vjjz/OhAkTmDNnDrt27fKpBfzoo49SrVo19+eAgABeffVV96rb559/3qt/WFiYz/Gjo6PdP+fl5TFmzBjKly/PyJEjCQwMdG8LDAzk9ddf591332Xq1Kl0797da5y4uDi/17pSpUq0a9eOOXPmsG7dOurVq+fe5iqL0bdvX3fbt99+y7Zt23jiiSdo1aqV11gVK1bkqaeeYvDgwfz444907tyZ7Oxspk6dSunSpcn/LzQaN25Mnz59zqv8Rt++fZk9ezZTpkzh1VdfdbcnJSWxfv16br/9dq8HScbGxvodp0GDBrRt25bZs2eTnZ1NcHCw1/aCrldBQkNDqVy5sk97ZGQk/fr14/HHH2fZsmX86U9/8ukzevRor7mPjY2le/fuvP/++2zatIlrr732nOMQERERERERESlqxT75bIx5Gked51VAB2ut/wKzMBfoA9wKfJxv25+AcGChtfb0pYpVfK1YsYKAgABat27ts61Vq1YEBgaycuVKn21Vq1Z116n11Lp1a4YOHep3n3Px888/07t3b0qWLMkXX3zhlaA9F9ZZSzd/3V5/VqxYAUDbtm19tgUFBfGnP/2JlJQUVq5c6ZN8zp+cBbjqqquoUqUKKSkpHDlyhKioKO6++25GjRpFjx496NWrF+3bt6dFixbUqFHDa9/Nmzdz6NAhrr76al566SW/8YaFhfmtZdygQQNCQ0P97pOQkMCcOXOYMmUK//nPfwBHDeNPPvmEcuXK0blzZ3ffJUuWALBz506fZDI4/nkIwIYNG+jcuTMbN24kMzOTli1b+v1jQOvWrc8r+Xz77bcTGRnJhx9+yPDhw90JeNcYCQkJPvt8++23TJgwgaSkJNLT08nJyfHanp6eToUKFbzaCrteBVm3bh2vvvoqCxcuJDU1laysLK/te/fu9dknMjKSmjVr+rRXqVIFgMOHD59XDCIiIiIiIiIiRa1YJ5+NMc/jqPG8HOhorfWp5ezhc2AEcI8x5i1rbZJzjBKAK9s2/lLGK76OHj1KmTJlCAkJ8dkWFBRETEwMBw74LmSPi4vzO1758uXd456vJUuW0KlTJwICAvj++++58cYbffq4kpwFje964F9BK6M9ucbIn5x0cbX7W0Vd2Pnv3LmTo0ePEhUVxY033siiRYt4+eWX+fzzz/nggw8AqF27NomJidx7770AHDp0CHAkeIcOLbjMz4kTJ/wesyC33347ERERfPjhh7zyyisEBgYyY8YMDh06xODBgwkK+t9XjCuGzz4rfKW9KwbX9Tvb78K5CgsL46677mLSpEnMnj2bTp06kZ2dzccff0y5cuXo1KmTV//Ro0fz6KOPEh0dTYcOHahatSrh4eEYY5g2bRqrV6/m9Gnfv2Wdb1xLly6lbdu25OTk0K5dO7p160ZERAQBAQGsWrWK6dOn+z1OVJT/Z6e6rnlubu55xSEiIiIiIiIiUtSKbfLZGNMXR+I5F1gEDPKz2jTFWvsegLX2mDHmIRxJ6PnGmE+ADKAbUNvZ/unliZ5LWsrityQyMpKMjAy/5QlycnJIT08nIiLCZ7+0tDS/4+3fv9897vlYtGgRt912GwEBAcyaNYtmzZr57Ve7dm3AsVLYH9fq3Fq1ap31mK4YXTHnl5qa6tXPU1pamjsWT/7O/6abbmLGjBmcPn2a5cuXM3PmTN566y169+5NuXLlaN++vbv/7bffzpdffnnW2D0VtsrbldB9++23mTNnDrfeeqvfkhueMU+fPp1u3bqd9biu/mf7XTgfffv2ZdKkSUyZMoVOnTq5E+WPPvqo1+9nTk4OiYmJ7gcA5P8DgmsVtz/nsire00svvcSpU6eYN2+ez78QeOWVV5g+ffp5jSciIiIiIiIiUlwEnL3LFeOquRAIDAYS/bwSPHew1k4DWgELgTuAgUA28Bhwj3XVTJAi4ypdUNAqyxtuuIG8vDwWLlzos23hwoXk5ubSsGFDn227du0iJSXFp33+/Pnucc/V3Llz6dSpE0FBQcyZM6fAxDNAmzZtAJg9ezZ5eXle244fP87ixYsJCwsrdAwXV4yumD3l5OTw008/Afg9/wULFvi0bd++nd27dxMfH+931WtoaCjNmzfnxRdfZPTo0QDuxGWdOnWIiopi6dKlZGdnnzX28+EqVzFlyhTS09P5/vvvqV+/Ptdff71XP9c1W7Ro0TmNW6dOHcLDw1m1apXflej+ruvZtGjRgquvvprp06dz9OjRAhPl6enpHDlyhObNm/sknk+cOOEuqVIUtm7dSpkyZfyWpvH3e3AhznafioiIiIiIiIhcCsU2+WytHWKtNWd5tfaz32JrbWdrbbS1Nsxae521dqS1VlmXSyA6OhpjDLt27fK7vV+/fgA8++yzZGZmutszMzN55plnAHjwwQd99svNzeXpp5/2SgDv2LGD0aNHExQUxH333XdO8c2ePZsuXbpQokQJfvzxR5o0aVJo/xo1atCxY0dSUlIYO3as17bExEROnjzJ/fffT8mSJc967B49elCmTBk+/vhjli5d6rXtzTffZPv27bRv396n3jPAqFGj2Llzp/tzXl4eTz75JHl5eTzwwAPu9kWLFvlNzLpWC4eHhwOOUgwDBw4kNTWVQYMGcerUKZ99UlNTWb9+/VnPKz/PhO748ePJzs72Wz+5e/fu1KhRg7Fjx/Ldd9/5HWvJkiXu35Pg4GD69OnD8ePHfWpEJyUlMXXq1POOFRyJ5qysLMaNG8d3331H/fr1ff6YERsbS3h4OMuXL/cqRZKdnc2jjz5KenpBpefPX3x8PBkZGSQnJ3u1v/POO8yaNatIjnG2+1RERERERERE5FIotmU35LehVKlSNG3alEWLFtGnTx9q1apFYGAg3bp1o379+vTu3Zvp06fz3//+l3r16tGjRw93zdwdO3Zw11130adPH59x69evzy+//EKjRo3o2LEjR48e5dNPP+XIkSP85z//8Xmgnj+bNm2ie/fuZGVl0blzZ6ZPn+63hEH+xOa4ceNo3rw5gwYN4scff6Ru3br88ssvzJs3j1q1avHyyy+f87WZPHkyd955J61ateLOO++katWqLF++nNmzZ1O+fHkmTpzod98WLVpw/fXXc/fddxMZGcmsWbNYvXo1jRo14qmnnnL3e/3115k9ezatW7fmqquuolSpUqxbt47vv/+e6OhoHn74YXff559/ntWrVzNhwgS++eYb2rZtS6VKlThw4ABbtmxh8eLFvPzyy1xzzTXndH6e7r//fp5//nmGDRtGUFAQvXv39ukTHBzMl19+yS233MJtt91G8+bNuf766wkPD2f37t0sW7aM7du3k5qa6k6a//vf/+bHH3/kzTffJCkpiZtvvpnU1FQ+/fRTOnfuzNdff31Bsb7wwgskJiaSnZ3ts+oZICAggEGDBjF8+HCuu+46unfvzpkzZ5g3bx4ZGRm0adOGefPmnfex/Rk8eDCzZs3i5ptv5q677iIyMpKkpCR++uknevXqxeeff37RxzjbfSoiIiIiIiIicklYa/XyeI0bN86OGzfOyrnbsmWL7dKliy1Tpow1xljAvvvuu+7tubm5duzYsbZRo0Y2LCzMhoWF2YYNG9oxY8bY3Nxcn/EA26pVK7t3717bp08fW65cORsaGmpvuOEGO3Xq1HOOa968eRY468ufXbt22YSEBFu+fHkbHBxsq1atagcNGmQPHTp03tfn119/tT169LAxMTE2ODjYVqlSxf7tb3+ze/fu9enbt29fC9ht27bZ1157zdauXduGhobaihUr2kcffdQePXrUq/+sWbNsQkKCrVu3ro2IiLDh4eG2Vq1aduDAgTYlJcVn/Ly8PPv+++/btm3b2ujoaBscHGwrVqxoW7RoYV9++WW7a9cud98dO3ZYwPbt2/es57hz504bEBBgAdulS5dC+6alpdmnn37a1qtXz4aFhdmSJUvamjVr2jvuuMN+8MEHNjs726t/amqqfeCBB2xMTIwtUaKEbdCggX333Xfd85uYmHjW+PJr166dBWxQUJDdv3+/3z7Z2dn29ddft3Xr1rUlSpSwcXFx9r777rMpKSnuedqxY4e7/7lcr1atWvn9nfvmm29s06ZNbalSpWxkZKTt0KGDXbBggX333Xd97idrra1WrZqtVq2a32MkJiZawM6bN8+r/Wz3qYiIiIiIiIj8bl2xXKuxKoPsZfz48Ragf//+VzqUPyxjDK1atbqgmr4iIiIiIiIiIiLixVypAxfbms8iIiLy/+zdeZzd1X0f/M8ZCQQS2lckwNh+vGWo5SR+TJ20YCObxqV4yYLTZvFOg/3U7hLavGogI+zWzUObxE4NNV6y8sTmSWxst3VrR9iQzUZCGxokQIAYLSDBsGhDEtKc/jFLZqQZje7oakYzer9fr/u697ece7/S4febqw9nzgEAAIDxS/gMAAAAAEDTCZ8BAAAAAGi6yWNdABzNPOQAAAAAMP4Z+QwAAAAAQNMJnwEAAAAAaDrhMwAAAAAATSd8BgAAAACg6YTPAAAAAAA0nfAZAAAAAICmEz4DAAAAANB0wmcAAAAAAJpO+AwAAAAAQNMJnwEAAAAAaDrhMwAAAAAATSd8BgAAAACg6YTPAAAAAAA0nfAZAAAAAICmEz4DAAAAANB0wmcAAAAAAJpO+AwAAAAAQNMJnwEAAAAAaDrhMwAAAAAATSd8BgAAAACg6YTPAAAAAAA0nfAZAAAAAICmEz4DAAAAANB0wmcAAAAAAJpO+AwAAAAAQNMJnwEAAAAAaDrhMwAAAAAATTd5JI1KKdOSLEkyL8mLSXbVWrc3szAAAAAAAMavEw6fSymvTPLBJG9P8uM5atR0KaUzyfeT/HmSr9daDzexTgAAAAAAxpFhw+dSyhuTfDrJ29IdOL+U5FtSdR0AACAASURBVIEkTyV5Nsm5SeYmeU2SX0jy80meLqV8Nslv11oPnprSAQAAAAA4XR03fC6l/GGSX0ryQpLbk3w1yX211gNDnH9xkn+U5H1J/kOSXyul/Eqt9d4m1gzABHT1p7913OPfvuGdo1QJAAAA0AzDLTh4ZZJ/nWRxrfWjtdZ7hwqek6TWuqXW+oVa608leUOS1Ukub165AJwJtnfu63sAAAAA49Nw0268otb64kjeuNa6Psl7SinnjqQ9AGeW/iOb+4+CNuIZAAAAxqfjjnweafDc7PcAAAAAAGB8GW7ajRNSSjmrlPLjpZTXNOP9AAAAAAAY3xoKn0sp15RS7iylzOm375VJ2pOsSvJgKeXrpZThpvMAAAAAAGACa3Tk8weTvLbW+my/ff8lyf+V5PtJ1id5V5IPNKc8AAAAAADGo0bD5x9LsrJ3o5QyI8k/TnJnrfVtSd6UZFOEzwAAAAAAZ7RGw+f5SZ7st/3mJJOTfDVJaq0vJfleklc2pToAAAAAAMalRsPnPUlm9tu+PElN8lf99h1IMv0k6wIAAAAAYBxrdGHAR5K8o5QyJd2h8y8kWV9rfabfOS9LsqtJ9QEAAAAAMA41OvL59iSvSHcIvbHn9VeOOufSJO0nXxoAAAAAAONVQ+FzrfUPk/ynJFPTPf3Gf+15JElKKVckuTjJ95tXIgAAAAAA402j026k1vrvk/z7IQ7/VZLZSfadTFEAAAAAAIxvDYfPx1NrPZTkUDPfEwAAAACA8ee4026UUs492Q9oxnsAAAAAADC+DDfn8+OllE+UUqY0+sallKWllG8m+fWRlQYAAAAAwHg1XPj83SS/neTJUsptpZS3Hm8kcynlFaWU60opf5tkdZKlsfggAAAAAMAZ57hzPtdaf7WU8rkk/zHJtT2PI6WUjUmeTPJcknOSzE3ymiTzkpQkO5N8Msnv1FoPnrryAQAAAAA4HQ274GCtdVWSK0spr0ryoSTLkrwhyd876tSnk3w9yZ8n+fNa60tNrhUAAAAAgHFi2PC5V631kSS/kSSllKlJlqR7xPOLSXbVWp88JRUCAAAAADDunHD43F+tdX+SR3oeAAAAAAAwwHALDgIAAAAAQMOEzwAAAAAANJ3wGQAAAACAphM+AwAAAADQdMJnAAAAAACaTvgMAAAAAEDTTW7k5FLKRUmer7XuPs4505PMrrV2nGxxAJx52js68/jO3Tl8pCuTJ7WkvaMzrRfNHeuyAAAAgAY1OvL58SSfGOacj/ecd9JKKT9fSvm9UspfllJ2l1JqKeVPhjj34p7jQz2+2oyaADg1Vqzbmss/+We55ON3pH3rs3lox/Np3/psLvn4Hbn8k3+WFeu2jnWJAAAAQAMaGvmcpPQ8RssNSZYm2ZtkW5LXnkCbdUnuGmT/hibWBUATffl77bn21rvTVeugx+9t35Er2+7KFz92RT74ttZRrg4AAAAYiUbD5xOxMMm+Jr3Xv0p36Lw5yeVJvn8CbdbWWtua9PkAnGIr1m09bvDcq6vWfOTzd+dl82dk2dILR6k6AAAAYKSGDZ9LKb961K43DLIvSSYluSjJryR5oAm1pdbaFzaXMpoDrgEYLTff+aNhg+deXbXmU3feJ3wGAACAceBERj7/QZLeVKAmeVfP42i96fD+JMtPurKRW1xK+edJ5ibpTPK3tdb1Y1gPAENo7+jMve07GmpzT/t2ixACAADAOHAi4fMHep5Lkq+kez7lbw5y3pH8Xdj7fHPKG5G39zz6lFJ+kOR9tdaOMakIgEGtWD+yRQRXrN8qfAYAAIDT3KS2trbjntDW1rau57F2+fLlb0ny1VrrH/Xb3/tY39bWtrmtre3AqSh0+fLlFyd5f5IH2travj7I8bOTHEjyr5P8RpLfSvKdJBcneWuSq5cvX/6Vtra2lwZ7/1LKtcuXL//CVVddtThJXv3qV2fjxo2ZN29e1qxZk8WLF2fVqlWZO3du2tvbM2XKlGzfvj379+/Pvn37smPHjpx99tl58MEHM3/+/KxevXpAm40bN+ass87Kjh07sm/fvrz44ovZtm1bpkyZkg0bNmThwoUD2syePTsPPfRQJk2alKeeeip79uzJgQMHsnXr1px77rlZv359Fi1alPvvv7+vzaxZs/Lwww+nlJJdu3Zl9+7dOXToUJ544olMnTo169evz+LFi7Ny5cosWbIkq1atysyZM7N58+bUWvPMM8/kueeey+HDh7Nly5acd955WbduXd/7937ejBkz8uijj6arqyvPPvtsOjs709XVlcceeyzTp08f0GbhwoVZs2ZNpk+fnscffzyHDx/Oc889l2eeeSZJsnnz5syaNStr167ta7NgwYKsW7cu06ZNy5YtW/LSSy/lhRdeyK5du1JKycMPP5zZs2cPaDN//vysX78+U6dOTUdHRw4ePJjdu3dn586dmTRpUjZt2pS5c+dm9erVfX/2efPmZcOGDZkyZUq2bduWAwcOZO/evXnyySczefLkvv7vbbNy5crMmzcvDz74YF9fHt3/7e3tWbBgwYC+nDNnTjZu3JjJkyf39eXR/X90X/b2f0tLS3bu3Jndu3fn4MGD6ejo6Ov/888/f8j+f/rpp/P888/npZdeOqb/V61a1fc8Y8aMY/r/yJEjeeyxx3Leeedl7dq1fX9fixYtyurVq4/b/zNmzOjrl5UrV2bRokVZu3Ztpk2bNmj/P/LII0P2/9SpU/PEE0/k0KFDx/T/nDlzBlyX8+fPzwMPPJBzzjknHR0dOXDgQPbs2ZOnnnpqQP/3b9O//3uv5b1792bHjh0566yzsnHjxkGv5f793/9a7t//999/f9/f25w5c7Jp06Yhr+UHHnhgQP+vXLly0P7vvZZ7+//ovuy9lpMM6P/BruXe/3Z6+7+rqyudnZ0D+v/oa7m3//tfy/37/9FHH83MmTMH9OXChQv7+r//tfz000/39f/qrftzT4Mjn5Pk4uld+Sdvfl1WrlyZBQsW9PX/1q1bh7yW3cvdy93L3cvdy0/NvXzWrFlZs2ZN3zW2YMGCIa/llpaWY/rfvdy93L3cvdy93L3cvdy93L381N7Lzz///DGbpaLUE5xnc6yVUt6S7gUH76i1/nID7SYn+asklyb5l7XWzx7v/Ntuu60myXXXXTfyYgE4IZ/772vziS/d23C7z374snz8n7zhFFQEAAAAE86YLabX0sjJpZSfKKXcVEpZOMTxRT3HT5tEoNZ6OMmXejYvG8taABho2etHtnDgSNsBAAAAo6eh8DnJryf5cJJdQxzfmeRD6Z764nTydM/ztDGtAoABWi+am8taFzfU5vLWJeZ7BgAAgHGg0fD5zUm+X4eYq6Nn/91JfvpkC2uyv9/z/NiYVgHAMW665tK0lBP7DaCWUnLjNW86xRUBAAAAzTC5wfMXJdk2zDk7kpw/snJGrpRyaZI1tdZDR+2/Ism/6tn8k9GuC4DjW7b0wtz+0Sty7a13p+s46xC0lJIvfuyKLFtqyg2YiK7+9LeOe/zbN7xzlCoBAACapdHweX+S+cOcMz/JwZGVM1Ap5d1J3t2zuajn+c2llD/oef1MrfXXe17/VpLWUsoP8ncB+euTXNHz+sZa6980oy4AmutDb2/NxQtm5FN33pd72rcfc/zy1iW58Zo3CZ7hDLG9c1/f6yVzzZoGAADjVRliBo3BTy5lRZJLkryy1rp3kOMzkjyS5MFa61tPurhS2pL85nFOeaLWenHPuR9K8p6e+uYlOSvdc1D/bZL/Wmv9yxP5zNtuu60myXXXXTfiugEYufaOzrz3lu/k8JGuTJ7Ukq9d/w5zPMMZpv8oaCOeAQDgpJ3YXJenQKMjn29P8qdJvldK+ee11vW9B0opS5N8Id3B7+3NKK7W2pak7QTP/XKSLzfjcwEYO60Xzc3LF84YsA0AAACMPw2Fz7XWr5VS3pHkV5OsKaXsTLI9yZIkC9Odov9hrfVPm14pAAAAAADjRkujDWqt70/ya0keTPc8zD/Z89ye5Npa6weaWSAAAAAAAONPo9NuJElqrbcnub2UMjXJrCTP11r3N7UyAAAAAADGrRGFz716AmehMwAAAAAAA4wofC6lzE/yc0lel2RarfXD/fa/PMkDtdYXm1YlAAAAAADjSsPhcynlQ0k+l+ScdC8wWJN8uOfwwiR/m+TaJF9uUo0AAAAAAIwzDS04WEp5e5Lbkzyc5D1Jbut/vNa6Id0LD767WQUCAAAAADD+NDry+d8leTLJ5bXW3aWUHx/knPVJ3nzSlQEAcMZp7+jM4zt35/CRrkye1JL2js60XjR3rMsCAABGoNHw+Y1Jvlpr3X2cc7YlWTTykgAAONOsWLc1N9/5o9zbvmPA/ks+fkcua12cm665NMuWXjhG1QEAACPR0LQbSc5Osm+Yc2YlOTKycgAAONN8+XvtubLtrmOC5173tu/IlW135St/0T7KlQEAACej0fB5S5KfHOacS5M8NKJqAAA4o6xYtzXX3np3umo97nldteYjn787K9ZtHaXKAACAk9Vo+PzNJP+wlPILgx0spXwgyeuT/PnJFgYAwMR3850/GjZ47tVVaz51532nuCIAAKBZGg2f/98kHUn+tJTytfQsLFhK+X96tm9P8kiS32tqlQAATDjtHZ1DTrUxlHvat6e9o/MUVQQAADRTQwsO1lqfK6VcnuSPkvQf/fy5nue/TPLPaq3DzQsNAMAZbsX6kU2hsWL91rReNLfJ1QAAAM3WUPicJLXWjiRvKaW8Pt0jn+cmeSHJD2ut9ze5PgAAJqjd+w+NajsAAGB0HTd8LqV8PclXa6139mxflmRLrbWj1ro+yfpRqBEAgAloxtSzR7UdAAAwuoab8/ndSV7bb/v7Sd5/yqoBAOCMsez1F45qOwAAYHQNFz4/n2RGv+1yCmsBAOAM0nrR3FzWurihNpe3LjHfMwAAjBPDzfm8Mck/LaWsTPJkz76Le6bfOK5a670nWxwAABPbTddcmivb7kpXrcOe21JKbrzmTaNQFQAA0AzDhc/Lk9yV5P/rt+99PY/hTBppUQAAnBmWLb0wt3/0ilx7693HDaBbSskXP3ZFli015QYAAIwXxw2fa63fLaW8LsnbkixJ0pbknp4HAACctA+9vTUXL5iRT915X+5p337M8ctbl+TGa94keAYAgHFmuJHPqbU+keTLSVJKaUvyg1rrzae4LgA4LVz96W8d9/i3b3jnKFUCE9uypRdm2dIL097Rmffe8p0cPtKVyZNa8rXr32GOZwAAGKeOGz6XUn47yf+qtX63Z9cHkqw55VUBwGloe+e+vtdL5k4bw0pg4mq9aG5evnDGgG0AAGB8Gm7k879M8nyS3vD5K+meB3r9qSwKAE4X/Uc29x8FbcQzAAAAHF/LMMf3Jpnab7ucwloAAAAAAJgghhv5vDnJz5ZSvpHkyZ59s0opFw33xrXWjpMtDgAAAACA8Wm48PmWJH+S5G/67ftEz+N46gm8NwAAAAAAE9RxA+Ja65+WUh5PclWSJUnen+75ntee+tIAAAAAABivhh2dXGv9YZIfJkkp5f1JvlFrvfkU1wUAAAAAwDjW6NQYH0iy5lQUAgAAAADAxNFQ+Fxr/cNTVQgAAAAAABNHS6MNSiktpZR/UUr5YSnlhVLK4X7HfryUcmsp5dXNLRMAAAAAgPGkofC5lHJ2ku8l+d0kr0yyJ0npd8rjST6Y5JeaVSAAAAAAAONPoyOfr0/y1iTLkyxM8qX+B2utzye5N8k/akp1AAAAAACMS42Gz7+U5K9rrTfXWruS1EHOeTzJRSddGQAAAAAA41ZDCw4meXmS/zHMOc8mmTOycgDg9NTe0ZnHd+7O4SNdmTypJe0dnWm9aO5YlwUAAACnrUbD5xeTzBrmnIuSPD+ycgDg9LJi3dbcfOePcm/7jgH7L/n4HbmsdXFuuubSLFt64RhVBxPH1Z/+Vt/r+x7Zecz+b9/wzlGvCQAAODmNTruxNsmVPQsPHqOUMjPd8z3fd7KFAcBY+/L32nNl213HBM+97m3fkSvb7spX/qJ9lCuDiW3JnPP6HgAAwPjV6MjnLya5I8kdpZQP9T9QSpmV5PeTzE7y35pTHgCMjRXrtubaW+9OVx1seYO/01VrPvL5u/Oy+TOMgIaTYGQzAABMPA2NfK61/mm6A+afS/J0kuuSpJSyKsmTSd6V5NZa6/9scp0AMKpuvvNHwwbPvbpqzafu9Es/AAAA0F+j026k1vqhJB9M8mCS+UlKkp9IsjnJh2qt/6KpFQLAKGvv6Bxyqo2h3NO+Pe0dnaeoIgAAABh/Gp12I0lSa/2DJH9QSjk33dNsvFBr3dfMwgBgrKxYv3XE7VovmtvkagAAAGB8GlH43KvW+mKSF5tUCwCcFnbvPzSq7QAAAGAianjaDQCY6GZMPXtU2wEAAMBEJHwGgKMse/2Fo9oOAAAAJqKTmnYDAJrl6k9/q+/1fY/sPGb/t29456jV0nrR3FzWurihRQcvb11ivmcAAADox8hnAE47S+ac1/cYKzddc2laSjmhc1tKyY3XvOkUVwQAAADji5HPAJwWRnNk84lYtvTC3P7RK3LtrXenq9Yhz2spJV/82BVZttSUGwAAANCf8BkAhvCht7fm4gUz8qk778s97duPOX5565LceM2bBM8AAAAwiKaEz6WUeUn+YZL9Sf6i1nqkGe8LAGNt2dILs2zphWnv6Mx7b/lODh/pyuRJLfna9e8wxzMAAAAcR0PhcynluiTvT/KOWuuzPft+Msn/SjKn57RVpZQraq37mlkoAIyl1ovm5uULZwzYBgAAAIbW6IKD701Se4PnHrckmZ3k95P8zyT/d5Jfa055AAAAAACMR42Gz69Ksr53o2e6jcuTfLnW+uFa69VJVib5Z80rEQAAAACA8abR8Hlukl39tn+65/kb/fb9ZZKXnUxRAAAAAACMb42Gz88mmddv+/IkXUn+pt++muSck6wLAAAAAIBxrNHweWOSq0spc0sps9I9B/TKWuvufudcnOSpJtUHAAAAAMA41Gj4/Nkk5yfZlmRrkkVJbu09WEqZlOQfJFnXrAIBAAAAABh/Jjdycq31W6WUX0tybc+uO2qtf9LvlLele8qN/92k+gAAAAAAGIcaCp+TpNZ6e5Lbhzj2v5PMPtmiAAAAAAAY3xqddgMAAAAAAIbV0MjnUsqPJ3lzuqfbeKFn37R0z/v8riT7k/xWrfWzzS4UgInr6k9/67jHv33DO0epEgAAAKBZGh35/O+SfLI3eO7xmSS/0vNec5P8dinlyibVB8AZZnvnvr4HAAAAMH41OufzG5P8oHejlHJWkvcluS/JW5LMSbImyceTfLcpFQIw4fUf2dx/FLQRzwAAADB+NTryeUGSrf2235hkepIv1FoP1Fp3JPlmktc3qT4AAAAAAMahRkc+16Pa/IOefff02/d0kvknWRcAAEBDbtl0+LjHr39to//8AQDgZDT67asjyd/vt/2uJNtqrY/127c4yXMnW9hYG2zxK7/+DQAA48Nzh2rf69lnlzGsBADgzNVo+HxnkuWllD9LciDJm5P87lHnXJLk0ZMtrJTy80kuT/KGJEvTPb3HHbXWXz5Om59KckO6A/JzkmxO8pUkv1drPdLI5/df6GrJ3GmNlg8AAIyy/iOb+4+CNuIZAGBsNPot7HeS/EySn+3ZXpvk5t6DpZQfS/KTSf5jE2q7Id2h894k25K89ngnl1LeleTP0x2Kfy3Js0mu7qn5p5P8QiMf3j9wNuIZAAAAAKAxDYXPtda9SX66lHJJz64Ha61d/U7Zn+Q9SVY1obZ/le7QeXO6R0B/f6gTSykzknwxyZEkb6m1rurZf2OSu5P8fCnlF2utX21CXQAAAAAADGNEv39Wa90wxP4tSbacRD3936svbC5l2Dnafj7dixz+UW/w3PMeB0opNyRZkeS6JMJnAAAAAIBRMKLwuZRyVpJlSV6X5Lxa66d69p+TZEaSZ44aEX2qXdHz/L8GOXZvukdk/1QpZUqt9eDolQUAAAAAcGZqabRBKeVn0j26+X8k+S9J2vodfkOSJ5O8twm1NeI1Pc8PH32g1no4yePpDtpfMZpFAQAAAACcqSa1tbWd8MmllDcm+W66RxLflO5F/f5eW1vbzUnS1ta2bfny5b+aZFZbW9v/36wily9ffnGS9yd5oK2t7euDHL8+ydwkt7S1tT07yPFfSnJhkjva2tq2DfYZpZRrly9f/oWrrrpqcZI8uGdq9u7dm7PPPjuvmbY3ixcvzqpVqzJ37ty0t7dnypQp2b59e/bv3599+/Zlx44dOfvss/Pggw9m/vz5Wb169YA2GzduzFlnnZUdO3Zk3759efHFF7Nt27ZMmTIlGzZsyMKFCwe0mT17dh566KFMmjQpTz31VPbs2ZMDBw5k69atOffcc7N+/fosWrQo999/f1+bWbNm5eGHH04pJbt27cru3btz6NChPPHEE5k6dWrWr1+fxYsXZ+XKlVmyZElWrVqVmTNnZvPmzam15plnnslzzz2Xw4cPZ8uWLTnvvPOybt26vvfv/bwZM2bk0UcfTVdXV5599tl0dnamq6srjz32WKZPnz6gzcKFC7NmzZpMnz49jz/+eA4fPpznnnsuzzzzTJJk8+bNmTVrVtauXdvXZsGCBVm3bl2mTZuWLVu25KWXXsoLL7yQXbt2pZSShx9+OLNnzx7QZv78+Vm/fn2mTp2ajo6OHDx4MLt3787OnTszadKkbNq0KXPnzs3q1av7/uzz5s3Lhg0bMmXKlGzbti0HDhzI3r178+STT2by5MnZuHFj5s2b19dm5cqVmTdvXh588MG+vjy6/9vb27NgwYIBfTlnzpxs3LgxkydP7uvLo/v/6L7s7f+Wlpbs3Lkzu3fvzsGDB9PR0dHX/+eff/6Q/f/000/n+eefz0svvXRM/69atarvecaMGcf0/5EjR/LYY4/lvPPOy9q1a/v+vhYtWpTVq1cft/9nzJjR1y8rV67MokWLsnbt2kybNm3Q/n/kkUeG7P+pU6fmiSeeyKFDh47p/zlz5mTNmjUD+v+BBx7IOeeck46Ojhw4cCB79uzJU089NaD/+7fp3/+91/LevXuzY8eOnHXWWdm4ceOg13L//u9/Lffv//vvv7/v723OnDnZtGnTkNfyAw88MKD/V65cOWj/917Lvf1/dF/2XstJBvT/YNdy7387vf3f1dWVzs7OAf1/9LXc2//9r+X+/f/oo49m5syZA/py4cKFff3f/1p++umnB/T/mjVr+q6xpw9Ozuf/+5p07j2Yzhf25pLzz83Z9VB27tyZlpaWY/p/5cqVWbBgQV//b926dchruX//j/Re/qX/vTYtLS05cPBgrnj1TPdy93L3cvdy9/JB7uULFiwY8loerXt555Gzsr79wXSeMzdnP7E2r7hgSR5ed7/v5e7l7uXu5e7l4+heLmNxL3cvb969/Pzzz18+TPx6ypRa64mfXMo3k/zDJD9Wa32qlPKbSW6qtU7qd86fJvmJWutrhnqfhoss5S3pXnDwjlrrLw9y/OEkr0ryqlrr5kGO/02SNyd5c631h8f7rNtuu60myecfnJ7DR7oyeVJLvnb9O9J60dwm/EkAGMyKdVtz850/yr3tO445dlnr4tx0zaVZtvTCMahsoKs//a2+19++4Z1jWAkAg9nwQle+vq0rm/Yc+2+c104v+dkLWnLJzIZ/+RMAYLwbdkG9U6XRb14/neSuWutTxzlna5LzR17SiLzQ8zxziOMzjjpvWO1bn81DO55P+9Znc8nH78jln/yzrFi39aSKBOBYX/5ee65su2vQ4DlJ7m3fkSvb7spX/qJ9lCsDYDz5/q6ufGbjkUGD5yTZtKfmMxuP5Ae7RnNpGgCAM1ujCw6el+SZYc6ZmhHMJX2SHkryxiSvTnJ//wOllMlJXp7kcJLHRvoBveHHFz92RT74ttaTqRWAHivWbc21t96drmF+C6er1nzk83fnZfNnjPoI6P6jne97ZOcx+42ABhh7G17oypceO5LhfqezJvniY0cyb0qMgAYAGAWNfuPanmS45PUNOYmQd4Tu7nn+mUGOXZbuQPxvaq0HT+ZDesMPI6ABmuPmO380bPDcq6vWfOrO+05xRce3ZM55fQ8ATh9f39Y1bPDcqyb5xjajnwEARkOjI5+/k+TXSin/oNb6V0cfLKW8I8lPJflPzSiuAX+W5LeS/GIp5fdqrat66jknyad7zrmtGR/UG36cDnOPAoxn7R2dQ061MZR72renvaNzVOfhN7IZ4PS2bX8dcqqNoWzcU7Ntf80FU8ds+kMAgDNCo+HzZ5L8YpLvllJ+L8nFSVJKuSrdI4w/luTJJL99soWVUt6d5N09m4t6nt9cSvmDntfP1Fp/PUlqrbtLKR9Jdwj9g1LKV5M8m+SdSV7Ts/9rJ1tTr7EIPwAmmhXrR/ZbJCvWb3X/BaDPhhdGNop5wwtduWDqpOFPBABgxBoKn2ut20spVya5M8n1/Q59K92rJj6a5GdrrcPNC30i3pDkfUfte0XPI0meSPLr/Wq7q5RyeZJPJvm5JOck2ZzkXyf5XK0n+HvdJ0j4AXBydu8/NKrtAJiYXjwyuu0AADhxjY58Tq11dSnlNUmuSvLmJHOTvJDkh0m+WWs93IzCaq1tSdoabPPXSf5xMz5/OMIPgJMzY+rZo9oOgInp3BEOXh5pOwAATlzD4XOS1FqPpHu087eaW874IfwAODnLXj+yufNH2g6AiemSmS1JGp96o7sdAACnkm9cIyT8ADg5rRfNzWWtixtqc3nrElMeATDABVNLXju9sYUDXze9WGwQAGAUjGjkcynljUnelGR2ksF+Ya3WWj91MoWdzoQfAM1x0zWX5sq2u9J1AtPyt5SSG6950yhUBcB487MXtOQzG4/kRBZ5KUnec4ExOAAAo6Gh8LmU+sH+1wAAIABJREFUMiPJ15O8Nd3f24ZSk0zI8Fn4AdA8y5ZemNs/ekWuvfXu4wbQLaXkix+7IsuW+q0TAI51ycyWfPgVyZceO34AXZJ85BWTTLkBADBKGh35fEuSK5L8ZZLfT7I1SVMWGBwPhB8Azfeht7fm4gUz8qk778s97duPOX5565LceM2b3HsBOK63LmjJ/CnJN7Z1ZeOeYyPo100vec8FLYJnAIBR1Gj4/K4kq5O8tdba+Koe45jwA+DUWbb0wixbemHaOzrz3lu+k8NHujJ5Uku+dv07THMEwAm7ZGZ3uLxtf81nHzmcrpq0lOQTr5psjmcAgDHQaPg8M8kfnwnBc+uFc4QfAKOs9aK5efnCGQO2AaBRF0wtWTClDNgGAGD0NRo+P5Jk4ako5HQj/AAAAAAAGLlGJzz7fJKrSylLTkUxAAAAAABMDI2OfP5Ouhcc/OtSyvIk9yd5frATa60dJ1kbAAAAAADjVKPh85YkNUlJ8qXjnFdH8N4AAAAAAEwQjQbEf5TuYBkAAAAAAIbUUPhca33/KaoDAAAAAIAJpNEFBwEAAAAAYFgNhc+llCOllBuHOeeTpZTDJ1cWAAAAAADjWaMjn0vP40TOAwAAAADgDHUqpt2YneTAKXhfAAAAAADGiWEXHCylXHbUrosH2Zckk5JclOSXkjzUhNoAAAAAABinhg2fk/wgSe15XZO8r+cxmJKkK8m/OenKAAAAAAAYt04kfL453aFzSXJTusPoewY570iSziTfr7VualaBAAAAAACMP8OGz7XWtt7XpZT3Jbmr1vq5U1kUAAAAAADj24mMfO5Ta335qSoEAAAAAICJo2WsCwAAAAAAYOI57sjnUsrd6VlksNa6rWf7RNRa67KTrg4AAADgOK7+9LeOe/zbN7xzlCoB4GjDTbvxlp7nqUdtD6eOpJjTyX2P7Ox73fuDzA8sAAA4fd2y6XDf60f31mP2X//ahmYdBMah7Z37+l4vmTttDCsBIBkmfK61thxveyJbMue8sS4BAAAYoTlnj3UFwGjpP1Cs/yhoA8gAxt5w0258PMkPa633jVI9p43Vv/NPx7oEAACgAUY2AwCcXoYbyfy7SX6md6OUcqSUcuOpLQkAAACgMe0dnXl85+48suP5PL5zd9o7Ose6JIAz3nBDAw4kmdJvu/Q8AAAAAMbcinVbc/OdP8q97TsG7L/k43fkstbFuemaS7Ns6YVjVB3AmW24kc+PJ/lHpZSF/faN+8UEAQAAgPHvy99rz5Vtdx0TPPe6t31Hrmy7K1/5i/ZRrgyAZPjw+QtJfiLJjlLKkZ59bT3Tbxzvcfg47wkAAABwUlas25prb707XfX4Y+S6as1HPn93VqzbOkqVAdDruNNu1Fo/V0rZleSqJIuTvDVJR5Itp740AAAAgMHdfOePhg2ee3XVmk/deZ/pNwBG2bDLQddav5rkq0lSSulK8vu11ptPdWEAAAAAg2nv6Bxyqo2h3NO+Pe0dnWm9aO4pqgqAow0bPh9leZIfnII6ADiDXf3pb/W9vu+Rncfs//YN7zxum8EM1gYAgIlhxfqRTaGxYv1W4TPAKGoofK61Lj9VhQBAkiyZc17DbbZ37vu79nOnNbMcAOAMd8um4y9pdP1rGx3TRTPs3n9oVNsBMDJ+SgIw5kYySrl/m/6joI14BgBOlecO/d38wrPPLmNYCTOmnj2q7QAYGeEzAONae0dnHt+5O4ePdGXypBbz+AEATdV/ZHP/UdBGPI+tZa8f2cKBI20HwMj4aQnAuLRi3dbcfOePjllo5pKP35HLWhfnpmsutZo5ANA02/bX7DpY01WTltK9fcFUo5/HSutFc3NZ6+KGFh28vHWJQQoAo6xlrAsAgEZ9+XvtubLtriH/sXFv+45c2XZXvvIX7aNcGQAw0Wx4oSs3tx/Ov11/ONtfTJ48kGx/Mfm36w/n5vbD2fBC11iXeMa66ZpL01JO7H8AtJSSG6950ymuCICjCZ8BGFdWrNuaa2+9O121Hve8rlrzkc/fnRXrRrYSOgDA93d15TMbj2TTnsG/d2zaU/OZjUfyg10C6LGwbOmFuf2jVwwbQLeUki9+7Aq/FQcwBoTPAIwrN9/5o2GD515dteZTd953iisCACaiDS905UuPHclw3zpqki8+dsQI6DHyobe35rtt787lrUsGPX5565J8t+3d+eDbWke5MgAScz4DMI60d3Q2NK9fktzTvt0ihABAw76+rWvY4LlXTfKNbV25ZKbxXWNh2dILs2zphWnv6Mx7b/lO30LUX7v+Hb4DAowx4TMA48aK9SObQmPF+q3+4QEAnLBt++uQU20MZeOeahHCMdZ60dy8fOGMAdsAjC3/WxaAcWP3/kOj2g4AODONdAoNU28AwEDCZwDGjRlTzx7VdgDAmenFI6PbDgAmKuEzAOPGstePbIXykbYDAM5M504a3XYAMFEJnwEYN1ovmpvLWhc31Oby1iXm+wMAGjLShQMtOAgAA/nJCMC4ctM1l6alnNhCPi2l5MZr3nSKKwIAJpoLppa8dnpjCwe+bnqx2CAAHEX4DMC4smzphbn9o1cMG0C3lJIvfuyKLFtqyg0AoHE/e0FLTjRKLknec4F/XgPA0fx0BGDc+dDbW/Pdtnfn8tYlgx6/vHVJvtv27nzwba2jXBkAMFFcMrMlH37FpGED6JLkI6+YZMoNABjE5LEuAABGYtnSC7Ns6YVp7+jMe2/5Tg4f6crkSS352vXvMMczANAUb13QkvlTkm9s68rGPfWY46+bXvKeC1oEzwAwBOEzAONa60Vz8/KFMwZsAwA0yyUzu8PlbftrPvvI4XTVpKUkn3jVZHM8A8AwhM8AAAAwjAumliyYUgZsAwDH53eDAAAAAABoOuEzAAAAAABNZ9oNAMalqz/9rb7X9z2y85j9377hnaNeEwAAo8/3QoDTl/AZgHFvyZzzxroEAABOA74XApxehM8AjEtGsAAAkPheCHA6M+czAAAAAABNJ3wGAAAAAKDphM8AAAAAADSd8BkAAAAAgKYTPgMAAAAA0HTCZwAAAAAAmk74DAAAAABA000e6wIAAADgdHXLpsN9rx/dW4/Zf/1r/bMaAIYy4X5KllK2JHnZEId31loXjWI5AAAATBBzzh7rCgBgfJlw4XOPF5L87iD79452IQAAAIxfRjYDwMhN1J+iz9da28a6CAAAAACAM5UFBwEAAAAAaLqJOvJ5Sinll5NclGRfkvVJ7q21HhnbsgAAAAAAzgwTNXxelOSPj9r3eCnlA7XWe8aiIAAAAACAM8mktra2sa6hqZYvXz4zyaeS/Lskv5nkq0kmJVmW5BeXL1/+39va2nYe3a6Ucu3y5cu/cNVVVy1Okle/+tXZuHFj5s2blzVr1mTx4sVZtWpV5s6dm/b29kyZMiXbt2/P/v37s2/fvuzYsSNnn312HnzwwcyfPz+rV68e0Gbjxo0566yzsmPHjuzbty8vvvhitm3blilTpmTDhg1ZuHDhgDazZ8/OQw89lEmTJuWpp57Knj17cuDAgWzdujXnnntu1q9fn0WLFuX+++/vazNr1qw8/PDDKaVk165d2b17dw4dOpQnnngiU6dOzfr167N48eKsXLkyS5YsyapVqzJz5sxs3rw5tdY888wzee6553L48OFs2bIl5513XtatW9f3/r2fN2PGjDz66KPp6urKs88+m87OznR1deWxxx7L9OnTB7RZuHBh1qxZk+nTp+fxxx/P4cOH89xzz+WZZ55JkmzevDmzZs3K2rVr+9osWLAg69aty7Rp07Jly5a89NJLeeGFF7Jr166UUvLwww9n9uzZA9rMnz8/69evz9SpU9PR0ZGDBw9m9+7d2blzZyZNmpRNmzZl7ty5Wb16dd+ffd68edmwYUOmTJmSbdu25cCBA9m7d2+efPLJTJ48ua//e9usXLky8+bNy4MPPtjXl0f3f3t7exYsWDCgL+fMmZONGzdm8uTJfX15dP8f3Ze9/d/S0pKdO3dm9+7dOXjwYDo6Ovr6//zzzx+y/59++uk8//zzeemll47p/1WrVvU9z5gx45j+P3LkSB577LGcd955Wbt2bd/f16JFi7J69erj9v+MGTP6+mXlypVZtGhR1q5dm2nTpg3a/4888siQ/T916tQ88cQTOXTo0DH9P2fOnAHX5fz58/PAAw/knHPOSUdHRw4cOJA9e/bkqaeeGtD//dv07//ea3nv3r3ZsWNHzjrrrGzcuHHQa7l///e/lvv3//3339/39zZnzpxs2rRpyGv5gQceGND/K1euHLT/e6/l3v4/ui97r+UkA/p/sGu597+d3v7v6upKZ2fngP4/+lru7f/+13L//n/00Uczc+bMAX25cOHCvv7vfy0//fTTA/p/zZo1fdfYggULhryWW1pajun/3ja9/b9169Yhr2X3cvdy93L3cvdy93L3cvdy93L3cvdy93L3cvfyM/Fefv755y8/6dB1hEqtdaw+e1SVUv5zkn+T5K5a63uGOu+2226rSXLdddeNVmkAAAAAAKdKGasPPpMWHPxvPc+XjWkVAAAAAABngDMpfN7V8zxtTKsAAAAAADgDnEnh85t7nh8b0yoAAAAAAM4AEyp8LqW0llLmDLL/ZUn+a8/mn4xuVQAAAAAAZ57JY11Ak/1Ckt8opXw/yeNJ9iR5ZZKrkpyT5H8m+c9jVx4AAAAAwJlhooXP30/ymiQ/nu5pNqYleT7JXyX54yR/XGutY1ceAAAAAMCZYUKFz7XWe5LcM9Z1AAAAAACc6SbUnM8AAAAAAJwehM8AAAAAADSd8BkAAAAAgKYTPgMAAAAA0HTCZwAAAAAAmk74DAAAAABA0wmfAQAAAABoOuEzAAAAAABNJ3wGAAAAAKDphM8AAAAAADSd8BkAAAAAgKYTPgMAAAAA0HTCZwAAAAAAmk74DAAAAABA0wmfAQAAAABoOuEzAAAAAABNJ3wGAAAAAKDphM8AAAAAADSd8BkAAAAAgKYTPgMAAAAA0HTCZwAAAAAAmk74DAAAAABA0wmfAQAAAABoOuEzAAAAAABNJ3wGAAAAAKDphM8AAAAAADSd8BkAAAAAgKYTPgMAAAAA0HTCZwAAAAAAmk74DAAAAABA0wmfAQAAAABoOuEzAAAAAABNJ3wGAAAAAKDphM8AAAAAADSd8BkAAAAAgKYTPgMAAAAA0HTCZwAAAAAAmk74DAAAAABA0wmfAQAAAABoOuEzAAAAAABNJ3wGAAAAAKDphM8AAAAAADSd8BkAAAAAgKYTPgMAAAAA0HTCZwAAAAAAmk74DAAAAABA0wmfAQAAAABoOuEzAAAAAABNJ3wGAAAAAKDphM8AAAAAADSd8BkAAAAAgKYTPgMAAAAA0HTCZwAAAAAAmk74DAAAAABA0wmfAQAAAABoOuEzAAAAAABNJ3wGAAAAAKDphM8AAAAAADSd8BkAAAAAgKYTPgMAAAAA0HTCZwAAAAAAmk74DAAAAABA0wmfAQAAAABoOuEzAAAAAABNJ3wGAAAAAKDphM8AAAAAADTdhAufSykXlFK+UkrZUUo5WErZUkr53VLK7LGuDQAAAADgTDF5rAtoplLKK5P8TZIFSb6ZZFOSNyX5RJKfKaX8dK21cwxLBAAAAAA4I0y0kc+3pjt4/nit9d211t+otV6R5HeSvCbJfxjT6gAAAAAAzhATJnwupbwiyZVJtiT5/FGHfzPJviS/UkqZNsqlAQAAAACccSZM+Jzkip7n79Zau/ofqLXuSfLXSaYm+fujXRgAAAAAwJlmIs35/Jqe54eHOP5IukdGvzrJiuHe7LbbbmtSWQAAAAAAY6Zed911ZSw+eCKNfJ7Z8/zCEMd7988a7GAp5dpSyqqmVwUAAAAAcAaaSCOfh9Ob7tfBDtZab09y++iVAwAAAAAwcU2kkc+9I5tnDnF8xlHnAQAAAABwikyk8PmhnudXD3H8VT3PQ80JDQAAAABAk5RaB52FYtwppbwyyeYkW5K8stba1e/Y9CRP5v+0d+dhUlVnHse/Lw10Iy7sqICggCyazIhEBBQx0STu6ACGPBp1ohNH0ThJHh2TEDWPk2ii0SQm0QwdcUZFA1HQkYhRgrhGMMYFRJulFZUdka2bprvP/HFOSVFW01XF7brVVb/P85ynus7dTvV5695b5557rm9s7+6c2x5LIUVERERERERERERKRNH0fHbOLQeeAvoBV6ZMvgnoCPyPGp5FREREREREREREWl7R9HyGT3s/vwj0AGYDbwMjgJPxw22Mcs5tjK+EIiIiIiIiIiIiIqWhqBqfAcysD/Bj4KtAV/xwG7OAm5xzm+Ism4iIiIiIiIiIiEipKLrGZxERERERERERERGJX9GM+SwiIiIiIiIiIiIihUONzyIiIiIiIiIiIiISubZxF6AQlZW1cY2NGo5ERERERKQYmBlRDjfYxiDKnwttgMboVpdXUf9vcy9HG5yL8L9oBnF8LmsDUX6OUhNXvYmIFLB27dqza1fdBudc9zi2r8bnNBobHW5R3KUQEREREZEo2HDHggULIlvfmDFjcHdFtjpsMrjPj49uhXlkb8zkrHe+GXcxeHxQJZNeqotsfdNHtufYGVWRrS9Tr04YCN9X42nOfmJMOn9J3KUQESko0x8eCvBeXNvXsBsiIiIiIiIiIiIiEjk1PouIiIiIiIiIiIhI5NT4LCIiIiIiIiIiIiKR05jPIiJSci6YAs+8Attr4eCucO034NJxcZdKRETitGrVKi655BJOOukkpkyZsk/rGnsnvFwNbUNXn16d4J0f7XsZc/HQ5lXctHYJ79ft4OB2FUzrM5wTO8byvKE9zDnmvj3eN9Q20O/rQ/jclJEtsr1tq6tZ9POr2fDWy5S1K6fPyecx7JrbadM2up/ENR8sY9XUm9i+4i3aHdiFXhdeR+cRX45s/Xm16C54YxqsfxOGToKzpu2etvIZmHslbHkfDh3hpx3UN6aCisTr3aoHWLlyFps/eZe+h53B8SN+AkBDQx0vvXwtmza9xfYdH/HFk6fRs8dxMZdWJB5qfBYRkZJz/cVQOQXK28PSahj7LThmEBw7JO6SiYhIXO644w4GDx4c2frumgiXjopsdTn5y9a1XLf6TR4+bATH7deF1fW18RYoyemvXfTp3/U7dvHU6Ac59Kv9Wmx7i35+NRWdu3Pu4+9Tt20zf736NKoeuZtBEydHsn7XUM/yn/073U+dxMAp09i65BWW3/otOvSZTcWhh0eyjbza/1AY/UNYMRfqa3bn79gAj5wHp0+FgWfBs1Pg0fPh4pfjK6tIjDp06MFRQ7/F6jUv0NCwc49p3boNY9CRF/L8i/8RU+lECoOG3RARkZJzVH/f8AxgIS3/IM4SiYhInJ555hkOOOAAhg0bFndRInXD2sX8qOcQju/YlTZm9GrXgV7tOsRdrM9YPbea8i4d6DL84BbbxvaPVnLYl8ZTVl5Bh64Hc8jxX+GTFUsiW3/thyvYtWkdPc68BCsr48DPjaTjoGFsXDArsm3k1eDzYNA46NB1z/x3HoFuR8GQCdC2Ak68Eda9DhuWxlJMkbj16X0qvXufQnl5pz3yy8raM3jQN+je/VjMymIqnUhhUOOziIiUpCtugf1Gw+DxcEg3OH103CUSEZE4bN++ncrKSq644opI13v9Y9DtOhj9C5j/bqSrzkiDcyyq+Zj19TsZsPTP9H77CSZ/+Bo1jQ35L0wzVj1aRe9xAzCzFtvGkROv4r2//JH62h3sWPchq19+kkOOj3BIDOfS5tW+XxXdNgrB+sXQ4592v2/fETr1hw2L4yuTiIgUNDU+i4hISfrtf8LWBfDcVDjv5N09oUVEpLRMnTqVM844g549e0a2zlvHwYob4cOb4d9GwVn3wPL1ka0+I2vra9nlHDM/+ZDn+o/lHwNP4bWazdy87u38FqQZOz7axsaFa+gzbmCLbqfHsBP5ZOUSZp7SldnnHE6XwcfS+6RzIlt/Ra8jaHtQF9Y+9t+4+l1sef05ti1ZSGNdTfMLtya7tkHFQXvmVRwEdVvjKY+IiBQ8NT6LiEjJKiuDE/4ZPlgHv5sZd2lERCTfqqqqePXVV5k4cWKk6x3RDw6ogPJ2cNHxMPoImJPnjqEdwm3eV3XtzyHtOtCtbTnf6T6QOVvW5LcgzfhgVhVdju3Jfn0OaLFtuMZG5l9zJn3GjmPCvM2c9+Rq6rZ+zD9+c31k27C27eh/7e/45NX5vH7ZKNY+/gc6jzqNdl1abiiRWLTbH3Zu2TNv5xZo33L1JyIirZseOCgiIiWvvkFjPouIlKLXXnuNNWvWMGHCBABqampoaGigurqaysrKyLZjBmkGZWhRndu2p3e7Di06lEUUPpi9jAGX/VPzM+6Dui2b2LF2FQPHX0FZ+3LK2pdzxBkX8cbvb+CYybdEtp39+g5m0I8f/PT90h9MpOvYcyNbf0HofhS8ed/u93Xb4ePlfhxoERGRNNTzWURESsq6TfDQXNi2AxoaYO5LMH0ufHF43CUTEZF8O/vss5k+fTqVlZVUVlZy9tlnM3LkSG6//fac17l5B8xdArW7/MXNBxbCgmXwlSERFjxDl3Tux683LGNdfS0f19dx5/oqzjywcHribvr7WmrX7uDQr/Zr0e2Ud+pGx0MPZ9kj99BYX0/d1s2snPO/dBrw+Ui3s+O9pTTW7aRxZw1rHpvKro/X03XseZFuI28a66G+FlyDT/W1Pu/Ic2H9W7D0Tz7v+R9Dj89Dt8Fxl1gkFo2N9TQ07KTRNeBcg/+7sR6AhoY6Ghp2hvl20dCwE5dufHiRIqeezyIiUlLM4Hd/gst/Co0O+h4Md34Xzhkbd8lERCTfKioqqKio+PR9hw4daN++PZ06dcp5nbsa4If/B0vXQlkbGNwTZl0Gg6IbUjpjU3oOYUPDTo5cOpeKNm2YeFAfftAjhlbwJqyaVcXBp/al7f4t/+CFE3/6MH+/83ssuf82rE0ZPYedxLBv3xbpNjY9O4sN82bg6uvZf8hwBk6ZRpt25ZFuI2+evxmev2n3+7fuhxNugDE3wnl/grmT4bEL4NARMO6h2IopErfFS+7mrcW//fR99XuPc/RRV/C5oyfzxJzT2b7jIwDmP3sZAGed+Rf279grlrKKxMV01eWzzMy5RXGXQkREREREomDDYcGCBZGtb8yYMbi7IlsdNhnc58dHt8I8sjdmctY734y7GDw+qJJJL9VFtr7pI9tz7IyqyNaXqVcnDITv6zd6zn5iTDp/SdylEBEpKNMfHgrwqnMulvt9NeyGiIiIiIiIiIiIiEROjc8iIiIiIiIiIiIiEjk1PouIiIiIiIiIiIhI5NT4LCIiIiIiIiIiIiKRU+OziIiIiIiIiIiIiETOnNOTdFOZ2VbgnbjLIQWlG7Ah7kJIwVFcSDqKC0mlmJB0FBeSjuJCUikmJB3FhaSjuJCmdAM6Oue6x7HxtnFstBV4xzk3PO5CSOEws0WKCUmluJB0FBeSSjEh6SguJB3FhaRSTEg6igtJR3EhTQmx0S+u7WvYDRERERERERERERGJnBqfRURERERERERERCRyanxO7/dxF0AKjmJC0lFcSDqKC0mlmJB0FBeSjuJCUikmJB3FhaSjuJCmxBobeuCgiIiIiIiIiIiIiEROPZ9FREREREREREREJHJqfBYRERERERERERGRyKnxOTCz3mb2BzP7yMx2mlm1md1pZp3jLpvsGzPramaXmtmjZrbMzGrM7BMze97Mvmlmab8HZjbKzOaY2SYz22Fmb5jZNWZWtpdtnWlm88P6t5nZ38zsopb7dBIlM7vQzFxIlzYxT9Z1bGYXmdkrYf5PwvJntsynkCiY2Ylm9iczWx2OCavN7CkzOz3NvNpXlAAzOyPEwAfhOLLCzGaY2cgm5ldcFAEzG29mvzaz58xsSzg+3N/MMnmpex1b4pFNTJjZQDO7zszmmdkqM6szs7VmNtvMTm5mO1nVr5mVhTh7I+yjNoU4HLWvn1mal8u+ImX5yqRz0AFNzJN1HZtZBzO7yczeMbNaM1tnZn80syG5fE7JTo7HEAvf//mhjmvMbGWotyObWEb7i1Yi25gws3IzuzLU74ZQx2+b2a/MrO9ellNMtDJWoO1WZjbazH5mZgvNbL3538YrzWxqU8ertJxzJZ+A/sBawAGzgFuAeeH9UqBr3GVU2qf6vTzU5UfAA8BPgT8Am0P+TML450nLnAPUA9uASuDnIRYcMKOJ7UwO0zcAvwHuAFaFvNvi/j8oNRsnfUJMbA11dmkUdQzcFqavCvP/BtgY8ibH/bmV0tbZD0P9rAfuBX6Cf0DDQuBnKfNqX1ECCbg1qc6mhvOEmUAd0AhcoLgozgT8I9TBVuDt8Pf9e5k/L3WvY0vriAngoTB9MXAP/hz0kRAjDrg6ivoFDJjB7t8uPw/xty1s65y4/2/FnrLdV6Qse1bSsg4YEEUdA+XA82GZheFY9iCwC9gOjIj7/1bsKYdjSAXweFI93xX2G/cBK4Az0yyj/UUrSlkeQ9omfYffBn4d6vvZkLcZGKqYKI5EgbZbAWuABuA54M4QXy+E+bcBIzP6fHH/gwshAXPDP+6qlPxfhPy74y6j0j7V7xfxJ3VtUvIPBt4PdfwvSfkHAuuAncDwpPwK4MUw/9dS1tUPqA079X5J+Z2BZWGZjL6USrHEiAFPA8vDDvszjc+51DEwKuQvAzqnrGtjWF+/lvpcSjnFwoRQZ38BDkgzvV3S39pXlEAKx4qGcOLVI2XayaHOViguijOFOh4YjhNj2fuPxLzUvY4trSomLgaOSZN/Ev7i1U7gkH2tX2BSWOYFoCIp/wthG+tIc0xTiicuUpbrHo4vDwHzabrxOes6Bq4Py8wg6XcQvrEicVGkTS6fV6ll4gLfEOTwHR8+UzcknYeG99pftLKU5TEk8bvk6dR4AG4K0/5Na9EVAAAL3UlEQVSgmCiORIG2WwHXAYemKe/3w/xvZvL5Sn7YDTM7AvgyUI3f2Se7AX9V+EIz65jnoklEnHPznHOPO+caU/LXAHeHt2OTJo3Hnwg+5JxblDR/Lb5HJMC/p2zmX/G9C+5yzlUnLfMx/uQB/JUsKUxX43f2l+C/8+nkUseJ9/8V5kssU43f35SHbUoBCLcy3QrsAL7unNuaOo9zblfSW+0rSkNf/DBlf3POrUue4Jz7K77nSvekbMVFEXHO/dU5V+XCWXYz8lX3OrbEKJuYcM5Nc869lib/WXxDY3t8Q0GyXOo3EVc/DPGWWGYh8DA+Lsc3V17JXZb7imS/D69XNjNfVnVsZsbuWLo2+XeQc242vgfbUPyFEGkh2cSFmfXH19lC4Aepv13D+nalZGl/0cpkua84Irw+kSYeZofX7in5iolWqlDbrZxztzrnPkpT5FuBGuBoM+va3Ocr+cZnfIMTwFNpKnkr/urPfsDx+S6Y5EXiAF6flJeIiSfTzL8A3zA1yszKM1zmzynzSAEJY97dAvzSObdgL7PmUseKi9ZlFHA4MAf42PwYv9eZ2bct/bi+2leUhip8D8XjzKxb8gQzGwMcgO+RkqC4KF35qnvFS3FIdw4KWdZviKdR+Ph6LpNlpDCY2cXAOOBy59zGvcyXSx33Bw4D3nXOrcxwGYnXJHz7zH3AgWZ2gZldb2b/tpdxVbW/KG6Lw+tpacb7TYzf/HRKvmKiOBViu5VLKk9DczO3zXClxWxQeH23ielV+J7RRwLP5KVEkhdm1hb4Rnib/OVrMiacc/VmthI4Cn8l8u0MllltZtuB3ma2n3NuRxTll30XYuB/8bexfL+Z2bOq43C3RC9gm3NudZr1VYXXtA8OkVh8IbyuBf4OfC55opktAMY759aHLO0rSoBzbpOZXYcfimuJmc3C36rWHzgbP0TLt5IWUVyUrhavex1bikN4SNSX8D8MFyTl51K/A4Ay/PA/qQ3ZTS0jMQsx8Ev87fazmpk9lzrO5Ddu6jISr8R56EH4oQCTexI6M/sdfpz4BtD+okQ8gX9OwHnAm2b2NL5DxLHACfgxoO9KzKyYKE4F3G41Ad8J52Xn3ObmPod6PvudO8AnTUxP5HfKQ1kkv24BjgbmOOfmJuXnEhOZLnNQE9MlHj8CjgEuds7VNDNvtnWsfUvr0yO8Xg50AE7BH1CPxj8bYAx+3MQE7StKhHPuTvyJf1vgMuA/8Sdcq4BpKcNxKC5KVz7qXseWVi70QHoAf9vrjcm3RdOyMaSYKBChB+N9+Ac1XZ3BIoqL0pA4D/0xsAjfCeIA/IWq5cAVwJSk+RUXRS4MzTEeuBHfaHg18D38uNELgAcTFyMCxURxKrh2KzM7HH/xox747t7mTVDjc/MsvGY7fpcUMDO7Gv8lWQpcmO3i4TWbmFAcFRgzOw7f2/l259xLUawyvGZbx4qJwlEWXg3fw/kZ59w259xi4FzgA+CkJobgSEf7iiJhZtfinzA9Dd/juSO+18kK4AEz+1k2qwuviovSk8+6V6wUIDMrw99xNRo/juZtOa5K+4/W7T/wYy1flnLxIVc6rhSHxHnoauBc59xb4Tx0Hr4BshH4jpm1z3K9iotWyswq8MeK7+HHhT8E3yh4Ov6ZJAvM7JwcVq2YaCUKsd3KzHrgh+joDnzbOfdiJitW43PzLfsHpswnrZyZXYm/zW0JcLJzblPKLLnERKbLbMmiqNJCkobbeJc9exDsTbZ13Nz8zV11lPxL/ABc4Zx7PXlC6BmfuNJ8XHjVvqIEmNlY/AM1HnPOfcc5t8I5t8M593f8RYkPge+GBxiD4qKU5aPudWxppULD8/34uyb+CFyQ5oFTudSvfsu0ImY2EPgv4F7n3JwMF2vJfYvionAkzkOfTL0jM5yXrsT3hB4SsrW/KH6JO+1+4Jy7xzm3xjm3xTn3Z/wFiXb4do0ExUQRKcR2q9DwPA/fE//bzrnfNrGez1DjM7wTXpsaw2ZgeG1qvCxpRczsGvy4SG/hv8Br0szWZEyERsvD8bcXrMhwmUPwveQ+0FidBWN/fF0NAWrNzCUScEOY579D3p3hfVZ17Jzbjm+U2j9MT6V9S+FJ1HFTY1YlfhR0SJlf+4rilnigy19TJ4R6egV/PnVMyFZclK4Wr3sdW1qnUP/Tga8BDwJfTze2Zo71uwz/oJ8jwnYyWUbicxR+yJVLks8/wznoSWGeqpA3LrzPpY71G7f1yeo8VPuLkrC3c9DXgU1AXzPrGvIUE0WiENutwvT5wFDgSufcrzL9PKDGZ9j9Rf5y6hNEzewA/G1xNcDL+S6YRCs8MOoO4B/4L/C6JmadF16/mmbaGGA/4EXn3M4MlzktZR6J306gson0Wpjn+fA+MSRHLnWsuGhdFuAP0AObuKXx6PBaHV61rygNiSdEd29ieiK/LrwqLkpXvupe8dKKhOPJTHzvtf8BLkwZozNVVvUb4ulFfHydmMkyEqtqmj4HTTQuzAjvqyHnOl6Of6D2kWFszkyWkXg9E16PTp0QxopPNAJWJ03S/qK4NXkOGmIi0Uu1LmmSYqKVK8R2KzPrDTwLDAYuz6bH86eccyWf8LdSO+CqlPxfhPy74y6j0j7X8ZRQl4uALs3MeyCwHt9AOTwpvwK/Y3bA11KWORyoBTYC/ZLyO+OvJjpgZNz/B6WMYuXGUF+X7msdA6NC/jKgc1J+v7Ce2uR1KcWf8LdEO+DmlPxT8WPtbQY6hTztK0ogARNDvawBeqVMOy3ERQ3QVXFR3AkYG+ri/iam56XudWwpnJRBTJQDT4R5pgJtMlhn1vULTArLvABUJOV/IcTjOuDAuP9fpZKai4u9LDc/LDcgzbSs6xi4PiwzIzn2gHNC/uJMYlIpP3EBtMdfNGgETk2ZdnNYdn5KvvYXrThlEBO/DdOfBspTpv00THtFMVE8iQJstwIOC/umBuCSXD+bhZWVNDPrj6+cHsBs4G1gBP4pou8Co5xzG+MroewLM7sI/5CoBvwTOdONV1TtnJuWtMw4fC+VWuAh/C0tZ+PHtpkJTHQpXx4zuwr4Ff6L/DD+CuR4oDf+oXbfi/JzScswsxvxQ29c5pybmjIt6zo2s9uB7+AfVjcTf2J5PtAVf8Hrrhb7MJK1MI7VC8AA4Dn8kAp98WP7Ovyt0jOS5te+osiFu6LmAqcAW4FH8Q3RQ/C3QxpwjXPul0nLKC6KRKjLxK3vBwNfwd+++FzI25BcN/mqex1b4pNNTJjZvcDFwAZ2NyKkmu+cm5+yjazq18wMP470ePxDiR4P856P/xH6L8652bl+ZmletvuKJtYxHz/0xkDn3LKUaVnXcegZOQ/fGLUI37P2MHwv/Drgi865v+XwcSVDORxDTgCewn/nHwXewzcAjsE3Mp3gnNtj+APtL1qXLI8hvfB34PfG93h/Et/hYTT+GTQ1wJecc4k7dRPbUEy0QoXabmVmK/EXL14F/q+J4k9zzlXv9QPG3bJfKAnoA9yLf7psHX5H/0uaudqgVPiJ3T1Z95bmp1luNDAHP75WDfAm/unUZXvZ1ln42xG2AtuBhcBFcf8PlHKKl0ubmJ51HQMXhfm2h+WeBc6M+7MqNVlfXfB3vqwMx4ON+AuTxzcxv/YVRZ7wD3S5Bv8DYAt+eJZ1+BOwLysuijdlcA5RHVfd69hS+DHB7p6se0s3RlG/QNsQZ2+GuPs4xOGouP9npZBy2VekWUciXj7T8znXOsaPD3wTUIXvHbce3xN6aNz/s1JIOR5DhuIbhNbhz0NXAfcAvfeyHe0vWknKNibwQ27chu8gWcvutqp7gcGKieJJGcRGLO1WGZTJAWOb+3zq+SwiIiIiIiIiIiIikdMDB0VEREREREREREQkcmp8FhEREREREREREZHIqfFZRERERERERERERCKnxmcRERERERERERERiZwan0VEREREREREREQkcmp8FhEREREREREREZHIqfFZRERERERERERERCKnxmcRERERERERERERiZwan0VEREREREREREQkcmp8FhEREREREREREZHI/T+LxQiG6vdqgQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 1728x576 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Plotting the predicted fitness effect for the variants seen more than once\n",
    "plt.close('all')\n",
    "f, (ax1, ax2) = plt.subplots(2, 1, sharey=True, sharex = True, figsize=(24, 8))\n",
    "gs = matplotlib.gridspec.GridSpec(2, 1, width_ratios=[1], height_ratios=[10,0.5])\n",
    "ax1 = plt.subplot(gs[0])\n",
    "ax2 = plt.subplot(gs[1])\n",
    "gs.update(hspace=0.05)\n",
    "\n",
    "nonsense_list = []\n",
    "y_list = []\n",
    "x_list=[]\n",
    "y_list_top20 = []\n",
    "x_list_top20=[]\n",
    "y_list_nonsense = []\n",
    "x_list_nonsense = []\n",
    "x=[]\n",
    "y=[]\n",
    "OBS_EXP_DICT={}\n",
    "top20_lower_err = []\n",
    "top20_upper_err = []\n",
    "missense_lower_err = []\n",
    "missense_upper_err = []\n",
    "nonsense_lower_err = []\n",
    "nonsense_upper_err = []\n",
    "\n",
    "top20 = c1\n",
    "missense = c6\n",
    "nonsense = c10\n",
    "synonymous = c3\n",
    "\n",
    "axisfont=20\n",
    "domainfont = 12\n",
    "titlefont=12\n",
    "axislabelfont=12\n",
    "\n",
    "top_20_TET2 = []\n",
    "\n",
    "#ERROR NEEDS TO BE IN THE FORMAT: [tuple(lower_err),tuple(upper_err)]\n",
    "\n",
    "for k, v in TET2_missense_s.items():\n",
    "    if k in top_20_TET2:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_top20.append(codon_number)\n",
    "        y_list_top20.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        top20_lower_err.append(error_lower)\n",
    "        top20_upper_err.append(error_upper)\n",
    "\n",
    "    if k not in top_20_TET2:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list.append(codon_number)\n",
    "        y_list.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        missense_lower_err.append(error_lower)\n",
    "        missense_upper_err.append(error_upper)\n",
    "\n",
    "    missense_err = [tuple(missense_lower_err), tuple(missense_upper_err)]   \n",
    "    \n",
    "for k, v in TET2_nonsense_s.items():\n",
    "    if k in top_20_TET2:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_top20.append(codon_number)\n",
    "        y_list_top20.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        top20_lower_err.append(error_lower)\n",
    "        top20_upper_err.append(error_upper)\n",
    "    \n",
    "    top20_err = [tuple(top20_lower_err), tuple(top20_upper_err)]\n",
    "        \n",
    "    if k not in top_20_TET2:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_nonsense.append(codon_number)\n",
    "        y_list_nonsense.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        nonsense_lower_err.append(error_lower)\n",
    "        nonsense_upper_err.append(error_upper)\n",
    "\n",
    "    nonsense_err = [tuple(nonsense_lower_err), tuple(nonsense_upper_err)]        \n",
    "    \n",
    "msize = 12\n",
    "medgewidth = 2\n",
    "linew = 3\n",
    "csize = 4\n",
    "cthick = 3\n",
    "\n",
    "markers1, caps1, bars1 = ax1.errorbar(x_list, y_list, yerr = missense_err, fmt = 'o', ecolor = missense, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = missense, markerfacecolor = missense, zorder = 1)    \n",
    "\n",
    "markers2, caps2, bars2 = ax1.errorbar(x_list_nonsense, y_list_nonsense, yerr = nonsense_err, fmt = 'o', ecolor = nonsense, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = nonsense, markerfacecolor = nonsense, zorder = 2)    \n",
    "\n",
    "markers3, caps3, bars3 = ax1.errorbar(x_list_top20, y_list_top20, yerr = top20_err, fmt = 'o', ecolor = top20, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = top20, markerfacecolor = top20, zorder = 3)    \n",
    "\n",
    "# loop through bars and caps and set the alpha value\n",
    "a = 0.7\n",
    "[bar.set_alpha(a) for bar in bars1]\n",
    "[cap.set_alpha(a) for cap in caps1]\n",
    "[bar.set_alpha(a) for bar in bars2]\n",
    "[cap.set_alpha(a) for cap in caps2]\n",
    "[bar.set_alpha(a) for bar in bars3]\n",
    "[cap.set_alpha(a) for cap in caps3]\n",
    "\n",
    "#Only show the required axis lines\n",
    "ax1.spines['top'].set_visible(False)\n",
    "ax1.spines['right'].set_visible(False)\n",
    "\n",
    "for axis in ['bottom','left']:\n",
    "    ax1.spines[axis].set_linewidth(2)\n",
    "\n",
    "for axis in ['bottom','left']:\n",
    "    ax1.spines[axis].set_color(grey3)\n",
    "    \n",
    "ax1.set_ylabel('fitness effect (s)', fontsize = axisfont, labelpad = 10)\n",
    "ax1.set_title('Fitness effects across TET2', fontsize = 18)\n",
    "\n",
    "# ax1.set_yscale('log')\n",
    "\n",
    "# STACKED BAR CHART\n",
    "# define the exons + create data\n",
    "exons = [{'3':(1137-1)+1, '4': (1167-1138)+1, '5': (1198-1168)+1, '6':(1268-1199)+1, '7':(1318-1269)+1, '8': (1348-1319)+1, '9': (1394-1349)+1,'10': (1513-1395)+1, '11': (2002-1514)+1}]\n",
    "exon_orders = [['3', '4', '5', '6', '7', '8', '9', '10', '11']]\n",
    "colors2 = [c4, c0, c3, c1, c5, c6,c7, c8, c9]\n",
    "names = exons[0].keys()\n",
    "values = np.array([[data[name] for name in order] for data,order in zip(exons, exon_orders)])\n",
    "lefts = np.insert(np.cumsum(values, axis=1),0,0, axis=1)[:, :-1]\n",
    "orders = np.array(exon_orders)\n",
    "\n",
    "#plot the bar chart\n",
    "for name, color in zip(names, colors2):\n",
    "    idx = np.where(orders == name)\n",
    "    value = values[idx]\n",
    "    left = lefts[idx]\n",
    "    ax2.bar(x=left, height=-10, width=value, bottom=-5, color=color, orientation = 'horizontal', label = name, edgecolor='k', linewidth=1, tick_label = name)\n",
    "    ax2.text(left + (value/2.), (-0.26), name, ha='center', va='center', fontsize = 12)\n",
    "\n",
    "# CONFIGURING THE GRAPH APPEARANCE\n",
    "#Set the x and y axis limits\n",
    "ax1.set_xlim(0, 2004)\n",
    "ax1.set_ylim(-0.01, 25)\n",
    "ax2.set_xlim(0, 2004)\n",
    "ax2.set_ylim(-0.5, 0)\n",
    "\n",
    "#x-axis ticks\n",
    "x1_major_ticks = []\n",
    "x1_major_tick_labels = []\n",
    "x2_major_ticks = [0, 200, 400, 600, 800, 1000, 1200, 1400, 1600, 1800, 2002]\n",
    "x2_major_tick_labels = [\"0\", \"200\", \"400\", \"600\", \"800\", \"1000\", \"1200\", \"1400\", \"1600\", \"1800\", \"2002\"]\n",
    "ax1.set_xticks(x1_major_ticks)\n",
    "ax1.set_xticklabels(x1_major_tick_labels, fontsize = axisfont)\n",
    "ax2.set_xticks(x2_major_ticks)\n",
    "ax2.set_xticklabels(x2_major_tick_labels, fontsize = axisfont)\n",
    "\n",
    "#y-axis ticks\n",
    "y1_major_ticks = [0, 5, 10, 15, 20, 25]\n",
    "y1_major_tick_labels = [\"0\", \"5\", \"10\", \"15\", \"20\", \"25\"]\n",
    "ax1.set_yticks(y1_major_ticks)\n",
    "ax1.set_yticklabels(y1_major_tick_labels, fontsize = axisfont)\n",
    "\n",
    "#set panel legends\n",
    "legend_marker_size = 2\n",
    "legend_elements = [Line2D([0], [0], marker = '_', color = missense, alpha = 1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='missense'),\n",
    "                   Line2D([0], [0], marker = '_', color=nonsense, alpha=1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='nonsense'),\n",
    "                  Line2D([0], [0], marker = '_', color=top20, alpha=1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='top 20 observed variant')]\n",
    "    \n",
    "ax1.legend(ncol=1, handles=legend_elements, loc='upper left', frameon=0, fontsize = 20)\n",
    "\n",
    "ax1.grid(which='major', linestyle=':', linewidth='0.7', color=grey3, zorder = 0)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ASXL1 variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {
    "code_folding": [
     0,
     86,
     97,
     109
    ]
   },
   "outputs": [],
   "source": [
    "def expected_number_variants_missense_ASXL1(data, params):\n",
    "    s_value = params[0]\n",
    "    \n",
    "    a = 0\n",
    "    expected_number_by_variant_missense = {}\n",
    "    \n",
    "    Jaiswal_fmin = 0.056\n",
    "    Jaiswal_mean = 58\n",
    "    Jaiswal_sigma = 11.5\n",
    "    Jaiswal_total = 17182\n",
    "    \n",
    "    Genovese_fmin = 0.1194\n",
    "    Genovese_mean = 55.1\n",
    "    Genovese_total = 12380\n",
    "    Genovese_sigma = 11.95\n",
    "    \n",
    "    McKerrel_fmin = 0.008\n",
    "    McKerrel_mean = 51.88\n",
    "    McKerrel_total = 4219\n",
    "    McKerrel_sigma = 18.36\n",
    "    \n",
    "    Zink_fmin = 0.2363\n",
    "    Zink_mean = 60\n",
    "    Zink_total = 11262\n",
    "    Zink_sigma = 1.18907720e+01\n",
    "    \n",
    "    Coombs_fmin = 0.019\n",
    "    Coombs_mean = 59.82\n",
    "    Coombs_total = 5649\n",
    "    Coombs_sigma = 15.14\n",
    "    \n",
    "    Young_fmin = 0.0009\n",
    "    Young_mean = 63\n",
    "    Young_total = 89\n",
    "    Young_sigma = 6.8\n",
    "    \n",
    "    Desai_fmin = 0.0169\n",
    "    Desai_mean = 66.1\n",
    "    Desai_total = 424\n",
    "    Desai_sigma = 6 #estimated\n",
    "    \n",
    "    Acuna_fmin = 0.0015\n",
    "    Acuna_mean = 4.5\n",
    "    Acuna_total = 2014\n",
    "    \n",
    "    N = 9.40166610e+04\n",
    "\n",
    "    for i in data:\n",
    "        variant_mutation_rate = mutation_rates_variants_ASXL1(i)\n",
    "\n",
    "        if i in Genovese_missense_ASXL1:\n",
    "            Genovese_number = prevalence_averaged_over_ages(Genovese_fmin, s_value, N, variant_mutation_rate, Genovese_mean, Genovese_sigma, Genovese_total)\n",
    "        else:\n",
    "            Genovese_number = 0 \n",
    "\n",
    "        if i in Zink_missense_ASXL1:\n",
    "            Zink_number = prevalence_averaged_over_ages(Zink_fmin, s_value, N, variant_mutation_rate, Zink_mean, Zink_sigma, Zink_total)\n",
    "        else:\n",
    "            Zink_number = 0\n",
    "\n",
    "        if i in Coombs_missense_ASXL1:\n",
    "            Coombs_number = prevalence_averaged_over_ages(Coombs_fmin, s_value, N, variant_mutation_rate, Coombs_mean, Coombs_sigma, Coombs_total)\n",
    "        else:\n",
    "            Coombs_number = 0\n",
    "\n",
    "        if i in Young_missense_ASXL1:\n",
    "            Young_number = prevalence_averaged_over_ages(Young_fmin, s_value, N, variant_mutation_rate, Young_mean, Young_sigma, Young_total)\n",
    "        else:\n",
    "            Young_number = 0\n",
    "\n",
    "        if i in Desai_missense_ASXL1:\n",
    "            Desai_number = prevalence_averaged_over_ages(Desai_fmin, s_value, N, variant_mutation_rate, Desai_mean, Desai_sigma, Desai_total)\n",
    "        else:\n",
    "            Desai_number = 0\n",
    "\n",
    "        if i in Acuna_missense_ASXL1:\n",
    "            Acuna_number = prevalence_averaged_over_uniform_ages(Acuna_fmin, s_value, N, variant_mutation_rate, Acuna_mean, Acuna_total)\n",
    "        else:\n",
    "            Acuna_number = 0\n",
    "\n",
    "        expected_total = Genovese_number + Zink_number + Coombs_number + Young_number + Desai_number + Acuna_number\n",
    "\n",
    "        expected_number_by_variant_missense[i] = expected_total\n",
    "    \n",
    "    return expected_number_by_variant_missense\n",
    "\n",
    "def prob_observed_given_expected_missense_ASXL1(params, data): #data = list of variant (e.g. p.R882H) (or all_missense_variants_ASXL1)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_ASXL1_counts[variant])\n",
    "        expected_number = float(expected_number_variants_missense_ASXL1(data, params)[variant])\n",
    "        square_distance = ((observed_number - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance\n",
    "\n",
    "def prob_observed_given_expected_missense_ASXL1_upper_error(params, data): #data = list of variant (e.g. p.R882H) (or all_missense_variants_ASXL1)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_ASXL1_counts[variant])\n",
    "        observed_number_upper_error = observed_number+np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_missense_ASXL1(data, params)[variant])\n",
    "        square_distance = ((observed_number_upper_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   \n",
    "\n",
    "def prob_observed_given_expected_missense_ASXL1_lower_error(params, data): #data = list of variant (e.g. p.R882H) (or all_missense_variants_ASXL1)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_ASXL1_counts[variant])\n",
    "        observed_number_lower_error = observed_number-np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_missense_ASXL1(data, params)[variant])\n",
    "        square_distance = ((observed_number_lower_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {
    "code_folding": [
     0,
     91,
     102,
     114
    ]
   },
   "outputs": [],
   "source": [
    "def expected_number_variants_nonsense_ASXL1(data, params):\n",
    "    s_value = params[0]\n",
    "    \n",
    "    a = 0\n",
    "    expected_number_by_variant_nonsense = {}\n",
    "    \n",
    "    Jaiswal_fmin = 0.056\n",
    "    Jaiswal_mean = 58\n",
    "    Jaiswal_sigma = 11.5\n",
    "    Jaiswal_total = 17182\n",
    "    \n",
    "    Genovese_fmin = 0.1194\n",
    "    Genovese_mean = 55.1\n",
    "    Genovese_total = 12380\n",
    "    Genovese_sigma = 11.95\n",
    "    \n",
    "    McKerrel_fmin = 0.008\n",
    "    McKerrel_mean = 51.88\n",
    "    McKerrel_total = 4219\n",
    "    McKerrel_sigma = 18.36\n",
    "    \n",
    "    Zink_fmin = 0.2363\n",
    "    Zink_mean = 60\n",
    "    Zink_total = 11262\n",
    "    Zink_sigma = 1.18907720e+01\n",
    "    \n",
    "    Coombs_fmin = 0.019\n",
    "    Coombs_mean = 59.82\n",
    "    Coombs_total = 5649\n",
    "    Coombs_sigma = 15.14\n",
    "    \n",
    "    Young_fmin = 0.0009\n",
    "    Young_mean = 63\n",
    "    Young_total = 89\n",
    "    Young_sigma = 6.8\n",
    "    \n",
    "    Desai_fmin = 0.0169\n",
    "    Desai_mean = 66.1\n",
    "    Desai_total = 424\n",
    "    Desai_sigma = 6 #estimated\n",
    "    \n",
    "    Acuna_fmin = 0.0015\n",
    "    Acuna_mean = 4.5\n",
    "    Acuna_total = 2014\n",
    "    \n",
    "    N = 9.40166610e+04\n",
    "\n",
    "    for i in data:\n",
    "        variant_mutation_rate = mutation_rates_variants_ASXL1(i)\n",
    "\n",
    "        if i in Jaiswal_nonsense_ASXL1:\n",
    "            Jaiswal_number = prevalence_averaged_over_ages(Jaiswal_fmin, s_value, N, variant_mutation_rate, Jaiswal_mean, Jaiswal_sigma, Jaiswal_total)\n",
    "        else:\n",
    "            Jaiswal_number = 0\n",
    "\n",
    "        if i in Genovese_nonsense_ASXL1:\n",
    "            Genovese_number = prevalence_averaged_over_ages(Genovese_fmin, s_value, N, variant_mutation_rate, Genovese_mean, Genovese_sigma, Genovese_total)\n",
    "        else:\n",
    "            Genovese_number = 0 \n",
    "\n",
    "        if i in Zink_nonsense_ASXL1:\n",
    "            Zink_number = prevalence_averaged_over_ages(Zink_fmin, s_value, N, variant_mutation_rate, Zink_mean, Zink_sigma, Zink_total)\n",
    "        else:\n",
    "            Zink_number = 0\n",
    "\n",
    "        if i in Coombs_nonsense_ASXL1:\n",
    "            Coombs_number = prevalence_averaged_over_ages(Coombs_fmin, s_value, N, variant_mutation_rate, Coombs_mean, Coombs_sigma, Coombs_total)\n",
    "        else:\n",
    "            Coombs_number = 0\n",
    "\n",
    "        if i in Young_nonsense_ASXL1:\n",
    "            Young_number = prevalence_averaged_over_ages(Young_fmin, s_value, N, variant_mutation_rate, Young_mean, Young_sigma, Young_total)\n",
    "        else:\n",
    "            Young_number = 0\n",
    "\n",
    "        if i in Desai_nonsense_ASXL1:\n",
    "            Desai_number = prevalence_averaged_over_ages(Desai_fmin, s_value, N, variant_mutation_rate, Desai_mean, Desai_sigma, Desai_total)\n",
    "        else:\n",
    "            Desai_number = 0\n",
    "\n",
    "        if i in Acuna_nonsense_ASXL1:\n",
    "            Acuna_number = prevalence_averaged_over_uniform_ages(Acuna_fmin, s_value, N, variant_mutation_rate, Acuna_mean, Acuna_total)\n",
    "        else:\n",
    "            Acuna_number = 0\n",
    "\n",
    "        expected_total = Jaiswal_number + Genovese_number + Zink_number + Coombs_number + Young_number + Desai_number + Acuna_number\n",
    "\n",
    "        expected_number_by_variant_nonsense[i] = expected_total\n",
    "    \n",
    "    return expected_number_by_variant_nonsense\n",
    "\n",
    "def prob_observed_given_expected_nonsense_ASXL1(params, data): #data = list of variant (e.g. p.R882H) (or all_nonsense_variants_ASXL1)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_ASXL1_counts[variant])\n",
    "        expected_number = float(expected_number_variants_nonsense_ASXL1(data, params)[variant])\n",
    "        square_distance = ((observed_number - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   \n",
    "\n",
    "def prob_observed_given_expected_nonsense_ASXL1_upper_error(params, data): #data = list of variant (e.g. p.R882H) (or all_nonsense_variants_ASXL1)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_ASXL1_counts[variant])\n",
    "        observed_number_upper_error = observed_number+np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_nonsense_ASXL1(data, params)[variant])\n",
    "        square_distance = ((observed_number_upper_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   \n",
    "\n",
    "def prob_observed_given_expected_nonsense_ASXL1_lower_error(params, data): #data = list of variant (e.g. p.R882H) (or all_nonsense_variants_ASXL1)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_ASXL1_counts[variant])\n",
    "        observed_number_lower_error = observed_number-np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_nonsense_ASXL1(data, params)[variant])\n",
    "        square_distance = ((observed_number_lower_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#CALCULATING S FOR MISSENSE VARIANTS\n",
    "ASXL1_missense_s = {}\n",
    "a = 0\n",
    "for var in all_missense_variants_ASXL1:\n",
    "    initial_guess=[0.05]\n",
    "    if var in all_studies_ASXL1_counts.keys():\n",
    "        observed_number = float(all_studies_ASXL1_counts[var])\n",
    "#         print(observed_number)\n",
    "        if observed_number >1:\n",
    "            variant_mutation_rate = mutation_rates_variants_ASXL1(var)\n",
    "            \n",
    "            MLE = scipy.optimize.minimize(prob_observed_given_expected_missense_ASXL1, initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value = MLE.x[0]\n",
    "\n",
    "            upper_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_missense_ASXL1_upper_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_upper = upper_error_MLE.x[0]\n",
    "\n",
    "            lower_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_missense_ASXL1_lower_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_lower = lower_error_MLE.x[0]\n",
    "\n",
    "#             print('a = ', a)\n",
    "            ASXL1_missense_s[var]=(s_value, observed_number, variant_mutation_rate, s_value_upper, s_value_lower) \n",
    "    a = a+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#CALCULATING S FOR NONSENSE VARIANTS\n",
    "ASXL1_nonsense_s = {}\n",
    "a = 0\n",
    "for var in all_nonsense_variants_ASXL1:\n",
    "    initial_guess=[0.05]\n",
    "    if var in all_studies_ASXL1_counts.keys():\n",
    "        observed_number = float(all_studies_ASXL1_counts[var])\n",
    "#         print(observed_number)\n",
    "        if observed_number >1:\n",
    "            variant_mutation_rate = mutation_rates_variants_ASXL1(var)\n",
    "            \n",
    "            MLE = scipy.optimize.minimize(prob_observed_given_expected_nonsense_ASXL1, initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value = MLE.x[0]\n",
    "\n",
    "            upper_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_nonsense_ASXL1_upper_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_upper = upper_error_MLE.x[0]\n",
    "\n",
    "            lower_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_nonsense_ASXL1_lower_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_lower = lower_error_MLE.x[0]\n",
    "\n",
    "#             print('a = ', a)\n",
    "            ASXL1_nonsense_s[var]=(s_value, observed_number, variant_mutation_rate, s_value_upper, s_value_lower) \n",
    "    a = a+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#Exac - retrieve the frequencies in Exac for each of the variants\n",
    "import csv\n",
    "filename=\"Data files/ASXL1_Exac.csv\"\n",
    "\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    Exac_ASXL1_results={} \n",
    "    \n",
    "    for row in read_reader:\n",
    "#         print(row)\n",
    "        if row_count>1:\n",
    "            variant = row[6]\n",
    "            allele_freq = float(row[14])\n",
    "            african_freq = float(row[15])/float(row[16])\n",
    "            east_asian_freq = float(row[18])/float(row[19])\n",
    "            european_freq = float(row[21])/float(row[22])\n",
    "            finnish_freq = float(row[24])/float(row[25])\n",
    "            latino_freq = float(row[27])/float(row[28])\n",
    "            south_asian_freq = float(row[33])/float(row[34])\n",
    "            Exac_ASXL1_results[variant]=(allele_freq, african_freq, east_asian_freq, european_freq, finnish_freq, latino_freq, south_asian_freq)\n",
    "\n",
    "        row_count=row_count+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#COSMIC - retrieve the frequencies in COSMIC for each of the variants\n",
    "import csv\n",
    "filename=\"Data files/ASXL1_COSMIC_variant_counts.csv\"\n",
    "\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    ASXL1_COSMIC_results={} \n",
    "    \n",
    "    for row in read_reader:\n",
    "#         print(row)\n",
    "        if row_count>1:\n",
    "            variant = row[1]\n",
    "            counts = int(row[2])\n",
    "            ASXL1_COSMIC_results[variant]=counts\n",
    "\n",
    "        row_count=row_count+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Create dictionary of results and save to .csv file\n",
    "ASXL1_s_dictionary={}\n",
    "for k, v in ASXL1_missense_s.items():\n",
    "#     print(k)\n",
    "    if k in Exac_ASXL1_results.keys():\n",
    "        exac_freq = round((Exac_ASXL1_results[k][0])*100, 4)\n",
    "        african_freq = round((Exac_ASXL1_results[k][1])*100, 4)\n",
    "        east_asian_freq = round((Exac_ASXL1_results[k][2])*100, 4)\n",
    "        european_freq = round((Exac_ASXL1_results[k][3])*100, 4)\n",
    "        finnish_freq = round((Exac_ASXL1_results[k][4])*100, 4)\n",
    "        latino_freq = round((Exac_ASXL1_results[k][5])*100, 4)\n",
    "        south_asian_freq = round((Exac_ASXL1_results[k][6])*100, 4)\n",
    "    if k not in Exac_ASXL1_results.keys():\n",
    "        exac_freq = '-'\n",
    "        african_freq = '-'\n",
    "        east_asian_freq = '-'\n",
    "        european_freq = '-'\n",
    "        finnish_freq = '-'\n",
    "        latino_freq = '-'\n",
    "        south_asian_freq = '-'\n",
    "    if k in ASXL1_COSMIC_results.keys():\n",
    "        COSMIC_number = ASXL1_COSMIC_results[k]\n",
    "    if k not in ASXL1_COSMIC_results.keys():\n",
    "        COSMIC_number = '-'\n",
    "    ASXL1_s_dictionary[k]=(round(v[0]*100, 2),round(v[4]*100, 2),round(v[3]*100, 2), v[1], v[2], COSMIC_number, exac_freq, \\\n",
    "                            african_freq, east_asian_freq, european_freq, finnish_freq, latino_freq, south_asian_freq)\n",
    "\n",
    "for k, v in ASXL1_nonsense_s.items():\n",
    "    if k in Exac_ASXL1_results.keys():\n",
    "        exac_freq = round((Exac_ASXL1_results[k][0])*100, 4)\n",
    "        african_freq = round((Exac_ASXL1_results[k][1])*100, 4)\n",
    "        east_asian_freq = round((Exac_ASXL1_results[k][2])*100, 4)\n",
    "        european_freq = round((Exac_ASXL1_results[k][3])*100, 4)\n",
    "        finnish_freq = round((Exac_ASXL1_results[k][4])*100, 4)\n",
    "        latino_freq = round((Exac_ASXL1_results[k][5])*100, 4)\n",
    "        south_asian_freq = round((Exac_ASXL1_results[k][6])*100, 4)\n",
    "    if k not in Exac_ASXL1_results.keys():\n",
    "        exac_freq = '-'\n",
    "        african_freq = '-'\n",
    "        east_asian_freq = '-'\n",
    "        european_freq = '-'\n",
    "        finnish_freq = '-'\n",
    "        latino_freq = '-'\n",
    "        south_asian_freq = '-'\n",
    "    if k in ASXL1_COSMIC_results.keys():\n",
    "        COSMIC_number = ASXL1_COSMIC_results[k]\n",
    "    if k not in ASXL1_COSMIC_results.keys():\n",
    "        COSMIC_number = '-'\n",
    "    ASXL1_s_dictionary[k]=(round(v[0]*100, 2),round(v[4]*100, 2),round(v[3]*100, 2), v[1], v[2], COSMIC_number, exac_freq, \\\n",
    "                            african_freq, east_asian_freq, european_freq, finnish_freq, latino_freq, south_asian_freq)\n",
    "\n",
    "ASXL1_df = pd.DataFrame.from_dict(ASXL1_s_dictionary, orient='index', columns=['s (%)', 's lower value (%)', 's upper value (%)','observed number', 'site-specific mutation rate', 'Number of times in COSMIC',\\\n",
    "                                                                                'Exac Overall freq (%)', 'Exac African freq (%)', 'Exac East Asian freq (%)',\\\n",
    "                                                                                'Exac European freq (%)', 'Exac Finnish freq (%)', 'Exac Latino freq (%)', 'Exac South Asian freq (%)'])\n",
    "ASXL1_df_sorted = ASXL1_df.sort_values(by = ['s (%)'], ascending = False)\n",
    "\n",
    "ASXL1_df_sorted.to_csv('Data files/ASXL1_s_across_whole_gene.csv')\n",
    "# ASXL1_s_dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s (%)</th>\n",
       "      <th>s lower value (%)</th>\n",
       "      <th>s upper value (%)</th>\n",
       "      <th>observed number</th>\n",
       "      <th>site-specific mutation rate</th>\n",
       "      <th>Number of times in COSMIC</th>\n",
       "      <th>Exac Overall freq (%)</th>\n",
       "      <th>Exac African freq (%)</th>\n",
       "      <th>Exac East Asian freq (%)</th>\n",
       "      <th>Exac European freq (%)</th>\n",
       "      <th>Exac Finnish freq (%)</th>\n",
       "      <th>Exac Latino freq (%)</th>\n",
       "      <th>Exac South Asian freq (%)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>p.I1220F</th>\n",
       "      <td>17.28</td>\n",
       "      <td>13.85</td>\n",
       "      <td>19.80</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.042010e-10</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0099</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0181</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.C856*</th>\n",
       "      <td>12.43</td>\n",
       "      <td>10.13</td>\n",
       "      <td>13.73</td>\n",
       "      <td>2.0</td>\n",
       "      <td>6.414520e-10</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.S392*</th>\n",
       "      <td>11.95</td>\n",
       "      <td>9.84</td>\n",
       "      <td>13.14</td>\n",
       "      <td>2.0</td>\n",
       "      <td>8.230880e-10</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.Y591*</th>\n",
       "      <td>11.07</td>\n",
       "      <td>9.34</td>\n",
       "      <td>12.10</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.431380e-09</td>\n",
       "      <td>23</td>\n",
       "      <td>0.0025</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0046</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.Q588*</th>\n",
       "      <td>10.99</td>\n",
       "      <td>8.59</td>\n",
       "      <td>12.20</td>\n",
       "      <td>2.0</td>\n",
       "      <td>9.919690e-10</td>\n",
       "      <td>11</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.Q708*</th>\n",
       "      <td>10.30</td>\n",
       "      <td>8.52</td>\n",
       "      <td>11.24</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.964580e-09</td>\n",
       "      <td>8</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.Q768*</th>\n",
       "      <td>10.30</td>\n",
       "      <td>8.52</td>\n",
       "      <td>11.24</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.964580e-09</td>\n",
       "      <td>3</td>\n",
       "      <td>0.0017</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.Q803*</th>\n",
       "      <td>10.08</td>\n",
       "      <td>8.45</td>\n",
       "      <td>11.01</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.322790e-09</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0025</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0174</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.W1037*</th>\n",
       "      <td>9.41</td>\n",
       "      <td>7.80</td>\n",
       "      <td>10.22</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.524850e-09</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.Q432*</th>\n",
       "      <td>8.80</td>\n",
       "      <td>6.73</td>\n",
       "      <td>9.77</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.211140e-09</td>\n",
       "      <td>-</td>\n",
       "      <td>0.0025</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0045</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.W583*</th>\n",
       "      <td>8.04</td>\n",
       "      <td>5.93</td>\n",
       "      <td>9.01</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.610690e-09</td>\n",
       "      <td>3</td>\n",
       "      <td>0.0017</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R404*</th>\n",
       "      <td>7.91</td>\n",
       "      <td>7.13</td>\n",
       "      <td>8.46</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.882290e-08</td>\n",
       "      <td>6</td>\n",
       "      <td>0.0058</td>\n",
       "      <td>0.0096</td>\n",
       "      <td>0.0116</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.0151</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R693*</th>\n",
       "      <td>7.77</td>\n",
       "      <td>6.95</td>\n",
       "      <td>8.34</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.882290e-08</td>\n",
       "      <td>75</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.0098</td>\n",
       "      <td>0.0233</td>\n",
       "      <td>0.0046</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R417*</th>\n",
       "      <td>6.98</td>\n",
       "      <td>5.61</td>\n",
       "      <td>7.73</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.415120e-08</td>\n",
       "      <td>8</td>\n",
       "      <td>0.0074</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0116</td>\n",
       "      <td>0.009</td>\n",
       "      <td>0.0302</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          s (%)  s lower value (%)  s upper value (%)  observed number  \\\n",
       "p.I1220F  17.28              13.85              19.80              3.0   \n",
       "p.C856*   12.43              10.13              13.73              2.0   \n",
       "p.S392*   11.95               9.84              13.14              2.0   \n",
       "p.Y591*   11.07               9.34              12.10              3.0   \n",
       "p.Q588*   10.99               8.59              12.20              2.0   \n",
       "p.Q708*   10.30               8.52              11.24              2.0   \n",
       "p.Q768*   10.30               8.52              11.24              2.0   \n",
       "p.Q803*   10.08               8.45              11.01              3.0   \n",
       "p.W1037*   9.41               7.80              10.22              2.0   \n",
       "p.Q432*    8.80               6.73               9.77              2.0   \n",
       "p.W583*    8.04               5.93               9.01              2.0   \n",
       "p.R404*    7.91               7.13               8.46              7.0   \n",
       "p.R693*    7.77               6.95               8.34              7.0   \n",
       "p.R417*    6.98               5.61               7.73              3.0   \n",
       "\n",
       "          site-specific mutation rate Number of times in COSMIC  \\\n",
       "p.I1220F                 5.042010e-10                         1   \n",
       "p.C856*                  6.414520e-10                         1   \n",
       "p.S392*                  8.230880e-10                         -   \n",
       "p.Y591*                  1.431380e-09                        23   \n",
       "p.Q588*                  9.919690e-10                        11   \n",
       "p.Q708*                  1.964580e-09                         8   \n",
       "p.Q768*                  1.964580e-09                         3   \n",
       "p.Q803*                  2.322790e-09                         2   \n",
       "p.W1037*                 3.524850e-09                         1   \n",
       "p.Q432*                  3.211140e-09                         -   \n",
       "p.W583*                  4.610690e-09                         3   \n",
       "p.R404*                  1.882290e-08                         6   \n",
       "p.R693*                  1.882290e-08                        75   \n",
       "p.R417*                  1.415120e-08                         8   \n",
       "\n",
       "         Exac Overall freq (%) Exac African freq (%) Exac East Asian freq (%)  \\\n",
       "p.I1220F                0.0099                     0                        0   \n",
       "p.C856*                 0.0008                     0                        0   \n",
       "p.S392*                      -                     -                        -   \n",
       "p.Y591*                 0.0025                     0                        0   \n",
       "p.Q588*                      -                     -                        -   \n",
       "p.Q708*                      -                     -                        -   \n",
       "p.Q768*                 0.0017                     0                        0   \n",
       "p.Q803*                 0.0025                     0                        0   \n",
       "p.W1037*                     -                     -                        -   \n",
       "p.Q432*                 0.0025                     0                        0   \n",
       "p.W583*                 0.0017                     0                        0   \n",
       "p.R404*                 0.0058                0.0096                   0.0116   \n",
       "p.R693*                  0.005                0.0098                   0.0233   \n",
       "p.R417*                 0.0074                     0                   0.0116   \n",
       "\n",
       "         Exac European freq (%) Exac Finnish freq (%) Exac Latino freq (%)  \\\n",
       "p.I1220F                 0.0181                     0                    0   \n",
       "p.C856*                  0.0015                     0                    0   \n",
       "p.S392*                       -                     -                    -   \n",
       "p.Y591*                  0.0046                     0                    0   \n",
       "p.Q588*                       -                     -                    -   \n",
       "p.Q708*                       -                     -                    -   \n",
       "p.Q768*                   0.003                     0                    0   \n",
       "p.Q803*                  0.0015                     0               0.0174   \n",
       "p.W1037*                      -                     -                    -   \n",
       "p.Q432*                  0.0045                     0                    0   \n",
       "p.W583*                   0.003                     0                    0   \n",
       "p.R404*                   0.006                0.0151                    0   \n",
       "p.R693*                  0.0046                     0                    0   \n",
       "p.R417*                   0.009                0.0302                    0   \n",
       "\n",
       "         Exac South Asian freq (%)  \n",
       "p.I1220F                         0  \n",
       "p.C856*                          0  \n",
       "p.S392*                          -  \n",
       "p.Y591*                          0  \n",
       "p.Q588*                          -  \n",
       "p.Q708*                          -  \n",
       "p.Q768*                          0  \n",
       "p.Q803*                          0  \n",
       "p.W1037*                         -  \n",
       "p.Q432*                          0  \n",
       "p.W583*                          0  \n",
       "p.R404*                          0  \n",
       "p.R693*                          0  \n",
       "p.R417*                          0  "
      ]
     },
     "execution_count": 116,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ASXL1_df_sorted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABZ8AAAHvCAYAAAA7GaHXAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAIABJREFUeJzs3Xd4VFX+x/H3SUJCAqRQQocACigKShdkCVVAiiIIwq4gruwPVpG1664GRFdcCxKpi6KoWNYGikpRQJAFJbTQe+gEQuihpJzfH3dmnMlMQjFAWD+v58kzybnnnvude+beB745+V5jrUVEREREREREREREpCAFXekAREREREREREREROR/j5LPIiIiIiIiIiIiIlLglHwWERERERERERERkQKn5LOIiIiIiIiIiIiIFDgln0VERERERERERESkwCn5LCIiIiIiIiIiIiIFTslnERERkauAMSbEGGONMW9d6VgKI2PMQ8aYjcaYs67zVMnV3sYY87Mx5oSr/Y9XOlYRERERkd8LJZ9FREREriBjTLwrKZrXV9N89g0yxgwzxnS9nDEXNsaYdkAisAYYCPwJSDfGlAa+AIoCj7jaF12iGOq75qLKpRhfPL+A2e+6Lp46R79+xphFxphUY8xpY8xuY8xcY8xwY0yoV9//GGNyjDEtA4xTzBizxRhz0BgT6zW2NcZMO494extj3jHGJBtjsrx/KSIiIiLyexFypQMQEREREQA+Ar4N0L4FwFqbZYwJB7K8tgUBCcDbwFeXPMLCq53r9T5r7TF3ozEmHogEnrXWXurzUx9nLr4Hdl7iY/1edQbKAluBAcDIPPp9AnQHFgKvAEeAKkAT4O/AKOCsq+9goCXwjjGmrrX2hNc4/wJqAD2ttQcuIt4HcT4Xq4BtwLUXMYaIiIjIVU3JZxEREZHCYbm19oP8OlhrT1+uYK4y5YBs78SzVztA+mWO56pjjDFAsVzJ18LmfmAT8DTwuTGmhbV2oXcHY0wTnMTzZ9banrkHMMaUAzzv0VqbZowZBHyOk2we7OrXBhgEfGyt/ewi4+0L7HH94mgCSj6LiIjI75DKboiIiIhcBXLXfDbGXANkujbf71WmIyt3f2PMrcaYhcaYDGNMmjHm38aYYgGOUdEYM8EYs8tVO3mP6+fSufqVMsaMNsZsc5U0OGSMSTLGPJKrX39jzFJjzBFjzEljzFZjzAfGmFLn+Z5vM8bMMcYcdR1nlTFmoNf2a4wxFqecRrDXOfjeGLMbZ0U4wELvc+Pat6gx5h/GmHWusQ8bY74yxtQLEIcxxvzFGPOLq3b0cVcphQTX9heASbmO5T1X4caY541TkzrDdazVxpi8Vu56HzvEFedCVwmJs8aYHcaYscaYknns09MY86PrvGUYYza45ivEtb2tK74/GadW9nrgDDDUa4x413k85hpjmTGmf4Bj3WiM+dz1WTljjNlnnPIWHb36XPT79xqjPNARmALMAA7hJKNzcyd4fwg0jrV2v7U2K1fbFzh/efB/xqkRHglMBlJxVi9fFGvtjtzHEhEREfm90cpnERERkcIhIneSFzhjrT2eR//9QD+cZNx8fk205uTq1wC407X9A6A18ABO+Y7B7k7GmGrAf4FgV193mYBBQCtjTCOvlcVfALcAE4BkoBhwHRAPvO4a7z6cBN6PwHPAKZzSB52A0jjJwzwZZzXqWFdMI4AM4DZgojGmmrX2adc5+JMrxqau8wGwD6fcxu04CcoROCtmc1xjhwKzccowvIdTLzrGdV7+a4y51Vq7wtXX4CQmewGLgRdxyjhcB9wFDAc+xVll7X0scJVMcZ2ne4F3XWMUcZ3b1vmdA5eiwKM4K3OnASeBxji1rZu75sX9SwiMMS8DTwBrgddwEqg1XLH+Ha9Vv65xY3Dmez+wwzXGHcBnrvP4iuuY9+CUpqhmrXUn3csAc4Fs13vciTO3jVwxflcA79+tP2CA9621Z40xHwIDjDEP5bpGtrpe7zbGfGytPXKe4z+I8/mdjPOZqwJ0s9bm+zkVERERkfwZa+2VjkFERETkd8s4dYnn5bH5E2ttb1e/EJyVzm9ba/+cV5vXuO5tOUATa22S17ZZOIm2aGvtKVfbNziJ6vrW2r1efZvgJOMSrLUvuFbbHgLetNYOyed9fQXcCpSx1maf5+lw71sJJ4n4ibX23lzbxgJ/AWpYa93J0g+A3tbakFx9/4yzIrmFtfYnr/bHgZeB9tba773ao3EeWrjBWtvW1dYHmIqT5B9grc3x6h/k/jmvY7m2HQPmW2sv+MGQxpggIMw9T17tf8FJ6t7lWrmLMaYZzgMVvwc6W2vPePU31vUPf2NMW2AOzjzWttamefUrAmwHIoDrrbX7Xe1hOL9IaARca63dZozpjpMU98SQx3u46PfvNcYmYKfXvDQAkoCB1tpJufp+i7NK+iTOZ/dn19dca21GPsfowq+109+z1vYL0Md9XU231t5xAfFPwPncVrbW7j7f/URERESudiq7ISIiIlI4/BvnwXneXy8UwLg/eSeeXeYCoUBVAFdCuSPOytqzxpjS7i+cJPB2oL1r3wyc5FtTY0zVfI57FCgBdHKtHr4QPV3xTfaOxRXP1zirs9tc4Jje/oizMnhlrrFDcMo1tHQlW8Gp22uBx7wTzwC5f87HUeBGY0ydCw3UWpvj9QuCYGNMtCvWua4uTby693W9PuWdeHaNE2jFybveiWeXRkBF4C134tm1/xngVZz/P7iTyEddr52MMSXyeRsX/f4BjDF/wFkpPcUrnmXAapwHD+Z2B/AwsA5ndfU/cD43+4wxQwP0dzvEr385MONiYhURERERX0o+i4iIiBQOm6213+f6WlMA424L0OYuJeCuvVwbp6TBX4CDAb5qAGXB89DDR4CbgBRjzBpjTKIxplWuY7wA7MZZSXrAGPOZMeZ+Y0zx84j5OtfrvACxuEs5lD2PcfJSG7ghwNgHccpDhADuesrXArsDJGkvxMM45SjWGGO2GGMmGWO6nG9S3hjT2xizFKd0yWFXnO7SHjFeXa/FKYGRfJ5xbQrQVs31ujbANvfnsbrrdS7wIU65kUPGmEXGmARjTO1c+/2m9+8a/yywyjh1vq8xTs3z2Ti/BLneu7O19qy1NtFa2xin/MofcFa6BwOjjDGBHkQYgVMWJB3Y4+oXfZ7xiYiIiEgeVPNZRERE5H9bfiUvTK7XKTh1oQPxlCuw1o4xxnyJU1P5D8DdwEPGmKnW2j+6+mx0JSHb4qxSbgm8BQwzxvzBWrv9POLqCxzIo8+WPNrPRxCwEng8nz7pXrH8pjp11tovjDE/4tS7bomzqv3PwHxjTHvvms25GWPuxqk5vQQYgpPQP42zMvwbfBeTXOgK80AlKM57DNdq6r6uBwd2BFrg1Jt+1lWLebyr3295/5FAD5z3uyqPbgOAx/KIMQNYiPMgyB+Bb3GS2Z/m6joSJ3nfCzju6jcKuC/fkyAiIiIi+VLyWUREROTqVVAP79jiGquIdw3kfA9s7R6cUiH/dtXBnYqTiHzN/bA+V6mGb1xfGGO6AtOBoTirYfOy2fV68HzjuUCbgTLAD3mUo/C2EaesROlzrH7OdxzXg+veB953rfh9BeeBf52BL/PZ9U84SeJWrlXnABhjbsgj1rbAjcDy/OLJh/uBfYFKZLhXGPusprfWrsYpgfEvVwmXX3CSueO9+lzs+++NU3/6ydzHdfkb8CdjzNP5JbFdlrheK3o3uuquPwh8aq39j6vtHeA+Y8yn1tpvzzGuiIiIiORBZTdERERErlKuB/md5tcSERc7TipOCYOexphGubcbRxnX9xHGmPBc+2fhJB9xx+KqS5zbcu8++fgEp8zC88aYogHiiTbGhJ5jjPy8h5OADJgAN8Z4l/SYirMa+F+5y0Tk+vmE67Vkrj4hxpgo7zZXwntloP4BZOPUIfb8u9113H8E6Puh6/Wl3OfnAkpcLMUpO3G/MSbWa/9QnNXFObgeymeMKZl7XGttOpACFDfGhBbA+78fSANes9Z+lvsLmAzE4iSxMcbUMsZUz2Ms9wMC13m9r+LAOzilTAZ79f0bzirzf6v8hoiIiMjF08pnERERkavbEuA2Y8wTwC4g27168wL9BfgJ+MkY8x6wAuffitVwknZv49Rxvh743lV2Yy1ODeLrgUE4q2YXucaba4w5iFPyYBdOkvE+nORlXqU9ALDW7jDGPAhMANYZYz4AduKsVq4LdANq4iQHL8brOCuERxlj2gLzcUotVMEpEXIcpzQEwMdAd1fsNY0xX+M8QK+mq289V79fcFY//8OVqD/Jrw9r3GmM+Qon4XoAp2byIJzSHt+cI9bPXO93rjHmfSAMuBPwS8pba/9rjHkNZ0VxkjHmU2A/zhz2NMbcbK09kXu/XGNkGWMewilLsdQYM8n1XnoDjYHnrbXuFcgDgAddn4UtQBbQynVePrTWnnX9EuKi3r9rdXdjnIcf5lU+ZjowESdJ/SVwMzDVGDMPWIDzGSkONMUp33EUGOG1/2tAHHCX98p2a+1RY8xA8i6/UdMYE+gXAOAkyk8ZY1rilCIBqO96HWKMOQbkWGv/mdd7FxEREflfoeSziIiIyNXt/4AxOCthS+CslL3g5LMr4VsfeAroivPgvQycxPE0nCQowA6c2tDxOEnQMJwE30TgZa/SEGOBnq74SuI85HA5MNhaO/884plkjFmPs9p2EBCFswJ2A/B3nJWqF8WVFO2AU2rhj8Bw16a9OMn8KV59rTGmF/BXnGRrAs453oZX3WBr7XZjzJ9xah6PB4rgJOz/CiTiJGTbAcWAfTiJ0pestfvPEesHrtW5D+MkStNxVh4/S4B62Nbax4wxK1zHfRJn1fYuYAbOKvnzOT9fGmPa4XymnsT5P8N6YIC19h2vrnNxfhnQBSiPk3zejpP8HuPqc/w3vP/7Xa9f5BNrmquWcwdjTAWch1Q+ifPLhftxVkWD8xmdDPzLnTw3xrQHBgIfWWv9jmGt/c4YMxkYEKD8xnX4JrG9TcB5OGQ7nM+qN3ed8WxAyWcRERH5n2fOXebuyjDGlML5D83tOHXrKuL8+eVqnD+Ne8dam+PVPw7nH7t5+cRa2/tSxSsiIiIiIiIiIiIivyrMK5974qwa2YezgmEnUBbnzx7fAjoaY3oGeEjMKpzVObmtuYSxioiIiIiIiIiIiIiXwrzyuTXOn+V9k2uFczmcmnqVgR7W2s9d7XE4K5+nWGv7X+54RURERERERERERORXQefucmVYa+daa7/2Tjy72vfj1FEDp9agiIiIiIiIiIiIiBQyhbnsRn4yXa9ZAbZVMMb8BSiF82Cbxdba5MsWmYiIiIiIiIiIiIhcfclnY0wIztPXAWYG6NLO9eW9z3ygn7V257nGHz9+vAUYNGiQ+W2RioiIiIiIiIiIiPx+FdqyG/kYCdwAfGutneXVngGMABoAMa6vljgPK4wHfjDGFMtrUGPMQGNMkvvno0eP2hUrVtjMzEyblJRkAZuUlGTPnDljV61aZQ8fPmy3bNlid+/ebfft22c3btxojx07ZlesWGGzsrL89lm9erVNT0+3W7dutbt27bL79++3GzZssMePH7fLli2z2dnZPvucOnXKrlmzxh46dMhu27bN7ty506amptr169fbEydO2KSkJJuTk+OzT0ZGhl27dq09ePCg3b59u92xY4c9cOCAXbdunT158qSn79KlSz37nDx50q5bt84eOHDA7tixw27fvt0ePHjQrl271mZkZPiM7z7eiRMn7Pr1621qaqrduXOn3bZtmz106JBds2aNPXXqlM8+2dnZdtmyZfb48eN2w4YNdv/+/XbXrl1269atNj093a5evdqeOXPGZ5+srCy7YsUKe+zYMbtx40a7b98+u3v3brtlyxZ7+PBhu2rVKr99MjMz7YoVK+zRo0ftpk2b7N69e+2ePXvs5s2b7ZEjR+zKlSttZmamz3s/e/asZy43b95s9+zZY/fu3Ws3bdpkveffvc/SpUvtmTNnbHJysmcuc8//8uXL/eby9OnTdvXq1T5zmXv+c8+le/7T0tI8c5mammrXrVvnmX9rbZ7zn5KSYlNSUuzBgwf95t/79cSJE37zn5aWZtesWWMzMjJ8zldOTo5dtmxZvvN/+vRpn89Zdna2Xb58uT127FjA+U9OTs5z/o8ePWo3btxo9+7d6zf/Z8+e9Zv/lStX2iNHjthNmzbZPXv22D179thNmzb5zL/3Pt7z776W9+7dazdu3OiZ/0DXsvf8e1/L3vPvfd5Onz6d77Wce/6XLl1qMzIy/ObffS275z/3XLqv5dzzH+hadn923POfmprqN/+5r2X3/Htfy97zv3r1ap/5d1//7vn3vpZzz7/3NZbftRxo/t37uOc/v2tZ93Ldy3Uv171c93Ldy3Uv171c93Ldy3Uv171c93Ldy3+P93KuoEL7wMFAjDFDgNHABqC5tTb9PPYJAX4CmgBDrbWj8+vvtfL5twcsIiIiIiIiIiIicmVdsQoPV83KZ2PMX3ESz+uAVueTeAaw1mYBb7l+/MMlCk9EREREREREREREvFwVyWdjzFBgDLAGJ/G8/wKHOOh6zbPshoiIiIiIiIiIiIgUnEKffDbGPAmMAlbiJJ4PXMQwTV2v2wosMBERERERERERERHJU6FOPhtjnsV5wOAyoI21Ni2fvk2MMaEB2lsDf3P9+MElCVREREREREREREREfIRc6QDyYozpBzwPZAMLgSHG+NXGTrHWvuv6/mWgjjFmPrDb1VYXaO36/llr7X8vZcwiIiIiIiIiIiIi4ii0yWegmus1GBiaR58fgXdd378P3Ak0AjoCRYBU4D/AGGvtwksWqYiIiIiIiIiIiIj4KLTJZ2vtMGDYBfR/G3j7UsUjIiIiIiIiIiIiIuevUNd8FhEREREREREREZGrk5LPIiIiIiIiIiIiIlLglHwWERERERERERERkQKn5LOIiIiIiIiIiIiIFDgln0VERERERERERESkwCn5LCIiIiIiIiIiIiIFTslnERERERERERERESlwSj6LiIiIiIiIiIiISIFT8llERERERERERERECpySz1KoDBs2DGMM8+fPv9KhiIiIiIiIiIiIyG+g5LOIiIiIiIiIiIiIFDgln6VQefDBB1m/fj2NGze+0qGIiIiIiIiIiIjIbxBypQMQ8Va6dGlKly59pcMQERERERERERGR30grn+U3S0lJwRhD//792bp1Kz169KBUqVKUKFGC9u3bs2bNGgAOHjzIwIEDKV++PEWLFqVRo0bMmzfPZ6y8aj4vXLiQLl26UKlSJcLCwihXrhxNmzZl+PDhPv1SU1N57LHHqFWrFsWKFSM6OppatWrRv39/tm3b5hf7rFmz6NSpE6VLlyYsLIwaNWrw+OOPc+TIEb++cXFxxMXFkZGRweOPP06VKlUICwvjmmuu4eWXX8Za67fPV199RZs2bShfvjxhYWFUqFCBli1bMm7cOL++6enpPP3001x33XWEh4cTFRVFmzZtmD179jnnQEREREREREREpLDRymcpMCkpKTRp0oTrrruO/v37k5KSwpdffkl8fDyLFy+mQ4cOREZG0qtXL9LT0/n444/p2LEjmzZtokqVKnmOO3PmTG6//XYiIyPp2rUrFStWJD09nfXr1zNu3DgSEhIAyMjIoHnz5mzdupV27drRpUsXrLXs2LGD6dOn06NHD6pXr+4Z9/nnnychIYGSJUvSuXNnYmNjSU5O5tVXX+Xbb79l8eLFREZG+sSSmZlJ+/bt2bt3Lx07diQkJIRp06bx1FNPcfr0aU8sAP/+97/5y1/+Qrly5ejSpQulS5fmwIEDJCcn88477zB48GBP3x07dhAfH09KSgotWrSgQ4cOnDx5khkzZtChQwcmTpzIAw88UFBTJSIiIiIiIiIicskp+SwF5scff+SFF17g73//u6dtxIgRPPfcczRp0oS7776bcePGERTkLLhv164d9957L6NGjWLUqFF5jjtp0iRycnKYP38+9erV89mWlpbm+f6HH35g69atDB061G+8s2fPcubMGc/P8+bNIyEhgVtuuYVvv/2W6Ohoz7Z3332X++67j4SEBL9x9u7dS7169ZgzZw7h4eEAJCQkULNmTUaNGsUzzzxDkSJFAJg4cSKhoaGsWrWK2NjYPOMG6NevHzt27OCjjz6id+/envYjR44QHx/PkCFD6Nq1K2XLls3zPImIiIiIiIiIiBQmSj5fIn2WZF7pEM7bh02LFMg4cXFxPPXUUz5t/fr147nnnuPMmTO88sornsQzQJ8+fRgwYAArV648r/HdyV5vgepDB+oXGhpKaGio5+fExETASWx7J54B+vfvz+jRo5k6dWrApHhiYqLPMWJjY+nWrRvvvfceGzdu5IYbbvBsCwkJ8SSj84p71apV/Pjjj/To0cMn8QwQHR3N8OHDueOOO/j88899VkuLiIiIiIiIiIgUZko+S4G56aabCA4O9mmrUKECADVr1qREiRI+24KDgylbtiy7d+/Od9y+ffvyxRdf0KRJE3r16kWrVq1o3rw5lSpV8unXsmVLKlasyMiRI1m+fDmdOnWiefPmAeNavHgxRYoU4dNPP+XTTz/1O+bZs2c5ePAghw4dolSpUp72qKgorrnmGr/+lStXBuDw4cM+cT/66KPUqVOHXr160bJlS5o3b06ZMmX8YgE4evQow4YN8xv74MGDAKxfvz6/0yQiIiIiIiIiIlKoKPksBSYqKsqvLSQkJM9t7u2ZmfmvEu/evTszZszgtddeY/LkyUycOBGABg0a8NJLL9GuXTsAIiMjWbJkCQkJCXz11VfMmjULcFYZDx48mH/84x+eVciHDh0iKyvL74GFuZ04ccIn+Zx7lXTu95mdne1pe+SRRyhdujTjxo0jMTGRN954A2MMLVu25JVXXqFhw4aeWADmzJnDnDlz8o1FRERERERERETkaqHk8yVSUKUsxHH77bdz++23c/LkSX7++WdmzJjB+PHj6dy5MytWrOD6668HoFKlSrz99ttYa1m3bh1z585l7NixPP/88+Tk5DBixAjASYbn5OSQnp5+SeO+9957uffeezly5Aj//e9/+fLLL5k8eTK33XYb69evJzY21pOYHz16NEOGDLmk8YiIiIiIiIiIiFwuQefuIlJ4FCtWjNatW/P666/zzDPPcPbsWb777ju/fsYY6tSpw0MPPeRZTTxt2jTP9qZNm3L48GHWrl17WeKOjo6mU6dOTJo0if79+5Oens7ChQs9sQCen0VERERERERERP4XKPkshd4PP/zAqVOn/NpTU1MBiIiIAGDNmjWkpKScsx/A3/72NwAeeOAB9u7d67fPyZMnWbJkyW+Ke+bMmWRlZfm1HzhwwCeehg0b0qJFC7744gsmT54ccKzVq1d79hMREREREREREbkaqOyGFHqPPvooKSkpxMfHExcXR2hoKMuWLWPu3LlUrVqV3r17A/D999/zyCOP0KxZM2rXrk1sbCy7d+9m+vTpBAUF8fjjj3vGbNOmDSNHjuTpp5/m2muvpVOnTlSrVo0TJ06wY8cOfvzxR2699VZmzpx50XH37t2bokWLcuuttxIXF4e1loULF7J06VIaNGhA27ZtPX0//PBDWrduzf33309iYiJNmjQhOjqa3bt3k5yczJo1a1i8eDGxsbEXfyJFREREREREREQuIyWfpdB75pln+PLLL0lKSuL7778nKCiIKlWq8MwzzzB06FBiYmIAuO222xg6dCgLFixg+vTpHDt2jPLly9OuXTtPUtrbk08+SfPmzUlMTOSnn35i+vTpREVFUbFiRQYOHEifPn1+U9wjR45k1qxZLF++nG+//ZaiRYtStWpVXn75ZQYNGuR5+CE4taqXLVvGm2++yeeff87UqVPJzs6mXLlyXH/99Tz00EPceOONvykeERERERERERGRy8lYa690DIXK+PHjLcCgQYOudCgiIiIiIiIiIiIiv5W5UgdWzWcRERERERERERERKXBKPouIiIiIiIiIiIhIgVPyWUREREREREREREQKnJLPIiIiIiIiIiIiIlLglHwWERERERERERERkQKn5LOIiIiIiIiIiIiIFDgln0VERERERERERESkwCn5LCIiIiIiIiIiIiIFTslnERERERERERERESlwSj6LiIiIiIiIiIiISIFT8llERERERERERERECpySzyIiIiIiIiIiIiJS4JR8FhEREREREREREZECp+SziIiIiIiIiIiIiBQ4JZ9FREREREREREREpMAp+SwiIiIiIiIiIiIiBU7JZxEREREREREREREpcEo+y2+WkpKCMYb+/fuTkpJC7969KV26NEWLFqVhw4bMmDHDb58zZ84wcuRI6tatS0REBJGRkbRo0YL//Oc/BTL+2bNnSUxMpH79+sTExBAREUFcXBzdunXj+++/9+u/YcMG+vfvT+XKlQkLC6Ns2bL06dOHjRs3+vXt378/xhhSUlKYOHEiN954I0WLFqVs2bIMHDiQo0eP+u2TnJzMPffcQ1xcHGFhYZQpU4b69eszdOhQMjMzffpmZWUxbtw4mjZtSmRkJBEREdx8882MGTOGnJycfOdCRERERERERESksDDW2isdQ6Eyfvx4CzBo0KArHcpVIyUlhWrVqhEfH8/atWupXr06t9xyC+np6XzyySdkZmby/fff06pVK8BJDLdv354ff/yR2rVr07lzZzIyMvjss884cOAATz/9NP/85z8venyAPn368NFHH3HDDTfQunVrwsPD2bt3Lz/99BPdu3fn1Vdf9fSdOXMm3bt3JzMzky5dunDNNdewe/duvvjiC8LCwpg3bx7169f39O/fvz9TpkyhZ8+ezJo1iy5dulC2bFnmzZvHihUraNWqFXPnzvX0T05OpkmTJhhj6Nq1K9WqVePYsWNs2bKFefPmkZ6eTvHixQE8McyaNYtatWoRHx9P0aJFmTdvHsnJyfzxj3/k/fffv2RzKSIiIiIiIiIi/3PMFTuytVZfXl/jxo2z48aNs3L+tm/fbgEL2GHDhvlsmzlzpgVsx44dPW3//Oc/PW2ZmZme9tTUVFu1alUn5sXVAAAgAElEQVQL2EWLFl30+EeOHLHGGNugQQOblZXlF29aWprn+/T0dBsdHW1LlSpl165d69NvzZo1tlixYvbmm2/2ae/Xr58FbOXKle2OHTs87ZmZmbZFixYWsD///LOn/ZFHHrGAnTZtml8s6enpNjs72/NzQkKCBeyDDz7oE3tWVpYdMGBAnuOIiIiIiIiIiIjk4YrlWkOuSMb7d8DckXilQzhvdtqQAhmnatWq/OMf//Bpu+2226hSpQq//PKLp23y5MkYY3j99dcJCfn1IxgbG8uzzz7Ln//8Z9566y2aNWt2UeMbY7DWEhYWRlCQf2WZUqVKeb5/7733OHLkCGPGjOH666/36VenTh0eeOAB3njjDdatW+e3/bnnnqNKlSqen0NCQrjvvvtYuHAhv/zyC40bN/bpHx4e7hdLTEyM5/ucnBzGjBlDuXLlGDVqFMHBwZ5twcHBvPbaa7zzzjtMnTqVbt26+Y0lIiIiIiIiIiJSmCj5LAXmpptu8kmYulWuXJnFixcDcPz4cbZs2ULFihWpXbu2X9/WrVsDsGLFiosaHyAyMpIuXbrw9ddfc9NNN3HXXXfRokULmjRpQkREhM++7v1WrVrFsGHD/MbetGkTAOvXr/dLPjds2DBgLACHDx/2tPXq1YvRo0dzxx130KNHD9q2bUvz5s2pUaOG37EOHTrEtddeywsvvOA3NjgJ7PXr1wfcJiIiIiIiIiIiUpgo+SwFJjo6OmB7SEiI50F57ofxlS9fPmBfd/uRI0cuany3Tz75hJdffpkPP/yQhIQEAIoWLUqPHj149dVXKVu2LACHDh0CYNKkSfm+txMnTpxXPO6V3NnZ2Z62xo0bs3DhQl588UU+++wzT83mWrVqkZCQwD333OMTy+bNmxk+fPgFxSIiIiIiIiIiIlLYKPl8iRRUKYv/NVFRUQDs378/4PZ9+/b59LtY4eHhDBs2jGHDhrFr1y4WLFjAu+++ywcffEBKSgoLFy70Oc6qVauoW7fubzpmfm655RZmzJjBmTNnWLZsGTNnzuTNN9+kT58+lClThrZt23piufPOO/niiy8uWSwiIiIiIiIiIiKXg39BXJFLqESJEtSoUYM9e/awefNmv+3z5s0DoH79+gV2zMqVK9O3b19mzZrFtddey08//eRZZdy0aVMATzL6UgsLC6NZs2Y8//zzJCY6dcGnT58OQO3atYmOjmbJkiVkZmZelnhEREREREREREQuFSWf5bIbMGAA1loef/xxn/IUaWlpjBgxwtPnYh08eJCff/7Zr/3kyZMcP36ckJAQQkNDAbjvvvuIjo5m+PDhPg8tdMvJyWH+/PkXHQs4iW13uRFvqampAJ461CEhITz00EPs27ePIUOGcOrUKb999u3bx7p1635TPCIiIiIiIiIiIpeDym7IZffYY4/x3XffMX36dOrVq0enTp3IyMjg008/5cCBAzzxxBPceuutFz3+nj17aNq0Kddddx3169encuXKHDt2jBkzZrB//36GDBlCiRIlAChVqhSfffYZd955J02bNqVNmzbUqVOHoKAgdu7cyeLFizl06BCnT5++6Hhee+01Zs+eTXx8PNWrV6d48eKsXbuW7777jpiYGAYOHOjp++yzz7Jq1SomTJjA119/TevWralYsSIHDhxg8+bNLFq0iBdffNHv4YciIiIiIiIiIiKFjZLPctmFhoYyZ84cXn/9dT788EPefPNNQkJCqFevHm+88YbnAXwXKy4ujuHDhzN//nzmzZtHWloaJUuWpFatWowcOZLevXv79G/Tpg3Jycm8+uqrzJo1i4ULFxIaGkqFChVo3bo1d91112+KZ/DgwcTExPDzzz+zaNEisrKyqFSpEoMHD+bRRx+latWqnr5FihRh2rRpfPDBB7z77rvMmDGDEydOUKZMGapVq8aIESPo27fvb4pHRERERERERETkcjDW2isdQ6Eyfvx4CzBo0KArHYqIiIiIiIiIiIjIb2Wu1IFV81lERERERERERERECpySzyIiIiIiIiIiIiJS4JR8FhEREREREREREZECp+SziIiIiIiIiIiIiBQ4JZ9FREREREREREREpMAp+SwiIiIiIiIiIiIiBU7JZxEREREREREREREpcEo+i4iIiIiIiIiIiEiBU/JZRERERERERERERAqcks8iIiIiIiIiIiIiUuCUfBYRERERERERERGRAldok8/GmFLGmD8bY740xmwxxpwyxhw1xvxkjLnfGBMwdmNMM2PMt8aYdGNMhjEm2Rgz1BgTfLnfg4iIiIiIiIiIiMjvVciVDiAfPYHxwD5gHrATKAt0B94COhpjelprrXsHY0w34HPgNPAJkA50AUYBzV1jioiIiIiIiIiIiMglVmhXPgObgK5AJWttX2vt09baAUBtYBdwF04iGgBjTCQwCcgG4q2191trHwduAhYDPYwxvS/3mxC5EP3798cYQ0pKypUOpdCbP38+xhiGDRt2pUM5b/Hx8RhjrnQYIiIiIiIiIiKXRaFNPltr51prv7bW5uRq3w9McP0Y77WpB1AG+Nham+TV/zTwD9ePgy5dxL9PKSkpGGPo37//lQ7FT2ZmJl9++SX3338/N9xwA5GRkURERHDjjTfy3HPPcfz48Tz33b17NwMGDKBChQqEhYURFxfH0KFDOXz48GV8ByIFozBfpyIiIiIiIiLyv6swl93IT6brNcurrbXrdWaA/guADKCZMSbMWnvmUgYnhcPWrVvp3r07xYoVo1WrVtx+++2cOHGCWbNmMWLECD755BMWLVpE6dKl/fZr1qwZBw4coFu3btSuXZtffvmF0aNHM3PmTBYtWkSpUqWu0LuSq9l7771HRkbGlQ5DREREREREROSyuOqSz8aYEOBe14/eieZartdNufex1mYZY7YDdYDqwPpLGqQUCiVKlGDs2LH069ePYsWKedrPnj1L9+7d+eabbxg+fDhvvvmmz36DBw/mwIEDJCYm8tBDD3naH3nkEUaNGsXf//53JkyYgMiFqlKlypUOQURERERERETksim0ZTfyMRK4AfjWWjvLqz3K9Xo0j/3c7dGXKrDfm2HDhlGtWjUApkyZgjHG8/Xuu+96+uXk5DBhwgQaNWpE8eLFKVasGI0aNWL8+PHk5OT4jWuMIT4+nr179/KnP/2J2NhYwsPDadCgAR9++OF5x1exYkUGDx7sk3gGCA0N5ZlnngGcusHetm3bxuzZs4mLi+Ovf/2rz7bhw4dTrFgx3n//fU6ePHnecSxbtoy77rqL2NhYwsLCqFq1KoMHD2bfvn157pOTk8Prr79O7dq1KVq0KJUqVeJvf/sbx44d8+ubnJzMPffcQ1xcHGFhYZQpU4b69eszdOhQMjMzffpmZWUxbtw4mjZt6ilDcvPNNzNmzBi/ufAu1bBp0yZ69epFbGwsQUFBzJ8/n9q1axMaGkpaWlrA9zBy5EiMMYwdO9anfffu3Tz44INUr16dsLAwSpUqRdeuXVm6dGnAcVJTU7n//vspW7Ys4eHh3HTTTUyZMiXPcxfISy+9hDGGxMTEgNv37t1LcHAwjRo18ml7/vnnad68OeXKlSM0NJQKFSrQp08f1q/3//3Vuc4XBK75fPbsWcaMGUOnTp2oWrUqYWFhlCxZkrZt2/Ldd98FjDcuLo64uDgyMjJ4/PHHqVKlCmFhYVxzzTW8/PLLeD2H9byvUxERERERERGRghZ8NT2syxgzBHge2AB0HjZs2Cn3tuHDhz8ElALeHDZsWHrufYcPH/5noDLw1rBhw3YHGHvg8OHDJ95+++0VAGrWrMn69espXbo0K1asoEKFCiQlJVGqVCnWrl1LWFgYe/bsISMjg5MnT7J3715CQ0NZt24dZcqUYfny5T77rF+/niJFirB3715OnjzJqVOn2L17N2FhYaxZs4ayZcv67BMTE8PGjRsJDg5m//79HD9+nNOnT7Nr1y7Cw8NJTk6mXLlyLFu2zLNPdHQ0mzZtwhjDgQMHOHbsGGfPnmXHjh1ERESQnJxMhQoVWLp0KRUrViQpKYmoqCi2bNmCtZa0tDQOHz5MVlYWKSkpFC9enFWrVnnGdx8vMjKSrVu3EhwcTPHixVmxYgU33HADd911F+3bt6d27dp06dKF3bt3U7ZsWTp37szo0aOx1tK9e3duuOEGVq5cyccff0xSUhI9evRg5cqVnuNMmjSJ0NBQJk6cyMGDB+ncuTPVq1dnyZIlfPTRRxw9epRWrVr57FOmTBmSk5OJiIhg586dnDlzhmPHjpGamkpwcDAbNmygVKlSLF++nIoVKzJ37lymTZtGTEwMffv2Zffu3Zw+fZpPP/2UOXPmcMcddxAXF0fp0qU9+yQnJ7Np0yY2btxIw4YNCQ0N9Zv/tWvXEhsb65nLUaNGcc8997Bhwwbat2/PrbfeytmzZ/n88895//336dixI7t27fKc219++YVVq1axdetWpkyZQtu2bWnYsCGpqanMmDGDr7/+mrvvvpt169ZRvnx5Pv74Y7p06cL69etp1aoVjRs3pmbNmpw4cYIvv/ySfv36sW/fPiIiIli+fDkDBw4kMTGR4sWL07RpU1q0aMHGjRv54IMPWL16NY0bNyY7O5tt27aRnZ3N2LFjKVGiBCNGjKBIkSI0bNiQRo0aUalSJWJiYvjxxx8JDw+nbt26bNu2jcjISM+83HvvvWRkZDB06FBKlSrF9u3bWb58OW3atGHhwoXUr1+fZs2acc011zBz5kwmTZpEZGQkDRs2ZNWqVURERLBy5Uo6dOjA/Pnzue666+jevTs5OTkkJiayefNmduzYQc2aNenQoQOrV6+maNGi7Ny5k9OnT3P8+HH2799PcHAw1lo+/PBDtm/fzoMPPkhSUhKlS5dmzZo1hIWF8eqrr7Jo0SIGDx5M2bJlKVKkCG+//TYvvvgiFSpUoFWrVtSqVYvSpUszbdo0Jk+eTN26dQkPD/dcy6dOnWL8+PEUL17c53w1adKEcuXKUa5cOd577z327NnDoEGDPNfyggUL+NOf/kSxYsW49dZbqVOnDjVq1GD+/PlMnjyZokWLUrlyZc+1HB4ezhtvvEFOTg4ff/wxK1eu9JzL5ORkpk+fztGjR6lduzaZmZkcOHCAihUrsnTpUurVq0fXrl3p3LkzNWvWpE2bNhw/fpycnBwOHTrE4cOHPfNfokQJv+t/+fLllChRgu3bt5OVlUV6ejqHDh0iJyeHrVu3EhUV5XNdli1blpUrV1KsWDFSUlLIzMzk6NGjHDx4EIDNmzcTHR3NihUrPDHGxsbmeS0HBQWxadMmSpYs6bkvu/dxz/+uXbs4ffo0J06cYN++fYSEhOhefo57eU5Ojs9cBpr/smXLsmLFCp/5P3z4sOeXT1u2bCE6Otpn/mNjY1m1apXf/B84cABjDJs2bSImJuY33ctzX8vue3mg+Xfvs3TpUkqXLs26des8c3mue3lSUhIlS5Zk/fr1hISEeOYy9/znnkv3/AcFBZGamsqxY8c4c+YMO3fu9Mx/+fLl85z/gwcPcuTIETIzM/3mPykpyfMaGRnpN//ua7l48eKsXLnSc77c13J+8+99L1+6dCnlypXzXMuB5t99LQea/4iICHbs2MHZs2f95t/7WnbPf373cvf8e+/jPf/ua/nEiRPs3buXIkWKsH79+oDXsvf8e1/L3vO/bNkyz3krWbIkGzZsyPNaXr16tc/8L126NOD8e9/LA82l+1oGfOY/0LXs/uy451/3ct3LdS/XvVz3ct3LdS/XvVz3ct3L85r/8uXLD7+4bGwBsNZeFV/AXwELrAXKBdi+1LW9QR77r3Ftvy6/44wbN86OGzfOyvnZvn27BWy/fv0Cbv/www8tYG+++WZ7/PhxT/uJEydsgwYNLGCnTp3qs49rnmzPnj1tdna2p33btm02JibGFilSxG7duvU3xf1///d/FrBPPfWUT/tjjz1mAfvqq68G3O+vf/2rBc7rM3L8+HFbqlQpGxQUZBcsWOCzbeTIkRaw7dq182nv16+fBWypUqVsSkqKpz07O9t2797dAvb555/3tD/yyCMWsNOmTfM7fnp6us/5S0hIsIB98MEHbVZWlqc9KyvLDhgwwG8c99wC9umnn/Ybf/fu3TYoKMg2aNDAb9svv/xiAdu9e3dPW2Zmpq1Ro4YNCwuz8+fP9+m/Z88eW6FCBVuuXDl7+vRpT/sDDzxgATt06FCf/kuXLrUhISEWsAkJCX7HD6R9+/YWsKtXr/bbdv3119siRYrYtLQ0T1tqaqo9duyYX9+VK1faYsWK2Q4dOvi0n+t8WWtty5YtrXPb/dXp06ftrl27/PoeOXLE1qlTx8bExNiMjAyfbVWrVrWA7dixo8+21NRUGxUVZaOiouzZs2f9YsvrOhURERERERGR/2lXLKd7VdR8NsYMBUbhJJDbWGsPBOi2EWgI1ASW5do/BKiG84DCbZc2Wpd6PS/LYQrEqk8v2dCTJ08GnBIMxYsX97QXK1aMl19+mbZt2/LWW2/Rp08fn/2Cg4N5+eWXCQr6tTJMtWrVGDJkCMOHD+f9998nISHhomL66quvmDhxIpUqVeKJJ57w2Xb0qFOdJSoqKtCunvYjR46c8zjTp0/n0KFD3HPPPbRo0cJn26OPPsqECROYM2cOO3fu9KsF/PDDD1O1alXPz0FBQbzyyiueVbfPPvusT//w8HC/48fExHi+z8nJYcyYMZQrV45Ro0YRHBzs2RYcHMxrr73GO++8w9SpU+nWrZvPOGXLlg14ritWrEibNm2YM2cOa9eupU6dOp5t7rIY/fr187R98803bN26lccee4yWLVv6jFWhQgWeeOIJhg4dyg8//ECnTp3IzMxk6tSplChRgtx/odGwYUP69u17QeU3+vXrx+zZs5kyZQqvvPKKpz0pKYl169Zx5513+jxIMjY2NuA49erVo3Xr1syePZvMzEyKFCnisz2v85WXsLAwKlWq5NceFRXFgAEDePTRR1m6dCl/+MMf/PokJib6zH1sbCzdunXjvffeY+PGjdxwww3nHYeIiIiIiFxer2zIynf747WvipSNiEi+Cv2dzBjzJE6d55VAO2tt4AKzMBfoC3QAPsq17Q9ABLDAWnvmUsUq/pYvX05QUBDx8fF+21q2bElwcDArVqzw21alShVPnVpv8fHxDB8+POA+5+O///0vffr0oVixYnz++ec+CdrzYV21dHPX7Q1k+fLlALRu3dpvW0hICH/4wx9ISUlhxYoVfsnn3MlZgOrVq1O5cmVSUlI4cuQI0dHR9OrVi9GjR3PHHXfQo0cP2rZtS/PmzalRo4bPvps2beLQoUNce+21vPDCCwHjDQ8PD1jLuF69eoSFhQXcp3///syZM4cpU6bwr3/9C3BqGH/88ceUKVOGTp06efouXrwYgB07dvglk8H58xCA9evX06lTJzZs2EBGRgYtWrQI+MuA+Pj4C0o+33nnnURFRfHBBx8wcuRITwLePUb//v399vnmm2+YMGECSUlJpKWlkZXl+4/DtLQ0ypcv79OW3/nKy9q1a3nllVdYsGAB+/bt4/Tp0z7b9+zZ47dPVFQU11xzjV975cqVATh8+PAFxSAiIiIiIlfO4bO/PrclJvTc/98UEblaFOrkszHmWZwaz8uA9tZav1rOXj4DXgZ6G2PetNYmucYoCrizbeMvZbzi7+jRo5QsWZLQ0FC/bSEhIZQuXZoDB/wXspctWzbgeOXKlfOMe6EWL15Mx44dCQoK4rvvvqNx48Z+fdxJzrzGdz/wL6+V0d7cY+ROTrq52wOtos7v/e/YsYOjR48SHR1N48aNWbhwIS+++CKfffYZ77//PgC1atUiISGBe+65B4BDhw4BToJ3+PC8y/ycOHEi4DHzcueddxIZGckHH3zASy+9RHBwMDNmzODQoUMMHTqUkJBfbzHuGD79NP+V9u4Y3OfvXJ+F8xUeHs7dd9/NpEmTmD17Nh07diQzM5OPPvqIMmXK0LFjR5/+iYmJPPzww8TExNCuXTuqVKlCREQExhimTZvGqlWrOHPG/3dZFxrXkiVLaN26NVlZWbRp04auXbsSGRlJUFAQK1euZPr06QGPEx0d+Nmp7nOenZ19QXGIiIiIiMjl5b2y2XsVtFY8i8j/kkJ7RzPG9MNJPGcDC4EhAVabplhr3wWw1h4zxjyAk4Seb4z5GEgHugK1XO2fXJ7ouaSlLK4mUVFRpKenByxPkJWVRVpaGpGRkX77paamBhxv//79nnEvxMKFC7n99tsJCgpi1qxZNG3aNGC/WrVqAc5K4UDcq3Nr1qx5zmO6Y3THnNu+fft8+nlLTU31xOIt0Pu/5ZZbmDFjBmfOnGHZsmXMnDmTN998kz59+lCmTBnatm3r6X/nnXfyxRdfnDN2b/mt8nYndN966y3mzJlDhw4dApbc8I55+vTpdO3a9ZzHdfc/12fhQvTr149JkyYxZcoUOnbs6EmUP/zwwz6fz6ysLBISEjwPAMj9CwT3Ku5AzmdVvLcXXniBU6dOMW/ePL+/EHjppZeYPn36BY0nIiIiIiIiIlJYBJ27yxXjrrkQDAwFEgJ89ffewVo7DWgJLADuAh4CMoFHgN7WXTNBCoy7dEFeqyxvvvlmcnJyWLBggd+2BQsWkJ2dTf369f227dy5k5SUFL/2+fPne8Y9X3PnzqVjx46EhIQwZ86cPBPPAK1atQJg9uzZ5OTk+Gw7fvw4ixYtIjw8PN8x3NwxumP2lpWVxU8//QQQ8P3/+OOPfm3btm1j165dxMXFBVz1GhYWRrNmzXj++edJTEwE8CQua9euTXR0NEuWLCEzM/OcsV8Id7mKKVOmkJaWxnfffUfdunW56aabfPq5z9nChQvPa9zatWsTERHBypUrA65ED3Rez6V58+Zce+21TJ8+naNHj+aZKE9LS+PIkSM0a9bML/F84sQJT0mVgrBlyxZKliwZsDRNoM/BxTjXdSoiIiIiIiIicikU2uSztXaYtdac4ys+wH6LrLWdrLUx1tpwa+2N1tpR1lplXS6BmJgYjDHs3Lkz4PYBAwYA8PTTT5ORkeFpz8jI4KmnngLg/vvv99svOzubJ5980icBvH37dhITEwkJCeGPf/zjecU3e/ZsOnfuTNGiRfnhhx9o1KhRvv1r1KhB+/btSUlJYezYsT7bEhISOHnyJPfeey/FihU757HvuOMOSpYsyUcffcSSJUt8tr3xxhts27aNtm3b+tV7Bhg9ejQ7duzw/JyTk8Pjjz9OTk4O9913n6d94cKFAROz7tXCERERgFOK4aGHHmLfvn0MGTKEU6dO+e2zb98+1q1bd873lZt3Qnf8+PFkZmYGrJ/crVs3atSowdixY/n2228DjrV48WLP56RIkSL07duX48eP+9WITkpKYurUqRccKziJ5tOnTzNu3Di+/fZb6tat6/fLjNjYWCIiIli2bJlPKZLMzEwefvhh0tLyKj1/4eLi4khPTyc5Odmn/e2332bWrFkFcoxzXaciIiIiIiIiIpdCoS27IVeH4sWL06RJExYuXEjfvn2pWbMmwcHBdO3albp169KnTx+mT5/Of/7zH+rUqcMdd9zhqZm7fft27r77bvr27es3bt26dfn5559p0KAB7du35+jRo3zyySccOXKEf/3rX34P1Atk48aNdOvWjdOnT9OpUyemT58esIRB7sTmuHHjaNasGUOGDOGHH37guuuu4+eff2bevHnUrFmTF1988bzPzeTJk+nZsyctW7akZ8+eVKlShWXLljF79mzKlSvHxIkTA+7bvHlzbrrpJnr16kVUVBSzZs1i1apVNGjQgCeeeMLT77XXXmP27NnEx8dTvXp1ihcvztq1a/nuu++IiYlh4MCBnr7PPvssq1atYsKECXz99de0bt2aihUrcuDAATZv3syiRYt48cUXuf7668/r/Xm79957efbZZxkxYgQhISH06dPHr0+RIkX44osvuO2227j99ttp1qwZN910ExEREezatYulS5eybds29u3b50ma//Of/+SHH37gjTfeICkpiVtvvZV9+/bxySef0KlTJ7766quLivW5554jISGBzMxMv1XPAEFBQQwZMoSRI0dy44030q1bN86ePcu8efNIT0+nVatWzJs374KPHcjQoUOZNWsWt956K3fffTdRUVEkJSXx008/0aNHDz777LPffIxzXaciIiIiIiIiIpeEtVZfXl/jxo2z48aNs3L+Nm/ebDt37mxLlixpjTEWsO+8845ne3Z2th07dqxt0KCBDQ8Pt+Hh4bZ+/fp2zJgxNjs72288wLZs2dLu2bPH9u3b15YpU8aGhYXZm2++2U79f/buPr7u8r4P/ueSbQy2kcHP2LWBtCGlYnH6MFjaFad2YMto0rRpnd5L2zw13CGsZOuadVsDkSG9u422W9MlrNA0bVPuJl6X0mR3syYFAn0EG2MbhAlPNvID2CDANgaDbV33H5ZUyU/ykY8kS3q/X6/zOuf8fr/r6Ht06fez/NF1ruv220+6rrvvvrsmGfR2LJ2dnfUDH/hAXbBgQZ0yZUpdsmRJve6662pXV1fD35/777+/vvvd765z5sypU6ZMqYsXL64f/ehH6/bt24869v3vf39NUp988sn667/+6/VNb3pTnTp1al24cGH9+Mc/Xnfv3j3g+L/4i7+oH/jAB+rFF19cW1tb67Rp0+pFF11Uf+EXfqFu2bLlqNfv7u6uf/iHf1iXL19ezz333DplypS6cOHC+kM/9EP1V3/1V2tnZ2ffsZs3b65J6vvf//5B3+PTTz9dW1paapL6oz/6oyc8dufOnfWXf/mXa1tbWz3rrLPq9OnT63d913fV97znPfWLX/xiPXDgwIDjn3nmmfrBD36wzpkzp5555pl16dKl9Qtf+EJf/37qU58atL4jrVixoiapkydPrs8+++wxjzlw4LNQhh0AACAASURBVED9jd/4jXrxxRfXM888s86fP7/+zM/8TN2yZUtfP23evLnv+JP5fi1btuyYP3Nf+9rX6mWXXVZnzJhRZ86cWa+44op6zz331C984QtHnU+11nr++efX888//5hf41Of+lRNUu++++4B2wc7TwEAgNHzXzYd6LsBDINRy1pLNQ3yALfccktNkmuuuWa0S5mwSilZtmzZkOb0BQAAgLHm5kcP9j3+xHf7kDrQdGW0vvBpO+czAAAAAABjl/AZAAAAAICmEz4DAAAAANB0JhLitGMecgAAAAAY+4x8BgAAAACg6YTPAAAAAAA0nfAZAAAAAICmEz4DAAAAANB0wmcAAAAAAJpO+AwAAAAAQNMJnwEAAAAAaDrhMwAAAAAATSd8BgAAAACg6YTPAAAAAAA0nfAZAAAAAICmEz4DAAAAANB0wmcAAAAAAJpO+AwAAAAAQNMJnwEAAAAAaDrhMwAAAAAATSd8BgAAAACg6YTPAAAAAAA0nfAZAAAAAICmEz4DAAAAANB0wmcAAAAAAJpO+AwAAAAAQNNNHu0CAAAAACaqba/U7HqtprsmLeXw8++YVka7LICmED4DAAAAjLCHd3fnK9u68+jeOmD7v9t4MN99dslPfEdLLpnpA+vA2OYqBgAAADCC7t7VnV/bdOio4LnXo3trfm3ToXxrV/cIVwbQXMJnAAAAgBHy8O7u/O5Th3Ls2Pkf1CS3PXUoD+8WQANjl/AZAAAAYIR8ZVv3oMFzr5rkT7cJn4Gxa0hzPpdSpidZlGROkleT7Kq1bm9mYQAAAADjybZX6nGn2jieTXurRQiBMeukw+dSyncm+VCSK5J8b44YNV1K6Upyd5L/leQrtdaDTawTAAAAYEwb6hQaD+/uzndMm9TkagCG36DhcynlB5J8OsnbczhwPpDkoSTPJnkhyVlJZid5U5KfSvKTSZ4rpfxWkt+stb42PKUDAAAAjB2vHhrZdgCj7YThcynlD5K8L8nuJLcm+VKS+2ut+49z/AVJ/lmS9yf51SQfLaX8bK313ibWDAAAADDmnDXEwctDbQcw2gZbcPDKJL+YZGGt9WO11nuPFzwnSa11S631d2qtP5jkLUnWJVnWvHIBAAAAxqZLZg4WwzS3HcBoG2zajTfUWl8dygvXWjcm+fFSyllDaQ8AAAAwnnzHtJLvPrs0tOjgxWcXiw0CY9YJ/3Q21OC52a8BAAAAMB78xHe05GSj5JLkx7/DqGdg7GrKFayUMqWU8r2llDc14/UAAAAAxqNLZrbk598wadAAuiT5yBsmmXIDGNMauoKVUlaWUlaXUmb12/adSTqSrE3ySCnlK6WUwabzAAAAAJiQfmReS/7DxZNy8dnHjqAvPrvkP1w8KW+bJ3gGxrZGQ+IP5fDigy/02/YbSb4ryV1JZif5sSQfTHJbUyoEAAAAGGcumdmSS2a2ZNsrNb/1+MF016SlJB9/42RzPAPjRqN/QvueJGt6n5RSWpP8iySra61vT3JpkkdzOHwGAAAA4AS+Y1rJvKklC848fC94BsaTRsPnuUme6ff8rTk8evpLSVJrPZDkm0m+synVAQAAAAAwJjUaPu9NMrPf82VJapK/7rdtf5KzT7EuAAAAAADGsEbnfH48yTtKKVNzOHT+qSQba63P9zvm/CS7mlQfAAAAAABjUKMjn29N8oYcDqE39Tz+vSOOuSxJx6mXBgAAAADAWNVQ+Fxr/YMk/ynJtByefuO/99ySJKWU5UkuSHJ380oEAAAAAGCsaXTajdRa/2OS/3ic3X+d5Nwk+06lKAAAAAAAxraGw+cTqbW+nuT1Zr4mAAAAAABjzwmn3SilnHWqX6AZrwEAAAAAwNgy2JzPm0spHy+lTG30hUspS0spf5bkl4ZWGgAAAAAAY9Vg4fM3kvxmkmdKKbeUUn7kRCOZSylvKKVcU0r5uyTrkiyNxQcBAAAAACacE875XGv9uVLKZ5L8P0mu7rkdKqVsSvJMkheTnJlkdpI3JZmTpCTZmeRXkvzXWutrw1c+AAAAAACno0EXHKy1rk1yZSnljUk+nGRFkrck+UdHHPpckq8k+V9J/let9UCTawUAAAAAYIwYNHzuVWt9PMm/T5JSyrQki3J4xPOrSXbVWp8ZlgoBAAAAABhzTjp87q/W+kqSx3tuAAAAAAAwwGALDgIAAAAAQMOEzwAAAAAANJ3wGQAAAACAphM+AwAAAADQdMJnAAAAAACaTvgMAAAAAEDTNRQ+l1KWlFJaBznm7FLKklMrCwAAAACAsazRkc+bk3x8kGOu6znulJVSfrKU8tullL8qpewppdRSyh8d59gLevYf7/alZtQEAAAAAMDgJjd4fOm5jZRPJlma5OUk25J890m02ZDkjmNsf7iJdQEAAAAAcAKNhs8nY36SfU16rX+Tw6HzE0mWJbn7JNqsr7W2N+nrAwAAAAAwBIOGz6WUnzti01uOsS1JJiVZkuRnkzzUhNpSa+0Lm0sZyQHXAAAAAACcipMZ+fz7SWrP45rkx3puR+pNh19JsuqUKxu6haWU/zvJ7CRdSf6u1rpxFOsBAAAAAJhwTiZ8/mDPfUnyezk8n/KfHeO4Q/mHsPel5pQ3JFf03PqUUr6V5P211s5RqQgAAAAAYIKZ1N7efsID2tvbN/Tc1q9ateptSb5Ua/3Dftt7bxvb29ufaG9v3z8cha5ateqCJB9I8lB7e/tXjrH/jCT7k/xikn+f5D8n+XqSC5L8SJJ3rlq16vfa29sPHOv1SylXr1q16neuuuqqhUly0UUXZdOmTZkzZ04efPDBLFy4MGvXrs3s2bPT0dGRqVOnZvv27XnllVeyb9++7NixI2eccUYeeeSRzJ07N+vWrRvQZtOmTZkyZUp27NiRffv25dVXX822bdsyderUPPzww5k/f/6ANueee26+/e1vZ9KkSXn22Wezd+/e7N+/P1u3bs1ZZ52VjRs3ZsGCBXnggQf62pxzzjl57LHHUkrJrl27smfPnrz++ut5+umnM23atGzcuDELFy7MmjVrsmjRoqxduzYzZ87ME088kVprnn/++bz44os5ePBgtmzZkhkzZmTDhg19r9/79VpbW/Pkk0+mu7s7L7zwQrq6utLd3Z2nnnoqZ5999oA28+fPz4MPPpizzz47mzdvzsGDB/Piiy/m+eefT5I88cQTOeecc7J+/fq+NvPmzcuGDRsyffr0bNmyJQcOHMju3buza9eulFLy2GOP5dxzzx3QZu7cudm4cWOmTZuWzs7OvPbaa9mzZ0927tyZSZMm5dFHH83s2bOzbt26vvc+Z86cPPzww5k6dWq2bduW/fv35+WXX84zzzyTyZMn9/V/b5s1a9Zkzpw5eeSRR/r68sj+7+joyLx58wb05axZs7Jp06ZMnjy5ry+P7P8j+7K3/1taWrJz587s2bMnr732Wjo7O/v6/7zzzjtu/z/33HN56aWXcuDAgaP6f+3atX33ra2tR/X/oUOH8tRTT2XGjBlZv3593/drwYIFWbdu3Qn7v7W1ta9f1qxZkwULFmT9+vWZPn36Mfv/8ccfP27/T5s2LU8//XRef/31o/p/1qxZA87LuXPn5qGHHsqZZ56Zzs7O7N+/P3v37s2zzz47oP/7t+nf/73n8ssvv5wdO3ZkypQp2bRp0zHP5f793/9c7t//DzzwQN/3bdasWXn00UePey4/9NBDA/p/zZo1x+z/3nO5t/+P7MvecznJgP4/1rnc+7PT2//d3d3p6uoa0P9Hnsu9/d//XO7f/08++WRmzpw5oC/nz5/f1//9z+XnnntuQP8/+OCDfefYvHnzjnsut7S0HNX/vW16+3/r1q3HPZddy13LXctdy13LXctdy13LXctdy0/Xa/n6R59IandaXu7KgkMvuZa7lruWu5Y39Vp+3nnnjdosFaXWOvhRp4FSyttyeMHB22utP9NAu8lJ/jrJZUn+da31t050/C233FKT5Jprrhl6sQAAAAAn6eZHD/Y9/sR3n8yH1AEaMmqL6bU0cnAp5ftKKTeUUuYfZ/+Cnv1vaU55p67WejDJ7/Y8vXw0awEAAAAAmCgaCp+T/FKSn0+y6zj7dyb5cA5PfXE6ea7nfvqoVgEAAAAAMEE0Gj6/Ncnd9ThzdfRsvyvJD51qYU32T3runxrVKgAAAAAAJohGw+cFSbYNcsyOJOcNrZyhK6VcVko54xjblyf5Nz1P/2hkqwIAAAAAmJgancX+lSRzBzlmbpLXhlbOQKWUdyd5d8/TBT33by2l/H7P4+drrb/U8/g/J2krpXwr/xCQvznJ8p7H19da/7YZdQEAAAAAcGKNhs/rk/xYKeUXa60vH7mzlNKa5Md6jmuGtyR5/xHb3tBzS5Knc3ge6iT5YpIfT/KPk7wjyZQcnoN6dZL/Xmv9qybVBAAAAADAIBqdduPWHB7Z/M1Sypv77yilLE3yjSRzeo47ZbXW9lprOcHtgn7Hfr7W+qO11gtqrTNqrVNrrUtqre8VPAMAAAAAjKyGRj7XWr9cSnlHkp9L8mApZWeS7UkWJZmfpCT5g1rrHze9UgAAAAAAxoxGRz6n1vqBJB9N8kgOz8P8/T33HUmurrV+sJkFAgAAAAAw9jQ653OSpNZ6a5JbSynTkpyT5KVa6ytNrQwAAAAAgDFrSOFzr57AWegMAAAAAMAAQwqfSylzk7wnycVJptdaf77f9guTPFRrfbVpVQIAAAAAMKY0HD6XUj6c5DNJzszhBQZrkp/v2T0/yd8luTrJ55tUIwAAAAAAY0xDCw6WUq5IcmuSx5L8eJJb+u+vtT6cwwsPvrtZBQIAAAAAMPY0OvL5l5M8k2RZrXVPKeV7j3HMxiRvPeXKAAAAAAAYsxoa+ZzkB5L871rrnhMcsy3JgqGXBAAAAADAWNfoyOczkuwb5JhzkhwaWjkAAAAA49/Njx7se/zky/Wo7Z/47oaX6QI47TR6JduS5PsHOeayJN8eUjUAAAAAE8ysM0a7AoDh0Wj4/GdJ/l0p5adqrf/zyJ2llA8meXOSX2lGcQAAwMTyzk9/9YT7v/bJd41QJQDDy8hmYCJo9Er3X5L8dJI/LqX8ZJKZSVJK+VdJfjjJTyR5PMlvN7NIAABg4tne9Q8z/i2aPX0UKwEAYCgaCp9rrS+WUpYl+cMkP9Vv12d67v8qyb+stQ42LzQAAMBR+o9s7j8K2ohnAICxp+HPeNRaO5O8rZTy5iRvTTI7ye4kf19rfaDJ9QEAAAAAMAadMHwupXwlyZdqrat7nl+eZEuttbPWujHJxhGoEQAAAACAMaZlkP3vTvLd/Z7fneQDw1YNAAAAAADjwmDh80tJWvs9L8NYCwAAAAAA48Rgcz5vSvJ/lVLWJHmmZ9sFPdNvnFCt9d5TLQ4AAAAAgLFpsPB5VZI7kvy//ba9v+c2mElDLQoAAAAAgLHthOFzrfUbpZSLk7w9yaIk7Unu6bkBAAAAAMAxDTbyObXWp5N8PklKKe1JvlVrvXGY6wJgDHrnp796wv1f++S7RqgSAAAAYLSdMHwupfxmkv9Ta/1Gz6YPJnlw2KsCYMzb3rWv7/Gi2dNHsRIAAABgNAw28vlfJ3kpSW/4/Hs5PA/0xuEsCoCxqf/I5v6joI14BgAAgImnZZD9LyeZ1u95GcZaAAAAAAAYJwYb+fxEkp8opfxpkmd6tp1TSlky2AvXWjtPtTgAAAAAAMamwcLnm5P8UZK/7bft4z23E6kn8doAAADH1NHZlc079+Tgoe5MntSSjs6utC2ZPdplAQDQgBMGxLXWPy6lbE5yVZJFST6Qw/M9rx/+0gAAgInmzg1bc+Pq+3Jvx44B2y+57vZc3rYwN6y8LCuWLh6l6gAAaMSgo5NrrX+f5O+TpJTygSR/Wmu9cZjrAgAAJpjPf7MjV3/urnTXesz993bsyJXtd+S2a5fnQ29vG+HqAABo1GALDh7pg0nuGI5CAACAievODVtPGDz36q41H/nsXblzw9YRqgwAgKFqKHyutf5BrXXjcBUDAABMTDeuvm/Q4LlXd625afX9w1wRAACnqtGRzymltJRSfqGU8vellN2llIP99n1vKeVzpZSLmlsmAAAwXnV0dh01x/Ng7unYno7OrmGqCACAZmgofC6lnJHkm0n+W5LvTLI3Sel3yOYkH0ryvmYVCAAAjG93bhzaFBpDbQcAwMhodOTzJ5L8SJJVSeYn+d3+O2utLyW5N8k/a0p1AADAuLfnlddHtB0AACOj0fD5fUn+ptZ6Y621O8mxJmXbnGTJKVcGwJjV0dmVzTv35PEdL2Xzzj0+Fg3ACbVOO2NE2wEAMDImN3j8hUn+v0GOeSHJrKGVA8BYdueGrblx9X1Hzdt5yXW35/K2hblh5WVZsXTxKFUHwOlqxZuH9m/DUNsBADAyGg2fX01yziDHLEny0tDKAWCs+vw3O3L15+5Kdz3Wh2KSezt25Mr2O3Lbtcvzobe3jXB18A/e+emvnnD/1z75rhGqBOjVtmR2Lm9b2NCig8vaFqVtyexhrAoAgFPVaPi8PsmVpZQzaq1HTbBWSpmZw/M9/20zigNgbLhzw9YTBs+9umvNRz57V86f22oENKeF/kHXmWdMSjIwnBZEw8i5YeVlubL9jkH/LUmSllJy/cpLR6AqAABORaNzPt+WZHGS20sprf13lFLOSfL7Sc5N8j+aUh0AY8KNq+87qbAgORxA37T6/mGuCI7va598V9/tzDMm9d0WzZqRRbNmZHvXvtEuESakFUsX59aPLU9LKSc8rqWU3Hbtcn/EBAAYAxoa+Vxr/eNSytuTfDDJu5K8mCSllLVJ2pJMTfLZWuufN7tQAE5PHZ1dDX1MOknu6diejs4uH5dm1F36xvnH3G7EM4yOD1/Rlgvmteam1ffnno7tR+1f1rYo16+8VPAMADBGNDrtRmqtHy6l/FWSjyd5c5KS5PuSdCT5zVrrF5pbIgCnszs3bh1yO+EzAEdasXRxVixdnI7Orrz35q/n4KHuTJ7Uki9/4h3+3QAAGGMaDp+TpNb6+0l+v5RyVg5Ps7G71uozqgAT0J5XjloCYFjbATAxtC2ZnQvntw54DgDA2DKk8LlXrfXVJK82qRYAxqDWaWeMaDsAAABgbGh0wUEAGGDFm4c27+ZQ2wEAAABjg/AZgFPStmR2Lm9b2FCbZW2LfHwaAAAAxjnhMwCn7IaVl6WllJM6tqWUXL/y0mGuCAAAABhtwmcATtmKpYtz68eWDxpAt5SS265dnhVLTbkBAAAA490pLTgIAL0+fEVbLpjXmptW3597OrYftX9Z26Jcv/JSwTMAAABMEMJnAJpmxdLFWbF0cTo6u/Lem7+eg4e6M3lSS778iXeY45nT2t5XX8/ze/b3/cx2dHb5mQUAADhFTQmfSylzkvxwkleS/GWt9VAzXheAsaltyexcOL91wHM43XR0duWhp7uya/crefX1gb+6XHLd7bm8bWFuWHmZ0foAAABD1NCcz6WUa0op95VSZvXb9v1JNiX5kyR/nuRvSynTm1smAEBz3Llha5b9yp/kkutuz9PP7T0qeO51b8eOXNl+R37vLztGuEIAAIDxodGRz+9NUmutL/TbdnOSc5N8Icn8JFcl+WiS32hKhQAATfL5b3bk6s/dle5aT+r47lrzkc/elfPnthoBPYLe+emvnnD/1z75rhGqpDnG2/sBAICT1dDI5yRvTLKx90nPdBvLkny+1vrztdZ3JlmT5F82r0QAgFN354atDQXPvbprzU2r7x+mqhjM9q59fbfxYLy9HwAAOJFGRz7PTrKr3/Mf6rn/037b/irJB06hJgCAprtx9X0NB8+97unYbhHCEdR/JHD/UcNjdYTweHs/AABwshod+fxCkjn9ni9L0p3kb/ttq0nOPMW6AACapqOzK/d27Dil17hz49YmVQMAADAxNDryeVOSd5ZSfiXJoRyeA3pNrXVPv2MuSPJsc8oDADh1zQiO97zyehMqgZFjrmkAAEZbo+HzbyW5I8m2JAeTTEvyy707SymTkvzTDBwJDQAwqpoRHLdOO6MJlcDo6D/H9KLZ00exEgAAJpKGwuda61dLKR9NcnXPpttrrX/U75C35/CUG3/RpPoAAE5ZM4LjFW9e3IRKYOSYaxoAgNHW6Mjn1FpvTXLrcfb9RZJzT7UoAIBmOtXgeFnbIosNAgAANKjRBQcBAMactiWzc3nbwiG1bSkl16+8tMkVAQAAjH8Nhc+llO8tpXyslDKz37bppZQ/KKW8VErZUUr5ePPLBAA4NTesvCwtpTTUpqWU3Hbt8qxYasqN0dDR2ZXNO/fk8R0vZfPOPeno7Brtkk7JeHs/AAAwmEan3fjlJD9ca/1cv22/luRnk7ycZHaS3yylbKq1fqNJNQIAnLIVSxfn1o8tz9WfuyvdtQ56/LK2Rbl+5aWC51Fw54atuXH1fbm3Y8eA7Zdcd3sub1uYG1ZeNqb6Zby9HwAAOFmNTrvxA0m+1fuklDIlyfuT3J9kXpILkzyf5Lom1QcA0DQfvqIt32h/d5a1LTrm/rPOmJTz556dhz/zvnzrV98jEBwFn/9mR65sv+OooLbXvR07cmX7Hfm9v+wY4cqGZry9HwAAaESjI5/nJdna7/kPJDk7ye/UWvcn2VFK+bMk/7xJ9QEANNWKpYuzYunidHR25b03fz0HD3Vn8qSWzGk9M2efdUaSWFxwlNy5YetJjUzvrjUf+exdOX9u62n9B4Lx9n4AAKBRjY58rhkYWP/Tnm339Nv2XJK5p1gXAMCwalsyOxfOb80bF56TC+e39gXPjJ4bV993UlOiJIcD25tW3z/MFZ2a0+X9mGsaAIDR0ujI584k/6Tf8x9Lsq3W+lS/bQuTvHiqhQEAMHF0dHYdd2qK47mnY3s6OrtOy5Hqp8P7Mdc0AACjrdGRz6uT/GAp5U9KKX+U5K1J/uSIYy5J8uSpFlZK+clSym+XUv6qlLKnlFJ7vuaJ2vxgKeXPSykvlFJeKaVsLKX861LKpFOtBwCA4XPnxq2DH9TEdsNttN+PuaYBADgdNBo+/9ckf5fkJ5L8yyQbktzYu7OU8j1Jvj8Dp+EYqk8m+VdJ3pJk+2AHl1J+LMm9SS5P8qdJPpvkjJ6av9SEegAAGCZ7Xnl9RNsNt9F8P43ONX3nhtMzwAcAYOxraNqNWuvLSX6olHJJz6ZHaq3d/Q55JcmPJ1nbhNr+TZJtSZ5IsizJ3cc7sJTSmuS2JIeSvK3WurZn+/VJ7kryk6WUn661CqEBAE5DrdOGNuf2UNsNt9F8P0OZa9r0G0xk7/z0V0+4/2uffNcIVQIA40+jcz4nSWqtDx9n+5YkW06hnv6v1Rc2l1IGO/wnc3iRwz/sDZ57XmN/KeWTSe5Mck2MgAYAOC2tePPQws+hthtuo/V+Toe5pk9V/yDw/sd3HrVdEMhw2t61r+/xotnTR7ESABgfhhQ+l1KmJFmR5OIkM2qtN/VsPzNJa5LnjxgRPdyW99z/n2PsuzeHR2T/YCllaq31tZErCwCAk9G2ZHYub1vYUHC6rG3RaROYHmm03s+pzDV9On4vF82aMdolMAH0/4NG/z9++EMHAJy6hsPnUso/T/L5JAuSlCQ1yU09u9+S5G+S/EySP25SjSfjTT33jx25o9Z6sJSyOUlbkjck2TSCdQEAcJJuWHlZrmy/46SmjGgpJdevvHQEqhq60Xg/42HubIEfAMD4Mam9vf2kDy6l/ECSb+TwSOIbkryQ5B+1t7ffmCTt7e3bVq1a9XNJzmlvb/+fzSpy1apVFyT5QJKH2tvbv3KM/Z9IMjvJze3t7S8cY//7kixOcnt7e/u2Y32NUsrVq1at+p2rrrpqYZJcdNFF2bRpU+bMmZMHH3wwCxcuzNq1azN79ux0dHRk6tSp2b59e1555ZXs27cvO3bsyBlnnJFHHnkkc+fOzbp16wa02bRpU6ZMmZIdO3Zk3759efXVV7Nt27ZMnTo1Dz/8cObPnz+gzbnnnptvf/vbmTRpUp599tns3bs3+/fvz9atW3PWWWdl48aNWbBgQR544IG+Nuecc04ee+yxlFKya9eu7NmzJ6+//nqefvrpTJs2LRs3bszChQuzZs2aLFq0KGvXrs3MmTPzxBNPpNaa559/Pi+++GIOHjyYLVu2ZMaMGdmwYUPf6/d+vdbW1jz55JPp7u7OCy+8kK6urnR3d+epp57K2WefPaDN/Pnz8+CDD+bss8/O5s2bc/Dgwbz44ot5/vnnkyRPPPFEzjnnnKxfv76vzbx587Jhw4ZMnz49W7ZsyYEDB7J79+7s2rUrpZQ89thjOffccwe0mTt3bjZu3Jhp06als7Mzr732Wvbs2ZOdO3dm0qRJefTRRzN79uysW7eu773PmTMnDz/8cKZOnZpt27Zl//79efnll/PMM89k8uTJff3f22bNmjWZM2dOHnnkkb6+PLL/Ozo6Mm/evAF9OWvWrGzatCmTJ0/u68sj+//Ivuzt/5aWluzcuTN79uzJa6+9ls7Ozr7+P++8847b/88991xeeumlHDhw4Kj+X7t2bd99a2vrUf1/6NChPPXUU5kxY0bWr1/f9/1asGBB1q1bd8L+b21t7euXNWvWZMGCBVm/fn2mT59+zP5//PHHj9v/06ZNy9NPP53XX3/9qP6fNWvWgPNy7ty5eeihh3LmmWems7Mz+/fvz969e/Pss88O6P/+bfr3f++5/PLLL2fHjh2ZMmVKNm3adMxzuX//9z+X+/f/Aw880Pd9mzVrVh599NHjnssPPfTQgP5fs2bNMfu/91zu7f8j+7L3XE4yoP+PdS73/uz09n93d3e6uroG9P+R53Jv//c/l/v3/5NPPpmZM2cO6MtvPb43e/bszuTJk/OPz2vpO5efe+65Af3/4IMP9p1j8+bNO+653NLSclT/97bp7f+tW7ce91x2LXctP9a1/PZ7Hk33EAwZmQAAIABJREFUoUM5eOhQ9r/2WlpaWrJ379783IpLXMtH4Vr+8q6n8z0XLsxfbNiWE8W1LaXkV656Q3727W8+ra/lB3Y/mzcunpdvbtx+Uu/np5ddfMrX8geeeDb3PPrcCb7asV1xyfycU3cfdS2fP39+X//3P5ddy13LT6dreTN/L7/7ib3ZvXt3zjzzzFw0ba9rud/Lx8Xv5a7lruUT7Vru9/Kj+/+8885b1fAviE1S6kkuRpIkpZQ/S/LDSb6n1vpsKeVTSW6otU7qd8wfJ/m+Wuubjvc6DRdZyttyeMHB22utP3OM/Y8leWOSN9ZanzjG/r9N8tYkb621/v2JvtYtt9xSk+Saa65pQuUAE5ePrTIWHG+RKT+zo+vODVtz0+r7c0/H9qP2LWtblOtXXjqmFsgbyffT0dmVS667veF2D3/mfafltBsw0vz+AsA4NeiCesOl0Wk3fijJHbXWZ09wzNYkVw29pCHZ3XM/8zj7W484DoBhYJEooBlWLF2cFUsXp6OzK++9+es5eKg7kye15MufeMeYDEhH8v2Mt7mzAQAY2xoNn2ckeX6QY6YlaRlaOUP27SQ/kOSiJA/031FKmZzkwiQHkzw1wnUBTFijvUjU8Ua09hKEw+mvbcnsXDi/dcDzsWyk3s94mzsbRkpHZ1c279zT9weijs6uMX/dAYDR1mj4vD2HF+47kbdk5EPeu5K8L8k/z9ELHV6ew4H4vbXW10a4LoAJ5XQNdLd37et7vGj29FGsBGD4rVi6OLd+bHmu/txdJwygW0rJbdcuH1NTmMBwuHPD1ty4+r6jPjFwyXW35/K2hblh5WXOEwAYokbD568n+Wgp5Z/WWv/6yJ2llHck+cEk/6kZxTXgT5L85yQ/XUr57Vrr2p56zkzy6Z5jbhnhmgAYRf2DcPM3AhPNh69oywXzWsfV3NkwHD7/zY4T/qHm3o4dubL9jtx27fJ86O2DjcMCAI7UaPj8a0l+Osk3Sim/neSCJCmlXJXDI4yvTfJMkt881cJKKe9O8u6epwt67t9aSvn9nsfP11p/KUlqrXtKKR/J4RD6W6WULyV5Icm7krypZ/uXT7UmAAAYK8bb3NnQbHdu2DroJwSSpLvWfOSzd+X8ua3+YAMADWoofK61bi+lXJlkdZJP9Nv11RxeNfHJJD9Rax1sXuiT8ZYk7z9i2xt6bknydJJf6lfbHaWUZUl+Jcl7kpyZ5Ikkv5jkM7WexKR3AAAwzoy3ubOhWW5cfd9JzY2eHA6gb1p9v/AZABrU6Mjn1FrXlVLelOSqJG9NMjvJ7iR/n+TPaq0Hm1FYrbU9SXuDbf4myb9oxtcHAABgfOro7DpqjufB3NOx3SKEANCghsPnJKm1Hsrh0c5fHexYAAAAOJ3cuXHrkNsJnwHg5A0pfAYAGA/uf3znMbf3LlJpgUqA8WnPK6+PaDsAmKiGFD6XUn4gyaVJzk0y6RiH1FrrTadSGADASFo0a8ZolwDACGmddsaItgOAiaqh8LmU0prkK0l+JIcXGDyemkT4DACc1i594/y+x0Y5A0wcK948tIUDh9oOACaqRkc+35xkeZK/SvKFJFuTNGWBQQAYDh2dXdm8c08OHurO5EktFgoCANK2ZHYub1vY0KKDy9oW+R0CABrUaPj8Y0nWJfmRWmv3MNQDAE1x54atuXH1fUf9p/KS627P5W0Lc8PKy7JiqdFLADBR3bDyslzZfke6ax302JZScv3KS0egKgAYXxoNn2cm+aLgGYDT2ee/2ZGrP3fXcf8zeW/HjlzZfkduu3Z5PvT2thGujpHWu3jgsdz/+M4BU28AMHGsWLo4t35s+Ql/Z0gOB8+3XbvcH60BYAgaDZ8fT+J/aACctu7csHXQ/0QmSXet+chn78r5c1v9Z3IC2d61r+/xotnTR7ESAE4HH76iLRfMa81Nq+/PPR3bj9q/rG1Rrl95qd8VAGCIGg2fP5vkP5VSFtVaj/6XGQBG2Y2r7zupj88mhwPom1bf7z+U41z/hQT7j4L+2iffdcJR0QBMDCuWLs6KpYvT0dmV99789b51Ir78iXeY4xkATlGj4fPXc3jBwb8ppaxK8kCSl451YK218xRrA4CGdHR2NbRwUJLc07HdIoQAQNqWzM6F81sHPAcATk2j4fOWJDVJSfK7JziuDuG1AeCU3Llx65Db+Q8mAAAANFejAfEf5nCwDACnnT2vvD6i7QAAAIDjayh8rrV+YJjqAIBT1jrtjBFtBwAAAByfqTEAGDdWvHloCwcOtR1j0/2P70xyePHB3se9z5OBCxQCAAAwdA2Fz6WUQ0naa603neCYX0myqtYq2AZgRLUtmZ3L2xY2tOjgsrZF5nuewBbNmjHaJXAMvX8ISDIu/kAw3t4PAACcrEYD4tJzO5njAOC4+ocxxzLUMOaGlZflyvY70l0HX6KgpZRcv/LSIX0dxq5L3zg/icBvrBhvfyAYyfcj9AYAYLQNx+jkc5PsH4bXBWCc2t61r+/xotnTT+m1VixdnFs/tjxXf+6uEwbQLaXktmuXZ8VSU25MFB2dXdm8c08OHurO5Ekt6ejsMur9NDXeQtHT4f2MtxAfAICxYdDwuZRy+RGbLjjGtiSZlGRJkvcl+XYTagNgHOsfxvQfndeMkObDV7TlgnmtuWn1/bmnY/tR+5e1Lcr1Ky8VPE8Qd27YmhtX33fUdCyXXHd7Lm9bmBtWXuZnYRgN16ccGJzvLQAAo+1kRj5/K0nv0LGa5P09t2MpSbqT/NtTrgyACWG4RqOuWLo4K5YuTkdnV95789f7Xv/Ln3iH0a4TyOe/2XHCUfD3duzIle135LZrl+dDb28b4eomnmZ+ygEAADj9nUz4fGMOh84lyQ05HEbfc4zjDiXpSnJ3rfXRZhUIwPg0UqNR25bMzoXzWwc8Z2K4c8PWQadfSZLuWvORz96V8+e2GgE9DIbzUw4AAMDpbdDwudba3vu4lPL+JHfUWj8znEUBML4ZjcpIuHH1fSe18GRyOIC+afX9wudhZM5tAACYeBpacLDWeuFwFQLAxGA0KiOho7PrqFH1g7mnY7tAdBiYcxsAACaultEuAICJZSijUaFRd27cOqLtOLbPf7MjV7bfcdw/BPR+yuH3/rJjhCsDAABGwglHPpdS7krPIoO11m09z09GrbWuOOXqABhXjEZlpOx55fURbcfRfMoBAAAYbNqNt/XcTzvi+WBObkgbABPKqYxGFT7TiNZpZ4xoO45mzm0AAOCE4XOtteVEzwGgEUajMlJWvHloIeZQ2zGQTzkAAADJ4NNuXJfk72utJtwE4JQZjcpIaVsyO5e3LWwoAF3Wtkjw2SQ+5QCMJe/89Ff7Ht//+M6jtn/tk+8a8ZoAYLwYbNqN/5akPcn9SVJKOZSkvdZ60zDXBcA4NNRRpV/5uyfyzfWdx93vP4Ucyw0rL8uV7Xec1NQPLaXk+pWXjkBVE4NPOQBj1aJZM0a7BAAYVwYLn/cnmdrveem5AUDDhjoa9eyzpvQ93961r+/xotnTm1of48uKpYtz68eWD7roXUspue3a5eYbbiKfcgDGEn/EBoDhM1j4vDnJPyulfKbW2vv5I4sJAjBkQxmN2j8U7P/RWP9ZZDAfvqItF8xrzU2r7889HduP2r+sbdFRP2OcOnNuAwAASTLYAoK/k+T7kuzomXIjSdpLKYcGuR0c3rIBGKt6R6O2lBN/kMZoVJplxdLF+davvicPf+Z9aVs8K29aeE7aFs/Kw595X771q+/xMzYMej/l0AhzbgMAwPhzwpHPtdbPlFJ2JbkqycIkP5KkM8mW4S8NgPHKaFRGQ9uS2blwfuuA5wwfc24DAACDTbuRWuuXknwpSUop3Um+UGu9cbgLA2B8W7F0cVYsXZyOzq689+av5+Ch7kye1JIvf+IdQkEYB8y5DQAADBo+H2FVkm8NQx0ATFCNjEbt6OzK5p17+oLqjs4uQTWcxnzKAQAAJraGwuda66rhKgQAjufODVtz4+r7cm/HjgHbL7nu9lzetjA3rLxMeAWnKZ9yAACAiavRkc8AMKI+/82OE35s/96OHbmy/Y7cdu3yfOjtbSNcHXCyzLkNAAATT8toFwAAx3Pnhq2DzhebJN215iOfvSt3btg6QpUBAAAAgzHyGYBR8c5Pf7Xv8f2P7zxq+9c++a7cuPq+QYPnXt215qbV95t+AwAAAE4TwmcARt2iWTOO2tbR2XXUHM+Duadju0UIAQAA4DQhfAZgVHztk+864f7P/O/1Q3rdOzduFT4DAADAaUD4DMBpac8rr49oO2B4nMwUOwAAwPgkfAbgtNQ67YwRbQcMv2NNsQMAAIxfwmcATksr3jy0hQOH2g4YHkY2AwDAxCV8BuC01LZkdi5vW9jQooPL2hb1zffso/4AAAAwulpGuwAAOJ4bVl6WllJO6tiWUnL9ykuPuW/RrBl9NwAAAGBkGPkMwGlrxdLFufVjy3P15+5Kd63HPa6llNx27fKsWPoPU24Y2QwAAACjS/gMwGntw1e05YJ5rblp9f25p2P7UfuXtS3K9SsvHRA8AwAAAKNP+AzAaW/F0sVZsXRxOjq78t6bv56Dh7ozeVJLvvyJd/TN8QwAAACcXoTPAIwZbUtm58L5rQOeAwAAAKcnCw4CAAAAANB0wmcAAAAAAJpO+AwAAAAAQNMJnwEAAAAAaDrhMwAAAAAATSd8BgAAAACg6YTPAAAAAAA03eTRLgAAYDi989Nf7Xt8/+M7j9r+tU++a8RrAgAAmAiEzwDAhLFo1ozRLgEAAGDCED4DAOOakc0AAACjw5zPAAAAAAA0nfAZAAAAAICmM+0GAKc9C8YBAADA2CN8BmBMsWAcAAAAjA3CZwBOe0Y2AwAAwNhjzmcAAAAAAJpO+AwAAAAAQNONu/C5lLKllFKPc3t2tOsDAAAAAJgIxuucz7uT/LdjbH95pAsBAAAAAJiIxmv4/FKttX20iwAAAAAAmKjG3bQbAAAAAACMvvE68nlqKeVnkixJsi/JxiT31loPjW5ZAAAAAAATw3gNnxck+eIR2zaXUj5Ya71nNAoCAAAAAJhIJrW3t492DU21atWqmUluSvLLST6V5EtJJiVZkeSnV61a9b/b29t3HtmulHL1qlWrfueqq65amCQXXXRRNm3alDlz5uTBBx/MwoULs3bt2syePTsdHR2ZOnVqtm/fnldeeSX79u3Ljh07csYZZ+SRRx7J3Llzs27dugFtNm3alClTpmTHjh3Zt29fXn311Wzbti1Tp07Nww8/nPnz5w9oc+655+bb3/52Jk2alGeffTZ79+7N/v37s3Xr1px11lnZuHFjFixYkAceeKCvzTnnnJPHHnsspZTs2rUre/bsyeuvv56nn34606ZNy8aNG7Nw4cKsWbMmixYtytq1azNz5sw88cQTqbXm+eefz4svvpiDBw9my5YtmTFjRjZs2ND3+r1fr7W1NU8++WS6u7vzwgsvpKurK93d3Xnqqady9tlnD2gzf/78PPjggzn77LOzefPmHDx4MC+++GKef/75JMkTTzyRc845J+vXr+9rM2/evGzYsCHTp0/Pli1bcuDAgezevTu7du1KKSWPPfZYzj333AFt5s6dm40bN2batGnp7OzMa6+9lj179mTnzp2ZNGlSHn300cyePTvr1q3re+9z5szJww8/nKlTp2bbtm3Zv39/Xn755TzzzDOZPHlyX//3tlmzZk3mzJmTRx55pK8vj+z/jo6OzJs3b0Bfzpo1K5s2bcrkyZP7+vLI/j+yL3v7v6WlJTt37syePXvy2muvpbOzs6//zzvvvOP2/3PPPZeXXnopBw4cOKr/165d23ff2tp6VP8fOnQoTz31VGbMmJH169f3fb8WLFiQdevWnbD/W1tb+/plzZo1WbBgQdavX5/p06cfs/8ff/zx4/b/tGnT8vTTT+f1118/qv9nzZo14LycO3duHnrooZx55pnp7OzM/v37s3fv3jz77LMD+r9/m/7933suv/zyy9mxY0emTJmSTZs2HfNc7t///c/l/v3/wAMP9H3fZs2alUcfffS45/JDDz00oP/XrFlzzP7vPZd7+//Ivuw9l5MM6P9jncu9Pzu9/d/d3Z2urq4B/X/kudzb//3P5f79/+STT2bmzJkD+nL+/Pl9/d//XH7uuecG9P+DDz7Yd47NmzfvuOdyS0vLUf3f26a3/7du3Xrcc9m13LXctdy13LXctdy13LXctdy13LXctdy13LV8Il7LzzvvvFWnHLoOUam1jtbXHlGllF9P8m+T3FFr/fHjHXfLLbfUJLnmmmtGqjQAAAAAgOFSRusLT6QFB/9Hz/3lo1oFAAAAAMAEMJHC510999NHtQoAAAAAgAlgIoXPb+25f2pUqwAAAAAAmADGVfhcSmkrpcw6xvbzk/z3nqd/NLJVAQAAAABMPJNHu4Am+6kk/76UcneSzUn2JvnOJFclOTPJnyf59dErDwAAAABgYhhv4fPdSd6U5HtzeJqN6UleSvLXSb6Y5Iu11jp65QEAAAAATAzjKnyutd6T5J7RrgMAAAAAYKIbV3M+AwAAAABwehA+AwAAAADQdMJnAAAAAACaTvgMAAAAAEDTCZ8BAAAAAGg64TMAAAAAAE0nfAYAAAAAoOmEzwAAAAAANJ3wGQAAAACAphM+AwAAAADQdMJnAAAAAACaTvgMAAAAAEDTCZ8BAAAAAGg64TMAAAAAAE0nfAYAAAAAoOmEzwAAAAAANJ3wGQAAAACAphM+AwAAAADQdMJnAAAAAACaTvgMAAAAAEDTCZ8BAAAAAGg64TMAAAAAAE0nfAYAAAAAoOmEzwAAAAAANJ3wGQAAAACAphM+AwAAAADQdMJnAAAAAACaTvgMAAAAAEDTCZ8BAAAAAGg64TMAAAAAAE0nfAYAAAAAoOmEzwAAAAAANJ3wGQAAAACAphM+AwAAAADQdMJnAAAAAACaTvgMAAAAAEDTCZ8BAAAAAGg64TMAAAAAAE0nfAYAAAAAoOmEzwAAAAAANJ3wGQAAAACAphM+AwAAAADQdMJnAAAAAACaTvgMAAAAAEDTCZ8BAAAAAGg64TMAAAAAAE0nfAYAAAAAoOmEzwAAAAAANJ3wGQAAAAD+//buO06q6v7/+OvDAktZytIVQaqNEgs2UASNRjG2RBNjLFijSawxPxONEfP9GvUbG8Ge2E1ixCQaTUxUEFE0doKVJkgRRBbBLbCFPb8/zhkZxpndmdk7M8vu+/l43Mfs3jLn3Pfcmbn3zL3nikjk1PgsIiIiIiIiIiIiIpFT47OIiIiIiIiIiIiIRE6NzyIiIiIiIiIiIiISOTU+i4iIiIiIiIiIiEjk1PgsIiIiIiIiIiIiIpFT47OIiIiIiIiIiIiIRE6NzyIiIiIiIiIiIiISOTU+i4iIiIiIiIiIiEjk1PgsIiIiIiIiIiIiIpFT47OIiIiIiIiIiIiIRE6NzyIiIiIiIiIiIiISOTU+i4iIiIiIiIiIiEjk1PgsIiIiIiIiIiIiIpFT47OIiIiIiIiIiIiIRE6NzyIiIiIiIiIiIiISOTU+i4iIiIiIiIiIiEjk1PgsIiIiIiIiIiIiIpFrcY3PZraDmd1rZp+YWbWZLTWzW8ystNB1ExEREREREREREWkt2ha6AlEys6HAy0Af4AngQ2Af4ELgcDMb55wrK2AVRURERERERERERFqFlnbm8+34hucLnHPHOud+5pw7GLgZ2Bm4pqC1ExEREREREREREWklWkzjs5kNAQ4DlgK3JUy+CqgETjGzznmumoiIiIiIiIiIiEir02Ian4GDw+Mzzrn6+AnOuXJgDtAJ2C/fFRMRERERERERERFpbVpSn887h8cFKaYvxJ8ZvRMwo7Enu+OOOyKqloiIiIiIiIiIiEjBuPPOO88KUXBLOvO5W3jckGJ6bHz3ZBPN7BwzeyPyWomIiIiIiIiIiIi0Qi3pzOfGxFr3XbKJzrm7gbvzVx0RERERERERERGRlqslnfkcO7O5W4rpXRPmExEREREREREREZEcaUmNz/PD404ppg8Pj6n6hBYRERERERERERGRiJhzSXuh2OaY2VBgEbAUGOqcq4+b1gVYhW9s7+2cqyxIJUVERERERERERERaiRZz5rNzbjHwDDAI+FHC5KuBzsCDangWERERERERERERyb0Wc+YzfHn288tAH+AJ4ANgX2AivruNsc65ssLVUERERERERERERKR1aFGNzwBmNgD4FXA40BPf3cbjwNXOuXWFrJuIiIiIiIiIiIhIa9HiGp9FREREREREREREpPBaTJ/PIiIiIiIiIiIiItJ8qPFZRERERERERERERCLXttAVaI6Kitq4+vrsuyMxM3LdnUkbgyZUsenlA/WFK34r+ch76/La4Fwe194MtqXucawN5DOfKBUkawPy//o2p/ewSJQK/f0oIiIiIiLSnLRv35aamrq1zrnehShfjc9J1Nc73BvZL29jHLNnz46uQkmMHz8ed2tOi2iQ/Rjc6OMLV4E4Nu8xjpp/Zt7Ke3Lne/jeKzV5K+9P+7dnr+kL81ZeU715wnC4fBtt+fm18b3vvp/XIv/0593gmKl5LROg/okLceN2z3u5Irlmc+Y26TtcRERERESkJbExdQAfF6p8dbshIiIiIiIiIiIiIpFT47OIiIiIiIiIiIiIRE7dbqRp9ltww0Pw5ofwyWdw31Uw+ajsn+/hhx9m9uzZLFu2jHbt2jFixAjOOecchgwZEl2l49z2Atw1B5au8/+P6Ae/OByOHJmT4tL26zUfcMXq9/hRz6Hc2n+PnJc3f9pbLLj17a3GFffqyGFzTspZmRvXrmLu7Vew6pV/UVtVTsn2g9n7p7fSZ8/xOSszlXd+OIGaz1Z+ZXzXPQ5i+OW/z3t90lK/GV6cAu8+DBWroGQ7GPF9GD8F2hT2I6y2tpJ57/yWFSufo7p6HaXdd2XPPX5Oz56joitk7SJY/DysXw6bNsAeJ8HAfbdMdw7m/wuWvgy1G6F0Rxh9PHTdLro6SFpmb6jghpVreLNyI5/U1HLfsAFM7tvzy+l/LVvPXavLeKuiirV1m3l+5FAmdOtSwBqLiIhIUzR0jFhbB7+4HZ5+GRavgK6dYeIYuO58GNivsPUWERHJJzU+p6miCkYOg1OPhFOvavrzvf322xx77LHssssuOOe49957ueSSS3jwwQfp2rVr0wtIsEMpXH8MDO8D9fXwwKtw7N3w5mUwun/kxaXlP5Vl/K5sCaM7dMtruZ0Hd2PsQ5O+/N+KLGdl1ZSv59kfTKD36LEcdMMTFJf2omLlEop7FKSPd3a59i9+Awhq16/hg8uOo8fYSQ0sVWCvXA9v3gZHPQC9R8GaefDUadC2GA64sqBVe+31K1m/fgH77XstnTr2ZenHT/L8C2cy6fAn6dSpbzSFbK6GLv1gwN7w1sNfnb5oBix6HvY8CUr6wPx/w8u3wyFXQLsO0dRB0lKxuZ6RnTtwap8enLrwq91pVW6uZ2yXzpzcu5RTFy4rQA1FREQkSg0dI1Ztgrc+hCvOgN13gg0V8JNb4PDzYd6foK2OxEVEpJXQV16aJh3gB4DJVzf9+W688cat/r/iiiuYNGkS77zzDuPGjWt6AQmOGb31/9ccDXe8BK8sKUzj84bNtXx/+WvcM2AMv/o0vzd4a9PW6NC7U17K+uDhG+jYsx/7X3Xfl+NKth+cl7KTadet51b/r505naKOJZTuf0SBapSGlS/D8KP8ANB9EAw/Gj55taDVqqvbxPIVz3LAuFvo22cfAEaN/DErP5nFosWPMHrUhdEU1HeEHwDe+sPW05yDxS/A8K/D9uHmgXt+H57+Bax8EwZF/1kiqU3q0ZVJPfyPh5OTNC6f0qcHAGtr6/JaLxEREcmNho4Ru5XAs7dvPe6uy2HEd+CDpTBqWF6qKCIiUnDq87mZqKqqor6+ni5dcn8J9uZ6eOQNqKiGsQVqBz1nxZsc360/B5f0yXvZlcvLefbAP/HcwX/mzYtnUrn8i5yVtWL23+k5Yh/m/OIk/jqpP0+fOoYF02/HOZezMtPlnGPtzOn0OPAY2hR3LHR1UtvhAPj4eVj7of//s/dh6UwYWtiztZ3bjHObKWpTvNX4oqIOfPbZW/mpRFUZVH8BfXaOq0B76DUU1i3JTx1EREREJC1fVPrHUvW6JSIirYjOfG4mpk2bxvDhwxkxYkTOynhnJex/I2yqg5Ji+NvZMKoAZz3/ruwjFtVU8NCAffJedvfRvdnj2vGUDOlO9bqNLLxjLnNOfIoJT32L9qXRd1FQ8ckSFv71Tnb+7gVMOOWnrF84jzdvugiAnU74YeTlZaJ83kvUrFlBr0NOKGg9GrX/ZVBTDnfvBm2KoL4Oxl4BexU2v3btOtOr5+689/6ddOs2nA4devHxsn9QVjaXkpKB+alEdbl/LE44ginuAhs35KcOIiIiItKomlr4yc1w1IGwQ0S9s4mIiGwL1PjcDNx6663MmzeP2267jaKiopyVs3NfmPtzWF8Ff5kLpz0Esy6EkdvnrMivmL+pnMtXv8uLQyfQvk3+T7zve9CArf4v/VofZnz9UZY/vpChp0d4k7iY+np67LIXu//wGgB67LwH5csXsvAvdxS88fmz5x6l09DRdBq8W0Hr0aj3/wzvPAjH/BF6j4BP58KzF0L3wbD7mQWt2n77Xcerr/2CJ56ciFkRpaW7MXDgJD7/PL9dyWAJ/ZY7B7nrylxEREREMlBXBydfCesr4O83Fbo2IiIi+aXG5wKbNm0aM2bMYOrUqWy/fW5bgdu3hWHhPndjdoTXl8HNz8M9389psVt5paqMtZtrGLng2S/HbcYxu3Itd5Z9ROXIYyluk7sG+ERtO7ejy7BSKpfmpuuNDr22o+vgXbca13XQLlQ+emtOyktX7YYyNrw+gwFnRXD3zFyb+VPY91IYcaL/v89+YIakAAAYDElEQVQo2PAxvHJtwRufu5QM5OsHP0hdXRW1tZV07NibOS9fQknnHfJTgdgZz5u+gI6lW8bXVHz1bGgRERERybu6OvjeFfDOIph1F/TsXugaiYiI5Jcanwto6tSpzJw5k6lTp7Ljjjvmvfx6B9V5vu/Vsd22Z0ynQ7cad/ryNxheXMLlfXahveX3bOjN1XVULFlPr323y8nz9x61P+XLFmw1rnzZQjr3y1O3DCmUzfoL1q49PcYdWdB6pKWuyne3Ea9NEbj6wtQnibZtO9G2bSdqajawavUcdv/aT/JTcKeeUNwVPpsPpeEzZHMtlC2GEcfkpw4iIiIiklRtHZz4c3h3Mcy6G/r1KnSNRERE8k+Nz2mqqIJFy/3f9fWwbDXMnQ89usHAfpk/30033cQzzzzDNddcQ5cuXSgrKwOgY8eOdOrUKcKaez97Ao4cAQNKoXwT/PENmLUQ/nFu5EU1qHtRe7oXtd9qXOc2RfQoas/IDt1yXv57179K34kD6bRdCdXrNrLg9rlsrqpjh+Nyc7vpnU+8kGfPGc9791/LwENO4PMFc1kw/TZGn/s/OSkvHc451s6YTo9xR1LUsaRg9UjbsKPglet8Nxu9RsCnb8OrN8GoUwtdM1ateglHPV27DKG8Yhlz//sbunYZxJDBx0VXSF01VH4W/nGw8XPYsALadYJOPWDoQbDgGSjpCyW9/d9FxdB/r+jqIGmp2LyZRRurAajHsay6lrkVVfRo15aBxe1ZV1vHsuoa1m/eDMCijTV0L6qiX/t29GvfrpBVFxERkSw0dIy4fS844TJ4/X148mbfI9rqtX7ebiXQMfrbzYiIiDRLanxO0xvvw8S4htqr7vLDad+E+6dk/nyPP/44ABdffPFW4ydPnswZZ5zRhJomt/oLOPkBWF0O3TrA6P7w9HnwjWbe3W/UNq2u5K1LZlGzfhPFpR3ovnsfDnj0KDr1z00XBT13G8OB1z/GvDuv5N37fk3nvgMYdc4Uhn87z63+cSree5XqVUsZfP4NBatDRg6bBrOvhH/9EKrWQMl2sMfZcMAvC10zamvL+e+8W6jauJr27bsxYIfDGD3qQtq0ibAhcf0ymBPXTcuHT/thwD6w5/dh2CH+bOd5j0FtlT8Deux50E5HNPn2RkUVE99d/OX/Vy1fzVXLV3Nan1LuH74jf1+3gdNjR6jA2Yv931cN6MuUgbm5+kJERERyp6FjxCnnwBMv+PF7nbz1cvddBZOPyl89RURECkmNz2maMAbcG9E93+zZs6N7sjTcf0pei8vIrKET8lbWXjcfnLeyYvqPm0T/cZPyXm4qXUbux17TFxa6Gukr7gKH3uKHZmbgwCMYOPCI3BbSazgcMzX1dDPY5Qg/SEFN6NYFN273lNMn9+3J5L4981gjERERyaXGjhGjPH4UERHZVuW3g10RERERERERERERaRXU+CwiIiIiIiIiIiIikVPjs4iIiIiIiIiIiIhETo3PIiIiIiIiIiIiIhI5NT6LiIiIiIiIiIiISOTMOVfoOjQ7ZlYOzC90PVqoXsDaQleihVK2uaFcc0fZ5o6yzR1lmzvKNneUbW4o19xRtrmjbHNH2eaOss0dZds69AI6O+d6F6LwtoUodBsw3zk3ptCVaInM7A1lmxvKNjeUa+4o29xRtrmjbHNH2eaOss0N5Zo7yjZ3lG3uKNvcUba5o2xbh/A6DypU+ep2Q0REREREREREREQip8ZnEREREREREREREYmcGp+Tu7vQFWjBlG3uKNvcUK65o2xzR9nmjrLNHWWbO8o2N5Rr7ijb3FG2uaNsc0fZ5o6ybR0K+jrrhoMiIiIiIiIiIiIiEjmd+SwiIiIiIiIiIiIikVPjs4iIiIiIiIiIiIhETo3PgZntYGb3mtknZlZtZkvN7BYzKy103ZoLMzvezKaZ2Ytm9oWZOTN7uJFlxprZP81snZlVmdk8M7vIzIoaWOabZjbLzDaYWYWZvWpmp0W/Rs2DmfU0s7PM7G9mtsjMNoZ1f8nMzjSzpO9TZZseM7vezGaY2fKQ7Toze9vMrjKznimWUbZZMLNTwueCM7OzUsyTcU5mdpqZvRbm3xCW/2Zu1qLwwvePSzGsTrGMttkMmNmBZvYXM1sVvvNXmdkzZjYpybzKthFmNrmBbTY2bE6ynLJNk5kdGbbRFeG77CMzm25m+6eYX9k2wrwzzOw/ZlYecnrbzC5IlZNy3Zo142MD28b3HTLJ1szamdmFZnafmc01sxprYF8sYdmMcjKzovB6zbMt+9X/NLOxTVnffMow2+FmdpmZzTR/LFFjZp+a2RNmNrGRclpVthnmOsDMbg/v5dXm98U+CcuebmbtGiinVeUK2X3WJix/j23ZHxuWYp6MczKzjmZ2tZnNN7NNZrbGzB41s12zWU/J+H00yBre934kjfKKzezdMP+KFPMcamY3mm9TWRfmfSnjlXPOtfoBGAp8CjjgceA6YGb4/0OgZ6Hr2BwGYG7IpBz4IPz9cAPzHwPUARXAPcBvQp4OmJ5imR+H6WuB24CbgeVh3A2FziBHuZ4b1u8T4A/AtcC9wPow/jFC/+zKNqt8a4D/hEyvA6YBr4f1XgkMULaR5DwgbLPlYb3PiiIn4IYwfXmY/zagLIz7caHXO0dZLg1ZTkkyXJpkfm2zmeX7i7CenwH3Ab/G34DjdeD/lG1Wme6eYnudAswI6/6Uss063+vj1vv3+O+yx/Dfb/XAyco2q1wfDOv3ach1KvAe2vfKJMNmeWxAC9h3yCRboHuY7oDVwDJS7Is1JSfAgOlsOUb+TXgdK8Lrekyhc8tBto+E6e8Bd+GP0/4a1tcBFyjbrHKdAGwAngHuxO+L3RW37T4PtFWumWebZNmj4pZ1wLAocgKKgZfCMq/j91X+CNQClcC+hc5tWxwyfB8NCtPnknwf/Pg0yrsxbttYkWKex8P0jcA74e+XMl63QofbHAbg3yHA8xPG3xTG31noOjaHAZgIDA8fThMaeSN0BdYA1cCYuPEdgJfDsicmLDMI2BS+QAbFjS8FFoVl9i90DjnI9eDwpdAmYXw/tnwBf1vZZp1vhxTjrwnrfbuybXLGBjwHLMbvrHzlgCebnICxYfwioDThucrC8w3K1XoVMM+lwNI059U2m1m2J4T1exbokmR6O2UbeeavhPU+WtlmlV8/YDO+QalPwrSJYb0/UrYZ53psLDugV9z4dsDfwrTJyrXRHJvdsQEtZN8hw2zbA0cA24X/p9BI43M2OQHfC8vMIW7/Gtg7vK5rSPLd2tyGDLOdDOyRZPxB+B8Aq2O5t/Zss9hm2yQZ3w7f8OyA7yjXzLNNWK43fv/hEWAWqRufM84J+HlYZnr8a4n/kdHhf7D5ymusIbrXmi2Nz/dnWdYE/EkMsZMhUzU+7w+MAIriylTjcxaBDwnhLUl8cwBd8L/2VAKdC13X5jSk8UY4I0x/IMm0g8O0FxLG/yqMvzqT52vJA3B5WO9pyjbybL8W1vtZZdvkLC8MX1zjSXHAk01ObDkr7fQky6R8vm19ILPGZ22z6efaBt/QVAn0VrZ5yXxkWOcVQJGyzSrDfcO6PZFi+hdAubLNONfY98uPkkyLbbdvKteMMp1AMzg2oAXuOzSWbZL5p9B443PGOQGzw/iJmTxfcx4yzTZh2WdIOElI2UaS64Vh2SuUa9Oyxf+YuhroScONzxnlhG8Y/TiMH5zJ82mI7rWmCY3P+B+ElxLaQmig8TlFmRk3PqvPZ7+zA/CMc64+foJzrhz/608nYL98V2wbF8v1X0mmzQaqgLFmVpzmMk8nzNNa1IbHurhxyjYaR4XHeXHjlG2GQp9e1wFTnXOzG5g1m5xac7bFZnaymV1uvh/HiZa8P0xts+kbCwwG/gl8br4P3ctCvsn6zVW2TfeD8HiPcy6+z2dlm76F+LPr9jGzXvETzGw8/kSJ5+JGK9v09AuPHyWZFhu3p5l1D38r16bLV4bKPT0Z5RRel7H41+nFdJZpBZIdp4GyzVrY143df2NewmTlmgEzm4y/yudc51xZA/Nlk9NQYCCwwDm3JM1lJHe2N7MfhOPGH5jZ6DSW+S3+KqIzc1y3L7XNV0HN2M7hcUGK6QuBw4Cd8P0WSnpS5uqcqzOzJfhT94fg+7JpbJlVZlYJ7GBmnZxzVTmoc7NiZm2BU8O/8V+yyjYLZnYpUAJ0A8YAB+B3aq6Lm03ZZiBsow/hu4e5vJHZM8rJzDoD/YEK59yqJM+3MDzulF3tm71++GzjLTGz051zL8SN0zabvr3D46fAW8Co+IlmNhvfN9pnYZSybQIz6wicjL8q4vcJk5Vtmpxz68zsMnxXcO+b2eP4y4uHAkfju5D5QdwiyjY9a8Pj4CTThsT9vQv+vhHKtelynqH2HdKTZU7D8Jdcf+ScS2xsTbVMi2VmOwKH4BvsZseNV7YZCD+q/hh/Jm1v4FB8Hn8EnoqbT7lmIGyfU/FnzD7eyOzZ5JROG1riMpI7h4bhS2Y2CzjNObcscWYzOw44DX91zFem54rOfPYNUeA7vE8mNr57iumSXDa5prtMtxTTW5rr8Jd+/tM59++48co2O5cCVwEX4Rue/wUcFtfQBMo2U78E9sD3i7mxkXkzzak1fzbfhz+g6Qd0xjeS3oW/zOlpM/ta3LzaZtPXJzyeC3QEvo4/a3Qk/t4P4/H91sUo26b5Dj6bp51zyxOmKdsMOOduAb6FP2nkbOBn+P7Ll+MvtVwTN7uyTU+sUeMSM+sRGxl+VL06br7S8Khcmy4fGbbmfYdM5PK1aPHZhjNF/4C/4doU59zncZOVbWZ64Y/Pfgmch/9h9Qb8sYWLm0+5psnM2gAP4LuPvSCNRZTttqsK+B9gL/z+Sim+P/rn8V12zAg/3HzJzPrijyufds7dk8/KqvG5cRYeXYNzSaayybXVvBZmdgHwE/zdZk/JdPHwqGzjOOf6OecM36D3LfyZNW+b2Z4ZPI2yDcxsH/zZzjc6516J4inDY6Y5tahcAZxzVzvnZjrnPnXOVTnn3nXOnYs/87Ejvi/HdGmb3SLWbYnhz3Ce4ZyrcM69BxyH75f4oBRdcCSjbBt2Tni8K4tllW0cM/t/wGPA/fgD8874A42PgD+Y2f9l8nThsbVn+wj+suCh+DPK7zazW/B3jJ/ElrO2NqdYPpFybbp8ZqjM06PtOUHoFuIhYBzwZ3xDaTaULeCc+zAcn7UFdgQuxu8/zI7/YTCTp8xg3paa68X4BsizE34YyZa+35op59wa59wvnXNvOefWh2E2vueGV/FntZ+VsNjv8Df2PDvP1VXjM42fddA1YT5JTza5prvMF02oV7NnZj/CXybzPr6T/nUJsyjbJggNen/Dfyj3xN9EIUbZpiGuu40FwJVpLpZpTo3N39gv7i3RneFxfNw4bbPpi+2Af+Sc+2/8hHDmfuwKk33Co7LNkpnthu8/cAW+j+1EyjZNZjYBuB74u3PuEufcR+FHqbfwP5qsBH5iZrGuIpRtGsJ9Xo7GXxW1Gv9D/xn4bfYAfNcmALGzypVr0+UjQ+07pCebnFr9MXNoeH4Yf+XJo8DJCWfngrLNinNus3NumXNuKr4rqf3wNxGMUa5pMLPhwDXAfc65ZPtfyeTys7nFZLstCd2nxLq8+/K40cxOxd/36kLn3Mp810uNzzA/PKbqj2Z4eEzVn40klzLX0HA1GH9zho/SXGY7/Jk+K1pyv3hmdhFwK/AuvuF5dZLZlG0EnHMf4xv4R8TdxEnZpqcEv767ApvMzMUG/KVzAL8L424J/2eUk3OuEt+oUhKmJ2qNn82xRpD4y6e0zaYvtt7rU0yPNU53TJhf2WYu1Y0GY5Rt+r4ZHp9PnBDW9TX8/vweYbSyTZNzrs45d6NzbnfnXEfnXFfn3OH4fYPdgY3Ae2F25dp0Oc9Q+w7pyTKnRfgrAYaE1yudZVqMsM5/Ak7E90d8UrL+cZVtJGI3q5sQG6Fc0zYC3x3M6fHHZ+EY7aAwz8Iw7tjwfzY5qQ2t+Yt1LRp/3Bi74vuBJNsHQP+4cZF3maLG5y0784eF/nG+ZGZd8JfUbMTfbETSNzM8Hp5k2nigE/Cyc646zWWOSJinxQk3FLoZf8nnxIQ+HOMp2+hsHx5jjSPKNj3VwD0phrfDPC+F/2NdcmSTU2vMtiGx7iDiD8y1zaZvNr5hY7iZtU8yfWR4XBoelW0WzKwD/izSevxnQDLKNn3F4bF3iumx8TXhUdk23SlAB+BR51xtGKdcmy5fGSr39GSUU3hdXsa/Tgems0xLEfYZHsOf8fwgcEqKH1ZjlG3T9A+PiY37yrVxS0l9jBY7qW16+H8pZJ3TYvwN53cys2Q37m2J2W5r9guP8ceNr5B6+wDfh3Ts//jv4mg451r9gL/U1gHnJ4y/KYy/s9B1bG4D/pdIh7+DarLpXfG/tlQDY+LGd8B/uDngxIRlBgOb8Jc6DoobX4r/Rc4B+xd63XOU55Vh/d4AejQyr7JNP9ddgH5JxrfBX5LkgDnKNtLMp4R1PqupOeEv23dhemnc+EHheTbFP1dLGPBnLHzlMwDfD97CkMflceO1zWaW78Nh/f43Yfyh+MbS9UB3ZdukjE8J6/lkA/Mo2/Tz/E5Yt9VA/4RpR4TtdiPQU9lmnG3XJOP2BtYB5cAQbbMZ5TmBZnBsQAvcd2gs2yTzTyHJvlhTcwK+F5aZA3SIG793eF3XJHtfNechje22GPhHmOf3QJs0nrPVZ5tGrvsCnZKMLwGeDcteo1wzz7aB5WaF5YYlmZZxTsDPwzLT498XwDFh/HvpvF80ZP9ah/dR+yTjDw7vBweMTbMsh7+SqLH5BoV5X8p0fSw8QatmZkPxOz19gCeAD/Av5ET8pQJjnXNlqZ+hdQiXZsQuz+gHfAP/S8qLYdxa59ylCfM/ht/wH8HvzB8N7BzGf8clbIBmdj7wW/wXyJ/xZ/IcD+yAv7HZpbQwZnYa/iZCm4FpJO8baalz7v64ZZRtGkI3Jr/Bn/G4GL/uffGXHQ3BH8wf4px7P24ZZdsEZjYF3/XG2c653ydMyzgnM7sRuATfD+djQHvgu/j+us93zt2as5UpgJDfz/BX5SzBN4AMBY7EH6D/EzjOOVcTt4y22TSZWR/8jvUw/HfXa/iG/ePwO1InOeemx82vbDNkZi/i+8w92jn3ZAPzKds0hKvy/g18Hf958Df8d9eu+C45DLjI+X4yY8so2zSY2av4hvt38dmOwN9ssBr4lnPu3wnzK9cEzfXYoCXsO2SR7c/wJ12A7zbma/jj29jNM19Ksl+WUU5mZvi+jo/H3xT9yTDvd/H7KN92zj3RpBXPg0yyNbP7gMnAWuB2kt9AbZZzblZCGa0u2wxzfRzfsPYC/gzaKmAA/kfV7vht9xvOuYqEMlpdrpD550GK55iFPwYe7pxblDAt45zMrBh/ZvNY/Al0M4CB+CsEaoCDnXOvZrG6rVqG76NZ+H2XWfj3BMBofOMzwJXOuf9Ns1wHrHTO7ZBk2gFsuXFhCfBt/A8SsS5ycM5NbrSQfLfeN9cB/2F3H7AK/2b5GH/TtwbPQm1NA1t+RU81LE2yzDh8Y8nn+B38d/B3YC1qoJyj8F9E5UAl8DpwWqHXv4C5OvxOjbLNPNuRwG34rkzW4i/f2hDWe0qq97eybVLmse056dk22eQEnBbmqwzLvQB8s9DrmqP8DsL3Kfgh/izcWvyZYs8Cp4L/0TjJctpm08+4B/7KpiX47/sy/A/P+ynbJme7a3j/L28oH2Wbca7tgIvwXcB9Eb7L1gBPAYcp26xz/SnwZvisrQ6fCXfSwFmxyvUr6zaFZnpswDa+75Bptmw5qzHVcH8UOQFtw+v1Tnj9Pg+vZ1pn1zWHIZNs08jVAVOUbca5Hgn8AX+i3wb8/u4a4DngHKBtA+W0qlwzzbaB54hty1858znbnPD3Sbka/yNXNf6YZTqwW6Ez21aHDN9HZ+L3BZcCFeE1WIb/ofbADMt1pDjzGf8DXIOfg+mUoTOfRURERERERERERCRyuuGgiIiIiIiIiIiIiEROjc8iIiIiIiIiIiIiEjk1PouIiIiIiIiIiIhI5NT4LCIiIiIiIiIiIiKRU+OziIiIiIiIiIiIiEROjc8iIiIiIiIiIiIiEjk1PouIiIiIiIiIiIhI5NT4LCIiIiIiIiIiIiKRU+OziIiIiIiIiIiIiEROjc8iIiIiIiIiIiIiErn/D8kIqlHmM1fQAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 1728x576 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Plotting the predicted fitness effect for the variants seen more than once\n",
    "plt.close('all')\n",
    "f, (ax1, ax2) = plt.subplots(2, 1, sharey=True, sharex = True, figsize=(24, 8))\n",
    "gs = matplotlib.gridspec.GridSpec(2, 1, width_ratios=[1], height_ratios=[10,0.5])\n",
    "ax1 = plt.subplot(gs[0])\n",
    "ax2 = plt.subplot(gs[1])\n",
    "gs.update(hspace=0.05)\n",
    "\n",
    "nonsense_list = []\n",
    "y_list = []\n",
    "x_list=[]\n",
    "y_list_top20 = []\n",
    "x_list_top20=[]\n",
    "y_list_nonsense = []\n",
    "x_list_nonsense = []\n",
    "x=[]\n",
    "y=[]\n",
    "OBS_EXP_DICT={}\n",
    "top20_lower_err = []\n",
    "top20_upper_err = []\n",
    "missense_lower_err = []\n",
    "missense_upper_err = []\n",
    "nonsense_lower_err = []\n",
    "nonsense_upper_err = []\n",
    "\n",
    "top20 = c1\n",
    "missense = c6\n",
    "nonsense = c10\n",
    "synonymous = c3\n",
    "\n",
    "axisfont=20\n",
    "domainfont = 12\n",
    "titlefont=12\n",
    "axislabelfont=12\n",
    "\n",
    "top_20_ASXL1 = []\n",
    "\n",
    "#ERROR NEEDS TO BE IN THE FORMAT: [tuple(lower_err),tuple(upper_err)]\n",
    "\n",
    "for k, v in ASXL1_missense_s.items():\n",
    "    if k in top_20_ASXL1:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_top20.append(codon_number)\n",
    "        y_list_top20.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        top20_lower_err.append(error_lower)\n",
    "        top20_upper_err.append(error_upper)\n",
    "\n",
    "    if k not in top_20_ASXL1:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list.append(codon_number)\n",
    "        y_list.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        missense_lower_err.append(error_lower)\n",
    "        missense_upper_err.append(error_upper)\n",
    "\n",
    "    missense_err = [tuple(missense_lower_err), tuple(missense_upper_err)]   \n",
    "    \n",
    "for k, v in ASXL1_nonsense_s.items():\n",
    "    if k in top_20_ASXL1:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_top20.append(codon_number)\n",
    "        y_list_top20.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        top20_lower_err.append(error_lower)\n",
    "        top20_upper_err.append(error_upper)\n",
    "    \n",
    "    top20_err = [tuple(top20_lower_err), tuple(top20_upper_err)]\n",
    "        \n",
    "    if k not in top_20_ASXL1:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_nonsense.append(codon_number)\n",
    "        y_list_nonsense.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        nonsense_lower_err.append(error_lower)\n",
    "        nonsense_upper_err.append(error_upper)\n",
    "\n",
    "    nonsense_err = [tuple(nonsense_lower_err), tuple(nonsense_upper_err)]        \n",
    "    \n",
    "msize = 12\n",
    "medgewidth = 2\n",
    "linew = 3\n",
    "csize = 4\n",
    "cthick = 3\n",
    "\n",
    "markers1, caps1, bars1 = ax1.errorbar(x_list, y_list, yerr = missense_err, fmt = 'o', ecolor = missense, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = missense, markerfacecolor = missense, zorder = 1)    \n",
    "\n",
    "markers2, caps2, bars2 = ax1.errorbar(x_list_nonsense, y_list_nonsense, yerr = nonsense_err, fmt = 'o', ecolor = nonsense, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = nonsense, markerfacecolor = nonsense, zorder = 2)    \n",
    "\n",
    "markers3, caps3, bars3 = ax1.errorbar(x_list_top20, y_list_top20, yerr = top20_err, fmt = 'o', ecolor = top20, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = top20, markerfacecolor = top20, zorder = 3)    \n",
    "\n",
    "# loop through bars and caps and set the alpha value\n",
    "a = 0.7\n",
    "[bar.set_alpha(a) for bar in bars1]\n",
    "[cap.set_alpha(a) for cap in caps1]\n",
    "[bar.set_alpha(a) for bar in bars2]\n",
    "[cap.set_alpha(a) for cap in caps2]\n",
    "[bar.set_alpha(a) for bar in bars3]\n",
    "[cap.set_alpha(a) for cap in caps3]\n",
    "\n",
    "#Only show the required axis lines\n",
    "ax1.spines['top'].set_visible(False)\n",
    "ax1.spines['right'].set_visible(False)\n",
    "\n",
    "for axis in ['bottom','left']:\n",
    "    ax1.spines[axis].set_linewidth(2)\n",
    "\n",
    "for axis in ['bottom','left']:\n",
    "    ax1.spines[axis].set_color(grey3)\n",
    "    \n",
    "ax1.set_ylabel('fitness effect (s)', fontsize = axisfont, labelpad = 10)\n",
    "ax1.set_title('Fitness effects across ASXL1', fontsize = 18)\n",
    "\n",
    "# ax1.set_yscale('log')\n",
    "\n",
    "# # STACKED BAR CHART\n",
    "# define the exons + create data\n",
    "exons = [{'1':(19-1)+1, '2': (47-20)+1, '3': (84-48)+1, '4':(125-85)+1, '5':(157-126)+1, '6': (189-158)+1,\\\n",
    "          '7': (240-190)+1,'8': (294-241)+1, '9': (327-295)+1, '10': (362-328)+1, '11': (573-363)+1, \\\n",
    "          '12': (1541-574)+1}]\n",
    "exon_orders = [['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12']]\n",
    "colors2 = [c4, c0, c3, c1, c5, c6,c7, c8, c9, c10, c2,c4]\n",
    "names = exons[0].keys()\n",
    "values = np.array([[data[name] for name in order] for data,order in zip(exons, exon_orders)])\n",
    "lefts = np.insert(np.cumsum(values, axis=1),0,0, axis=1)[:, :-1]\n",
    "orders = np.array(exon_orders)\n",
    "\n",
    "# #plot the bar chart\n",
    "for name, color in zip(names, colors2):\n",
    "    idx = np.where(orders == name)\n",
    "    value = values[idx]\n",
    "    left = lefts[idx]\n",
    "    ax2.bar(x=left, height=-10, width=value, bottom=-5, color=color, orientation = 'horizontal', label = name, edgecolor='k', linewidth=1, tick_label = name)\n",
    "    ax2.text(left + (value/2.), (-0.26), name, ha='center', va='center', fontsize = 14)\n",
    "    \n",
    "# # CONFIGURING THE GRAPH APPEARANCE\n",
    "# #Set the x and y axis limits\n",
    "ax1.set_xlim(0, 1542)\n",
    "ax1.set_ylim(-0.01, 25)\n",
    "ax2.set_xlim(0, 1542)\n",
    "ax2.set_ylim(-0.5, 0)\n",
    "\n",
    "#x-axis ticks\n",
    "x1_major_ticks = []\n",
    "x1_major_tick_labels = []\n",
    "x2_major_ticks = [0, 100, 200, 300, 400, 500, 600, 700, 800, 900, 1000, 1100, 1200, 1300, 1400, 1541]\n",
    "x2_major_tick_labels = [\"0\", \"100\", \"200\", \"300\", \"400\", \"500\", \"600\", \"700\", \"800\", \"900\", \"1000\", \"1100\", \"1200\", \"1300\", \"1400\", \"1541\"]\n",
    "ax1.set_xticks(x1_major_ticks)\n",
    "ax1.set_xticklabels(x1_major_tick_labels, fontsize = axisfont)\n",
    "ax2.set_xticks(x2_major_ticks)\n",
    "ax2.set_xticklabels(x2_major_tick_labels, fontsize = axisfont)\n",
    "\n",
    "#y-axis ticks\n",
    "y1_major_ticks = [0, 5, 10, 15, 20, 25]\n",
    "y1_major_tick_labels = [\"0\", \"5\", \"10\", \"15\", \"20\", \"25\"]\n",
    "ax1.set_yticks(y1_major_ticks)\n",
    "ax1.set_yticklabels(y1_major_tick_labels, fontsize = axisfont)\n",
    "\n",
    "#set panel legends\n",
    "legend_marker_size = 2\n",
    "legend_elements = [Line2D([0], [0], marker = '_', color = missense, alpha = 1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='missense'),\n",
    "                   Line2D([0], [0], marker = '_', color=nonsense, alpha=1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='nonsense'),\n",
    "                  Line2D([0], [0], marker = '_', color=top20, alpha=1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='top 20 observed variant')]\n",
    "    \n",
    "ax1.legend(ncol=1, handles=legend_elements, loc='upper left', frameon=0, fontsize = 20)\n",
    "\n",
    "ax1.grid(which='major', linestyle=':', linewidth='0.7', color=grey3, zorder = 0)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TP53 variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {
    "code_folding": [
     0,
     86,
     97,
     109
    ]
   },
   "outputs": [],
   "source": [
    "def expected_number_variants_missense_TP53(data, params):\n",
    "    s_value = params[0]\n",
    "    \n",
    "    a = 0\n",
    "    expected_number_by_variant_missense = {}\n",
    "    \n",
    "    Jaiswal_fmin = 0.056\n",
    "    Jaiswal_mean = 58\n",
    "    Jaiswal_sigma = 11.5\n",
    "    Jaiswal_total = 17182\n",
    "    \n",
    "    Genovese_fmin = 0.1194\n",
    "    Genovese_mean = 55.1\n",
    "    Genovese_total = 12380\n",
    "    Genovese_sigma = 11.95\n",
    "    \n",
    "    McKerrel_fmin = 0.008\n",
    "    McKerrel_mean = 51.88\n",
    "    McKerrel_total = 4219\n",
    "    McKerrel_sigma = 18.36\n",
    "    \n",
    "    Zink_fmin = 0.2363\n",
    "    Zink_mean = 60\n",
    "    Zink_total = 11262\n",
    "    Zink_sigma = 1.18907720e+01\n",
    "    \n",
    "    Coombs_fmin = 0.019\n",
    "    Coombs_mean = 59.82\n",
    "    Coombs_total = 5649\n",
    "    Coombs_sigma = 15.14\n",
    "    \n",
    "    Young_fmin = 0.0009\n",
    "    Young_mean = 63\n",
    "    Young_total = 89\n",
    "    Young_sigma = 6.8\n",
    "    \n",
    "    Desai_fmin = 0.0169\n",
    "    Desai_mean = 66.1\n",
    "    Desai_total = 424\n",
    "    Desai_sigma = 6 #estimated\n",
    "    \n",
    "    Acuna_fmin = 0.0015\n",
    "    Acuna_mean = 4.5\n",
    "    Acuna_total = 2014\n",
    "    \n",
    "    N = 9.40166610e+04\n",
    "\n",
    "    for i in data:\n",
    "        variant_mutation_rate = mutation_rates_variants_TP53(i)\n",
    "\n",
    "        if i in Genovese_missense_TP53:\n",
    "            Genovese_number = prevalence_averaged_over_ages(Genovese_fmin, s_value, N, variant_mutation_rate, Genovese_mean, Genovese_sigma, Genovese_total)\n",
    "        else:\n",
    "            Genovese_number = 0 \n",
    "\n",
    "        if i in Zink_missense_TP53:\n",
    "            Zink_number = prevalence_averaged_over_ages(Zink_fmin, s_value, N, variant_mutation_rate, Zink_mean, Zink_sigma, Zink_total)\n",
    "        else:\n",
    "            Zink_number = 0\n",
    "\n",
    "        if i in Coombs_missense_TP53:\n",
    "            Coombs_number = prevalence_averaged_over_ages(Coombs_fmin, s_value, N, variant_mutation_rate, Coombs_mean, Coombs_sigma, Coombs_total)\n",
    "        else:\n",
    "            Coombs_number = 0\n",
    "\n",
    "        if i in Young_missense_TP53:\n",
    "            Young_number = prevalence_averaged_over_ages(Young_fmin, s_value, N, variant_mutation_rate, Young_mean, Young_sigma, Young_total)\n",
    "        else:\n",
    "            Young_number = 0\n",
    "\n",
    "        if i in Desai_missense_TP53:\n",
    "            Desai_number = prevalence_averaged_over_ages(Desai_fmin, s_value, N, variant_mutation_rate, Desai_mean, Desai_sigma, Desai_total)\n",
    "        else:\n",
    "            Desai_number = 0\n",
    "\n",
    "        if i in Acuna_missense_TP53:\n",
    "            Acuna_number = prevalence_averaged_over_uniform_ages(Acuna_fmin, s_value, N, variant_mutation_rate, Acuna_mean, Acuna_total)\n",
    "        else:\n",
    "            Acuna_number = 0\n",
    "\n",
    "        expected_total = Genovese_number + Zink_number + Coombs_number + Young_number + Desai_number + Acuna_number\n",
    "\n",
    "        expected_number_by_variant_missense[i] = expected_total\n",
    "    \n",
    "    return expected_number_by_variant_missense\n",
    "\n",
    "def prob_observed_given_expected_missense_TP53(params, data): #data = list of variant (e.g. p.R882H) (or all_missense_variants_TP53)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_TP53_counts[variant])\n",
    "        expected_number = float(expected_number_variants_missense_TP53(data, params)[variant])\n",
    "        square_distance = ((observed_number - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance\n",
    "\n",
    "def prob_observed_given_expected_missense_TP53_upper_error(params, data): #data = list of variant (e.g. p.R882H) (or all_missense_variants_TP53)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_TP53_counts[variant])\n",
    "        observed_number_upper_error = observed_number+np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_missense_TP53(data, params)[variant])\n",
    "        square_distance = ((observed_number_upper_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   \n",
    "\n",
    "def prob_observed_given_expected_missense_TP53_lower_error(params, data): #data = list of variant (e.g. p.R882H) (or all_missense_variants_TP53)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_TP53_counts[variant])\n",
    "        observed_number_lower_error = observed_number-np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_missense_TP53(data, params)[variant])\n",
    "        square_distance = ((observed_number_lower_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {
    "code_folding": [
     0,
     91,
     102,
     114
    ]
   },
   "outputs": [],
   "source": [
    "def expected_number_variants_nonsense_TP53(data, params):\n",
    "    s_value = params[0]\n",
    "    \n",
    "    a = 0\n",
    "    expected_number_by_variant_nonsense = {}\n",
    "    \n",
    "    Jaiswal_fmin = 0.056\n",
    "    Jaiswal_mean = 58\n",
    "    Jaiswal_sigma = 11.5\n",
    "    Jaiswal_total = 17182\n",
    "    \n",
    "    Genovese_fmin = 0.1194\n",
    "    Genovese_mean = 55.1\n",
    "    Genovese_total = 12380\n",
    "    Genovese_sigma = 11.95\n",
    "    \n",
    "    McKerrel_fmin = 0.008\n",
    "    McKerrel_mean = 51.88\n",
    "    McKerrel_total = 4219\n",
    "    McKerrel_sigma = 18.36\n",
    "    \n",
    "    Zink_fmin = 0.2363\n",
    "    Zink_mean = 60\n",
    "    Zink_total = 11262\n",
    "    Zink_sigma = 1.18907720e+01\n",
    "    \n",
    "    Coombs_fmin = 0.019\n",
    "    Coombs_mean = 59.82\n",
    "    Coombs_total = 5649\n",
    "    Coombs_sigma = 15.14\n",
    "    \n",
    "    Young_fmin = 0.0009\n",
    "    Young_mean = 63\n",
    "    Young_total = 89\n",
    "    Young_sigma = 6.8\n",
    "    \n",
    "    Desai_fmin = 0.0169\n",
    "    Desai_mean = 66.1\n",
    "    Desai_total = 424\n",
    "    Desai_sigma = 6 #estimated\n",
    "    \n",
    "    Acuna_fmin = 0.0015\n",
    "    Acuna_mean = 4.5\n",
    "    Acuna_total = 2014\n",
    "\n",
    "    N = 9.40166610e+04\n",
    "    \n",
    "    for i in data:\n",
    "        variant_mutation_rate = mutation_rates_variants_TP53(i)\n",
    "\n",
    "        if i in Jaiswal_nonsense_TP53:\n",
    "            Jaiswal_number = prevalence_averaged_over_ages(Jaiswal_fmin, s_value, N, variant_mutation_rate, Jaiswal_mean, Jaiswal_sigma, Jaiswal_total)\n",
    "        else:\n",
    "            Jaiswal_number = 0\n",
    "\n",
    "        if i in Genovese_nonsense_TP53:\n",
    "            Genovese_number = prevalence_averaged_over_ages(Genovese_fmin, s_value, N, variant_mutation_rate, Genovese_mean, Genovese_sigma, Genovese_total)\n",
    "        else:\n",
    "            Genovese_number = 0 \n",
    "\n",
    "        if i in Zink_nonsense_TP53:\n",
    "            Zink_number = prevalence_averaged_over_ages(Zink_fmin, s_value, N, variant_mutation_rate, Zink_mean, Zink_sigma, Zink_total)\n",
    "        else:\n",
    "            Zink_number = 0\n",
    "\n",
    "        if i in Coombs_nonsense_TP53:\n",
    "            Coombs_number = prevalence_averaged_over_ages(Coombs_fmin, s_value, N, variant_mutation_rate, Coombs_mean, Coombs_sigma, Coombs_total)\n",
    "        else:\n",
    "            Coombs_number = 0\n",
    "\n",
    "        if i in Young_nonsense_TP53:\n",
    "            Young_number = prevalence_averaged_over_ages(Young_fmin, s_value, N, variant_mutation_rate, Young_mean, Young_sigma, Young_total)\n",
    "        else:\n",
    "            Young_number = 0\n",
    "\n",
    "        if i in Desai_nonsense_TP53:\n",
    "            Desai_number = prevalence_averaged_over_ages(Desai_fmin, s_value, N, variant_mutation_rate, Desai_mean, Desai_sigma, Desai_total)\n",
    "        else:\n",
    "            Desai_number = 0\n",
    "\n",
    "        if i in Acuna_nonsense_TP53:\n",
    "            Acuna_number = prevalence_averaged_over_uniform_ages(Acuna_fmin, s_value, N, variant_mutation_rate, Acuna_mean, Acuna_total)\n",
    "        else:\n",
    "            Acuna_number = 0\n",
    "\n",
    "        expected_total = Jaiswal_number + Genovese_number + Zink_number + Coombs_number + Young_number + Desai_number + Acuna_number\n",
    "\n",
    "        expected_number_by_variant_nonsense[i] = expected_total\n",
    "    \n",
    "    return expected_number_by_variant_nonsense\n",
    "\n",
    "def prob_observed_given_expected_nonsense_TP53(params, data): #data = list of variant (e.g. p.R882H) (or all_nonsense_variants_TP53)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_TP53_counts[variant])\n",
    "        expected_number = float(expected_number_variants_nonsense_TP53(data, params)[variant])\n",
    "        square_distance = ((observed_number - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance\n",
    "\n",
    "def prob_observed_given_expected_nonsense_TP53_upper_error(params, data): #data = list of variant (e.g. p.R882H) (or all_nonsense_variants_TP53)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_TP53_counts[variant])\n",
    "        observed_number_upper_error = observed_number+np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_nonsense_TP53(data, params)[variant])\n",
    "        square_distance = ((observed_number_upper_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   \n",
    "\n",
    "def prob_observed_given_expected_nonsense_TP53_lower_error(params, data): #data = list of variant (e.g. p.R882H) (or all_nonsense_variants_TP53)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_TP53_counts[variant])\n",
    "        observed_number_lower_error = observed_number-np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_nonsense_TP53(data, params)[variant])\n",
    "        square_distance = ((observed_number_lower_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#CALCULATING S FOR MISSENSE VARIANTS\n",
    "TP53_missense_s = {}\n",
    "a = 0\n",
    "for var in all_missense_variants_TP53:\n",
    "    initial_guess=[0.05]\n",
    "    if var in all_studies_TP53_counts.keys():\n",
    "        observed_number = float(all_studies_TP53_counts[var])\n",
    "#         print(observed_number)\n",
    "        if observed_number >1:\n",
    "            variant_mutation_rate = mutation_rates_variants_TP53(var)\n",
    "            \n",
    "            MLE = scipy.optimize.minimize(prob_observed_given_expected_missense_TP53, initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value = MLE.x[0]\n",
    "\n",
    "            upper_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_missense_TP53_upper_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_upper = upper_error_MLE.x[0]\n",
    "\n",
    "            lower_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_missense_TP53_lower_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_lower = lower_error_MLE.x[0]\n",
    "\n",
    "#             print('a = ', a)\n",
    "            TP53_missense_s[var]=(s_value, observed_number, variant_mutation_rate, s_value_upper, s_value_lower) \n",
    "    a = a+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#CALCULATING S FOR NONSENSE VARIANTS\n",
    "TP53_nonsense_s = {}\n",
    "a = 0\n",
    "for var in all_nonsense_variants_TP53:\n",
    "    initial_guess=[0.05]\n",
    "    if var in all_studies_TP53_counts.keys():\n",
    "        observed_number = float(all_studies_TP53_counts[var])\n",
    "#         print(observed_number)\n",
    "        if observed_number >1:\n",
    "            variant_mutation_rate = mutation_rates_variants_TP53(var)\n",
    "            \n",
    "            MLE = scipy.optimize.minimize(prob_observed_given_expected_nonsense_TP53, initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value = MLE.x[0]\n",
    "\n",
    "            upper_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_nonsense_TP53_upper_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_upper = upper_error_MLE.x[0]\n",
    "\n",
    "            lower_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_nonsense_TP53_lower_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_lower = lower_error_MLE.x[0]\n",
    "\n",
    "#             print('a = ', a)\n",
    "            TP53_nonsense_s[var]=(s_value, observed_number, variant_mutation_rate, s_value_upper, s_value_lower) \n",
    "    a = a+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#Exac - retrieve the frequencies in Exac for each of the variants\n",
    "import csv\n",
    "filename=\"Data files/TP53_Exac.csv\"\n",
    "\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    Exac_TP53_results={} \n",
    "    \n",
    "    for row in read_reader:\n",
    "#         print(row)\n",
    "        if row_count>1:\n",
    "            variant = row[6]\n",
    "            allele_freq = float(row[14])\n",
    "            african_freq = float(row[15])/float(row[16])\n",
    "            east_asian_freq = float(row[18])/float(row[19])\n",
    "            european_freq = float(row[21])/float(row[22])\n",
    "#             print(row[24])\n",
    "            if row[25]=='0':\n",
    "                finnish_freq = '-'\n",
    "            if row[25]!='0':\n",
    "#                 print(row)\n",
    "                finnish_freq = float(row[24])/float(row[25])\n",
    "            latino_freq = float(row[27])/float(row[28])\n",
    "            south_asian_freq = float(row[33])/float(row[34])\n",
    "            Exac_TP53_results[variant]=(allele_freq, african_freq, east_asian_freq, european_freq, finnish_freq, latino_freq, south_asian_freq)\n",
    "\n",
    "        row_count=row_count+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#COSMIC - retrieve the frequencies in COSMIC for each of the variants\n",
    "import csv\n",
    "filename=\"Data files/TP53_COSMIC_variant_counts.csv\"\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    TP53_COSMIC_results={} \n",
    "    \n",
    "    for row in read_reader:\n",
    "#         print(row)\n",
    "        if row_count>1:\n",
    "            variant = row[1]\n",
    "            counts = int(row[2])\n",
    "            TP53_COSMIC_results[variant]=counts\n",
    "\n",
    "        row_count=row_count+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Create dictionary of results and save to .csv file\n",
    "TP53_s_dictionary={}\n",
    "for k, v in TP53_missense_s.items():\n",
    "#     print(k)\n",
    "    if k in Exac_TP53_results.keys():\n",
    "        exac_freq = round((Exac_TP53_results[k][0])*100, 4)\n",
    "        african_freq = round((Exac_TP53_results[k][1])*100, 4)\n",
    "        east_asian_freq = round((Exac_TP53_results[k][2])*100, 4)\n",
    "        european_freq = round((Exac_TP53_results[k][3])*100, 4)\n",
    "        finnish_freq = round((Exac_TP53_results[k][4])*100, 4)\n",
    "        latino_freq = round((Exac_TP53_results[k][5])*100, 4)\n",
    "        south_asian_freq = round((Exac_TP53_results[k][6])*100, 4)\n",
    "    if k not in Exac_TP53_results.keys():\n",
    "        exac_freq = '-'\n",
    "        african_freq = '-'\n",
    "        east_asian_freq = '-'\n",
    "        european_freq = '-'\n",
    "        finnish_freq = '-'\n",
    "        latino_freq = '-'\n",
    "        south_asian_freq = '-'\n",
    "    if k in TP53_COSMIC_results.keys():\n",
    "        COSMIC_number = TP53_COSMIC_results[k]\n",
    "    if k not in TP53_COSMIC_results.keys():\n",
    "        COSMIC_number = '-'\n",
    "    TP53_s_dictionary[k]=(round(v[0]*100, 2),round(v[4]*100, 2),round(v[3]*100, 2), v[1], v[2], COSMIC_number, exac_freq, \\\n",
    "                            african_freq, east_asian_freq, european_freq, finnish_freq, latino_freq, south_asian_freq)\n",
    "\n",
    "for k, v in TP53_nonsense_s.items():\n",
    "    if k in Exac_TP53_results.keys():\n",
    "        exac_freq = round((Exac_TP53_results[k][0])*100, 4)\n",
    "        african_freq = round((Exac_TP53_results[k][1])*100, 4)\n",
    "        east_asian_freq = round((Exac_TP53_results[k][2])*100, 4)\n",
    "        european_freq = round((Exac_TP53_results[k][3])*100, 4)\n",
    "        finnish_freq = round((Exac_TP53_results[k][4])*100, 4)\n",
    "        latino_freq = round((Exac_TP53_results[k][5])*100, 4)\n",
    "        south_asian_freq = round((Exac_TP53_results[k][6])*100, 4)\n",
    "    if k not in Exac_TP53_results.keys():\n",
    "        exac_freq = '-'\n",
    "        african_freq = '-'\n",
    "        east_asian_freq = '-'\n",
    "        european_freq = '-'\n",
    "        finnish_freq = '-'\n",
    "        latino_freq = '-'\n",
    "        south_asian_freq = '-'\n",
    "    if k in TP53_COSMIC_results.keys():\n",
    "        COSMIC_number = TP53_COSMIC_results[k]\n",
    "    if k not in TP53_COSMIC_results.keys():\n",
    "        COSMIC_number = '-'\n",
    "    TP53_s_dictionary[k]=(round(v[0]*100, 2),round(v[4]*100, 2),round(v[3]*100, 2), v[1], v[2], COSMIC_number, exac_freq, \\\n",
    "                            african_freq, east_asian_freq, european_freq, finnish_freq, latino_freq, south_asian_freq)\n",
    "\n",
    "TP53_df = pd.DataFrame.from_dict(TP53_s_dictionary, orient='index', columns=['s (%)', 's lower value (%)', 's upper value (%)','observed number', 'site-specific mutation rate', 'Number of times in COSMIC',\\\n",
    "                                                                                'Exac Overall freq (%)', 'Exac African freq (%)', 'Exac East Asian freq (%)',\\\n",
    "                                                                                'Exac European freq (%)', 'Exac Finnish freq (%)', 'Exac Latino freq (%)', 'Exac South Asian freq (%)'])\n",
    "TP53_df_sorted = TP53_df.sort_values(by = ['s (%)'], ascending = False)\n",
    "\n",
    "TP53_df_sorted.to_csv('Data files/TP53_s_across_whole_gene.csv')\n",
    "# TP53_s_dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s (%)</th>\n",
       "      <th>s lower value (%)</th>\n",
       "      <th>s upper value (%)</th>\n",
       "      <th>observed number</th>\n",
       "      <th>site-specific mutation rate</th>\n",
       "      <th>Number of times in COSMIC</th>\n",
       "      <th>Exac Overall freq (%)</th>\n",
       "      <th>Exac African freq (%)</th>\n",
       "      <th>Exac East Asian freq (%)</th>\n",
       "      <th>Exac European freq (%)</th>\n",
       "      <th>Exac Finnish freq (%)</th>\n",
       "      <th>Exac Latino freq (%)</th>\n",
       "      <th>Exac South Asian freq (%)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>p.N235S</th>\n",
       "      <td>16.09</td>\n",
       "      <td>13.87</td>\n",
       "      <td>17.80</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.094130e-09</td>\n",
       "      <td>6</td>\n",
       "      <td>0.0239</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0345</td>\n",
       "      <td>0.0913</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.S241C</th>\n",
       "      <td>15.00</td>\n",
       "      <td>11.67</td>\n",
       "      <td>16.96</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.368770e-10</td>\n",
       "      <td>9</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0152</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.K132E</th>\n",
       "      <td>13.60</td>\n",
       "      <td>10.13</td>\n",
       "      <td>15.45</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.441820e-10</td>\n",
       "      <td>9</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.Y220C</th>\n",
       "      <td>12.67</td>\n",
       "      <td>10.73</td>\n",
       "      <td>13.94</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.472060e-09</td>\n",
       "      <td>145</td>\n",
       "      <td>0.0025</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0045</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R110L</th>\n",
       "      <td>12.48</td>\n",
       "      <td>9.77</td>\n",
       "      <td>13.91</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.064910e-09</td>\n",
       "      <td>7</td>\n",
       "      <td>0.0017</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.G245C</th>\n",
       "      <td>12.27</td>\n",
       "      <td>9.61</td>\n",
       "      <td>13.66</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.163110e-09</td>\n",
       "      <td>10</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.V216M</th>\n",
       "      <td>8.85</td>\n",
       "      <td>6.54</td>\n",
       "      <td>10.10</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.211140e-09</td>\n",
       "      <td>60</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R248Q</th>\n",
       "      <td>8.63</td>\n",
       "      <td>7.67</td>\n",
       "      <td>9.30</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.415120e-08</td>\n",
       "      <td>189</td>\n",
       "      <td>0.0058</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0231</td>\n",
       "      <td>0.0075</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.W146*</th>\n",
       "      <td>8.04</td>\n",
       "      <td>5.93</td>\n",
       "      <td>9.01</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.610690e-09</td>\n",
       "      <td>9</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R273H</th>\n",
       "      <td>7.77</td>\n",
       "      <td>7.07</td>\n",
       "      <td>8.27</td>\n",
       "      <td>5.0</td>\n",
       "      <td>3.269290e-08</td>\n",
       "      <td>106</td>\n",
       "      <td>0.0026</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0047</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R110C</th>\n",
       "      <td>7.67</td>\n",
       "      <td>6.30</td>\n",
       "      <td>8.39</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.415120e-08</td>\n",
       "      <td>-</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0151</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R248W</th>\n",
       "      <td>7.67</td>\n",
       "      <td>6.30</td>\n",
       "      <td>8.39</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.415120e-08</td>\n",
       "      <td>129</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R283H</th>\n",
       "      <td>7.33</td>\n",
       "      <td>5.99</td>\n",
       "      <td>7.99</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.882290e-08</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0066</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0262</td>\n",
       "      <td>0.0183</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         s (%)  s lower value (%)  s upper value (%)  observed number  \\\n",
       "p.N235S  16.09              13.87              17.80              5.0   \n",
       "p.S241C  15.00              11.67              16.96              2.0   \n",
       "p.K132E  13.60              10.13              15.45              2.0   \n",
       "p.Y220C  12.67              10.73              13.94              4.0   \n",
       "p.R110L  12.48               9.77              13.91              2.0   \n",
       "p.G245C  12.27               9.61              13.66              2.0   \n",
       "p.V216M   8.85               6.54              10.10              2.0   \n",
       "p.R248Q   8.63               7.67               9.30              4.0   \n",
       "p.W146*   8.04               5.93               9.01              2.0   \n",
       "p.R273H   7.77               7.07               8.27              5.0   \n",
       "p.R110C   7.67               6.30               8.39              2.0   \n",
       "p.R248W   7.67               6.30               8.39              2.0   \n",
       "p.R283H   7.33               5.99               7.99              2.0   \n",
       "\n",
       "         site-specific mutation rate Number of times in COSMIC  \\\n",
       "p.N235S                 1.094130e-09                         6   \n",
       "p.S241C                 4.368770e-10                         9   \n",
       "p.K132E                 5.441820e-10                         9   \n",
       "p.Y220C                 1.472060e-09                       145   \n",
       "p.R110L                 1.064910e-09                         7   \n",
       "p.G245C                 1.163110e-09                        10   \n",
       "p.V216M                 3.211140e-09                        60   \n",
       "p.R248Q                 1.415120e-08                       189   \n",
       "p.W146*                 4.610690e-09                         9   \n",
       "p.R273H                 3.269290e-08                       106   \n",
       "p.R110C                 1.415120e-08                         -   \n",
       "p.R248W                 1.415120e-08                       129   \n",
       "p.R283H                 1.882290e-08                         2   \n",
       "\n",
       "        Exac Overall freq (%) Exac African freq (%) Exac East Asian freq (%)  \\\n",
       "p.N235S                0.0239                     0                        0   \n",
       "p.S241C                0.0008                     0                        0   \n",
       "p.K132E                0.0008                     0                        0   \n",
       "p.Y220C                0.0025                     0                        0   \n",
       "p.R110L                0.0017                     0                        0   \n",
       "p.G245C                     -                     -                        -   \n",
       "p.V216M                     -                     -                        -   \n",
       "p.R248Q                0.0058                     0                   0.0231   \n",
       "p.W146*                     -                     -                        -   \n",
       "p.R273H                0.0026                     0                        0   \n",
       "p.R110C                0.0008                     0                        0   \n",
       "p.R248W                0.0008                     0                        0   \n",
       "p.R283H                0.0066                     0                        0   \n",
       "\n",
       "        Exac European freq (%) Exac Finnish freq (%) Exac Latino freq (%)  \\\n",
       "p.N235S                 0.0345                0.0913                    0   \n",
       "p.S241C                      0                0.0152                    0   \n",
       "p.K132E                 0.0015                     0                    0   \n",
       "p.Y220C                 0.0045                     0                    0   \n",
       "p.R110L                 0.0015                     0                    0   \n",
       "p.G245C                      -                     -                    -   \n",
       "p.V216M                      -                     -                    -   \n",
       "p.R248Q                 0.0075                     0                    0   \n",
       "p.W146*                      -                     -                    -   \n",
       "p.R273H                 0.0047                     0                    0   \n",
       "p.R110C                      0                0.0151                    0   \n",
       "p.R248W                 0.0015                     0                    0   \n",
       "p.R283H                  0.003                     0               0.0262   \n",
       "\n",
       "        Exac South Asian freq (%)  \n",
       "p.N235S                         0  \n",
       "p.S241C                         0  \n",
       "p.K132E                         0  \n",
       "p.Y220C                         0  \n",
       "p.R110L                    0.0061  \n",
       "p.G245C                         -  \n",
       "p.V216M                         -  \n",
       "p.R248Q                         0  \n",
       "p.W146*                         -  \n",
       "p.R273H                         0  \n",
       "p.R110C                         0  \n",
       "p.R248W                         0  \n",
       "p.R283H                    0.0183  "
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TP53_df_sorted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABZoAAAHvCAYAAADdMGqTAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAIABJREFUeJzs3Xd4VFX+x/H3SQIJAVIooQkEUUBZBUEEKUuoCipNVAQXYt2FFWRVbKsGUFdYK4igPysquq4NFIWAEqqA9N4h1BBIQmghkHJ+f9yZ2ZnMhBZKlM/refJM5txzz/3ee3LngW9OvtdYaxEREREREREREREROVtBFzsAEREREREREREREfl9U6JZRERERERERERERIpEiWYRERERERERERERKRIlmkVERERERERERESkSJRoFhEREREREREREZEiUaJZRERERERERERERIpEiWYRERGR3zFjTIgxxhpj3r/YsRRHxpiBxpgNxpgTrut0mau9nTFmoTHmiKv9nosdq4iIiIjI75kSzSIiIiLFkDEmzpUALeyr2Un2DTLGDDXGdLmQMRc3xpgOwGhgNfAQ8BcgwxhTAfgWCAMedbXPO08xNHLNRY3zMf6lyBiz6xT3hvdXS69fxnh/ZRtjNhpjXjXGRJ/B+FEF+t5rjJnm2ifbGLPfGPOrMaavMUb/1xIREZFLSsjFDkBERERETuoL4KcA7ZsBrLW5xphSQK7XtiAgAfgA+P68R1h8dXC93mutPeRuNMbEARHAc9ba8319GuHMxc/AjvN8rEvFQKC01/v6wFPA18CkAn03eH2/BHjT9X054FbgMaC9MaaJtTbHq+8aYESAYx8t8L4xkA6MAfYDZV3jjgdaAH89vVMSERER+f1TollERESkeFtqrf3sZB2stdkXKpjfmcpAnneS2asdIOMCx/O7Y4wxQGlr7ZGLHYubtfY77/fGmPY4ieYVge4VY4z7/zy7CmwfbYz5CeiEkxz2Hnfvqe47VywPB2h+0xgzFXjAGPNPa23aqcYRERER+SPQn3OJiIiI/I4VrNFsjLkCcK/MvN/rT/5zC/Z3lRWYY4zJMsakGWP+zxhTOsAxqhlj3jHG7HTVOt7tel+hQL/yxphRxpitrjIC6caYxcaYRwv0izfGLDLGZBpjjhpjthhjPjPGlD/Nc77JGDPdGHPQdZwVxpiHvLZfYYyxOCUxgr2uwc/GmF04K70B5nhfG9e+YcaYZ40xa11jHzDGfG+MaRAgDmOM+asx5jfj1Ho+bIxZaYxJcG1/EXivwLG856qUMWa4cWpIZ7mOtcoYE2glbcFjh7jinGOMSXXNy3ZjzNvGmHKF7HOHMWaW67plGWPWu+YrxLW9vSu+vxintvU64Dgw2GuMONd1POQaY4kxJj7Asa4xxnzj+lk5boxJMcbMMMZ08upz1ud/Dk1zvV5RcIPrGpc9y3F34PxfK/JsAxMRERH5vdGKZhEREZHiLbxgQhc4bq09XEj/vUA/nD/dn8n/kqr5Bfo1Brq7tn8GtAUexCnBMcDdyRhTC/gVCHb13QpcCfQH2hin5IB7xfC3wI3AO8BKnPIGVwFxwOuu8e4FPgRmAc8Dx4AaQGegAk4ZgkIZY/oDb7tiegHIAm4C3jXG1LLWPu26Bn9xxdjMdT0AUnBKZtwC3O/af6P72hhjSuIkHpsCn+DUd452XZdfjTEtrbXLXH0NTlmTu4D5wEtAput8bweGAV/hrJ72Pha4yp64rlNf4GPXGCVc17btya6BSxhO2YdvgIk4JR1uwKlF3cIUKAVhjBkJPIFTEuI1IBWo7Yr1n4D3iuXHXOf9getabneN0Q2nPEUK8IrrmHcDH7muvTvBXhGYAeS5znEHztw2ccU45Ryc/7lypeu14KrjFjg/WyWMMZk41/gZa21KoEGMU7s5BOe6dcI5r3XAtvMRtIiIiEhxZKy1FzsGERERESnAOHWEkwrZ/KW1tperXwjOCuYPrLUPFNbmNa57Wz7Q1Fq72GtbIk5SOMpae8zV9iNOUrqRtXaPV9+mOMneBGvti65VtOnAW9baQSc5r++BlkBFa23eaV4O976XAVtc59+3wLa3cerh1rbWuhOjnwG9rLUhBfo+gLPSuJW1dq5X+xBgJNDRWvuzV3sUzgMF11tr27vaegMTcBL691lr8736B7nfF3Ys17ZDwExr7Rk/tNE4D5oLdc+TV/tfcRK4t1trv3W1Ncd52OHPwK3W2uNe/Y11/YfAOCUopuPMYz3vkg/GmBI4SdNw4Gpr7V5XeyjOLw2aAFdaa7caY3rgJMA9MRRyDmd9/gXGccf9nLX2xQDb3T/zU3ASwOAkhLvg1GE+4oo9zdV/CjAXWA+UBNrg/LJgN3CD+9wLHGM54F71nu+K52/W2uSinJuIiIjI74lKZ4iIiIgUb/+H81A77y+/ZNpZmOudZHaZgZNYqwngSh53wlnNecIYU8H9hZPw3QZ0dO2bhZPMa2aMqXmS4x7EeWBaZ9eq4DNxhyu+D71jccXzA86q63ZnOKa3e3BW/C4vMHYI8AvQ2pVYBegDWOBx7yQzQMH3J3EQuMYYU/9MA7XW5nv9MiDYGBPlinWGq0tTr+59XK9PeSeZXeMEWnXycYC6wk2AasD73olW13iv4vy/wp0wPuh67XyK0hNnff5nqRPOA/v246wufxVYBXTwPl9rbSdr7UvW2m+stV9Yax8C7gWq4zzYMZC/4twL/XBWfZcEos7bmYiIiIgUQyqdISIiIlK8bfJeXXsObQ3Q5i5b4a6VXA8wOEm0vxYyTh44DyQ0Ti3m14FkY8wanKTnd9Za75XZL+KsaP4eSDPGzMJZafrlaTxw7irXa2ErvQEqnWKMk6mHkyDcf5I+5XBKR1yJ83C5ojzo7RGcFdGrjTFbcM7re2ByIQlgH8aYXjhlLhrglJ3wFu31/ZU487TyNOPaGKCtlut1TYBtq12vl7teZwCf46wC7muMWYRTkuRLa+16r/2KdP5n4Vf+lyg+DiRba3eezo7W2k+MU3P7lkK2L/R6+4kx5hVgrjHmGmutymeIiIjIJUGJZhEREZFL08nKVpgCr+Nx6jgHkuX+xlo7xhjzHU4y7s/AncBAY8wEa+09rj4bjDH1gPY4q49bA+8DQ40xfz5FUs4dTx9gXyF9NhfSfjqCgOXAkJP0yfCKpUjJUGvtt65Ee2ec69ABeACYaYzp6F1juSBjzJ04NaIXAIOAXUA2TqL8R3z/cvFMV45nBWg77TFcSeI+rof6dQJa4dSHfs4YM9BaO87V76zP/yztL+IvbbbjlJE5HeOBx3FKdQwrwjFFREREfjeUaBYRERH54zlXq0E3u8YqcboJOmvtbpxyH//nqo07ASfp+Jr7QXqucgs/ur4wxnQBJgGDcVa5FmaT67WoCcOTjV8R+OU0VtRuwCkNUeEUq5pPOo61Nh34FPjUVUrkFZxVyrcC351k17/gJITbWGuz3Y3GmD8VEmt74Bpg6cniOYktrtdAZS6udr36rJK31q7CKU3xb1cZlt9waiKP8+pztud/Qbliq43zEMXTUcr1Wu78RCQiIiJS/KhGs4iIiMgfjOshe9kUMcllrU3FKXlwhzGmScHtxlHR9X24MaZUgf1zcRKNuGNx1REuaKl3n5P4EjgBDDfGhAWIJ8oYU/IUY5zMJzh1iAMmu40x3mU5JuCs8v13wVrTBd67y4GUK9AnxBgT6d3mSm4vD9Q/gDych855/j3vOu6zAfp+7np9ueD1OYM62YtwHoZ3vzEmxmv/kjgrd/Nxyl5gjClXcFxrbQaQDJQxxpQ8B+d/XhhjyheyaRBQBacWuLtvCVcCPZCBrtcF5zA8ERERkWJNK5pFRERE/pgWADcZY54AdgJ51tr/nsU4fwXm4tSb/QRYhvNvyFpAN+ADnLrLVwM/u0pnrAEOuNr646yGnecab4YxZj8wxxVXOZwHreVTeHkOAKy1240xDwPvAGuNMZ8BO3BWIV8LdAXq4JSROBuv46z8fcMY0x6YCRwGauCU+TiMU94B4D9AD1fsdYwxP+A83K6Oq28DV7/fcFY1P+tKyh/lfw9S3GGM+R4nuboPp8Zxf5zyHD+eItavXec7wxjzKRAKdAf8EvDW2l+NMa/hrBRebIz5CtiLM4d3GGOuO1V9bGttrjFmIPAVsMgY857rXHoBNwDDrbXuFc33AQ+7fhY2A7lAG9d1+dxae8L1C4einP/5cq8xpi8wFScxXhKIw7nWm4DhXn0jcc7hO5w61fuAyjjz0BjnlzRnc8+JiIiI/C4p0SwiIiLyx/Q3YAzOCteyOCtgzzjp5UruNgKeArrg1JzNwkkST8RJeIJTv3Y8TlKuO07icxfwLjDSq7zD28AdrvjK4TyAcCkwwFo78zTiec8Ysw5nFW1/nGRfGrAe+Ccnf5DfqcY+YYy5GXgYuIf/1dbdg5O4H+/V1xpj7gL+jpNYTcC5xltxkrHuftuMMQ/g1Cgeh/PQvg9c+43GSb52AErjPGTwO+Bla+3eU8T6mTGmDM7q69dwkrPfA88RoH61tfZxY8wy13GfxFmNvROYjLP6/XSuz3fGmA44P1NP4vxfYh1wn7X2I6+uM3AS/7fhrALOxUmsP4bzMwlO0v6sz/88+g2nXnQvnF9ggBP7yzg/xwe9+h7B+aVHK+AmIArnvFYDA4D3XH9dICIiInJJMOfngc5F5/qzte44D5O5BufPGE/g/PnlR8BH1tp8r/6xOP8ILMyX1tpe5yteERERERERERERkUtVcV7RfAfOqo8UIAnnzyIr4fyJ4vtAJ2PMHQEe1LICZ3VNQavPY6wiIiIiIiIiIiIil6zivKK5Lc6f0P1YYOVyZZw/aasO9LTWfuNqj8VZ0TzeWht/oeMVERERERERERERuVQFnbrLxWGtnWGt/cE7yexq34tTCw2cGoAiIiIiIiIiIiIichEV59IZJ5Pjes0NsK2qMeavQHmch8vMt9auvGCRiYiIiIiIiIiIiFxifneJZmNMCM7TzgGmBujSwfXlvc9MoJ+1dsepxh83bpwF6N+/vylapCIiIiIiIiIiIiKXhmJbOuMkRgB/An6y1iZ6tWcBLwCNgWjXV2ucBwnGAb8YY0oXNqgx5iFjzGL3+4MHD9ply5bZnJwcu3jxYgvYxYsX2+PHj9sVK1bYAwcO2M2bN9tdu3bZlJQUu2HDBnvo0CG7bNkym5ub67fPqlWrbEZGht2yZYvduXOn3bt3r12/fr09fPiwXbJkic3Ly/PZ59ixY3b16tU2PT3dbt261e7YscOmpqbadevW2SNHjtjFixfb/Px8n32ysrLsmjVr7P79++22bdvs9u3b7b59++zatWvt0aNHPX0XLVrk2efo0aN27dq1dt++fXb79u1227Ztdv/+/XbNmjU2KyvLZ3z38Y4cOWLXrVtnU1NT7Y4dO+zWrVttenq6Xb16tT127JjPPnl5eXbJkiX28OHDdv369Xbv3r12586ddsuWLTYjI8OuWrXKHj9+3Gef3Nxcu2zZMnvo0CG7YcMGm5KSYnft2mU3b95sDxw4YFesWOG3T05Ojl22bJk9ePCg3bhxo92zZ4/dvXu33bRpk83MzLTLly+3OTk5Pud+4sQJz1xu2rTJ7t692+7Zs8du3LjRes+/e59FixbZ48eP25UrV3rmsuD8L1261G8us7Oz7apVq3zmsuD8F5xL9/ynpaV55jI1NdWuXbvWM//W2kLnPzk52SYnJ9v9+/f7zb/365EjR/zmPy0tza5evdpmZWX5XK/8/Hy7ZMmSk85/dna2z89ZXl6eXbp0qT106FDA+V+5cmWh83/w4EG7YcMGu2fPHr/5P3HihN/8L1++3GZmZtqNGzfa3bt32927d9uNGzf6zL/3Pt7z776X9+zZYzds2OCZ/0D3svf8e9/L3vPvfd2ys7NPei8XnP9FixbZrKwsv/l338vu+S84l+57ueD8B7qX3T877vlPTU31m/+C97J7/r3vZe/5X7Vqlc/8u+9/9/x738sF59/7HjvZvRxo/t37uOf/ZPeyPsv1Wa7Pcn2W67Ncn+X6LNdnuT7L9Vmuz3J9luuzXJ/ll+JnORdIsX0YYCDGmEHAKGA90MJam3Ea+4QAc4GmwGBr7aiT9fda0Vz0gEVEREREREREREQurgtSueF3s6LZGPN3nCTzWqDN6SSZAay1ucD7rrd/Pk/hiYiIiIiIiIiIiFyyfheJZmPMYGAMsBonybz3DIfY73ottHSGiIiIiIiIiIiIiJydYp9oNsY8CbwBLMdJMu87i2GauV63nrPARERERERERERERAQo5olmY8xzOA//WwK0s9amnaRvU2NMyQDtbYF/uN5+dl4CFREREREREREREbmEhVzsAApjjOkHDAfygDnAIGP86lYnW2s/dn0/EqhvjJkJ7HK1XQu0dX3/nLX21/MZs4iIiIiIiIiIiMilqNgmmoFartdgYHAhfWYBH7u+/xToDjQBOgElgFTgv8AYa+2c8xapiIiIiIiIiIiIyCWs2CaarbVDgaFn0P8D4IPzFY+IiIiIiIiIiIiIBFasazSLiIiIiIiIiIiISPGnRLOIiIiIiIiIiIiIFIkSzSIiIiIiIiIiIiJSJEo0i4iIiIiIiIiIiEiRKNEsIiIiIiIiIiIiIkWiRLOIiIiIiIiIiIiIFIkSzSIiIiIiIiIiIiJSJEo0i4iIiIiIiIiIiEiRKNEsIiIiIiIiIiIiIkWiRLMUK0OHDsUYw8yZMy92KCIiIiIiIiIiInKalGgWERERERERERERkSJRolmKlYcffph169Zxww03XOxQRERERERERERE5DSFXOwARLxVqFCBChUqXOwwRERERERERERE5AxoRbMUWXJyMsYY4uPj2bJlCz179qR8+fKULVuWjh07snr1agD279/PQw89RJUqVQgLC6NJkyYkJSX5jFVYjeY5c+Zw2223cdlllxEaGkrlypVp1qwZw4YN8+mXmprK448/Tt26dSldujRRUVHUrVuX+Ph4tm7d6hd7YmIinTt3pkKFCoSGhlK7dm2GDBlCZmamX9/Y2FhiY2PJyspiyJAh1KhRg9DQUK644gpGjhyJtdZvn++//5527dpRpUoVQkNDqVq1Kq1bt2bs2LF+fTMyMnj66ae56qqrKFWqFJGRkbRr145p06adcg5EREREREREREQuJq1olnMmOTmZpk2bctVVVxEfH09ycjLfffcdcXFxzJ8/n5tvvpmIiAjuuusuMjIy+M9//kOnTp3YuHEjNWrUKHTcqVOncssttxAREUGXLl2oVq0aGRkZrFu3jrFjx5KQkABAVlYWLVq0YMuWLXTo0IHbbrsNay3bt29n0qRJ9OzZk8svv9wz7vDhw0lISKBcuXLceuutxMTEsHLlSl599VV++ukn5s+fT0REhE8sOTk5dOzYkT179tCpUydCQkKYOHEiTz31FNnZ2Z5YAP7v//6Pv/71r1SuXJnbbruNChUqsG/fPlauXMlHH33EgAEDPH23b99OXFwcycnJtGrViptvvpmjR48yefJkbr75Zt59910efPDBczVVIiIiIiIiIiIi55QSzXLOzJo1ixdffJF//vOfnrYXXniB559/nqZNm3LnnXcyduxYgoKchfQdOnSgb9++vPHGG7zxxhuFjvvee++Rn5/PzJkzadCggc+2tLQ0z/e//PILW7ZsYfDgwX7jnThxguPHj3veJyUlkZCQwI033shPP/1EVFSUZ9vHH3/MvffeS0JCgt84e/bsoUGDBkyfPp1SpUoBkJCQQJ06dXjjjTd45plnKFGiBADvvvsuJUuWZMWKFcTExBQaN0C/fv3Yvn07X3zxBb169fK0Z2ZmEhcXx6BBg+jSpQuVKlUq9DqJiIiIiIiIiIhcLEo0nye9F+Rc7BBO2+fNSpyTcWJjY3nqqad82vr168fzzz/P8ePHeeWVVzxJZoDevXtz3333sXz58tMa353Y9RaonnOgfiVLlqRkyZKe96NHjwacJLZ3khkgPj6eUaNGMWHChIAJ8NGjR/scIyYmhq5du/LJJ5+wYcMG/vSnP3m2hYSEeBLPhcW9YsUKZs2aRc+ePX2SzABRUVEMGzaMbt268c033/isghYRERERERERESkulGiWc6Zhw4YEBwf7tFWtWhWAOnXqULZsWZ9twcHBVKpUiV27dp103D59+vDtt9/StGlT7rrrLtq0aUOLFi247LLLfPq1bt2aatWqMWLECJYuXUrnzp1p0aJFwLjmz59PiRIl+Oqrr/jqq6/8jnnixAn2799Peno65cuX97RHRkZyxRVX+PWvXr06AAcOHPCJ+7HHHqN+/frcddddtG7dmhYtWlCxYkW/WAAOHjzI0KFD/cbev38/AOvWrTvZZRIREREREREREblolGiWcyYyMtKvLSQkpNBt7u05OSdf/d2jRw8mT57Ma6+9xocffsi7774LQOPGjXn55Zfp0KEDABERESxYsICEhAS+//57EhMTAWf18IABA3j22Wc9q4vT09PJzc31e5hgQUeOHPFJNBdc/VzwPPPy8jxtjz76KBUqVGDs2LGMHj2aN998E2MMrVu35pVXXuH666/3xAIwffp0pk+fftJYREREREREREREiiMlms+Tc1WOQhy33HILt9xyC0ePHmXhwoVMnjyZcePGceutt7Js2TKuvvpqAC677DI++OADrLWsXbuWGTNm8PbbbzN8+HDy8/N54YUXACfxnZ+fT0ZGxnmNu2/fvvTt25fMzEx+/fVXvvvuOz788ENuuukm1q1bR0xMjCcJP2rUKAYNGnRe4xERERERERERETkfgk7dRaT4KF26NG3btuX111/nmWee4cSJE0yZMsWvnzGG+vXrM3DgQM8q4YkTJ3q2N2vWjAMHDrBmzZoLEndUVBSdO3fmvffeIz4+noyMDObMmeOJBfC8FxERERERERER+b1RolmKvV9++YVjx475taempgIQHh4OwOrVq0lOTj5lP4B//OMfADz44IPs2bPHb5+jR4+yYMGCIsU9depUcnNz/dr37dvnE8/1119Pq1at+Pbbb/nwww8DjrVq1SrPfiIiIiIiIiIiIsWNSmdIsffYY4+RnJxMXFwcsbGxlCxZkiVLljBjxgxq1qxJr169APj555959NFHad68OfXq1SMmJoZdu3YxadIkgoKCGDJkiGfMdu3aMWLECJ5++mmuvPJKOnfuTK1atThy5Ajbt29n1qxZtGzZkqlTp5513L169SIsLIyWLVsSGxuLtZY5c+awaNEiGjduTPv27T19P//8c9q2bcv999/P6NGjadq0KVFRUezatYuVK1eyevVq5s+fT0xMzNlfSBERERERERERkfNEiWYp9p555hm+++47Fi9ezM8//0xQUBA1atTgmWeeYfDgwURHRwNw0003MXjwYGbPns2kSZM4dOgQVapUoUOHDp4EtLcnn3ySFi1aMHr0aObOncukSZOIjIykWrVqPPTQQ/Tu3btIcY8YMYLExESWLl3KTz/9RFhYGDVr1mTkyJH079/f82BCcGpLL1myhLfeeotvvvmGCRMmkJeXR+XKlbn66qsZOHAg11xzTZHiEREREREREREROV+MtfZix1CsjBs3zgL079//YociIiIiIiIiIiIiUlTmQhxENZpFREREREREREREpEiUaBYRERERERERERGRIlGiWURERERERERERESKRIlmERERERERERERESkSJZpFREREREREREREpEiUaBYRERERERERERGRIlGiWURERERERERERESKRIlmERERERERERERESkSJZpFREREREREREREpEiUaBYRERERERERERGRIlGiWURERERERERERESKRIlmERERERERERERESkSJZpFREREREREREREpEiUaBYRERERERERERGRIlGiWURERERERERERESKRIlmERERERERERERESkSJZpFREREREREREREpEiUaJYiS05OxhhDfHw8ycnJ9OrViwoVKhAWFsb111/P5MmT/fY5fvw4I0aM4NprryU8PJyIiAhatWrFf//733My/okTJxg9ejSNGjUiOjqa8PBwYmNj6dq1Kz///LNf//Xr1xMfH0/16tUJDQ2lUqVK9O7dmw0bNvj1jY+PxxhDcnIy7777Ltdccw1hYWFUqlSJhx56iIMHD/rts3LlSu6++25iY2MJDQ2lYsWKNGrUiMGDB5OTk+PTNzc3l7Fjx9KsWTMiIiIIDw/nuuuuY8yYMeTn5590LkRERERERERERC4GY6292DEUK+PGjbMA/fv3v9ih/G4kJydTq1Yt4uLiWLNmDZdffjk33ngjGRkZfPnll+Tk5PDzzz/Tpk0bwEkCd+zYkVmzZlGvXj1uvfVWsrKy+Prrr9m3bx9PP/00//rXv856fIDevXvzxRdf8Kc//Ym2bdtSqlQp9uzZw9y5c+nRowevvvqqp+/UqVPp0aMHOTk53HbbbVxxxRXs2rWLb7/9ltDQUJKSkmjUqJGnf3x8POPHj+eOO+4gMTGR2267jUqVKpGUlMSyZcto06YNM2bM8PRfuXIlTZs2xRhDly5dqFWrFocOHWLz5s0kJSWRkZFBmTJlADwxJCYmUrduXeLi4ggLCyMpKYmVK1dyzz338Omnn563uRQRERERERERkT8cc0GOYq3Vl9fX2LFj7dixY62cvm3btlnAAnbo0KE+26ZOnWoB26lTJ0/bv/71L09bTk6Opz01NdXWrFnTAnbevHlnPX5mZqY1xtjGjRvb3Nxcv3jT0tI832dkZNioqChbvnx5u2bNGp9+q1evtqVLl7bXXXedT3u/fv0sYKtXr263b9/uac/JybGtWrWygF24cKGn/dFHH7WAnThxol8sGRkZNi8vz/M+ISHBAvbhhx/2iT03N9fed999hY4jIiIiIiIiIiJSiAuSVw25INnsS5DpNvpih3Da7MRB52ScmjVr8uyzz/q03XTTTdSoUYPffvvN0/bhhx9ijOH1118nJOR/P4IxMTE899xzPPDAA7z//vs0b978rMY3xmCtJTQ0lKAg/+ow5cuX93z/ySefkJmZyZgxY7j66qt9+tWvX58HH3yQN998k7VxlD30AAAgAElEQVRr1/ptf/7556lRo4bnfUhICPfeey9z5szht99+44YbbvDpX6pUKb9YoqOjPd/n5+czZswYKleuzBtvvEFwcLBnW3BwMK+99hofffQREyZMoGvXrn5jiYiIiIiIiIiIXCxKNMs507BhQ5/kqFv16tWZP38+AIcPH2bz5s1Uq1aNevXq+fVt27YtAMuWLTur8QEiIiK47bbb+OGHH2jYsCG33347rVq1omnTpoSHh/vs695vxYoVDB061G/sjRs3ArBu3Tq/RPP1118fMBaAAwcOeNruuusuRo0aRbdu3ejZsyft27enRYsW1K5d2+9Y6enpXHnllbz44ot+Y4OTrF63bl3AbSIiIiIiIiIiIheLEs1yzkRFRQVsDwkJ8TzEzv2gvCpVqgTs627PzMw8q/HdvvzyS0aOHMnnn39OQkICAGFhYfTs2ZNXX32VSpUqAZCeng7Ae++9d9JzO3LkyGnF416hnZeX52m74YYbmDNnDi+99BJff/21p8Zy3bp1SUhI4O677/aJZdOmTQwbNuyMYhEREREREREREbmYlGg+T85VOYo/msjISAD27t0bcHtKSopPv7NVqlQphg4dytChQ9m5cyezZ8/m448/5rPPPiM5OZk5c+b4HGfFihVce+21RTrmydx4441MnjyZ48ePs2TJEqZOncpbb71F7969qVixIu3bt/fE0r17d7799tvzFouIiIiIiIiIiMi55l/AVuQ8Klu2LLVr12b37t1s2rTJb3tSUhIAjRo1OmfHrF69On369CExMZErr7ySuXPnelYPN2vWDMCTeD7fQkNDad68OcOHD2f0aKeO96RJkwCoV68eUVFRLFiwgJycnAsSj4iIiIiIiIiIyLmgRLNccPfddx/WWoYMGeJTYiItLY0XXnjB0+ds7d+/n4ULF/q1Hz16lMOHDxMSEkLJkiUBuPfee4mKimLYsGE+DxR0y8/PZ+bMmWcdCzhJbHfJEG+pqakAnrrRISEhDBw4kJSUFAYNGsSxY8f89klJSWHt2rVFikdERERERERERORcU+kMueAef/xxpkyZwqRJk2jQoAGdO3cmKyuLr776in379vHEE0/QsmXLsx5/9+7dNGvWjKuuuopGjRpRvXp1Dh06xOTJk9m7dy+DBg2ibNmyAJQvX56vv/6a7t2706xZM9q1a0f9+vUJCgpix44dzJ8/n/T0dLKzs886ntdee41p06YRFxfH5ZdfTpkyZVizZg1TpkwhOjqahx56yNP3ueeeY8WKFbzzzjv88MMPtG3blmrVqrFv3z42bdrEvHnzeOmll/weTCgiIiIiIiIiInIxKdEsF1zJkiWZPn06r7/+Op9//jlvvfUWISEhNGjQgDfffNPzcLyzFRsby7Bhw5g5cyZJSUmkpaVRrlw56taty4gRI+jVq5dP/3bt2rFy5UpeffVVEhMTmTNnDiVLlqRq1aq0bduW22+/vUjxDBgwgOjoaBYuXMi8efPIzc3lsssuY8CAATz22GPUrFnT07dEiRJMnDiRzz77jI8//pjJkydz5MgRKlasSK1atXjhhRfo06dPkeIRERERERERERE514y19mLHUKyMGzfOAvTv3/9ihyIiIiIiIiIiIiJSVOZCHEQ1mkVERERERERERESkSJRoFhEREREREREREZEiUaJZRERERERERERERIpEiWYRERERERERERERKRIlmkVERERERERERESkSJRoFhEREREREREREZEiUaJZRERERERERERERIqk2CaajTHljTEPGGO+M8ZsNsYcM8YcNMbMNcbcb4wJGLsxprkx5idjTIYxJssYs9IYM9gYE3yhz0FERERERERERETkUhBysQM4iTuAcUAKkATsACoBPYD3gU7GmDustda9gzGmK/ANkA18CWQAtwFvAC1cY4qIiIiIiIiIiIjIOVScE80bgS7Aj9bafHejMeYZ4Dfgdpyk8zeu9gjgPSAPiLPWLna1PwfMAHoaY3pZa/9zQc9CRERERERERERE5A+u2JbOsNbOsNb+4J1kdrXvBd5xvY3z2tQTqAj8x51kdvXPBp51ve1//iIWERERERERERERuTQV20TzKeS4XnO92tq6XqcG6D8byAKaG2NCz2dgIiIiIiIiIiIiIpea312i2RgTAvR1vfVOKtd1vW4suI+1NhfYhlMq5PLzGqCIiIiIiIiIiIjIJSZ46NChFzuGMzJs2LB/A92An6y1w73ahwDlgVeGDh2aEWC/PkB1YMLQoUN3FdxujHlo2LBh795yyy1VAerUqcO6deuoUKECy5Yto2rVqixevJjy5cuzZs0aQkND2b17N1lZWRw9epQ9e/ZQsmRJ1q5dS8WKFVm6dKnPPuvWraNEiRLs2bOHo0ePcuzYMXbt2kVoaCirV6+mUqVKPvtER0ezYcMGgoOD2bt3L4cPHyY7O5udO3dSqlQpVq5cSeXKlVmyZIlnn6ioKDZu3Igxhn379nHo0CFOnDjB9u3bCQ8PZ+XKlVStWpVFixZRrVo1Fi9eTGRkJJs3b8ZaS1paGgcOHCA3N5fk5GTKlCnDihUrPOO7jxcREcGWLVvIz88nIyOD9PR08vPz2bp1K2XLlvXZp1KlSixbtoyyZcuybds2cnNzOXDgAGlpaQBs3ryZqKgoli9f7tknJiaGFStWULp0aZKTk8nJyeHgwYPs27cPYwwbN24kOjraZ5+KFSuycuVKwsPD2bFjB8ePH+fQoUOkpqYSHBzM+vXrKV++PEuXLvWce4UKFVi9ejWhoaHs2rWL7Oxsjhw5QkpKCiEhIZ75d++zaNEiKlSowNq1az1zWXD+16xZQ0xMjM9clitXjnXr1hESEuKZy4Lz7762zzzzDN27d6dXr17s3buXoKAgUlNTOXToEMePH2fHjh2e+a9SpUqh879//34yMzPJycnxm//Fixd7XiMiIvzmPy8vj61bt1KmTBmWL1/uuV6VK1dm6dKlJ53/iIgIz7wsWrSIypUrs3z5ckqXLh1w/jdt2lTo/IeHh7N9+3ZOnDjhN//lypXj/fffp0mTJuzZs4ebb76ZVatWERYWxo4dO8jOzubw4cPs3bvXZ/6972Xv+Xffy0eOHGHPnj2UKFGCdevWBbyXveff+172nv8lS5Z4rlu5cuVYv349wcHBtG3blgcffJD+/ft77uVVq1b53MuLFi3y3P/e8+++l93zX3Au3fcy4DP/ge5l98+Oe/7z8/NJT0/3mf+C97J7/r3vZe/537JlC5GRkT5zWalSJc/8e9/L+/fv95n/ZcuWee6xmJiYQu/loKAgz/y759K9j3v+d+7cWei9rM9yfZZfqM/ygvOvz/LCP8u978uKFSv+Lj7LA93L+izXZ7k+y/VZrs9yfZbrs1yf5fos12d5cf4sr1KlyrAzy8CeHWOtvRDHOSeMMYOAUcB6oIW1NsNr20bgSuBKa+3mAPv+CtwI3GitXVDYMcaNG2cB+vdXOefTkZycTK1atejXrx8ff/zxxQ7HR05ODpMnT2by5MksXLiQHTt2kJubS+3atenevTtDhgyhbNmyAffdtWsXzz//PFOnTiU9PZ0qVarQrVs3EhISiI6OPm8xx8fHM378eLZt20ZsbOx5O84fwcyZM2nTpg0JCQn8Xn5hFhcXx6xZs7jQn7vF+T4VERERERERkfPOXIiDhFyIg5wLxpi/4ySZ1wLtvJPMLgddr5GFDBFRoJ/8wW3ZsoUePXpQunRp2rRpwy233MKRI0dITEzkhRde4Msvv2TevHlUqFDBb7/mzZuzb98+unbtSr169fjtt98YNWoUU6dOZd68eZQvX/4inZX8nn3yySdkZWVd7DBERERERERERM6530Wi2RgzGHgDWI2TZN4XoNsG4HqgDrCkwP4hQC2chwduPb/RSnFRtmxZ3n77bfr160fp0qU97SdOnKBHjx78+OOPDBs2jLfeestnvwEDBrBv3z5Gjx7NwIEDPe2PPvoob7zxBv/85z955513Lth5yB9HjRo1LnYIIiIiIiIiIiLnRbF/GKAx5kmcJPNyoE0hSWaAGa7XmwNs+zMQDvxqrT1+7qO8NA0dOpRatWoBMH78eIwxni/vP8/Pz8/nnXfeoUmTJpQpU4bSpUvTpEkTxo0bR35+vt+4xhji4uLYs2cPf/nLX4iJiaFUqVI0btyYzz///LTjq1atGgMGDPBJMgOULFmSZ555BnDKL3jbunUr06ZNIzY2lr///e8+24YNG0bp0qX59NNPOXr06GnHsWTJEm6//XZiYmIIDQ2lZs2aDBgwgJSUlEL3yc/P5/XXX6devXqEhYVx2WWX8Y9//INDhw759V25ciV33303sbGxhIaGUrFiRRo1asTgwYPJycnx6Zubm8vYsWNp1qwZERERhIeHc9111zFmzBi/uUhOTsYYQ3x8PBs3buSuu+4iJiaGoKAgZs6cSb169ShZsqSnFlBBI0aMwBjD22+/7dO+a9cuHn74YS6//HJCQ0MpX748Xbp0YdGiRQHHSU1N5f7776dSpUqUKlWKhg0bMn78+EKvXSAvv/wyxhhGjx4dcPuePXsIDg6mSZMmPm3Dhw+nRYsWVK5cmZIlS1K1alV69+7NunXr/MY41fUCp3SGMb5/rXLixAnGjBlD586dqVmzJqGhoZQrV4727dszZcqUgPHGxsYSGxtLVlYWQ4YMoUaNGoSGhnLFFVcwcuRIn9Icp3ufioiIiIiIiIgURbFe0WyMeQ4YjrNCuWOAchnevgZGAr2MMW9Zaxe7xggDXnT1GXc+473UxMXFkZmZyahRo2jQoAHdunXzbGvYsKHn+7/85S98/vnnVK9enQceeABjDN999x0DBgxg7ty5TJgwwW/sAwcO0Lx5c6Kiorj33nvJzMzkv//9L3369GH37t0MGTKkSLGXKFECgJAQ31tgxgzn9xUdO3YkKMj39zBly5alRYsWTJs2jQULFtCuXbtTHmfy5MncfvvtWGvp2bMnNWvWZMmSJYwbN45JkyYxb968gLWY//GPfzB79mzuvPNOunbtSmJiIm+++SZz5sxh7ty5hIWFAU6SuWnTphhj6NKlC7Vq1eLQoUNs3ryZsWPH8uKLL3rONScnh9tuu43ExETq1q1L7969CQsLIykpiYEDB7Jw4UI+/fRTv1i2bNlC06ZNqVOnDn369OHYsWNERETQr18/nnnmGb744gufld9un3zyCSVLlqRXr16etqVLl9KxY0cyMjK46aab6NGjB2lpaUycOJGWLVvy3Xff0blzZ0//9PR0mjdvztatW2nZsiUtW7YkJSWFv/3tb3Ts2PGU19+tb9++PPvss4wfP55Bgwb5bf/ss8/Iz8+nX79+nrbZs2czYsQI2rRpw+23306ZMmXYtGkTX3/9Nd9//z3z5s2jQYMGp329CpORkcEjjzxC8+bN6dChAxUrViQlJYUffviBzp0789577/HAAw/47ZeTk0PHjh3Zs2cPnTp1IiQkhIkTJ/LUU0+RnZ1NQkICcPr3qYiIiIiIiIhIkVhri+UX0A+wOOUu3gCGBviKL7BPN1f/I8D7wL9xHhxoga9wPfzwZF9jx461Y8eOtXJ6tm3bZgHbr1+/gNs///xzC9jrrrvOHj582NN+5MgR27hxYwvYCRMm+Ozjmi97xx132Ly8PE/71q1bbXR0tC1RooTdsmVLkeL+29/+ZgH71FNP+bQ//vjjFrCvvvpqwP3+/ve/W+C0fkYOHz5sy5cvb4OCguzs2bN9to0YMcICtkOHDj7t/fr1s4AtX768TU5O9rTn5eXZHj16WMAOHz7c0/7oo49awE6cONHv+BkZGT7XLyEhwQL24Ycftrm5uZ723Nxce9999/mN455bwD799NN+4+/atcsGBQXZxo0b+2377bffLGB79OjhacvJybG1a9e2oaGhdubMmT79d+/ebatWrWorV65ss7OzPe0PPvigBezgwYN9+i9atMiGhIRYwCYkJPgdP5COHTtawK5atcpv29VXX21LlChh09LSPG2pqan20KFDfn2XL19uS5cubW+++Waf9lNdL2utbd26tXU+dv8nOzvb7ty5069vZmamrV+/vo2OjrZZWVk+22rWrGkB26lTJ59tqampNjIy0kZGRtoTJ074xVbYfSoiIiIiIiIif2gXJJ9bnFc013K9BgODC+kzC/jY/cZaO9EY0xr4J3A7EAZsBh4FRlvr9ffk51uDOy7YoYpsxVfnbegPP/wQcMoolClTxtNeunRpRo4cSfv27Xn//ffp3bu3z37BwcGMHDnSZ1VxrVq1GDRoEMOGDePTTz/1rNg8U99//z3vvvsul112GU888YTPtoMHnWdFRkYGfqakuz0zM/OUx5k0aRLp6encfffdtGrVymfbY489xjvvvMP06dPZsWOHX+3eRx55hJo1a3reBwUF8corrzBx4kQ+/PBDnnvuOZ/+pUqV8jt+dHS05/v8/HzGjBlD5cqVeeONNwgODvZsCw4O5rXXXuOjjz5iwoQJdO3a1WecSpUqBbzW1apVo127dkyfPp01a9ZQv359zzZ3aQvvFcI//vgjW7Zs4fHHH6d169Y+Y1WtWpUnnniCwYMH88svv9C5c2dycnKYMGECZcuWZejQoT79r7/+evr06XNGJTT69evHtGnTGD9+PK+88oqnffHixaxdu5bu3bv7POQxJiYm4DgNGjSgbdu2TJs2jZycHM+KcbfCrldhQkNDueyyy/zaIyMjue+++3jsscdYtGgRf/7zn/36jB492mfuY2Ji6Nq1K5988gkbNmzgT3/602nHISIiIiIiIiJSFMU20WytHYqzavlM95sHdD5lR7kgli5dSlBQEHFxcX7bWrduTXBwMMuWLfPbVqNGDU9dWW9xcXEMGzYs4D6n49dff6V3796ULl2ab775xicZezrcv6soWGc3kKVLlwLQtm1bv20hISH8+c9/Jjk5mWXLlvklmgsmYgEuv/xyqlevTnJyMpmZmURFRXHXXXcxatQounXrRs+ePWnfvj0tWrSgdu3aPvtu3LiR9PR0rrzySl588UW/scFJVgeqPdygQQNCQ0MD7hMfH8/06dMZP348//73vwGn5vB//vMfKlas6FMGY/78+QBs377dL3EMsGnTJgDWrVtH586dWb9+PVlZWbRq1Spg4j8uLu6MEs3du3cnMjKSzz77jBEjRniS7e4x4uPj/fb58ccfeeedd1i8eDFpaWnk5ub6bE9LS6NKlSo+bSe7XoVZs2YNr7zyCrNnzyYlJYXs7Gyf7bt37/bbJzIykiuuuMKvvXr16oBTfkZERERERERE5EIptolm+WM4ePAg5cqVo2TJkn7bQkJCqFChAvv2+T/fsVKlSgHHq1y5smfcMzV//nw6depEUFAQU6ZM4YYbbvDr405oFja++2F8ha149uYeo2Ai0s3dHmh19MnOf/v27Rw8eJCoqChuuOEG5syZw0svvcTXX3/tqbFct25dEhISuPvuuwGn1jE4ydxhw4YVGvORI0cCHrMw3bt3JyIigs8++4yXX36Z4OBgJk+eTHp6OoMHD/apge2O4auvTr6C3h2D+/qd6mfhdJUqVYo777yT9957j2nTptGpUydycnL44osvqFixIp06dfLpP3r0aB555BGio6Pp0KEDNWrUIDw8HGMMEydOZMWKFRw/7v9s0TONa8GCBbRt25bc3FzatWtHly5diIiIICgoiOXLlzNp0qSAx4mKigo4nvua5+XlnVEcIiIiIiIiIiJFoUTz+XIey1H8nkRGRpKRkRGwxEBubi5paWkBH5SWmpoacLy9e/d6xj0Tc+bM4ZZbbiEoKIjExESaNWsWsF/dunUBZwVwIO5Vt3Xq1DnlMd0xumMuKCUlxaeft9TUVE8s3gKd/4033sjkyZM5fvw4S5YsYerUqbz11lv07t2bihUr0r59e0//7t278+23354ydm8nW73tTt6+//77TJ8+nZtvvjlg2QzvmCdNmkSXLl1OeVx3/1P9LJyJfv368d577zF+/Hg6derkSYo/8sgjPj+fubm5JCQkULlyZZYuXer3ywL36uxATme1u7cXX3yRY8eOkZSU5Lfy/+WXX2bSpElnNJ6IiIiIiIiIyMUQdOouIoVzlx8obPXkddddR35+PrNnz/bbNnv2bPLy8mjUqJHfth07dpCcnOzXPnPmTM+4p2vGjBl06tSJkJAQpk+fXmiSGaBNmzYATJs2jfz8fJ9thw8fZt68eZQqVeqkY7i5Y3TH7C03N5e5c+cCBDz/WbNm+bVt3bqVnTt3EhsbG3A1a2hoKM2bN2f48OGMHj0awJOkrFevHlFRUSxYsICcnJxTxn4m3CUnxo8fT1paGlOmTOHaa6+lYcOGPv3c12zOnDmnNW69evUIDw9n+fLlAVeYB7qup9KiRQuuvPJKJk2axMGDBwtNiqelpZGZmUnz5s39ksxHjhzxlEU5FzZv3ky5cuUClpcJ9HNwNk51n4qIiIiIiIiIFJUSzVIk0dHRGGPYsWNHwO333XcfAE8//TRZWVme9qysLJ566ikA7r//fr/98vLyePLJJ32Svdu2bWP06NGEhIRwzz33nFZ806ZN49ZbbyUsLIxffvmFJk2anLR/7dq16dixI8nJybz99ts+2xISEjh69Ch9+/aldOnSpzx2t27dKFeuHF988QULFizw2fbmm2+ydetW2rdv71efGWDUqFFs377d8z4/P58hQ4aQn5/Pvffe62mfM2dOwCSsexVweHg44JRTGDhwICkpKQwaNIhjx4757ZOSksLatWtPeV4FeSdvx40bR05OTsB6x127dqV27dq8/fbb/PTTTwHHmj9/vufnpESJEvTp04fDhw/71XRevHgxEyZMOONYwUkqZ2dnM3bsWH766SeuvfZav19cxMTEEB4ezpIlS3zKieTk5PDII4+QlpZ2VscOJDY2loyMDFauXOnT/sEHH5CYmHhOjnGq+1REREREREREpKhUOkOKpEyZMjRt2pQ5c+bQp08f6tSpQ3BwMF26dOHaa6+ld+/eTJo0if/+97/Ur1+fbt26eWrcbtu2jTvvvJM+ffr4jXvttdeycOFCGjduTMeOHTl48CBffvklmZmZ/Pvf//Z72F0gGzZsoGvXrmRnZ9O5c2cmTZoUsAxBwSTm2LFjad68OYMGDeKXX37hqquuYuHChSQlJVGnTh1eeuml0742H374IXfccQetW7fmjjvuoEaNGixZsoRp06ZRuXJl3n333YD7tmjRgoYNG3LXXXcRGRlJYmIiK1asoHHjxjzxxBOefq+99hrTpk0jLi6Oyy+/nDJlyrBmzRqmTJlCdHQ0Dz30kKfvc889x4oVK3jnnXf44YcfaNu2LdWqVWPfvn1s2rSJefPm8dJLL3H11Vef1vl569u3L8899xwvvPACISEh9O7d269PiRIl+Pbbb7npppu45ZZbaN68OQ0bNiQ8PJydO3eyaNEitm7dSkpKiidB/q9//YtffvmFN998k8WLF9OyZUtSUlL48ssv6dy5M99///1Zxfr888+TkJBATk6O32pmgKCgIAYNGsSIESO45ppr6Nq1KydOnCApKYmMjAzatGlDUlLSGR87kMGDB5OYmEjLli258847iYyMZPHixcydO5eePXvy9ddfF/kYp7pPRURERERERESKzFqrL6+vsWPH2rFjx1o5fZs2bbK33nqrLVeunDXGWMB+9NFHnu15eXn27bffto0bN7alSpWypUqVso0aNbJjxoyxeXl5fuMBtnXr1nb37t22T58+tmLFijY0NNRed911dsKECacdV1JSkgVO+RXIjh07bHx8vK1cubItUaKErVGjhh00aJBNT08/4+vz22+/2W7dutkKFSrYEiVK2OrVq9u//e1vdvfu3X59+/XrZwG7ZcsW++qrr9q6deva0NBQW7VqVfvII4/YgwcP+vRPTEy08fHx9qqrrrIRERE2PDzc1qlTxw4cONAmJyf7jZ+fn28/+eQT27ZtWxsdHW1LlChhq1atalu0aGFfeuklu2PHDk/fbdu2WcD269fvlOe4fft2GxQUZAF76623nrRvamqqffLJJ239+vVtqVKlbOnSpe0VV1xhb7/9dvvpp5/anJwcn/4pKSn23nvvtRUqVLBhYWG2QYMG9qOPPvLMb0JCwinjK6hdu3YWsCEhIXbv3r0B++Tk5NjXXnvNXnXVVTYsLMxWqlTJ3nPPPTY5OdkzT9u2bfP0P53r1bp164A/cz/88INt2rSpLVOmjI2MjLQdOnSws2bNsh999JHf/WSttTVr1rQ1a9YMeIyEhAQL2KSkJJ/2U92nIiIiIiIiIvKHdUHyqsZaeyHz2sXeuHHjLED//v0vdiiXLGMMrVu3PqsavCIiIiIiIiIiIuLDXIiDqEaziIiIiIiIiIiIiBSJEs0iIiIiIiIiIiIiUiRKNIuIiIiIiIiIiIhIkYRc7ABEClLdcBERERERERERkd8XrWgWEREREfl/9u48zs7rrg//58zIkj3yaN8sK07i1FvtZoGUJKRNYgfSUBMSuwkNEAghyw+XlpQWB1JIKyVAoG6BhMXNxhJIE1KwHfr7QRsKTgwJaSRLsix55EWyJI9lS7Ysa18szfn9MYtH0mhGd+6dVe/363Vfd+6z3Plqjp7nPvdzzz0HAABoiqAZAAAAAICmCJoBAAAAAGiKoBkAAAAAgKYImgEAAAAAaIqgGQAAAACApgiaAQAAAABoiqAZAAAAAICmCJoBAAAAAGiKoBkAAAAAgKYImgEAAAAAaIqgGQAAAACApgiaAQAAAABoyoyJLgAAAOB8c9vmE8Ouv/Vqb9UAgKnF1QsAAMAE2nu8Dvw8f2aZwEoAAEZP0AwAADDOBvdYHty7WU9mAGCqGtVVTClldpJLkyxKciTJ7lrr460sDAAAAACAqeGcg+ZSykuS/ESS703yistPw7oAACAASURBVJw2kWApZU+Su5P8WZI7aq3DDzoGAAAAAMC0MGLQXEp5ZZJfSvI96Q2Xn0tyf5InkzyT5KIkC5NcleQdSd6e5KlSyieS/Hqt9djYlA4AAAAAwGQwbNBcSvnDJD+SZF+STyf5UpJv11qPnmX7FyX5Z0neneSXk/xkKeVHa633tLBmAAAAAAAmkbYR1r8pyb9LsrzW+q9qrfecLWROklrrtlrrp2qt353k5UnWJnl968oFAAAAAGCyGWnojMtrrUdG88S11g1JbiqlXDSa/QEAAAAAmBqG7dE82pC51c8BAAAAAMDkNdLQGeeklHJBKeUVpZSrWvF8AAAAAABMHQ0FzaWUHyylfLmUsmDQspck2ZRkTZIHSil3lFJGGpIDAAAAAIBpotEezT+R5Opa6zODlv3XJP8gyd1JNiR5a5L3tKY8AAAAAAAmu0Z7Hv/DJH/V/6CUMifJP0/y5VrrO0spFyRZn96g+TMtqxIAAICzum3ziWHX33q1L50CAGOr0R7Ni5M8Mejxa9IbVn8pSWqtz6U3iH5JS6oDAACgIXuP14EbAMB4afRj7QNJ5g56/PokNcnfDVp2NElnk3UBAABwjgb3WB7cu1lPZgBgvDR61fFwku8rpcxKb8D8jiQbaq1PD9rmhUl2t6g+AAAAAAAmuUaHzvh0ksvTGzh39f38e6dt86okm5ovDQAAAACAqaChoLnW+odJfjVJR3qH0PjtvluSpJRyQ5IXJbm7dSUCAAAAADCZNTxgV631PyT5D2dZ/XdJ5ic51ExRAAAAAABMHS2dGaLWejzJ8VY+JwAAwHTVfbhm97Ganpq0ld7HKzrKRJcFANCwYYPmUspFtdYjzfyCVjwHAADAdLJxX0/u6O7J5gP1lOUf2nAiV3eW3LyiLdfNbXRKHQCAiTPSlcujpZQPllJmNfrEpZSXlVK+kuRnR1caAADA9HP37p58vOvkGSFzv80Haj7edTJf290zzpUBAIzeSEHzV5P8epInSim3l1KuL6VcdLaNSymXl1JuKaX8fZK1SV4WEwMCAAAk6e3J/NmtJzN0xPy8muQzW09m4z5hMwAwNQw7dEat9cdKKZ9M8itJPtB3O1lK6UryRJK9SS5MsjDJVUkWJSlJdiX5hSS/UWs9NnblAwAATB13dPeMGDL3q0nu7O4xhAYAMCWMOBlgrXVNkjeVUq5I8t4kb0zy8iT/6LRNn0pyR5I/S/JntdbnWlwrAADAlNV9uJ51uIyz6TpQTRAIAEwJIwbN/WqtDyf5+SQppXQkuTS9PZmPJNlda31iTCoEAACYBkY7DMbGfT1Z0dHe4moAAFrrnIPmwWqth5M83HcDAABgBEdOju9+AADjyWBfAAAA4+CiUXZKHu1+AADjSdAMAAAwDkY7qZ/JAAGAqcAVCwAAwDhY0VFydWdjk/pd01lMBAgATAmCZgAAgHFy84q2nGtsXJLctMJbNgBganDVAgAAME6um9uW913ePmLYXJK8//J2w2YAAFPGjIkuAAAA4Hxy/ZK2LJ6V3Nndk64D9Yz113SW3LSiTcgMAEwpDQXNpZTLkjxba90/zDadSebXWnc0WxwAAMB0dN3c3iC5+3DNJx4+kZ6atJXkg1fMMCYzADAlNfoR+aNJPjjCNj/dtx0AAADDWNFRsmRWybILe++FzADAVNXo0Bml7wYATBO3bT4x7PpbrzbSFgAAAMMbi3eOS5McGoPnBQDG2N7jz48VOn+mz5YBppruwzW7j9WBoTi6D1e9pAGAcTFi0FxK+bHTFr18iGVJ0p7ksiQ/muT+FtSWUsrbk7w+ycuTvCxJZ5Iv1FrfNcS2L8rwQ3b8Sa31na2oCwCmk8E9lgf3btaTGWDq2LivJ3d092TzaZMLfmjDiVzdWXKzyQUBgDF2Lu8g/yBJ/9VKTfLWvtvp+j8mP5xkVdOV9frF9AbMB5N0J7n6HPa5L8ldQyzf2KKaAAAAJo27d/fks1tPpp5l/eYDNR/vOpn3X568YYmwGQAYG+cSNL+n774k+b30hrhfGWK7k0n2JPn7WuuzrSkvP5PegPmR9PZsvvsc9llfa13Zot8PAAAwaW3cN3zI3K8m+czWk1k0K3o2AwBjYsSgudb6h/0/l1LeneSuWuvnx7Sq53/3QLBcinHFAAAABruju2fEkLlfTXJnd4+gGQAYEw0NvlhrvX6sCmmh5aWU/yfJwjzfw3rDBNcEAADQUt2H6xljMo+k60A1QSAAMCYaCppLKd+R5PuTfKrWumuI9cuSfCDJn9da17emxIZ9b99tQCnla0neXWvdMSEVAQAAtNjGfT2j3m9FR3uLqwEAznftK1euPOeNV61a9V+TvC3JLw+136pVqw4l+XyS5StXrryzNSUOPPeLkvx4kvtXrlx5xxDrZyY5muTfJfn5JL+W5C+TvCjJ9UnesmrVqt9buXLlc0M9fynlA6tWrfrUjTfeuDxJrrzyynR1dWXRokVZt25dli9fnjVr1mThwoXZtGlTZs2alccffzyHDx/OoUOHsnPnzsycOTMPPPBAFi9enLVr156yT1dXVy644ILs3Lkzhw4dypEjR9Ld3Z1Zs2Zl48aNWbp06Sn7zJ8/Pw8++GDa29vz5JNP5sCBAzl69Ggee+yxXHTRRdmwYUOWLVuWe++9d2CfefPm5aGHHkopJbt3787+/ftz/PjxbN++PR0dHdmwYUOWL1+e1atX59JLL82aNWsyd+7cPPLII6m15umnn87evXtz4sSJbNu2LRdffHHuu+++gefv/31z5szJli1b0tPTk2eeeSZ79uxJT09Ptm7dms7OzlP2Wbp0adatW5fOzs48+uijOXHiRPbu3Zunn346SfLII49k3rx5Wb9+/cA+S5YsyX333ZfZs2dn27Ztee6557Jv377s3r07pZQ89NBDmT9//in7LF68OBs2bEhHR0d27NiRY8eOZf/+/dm1a1fa29uzefPmLFy4MGvXrh34ty9atCgbN27MrFmz0t3dnaNHj+bgwYN54oknMmPGjIH2799n9erVWbRoUR544IGBtjy9/Tdt2pQlS5ac0pYLFixIV1dXZsyYMdCWp7f/6W3Z3/5tbW3ZtWtX9u/fn2PHjmXHjh0D7X/JJZectf2feuqpPPvss3nuuefOaP81a9YM3M+ZM+eM9j958mS2bt2aiy++OOvXrx/4ey1btixr164dtv3nzJkz0C6rV6/OsmXLsn79+syePXvI9n/44YfP2v4dHR3Zvn17jh8/fkb7L1iw4JTjcvHixbn//vtz4YUXZseOHTl69GgOHDiQJ5988pT2H7zP4PbvP5YPHjyYnTt35oILLkhXV9eQx/Lg9h98LA9u/3vvvXfg77ZgwYJs3rz5rMfy/ffff0r7r169esj27z+W+9v/9LbsP5aTnNL+Qx3L/f93+tu/p6cne/bsOaX9Tz+W+9t/8LE8uP23bNmSuXPnntKWS5cuHWj/wcfyU089dUr7r1u3buAYW7JkyVmP5ba2tjPav3+f/vZ/7LHHznosO5ef27l840Nb0nPhnFywY0NefcWlzuXO5c7lzuXO5eNwLl+9ek1qx5zM2L01L5mdcz6X727rzDPdW9NWezL76DOZfWRvSj2ZRc9uy9FZnVmxa0P2dV6Sy55Ym/0XL80Ldt2XYzNn55ID27PkgpPO5c7lzuXO5c7lk/i6XMbiXN7Kc/kll1yy6lwz2GaUWs/9q1allEeT3FNrffcw2/x+ktfVWl/SgvoGP+8b0jsZ4Bdqre9qYL8ZSf4uyauS/Nta6yeG2/7222+vSXLLLbeMvlgAmKJu23xi4Odbr27oi08AjNJoz73/64mT+fz2xns1/9gL2/LmS/RoBoDzyLiMmdXoLBDLknSPsM3OJJeMrpzWq7WeSPLZvoevm8haAAAAWmW0k/qZDBAAGAuNXmEcTrJ4hG0WJzk2unLGzFN997MntAoAAIAWWdFRcnVnYx2UruksJgIEAMZEo0Hz+iRvLaVcPNTKUsqcJG/t224yeXXf/dYJrQIAAKCFbl7Rds7fhS1JblqhNzMAMDYavcr4dHp7LP9VKeWlg1eUUl6W5KtJFvVtN65KKa8qpcwcYvkNSX6m7+Efj29VAAAAY+e6uW153+XtI4bNJcn7L283bAYAMGYamuWn1vonpZTvS/JjSdaVUnYleTzJpUmWpvf65Q9rrV9sRXGllLcleVvfw2V9968ppfxB389P11p/tu/nX0tybSnla3l+HOmXJrmh7+eP1Fq/2Yq6AAAAJovrl7Rl8azkzu6edB04c7L3azpLblrRJmQGAMZUw9PJ11p/vJTyzST/Jsm1eT4A3pjkk7XWz55158a9PMm7T1t2ed8tSbYn6Q+a/yjJTUn+cZLvS3JBkl1Jvpzkt2utf9vCugAAACaN6+b2Bsndh2s+8fCJ9NSkrSQfvGKGMZkBgHHRcNCcJLXWTyf5dCmlI8m8JM/WWg+3tLLe37Myycpz3PZzST7X6hoAAACmihUdJUtmlVMeAwCMh1EFzf36wuWWB8wAAAAAAEwdowqaSymLk/yLJNckmV1rfd+g5S9Ocn+t9UjLqgQAAAAAYNJqOGgupbw3ySeTXJjeyf9qkvf1rV6a5O+TfCCGsQAAAAAAOC80NO1wKeV7k3w6yUPpnXjv9sHra60bk2xK8rZWFQgAAAAAwOTWaI/mn0vyRJLX11r3l1JeMcQ2G5K8punKAAAAAACYEhoNml+Z5Eu11v3DbNOdZNnoSwIAJkL34Zrdx2p6atJWeh+v6CgTXRYAAABTQKNB88wkh0bYZl6Sk6MrBwAYbxv39eSO7p5sPlBPWf6hDSdydWfJzSvact3chkbbAhgTt20+Mez6W68e1VznAAC0QKNXYtuSfOcI27wqyYOjqgYAWkQYcW7u3t2Tz249mXqW9ZsP1Hy862Tef3nyhiXCZmDy2Hv8+TPX/JlT79sXg1+nthysZyz3OgUATDWNXr18JcmHSinvqLX+j9NXllLek+SlSX6hFcUBQCtM9TBirGzcN3zI3K8m+czWk1k0K3o2AxNqcPg6OKid6qHsgpkTXQEAQPMavSL7z0nemeSLpZS3J5mbJKWUf53knya5OcnDSX6rlUUCQKOmaxjRSnd094wYMverSe7s7hE0A7SI1yMAYLpp6Oqm1rq3lPL6JJ9P8o5Bqz7Zd/+3SX641jrSOM4AwATqPlzPGJN5JF0HqgkCAQAAGFLDH6PXWnckeUMp5aVJXpNkYZJ9Sb5Va723xfUBAGNg476eUe+3oqO9xdUAMFHMaQAAtMqwVw2llDuSfKnW+uW+x69Lsq3WuqPWuiHJhnGoEQBosSMnx3c/ACY/cxoAAM0Y6ePptyVZP+jx3UlWJfnomFUEAIy5i0bZKXm0+wEwOZnTAABolZFm9Hk2yZxBj32sDQDTwGgn9TMZIAAAAEMZ6WPqriQ/VEpZneSJvmUv6htCY1i11nuaLQ4AGBsrOkqu7iwNTQh4TWcxESAAAABDGiloXpXkriT/fdCyd/fdRuLLtQAwid28oi0f7zqZc4maS5KbVujNDAAAwNCGDZprrV8tpVyT5HuSXJpkZZKv990AgCnsurlted/lyWe3Dh82lyTvv7zdsBkAAACc1YgzPNRatyf5XJKUUlYm+Vqt1WSAADANXL+kLYtnJXd296RriGE0ruksuWlFm5AZAACAYQ0bNJdSfj3J/6q1frVv0XuSrBvzqgCAcXPd3N4guftwzScePpGemrSV5INXzDAmMwAAAOdkpO5J/zbJqwc9/r0kbxu7cgCAibKio2TJrJJlF/beC5mByar7cM3uYzVPHu297z587hObAgAwNkYaOuNgko5Bj73jBAAAJsTGfT25o7snm08b6udDG07k6s6Smw31AwAwYUYKmh9JcnMp5c4kT/Qtm1dKuWykJ6617mi2OABoVn+vt/7hILoPVz11Aaagu3f3DDt56eYDNR/vOpn3X568YYmwGQBgvI0UNN+W5I+TfHPQsg/23YZTz+G5AWDM6PUGMH1s3Dd8yNyvJvnM1pNZNCvO8QAA42zYMLjW+sVSyqNJbkxyaZIfT7IhyfqxLw0ARkevN4Dp5Y7unhFD5n41yZ3dPYJmAIBxNmKv41rrt5J8K0lKKT+e5M5a60fHuC4AGBW93gCml+7D9Yxvp4yk60A1VBIAwDhr9J31e5LcNRaFAEArjKbXGwCT18Z9oztPj3a/81X/nAZPHu297z7cWLgPANDQOMq11j8cq0IAoFl6vQFMP0dOju9+5xtzGgAArdLwFUMppa2U8m9KKd8qpewrpZwYtO4VpZTfLaVc2doyAWBker0BTD8XtY/vfueTu3f35ONdJ8/6IW3/nAZf2+11EgAYWUNBcyllZpK/SvKbSV6S5ECSwV3AHk3yE0l+pFUFAsC50usNYPoZbW9avXCH1+icBj6UBQBG0ujV161Jrk+yKsnSJJ8dvLLW+mySe5L8s5ZUBwAN0OsNYPpZ0VFydWdjwxtd01kMiTQCcxoAAK3WaND8I0m+UWv9aK21Jxny2uTRJJc1XRkANEivN4Dp6eYVbTnX2LgkuWmF8/pwmpnTAADgbBq9Antxkm+NsM0zSRaMrhwAGD293gCmp+vmtuV9l7ePGDaXJO+/vN0HiCMwpwEAMBYavQI7kmTeCNtcluTZ0ZUDAM3R6w1gerp+SVs+fE17rjnLB4rXdJZ8+Jr2vGGJ8/pIzGkAAIyFGQ1uvz7Jm0opM2utx09fWUqZm97xmb/ZiuIAoFG9vd4y4gRHer0BTD3XzW3LdXPb0n245hMPn0hPTdpK8sErZvh2SgPMaQAAjIVGg+bPJPlCki+UUt47eEUpZV6S308yP8l/a015ANC465e0ZfGs3omLuoYYg/KazpKbVrQJmQGmqBUdJUtmlVMec+56X/8aHwbD6yYAMJyGguZa6xdLKd+T5D1JfiDJ3iQppaxJcm2SWUl+p9b6F60uFAAaodcbAAytf06DRiYENKcBADCShj+SrrW+N8lPJHkgyeL0fvv4O5I8kuS9tdZ/09IKAaAJ/b3ell3Ye+9NMgCY0wAAaL1RXS3UWv+g1vqKJBcnWZGks9b6j2qtv9/S6gAAAGi53jkN2kcMm81pAACcq0bHaD5FrfVIkiMtqgUAAIBxYk4DAKCVmgqaAQAAmLrMaQAAtIqgGQAAYIq7bfOJgZ+3HKxnLL/16uHf+vXPaTD4MQBAIwTNAAAA08iCmRNdAQBwPhI0AwAATHEj9VgGABhrZnUAAAAAAKApgmYAAAAAAJrSku9XlVIWJfmnSQ4n+T+11pOteF4AAAAAACa/hno0l1JuKaX831LKgkHLvjNJV5I/TfIXSb5ZSpnd2jIBAAAAAJisGh06418mqbXWZwYtuy3J/CS/n96g+R8n+cnWlAcAAAAAwGTXaNB8RZIN/Q/6hsx4fZLP1VrfV2t9S5LVSX64dSUCAAAAADCZNRo0L0yye9Dj1/bd3zlo2d8meWEzRQEAAAAAMHU0GjQ/k2TRoMevT9KT5JuDltUkFzZZFwAAAAAAU0SjQXNXkreUUhaWUuald8zm1bXW/YO2eVGSJ1tUHwAAAAAAk1yjQfMnklySpDvJY0mWJfnd/pWllPYk/yTJfa0qEAAAAACAyW1GIxvXWv+8lPKTST7Qt+gLtdY/HrTJ96R32Iz/3aL6AAAAAACY5BoKmpOk1vrpJJ8+y7r/nWR+s0UBAACc7rbNJwZ+3nKwnrH81qsbfnsDAECLuBIDAACmnAUzJ7oCAAAGayhoLqW8Islr0jtkxr6+ZbPTO07zW5McTvJrtdZPtLpQAADg/KbHMgDA5NXoZIA/l+QX+kPmPh9P8qN9z7Uwya+XUt7UovoAAAAAAJjkGg2aX5nka/0PSikXJHl3km8nWZLkxUmeTvLTLaoPAAAAAIBJrtGgeUmSxwY9fmWSziSfqrUerbXuTPKVJC9tUX0AAAAAAExyjQbNNaeO6/xP+pZ9fdCyp5IsbrKulFLeXkr5rVLK35ZS9pdSainlj0fY57tLKX9RSnmmlHK4lLKhlPJvSyntzdYDAAAAAMDQGp1NY0eSVw96/NYk3bXWrYOWLU+yt9nCkvxikpclOZikO8nVw21cSnlrkj9LcjTJnyR5JslbkvxGktcmeUcLagIAAAAA4DSN9mj+cpLvLqX8aV/v4tck+dPTtrkuyZYW1PYzSa5MMifJLcNtWEqZk+QzSU4meUOt9b211luTvDzJ3yd5eynlnS2oCQAAAACA0zTao/k3krw5yc19j9cn+Wj/ylLKP0zynUl+pdnCaq13D3rekTZ/e3qH6/h8rXXNoOc4Wkr5xSR/nd6w+kvN1gUA081tm08M/LzlYD1j+a1XN3q5AAAAwPmmoXeOtdaDSV5bSrmub9EDtdaeQZscTnJTkjVn7Dy2bui7/19DrLsnvXV9dyllVq312PiVBQBTy4KZE10BcD4Y/AHXUHzABQAw9YzqCq7WuvEsy7cl2dZEPaN1Vd/9Q6evqLWeKKU8muTaJJcn6RrPwgBgshPoABNp7/Hnv0kxf+aI32QEAGCSal+5cmXDO5VSLli1atWbVq1a9ZZVq1Z978qVK+/pW37hqlWrFqxaterIypUr60jPc65WrVr1oiQ/nuT+lStX3jHE+luTLExy28qVK58ZYv2PJHlBki+sXLmy+yz/pg+sWrXqUzfeeOPyJLnyyivT1dWVRYsWZd26dVm+fHnWrFmThQsXZtOmTZk1a1Yef/zxHD58OIcOHcrOnTszc+bMPPDAA1m8eHHWrl17yj5dXV254IILsnPnzhw6dChHjhxJd3d3Zs2alY0bN2bp0qWn7DN//vw8+OCDaW9vz5NPPpkDBw7k6NGjeeyxx3LRRRdlw4YNWbZsWe69996BfebNm5eHHnoopZTs3r07+/fvz/Hjx7N9+/Z0dHRkw4YNWb58eVavXp1LL700a9asydy5c/PII4+k1pqnn346e/fuzYkTJ7Jt27ZcfPHFue+++waev//3zZkzJ1u2bElPT0+eeeaZ7NmzJz09Pdm6dWs6OztP2Wfp0qVZt25dOjs78+ijj+bEiRPZu3dvnn766STJI488knnz5mX9+vUD+yxZsiT33XdfZs+enW3btuW5557Lvn37snv37pRS8tBDD2X+/Pmn7LN48eJs2LAhHR0d2bFjR44dO5b9+/dn165daW9vz+bNm7Nw4cKsXbt24N++aNGibNy4MbNmzUp3d3eOHj2agwcP5oknnsiMGTMG2r9/n9WrV2fRokV54IEHBtry9PbftGlTlixZckpbLliwIF1dXZkxY8ZAW57e/qe3ZX/7t7W1ZdeuXdm/f3+OHTuWHTt2DLT/JZdcctb2f+qpp/Lss8/mueeeO6P916xZM3A/Z86cM9r/5MmT2bp1ay6++OKsX79+4O+1bNmyrF27dtj2nzNnzkC7rF69OsuWLcv69esze/bsIdv/4YcfPmv7d3R0ZPv27Tl+/PgZ7b9gwYJTjsvFixfn/vvvz4UXXpgdO3bk6NGjOXDgQJ588slT2n/wPoPbv/9YPnjwYHbu3JkLLrggXV1dQx7Lg9t/8LE8uP3vvffegb/bggULsnnz5rMey/fff/8p7b969eoh27//WO5v/9Pbsv9YTnJK+w91LPf/3+lv/56enuzZs+eU9j/9WO5v/8HH8uD237JlS+bOnXtKWy5dunSg/Qcfy0899dQp7b9u3bqBY2zJkiVnPZbb2trOaP/+ffrb/7HHHhvyWN7atjAXbF+fnvmXZNa2tc7lzuXO5c7lzuUTfC7//isXpeOx+/KOl67I/WvXZOa8xZnzxKa896rZuezI42c9ll2Xj+25/JGeOblgx4aB10vncudy53Ln8lZflzuXuy53Lp+Yc/kll1yyatjwtUVKrY3lwaWUNyf5XJJlSUqSWmtt71v36iTfSPKuWusXW1ZkKW9IcneSL9Ra3zXE+oeSXJHkilrrI0Os/2Z6Jy58Ta31W8P9rttvv70myS23DDv/IABTyOCvaOu9CzC5OEdPHtoCAKatcfnaWFsjG5dSXpnkriQ1yc8k+e+D1/eFuI+md5zm8bSv737uWdbPOW07AAAAAABapNGPqT+S3on1XllrfbKU8p+G2GZ1ku9ourLGPJjklUmuTHLv4BWllBlJXpzkRJKt41wXABNkcK+sLQfrGcv11AIAAIDWaahHc5LXJrmr1vrkMNs8luSS0Zc0Kn/Td//mIda9LklHkm/WWo+NX0kATBYLZj5/AwAAAFqv0e5cFyd5eoRtOtJ4gN2sP03ya0neWUr5rVrrmqR3csIkv9S3ze3jXBMAE0iPZQAYmW8AAQCt0uhVw+NJrh1hm5enBUNUlFLeluRtfQ+X9d2/ppTyB30/P11r/dkkqbXuL6W8P72B89dKKV9K8kySH0hyVd/yP2m2JgAAgOnKN38AgGY0GjT/ZZKfLKX8k1rr352+spTyfUm+O8mvtqC2lyd592nLLu+7Jcn2JD/bv6LWelcp5fVJfiHJv0hyYZJHkvy7JJ+stdYAAAAwQI9lAKBVGr2q+HiSdyb5ainlt5K8KElKKTemdyzkn0ryRJJfb7awWuvKJCsb3OcbSf55s78bAAAAAIBz11DQXGt9vJTypiRfTnLroFV/nqQk2ZLk5lrrSOM4AwAAAAAwTTT8Pala69pSylVJbkzymiQLk+xL8q0kX6m1nhhufwAAAKaWwZMGDsUQHADAqK4Gaq0n09uL+c9bWw4AAHC+6D5cs/tYTU9N2krv4xUdZaLLYgR7jz8//c38mdoLAOjlY2cAAGBcbdzXkzu6e7L5wKnzdX9ow4lc3Vly84q2XDe3bYKqYyiDeywP7t2sJzMA0G9UVwWllFcm+a4k85O0D7FJrbV+rJnCAACA6efu3T357NaTqWdZv/lAzce7Tub9lydvWCJsBgCYKhoKmkspNBiOJgAAIABJREFUc5LckeT69E7+dzY1iaAZAAAYsHHf8CFzv5rkM1tPZtGs6Nk8yRjuBAA4m0Z7NN+W5IYkf5vk95M8lsTkfwDAhHjLLw0/XcT//MUfGKdKgHNxR3fPiCFzv5rkzu4eQfMkYbgTAGAkjQbNb02yNsn1tdaeMagHAGBUHt9zaODnSxfOnsBKgKF0H65nhJQj6TpQ9ZidBAx3AgCci0aD5rlJ/kjIDABMBoN7LA/u3awnM0w+G/eN7i3Exn09WdEx1LQwjAfDnQAA56rRK4CHkywdi0IAAIDp68jJ8d2P1hjNcCcAwPmp0aD5d5K8pZRy6VgUAwAATE8XjbJT8mj3o3nNDHcCAJx/Gh064y/TOxngN0opq5Lcm+TZoTaste5osjYAAGCa6B1OofHeroZhmDiGOwEAGtFo0Lwtvd+IKkk+O8x2dRTPDQAATFMrOkqu7iwN9ZC9prOYCHACGe4EAGhEo2Hw55NzHqILAABgwM0r2vLxrpEnlkt6e7bctEJv5olkuBMAoBENBc211h8fozoAAEZt0449eXTX/pw42ZMZ7W3ZtGNPrr1s4USXBZzmurlted/lyWe3Dh82lyTvv7zdsBkTzHAnAEAjDG8BwLR02+YTw66/9WovgdPBX9/3WD765f+bezbtPGX5dT/9hbzu2uX5jz/4qrzxZS+YoOqAoVy/pC2LZyV3dveka4hhNK7pLLlpRZuwchIw3AkA0IiGrt5KKSdLKR8ZYZtfKKUM/+4eAMbR3uN14Mb08bm/2pQ3rbzrjJC53z2bduZNK+/K7/2fTeNcGTCS6+a25SPXzsh/fumMXHpRcsmFyaUXJf/5pTPykWtnCJknkZtXtOVcY2PDnQDA+a3R7lyl73Yu2wHAhBncY3lw72Y9maeHv77vsXzgd/8mPXX4Dw96as37f+dv8sLFc/RshkloRUfJklnllMdMLoY7GTu+fQXAdDMWVwHzkxwdg+cFAEiSfPTL/3fEkLlfT6352Je/PcYVAUxf1y9py4evac81nUN/EHBNZ8mHr2nPG5YImUfLt68AmA5G/Ii0lPK60xa9aIhlSdKe5LIkP5LkwRbUBgBwhk079px1uIyz+fqmx00QCNCE6+b2jpvdfbjmIxuf74n7setm6Ik+Sv09lrsP13zi4RPpqUlbSX7osnZ/UwCmpHP5Ls7XkoFvSdUk7+67DaWkd1rif990ZQDQAt2Ha3YfqwNv3roPV2/epri/3vDYqPcTNAMwWWzc15M7unvOmGzxQxtO5OrOkptNignAFHMuQfNH0xswlyT/Mb3B89eH2O5kkj1J7q61bm5VgQAwGt68TV/7Dx8f1/0A8Lraanfv7hl23OvNB2o+3nUy7788Yz4kibGiAWiVEV8xaq0r+38upbw7yV211k+OZVEA0IzJ9OaN1pvTMXNc9wM433ldba2N+4b/e/arST6z9WQWzcq4hfiDx4ieP9M3wABoTEMfTdZaXzxWhQBAK0zmN2+0xhtf+oJx3Q/gfOZ1tfXu6O4Z8e/Zrya5s7tnTP+mg3ssD+7drCczAI1yBQDAtDKaN29MLddetjCvu3Z5Q/u8/tpLjc8MMApeV1ur+3A9Y/iRkXQdqOk+3Ng+ADARhv2IspTyN+mbALDW2t33+FzUWusbm64OABrQzJs3EwROLf/xB1+VN628Kz115PZuKyUf+cHvGoeqAKYXr6utt3Hf6IL4jft6sqKjvcXVnMoEygA0a6Tvwryh777jtMcj8XErAONuMr95o7Xe+LIX5NP/6oZ84Hf/Ztiwua2UfOanbsgbX2bYDIBGeV1tvSMnx3e/c2GiRwBaZdigudbaNtxjAJhMJuObN8bOe7/32rxoyZx87Mvfztc3PX7G+tdfe2k+8oPfJWQGGCWvq6130Sjz99HuNxITPQLQSiMNnfHTSb5Va/32ONUDAKM22d68Mfbe+LIX5I0ve0E27diTf3nbX+bEyZ7MaG/Ln9z6fcZkBmiS19XW6+0Z3HhP8bHoUWyiRwBabaRXid9M8ub+B6WUk6WUj4xtSQAwOqN98+NN09R37WUL8+Klc3LF8nl58dI5QmaAFvC62norOkqu7mxs3ONrOsuYjJVsokcAWm2kK4CjSWYNelz6bgAw6UymN28AMNV5XR0bN69oO+c31SXJTStaH9w3M9EjAJzNSK9Yjyb5Z6WUpYOWeWUBYNKaDG/eAGC68LraetfNbcv7Lm8f8e9akrz/8vYxGzZjPPcD4Pww0ivWp5J8R5KdpZT+KR1W9g2hMdztxNiWDQBDmwxv3gBguvC6OjauX9KWD1/TnmvO0mP8ms6SD1/TPmYT8JnoEYCxMOxkgLXWT5ZSdie5McnyJNcn2ZFk29iXBgCjc/2Stiye1TuWYNcQXwu9prPkphVt3gwDwDnwujo2rpvb+zfrPlzziYdPpKcmbSX54BUzxnz4ERM9AjAWhg2ak6TW+qUkX0qSUkpPkt+vtX50rAsDgGZM5Js3AJhuvK6OnRUdJUtmlVMej7XeDwUaHwbDhwkADGfEoPk0q5J8bQzqAIAxMRFv3gBguvK62jq3bX5+xMktB+sZy2+9utG36+euf6LHRiYENNEjACNp6OPIWuuqWus9Y1UMAAAAnG8WzHz+Nl5M9AhAq43dR6QAAADAkMayx/K56J3oMfns1pMZrl+ziR4BOFeCZgAAADgPmegRgFYSNAMAAMB5ykSPALSKoBkAAADOcyZ6BKBZvv8CAAAAAEBTBM0AAAAAADTF0BkAAMC4um3ziYGftxysZyy/9WpvUwAAphpXcAAAwIRZMHOiKwAAoBUEzQAAwLjSYxkAYPoxRjMAAAAAAE3RlQCAacn4nwDQOl5XAYCRuBoAYNoz/icAtI7XVQBgKIJmAKYlPasAoHW8rgIAIzFGMwAAAAAATRE0AwAAAADQFN9/AgAAgPOUiR4BaBWvGADAlPWWX/rzgZ+//fCuM5b/z1/8gXGvCQCmKhM9AtAMQTMAMC1cuuDiiS4BAKYcPZYBaBWvKADAlKXHMgAAwORgMkAAAAAAAJoiaAYAAAAAoCmCZgAAAAAAmiJoBgAAAACgKYJmAAAAAACaImgGAAAAAKAp0y5oLqVsK6XUs9yenOj6AAAAAACmmxkTXcAY2ZfkN4dYfnC8CwEAAAAAmO6ma9D8bK115UQXAQAAAABwPph2Q2cAAAAAADC+pmuP5lmllHcluSzJoSQbktxTaz05sWUBAAAAAEw/0zVoXpbkj05b9mgp5T211q9PREEAAAAAANNV+8qVKye6hpZatWrV3CQfS/JzSf5Tki8laU/yxiTvXLVq1f+7cuXKXafvV0r5wKpVqz514403Lk+SK6+8Ml1dXVm0aFHWrVuX5cuXZ82aNVm4cGE2bdqUWbNm5fHHH8/hw4dz6NCh7Ny5MzNnzswDDzyQxYsXZ+3atafs09XVlQsuuCA7d+7MoUOHcuTIkXR3d2fWrFnZuHFjli5deso+8+fPz4MPPpj29vY8+eSTOXDgQI4ePZrHHnssF110UTZs2JBly5bl3nvvHdhn3rx5eeihh1JKye7du7N///4cP34827dvT0dHRzZs2JDly5dn9erVufTSS7NmzZrMnTs3jzzySGqtefrpp7N3796cOHEi27Zty8UXX5z77rtv4Pn7f9+cOXOyZcuW9PT05JlnnsmePXvS09OTrVu3prOz85R9li5dmnXr1qWzszOPPvpoTpw4kb179+bpp59OkjzyyCOZN29e1q9fP7DPkiVLct9992X27NnZtm1bnnvuuezbty+7d+9OKSUPPfRQ5s+ff8o+ixcvzoYNG9LR0ZEdO3bk2LFj2b9/f3bt2pX29vZs3rw5CxcuzNq1awf+7YsWLcrGjRsza9asdHd35+jRozl48GCeeOKJzJgxY6D9+/dZvXp1Fi1alAceeGCgLU9v/02bNmXJkiWntOWCBQvS1dWVGTNmDLTl6e1/elv2t39bW1t27dqV/fv359ixY9mxY8dA+19yySVnbf+nnnoqzz77bJ577rkz2n/NmjUD93PmzDmj/U+ePJmtW7fm4osvzvr16wf+XsuWLcvatWuHbf85c+YMtMvq1auzbNmyrF+/PrNnzx6y/R9++OGztn9HR0e2b9+e48ePn9H+CxYsOOW4XLx4ce6///5ceOGF2bFjR44ePZoDBw7kySefPKX9B+8zuP37j+WDBw9m586dueCCC9LV1TXksTy4/Qcfy4Pb/9577x34uy1YsCCbN28+67F8//33n9L+q1evHrL9+4/l/vY/vS37j+Ukp7T/UMdy//+d/vbv6enJnj17Tmn/04/l/vYffCwPbv8tW7Zk7ty5p7Tl0qVLB9p/8LH81FNPndL+69atGzjGlixZctZjua2t7Yz279+nv/0fe+yxsx7LzuXO5c7lzuXO5c7lzuXO5c7lzuXO5c7lzuXO5efjufySSy5Z1XToeg5KrXU8fs+EK6X8lyT/Psldtdabzrbd7bffXpPklltuGa/SAAAAAADGShmPX3I+TQb43/ruXzehVQAAAAAATDPnU9C8u+9+9oRWAQAAAAAwzZxPQfNr+u63TmgVAAAAAADTzLQKmksp15ZSFgyx/IVJfrvv4R+Pb1UAAAAAANPbjIkuoMXekeTnSyl3J3k0yYEkL0lyY5ILk/xFkv8yceUBAAAAAEw/0y1ovjvJVUlekd6hMmYneTbJ3yX5oyR/VGutE1ceAAAAAMD0M62C5lrr15N8faLrAAAAAAA4n0yrMZoBAAAAABh/gmYAAAAAAJoiaAYAAAAAoCmCZgAAAAAAmiJoBgAAAACgKYJmAAAAAACaImgGAAAAAKApgmYAAAAAAJoiaAYAAAAAoCmCZgAAAAAAmiJoBgAAAACgKYJmAAAAAACaImgGAAAAAKApgmYAAAAAAJoiaAYAAAAAoCmCZgAAAAAAmiJoBgAAAACgKYJmAAAAAACaImgGAAAAAKApgmYAAAAAAJoiaAYAAAAAoCmCZgAAAAAAmiJoBgAAAACgKYJmAAAAAACaImgGAAAAAKApgmYAAAAAAJoiaAYAAAAAoCmCZgAAAAAAmiJoBgAAAACgKYJmAAAAAACaImgGAAAAAKApgmYAAAAAAJoiaAYAAAAAoCmCZgAAAAAAmiJoBgAAAACgKYJmAAAAAACaImgGAAAAAKApgmYAAAAAAJoiaAYAAAAAoCmCZgAAAAAAmiJoBgAAAACgKYJmAAAAAACaImgGAAAAAKApgmYAAAAAAJoiaAYAAAAAoCmCZgAAAAAAmiJoBgAAAACgKYJmAAAAAACaImgGAAAAAKApgmYAAAAAAJoiaAYAAAAAoCmCZgAAAAAAmiJoBgAAAACgKYJmAAAAAACaImgGAAAAAKApgmYAAAAAAJoiaAYAAAAAoCmCZgAAAAAAmiJoBgAAAACgKYJmAAAAAACaImgGAAAAAKAp0y5oLqWsKKX8XillZynlWCllWynlN0sp8ye6NgAAAACA6WjGRBfQSqWUlyT5ZpIlSb6SZHOS70rywSRvLqW8tta6ZwJLBAAAAACYdqZbj+bfTW/I/NO11rfVWn++1npDkt9IclWSX57Q6gAAAAAApqFpEzSXUi5P8qYk25L8zmmr/1OSQ0l+tJQye5xLAwAAAACY1qZN0Jzkhr77r9ZaewavqLUeSPKNJB1JXj3ehQEAAAAATGfTKWi+qu/+obOsf7jv/spxqAUAAAAA4LwxnSYDnNt3v+8s6/uXzzuXJ7v99tubLggAAAAAYILVW265pYz1L5lOPZpH0v/HrEOuLOUDpZQ141gPAAAAAMC0MJ16NPf3WJ57lvVzTtvuFLXWTyf5dKuLAgAAAACY7qZTj+YH++7PNgbzFX33ZxvDGQAAAACAUSi1DjmSxJRTSnlJkkeSbEvyklprz6B1nUmeSG+wvrjWemhCigQAAAAAmIamTY/mWuuWJF9N8qIkP3Xa6lVJZif5vJAZAAAAAKC1pk2P5mSgV/M3kyxJ8pUkXUleleT69A6Z8d211j0TVyEAAAAAwPQzrYLmJCmlvCDJR5O8OcnC9A6ZcVeSVbXWZyayNgAAAACA6WjaBc2t0N7eVnt6/F1aqZSSsfy/1lYSTQacL9qS9Iy4FTCSsb4+YXIqpS2DpnPhfFFK4nhnPJW2xLmmeY7d80RJop3H2NO11sVj/UtmjPUvmIp6emrqmomuYnopr6y55557xuz5X/e616X+9pg9PcCkUv51Ul/69okuA6a8suFP85YH3zvRZTDO/udVn8sP/f3xiS6DcfbF18zMd/6Phye6DM4j977jiuQ/CM6a9islP/QvH5joKhhjX/yTf5i89RMTXcb09pUPbh+PXzNtJgMEAAAAAGBiCJoBAAAAAGiKoTOmmGPHk3/1q8n/+XbyzP7kH6xIfuWnku977URX1hof+9jHcu+99+bo0aNZsGBBfviHfzjf//3fP9FlwZAe3p38o19J3v6K5I/fPdHVAJPFG7Z8Ld86/ExmlJIkuXTGRXnw6jdPcFXwvMf/vy156LfX5cgThzJr0UV5+a++LgtfuWyiy2KM/I8b5p/y+OSxI/kHN/9kXvnvf3OCKmKqOLa7Ozs+uzKHHlqXMmNm5r/6zXnBe34hpV2MMOU93ZX8759Knrw36Vic3HBbctVNE13Veeehh7+QRx+9K8/ueygvvOzGvPpVvzKw7sldf5819/5SDh9+IgsXvjSv/q5fzuzZl05gteeprfckO76dHNiZXPqdyXf8SO/ynhPJms8nzz6WHHkmee2/ThZdMbG1ThJeIaaYEyeTFyxNvv7p5LJlyV98I/nBDyf3fyl50fKJrq5573rXu/JzP/dzmTlzZrZv354PfvCDueKKK3LVVVdNdGlwhp/6cvKPXzjRVQCT0W8vf0Xet/DFE10GnOGpbzyerv+yOt/5Gzdk3ksX5+hThye6JMbYO/5m78DPJ44cyp03rshlN/yLCayIqWLHZ1fm/2/vvuOrKvL/j78mvZBQQkLAoEGqAgEFGwhSBAHB7iqurG1VFkFc18Li2vtXwcZPFF27rrq4gtIUQaoKSu9FQAEDhJZCSLt3fn+cm5CEJKSfhLyfj8d53Nw5M3M/N/lkcjLn3DmBkY1ImPQDniMpbH7yZpK++ZiYQbrColbz5sDky+Gs4TB0Nvw+H/47BG5dAVFt3I6uTgkNjaH9mXeSuGcxHk9mXnlm5iEWLR7Nuec8wSnNerN6zass/uEf9O/3qYvR1lEh9aFtf9i3ETzZBfdFnQ4tL4Kf33MltJpKS2fUMuGh8NidzqSynx8M7gEtmsGyDW5HVjlatGhBUFAQ4NwJHmD37t1uhiRSpE9/gQah0FfHYiIiUotsem05bUacRcPOMRg/Q2iTcEKbhLsdllST3+d+QXDDGKI7X+h2KFILZO3bScNug/ALCiawYTT1O/fg6E7dULHW27/RuTrz3L+Dnz/E94G47rD2Q7cjq3Oax/UjLu5igoMbFCjfuWs29SNbcWrzAfj7B9Oxw10cTt5ESso2lyKtw5p1gqYJEFToWMkvAFr2gqiW4Ju7Eocmmmu5vQdg8+/QvqXbkVSe8ePH069fP2688UaioqI4//zz3Q5JpICUo/DIdBh3lduRiEhN9c89a2i87iu6b/2eeWn73A5HBADr8XJ47X4yD2Uwp9/nzO75H9Y88QOejBy3Q5Nqsn3mR7QY+Oe8CzpEShIz6CYOLp6ON/MoWQf2kLxyAZGde7odllSYLaLIQtLa6g9FipScvJUGDY59qjsgIIx64c1JTt7qYlQipaOJ5losOwf+/DDcdCm0i3c7mspz7733MmvWLCZMmEDPnj3zrnAWqSkeng63dYPmDU9cV0TqnuebdmRbu4HsPuNS7mjUgiE7fuDXzDS3wxIhc/9RbLaXxFk76P7xYHpOuZLk9QfYPHGl26FJNTiy53eSViygxaBhbocitUS9M88lY+cWVvzlLNYM70H46R1ocG4/t8OSiopqB+Ex8NMLzlIA2751ls/I1lJKNUVOTjqBgREFygIDI8jOOeJSRCKlp4nmWsrrhWEPQ1AATHjQ7Wgqn7+/PwkJCSQlJTFlyhS3wxHJs3IXfLcR/t7b7UhEpKY6LyyKCP9Agv38ualRPN3DopiRusftsETwC3Fuz9Ji2JmExIQR3CiE02/pwL75O12OTKrD9pkf0TihO/Waaf14OTHr9bLlqVtpcF5/zvpoFZ3eWUrOkRR2f/R/bocmFeUfCNdMga3T4dVYWDIOzvgTRMa5HZn4BASEkZ1d8CKF7Jw0AgO01JXUfLoZYC1kLdz2JOw9CDNegcCT+Kfo8Xi0RrPUKPO2wI6DcOrDzvO0TPBYWJ8Iy8e4G5uI1EzGgC3qY6oi1SyofjAhseGgVRPqpB0zP+KMYQ+4HYbUEp60w2QfSCRmwDD8AoPxCwymce+r2f3pS8QNOwmvdKprYhJg2Pxjz9/vBh11k8eaon79VmzfPjXveU5OOmlpO6lfv5WLUYmUjq5oroX+9ixs2A5fvwShIW5HU3kOHTrEnDlzSE9Px+PxsHTpUubMmcPZZ5/tdmgiee7oDr8+Biv/6WzDL4RL28M3d7kdmYjUBIc9WXyTuocMr4cc6+XjQ7+zIG0/l9SLdTs0EQCaX9Wa7R+uJ/PAUbKSM9n+/jqa9DrV7bCkiiWt/pH0pD84tc/VbocitURAZCOCYuJI+vYTrCeHnCMpHJj3JWGntXM7NKkM+1ZDToazXMZPL0JaIiTc7HZUdY7Xm4PHk4nXerDW43ztzSHulItJTtnCzp3f4vFksnbdRBrUb0Nk5Oluh1z3eD3OEjPW62yebKcMwJPjPC9QTxeXnMTXwp6cfkuEN/8HwUEQe8mx8jfHwp8HuhdXZTDGMGXKFMaNG4fX66VJkyaMGjWKHj16uB2aSJ6wIGfLVS8YQgIhOqL4NiJSd2Rby7/2rGNjZir+GNqFRDAlvhttQzRISM3QZsRZZB3KYO4lk/EP9qfZwBa0/lsnt8OSKrZ9xoc073UFgeEai6T0Wt73/9j53tPsmToJ/PyJaH8ecTc/5HZYUhnWfAir3nYmxpr3gKGzISDY7ajqnHXr32Dtutfznu/47Ws6tB9Bxw4jubDby/yy/Gl+XPIgUY0S6NZtnIuR1mGbv4VNs4493/ULtB0A7QbCnKfh6EGn/MeJzmO/RyAsqvrjrEGM1Wz7cYwx1v7idhQnF9MVFixYUGX99+zZEzuhyroXEalRzEiwCde4HYZIrWdWT2bIptvcDkOq2ddt/83QH7PcDkOq2X8uCKLLf7e4HYbUIcuubQ1jNd9SYc8Yhl633u0opIr957Mz4fJX3A7j5DZ19DJrbdeqfhktnSEiIiIiIiIiIiIiFaKJZhERERERERERERGpEE00i4iIiIiIiIiIiEiFaKJZRERERERERERERCpEE80iIiIiIiIiIiIiUiHGWt0FtTBjTBLwm9txiIiIiIiIiIiIiFTQadba6Kp+EU00i4iIiIiIiIiIiEiFaOkMEREREREREREREakQTTSLiIiIiIiIiIiISIVoollEREREREREREREKkQTzSIiIiIiIiIiIiJSIZpoFhEREREREREREZEKCXA7gJooICDAuh2D1E7WC35+xu0waiwPFj+j70+NYy1+flV33jEnx6PfCxERERGpcaryGFjAenMw+v+vRrPWYKrw98DmKAdqimyvd7+1NrqqX0cTzcXo2PEht0OQWmjD6nGM77HJ7TBqrFE/tabL1M/cDkMKWX/1DSQlJVVZ/+Hh9enYcUyV9S8iIiIiUlZbt46v0mNggegGYaS+3MTtMKQEEfenkJS0v8r6jw4LI7WJcqAmMImJv1XH6+j0nYiIiIiIiIiIiIhUiCaaRURERERK6fHHr2DEiD7F7l+8eCynnNKwXH2/9dbNXHnl2QAMHNiR118fVq5+yupE76mqTJhwI0OGdKr21xURERGRqqGlM+qYxYvH5n0dEhJIVpYHr9cLwFNPfc3MmWsAGDKkM088cSUPPPA5s2evy2vTpUs8kybdREZGNgCpqRmsWrWT999fzPr1f+TVGzGiD716taNFi8a8/fYC3nxzXjW8u5PL7TOa5n0dEGzwZIP1OsuHzxt/mC3fHQWg7SVh9B3TkG8eP8iv847mtWnWKYjLxzcmJ9Npk5lm2bMuk5WfprFvk/PzC23gx4Uj69OsUzABIYaDO7JZ/Hoy+zZkV9fbrDG+7NGXRkHBeK0lw+Phh/17eXHjWsaddS4d6jfEYy1ZXi8rDh3ghQ1rOJCVmde2RXg9RrQ+g7MbRmGMYWPKYSZu2cia5EMuviMRkZPb9On30KhRPTweL16vl23bkpg2bRVffLEMa52/fY8/fgWXXXYWN944iXXrdgPQvHkjvvpqNGed9WiB/h5//AoGDUpg0KCXSEpKLXdc3bs/U/43lc/MmWvyjstOViNHfuR2CCIiIsdr+ydIuBMi4yE7FbbPhCXPQlYKdP0HnHU3eHz/D6bvhV0LYPkrkL7P1bClnK68Eu64A1q1grQ0WLcOAgPhbOfkP4GBYAxkZTnPlyyBMWNg6VI4csQpS0+HlSvh3/+GBQvceR81hCaa65j8//xMn34PTzzxFUuWbDuu3pAhnTl8OJ0hQzoXmGgGSEpKZcCA8QDExERy9dVdeOedW7n77o9ZunQ7ADt3HuCVV77lmmvOqcJ3c3J7a1Bi3tc3/qcJ8144zK7lmcfVa3dJGBnJXtpdElZgohngyAEvH/xpDwDhjf04c0g4V7wazfR/HmD38kwCQw37NjmTy0cPezljUBiXPhvFh9fvJSej7t0T874VS/n54H6ig0N4pct53HJ6awBe3LiWr3b/Tr2AAJ5J6Mrdbc/k0TUrADglNIxJ53bni52/8eTaleRYL4NPOZVXu5zPqGU/sVaTzSIiVeaeez5hyZJt1KsXTJcu8dx7BsiZAAAan0lEQVR//0A6dIjjscem5NU5fDidu+7qw4gRHxbbT0hIIH37nklaWiYDBybwwQeLqyN8ERERqWkS7oTOI+D70bB7EYTHQo9nYfCnMOVyp86vX8HcUeAXAPVPh673wdWz4IsBmmyube64A0aOhAcfhHnzIDsbeveG88+H665z6vzjHxAfD6NGHWsXF+c8tmsHHg9ER8NllzkTzQ89BJ9/Xt3vpMbQ0hlynKZN69Oly2k89dRXXHBBSxo1Ci+27r59KUyc+D1ffrmc0aP75ZV//fUqFi/eypEjx0+MSuWp18SfZp2CmDf+EM3PCSa0YfG/0kf2e/n53VQ2TD/CBXdEApCS6GHVf9NIP+jFemH9tHT8AwwNT63b56CSMjP4cf8+WtaLLFCelpPD/KRE2kTUzyu7vWVb1h4+xBtbN5KSk026x8Pnv29nZuIuRrY5o7pDFxGpk9LSMpk/fxMPPvhfhgzpRMuWMXn7pk1bSevWTejS5bRi21988ZmkpmYwadL8Ui3l0LBhGBMn/oVFi8by9tu30LTpsb8LK1Y8TvPmjQDnKukxYy7l1Vf/zKJFY/ngg9uJizu2rMZ5553O//43kgUL/smDDw4qcFf2IUM68847txbo95prujJ16t3Mnz+GMWMuzdvn52e4995LmDv3AaZNu4frrjuXFSsex9+/6OOCtm1j+eSTO1m0aCzPPXctwcEF/+5feWUXpk69m3nzHuTll4cSHR1RII5rrz2HqVPvZtGisYwY0Ye4uIa8//5fWbjwnzz//LUEBPgDEBERwiuv3MDcuQ8wf/4YXnnlBmJijv1tzb9USO77/fvf+zN//himTbuH7t1bnfBnISIiUmkC68E598Hif8HOeeDNgdRd8O2dEBEHra8uWN+bA4c2w3fD4egBSBjuSthSThERcP/9MHYszJwJR49CTg7Mng1PPlm2vpKSnEnmceOcieZ8x3R1jSaa5TiDB3dm/fo/mDNnA9u372fQoIQTtpk7dwPt2jUlJCSwGiKUXG37h7FvUzbbFmRw6Lcc2lwcdsI22xYeJbp1IAEhxw98US0D8Qs0JO/OqYpwa42Y4BC6NW7C5pTkAuWRgYH0imnKrvQjeWXnRjVmzt7Ewl0wZ88fJDRoRLCff5XHKyIijnXrdrN3bwpnn31sUvno0WzeeWchd93Vt9h2gwd3ZtasNXzzzRri4xvTrl3TYusCDByYwFtvzad37+fZtGkPTz99dbF1BwzowJtvzuOii55j586DjBzpxNGgQRgvvngdr78+lz59nmfXrkN06tS8xNft0aMNf/7zJK67biL9+7enWzdnIvaqq7rQvXsrrr/+DYYOfYPevdsV20dAgD8vvTSU6dNX06vXc3z33Tr69Dkzb/8557Rg1Ki+PPjgf+nX70USE5N57rlrCvTRvXsrbrjhTf7yl7e46abu/OtflzF27BcMHDieVq1iGDiwA+BMgH/11QoGDXqJgQPHk5mZw5gxg4qNrUOHOHbsOEDv3s/z/vuLefTRy0v8foiIiFSq2K7gHwzbZhQsz0mH3+dCXM+i21kv7PgWmp5b9TFK5enSBYKDnUnmyjJjhnN1c8uWlddnLaOJZjnO4MGd8tYEnDVrDUOGdD5hm6SkVPz8/IiICKnq8CSftv3D2DInHYAtc9Jpe8mJJ5qP7Pdi/AzB9QpONAeGGS4e25Bf3k8h60jdWzYD4P86n8N3vQcw6dzuLD90gPe2bwHgH+06MKf3AGb3HkCDwCBe3Lg2r039wCD2Z2Yc19f+zEz8jSEyUCdfRESqU1JSKpGRoQXKJk/+hdjY+kVeIRsbW59zzoln5sw1HDx4hKVLt5/w2GfRos0sX/4b2dkeJkyYQ0JCc5o0iSyy7ty5G1i3bjcej5eZM1fTpk0sABde2Jpt25L47rv15OR4+fjjHzlwIK3E13333UWkpWWwZ08yP/+8nbZtnb769WvPJ58sYd++FFJTM3j33UXF9pGQEEdAgB8ff/wjOTlevvtuPevX787bP2hQAlOnrmDjxkSysz28+up3dOzYnKZNGxSI48iRTLZtS2Lr1n389NOv7N59iLS0TBYv3krbts5EfXLyUebM2UBGRjbp6Vm8/fYCunSJLza2xMTDfPnlMrxey9dfryQ6OpKoqHolfk9EREQqTUgjyDgI1nP8vvR9ENqo+LbpeyCkQfH7peZp2BAOHnSWvqgse/ce67uO0kSzFNCpU3OaNWvAN984E2kzZ66mVauYvH+KihMTE4HX6yU19fgJN6kasR2CiGzqz5a5zrrMm+ccJapFAFEtS57YDI/2x3otmWnHJpP9g2DQM1HsXZ/F8k9K/if3ZPbAyp+5+PtZXLFwDi9sWEOm70aZ4zaupe/3s7jhh3lEBAYSE3zshEpydhaNg48/wdI4OBiPtaRk170bK4qIuCkmJpKUlIL3LMjO9vDWWwsYMaLPcZ9kvPTSTmzfvp/Nm517GsycuZqBAzsSEFD8YfKePSl5Xx89mkVKytECy0vkt39/Wr662YSFBQEQHR3B3r0pBeoWfl5Y/onojIxsQkPz93XsUzh79iQf1zZXdHQE+/YVvNlhYmJygf35nx89mkVycjoxMcfe38GDxz7Zk5mZfVxcue8xJCSQhx4awowZf2fhwn/y73/fQmRkKH5+RX+ctHA/QF5fIiIiVS7joDPZbIr4VGpYDBw9WHzb8FjIOFx1sUnlO3QIGjUC/0r8FHJs7LG+6yhNNEsBQ4Z0xhjDp58OZ/bs+/jgg9t95SWvV9i79xls3JiY90+BVL3cq5evezuGm7+I5ZrXowuUF+f0C0NI2pKdd7M/v0AY+FQUR/Z7mDdefxhL8mtaKu9u28L9Z3TMK1t6YD99mxz/Eeu+sc1Ye/gQmd5KPDsqIiIlOvPMZsTERLBixW/H7fvqqxXUqxdC794F188fPLgTp5zSkNmz72P27Pu4995LaNgwnO7dWxf7OrGxx65eDg0NIjIylKSk1GLrF2X//tTjroIu7qroE/eVVmDt49jY+iXUTS0waVy4flJSaoE1p0NCAqlfP+y4yenSGDasG/HxUQwb9hY9ejzLbbe9C1BgLWoREZEaY+8y8GTB6YWWeQoIhVP7ODcHLJKB0/pD4tIqD1Eq0bJlkJkJAwZUXp8DBzrrNf/6a+X1WctoolnyBAUF0L9/e5566muuv/6NvO3552cycGBCkTeUiY6O4I47enHllWfz2mtz8soDAvwICgrAz8/g73/sa6kc/oHQqlco88cf5rO/7svbFr6WTJu+oZgifrPDG/vR9S8RnHFpOEvedq6Y8vOHAY81IifTMueZQ1A3V8wok+l/7KRhUBA9o5sA8O9tm+nYoBHDW7UjMiCQMH9/rm0ez6BmcUzYst7laEVE6obw8GB69GjDc89dy4wZq9m69fg7vns8Xt5443tuvvnCvLKEhDji4hoybNikvOOea699nRkzVpe4fEb37q3p3PlUAgL8GTGiD2vX7jrh1ciFLVy4hZYtY+jT5wz8/f0YOvT8ci8TMXv2Om644XyioyOoVy+kwHssbNWqXeTkeBk69Hz8/f3o0+cM2rc/JW//zJmrufzys2jTJpbAQH9GjbqYtWt3kZhY9pPRYWFBZGbmkJqaQWRkKHfe2as8b09ERKR6ZKXCsvHQ/Slo3gv8ApybAPafBGmJsHlywfp+AdCgFVw8EcKiYfWbroQt5ZSaCi+8AM8840w2h4ZCQAD07g3/+lfZ+mrcGG65Be69F559FmzdnVwJOHEVqSt6925HRkYO06atJCfHm1c+Zcpyhg/vRbdurUhPzyI6OoLFi8dijCEtLYNVq3Zy++3vsWbNrrw2Dz98GZdddlbe89tvv4hHHvmSr79eWZ1v6aTV4sJQcjItm75JJ/8FsxumH+GcmyM49dwQso96CY/y4/YZztW2WUcse9ZlMfWeJPZucK48j+0QRHy3ULIzvPx12rGrcqc9eIDENVnV+p5qixxr+fz37dx6ehsWJO1lZ/oR7vh5MXe1PoMpPftiMGxIOczoZUtYfbjuflxGRKQ6vPzyDXg8Xrxey/btSXz00Y9MnvxzsfVnzVrLrbf2oEED59M/Q4Z0Zv78TcdNTH/yyU+8886tREaGHrcMh9PPGu68sxcJCXFs3JjI2LFflDn2w4fTeeCBz3nggYE89tgVTJ++ilWrdpa5H4D//W8Zp54axeefj+DIkUz+858ldOkSj8fjPa5uTo6H++77jIcfvoy77urDokVbmDt3Q97+pUu38/rrc3nxxeuIjAxh1aqdjBkz+bh+SuOTT37imWeu5vvvHyApKZUPP/yRPn3OOHFDERERt6x8HTIOwQWPQORpkJUGO2bBdyPB6/sfueVl0GIAYJy1mXctgC8GQPpeV0OXcpg0ybkCefRomDAB0tJg9Wp49dXStd+4EYyB9HRYtQruuAPmzavSkGs6Y+vwLHtxAgICbMeOD7kdhtRCG1aPY3zPTW6HUWON+qk1Xad+5nYYUsj6q28gKSmpyvoPD69PQsKYKutfREQK6t69FQ89NIRBg15yOxQRkRpr69bxVXoMLBDdIIzUl0u+35O4K+L+FJKS9ldZ/9FhYaTGKgdqApOYuMxa27WqX0dLZ4iIiIiI1GLBwQFceGFr/P398pY1y3+VsoiIiIhIddDSGSIiIiIitZgxhuHDe/Pcc9eSmZnNwoVbmDjxe7fDEhEREZE6RhPNIiIiIiK1WEZGNjfeOMntMERERESkjtPSGSIiIiIiIiIiIiJSIZpoFhEREREREREREZEKMdZat2OocYwxqcAmt+MQOYHGQNXdHlak8ihXpTZQnkptoDyV2kK5KrWB8lRqA+WpVJa21tqIqn4RrdFctE3W2q5uByFSEmPML8pTqQ2Uq1IbKE+lNlCeSm2hXJXaQHkqtYHyVCqLMeaX6ngdLZ0hIiIiIiIiIiIiIhWiiWYRERERERERERERqRBNNBdtktsBiJSC8lRqC+Wq1AbKU6kNlKdSWyhXpTZQnkptoDyVylItuaSbAYqIiIiIiIiIiIhIheiKZhERERERERERERGpEE00i4iIiIiIiIiIiEiFaKLZxxgTZ4x5xxjzhzEm0xizwxjzsjGmoduxSd3iyz1bzLanmDbdjDEzjDEHjTHpxpjVxph7jDH+1R2/nFyMMdcYY14zxiw0xqT48vCjE7Qpcz4aYwYbY+YZY5KNMWnGmCXGmJsq/x3JyagseWqMiS9hjLXGmE9LeJ2bjDFLfTma7MvZwVX3zuRkYoyJMsb81RjzpTFmqzHmqC+PFhljbjPGFHlcrjFVqlNZ81RjqrjFGPO8MWaOMWanL08PGmNWGGMeNcZEFdNG46lUu7LkqsZUKU5ZxzxjTD1jzJPGmA3GmAxjzGFf+0HF9N/TGPOhMWatMeaAr812Y8xXxpi+ZYpVazSDMaYl8AMQA0wFNgLnAr2BTUB3a+0B9yKUusQYswNoALxcxO40a+2LhepfDnwBZACfAQeBIUBbYLK19toqDVhOasaYlUAnIA3YBbQDPrbW3lhM/TLnozFmJPAacMDXJgu4BogDxllr76vktyUnmbLkqTEmHtgOrAKmFNHdWmvt5CLavQj8w9f/ZCAIuB5oBIyy1k6ojPciJy9jzHBgIpAIfA/8DjQBrgLq44yd19p8B+caU6W6lTVPNaaKW4wxWcByYD2wDwgHzge6An8A51trd+arr/FUXFGWXNWYKsUpYx41ABYCHYB1wBxf/cuAaGC0tfbVQv0/BtwOLMHJoyPAqb429YCnrLUPlypYa22d34BvAIvzC5i/fLyv/A23Y9RWdzZgB7CjlHUjfYNMJtA1X3kIzskTC1zv9nvSVns3nBNurQED9PLl1EfF1C1zPgLxOAf8B4D4fOUNga2+Nhe4/X3QVrO3MuZpvG//e2Xov5uvzVagYaG+DvhyON7t74O2mr0BfXAmNfwKlcfiTOZZ4Op85RpTtVX7Vo481ZiqzZUNCCmm/Glffr2er0zjqTbXtjLmqsZUbZWRRy/7yr4AAvKVR+OcyMgCWpey/1OAvYAHaFqaWOv80hnGmNOB/jiTe/+v0O5HcWbxhxljwqs5NJHSuAZnsPjUWvtLbqG1NgP4l+/p39wITE4O1trvrbVbrO+vzAmUJx9vBYKBCdbaHfnaHAKe8T0dXs7wpY4oY56WR24OPu3LzdzX3YFz7BAM3FJFry0nCWvtXGvt19Zab6HyPcAbvqe98u3SmCrVrhx5Wh4aU6XCfGNhUT73PbbOV6bxVFxTxlwtD42pdUAZ8+gq3+Mj1tqcfH0kAeOAQAqNX8X1b63djXNCzg84vTSx1vmJZpyz9gDfFnFAlQosBsJwLkkXqS7BxpgbjTFjjTGjjTG9i1k7LDd/ZxWxbwGQDnQzxgRXWaQix5QnH0tqM7NQHZHK1MwYc6dvnL3TGJNQQl3lqVS1bN9jTr4yjalS0xSVp7k0pkpNMcT3uDpfmcZTqYmKytVcGlOltIrKo1jf47Yi6ueWlWrdZWNMDHAezidCNpWmTUBpKp3k2voeNxezfwvOFc9tcNY1EakOscCHhcq2G2NusdbOz1dWbP5aa3OMMduB9jhnnjZUSaQix5QnH0tqk2iMOQLEGWPCrLXpVRCz1F39fFseY8w84CZr7e/5ysJxPjKWZq1NLKKfLb7HNlUUp5zkjDEBwF98T/P/k6gxVWqMEvI0l8ZUcYUx5j6c9UPr46xVeiHOhMtz+appPBXXlTJXc2lMlSKVMo/2A02BFjhrOueXe1Vyu2L67woMxpkvjsNZozkSZ6nh/aWJURPNzg8HILmY/bnlDaohFhGAd3EWbl8HpOIMBCOBO4CZxpgLrLWrfHWVv1KTlCcfS9Mm3FdPB/FSGdKBJ3FusJJ7Rj8BeAxnrec5xpjO1tojvn0aZ6WqPYdzs5YZ1tpv8pVrTJWapLg81ZgqbrsP54aVuWYBN/s+Ip5L46nUBKXJVY2pciKlyaNpODf2e8wYM9Ra6wEwxkQB9/rqBBtjQq21Rwv13xVnGeFcqcAt1trCF0IWS0tnnJjxPVbVuo8iBVhrH/etj7fXWpturV1rrR2Oc3PKUJw/MqWl/JWapDz5qByWSmWt3WetfcRau9xae9i3LcD59NISoBXw1/J0XamBSp1gjLkb5y7xG4FhZW3ue9SYKlWqpDzVmCpus9bGWmsNzidCr8K5SGeFMebsMnSj8VSqXGlyVWOqnEgpx7xHgN+Aa4GVxpiXjTGTcK5u9nLs5JiniP7f8PUfCpyJcyHkB8aYNwrXLY4mmo+d4alfzP7IQvVE3JL7i90zX5nyV2qS8uRjadukVCAukRPy3Sjjbd/TsoyzJ7qSRKRIxpi7gFdwDvp7W2sPFqqiMVVcV4o8LZLGVKluvot0vsSZkIsCPsi3W+Op1BgnyNXi2mhMlQJKyiPfDXzPAV7F+eTFCOBynCudL8aZRE621maV0H+GtXaDtXY08CZwpzHmmtLEponmY4tZF7dmTe6dG4tbw1mkuuzzPYbnKys2f31r6bXAuWFLUYvAi1S28uRjSW2a4uT7Lq19J9Uk9yNneeOs76OJu4F6vpwsTMcJUmbGmHuACcBanMm7PUVU05gqriplnpZEY6pUO2vtbzgnRtobYxr7ijWeSo1TTK6WRGOqHKe4PLLWJllrR1trT7fWBllrm1hrb8MZ7wzwcxleJvemkr1KU1kTzfC977G/MabA98MYEwF0B44CP1V3YCKFXOB7zH8ANNf3OKCI+j2BMOAHa21mVQYm4lOefCypzcBCdUSq2vm+x8In55SnUmmMMQ8CLwErcSbv9hVTVWOquKYMeVoSjanilma+x9yPhWs8lZqqcK6WRGOqFKcseXS77/HjMvR/iu8xpzSV6/xEs7X2V+BbIB64q9Dux3HOFn2Qb7F1kSpjjGlvjGlURPlpOFeUAHyUb9dknDuKXu+7O2hu/RDgKd/TiVUUrkhh5cnHd4FMYKQxJj5fm4bAWN/TUq8HJXIixpjzjDFBRZT3Af7ue/pRod25OfiQLzdz28TjHDtk4uSySImMMQ/j3FRtGdD3BHfv1pgqrihLnmpMFTcYY9oZY2KLKPczxjwNxOBMHB/y7dJ4Kq4oa65qTJWilCOP/Iwx9Yqo/1dgKM5J5I8L7buo8MW3vvKWwEO+p9NLFa+1WhPc9437AeeHMxXYAJyHc1fPzUA3a+0B9yKUusIY8xgwBudK++04d/hsCVwKhAAzgCvzr6VjjLkC5+ApA/gUOAhcBrT1lf/J6hddysmXX1f4nsYCl+CcRV/oK9tvrb2vUP0y5aMxZhTO+lEHgM+ALOAaIA4Yl79/kaKUJU+NMfOA9sA8YJdvfwLQx/f1w9ba3H8687/GOJy7NO/CyeUg4DqcNdFGWWsnFG4jkp8x5ibgPZyrTV6j6PUSd1hr38vXRmOqVKuy5qnGVHGDb1mXF4AFwK84410T4CKcG2PtwTlJsj5fG42nUu3KmqsaU6Uo5cijesBeYDaw1ddND+BcX/uLrbU7Cr3GYeAwzk0ndwIBOHNRA3xfv2atvbtU8Wr+yWGMaQ48gfNNjAISgSnA46W96YVIRRljLgKGA2fhTJaE4/yyrwQ+BD4satLYGNMd5yzTBTgT0luBd4BXrbWl+fiESJF8Jz8eLaHKb9ba+EJtypyPxpghwH3A2TiftlkPTLDWvl/BtyB1QFny1BhzG3Al0AFoDATiHIj9iJNzC4vrxDcBMxLnDsxeYDnwgrV2WsXfhZzsSpGnAPOttb0KtdOYKtWmrHmqMVXcYIzpAPwNZ5nLOKABcATnIrHpOOPjcf/DazyV6lbWXNWYKkUpRx4F4lzpfqGvPjgTzJOB8dbatCJeYzTOjQU7ANGAP07uLQHettZ+U+p4NdEsIiIiIiIiIiIiIhVR59doFhEREREREREREZGK0USziIiIiIiIiIiIiFSIJppFREREREREREREpEI00SwiIiIiIiIiIiIiFaKJZhERERERERERERGpEE00i4iIiIiIiIiIiEiFaKJZRERERERERERERCpEE80iIiIiIiIiIiIiUiGaaBYRERERERERERGRCtFEs4iIiIiIiIiIiIhUyP8Hs0OeLzyzNU0AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 1728x576 with 3 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Plotting the predicted fitness effect for the variants seen more than once\n",
    "plt.close('all')\n",
    "f, (ax1, ax2) = plt.subplots(2, 1, sharey=True, sharex = True, figsize=(24, 8))\n",
    "gs = matplotlib.gridspec.GridSpec(3, 1, width_ratios=[1], height_ratios=[10,0.5, 0.5])\n",
    "ax1 = plt.subplot(gs[0])\n",
    "ax2 = plt.subplot(gs[1])\n",
    "ax3 = plt.subplot(gs[2])\n",
    "gs.update(hspace=0.05)\n",
    "\n",
    "nonsense_list = []\n",
    "y_list = []\n",
    "x_list=[]\n",
    "y_list_top20 = []\n",
    "x_list_top20=[]\n",
    "y_list_nonsense = []\n",
    "x_list_nonsense = []\n",
    "x=[]\n",
    "y=[]\n",
    "OBS_EXP_DICT={}\n",
    "top20_lower_err = []\n",
    "top20_upper_err = []\n",
    "missense_lower_err = []\n",
    "missense_upper_err = []\n",
    "nonsense_lower_err = []\n",
    "nonsense_upper_err = []\n",
    "\n",
    "top20 = c1\n",
    "missense = c6\n",
    "nonsense = c10\n",
    "synonymous = c3\n",
    "\n",
    "TAD1_col = 'midnightblue'\n",
    "TAD2_col = 'blueviolet'\n",
    "PRO_col = 'lightseagreen'\n",
    "DBD_col = 'midnightblue'\n",
    "OD_col = 'darkorange'\n",
    "CTD_col = 'red'\n",
    "\n",
    "axisfont=20\n",
    "domainfont = 12\n",
    "titlefont=12\n",
    "axislabelfont=12\n",
    "\n",
    "top_20_TP53 = []\n",
    "\n",
    "for k, v in TP53_missense_s.items():\n",
    "    if k in top_20_TP53:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_top20.append(codon_number)\n",
    "        y_list_top20.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        top20_lower_err.append(error_lower)\n",
    "        top20_upper_err.append(error_upper)\n",
    "\n",
    "    if k not in top_20_TP53:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list.append(codon_number)\n",
    "        y_list.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        missense_lower_err.append(error_lower)\n",
    "        missense_upper_err.append(error_upper)\n",
    "\n",
    "    missense_err = [tuple(missense_lower_err), tuple(missense_upper_err)]   \n",
    "    \n",
    "for k, v in TP53_nonsense_s.items():\n",
    "    if k in top_20_TP53:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_top20.append(codon_number)\n",
    "        y_list_top20.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        top20_lower_err.append(error_lower)\n",
    "        top20_upper_err.append(error_upper)\n",
    "    \n",
    "    top20_err = [tuple(top20_lower_err), tuple(top20_upper_err)]\n",
    "        \n",
    "    if k not in top_20_TP53:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_nonsense.append(codon_number)\n",
    "        y_list_nonsense.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        nonsense_lower_err.append(error_lower)\n",
    "        nonsense_upper_err.append(error_upper)\n",
    "\n",
    "    nonsense_err = [tuple(nonsense_lower_err), tuple(nonsense_upper_err)]        \n",
    "    \n",
    "msize = 12\n",
    "medgewidth = 2\n",
    "linew = 3\n",
    "csize = 4\n",
    "cthick = 3\n",
    "\n",
    "markers1, caps1, bars1 = ax1.errorbar(x_list, y_list, yerr = missense_err, fmt = 'o', ecolor = missense, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = missense, markerfacecolor = missense, zorder = 1)    \n",
    "\n",
    "markers2, caps2, bars2 = ax1.errorbar(x_list_nonsense, y_list_nonsense, yerr = nonsense_err, fmt = 'o', ecolor = nonsense, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = nonsense, markerfacecolor = nonsense, zorder = 2)    \n",
    "\n",
    "markers3, caps3, bars3 = ax1.errorbar(x_list_top20, y_list_top20, yerr = top20_err, fmt = 'o', ecolor = top20, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = top20, markerfacecolor = top20, zorder = 3)    \n",
    "\n",
    "# loop through bars and caps and set the alpha value\n",
    "a = 0.7\n",
    "[bar.set_alpha(a) for bar in bars1]\n",
    "[cap.set_alpha(a) for cap in caps1]\n",
    "[bar.set_alpha(a) for bar in bars2]\n",
    "[cap.set_alpha(a) for cap in caps2]\n",
    "[bar.set_alpha(a) for bar in bars3]\n",
    "[cap.set_alpha(a) for cap in caps3]\n",
    "\n",
    "#Only show the required axis lines\n",
    "ax1.spines['top'].set_visible(False)\n",
    "ax1.spines['right'].set_visible(False)\n",
    "\n",
    "for axis in ['bottom','left']:\n",
    "    ax1.spines[axis].set_linewidth(2)\n",
    "\n",
    "for axis in ['bottom','left']:\n",
    "    ax1.spines[axis].set_color(grey3)\n",
    "    \n",
    "ax1.set_ylabel('fitness effect (s)', fontsize = axisfont, labelpad = 10)\n",
    "ax1.set_title('Fitness effects across TP53', fontsize = 18)\n",
    "\n",
    "# ax1.set_yscale('log')\n",
    "\n",
    "# STACKED BAR CHART\n",
    "# define the exons + create data\n",
    "exons = [{'2':(25-1)+1, '3': (32-26)+1, '4': (125-33)+1, '5':(187-126)+1, '6':(224-188)+1, '7': (261-225)+1, '8': (307-262)+1,'9': (331-308)+1, '10': (367-332)+1, '11': (393-368)+1}]\n",
    "exon_orders = [['2', '3', '4', '5', '6', '7', '8', '9', '10', '11']]\n",
    "colors2 = [c4, c0, c3, c1, c5, c6,c7, c8, c9, c10]\n",
    "names = exons[0].keys()\n",
    "values = np.array([[data[name] for name in order] for data,order in zip(exons, exon_orders)])\n",
    "lefts = np.insert(np.cumsum(values, axis=1),0,0, axis=1)[:, :-1]\n",
    "orders = np.array(exon_orders)\n",
    "\n",
    "#plot the bar chart\n",
    "for name, color in zip(names, colors2):\n",
    "    idx = np.where(orders == name)\n",
    "    value = values[idx]\n",
    "    left = lefts[idx]\n",
    "    ax2.bar(x=left, height=-10, width=value, bottom=-5, color=color, orientation = 'horizontal', label = name, edgecolor='k', linewidth=1, tick_label = name)\n",
    "    ax2.text(left + (value/2.), (-0.26), name, ha='center', va='center', fontsize = 12)\n",
    "# STACKED BAR CHART of domains    \n",
    "domains = [{'TAD1': (42-1)+1, 'TAD2' :(63-43)+1, 'PRO':(92-63)+1, ' ':(99-93)+1, 'DNA binding domain': (299-100)+1, '  ':(325-300)+1, 'OD':(356-326)+1, '   ':(363-357)+1, 'CTD':(393-364)+1}]\n",
    "domain_orders = [['TAD1','TAD2','PRO', ' ','DNA binding domain','  ','OD','   ','CTD']]\n",
    "# blank_col = c13\n",
    "blank_col = grey1\n",
    "colors3 = [TAD1_col, TAD2_col, PRO_col, blank_col, DBD_col, blank_col, OD_col, blank_col, CTD_col]\n",
    "names_domains = domains[0].keys()\n",
    "values = np.array([[data[name] for name in order] for data,order in zip(domains, domain_orders)])\n",
    "lefts = np.insert(np.cumsum(values, axis=1),0,0, axis=1)[:, :-1]\n",
    "orders = np.array(domain_orders)\n",
    "\n",
    "# #plot the bar chart\n",
    "for name, color in zip(names_domains, colors3):\n",
    "    idx = np.where(orders == name)\n",
    "    value = values[idx]\n",
    "    left = lefts[idx]\n",
    "    ax3.bar(x=left, height=-10, width=value, bottom=-5, color=color, orientation = 'horizontal', alpha = 0.9, label = name, edgecolor='k', linewidth=1)\n",
    "    ax3.text(left + (value/2.), (-0.25), name, ha='center', va='center', fontsize = domainfont, color = 'white', fontweight = 530)\n",
    "\n",
    "    \n",
    "# CONFIGURING THE GRAPH APPEARANCE\n",
    "#Set the x and y axis limits\n",
    "ax1.set_xlim(0, 393)\n",
    "ax1.set_ylim(0, 25)\n",
    "# ax1.set_ylim(10**1, 10**5)\n",
    "ax2.set_xlim(0, 393)\n",
    "ax2.set_ylim(-0.5, 0)\n",
    "ax3.set_xlim(0, 393)\n",
    "ax3.set_ylim(-0.5, 0)\n",
    "\n",
    "#x-axis ticks\n",
    "x1_major_ticks = []\n",
    "x1_major_tick_labels = []\n",
    "x2_major_ticks = [0, 50, 100, 150, 200, 250, 300, 350, 393]\n",
    "x2_major_tick_labels = [\"0\", \"50\", \"100\", \"150\", \"200\", \"250\", \"300\", \"350\", \"393\"]\n",
    "ax1.set_xticks(x1_major_ticks)\n",
    "ax1.set_xticklabels(x1_major_tick_labels, fontsize = axisfont)\n",
    "ax2.set_xticklabels([])\n",
    "ax2.set_xticks([])\n",
    "ax3.set_xticks(x2_major_ticks)\n",
    "ax3.set_xticklabels(x2_major_tick_labels, fontsize = axisfont)\n",
    "\n",
    "\n",
    "#y-axis ticks\n",
    "y1_major_ticks = [0, 5, 10, 15, 20, 25]\n",
    "y1_major_tick_labels = [\"0\", \"5\", \"10\", \"15\", \"20\", \"25\"]\n",
    "ax1.set_yticks(y1_major_ticks)\n",
    "ax1.set_yticklabels(y1_major_tick_labels, fontsize = axisfont)\n",
    "\n",
    "ax2.set_yticklabels([])\n",
    "ax2.set_yticks([])\n",
    "ax3.set_yticklabels([])\n",
    "ax3.set_yticks([])\n",
    "\n",
    "#set panel legends\n",
    "legend_marker_size = 2\n",
    "legend_elements = [Line2D([0], [0], marker = '_', color = missense, alpha = 1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='missense'),\n",
    "                   Line2D([0], [0], marker = '_', color=nonsense, alpha=1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='nonsense'),\n",
    "                  Line2D([0], [0], marker = '_', color=top20, alpha=1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='top 20 observed variant')]\n",
    "    \n",
    "ax1.legend(ncol=1, handles=legend_elements, loc='upper left', frameon=0, fontsize = 20)\n",
    "\n",
    "ax1.grid(which='major', linestyle=':', linewidth='0.7', color=grey3, zorder = 0)\n",
    "\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
