{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "toc": true
   },
   "source": [
    "<h1>Table of Contents<span class=\"tocSkip\"></span></h1>\n",
    "<div class=\"toc\"><ul class=\"toc-item\"><li><span><a href=\"#Python-Setup\" data-toc-modified-id=\"Python-Setup-1\"><span class=\"toc-item-num\">1&nbsp;&nbsp;</span>Python Setup</a></span><ul class=\"toc-item\"><li><span><a href=\"#Change-the-width-of-the-page\" data-toc-modified-id=\"Change-the-width-of-the-page-1.1\"><span class=\"toc-item-num\">1.1&nbsp;&nbsp;</span>Change the width of the page</a></span></li><li><span><a href=\"#Import-packages\" data-toc-modified-id=\"Import-packages-1.2\"><span class=\"toc-item-num\">1.2&nbsp;&nbsp;</span>Import packages</a></span></li></ul></li><li><span><a href=\"#Colours\" data-toc-modified-id=\"Colours-2\"><span class=\"toc-item-num\">2&nbsp;&nbsp;</span>Colours</a></span><ul class=\"toc-item\"><li><span><a href=\"#Colour-line-graph\" data-toc-modified-id=\"Colour-line-graph-2.1\"><span class=\"toc-item-num\">2.1&nbsp;&nbsp;</span>Colour line graph</a></span></li></ul></li><li><span><a href=\"#Totals-for-studies\" data-toc-modified-id=\"Totals-for-studies-3\"><span class=\"toc-item-num\">3&nbsp;&nbsp;</span>Totals for studies</a></span></li><li><span><a href=\"#Functions-for-calculating-trinucleotide-context-specific-mutation-rates\" data-toc-modified-id=\"Functions-for-calculating-trinucleotide-context-specific-mutation-rates-4\"><span class=\"toc-item-num\">4&nbsp;&nbsp;</span>Functions for calculating trinucleotide-context specific mutation rates</a></span><ul class=\"toc-item\"><li><span><a href=\"#Calculating-mutation-rates-for-individual-variants\" data-toc-modified-id=\"Calculating-mutation-rates-for-individual-variants-4.1\"><span class=\"toc-item-num\">4.1&nbsp;&nbsp;</span>Calculating mutation rates for individual variants</a></span><ul class=\"toc-item\"><li><span><a href=\"#DNMT3A\" data-toc-modified-id=\"DNMT3A-4.1.1\"><span class=\"toc-item-num\">4.1.1&nbsp;&nbsp;</span>DNMT3A</a></span></li><li><span><a href=\"#TET2\" data-toc-modified-id=\"TET2-4.1.2\"><span class=\"toc-item-num\">4.1.2&nbsp;&nbsp;</span>TET2</a></span></li><li><span><a href=\"#ASXL1\" data-toc-modified-id=\"ASXL1-4.1.3\"><span class=\"toc-item-num\">4.1.3&nbsp;&nbsp;</span>ASXL1</a></span></li><li><span><a href=\"#TP53\" data-toc-modified-id=\"TP53-4.1.4\"><span class=\"toc-item-num\">4.1.4&nbsp;&nbsp;</span>TP53</a></span></li></ul></li><li><span><a href=\"#Calculating-mutation-rates-from-a-.csv-file-of-variants\" data-toc-modified-id=\"Calculating-mutation-rates-from-a-.csv-file-of-variants-4.2\"><span class=\"toc-item-num\">4.2&nbsp;&nbsp;</span>Calculating mutation rates from a .csv file of variants</a></span><ul class=\"toc-item\"><li><span><a href=\"#DNMT3A\" data-toc-modified-id=\"DNMT3A-4.2.1\"><span class=\"toc-item-num\">4.2.1&nbsp;&nbsp;</span>DNMT3A</a></span></li><li><span><a href=\"#TET2\" data-toc-modified-id=\"TET2-4.2.2\"><span class=\"toc-item-num\">4.2.2&nbsp;&nbsp;</span>TET2</a></span></li><li><span><a href=\"#ASXL1\" data-toc-modified-id=\"ASXL1-4.2.3\"><span class=\"toc-item-num\">4.2.3&nbsp;&nbsp;</span>ASXL1</a></span></li><li><span><a href=\"#TP53\" data-toc-modified-id=\"TP53-4.2.4\"><span class=\"toc-item-num\">4.2.4&nbsp;&nbsp;</span>TP53</a></span></li></ul></li><li><span><a href=\"#Calculating-mutation-rates-from-a-list-of-variants\" data-toc-modified-id=\"Calculating-mutation-rates-from-a-list-of-variants-4.3\"><span class=\"toc-item-num\">4.3&nbsp;&nbsp;</span>Calculating mutation rates from a list of variants</a></span><ul class=\"toc-item\"><li><span><a href=\"#DNMT3A\" data-toc-modified-id=\"DNMT3A-4.3.1\"><span class=\"toc-item-num\">4.3.1&nbsp;&nbsp;</span>DNMT3A</a></span></li><li><span><a href=\"#TET2\" data-toc-modified-id=\"TET2-4.3.2\"><span class=\"toc-item-num\">4.3.2&nbsp;&nbsp;</span>TET2</a></span></li><li><span><a href=\"#ASXL1\" data-toc-modified-id=\"ASXL1-4.3.3\"><span class=\"toc-item-num\">4.3.3&nbsp;&nbsp;</span>ASXL1</a></span></li><li><span><a href=\"#TP53\" data-toc-modified-id=\"TP53-4.3.4\"><span class=\"toc-item-num\">4.3.4&nbsp;&nbsp;</span>TP53</a></span></li></ul></li></ul></li><li><span><a href=\"#Lists-of-variants-targeted-by-each-study\" data-toc-modified-id=\"Lists-of-variants-targeted-by-each-study-5\"><span class=\"toc-item-num\">5&nbsp;&nbsp;</span>Lists of variants targeted by each study</a></span><ul class=\"toc-item\"><li><span><a href=\"#Jaiswal-2014\" data-toc-modified-id=\"Jaiswal-2014-5.1\"><span class=\"toc-item-num\">5.1&nbsp;&nbsp;</span>Jaiswal 2014</a></span></li><li><span><a href=\"#Genovese-2014\" data-toc-modified-id=\"Genovese-2014-5.2\"><span class=\"toc-item-num\">5.2&nbsp;&nbsp;</span>Genovese 2014</a></span></li><li><span><a href=\"#McKerrel-2015\" data-toc-modified-id=\"McKerrel-2015-5.3\"><span class=\"toc-item-num\">5.3&nbsp;&nbsp;</span>McKerrel 2015</a></span></li><li><span><a href=\"#Zink-2017\" data-toc-modified-id=\"Zink-2017-5.4\"><span class=\"toc-item-num\">5.4&nbsp;&nbsp;</span>Zink 2017</a></span></li><li><span><a href=\"#Coombs-2017\" data-toc-modified-id=\"Coombs-2017-5.5\"><span class=\"toc-item-num\">5.5&nbsp;&nbsp;</span>Coombs 2017</a></span></li><li><span><a href=\"#Young-2016-&amp;-2019\" data-toc-modified-id=\"Young-2016-&amp;-2019-5.6\"><span class=\"toc-item-num\">5.6&nbsp;&nbsp;</span>Young 2016 &amp; 2019</a></span></li><li><span><a href=\"#Desai-2018\" data-toc-modified-id=\"Desai-2018-5.7\"><span class=\"toc-item-num\">5.7&nbsp;&nbsp;</span>Desai 2018</a></span></li><li><span><a href=\"#Acuna-Hidalgo-2017\" data-toc-modified-id=\"Acuna-Hidalgo-2017-5.8\"><span class=\"toc-item-num\">5.8&nbsp;&nbsp;</span>Acuna-Hidalgo 2017</a></span></li></ul></li><li><span><a href=\"#Lists-of-all-possible-variants-in-DNMT3A,-TET2,-ASXL1,-TP53\" data-toc-modified-id=\"Lists-of-all-possible-variants-in-DNMT3A,-TET2,-ASXL1,-TP53-6\"><span class=\"toc-item-num\">6&nbsp;&nbsp;</span>Lists of all possible variants in DNMT3A, TET2, ASXL1, TP53</a></span><ul class=\"toc-item\"><li><span><a href=\"#DNMT3A\" data-toc-modified-id=\"DNMT3A-6.1\"><span class=\"toc-item-num\">6.1&nbsp;&nbsp;</span>DNMT3A</a></span></li><li><span><a href=\"#TET2\" data-toc-modified-id=\"TET2-6.2\"><span class=\"toc-item-num\">6.2&nbsp;&nbsp;</span>TET2</a></span></li><li><span><a href=\"#ASXL1\" data-toc-modified-id=\"ASXL1-6.3\"><span class=\"toc-item-num\">6.3&nbsp;&nbsp;</span>ASXL1</a></span></li><li><span><a href=\"#TP53\" data-toc-modified-id=\"TP53-6.4\"><span class=\"toc-item-num\">6.4&nbsp;&nbsp;</span>TP53</a></span></li></ul></li><li><span><a href=\"#Actual-number-of-observations-of-each-variant\" data-toc-modified-id=\"Actual-number-of-observations-of-each-variant-7\"><span class=\"toc-item-num\">7&nbsp;&nbsp;</span>Actual number of observations of each variant</a></span><ul class=\"toc-item\"><li><ul class=\"toc-item\"><li><span><a href=\"#DNMT3A\" data-toc-modified-id=\"DNMT3A-7.0.1\"><span class=\"toc-item-num\">7.0.1&nbsp;&nbsp;</span>DNMT3A</a></span></li><li><span><a href=\"#TET2\" data-toc-modified-id=\"TET2-7.0.2\"><span class=\"toc-item-num\">7.0.2&nbsp;&nbsp;</span>TET2</a></span></li><li><span><a href=\"#ASXL1\" data-toc-modified-id=\"ASXL1-7.0.3\"><span class=\"toc-item-num\">7.0.3&nbsp;&nbsp;</span>ASXL1</a></span></li><li><span><a href=\"#TP53\" data-toc-modified-id=\"TP53-7.0.4\"><span class=\"toc-item-num\">7.0.4&nbsp;&nbsp;</span>TP53</a></span></li></ul></li></ul></li><li><span><a href=\"#Functions-for-calculating-the-expected-number-of-observations-of-a-variant\" data-toc-modified-id=\"Functions-for-calculating-the-expected-number-of-observations-of-a-variant-8\"><span class=\"toc-item-num\">8&nbsp;&nbsp;</span>Functions for calculating the expected number of observations of a variant</a></span></li><li><span><a href=\"#Maximum-Likelihood-Estimation-for-s\" data-toc-modified-id=\"Maximum-Likelihood-Estimation-for-s-9\"><span class=\"toc-item-num\">9&nbsp;&nbsp;</span>Maximum Likelihood Estimation for s</a></span><ul class=\"toc-item\"><li><span><a href=\"#DNMT3A-variants\" data-toc-modified-id=\"DNMT3A-variants-9.1\"><span class=\"toc-item-num\">9.1&nbsp;&nbsp;</span>DNMT3A variants</a></span></li><li><span><a href=\"#TET2-variants\" data-toc-modified-id=\"TET2-variants-9.2\"><span class=\"toc-item-num\">9.2&nbsp;&nbsp;</span>TET2 variants</a></span></li><li><span><a href=\"#ASXL1-variants\" data-toc-modified-id=\"ASXL1-variants-9.3\"><span class=\"toc-item-num\">9.3&nbsp;&nbsp;</span>ASXL1 variants</a></span></li><li><span><a href=\"#TP53-variants\" data-toc-modified-id=\"TP53-variants-9.4\"><span class=\"toc-item-num\">9.4&nbsp;&nbsp;</span>TP53 variants</a></span></li></ul></li></ul></div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Python Setup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Change the width of the page"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>.container { width:80% !important; }</style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.core.display import display, HTML\n",
    "display(HTML(\"<style>.container { width:80% !important; }</style>\"))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Import packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "# imported packages\n",
    "import random\n",
    "import numpy as np\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import math\n",
    "import matplotlib.ticker as plticker\n",
    "from matplotlib.ticker import LinearLocator, FormatStrFormatter\n",
    "from matplotlib.patches import Polygon\n",
    "from matplotlib.patches import Patch\n",
    "from matplotlib.lines import Line2D\n",
    "from matplotlib import cm\n",
    "import scipy.special\n",
    "import scipy.integrate as it\n",
    "from scipy import integrate\n",
    "from scipy.interpolate import interp1d\n",
    "from scipy.stats import kde\n",
    "import copy\n",
    "import glob, os\n",
    "import re\n",
    "from sklearn import datasets, linear_model\n",
    "import pandas as pd\n",
    "from decimal import *\n",
    "from operator import itemgetter    \n",
    "from collections import OrderedDict\n",
    "import timeit"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Colours"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "# Lists of colors for plots\n",
    "c0 = (0.76, 0.76, 0.76)\n",
    "c1 = (1.00, 0.18, 0.33);\n",
    "c2 = (1.00, 0.23, 0.19);\n",
    "c3 = (1.00, 0.58, 0.00);\n",
    "c4 = (1.00, 0.80, 0.00);\n",
    "c5 = (0.30, 0.85, 0.39);\n",
    "c6 = (0.35, 0.78, 0.98);\n",
    "c7 = (0.20, 0.67, 0.86);\n",
    "c8 = (0.00, 0.48, 1.00);\n",
    "c9 = (0.35, 0.34, 0.84);\n",
    "c10 = (0.00, 0.31, 0.57);\n",
    "c11 = (0.12, 0.29, 0.69);\n",
    "c12 = (0.17, 0.17, 0.42);\n",
    "c13 = (1.00, 1.00, 1.00);\n",
    "c14 = (0.77, 0.04, 0.00);"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#define the colors from colorbrewer2\n",
    "orange1 = '#feedde'\n",
    "orange2 = '#fdbe85'\n",
    "orange3 = '#fd8d3c'\n",
    "orange4 = '#e6550d'\n",
    "orange5 = '#a63603'\n",
    "blue1 = '#eff3ff'\n",
    "blue2 = '#bdd7e7'\n",
    "blue3 = '#6baed6'\n",
    "blue4 = '#3182bd'\n",
    "blue5 = '#08519c'\n",
    "green1 = '#edf8e9'\n",
    "green2 = '#bae4b3'\n",
    "green3 = '#74c476'\n",
    "green4 = '#31a354'\n",
    "green5 = '#006d2c'\n",
    "grey1 = '#f7f7f7'\n",
    "grey2 = '#cccccc'\n",
    "grey3 = '#969696'\n",
    "grey4 = '#636363'\n",
    "grey5 = '#252525'\n",
    "purple1 = '#f2f0f7'\n",
    "purple2 = '#cbc9e2'\n",
    "purple3 = '#9e9ac8'\n",
    "purple4 = '#756bb1'\n",
    "purple5 = '#54278f'\n",
    "red1 = '#fee5d9'\n",
    "red2 = '#fcae91'\n",
    "red3 = '#fb6a4a'\n",
    "red4 = '#de2d26'\n",
    "red5 = '#a50f15'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Colour line graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAlkAAAExCAYAAABPt7ftAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAF0xJREFUeJzt3X2QnWd53/Hvj2KcENtgrTAOkW11hpeCiIDhOIEqGEcwQNrBocUW5EWyOmZ2mBiHiSs6oetpAq5CQsDQDs6MRcuwQMybGqfEaUiYJVZCB5geWfbSJVGdpn6R7WLWshkgDl7JV//Y48x22dUenec82j2r72dmZ3Tu+36uc9l//ea+n/M8qSokSZI0XE9Z7QYkSZLWI0OWJElSCwxZkiRJLTBkSZIktcCQJUmS1AJDliRJUgsMWZJGVpJLkxxZ7T4kaSmGLElrQpJfTNJN8r0kDyb5kyQ/s9p9SdKgDFmSVl2Sa4EPA78FPBu4EPg94OdXqZ+nrsb3SlpfDFmSVlWSZwDvBa6uqj+oqu9X1VxV/VFVvSvJmUk+nOSB3t+Hk5y5TK0XJrktyaNJZpJctmDutiRvW/B5d5KvLPhcSa5OchdwV+Z9KMlDSb6TZDrJi1v8XyFpnTFkSVptrwR+BLhlmfkJ4BXAS4GXAD8FXLd4UZIzgD8C/gw4D7gG+P0kLziJXt4E/DTwIuB1wCXA84FnAm8BHj6JWpJOc4YsSattDJitqmPLzP8S8N6qeqiqvg28B9i5xLpXAGcBv11Vj1fVl4FbgV84iV7eV1VHq+oxYA44G/gnQKrqr6rqwZOoJek0Z8iStNoeBjae4D6o5wD3LPh8T29sqXX3VdUTi9b+xEn0ct+T/+iFtI8ANwLfSrIvyTknUUvSac6QJWm1fRX4e+aP6pbyAHDRgs8X9saWWndBkqcsWnt/79/fB56+YO78JWrU//eh6j9W1cuBLcwfG75rmR4l6YcYsiStqqr6DvDvgBuTvCnJ05OckeTnkrwf+DRwXZJnJdnYW/upJUp9nfkg9W96118KvBH4TG/+DuBf9uo/F7jqRH0luTjJT/fu9fo+80HwePP/YkmnC0OWpFVXVTcA1zJ/Q/u3mT+2ewfwh8C/B7rANPAN4Pbe2OIajwOXAT8HzDL/CIhdVfXXvSUfAh4HvgVMAr+/QlvnAB8FHmH+2PFh4AOD/jdKOv2kqlZeJUmSpJPiTpYkSVILDFmSJEktMGRJkiS1wJAlSZLUAkOWJElSC1b9TfMbN26szZs3r3YbkiRJKzp48OBsVT2rn7WrHrI2b95Mt9td7TYkSZJWlOSelVfN87hQkiSpBYYsSZKkFhiyJEmSWmDIkiRJaoEhS5IkqQWGLEmSpBYYsiRJklpgyJIkSWqBIUuSJKkFhixJkqQWGLIkSZJaYMiSJElqwUAhK8klSW5PcizJ5UvMn5Pk/iQfad6iJEnS6Bl0J+teYDdw8zLz1wMHBqwtSZI08p7az6Iku4A9QAHTVbWzN/7EEmtfDjwb+CLQWan2Y3cd5htvuPQkWpYkSVr7VgxZSbYAE8C2qppNsuEEa58CfBDYCbzmBOvGgXGA5/3omSfbsyRJ0prXz07WdmB/Vc0CVNXRE6z9FeC/VdV9SZZdVFX7gH0AnU6nfvKLt/XdsCRJ0qo5Qb5ZrJ+QFeaPCfvxSuBVSX4FOAt4WpLvVdWv992RJEnSOtDPje9TwI4kYwAnOi6sql+qqgurajPz93B9woAlSZJORyuGrKqaAfYCB5LcCdyQ5OIkR4ArgJuSzLTcpyRJ0kjp69eFVTUJTC4a3rTCNR8HPj5QV5IkSSPOJ75LkiS1wJAlSZLUAkOWJElSCwxZkiRJLTBkSZIktcCQJUmS1AJDliRJUgsMWZIkSS0YKGQluSTJ7UmOJbl8wfhFSQ4muSPJTJK3D69VSZKk0dHXE9+XcC+wm/n3Ey70IPBPq+oHSc4C/meSL1TVAw16lCRJGjl9hawku5gPVAVMV9XO3vgTC9dV1eMLPp6Jx5GSJOk0tWLISrIFmAC2VdVskg0rrL8A+GPgucC73MWSJEmno352mrYD+6tqFqCqjp5ocVXdV1VbmQ9ZVyZ59uI1ScaTdJN05+bmBulbkiRpTesnZIX5Y8KT0tvBmgFetcTcvqrqVFXnjDPOONnSkiRJa14/IWsK2JFkDOBEx4VJNiX50d6/zwW2AYeH0agkSdIoWfGerKqaSbIXOJDkOHAoyY3ALcC5wBuTvKeqtgAvBD6YpJjfAftAVX2jxf4lSZLWpL5+XVhVk8DkouFNS6z7ErB1CH1JkiSNNB+xIEmS1AJDliRJUgsMWZIkSS0wZEmSJLXAkCVJktQCQ5YkSVILDFmSJEktMGRJkiS1wJAlSZLUgoFCVpJLktye5FiSyxeMvzTJV5PMJJlO8pbhtSpJkjQ6+nqtzhLuBXYDexaN/x2wq6ruSvIc4GCSP62qRxv0KEmSNHL6CllJdjEfqAqYrqqdvfEnFq6rqv+14N8PJHkIeBawbMi6++5vsXv3hwdoXZIkae1aMWQl2QJMANuqajbJhn4KJ/kp4GnA/15ibhwYBzjnnGefVMOSJEmjIFV14gXJNcD5VTWxxNzHgVurav+i8R8HbgOurKqvnah+p9Opbrd7km1LkiSdekkOVlWnn7X93Pge5o8J+/3yc4A/Bq5bKWBJkiStV/2ErClgR5IxgBMdFyZ5GnAL8Imq+vxwWpQkSRo9K4asqpoB9gIHktwJ3JDk4iRHgCuAm5LM9JbvAC4Bdie5o/f30raalyRJWqv6+nVhVU0Ck4uGNy2x7lPAp4bQlyRJ0kjzie+SJEktMGRJkiS1wJAlSZLUAkOWJElSCwxZkiRJLTBkSZIktcCQJUmS1AJDliRJUgsGCllJLklye5JjSS5fNPfFJI8muXU4LUqSJI2eQXey7gV2AzcvMfe7wM5BG5IkSVoP+gpZSXYlmU5yZ5JPVtXdVTUNPLF4bVVNAd8ddqOSJEmjZMV3FybZAkwA26pqNsmGYTbwt/d/j7f8+leGWVKSJGnV9bOTtR3YX1WzAFV1tOmXJhlP0k3SnXt8rmk5SZKkNWfFnSwgQA3zS6tqH7APoNPp1Gd/+2eGWV6SJKkVn/ud/tf2s5M1BexIMgYw7ONCSZKk9WjFkFVVM8Be4ECSO4Ebklyc5AhwBXBTkpkn1yf5S+DzwGuSHEny+pZ6lyRJWrP6OS6kqiaByUXDm5ZZ+6qmTUmSJI06n/guSZLUAkOWJElSCwxZkiRJLTBkSZIktcCQJUmS1AJDliRJUgsMWZIkSS0wZEmSJLXAkCVJktSCgUJWkkuS3J7kWJLLF81dmeSu3t+Vw2lTkiRptPT1Wp0l3AvsBvYsHOy9PPo3gA5QwMEkX6iqR5o0KUmSNGr6CllJdjEfqAqYrqqdvfEnFi19PfClqjram/8S8Abg08vVPnz/I1w68V8GaF2SJGntWjFkJdkCTADbqmq2t1u1nJ8A7lvw+UhvbHHNcWAc4MyxC06qYUmSpFHQz07WdmB/Vc0CPLlLtYwsMVY/NFC1D9gH0Ol06ra9b+6jDUmSpNWV3+p/bT83voclgtIyjgALt6Y2AQ/0344kSdL60E/ImgJ2JBmDf7i5fTl/CrwuyblJzgVe1xuTJEk6rax4XFhVM0n2AgeSHAcOJbkRuAU4F3hjkvdU1ZaqOprkeuB/9C5/7wrHi5IkSetSX78urKpJYHLR8KZl1n4M+FjDviRJkkaaT3yXJElqgSFLkiSpBYYsSZKkFhiyJEmSWmDIkiRJaoEhS5IkqQWGLEmSpBYYsiRJklow1JCV5KIkU0mmk9yWZMkHlkqSJK13w97J+gDwiaraCrwXeN+Q60uSJI2ERiErya7ertWdST4JvIj5F0oD/Dnw800blCRJGkV9vbtwKUm2ABPAtqqaTbIB+AjwZuA/AP8CODvJWFU9vFydBx94nOvfc8+gbUiSJK1JTXaytgP7q2oWoKqOAnuAVyc5BLwauB84tvjCJONJukm6c3NzDVqQJElam1JVg12Y/CpwXlVdt8z8WcBfV9UJb37vdDrV7XYH6kGSJOlUSnKwqjr9rG2ykzUF7Egy1vvSDUk2Jnmy5ruBjzWoL0mSNLIGvierqmaS7AUOJDkOHAJuBd6XpIC/AK4eTpuSJEmjZeCQBVBVk8DkouH9TWpKkiStBz7xXZIkqQWGLEmSpBYYsiRJklpgyJIkSWqBIUuSJKkFhixJkqQWGLIkSZJaYMiSJElqgSFLkiSpBUMNWUkuTPLnSQ4lmU7yz4ZZX5IkaVQMeyfrOuBzVfUy4K3A7w25viRJ0kho9O7CJLuAPUAB08DfAef0pp8BPLBSjcOzcOl/atKFJEnS2jNwyEqyBZgAtlXVbJINwJnAnyW5Bvgx4LXLXDsOjAOc+Zytg7YgSZK0ZqWqBrtwPkidX1UTC8au7dX8YJJXAv8ZeHFVPbFcnU6nU91ud6AeJEmSTqUkB6uq08/aJvdkhfljwoWuAj4HUFVfBX4E2NjgOyRJkkZSk5A1BexIMgbQOy68F3hN7/MLmQ9Z327apCRJ0qgZ+J6sqppJshc4kOQ4cAj418BHk/wa87tcu2vQ80hJkqQR1ujXhVU1CUwuGt7WpKYkSdJ64BPfJUmSWmDIkiRJaoEhS5IkqQWGLEmSpBYYsiRJklpgyJIkSWqBIUuSJKkFhixJkqQWNHoY6WJJPgT8bO/j04HzquqZw/wOSZKkUTDUkFVVv/bkv5NcA7xsmPUlSZJGRaPjwiS7kkwnuTPJJxdN/wLw6Sb1JUmSRtXAO1lJtgATwLaqmk2yYcHcRcA/Br68Up17vvs447cdGbQNSZKkNanJTtZ2YH9VzQJU1dEFc2/tzR1f6sIk40m6Sbpzc3MNWpAkSVqbmtyTFaCWmXsrcPVyF1bVPmAfQKfTqX2XbmrQhiRJ0qnx0ZNY22QnawrYkWQM4MnjwiQvAM4FvtqgtiRJ0kgbeCerqmaS7AUOJDkOHAJ2M3/D+2eqarldLkmSpHWv0SMcqmoSmFw09ptNakqSJK0HPvFdkiSpBYYsSZKkFhiyJEmSWmDIkiRJaoEhS5IkqQWGLEmSpBYYsiRJklpgyJIkSWqBIUuSJKkFQw9ZSXYk+WaSmSQ3D7u+JEnSKGj0Wp3FkjwPeDewraoeSXLeMOtLkiSNikYhK8kuYA9QwDTwIHBjVT0CUFUPrVTjwceK62eONWlDkiRpzRn4uDDJFmAC2F5VLwHeCTwfeH6S/57ka0nesMy140m6Sbpzc3ODtiBJkrRmpaoGuzC5Bji/qiYWjN0KzAE7gE3AXwIvrqpHl6vT6XSq2+0O1IMkSdKplORgVXX6Wdvkxvcwf0y40BHgv1bVXFX9H+Aw8LwG3yFJkjSSmoSsKWBHkjGAJBuAPwR+tvd5I/PHh3/btElJkqRRM/CN71U1k2QvcCDJceAQ8K+A1yX5JnAceFdVPTycViVJkkZHo18XVtUkMLlo+NrenyRJ0mnLJ75LkiS1wJAlSZLUAkOWJElSCwxZkiRJLTBkSZIktcCQJUmS1AJDliRJUgsMWZIkSS0wZEmSJLVgqCErye4k305yR+/vbcOsL0mSNCoavVZnGZ+tqne0UFeSJGlkNApZSXYBe4ACpoGpk61x/+P/l3973/ubtCFJkrTmDHxcmGQLMAFsr6qXAO/sTb05yXSS/UkuWOba8STdJN25ublBW5AkSVqzUlWDXZhcA5xfVRMLxsaA71XVD5K8HdhRVdtPVKfT6VS32x2oB0mSpFMpycGq6vSztsmN72H+mPAfVNXDVfWD3sePAi9vUF+SJGlkNQlZU8CO3u4VSTYk+fEF85cBf9WkOUmSpFE18I3vVTWTZC9wIMlx4BDwYJLLgGPAUWD3ULqUJEkaMY1+XVhVk8DkouF3N6kpSZK0HvjEd0mSpBYYsiRJklpgyJIkSWqBIUuSJKkFhixJkqQWGLIkSZJaYMiSJElqgSFLkiSpBa2ErCSXJ6kkfb1AUZIkab0ZeshKcjbwq8DXh11bkiRpVDR6rU6SXcAeoIDpqtoJXA+8vze+sr8/DIcvbdKGJEnSmjPwTlaSLcAEsL2qXgK8M8nLgAuq6tYVrh1P0k3SnZubG7QFSZKkNavJTtZ2YH9VzfY+Pwr8AbB7pQurah+wD6DT6RQvuK1BG5IkSadK+l7Z5J6sMH9M+KSzgRcDtyW5G3gF8AVvfpckSaejJiFrCtiRZKz3+R9V1caq2lxVm4GvAZdVVbdpk5IkSaNm4OPCqppJshc4kOQ4cIg+jgolSZJOB41+XVhVk8DkMnOXNqktSZI0ynziuyRJUgsMWZIkSS0wZEmSJLXAkCVJktQCQ5YkSVILDFmSJEktMGRJkiS1wJAlSZLUAkOWJElSC4YaspK8Pck3ktyR5CtJXjTM+pIkSaNi2DtZN1fVT1bVS4H3AzcMub4kSdJIaPTuwiS7gD1AAdNVtXPB9I/1xk/s0cNwy6VN2pAkSVpzBg5ZSbYAE8C2qppNsqE3fjVwLfA0YPsy144D4wBbLzxz0BYkSZLWrFStvNm05IXJNcD5VTWxzPwvAq+vqitPVKfT6VS32x2oB0mSpFMpycGq6vSztsk9WeHEx4GfAd7UoL4kSdLIahKypoAdScYAkmxI8rwF8/8cuKtJc5IkSaNq4HuyqmomyV7gQJLjwCHgO0leC8wBjwAnPCqUJElarxr9urCqJoHJIfUiSZK0bvjEd0mSpBYYsiRJklpgyJIkSWqBIUuSJKkFhixJkqQWGLIkSZJaYMiSJElqgSFLkiSpBUMNWUmuTfLNJNNJppJcNMz6kiRJo2LYO1mHgE5VbQX2A+8fcn1JkqSR0Oi1Okl2AXuAAqaraueC6a8Bv7xikXvvgXe8rUkbkiRJa87AISvJFmAC2FZVs0k2LFpyFfAny1w7DowDbH3mOYO2IEmStGalqga7MLkGOL+qJpaY+2XgHcCrq+oHJ6rT6XSq2+0O1IMkSdKplORgVXX6WdvkuDDMHxMu/vLXMr/DtWLAkiRJWq+a3Pg+BexIMgaQZEOSlwE3AZdV1UPDaFCSJGkUDbyTVVUzSfYCB5IcZ/6XhZuAs4DPJwG4t6ouG0qnkiRJI6TRrwurahKYHFIvkiRJ64ZPfJckSWqBIUuSJKkFhixJkqQWGLIkSZJaYMiSJElqgSFLkiSpBYYsSZKkFhiyJEmSWmDIkiRJasFQQ1aSS5LcnuRYksuHWVuSJGmUDHsn615gN3DzkOtKkiSNlEbvLkyyC9gDFDBdVTt740/0XeTuB+Cq32jShiRJ0pozcMhKsgWYALZV1WySDSdx7TgwDrD1nI2DtiBJkrRmpaoGuzC5Bji/qiaWmPs4cGtV7V+pTqfTqW63O1APkiRJp1KSg1XV6Wdtk3uywvwxoSRJkhZpErKmgB1JxgBO5rhQkiRpvRs4ZFXVDLAXOJDkTuCGJBcnOQJcAdyUZGZIfUqSJI2URr8urKpJYHLR8KYmNSVJktYDn/guSZLUAkOWJElSCwxZkiRJLTBkSZIktcCQJUmS1AJDliRJUgsMWZIkSS0wZEmSJLVgqCEryZlJPpvkb5J8PcnmYdaXJEkaFcPeyboKeKSqngt8CPidIdeXJEkaCY1eq5NkF7AHKGAaOA/4zd70fuAjSVJV1eR7JEmSRs3AISvJFmAC2FZVs0k2AH8B3AdQVceSfAcYA2aXq/PYY48xPT09aBuSJElrUpPjwu3A/qqaBaiqo0CWWPdDu1hJxpN0k3Tn5uYatCBJkrQ2NTkuDD8coI4AFwBHkjwVeAZwdPGFVbUP2AfQ6XRq69atDdqQJElae5rsZE0BO5KMAfSOC78AXNmbvxz4svdjSZKk09HAO1lVNZNkL3AgyXHgEPB24JNJ/ob5Hay3DqdNSZKk0dLo14VVNQlMLhq+oklNSZKk9cAnvkuSJLXAkCVJktQCQ5YkSVILsto//kvyXeDwqjYhSZLUnxdU1dn9LGx04/uQHK6qzmo3IUmStJIk3X7XelwoSZLUAkOWJElSC9ZCyNq32g1IkiT1qe/csuo3vkuSJK1Ha2EnS5Ikad0xZEmSJLXAkCVJktQCQ5YkSVILDFmSJEkt+H8LopM0sc1fuQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 720x360 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Colour graph\n",
    "import matplotlib.pyplot as plt\n",
    "plt.close('all')\n",
    "f, (col) = plt.subplots(1, 1, sharey=True, figsize=(10, 5))\n",
    "col.set_xlim(1, 5)\n",
    "col.set_ylim(-0.5,15)\n",
    "col.set_title('Colours')\n",
    "\n",
    "x_major_ticks = [1, 5]\n",
    "x_major_tick_labels = [\"\", \"\"]\n",
    "col.set_xticks(x_major_ticks)\n",
    "col.set_xticklabels(x_major_tick_labels)\n",
    "\n",
    "y_major_ticks = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14]\n",
    "y_major_tick_labels = [\"c0\", \"c1\", \"c2\", \"c3\", \"c4\", \"c5\", \"c6\", \"c7\", \"c8\", \"c9\", \"c10\", \"c11\", \"c12\", \"c13\", \"c14\"]\n",
    "col.set_yticks(y_major_ticks)\n",
    "col.set_yticklabels(y_major_tick_labels)\n",
    " \n",
    "# c0 points\n",
    "x0 = [1,5]\n",
    "y0 = [0,0]\n",
    "# plotting the c1 points \n",
    "plt.plot(x0, y0, label = \"c0\", color=c0)\n",
    " \n",
    "# c1 points\n",
    "x1 = [1,5]\n",
    "y1 = [1,1]\n",
    "# plotting the c1 points \n",
    "plt.plot(x1, y1, label = \"c1\", color=c1)\n",
    "\n",
    "# c2 points\n",
    "x2 = [1,5]\n",
    "y2 = [2,2]\n",
    "# plotting the c2 points \n",
    "plt.plot(x2, y2, label = \"c2\", color=c2)\n",
    "\n",
    "# c3 points\n",
    "x3 = [1,5]\n",
    "y3 = [3,3]\n",
    "# plotting the c3 points \n",
    "plt.plot(x3, y3, label = \"c3\", color=c3)\n",
    "\n",
    "# c4 points\n",
    "x4 = [1,5]\n",
    "y4 = [4, 4]\n",
    "# plotting the c4 points \n",
    "plt.plot(x4, y4, label = \"c4\", color=c4)\n",
    "\n",
    "# c5 points\n",
    "x5 = [1,5]\n",
    "y5 = [5, 5]\n",
    "# plotting the c5 points \n",
    "plt.plot(x5, y5, label = \"c5\", color=c5)\n",
    "\n",
    "# c6 points\n",
    "x6 = [1,5]\n",
    "y6 = [6, 6]\n",
    "# plotting the c6 points \n",
    "plt.plot(x6, y6, label = \"c6\", color=c6)\n",
    "\n",
    "# c7 points\n",
    "x7 = [1,5]\n",
    "y7 = [7, 7]\n",
    "# plotting the c7 points \n",
    "plt.plot(x7, y7, label = \"c7\", color=c7)\n",
    "\n",
    "# c8 points\n",
    "x8 = [1,5]\n",
    "y8 = [8, 8]\n",
    "# plotting the c8 points \n",
    "plt.plot(x8, y8, label = \"c8\", color=c8)\n",
    "\n",
    "# c9 points\n",
    "x9 = [1,5]\n",
    "y9 = [9, 9]\n",
    "# plotting the c9 points \n",
    "plt.plot(x9, y9, label = \"c9\", color=c9)\n",
    "\n",
    "# c10 points\n",
    "x10 = [1,5]\n",
    "y10 = [10, 10]\n",
    "# plotting the c4 points \n",
    "plt.plot(x10, y10, label = \"c10\", color=c10)\n",
    "\n",
    "# c11 points\n",
    "x11 = [1,5]\n",
    "y11 = [11, 11]\n",
    "# plotting the c11 points \n",
    "plt.plot(x11, y11, label = \"c11\", color=c11)\n",
    "\n",
    "# c12 points\n",
    "x12 = [1,5]\n",
    "y12 = [12, 12]\n",
    "# plotting the c12 points \n",
    "plt.plot(x12, y12, label = \"c12\", color=c12)\n",
    "\n",
    "# c13 points\n",
    "x13 = [1,5]\n",
    "y13 = [13, 13]\n",
    "# plotting the c13 points \n",
    "plt.plot(x13, y13, label = \"c13\", color=c13)\n",
    "\n",
    "# c14 points\n",
    "x14 = [1,5]\n",
    "y14 = [14, 14]\n",
    "# plotting the c14 points \n",
    "plt.plot(x14, y14, label = \"c14\", color=c14)\n",
    "\n",
    "# # show a legend on the plot\n",
    "# plt.legend()\n",
    " \n",
    "# function to show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "missense = c6\n",
    "nonsense = c1\n",
    "synonymous = c4"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Totals for studies"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Number of individuals in each study\n",
    "Jaiswaltotal = 17182\n",
    "Genovesetotal = 12380\n",
    "McKerreltotal = 4219\n",
    "Youngtotal = 20\n",
    "Young2019totalcontrols = 69 #controls only\n",
    "ZinkWGStotal = 11262\n",
    "Acunatotal = 2006\n",
    "Coombstotal = 1591 #chemotherapy and radiotherapy naive individuals\n",
    "Desaitotal = 181 #controls only"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functions for calculating trinucleotide-context specific mutation rates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "# Amino Acid Code:\n",
    "\n",
    "Amino_acid_abbreviations = {'Ala': 'A', 'Arg': 'R', 'Asn': 'N', 'Asp': 'D',\\\n",
    "                           'Cys': 'C', 'Glu': 'E', 'Gln': 'Q', 'Gly': 'G',\\\n",
    "                           'His': 'H', 'Ile': 'I', 'Leu': 'L', 'Lys': 'K',\\\n",
    "                           'Met': 'M', 'Phe': 'F', 'Pro': 'P', 'Ser': 'S',\\\n",
    "                           'Thr': 'T', 'Trp': 'W', 'Tyr': 'Y', 'Val': 'V', '*': '*'}\n",
    "\n",
    "Genetic_Code = {'TTT': 'Phe', 'TTC': 'Phe', 'TTA': 'Leu', 'TTG': 'Leu',\\\n",
    "               'TCT': 'Ser', 'TCC': 'Ser', 'TCA': 'Ser', 'TCG': 'Ser',\\\n",
    "               'TAT': 'Tyr', 'TAC': 'Tyr', 'TAA': '*', 'TAG': '*',\\\n",
    "               'TGT': 'Cys', 'TGC': 'Cys', 'TGA': '*', 'TGG': 'Trp',\\\n",
    "               'CTT': 'Leu', 'CTC': 'Leu', 'CTA': 'Leu', 'CTG': 'Leu',\\\n",
    "               'CCT': 'Pro', 'CCC': 'Pro', 'CCA': 'Pro', 'CCG': 'Pro',\\\n",
    "               'CAT': 'His', 'CAC': 'His', 'CAA': 'Gln', 'CAG': 'Gln',\\\n",
    "               'CGT': 'Arg', 'CGC': 'Arg', 'CGA': 'Arg', 'CGG': 'Arg',\\\n",
    "               'ATT': 'Ile', 'ATC': 'Ile', 'ATA': 'Ile', 'ATG': 'Met',\\\n",
    "               'ACT': 'Thr', 'ACC': 'Thr', 'ACA': 'Thr', 'ACG': 'Thr',\\\n",
    "               'AAT': 'Asn', 'AAC': 'Asn', 'AAA': 'Lys', 'AAG': 'Lys',\\\n",
    "               'AGT': 'Ser', 'AGC': 'Ser', 'AGA': 'Arg', 'AGG': 'Arg',\\\n",
    "               'GTT': 'Val', 'GTC': 'Val', 'GTA': 'Val', 'GTG': 'Val',\\\n",
    "               'GCT': 'Ala', 'GCC': 'Ala', 'GCA': 'Ala', 'GCG': 'Ala',\\\n",
    "               'GAT': 'Asp', 'GAC': 'Asp', 'GAA': 'Glu', 'GAG': 'Glu',\\\n",
    "               'GGT': 'Gly', 'GGC': 'Gly', 'GGA': 'Gly', 'GGG': 'Gly'}\n",
    "\n",
    "base_alternative = {'A': ['T', 'C', 'G'], 'T': ['A', 'C', 'G'], 'G':['A', 'C', 'T'], 'C':['A', 'G', 'T']}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "Possible_AA_Changes = {}\n",
    "Possible_Base_Changes = {}\n",
    "Possible_Codon_Changes = {}\n",
    "for k, v in Genetic_Code.items():\n",
    "    codon = k\n",
    "    base_1 = str(codon[0])\n",
    "    base_2 = str(codon[1])\n",
    "    base_3 = str(codon[2])\n",
    "    new_codons = []\n",
    "    amino_acid_changes = []\n",
    "    base_changes = []\n",
    "    if base_1 in base_alternative.keys():\n",
    "        alternate_base = base_alternative[base_1]\n",
    "        for i in alternate_base:\n",
    "            base_changes.append(i)\n",
    "            new_codon = i+base_2+base_3\n",
    "            new_codons.append(new_codon)\n",
    "    if base_2 in base_alternative.keys():\n",
    "        alternate_base = base_alternative[base_2]\n",
    "        for i in alternate_base:\n",
    "            base_changes.append(i)\n",
    "            new_codon = base_1+i+base_3\n",
    "            new_codons.append(new_codon)\n",
    "    if base_3 in base_alternative.keys():\n",
    "        alternate_base = base_alternative[base_3]\n",
    "        for i in alternate_base:\n",
    "            base_changes.append(i)\n",
    "            new_codon = base_1+base_2+i\n",
    "            new_codons.append(new_codon)\n",
    "    Possible_Codon_Changes[k]=(new_codons[0],new_codons[1], new_codons[2], new_codons[3], new_codons[4],\\\n",
    "                               new_codons[5], new_codons[6], new_codons[7], new_codons[8]) #i.e. {'TTT': ['ATT', 'CTT', 'GTT'...etc]}\n",
    "    Possible_Base_Changes[k]=(base_changes[0], base_changes[1], base_changes[2], base_changes[3], base_changes[4],\\\n",
    "                              base_changes[5], base_changes[6], base_changes[7], base_changes[8]) #i.e. {'TTT': ['A', 'C', 'G'... etc]}\n",
    "    \n",
    "Possible_Amino_Acid_Changes = {}\n",
    "for k, v in Possible_Codon_Changes.items():\n",
    "    codon = k\n",
    "    codon_changes = v\n",
    "    amino_acid_changes = []\n",
    "    for i in codon_changes:\n",
    "        amino_acid_change = Genetic_Code[i]\n",
    "        amino_acid_changes.append(amino_acid_change)\n",
    "    Possible_Amino_Acid_Changes[k]=(amino_acid_changes[0], amino_acid_changes[1], amino_acid_changes[2],\\\n",
    "                                    amino_acid_changes[3], amino_acid_changes[4], amino_acid_changes[5],\\\n",
    "                                    amino_acid_changes[6], amino_acid_changes[7], amino_acid_changes[8]) #i.e. {'TTT': ['Ile', 'Leu', 'Val'...etc]}\n",
    "    \n",
    "Possible_Amino_Acid_Changes_letters = {}\n",
    "for k, v in Possible_Amino_Acid_Changes.items():\n",
    "    codon = k\n",
    "    codon_changes = v\n",
    "    amino_acid_changes_letters = []\n",
    "    for i in codon_changes:\n",
    "        amino_acid_change_letter = Amino_acid_abbreviations[i]\n",
    "        amino_acid_changes_letters.append(amino_acid_change_letter)\n",
    "    Possible_Amino_Acid_Changes_letters[k]=(amino_acid_changes_letters[0], amino_acid_changes_letters[1],\\\n",
    "                                            amino_acid_changes_letters[2], amino_acid_changes_letters[3],\\\n",
    "                                            amino_acid_changes_letters[4], amino_acid_changes_letters[5],\\\n",
    "                                            amino_acid_changes_letters[6], amino_acid_changes_letters[7],\\\n",
    "                                            amino_acid_changes_letters[8]) #i.e. {'TTT': ['I', 'L', 'V'...etc]}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'TTT': ('ATT', 'CTT', 'GTT', 'TAT', 'TCT', 'TGT', 'TTA', 'TTC', 'TTG'),\n",
       " 'TTC': ('ATC', 'CTC', 'GTC', 'TAC', 'TCC', 'TGC', 'TTA', 'TTG', 'TTT'),\n",
       " 'TTA': ('ATA', 'CTA', 'GTA', 'TAA', 'TCA', 'TGA', 'TTT', 'TTC', 'TTG'),\n",
       " 'TTG': ('ATG', 'CTG', 'GTG', 'TAG', 'TCG', 'TGG', 'TTA', 'TTC', 'TTT'),\n",
       " 'TCT': ('ACT', 'CCT', 'GCT', 'TAT', 'TGT', 'TTT', 'TCA', 'TCC', 'TCG'),\n",
       " 'TCC': ('ACC', 'CCC', 'GCC', 'TAC', 'TGC', 'TTC', 'TCA', 'TCG', 'TCT'),\n",
       " 'TCA': ('ACA', 'CCA', 'GCA', 'TAA', 'TGA', 'TTA', 'TCT', 'TCC', 'TCG'),\n",
       " 'TCG': ('ACG', 'CCG', 'GCG', 'TAG', 'TGG', 'TTG', 'TCA', 'TCC', 'TCT'),\n",
       " 'TAT': ('AAT', 'CAT', 'GAT', 'TTT', 'TCT', 'TGT', 'TAA', 'TAC', 'TAG'),\n",
       " 'TAC': ('AAC', 'CAC', 'GAC', 'TTC', 'TCC', 'TGC', 'TAA', 'TAG', 'TAT'),\n",
       " 'TAA': ('AAA', 'CAA', 'GAA', 'TTA', 'TCA', 'TGA', 'TAT', 'TAC', 'TAG'),\n",
       " 'TAG': ('AAG', 'CAG', 'GAG', 'TTG', 'TCG', 'TGG', 'TAA', 'TAC', 'TAT'),\n",
       " 'TGT': ('AGT', 'CGT', 'GGT', 'TAT', 'TCT', 'TTT', 'TGA', 'TGC', 'TGG'),\n",
       " 'TGC': ('AGC', 'CGC', 'GGC', 'TAC', 'TCC', 'TTC', 'TGA', 'TGG', 'TGT'),\n",
       " 'TGA': ('AGA', 'CGA', 'GGA', 'TAA', 'TCA', 'TTA', 'TGT', 'TGC', 'TGG'),\n",
       " 'TGG': ('AGG', 'CGG', 'GGG', 'TAG', 'TCG', 'TTG', 'TGA', 'TGC', 'TGT'),\n",
       " 'CTT': ('ATT', 'GTT', 'TTT', 'CAT', 'CCT', 'CGT', 'CTA', 'CTC', 'CTG'),\n",
       " 'CTC': ('ATC', 'GTC', 'TTC', 'CAC', 'CCC', 'CGC', 'CTA', 'CTG', 'CTT'),\n",
       " 'CTA': ('ATA', 'GTA', 'TTA', 'CAA', 'CCA', 'CGA', 'CTT', 'CTC', 'CTG'),\n",
       " 'CTG': ('ATG', 'GTG', 'TTG', 'CAG', 'CCG', 'CGG', 'CTA', 'CTC', 'CTT'),\n",
       " 'CCT': ('ACT', 'GCT', 'TCT', 'CAT', 'CGT', 'CTT', 'CCA', 'CCC', 'CCG'),\n",
       " 'CCC': ('ACC', 'GCC', 'TCC', 'CAC', 'CGC', 'CTC', 'CCA', 'CCG', 'CCT'),\n",
       " 'CCA': ('ACA', 'GCA', 'TCA', 'CAA', 'CGA', 'CTA', 'CCT', 'CCC', 'CCG'),\n",
       " 'CCG': ('ACG', 'GCG', 'TCG', 'CAG', 'CGG', 'CTG', 'CCA', 'CCC', 'CCT'),\n",
       " 'CAT': ('AAT', 'GAT', 'TAT', 'CTT', 'CCT', 'CGT', 'CAA', 'CAC', 'CAG'),\n",
       " 'CAC': ('AAC', 'GAC', 'TAC', 'CTC', 'CCC', 'CGC', 'CAA', 'CAG', 'CAT'),\n",
       " 'CAA': ('AAA', 'GAA', 'TAA', 'CTA', 'CCA', 'CGA', 'CAT', 'CAC', 'CAG'),\n",
       " 'CAG': ('AAG', 'GAG', 'TAG', 'CTG', 'CCG', 'CGG', 'CAA', 'CAC', 'CAT'),\n",
       " 'CGT': ('AGT', 'GGT', 'TGT', 'CAT', 'CCT', 'CTT', 'CGA', 'CGC', 'CGG'),\n",
       " 'CGC': ('AGC', 'GGC', 'TGC', 'CAC', 'CCC', 'CTC', 'CGA', 'CGG', 'CGT'),\n",
       " 'CGA': ('AGA', 'GGA', 'TGA', 'CAA', 'CCA', 'CTA', 'CGT', 'CGC', 'CGG'),\n",
       " 'CGG': ('AGG', 'GGG', 'TGG', 'CAG', 'CCG', 'CTG', 'CGA', 'CGC', 'CGT'),\n",
       " 'ATT': ('TTT', 'CTT', 'GTT', 'AAT', 'ACT', 'AGT', 'ATA', 'ATC', 'ATG'),\n",
       " 'ATC': ('TTC', 'CTC', 'GTC', 'AAC', 'ACC', 'AGC', 'ATA', 'ATG', 'ATT'),\n",
       " 'ATA': ('TTA', 'CTA', 'GTA', 'AAA', 'ACA', 'AGA', 'ATT', 'ATC', 'ATG'),\n",
       " 'ATG': ('TTG', 'CTG', 'GTG', 'AAG', 'ACG', 'AGG', 'ATA', 'ATC', 'ATT'),\n",
       " 'ACT': ('TCT', 'CCT', 'GCT', 'AAT', 'AGT', 'ATT', 'ACA', 'ACC', 'ACG'),\n",
       " 'ACC': ('TCC', 'CCC', 'GCC', 'AAC', 'AGC', 'ATC', 'ACA', 'ACG', 'ACT'),\n",
       " 'ACA': ('TCA', 'CCA', 'GCA', 'AAA', 'AGA', 'ATA', 'ACT', 'ACC', 'ACG'),\n",
       " 'ACG': ('TCG', 'CCG', 'GCG', 'AAG', 'AGG', 'ATG', 'ACA', 'ACC', 'ACT'),\n",
       " 'AAT': ('TAT', 'CAT', 'GAT', 'ATT', 'ACT', 'AGT', 'AAA', 'AAC', 'AAG'),\n",
       " 'AAC': ('TAC', 'CAC', 'GAC', 'ATC', 'ACC', 'AGC', 'AAA', 'AAG', 'AAT'),\n",
       " 'AAA': ('TAA', 'CAA', 'GAA', 'ATA', 'ACA', 'AGA', 'AAT', 'AAC', 'AAG'),\n",
       " 'AAG': ('TAG', 'CAG', 'GAG', 'ATG', 'ACG', 'AGG', 'AAA', 'AAC', 'AAT'),\n",
       " 'AGT': ('TGT', 'CGT', 'GGT', 'AAT', 'ACT', 'ATT', 'AGA', 'AGC', 'AGG'),\n",
       " 'AGC': ('TGC', 'CGC', 'GGC', 'AAC', 'ACC', 'ATC', 'AGA', 'AGG', 'AGT'),\n",
       " 'AGA': ('TGA', 'CGA', 'GGA', 'AAA', 'ACA', 'ATA', 'AGT', 'AGC', 'AGG'),\n",
       " 'AGG': ('TGG', 'CGG', 'GGG', 'AAG', 'ACG', 'ATG', 'AGA', 'AGC', 'AGT'),\n",
       " 'GTT': ('ATT', 'CTT', 'TTT', 'GAT', 'GCT', 'GGT', 'GTA', 'GTC', 'GTG'),\n",
       " 'GTC': ('ATC', 'CTC', 'TTC', 'GAC', 'GCC', 'GGC', 'GTA', 'GTG', 'GTT'),\n",
       " 'GTA': ('ATA', 'CTA', 'TTA', 'GAA', 'GCA', 'GGA', 'GTT', 'GTC', 'GTG'),\n",
       " 'GTG': ('ATG', 'CTG', 'TTG', 'GAG', 'GCG', 'GGG', 'GTA', 'GTC', 'GTT'),\n",
       " 'GCT': ('ACT', 'CCT', 'TCT', 'GAT', 'GGT', 'GTT', 'GCA', 'GCC', 'GCG'),\n",
       " 'GCC': ('ACC', 'CCC', 'TCC', 'GAC', 'GGC', 'GTC', 'GCA', 'GCG', 'GCT'),\n",
       " 'GCA': ('ACA', 'CCA', 'TCA', 'GAA', 'GGA', 'GTA', 'GCT', 'GCC', 'GCG'),\n",
       " 'GCG': ('ACG', 'CCG', 'TCG', 'GAG', 'GGG', 'GTG', 'GCA', 'GCC', 'GCT'),\n",
       " 'GAT': ('AAT', 'CAT', 'TAT', 'GTT', 'GCT', 'GGT', 'GAA', 'GAC', 'GAG'),\n",
       " 'GAC': ('AAC', 'CAC', 'TAC', 'GTC', 'GCC', 'GGC', 'GAA', 'GAG', 'GAT'),\n",
       " 'GAA': ('AAA', 'CAA', 'TAA', 'GTA', 'GCA', 'GGA', 'GAT', 'GAC', 'GAG'),\n",
       " 'GAG': ('AAG', 'CAG', 'TAG', 'GTG', 'GCG', 'GGG', 'GAA', 'GAC', 'GAT'),\n",
       " 'GGT': ('AGT', 'CGT', 'TGT', 'GAT', 'GCT', 'GTT', 'GGA', 'GGC', 'GGG'),\n",
       " 'GGC': ('AGC', 'CGC', 'TGC', 'GAC', 'GCC', 'GTC', 'GGA', 'GGG', 'GGT'),\n",
       " 'GGA': ('AGA', 'CGA', 'TGA', 'GAA', 'GCA', 'GTA', 'GGT', 'GGC', 'GGG'),\n",
       " 'GGG': ('AGG', 'CGG', 'TGG', 'GAG', 'GCG', 'GTG', 'GGA', 'GGC', 'GGT')}"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Possible_Codon_Changes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'TTT': ('A', 'C', 'G', 'A', 'C', 'G', 'A', 'C', 'G'),\n",
       " 'TTC': ('A', 'C', 'G', 'A', 'C', 'G', 'A', 'G', 'T'),\n",
       " 'TTA': ('A', 'C', 'G', 'A', 'C', 'G', 'T', 'C', 'G'),\n",
       " 'TTG': ('A', 'C', 'G', 'A', 'C', 'G', 'A', 'C', 'T'),\n",
       " 'TCT': ('A', 'C', 'G', 'A', 'G', 'T', 'A', 'C', 'G'),\n",
       " 'TCC': ('A', 'C', 'G', 'A', 'G', 'T', 'A', 'G', 'T'),\n",
       " 'TCA': ('A', 'C', 'G', 'A', 'G', 'T', 'T', 'C', 'G'),\n",
       " 'TCG': ('A', 'C', 'G', 'A', 'G', 'T', 'A', 'C', 'T'),\n",
       " 'TAT': ('A', 'C', 'G', 'T', 'C', 'G', 'A', 'C', 'G'),\n",
       " 'TAC': ('A', 'C', 'G', 'T', 'C', 'G', 'A', 'G', 'T'),\n",
       " 'TAA': ('A', 'C', 'G', 'T', 'C', 'G', 'T', 'C', 'G'),\n",
       " 'TAG': ('A', 'C', 'G', 'T', 'C', 'G', 'A', 'C', 'T'),\n",
       " 'TGT': ('A', 'C', 'G', 'A', 'C', 'T', 'A', 'C', 'G'),\n",
       " 'TGC': ('A', 'C', 'G', 'A', 'C', 'T', 'A', 'G', 'T'),\n",
       " 'TGA': ('A', 'C', 'G', 'A', 'C', 'T', 'T', 'C', 'G'),\n",
       " 'TGG': ('A', 'C', 'G', 'A', 'C', 'T', 'A', 'C', 'T'),\n",
       " 'CTT': ('A', 'G', 'T', 'A', 'C', 'G', 'A', 'C', 'G'),\n",
       " 'CTC': ('A', 'G', 'T', 'A', 'C', 'G', 'A', 'G', 'T'),\n",
       " 'CTA': ('A', 'G', 'T', 'A', 'C', 'G', 'T', 'C', 'G'),\n",
       " 'CTG': ('A', 'G', 'T', 'A', 'C', 'G', 'A', 'C', 'T'),\n",
       " 'CCT': ('A', 'G', 'T', 'A', 'G', 'T', 'A', 'C', 'G'),\n",
       " 'CCC': ('A', 'G', 'T', 'A', 'G', 'T', 'A', 'G', 'T'),\n",
       " 'CCA': ('A', 'G', 'T', 'A', 'G', 'T', 'T', 'C', 'G'),\n",
       " 'CCG': ('A', 'G', 'T', 'A', 'G', 'T', 'A', 'C', 'T'),\n",
       " 'CAT': ('A', 'G', 'T', 'T', 'C', 'G', 'A', 'C', 'G'),\n",
       " 'CAC': ('A', 'G', 'T', 'T', 'C', 'G', 'A', 'G', 'T'),\n",
       " 'CAA': ('A', 'G', 'T', 'T', 'C', 'G', 'T', 'C', 'G'),\n",
       " 'CAG': ('A', 'G', 'T', 'T', 'C', 'G', 'A', 'C', 'T'),\n",
       " 'CGT': ('A', 'G', 'T', 'A', 'C', 'T', 'A', 'C', 'G'),\n",
       " 'CGC': ('A', 'G', 'T', 'A', 'C', 'T', 'A', 'G', 'T'),\n",
       " 'CGA': ('A', 'G', 'T', 'A', 'C', 'T', 'T', 'C', 'G'),\n",
       " 'CGG': ('A', 'G', 'T', 'A', 'C', 'T', 'A', 'C', 'T'),\n",
       " 'ATT': ('T', 'C', 'G', 'A', 'C', 'G', 'A', 'C', 'G'),\n",
       " 'ATC': ('T', 'C', 'G', 'A', 'C', 'G', 'A', 'G', 'T'),\n",
       " 'ATA': ('T', 'C', 'G', 'A', 'C', 'G', 'T', 'C', 'G'),\n",
       " 'ATG': ('T', 'C', 'G', 'A', 'C', 'G', 'A', 'C', 'T'),\n",
       " 'ACT': ('T', 'C', 'G', 'A', 'G', 'T', 'A', 'C', 'G'),\n",
       " 'ACC': ('T', 'C', 'G', 'A', 'G', 'T', 'A', 'G', 'T'),\n",
       " 'ACA': ('T', 'C', 'G', 'A', 'G', 'T', 'T', 'C', 'G'),\n",
       " 'ACG': ('T', 'C', 'G', 'A', 'G', 'T', 'A', 'C', 'T'),\n",
       " 'AAT': ('T', 'C', 'G', 'T', 'C', 'G', 'A', 'C', 'G'),\n",
       " 'AAC': ('T', 'C', 'G', 'T', 'C', 'G', 'A', 'G', 'T'),\n",
       " 'AAA': ('T', 'C', 'G', 'T', 'C', 'G', 'T', 'C', 'G'),\n",
       " 'AAG': ('T', 'C', 'G', 'T', 'C', 'G', 'A', 'C', 'T'),\n",
       " 'AGT': ('T', 'C', 'G', 'A', 'C', 'T', 'A', 'C', 'G'),\n",
       " 'AGC': ('T', 'C', 'G', 'A', 'C', 'T', 'A', 'G', 'T'),\n",
       " 'AGA': ('T', 'C', 'G', 'A', 'C', 'T', 'T', 'C', 'G'),\n",
       " 'AGG': ('T', 'C', 'G', 'A', 'C', 'T', 'A', 'C', 'T'),\n",
       " 'GTT': ('A', 'C', 'T', 'A', 'C', 'G', 'A', 'C', 'G'),\n",
       " 'GTC': ('A', 'C', 'T', 'A', 'C', 'G', 'A', 'G', 'T'),\n",
       " 'GTA': ('A', 'C', 'T', 'A', 'C', 'G', 'T', 'C', 'G'),\n",
       " 'GTG': ('A', 'C', 'T', 'A', 'C', 'G', 'A', 'C', 'T'),\n",
       " 'GCT': ('A', 'C', 'T', 'A', 'G', 'T', 'A', 'C', 'G'),\n",
       " 'GCC': ('A', 'C', 'T', 'A', 'G', 'T', 'A', 'G', 'T'),\n",
       " 'GCA': ('A', 'C', 'T', 'A', 'G', 'T', 'T', 'C', 'G'),\n",
       " 'GCG': ('A', 'C', 'T', 'A', 'G', 'T', 'A', 'C', 'T'),\n",
       " 'GAT': ('A', 'C', 'T', 'T', 'C', 'G', 'A', 'C', 'G'),\n",
       " 'GAC': ('A', 'C', 'T', 'T', 'C', 'G', 'A', 'G', 'T'),\n",
       " 'GAA': ('A', 'C', 'T', 'T', 'C', 'G', 'T', 'C', 'G'),\n",
       " 'GAG': ('A', 'C', 'T', 'T', 'C', 'G', 'A', 'C', 'T'),\n",
       " 'GGT': ('A', 'C', 'T', 'A', 'C', 'T', 'A', 'C', 'G'),\n",
       " 'GGC': ('A', 'C', 'T', 'A', 'C', 'T', 'A', 'G', 'T'),\n",
       " 'GGA': ('A', 'C', 'T', 'A', 'C', 'T', 'T', 'C', 'G'),\n",
       " 'GGG': ('A', 'C', 'T', 'A', 'C', 'T', 'A', 'C', 'T')}"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Possible_Base_Changes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'TTT': ('Ile', 'Leu', 'Val', 'Tyr', 'Ser', 'Cys', 'Leu', 'Phe', 'Leu'),\n",
       " 'TTC': ('Ile', 'Leu', 'Val', 'Tyr', 'Ser', 'Cys', 'Leu', 'Leu', 'Phe'),\n",
       " 'TTA': ('Ile', 'Leu', 'Val', '*', 'Ser', '*', 'Phe', 'Phe', 'Leu'),\n",
       " 'TTG': ('Met', 'Leu', 'Val', '*', 'Ser', 'Trp', 'Leu', 'Phe', 'Phe'),\n",
       " 'TCT': ('Thr', 'Pro', 'Ala', 'Tyr', 'Cys', 'Phe', 'Ser', 'Ser', 'Ser'),\n",
       " 'TCC': ('Thr', 'Pro', 'Ala', 'Tyr', 'Cys', 'Phe', 'Ser', 'Ser', 'Ser'),\n",
       " 'TCA': ('Thr', 'Pro', 'Ala', '*', '*', 'Leu', 'Ser', 'Ser', 'Ser'),\n",
       " 'TCG': ('Thr', 'Pro', 'Ala', '*', 'Trp', 'Leu', 'Ser', 'Ser', 'Ser'),\n",
       " 'TAT': ('Asn', 'His', 'Asp', 'Phe', 'Ser', 'Cys', '*', 'Tyr', '*'),\n",
       " 'TAC': ('Asn', 'His', 'Asp', 'Phe', 'Ser', 'Cys', '*', '*', 'Tyr'),\n",
       " 'TAA': ('Lys', 'Gln', 'Glu', 'Leu', 'Ser', '*', 'Tyr', 'Tyr', '*'),\n",
       " 'TAG': ('Lys', 'Gln', 'Glu', 'Leu', 'Ser', 'Trp', '*', 'Tyr', 'Tyr'),\n",
       " 'TGT': ('Ser', 'Arg', 'Gly', 'Tyr', 'Ser', 'Phe', '*', 'Cys', 'Trp'),\n",
       " 'TGC': ('Ser', 'Arg', 'Gly', 'Tyr', 'Ser', 'Phe', '*', 'Trp', 'Cys'),\n",
       " 'TGA': ('Arg', 'Arg', 'Gly', '*', 'Ser', 'Leu', 'Cys', 'Cys', 'Trp'),\n",
       " 'TGG': ('Arg', 'Arg', 'Gly', '*', 'Ser', 'Leu', '*', 'Cys', 'Cys'),\n",
       " 'CTT': ('Ile', 'Val', 'Phe', 'His', 'Pro', 'Arg', 'Leu', 'Leu', 'Leu'),\n",
       " 'CTC': ('Ile', 'Val', 'Phe', 'His', 'Pro', 'Arg', 'Leu', 'Leu', 'Leu'),\n",
       " 'CTA': ('Ile', 'Val', 'Leu', 'Gln', 'Pro', 'Arg', 'Leu', 'Leu', 'Leu'),\n",
       " 'CTG': ('Met', 'Val', 'Leu', 'Gln', 'Pro', 'Arg', 'Leu', 'Leu', 'Leu'),\n",
       " 'CCT': ('Thr', 'Ala', 'Ser', 'His', 'Arg', 'Leu', 'Pro', 'Pro', 'Pro'),\n",
       " 'CCC': ('Thr', 'Ala', 'Ser', 'His', 'Arg', 'Leu', 'Pro', 'Pro', 'Pro'),\n",
       " 'CCA': ('Thr', 'Ala', 'Ser', 'Gln', 'Arg', 'Leu', 'Pro', 'Pro', 'Pro'),\n",
       " 'CCG': ('Thr', 'Ala', 'Ser', 'Gln', 'Arg', 'Leu', 'Pro', 'Pro', 'Pro'),\n",
       " 'CAT': ('Asn', 'Asp', 'Tyr', 'Leu', 'Pro', 'Arg', 'Gln', 'His', 'Gln'),\n",
       " 'CAC': ('Asn', 'Asp', 'Tyr', 'Leu', 'Pro', 'Arg', 'Gln', 'Gln', 'His'),\n",
       " 'CAA': ('Lys', 'Glu', '*', 'Leu', 'Pro', 'Arg', 'His', 'His', 'Gln'),\n",
       " 'CAG': ('Lys', 'Glu', '*', 'Leu', 'Pro', 'Arg', 'Gln', 'His', 'His'),\n",
       " 'CGT': ('Ser', 'Gly', 'Cys', 'His', 'Pro', 'Leu', 'Arg', 'Arg', 'Arg'),\n",
       " 'CGC': ('Ser', 'Gly', 'Cys', 'His', 'Pro', 'Leu', 'Arg', 'Arg', 'Arg'),\n",
       " 'CGA': ('Arg', 'Gly', '*', 'Gln', 'Pro', 'Leu', 'Arg', 'Arg', 'Arg'),\n",
       " 'CGG': ('Arg', 'Gly', 'Trp', 'Gln', 'Pro', 'Leu', 'Arg', 'Arg', 'Arg'),\n",
       " 'ATT': ('Phe', 'Leu', 'Val', 'Asn', 'Thr', 'Ser', 'Ile', 'Ile', 'Met'),\n",
       " 'ATC': ('Phe', 'Leu', 'Val', 'Asn', 'Thr', 'Ser', 'Ile', 'Met', 'Ile'),\n",
       " 'ATA': ('Leu', 'Leu', 'Val', 'Lys', 'Thr', 'Arg', 'Ile', 'Ile', 'Met'),\n",
       " 'ATG': ('Leu', 'Leu', 'Val', 'Lys', 'Thr', 'Arg', 'Ile', 'Ile', 'Ile'),\n",
       " 'ACT': ('Ser', 'Pro', 'Ala', 'Asn', 'Ser', 'Ile', 'Thr', 'Thr', 'Thr'),\n",
       " 'ACC': ('Ser', 'Pro', 'Ala', 'Asn', 'Ser', 'Ile', 'Thr', 'Thr', 'Thr'),\n",
       " 'ACA': ('Ser', 'Pro', 'Ala', 'Lys', 'Arg', 'Ile', 'Thr', 'Thr', 'Thr'),\n",
       " 'ACG': ('Ser', 'Pro', 'Ala', 'Lys', 'Arg', 'Met', 'Thr', 'Thr', 'Thr'),\n",
       " 'AAT': ('Tyr', 'His', 'Asp', 'Ile', 'Thr', 'Ser', 'Lys', 'Asn', 'Lys'),\n",
       " 'AAC': ('Tyr', 'His', 'Asp', 'Ile', 'Thr', 'Ser', 'Lys', 'Lys', 'Asn'),\n",
       " 'AAA': ('*', 'Gln', 'Glu', 'Ile', 'Thr', 'Arg', 'Asn', 'Asn', 'Lys'),\n",
       " 'AAG': ('*', 'Gln', 'Glu', 'Met', 'Thr', 'Arg', 'Lys', 'Asn', 'Asn'),\n",
       " 'AGT': ('Cys', 'Arg', 'Gly', 'Asn', 'Thr', 'Ile', 'Arg', 'Ser', 'Arg'),\n",
       " 'AGC': ('Cys', 'Arg', 'Gly', 'Asn', 'Thr', 'Ile', 'Arg', 'Arg', 'Ser'),\n",
       " 'AGA': ('*', 'Arg', 'Gly', 'Lys', 'Thr', 'Ile', 'Ser', 'Ser', 'Arg'),\n",
       " 'AGG': ('Trp', 'Arg', 'Gly', 'Lys', 'Thr', 'Met', 'Arg', 'Ser', 'Ser'),\n",
       " 'GTT': ('Ile', 'Leu', 'Phe', 'Asp', 'Ala', 'Gly', 'Val', 'Val', 'Val'),\n",
       " 'GTC': ('Ile', 'Leu', 'Phe', 'Asp', 'Ala', 'Gly', 'Val', 'Val', 'Val'),\n",
       " 'GTA': ('Ile', 'Leu', 'Leu', 'Glu', 'Ala', 'Gly', 'Val', 'Val', 'Val'),\n",
       " 'GTG': ('Met', 'Leu', 'Leu', 'Glu', 'Ala', 'Gly', 'Val', 'Val', 'Val'),\n",
       " 'GCT': ('Thr', 'Pro', 'Ser', 'Asp', 'Gly', 'Val', 'Ala', 'Ala', 'Ala'),\n",
       " 'GCC': ('Thr', 'Pro', 'Ser', 'Asp', 'Gly', 'Val', 'Ala', 'Ala', 'Ala'),\n",
       " 'GCA': ('Thr', 'Pro', 'Ser', 'Glu', 'Gly', 'Val', 'Ala', 'Ala', 'Ala'),\n",
       " 'GCG': ('Thr', 'Pro', 'Ser', 'Glu', 'Gly', 'Val', 'Ala', 'Ala', 'Ala'),\n",
       " 'GAT': ('Asn', 'His', 'Tyr', 'Val', 'Ala', 'Gly', 'Glu', 'Asp', 'Glu'),\n",
       " 'GAC': ('Asn', 'His', 'Tyr', 'Val', 'Ala', 'Gly', 'Glu', 'Glu', 'Asp'),\n",
       " 'GAA': ('Lys', 'Gln', '*', 'Val', 'Ala', 'Gly', 'Asp', 'Asp', 'Glu'),\n",
       " 'GAG': ('Lys', 'Gln', '*', 'Val', 'Ala', 'Gly', 'Glu', 'Asp', 'Asp'),\n",
       " 'GGT': ('Ser', 'Arg', 'Cys', 'Asp', 'Ala', 'Val', 'Gly', 'Gly', 'Gly'),\n",
       " 'GGC': ('Ser', 'Arg', 'Cys', 'Asp', 'Ala', 'Val', 'Gly', 'Gly', 'Gly'),\n",
       " 'GGA': ('Arg', 'Arg', '*', 'Glu', 'Ala', 'Val', 'Gly', 'Gly', 'Gly'),\n",
       " 'GGG': ('Arg', 'Arg', 'Trp', 'Glu', 'Ala', 'Val', 'Gly', 'Gly', 'Gly')}"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Possible_Amino_Acid_Changes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'TTT': ('I', 'L', 'V', 'Y', 'S', 'C', 'L', 'F', 'L'),\n",
       " 'TTC': ('I', 'L', 'V', 'Y', 'S', 'C', 'L', 'L', 'F'),\n",
       " 'TTA': ('I', 'L', 'V', '*', 'S', '*', 'F', 'F', 'L'),\n",
       " 'TTG': ('M', 'L', 'V', '*', 'S', 'W', 'L', 'F', 'F'),\n",
       " 'TCT': ('T', 'P', 'A', 'Y', 'C', 'F', 'S', 'S', 'S'),\n",
       " 'TCC': ('T', 'P', 'A', 'Y', 'C', 'F', 'S', 'S', 'S'),\n",
       " 'TCA': ('T', 'P', 'A', '*', '*', 'L', 'S', 'S', 'S'),\n",
       " 'TCG': ('T', 'P', 'A', '*', 'W', 'L', 'S', 'S', 'S'),\n",
       " 'TAT': ('N', 'H', 'D', 'F', 'S', 'C', '*', 'Y', '*'),\n",
       " 'TAC': ('N', 'H', 'D', 'F', 'S', 'C', '*', '*', 'Y'),\n",
       " 'TAA': ('K', 'Q', 'E', 'L', 'S', '*', 'Y', 'Y', '*'),\n",
       " 'TAG': ('K', 'Q', 'E', 'L', 'S', 'W', '*', 'Y', 'Y'),\n",
       " 'TGT': ('S', 'R', 'G', 'Y', 'S', 'F', '*', 'C', 'W'),\n",
       " 'TGC': ('S', 'R', 'G', 'Y', 'S', 'F', '*', 'W', 'C'),\n",
       " 'TGA': ('R', 'R', 'G', '*', 'S', 'L', 'C', 'C', 'W'),\n",
       " 'TGG': ('R', 'R', 'G', '*', 'S', 'L', '*', 'C', 'C'),\n",
       " 'CTT': ('I', 'V', 'F', 'H', 'P', 'R', 'L', 'L', 'L'),\n",
       " 'CTC': ('I', 'V', 'F', 'H', 'P', 'R', 'L', 'L', 'L'),\n",
       " 'CTA': ('I', 'V', 'L', 'Q', 'P', 'R', 'L', 'L', 'L'),\n",
       " 'CTG': ('M', 'V', 'L', 'Q', 'P', 'R', 'L', 'L', 'L'),\n",
       " 'CCT': ('T', 'A', 'S', 'H', 'R', 'L', 'P', 'P', 'P'),\n",
       " 'CCC': ('T', 'A', 'S', 'H', 'R', 'L', 'P', 'P', 'P'),\n",
       " 'CCA': ('T', 'A', 'S', 'Q', 'R', 'L', 'P', 'P', 'P'),\n",
       " 'CCG': ('T', 'A', 'S', 'Q', 'R', 'L', 'P', 'P', 'P'),\n",
       " 'CAT': ('N', 'D', 'Y', 'L', 'P', 'R', 'Q', 'H', 'Q'),\n",
       " 'CAC': ('N', 'D', 'Y', 'L', 'P', 'R', 'Q', 'Q', 'H'),\n",
       " 'CAA': ('K', 'E', '*', 'L', 'P', 'R', 'H', 'H', 'Q'),\n",
       " 'CAG': ('K', 'E', '*', 'L', 'P', 'R', 'Q', 'H', 'H'),\n",
       " 'CGT': ('S', 'G', 'C', 'H', 'P', 'L', 'R', 'R', 'R'),\n",
       " 'CGC': ('S', 'G', 'C', 'H', 'P', 'L', 'R', 'R', 'R'),\n",
       " 'CGA': ('R', 'G', '*', 'Q', 'P', 'L', 'R', 'R', 'R'),\n",
       " 'CGG': ('R', 'G', 'W', 'Q', 'P', 'L', 'R', 'R', 'R'),\n",
       " 'ATT': ('F', 'L', 'V', 'N', 'T', 'S', 'I', 'I', 'M'),\n",
       " 'ATC': ('F', 'L', 'V', 'N', 'T', 'S', 'I', 'M', 'I'),\n",
       " 'ATA': ('L', 'L', 'V', 'K', 'T', 'R', 'I', 'I', 'M'),\n",
       " 'ATG': ('L', 'L', 'V', 'K', 'T', 'R', 'I', 'I', 'I'),\n",
       " 'ACT': ('S', 'P', 'A', 'N', 'S', 'I', 'T', 'T', 'T'),\n",
       " 'ACC': ('S', 'P', 'A', 'N', 'S', 'I', 'T', 'T', 'T'),\n",
       " 'ACA': ('S', 'P', 'A', 'K', 'R', 'I', 'T', 'T', 'T'),\n",
       " 'ACG': ('S', 'P', 'A', 'K', 'R', 'M', 'T', 'T', 'T'),\n",
       " 'AAT': ('Y', 'H', 'D', 'I', 'T', 'S', 'K', 'N', 'K'),\n",
       " 'AAC': ('Y', 'H', 'D', 'I', 'T', 'S', 'K', 'K', 'N'),\n",
       " 'AAA': ('*', 'Q', 'E', 'I', 'T', 'R', 'N', 'N', 'K'),\n",
       " 'AAG': ('*', 'Q', 'E', 'M', 'T', 'R', 'K', 'N', 'N'),\n",
       " 'AGT': ('C', 'R', 'G', 'N', 'T', 'I', 'R', 'S', 'R'),\n",
       " 'AGC': ('C', 'R', 'G', 'N', 'T', 'I', 'R', 'R', 'S'),\n",
       " 'AGA': ('*', 'R', 'G', 'K', 'T', 'I', 'S', 'S', 'R'),\n",
       " 'AGG': ('W', 'R', 'G', 'K', 'T', 'M', 'R', 'S', 'S'),\n",
       " 'GTT': ('I', 'L', 'F', 'D', 'A', 'G', 'V', 'V', 'V'),\n",
       " 'GTC': ('I', 'L', 'F', 'D', 'A', 'G', 'V', 'V', 'V'),\n",
       " 'GTA': ('I', 'L', 'L', 'E', 'A', 'G', 'V', 'V', 'V'),\n",
       " 'GTG': ('M', 'L', 'L', 'E', 'A', 'G', 'V', 'V', 'V'),\n",
       " 'GCT': ('T', 'P', 'S', 'D', 'G', 'V', 'A', 'A', 'A'),\n",
       " 'GCC': ('T', 'P', 'S', 'D', 'G', 'V', 'A', 'A', 'A'),\n",
       " 'GCA': ('T', 'P', 'S', 'E', 'G', 'V', 'A', 'A', 'A'),\n",
       " 'GCG': ('T', 'P', 'S', 'E', 'G', 'V', 'A', 'A', 'A'),\n",
       " 'GAT': ('N', 'H', 'Y', 'V', 'A', 'G', 'E', 'D', 'E'),\n",
       " 'GAC': ('N', 'H', 'Y', 'V', 'A', 'G', 'E', 'E', 'D'),\n",
       " 'GAA': ('K', 'Q', '*', 'V', 'A', 'G', 'D', 'D', 'E'),\n",
       " 'GAG': ('K', 'Q', '*', 'V', 'A', 'G', 'E', 'D', 'D'),\n",
       " 'GGT': ('S', 'R', 'C', 'D', 'A', 'V', 'G', 'G', 'G'),\n",
       " 'GGC': ('S', 'R', 'C', 'D', 'A', 'V', 'G', 'G', 'G'),\n",
       " 'GGA': ('R', 'R', '*', 'E', 'A', 'V', 'G', 'G', 'G'),\n",
       " 'GGG': ('R', 'R', 'W', 'E', 'A', 'V', 'G', 'G', 'G')}"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Possible_Amino_Acid_Changes_letters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "#Import mutation rates\n",
    "import csv\n",
    "filename=\"Data files/Trinucleotide_context_mutation_rates.csv\" #csv file containing trinucleotide-context specific mutation rates (calculated from Lee-Six et al 2018 data)\n",
    "\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    Mutation_rates = {} #create a dictionary where key = trincucleotide site, value = mutation rate\n",
    "\n",
    "    for row in read_reader:\n",
    "        if row_count>0:\n",
    "            site = row[0] #trinucleotide site\n",
    "            comp_site = row[2] #complementary site\n",
    "            if site != '':\n",
    "                mutation_rate = float(row[7])\n",
    "                Mutation_rates[site] = mutation_rate\n",
    "                Mutation_rates[comp_site] = mutation_rate\n",
    "        row_count = row_count+1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculating mutation rates for individual variants"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DNMT3A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_variants_DNMT3A(variant):\n",
    "    intron_exon_boundary_caution_base_DNMT3A_NP072046 = [1279, 1429, 1474, 1554, 1852, 2083, 2082, 2173, 2322, 2478]\n",
    "\n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    def intron_exon_check_DNMT3A_NP072046(position_in_gene_number):\n",
    "        if position_in_gene_number == 1279:\n",
    "            gene_pos = 1279\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "    #             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1429:\n",
    "            gene_pos = 1429\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "    #             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1474:\n",
    "            gene_pos = 1474\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "    #             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1554:\n",
    "            gene_pos = 1554\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "    #             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1852:\n",
    "            gene_pos = 1852\n",
    "            first_base = 'G'\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "    #             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2083:\n",
    "            gene_pos = 2083\n",
    "            first_base = 'G'\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "    #             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2082:\n",
    "            gene_pos = 2082\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "    #             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2173:\n",
    "            gene_pos = 2173\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "    #             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2322:\n",
    "            gene_pos = 2322\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "    #             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2478:\n",
    "            gene_pos = 2478\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "    #             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "\n",
    "    #DNMT3A transcript used:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000119772;r=2:25455845-25564754;t=ENST00000321117\n",
    "    DNMT3A_exons_NP_072046 = 'ATGCCCGCCATGCCCTCCAGCGGCCCCGGGGACACCAGCAGCTCTGCTGCGGAGCGGGAGGAGGACCGAAAGGACGGAGAGGAGCAGGAGGAGCCGCGTGGCAAGGAGGAGCGCCAAGAGCCCAGCACCACGGCACGGAAGGTGGGGCGGCCTGGGAGGAAGCGCAAGCACCCCCCGGTGGAAAGCGGTGACACGCCAAAGGACCCTGCGGTGATCTCCAAGTCCCCATCCATGGCCCAGGACTCAGGCGCCTCAGAGCTATTACCCAATGGGGACTTGGAGAAGCGGAGTGAGCCCCAGCCAGAGGAGGGGAGCCCTGCTGGGGGGCAGAAGGGCGGGGCCCCAGCAGAGGGAGAGGGTGCAGCTGAGACCCTGCCTGAAGCCTCAAGAGCAGTGGAAAATGGCTGCTGCACCCCCAAGGAGGGCCGAGGAGCCCCTGCAGAAGCGGGCAAAGAACAGAAGGAGACCAACATCGAATCCATGAAAATGGAGGGCTCCCGGGGCCGGCTGCGGGGTGGCTTGGGCTGGGAGTCCAGCCTCCGTCAGCGGCCCATGCCGAGGCTCACCTTCCAGGCGGGGGACCCCTACTACATCAGCAAGCGCAAGCGGGACGAGTGGCTGGCACGCTGGAAAAGGGAGGCTGAGAAGAAAGCCAAGGTCATTGCAGGAATGAATGCTGTGGAAGAAAACCAGGGGCCCGGGGAGTCTCAGAAGGTGGAGGAGGCCAGCCCTCCTGCTGTGCAGCAGCCCACTGACCCCGCATCCCCCACTGTGGCTACCACGCCTGAGCCCGTGGGGTCCGATGCTGGGGACAAGAATGCCACCAAAGCAGGCGATGACGAGCCAGAGTACGAGGACGGCCGGGGCTTTGGCATTGGGGAGCTGGTGTGGGGGAAACTGCGGGGCTTCTCCTGGTGGCCAGGCCGCATTGTGTCTTGGTGGATGACGGGCCGGAGCCGAGCAGCTGAAGGCACCCGCTGGGTCATGTGGTTCGGAGACGGCAAATTCTCAGTGGTGTGTGTTGAGAAGCTGATGCCGCTGAGCTCGTTTTGCAGTGCGTTCCACCAGGCCACGTACAACAAGCAGCCCATGTACCGCAAAGCCATCTACGAGGTCCTGCAGGTGGCCAGCAGCCGCGCGGGGAAGCTGTTCCCGGTGTGCCACGACAGCGATGAGAGTGACACTGCCAAGGCCGTGGAGGTGCAGAACAAGCCCATGATTGAATGGGCCCTGGGGGGCTTCCAGCCTTCTGGCCCTAAGGGCCTGGAGCCACCAGAAGAAGAGAAGAATCCCTACAAAGAAGTGTACACGGACATGTGGGTGGAACCTGAGGCAGCTGCCTACGCACCACCTCCACCAGCCAAAAAGCCCCGGAAGAGCACAGCGGAGAAGCCCAAGGTCAAGGAGATTATTGATGAGCGCACAAGAGAGCGGCTGGTGTACGAGGTGCGGCAGAAGTGCCGGAACATTGAGGACATCTGCATCTCCTGTGGGAGCCTCAATGTTACCCTGGAACACCCCCTCTTCGTTGGAGGAATGTGCCAAAACTGCAAGAACTGCTTTCTGGAGTGTGCGTACCAGTACGACGACGACGGCTACCAGTCCTACTGCACCATCTGCTGTGGGGGCCGTGAGGTGCTCATGTGCGGAAACAACAACTGCTGCAGGTGCTTTTGCGTGGAGTGTGTGGACCTCTTGGTGGGGCCGGGGGCTGCCCAGGCAGCCATTAAGGAAGACCCCTGGAACTGCTACATGTGCGGGCACAAGGGTACCTACGGGCTGCTGCGGCGGCGAGAGGACTGGCCCTCCCGGCTCCAGATGTTCTTCGCTAATAACCACGACCAGGAATTTGACCCTCCAAAGGTTTACCCACCTGTCCCAGCTGAGAAGAGGAAGCCCATCCGGGTGCTGTCTCTCTTTGATGGAATCGCTACAGGGCTCCTGGTGCTGAAGGACTTGGGCATTCAGGTGGACCGCTACATTGCCTCGGAGGTGTGTGAGGACTCCATCACGGTGGGCATGGTGCGGCACCAGGGGAAGATCATGTACGTCGGGGACGTCCGCAGCGTCACACAGAAGCATATCCAGGAGTGGGGCCCATTCGATCTGGTGATTGGGGGCAGTCCCTGCAATGACCTCTCCATCGTCAACCCTGCTCGCAAGGGCCTCTACGAGGGCACTGGCCGGCTCTTCTTTGAGTTCTACCGCCTCCTGCATGATGCGCGGCCCAAGGAGGGAGATGATCGCCCCTTCTTCTGGCTCTTTGAGAATGTGGTGGCCATGGGCGTTAGTGACAAGAGGGACATCTCGCGATTTCTCGAGTCCAACCCTGTGATGATTGATGCCAAAGAAGTGTCAGCTGCACACAGGGCCCGCTACTTCTGGGGTAACCTTCCCGGTATGAACAGGCCGTTGGCATCCACTGTGAATGATAAGCTGGAGCTGCAGGAGTGTCTGGAGCATGGCAGGATAGCCAAGTTCAGCAAAGTGAGGACCATTACTACGAGGTCAAACTCCATAAAGCAGGGCAAAGACCAGCATTTTCCTGTCTTCATGAATGAGAAAGAGGACATCTTATGGTGCACTGAAATGGAAAGGGTATTTGGTTTCCCAGTCCACTATACTGACGTCTCCAACATGAGCCGCTTGGCGAGGCAGAGACTGCTGGGCCGGTCATGGAGCGTGCCAGTCATCCGCCACCTCTTCGCTCCGCTGAAGGAGTATTTTGCGTGTGTGTAA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    DNMT3A_codons_NP_072046 = [DNMT3A_exons_NP_072046[i:i+3] for i in range(0, len(DNMT3A_exons_NP_072046), 3)]\n",
    "\n",
    "    #DNMT3A transcript used:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000119772;r=2:25455845-25564754;t=ENST00000321117\n",
    "    DNMT3A_exons_extra_bases_NP_072046 = 'GATGCCCGCCATGCCCTCCAGCGGCCCCGGGGACACCAGCAGCTCTGCTGCGGAGCGGGAGGAGGACCGAAAGGACGGAGAGGAGCAGGAGGAGCCGCGTGGCAAGGAGGAGCGCCAAGAGCCCAGCACCACGGCACGGAAGGTGGGGCGGCCTGGGAGGAAGCGCAAGCACCCCCCGGTGGAAAGCGGTGACACGCCAAAGGACCCTGCGGTGATCTCCAAGTCCCCATCCATGGCCCAGGACTCAGGCGCCTCAGAGCTATTACCCAATGGGGACTTGGAGAAGCGGAGTGAGCCCCAGCCAGAGGAGGGGAGCCCTGCTGGGGGGCAGAAGGGCGGGGCCCCAGCAGAGGGAGAGGGTGCAGCTGAGACCCTGCCTGAAGCCTCAAGAGCAGTGGAAAATGGCTGCTGCACCCCCAAGGAGGGCCGAGGAGCCCCTGCAGAAGCGGGCAAAGAACAGAAGGAGACCAACATCGAATCCATGAAAATGGAGGGCTCCCGGGGCCGGCTGCGGGGTGGCTTGGGCTGGGAGTCCAGCCTCCGTCAGCGGCCCATGCCGAGGCTCACCTTCCAGGCGGGGGACCCCTACTACATCAGCAAGCGCAAGCGGGACGAGTGGCTGGCACGCTGGAAAAGGGAGGCTGAGAAGAAAGCCAAGGTCATTGCAGGAATGAATGCTGTGGAAGAAAACCAGGGGCCCGGGGAGTCTCAGAAGGTGGAGGAGGCCAGCCCTCCTGCTGTGCAGCAGCCCACTGACCCCGCATCCCCCACTGTGGCTACCACGCCTGAGCCCGTGGGGTCCGATGCTGGGGACAAGAATGCCACCAAAGCAGGCGATGACGAGCCAGAGTACGAGGACGGCCGGGGCTTTGGCATTGGGGAGCTGGTGTGGGGGAAACTGCGGGGCTTCTCCTGGTGGCCAGGCCGCATTGTGTCTTGGTGGATGACGGGCCGGAGCCGAGCAGCTGAAGGCACCCGCTGGGTCATGTGGTTCGGAGACGGCAAATTCTCAGTGGTGTGTGTTGAGAAGCTGATGCCGCTGAGCTCGTTTTGCAGTGCGTTCCACCAGGCCACGTACAACAAGCAGCCCATGTACCGCAAAGCCATCTACGAGGTCCTGCAGGTGGCCAGCAGCCGCGCGGGGAAGCTGTTCCCGGTGTGCCACGACAGCGATGAGAGTGACACTGCCAAGGCCGTGGAGGTGCAGAACAAGCCCATGATTGAATGGGCCCTGGGGGGCTTCCAGCCTTCTGGCCCTAAGGGCCTGGAGCCACCAGAAGAAGAGAAGAATCCCTACAAAGAAGTGTACACGGACATGTGGGTGGAACCTGAGGCAGCTGCCTACGCACCACCTCCACCAGCCAAAAAGCCCCGGAAGAGCACAGCGGAGAAGCCCAAGGTCAAGGAGATTATTGATGAGCGCACAAGAGAGCGGCTGGTGTACGAGGTGCGGCAGAAGTGCCGGAACATTGAGGACATCTGCATCTCCTGTGGGAGCCTCAATGTTACCCTGGAACACCCCCTCTTCGTTGGAGGAATGTGCCAAAACTGCAAGAACTGCTTTCTGGAGTGTGCGTACCAGTACGACGACGACGGCTACCAGTCCTACTGCACCATCTGCTGTGGGGGCCGTGAGGTGCTCATGTGCGGAAACAACAACTGCTGCAGGTGCTTTTGCGTGGAGTGTGTGGACCTCTTGGTGGGGCCGGGGGCTGCCCAGGCAGCCATTAAGGAAGACCCCTGGAACTGCTACATGTGCGGGCACAAGGGTACCTACGGGCTGCTGCGGCGGCGAGAGGACTGGCCCTCCCGGCTCCAGATGTTCTTCGCTAATAACCACGACCAGGAATTTGACCCTCCAAAGGTTTACCCACCTGTCCCAGCTGAGAAGAGGAAGCCCATCCGGGTGCTGTCTCTCTTTGATGGAATCGCTACAGGGCTCCTGGTGCTGAAGGACTTGGGCATTCAGGTGGACCGCTACATTGCCTCGGAGGTGTGTGAGGACTCCATCACGGTGGGCATGGTGCGGCACCAGGGGAAGATCATGTACGTCGGGGACGTCCGCAGCGTCACACAGAAGCATATCCAGGAGTGGGGCCCATTCGATCTGGTGATTGGGGGCAGTCCCTGCAATGACCTCTCCATCGTCAACCCTGCTCGCAAGGGCCTCTACGAGGGCACTGGCCGGCTCTTCTTTGAGTTCTACCGCCTCCTGCATGATGCGCGGCCCAAGGAGGGAGATGATCGCCCCTTCTTCTGGCTCTTTGAGAATGTGGTGGCCATGGGCGTTAGTGACAAGAGGGACATCTCGCGATTTCTCGAGTCCAACCCTGTGATGATTGATGCCAAAGAAGTGTCAGCTGCACACAGGGCCCGCTACTTCTGGGGTAACCTTCCCGGTATGAACAGGCCGTTGGCATCCACTGTGAATGATAAGCTGGAGCTGCAGGAGTGTCTGGAGCATGGCAGGATAGCCAAGTTCAGCAAAGTGAGGACCATTACTACGAGGTCAAACTCCATAAAGCAGGGCAAAGACCAGCATTTTCCTGTCTTCATGAATGAGAAAGAGGACATCTTATGGTGCACTGAAATGGAAAGGGTATTTGGTTTCCCAGTCCACTATACTGACGTCTCCAACATGAGCCGCTTGGCGAGGCAGAGACTGCTGGGCCGGTCATGGAGCGTGCCAGTCATCCGCCACCTCTTCGCTCCGCTGAAGGAGTATTTTGCGTGTGTGTAAG'\n",
    "\n",
    "    #Create a list of the codons they called missense variants at\n",
    "    DNMT3A_variant_list_NP_072046 = []\n",
    "    DNMT3A_missense_codons_NP_072046 =[]\n",
    "    DNMT3A_missense_variant_DICT_NP_072046 = {}\n",
    "\n",
    "    variant_codon = int(((re.findall('\\d+', variant)))[0])\n",
    "\n",
    "    codon0_NP_072046 = variant_codon-1\n",
    "    codon1_NP_072046 = variant_codon\n",
    "    variant_NP_072046 = variant\n",
    "    DNMT3A_missense_codons_NP_072046.append(codon0_NP_072046) #List of codon numbers that they called variants at\n",
    "    DNMT3A_missense_variant_DICT_NP_072046[row_count]=(codon1_NP_072046, variant_NP_072046) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    DNMT3A_missense_codons_NP_072046_DICT = {}\n",
    "    calls_DNMT3A_missense_NP_072046 = {}\n",
    "\n",
    "    for i in DNMT3A_missense_codons_NP_072046: #Iterates over the list of codon numbers\n",
    "        DNMT3A_missense_codons_NP_072046_DICT[i+1]= (DNMT3A_codons_NP_072046[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "\n",
    "    for k,v in DNMT3A_missense_codons_NP_072046_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_DNMT3A_missense_NP_072046[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_DNMT3A_missense_NP_072046_extra = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_missense_NP_072046.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_DNMT3A_missense_NP_072046_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_DNMT3A_missense_NP_072046_extra2 = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_missense_NP_072046_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_DNMT3A_missense_NP_072046_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    DNMT3A_position_in_list_dict_missense_NP_072046={}\n",
    "    for k, v in DNMT3A_missense_variant_DICT_NP_072046.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_DNMT3A_missense_NP_072046[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_DNMT3A_missense_NP_072046[v[0]]\n",
    "        codon_seq=calls_DNMT3A_missense_NP_072046[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_DNMT3A_missense_NP_072046_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                DNMT3A_position_in_list_dict_missense_NP_072046[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "\n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in DNMT3A_position_in_list_dict_missense_NP_072046.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "    #     print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    DNMT3A_bases_missense_NP_072046 = {}\n",
    "    for k, v in DNMT3A_position_in_list_dict_missense_NP_072046.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        DNMT3A_bases_missense_NP_072046[k]=(v, base_changes)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    DNMT3A_bases_missense2_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_missense_NP_072046.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        DNMT3A_bases_missense2_NP_072046[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    DNMT3A_bases_missense3_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_missense2_NP_072046.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        DNMT3A_bases_missense3_NP_072046[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    DNMT3A_bases_missense4_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_missense3_NP_072046.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_DNMT3A_NP072046:\n",
    "                adapted_trio = intron_exon_check_DNMT3A_NP072046(i)\n",
    "                sites_trio_missense.append([adapted_trio])\n",
    "                DNMT3A_bases_missense4_NP_072046[k] = (v, sites_trio_missense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_missense = [DNMT3A_exons_extra_bases_NP_072046[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_missense.append(site_trio_missense)\n",
    "                    DNMT3A_bases_missense4_NP_072046[k] = (v, sites_trio_missense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    DNMT3A_sites_missense5_NP_072046 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in DNMT3A_bases_missense4_NP_072046.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        DNMT3A_sites_missense5_NP_072046[a]=info\n",
    "        a=a+1\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    DNMT3A_sites_missense6_NP_072046={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in DNMT3A_sites_missense5_NP_072046.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (original_base,new_base)\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        DNMT3A_sites_missense6_NP_072046[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    DNMT3A_mutation_rates_missense_NP_072046 = {}\n",
    "    for k, v in DNMT3A_sites_missense6_NP_072046.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            DNMT3A_mutation_rates_missense_NP_072046[k] = (v, mutation_rate)\n",
    "\n",
    "    DNMT3A_mutation_rate_missense_NP_072046_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in DNMT3A_mutation_rates_missense_NP_072046.items():\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        DNMT3A_mutation_rate_missense_NP_072046_sum.append(mutation_rate_value)\n",
    "\n",
    "    DNMT3A_Summed_mutation_rate_missense_NP_072046_added = np.sum(DNMT3A_mutation_rate_missense_NP_072046_sum)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return DNMT3A_Summed_mutation_rate_missense_NP_072046_added"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TET2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_variants_TET2(variant):\n",
    "    intron_exon_boundary_caution_base_TET2 = [3409, 3804, 4044, 4537]\n",
    "\n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "\n",
    "    def intron_exon_check_TET2(position_in_gene_number):\n",
    "        if position_in_gene_number == 3409:\n",
    "            gene_pos = 3409\n",
    "            first_base = TET2_exons_extra_bases_NP[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 3804:\n",
    "            gene_pos = 3804\n",
    "            first_base = 'G'\n",
    "            middle_base = TET2_exons_extra_bases_NP[gene_pos]\n",
    "            third_base = TET2_exons_extra_bases_NP[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 4044:\n",
    "            gene_pos = 4044\n",
    "            first_base = TET2_exons_extra_bases_NP[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 4537:\n",
    "            gene_pos = 4537\n",
    "            first_base = TET2_exons_extra_bases_NP[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #TET2 transcript (NP):\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000168769;r=4:106067032-106163928;t=ENST00000380013\n",
    "    TET2_exons_NP = 'ATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGAGCAAATTATTGAAAAAGATGAAGGTCCTTTTTATACCCATCTAGGAGCAGGTCCTAATGTGGCAGCTATTAGAGAAATCATGGAAGAAAGGTTTGGACAGAAGGGTAAAGCTATTAGGATTGAAAGAGTCATCTATACTGGTAAAGAAGGCAAAAGTTCTCAGGGATGTCCTATTGCTAAGTGGGTGGTTCGCAGAAGCAGCAGTGAAGAGAAGCTACTGTGTTTGGTGCGGGAGCGAGCTGGCCACACCTGTGAGGCTGCAGTGATTGTGATTCTCATCCTGGTGTGGGAAGGAATCCCGCTGTCTCTGGCTGACAAACTCTACTCGGAGCTTACCGAGACGCTGAGGAAATACGGCACGCTCACCAATCGCCGGTGTGCCTTGAATGAAGAGAGAACTTGCGCCTGTCAGGGGCTGGATCCAGAAACCTGTGGTGCCTCCTTCTCTTTTGGTTGTTCATGGAGCATGTACTACAATGGATGTAAGTTTGCCAGAAGCAAGATCCCAAGGAAGTTTAAGCTGCTTGGGGATGACCCAAAAGAGGAAGAGAAACTGGAGTCTCATTTGCAAAACCTGTCCACTCTTATGGCACCAACATATAAGAAACTTGCACCTGATGCATATAATAATCAGATTGAATATGAACACAGAGCACCAGAGTGCCGTCTGGGTCTGAAGGAAGGCCGTCCATTCTCAGGGGTCACTGCATGTTTGGACTTCTGTGCTCATGCCCACAGAGACTTGCACAACATGCAGAATGGCAGCACATTGGTATGCACTCTCACTAGAGAAGACAATCGAGAATTTGGAGGAAAACCTGAGGATGAGCAGCTTCACGTTCTGCCTTTATACAAAGTCTCTGACGTGGATGAGTTTGGGAGTGTGGAAGCTCAGGAGGAGAAAAAACGGAGTGGTGCCATTCAGGTACTGAGTTCTTTTCGGCGAAAAGTCAGGATGTTAGCAGAGCCAGTCAAGACTTGCCGACAAAGGAAACTAGAAGCCAAGAAAGCTGCAGCTGAAAAGCTTTCCTCCCTGGAGAACAGCTCAAATAAAAATGAAAAGGAAAAGTCAGCCCCATCACGTACAAAACAAACTGAAAACGCAAGCCAGGCTAAACAGTTGGCAGAACTTTTGCGACTTTCAGGACCAGTCATGCAGCAGTCCCAGCAGCCCCAGCCTCTACAGAAGCAGCCACCACAGCCCCAGCAGCAGCAGAGACCCCAGCAGCAGCAGCCACATCACCCTCAGACAGAGTCTGTCAACTCTTATTCTGCTTCTGGATCCACCAATCCATACATGAGACGGCCCAATCCAGTTAGTCCTTATCCAAACTCTTCACACACTTCAGATATCTATGGAAGCACCAGCCCTATGAACTTCTATTCCACCTCATCTCAAGCTGCAGGTTCATATTTGAATTCTTCTAATCCCATGAACCCTTACCCTGGGCTTTTGAATCAGAATACCCAATATCCATCATATCAATGCAATGGAAACCTATCAGTGGACAACTGCTCCCCATATCTGGGTTCCTATTCTCCCCAGTCTCAGCCGATGGATCTGTATAGGTATCCAAGCCAAGACCCTCTGTCTAAGCTCAGTCTACCACCCATCCATACACTTTACCAGCCAAGGTTTGGAAATAGCCAGAGTTTTACATCTAAATACTTAGGTTATGGAAACCAAAATATGCAGGGAGATGGTTTCAGCAGTTGTACCATTAGACCAAATGTACATCATGTAGGGAAATTGCCTCCTTATCCCACTCATGAGATGGATGGCCACTTCATGGGAGCCACCTCTAGATTACCACCCAATCTGAGCAATCCAAACATGGACTATAAAAATGGTGAACATCATTCACCTTCTCACATAATCCATAACTACAGTGCAGCTCCGGGCATGTTCAACAGCTCTCTTCATGCCCTGCATCTCCAAAACAAGGAGAATGACATGCTTTCCCACACAGCTAATGGGTTATCAAAGATGCTTCCAGCTCTTAACCATGATAGAACTGCTTGTGTCCAAGGAGGCTTACACAAATTAAGTGATGCTAATGGTCAGGAAAAGCAGCCATTGGCACTAGTCCAGGGTGTGGCTTCTGGTGCAGAGGACAACGATGAGGTCTGGTCAGACAGCGAGCAGAGCTTTCTGGATCCTGACATTGGGGGAGTGGCCGTGGCTCCAACTCATGGGTCAATTCTCATTGAGTGTGCAAAGCGTGAGCTGCATGCCACAACCCCTTTAAAGAATCCCAATAGGAATCACCCCACCAGGATCTCCCTCGTCTTTTACCAGCATAAGAGCATGAATGAGCCAAAACATGGCTTGGCTCTTTGGGAAGCCAAAATGGCTGAAAAAGCCCGTGAGAAAGAGGAAGAGTGTGAAAAGTATGGCCCAGACTATGTGCCTCAGAAATCCCATGGCAAAAAAGTGAAACGGGAGCCTGCTGAGCCACATGAAACTTCAGAGCCCACTTACCTGCGTTTCATCAAGTCTCTTGCCGAAAGGACCATGTCCGTGACCACAGACTCCACAGTAACTACATCTCCATATGCCTTCACTCGGGTCACAGGGCCTTACAACAGATATATATGA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    TET2_codons_NP = [TET2_exons_NP[i:i+3] for i in range(0, len(TET2_exons_NP), 3)]\n",
    "\n",
    "    #TET2 transcript (transcript variant 2 (TET2-001), NM_017628, NP_001120680):\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000168769;r=4:106067032-106163928;t=ENST00000380013\n",
    "    TET2_exons_extra_bases_NP = 'GATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGAGCAAATTATTGAAAAAGATGAAGGTCCTTTTTATACCCATCTAGGAGCAGGTCCTAATGTGGCAGCTATTAGAGAAATCATGGAAGAAAGGTTTGGACAGAAGGGTAAAGCTATTAGGATTGAAAGAGTCATCTATACTGGTAAAGAAGGCAAAAGTTCTCAGGGATGTCCTATTGCTAAGTGGGTGGTTCGCAGAAGCAGCAGTGAAGAGAAGCTACTGTGTTTGGTGCGGGAGCGAGCTGGCCACACCTGTGAGGCTGCAGTGATTGTGATTCTCATCCTGGTGTGGGAAGGAATCCCGCTGTCTCTGGCTGACAAACTCTACTCGGAGCTTACCGAGACGCTGAGGAAATACGGCACGCTCACCAATCGCCGGTGTGCCTTGAATGAAGAGAGAACTTGCGCCTGTCAGGGGCTGGATCCAGAAACCTGTGGTGCCTCCTTCTCTTTTGGTTGTTCATGGAGCATGTACTACAATGGATGTAAGTTTGCCAGAAGCAAGATCCCAAGGAAGTTTAAGCTGCTTGGGGATGACCCAAAAGAGGAAGAGAAACTGGAGTCTCATTTGCAAAACCTGTCCACTCTTATGGCACCAACATATAAGAAACTTGCACCTGATGCATATAATAATCAGATTGAATATGAACACAGAGCACCAGAGTGCCGTCTGGGTCTGAAGGAAGGCCGTCCATTCTCAGGGGTCACTGCATGTTTGGACTTCTGTGCTCATGCCCACAGAGACTTGCACAACATGCAGAATGGCAGCACATTGGTATGCACTCTCACTAGAGAAGACAATCGAGAATTTGGAGGAAAACCTGAGGATGAGCAGCTTCACGTTCTGCCTTTATACAAAGTCTCTGACGTGGATGAGTTTGGGAGTGTGGAAGCTCAGGAGGAGAAAAAACGGAGTGGTGCCATTCAGGTACTGAGTTCTTTTCGGCGAAAAGTCAGGATGTTAGCAGAGCCAGTCAAGACTTGCCGACAAAGGAAACTAGAAGCCAAGAAAGCTGCAGCTGAAAAGCTTTCCTCCCTGGAGAACAGCTCAAATAAAAATGAAAAGGAAAAGTCAGCCCCATCACGTACAAAACAAACTGAAAACGCAAGCCAGGCTAAACAGTTGGCAGAACTTTTGCGACTTTCAGGACCAGTCATGCAGCAGTCCCAGCAGCCCCAGCCTCTACAGAAGCAGCCACCACAGCCCCAGCAGCAGCAGAGACCCCAGCAGCAGCAGCCACATCACCCTCAGACAGAGTCTGTCAACTCTTATTCTGCTTCTGGATCCACCAATCCATACATGAGACGGCCCAATCCAGTTAGTCCTTATCCAAACTCTTCACACACTTCAGATATCTATGGAAGCACCAGCCCTATGAACTTCTATTCCACCTCATCTCAAGCTGCAGGTTCATATTTGAATTCTTCTAATCCCATGAACCCTTACCCTGGGCTTTTGAATCAGAATACCCAATATCCATCATATCAATGCAATGGAAACCTATCAGTGGACAACTGCTCCCCATATCTGGGTTCCTATTCTCCCCAGTCTCAGCCGATGGATCTGTATAGGTATCCAAGCCAAGACCCTCTGTCTAAGCTCAGTCTACCACCCATCCATACACTTTACCAGCCAAGGTTTGGAAATAGCCAGAGTTTTACATCTAAATACTTAGGTTATGGAAACCAAAATATGCAGGGAGATGGTTTCAGCAGTTGTACCATTAGACCAAATGTACATCATGTAGGGAAATTGCCTCCTTATCCCACTCATGAGATGGATGGCCACTTCATGGGAGCCACCTCTAGATTACCACCCAATCTGAGCAATCCAAACATGGACTATAAAAATGGTGAACATCATTCACCTTCTCACATAATCCATAACTACAGTGCAGCTCCGGGCATGTTCAACAGCTCTCTTCATGCCCTGCATCTCCAAAACAAGGAGAATGACATGCTTTCCCACACAGCTAATGGGTTATCAAAGATGCTTCCAGCTCTTAACCATGATAGAACTGCTTGTGTCCAAGGAGGCTTACACAAATTAAGTGATGCTAATGGTCAGGAAAAGCAGCCATTGGCACTAGTCCAGGGTGTGGCTTCTGGTGCAGAGGACAACGATGAGGTCTGGTCAGACAGCGAGCAGAGCTTTCTGGATCCTGACATTGGGGGAGTGGCCGTGGCTCCAACTCATGGGTCAATTCTCATTGAGTGTGCAAAGCGTGAGCTGCATGCCACAACCCCTTTAAAGAATCCCAATAGGAATCACCCCACCAGGATCTCCCTCGTCTTTTACCAGCATAAGAGCATGAATGAGCCAAAACATGGCTTGGCTCTTTGGGAAGCCAAAATGGCTGAAAAAGCCCGTGAGAAAGAGGAAGAGTGTGAAAAGTATGGCCCAGACTATGTGCCTCAGAAATCCCATGGCAAAAAAGTGAAACGGGAGCCTGCTGAGCCACATGAAACTTCAGAGCCCACTTACCTGCGTTTCATCAAGTCTCTTGCCGAAAGGACCATGTCCGTGACCACAGACTCCACAGTAACTACATCTCCATATGCCTTCACTCGGGTCACAGGGCCTTACAACAGATATATATGAT'\n",
    "\n",
    "    #Create a list of the codons they called missense variants at\n",
    "    TET2_variant_list_NP = []\n",
    "    TET2_missense_codons_NP =[]\n",
    "    TET2_missense_variant_DICT_NP = {}\n",
    "\n",
    "    variant_codon = int(((re.findall('\\d+', variant)))[0])\n",
    "\n",
    "    codon0_NP = variant_codon-1\n",
    "    codon1_NP = variant_codon\n",
    "    variant_NP = variant\n",
    "    TET2_missense_codons_NP.append(codon0_NP) #List of codon numbers that they called variants at\n",
    "    TET2_missense_variant_DICT_NP[row_count]=(codon1_NP, variant_NP) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    TET2_missense_codons_NP_DICT = {}\n",
    "    calls_TET2_missense_NP = {}\n",
    "\n",
    "    for i in TET2_missense_codons_NP: #Iterates over the list of codon numbers\n",
    "        TET2_missense_codons_NP_DICT[i+1]= (TET2_codons_NP[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "\n",
    "    for k,v in TET2_missense_codons_NP_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_TET2_missense_NP[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_TET2_missense_NP_extra = {}\n",
    "\n",
    "    for k, v in calls_TET2_missense_NP.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_TET2_missense_NP_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_TET2_missense_NP_extra2 = {}\n",
    "\n",
    "    for k, v in calls_TET2_missense_NP_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_TET2_missense_NP_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    TET2_position_in_list_dict_missense_NP={}\n",
    "    for k, v in TET2_missense_variant_DICT_NP.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_TET2_missense_NP[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_TET2_missense_NP[v[0]]\n",
    "        codon_seq=calls_TET2_missense_NP[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_TET2_missense_NP_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                TET2_position_in_list_dict_missense_NP[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "\n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in TET2_position_in_list_dict_missense_NP.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "    #     print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TET2_bases_missense_NP = {}\n",
    "    for k, v in TET2_position_in_list_dict_missense_NP.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TET2_bases_missense_NP[k]=(v, base_changes)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TET2_bases_missense2_NP = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense_NP.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TET2_bases_missense2_NP[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TET2_bases_missense3_NP = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense2_NP.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TET2_bases_missense3_NP[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TET2_bases_missense4_NP = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense3_NP.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_TET2:\n",
    "    #                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check_TET2(i)\n",
    "                sites_trio_missense.append([adapted_trio])\n",
    "                TET2_bases_missense4_NP[k] = (v, sites_trio_missense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_missense = [TET2_exons_extra_bases_NP[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_missense.append(site_trio_missense)\n",
    "                    TET2_bases_missense4_NP[k] = (v, sites_trio_missense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TET2_sites_missense5_NP = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TET2_bases_missense4_NP.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        TET2_sites_missense5_NP[a]=info\n",
    "        a=a+1\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TET2_sites_missense6_NP={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TET2_sites_missense5_NP.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (original_base,new_base)\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TET2_sites_missense6_NP[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    TET2_mutation_rates_missense_NP = {}\n",
    "    for k, v in TET2_sites_missense6_NP.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            TET2_mutation_rates_missense_NP[k] = (v, mutation_rate)\n",
    "\n",
    "    TET2_mutation_rate_missense_NP_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TET2_mutation_rates_missense_NP.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        TET2_mutation_rate_missense_NP_sum.append(mutation_rate_value)\n",
    "\n",
    "    TET2_Summed_mutation_rate_missense_NP_added = np.sum(TET2_mutation_rate_missense_NP_sum)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return TET2_Summed_mutation_rate_missense_NP_added"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ASXL1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_variants_ASXL1(variant):\n",
    "    intron_exon_boundary_caution_base_ASXL1 = [140, 143, 252, 373, 1086, 1085]\n",
    "   \n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "\n",
    "    def intron_exon_check(position_in_gene_number):\n",
    "        if position_in_gene_number == 140:\n",
    "            gene_pos = 140\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 143:\n",
    "            gene_pos = 143\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 252:\n",
    "            gene_pos = 252\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 373:\n",
    "            gene_pos = 373\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1086:\n",
    "            gene_pos = 1086\n",
    "            first_base = 'G'\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = ASXL1_exons_extra_bases[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1085:\n",
    "            gene_pos = 1085\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #ASXL1 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    ASXL1_exons = 'ATGAAGGACAAACAGAAGAAGAAGAAGGAGCGCACGTGGGCCGAGGCCGCGCGCCTGGTATTAGAAAACTACTCGGATGCTCCAATGACACCAAAACAGATTCTGCAGGTCATAGAGGCAGAAGGACTAAAGGAAATGAGAAGTGGGACTTCCCCTCTCGCATGCCTCAATGCTATGCTACATTCCAATTCAAGAGGAGGAGAGGGGTTGTTTTATAAACTGCCTGGCCGAATCAGCCTTTTCACGCTCAAGAAGGATGCCCTGCAGTGGTCTCGCCATCCAGCTACAGTGGAGGGAGAGGAGCCAGAGGACACGGCTGATGTGGAGAGCTGTGGGTCTAATGAAGCCAGCACTGTGAGTGGTGAAAACGATGTATCTCTTGATGAAACATCTTCGAACGCATCCTGTTCTACAGAATCTCAGAGTCGACCTCTTTCCAATCCCAGGGACAGCTACAGAGCTTCCTCACAGGCGAACAAACAAAAGAAAAAGACTGGGGTGATGCTGCCTCGAGTTGTCCTGACTCCTCTGAAGGTAAACGGGGCCCACGTGGAATCTGCATCAGGGTTCTCGGGCTGCCACGCCGATGGCGAGAGCGGCAGCCCGTCCAGCAGCAGCAGCGGCTCTCTGGCCCTGGGCAGCGCTGCTATTCGTGGCCAGGCCGAGGTCACCCAGGACCCTGCCCCGCTCCTGAGAGGCTTCCGGAAGCCAGCCACAGGTCAAATGAAGCGCAACAGAGGGGAAGAAATAGATTTTGAGACACCTGGGTCCATTCTTGTCAACACCAACCTCCGTGCCCTGATCAACTCTCGGACCTTCCATGCCTTACCATCACACTTCCAGCAGCAGCTCCTCTTCCTCCTGCCTGAAGTAGACAGACAGGTGGGGACGGATGGCCTGTTGCGTCTCAGCAGCAGTGCACTAAATAACGAGTTTTTTACCCATGCGGCTCAGAGCTGGCGGGAGCGCCTGGCTGATGGTGAATTTACTCATGAGATGCAAGTCAGGATACGACAGGAAATGGAGAAGGAAAAGAAGGTGGAACAATGGAAAGAAAAGTTCTTTGAAGACTACTATGGACAGAAGCTGGGTTTGACCAAAGAAGAGTCATTGCAGCAGAACGTGGGCCAGGAGGAGGCTGAAATCAAAAGTGGCTTGTGTGTCCCAGGAGAATCAGTGCGTATACAGCGTGGTCCAGCCACCCGACAGCGAGATGGGCATTTTAAGAAACGCTCTCGGCCAGATCTCCGAACCAGAGCCAGAAGGAATCTGTACAAAAAACAGGAGTCAGAACAAGCAGGGGTTGCTAAGGATGCAAAATCTGTGGCCTCAGATGTTCCCCTCTACAAGGATGGGGAGGCTAAGACTGACCCAGCAGGGCTGAGCAGTCCCCATCTGCCAGGCACATCCTCTGCAGCACCCGACCTGGAGGGTCCCGAATTCCCAGTTGAGTCTGTGGCTTCTCGGATCCAGGCTGAGCCAGACAACTTGGCACGTGCCTCTGCATCTCCAGACAGAATTCCTAGCCTGCCTCAGGAAACTGTGGATCAGGAACCCAAGGATCAGAAGAGGAAATCCTTTGAGCAGGCGGCCTCTGCATCCTTTCCCGAAAAGAAGCCCCGGCTTGAAGATCGTCAGTCCTTTCGTAACACAATTGAAAGTGTTCACACCGAAAAGCCACAGCCCACTAAAGAGGAGCCCAAAGTCCCGCCCATCCGGATTCAACTTTCACGTATCAAACCACCCTGGGTGGTTAAAGGTCAGCCCACTTACCAGATATGCCCCCGGATCATCCCCACCACGGAGTCCTCCTGCCGGGGTTGGACTGGCGCCAGGACCCTCGCAGACATTAAAGCCCGTGCTCTGCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCCACCACTGCCATCGGAGGGGGGGGTGGCCCGGGTGGAGGTGGCGGCGGGGCCACCGATGAGGGAGGTGGCAGAGGCAGCAGCAGTGGTGATGGTGGTGAGGCCTGTGGCCACCCTGAGCCCAGGGGAGGCCCGAGCACCCCTGGAAAGTGTACGTCAGATCTACAGCGAACACAACTACTGCCGCCTTATCCTCTAAATGGGGAGCATACCCAGGCCGGAACTGCCATGTCCAGAGCTAGGAGAGAGGACCTGCCTTCTCTGAGAAAGGAGGAAAGCTGCCTACTACAGAGGGCTACAGTTGGACTCACAGATGGGCTAGGAGATGCCTCCCAACTCCCCGTTGCTCCCACTGGGGACCAGCCATGCCAGGCCTTGCCCCTACTGTCCTCCCAAACCTCAGTAGCTGAGAGATTAGTGGAGCAGCCTCAGTTGCATCCGGATGTTAGAACTGAATGTGAGTCTGGCACCACTTCCTGGGAAAGTGATGATGAGGAGCAAGGACCCACCGTTCCTGCAGACAATGGTCCCATTCTGTCTCTAGTGGGAGATGATACATTAGAGAAAGGAACTGGCCAAGCTCTTGACAGTCATCCCACTATGAAGGATCCTGTAAATGTGACCCCCAGTTCCACACCTGAATCCTCACCGACTGATTGCCTGCAGAACAGAGCATTTGATGACGAATTAGGGCTTGGTGGCTCATGCCCTCCTATGAGGGAAAGTGATACTAGACAAGAAAACTTGAAAACCAAGGCTCTCGTTTCTAACAGTTCTTTGCATTGGATACCCATCCCATCGAATGATGAGGTAGTGAAACAGCCCAAACCAGAATCCAGAGAACACATACCATCTGTTGAGCCCCAGGTTGGAGAGGAGTGGGAGAAAGCTGCTCCCACCCCTCCTGCATTGCCTGGGGATTTGACAGCTGAGGAGGGTCTAGATCCTCTTGACAGCCTTACTTCACTCTGGACTGTGCCATCTCGAGGAGGCAGTGACAGCAATGGCAGTTACTGTCAACAGGTGGACATTGAAAAGCTGAAAATCAACGGAGACTCTGAAGCACTGAGTCCTCACGGTGAGTCCACGGATACAGCCTCTGACTTTGAAGGTCACCTCACGGAGGACAGCAGTGAGGCTGACACTAGAGAAGCTGCAGTGACAAAGGGATCTTCGGTGGACAAGGATGAGAAACCCAATTGGAACCAATCTGCCCCACTGTCCAAGGTGAATGGTGACATGCGTCTGGTTACAAGGACAGATGGGATGGTTGCTCCTCAGAGCTGGGTGTCTCGAGTATGTGCGGTCCGCCAAAAGATCCCAGATTCCCTACTGCTGGCCAGTACTGAGTACCAGCCAAGAGCCGTGTGCCTGTCCATGCCTGGGTCCTCAGTGGAGGCCACTAACCCACTTGTGATGCAGTTGCTGCAGGGTAGCTTGCCCCTAGAGAAGGTTCTTCCACCAGCCCACGATGACAGCATGTCAGAATCCCCACAAGTACCACTTACAAAAGACCAGAGCCATGGCTCGCTACGCATGGGATCTTTACATGGTCTTGGAAAAAACAGTGGCATGGTTGATGGAAGCAGCCCCAGTTCTTTAAGGGCTTTGAAGGAGCCTCTTCTGCCAGATAGCTGTGAAACAGGCACTGGTCTTGCCAGGATTGAGGCCACCCAGGCTCCTGGAGCACCCCAAAAGAATTGCAAGGCAGTCCCAAGTTTTGACTCCCTCCATCCAGTGACAAATCCCATTACATCCTCTAGGAAACTGGAAGAAATGGATTCCAAAGAGCAGTTCTCTTCCTTTAGTTGTGAAGATCAGAAGGAAGTCCGTGCTATGTCACAGGACAGTAATTCAAATGCTGCTCCAGGAAAGAGCCCAGGAGATCTTACTACCTCGAGAACACCTCGTTTCTCATCTCCAAATGTGATCTCCTTTGGTCCAGAGCAGACAGGTCGGGCCCTGGGTGATCAGAGCAATGTTACAGGCCAAGGGAAGAAGCTTTTTGGCTCTGGGAATGTGGCTGCAACCCTTCAGCGCCCCAGGCCTGCGGACCCGATGCCTCTTCCTGCTGAGATCCCTCCAGTTTTTCCCAGTGGGAAGTTGGGACCAAGCACAAACTCCATGTCTGGTGGGGTACAGACTCCAAGGGAAGACTGGGCTCCAAAGCCACATGCCTTTGTTGGCAGCGTCAAGAATGAGAAGACTTTTGTGGGGGGTCCTCTTAAGGCAAATGCCGAGAACAGGAAAGCTACTGGGCATAGTCCCCTGGAACTGGTGGGTCACTTGGAAGGGATGCCCTTTGTCATGGACTTGCCCTTCTGGAAATTACCCCGAGAGCCAGGGAAGGGGCTCAGTGAGCCTCTGGAGCCTTCTTCTCTCCCCTCCCAACTCAGCATCAAGCAGGCATTTTATGGGAAGCTTTCTAAACTCCAACTGAGTTCCACCAGCTTTAATTATTCCTCTAGCTCTCCCACCTTTCCCAAAGGCCTTGCTGGAAGTGTGGTGCAGCTGAGCCACAAAGCAAACTTTGGTGCGAGCCACAGTGCATCACTTTCCTTGCAAATGTTCACTGACAGCAGCACGGTGGAAAGCATCTCGCTCCAGTGTGCGTGCAGCCTGAAAGCCATGATCATGTGCCAAGGCTGCGGTGCGTTCTGTCACGATGACTGTATTGGACCCTCAAAGCTCTGTGTATTGTGCCTTGTGGTGAGATAA'\n",
    "    \n",
    "    #Split the sequence into codons \n",
    "    ASXL1_codons = [ASXL1_exons[i:i+3] for i in range(0, len(ASXL1_exons), 3)]\n",
    "\n",
    "    #ASXL1 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    ASXL1_exons_extra_bases = 'GATGAAGGACAAACAGAAGAAGAAGAAGGAGCGCACGTGGGCCGAGGCCGCGCGCCTGGTATTAGAAAACTACTCGGATGCTCCAATGACACCAAAACAGATTCTGCAGGTCATAGAGGCAGAAGGACTAAAGGAAATGAGAAGTGGGACTTCCCCTCTCGCATGCCTCAATGCTATGCTACATTCCAATTCAAGAGGAGGAGAGGGGTTGTTTTATAAACTGCCTGGCCGAATCAGCCTTTTCACGCTCAAGAAGGATGCCCTGCAGTGGTCTCGCCATCCAGCTACAGTGGAGGGAGAGGAGCCAGAGGACACGGCTGATGTGGAGAGCTGTGGGTCTAATGAAGCCAGCACTGTGAGTGGTGAAAACGATGTATCTCTTGATGAAACATCTTCGAACGCATCCTGTTCTACAGAATCTCAGAGTCGACCTCTTTCCAATCCCAGGGACAGCTACAGAGCTTCCTCACAGGCGAACAAACAAAAGAAAAAGACTGGGGTGATGCTGCCTCGAGTTGTCCTGACTCCTCTGAAGGTAAACGGGGCCCACGTGGAATCTGCATCAGGGTTCTCGGGCTGCCACGCCGATGGCGAGAGCGGCAGCCCGTCCAGCAGCAGCAGCGGCTCTCTGGCCCTGGGCAGCGCTGCTATTCGTGGCCAGGCCGAGGTCACCCAGGACCCTGCCCCGCTCCTGAGAGGCTTCCGGAAGCCAGCCACAGGTCAAATGAAGCGCAACAGAGGGGAAGAAATAGATTTTGAGACACCTGGGTCCATTCTTGTCAACACCAACCTCCGTGCCCTGATCAACTCTCGGACCTTCCATGCCTTACCATCACACTTCCAGCAGCAGCTCCTCTTCCTCCTGCCTGAAGTAGACAGACAGGTGGGGACGGATGGCCTGTTGCGTCTCAGCAGCAGTGCACTAAATAACGAGTTTTTTACCCATGCGGCTCAGAGCTGGCGGGAGCGCCTGGCTGATGGTGAATTTACTCATGAGATGCAAGTCAGGATACGACAGGAAATGGAGAAGGAAAAGAAGGTGGAACAATGGAAAGAAAAGTTCTTTGAAGACTACTATGGACAGAAGCTGGGTTTGACCAAAGAAGAGTCATTGCAGCAGAACGTGGGCCAGGAGGAGGCTGAAATCAAAAGTGGCTTGTGTGTCCCAGGAGAATCAGTGCGTATACAGCGTGGTCCAGCCACCCGACAGCGAGATGGGCATTTTAAGAAACGCTCTCGGCCAGATCTCCGAACCAGAGCCAGAAGGAATCTGTACAAAAAACAGGAGTCAGAACAAGCAGGGGTTGCTAAGGATGCAAAATCTGTGGCCTCAGATGTTCCCCTCTACAAGGATGGGGAGGCTAAGACTGACCCAGCAGGGCTGAGCAGTCCCCATCTGCCAGGCACATCCTCTGCAGCACCCGACCTGGAGGGTCCCGAATTCCCAGTTGAGTCTGTGGCTTCTCGGATCCAGGCTGAGCCAGACAACTTGGCACGTGCCTCTGCATCTCCAGACAGAATTCCTAGCCTGCCTCAGGAAACTGTGGATCAGGAACCCAAGGATCAGAAGAGGAAATCCTTTGAGCAGGCGGCCTCTGCATCCTTTCCCGAAAAGAAGCCCCGGCTTGAAGATCGTCAGTCCTTTCGTAACACAATTGAAAGTGTTCACACCGAAAAGCCACAGCCCACTAAAGAGGAGCCCAAAGTCCCGCCCATCCGGATTCAACTTTCACGTATCAAACCACCCTGGGTGGTTAAAGGTCAGCCCACTTACCAGATATGCCCCCGGATCATCCCCACCACGGAGTCCTCCTGCCGGGGTTGGACTGGCGCCAGGACCCTCGCAGACATTAAAGCCCGTGCTCTGCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCCACCACTGCCATCGGAGGGGGGGGTGGCCCGGGTGGAGGTGGCGGCGGGGCCACCGATGAGGGAGGTGGCAGAGGCAGCAGCAGTGGTGATGGTGGTGAGGCCTGTGGCCACCCTGAGCCCAGGGGAGGCCCGAGCACCCCTGGAAAGTGTACGTCAGATCTACAGCGAACACAACTACTGCCGCCTTATCCTCTAAATGGGGAGCATACCCAGGCCGGAACTGCCATGTCCAGAGCTAGGAGAGAGGACCTGCCTTCTCTGAGAAAGGAGGAAAGCTGCCTACTACAGAGGGCTACAGTTGGACTCACAGATGGGCTAGGAGATGCCTCCCAACTCCCCGTTGCTCCCACTGGGGACCAGCCATGCCAGGCCTTGCCCCTACTGTCCTCCCAAACCTCAGTAGCTGAGAGATTAGTGGAGCAGCCTCAGTTGCATCCGGATGTTAGAACTGAATGTGAGTCTGGCACCACTTCCTGGGAAAGTGATGATGAGGAGCAAGGACCCACCGTTCCTGCAGACAATGGTCCCATTCTGTCTCTAGTGGGAGATGATACATTAGAGAAAGGAACTGGCCAAGCTCTTGACAGTCATCCCACTATGAAGGATCCTGTAAATGTGACCCCCAGTTCCACACCTGAATCCTCACCGACTGATTGCCTGCAGAACAGAGCATTTGATGACGAATTAGGGCTTGGTGGCTCATGCCCTCCTATGAGGGAAAGTGATACTAGACAAGAAAACTTGAAAACCAAGGCTCTCGTTTCTAACAGTTCTTTGCATTGGATACCCATCCCATCGAATGATGAGGTAGTGAAACAGCCCAAACCAGAATCCAGAGAACACATACCATCTGTTGAGCCCCAGGTTGGAGAGGAGTGGGAGAAAGCTGCTCCCACCCCTCCTGCATTGCCTGGGGATTTGACAGCTGAGGAGGGTCTAGATCCTCTTGACAGCCTTACTTCACTCTGGACTGTGCCATCTCGAGGAGGCAGTGACAGCAATGGCAGTTACTGTCAACAGGTGGACATTGAAAAGCTGAAAATCAACGGAGACTCTGAAGCACTGAGTCCTCACGGTGAGTCCACGGATACAGCCTCTGACTTTGAAGGTCACCTCACGGAGGACAGCAGTGAGGCTGACACTAGAGAAGCTGCAGTGACAAAGGGATCTTCGGTGGACAAGGATGAGAAACCCAATTGGAACCAATCTGCCCCACTGTCCAAGGTGAATGGTGACATGCGTCTGGTTACAAGGACAGATGGGATGGTTGCTCCTCAGAGCTGGGTGTCTCGAGTATGTGCGGTCCGCCAAAAGATCCCAGATTCCCTACTGCTGGCCAGTACTGAGTACCAGCCAAGAGCCGTGTGCCTGTCCATGCCTGGGTCCTCAGTGGAGGCCACTAACCCACTTGTGATGCAGTTGCTGCAGGGTAGCTTGCCCCTAGAGAAGGTTCTTCCACCAGCCCACGATGACAGCATGTCAGAATCCCCACAAGTACCACTTACAAAAGACCAGAGCCATGGCTCGCTACGCATGGGATCTTTACATGGTCTTGGAAAAAACAGTGGCATGGTTGATGGAAGCAGCCCCAGTTCTTTAAGGGCTTTGAAGGAGCCTCTTCTGCCAGATAGCTGTGAAACAGGCACTGGTCTTGCCAGGATTGAGGCCACCCAGGCTCCTGGAGCACCCCAAAAGAATTGCAAGGCAGTCCCAAGTTTTGACTCCCTCCATCCAGTGACAAATCCCATTACATCCTCTAGGAAACTGGAAGAAATGGATTCCAAAGAGCAGTTCTCTTCCTTTAGTTGTGAAGATCAGAAGGAAGTCCGTGCTATGTCACAGGACAGTAATTCAAATGCTGCTCCAGGAAAGAGCCCAGGAGATCTTACTACCTCGAGAACACCTCGTTTCTCATCTCCAAATGTGATCTCCTTTGGTCCAGAGCAGACAGGTCGGGCCCTGGGTGATCAGAGCAATGTTACAGGCCAAGGGAAGAAGCTTTTTGGCTCTGGGAATGTGGCTGCAACCCTTCAGCGCCCCAGGCCTGCGGACCCGATGCCTCTTCCTGCTGAGATCCCTCCAGTTTTTCCCAGTGGGAAGTTGGGACCAAGCACAAACTCCATGTCTGGTGGGGTACAGACTCCAAGGGAAGACTGGGCTCCAAAGCCACATGCCTTTGTTGGCAGCGTCAAGAATGAGAAGACTTTTGTGGGGGGTCCTCTTAAGGCAAATGCCGAGAACAGGAAAGCTACTGGGCATAGTCCCCTGGAACTGGTGGGTCACTTGGAAGGGATGCCCTTTGTCATGGACTTGCCCTTCTGGAAATTACCCCGAGAGCCAGGGAAGGGGCTCAGTGAGCCTCTGGAGCCTTCTTCTCTCCCCTCCCAACTCAGCATCAAGCAGGCATTTTATGGGAAGCTTTCTAAACTCCAACTGAGTTCCACCAGCTTTAATTATTCCTCTAGCTCTCCCACCTTTCCCAAAGGCCTTGCTGGAAGTGTGGTGCAGCTGAGCCACAAAGCAAACTTTGGTGCGAGCCACAGTGCATCACTTTCCTTGCAAATGTTCACTGACAGCAGCACGGTGGAAAGCATCTCGCTCCAGTGTGCGTGCAGCCTGAAAGCCATGATCATGTGCCAAGGCTGCGGTGCGTTCTGTCACGATGACTGTATTGGACCCTCAAAGCTCTGTGTATTGTGCCTTGTGGTGAGATAAT'\n",
    "  \n",
    "    #Create a list of the codons they called missense variants at\n",
    "    ASXL1_variant_list = []\n",
    "    ASXL1_missense_codons =[]\n",
    "    ASXL1_missense_variant_DICT = {}\n",
    "\n",
    "    variant_codon = int(((re.findall('\\d+', variant)))[0])\n",
    "\n",
    "    codon0 = variant_codon-1\n",
    "    codon1 = variant_codon\n",
    "    variant = variant\n",
    "    ASXL1_missense_codons.append(codon0) #List of codon numbers that they called variants at\n",
    "    ASXL1_missense_variant_DICT[row_count]=(codon1, variant) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    ASXL1_missense_codons_DICT = {}\n",
    "    calls_ASXL1_missense = {}\n",
    "\n",
    "    for i in ASXL1_missense_codons: #Iterates over the list of codon numbers\n",
    "        ASXL1_missense_codons_DICT[i+1]= (ASXL1_codons[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "\n",
    "    for k,v in ASXL1_missense_codons_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_ASXL1_missense[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_ASXL1_missense_extra = {}\n",
    "\n",
    "    for k, v in calls_ASXL1_missense.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_ASXL1_missense_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_ASXL1_missense_extra2 = {}\n",
    "\n",
    "    for k, v in calls_ASXL1_missense_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_ASXL1_missense_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    ASXL1_position_in_list_dict_missense={}\n",
    "    for k, v in ASXL1_missense_variant_DICT.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_ASXL1_missense[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_ASXL1_missense[v[0]]\n",
    "        codon_seq=calls_ASXL1_missense[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_ASXL1_missense_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                ASXL1_position_in_list_dict_missense[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "\n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in ASXL1_position_in_list_dict_missense.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "    #     print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    ASXL1_bases_missense = {}\n",
    "    for k, v in ASXL1_position_in_list_dict_missense.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        ASXL1_bases_missense[k]=(v, base_changes)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    ASXL1_bases_missense2 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_missense.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        ASXL1_bases_missense2[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    ASXL1_bases_missense3 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_missense2.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        ASXL1_bases_missense3[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    ASXL1_bases_missense4 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_missense3.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_ASXL1:\n",
    "    #                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check_ASXL1(i)\n",
    "                sites_trio_missense.append([adapted_trio])\n",
    "                ASXL1_bases_missense4[k] = (v, sites_trio_missense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_missense = [ASXL1_exons_extra_bases[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_missense.append(site_trio_missense)\n",
    "                    ASXL1_bases_missense4[k] = (v, sites_trio_missense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    ASXL1_sites_missense5 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in ASXL1_bases_missense4.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        ASXL1_sites_missense5[a]=info\n",
    "        a=a+1\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    ASXL1_sites_missense6={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in ASXL1_sites_missense5.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (original_base,new_base)\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        ASXL1_sites_missense6[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    ASXL1_mutation_rates_missense = {}\n",
    "    for k, v in ASXL1_sites_missense6.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            ASXL1_mutation_rates_missense[k] = (v, mutation_rate)\n",
    "\n",
    "    ASXL1_mutation_rate_missense_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in ASXL1_mutation_rates_missense.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        ASXL1_mutation_rate_missense_sum.append(mutation_rate_value)\n",
    "\n",
    "    ASXL1_Summed_mutation_rate_missense_added = np.sum(ASXL1_mutation_rate_missense_sum)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return ASXL1_Summed_mutation_rate_missense_added"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TP53"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_variants_TP53(variant):\n",
    "    intron_exon_boundary_caution_base_TP53 = [74, 75, 96, 375, 672, 782, 919, 993, 1100]\n",
    "\n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    def intron_exon_check(position_in_gene_number):\n",
    "        if position_in_gene_number == 74:\n",
    "            gene_pos = 74\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 75:\n",
    "            gene_pos = 75\n",
    "            first_base = 'G'\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = TP53_exons_extra_bases[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 96:\n",
    "            gene_pos = 96\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 375:\n",
    "            gene_pos = 375\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 672:\n",
    "            gene_pos = 672\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 782:\n",
    "            gene_pos = 782\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 919:\n",
    "            gene_pos = 919\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 993:\n",
    "            gene_pos = 993\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1100:\n",
    "            gene_pos = 1100\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #TP53 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    TP53_exons = 'ATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGACATAGTGTGGTGGTGCCCTATGAGCCGCCTGAGGTTGGCTCTGACTGTACCACCATCCACTACAACTACATGTGTAACAGTTCCTGCATGGGCGGCATGAACCGGAGGCCCATCCTCACCATCATCACACTGGAAGACTCCAGTGGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGAGGAAGAGAATCTCCGCAAGAAAGGGGAGCCTCACCACGAGCTGCCCCCAGGGAGCACTAAGCGAGCACTGCCCAACAACACCAGCTCCTCTCCCCAGCCAAAGAAGAAACCACTGGATGGAGAATATTTCACCCTTCAGATCCGTGGGCGTGAGCGCTTCGAGATGTTCCGAGAGCTGAATGAGGCCTTGGAACTCAAGGATGCCCAGGCTGGGAAGGAGCCAGGGGGGAGCAGGGCTCACTCCAGCCACCTGAAGTCCAAAAAGGGTCAGTCTACCTCCCGCCATAAAAAACTCATGTTCAAGACAGAAGGGCCTGACTCAGACTGA'\n",
    "    \n",
    "    #Split the sequence into codons \n",
    "    TP53_codons = [TP53_exons[i:i+3] for i in range(0, len(TP53_exons), 3)]\n",
    "\n",
    "    #TP53 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    TP53_exons_extra_bases = 'CATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGACATAGTGTGGTGGTGCCCTATGAGCCGCCTGAGGTTGGCTCTGACTGTACCACCATCCACTACAACTACATGTGTAACAGTTCCTGCATGGGCGGCATGAACCGGAGGCCCATCCTCACCATCATCACACTGGAAGACTCCAGTGGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGAGGAAGAGAATCTCCGCAAGAAAGGGGAGCCTCACCACGAGCTGCCCCCAGGGAGCACTAAGCGAGCACTGCCCAACAACACCAGCTCCTCTCCCCAGCCAAAGAAGAAACCACTGGATGGAGAATATTTCACCCTTCAGATCCGTGGGCGTGAGCGCTTCGAGATGTTCCGAGAGCTGAATGAGGCCTTGGAACTCAAGGATGCCCAGGCTGGGAAGGAGCCAGGGGGGAGCAGGGCTCACTCCAGCCACCTGAAGTCCAAAAAGGGTCAGTCTACCTCCCGCCATAAAAAACTCATGTTCAAGACAGAAGGGCCTGACTCAGACTGAC'\n",
    "    \n",
    "    #Create a list of the codons they called missense variants at (from COSMIC)\n",
    "    TP53_variant_list = []\n",
    "    TP53_missense_codons =[]\n",
    "    TP53_missense_variant_DICT = {}\n",
    "\n",
    "    variant_codon = int(((re.findall('\\d+', variant)))[0])\n",
    "\n",
    "    codon0 = variant_codon-1\n",
    "    codon1 = variant_codon\n",
    "    variant = variant\n",
    "    TP53_missense_codons.append(codon0) #List of codon numbers that they called variants at\n",
    "    TP53_missense_variant_DICT[row_count]=(codon1, variant) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    TP53_missense_codons_DICT = {}\n",
    "    calls_TP53_missense = {}\n",
    "\n",
    "    for i in TP53_missense_codons: #Iterates over the list of codon numbers\n",
    "        TP53_missense_codons_DICT[i+1]= (TP53_codons[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "\n",
    "    for k,v in TP53_missense_codons_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_TP53_missense[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_TP53_missense_extra = {}\n",
    "\n",
    "    for k, v in calls_TP53_missense.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_TP53_missense_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_TP53_missense_extra2 = {}\n",
    "\n",
    "    for k, v in calls_TP53_missense_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_TP53_missense_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    TP53_position_in_list_dict_missense={}\n",
    "    for k, v in TP53_missense_variant_DICT.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_TP53_missense[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_TP53_missense[v[0]]\n",
    "        codon_seq=calls_TP53_missense[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_TP53_missense_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                TP53_position_in_list_dict_missense[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "\n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in TP53_position_in_list_dict_missense.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "    #     print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TP53_bases_missense = {}\n",
    "    for k, v in TP53_position_in_list_dict_missense.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TP53_bases_missense[k]=(v, base_changes)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TP53_bases_missense2 = {}\n",
    "\n",
    "    for k, v in TP53_bases_missense.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TP53_bases_missense2[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TP53_bases_missense3 = {}\n",
    "\n",
    "    for k, v in TP53_bases_missense2.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TP53_bases_missense3[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TP53_bases_missense4 = {}\n",
    "\n",
    "    for k, v in TP53_bases_missense3.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_TP53:\n",
    "    #                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check(i)\n",
    "                sites_trio_missense.append([adapted_trio])\n",
    "                TP53_bases_missense4[k] = (v, sites_trio_missense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_missense = [TP53_exons_extra_bases[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_missense.append(site_trio_missense)\n",
    "                    TP53_bases_missense4[k] = (v, sites_trio_missense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TP53_sites_missense5 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TP53_bases_missense4.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        TP53_sites_missense5[a]=info\n",
    "        a=a+1\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TP53_sites_missense6={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TP53_sites_missense5.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (original_base,new_base)\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TP53_sites_missense6[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    TP53_mutation_rates_missense = {}\n",
    "    for k, v in TP53_sites_missense6.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            TP53_mutation_rates_missense[k] = (v, mutation_rate)\n",
    "\n",
    "    TP53_mutation_rate_missense_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TP53_mutation_rates_missense.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        TP53_mutation_rate_missense_sum.append(mutation_rate_value)\n",
    "\n",
    "    TP53_Summed_mutation_rate_missense_added = np.sum(TP53_mutation_rate_missense_sum)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return TP53_Summed_mutation_rate_missense_added"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculating mutation rates from a .csv file of variants"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DNMT3A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_list_nonsense_variants_DNMT3A_NP072046(csv_file_variants):\n",
    "   \n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "\n",
    "    intron_exon_boundary_caution_base_DNMT3A_NP072046 = [1279, 1429, 1474, 1554, 1852, 2083, 2082, 2173, 2322, 2478]\n",
    "   \n",
    "    def intron_exon_check_DNMT3A_NP072046(position_in_gene_number):\n",
    "        if position_in_gene_number == 1279:\n",
    "            gene_pos = 1279\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1429:\n",
    "            gene_pos = 1429\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1474:\n",
    "            gene_pos = 1474\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1554:\n",
    "            gene_pos = 1554\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1852:\n",
    "            gene_pos = 1852\n",
    "            first_base = 'G'\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2083:\n",
    "            gene_pos = 2083\n",
    "            first_base = 'G'\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2082:\n",
    "            gene_pos = 2082\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2173:\n",
    "            gene_pos = 2173\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2322:\n",
    "            gene_pos = 2322\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2478:\n",
    "            gene_pos = 2478\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #DNMT3A transcript used:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000119772;r=2:25455845-25564754;t=ENST00000321117\n",
    "    DNMT3A_exons_NP_072046 = 'ATGCCCGCCATGCCCTCCAGCGGCCCCGGGGACACCAGCAGCTCTGCTGCGGAGCGGGAGGAGGACCGAAAGGACGGAGAGGAGCAGGAGGAGCCGCGTGGCAAGGAGGAGCGCCAAGAGCCCAGCACCACGGCACGGAAGGTGGGGCGGCCTGGGAGGAAGCGCAAGCACCCCCCGGTGGAAAGCGGTGACACGCCAAAGGACCCTGCGGTGATCTCCAAGTCCCCATCCATGGCCCAGGACTCAGGCGCCTCAGAGCTATTACCCAATGGGGACTTGGAGAAGCGGAGTGAGCCCCAGCCAGAGGAGGGGAGCCCTGCTGGGGGGCAGAAGGGCGGGGCCCCAGCAGAGGGAGAGGGTGCAGCTGAGACCCTGCCTGAAGCCTCAAGAGCAGTGGAAAATGGCTGCTGCACCCCCAAGGAGGGCCGAGGAGCCCCTGCAGAAGCGGGCAAAGAACAGAAGGAGACCAACATCGAATCCATGAAAATGGAGGGCTCCCGGGGCCGGCTGCGGGGTGGCTTGGGCTGGGAGTCCAGCCTCCGTCAGCGGCCCATGCCGAGGCTCACCTTCCAGGCGGGGGACCCCTACTACATCAGCAAGCGCAAGCGGGACGAGTGGCTGGCACGCTGGAAAAGGGAGGCTGAGAAGAAAGCCAAGGTCATTGCAGGAATGAATGCTGTGGAAGAAAACCAGGGGCCCGGGGAGTCTCAGAAGGTGGAGGAGGCCAGCCCTCCTGCTGTGCAGCAGCCCACTGACCCCGCATCCCCCACTGTGGCTACCACGCCTGAGCCCGTGGGGTCCGATGCTGGGGACAAGAATGCCACCAAAGCAGGCGATGACGAGCCAGAGTACGAGGACGGCCGGGGCTTTGGCATTGGGGAGCTGGTGTGGGGGAAACTGCGGGGCTTCTCCTGGTGGCCAGGCCGCATTGTGTCTTGGTGGATGACGGGCCGGAGCCGAGCAGCTGAAGGCACCCGCTGGGTCATGTGGTTCGGAGACGGCAAATTCTCAGTGGTGTGTGTTGAGAAGCTGATGCCGCTGAGCTCGTTTTGCAGTGCGTTCCACCAGGCCACGTACAACAAGCAGCCCATGTACCGCAAAGCCATCTACGAGGTCCTGCAGGTGGCCAGCAGCCGCGCGGGGAAGCTGTTCCCGGTGTGCCACGACAGCGATGAGAGTGACACTGCCAAGGCCGTGGAGGTGCAGAACAAGCCCATGATTGAATGGGCCCTGGGGGGCTTCCAGCCTTCTGGCCCTAAGGGCCTGGAGCCACCAGAAGAAGAGAAGAATCCCTACAAAGAAGTGTACACGGACATGTGGGTGGAACCTGAGGCAGCTGCCTACGCACCACCTCCACCAGCCAAAAAGCCCCGGAAGAGCACAGCGGAGAAGCCCAAGGTCAAGGAGATTATTGATGAGCGCACAAGAGAGCGGCTGGTGTACGAGGTGCGGCAGAAGTGCCGGAACATTGAGGACATCTGCATCTCCTGTGGGAGCCTCAATGTTACCCTGGAACACCCCCTCTTCGTTGGAGGAATGTGCCAAAACTGCAAGAACTGCTTTCTGGAGTGTGCGTACCAGTACGACGACGACGGCTACCAGTCCTACTGCACCATCTGCTGTGGGGGCCGTGAGGTGCTCATGTGCGGAAACAACAACTGCTGCAGGTGCTTTTGCGTGGAGTGTGTGGACCTCTTGGTGGGGCCGGGGGCTGCCCAGGCAGCCATTAAGGAAGACCCCTGGAACTGCTACATGTGCGGGCACAAGGGTACCTACGGGCTGCTGCGGCGGCGAGAGGACTGGCCCTCCCGGCTCCAGATGTTCTTCGCTAATAACCACGACCAGGAATTTGACCCTCCAAAGGTTTACCCACCTGTCCCAGCTGAGAAGAGGAAGCCCATCCGGGTGCTGTCTCTCTTTGATGGAATCGCTACAGGGCTCCTGGTGCTGAAGGACTTGGGCATTCAGGTGGACCGCTACATTGCCTCGGAGGTGTGTGAGGACTCCATCACGGTGGGCATGGTGCGGCACCAGGGGAAGATCATGTACGTCGGGGACGTCCGCAGCGTCACACAGAAGCATATCCAGGAGTGGGGCCCATTCGATCTGGTGATTGGGGGCAGTCCCTGCAATGACCTCTCCATCGTCAACCCTGCTCGCAAGGGCCTCTACGAGGGCACTGGCCGGCTCTTCTTTGAGTTCTACCGCCTCCTGCATGATGCGCGGCCCAAGGAGGGAGATGATCGCCCCTTCTTCTGGCTCTTTGAGAATGTGGTGGCCATGGGCGTTAGTGACAAGAGGGACATCTCGCGATTTCTCGAGTCCAACCCTGTGATGATTGATGCCAAAGAAGTGTCAGCTGCACACAGGGCCCGCTACTTCTGGGGTAACCTTCCCGGTATGAACAGGCCGTTGGCATCCACTGTGAATGATAAGCTGGAGCTGCAGGAGTGTCTGGAGCATGGCAGGATAGCCAAGTTCAGCAAAGTGAGGACCATTACTACGAGGTCAAACTCCATAAAGCAGGGCAAAGACCAGCATTTTCCTGTCTTCATGAATGAGAAAGAGGACATCTTATGGTGCACTGAAATGGAAAGGGTATTTGGTTTCCCAGTCCACTATACTGACGTCTCCAACATGAGCCGCTTGGCGAGGCAGAGACTGCTGGGCCGGTCATGGAGCGTGCCAGTCATCCGCCACCTCTTCGCTCCGCTGAAGGAGTATTTTGCGTGTGTGTAA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    DNMT3A_codons_NP_072046 = [DNMT3A_exons_NP_072046[i:i+3] for i in range(0, len(DNMT3A_exons_NP_072046), 3)]\n",
    "\n",
    "    #DNMT3A transcript used:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000119772;r=2:25455845-25564754;t=ENST00000321117\n",
    "    DNMT3A_exons_extra_bases_NP_072046 = 'GATGCCCGCCATGCCCTCCAGCGGCCCCGGGGACACCAGCAGCTCTGCTGCGGAGCGGGAGGAGGACCGAAAGGACGGAGAGGAGCAGGAGGAGCCGCGTGGCAAGGAGGAGCGCCAAGAGCCCAGCACCACGGCACGGAAGGTGGGGCGGCCTGGGAGGAAGCGCAAGCACCCCCCGGTGGAAAGCGGTGACACGCCAAAGGACCCTGCGGTGATCTCCAAGTCCCCATCCATGGCCCAGGACTCAGGCGCCTCAGAGCTATTACCCAATGGGGACTTGGAGAAGCGGAGTGAGCCCCAGCCAGAGGAGGGGAGCCCTGCTGGGGGGCAGAAGGGCGGGGCCCCAGCAGAGGGAGAGGGTGCAGCTGAGACCCTGCCTGAAGCCTCAAGAGCAGTGGAAAATGGCTGCTGCACCCCCAAGGAGGGCCGAGGAGCCCCTGCAGAAGCGGGCAAAGAACAGAAGGAGACCAACATCGAATCCATGAAAATGGAGGGCTCCCGGGGCCGGCTGCGGGGTGGCTTGGGCTGGGAGTCCAGCCTCCGTCAGCGGCCCATGCCGAGGCTCACCTTCCAGGCGGGGGACCCCTACTACATCAGCAAGCGCAAGCGGGACGAGTGGCTGGCACGCTGGAAAAGGGAGGCTGAGAAGAAAGCCAAGGTCATTGCAGGAATGAATGCTGTGGAAGAAAACCAGGGGCCCGGGGAGTCTCAGAAGGTGGAGGAGGCCAGCCCTCCTGCTGTGCAGCAGCCCACTGACCCCGCATCCCCCACTGTGGCTACCACGCCTGAGCCCGTGGGGTCCGATGCTGGGGACAAGAATGCCACCAAAGCAGGCGATGACGAGCCAGAGTACGAGGACGGCCGGGGCTTTGGCATTGGGGAGCTGGTGTGGGGGAAACTGCGGGGCTTCTCCTGGTGGCCAGGCCGCATTGTGTCTTGGTGGATGACGGGCCGGAGCCGAGCAGCTGAAGGCACCCGCTGGGTCATGTGGTTCGGAGACGGCAAATTCTCAGTGGTGTGTGTTGAGAAGCTGATGCCGCTGAGCTCGTTTTGCAGTGCGTTCCACCAGGCCACGTACAACAAGCAGCCCATGTACCGCAAAGCCATCTACGAGGTCCTGCAGGTGGCCAGCAGCCGCGCGGGGAAGCTGTTCCCGGTGTGCCACGACAGCGATGAGAGTGACACTGCCAAGGCCGTGGAGGTGCAGAACAAGCCCATGATTGAATGGGCCCTGGGGGGCTTCCAGCCTTCTGGCCCTAAGGGCCTGGAGCCACCAGAAGAAGAGAAGAATCCCTACAAAGAAGTGTACACGGACATGTGGGTGGAACCTGAGGCAGCTGCCTACGCACCACCTCCACCAGCCAAAAAGCCCCGGAAGAGCACAGCGGAGAAGCCCAAGGTCAAGGAGATTATTGATGAGCGCACAAGAGAGCGGCTGGTGTACGAGGTGCGGCAGAAGTGCCGGAACATTGAGGACATCTGCATCTCCTGTGGGAGCCTCAATGTTACCCTGGAACACCCCCTCTTCGTTGGAGGAATGTGCCAAAACTGCAAGAACTGCTTTCTGGAGTGTGCGTACCAGTACGACGACGACGGCTACCAGTCCTACTGCACCATCTGCTGTGGGGGCCGTGAGGTGCTCATGTGCGGAAACAACAACTGCTGCAGGTGCTTTTGCGTGGAGTGTGTGGACCTCTTGGTGGGGCCGGGGGCTGCCCAGGCAGCCATTAAGGAAGACCCCTGGAACTGCTACATGTGCGGGCACAAGGGTACCTACGGGCTGCTGCGGCGGCGAGAGGACTGGCCCTCCCGGCTCCAGATGTTCTTCGCTAATAACCACGACCAGGAATTTGACCCTCCAAAGGTTTACCCACCTGTCCCAGCTGAGAAGAGGAAGCCCATCCGGGTGCTGTCTCTCTTTGATGGAATCGCTACAGGGCTCCTGGTGCTGAAGGACTTGGGCATTCAGGTGGACCGCTACATTGCCTCGGAGGTGTGTGAGGACTCCATCACGGTGGGCATGGTGCGGCACCAGGGGAAGATCATGTACGTCGGGGACGTCCGCAGCGTCACACAGAAGCATATCCAGGAGTGGGGCCCATTCGATCTGGTGATTGGGGGCAGTCCCTGCAATGACCTCTCCATCGTCAACCCTGCTCGCAAGGGCCTCTACGAGGGCACTGGCCGGCTCTTCTTTGAGTTCTACCGCCTCCTGCATGATGCGCGGCCCAAGGAGGGAGATGATCGCCCCTTCTTCTGGCTCTTTGAGAATGTGGTGGCCATGGGCGTTAGTGACAAGAGGGACATCTCGCGATTTCTCGAGTCCAACCCTGTGATGATTGATGCCAAAGAAGTGTCAGCTGCACACAGGGCCCGCTACTTCTGGGGTAACCTTCCCGGTATGAACAGGCCGTTGGCATCCACTGTGAATGATAAGCTGGAGCTGCAGGAGTGTCTGGAGCATGGCAGGATAGCCAAGTTCAGCAAAGTGAGGACCATTACTACGAGGTCAAACTCCATAAAGCAGGGCAAAGACCAGCATTTTCCTGTCTTCATGAATGAGAAAGAGGACATCTTATGGTGCACTGAAATGGAAAGGGTATTTGGTTTCCCAGTCCACTATACTGACGTCTCCAACATGAGCCGCTTGGCGAGGCAGAGACTGCTGGGCCGGTCATGGAGCGTGCCAGTCATCCGCCACCTCTTCGCTCCGCTGAAGGAGTATTTTGCGTGTGTGTAAG'\n",
    "\n",
    "    #Create a list of the codons they called nonsense variants at\n",
    "    import csv\n",
    "    filename = csv_file_variants\n",
    "    with open(filename, 'r') as csvfile:\n",
    "        read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "        row_count=0\n",
    "\n",
    "        DNMT3A_nonsense_variant_list_NP_072046 = []\n",
    "        DNMT3A_nonsense_codons_NP_072046 =[]\n",
    "        DNMT3A_nonsense_variant_DICT_NP_072046 = {}\n",
    "\n",
    "        for row in read_reader: #each row contains aa number (e.g. 11) and variant (e.g. p.D11A)\n",
    "    #         print(row)\n",
    "            if row_count>0: \n",
    "                codon0_NP_072046 = int(row[0])-1\n",
    "                codon1_NP_072046 = int(row[0])\n",
    "                variant_NP_072046 = row[1]\n",
    "                DNMT3A_nonsense_codons_NP_072046.append(codon0_NP_072046) #List of codon numbers that they called variants at\n",
    "                DNMT3A_nonsense_variant_DICT_NP_072046[row_count]=(codon1_NP_072046, variant_NP_072046) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "            row_count=row_count+1\n",
    "            \n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    DNMT3A_nonsense_codons_NP_072046_DICT = {}\n",
    "    calls_DNMT3A_nonsense_NP_072046 = {}\n",
    "\n",
    "    for i in DNMT3A_nonsense_codons_NP_072046: #Iterates over the list of codon numbers\n",
    "        DNMT3A_nonsense_codons_NP_072046_DICT[i+1]= (DNMT3A_codons_NP_072046[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "        \n",
    "    for k,v in DNMT3A_nonsense_codons_NP_072046_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_DNMT3A_nonsense_NP_072046[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "        \n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_DNMT3A_nonsense_NP_072046_extra = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_nonsense_NP_072046.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_DNMT3A_nonsense_NP_072046_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_DNMT3A_nonsense_NP_072046_extra2 = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_nonsense_NP_072046_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_DNMT3A_nonsense_NP_072046_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "            \n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    DNMT3A_position_in_list_dict_nonsense_NP_072046={}\n",
    "    for k, v in DNMT3A_nonsense_variant_DICT_NP_072046.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_DNMT3A_nonsense_NP_072046[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_DNMT3A_nonsense_NP_072046[v[0]]\n",
    "        codon_seq=calls_DNMT3A_nonsense_NP_072046[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_DNMT3A_nonsense_NP_072046_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                DNMT3A_position_in_list_dict_nonsense_NP_072046[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "            \n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in DNMT3A_position_in_list_dict_nonsense_NP_072046.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "#     print('positions at which nonsense variants called = ', aa_positions_variants_called)\n",
    "                \n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    DNMT3A_bases_nonsense_NP_072046 = {}\n",
    "    for k, v in DNMT3A_position_in_list_dict_nonsense_NP_072046.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        DNMT3A_bases_nonsense_NP_072046[k]=(v, base_changes)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    DNMT3A_bases_nonsense2_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_nonsense_NP_072046.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        DNMT3A_bases_nonsense2_NP_072046[k] = (v, codon_positions)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    DNMT3A_bases_nonsense3_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_nonsense2_NP_072046.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        DNMT3A_bases_nonsense3_NP_072046[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    DNMT3A_bases_nonsense4_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_nonsense3_NP_072046.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_nonsense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_DNMT3A_NP072046:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check_DNMT3A_NP072046(i)\n",
    "                sites_trio_nonsense.append([adapted_trio])\n",
    "                DNMT3A_bases_nonsense4_NP_072046[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_nonsense = [DNMT3A_exons_extra_bases_NP_072046[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_nonsense.append(site_trio_nonsense)\n",
    "                    DNMT3A_bases_nonsense4_NP_072046[k] = (v, sites_trio_nonsense)\n",
    "                    \n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    DNMT3A_sites_nonsense5_NP_072046 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in DNMT3A_bases_nonsense4_NP_072046.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        DNMT3A_sites_nonsense5_NP_072046[a]=info\n",
    "        a=a+1\n",
    "#     print(DNMT3A_sites_nonsense5_NP_072046)\n",
    "        \n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    DNMT3A_sites_nonsense6_NP_072046={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in DNMT3A_sites_nonsense5_NP_072046.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "    \n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        DNMT3A_sites_nonsense6_NP_072046[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "        \n",
    "        \n",
    "    #Assign the mutation rates to each site...\n",
    "    DNMT3A_mutation_rates_nonsense_NP_072046 = {}\n",
    "    for k, v in DNMT3A_sites_nonsense6_NP_072046.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            DNMT3A_mutation_rates_nonsense_NP_072046[k] = (v, mutation_rate)\n",
    "\n",
    "    DNMT3A_mutation_rate_nonsense_NP_072046_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in DNMT3A_mutation_rates_nonsense_NP_072046.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        DNMT3A_mutation_rate_nonsense_NP_072046_sum.append(mutation_rate_value)\n",
    "\n",
    "    DNMT3A_Summed_mutation_rate_nonsense_NP_072046_added = np.sum(DNMT3A_mutation_rate_nonsense_NP_072046_sum)\n",
    "\n",
    "    print('Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study = ', DNMT3A_Summed_mutation_rate_nonsense_NP_072046_added)\n",
    "    \n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all #in format: column 1 = 882, column 2 = p.R882*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_list_missense_variants_DNMT3A_NP072046(csv_file_variants):\n",
    "    \n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base_DNMT3A_NP072046 = [1279, 1429, 1474, 1554, 1852, 2083, 2082, 2173, 2322, 2478]\n",
    "\n",
    "    def intron_exon_check_DNMT3A_NP072046(position_in_gene_number):\n",
    "        if position_in_gene_number == 1279:\n",
    "            gene_pos = 1279\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1429:\n",
    "            gene_pos = 1429\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1474:\n",
    "            gene_pos = 1474\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1554:\n",
    "            gene_pos = 1554\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1852:\n",
    "            gene_pos = 1852\n",
    "            first_base = 'G'\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2083:\n",
    "            gene_pos = 2083\n",
    "            first_base = 'G'\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2082:\n",
    "            gene_pos = 2082\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2173:\n",
    "            gene_pos = 2173\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2322:\n",
    "            gene_pos = 2322\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2478:\n",
    "            gene_pos = 2478\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #DNMT3A transcript used:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000119772;r=2:25455845-25564754;t=ENST00000321117\n",
    "    DNMT3A_exons_NP_072046 = 'ATGCCCGCCATGCCCTCCAGCGGCCCCGGGGACACCAGCAGCTCTGCTGCGGAGCGGGAGGAGGACCGAAAGGACGGAGAGGAGCAGGAGGAGCCGCGTGGCAAGGAGGAGCGCCAAGAGCCCAGCACCACGGCACGGAAGGTGGGGCGGCCTGGGAGGAAGCGCAAGCACCCCCCGGTGGAAAGCGGTGACACGCCAAAGGACCCTGCGGTGATCTCCAAGTCCCCATCCATGGCCCAGGACTCAGGCGCCTCAGAGCTATTACCCAATGGGGACTTGGAGAAGCGGAGTGAGCCCCAGCCAGAGGAGGGGAGCCCTGCTGGGGGGCAGAAGGGCGGGGCCCCAGCAGAGGGAGAGGGTGCAGCTGAGACCCTGCCTGAAGCCTCAAGAGCAGTGGAAAATGGCTGCTGCACCCCCAAGGAGGGCCGAGGAGCCCCTGCAGAAGCGGGCAAAGAACAGAAGGAGACCAACATCGAATCCATGAAAATGGAGGGCTCCCGGGGCCGGCTGCGGGGTGGCTTGGGCTGGGAGTCCAGCCTCCGTCAGCGGCCCATGCCGAGGCTCACCTTCCAGGCGGGGGACCCCTACTACATCAGCAAGCGCAAGCGGGACGAGTGGCTGGCACGCTGGAAAAGGGAGGCTGAGAAGAAAGCCAAGGTCATTGCAGGAATGAATGCTGTGGAAGAAAACCAGGGGCCCGGGGAGTCTCAGAAGGTGGAGGAGGCCAGCCCTCCTGCTGTGCAGCAGCCCACTGACCCCGCATCCCCCACTGTGGCTACCACGCCTGAGCCCGTGGGGTCCGATGCTGGGGACAAGAATGCCACCAAAGCAGGCGATGACGAGCCAGAGTACGAGGACGGCCGGGGCTTTGGCATTGGGGAGCTGGTGTGGGGGAAACTGCGGGGCTTCTCCTGGTGGCCAGGCCGCATTGTGTCTTGGTGGATGACGGGCCGGAGCCGAGCAGCTGAAGGCACCCGCTGGGTCATGTGGTTCGGAGACGGCAAATTCTCAGTGGTGTGTGTTGAGAAGCTGATGCCGCTGAGCTCGTTTTGCAGTGCGTTCCACCAGGCCACGTACAACAAGCAGCCCATGTACCGCAAAGCCATCTACGAGGTCCTGCAGGTGGCCAGCAGCCGCGCGGGGAAGCTGTTCCCGGTGTGCCACGACAGCGATGAGAGTGACACTGCCAAGGCCGTGGAGGTGCAGAACAAGCCCATGATTGAATGGGCCCTGGGGGGCTTCCAGCCTTCTGGCCCTAAGGGCCTGGAGCCACCAGAAGAAGAGAAGAATCCCTACAAAGAAGTGTACACGGACATGTGGGTGGAACCTGAGGCAGCTGCCTACGCACCACCTCCACCAGCCAAAAAGCCCCGGAAGAGCACAGCGGAGAAGCCCAAGGTCAAGGAGATTATTGATGAGCGCACAAGAGAGCGGCTGGTGTACGAGGTGCGGCAGAAGTGCCGGAACATTGAGGACATCTGCATCTCCTGTGGGAGCCTCAATGTTACCCTGGAACACCCCCTCTTCGTTGGAGGAATGTGCCAAAACTGCAAGAACTGCTTTCTGGAGTGTGCGTACCAGTACGACGACGACGGCTACCAGTCCTACTGCACCATCTGCTGTGGGGGCCGTGAGGTGCTCATGTGCGGAAACAACAACTGCTGCAGGTGCTTTTGCGTGGAGTGTGTGGACCTCTTGGTGGGGCCGGGGGCTGCCCAGGCAGCCATTAAGGAAGACCCCTGGAACTGCTACATGTGCGGGCACAAGGGTACCTACGGGCTGCTGCGGCGGCGAGAGGACTGGCCCTCCCGGCTCCAGATGTTCTTCGCTAATAACCACGACCAGGAATTTGACCCTCCAAAGGTTTACCCACCTGTCCCAGCTGAGAAGAGGAAGCCCATCCGGGTGCTGTCTCTCTTTGATGGAATCGCTACAGGGCTCCTGGTGCTGAAGGACTTGGGCATTCAGGTGGACCGCTACATTGCCTCGGAGGTGTGTGAGGACTCCATCACGGTGGGCATGGTGCGGCACCAGGGGAAGATCATGTACGTCGGGGACGTCCGCAGCGTCACACAGAAGCATATCCAGGAGTGGGGCCCATTCGATCTGGTGATTGGGGGCAGTCCCTGCAATGACCTCTCCATCGTCAACCCTGCTCGCAAGGGCCTCTACGAGGGCACTGGCCGGCTCTTCTTTGAGTTCTACCGCCTCCTGCATGATGCGCGGCCCAAGGAGGGAGATGATCGCCCCTTCTTCTGGCTCTTTGAGAATGTGGTGGCCATGGGCGTTAGTGACAAGAGGGACATCTCGCGATTTCTCGAGTCCAACCCTGTGATGATTGATGCCAAAGAAGTGTCAGCTGCACACAGGGCCCGCTACTTCTGGGGTAACCTTCCCGGTATGAACAGGCCGTTGGCATCCACTGTGAATGATAAGCTGGAGCTGCAGGAGTGTCTGGAGCATGGCAGGATAGCCAAGTTCAGCAAAGTGAGGACCATTACTACGAGGTCAAACTCCATAAAGCAGGGCAAAGACCAGCATTTTCCTGTCTTCATGAATGAGAAAGAGGACATCTTATGGTGCACTGAAATGGAAAGGGTATTTGGTTTCCCAGTCCACTATACTGACGTCTCCAACATGAGCCGCTTGGCGAGGCAGAGACTGCTGGGCCGGTCATGGAGCGTGCCAGTCATCCGCCACCTCTTCGCTCCGCTGAAGGAGTATTTTGCGTGTGTGTAA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    DNMT3A_codons_NP_072046 = [DNMT3A_exons_NP_072046[i:i+3] for i in range(0, len(DNMT3A_exons_NP_072046), 3)]\n",
    "\n",
    "    #DNMT3A transcript used:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000119772;r=2:25455845-25564754;t=ENST00000321117\n",
    "    DNMT3A_exons_extra_bases_NP_072046 = 'GATGCCCGCCATGCCCTCCAGCGGCCCCGGGGACACCAGCAGCTCTGCTGCGGAGCGGGAGGAGGACCGAAAGGACGGAGAGGAGCAGGAGGAGCCGCGTGGCAAGGAGGAGCGCCAAGAGCCCAGCACCACGGCACGGAAGGTGGGGCGGCCTGGGAGGAAGCGCAAGCACCCCCCGGTGGAAAGCGGTGACACGCCAAAGGACCCTGCGGTGATCTCCAAGTCCCCATCCATGGCCCAGGACTCAGGCGCCTCAGAGCTATTACCCAATGGGGACTTGGAGAAGCGGAGTGAGCCCCAGCCAGAGGAGGGGAGCCCTGCTGGGGGGCAGAAGGGCGGGGCCCCAGCAGAGGGAGAGGGTGCAGCTGAGACCCTGCCTGAAGCCTCAAGAGCAGTGGAAAATGGCTGCTGCACCCCCAAGGAGGGCCGAGGAGCCCCTGCAGAAGCGGGCAAAGAACAGAAGGAGACCAACATCGAATCCATGAAAATGGAGGGCTCCCGGGGCCGGCTGCGGGGTGGCTTGGGCTGGGAGTCCAGCCTCCGTCAGCGGCCCATGCCGAGGCTCACCTTCCAGGCGGGGGACCCCTACTACATCAGCAAGCGCAAGCGGGACGAGTGGCTGGCACGCTGGAAAAGGGAGGCTGAGAAGAAAGCCAAGGTCATTGCAGGAATGAATGCTGTGGAAGAAAACCAGGGGCCCGGGGAGTCTCAGAAGGTGGAGGAGGCCAGCCCTCCTGCTGTGCAGCAGCCCACTGACCCCGCATCCCCCACTGTGGCTACCACGCCTGAGCCCGTGGGGTCCGATGCTGGGGACAAGAATGCCACCAAAGCAGGCGATGACGAGCCAGAGTACGAGGACGGCCGGGGCTTTGGCATTGGGGAGCTGGTGTGGGGGAAACTGCGGGGCTTCTCCTGGTGGCCAGGCCGCATTGTGTCTTGGTGGATGACGGGCCGGAGCCGAGCAGCTGAAGGCACCCGCTGGGTCATGTGGTTCGGAGACGGCAAATTCTCAGTGGTGTGTGTTGAGAAGCTGATGCCGCTGAGCTCGTTTTGCAGTGCGTTCCACCAGGCCACGTACAACAAGCAGCCCATGTACCGCAAAGCCATCTACGAGGTCCTGCAGGTGGCCAGCAGCCGCGCGGGGAAGCTGTTCCCGGTGTGCCACGACAGCGATGAGAGTGACACTGCCAAGGCCGTGGAGGTGCAGAACAAGCCCATGATTGAATGGGCCCTGGGGGGCTTCCAGCCTTCTGGCCCTAAGGGCCTGGAGCCACCAGAAGAAGAGAAGAATCCCTACAAAGAAGTGTACACGGACATGTGGGTGGAACCTGAGGCAGCTGCCTACGCACCACCTCCACCAGCCAAAAAGCCCCGGAAGAGCACAGCGGAGAAGCCCAAGGTCAAGGAGATTATTGATGAGCGCACAAGAGAGCGGCTGGTGTACGAGGTGCGGCAGAAGTGCCGGAACATTGAGGACATCTGCATCTCCTGTGGGAGCCTCAATGTTACCCTGGAACACCCCCTCTTCGTTGGAGGAATGTGCCAAAACTGCAAGAACTGCTTTCTGGAGTGTGCGTACCAGTACGACGACGACGGCTACCAGTCCTACTGCACCATCTGCTGTGGGGGCCGTGAGGTGCTCATGTGCGGAAACAACAACTGCTGCAGGTGCTTTTGCGTGGAGTGTGTGGACCTCTTGGTGGGGCCGGGGGCTGCCCAGGCAGCCATTAAGGAAGACCCCTGGAACTGCTACATGTGCGGGCACAAGGGTACCTACGGGCTGCTGCGGCGGCGAGAGGACTGGCCCTCCCGGCTCCAGATGTTCTTCGCTAATAACCACGACCAGGAATTTGACCCTCCAAAGGTTTACCCACCTGTCCCAGCTGAGAAGAGGAAGCCCATCCGGGTGCTGTCTCTCTTTGATGGAATCGCTACAGGGCTCCTGGTGCTGAAGGACTTGGGCATTCAGGTGGACCGCTACATTGCCTCGGAGGTGTGTGAGGACTCCATCACGGTGGGCATGGTGCGGCACCAGGGGAAGATCATGTACGTCGGGGACGTCCGCAGCGTCACACAGAAGCATATCCAGGAGTGGGGCCCATTCGATCTGGTGATTGGGGGCAGTCCCTGCAATGACCTCTCCATCGTCAACCCTGCTCGCAAGGGCCTCTACGAGGGCACTGGCCGGCTCTTCTTTGAGTTCTACCGCCTCCTGCATGATGCGCGGCCCAAGGAGGGAGATGATCGCCCCTTCTTCTGGCTCTTTGAGAATGTGGTGGCCATGGGCGTTAGTGACAAGAGGGACATCTCGCGATTTCTCGAGTCCAACCCTGTGATGATTGATGCCAAAGAAGTGTCAGCTGCACACAGGGCCCGCTACTTCTGGGGTAACCTTCCCGGTATGAACAGGCCGTTGGCATCCACTGTGAATGATAAGCTGGAGCTGCAGGAGTGTCTGGAGCATGGCAGGATAGCCAAGTTCAGCAAAGTGAGGACCATTACTACGAGGTCAAACTCCATAAAGCAGGGCAAAGACCAGCATTTTCCTGTCTTCATGAATGAGAAAGAGGACATCTTATGGTGCACTGAAATGGAAAGGGTATTTGGTTTCCCAGTCCACTATACTGACGTCTCCAACATGAGCCGCTTGGCGAGGCAGAGACTGCTGGGCCGGTCATGGAGCGTGCCAGTCATCCGCCACCTCTTCGCTCCGCTGAAGGAGTATTTTGCGTGTGTGTAAG'\n",
    "\n",
    "    #Create a list of the codons they called missense variants at\n",
    "    import csv\n",
    "    filename = csv_file_variants\n",
    "    with open(filename, 'r') as csvfile:\n",
    "        read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "        row_count=0\n",
    "\n",
    "        DNMT3A_missense_variant_list_NP_072046 = []\n",
    "        DNMT3A_missense_codons_NP_072046 =[]\n",
    "        DNMT3A_missense_variant_DICT_NP_072046 = {}\n",
    "\n",
    "        for row in read_reader: #each row contains aa number (e.g. 11) and variant (e.g. p.D11A)\n",
    "    #         print(row)\n",
    "            if row_count>0: \n",
    "                codon0_NP_072046 = int(row[0])-1\n",
    "                codon1_NP_072046 = int(row[0])\n",
    "                variant_NP_072046 = row[1]\n",
    "                DNMT3A_missense_codons_NP_072046.append(codon0_NP_072046) #List of codon numbers that they called variants at\n",
    "                DNMT3A_missense_variant_DICT_NP_072046[row_count]=(codon1_NP_072046, variant_NP_072046) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "            row_count=row_count+1\n",
    "            \n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    DNMT3A_missense_codons_NP_072046_DICT = {}\n",
    "    calls_DNMT3A_missense_NP_072046 = {}\n",
    "\n",
    "    for i in DNMT3A_missense_codons_NP_072046: #Iterates over the list of codon numbers\n",
    "        DNMT3A_missense_codons_NP_072046_DICT[i+1]= (DNMT3A_codons_NP_072046[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "        \n",
    "    for k,v in DNMT3A_missense_codons_NP_072046_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_DNMT3A_missense_NP_072046[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "        \n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_DNMT3A_missense_NP_072046_extra = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_missense_NP_072046.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_DNMT3A_missense_NP_072046_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_DNMT3A_missense_NP_072046_extra2 = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_missense_NP_072046_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_DNMT3A_missense_NP_072046_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "            \n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    DNMT3A_position_in_list_dict_missense_NP_072046={}\n",
    "    for k, v in DNMT3A_missense_variant_DICT_NP_072046.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_DNMT3A_missense_NP_072046[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_DNMT3A_missense_NP_072046[v[0]]\n",
    "        codon_seq=calls_DNMT3A_missense_NP_072046[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_DNMT3A_missense_NP_072046_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                DNMT3A_position_in_list_dict_missense_NP_072046[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "            \n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in DNMT3A_position_in_list_dict_missense_NP_072046.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "#     print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "                \n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    DNMT3A_bases_missense_NP_072046 = {}\n",
    "    for k, v in DNMT3A_position_in_list_dict_missense_NP_072046.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        DNMT3A_bases_missense_NP_072046[k]=(v, base_changes)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    DNMT3A_bases_missense2_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_missense_NP_072046.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        DNMT3A_bases_missense2_NP_072046[k] = (v, codon_positions)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    DNMT3A_bases_missense3_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_missense2_NP_072046.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        DNMT3A_bases_missense3_NP_072046[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    DNMT3A_bases_missense4_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_missense3_NP_072046.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_DNMT3A_NP072046:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check_DNMT3A_NP072046(i)\n",
    "                sites_trio_missense.append([adapted_trio])\n",
    "                DNMT3A_bases_missense4_NP_072046[k] = (v, sites_trio_missense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_missense = [DNMT3A_exons_extra_bases_NP_072046[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_missense.append(site_trio_missense)\n",
    "                    DNMT3A_bases_missense4_NP_072046[k] = (v, sites_trio_missense)\n",
    "                    \n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    DNMT3A_sites_missense5_NP_072046 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in DNMT3A_bases_missense4_NP_072046.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        DNMT3A_sites_missense5_NP_072046[a]=info\n",
    "        a=a+1\n",
    "        \n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    DNMT3A_sites_missense6_NP_072046={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in DNMT3A_sites_missense5_NP_072046.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "    \n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        DNMT3A_sites_missense6_NP_072046[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "        \n",
    "        \n",
    "    #Assign the mutation rates to each site...\n",
    "    DNMT3A_mutation_rates_missense_NP_072046 = {}\n",
    "    for k, v in DNMT3A_sites_missense6_NP_072046.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            DNMT3A_mutation_rates_missense_NP_072046[k] = (v, mutation_rate)\n",
    "\n",
    "    DNMT3A_mutation_rate_missense_NP_072046_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in DNMT3A_mutation_rates_missense_NP_072046.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        DNMT3A_mutation_rate_missense_NP_072046_sum.append(mutation_rate_value)\n",
    "        \n",
    "    DNMT3A_Summed_mutation_rate_missense_NP_072046_added = np.sum(DNMT3A_mutation_rate_missense_NP_072046_sum)\n",
    "\n",
    "    print('Total summed mutation rate across DNMT3A missense_NP_072046 for this study = ', DNMT3A_Summed_mutation_rate_missense_NP_072046_added)\n",
    "    \n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all #in format: column 1 = 882, column 2 = p.R882H"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TET2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_list_nonsense_variants_TET2(csv_file_variants):\n",
    "\n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base_TET2_NP01120680 = [3409, 3804, 4044, 4537]\n",
    "   \n",
    "    def intron_exon_check_TET2_NP01120680(position_in_gene_number):\n",
    "        if position_in_gene_number == 3409:\n",
    "            gene_pos = 3409\n",
    "            first_base = TET2_exons_extra_bases_NP_01120680[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 3804:\n",
    "            gene_pos = 3804\n",
    "            first_base = 'G'\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = TET2_exons_extra_bases_NP_01120680[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 4044:\n",
    "            gene_pos = 4044\n",
    "            first_base = TET2_exons_extra_bases_NP_01120680[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 4537:\n",
    "            gene_pos = 4537\n",
    "            first_base = TET2_exons_extra_bases_NP_01120680[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #TET2 transcript (NP_01120680):\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000168769;r=4:106067032-106163928;t=ENST00000380013\n",
    "    TET2_exons_NP_01120680 = 'ATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGAGCAAATTATTGAAAAAGATGAAGGTCCTTTTTATACCCATCTAGGAGCAGGTCCTAATGTGGCAGCTATTAGAGAAATCATGGAAGAAAGGTTTGGACAGAAGGGTAAAGCTATTAGGATTGAAAGAGTCATCTATACTGGTAAAGAAGGCAAAAGTTCTCAGGGATGTCCTATTGCTAAGTGGGTGGTTCGCAGAAGCAGCAGTGAAGAGAAGCTACTGTGTTTGGTGCGGGAGCGAGCTGGCCACACCTGTGAGGCTGCAGTGATTGTGATTCTCATCCTGGTGTGGGAAGGAATCCCGCTGTCTCTGGCTGACAAACTCTACTCGGAGCTTACCGAGACGCTGAGGAAATACGGCACGCTCACCAATCGCCGGTGTGCCTTGAATGAAGAGAGAACTTGCGCCTGTCAGGGGCTGGATCCAGAAACCTGTGGTGCCTCCTTCTCTTTTGGTTGTTCATGGAGCATGTACTACAATGGATGTAAGTTTGCCAGAAGCAAGATCCCAAGGAAGTTTAAGCTGCTTGGGGATGACCCAAAAGAGGAAGAGAAACTGGAGTCTCATTTGCAAAACCTGTCCACTCTTATGGCACCAACATATAAGAAACTTGCACCTGATGCATATAATAATCAGATTGAATATGAACACAGAGCACCAGAGTGCCGTCTGGGTCTGAAGGAAGGCCGTCCATTCTCAGGGGTCACTGCATGTTTGGACTTCTGTGCTCATGCCCACAGAGACTTGCACAACATGCAGAATGGCAGCACATTGGTATGCACTCTCACTAGAGAAGACAATCGAGAATTTGGAGGAAAACCTGAGGATGAGCAGCTTCACGTTCTGCCTTTATACAAAGTCTCTGACGTGGATGAGTTTGGGAGTGTGGAAGCTCAGGAGGAGAAAAAACGGAGTGGTGCCATTCAGGTACTGAGTTCTTTTCGGCGAAAAGTCAGGATGTTAGCAGAGCCAGTCAAGACTTGCCGACAAAGGAAACTAGAAGCCAAGAAAGCTGCAGCTGAAAAGCTTTCCTCCCTGGAGAACAGCTCAAATAAAAATGAAAAGGAAAAGTCAGCCCCATCACGTACAAAACAAACTGAAAACGCAAGCCAGGCTAAACAGTTGGCAGAACTTTTGCGACTTTCAGGACCAGTCATGCAGCAGTCCCAGCAGCCCCAGCCTCTACAGAAGCAGCCACCACAGCCCCAGCAGCAGCAGAGACCCCAGCAGCAGCAGCCACATCACCCTCAGACAGAGTCTGTCAACTCTTATTCTGCTTCTGGATCCACCAATCCATACATGAGACGGCCCAATCCAGTTAGTCCTTATCCAAACTCTTCACACACTTCAGATATCTATGGAAGCACCAGCCCTATGAACTTCTATTCCACCTCATCTCAAGCTGCAGGTTCATATTTGAATTCTTCTAATCCCATGAACCCTTACCCTGGGCTTTTGAATCAGAATACCCAATATCCATCATATCAATGCAATGGAAACCTATCAGTGGACAACTGCTCCCCATATCTGGGTTCCTATTCTCCCCAGTCTCAGCCGATGGATCTGTATAGGTATCCAAGCCAAGACCCTCTGTCTAAGCTCAGTCTACCACCCATCCATACACTTTACCAGCCAAGGTTTGGAAATAGCCAGAGTTTTACATCTAAATACTTAGGTTATGGAAACCAAAATATGCAGGGAGATGGTTTCAGCAGTTGTACCATTAGACCAAATGTACATCATGTAGGGAAATTGCCTCCTTATCCCACTCATGAGATGGATGGCCACTTCATGGGAGCCACCTCTAGATTACCACCCAATCTGAGCAATCCAAACATGGACTATAAAAATGGTGAACATCATTCACCTTCTCACATAATCCATAACTACAGTGCAGCTCCGGGCATGTTCAACAGCTCTCTTCATGCCCTGCATCTCCAAAACAAGGAGAATGACATGCTTTCCCACACAGCTAATGGGTTATCAAAGATGCTTCCAGCTCTTAACCATGATAGAACTGCTTGTGTCCAAGGAGGCTTACACAAATTAAGTGATGCTAATGGTCAGGAAAAGCAGCCATTGGCACTAGTCCAGGGTGTGGCTTCTGGTGCAGAGGACAACGATGAGGTCTGGTCAGACAGCGAGCAGAGCTTTCTGGATCCTGACATTGGGGGAGTGGCCGTGGCTCCAACTCATGGGTCAATTCTCATTGAGTGTGCAAAGCGTGAGCTGCATGCCACAACCCCTTTAAAGAATCCCAATAGGAATCACCCCACCAGGATCTCCCTCGTCTTTTACCAGCATAAGAGCATGAATGAGCCAAAACATGGCTTGGCTCTTTGGGAAGCCAAAATGGCTGAAAAAGCCCGTGAGAAAGAGGAAGAGTGTGAAAAGTATGGCCCAGACTATGTGCCTCAGAAATCCCATGGCAAAAAAGTGAAACGGGAGCCTGCTGAGCCACATGAAACTTCAGAGCCCACTTACCTGCGTTTCATCAAGTCTCTTGCCGAAAGGACCATGTCCGTGACCACAGACTCCACAGTAACTACATCTCCATATGCCTTCACTCGGGTCACAGGGCCTTACAACAGATATATATGA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    TET2_codons_NP_01120680 = [TET2_exons_NP_01120680[i:i+3] for i in range(0, len(TET2_exons_NP_01120680), 3)]\n",
    "\n",
    "    #TET2 transcript (transcript variant 2 (TET2-001), NM_017628, NP_001120680):\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000168769;r=4:106067032-106163928;t=ENST00000380013\n",
    "    TET2_exons_extra_bases_NP_01120680 = 'GATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGAGCAAATTATTGAAAAAGATGAAGGTCCTTTTTATACCCATCTAGGAGCAGGTCCTAATGTGGCAGCTATTAGAGAAATCATGGAAGAAAGGTTTGGACAGAAGGGTAAAGCTATTAGGATTGAAAGAGTCATCTATACTGGTAAAGAAGGCAAAAGTTCTCAGGGATGTCCTATTGCTAAGTGGGTGGTTCGCAGAAGCAGCAGTGAAGAGAAGCTACTGTGTTTGGTGCGGGAGCGAGCTGGCCACACCTGTGAGGCTGCAGTGATTGTGATTCTCATCCTGGTGTGGGAAGGAATCCCGCTGTCTCTGGCTGACAAACTCTACTCGGAGCTTACCGAGACGCTGAGGAAATACGGCACGCTCACCAATCGCCGGTGTGCCTTGAATGAAGAGAGAACTTGCGCCTGTCAGGGGCTGGATCCAGAAACCTGTGGTGCCTCCTTCTCTTTTGGTTGTTCATGGAGCATGTACTACAATGGATGTAAGTTTGCCAGAAGCAAGATCCCAAGGAAGTTTAAGCTGCTTGGGGATGACCCAAAAGAGGAAGAGAAACTGGAGTCTCATTTGCAAAACCTGTCCACTCTTATGGCACCAACATATAAGAAACTTGCACCTGATGCATATAATAATCAGATTGAATATGAACACAGAGCACCAGAGTGCCGTCTGGGTCTGAAGGAAGGCCGTCCATTCTCAGGGGTCACTGCATGTTTGGACTTCTGTGCTCATGCCCACAGAGACTTGCACAACATGCAGAATGGCAGCACATTGGTATGCACTCTCACTAGAGAAGACAATCGAGAATTTGGAGGAAAACCTGAGGATGAGCAGCTTCACGTTCTGCCTTTATACAAAGTCTCTGACGTGGATGAGTTTGGGAGTGTGGAAGCTCAGGAGGAGAAAAAACGGAGTGGTGCCATTCAGGTACTGAGTTCTTTTCGGCGAAAAGTCAGGATGTTAGCAGAGCCAGTCAAGACTTGCCGACAAAGGAAACTAGAAGCCAAGAAAGCTGCAGCTGAAAAGCTTTCCTCCCTGGAGAACAGCTCAAATAAAAATGAAAAGGAAAAGTCAGCCCCATCACGTACAAAACAAACTGAAAACGCAAGCCAGGCTAAACAGTTGGCAGAACTTTTGCGACTTTCAGGACCAGTCATGCAGCAGTCCCAGCAGCCCCAGCCTCTACAGAAGCAGCCACCACAGCCCCAGCAGCAGCAGAGACCCCAGCAGCAGCAGCCACATCACCCTCAGACAGAGTCTGTCAACTCTTATTCTGCTTCTGGATCCACCAATCCATACATGAGACGGCCCAATCCAGTTAGTCCTTATCCAAACTCTTCACACACTTCAGATATCTATGGAAGCACCAGCCCTATGAACTTCTATTCCACCTCATCTCAAGCTGCAGGTTCATATTTGAATTCTTCTAATCCCATGAACCCTTACCCTGGGCTTTTGAATCAGAATACCCAATATCCATCATATCAATGCAATGGAAACCTATCAGTGGACAACTGCTCCCCATATCTGGGTTCCTATTCTCCCCAGTCTCAGCCGATGGATCTGTATAGGTATCCAAGCCAAGACCCTCTGTCTAAGCTCAGTCTACCACCCATCCATACACTTTACCAGCCAAGGTTTGGAAATAGCCAGAGTTTTACATCTAAATACTTAGGTTATGGAAACCAAAATATGCAGGGAGATGGTTTCAGCAGTTGTACCATTAGACCAAATGTACATCATGTAGGGAAATTGCCTCCTTATCCCACTCATGAGATGGATGGCCACTTCATGGGAGCCACCTCTAGATTACCACCCAATCTGAGCAATCCAAACATGGACTATAAAAATGGTGAACATCATTCACCTTCTCACATAATCCATAACTACAGTGCAGCTCCGGGCATGTTCAACAGCTCTCTTCATGCCCTGCATCTCCAAAACAAGGAGAATGACATGCTTTCCCACACAGCTAATGGGTTATCAAAGATGCTTCCAGCTCTTAACCATGATAGAACTGCTTGTGTCCAAGGAGGCTTACACAAATTAAGTGATGCTAATGGTCAGGAAAAGCAGCCATTGGCACTAGTCCAGGGTGTGGCTTCTGGTGCAGAGGACAACGATGAGGTCTGGTCAGACAGCGAGCAGAGCTTTCTGGATCCTGACATTGGGGGAGTGGCCGTGGCTCCAACTCATGGGTCAATTCTCATTGAGTGTGCAAAGCGTGAGCTGCATGCCACAACCCCTTTAAAGAATCCCAATAGGAATCACCCCACCAGGATCTCCCTCGTCTTTTACCAGCATAAGAGCATGAATGAGCCAAAACATGGCTTGGCTCTTTGGGAAGCCAAAATGGCTGAAAAAGCCCGTGAGAAAGAGGAAGAGTGTGAAAAGTATGGCCCAGACTATGTGCCTCAGAAATCCCATGGCAAAAAAGTGAAACGGGAGCCTGCTGAGCCACATGAAACTTCAGAGCCCACTTACCTGCGTTTCATCAAGTCTCTTGCCGAAAGGACCATGTCCGTGACCACAGACTCCACAGTAACTACATCTCCATATGCCTTCACTCGGGTCACAGGGCCTTACAACAGATATATATGAT'\n",
    "\n",
    "    #Create a list of the codons they called nonsense variants at\n",
    "    import csv\n",
    "    filename = csv_file_variants\n",
    "    with open(filename, 'r') as csvfile:\n",
    "        read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "        row_count=0\n",
    "\n",
    "        TET2_nonsense_variant_list_NP_01120680 = []\n",
    "        TET2_nonsense_codons_NP_01120680 =[]\n",
    "        TET2_nonsense_variant_DICT_NP_01120680 = {}\n",
    "\n",
    "        for row in read_reader: #each row contains aa number (e.g. 11) and variant (e.g. p.D11A)\n",
    "    #         print(row)\n",
    "            if row_count>0: \n",
    "                codon0_NP_01120680 = int(row[0])-1\n",
    "                codon1_NP_01120680 = int(row[0])\n",
    "                variant_NP_01120680 = row[1]\n",
    "                TET2_nonsense_codons_NP_01120680.append(codon0_NP_01120680) #List of codon numbers that they called variants at\n",
    "                TET2_nonsense_variant_DICT_NP_01120680[row_count]=(codon1_NP_01120680, variant_NP_01120680) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "            row_count=row_count+1\n",
    "            \n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    TET2_nonsense_codons_NP_01120680_DICT = {}\n",
    "    calls_TET2_nonsense_NP_01120680 = {}\n",
    "\n",
    "    for i in TET2_nonsense_codons_NP_01120680: #Iterates over the list of codon numbers\n",
    "        TET2_nonsense_codons_NP_01120680_DICT[i+1]= (TET2_codons_NP_01120680[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "        \n",
    "    for k,v in TET2_nonsense_codons_NP_01120680_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_TET2_nonsense_NP_01120680[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "        \n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_TET2_nonsense_NP_01120680_extra = {}\n",
    "\n",
    "    for k, v in calls_TET2_nonsense_NP_01120680.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_TET2_nonsense_NP_01120680_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_TET2_nonsense_NP_01120680_extra2 = {}\n",
    "\n",
    "    for k, v in calls_TET2_nonsense_NP_01120680_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_TET2_nonsense_NP_01120680_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "            \n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    TET2_position_in_list_dict_nonsense_NP_01120680={}\n",
    "    for k, v in TET2_nonsense_variant_DICT_NP_01120680.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_TET2_nonsense_NP_01120680[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_TET2_nonsense_NP_01120680[v[0]]\n",
    "        codon_seq=calls_TET2_nonsense_NP_01120680[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_TET2_nonsense_NP_01120680_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                TET2_position_in_list_dict_nonsense_NP_01120680[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "            \n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in TET2_position_in_list_dict_nonsense_NP_01120680.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "#     print('positions at which nonsense variants called = ', aa_positions_variants_called)\n",
    "                \n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TET2_bases_nonsense_NP_01120680 = {}\n",
    "    for k, v in TET2_position_in_list_dict_nonsense_NP_01120680.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TET2_bases_nonsense_NP_01120680[k]=(v, base_changes)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TET2_bases_nonsense2_NP_01120680 = {}\n",
    "\n",
    "    for k, v in TET2_bases_nonsense_NP_01120680.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TET2_bases_nonsense2_NP_01120680[k] = (v, codon_positions)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TET2_bases_nonsense3_NP_01120680 = {}\n",
    "\n",
    "    for k, v in TET2_bases_nonsense2_NP_01120680.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TET2_bases_nonsense3_NP_01120680[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TET2_bases_nonsense4_NP_01120680 = {}\n",
    "\n",
    "    for k, v in TET2_bases_nonsense3_NP_01120680.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_nonsense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_TET2_NP01120680:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check_TET2_NP01120680(i)\n",
    "                sites_trio_nonsense.append([adapted_trio])\n",
    "                TET2_bases_nonsense4_NP_01120680[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_nonsense = [TET2_exons_extra_bases_NP_01120680[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_nonsense.append(site_trio_nonsense)\n",
    "                    TET2_bases_nonsense4_NP_01120680[k] = (v, sites_trio_nonsense)\n",
    "                    \n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TET2_sites_nonsense5_NP_01120680 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TET2_bases_nonsense4_NP_01120680.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        TET2_sites_nonsense5_NP_01120680[a]=info\n",
    "        a=a+1\n",
    "#     print(TET2_sites_nonsense5_NP_01120680)\n",
    "        \n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TET2_sites_nonsense6_NP_01120680={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TET2_sites_nonsense5_NP_01120680.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "    \n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TET2_sites_nonsense6_NP_01120680[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "        \n",
    "        \n",
    "    #Assign the mutation rates to each site...\n",
    "    TET2_mutation_rates_nonsense_NP_01120680 = {}\n",
    "    for k, v in TET2_sites_nonsense6_NP_01120680.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            TET2_mutation_rates_nonsense_NP_01120680[k] = (v, mutation_rate)\n",
    "\n",
    "    TET2_mutation_rate_nonsense_NP_01120680_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TET2_mutation_rates_nonsense_NP_01120680.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        TET2_mutation_rate_nonsense_NP_01120680_sum.append(mutation_rate_value)\n",
    "        \n",
    "    TET2_Summed_mutation_rate_nonsense_NP_01120680_added = np.sum(TET2_mutation_rate_nonsense_NP_01120680_sum)\n",
    "\n",
    "    print('Total summed mutation rate across TET2 nonsense_NP_01120680 for this study = ', TET2_Summed_mutation_rate_nonsense_NP_01120680_added)\n",
    "    \n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all #in format: column 1 = 882, column 2 = p.R882* #2002 aa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_list_missense_variants_TET2(csv_file_variants):\n",
    "    \n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base_TET2_NP01120680 = [3409, 3804, 4044, 4537]\n",
    "   \n",
    "    def intron_exon_check_TET2_NP01120680(position_in_gene_number):\n",
    "        if position_in_gene_number == 3409:\n",
    "            gene_pos = 3409\n",
    "            first_base = TET2_exons_extra_bases_NP_01120680[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 3804:\n",
    "            gene_pos = 3804\n",
    "            first_base = 'G'\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = TET2_exons_extra_bases_NP_01120680[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 4044:\n",
    "            gene_pos = 4044\n",
    "            first_base = TET2_exons_extra_bases_NP_01120680[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 4537:\n",
    "            gene_pos = 4537\n",
    "            first_base = TET2_exons_extra_bases_NP_01120680[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #TET2 transcript (NP_01120680):\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000168769;r=4:106067032-106163928;t=ENST00000380013\n",
    "    TET2_exons_NP_01120680 = 'ATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGAGCAAATTATTGAAAAAGATGAAGGTCCTTTTTATACCCATCTAGGAGCAGGTCCTAATGTGGCAGCTATTAGAGAAATCATGGAAGAAAGGTTTGGACAGAAGGGTAAAGCTATTAGGATTGAAAGAGTCATCTATACTGGTAAAGAAGGCAAAAGTTCTCAGGGATGTCCTATTGCTAAGTGGGTGGTTCGCAGAAGCAGCAGTGAAGAGAAGCTACTGTGTTTGGTGCGGGAGCGAGCTGGCCACACCTGTGAGGCTGCAGTGATTGTGATTCTCATCCTGGTGTGGGAAGGAATCCCGCTGTCTCTGGCTGACAAACTCTACTCGGAGCTTACCGAGACGCTGAGGAAATACGGCACGCTCACCAATCGCCGGTGTGCCTTGAATGAAGAGAGAACTTGCGCCTGTCAGGGGCTGGATCCAGAAACCTGTGGTGCCTCCTTCTCTTTTGGTTGTTCATGGAGCATGTACTACAATGGATGTAAGTTTGCCAGAAGCAAGATCCCAAGGAAGTTTAAGCTGCTTGGGGATGACCCAAAAGAGGAAGAGAAACTGGAGTCTCATTTGCAAAACCTGTCCACTCTTATGGCACCAACATATAAGAAACTTGCACCTGATGCATATAATAATCAGATTGAATATGAACACAGAGCACCAGAGTGCCGTCTGGGTCTGAAGGAAGGCCGTCCATTCTCAGGGGTCACTGCATGTTTGGACTTCTGTGCTCATGCCCACAGAGACTTGCACAACATGCAGAATGGCAGCACATTGGTATGCACTCTCACTAGAGAAGACAATCGAGAATTTGGAGGAAAACCTGAGGATGAGCAGCTTCACGTTCTGCCTTTATACAAAGTCTCTGACGTGGATGAGTTTGGGAGTGTGGAAGCTCAGGAGGAGAAAAAACGGAGTGGTGCCATTCAGGTACTGAGTTCTTTTCGGCGAAAAGTCAGGATGTTAGCAGAGCCAGTCAAGACTTGCCGACAAAGGAAACTAGAAGCCAAGAAAGCTGCAGCTGAAAAGCTTTCCTCCCTGGAGAACAGCTCAAATAAAAATGAAAAGGAAAAGTCAGCCCCATCACGTACAAAACAAACTGAAAACGCAAGCCAGGCTAAACAGTTGGCAGAACTTTTGCGACTTTCAGGACCAGTCATGCAGCAGTCCCAGCAGCCCCAGCCTCTACAGAAGCAGCCACCACAGCCCCAGCAGCAGCAGAGACCCCAGCAGCAGCAGCCACATCACCCTCAGACAGAGTCTGTCAACTCTTATTCTGCTTCTGGATCCACCAATCCATACATGAGACGGCCCAATCCAGTTAGTCCTTATCCAAACTCTTCACACACTTCAGATATCTATGGAAGCACCAGCCCTATGAACTTCTATTCCACCTCATCTCAAGCTGCAGGTTCATATTTGAATTCTTCTAATCCCATGAACCCTTACCCTGGGCTTTTGAATCAGAATACCCAATATCCATCATATCAATGCAATGGAAACCTATCAGTGGACAACTGCTCCCCATATCTGGGTTCCTATTCTCCCCAGTCTCAGCCGATGGATCTGTATAGGTATCCAAGCCAAGACCCTCTGTCTAAGCTCAGTCTACCACCCATCCATACACTTTACCAGCCAAGGTTTGGAAATAGCCAGAGTTTTACATCTAAATACTTAGGTTATGGAAACCAAAATATGCAGGGAGATGGTTTCAGCAGTTGTACCATTAGACCAAATGTACATCATGTAGGGAAATTGCCTCCTTATCCCACTCATGAGATGGATGGCCACTTCATGGGAGCCACCTCTAGATTACCACCCAATCTGAGCAATCCAAACATGGACTATAAAAATGGTGAACATCATTCACCTTCTCACATAATCCATAACTACAGTGCAGCTCCGGGCATGTTCAACAGCTCTCTTCATGCCCTGCATCTCCAAAACAAGGAGAATGACATGCTTTCCCACACAGCTAATGGGTTATCAAAGATGCTTCCAGCTCTTAACCATGATAGAACTGCTTGTGTCCAAGGAGGCTTACACAAATTAAGTGATGCTAATGGTCAGGAAAAGCAGCCATTGGCACTAGTCCAGGGTGTGGCTTCTGGTGCAGAGGACAACGATGAGGTCTGGTCAGACAGCGAGCAGAGCTTTCTGGATCCTGACATTGGGGGAGTGGCCGTGGCTCCAACTCATGGGTCAATTCTCATTGAGTGTGCAAAGCGTGAGCTGCATGCCACAACCCCTTTAAAGAATCCCAATAGGAATCACCCCACCAGGATCTCCCTCGTCTTTTACCAGCATAAGAGCATGAATGAGCCAAAACATGGCTTGGCTCTTTGGGAAGCCAAAATGGCTGAAAAAGCCCGTGAGAAAGAGGAAGAGTGTGAAAAGTATGGCCCAGACTATGTGCCTCAGAAATCCCATGGCAAAAAAGTGAAACGGGAGCCTGCTGAGCCACATGAAACTTCAGAGCCCACTTACCTGCGTTTCATCAAGTCTCTTGCCGAAAGGACCATGTCCGTGACCACAGACTCCACAGTAACTACATCTCCATATGCCTTCACTCGGGTCACAGGGCCTTACAACAGATATATATGA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    TET2_codons_NP_01120680 = [TET2_exons_NP_01120680[i:i+3] for i in range(0, len(TET2_exons_NP_01120680), 3)]\n",
    "\n",
    "    #TET2 transcript (transcript variant 2 (TET2-001), NM_017628, NP_001120680):\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000168769;r=4:106067032-106163928;t=ENST00000380013\n",
    "    TET2_exons_extra_bases_NP_01120680 = 'GATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGAGCAAATTATTGAAAAAGATGAAGGTCCTTTTTATACCCATCTAGGAGCAGGTCCTAATGTGGCAGCTATTAGAGAAATCATGGAAGAAAGGTTTGGACAGAAGGGTAAAGCTATTAGGATTGAAAGAGTCATCTATACTGGTAAAGAAGGCAAAAGTTCTCAGGGATGTCCTATTGCTAAGTGGGTGGTTCGCAGAAGCAGCAGTGAAGAGAAGCTACTGTGTTTGGTGCGGGAGCGAGCTGGCCACACCTGTGAGGCTGCAGTGATTGTGATTCTCATCCTGGTGTGGGAAGGAATCCCGCTGTCTCTGGCTGACAAACTCTACTCGGAGCTTACCGAGACGCTGAGGAAATACGGCACGCTCACCAATCGCCGGTGTGCCTTGAATGAAGAGAGAACTTGCGCCTGTCAGGGGCTGGATCCAGAAACCTGTGGTGCCTCCTTCTCTTTTGGTTGTTCATGGAGCATGTACTACAATGGATGTAAGTTTGCCAGAAGCAAGATCCCAAGGAAGTTTAAGCTGCTTGGGGATGACCCAAAAGAGGAAGAGAAACTGGAGTCTCATTTGCAAAACCTGTCCACTCTTATGGCACCAACATATAAGAAACTTGCACCTGATGCATATAATAATCAGATTGAATATGAACACAGAGCACCAGAGTGCCGTCTGGGTCTGAAGGAAGGCCGTCCATTCTCAGGGGTCACTGCATGTTTGGACTTCTGTGCTCATGCCCACAGAGACTTGCACAACATGCAGAATGGCAGCACATTGGTATGCACTCTCACTAGAGAAGACAATCGAGAATTTGGAGGAAAACCTGAGGATGAGCAGCTTCACGTTCTGCCTTTATACAAAGTCTCTGACGTGGATGAGTTTGGGAGTGTGGAAGCTCAGGAGGAGAAAAAACGGAGTGGTGCCATTCAGGTACTGAGTTCTTTTCGGCGAAAAGTCAGGATGTTAGCAGAGCCAGTCAAGACTTGCCGACAAAGGAAACTAGAAGCCAAGAAAGCTGCAGCTGAAAAGCTTTCCTCCCTGGAGAACAGCTCAAATAAAAATGAAAAGGAAAAGTCAGCCCCATCACGTACAAAACAAACTGAAAACGCAAGCCAGGCTAAACAGTTGGCAGAACTTTTGCGACTTTCAGGACCAGTCATGCAGCAGTCCCAGCAGCCCCAGCCTCTACAGAAGCAGCCACCACAGCCCCAGCAGCAGCAGAGACCCCAGCAGCAGCAGCCACATCACCCTCAGACAGAGTCTGTCAACTCTTATTCTGCTTCTGGATCCACCAATCCATACATGAGACGGCCCAATCCAGTTAGTCCTTATCCAAACTCTTCACACACTTCAGATATCTATGGAAGCACCAGCCCTATGAACTTCTATTCCACCTCATCTCAAGCTGCAGGTTCATATTTGAATTCTTCTAATCCCATGAACCCTTACCCTGGGCTTTTGAATCAGAATACCCAATATCCATCATATCAATGCAATGGAAACCTATCAGTGGACAACTGCTCCCCATATCTGGGTTCCTATTCTCCCCAGTCTCAGCCGATGGATCTGTATAGGTATCCAAGCCAAGACCCTCTGTCTAAGCTCAGTCTACCACCCATCCATACACTTTACCAGCCAAGGTTTGGAAATAGCCAGAGTTTTACATCTAAATACTTAGGTTATGGAAACCAAAATATGCAGGGAGATGGTTTCAGCAGTTGTACCATTAGACCAAATGTACATCATGTAGGGAAATTGCCTCCTTATCCCACTCATGAGATGGATGGCCACTTCATGGGAGCCACCTCTAGATTACCACCCAATCTGAGCAATCCAAACATGGACTATAAAAATGGTGAACATCATTCACCTTCTCACATAATCCATAACTACAGTGCAGCTCCGGGCATGTTCAACAGCTCTCTTCATGCCCTGCATCTCCAAAACAAGGAGAATGACATGCTTTCCCACACAGCTAATGGGTTATCAAAGATGCTTCCAGCTCTTAACCATGATAGAACTGCTTGTGTCCAAGGAGGCTTACACAAATTAAGTGATGCTAATGGTCAGGAAAAGCAGCCATTGGCACTAGTCCAGGGTGTGGCTTCTGGTGCAGAGGACAACGATGAGGTCTGGTCAGACAGCGAGCAGAGCTTTCTGGATCCTGACATTGGGGGAGTGGCCGTGGCTCCAACTCATGGGTCAATTCTCATTGAGTGTGCAAAGCGTGAGCTGCATGCCACAACCCCTTTAAAGAATCCCAATAGGAATCACCCCACCAGGATCTCCCTCGTCTTTTACCAGCATAAGAGCATGAATGAGCCAAAACATGGCTTGGCTCTTTGGGAAGCCAAAATGGCTGAAAAAGCCCGTGAGAAAGAGGAAGAGTGTGAAAAGTATGGCCCAGACTATGTGCCTCAGAAATCCCATGGCAAAAAAGTGAAACGGGAGCCTGCTGAGCCACATGAAACTTCAGAGCCCACTTACCTGCGTTTCATCAAGTCTCTTGCCGAAAGGACCATGTCCGTGACCACAGACTCCACAGTAACTACATCTCCATATGCCTTCACTCGGGTCACAGGGCCTTACAACAGATATATATGAT'\n",
    "\n",
    "\n",
    "    #Create a list of the codons they called missense variants at\n",
    "    import csv\n",
    "    filename = csv_file_variants\n",
    "    with open(filename, 'r') as csvfile:\n",
    "        read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "        row_count=0\n",
    "\n",
    "        TET2_missense_variant_list_NP_01120680 = []\n",
    "        TET2_missense_codons_NP_01120680 =[]\n",
    "        TET2_missense_variant_DICT_NP_01120680 = {}\n",
    "\n",
    "        for row in read_reader: #each row contains aa number (e.g. 11) and variant (e.g. p.D11A)\n",
    "    #         print(row)\n",
    "            if row_count>0: \n",
    "                codon0_NP_01120680 = int(row[0])-1\n",
    "                codon1_NP_01120680 = int(row[0])\n",
    "                variant_NP_01120680 = row[1]\n",
    "                TET2_missense_codons_NP_01120680.append(codon0_NP_01120680) #List of codon numbers that they called variants at\n",
    "                TET2_missense_variant_DICT_NP_01120680[row_count]=(codon1_NP_01120680, variant_NP_01120680) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "            row_count=row_count+1\n",
    "            \n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    TET2_missense_codons_NP_01120680_DICT = {}\n",
    "    calls_TET2_missense_NP_01120680 = {}\n",
    "\n",
    "    for i in TET2_missense_codons_NP_01120680: #Iterates over the list of codon numbers\n",
    "        TET2_missense_codons_NP_01120680_DICT[i+1]= (TET2_codons_NP_01120680[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "        \n",
    "    for k,v in TET2_missense_codons_NP_01120680_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_TET2_missense_NP_01120680[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "        \n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_TET2_missense_NP_01120680_extra = {}\n",
    "\n",
    "    for k, v in calls_TET2_missense_NP_01120680.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_TET2_missense_NP_01120680_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_TET2_missense_NP_01120680_extra2 = {}\n",
    "\n",
    "    for k, v in calls_TET2_missense_NP_01120680_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_TET2_missense_NP_01120680_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "            \n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    TET2_position_in_list_dict_missense_NP_01120680={}\n",
    "    for k, v in TET2_missense_variant_DICT_NP_01120680.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_TET2_missense_NP_01120680[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_TET2_missense_NP_01120680[v[0]]\n",
    "        codon_seq=calls_TET2_missense_NP_01120680[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_TET2_missense_NP_01120680_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                TET2_position_in_list_dict_missense_NP_01120680[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "            \n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in TET2_position_in_list_dict_missense_NP_01120680.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "#     print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "                \n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TET2_bases_missense_NP_01120680 = {}\n",
    "    for k, v in TET2_position_in_list_dict_missense_NP_01120680.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TET2_bases_missense_NP_01120680[k]=(v, base_changes)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TET2_bases_missense2_NP_01120680 = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense_NP_01120680.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TET2_bases_missense2_NP_01120680[k] = (v, codon_positions)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TET2_bases_missense3_NP_01120680 = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense2_NP_01120680.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TET2_bases_missense3_NP_01120680[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TET2_bases_missense4_NP_01120680 = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense3_NP_01120680.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_TET2_NP01120680:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check_TET2_NP01120680(i)\n",
    "                sites_trio_missense.append([adapted_trio])\n",
    "                TET2_bases_missense4_NP_01120680[k] = (v, sites_trio_missense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_missense = [TET2_exons_extra_bases_NP_01120680[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_missense.append(site_trio_missense)\n",
    "                    TET2_bases_missense4_NP_01120680[k] = (v, sites_trio_missense)\n",
    "                    \n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TET2_sites_missense5_NP_01120680 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TET2_bases_missense4_NP_01120680.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        TET2_sites_missense5_NP_01120680[a]=info\n",
    "        a=a+1\n",
    "        \n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TET2_sites_missense6_NP_01120680={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TET2_sites_missense5_NP_01120680.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "    \n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TET2_sites_missense6_NP_01120680[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "        \n",
    "        \n",
    "    #Assign the mutation rates to each site...\n",
    "    TET2_mutation_rates_missense_NP_01120680 = {}\n",
    "    for k, v in TET2_sites_missense6_NP_01120680.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            TET2_mutation_rates_missense_NP_01120680[k] = (v, mutation_rate)\n",
    "\n",
    "    TET2_mutation_rate_missense_NP_01120680_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TET2_mutation_rates_missense_NP_01120680.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        TET2_mutation_rate_missense_NP_01120680_sum.append(mutation_rate_value)\n",
    "        \n",
    "    TET2_Summed_mutation_rate_missense_NP_01120680_added = np.sum(TET2_mutation_rate_missense_NP_01120680_sum)\n",
    "\n",
    "    print('Total summed mutation rate across TET2 missense_NP_01120680 for this study = ', TET2_Summed_mutation_rate_missense_NP_01120680_added)\n",
    "    \n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all #in format: column 1 = 882, column 2 = p.R882H  #2002 aa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_list_nonsense_variants_TET2_shorter(csv_file_variants):\n",
    "\n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    #TET2 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Summary?db=core;g=ENSG00000168769;r=4:106067032-106200973;t=ENST00000305737\n",
    "    TET2_exons_NP_060098 = 'ATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGGTAAGTGCCAGAAATGTACTGAGACACATGGCGTTTATCCAGAATTAGCAAATTTATCTTCAGATATGGGATTTTCCTTCTTTTTTTAA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    TET2_codons_NP_060098 = [TET2_exons_NP_060098[i:i+3] for i in range(0, len(TET2_exons_NP_060098), 3)]\n",
    "\n",
    "    #TET2 transcript 1 extra base at each end:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Summary?db=core;g=ENSG00000168769;r=4:106067032-106200973;t=ENST00000305737\n",
    "    TET2_exons_extra_bases_NP_060098 = 'GATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGGTAAGTGCCAGAAATGTACTGAGACACATGGCGTTTATCCAGAATTAGCAAATTTATCTTCAGATATGGGATTTTCCTTCTTTTTTTAAA'\n",
    "\n",
    "    #Create a list of the codons they called nonsense variants at\n",
    "    import csv\n",
    "    filename = csv_file_variants\n",
    "    with open(filename, 'r') as csvfile:\n",
    "        read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "        row_count=0\n",
    "\n",
    "        TET2_nonsense_variant_list_NP_060098 = []\n",
    "        TET2_nonsense_codons_NP_060098 =[]\n",
    "        TET2_nonsense_variant_DICT_NP_060098 = {}\n",
    "\n",
    "        for row in read_reader: #each row contains aa number (e.g. 11) and variant (e.g. p.D11A)\n",
    "    #         print(row)\n",
    "            if row_count>0: \n",
    "                codon0_NP_060098 = int(row[0])-1\n",
    "                codon1_NP_060098 = int(row[0])\n",
    "                variant_NP_060098 = row[1]\n",
    "                if codon1_NP_060098 < 1167:\n",
    "                    TET2_nonsense_codons_NP_060098.append(codon0_NP_060098) #List of codon numbers that they called variants at\n",
    "                    TET2_nonsense_variant_DICT_NP_060098[row_count]=(codon1_NP_060098, variant_NP_060098) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "            row_count=row_count+1\n",
    "            \n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    TET2_nonsense_codons_NP_060098_DICT = {}\n",
    "    calls_TET2_nonsense_NP_060098 = {}\n",
    "\n",
    "    for i in TET2_nonsense_codons_NP_060098: #Iterates over the list of codon numbers\n",
    "        TET2_nonsense_codons_NP_060098_DICT[i+1]= (TET2_codons_NP_060098[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "        \n",
    "    for k,v in TET2_nonsense_codons_NP_060098_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_TET2_nonsense_NP_060098[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "        \n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_TET2_nonsense_NP_060098_extra = {}\n",
    "\n",
    "    for k, v in calls_TET2_nonsense_NP_060098.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_TET2_nonsense_NP_060098_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_TET2_nonsense_NP_060098_extra2 = {}\n",
    "\n",
    "    for k, v in calls_TET2_nonsense_NP_060098_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_TET2_nonsense_NP_060098_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "            \n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    TET2_position_in_list_dict_nonsense_NP_060098={}\n",
    "    for k, v in TET2_nonsense_variant_DICT_NP_060098.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_TET2_nonsense_NP_060098[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_TET2_nonsense_NP_060098[v[0]]\n",
    "        codon_seq=calls_TET2_nonsense_NP_060098[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_TET2_nonsense_NP_060098_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                TET2_position_in_list_dict_nonsense_NP_060098[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "            \n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in TET2_position_in_list_dict_nonsense_NP_060098.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "#     print('positions at which nonsense variants called = ', aa_positions_variants_called)\n",
    "                \n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TET2_bases_nonsense_NP_060098 = {}\n",
    "    for k, v in TET2_position_in_list_dict_nonsense_NP_060098.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TET2_bases_nonsense_NP_060098[k]=(v, base_changes)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TET2_bases_nonsense2_NP_060098 = {}\n",
    "\n",
    "    for k, v in TET2_bases_nonsense_NP_060098.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TET2_bases_nonsense2_NP_060098[k] = (v, codon_positions)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TET2_bases_nonsense3_NP_060098 = {}\n",
    "\n",
    "    for k, v in TET2_bases_nonsense2_NP_060098.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TET2_bases_nonsense3_NP_060098[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TET2_bases_nonsense4_NP_060098 = {}\n",
    "\n",
    "    for k, v in TET2_bases_nonsense3_NP_060098.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_nonsense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "            gene_pos = i\n",
    "            site_trio_nonsense = [TET2_exons_extra_bases_NP_060098[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "            sites_trio_nonsense.append(site_trio_nonsense)\n",
    "            TET2_bases_nonsense4_NP_060098[k] = (v, sites_trio_nonsense)\n",
    "                    \n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TET2_sites_nonsense5_NP_060098 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TET2_bases_nonsense4_NP_060098.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        TET2_sites_nonsense5_NP_060098[a]=info\n",
    "        a=a+1\n",
    "#     print(TET2_sites_nonsense5_NP_060098)\n",
    "        \n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TET2_sites_nonsense6_NP_060098={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TET2_sites_nonsense5_NP_060098.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "    \n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TET2_sites_nonsense6_NP_060098[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "        \n",
    "        \n",
    "    #Assign the mutation rates to each site...\n",
    "    TET2_mutation_rates_nonsense_NP_060098 = {}\n",
    "    for k, v in TET2_sites_nonsense6_NP_060098.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            TET2_mutation_rates_nonsense_NP_060098[k] = (v, mutation_rate)\n",
    "\n",
    "    TET2_mutation_rate_nonsense_NP_060098_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TET2_mutation_rates_nonsense_NP_060098.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        TET2_mutation_rate_nonsense_NP_060098_sum.append(mutation_rate_value)\n",
    "        \n",
    "    TET2_Summed_mutation_rate_nonsense_NP_060098_added = np.sum(TET2_mutation_rate_nonsense_NP_060098_sum)\n",
    "\n",
    "    print('Total summed mutation rate across TET2 nonsense_NP_060098 for this study = ', TET2_Summed_mutation_rate_nonsense_NP_060098_added)\n",
    "    \n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all #in format: column 1 = 882, column 2 = p.R882*  #1165 aa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_list_missense_variants_TET2_shorter(csv_file_variants):\n",
    "    \n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    #TET2 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Summary?db=core;g=ENSG00000168769;r=4:106067032-106200973;t=ENST00000305737\n",
    "    TET2_exons_NP_060098 = 'ATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGGTAAGTGCCAGAAATGTACTGAGACACATGGCGTTTATCCAGAATTAGCAAATTTATCTTCAGATATGGGATTTTCCTTCTTTTTTTAA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    TET2_codons_NP_060098 = [TET2_exons_NP_060098[i:i+3] for i in range(0, len(TET2_exons_NP_060098), 3)]\n",
    "\n",
    "    #TET2 transcript 1 extra base at each end:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Summary?db=core;g=ENSG00000168769;r=4:106067032-106200973;t=ENST00000305737\n",
    "    TET2_exons_extra_bases_NP_060098 = 'GATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGGTAAGTGCCAGAAATGTACTGAGACACATGGCGTTTATCCAGAATTAGCAAATTTATCTTCAGATATGGGATTTTCCTTCTTTTTTTAAA'\n",
    "\n",
    "    #Create a list of the codons they called missense variants at\n",
    "    import csv\n",
    "    filename = csv_file_variants\n",
    "    with open(filename, 'r') as csvfile:\n",
    "        read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "        row_count=0\n",
    "\n",
    "        TET2_missense_variant_list_NP_060098 = []\n",
    "        TET2_missense_codons_NP_060098 =[]\n",
    "        TET2_missense_variant_DICT_NP_060098 = {}\n",
    "\n",
    "        for row in read_reader: #each row contains aa number (e.g. 11) and variant (e.g. p.D11A)\n",
    "    #         print(row)\n",
    "            if row_count>0: \n",
    "                codon0_NP_060098 = int(row[0])-1\n",
    "                codon1_NP_060098 = int(row[0])\n",
    "                variant_NP_060098 = row[1]\n",
    "                if codon1_NP_060098 < 1167:\n",
    "                    TET2_missense_codons_NP_060098.append(codon0_NP_060098) #List of codon numbers that they called variants at\n",
    "                    TET2_missense_variant_DICT_NP_060098[row_count]=(codon1_NP_060098, variant_NP_060098) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "            row_count=row_count+1\n",
    "            \n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    TET2_missense_codons_NP_060098_DICT = {}\n",
    "    calls_TET2_missense_NP_060098 = {}\n",
    "\n",
    "    for i in TET2_missense_codons_NP_060098: #Iterates over the list of codon numbers\n",
    "        TET2_missense_codons_NP_060098_DICT[i+1]= (TET2_codons_NP_060098[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "        \n",
    "    for k,v in TET2_missense_codons_NP_060098_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_TET2_missense_NP_060098[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "        \n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_TET2_missense_NP_060098_extra = {}\n",
    "\n",
    "    for k, v in calls_TET2_missense_NP_060098.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_TET2_missense_NP_060098_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_TET2_missense_NP_060098_extra2 = {}\n",
    "\n",
    "    for k, v in calls_TET2_missense_NP_060098_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_TET2_missense_NP_060098_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "            \n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    TET2_position_in_list_dict_missense_NP_060098={}\n",
    "    for k, v in TET2_missense_variant_DICT_NP_060098.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_TET2_missense_NP_060098[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_TET2_missense_NP_060098[v[0]]\n",
    "        codon_seq=calls_TET2_missense_NP_060098[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_TET2_missense_NP_060098_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                TET2_position_in_list_dict_missense_NP_060098[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "            \n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in TET2_position_in_list_dict_missense_NP_060098.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "#     print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "                \n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TET2_bases_missense_NP_060098 = {}\n",
    "    for k, v in TET2_position_in_list_dict_missense_NP_060098.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TET2_bases_missense_NP_060098[k]=(v, base_changes)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TET2_bases_missense2_NP_060098 = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense_NP_060098.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TET2_bases_missense2_NP_060098[k] = (v, codon_positions)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TET2_bases_missense3_NP_060098 = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense2_NP_060098.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TET2_bases_missense3_NP_060098[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TET2_bases_missense4_NP_060098 = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense3_NP_060098.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "            gene_pos = i\n",
    "            site_trio_missense = [TET2_exons_extra_bases_NP_060098[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "            sites_trio_missense.append(site_trio_missense)\n",
    "            TET2_bases_missense4_NP_060098[k] = (v, sites_trio_missense)\n",
    "                    \n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TET2_sites_missense5_NP_060098 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TET2_bases_missense4_NP_060098.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        TET2_sites_missense5_NP_060098[a]=info\n",
    "        a=a+1\n",
    "        \n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TET2_sites_missense6_NP_060098={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TET2_sites_missense5_NP_060098.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "    \n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TET2_sites_missense6_NP_060098[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "        \n",
    "        \n",
    "    #Assign the mutation rates to each site...\n",
    "    TET2_mutation_rates_missense_NP_060098 = {}\n",
    "    for k, v in TET2_sites_missense6_NP_060098.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            TET2_mutation_rates_missense_NP_060098[k] = (v, mutation_rate)\n",
    "\n",
    "    TET2_mutation_rate_missense_NP_060098_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TET2_mutation_rates_missense_NP_060098.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        TET2_mutation_rate_missense_NP_060098_sum.append(mutation_rate_value)\n",
    "        \n",
    "    TET2_Summed_mutation_rate_missense_NP_060098_added = np.sum(TET2_mutation_rate_missense_NP_060098_sum)\n",
    "\n",
    "    print('Total summed mutation rate across TET2 missense_NP_060098 for this study = ', TET2_Summed_mutation_rate_missense_NP_060098_added)\n",
    "    \n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all #in format: column 1 = 882, column 2 = p.R882H  #1165 aa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ASXL1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_list_nonsense_variants_ASXL1(csv_file_variants):\n",
    "    \n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base = [140, 143, 252, 373, 1086, 1085]\n",
    "   \n",
    "    def intron_exon_check(position_in_gene_number):\n",
    "        if position_in_gene_number == 140:\n",
    "            gene_pos = 140\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 143:\n",
    "            gene_pos = 143\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 252:\n",
    "            gene_pos = 252\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 373:\n",
    "            gene_pos = 373\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1086:\n",
    "            gene_pos = 1086\n",
    "            first_base = 'G'\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = ASXL1_exons_extra_bases[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1085:\n",
    "            gene_pos = 1085\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #ASXL1 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    ASXL1_exons = 'ATGAAGGACAAACAGAAGAAGAAGAAGGAGCGCACGTGGGCCGAGGCCGCGCGCCTGGTATTAGAAAACTACTCGGATGCTCCAATGACACCAAAACAGATTCTGCAGGTCATAGAGGCAGAAGGACTAAAGGAAATGAGAAGTGGGACTTCCCCTCTCGCATGCCTCAATGCTATGCTACATTCCAATTCAAGAGGAGGAGAGGGGTTGTTTTATAAACTGCCTGGCCGAATCAGCCTTTTCACGCTCAAGAAGGATGCCCTGCAGTGGTCTCGCCATCCAGCTACAGTGGAGGGAGAGGAGCCAGAGGACACGGCTGATGTGGAGAGCTGTGGGTCTAATGAAGCCAGCACTGTGAGTGGTGAAAACGATGTATCTCTTGATGAAACATCTTCGAACGCATCCTGTTCTACAGAATCTCAGAGTCGACCTCTTTCCAATCCCAGGGACAGCTACAGAGCTTCCTCACAGGCGAACAAACAAAAGAAAAAGACTGGGGTGATGCTGCCTCGAGTTGTCCTGACTCCTCTGAAGGTAAACGGGGCCCACGTGGAATCTGCATCAGGGTTCTCGGGCTGCCACGCCGATGGCGAGAGCGGCAGCCCGTCCAGCAGCAGCAGCGGCTCTCTGGCCCTGGGCAGCGCTGCTATTCGTGGCCAGGCCGAGGTCACCCAGGACCCTGCCCCGCTCCTGAGAGGCTTCCGGAAGCCAGCCACAGGTCAAATGAAGCGCAACAGAGGGGAAGAAATAGATTTTGAGACACCTGGGTCCATTCTTGTCAACACCAACCTCCGTGCCCTGATCAACTCTCGGACCTTCCATGCCTTACCATCACACTTCCAGCAGCAGCTCCTCTTCCTCCTGCCTGAAGTAGACAGACAGGTGGGGACGGATGGCCTGTTGCGTCTCAGCAGCAGTGCACTAAATAACGAGTTTTTTACCCATGCGGCTCAGAGCTGGCGGGAGCGCCTGGCTGATGGTGAATTTACTCATGAGATGCAAGTCAGGATACGACAGGAAATGGAGAAGGAAAAGAAGGTGGAACAATGGAAAGAAAAGTTCTTTGAAGACTACTATGGACAGAAGCTGGGTTTGACCAAAGAAGAGTCATTGCAGCAGAACGTGGGCCAGGAGGAGGCTGAAATCAAAAGTGGCTTGTGTGTCCCAGGAGAATCAGTGCGTATACAGCGTGGTCCAGCCACCCGACAGCGAGATGGGCATTTTAAGAAACGCTCTCGGCCAGATCTCCGAACCAGAGCCAGAAGGAATCTGTACAAAAAACAGGAGTCAGAACAAGCAGGGGTTGCTAAGGATGCAAAATCTGTGGCCTCAGATGTTCCCCTCTACAAGGATGGGGAGGCTAAGACTGACCCAGCAGGGCTGAGCAGTCCCCATCTGCCAGGCACATCCTCTGCAGCACCCGACCTGGAGGGTCCCGAATTCCCAGTTGAGTCTGTGGCTTCTCGGATCCAGGCTGAGCCAGACAACTTGGCACGTGCCTCTGCATCTCCAGACAGAATTCCTAGCCTGCCTCAGGAAACTGTGGATCAGGAACCCAAGGATCAGAAGAGGAAATCCTTTGAGCAGGCGGCCTCTGCATCCTTTCCCGAAAAGAAGCCCCGGCTTGAAGATCGTCAGTCCTTTCGTAACACAATTGAAAGTGTTCACACCGAAAAGCCACAGCCCACTAAAGAGGAGCCCAAAGTCCCGCCCATCCGGATTCAACTTTCACGTATCAAACCACCCTGGGTGGTTAAAGGTCAGCCCACTTACCAGATATGCCCCCGGATCATCCCCACCACGGAGTCCTCCTGCCGGGGTTGGACTGGCGCCAGGACCCTCGCAGACATTAAAGCCCGTGCTCTGCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCCACCACTGCCATCGGAGGGGGGGGTGGCCCGGGTGGAGGTGGCGGCGGGGCCACCGATGAGGGAGGTGGCAGAGGCAGCAGCAGTGGTGATGGTGGTGAGGCCTGTGGCCACCCTGAGCCCAGGGGAGGCCCGAGCACCCCTGGAAAGTGTACGTCAGATCTACAGCGAACACAACTACTGCCGCCTTATCCTCTAAATGGGGAGCATACCCAGGCCGGAACTGCCATGTCCAGAGCTAGGAGAGAGGACCTGCCTTCTCTGAGAAAGGAGGAAAGCTGCCTACTACAGAGGGCTACAGTTGGACTCACAGATGGGCTAGGAGATGCCTCCCAACTCCCCGTTGCTCCCACTGGGGACCAGCCATGCCAGGCCTTGCCCCTACTGTCCTCCCAAACCTCAGTAGCTGAGAGATTAGTGGAGCAGCCTCAGTTGCATCCGGATGTTAGAACTGAATGTGAGTCTGGCACCACTTCCTGGGAAAGTGATGATGAGGAGCAAGGACCCACCGTTCCTGCAGACAATGGTCCCATTCTGTCTCTAGTGGGAGATGATACATTAGAGAAAGGAACTGGCCAAGCTCTTGACAGTCATCCCACTATGAAGGATCCTGTAAATGTGACCCCCAGTTCCACACCTGAATCCTCACCGACTGATTGCCTGCAGAACAGAGCATTTGATGACGAATTAGGGCTTGGTGGCTCATGCCCTCCTATGAGGGAAAGTGATACTAGACAAGAAAACTTGAAAACCAAGGCTCTCGTTTCTAACAGTTCTTTGCATTGGATACCCATCCCATCGAATGATGAGGTAGTGAAACAGCCCAAACCAGAATCCAGAGAACACATACCATCTGTTGAGCCCCAGGTTGGAGAGGAGTGGGAGAAAGCTGCTCCCACCCCTCCTGCATTGCCTGGGGATTTGACAGCTGAGGAGGGTCTAGATCCTCTTGACAGCCTTACTTCACTCTGGACTGTGCCATCTCGAGGAGGCAGTGACAGCAATGGCAGTTACTGTCAACAGGTGGACATTGAAAAGCTGAAAATCAACGGAGACTCTGAAGCACTGAGTCCTCACGGTGAGTCCACGGATACAGCCTCTGACTTTGAAGGTCACCTCACGGAGGACAGCAGTGAGGCTGACACTAGAGAAGCTGCAGTGACAAAGGGATCTTCGGTGGACAAGGATGAGAAACCCAATTGGAACCAATCTGCCCCACTGTCCAAGGTGAATGGTGACATGCGTCTGGTTACAAGGACAGATGGGATGGTTGCTCCTCAGAGCTGGGTGTCTCGAGTATGTGCGGTCCGCCAAAAGATCCCAGATTCCCTACTGCTGGCCAGTACTGAGTACCAGCCAAGAGCCGTGTGCCTGTCCATGCCTGGGTCCTCAGTGGAGGCCACTAACCCACTTGTGATGCAGTTGCTGCAGGGTAGCTTGCCCCTAGAGAAGGTTCTTCCACCAGCCCACGATGACAGCATGTCAGAATCCCCACAAGTACCACTTACAAAAGACCAGAGCCATGGCTCGCTACGCATGGGATCTTTACATGGTCTTGGAAAAAACAGTGGCATGGTTGATGGAAGCAGCCCCAGTTCTTTAAGGGCTTTGAAGGAGCCTCTTCTGCCAGATAGCTGTGAAACAGGCACTGGTCTTGCCAGGATTGAGGCCACCCAGGCTCCTGGAGCACCCCAAAAGAATTGCAAGGCAGTCCCAAGTTTTGACTCCCTCCATCCAGTGACAAATCCCATTACATCCTCTAGGAAACTGGAAGAAATGGATTCCAAAGAGCAGTTCTCTTCCTTTAGTTGTGAAGATCAGAAGGAAGTCCGTGCTATGTCACAGGACAGTAATTCAAATGCTGCTCCAGGAAAGAGCCCAGGAGATCTTACTACCTCGAGAACACCTCGTTTCTCATCTCCAAATGTGATCTCCTTTGGTCCAGAGCAGACAGGTCGGGCCCTGGGTGATCAGAGCAATGTTACAGGCCAAGGGAAGAAGCTTTTTGGCTCTGGGAATGTGGCTGCAACCCTTCAGCGCCCCAGGCCTGCGGACCCGATGCCTCTTCCTGCTGAGATCCCTCCAGTTTTTCCCAGTGGGAAGTTGGGACCAAGCACAAACTCCATGTCTGGTGGGGTACAGACTCCAAGGGAAGACTGGGCTCCAAAGCCACATGCCTTTGTTGGCAGCGTCAAGAATGAGAAGACTTTTGTGGGGGGTCCTCTTAAGGCAAATGCCGAGAACAGGAAAGCTACTGGGCATAGTCCCCTGGAACTGGTGGGTCACTTGGAAGGGATGCCCTTTGTCATGGACTTGCCCTTCTGGAAATTACCCCGAGAGCCAGGGAAGGGGCTCAGTGAGCCTCTGGAGCCTTCTTCTCTCCCCTCCCAACTCAGCATCAAGCAGGCATTTTATGGGAAGCTTTCTAAACTCCAACTGAGTTCCACCAGCTTTAATTATTCCTCTAGCTCTCCCACCTTTCCCAAAGGCCTTGCTGGAAGTGTGGTGCAGCTGAGCCACAAAGCAAACTTTGGTGCGAGCCACAGTGCATCACTTTCCTTGCAAATGTTCACTGACAGCAGCACGGTGGAAAGCATCTCGCTCCAGTGTGCGTGCAGCCTGAAAGCCATGATCATGTGCCAAGGCTGCGGTGCGTTCTGTCACGATGACTGTATTGGACCCTCAAAGCTCTGTGTATTGTGCCTTGTGGTGAGATAA'\n",
    "    \n",
    "    #Split the sequence into codons \n",
    "    ASXL1_codons = [ASXL1_exons[i:i+3] for i in range(0, len(ASXL1_exons), 3)]\n",
    "\n",
    "    #ASXL1 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    ASXL1_exons_extra_bases = 'GATGAAGGACAAACAGAAGAAGAAGAAGGAGCGCACGTGGGCCGAGGCCGCGCGCCTGGTATTAGAAAACTACTCGGATGCTCCAATGACACCAAAACAGATTCTGCAGGTCATAGAGGCAGAAGGACTAAAGGAAATGAGAAGTGGGACTTCCCCTCTCGCATGCCTCAATGCTATGCTACATTCCAATTCAAGAGGAGGAGAGGGGTTGTTTTATAAACTGCCTGGCCGAATCAGCCTTTTCACGCTCAAGAAGGATGCCCTGCAGTGGTCTCGCCATCCAGCTACAGTGGAGGGAGAGGAGCCAGAGGACACGGCTGATGTGGAGAGCTGTGGGTCTAATGAAGCCAGCACTGTGAGTGGTGAAAACGATGTATCTCTTGATGAAACATCTTCGAACGCATCCTGTTCTACAGAATCTCAGAGTCGACCTCTTTCCAATCCCAGGGACAGCTACAGAGCTTCCTCACAGGCGAACAAACAAAAGAAAAAGACTGGGGTGATGCTGCCTCGAGTTGTCCTGACTCCTCTGAAGGTAAACGGGGCCCACGTGGAATCTGCATCAGGGTTCTCGGGCTGCCACGCCGATGGCGAGAGCGGCAGCCCGTCCAGCAGCAGCAGCGGCTCTCTGGCCCTGGGCAGCGCTGCTATTCGTGGCCAGGCCGAGGTCACCCAGGACCCTGCCCCGCTCCTGAGAGGCTTCCGGAAGCCAGCCACAGGTCAAATGAAGCGCAACAGAGGGGAAGAAATAGATTTTGAGACACCTGGGTCCATTCTTGTCAACACCAACCTCCGTGCCCTGATCAACTCTCGGACCTTCCATGCCTTACCATCACACTTCCAGCAGCAGCTCCTCTTCCTCCTGCCTGAAGTAGACAGACAGGTGGGGACGGATGGCCTGTTGCGTCTCAGCAGCAGTGCACTAAATAACGAGTTTTTTACCCATGCGGCTCAGAGCTGGCGGGAGCGCCTGGCTGATGGTGAATTTACTCATGAGATGCAAGTCAGGATACGACAGGAAATGGAGAAGGAAAAGAAGGTGGAACAATGGAAAGAAAAGTTCTTTGAAGACTACTATGGACAGAAGCTGGGTTTGACCAAAGAAGAGTCATTGCAGCAGAACGTGGGCCAGGAGGAGGCTGAAATCAAAAGTGGCTTGTGTGTCCCAGGAGAATCAGTGCGTATACAGCGTGGTCCAGCCACCCGACAGCGAGATGGGCATTTTAAGAAACGCTCTCGGCCAGATCTCCGAACCAGAGCCAGAAGGAATCTGTACAAAAAACAGGAGTCAGAACAAGCAGGGGTTGCTAAGGATGCAAAATCTGTGGCCTCAGATGTTCCCCTCTACAAGGATGGGGAGGCTAAGACTGACCCAGCAGGGCTGAGCAGTCCCCATCTGCCAGGCACATCCTCTGCAGCACCCGACCTGGAGGGTCCCGAATTCCCAGTTGAGTCTGTGGCTTCTCGGATCCAGGCTGAGCCAGACAACTTGGCACGTGCCTCTGCATCTCCAGACAGAATTCCTAGCCTGCCTCAGGAAACTGTGGATCAGGAACCCAAGGATCAGAAGAGGAAATCCTTTGAGCAGGCGGCCTCTGCATCCTTTCCCGAAAAGAAGCCCCGGCTTGAAGATCGTCAGTCCTTTCGTAACACAATTGAAAGTGTTCACACCGAAAAGCCACAGCCCACTAAAGAGGAGCCCAAAGTCCCGCCCATCCGGATTCAACTTTCACGTATCAAACCACCCTGGGTGGTTAAAGGTCAGCCCACTTACCAGATATGCCCCCGGATCATCCCCACCACGGAGTCCTCCTGCCGGGGTTGGACTGGCGCCAGGACCCTCGCAGACATTAAAGCCCGTGCTCTGCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCCACCACTGCCATCGGAGGGGGGGGTGGCCCGGGTGGAGGTGGCGGCGGGGCCACCGATGAGGGAGGTGGCAGAGGCAGCAGCAGTGGTGATGGTGGTGAGGCCTGTGGCCACCCTGAGCCCAGGGGAGGCCCGAGCACCCCTGGAAAGTGTACGTCAGATCTACAGCGAACACAACTACTGCCGCCTTATCCTCTAAATGGGGAGCATACCCAGGCCGGAACTGCCATGTCCAGAGCTAGGAGAGAGGACCTGCCTTCTCTGAGAAAGGAGGAAAGCTGCCTACTACAGAGGGCTACAGTTGGACTCACAGATGGGCTAGGAGATGCCTCCCAACTCCCCGTTGCTCCCACTGGGGACCAGCCATGCCAGGCCTTGCCCCTACTGTCCTCCCAAACCTCAGTAGCTGAGAGATTAGTGGAGCAGCCTCAGTTGCATCCGGATGTTAGAACTGAATGTGAGTCTGGCACCACTTCCTGGGAAAGTGATGATGAGGAGCAAGGACCCACCGTTCCTGCAGACAATGGTCCCATTCTGTCTCTAGTGGGAGATGATACATTAGAGAAAGGAACTGGCCAAGCTCTTGACAGTCATCCCACTATGAAGGATCCTGTAAATGTGACCCCCAGTTCCACACCTGAATCCTCACCGACTGATTGCCTGCAGAACAGAGCATTTGATGACGAATTAGGGCTTGGTGGCTCATGCCCTCCTATGAGGGAAAGTGATACTAGACAAGAAAACTTGAAAACCAAGGCTCTCGTTTCTAACAGTTCTTTGCATTGGATACCCATCCCATCGAATGATGAGGTAGTGAAACAGCCCAAACCAGAATCCAGAGAACACATACCATCTGTTGAGCCCCAGGTTGGAGAGGAGTGGGAGAAAGCTGCTCCCACCCCTCCTGCATTGCCTGGGGATTTGACAGCTGAGGAGGGTCTAGATCCTCTTGACAGCCTTACTTCACTCTGGACTGTGCCATCTCGAGGAGGCAGTGACAGCAATGGCAGTTACTGTCAACAGGTGGACATTGAAAAGCTGAAAATCAACGGAGACTCTGAAGCACTGAGTCCTCACGGTGAGTCCACGGATACAGCCTCTGACTTTGAAGGTCACCTCACGGAGGACAGCAGTGAGGCTGACACTAGAGAAGCTGCAGTGACAAAGGGATCTTCGGTGGACAAGGATGAGAAACCCAATTGGAACCAATCTGCCCCACTGTCCAAGGTGAATGGTGACATGCGTCTGGTTACAAGGACAGATGGGATGGTTGCTCCTCAGAGCTGGGTGTCTCGAGTATGTGCGGTCCGCCAAAAGATCCCAGATTCCCTACTGCTGGCCAGTACTGAGTACCAGCCAAGAGCCGTGTGCCTGTCCATGCCTGGGTCCTCAGTGGAGGCCACTAACCCACTTGTGATGCAGTTGCTGCAGGGTAGCTTGCCCCTAGAGAAGGTTCTTCCACCAGCCCACGATGACAGCATGTCAGAATCCCCACAAGTACCACTTACAAAAGACCAGAGCCATGGCTCGCTACGCATGGGATCTTTACATGGTCTTGGAAAAAACAGTGGCATGGTTGATGGAAGCAGCCCCAGTTCTTTAAGGGCTTTGAAGGAGCCTCTTCTGCCAGATAGCTGTGAAACAGGCACTGGTCTTGCCAGGATTGAGGCCACCCAGGCTCCTGGAGCACCCCAAAAGAATTGCAAGGCAGTCCCAAGTTTTGACTCCCTCCATCCAGTGACAAATCCCATTACATCCTCTAGGAAACTGGAAGAAATGGATTCCAAAGAGCAGTTCTCTTCCTTTAGTTGTGAAGATCAGAAGGAAGTCCGTGCTATGTCACAGGACAGTAATTCAAATGCTGCTCCAGGAAAGAGCCCAGGAGATCTTACTACCTCGAGAACACCTCGTTTCTCATCTCCAAATGTGATCTCCTTTGGTCCAGAGCAGACAGGTCGGGCCCTGGGTGATCAGAGCAATGTTACAGGCCAAGGGAAGAAGCTTTTTGGCTCTGGGAATGTGGCTGCAACCCTTCAGCGCCCCAGGCCTGCGGACCCGATGCCTCTTCCTGCTGAGATCCCTCCAGTTTTTCCCAGTGGGAAGTTGGGACCAAGCACAAACTCCATGTCTGGTGGGGTACAGACTCCAAGGGAAGACTGGGCTCCAAAGCCACATGCCTTTGTTGGCAGCGTCAAGAATGAGAAGACTTTTGTGGGGGGTCCTCTTAAGGCAAATGCCGAGAACAGGAAAGCTACTGGGCATAGTCCCCTGGAACTGGTGGGTCACTTGGAAGGGATGCCCTTTGTCATGGACTTGCCCTTCTGGAAATTACCCCGAGAGCCAGGGAAGGGGCTCAGTGAGCCTCTGGAGCCTTCTTCTCTCCCCTCCCAACTCAGCATCAAGCAGGCATTTTATGGGAAGCTTTCTAAACTCCAACTGAGTTCCACCAGCTTTAATTATTCCTCTAGCTCTCCCACCTTTCCCAAAGGCCTTGCTGGAAGTGTGGTGCAGCTGAGCCACAAAGCAAACTTTGGTGCGAGCCACAGTGCATCACTTTCCTTGCAAATGTTCACTGACAGCAGCACGGTGGAAAGCATCTCGCTCCAGTGTGCGTGCAGCCTGAAAGCCATGATCATGTGCCAAGGCTGCGGTGCGTTCTGTCACGATGACTGTATTGGACCCTCAAAGCTCTGTGTATTGTGCCTTGTGGTGAGATAAT'\n",
    "    \n",
    "    #Create a list of the codons they called nonsense variants at\n",
    "    import csv\n",
    "    filename = csv_file_variants\n",
    "    with open(filename, 'r') as csvfile:\n",
    "        read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "        row_count=0\n",
    "\n",
    "        ASXL1_nonsense_variant_list = []\n",
    "        ASXL1_nonsense_codons =[]\n",
    "        ASXL1_nonsense_variant_DICT = {}\n",
    "\n",
    "        for row in read_reader: #each row contains aa number (e.g. 11) and variant (e.g. p.D11A)\n",
    "    #         print(row)\n",
    "            if row_count>0: \n",
    "                codon0 = int(row[0])-1\n",
    "                codon1 = int(row[0])\n",
    "                variant = row[1]\n",
    "                ASXL1_nonsense_codons.append(codon0) #List of codon numbers that they called variants at\n",
    "                ASXL1_nonsense_variant_DICT[row_count]=(codon1, variant) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "            row_count=row_count+1\n",
    "            \n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    ASXL1_nonsense_codons_DICT = {}\n",
    "    calls_nonsense = {}\n",
    "\n",
    "    for i in ASXL1_nonsense_codons: #Iterates over the list of codon numbers\n",
    "        ASXL1_nonsense_codons_DICT[i+1]= (ASXL1_codons[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "        \n",
    "    for k,v in ASXL1_nonsense_codons_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_nonsense[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "        \n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_nonsense_extra = {}\n",
    "\n",
    "    for k, v in calls_nonsense.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_nonsense_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_nonsense_extra2 = {}\n",
    "\n",
    "    for k, v in calls_nonsense_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_nonsense_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "            \n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    ASXL1_position_in_list_dict_nonsense={}\n",
    "    for k, v in ASXL1_nonsense_variant_DICT.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_nonsense[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_nonsense[v[0]]\n",
    "        codon_seq=calls_nonsense[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_nonsense_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                ASXL1_position_in_list_dict_nonsense[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "            \n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in ASXL1_position_in_list_dict_nonsense.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "#     print('positions at which nonsense variants called = ', aa_positions_variants_called)\n",
    "                \n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    ASXL1_bases_nonsense = {}\n",
    "    for k, v in ASXL1_position_in_list_dict_nonsense.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        ASXL1_bases_nonsense[k]=(v, base_changes)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    ASXL1_bases_nonsense2 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_nonsense.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        ASXL1_bases_nonsense2[k] = (v, codon_positions)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    ASXL1_bases_nonsense3 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_nonsense2.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        ASXL1_bases_nonsense3[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    ASXL1_bases_nonsense4 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_nonsense3.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_nonsense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check(i)\n",
    "                sites_trio_nonsense.append([adapted_trio])\n",
    "                ASXL1_bases_nonsense4[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_nonsense = [ASXL1_exons_extra_bases[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_nonsense.append(site_trio_nonsense)\n",
    "                    ASXL1_bases_nonsense4[k] = (v, sites_trio_nonsense)\n",
    "                    \n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    ASXL1_sites_nonsense5 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in ASXL1_bases_nonsense4.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        ASXL1_sites_nonsense5[a]=info\n",
    "        a=a+1\n",
    "#     print(ASXL1_sites_nonsense5)\n",
    "        \n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    ASXL1_sites_nonsense6={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in ASXL1_sites_nonsense5.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "    \n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        ASXL1_sites_nonsense6[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "        \n",
    "        \n",
    "    #Assign the mutation rates to each site...\n",
    "    ASXL1_mutation_rates_nonsense = {}\n",
    "    for k, v in ASXL1_sites_nonsense6.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            ASXL1_mutation_rates_nonsense[k] = (v, mutation_rate)\n",
    "\n",
    "    ASXL1_mutation_rate_nonsense_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in ASXL1_mutation_rates_nonsense.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        ASXL1_mutation_rate_nonsense_sum.append(mutation_rate_value)\n",
    "        \n",
    "    ASXL1_Summed_mutation_rate_nonsense_added = np.sum(ASXL1_mutation_rate_nonsense_sum)\n",
    "\n",
    "    print('Total summed mutation rate across ASXL1 nonsense for this study = ', ASXL1_Summed_mutation_rate_nonsense_added)\n",
    "    \n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all #in format: column 1 = 882, column 2 = p.R882*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_list_missense_variants_ASXL1(csv_file_variants):\n",
    "    \n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base = [140, 143, 252, 373, 1086, 1085]\n",
    "   \n",
    "    def intron_exon_check(position_in_gene_number):\n",
    "        if position_in_gene_number == 140:\n",
    "            gene_pos = 140\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 143:\n",
    "            gene_pos = 143\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 252:\n",
    "            gene_pos = 252\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 373:\n",
    "            gene_pos = 373\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1086:\n",
    "            gene_pos = 1086\n",
    "            first_base = 'G'\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = ASXL1_exons_extra_bases[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1085:\n",
    "            gene_pos = 1085\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #ASXL1 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    ASXL1_exons = 'ATGAAGGACAAACAGAAGAAGAAGAAGGAGCGCACGTGGGCCGAGGCCGCGCGCCTGGTATTAGAAAACTACTCGGATGCTCCAATGACACCAAAACAGATTCTGCAGGTCATAGAGGCAGAAGGACTAAAGGAAATGAGAAGTGGGACTTCCCCTCTCGCATGCCTCAATGCTATGCTACATTCCAATTCAAGAGGAGGAGAGGGGTTGTTTTATAAACTGCCTGGCCGAATCAGCCTTTTCACGCTCAAGAAGGATGCCCTGCAGTGGTCTCGCCATCCAGCTACAGTGGAGGGAGAGGAGCCAGAGGACACGGCTGATGTGGAGAGCTGTGGGTCTAATGAAGCCAGCACTGTGAGTGGTGAAAACGATGTATCTCTTGATGAAACATCTTCGAACGCATCCTGTTCTACAGAATCTCAGAGTCGACCTCTTTCCAATCCCAGGGACAGCTACAGAGCTTCCTCACAGGCGAACAAACAAAAGAAAAAGACTGGGGTGATGCTGCCTCGAGTTGTCCTGACTCCTCTGAAGGTAAACGGGGCCCACGTGGAATCTGCATCAGGGTTCTCGGGCTGCCACGCCGATGGCGAGAGCGGCAGCCCGTCCAGCAGCAGCAGCGGCTCTCTGGCCCTGGGCAGCGCTGCTATTCGTGGCCAGGCCGAGGTCACCCAGGACCCTGCCCCGCTCCTGAGAGGCTTCCGGAAGCCAGCCACAGGTCAAATGAAGCGCAACAGAGGGGAAGAAATAGATTTTGAGACACCTGGGTCCATTCTTGTCAACACCAACCTCCGTGCCCTGATCAACTCTCGGACCTTCCATGCCTTACCATCACACTTCCAGCAGCAGCTCCTCTTCCTCCTGCCTGAAGTAGACAGACAGGTGGGGACGGATGGCCTGTTGCGTCTCAGCAGCAGTGCACTAAATAACGAGTTTTTTACCCATGCGGCTCAGAGCTGGCGGGAGCGCCTGGCTGATGGTGAATTTACTCATGAGATGCAAGTCAGGATACGACAGGAAATGGAGAAGGAAAAGAAGGTGGAACAATGGAAAGAAAAGTTCTTTGAAGACTACTATGGACAGAAGCTGGGTTTGACCAAAGAAGAGTCATTGCAGCAGAACGTGGGCCAGGAGGAGGCTGAAATCAAAAGTGGCTTGTGTGTCCCAGGAGAATCAGTGCGTATACAGCGTGGTCCAGCCACCCGACAGCGAGATGGGCATTTTAAGAAACGCTCTCGGCCAGATCTCCGAACCAGAGCCAGAAGGAATCTGTACAAAAAACAGGAGTCAGAACAAGCAGGGGTTGCTAAGGATGCAAAATCTGTGGCCTCAGATGTTCCCCTCTACAAGGATGGGGAGGCTAAGACTGACCCAGCAGGGCTGAGCAGTCCCCATCTGCCAGGCACATCCTCTGCAGCACCCGACCTGGAGGGTCCCGAATTCCCAGTTGAGTCTGTGGCTTCTCGGATCCAGGCTGAGCCAGACAACTTGGCACGTGCCTCTGCATCTCCAGACAGAATTCCTAGCCTGCCTCAGGAAACTGTGGATCAGGAACCCAAGGATCAGAAGAGGAAATCCTTTGAGCAGGCGGCCTCTGCATCCTTTCCCGAAAAGAAGCCCCGGCTTGAAGATCGTCAGTCCTTTCGTAACACAATTGAAAGTGTTCACACCGAAAAGCCACAGCCCACTAAAGAGGAGCCCAAAGTCCCGCCCATCCGGATTCAACTTTCACGTATCAAACCACCCTGGGTGGTTAAAGGTCAGCCCACTTACCAGATATGCCCCCGGATCATCCCCACCACGGAGTCCTCCTGCCGGGGTTGGACTGGCGCCAGGACCCTCGCAGACATTAAAGCCCGTGCTCTGCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCCACCACTGCCATCGGAGGGGGGGGTGGCCCGGGTGGAGGTGGCGGCGGGGCCACCGATGAGGGAGGTGGCAGAGGCAGCAGCAGTGGTGATGGTGGTGAGGCCTGTGGCCACCCTGAGCCCAGGGGAGGCCCGAGCACCCCTGGAAAGTGTACGTCAGATCTACAGCGAACACAACTACTGCCGCCTTATCCTCTAAATGGGGAGCATACCCAGGCCGGAACTGCCATGTCCAGAGCTAGGAGAGAGGACCTGCCTTCTCTGAGAAAGGAGGAAAGCTGCCTACTACAGAGGGCTACAGTTGGACTCACAGATGGGCTAGGAGATGCCTCCCAACTCCCCGTTGCTCCCACTGGGGACCAGCCATGCCAGGCCTTGCCCCTACTGTCCTCCCAAACCTCAGTAGCTGAGAGATTAGTGGAGCAGCCTCAGTTGCATCCGGATGTTAGAACTGAATGTGAGTCTGGCACCACTTCCTGGGAAAGTGATGATGAGGAGCAAGGACCCACCGTTCCTGCAGACAATGGTCCCATTCTGTCTCTAGTGGGAGATGATACATTAGAGAAAGGAACTGGCCAAGCTCTTGACAGTCATCCCACTATGAAGGATCCTGTAAATGTGACCCCCAGTTCCACACCTGAATCCTCACCGACTGATTGCCTGCAGAACAGAGCATTTGATGACGAATTAGGGCTTGGTGGCTCATGCCCTCCTATGAGGGAAAGTGATACTAGACAAGAAAACTTGAAAACCAAGGCTCTCGTTTCTAACAGTTCTTTGCATTGGATACCCATCCCATCGAATGATGAGGTAGTGAAACAGCCCAAACCAGAATCCAGAGAACACATACCATCTGTTGAGCCCCAGGTTGGAGAGGAGTGGGAGAAAGCTGCTCCCACCCCTCCTGCATTGCCTGGGGATTTGACAGCTGAGGAGGGTCTAGATCCTCTTGACAGCCTTACTTCACTCTGGACTGTGCCATCTCGAGGAGGCAGTGACAGCAATGGCAGTTACTGTCAACAGGTGGACATTGAAAAGCTGAAAATCAACGGAGACTCTGAAGCACTGAGTCCTCACGGTGAGTCCACGGATACAGCCTCTGACTTTGAAGGTCACCTCACGGAGGACAGCAGTGAGGCTGACACTAGAGAAGCTGCAGTGACAAAGGGATCTTCGGTGGACAAGGATGAGAAACCCAATTGGAACCAATCTGCCCCACTGTCCAAGGTGAATGGTGACATGCGTCTGGTTACAAGGACAGATGGGATGGTTGCTCCTCAGAGCTGGGTGTCTCGAGTATGTGCGGTCCGCCAAAAGATCCCAGATTCCCTACTGCTGGCCAGTACTGAGTACCAGCCAAGAGCCGTGTGCCTGTCCATGCCTGGGTCCTCAGTGGAGGCCACTAACCCACTTGTGATGCAGTTGCTGCAGGGTAGCTTGCCCCTAGAGAAGGTTCTTCCACCAGCCCACGATGACAGCATGTCAGAATCCCCACAAGTACCACTTACAAAAGACCAGAGCCATGGCTCGCTACGCATGGGATCTTTACATGGTCTTGGAAAAAACAGTGGCATGGTTGATGGAAGCAGCCCCAGTTCTTTAAGGGCTTTGAAGGAGCCTCTTCTGCCAGATAGCTGTGAAACAGGCACTGGTCTTGCCAGGATTGAGGCCACCCAGGCTCCTGGAGCACCCCAAAAGAATTGCAAGGCAGTCCCAAGTTTTGACTCCCTCCATCCAGTGACAAATCCCATTACATCCTCTAGGAAACTGGAAGAAATGGATTCCAAAGAGCAGTTCTCTTCCTTTAGTTGTGAAGATCAGAAGGAAGTCCGTGCTATGTCACAGGACAGTAATTCAAATGCTGCTCCAGGAAAGAGCCCAGGAGATCTTACTACCTCGAGAACACCTCGTTTCTCATCTCCAAATGTGATCTCCTTTGGTCCAGAGCAGACAGGTCGGGCCCTGGGTGATCAGAGCAATGTTACAGGCCAAGGGAAGAAGCTTTTTGGCTCTGGGAATGTGGCTGCAACCCTTCAGCGCCCCAGGCCTGCGGACCCGATGCCTCTTCCTGCTGAGATCCCTCCAGTTTTTCCCAGTGGGAAGTTGGGACCAAGCACAAACTCCATGTCTGGTGGGGTACAGACTCCAAGGGAAGACTGGGCTCCAAAGCCACATGCCTTTGTTGGCAGCGTCAAGAATGAGAAGACTTTTGTGGGGGGTCCTCTTAAGGCAAATGCCGAGAACAGGAAAGCTACTGGGCATAGTCCCCTGGAACTGGTGGGTCACTTGGAAGGGATGCCCTTTGTCATGGACTTGCCCTTCTGGAAATTACCCCGAGAGCCAGGGAAGGGGCTCAGTGAGCCTCTGGAGCCTTCTTCTCTCCCCTCCCAACTCAGCATCAAGCAGGCATTTTATGGGAAGCTTTCTAAACTCCAACTGAGTTCCACCAGCTTTAATTATTCCTCTAGCTCTCCCACCTTTCCCAAAGGCCTTGCTGGAAGTGTGGTGCAGCTGAGCCACAAAGCAAACTTTGGTGCGAGCCACAGTGCATCACTTTCCTTGCAAATGTTCACTGACAGCAGCACGGTGGAAAGCATCTCGCTCCAGTGTGCGTGCAGCCTGAAAGCCATGATCATGTGCCAAGGCTGCGGTGCGTTCTGTCACGATGACTGTATTGGACCCTCAAAGCTCTGTGTATTGTGCCTTGTGGTGAGATAA'\n",
    "    \n",
    "    #Split the sequence into codons \n",
    "    ASXL1_codons = [ASXL1_exons[i:i+3] for i in range(0, len(ASXL1_exons), 3)]\n",
    "\n",
    "    #ASXL1 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    ASXL1_exons_extra_bases = 'GATGAAGGACAAACAGAAGAAGAAGAAGGAGCGCACGTGGGCCGAGGCCGCGCGCCTGGTATTAGAAAACTACTCGGATGCTCCAATGACACCAAAACAGATTCTGCAGGTCATAGAGGCAGAAGGACTAAAGGAAATGAGAAGTGGGACTTCCCCTCTCGCATGCCTCAATGCTATGCTACATTCCAATTCAAGAGGAGGAGAGGGGTTGTTTTATAAACTGCCTGGCCGAATCAGCCTTTTCACGCTCAAGAAGGATGCCCTGCAGTGGTCTCGCCATCCAGCTACAGTGGAGGGAGAGGAGCCAGAGGACACGGCTGATGTGGAGAGCTGTGGGTCTAATGAAGCCAGCACTGTGAGTGGTGAAAACGATGTATCTCTTGATGAAACATCTTCGAACGCATCCTGTTCTACAGAATCTCAGAGTCGACCTCTTTCCAATCCCAGGGACAGCTACAGAGCTTCCTCACAGGCGAACAAACAAAAGAAAAAGACTGGGGTGATGCTGCCTCGAGTTGTCCTGACTCCTCTGAAGGTAAACGGGGCCCACGTGGAATCTGCATCAGGGTTCTCGGGCTGCCACGCCGATGGCGAGAGCGGCAGCCCGTCCAGCAGCAGCAGCGGCTCTCTGGCCCTGGGCAGCGCTGCTATTCGTGGCCAGGCCGAGGTCACCCAGGACCCTGCCCCGCTCCTGAGAGGCTTCCGGAAGCCAGCCACAGGTCAAATGAAGCGCAACAGAGGGGAAGAAATAGATTTTGAGACACCTGGGTCCATTCTTGTCAACACCAACCTCCGTGCCCTGATCAACTCTCGGACCTTCCATGCCTTACCATCACACTTCCAGCAGCAGCTCCTCTTCCTCCTGCCTGAAGTAGACAGACAGGTGGGGACGGATGGCCTGTTGCGTCTCAGCAGCAGTGCACTAAATAACGAGTTTTTTACCCATGCGGCTCAGAGCTGGCGGGAGCGCCTGGCTGATGGTGAATTTACTCATGAGATGCAAGTCAGGATACGACAGGAAATGGAGAAGGAAAAGAAGGTGGAACAATGGAAAGAAAAGTTCTTTGAAGACTACTATGGACAGAAGCTGGGTTTGACCAAAGAAGAGTCATTGCAGCAGAACGTGGGCCAGGAGGAGGCTGAAATCAAAAGTGGCTTGTGTGTCCCAGGAGAATCAGTGCGTATACAGCGTGGTCCAGCCACCCGACAGCGAGATGGGCATTTTAAGAAACGCTCTCGGCCAGATCTCCGAACCAGAGCCAGAAGGAATCTGTACAAAAAACAGGAGTCAGAACAAGCAGGGGTTGCTAAGGATGCAAAATCTGTGGCCTCAGATGTTCCCCTCTACAAGGATGGGGAGGCTAAGACTGACCCAGCAGGGCTGAGCAGTCCCCATCTGCCAGGCACATCCTCTGCAGCACCCGACCTGGAGGGTCCCGAATTCCCAGTTGAGTCTGTGGCTTCTCGGATCCAGGCTGAGCCAGACAACTTGGCACGTGCCTCTGCATCTCCAGACAGAATTCCTAGCCTGCCTCAGGAAACTGTGGATCAGGAACCCAAGGATCAGAAGAGGAAATCCTTTGAGCAGGCGGCCTCTGCATCCTTTCCCGAAAAGAAGCCCCGGCTTGAAGATCGTCAGTCCTTTCGTAACACAATTGAAAGTGTTCACACCGAAAAGCCACAGCCCACTAAAGAGGAGCCCAAAGTCCCGCCCATCCGGATTCAACTTTCACGTATCAAACCACCCTGGGTGGTTAAAGGTCAGCCCACTTACCAGATATGCCCCCGGATCATCCCCACCACGGAGTCCTCCTGCCGGGGTTGGACTGGCGCCAGGACCCTCGCAGACATTAAAGCCCGTGCTCTGCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCCACCACTGCCATCGGAGGGGGGGGTGGCCCGGGTGGAGGTGGCGGCGGGGCCACCGATGAGGGAGGTGGCAGAGGCAGCAGCAGTGGTGATGGTGGTGAGGCCTGTGGCCACCCTGAGCCCAGGGGAGGCCCGAGCACCCCTGGAAAGTGTACGTCAGATCTACAGCGAACACAACTACTGCCGCCTTATCCTCTAAATGGGGAGCATACCCAGGCCGGAACTGCCATGTCCAGAGCTAGGAGAGAGGACCTGCCTTCTCTGAGAAAGGAGGAAAGCTGCCTACTACAGAGGGCTACAGTTGGACTCACAGATGGGCTAGGAGATGCCTCCCAACTCCCCGTTGCTCCCACTGGGGACCAGCCATGCCAGGCCTTGCCCCTACTGTCCTCCCAAACCTCAGTAGCTGAGAGATTAGTGGAGCAGCCTCAGTTGCATCCGGATGTTAGAACTGAATGTGAGTCTGGCACCACTTCCTGGGAAAGTGATGATGAGGAGCAAGGACCCACCGTTCCTGCAGACAATGGTCCCATTCTGTCTCTAGTGGGAGATGATACATTAGAGAAAGGAACTGGCCAAGCTCTTGACAGTCATCCCACTATGAAGGATCCTGTAAATGTGACCCCCAGTTCCACACCTGAATCCTCACCGACTGATTGCCTGCAGAACAGAGCATTTGATGACGAATTAGGGCTTGGTGGCTCATGCCCTCCTATGAGGGAAAGTGATACTAGACAAGAAAACTTGAAAACCAAGGCTCTCGTTTCTAACAGTTCTTTGCATTGGATACCCATCCCATCGAATGATGAGGTAGTGAAACAGCCCAAACCAGAATCCAGAGAACACATACCATCTGTTGAGCCCCAGGTTGGAGAGGAGTGGGAGAAAGCTGCTCCCACCCCTCCTGCATTGCCTGGGGATTTGACAGCTGAGGAGGGTCTAGATCCTCTTGACAGCCTTACTTCACTCTGGACTGTGCCATCTCGAGGAGGCAGTGACAGCAATGGCAGTTACTGTCAACAGGTGGACATTGAAAAGCTGAAAATCAACGGAGACTCTGAAGCACTGAGTCCTCACGGTGAGTCCACGGATACAGCCTCTGACTTTGAAGGTCACCTCACGGAGGACAGCAGTGAGGCTGACACTAGAGAAGCTGCAGTGACAAAGGGATCTTCGGTGGACAAGGATGAGAAACCCAATTGGAACCAATCTGCCCCACTGTCCAAGGTGAATGGTGACATGCGTCTGGTTACAAGGACAGATGGGATGGTTGCTCCTCAGAGCTGGGTGTCTCGAGTATGTGCGGTCCGCCAAAAGATCCCAGATTCCCTACTGCTGGCCAGTACTGAGTACCAGCCAAGAGCCGTGTGCCTGTCCATGCCTGGGTCCTCAGTGGAGGCCACTAACCCACTTGTGATGCAGTTGCTGCAGGGTAGCTTGCCCCTAGAGAAGGTTCTTCCACCAGCCCACGATGACAGCATGTCAGAATCCCCACAAGTACCACTTACAAAAGACCAGAGCCATGGCTCGCTACGCATGGGATCTTTACATGGTCTTGGAAAAAACAGTGGCATGGTTGATGGAAGCAGCCCCAGTTCTTTAAGGGCTTTGAAGGAGCCTCTTCTGCCAGATAGCTGTGAAACAGGCACTGGTCTTGCCAGGATTGAGGCCACCCAGGCTCCTGGAGCACCCCAAAAGAATTGCAAGGCAGTCCCAAGTTTTGACTCCCTCCATCCAGTGACAAATCCCATTACATCCTCTAGGAAACTGGAAGAAATGGATTCCAAAGAGCAGTTCTCTTCCTTTAGTTGTGAAGATCAGAAGGAAGTCCGTGCTATGTCACAGGACAGTAATTCAAATGCTGCTCCAGGAAAGAGCCCAGGAGATCTTACTACCTCGAGAACACCTCGTTTCTCATCTCCAAATGTGATCTCCTTTGGTCCAGAGCAGACAGGTCGGGCCCTGGGTGATCAGAGCAATGTTACAGGCCAAGGGAAGAAGCTTTTTGGCTCTGGGAATGTGGCTGCAACCCTTCAGCGCCCCAGGCCTGCGGACCCGATGCCTCTTCCTGCTGAGATCCCTCCAGTTTTTCCCAGTGGGAAGTTGGGACCAAGCACAAACTCCATGTCTGGTGGGGTACAGACTCCAAGGGAAGACTGGGCTCCAAAGCCACATGCCTTTGTTGGCAGCGTCAAGAATGAGAAGACTTTTGTGGGGGGTCCTCTTAAGGCAAATGCCGAGAACAGGAAAGCTACTGGGCATAGTCCCCTGGAACTGGTGGGTCACTTGGAAGGGATGCCCTTTGTCATGGACTTGCCCTTCTGGAAATTACCCCGAGAGCCAGGGAAGGGGCTCAGTGAGCCTCTGGAGCCTTCTTCTCTCCCCTCCCAACTCAGCATCAAGCAGGCATTTTATGGGAAGCTTTCTAAACTCCAACTGAGTTCCACCAGCTTTAATTATTCCTCTAGCTCTCCCACCTTTCCCAAAGGCCTTGCTGGAAGTGTGGTGCAGCTGAGCCACAAAGCAAACTTTGGTGCGAGCCACAGTGCATCACTTTCCTTGCAAATGTTCACTGACAGCAGCACGGTGGAAAGCATCTCGCTCCAGTGTGCGTGCAGCCTGAAAGCCATGATCATGTGCCAAGGCTGCGGTGCGTTCTGTCACGATGACTGTATTGGACCCTCAAAGCTCTGTGTATTGTGCCTTGTGGTGAGATAAT'\n",
    "  \n",
    "    #Create a list of the codons they called missense variants at\n",
    "    import csv\n",
    "    filename = csv_file_variants\n",
    "    with open(filename, 'r') as csvfile:\n",
    "        read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "        row_count=0\n",
    "\n",
    "        ASXL1_missense_variant_list = []\n",
    "        ASXL1_missense_codons =[]\n",
    "        ASXL1_missense_variant_DICT = {}\n",
    "\n",
    "        for row in read_reader: #each row contains aa number (e.g. 11) and variant (e.g. p.D11A)\n",
    "    #         print(row)\n",
    "            if row_count>0: \n",
    "                codon0 = int(row[0])-1\n",
    "                codon1 = int(row[0])\n",
    "                variant = row[1]\n",
    "                ASXL1_missense_codons.append(codon0) #List of codon numbers that they called variants at\n",
    "                ASXL1_missense_variant_DICT[row_count]=(codon1, variant) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "            row_count=row_count+1\n",
    "            \n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    ASXL1_missense_codons_DICT = {}\n",
    "    calls_ASXL1_missense = {}\n",
    "\n",
    "    for i in ASXL1_missense_codons: #Iterates over the list of codon numbers\n",
    "        ASXL1_missense_codons_DICT[i+1]= (ASXL1_codons[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "        \n",
    "    for k,v in ASXL1_missense_codons_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_ASXL1_missense[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "        \n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_ASXL1_missense_extra = {}\n",
    "\n",
    "    for k, v in calls_ASXL1_missense.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_ASXL1_missense_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_ASXL1_missense_extra2 = {}\n",
    "\n",
    "    for k, v in calls_ASXL1_missense_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_ASXL1_missense_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "            \n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    ASXL1_position_in_list_dict_missense={}\n",
    "    for k, v in ASXL1_missense_variant_DICT.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_ASXL1_missense[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_ASXL1_missense[v[0]]\n",
    "        codon_seq=calls_ASXL1_missense[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_ASXL1_missense_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                ASXL1_position_in_list_dict_missense[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "            \n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in ASXL1_position_in_list_dict_missense.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "#     print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "                \n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    ASXL1_bases_missense = {}\n",
    "    for k, v in ASXL1_position_in_list_dict_missense.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        ASXL1_bases_missense[k]=(v, base_changes)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    ASXL1_bases_missense2 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_missense.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        ASXL1_bases_missense2[k] = (v, codon_positions)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    ASXL1_bases_missense3 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_missense2.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        ASXL1_bases_missense3[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    ASXL1_bases_missense4 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_missense3.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check(i)\n",
    "                sites_trio_missense.append([adapted_trio])\n",
    "                ASXL1_bases_missense4[k] = (v, sites_trio_missense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_missense = [ASXL1_exons_extra_bases[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_missense.append(site_trio_missense)\n",
    "                    ASXL1_bases_missense4[k] = (v, sites_trio_missense)\n",
    "                    \n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    ASXL1_sites_missense5 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in ASXL1_bases_missense4.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        ASXL1_sites_missense5[a]=info\n",
    "        a=a+1\n",
    "        \n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    ASXL1_sites_missense6={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in ASXL1_sites_missense5.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "    \n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        ASXL1_sites_missense6[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "        \n",
    "        \n",
    "    #Assign the mutation rates to each site...\n",
    "    ASXL1_mutation_rates_missense = {}\n",
    "    for k, v in ASXL1_sites_missense6.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            ASXL1_mutation_rates_missense[k] = (v, mutation_rate)\n",
    "\n",
    "    ASXL1_mutation_rate_missense_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in ASXL1_mutation_rates_missense.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        ASXL1_mutation_rate_missense_sum.append(mutation_rate_value)\n",
    "        \n",
    "    ASXL1_Summed_mutation_rate_missense_added = np.sum(ASXL1_mutation_rate_missense_sum)\n",
    "\n",
    "    print('Total summed mutation rate across ASXL1 missense for this study = ', ASXL1_Summed_mutation_rate_missense_added)\n",
    "    \n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all #in format: column 1 = 882, column 2 = p.R882H"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TP53"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_list_nonsense_variants_TP53(csv_file_variants):\n",
    "\n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base = [74, 75, 96, 375, 672, 782, 919, 993, 1100]\n",
    "   \n",
    "    def intron_exon_check(position_in_gene_number):\n",
    "        if position_in_gene_number == 74:\n",
    "            gene_pos = 74\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 75:\n",
    "            gene_pos = 75\n",
    "            first_base = 'G'\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = TP53_exons_extra_bases[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 96:\n",
    "            gene_pos = 96\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 375:\n",
    "            gene_pos = 375\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 672:\n",
    "            gene_pos = 672\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 782:\n",
    "            gene_pos = 782\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 919:\n",
    "            gene_pos = 919\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 993:\n",
    "            gene_pos = 993\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1100:\n",
    "            gene_pos = 1100\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #TP53 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    TP53_exons = 'ATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGACATAGTGTGGTGGTGCCCTATGAGCCGCCTGAGGTTGGCTCTGACTGTACCACCATCCACTACAACTACATGTGTAACAGTTCCTGCATGGGCGGCATGAACCGGAGGCCCATCCTCACCATCATCACACTGGAAGACTCCAGTGGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGAGGAAGAGAATCTCCGCAAGAAAGGGGAGCCTCACCACGAGCTGCCCCCAGGGAGCACTAAGCGAGCACTGCCCAACAACACCAGCTCCTCTCCCCAGCCAAAGAAGAAACCACTGGATGGAGAATATTTCACCCTTCAGATCCGTGGGCGTGAGCGCTTCGAGATGTTCCGAGAGCTGAATGAGGCCTTGGAACTCAAGGATGCCCAGGCTGGGAAGGAGCCAGGGGGGAGCAGGGCTCACTCCAGCCACCTGAAGTCCAAAAAGGGTCAGTCTACCTCCCGCCATAAAAAACTCATGTTCAAGACAGAAGGGCCTGACTCAGACTGA'\n",
    "    \n",
    "    #Split the sequence into codons \n",
    "    TP53_codons = [TP53_exons[i:i+3] for i in range(0, len(TP53_exons), 3)]\n",
    "\n",
    "    #TP53 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    TP53_exons_extra_bases = 'CATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGACATAGTGTGGTGGTGCCCTATGAGCCGCCTGAGGTTGGCTCTGACTGTACCACCATCCACTACAACTACATGTGTAACAGTTCCTGCATGGGCGGCATGAACCGGAGGCCCATCCTCACCATCATCACACTGGAAGACTCCAGTGGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGAGGAAGAGAATCTCCGCAAGAAAGGGGAGCCTCACCACGAGCTGCCCCCAGGGAGCACTAAGCGAGCACTGCCCAACAACACCAGCTCCTCTCCCCAGCCAAAGAAGAAACCACTGGATGGAGAATATTTCACCCTTCAGATCCGTGGGCGTGAGCGCTTCGAGATGTTCCGAGAGCTGAATGAGGCCTTGGAACTCAAGGATGCCCAGGCTGGGAAGGAGCCAGGGGGGAGCAGGGCTCACTCCAGCCACCTGAAGTCCAAAAAGGGTCAGTCTACCTCCCGCCATAAAAAACTCATGTTCAAGACAGAAGGGCCTGACTCAGACTGAC'\n",
    "    \n",
    "    #Create a list of the codons they called nonsense variants at\n",
    "    import csv\n",
    "    filename = csv_file_variants\n",
    "    with open(filename, 'r') as csvfile:\n",
    "        read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "        row_count=0\n",
    "\n",
    "        TP53_nonsense_variant_list = []\n",
    "        TP53_nonsense_codons =[]\n",
    "        TP53_nonsense_variant_DICT = {}\n",
    "\n",
    "        for row in read_reader: #each row contains aa number (e.g. 11) and variant (e.g. p.D11A)\n",
    "    #         print(row)\n",
    "            if row_count>0: \n",
    "                codon0 = int(row[0])-1\n",
    "                codon1 = int(row[0])\n",
    "                variant = row[1]\n",
    "                TP53_nonsense_codons.append(codon0) #List of codon numbers that they called variants at\n",
    "                TP53_nonsense_variant_DICT[row_count]=(codon1, variant) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "            row_count=row_count+1\n",
    "            \n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    TP53_nonsense_codons_DICT = {}\n",
    "    calls_nonsense = {}\n",
    "\n",
    "    for i in TP53_nonsense_codons: #Iterates over the list of codon numbers\n",
    "        TP53_nonsense_codons_DICT[i+1]= (TP53_codons[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "        \n",
    "    for k,v in TP53_nonsense_codons_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_nonsense[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "        \n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_nonsense_extra = {}\n",
    "\n",
    "    for k, v in calls_nonsense.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_nonsense_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_nonsense_extra2 = {}\n",
    "\n",
    "    for k, v in calls_nonsense_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_nonsense_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "            \n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    TP53_position_in_list_dict_nonsense={}\n",
    "    for k, v in TP53_nonsense_variant_DICT.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_nonsense[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_nonsense[v[0]]\n",
    "        codon_seq=calls_nonsense[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_nonsense_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                TP53_position_in_list_dict_nonsense[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "            \n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in TP53_position_in_list_dict_nonsense.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "#     print('positions at which nonsense variants called = ', aa_positions_variants_called)\n",
    "                \n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TP53_bases_nonsense = {}\n",
    "    for k, v in TP53_position_in_list_dict_nonsense.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TP53_bases_nonsense[k]=(v, base_changes)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TP53_bases_nonsense2 = {}\n",
    "\n",
    "    for k, v in TP53_bases_nonsense.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TP53_bases_nonsense2[k] = (v, codon_positions)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TP53_bases_nonsense3 = {}\n",
    "\n",
    "    for k, v in TP53_bases_nonsense2.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TP53_bases_nonsense3[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TP53_bases_nonsense4 = {}\n",
    "\n",
    "    for k, v in TP53_bases_nonsense3.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_nonsense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check(i)\n",
    "                sites_trio_nonsense.append([adapted_trio])\n",
    "                TP53_bases_nonsense4[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_nonsense = [TP53_exons_extra_bases[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_nonsense.append(site_trio_nonsense)\n",
    "                    TP53_bases_nonsense4[k] = (v, sites_trio_nonsense)\n",
    "                    \n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TP53_sites_nonsense5 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TP53_bases_nonsense4.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        TP53_sites_nonsense5[a]=info\n",
    "        a=a+1\n",
    "#     print(TP53_sites_nonsense5)\n",
    "        \n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TP53_sites_nonsense6={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TP53_sites_nonsense5.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "    \n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TP53_sites_nonsense6[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "        \n",
    "        \n",
    "    #Assign the mutation rates to each site...\n",
    "    TP53_mutation_rates_nonsense = {}\n",
    "    for k, v in TP53_sites_nonsense6.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            TP53_mutation_rates_nonsense[k] = (v, mutation_rate)\n",
    "\n",
    "    TP53_mutation_rate_nonsense_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TP53_mutation_rates_nonsense.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        TP53_mutation_rate_nonsense_sum.append(mutation_rate_value)\n",
    "    \n",
    "    TP53_Summed_mutation_rate_nonsense_added = np.sum(TP53_mutation_rate_nonsense_sum)\n",
    "\n",
    "    print('Total summed mutation rate across TP53 nonsense for this study = ', TP53_Summed_mutation_rate_nonsense_added)\n",
    "    \n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all #in format: column 1 = 882, column 2 = p.R882*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_list_missense_variants_TP53(csv_file_variants):\n",
    "    \n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base_TP53 = [74, 75, 96, 375, 672, 782, 919, 993, 1100]\n",
    "   \n",
    "    def intron_exon_check(position_in_gene_number):\n",
    "        if position_in_gene_number == 74:\n",
    "            gene_pos = 74\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 75:\n",
    "            gene_pos = 75\n",
    "            first_base = 'G'\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = TP53_exons_extra_bases[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 96:\n",
    "            gene_pos = 96\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 375:\n",
    "            gene_pos = 375\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 672:\n",
    "            gene_pos = 672\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 782:\n",
    "            gene_pos = 782\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 919:\n",
    "            gene_pos = 919\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 993:\n",
    "            gene_pos = 993\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1100:\n",
    "            gene_pos = 1100\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #TP53 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    TP53_exons = 'ATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGACATAGTGTGGTGGTGCCCTATGAGCCGCCTGAGGTTGGCTCTGACTGTACCACCATCCACTACAACTACATGTGTAACAGTTCCTGCATGGGCGGCATGAACCGGAGGCCCATCCTCACCATCATCACACTGGAAGACTCCAGTGGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGAGGAAGAGAATCTCCGCAAGAAAGGGGAGCCTCACCACGAGCTGCCCCCAGGGAGCACTAAGCGAGCACTGCCCAACAACACCAGCTCCTCTCCCCAGCCAAAGAAGAAACCACTGGATGGAGAATATTTCACCCTTCAGATCCGTGGGCGTGAGCGCTTCGAGATGTTCCGAGAGCTGAATGAGGCCTTGGAACTCAAGGATGCCCAGGCTGGGAAGGAGCCAGGGGGGAGCAGGGCTCACTCCAGCCACCTGAAGTCCAAAAAGGGTCAGTCTACCTCCCGCCATAAAAAACTCATGTTCAAGACAGAAGGGCCTGACTCAGACTGA'\n",
    "    \n",
    "    #Split the sequence into codons \n",
    "    TP53_codons = [TP53_exons[i:i+3] for i in range(0, len(TP53_exons), 3)]\n",
    "\n",
    "    #TP53 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    TP53_exons_extra_bases = 'CATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGACATAGTGTGGTGGTGCCCTATGAGCCGCCTGAGGTTGGCTCTGACTGTACCACCATCCACTACAACTACATGTGTAACAGTTCCTGCATGGGCGGCATGAACCGGAGGCCCATCCTCACCATCATCACACTGGAAGACTCCAGTGGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGAGGAAGAGAATCTCCGCAAGAAAGGGGAGCCTCACCACGAGCTGCCCCCAGGGAGCACTAAGCGAGCACTGCCCAACAACACCAGCTCCTCTCCCCAGCCAAAGAAGAAACCACTGGATGGAGAATATTTCACCCTTCAGATCCGTGGGCGTGAGCGCTTCGAGATGTTCCGAGAGCTGAATGAGGCCTTGGAACTCAAGGATGCCCAGGCTGGGAAGGAGCCAGGGGGGAGCAGGGCTCACTCCAGCCACCTGAAGTCCAAAAAGGGTCAGTCTACCTCCCGCCATAAAAAACTCATGTTCAAGACAGAAGGGCCTGACTCAGACTGAC'\n",
    "    \n",
    "    #Create a list of the codons they called missense variants at\n",
    "    import csv\n",
    "    filename = csv_file_variants\n",
    "    with open(filename, 'r') as csvfile:\n",
    "        read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "        row_count=0\n",
    "\n",
    "        TP53_missense_variant_list = []\n",
    "        TP53_missense_codons =[]\n",
    "        TP53_missense_variant_DICT = {}\n",
    "\n",
    "        for row in read_reader: #each row contains aa number (e.g. 11) and variant (e.g. p.D11A)\n",
    "    #         print(row)\n",
    "            if row_count>0: \n",
    "                codon0 = int(row[0])-1\n",
    "                codon1 = int(row[0])\n",
    "                variant = row[1]\n",
    "                TP53_missense_codons.append(codon0) #List of codon numbers that they called variants at\n",
    "                TP53_missense_variant_DICT[row_count]=(codon1, variant) #Dictionary:  K = codon number, V = variant at that codon (e.g. 76: p.V76L)\n",
    "\n",
    "            row_count=row_count+1\n",
    "            \n",
    "    #Looking at what possible amino acid changes are possible at those codons\n",
    "    TP53_missense_codons_DICT = {}\n",
    "    calls_TP53_missense = {}\n",
    "\n",
    "    for i in TP53_missense_codons: #Iterates over the list of codon numbers\n",
    "        TP53_missense_codons_DICT[i+1]= (TP53_codons[i]) #Creates of dicitonary of key = codon number, value = bases (sequence) at that codon\n",
    "        #e.g. 12: CGT, 882: GCA\n",
    "        \n",
    "    for k,v in TP53_missense_codons_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes_letters.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_TP53_missense[k]=(v, changes)\n",
    "        #e.g. 12: (CGT, 'K', '*', 'T'etc...)\n",
    "        \n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_TP53_missense_extra = {}\n",
    "\n",
    "    for k, v in calls_TP53_missense.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_TP53_missense_extra[k] = (Amino_acid_change, v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_TP53_missense_extra2 = {}\n",
    "\n",
    "    for k, v in calls_TP53_missense_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_TP53_missense_extra2[k] = (Amino_acid_abbrev, v[1])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "            \n",
    "    #Create a dictionary containing info on which position in the codon is affected to create the variant\n",
    "    TP53_position_in_list_dict_missense={}\n",
    "    for k, v in TP53_missense_variant_DICT.items(): #Iterates over the dictionary of (key = row_count, value = aa_pos, variant (e.g. 3: (76, p.V76L))\n",
    "        observed_change=(v[1])[-1] #Selects the amino acid change (e.g. the L from p.V76L)\n",
    "        possible_changes=calls_TP53_missense[v[0]][1] #possible changes are the second value in the dictionary of (k = codon number, value = (original amino acid, possible changes))\n",
    "        possible_changes_info=calls_TP53_missense[v[0]]\n",
    "        codon_seq=calls_TP53_missense[v[0]][0]\n",
    "        possible_changes_string = \"\".join(possible_changes)\n",
    "        if v[1][2] == calls_TP53_missense_extra2[v[0]][0]:  #v[1][2] = the original amino acid letter (i.e if this aa is in this transcript)\n",
    "            position_in_list = -1\n",
    "            positions_in_list = []\n",
    "            while True:\n",
    "                position_in_list = possible_changes_string.find(observed_change, position_in_list+1) #counts position in possible changes (if 0-2 = 1st base of codon, if 3-5 = 2nd base, if 6-8 = 3rd base)\n",
    "                if position_in_list == -1: break\n",
    "                positions_in_list.append(position_in_list)\n",
    "                TP53_position_in_list_dict_missense[k]=(positions_in_list, possible_changes_info, codon_seq, v[0], v[1])\n",
    "            \n",
    "    aa_positions_variants_called = []\n",
    "    for k, v in TP53_position_in_list_dict_missense.items():\n",
    "        aa_positions_variants_called.append(v[3])\n",
    "#     print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "                \n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TP53_bases_missense = {}\n",
    "    for k, v in TP53_position_in_list_dict_missense.items():\n",
    "        codon_seq = v[2]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TP53_bases_missense[k]=(v, base_changes)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TP53_bases_missense2 = {}\n",
    "\n",
    "    for k, v in TP53_bases_missense.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TP53_bases_missense2[k] = (v, codon_positions)\n",
    "        \n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TP53_bases_missense3 = {}\n",
    "\n",
    "    for k, v in TP53_bases_missense2.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = v[0][0][3]\n",
    "        variant = v[0][0][4]\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(v[0][0][3])-1)*3)+codon_pos #codon number (v[0][0][3]) minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][2][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TP53_bases_missense3[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, variant)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TP53_bases_missense4 = {}\n",
    "\n",
    "    for k, v in TP53_bases_missense3.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_TP53:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check(i)\n",
    "                sites_trio_missense.append([adapted_trio])\n",
    "                TP53_bases_missense4[k] = (v, sites_trio_missense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_missense = [TP53_exons_extra_bases[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_missense.append(site_trio_missense)\n",
    "                    TP53_bases_missense4[k] = (v, sites_trio_missense)\n",
    "                    \n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TP53_sites_missense5 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TP53_bases_missense4.items():\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "        variant = v[0][5]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, variant])\n",
    "            change = change+1\n",
    "        TP53_sites_missense5[a]=info\n",
    "        a=a+1\n",
    "        \n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TP53_sites_missense6={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TP53_sites_missense5.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "    \n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        variant = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TP53_sites_missense6[k] = (v[0],changes3, codon_n_912, variant) #site_trio = the affected base and its neighbours\n",
    "        \n",
    "        \n",
    "    #Assign the mutation rates to each site...\n",
    "    TP53_mutation_rates_missense = {}\n",
    "    for k, v in TP53_sites_missense6.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            TP53_mutation_rates_missense[k] = (v, mutation_rate)\n",
    "\n",
    "    TP53_mutation_rate_missense_sum = []\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TP53_mutation_rates_missense.items():\n",
    "    #     print(v[1])\n",
    "        aa_pos = v[0][2]\n",
    "        site = v[0][1]\n",
    "        mutation_rate_value = v[1] \n",
    "        variant = v[0][3]\n",
    "        mutation_rates_list.append((aa_pos, site, mutation_rate_value, variant))\n",
    "        TP53_mutation_rate_missense_sum.append(mutation_rate_value)\n",
    "        \n",
    "    TP53_Summed_mutation_rate_missense_added = np.sum(TP53_mutation_rate_missense_sum)\n",
    "\n",
    "    print('Total summed mutation rate across TP53 missense for this study = ', TP53_Summed_mutation_rate_missense_added)\n",
    "    \n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all #in format: column 1 = 882, column 2 = p.R882H"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Calculating mutation rates from a list of variants"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DNMT3A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(lower_range, upper_range):\n",
    "    \n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base_DNMT3A_NP072046 = [1279, 1431, 1476, 1554, 1852, 2083, 2085, 2175, 2322, 2478]\n",
    "\n",
    "    def intron_exon_check_DNMT3A_NP072046(position_in_gene_number):\n",
    "        if position_in_gene_number == 1279:\n",
    "            gene_pos = 1279\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1429:\n",
    "            gene_pos = 1429\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1474:\n",
    "            gene_pos = 1474\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1554:\n",
    "            gene_pos = 1554\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1852:\n",
    "            gene_pos = 1852\n",
    "            first_base = 'G'\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2083:\n",
    "            gene_pos = 2083\n",
    "            first_base = 'G'\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2082:\n",
    "            gene_pos = 2082\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2173:\n",
    "            gene_pos = 2173\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2322:\n",
    "            gene_pos = 2322\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2478:\n",
    "            gene_pos = 2478\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "\n",
    "    #DNMT3A transcript used:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000119772;r=2:25455845-25564754;t=ENST00000321117\n",
    "    DNMT3A_exons_NP_072046 = 'ATGCCCGCCATGCCCTCCAGCGGCCCCGGGGACACCAGCAGCTCTGCTGCGGAGCGGGAGGAGGACCGAAAGGACGGAGAGGAGCAGGAGGAGCCGCGTGGCAAGGAGGAGCGCCAAGAGCCCAGCACCACGGCACGGAAGGTGGGGCGGCCTGGGAGGAAGCGCAAGCACCCCCCGGTGGAAAGCGGTGACACGCCAAAGGACCCTGCGGTGATCTCCAAGTCCCCATCCATGGCCCAGGACTCAGGCGCCTCAGAGCTATTACCCAATGGGGACTTGGAGAAGCGGAGTGAGCCCCAGCCAGAGGAGGGGAGCCCTGCTGGGGGGCAGAAGGGCGGGGCCCCAGCAGAGGGAGAGGGTGCAGCTGAGACCCTGCCTGAAGCCTCAAGAGCAGTGGAAAATGGCTGCTGCACCCCCAAGGAGGGCCGAGGAGCCCCTGCAGAAGCGGGCAAAGAACAGAAGGAGACCAACATCGAATCCATGAAAATGGAGGGCTCCCGGGGCCGGCTGCGGGGTGGCTTGGGCTGGGAGTCCAGCCTCCGTCAGCGGCCCATGCCGAGGCTCACCTTCCAGGCGGGGGACCCCTACTACATCAGCAAGCGCAAGCGGGACGAGTGGCTGGCACGCTGGAAAAGGGAGGCTGAGAAGAAAGCCAAGGTCATTGCAGGAATGAATGCTGTGGAAGAAAACCAGGGGCCCGGGGAGTCTCAGAAGGTGGAGGAGGCCAGCCCTCCTGCTGTGCAGCAGCCCACTGACCCCGCATCCCCCACTGTGGCTACCACGCCTGAGCCCGTGGGGTCCGATGCTGGGGACAAGAATGCCACCAAAGCAGGCGATGACGAGCCAGAGTACGAGGACGGCCGGGGCTTTGGCATTGGGGAGCTGGTGTGGGGGAAACTGCGGGGCTTCTCCTGGTGGCCAGGCCGCATTGTGTCTTGGTGGATGACGGGCCGGAGCCGAGCAGCTGAAGGCACCCGCTGGGTCATGTGGTTCGGAGACGGCAAATTCTCAGTGGTGTGTGTTGAGAAGCTGATGCCGCTGAGCTCGTTTTGCAGTGCGTTCCACCAGGCCACGTACAACAAGCAGCCCATGTACCGCAAAGCCATCTACGAGGTCCTGCAGGTGGCCAGCAGCCGCGCGGGGAAGCTGTTCCCGGTGTGCCACGACAGCGATGAGAGTGACACTGCCAAGGCCGTGGAGGTGCAGAACAAGCCCATGATTGAATGGGCCCTGGGGGGCTTCCAGCCTTCTGGCCCTAAGGGCCTGGAGCCACCAGAAGAAGAGAAGAATCCCTACAAAGAAGTGTACACGGACATGTGGGTGGAACCTGAGGCAGCTGCCTACGCACCACCTCCACCAGCCAAAAAGCCCCGGAAGAGCACAGCGGAGAAGCCCAAGGTCAAGGAGATTATTGATGAGCGCACAAGAGAGCGGCTGGTGTACGAGGTGCGGCAGAAGTGCCGGAACATTGAGGACATCTGCATCTCCTGTGGGAGCCTCAATGTTACCCTGGAACACCCCCTCTTCGTTGGAGGAATGTGCCAAAACTGCAAGAACTGCTTTCTGGAGTGTGCGTACCAGTACGACGACGACGGCTACCAGTCCTACTGCACCATCTGCTGTGGGGGCCGTGAGGTGCTCATGTGCGGAAACAACAACTGCTGCAGGTGCTTTTGCGTGGAGTGTGTGGACCTCTTGGTGGGGCCGGGGGCTGCCCAGGCAGCCATTAAGGAAGACCCCTGGAACTGCTACATGTGCGGGCACAAGGGTACCTACGGGCTGCTGCGGCGGCGAGAGGACTGGCCCTCCCGGCTCCAGATGTTCTTCGCTAATAACCACGACCAGGAATTTGACCCTCCAAAGGTTTACCCACCTGTCCCAGCTGAGAAGAGGAAGCCCATCCGGGTGCTGTCTCTCTTTGATGGAATCGCTACAGGGCTCCTGGTGCTGAAGGACTTGGGCATTCAGGTGGACCGCTACATTGCCTCGGAGGTGTGTGAGGACTCCATCACGGTGGGCATGGTGCGGCACCAGGGGAAGATCATGTACGTCGGGGACGTCCGCAGCGTCACACAGAAGCATATCCAGGAGTGGGGCCCATTCGATCTGGTGATTGGGGGCAGTCCCTGCAATGACCTCTCCATCGTCAACCCTGCTCGCAAGGGCCTCTACGAGGGCACTGGCCGGCTCTTCTTTGAGTTCTACCGCCTCCTGCATGATGCGCGGCCCAAGGAGGGAGATGATCGCCCCTTCTTCTGGCTCTTTGAGAATGTGGTGGCCATGGGCGTTAGTGACAAGAGGGACATCTCGCGATTTCTCGAGTCCAACCCTGTGATGATTGATGCCAAAGAAGTGTCAGCTGCACACAGGGCCCGCTACTTCTGGGGTAACCTTCCCGGTATGAACAGGCCGTTGGCATCCACTGTGAATGATAAGCTGGAGCTGCAGGAGTGTCTGGAGCATGGCAGGATAGCCAAGTTCAGCAAAGTGAGGACCATTACTACGAGGTCAAACTCCATAAAGCAGGGCAAAGACCAGCATTTTCCTGTCTTCATGAATGAGAAAGAGGACATCTTATGGTGCACTGAAATGGAAAGGGTATTTGGTTTCCCAGTCCACTATACTGACGTCTCCAACATGAGCCGCTTGGCGAGGCAGAGACTGCTGGGCCGGTCATGGAGCGTGCCAGTCATCCGCCACCTCTTCGCTCCGCTGAAGGAGTATTTTGCGTGTGTGTAA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    DNMT3A_codons_NP_072046 = [DNMT3A_exons_NP_072046[i:i+3] for i in range(0, len(DNMT3A_exons_NP_072046), 3)]\n",
    "\n",
    "    #DNMT3A transcript used:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000119772;r=2:25455845-25564754;t=ENST00000321117\n",
    "    DNMT3A_exons_extra_bases_NP_072046 = 'GATGCCCGCCATGCCCTCCAGCGGCCCCGGGGACACCAGCAGCTCTGCTGCGGAGCGGGAGGAGGACCGAAAGGACGGAGAGGAGCAGGAGGAGCCGCGTGGCAAGGAGGAGCGCCAAGAGCCCAGCACCACGGCACGGAAGGTGGGGCGGCCTGGGAGGAAGCGCAAGCACCCCCCGGTGGAAAGCGGTGACACGCCAAAGGACCCTGCGGTGATCTCCAAGTCCCCATCCATGGCCCAGGACTCAGGCGCCTCAGAGCTATTACCCAATGGGGACTTGGAGAAGCGGAGTGAGCCCCAGCCAGAGGAGGGGAGCCCTGCTGGGGGGCAGAAGGGCGGGGCCCCAGCAGAGGGAGAGGGTGCAGCTGAGACCCTGCCTGAAGCCTCAAGAGCAGTGGAAAATGGCTGCTGCACCCCCAAGGAGGGCCGAGGAGCCCCTGCAGAAGCGGGCAAAGAACAGAAGGAGACCAACATCGAATCCATGAAAATGGAGGGCTCCCGGGGCCGGCTGCGGGGTGGCTTGGGCTGGGAGTCCAGCCTCCGTCAGCGGCCCATGCCGAGGCTCACCTTCCAGGCGGGGGACCCCTACTACATCAGCAAGCGCAAGCGGGACGAGTGGCTGGCACGCTGGAAAAGGGAGGCTGAGAAGAAAGCCAAGGTCATTGCAGGAATGAATGCTGTGGAAGAAAACCAGGGGCCCGGGGAGTCTCAGAAGGTGGAGGAGGCCAGCCCTCCTGCTGTGCAGCAGCCCACTGACCCCGCATCCCCCACTGTGGCTACCACGCCTGAGCCCGTGGGGTCCGATGCTGGGGACAAGAATGCCACCAAAGCAGGCGATGACGAGCCAGAGTACGAGGACGGCCGGGGCTTTGGCATTGGGGAGCTGGTGTGGGGGAAACTGCGGGGCTTCTCCTGGTGGCCAGGCCGCATTGTGTCTTGGTGGATGACGGGCCGGAGCCGAGCAGCTGAAGGCACCCGCTGGGTCATGTGGTTCGGAGACGGCAAATTCTCAGTGGTGTGTGTTGAGAAGCTGATGCCGCTGAGCTCGTTTTGCAGTGCGTTCCACCAGGCCACGTACAACAAGCAGCCCATGTACCGCAAAGCCATCTACGAGGTCCTGCAGGTGGCCAGCAGCCGCGCGGGGAAGCTGTTCCCGGTGTGCCACGACAGCGATGAGAGTGACACTGCCAAGGCCGTGGAGGTGCAGAACAAGCCCATGATTGAATGGGCCCTGGGGGGCTTCCAGCCTTCTGGCCCTAAGGGCCTGGAGCCACCAGAAGAAGAGAAGAATCCCTACAAAGAAGTGTACACGGACATGTGGGTGGAACCTGAGGCAGCTGCCTACGCACCACCTCCACCAGCCAAAAAGCCCCGGAAGAGCACAGCGGAGAAGCCCAAGGTCAAGGAGATTATTGATGAGCGCACAAGAGAGCGGCTGGTGTACGAGGTGCGGCAGAAGTGCCGGAACATTGAGGACATCTGCATCTCCTGTGGGAGCCTCAATGTTACCCTGGAACACCCCCTCTTCGTTGGAGGAATGTGCCAAAACTGCAAGAACTGCTTTCTGGAGTGTGCGTACCAGTACGACGACGACGGCTACCAGTCCTACTGCACCATCTGCTGTGGGGGCCGTGAGGTGCTCATGTGCGGAAACAACAACTGCTGCAGGTGCTTTTGCGTGGAGTGTGTGGACCTCTTGGTGGGGCCGGGGGCTGCCCAGGCAGCCATTAAGGAAGACCCCTGGAACTGCTACATGTGCGGGCACAAGGGTACCTACGGGCTGCTGCGGCGGCGAGAGGACTGGCCCTCCCGGCTCCAGATGTTCTTCGCTAATAACCACGACCAGGAATTTGACCCTCCAAAGGTTTACCCACCTGTCCCAGCTGAGAAGAGGAAGCCCATCCGGGTGCTGTCTCTCTTTGATGGAATCGCTACAGGGCTCCTGGTGCTGAAGGACTTGGGCATTCAGGTGGACCGCTACATTGCCTCGGAGGTGTGTGAGGACTCCATCACGGTGGGCATGGTGCGGCACCAGGGGAAGATCATGTACGTCGGGGACGTCCGCAGCGTCACACAGAAGCATATCCAGGAGTGGGGCCCATTCGATCTGGTGATTGGGGGCAGTCCCTGCAATGACCTCTCCATCGTCAACCCTGCTCGCAAGGGCCTCTACGAGGGCACTGGCCGGCTCTTCTTTGAGTTCTACCGCCTCCTGCATGATGCGCGGCCCAAGGAGGGAGATGATCGCCCCTTCTTCTGGCTCTTTGAGAATGTGGTGGCCATGGGCGTTAGTGACAAGAGGGACATCTCGCGATTTCTCGAGTCCAACCCTGTGATGATTGATGCCAAAGAAGTGTCAGCTGCACACAGGGCCCGCTACTTCTGGGGTAACCTTCCCGGTATGAACAGGCCGTTGGCATCCACTGTGAATGATAAGCTGGAGCTGCAGGAGTGTCTGGAGCATGGCAGGATAGCCAAGTTCAGCAAAGTGAGGACCATTACTACGAGGTCAAACTCCATAAAGCAGGGCAAAGACCAGCATTTTCCTGTCTTCATGAATGAGAAAGAGGACATCTTATGGTGCACTGAAATGGAAAGGGTATTTGGTTTCCCAGTCCACTATACTGACGTCTCCAACATGAGCCGCTTGGCGAGGCAGAGACTGCTGGGCCGGTCATGGAGCGTGCCAGTCATCCGCCACCTCTTCGCTCCGCTGAAGGAGTATTTTGCGTGTGTGTAAG'\n",
    "\n",
    "    #List of all amino acid positions (across whole coding region) covered (amino acid position minus 1 (as python starts at 0))\n",
    "    DNMT3A_nonsense_codons_DICT_NP_072046={}\n",
    "    for i in range(lower_range, upper_range):\n",
    "        DNMT3A_nonsense_codons_DICT_NP_072046[i+1]= (DNMT3A_codons_NP_072046[i])\n",
    "\n",
    "    calls_DNMT3A_nonsense_NP_072046 = {}\n",
    "    for k,v in DNMT3A_nonsense_codons_DICT_NP_072046.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_DNMT3A_nonsense_NP_072046[k]=(v, changes)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_DNMT3A_nonsense_NP_072046_extra = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_nonsense_NP_072046.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_DNMT3A_nonsense_NP_072046_extra[k] = (Amino_acid_change, v[0], v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_DNMT3A_nonsense_NP_072046_extra2 = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_nonsense_NP_072046_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_DNMT3A_nonsense_NP_072046_extra2[k] = (Amino_acid_abbrev, v[1], v[2])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary that contains details of the positions of nonsense variants at that codon\n",
    "    DNMT3A_position_in_list_dict_nonsense_NP_072046 = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_nonsense_NP_072046_extra2.items():\n",
    "        original_AA = v[0]\n",
    "        possible_AA_changes = v[2]\n",
    "        possible_AA_changes_string = ''.join(v[2])\n",
    "        positions_in_list = []\n",
    "        for i in possible_AA_changes:\n",
    "            if i == '*': #nonsense\n",
    "                position_in_list=-1\n",
    "                nonsense_change = i    \n",
    "                while True:\n",
    "                    position_in_list = possible_AA_changes_string.find(nonsense_change, position_in_list+1)\n",
    "    #               print(position_in_list)\n",
    "                    if position_in_list == -1: break\n",
    "                    positions_in_list.append(position_in_list)\n",
    "                    positions_in_list = set(positions_in_list) #convert into a 'set' to remove duplicate positions\n",
    "                    positions_in_list = list(positions_in_list) #convert back to a list so can be added to dictionary\n",
    "                DNMT3A_position_in_list_dict_nonsense_NP_072046[k]=(positions_in_list, v)\n",
    "\n",
    "        aa_positions_variants_called = []\n",
    "        for k, v in DNMT3A_position_in_list_dict_nonsense_NP_072046.items():\n",
    "            aa_positions_variants_called.append(k)\n",
    "    # print('positions at which nonsense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    DNMT3A_bases_nonsense_NP_072046 = {}\n",
    "    for k, v in DNMT3A_position_in_list_dict_nonsense_NP_072046.items():\n",
    "        original_a_a = v[1][0]\n",
    "        codon_seq = v[1][1]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        DNMT3A_bases_nonsense_NP_072046[k]=(v, base_changes, original_a_a)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    DNMT3A_bases_nonsense2_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_nonsense_NP_072046.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        DNMT3A_bases_nonsense2_NP_072046[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    DNMT3A_bases_nonsense3_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_nonsense2_NP_072046.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_a_a = v[0][2]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = k\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(k)-1)*3)+codon_pos #codon number minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][1][1][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        DNMT3A_bases_nonsense3_NP_072046[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, original_a_a)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    DNMT3A_bases_nonsense4_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_nonsense3_NP_072046.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_nonsense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_DNMT3A_NP072046:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check_DNMT3A_NP072046(i)\n",
    "                sites_trio_nonsense.append([adapted_trio])\n",
    "                DNMT3A_bases_nonsense4_NP_072046[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_nonsense = [DNMT3A_exons_extra_bases_NP_072046[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_nonsense.append(site_trio_nonsense)\n",
    "                    DNMT3A_bases_nonsense4_NP_072046[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    DNMT3A_sites_nonsense5_NP_072046 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in DNMT3A_bases_nonsense4_NP_072046.items():\n",
    "        original_a_a = v[0][0][0]\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, original_a_a])\n",
    "            change = change+1\n",
    "        DNMT3A_sites_nonsense5_NP_072046[a]=info\n",
    "        a=a+1\n",
    "    #     print(DNMT3A_sites_nonsense5_NP_072046)\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    DNMT3A_sites_nonsense6_NP_072046={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in DNMT3A_sites_nonsense5_NP_072046.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        original_a_a = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        DNMT3A_sites_nonsense6_NP_072046[k] = (v[0],changes3, codon_n_912, original_a_a) #site_trio = the affected base and its neighbours\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    DNMT3A_mutation_rates_nonsense_NP_072046 = {}\n",
    "    for k, v in DNMT3A_sites_nonsense6_NP_072046.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            original_a_a = v[3]\n",
    "            aa_positioning = v[2]\n",
    "            variant_type = '*'\n",
    "            variant_id = str(original_a_a)+str(aa_positioning)+variant_type\n",
    "            DNMT3A_mutation_rates_nonsense_NP_072046[k] = (v[0],v[1], mutation_rate, variant_id)\n",
    "\n",
    "    DNMT3A_mutation_rate_nonsense_NP_072046_sum = []\n",
    "\n",
    "    mutation_rates_list = []\n",
    "    for k, v in DNMT3A_mutation_rates_nonsense_NP_072046.items():\n",
    "        codon_position_in_gene=(v[0])\n",
    "        variant = v[3]\n",
    "        site = v[1]\n",
    "        mutation_rate_value = v[2] \n",
    "        mutation_rates_list.append((v[0], variant, site, mutation_rate_value))\n",
    "        DNMT3A_mutation_rate_nonsense_NP_072046_sum.append(mutation_rate_value)\n",
    "\n",
    "    #Get the average of all the mutation rates\n",
    "    DNMT3A_mutation_rate_nonsense_NP_072046_sum = []\n",
    "    for k, v in DNMT3A_mutation_rates_nonsense_NP_072046.items():\n",
    "        mut_rate = v[2]\n",
    "        DNMT3A_mutation_rate_nonsense_NP_072046_sum.append(mut_rate)\n",
    "\n",
    "    DNMT3A_Summed_mutation_rate_nonsense_NP_072046_added = np.sum(DNMT3A_mutation_rate_nonsense_NP_072046_sum)\n",
    "\n",
    "    print('Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study = ', DNMT3A_Summed_mutation_rate_nonsense_NP_072046_added)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all # e.g. lower= 0, upper = 913 for whole gene\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_missense_range_of_variants_DNMT3A_NP072046(lower_range, upper_range):\n",
    "    intron_exon_boundary_caution_base_DNMT3A_NP072046 = [1279, 1429, 1474, 1554, 1852, 2083, 2082, 2173, 2322, 2478]\n",
    "\n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    def intron_exon_check_DNMT3A_NP072046(position_in_gene_number):\n",
    "        if position_in_gene_number == 1279:\n",
    "            gene_pos = 1279\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1429:\n",
    "            gene_pos = 1429\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1474:\n",
    "            gene_pos = 1474\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1554:\n",
    "            gene_pos = 1554\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1852:\n",
    "            gene_pos = 1852\n",
    "            first_base = 'G'\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2083:\n",
    "            gene_pos = 2083\n",
    "            first_base = 'G'\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2082:\n",
    "            gene_pos = 2082\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2173:\n",
    "            gene_pos = 2173\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2322:\n",
    "            gene_pos = 2322\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 2478:\n",
    "            gene_pos = 2478\n",
    "            first_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos-1]\n",
    "            middle_base = DNMT3A_exons_extra_bases_NP_072046[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "\n",
    "    #DNMT3A transcript used (NP_072046):\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000119772;r=2:25455845-25564754;t=ENST00000321117\n",
    "    DNMT3A_exons_NP_072046 = 'ATGCCCGCCATGCCCTCCAGCGGCCCCGGGGACACCAGCAGCTCTGCTGCGGAGCGGGAGGAGGACCGAAAGGACGGAGAGGAGCAGGAGGAGCCGCGTGGCAAGGAGGAGCGCCAAGAGCCCAGCACCACGGCACGGAAGGTGGGGCGGCCTGGGAGGAAGCGCAAGCACCCCCCGGTGGAAAGCGGTGACACGCCAAAGGACCCTGCGGTGATCTCCAAGTCCCCATCCATGGCCCAGGACTCAGGCGCCTCAGAGCTATTACCCAATGGGGACTTGGAGAAGCGGAGTGAGCCCCAGCCAGAGGAGGGGAGCCCTGCTGGGGGGCAGAAGGGCGGGGCCCCAGCAGAGGGAGAGGGTGCAGCTGAGACCCTGCCTGAAGCCTCAAGAGCAGTGGAAAATGGCTGCTGCACCCCCAAGGAGGGCCGAGGAGCCCCTGCAGAAGCGGGCAAAGAACAGAAGGAGACCAACATCGAATCCATGAAAATGGAGGGCTCCCGGGGCCGGCTGCGGGGTGGCTTGGGCTGGGAGTCCAGCCTCCGTCAGCGGCCCATGCCGAGGCTCACCTTCCAGGCGGGGGACCCCTACTACATCAGCAAGCGCAAGCGGGACGAGTGGCTGGCACGCTGGAAAAGGGAGGCTGAGAAGAAAGCCAAGGTCATTGCAGGAATGAATGCTGTGGAAGAAAACCAGGGGCCCGGGGAGTCTCAGAAGGTGGAGGAGGCCAGCCCTCCTGCTGTGCAGCAGCCCACTGACCCCGCATCCCCCACTGTGGCTACCACGCCTGAGCCCGTGGGGTCCGATGCTGGGGACAAGAATGCCACCAAAGCAGGCGATGACGAGCCAGAGTACGAGGACGGCCGGGGCTTTGGCATTGGGGAGCTGGTGTGGGGGAAACTGCGGGGCTTCTCCTGGTGGCCAGGCCGCATTGTGTCTTGGTGGATGACGGGCCGGAGCCGAGCAGCTGAAGGCACCCGCTGGGTCATGTGGTTCGGAGACGGCAAATTCTCAGTGGTGTGTGTTGAGAAGCTGATGCCGCTGAGCTCGTTTTGCAGTGCGTTCCACCAGGCCACGTACAACAAGCAGCCCATGTACCGCAAAGCCATCTACGAGGTCCTGCAGGTGGCCAGCAGCCGCGCGGGGAAGCTGTTCCCGGTGTGCCACGACAGCGATGAGAGTGACACTGCCAAGGCCGTGGAGGTGCAGAACAAGCCCATGATTGAATGGGCCCTGGGGGGCTTCCAGCCTTCTGGCCCTAAGGGCCTGGAGCCACCAGAAGAAGAGAAGAATCCCTACAAAGAAGTGTACACGGACATGTGGGTGGAACCTGAGGCAGCTGCCTACGCACCACCTCCACCAGCCAAAAAGCCCCGGAAGAGCACAGCGGAGAAGCCCAAGGTCAAGGAGATTATTGATGAGCGCACAAGAGAGCGGCTGGTGTACGAGGTGCGGCAGAAGTGCCGGAACATTGAGGACATCTGCATCTCCTGTGGGAGCCTCAATGTTACCCTGGAACACCCCCTCTTCGTTGGAGGAATGTGCCAAAACTGCAAGAACTGCTTTCTGGAGTGTGCGTACCAGTACGACGACGACGGCTACCAGTCCTACTGCACCATCTGCTGTGGGGGCCGTGAGGTGCTCATGTGCGGAAACAACAACTGCTGCAGGTGCTTTTGCGTGGAGTGTGTGGACCTCTTGGTGGGGCCGGGGGCTGCCCAGGCAGCCATTAAGGAAGACCCCTGGAACTGCTACATGTGCGGGCACAAGGGTACCTACGGGCTGCTGCGGCGGCGAGAGGACTGGCCCTCCCGGCTCCAGATGTTCTTCGCTAATAACCACGACCAGGAATTTGACCCTCCAAAGGTTTACCCACCTGTCCCAGCTGAGAAGAGGAAGCCCATCCGGGTGCTGTCTCTCTTTGATGGAATCGCTACAGGGCTCCTGGTGCTGAAGGACTTGGGCATTCAGGTGGACCGCTACATTGCCTCGGAGGTGTGTGAGGACTCCATCACGGTGGGCATGGTGCGGCACCAGGGGAAGATCATGTACGTCGGGGACGTCCGCAGCGTCACACAGAAGCATATCCAGGAGTGGGGCCCATTCGATCTGGTGATTGGGGGCAGTCCCTGCAATGACCTCTCCATCGTCAACCCTGCTCGCAAGGGCCTCTACGAGGGCACTGGCCGGCTCTTCTTTGAGTTCTACCGCCTCCTGCATGATGCGCGGCCCAAGGAGGGAGATGATCGCCCCTTCTTCTGGCTCTTTGAGAATGTGGTGGCCATGGGCGTTAGTGACAAGAGGGACATCTCGCGATTTCTCGAGTCCAACCCTGTGATGATTGATGCCAAAGAAGTGTCAGCTGCACACAGGGCCCGCTACTTCTGGGGTAACCTTCCCGGTATGAACAGGCCGTTGGCATCCACTGTGAATGATAAGCTGGAGCTGCAGGAGTGTCTGGAGCATGGCAGGATAGCCAAGTTCAGCAAAGTGAGGACCATTACTACGAGGTCAAACTCCATAAAGCAGGGCAAAGACCAGCATTTTCCTGTCTTCATGAATGAGAAAGAGGACATCTTATGGTGCACTGAAATGGAAAGGGTATTTGGTTTCCCAGTCCACTATACTGACGTCTCCAACATGAGCCGCTTGGCGAGGCAGAGACTGCTGGGCCGGTCATGGAGCGTGCCAGTCATCCGCCACCTCTTCGCTCCGCTGAAGGAGTATTTTGCGTGTGTGTAA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    DNMT3A_codons_NP_072046 = [DNMT3A_exons_NP_072046[i:i+3] for i in range(0, len(DNMT3A_exons_NP_072046), 3)]\n",
    "\n",
    "    #DNMT3A transcript used (NP_072046):\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000119772;r=2:25455845-25564754;t=ENST00000321117\n",
    "    DNMT3A_exons_extra_bases_NP_072046 = 'GATGCCCGCCATGCCCTCCAGCGGCCCCGGGGACACCAGCAGCTCTGCTGCGGAGCGGGAGGAGGACCGAAAGGACGGAGAGGAGCAGGAGGAGCCGCGTGGCAAGGAGGAGCGCCAAGAGCCCAGCACCACGGCACGGAAGGTGGGGCGGCCTGGGAGGAAGCGCAAGCACCCCCCGGTGGAAAGCGGTGACACGCCAAAGGACCCTGCGGTGATCTCCAAGTCCCCATCCATGGCCCAGGACTCAGGCGCCTCAGAGCTATTACCCAATGGGGACTTGGAGAAGCGGAGTGAGCCCCAGCCAGAGGAGGGGAGCCCTGCTGGGGGGCAGAAGGGCGGGGCCCCAGCAGAGGGAGAGGGTGCAGCTGAGACCCTGCCTGAAGCCTCAAGAGCAGTGGAAAATGGCTGCTGCACCCCCAAGGAGGGCCGAGGAGCCCCTGCAGAAGCGGGCAAAGAACAGAAGGAGACCAACATCGAATCCATGAAAATGGAGGGCTCCCGGGGCCGGCTGCGGGGTGGCTTGGGCTGGGAGTCCAGCCTCCGTCAGCGGCCCATGCCGAGGCTCACCTTCCAGGCGGGGGACCCCTACTACATCAGCAAGCGCAAGCGGGACGAGTGGCTGGCACGCTGGAAAAGGGAGGCTGAGAAGAAAGCCAAGGTCATTGCAGGAATGAATGCTGTGGAAGAAAACCAGGGGCCCGGGGAGTCTCAGAAGGTGGAGGAGGCCAGCCCTCCTGCTGTGCAGCAGCCCACTGACCCCGCATCCCCCACTGTGGCTACCACGCCTGAGCCCGTGGGGTCCGATGCTGGGGACAAGAATGCCACCAAAGCAGGCGATGACGAGCCAGAGTACGAGGACGGCCGGGGCTTTGGCATTGGGGAGCTGGTGTGGGGGAAACTGCGGGGCTTCTCCTGGTGGCCAGGCCGCATTGTGTCTTGGTGGATGACGGGCCGGAGCCGAGCAGCTGAAGGCACCCGCTGGGTCATGTGGTTCGGAGACGGCAAATTCTCAGTGGTGTGTGTTGAGAAGCTGATGCCGCTGAGCTCGTTTTGCAGTGCGTTCCACCAGGCCACGTACAACAAGCAGCCCATGTACCGCAAAGCCATCTACGAGGTCCTGCAGGTGGCCAGCAGCCGCGCGGGGAAGCTGTTCCCGGTGTGCCACGACAGCGATGAGAGTGACACTGCCAAGGCCGTGGAGGTGCAGAACAAGCCCATGATTGAATGGGCCCTGGGGGGCTTCCAGCCTTCTGGCCCTAAGGGCCTGGAGCCACCAGAAGAAGAGAAGAATCCCTACAAAGAAGTGTACACGGACATGTGGGTGGAACCTGAGGCAGCTGCCTACGCACCACCTCCACCAGCCAAAAAGCCCCGGAAGAGCACAGCGGAGAAGCCCAAGGTCAAGGAGATTATTGATGAGCGCACAAGAGAGCGGCTGGTGTACGAGGTGCGGCAGAAGTGCCGGAACATTGAGGACATCTGCATCTCCTGTGGGAGCCTCAATGTTACCCTGGAACACCCCCTCTTCGTTGGAGGAATGTGCCAAAACTGCAAGAACTGCTTTCTGGAGTGTGCGTACCAGTACGACGACGACGGCTACCAGTCCTACTGCACCATCTGCTGTGGGGGCCGTGAGGTGCTCATGTGCGGAAACAACAACTGCTGCAGGTGCTTTTGCGTGGAGTGTGTGGACCTCTTGGTGGGGCCGGGGGCTGCCCAGGCAGCCATTAAGGAAGACCCCTGGAACTGCTACATGTGCGGGCACAAGGGTACCTACGGGCTGCTGCGGCGGCGAGAGGACTGGCCCTCCCGGCTCCAGATGTTCTTCGCTAATAACCACGACCAGGAATTTGACCCTCCAAAGGTTTACCCACCTGTCCCAGCTGAGAAGAGGAAGCCCATCCGGGTGCTGTCTCTCTTTGATGGAATCGCTACAGGGCTCCTGGTGCTGAAGGACTTGGGCATTCAGGTGGACCGCTACATTGCCTCGGAGGTGTGTGAGGACTCCATCACGGTGGGCATGGTGCGGCACCAGGGGAAGATCATGTACGTCGGGGACGTCCGCAGCGTCACACAGAAGCATATCCAGGAGTGGGGCCCATTCGATCTGGTGATTGGGGGCAGTCCCTGCAATGACCTCTCCATCGTCAACCCTGCTCGCAAGGGCCTCTACGAGGGCACTGGCCGGCTCTTCTTTGAGTTCTACCGCCTCCTGCATGATGCGCGGCCCAAGGAGGGAGATGATCGCCCCTTCTTCTGGCTCTTTGAGAATGTGGTGGCCATGGGCGTTAGTGACAAGAGGGACATCTCGCGATTTCTCGAGTCCAACCCTGTGATGATTGATGCCAAAGAAGTGTCAGCTGCACACAGGGCCCGCTACTTCTGGGGTAACCTTCCCGGTATGAACAGGCCGTTGGCATCCACTGTGAATGATAAGCTGGAGCTGCAGGAGTGTCTGGAGCATGGCAGGATAGCCAAGTTCAGCAAAGTGAGGACCATTACTACGAGGTCAAACTCCATAAAGCAGGGCAAAGACCAGCATTTTCCTGTCTTCATGAATGAGAAAGAGGACATCTTATGGTGCACTGAAATGGAAAGGGTATTTGGTTTCCCAGTCCACTATACTGACGTCTCCAACATGAGCCGCTTGGCGAGGCAGAGACTGCTGGGCCGGTCATGGAGCGTGCCAGTCATCCGCCACCTCTTCGCTCCGCTGAAGGAGTATTTTGCGTGTGTGTAAG'\n",
    "\n",
    "    #List of all amino acid positions (across whole coding region) covered (amino acid position minus 1 (as python starts at 0))\n",
    "    DNMT3A_missense_codons_DICT_NP_072046={}\n",
    "    for i in range(lower_range, upper_range):\n",
    "        DNMT3A_missense_codons_DICT_NP_072046[i+1]= (DNMT3A_codons_NP_072046[i])\n",
    "\n",
    "    calls_DNMT3A_missense_NP_072046 = {}\n",
    "    for k,v in DNMT3A_missense_codons_DICT_NP_072046.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_DNMT3A_missense_NP_072046[k]=(v, changes)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_DNMT3A_missense_NP_072046_extra = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_missense_NP_072046.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_DNMT3A_missense_NP_072046_extra[k] = (Amino_acid_change, v[0], v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_DNMT3A_missense_NP_072046_extra2 = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_missense_NP_072046_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_DNMT3A_missense_NP_072046_extra2[k] = (Amino_acid_abbrev, v[1], v[2])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary that contains details of the positions of missense variants at that codon\n",
    "    DNMT3A_position_in_list_dict_missense_NP_072046 = {}\n",
    "\n",
    "    for k, v in calls_DNMT3A_missense_NP_072046_extra2.items():\n",
    "        original_AA = v[0]\n",
    "        possible_AA_changes = v[2]\n",
    "        possible_AA_changes_string = ''.join(v[2])\n",
    "        positions_in_list = []\n",
    "        for i in possible_AA_changes:\n",
    "            if i != original_AA: #not synonymous\n",
    "                if i != '*': #not missense\n",
    "                    position_in_list=-1\n",
    "                    missense_change = i\n",
    "                    new_aa_list = []\n",
    "                    while True:\n",
    "                        position_in_list = possible_AA_changes_string.find(missense_change, position_in_list+1)\n",
    "        #               print(position_in_list)\n",
    "                        if position_in_list == -1: break\n",
    "                        positions_in_list.append(position_in_list)\n",
    "                        positions_in_list = set(positions_in_list) #convert into a 'set' to remove duplicate positions\n",
    "                        positions_in_list = list(positions_in_list) #convert back to a list so can be added to dictionary\n",
    "                    for pos in positions_in_list:\n",
    "                        new_aa = possible_AA_changes_string[pos]\n",
    "                        new_aa_list.append(new_aa)\n",
    "                    DNMT3A_position_in_list_dict_missense_NP_072046[k]=(positions_in_list, v, new_aa_list)\n",
    "\n",
    "        aa_positions_variants_called = []\n",
    "        for k, v in DNMT3A_position_in_list_dict_missense_NP_072046.items():\n",
    "            aa_positions_variants_called.append(k)\n",
    "    # print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    DNMT3A_bases_missense_NP_072046 = {}\n",
    "    for k, v in DNMT3A_position_in_list_dict_missense_NP_072046.items():\n",
    "        original_a_a = v[1][0]\n",
    "        codon_seq = v[1][1]\n",
    "        base_changes=[]\n",
    "        new_a_a_list = v[2]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        DNMT3A_bases_missense_NP_072046[k]=(v, base_changes, original_a_a, new_a_a_list)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    DNMT3A_bases_missense2_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_missense_NP_072046.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        new_a_a_list = v[3]\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        DNMT3A_bases_missense2_NP_072046[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    DNMT3A_bases_missense3_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_missense2_NP_072046.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_a_a = v[0][2]\n",
    "        new_aa_list = v[0][0][2]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = k\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(k)-1)*3)+codon_pos #codon number minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][1][1][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        DNMT3A_bases_missense3_NP_072046[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, original_a_a, new_aa_list)\n",
    "\n",
    "    # Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    DNMT3A_bases_missense4_NP_072046 = {}\n",
    "\n",
    "    for k, v in DNMT3A_bases_missense3_NP_072046.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "        new_a_a_list = v[6]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_DNMT3A_NP072046:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check_DNMT3A_NP072046(i)\n",
    "                sites_trio_missense.append([adapted_trio])\n",
    "                DNMT3A_bases_missense4_NP_072046[k] = (v, sites_trio_missense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_missense = [DNMT3A_exons_extra_bases_NP_072046[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_missense.append(site_trio_missense)\n",
    "                    DNMT3A_bases_missense4_NP_072046[k] = (v, sites_trio_missense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    DNMT3A_sites_missense5_NP_072046 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in DNMT3A_bases_missense4_NP_072046.items():\n",
    "        original_a_a = v[0][0][0]\n",
    "        new_a_a = v[0][6]\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            new_amino_acid = new_a_a[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, original_a_a, new_amino_acid])\n",
    "            change = change+1\n",
    "        DNMT3A_sites_missense5_NP_072046[a]=info\n",
    "        a=a+1\n",
    "    #     print(DNMT3A_sites_missense5_NP_072046)\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    DNMT3A_sites_missense6_NP_072046={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in DNMT3A_sites_missense5_NP_072046.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        original_a_a = v[5]\n",
    "        new_a_a = v[6]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        DNMT3A_sites_missense6_NP_072046[k] = (v[0],changes3, codon_n_912, original_a_a, new_a_a) #site_trio = the affected base and its neighbours\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    DNMT3A_mutation_rates_missense_NP_072046 = {}\n",
    "    for k, v in DNMT3A_sites_missense6_NP_072046.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            original_a_a = v[3]\n",
    "            new_a_a = v[4]\n",
    "            aa_positioning = v[2]\n",
    "            variant_id = str(original_a_a)+str(aa_positioning)+str(new_a_a)\n",
    "            DNMT3A_mutation_rates_missense_NP_072046[k] = (v[0],v[1], mutation_rate, variant_id)\n",
    "\n",
    "    DNMT3A_mutation_rate_missense_NP_072046_sum = []\n",
    "\n",
    "    mutation_rates_list = []\n",
    "    for k, v in DNMT3A_mutation_rates_missense_NP_072046.items():\n",
    "        codon_position_in_gene=(v[0])\n",
    "        variant = v[3]\n",
    "        site = v[1]\n",
    "        mutation_rate_value = v[2] \n",
    "        mutation_rates_list.append((v[0], variant, site, mutation_rate_value))\n",
    "        DNMT3A_mutation_rate_missense_NP_072046_sum.append(mutation_rate_value)\n",
    "\n",
    "    #Get the average of all the mutation rates\n",
    "    DNMT3A_mutation_rate_missense_NP_072046_sum = []\n",
    "    for k, v in DNMT3A_mutation_rates_missense_NP_072046.items():\n",
    "        mut_rate = v[2]\n",
    "        DNMT3A_mutation_rate_missense_NP_072046_sum.append(mut_rate)\n",
    "\n",
    "    DNMT3A_Summed_mutation_rate_missense_NP_072046_added = np.sum(DNMT3A_mutation_rate_missense_NP_072046_sum)\n",
    "\n",
    "    print('Total summed mutation rate across DNMT3A missense_NP_072046 for this study = ', DNMT3A_Summed_mutation_rate_missense_NP_072046_added)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all # e.g. lower= 0, upper = 913 for whole gene\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TET2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_nonsense_range_of_variants_TET2(lower_range, upper_range):\n",
    "    \n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base_TET2_NP01120680 = [3409, 3804, 4044, 4537]\n",
    "   \n",
    "    def intron_exon_check_TET2_NP01120680(position_in_gene_number):\n",
    "        if position_in_gene_number == 3409:\n",
    "            gene_pos = 3409\n",
    "            first_base = TET2_exons_extra_bases_NP_01120680[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 3804:\n",
    "            gene_pos = 3804\n",
    "            first_base = 'G'\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = TET2_exons_extra_bases_NP_01120680[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 4044:\n",
    "            gene_pos = 4044\n",
    "            first_base = TET2_exons_extra_bases_NP_01120680[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 4537:\n",
    "            gene_pos = 4537\n",
    "            first_base = TET2_exons_extra_bases_NP_01120680[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #TET2 transcript (NP_01120680):\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000168769;r=4:106067032-106163928;t=ENST00000380013\n",
    "    TET2_exons_NP_01120680 = 'ATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGAGCAAATTATTGAAAAAGATGAAGGTCCTTTTTATACCCATCTAGGAGCAGGTCCTAATGTGGCAGCTATTAGAGAAATCATGGAAGAAAGGTTTGGACAGAAGGGTAAAGCTATTAGGATTGAAAGAGTCATCTATACTGGTAAAGAAGGCAAAAGTTCTCAGGGATGTCCTATTGCTAAGTGGGTGGTTCGCAGAAGCAGCAGTGAAGAGAAGCTACTGTGTTTGGTGCGGGAGCGAGCTGGCCACACCTGTGAGGCTGCAGTGATTGTGATTCTCATCCTGGTGTGGGAAGGAATCCCGCTGTCTCTGGCTGACAAACTCTACTCGGAGCTTACCGAGACGCTGAGGAAATACGGCACGCTCACCAATCGCCGGTGTGCCTTGAATGAAGAGAGAACTTGCGCCTGTCAGGGGCTGGATCCAGAAACCTGTGGTGCCTCCTTCTCTTTTGGTTGTTCATGGAGCATGTACTACAATGGATGTAAGTTTGCCAGAAGCAAGATCCCAAGGAAGTTTAAGCTGCTTGGGGATGACCCAAAAGAGGAAGAGAAACTGGAGTCTCATTTGCAAAACCTGTCCACTCTTATGGCACCAACATATAAGAAACTTGCACCTGATGCATATAATAATCAGATTGAATATGAACACAGAGCACCAGAGTGCCGTCTGGGTCTGAAGGAAGGCCGTCCATTCTCAGGGGTCACTGCATGTTTGGACTTCTGTGCTCATGCCCACAGAGACTTGCACAACATGCAGAATGGCAGCACATTGGTATGCACTCTCACTAGAGAAGACAATCGAGAATTTGGAGGAAAACCTGAGGATGAGCAGCTTCACGTTCTGCCTTTATACAAAGTCTCTGACGTGGATGAGTTTGGGAGTGTGGAAGCTCAGGAGGAGAAAAAACGGAGTGGTGCCATTCAGGTACTGAGTTCTTTTCGGCGAAAAGTCAGGATGTTAGCAGAGCCAGTCAAGACTTGCCGACAAAGGAAACTAGAAGCCAAGAAAGCTGCAGCTGAAAAGCTTTCCTCCCTGGAGAACAGCTCAAATAAAAATGAAAAGGAAAAGTCAGCCCCATCACGTACAAAACAAACTGAAAACGCAAGCCAGGCTAAACAGTTGGCAGAACTTTTGCGACTTTCAGGACCAGTCATGCAGCAGTCCCAGCAGCCCCAGCCTCTACAGAAGCAGCCACCACAGCCCCAGCAGCAGCAGAGACCCCAGCAGCAGCAGCCACATCACCCTCAGACAGAGTCTGTCAACTCTTATTCTGCTTCTGGATCCACCAATCCATACATGAGACGGCCCAATCCAGTTAGTCCTTATCCAAACTCTTCACACACTTCAGATATCTATGGAAGCACCAGCCCTATGAACTTCTATTCCACCTCATCTCAAGCTGCAGGTTCATATTTGAATTCTTCTAATCCCATGAACCCTTACCCTGGGCTTTTGAATCAGAATACCCAATATCCATCATATCAATGCAATGGAAACCTATCAGTGGACAACTGCTCCCCATATCTGGGTTCCTATTCTCCCCAGTCTCAGCCGATGGATCTGTATAGGTATCCAAGCCAAGACCCTCTGTCTAAGCTCAGTCTACCACCCATCCATACACTTTACCAGCCAAGGTTTGGAAATAGCCAGAGTTTTACATCTAAATACTTAGGTTATGGAAACCAAAATATGCAGGGAGATGGTTTCAGCAGTTGTACCATTAGACCAAATGTACATCATGTAGGGAAATTGCCTCCTTATCCCACTCATGAGATGGATGGCCACTTCATGGGAGCCACCTCTAGATTACCACCCAATCTGAGCAATCCAAACATGGACTATAAAAATGGTGAACATCATTCACCTTCTCACATAATCCATAACTACAGTGCAGCTCCGGGCATGTTCAACAGCTCTCTTCATGCCCTGCATCTCCAAAACAAGGAGAATGACATGCTTTCCCACACAGCTAATGGGTTATCAAAGATGCTTCCAGCTCTTAACCATGATAGAACTGCTTGTGTCCAAGGAGGCTTACACAAATTAAGTGATGCTAATGGTCAGGAAAAGCAGCCATTGGCACTAGTCCAGGGTGTGGCTTCTGGTGCAGAGGACAACGATGAGGTCTGGTCAGACAGCGAGCAGAGCTTTCTGGATCCTGACATTGGGGGAGTGGCCGTGGCTCCAACTCATGGGTCAATTCTCATTGAGTGTGCAAAGCGTGAGCTGCATGCCACAACCCCTTTAAAGAATCCCAATAGGAATCACCCCACCAGGATCTCCCTCGTCTTTTACCAGCATAAGAGCATGAATGAGCCAAAACATGGCTTGGCTCTTTGGGAAGCCAAAATGGCTGAAAAAGCCCGTGAGAAAGAGGAAGAGTGTGAAAAGTATGGCCCAGACTATGTGCCTCAGAAATCCCATGGCAAAAAAGTGAAACGGGAGCCTGCTGAGCCACATGAAACTTCAGAGCCCACTTACCTGCGTTTCATCAAGTCTCTTGCCGAAAGGACCATGTCCGTGACCACAGACTCCACAGTAACTACATCTCCATATGCCTTCACTCGGGTCACAGGGCCTTACAACAGATATATATGA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    TET2_codons_NP_01120680 = [TET2_exons_NP_01120680[i:i+3] for i in range(0, len(TET2_exons_NP_01120680), 3)]\n",
    "\n",
    "    #TET2 transcript (transcript variant 2 (TET2-001), NM_017628, NP_001120680):\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000168769;r=4:106067032-106163928;t=ENST00000380013\n",
    "    TET2_exons_extra_bases_NP_01120680 = 'GATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGAGCAAATTATTGAAAAAGATGAAGGTCCTTTTTATACCCATCTAGGAGCAGGTCCTAATGTGGCAGCTATTAGAGAAATCATGGAAGAAAGGTTTGGACAGAAGGGTAAAGCTATTAGGATTGAAAGAGTCATCTATACTGGTAAAGAAGGCAAAAGTTCTCAGGGATGTCCTATTGCTAAGTGGGTGGTTCGCAGAAGCAGCAGTGAAGAGAAGCTACTGTGTTTGGTGCGGGAGCGAGCTGGCCACACCTGTGAGGCTGCAGTGATTGTGATTCTCATCCTGGTGTGGGAAGGAATCCCGCTGTCTCTGGCTGACAAACTCTACTCGGAGCTTACCGAGACGCTGAGGAAATACGGCACGCTCACCAATCGCCGGTGTGCCTTGAATGAAGAGAGAACTTGCGCCTGTCAGGGGCTGGATCCAGAAACCTGTGGTGCCTCCTTCTCTTTTGGTTGTTCATGGAGCATGTACTACAATGGATGTAAGTTTGCCAGAAGCAAGATCCCAAGGAAGTTTAAGCTGCTTGGGGATGACCCAAAAGAGGAAGAGAAACTGGAGTCTCATTTGCAAAACCTGTCCACTCTTATGGCACCAACATATAAGAAACTTGCACCTGATGCATATAATAATCAGATTGAATATGAACACAGAGCACCAGAGTGCCGTCTGGGTCTGAAGGAAGGCCGTCCATTCTCAGGGGTCACTGCATGTTTGGACTTCTGTGCTCATGCCCACAGAGACTTGCACAACATGCAGAATGGCAGCACATTGGTATGCACTCTCACTAGAGAAGACAATCGAGAATTTGGAGGAAAACCTGAGGATGAGCAGCTTCACGTTCTGCCTTTATACAAAGTCTCTGACGTGGATGAGTTTGGGAGTGTGGAAGCTCAGGAGGAGAAAAAACGGAGTGGTGCCATTCAGGTACTGAGTTCTTTTCGGCGAAAAGTCAGGATGTTAGCAGAGCCAGTCAAGACTTGCCGACAAAGGAAACTAGAAGCCAAGAAAGCTGCAGCTGAAAAGCTTTCCTCCCTGGAGAACAGCTCAAATAAAAATGAAAAGGAAAAGTCAGCCCCATCACGTACAAAACAAACTGAAAACGCAAGCCAGGCTAAACAGTTGGCAGAACTTTTGCGACTTTCAGGACCAGTCATGCAGCAGTCCCAGCAGCCCCAGCCTCTACAGAAGCAGCCACCACAGCCCCAGCAGCAGCAGAGACCCCAGCAGCAGCAGCCACATCACCCTCAGACAGAGTCTGTCAACTCTTATTCTGCTTCTGGATCCACCAATCCATACATGAGACGGCCCAATCCAGTTAGTCCTTATCCAAACTCTTCACACACTTCAGATATCTATGGAAGCACCAGCCCTATGAACTTCTATTCCACCTCATCTCAAGCTGCAGGTTCATATTTGAATTCTTCTAATCCCATGAACCCTTACCCTGGGCTTTTGAATCAGAATACCCAATATCCATCATATCAATGCAATGGAAACCTATCAGTGGACAACTGCTCCCCATATCTGGGTTCCTATTCTCCCCAGTCTCAGCCGATGGATCTGTATAGGTATCCAAGCCAAGACCCTCTGTCTAAGCTCAGTCTACCACCCATCCATACACTTTACCAGCCAAGGTTTGGAAATAGCCAGAGTTTTACATCTAAATACTTAGGTTATGGAAACCAAAATATGCAGGGAGATGGTTTCAGCAGTTGTACCATTAGACCAAATGTACATCATGTAGGGAAATTGCCTCCTTATCCCACTCATGAGATGGATGGCCACTTCATGGGAGCCACCTCTAGATTACCACCCAATCTGAGCAATCCAAACATGGACTATAAAAATGGTGAACATCATTCACCTTCTCACATAATCCATAACTACAGTGCAGCTCCGGGCATGTTCAACAGCTCTCTTCATGCCCTGCATCTCCAAAACAAGGAGAATGACATGCTTTCCCACACAGCTAATGGGTTATCAAAGATGCTTCCAGCTCTTAACCATGATAGAACTGCTTGTGTCCAAGGAGGCTTACACAAATTAAGTGATGCTAATGGTCAGGAAAAGCAGCCATTGGCACTAGTCCAGGGTGTGGCTTCTGGTGCAGAGGACAACGATGAGGTCTGGTCAGACAGCGAGCAGAGCTTTCTGGATCCTGACATTGGGGGAGTGGCCGTGGCTCCAACTCATGGGTCAATTCTCATTGAGTGTGCAAAGCGTGAGCTGCATGCCACAACCCCTTTAAAGAATCCCAATAGGAATCACCCCACCAGGATCTCCCTCGTCTTTTACCAGCATAAGAGCATGAATGAGCCAAAACATGGCTTGGCTCTTTGGGAAGCCAAAATGGCTGAAAAAGCCCGTGAGAAAGAGGAAGAGTGTGAAAAGTATGGCCCAGACTATGTGCCTCAGAAATCCCATGGCAAAAAAGTGAAACGGGAGCCTGCTGAGCCACATGAAACTTCAGAGCCCACTTACCTGCGTTTCATCAAGTCTCTTGCCGAAAGGACCATGTCCGTGACCACAGACTCCACAGTAACTACATCTCCATATGCCTTCACTCGGGTCACAGGGCCTTACAACAGATATATATGAT'\n",
    "\n",
    "    #List of all amino acid positions (across whole coding region) covered by Coombs (amino acid position minus 1 (as python starts at 0))\n",
    "    TET2_nonsense_codons_DICT_NP_01120680={}\n",
    "    for i in range(lower_range, upper_range):\n",
    "        TET2_nonsense_codons_DICT_NP_01120680[i+1]= (TET2_codons_NP_01120680[i])\n",
    "\n",
    "    calls_TET2_nonsense_NP_01120680 = {}\n",
    "    for k,v in TET2_nonsense_codons_DICT_NP_01120680.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_TET2_nonsense_NP_01120680[k]=(v, changes)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_TET2_nonsense_NP_01120680_extra = {}\n",
    "\n",
    "    for k, v in calls_TET2_nonsense_NP_01120680.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_TET2_nonsense_NP_01120680_extra[k] = (Amino_acid_change, v[0], v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_TET2_nonsense_NP_01120680_extra2 = {}\n",
    "\n",
    "    for k, v in calls_TET2_nonsense_NP_01120680_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_TET2_nonsense_NP_01120680_extra2[k] = (Amino_acid_abbrev, v[1], v[2])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary that contains details of the positions of nonsense variants at that codon\n",
    "    TET2_position_in_list_dict_nonsense_NP_01120680 = {}\n",
    "\n",
    "    for k, v in calls_TET2_nonsense_NP_01120680_extra2.items():\n",
    "        original_AA = v[0]\n",
    "        possible_AA_changes = v[2]\n",
    "        possible_AA_changes_string = ''.join(v[2])\n",
    "        positions_in_list = []\n",
    "        for i in possible_AA_changes:\n",
    "            if i == '*': #nonsense\n",
    "                position_in_list=-1\n",
    "                nonsense_change = i    \n",
    "                while True:\n",
    "                    position_in_list = possible_AA_changes_string.find(nonsense_change, position_in_list+1)\n",
    "    #               print(position_in_list)\n",
    "                    if position_in_list == -1: break\n",
    "                    positions_in_list.append(position_in_list)\n",
    "                    positions_in_list = set(positions_in_list) #convert into a 'set' to remove duplicate positions\n",
    "                    positions_in_list = list(positions_in_list) #convert back to a list so can be added to dictionary\n",
    "                TET2_position_in_list_dict_nonsense_NP_01120680[k]=(positions_in_list, v)\n",
    "\n",
    "        aa_positions_variants_called = []\n",
    "        for k, v in TET2_position_in_list_dict_nonsense_NP_01120680.items():\n",
    "            aa_positions_variants_called.append(k)\n",
    "    # print('positions at which nonsense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TET2_bases_nonsense_NP_01120680 = {}\n",
    "    for k, v in TET2_position_in_list_dict_nonsense_NP_01120680.items():\n",
    "        original_a_a = v[1][0]\n",
    "        codon_seq = v[1][1]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TET2_bases_nonsense_NP_01120680[k]=(v, base_changes, original_a_a)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TET2_bases_nonsense2_NP_01120680 = {}\n",
    "\n",
    "    for k, v in TET2_bases_nonsense_NP_01120680.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TET2_bases_nonsense2_NP_01120680[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TET2_bases_nonsense3_NP_01120680 = {}\n",
    "\n",
    "    for k, v in TET2_bases_nonsense2_NP_01120680.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_a_a = v[0][2]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = k\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(k)-1)*3)+codon_pos #codon number minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][1][1][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TET2_bases_nonsense3_NP_01120680[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, original_a_a)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TET2_bases_nonsense4_NP_01120680 = {}\n",
    "\n",
    "    for k, v in TET2_bases_nonsense3_NP_01120680.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_nonsense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_TET2_NP01120680:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check_TET2_NP01120680(i)\n",
    "                sites_trio_nonsense.append([adapted_trio])\n",
    "                TET2_bases_nonsense4_NP_01120680[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_nonsense = [TET2_exons_extra_bases_NP_01120680[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_nonsense.append(site_trio_nonsense)\n",
    "                    TET2_bases_nonsense4_NP_01120680[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TET2_sites_nonsense5_NP_01120680 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TET2_bases_nonsense4_NP_01120680.items():\n",
    "        original_a_a = v[0][0][0]\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, original_a_a])\n",
    "            change = change+1\n",
    "        TET2_sites_nonsense5_NP_01120680[a]=info\n",
    "        a=a+1\n",
    "    #     print(TET2_sites_nonsense5_NP_01120680)\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TET2_sites_nonsense6_NP_01120680={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TET2_sites_nonsense5_NP_01120680.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        original_a_a = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TET2_sites_nonsense6_NP_01120680[k] = (v[0],changes3, codon_n_912, original_a_a) #site_trio = the affected base and its neighbours\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    TET2_mutation_rates_nonsense_NP_01120680 = {}\n",
    "    for k, v in TET2_sites_nonsense6_NP_01120680.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            original_a_a = v[3]\n",
    "            aa_positioning = v[2]\n",
    "            variant_type = '*'\n",
    "            variant_id = str(original_a_a)+str(aa_positioning)+variant_type\n",
    "            TET2_mutation_rates_nonsense_NP_01120680[k] = (v[0],v[1], mutation_rate, variant_id)\n",
    "\n",
    "    TET2_mutation_rate_nonsense_NP_01120680_sum = []\n",
    "\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TET2_mutation_rates_nonsense_NP_01120680.items():\n",
    "        codon_position_in_gene=(v[0])\n",
    "        variant = v[3]\n",
    "        site = v[1]\n",
    "        mutation_rate_value = v[2] \n",
    "        mutation_rates_list.append((v[0], variant, site, mutation_rate_value))\n",
    "        TET2_mutation_rate_nonsense_NP_01120680_sum.append(mutation_rate_value)\n",
    "\n",
    "    #Get the average of all the mutation rates\n",
    "    TET2_mutation_rate_nonsense_NP_01120680_sum = []\n",
    "    for k, v in TET2_mutation_rates_nonsense_NP_01120680.items():\n",
    "        mut_rate = v[2]\n",
    "        TET2_mutation_rate_nonsense_NP_01120680_sum.append(mut_rate)\n",
    "\n",
    "    TET2_Summed_mutation_rate_nonsense_NP_01120680_added = np.sum(TET2_mutation_rate_nonsense_NP_01120680_sum)\n",
    "\n",
    "    print('Total summed mutation rate across TET2 nonsense_NP_01120680 for this study = ', TET2_Summed_mutation_rate_nonsense_NP_01120680_added)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all # e.g. lower= 0, upper = 913 for whole gene\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_missense_range_of_variants_TET2(lower_range, upper_range):\n",
    "\n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base_TET2_NP01120680 = [3409, 3804, 4044, 4537]\n",
    "   \n",
    "    def intron_exon_check_TET2_NP01120680(position_in_gene_number):\n",
    "        if position_in_gene_number == 3409:\n",
    "            gene_pos = 3409\n",
    "            first_base = TET2_exons_extra_bases_NP_01120680[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 3804:\n",
    "            gene_pos = 3804\n",
    "            first_base = 'G'\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = TET2_exons_extra_bases_NP_01120680[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 4044:\n",
    "            gene_pos = 4044\n",
    "            first_base = TET2_exons_extra_bases_NP_01120680[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 4537:\n",
    "            gene_pos = 4537\n",
    "            first_base = TET2_exons_extra_bases_NP_01120680[gene_pos-1]\n",
    "            middle_base = TET2_exons_extra_bases_NP_01120680[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #TET2 transcript (NP_01120680):\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000168769;r=4:106067032-106163928;t=ENST00000380013\n",
    "    TET2_exons_NP_01120680 = 'ATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGAGCAAATTATTGAAAAAGATGAAGGTCCTTTTTATACCCATCTAGGAGCAGGTCCTAATGTGGCAGCTATTAGAGAAATCATGGAAGAAAGGTTTGGACAGAAGGGTAAAGCTATTAGGATTGAAAGAGTCATCTATACTGGTAAAGAAGGCAAAAGTTCTCAGGGATGTCCTATTGCTAAGTGGGTGGTTCGCAGAAGCAGCAGTGAAGAGAAGCTACTGTGTTTGGTGCGGGAGCGAGCTGGCCACACCTGTGAGGCTGCAGTGATTGTGATTCTCATCCTGGTGTGGGAAGGAATCCCGCTGTCTCTGGCTGACAAACTCTACTCGGAGCTTACCGAGACGCTGAGGAAATACGGCACGCTCACCAATCGCCGGTGTGCCTTGAATGAAGAGAGAACTTGCGCCTGTCAGGGGCTGGATCCAGAAACCTGTGGTGCCTCCTTCTCTTTTGGTTGTTCATGGAGCATGTACTACAATGGATGTAAGTTTGCCAGAAGCAAGATCCCAAGGAAGTTTAAGCTGCTTGGGGATGACCCAAAAGAGGAAGAGAAACTGGAGTCTCATTTGCAAAACCTGTCCACTCTTATGGCACCAACATATAAGAAACTTGCACCTGATGCATATAATAATCAGATTGAATATGAACACAGAGCACCAGAGTGCCGTCTGGGTCTGAAGGAAGGCCGTCCATTCTCAGGGGTCACTGCATGTTTGGACTTCTGTGCTCATGCCCACAGAGACTTGCACAACATGCAGAATGGCAGCACATTGGTATGCACTCTCACTAGAGAAGACAATCGAGAATTTGGAGGAAAACCTGAGGATGAGCAGCTTCACGTTCTGCCTTTATACAAAGTCTCTGACGTGGATGAGTTTGGGAGTGTGGAAGCTCAGGAGGAGAAAAAACGGAGTGGTGCCATTCAGGTACTGAGTTCTTTTCGGCGAAAAGTCAGGATGTTAGCAGAGCCAGTCAAGACTTGCCGACAAAGGAAACTAGAAGCCAAGAAAGCTGCAGCTGAAAAGCTTTCCTCCCTGGAGAACAGCTCAAATAAAAATGAAAAGGAAAAGTCAGCCCCATCACGTACAAAACAAACTGAAAACGCAAGCCAGGCTAAACAGTTGGCAGAACTTTTGCGACTTTCAGGACCAGTCATGCAGCAGTCCCAGCAGCCCCAGCCTCTACAGAAGCAGCCACCACAGCCCCAGCAGCAGCAGAGACCCCAGCAGCAGCAGCCACATCACCCTCAGACAGAGTCTGTCAACTCTTATTCTGCTTCTGGATCCACCAATCCATACATGAGACGGCCCAATCCAGTTAGTCCTTATCCAAACTCTTCACACACTTCAGATATCTATGGAAGCACCAGCCCTATGAACTTCTATTCCACCTCATCTCAAGCTGCAGGTTCATATTTGAATTCTTCTAATCCCATGAACCCTTACCCTGGGCTTTTGAATCAGAATACCCAATATCCATCATATCAATGCAATGGAAACCTATCAGTGGACAACTGCTCCCCATATCTGGGTTCCTATTCTCCCCAGTCTCAGCCGATGGATCTGTATAGGTATCCAAGCCAAGACCCTCTGTCTAAGCTCAGTCTACCACCCATCCATACACTTTACCAGCCAAGGTTTGGAAATAGCCAGAGTTTTACATCTAAATACTTAGGTTATGGAAACCAAAATATGCAGGGAGATGGTTTCAGCAGTTGTACCATTAGACCAAATGTACATCATGTAGGGAAATTGCCTCCTTATCCCACTCATGAGATGGATGGCCACTTCATGGGAGCCACCTCTAGATTACCACCCAATCTGAGCAATCCAAACATGGACTATAAAAATGGTGAACATCATTCACCTTCTCACATAATCCATAACTACAGTGCAGCTCCGGGCATGTTCAACAGCTCTCTTCATGCCCTGCATCTCCAAAACAAGGAGAATGACATGCTTTCCCACACAGCTAATGGGTTATCAAAGATGCTTCCAGCTCTTAACCATGATAGAACTGCTTGTGTCCAAGGAGGCTTACACAAATTAAGTGATGCTAATGGTCAGGAAAAGCAGCCATTGGCACTAGTCCAGGGTGTGGCTTCTGGTGCAGAGGACAACGATGAGGTCTGGTCAGACAGCGAGCAGAGCTTTCTGGATCCTGACATTGGGGGAGTGGCCGTGGCTCCAACTCATGGGTCAATTCTCATTGAGTGTGCAAAGCGTGAGCTGCATGCCACAACCCCTTTAAAGAATCCCAATAGGAATCACCCCACCAGGATCTCCCTCGTCTTTTACCAGCATAAGAGCATGAATGAGCCAAAACATGGCTTGGCTCTTTGGGAAGCCAAAATGGCTGAAAAAGCCCGTGAGAAAGAGGAAGAGTGTGAAAAGTATGGCCCAGACTATGTGCCTCAGAAATCCCATGGCAAAAAAGTGAAACGGGAGCCTGCTGAGCCACATGAAACTTCAGAGCCCACTTACCTGCGTTTCATCAAGTCTCTTGCCGAAAGGACCATGTCCGTGACCACAGACTCCACAGTAACTACATCTCCATATGCCTTCACTCGGGTCACAGGGCCTTACAACAGATATATATGA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    TET2_codons_NP_01120680 = [TET2_exons_NP_01120680[i:i+3] for i in range(0, len(TET2_exons_NP_01120680), 3)]\n",
    "\n",
    "    #TET2 transcript (transcript variant 2 (TET2-001), NM_017628, NP_001120680):\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Exons?db=core;g=ENSG00000168769;r=4:106067032-106163928;t=ENST00000380013\n",
    "    TET2_exons_extra_bases_NP_01120680 = 'GATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGAGCAAATTATTGAAAAAGATGAAGGTCCTTTTTATACCCATCTAGGAGCAGGTCCTAATGTGGCAGCTATTAGAGAAATCATGGAAGAAAGGTTTGGACAGAAGGGTAAAGCTATTAGGATTGAAAGAGTCATCTATACTGGTAAAGAAGGCAAAAGTTCTCAGGGATGTCCTATTGCTAAGTGGGTGGTTCGCAGAAGCAGCAGTGAAGAGAAGCTACTGTGTTTGGTGCGGGAGCGAGCTGGCCACACCTGTGAGGCTGCAGTGATTGTGATTCTCATCCTGGTGTGGGAAGGAATCCCGCTGTCTCTGGCTGACAAACTCTACTCGGAGCTTACCGAGACGCTGAGGAAATACGGCACGCTCACCAATCGCCGGTGTGCCTTGAATGAAGAGAGAACTTGCGCCTGTCAGGGGCTGGATCCAGAAACCTGTGGTGCCTCCTTCTCTTTTGGTTGTTCATGGAGCATGTACTACAATGGATGTAAGTTTGCCAGAAGCAAGATCCCAAGGAAGTTTAAGCTGCTTGGGGATGACCCAAAAGAGGAAGAGAAACTGGAGTCTCATTTGCAAAACCTGTCCACTCTTATGGCACCAACATATAAGAAACTTGCACCTGATGCATATAATAATCAGATTGAATATGAACACAGAGCACCAGAGTGCCGTCTGGGTCTGAAGGAAGGCCGTCCATTCTCAGGGGTCACTGCATGTTTGGACTTCTGTGCTCATGCCCACAGAGACTTGCACAACATGCAGAATGGCAGCACATTGGTATGCACTCTCACTAGAGAAGACAATCGAGAATTTGGAGGAAAACCTGAGGATGAGCAGCTTCACGTTCTGCCTTTATACAAAGTCTCTGACGTGGATGAGTTTGGGAGTGTGGAAGCTCAGGAGGAGAAAAAACGGAGTGGTGCCATTCAGGTACTGAGTTCTTTTCGGCGAAAAGTCAGGATGTTAGCAGAGCCAGTCAAGACTTGCCGACAAAGGAAACTAGAAGCCAAGAAAGCTGCAGCTGAAAAGCTTTCCTCCCTGGAGAACAGCTCAAATAAAAATGAAAAGGAAAAGTCAGCCCCATCACGTACAAAACAAACTGAAAACGCAAGCCAGGCTAAACAGTTGGCAGAACTTTTGCGACTTTCAGGACCAGTCATGCAGCAGTCCCAGCAGCCCCAGCCTCTACAGAAGCAGCCACCACAGCCCCAGCAGCAGCAGAGACCCCAGCAGCAGCAGCCACATCACCCTCAGACAGAGTCTGTCAACTCTTATTCTGCTTCTGGATCCACCAATCCATACATGAGACGGCCCAATCCAGTTAGTCCTTATCCAAACTCTTCACACACTTCAGATATCTATGGAAGCACCAGCCCTATGAACTTCTATTCCACCTCATCTCAAGCTGCAGGTTCATATTTGAATTCTTCTAATCCCATGAACCCTTACCCTGGGCTTTTGAATCAGAATACCCAATATCCATCATATCAATGCAATGGAAACCTATCAGTGGACAACTGCTCCCCATATCTGGGTTCCTATTCTCCCCAGTCTCAGCCGATGGATCTGTATAGGTATCCAAGCCAAGACCCTCTGTCTAAGCTCAGTCTACCACCCATCCATACACTTTACCAGCCAAGGTTTGGAAATAGCCAGAGTTTTACATCTAAATACTTAGGTTATGGAAACCAAAATATGCAGGGAGATGGTTTCAGCAGTTGTACCATTAGACCAAATGTACATCATGTAGGGAAATTGCCTCCTTATCCCACTCATGAGATGGATGGCCACTTCATGGGAGCCACCTCTAGATTACCACCCAATCTGAGCAATCCAAACATGGACTATAAAAATGGTGAACATCATTCACCTTCTCACATAATCCATAACTACAGTGCAGCTCCGGGCATGTTCAACAGCTCTCTTCATGCCCTGCATCTCCAAAACAAGGAGAATGACATGCTTTCCCACACAGCTAATGGGTTATCAAAGATGCTTCCAGCTCTTAACCATGATAGAACTGCTTGTGTCCAAGGAGGCTTACACAAATTAAGTGATGCTAATGGTCAGGAAAAGCAGCCATTGGCACTAGTCCAGGGTGTGGCTTCTGGTGCAGAGGACAACGATGAGGTCTGGTCAGACAGCGAGCAGAGCTTTCTGGATCCTGACATTGGGGGAGTGGCCGTGGCTCCAACTCATGGGTCAATTCTCATTGAGTGTGCAAAGCGTGAGCTGCATGCCACAACCCCTTTAAAGAATCCCAATAGGAATCACCCCACCAGGATCTCCCTCGTCTTTTACCAGCATAAGAGCATGAATGAGCCAAAACATGGCTTGGCTCTTTGGGAAGCCAAAATGGCTGAAAAAGCCCGTGAGAAAGAGGAAGAGTGTGAAAAGTATGGCCCAGACTATGTGCCTCAGAAATCCCATGGCAAAAAAGTGAAACGGGAGCCTGCTGAGCCACATGAAACTTCAGAGCCCACTTACCTGCGTTTCATCAAGTCTCTTGCCGAAAGGACCATGTCCGTGACCACAGACTCCACAGTAACTACATCTCCATATGCCTTCACTCGGGTCACAGGGCCTTACAACAGATATATATGAT'\n",
    "\n",
    "    #List of all amino acid positions (across whole coding region) covered (amino acid position minus 1 (as python starts at 0))\n",
    "    TET2_missense_codons_DICT_NP_01120680={}\n",
    "    for i in range(lower_range, upper_range):\n",
    "        TET2_missense_codons_DICT_NP_01120680[i+1]= (TET2_codons_NP_01120680[i])\n",
    "\n",
    "    calls_TET2_missense_NP_01120680 = {}\n",
    "    for k,v in TET2_missense_codons_DICT_NP_01120680.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_TET2_missense_NP_01120680[k]=(v, changes)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_TET2_missense_NP_01120680_extra = {}\n",
    "\n",
    "    for k, v in calls_TET2_missense_NP_01120680.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_TET2_missense_NP_01120680_extra[k] = (Amino_acid_change, v[0], v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_TET2_missense_NP_01120680_extra2 = {}\n",
    "\n",
    "    for k, v in calls_TET2_missense_NP_01120680_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_TET2_missense_NP_01120680_extra2[k] = (Amino_acid_abbrev, v[1], v[2])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary that contains details of the positions of missense variants at that codon\n",
    "    TET2_position_in_list_dict_missense_NP_01120680 = {}\n",
    "\n",
    "    for k, v in calls_TET2_missense_NP_01120680_extra2.items():\n",
    "        original_AA = v[0]\n",
    "        possible_AA_changes = v[2]\n",
    "        possible_AA_changes_string = ''.join(v[2])\n",
    "        positions_in_list = []\n",
    "        for i in possible_AA_changes:\n",
    "            if i != original_AA: #not synonymous\n",
    "                if i != '*': #not missense\n",
    "                    position_in_list=-1\n",
    "                    missense_change = i\n",
    "                    new_aa_list = []\n",
    "                    while True:\n",
    "                        position_in_list = possible_AA_changes_string.find(missense_change, position_in_list+1)\n",
    "        #               print(position_in_list)\n",
    "                        if position_in_list == -1: break\n",
    "                        positions_in_list.append(position_in_list)\n",
    "                        positions_in_list = set(positions_in_list) #convert into a 'set' to remove duplicate positions\n",
    "                        positions_in_list = list(positions_in_list) #convert back to a list so can be added to dictionary\n",
    "                    for pos in positions_in_list:\n",
    "                        new_aa = possible_AA_changes_string[pos]\n",
    "                        new_aa_list.append(new_aa)\n",
    "                    TET2_position_in_list_dict_missense_NP_01120680[k]=(positions_in_list, v, new_aa_list)\n",
    "\n",
    "        aa_positions_variants_called = []\n",
    "        for k, v in TET2_position_in_list_dict_missense_NP_01120680.items():\n",
    "            aa_positions_variants_called.append(k)\n",
    "    # print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TET2_bases_missense_NP_01120680 = {}\n",
    "    for k, v in TET2_position_in_list_dict_missense_NP_01120680.items():\n",
    "        original_a_a = v[1][0]\n",
    "        codon_seq = v[1][1]\n",
    "        base_changes=[]\n",
    "        new_a_a_list = v[2]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TET2_bases_missense_NP_01120680[k]=(v, base_changes, original_a_a, new_a_a_list)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TET2_bases_missense2_NP_01120680 = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense_NP_01120680.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        new_a_a_list = v[3]\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TET2_bases_missense2_NP_01120680[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TET2_bases_missense3_NP_01120680 = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense2_NP_01120680.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_a_a = v[0][2]\n",
    "        new_aa_list = v[0][0][2]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = k\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(k)-1)*3)+codon_pos #codon number minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][1][1][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TET2_bases_missense3_NP_01120680[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, original_a_a, new_aa_list)\n",
    "\n",
    "    # Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TET2_bases_missense4_NP_01120680 = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense3_NP_01120680.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "        new_a_a_list = v[6]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_TET2_NP01120680:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check_TET2_NP01120680(i)\n",
    "                sites_trio_missense.append([adapted_trio])\n",
    "                TET2_bases_missense4_NP_01120680[k] = (v, sites_trio_missense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_missense = [TET2_exons_extra_bases_NP_01120680[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_missense.append(site_trio_missense)\n",
    "                    TET2_bases_missense4_NP_01120680[k] = (v, sites_trio_missense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TET2_sites_missense5_NP_01120680 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TET2_bases_missense4_NP_01120680.items():\n",
    "        original_a_a = v[0][0][0]\n",
    "        new_a_a = v[0][6]\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            new_amino_acid = new_a_a[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, original_a_a, new_amino_acid])\n",
    "            change = change+1\n",
    "        TET2_sites_missense5_NP_01120680[a]=info\n",
    "        a=a+1\n",
    "    #     print(TET2_sites_missense5_NP_01120680)\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TET2_sites_missense6_NP_01120680={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TET2_sites_missense5_NP_01120680.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        original_a_a = v[5]\n",
    "        new_a_a = v[6]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TET2_sites_missense6_NP_01120680[k] = (v[0],changes3, codon_n_912, original_a_a, new_a_a) #site_trio = the affected base and its neighbours\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    TET2_mutation_rates_missense_NP_01120680 = {}\n",
    "    for k, v in TET2_sites_missense6_NP_01120680.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            original_a_a = v[3]\n",
    "            new_a_a = v[4]\n",
    "            aa_positioning = v[2]\n",
    "            variant_id = str(original_a_a)+str(aa_positioning)+str(new_a_a)\n",
    "            TET2_mutation_rates_missense_NP_01120680[k] = (v[0],v[1], mutation_rate, variant_id)\n",
    "\n",
    "    TET2_mutation_rate_missense_NP_01120680_sum = []\n",
    "\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TET2_mutation_rates_missense_NP_01120680.items():\n",
    "        codon_position_in_gene=(v[0])\n",
    "        variant = v[3]\n",
    "        site = v[1]\n",
    "        mutation_rate_value = v[2] \n",
    "        mutation_rates_list.append((v[0], variant, site, mutation_rate_value))\n",
    "        TET2_mutation_rate_missense_NP_01120680_sum.append(mutation_rate_value)\n",
    "\n",
    "    #Get the average of all the mutation rates\n",
    "    TET2_mutation_rate_missense_NP_01120680_sum = []\n",
    "    for k, v in TET2_mutation_rates_missense_NP_01120680.items():\n",
    "        mut_rate = v[2]\n",
    "        TET2_mutation_rate_missense_NP_01120680_sum.append(mut_rate)\n",
    "\n",
    "    TET2_Summed_mutation_rate_missense_NP_01120680_added = np.sum(TET2_mutation_rate_missense_NP_01120680_sum)\n",
    "\n",
    "    print('Total summed mutation rate across TET2 missense_NP_01120680 for this study = ', TET2_Summed_mutation_rate_missense_NP_01120680_added)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all # e.g. lower= 0, upper = 913 for whole gene\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_nonsense_range_of_variants_TET2_shorter(lower_range, upper_range):\n",
    "    \n",
    "    #TET2 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Summary?db=core;g=ENSG00000168769;r=4:106067032-106200973;t=ENST00000305737\n",
    "    TET2_exons_NP_060098 = 'ATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGGTAAGTGCCAGAAATGTACTGAGACACATGGCGTTTATCCAGAATTAGCAAATTTATCTTCAGATATGGGATTTTCCTTCTTTTTTTAA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    TET2_codons_NP_060098 = [TET2_exons_NP_060098[i:i+3] for i in range(0, len(TET2_exons_NP_060098), 3)]\n",
    "\n",
    "    #TET2 transcript 1 extra base at each end:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Summary?db=core;g=ENSG00000168769;r=4:106067032-106200973;t=ENST00000305737\n",
    "    TET2_exons_extra_bases_NP_060098 = 'GATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGGTAAGTGCCAGAAATGTACTGAGACACATGGCGTTTATCCAGAATTAGCAAATTTATCTTCAGATATGGGATTTTCCTTCTTTTTTTAAA'\n",
    "\n",
    "    #List of all amino acid positions (across whole coding region) covered (amino acid position minus 1 (as python starts at 0))\n",
    "    TET2_nonsense_codons_DICT_NP_060098={}\n",
    "    for i in range(lower_range, upper_range):\n",
    "        TET2_nonsense_codons_DICT_NP_060098[i+1]= (TET2_codons_NP_060098[i])\n",
    "\n",
    "    calls_TET2_nonsense_NP_060098 = {}\n",
    "    for k,v in TET2_nonsense_codons_DICT_NP_060098.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_TET2_nonsense_NP_060098[k]=(v, changes)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_TET2_nonsense_NP_060098_extra = {}\n",
    "\n",
    "    for k, v in calls_TET2_nonsense_NP_060098.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_TET2_nonsense_NP_060098_extra[k] = (Amino_acid_change, v[0], v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_TET2_nonsense_NP_060098_extra2 = {}\n",
    "\n",
    "    for k, v in calls_TET2_nonsense_NP_060098_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_TET2_nonsense_NP_060098_extra2[k] = (Amino_acid_abbrev, v[1], v[2])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary that contains details of the positions of nonsense variants at that codon\n",
    "    TET2_position_in_list_dict_nonsense_NP_060098 = {}\n",
    "\n",
    "    for k, v in calls_TET2_nonsense_NP_060098_extra2.items():\n",
    "        original_AA = v[0]\n",
    "        possible_AA_changes = v[2]\n",
    "        possible_AA_changes_string = ''.join(v[2])\n",
    "        positions_in_list = []\n",
    "        for i in possible_AA_changes:\n",
    "            if i == '*': #nonsense\n",
    "                position_in_list=-1\n",
    "                nonsense_change = i    \n",
    "                while True:\n",
    "                    position_in_list = possible_AA_changes_string.find(nonsense_change, position_in_list+1)\n",
    "    #               print(position_in_list)\n",
    "                    if position_in_list == -1: break\n",
    "                    positions_in_list.append(position_in_list)\n",
    "                    positions_in_list = set(positions_in_list) #convert into a 'set' to remove duplicate positions\n",
    "                    positions_in_list = list(positions_in_list) #convert back to a list so can be added to dictionary\n",
    "                TET2_position_in_list_dict_nonsense_NP_060098[k]=(positions_in_list, v)\n",
    "\n",
    "        aa_positions_variants_called = []\n",
    "        for k, v in TET2_position_in_list_dict_nonsense_NP_060098.items():\n",
    "            aa_positions_variants_called.append(k)\n",
    "    # print('positions at which nonsense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TET2_bases_nonsense_NP_060098 = {}\n",
    "    for k, v in TET2_position_in_list_dict_nonsense_NP_060098.items():\n",
    "        original_a_a = v[1][0]\n",
    "        codon_seq = v[1][1]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TET2_bases_nonsense_NP_060098[k]=(v, base_changes, original_a_a)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TET2_bases_nonsense2_NP_060098 = {}\n",
    "\n",
    "    for k, v in TET2_bases_nonsense_NP_060098.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TET2_bases_nonsense2_NP_060098[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TET2_bases_nonsense3_NP_060098 = {}\n",
    "\n",
    "    for k, v in TET2_bases_nonsense2_NP_060098.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_a_a = v[0][2]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = k\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(k)-1)*3)+codon_pos #codon number minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][1][1][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TET2_bases_nonsense3_NP_060098[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, original_a_a)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TET2_bases_nonsense4_NP_060098 = {}\n",
    "\n",
    "    for k, v in TET2_bases_nonsense3_NP_060098.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_nonsense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "            gene_pos = i\n",
    "            site_trio_nonsense = [TET2_exons_extra_bases_NP_060098[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "            sites_trio_nonsense.append(site_trio_nonsense)\n",
    "            TET2_bases_nonsense4_NP_060098[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TET2_sites_nonsense5_NP_060098 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TET2_bases_nonsense4_NP_060098.items():\n",
    "        original_a_a = v[0][0][0]\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, original_a_a])\n",
    "            change = change+1\n",
    "        TET2_sites_nonsense5_NP_060098[a]=info\n",
    "        a=a+1\n",
    "    #     print(TET2_sites_nonsense5_NP_060098)\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TET2_sites_nonsense6_NP_060098={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TET2_sites_nonsense5_NP_060098.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        original_a_a = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TET2_sites_nonsense6_NP_060098[k] = (v[0],changes3, codon_n_912, original_a_a) #site_trio = the affected base and its neighbours\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    TET2_mutation_rates_nonsense_NP_060098 = {}\n",
    "    for k, v in TET2_sites_nonsense6_NP_060098.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            original_a_a = v[3]\n",
    "            aa_positioning = v[2]\n",
    "            variant_type = '*'\n",
    "            variant_id = str(original_a_a)+str(aa_positioning)+variant_type\n",
    "            TET2_mutation_rates_nonsense_NP_060098[k] = (v[0],v[1], mutation_rate, variant_id)\n",
    "\n",
    "    TET2_mutation_rate_nonsense_NP_060098_sum = []\n",
    "\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TET2_mutation_rates_nonsense_NP_060098.items():\n",
    "        codon_position_in_gene=(v[0])\n",
    "        variant = v[3]\n",
    "        site = v[1]\n",
    "        mutation_rate_value = v[2] \n",
    "        mutation_rates_list.append((v[0], variant, site, mutation_rate_value))\n",
    "        TET2_mutation_rate_nonsense_NP_060098_sum.append(mutation_rate_value)\n",
    "\n",
    "    #Get the average of all the mutation rates\n",
    "    TET2_mutation_rate_nonsense_NP_060098_sum = []\n",
    "    for k, v in TET2_mutation_rates_nonsense_NP_060098.items():\n",
    "        mut_rate = v[2]\n",
    "        TET2_mutation_rate_nonsense_NP_060098_sum.append(mut_rate)\n",
    "\n",
    "    TET2_Summed_mutation_rate_nonsense_NP_060098_added = np.sum(TET2_mutation_rate_nonsense_NP_060098_sum)\n",
    "\n",
    "    print('Total summed mutation rate across TET2 nonsense_NP_060098 for this study = ', TET2_Summed_mutation_rate_nonsense_NP_060098_added)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all # e.g. lower= 0, upper = 913 for whole gene\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_missense_range_of_variants_TET2_shorter(lower_range, upper_range):\n",
    "    \n",
    "    #TET2 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Summary?db=core;g=ENSG00000168769;r=4:106067032-106200973;t=ENST00000305737\n",
    "    TET2_exons_NP_060098 = 'ATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGGTAAGTGCCAGAAATGTACTGAGACACATGGCGTTTATCCAGAATTAGCAAATTTATCTTCAGATATGGGATTTTCCTTCTTTTTTTAA'\n",
    "\n",
    "    #Split the sequence into codons \n",
    "    TET2_codons_NP_060098 = [TET2_exons_NP_060098[i:i+3] for i in range(0, len(TET2_exons_NP_060098), 3)]\n",
    "\n",
    "    #TET2 transcript 1 extra base at each end:\n",
    "    #Downloaded from http://grch37.ensembl.org/Homo_sapiens/Transcript/Summary?db=core;g=ENSG00000168769;r=4:106067032-106200973;t=ENST00000305737\n",
    "    TET2_exons_extra_bases_NP_060098 = 'GATGGAACAGGATAGAACCAACCATGTTGAGGGCAACAGACTAAGTCCATTCCTGATACCATCACCTCCCATTTGCCAGACAGAACCTCTGGCTACAAAGCTCCAGAATGGAAGCCCACTGCCTGAGAGAGCTCATCCAGAAGTAAATGGAGACACCAAGTGGCACTCTTTCAAAAGTTATTATGGAATACCCTGTATGAAGGGAAGCCAGAATAGTCGTGTGAGTCCTGACTTTACACAAGAAAGTAGAGGGTATTCCAAGTGTTTGCAAAATGGAGGAATAAAACGCACAGTTAGTGAACCTTCTCTCTCTGGGCTCCTTCAGATCAAGAAATTGAAACAAGACCAAAAGGCTAATGGAGAAAGACGTAACTTCGGGGTAAGCCAAGAAAGAAATCCAGGTGAAAGCAGTCAACCAAATGTCTCCGATTTGAGTGATAAGAAAGAATCTGTGAGTTCTGTAGCCCAAGAAAATGCAGTTAAAGATTTCACCAGTTTTTCAACACATAACTGCAGTGGGCCTGAAAATCCAGAGCTTCAGATTCTGAATGAGCAGGAGGGGAAAAGTGCTAATTACCATGACAAGAACATTGTATTACTTAAAAACAAGGCAGTGCTAATGCCTAATGGTGCTACAGTTTCTGCCTCTTCCGTGGAACACACACATGGTGAACTCCTGGAAAAAACACTGTCTCAATATTATCCAGATTGTGTTTCCATTGCGGTGCAGAAAACCACATCTCACATAAATGCCATTAACAGTCAGGCTACTAATGAGTTGTCCTGTGAGATCACTCACCCATCGCATACCTCAGGGCAGATCAATTCCGCACAGACCTCTAACTCTGAGCTGCCTCCAAAGCCAGCTGCAGTGGTGAGTGAGGCCTGTGATGCTGATGATGCTGATAATGCCAGTAAACTAGCTGCAATGCTAAATACCTGTTCCTTTCAGAAACCAGAACAACTACAACAACAAAAATCAGTTTTTGAGATATGCCCATCTCCTGCAGAAAATAACATCCAGGGAACCACAAAGCTAGCGTCTGGTGAAGAATTCTGTTCAGGTTCCAGCAGCAATTTGCAAGCTCCTGGTGGCAGCTCTGAACGGTATTTAAAACAAAATGAAATGAATGGTGCTTACTTCAAGCAAAGCTCAGTGTTCACTAAGGATTCCTTTTCTGCCACTACCACACCACCACCACCATCACAATTGCTTCTTTCTCCCCCTCCTCCTCTTCCACAGGTTCCTCAGCTTCCTTCAGAAGGAAAAAGCACTCTGAATGGTGGAGTTTTAGAAGAACACCACCACTACCCCAACCAAAGTAACACAACACTTTTAAGGGAAGTGAAAATAGAGGGTAAACCTGAGGCACCACCTTCCCAGAGTCCTAATCCATCTACACATGTATGCAGCCCTTCTCCGATGCTTTCTGAAAGGCCTCAGAATAATTGTGTGAACAGGAATGACATACAGACTGCAGGGACAATGACTGTTCCATTGTGTTCTGAGAAAACAAGACCAATGTCAGAACACCTCAAGCATAACCCACCAATTTTTGGTAGCAGTGGAGAGCTACAGGACAACTGCCAGCAGTTGATGAGAAACAAAGAGCAAGAGATTCTGAAGGGTCGAGACAAGGAGCAAACACGAGATCTTGTGCCCCCAACACAGCACTATCTGAAACCAGGATGGATTGAATTGAAGGCCCCTCGTTTTCACCAAGCGGAATCCCATCTAAAACGTAATGAGGCATCACTGCCATCAATTCTTCAGTATCAACCCAATCTCTCCAATCAAATGACCTCCAAACAATACACTGGAAATTCCAACATGCCTGGGGGGCTCCCAAGGCAAGCTTACACCCAGAAAACAACACAGCTGGAGCACAAGTCACAAATGTACCAAGTTGAAATGAATCAAGGGCAGTCCCAAGGTACAGTGGACCAACATCTCCAGTTCCAAAAACCCTCACACCAGGTGCACTTCTCCAAAACAGACCATTTACCAAAAGCTCATGTGCAGTCACTGTGTGGCACTAGATTTCATTTTCAACAAAGAGCAGATTCCCAAACTGAAAAACTTATGTCCCCAGTGTTGAAACAGCACTTGAATCAACAGGCTTCAGAGACTGAGCCATTTTCAAACTCACACCTTTTGCAACATAAGCCTCATAAACAGGCAGCACAAACACAACCATCCCAGAGTTCACATCTCCCTCAAAACCAGCAACAGCAGCAAAAATTACAAATAAAGAATAAAGAGGAAATACTCCAGACTTTTCCTCACCCCCAAAGCAACAATGATCAGCAAAGAGAAGGATCATTCTTTGGCCAGACTAAAGTGGAAGAATGTTTTCATGGTGAAAATCAGTATTCAAAATCAAGCGAGTTCGAGACTCATAATGTCCAAATGGGACTGGAGGAAGTACAGAATATAAATCGTAGAAATTCCCCTTATAGTCAGACCATGAAATCAAGTGCATGCAAAATACAGGTTTCTTGTTCAAACAATACACACCTAGTTTCAGAGAATAAAGAACAGACTACACATCCTGAACTTTTTGCAGGAAACAAGACCCAAAACTTGCATCACATGCAATATTTTCCAAATAATGTGATCCCAAAGCAAGATCTTCTTCACAGGTGCTTTCAAGAACAGGAGCAGAAGTCACAACAAGCTTCAGTTCTACAGGGATATAAAAATAGAAACCAAGATATGTCTGGTCAACAAGCTGCGCAACTTGCTCAGCAAAGGTACTTGATACATAACCATGCAAATGTTTTTCCTGTGCCTGACCAGGGAGGAAGTCACACTCAGACCCCTCCCCAGAAGGACACTCAAAAGCATGCTGCTCTAAGGTGGCATCTCTTACAGAAGCAAGAACAGCAGCAAACACAGCAACCCCAAACTGAGTCTTGCCATAGTCAGATGCACAGGCCAATTAAGGTGGAACCTGGATGCAAGCCACATGCCTGTATGCACACAGCACCACCAGAAAACAAAACATGGAAAAAGGTAACTAAGCAAGAGAATCCACCTGCAAGCTGTGATAATGTGCAGCAAAAGAGCATCATTGAGACCATGGAGCAGCATCTGAAGCAGTTTCACGCCAAGTCGTTATTTGACCATAAGGCTCTTACTCTCAAATCACAGAAGCAAGTAAAAGTTGAAATGTCAGGGCCAGTCACAGTTTTGACTAGACAAACCACTGCTGCAGAACTTGATAGCCACACCCCAGCTTTAGAGCAGCAAACAACTTCTTCAGAAAAGACACCAACCAAAAGAACAGCTGCTTCTGTTCTCAATAATTTTATAGAGTCACCTTCCAAATTACTAGATACTCCTATAAAAAATTTATTGGATACACCTGTCAAGACTCAATATGATTTCCCATCTTGCAGATGTGTAGGTAAGTGCCAGAAATGTACTGAGACACATGGCGTTTATCCAGAATTAGCAAATTTATCTTCAGATATGGGATTTTCCTTCTTTTTTTAAA'\n",
    "\n",
    "    #List of all amino acid positions (across whole coding region) covered (amino acid position minus 1 (as python starts at 0))\n",
    "    TET2_missense_codons_DICT_NP_072046={}\n",
    "    for i in range(lower_range, upper_range):\n",
    "        TET2_missense_codons_DICT_NP_072046[i+1]= (TET2_codons_NP_072046[i])\n",
    "\n",
    "    calls_TET2_missense_NP_072046 = {}\n",
    "    for k,v in TET2_missense_codons_DICT_NP_072046.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_TET2_missense_NP_072046[k]=(v, changes)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_TET2_missense_NP_072046_extra = {}\n",
    "\n",
    "    for k, v in calls_TET2_missense_NP_072046.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_TET2_missense_NP_072046_extra[k] = (Amino_acid_change, v[0], v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_TET2_missense_NP_072046_extra2 = {}\n",
    "\n",
    "    for k, v in calls_TET2_missense_NP_072046_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_TET2_missense_NP_072046_extra2[k] = (Amino_acid_abbrev, v[1], v[2])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary that contains details of the positions of missense variants at that codon\n",
    "    TET2_position_in_list_dict_missense_NP_072046 = {}\n",
    "\n",
    "    for k, v in calls_TET2_missense_NP_072046_extra2.items():\n",
    "        original_AA = v[0]\n",
    "        possible_AA_changes = v[2]\n",
    "        possible_AA_changes_string = ''.join(v[2])\n",
    "        positions_in_list = []\n",
    "        for i in possible_AA_changes:\n",
    "            if i != original_AA: #not synonymous\n",
    "                if i != '*': #not missense\n",
    "                    position_in_list=-1\n",
    "                    missense_change = i\n",
    "                    new_aa_list = []\n",
    "                    while True:\n",
    "                        position_in_list = possible_AA_changes_string.find(missense_change, position_in_list+1)\n",
    "        #               print(position_in_list)\n",
    "                        if position_in_list == -1: break\n",
    "                        positions_in_list.append(position_in_list)\n",
    "                        positions_in_list = set(positions_in_list) #convert into a 'set' to remove duplicate positions\n",
    "                        positions_in_list = list(positions_in_list) #convert back to a list so can be added to dictionary\n",
    "                    for pos in positions_in_list:\n",
    "                        new_aa = possible_AA_changes_string[pos]\n",
    "                        new_aa_list.append(new_aa)\n",
    "                    TET2_position_in_list_dict_missense_NP_072046[k]=(positions_in_list, v, new_aa_list)\n",
    "\n",
    "        aa_positions_variants_called = []\n",
    "        for k, v in TET2_position_in_list_dict_missense_NP_072046.items():\n",
    "            aa_positions_variants_called.append(k)\n",
    "    # print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TET2_bases_missense_NP_072046 = {}\n",
    "    for k, v in TET2_position_in_list_dict_missense_NP_072046.items():\n",
    "        original_a_a = v[1][0]\n",
    "        codon_seq = v[1][1]\n",
    "        base_changes=[]\n",
    "        new_a_a_list = v[2]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TET2_bases_missense_NP_072046[k]=(v, base_changes, original_a_a, new_a_a_list)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TET2_bases_missense2_NP_072046 = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense_NP_072046.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        new_a_a_list = v[3]\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TET2_bases_missense2_NP_072046[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TET2_bases_missense3_NP_072046 = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense2_NP_072046.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_a_a = v[0][2]\n",
    "        new_aa_list = v[0][0][2]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = k\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(k)-1)*3)+codon_pos #codon number minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][1][1][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TET2_bases_missense3_NP_072046[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, original_a_a, new_aa_list)\n",
    "\n",
    "    # Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TET2_bases_missense4_NP_072046 = {}\n",
    "\n",
    "    for k, v in TET2_bases_missense3_NP_072046.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "        new_a_a_list = v[6]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "            gene_pos = i\n",
    "            site_trio_missense = [TET2_exons_extra_bases_NP_072046[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "            sites_trio_missense.append(site_trio_missense)\n",
    "            TET2_bases_missense4_NP_072046[k] = (v, sites_trio_missense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TET2_sites_missense5_NP_072046 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TET2_bases_missense4_NP_072046.items():\n",
    "        original_a_a = v[0][0][0]\n",
    "        new_a_a = v[0][6]\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            new_amino_acid = new_a_a[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, original_a_a, new_amino_acid])\n",
    "            change = change+1\n",
    "        TET2_sites_missense5_NP_072046[a]=info\n",
    "        a=a+1\n",
    "    #     print(TET2_sites_missense5_NP_072046)\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TET2_sites_missense6_NP_072046={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TET2_sites_missense5_NP_072046.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        original_a_a = v[5]\n",
    "        new_a_a = v[6]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TET2_sites_missense6_NP_072046[k] = (v[0],changes3, codon_n_912, original_a_a, new_a_a) #site_trio = the affected base and its neighbours\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    TET2_mutation_rates_missense_NP_072046 = {}\n",
    "    for k, v in TET2_sites_missense6_NP_072046.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            original_a_a = v[3]\n",
    "            new_a_a = v[4]\n",
    "            aa_positioning = v[2]\n",
    "            variant_id = str(original_a_a)+str(aa_positioning)+str(new_a_a)\n",
    "            TET2_mutation_rates_missense_NP_072046[k] = (v[0],v[1], mutation_rate, variant_id)\n",
    "\n",
    "    TET2_mutation_rate_missense_NP_072046_sum = []\n",
    "\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TET2_mutation_rates_missense_NP_072046.items():\n",
    "        codon_position_in_gene=(v[0])\n",
    "        variant = v[3]\n",
    "        site = v[1]\n",
    "        mutation_rate_value = v[2] \n",
    "        mutation_rates_list.append((v[0], variant, site, mutation_rate_value))\n",
    "        TET2_mutation_rate_missense_NP_072046_sum.append(mutation_rate_value)\n",
    "\n",
    "    #Get the average of all the mutation rates\n",
    "    TET2_mutation_rate_missense_NP_072046_sum = []\n",
    "    for k, v in TET2_mutation_rates_missense_NP_072046.items():\n",
    "        mut_rate = v[2]\n",
    "        TET2_mutation_rate_missense_NP_072046_sum.append(mut_rate)\n",
    "\n",
    "    TET2_Summed_mutation_rate_missense_NP_072046_added = np.sum(TET2_mutation_rate_missense_NP_072046_sum)\n",
    "\n",
    "    print('Total summed mutation rate across TET2 missense_NP_072046 for this study = ', TET2_Summed_mutation_rate_missense_NP_072046_added)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all # e.g. lower= 0, upper = 913 for whole gene\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ASXL1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_nonsense_range_of_variants_ASXL1(lower_range, upper_range):\n",
    "    \n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base = [140, 143, 252, 373, 1086, 1085]\n",
    "   \n",
    "    def intron_exon_check(position_in_gene_number):\n",
    "        if position_in_gene_number == 140:\n",
    "            gene_pos = 140\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 143:\n",
    "            gene_pos = 143\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 252:\n",
    "            gene_pos = 252\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 373:\n",
    "            gene_pos = 373\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1086:\n",
    "            gene_pos = 1086\n",
    "            first_base = 'G'\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = ASXL1_exons_extra_bases[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1085:\n",
    "            gene_pos = 1085\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #ASXL1 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    ASXL1_exons = 'ATGAAGGACAAACAGAAGAAGAAGAAGGAGCGCACGTGGGCCGAGGCCGCGCGCCTGGTATTAGAAAACTACTCGGATGCTCCAATGACACCAAAACAGATTCTGCAGGTCATAGAGGCAGAAGGACTAAAGGAAATGAGAAGTGGGACTTCCCCTCTCGCATGCCTCAATGCTATGCTACATTCCAATTCAAGAGGAGGAGAGGGGTTGTTTTATAAACTGCCTGGCCGAATCAGCCTTTTCACGCTCAAGAAGGATGCCCTGCAGTGGTCTCGCCATCCAGCTACAGTGGAGGGAGAGGAGCCAGAGGACACGGCTGATGTGGAGAGCTGTGGGTCTAATGAAGCCAGCACTGTGAGTGGTGAAAACGATGTATCTCTTGATGAAACATCTTCGAACGCATCCTGTTCTACAGAATCTCAGAGTCGACCTCTTTCCAATCCCAGGGACAGCTACAGAGCTTCCTCACAGGCGAACAAACAAAAGAAAAAGACTGGGGTGATGCTGCCTCGAGTTGTCCTGACTCCTCTGAAGGTAAACGGGGCCCACGTGGAATCTGCATCAGGGTTCTCGGGCTGCCACGCCGATGGCGAGAGCGGCAGCCCGTCCAGCAGCAGCAGCGGCTCTCTGGCCCTGGGCAGCGCTGCTATTCGTGGCCAGGCCGAGGTCACCCAGGACCCTGCCCCGCTCCTGAGAGGCTTCCGGAAGCCAGCCACAGGTCAAATGAAGCGCAACAGAGGGGAAGAAATAGATTTTGAGACACCTGGGTCCATTCTTGTCAACACCAACCTCCGTGCCCTGATCAACTCTCGGACCTTCCATGCCTTACCATCACACTTCCAGCAGCAGCTCCTCTTCCTCCTGCCTGAAGTAGACAGACAGGTGGGGACGGATGGCCTGTTGCGTCTCAGCAGCAGTGCACTAAATAACGAGTTTTTTACCCATGCGGCTCAGAGCTGGCGGGAGCGCCTGGCTGATGGTGAATTTACTCATGAGATGCAAGTCAGGATACGACAGGAAATGGAGAAGGAAAAGAAGGTGGAACAATGGAAAGAAAAGTTCTTTGAAGACTACTATGGACAGAAGCTGGGTTTGACCAAAGAAGAGTCATTGCAGCAGAACGTGGGCCAGGAGGAGGCTGAAATCAAAAGTGGCTTGTGTGTCCCAGGAGAATCAGTGCGTATACAGCGTGGTCCAGCCACCCGACAGCGAGATGGGCATTTTAAGAAACGCTCTCGGCCAGATCTCCGAACCAGAGCCAGAAGGAATCTGTACAAAAAACAGGAGTCAGAACAAGCAGGGGTTGCTAAGGATGCAAAATCTGTGGCCTCAGATGTTCCCCTCTACAAGGATGGGGAGGCTAAGACTGACCCAGCAGGGCTGAGCAGTCCCCATCTGCCAGGCACATCCTCTGCAGCACCCGACCTGGAGGGTCCCGAATTCCCAGTTGAGTCTGTGGCTTCTCGGATCCAGGCTGAGCCAGACAACTTGGCACGTGCCTCTGCATCTCCAGACAGAATTCCTAGCCTGCCTCAGGAAACTGTGGATCAGGAACCCAAGGATCAGAAGAGGAAATCCTTTGAGCAGGCGGCCTCTGCATCCTTTCCCGAAAAGAAGCCCCGGCTTGAAGATCGTCAGTCCTTTCGTAACACAATTGAAAGTGTTCACACCGAAAAGCCACAGCCCACTAAAGAGGAGCCCAAAGTCCCGCCCATCCGGATTCAACTTTCACGTATCAAACCACCCTGGGTGGTTAAAGGTCAGCCCACTTACCAGATATGCCCCCGGATCATCCCCACCACGGAGTCCTCCTGCCGGGGTTGGACTGGCGCCAGGACCCTCGCAGACATTAAAGCCCGTGCTCTGCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCCACCACTGCCATCGGAGGGGGGGGTGGCCCGGGTGGAGGTGGCGGCGGGGCCACCGATGAGGGAGGTGGCAGAGGCAGCAGCAGTGGTGATGGTGGTGAGGCCTGTGGCCACCCTGAGCCCAGGGGAGGCCCGAGCACCCCTGGAAAGTGTACGTCAGATCTACAGCGAACACAACTACTGCCGCCTTATCCTCTAAATGGGGAGCATACCCAGGCCGGAACTGCCATGTCCAGAGCTAGGAGAGAGGACCTGCCTTCTCTGAGAAAGGAGGAAAGCTGCCTACTACAGAGGGCTACAGTTGGACTCACAGATGGGCTAGGAGATGCCTCCCAACTCCCCGTTGCTCCCACTGGGGACCAGCCATGCCAGGCCTTGCCCCTACTGTCCTCCCAAACCTCAGTAGCTGAGAGATTAGTGGAGCAGCCTCAGTTGCATCCGGATGTTAGAACTGAATGTGAGTCTGGCACCACTTCCTGGGAAAGTGATGATGAGGAGCAAGGACCCACCGTTCCTGCAGACAATGGTCCCATTCTGTCTCTAGTGGGAGATGATACATTAGAGAAAGGAACTGGCCAAGCTCTTGACAGTCATCCCACTATGAAGGATCCTGTAAATGTGACCCCCAGTTCCACACCTGAATCCTCACCGACTGATTGCCTGCAGAACAGAGCATTTGATGACGAATTAGGGCTTGGTGGCTCATGCCCTCCTATGAGGGAAAGTGATACTAGACAAGAAAACTTGAAAACCAAGGCTCTCGTTTCTAACAGTTCTTTGCATTGGATACCCATCCCATCGAATGATGAGGTAGTGAAACAGCCCAAACCAGAATCCAGAGAACACATACCATCTGTTGAGCCCCAGGTTGGAGAGGAGTGGGAGAAAGCTGCTCCCACCCCTCCTGCATTGCCTGGGGATTTGACAGCTGAGGAGGGTCTAGATCCTCTTGACAGCCTTACTTCACTCTGGACTGTGCCATCTCGAGGAGGCAGTGACAGCAATGGCAGTTACTGTCAACAGGTGGACATTGAAAAGCTGAAAATCAACGGAGACTCTGAAGCACTGAGTCCTCACGGTGAGTCCACGGATACAGCCTCTGACTTTGAAGGTCACCTCACGGAGGACAGCAGTGAGGCTGACACTAGAGAAGCTGCAGTGACAAAGGGATCTTCGGTGGACAAGGATGAGAAACCCAATTGGAACCAATCTGCCCCACTGTCCAAGGTGAATGGTGACATGCGTCTGGTTACAAGGACAGATGGGATGGTTGCTCCTCAGAGCTGGGTGTCTCGAGTATGTGCGGTCCGCCAAAAGATCCCAGATTCCCTACTGCTGGCCAGTACTGAGTACCAGCCAAGAGCCGTGTGCCTGTCCATGCCTGGGTCCTCAGTGGAGGCCACTAACCCACTTGTGATGCAGTTGCTGCAGGGTAGCTTGCCCCTAGAGAAGGTTCTTCCACCAGCCCACGATGACAGCATGTCAGAATCCCCACAAGTACCACTTACAAAAGACCAGAGCCATGGCTCGCTACGCATGGGATCTTTACATGGTCTTGGAAAAAACAGTGGCATGGTTGATGGAAGCAGCCCCAGTTCTTTAAGGGCTTTGAAGGAGCCTCTTCTGCCAGATAGCTGTGAAACAGGCACTGGTCTTGCCAGGATTGAGGCCACCCAGGCTCCTGGAGCACCCCAAAAGAATTGCAAGGCAGTCCCAAGTTTTGACTCCCTCCATCCAGTGACAAATCCCATTACATCCTCTAGGAAACTGGAAGAAATGGATTCCAAAGAGCAGTTCTCTTCCTTTAGTTGTGAAGATCAGAAGGAAGTCCGTGCTATGTCACAGGACAGTAATTCAAATGCTGCTCCAGGAAAGAGCCCAGGAGATCTTACTACCTCGAGAACACCTCGTTTCTCATCTCCAAATGTGATCTCCTTTGGTCCAGAGCAGACAGGTCGGGCCCTGGGTGATCAGAGCAATGTTACAGGCCAAGGGAAGAAGCTTTTTGGCTCTGGGAATGTGGCTGCAACCCTTCAGCGCCCCAGGCCTGCGGACCCGATGCCTCTTCCTGCTGAGATCCCTCCAGTTTTTCCCAGTGGGAAGTTGGGACCAAGCACAAACTCCATGTCTGGTGGGGTACAGACTCCAAGGGAAGACTGGGCTCCAAAGCCACATGCCTTTGTTGGCAGCGTCAAGAATGAGAAGACTTTTGTGGGGGGTCCTCTTAAGGCAAATGCCGAGAACAGGAAAGCTACTGGGCATAGTCCCCTGGAACTGGTGGGTCACTTGGAAGGGATGCCCTTTGTCATGGACTTGCCCTTCTGGAAATTACCCCGAGAGCCAGGGAAGGGGCTCAGTGAGCCTCTGGAGCCTTCTTCTCTCCCCTCCCAACTCAGCATCAAGCAGGCATTTTATGGGAAGCTTTCTAAACTCCAACTGAGTTCCACCAGCTTTAATTATTCCTCTAGCTCTCCCACCTTTCCCAAAGGCCTTGCTGGAAGTGTGGTGCAGCTGAGCCACAAAGCAAACTTTGGTGCGAGCCACAGTGCATCACTTTCCTTGCAAATGTTCACTGACAGCAGCACGGTGGAAAGCATCTCGCTCCAGTGTGCGTGCAGCCTGAAAGCCATGATCATGTGCCAAGGCTGCGGTGCGTTCTGTCACGATGACTGTATTGGACCCTCAAAGCTCTGTGTATTGTGCCTTGTGGTGAGATAA'\n",
    "    \n",
    "    #Split the sequence into codons \n",
    "    ASXL1_codons = [ASXL1_exons[i:i+3] for i in range(0, len(ASXL1_exons), 3)]\n",
    "\n",
    "    #ASXL1 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    ASXL1_exons_extra_bases = 'GATGAAGGACAAACAGAAGAAGAAGAAGGAGCGCACGTGGGCCGAGGCCGCGCGCCTGGTATTAGAAAACTACTCGGATGCTCCAATGACACCAAAACAGATTCTGCAGGTCATAGAGGCAGAAGGACTAAAGGAAATGAGAAGTGGGACTTCCCCTCTCGCATGCCTCAATGCTATGCTACATTCCAATTCAAGAGGAGGAGAGGGGTTGTTTTATAAACTGCCTGGCCGAATCAGCCTTTTCACGCTCAAGAAGGATGCCCTGCAGTGGTCTCGCCATCCAGCTACAGTGGAGGGAGAGGAGCCAGAGGACACGGCTGATGTGGAGAGCTGTGGGTCTAATGAAGCCAGCACTGTGAGTGGTGAAAACGATGTATCTCTTGATGAAACATCTTCGAACGCATCCTGTTCTACAGAATCTCAGAGTCGACCTCTTTCCAATCCCAGGGACAGCTACAGAGCTTCCTCACAGGCGAACAAACAAAAGAAAAAGACTGGGGTGATGCTGCCTCGAGTTGTCCTGACTCCTCTGAAGGTAAACGGGGCCCACGTGGAATCTGCATCAGGGTTCTCGGGCTGCCACGCCGATGGCGAGAGCGGCAGCCCGTCCAGCAGCAGCAGCGGCTCTCTGGCCCTGGGCAGCGCTGCTATTCGTGGCCAGGCCGAGGTCACCCAGGACCCTGCCCCGCTCCTGAGAGGCTTCCGGAAGCCAGCCACAGGTCAAATGAAGCGCAACAGAGGGGAAGAAATAGATTTTGAGACACCTGGGTCCATTCTTGTCAACACCAACCTCCGTGCCCTGATCAACTCTCGGACCTTCCATGCCTTACCATCACACTTCCAGCAGCAGCTCCTCTTCCTCCTGCCTGAAGTAGACAGACAGGTGGGGACGGATGGCCTGTTGCGTCTCAGCAGCAGTGCACTAAATAACGAGTTTTTTACCCATGCGGCTCAGAGCTGGCGGGAGCGCCTGGCTGATGGTGAATTTACTCATGAGATGCAAGTCAGGATACGACAGGAAATGGAGAAGGAAAAGAAGGTGGAACAATGGAAAGAAAAGTTCTTTGAAGACTACTATGGACAGAAGCTGGGTTTGACCAAAGAAGAGTCATTGCAGCAGAACGTGGGCCAGGAGGAGGCTGAAATCAAAAGTGGCTTGTGTGTCCCAGGAGAATCAGTGCGTATACAGCGTGGTCCAGCCACCCGACAGCGAGATGGGCATTTTAAGAAACGCTCTCGGCCAGATCTCCGAACCAGAGCCAGAAGGAATCTGTACAAAAAACAGGAGTCAGAACAAGCAGGGGTTGCTAAGGATGCAAAATCTGTGGCCTCAGATGTTCCCCTCTACAAGGATGGGGAGGCTAAGACTGACCCAGCAGGGCTGAGCAGTCCCCATCTGCCAGGCACATCCTCTGCAGCACCCGACCTGGAGGGTCCCGAATTCCCAGTTGAGTCTGTGGCTTCTCGGATCCAGGCTGAGCCAGACAACTTGGCACGTGCCTCTGCATCTCCAGACAGAATTCCTAGCCTGCCTCAGGAAACTGTGGATCAGGAACCCAAGGATCAGAAGAGGAAATCCTTTGAGCAGGCGGCCTCTGCATCCTTTCCCGAAAAGAAGCCCCGGCTTGAAGATCGTCAGTCCTTTCGTAACACAATTGAAAGTGTTCACACCGAAAAGCCACAGCCCACTAAAGAGGAGCCCAAAGTCCCGCCCATCCGGATTCAACTTTCACGTATCAAACCACCCTGGGTGGTTAAAGGTCAGCCCACTTACCAGATATGCCCCCGGATCATCCCCACCACGGAGTCCTCCTGCCGGGGTTGGACTGGCGCCAGGACCCTCGCAGACATTAAAGCCCGTGCTCTGCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCCACCACTGCCATCGGAGGGGGGGGTGGCCCGGGTGGAGGTGGCGGCGGGGCCACCGATGAGGGAGGTGGCAGAGGCAGCAGCAGTGGTGATGGTGGTGAGGCCTGTGGCCACCCTGAGCCCAGGGGAGGCCCGAGCACCCCTGGAAAGTGTACGTCAGATCTACAGCGAACACAACTACTGCCGCCTTATCCTCTAAATGGGGAGCATACCCAGGCCGGAACTGCCATGTCCAGAGCTAGGAGAGAGGACCTGCCTTCTCTGAGAAAGGAGGAAAGCTGCCTACTACAGAGGGCTACAGTTGGACTCACAGATGGGCTAGGAGATGCCTCCCAACTCCCCGTTGCTCCCACTGGGGACCAGCCATGCCAGGCCTTGCCCCTACTGTCCTCCCAAACCTCAGTAGCTGAGAGATTAGTGGAGCAGCCTCAGTTGCATCCGGATGTTAGAACTGAATGTGAGTCTGGCACCACTTCCTGGGAAAGTGATGATGAGGAGCAAGGACCCACCGTTCCTGCAGACAATGGTCCCATTCTGTCTCTAGTGGGAGATGATACATTAGAGAAAGGAACTGGCCAAGCTCTTGACAGTCATCCCACTATGAAGGATCCTGTAAATGTGACCCCCAGTTCCACACCTGAATCCTCACCGACTGATTGCCTGCAGAACAGAGCATTTGATGACGAATTAGGGCTTGGTGGCTCATGCCCTCCTATGAGGGAAAGTGATACTAGACAAGAAAACTTGAAAACCAAGGCTCTCGTTTCTAACAGTTCTTTGCATTGGATACCCATCCCATCGAATGATGAGGTAGTGAAACAGCCCAAACCAGAATCCAGAGAACACATACCATCTGTTGAGCCCCAGGTTGGAGAGGAGTGGGAGAAAGCTGCTCCCACCCCTCCTGCATTGCCTGGGGATTTGACAGCTGAGGAGGGTCTAGATCCTCTTGACAGCCTTACTTCACTCTGGACTGTGCCATCTCGAGGAGGCAGTGACAGCAATGGCAGTTACTGTCAACAGGTGGACATTGAAAAGCTGAAAATCAACGGAGACTCTGAAGCACTGAGTCCTCACGGTGAGTCCACGGATACAGCCTCTGACTTTGAAGGTCACCTCACGGAGGACAGCAGTGAGGCTGACACTAGAGAAGCTGCAGTGACAAAGGGATCTTCGGTGGACAAGGATGAGAAACCCAATTGGAACCAATCTGCCCCACTGTCCAAGGTGAATGGTGACATGCGTCTGGTTACAAGGACAGATGGGATGGTTGCTCCTCAGAGCTGGGTGTCTCGAGTATGTGCGGTCCGCCAAAAGATCCCAGATTCCCTACTGCTGGCCAGTACTGAGTACCAGCCAAGAGCCGTGTGCCTGTCCATGCCTGGGTCCTCAGTGGAGGCCACTAACCCACTTGTGATGCAGTTGCTGCAGGGTAGCTTGCCCCTAGAGAAGGTTCTTCCACCAGCCCACGATGACAGCATGTCAGAATCCCCACAAGTACCACTTACAAAAGACCAGAGCCATGGCTCGCTACGCATGGGATCTTTACATGGTCTTGGAAAAAACAGTGGCATGGTTGATGGAAGCAGCCCCAGTTCTTTAAGGGCTTTGAAGGAGCCTCTTCTGCCAGATAGCTGTGAAACAGGCACTGGTCTTGCCAGGATTGAGGCCACCCAGGCTCCTGGAGCACCCCAAAAGAATTGCAAGGCAGTCCCAAGTTTTGACTCCCTCCATCCAGTGACAAATCCCATTACATCCTCTAGGAAACTGGAAGAAATGGATTCCAAAGAGCAGTTCTCTTCCTTTAGTTGTGAAGATCAGAAGGAAGTCCGTGCTATGTCACAGGACAGTAATTCAAATGCTGCTCCAGGAAAGAGCCCAGGAGATCTTACTACCTCGAGAACACCTCGTTTCTCATCTCCAAATGTGATCTCCTTTGGTCCAGAGCAGACAGGTCGGGCCCTGGGTGATCAGAGCAATGTTACAGGCCAAGGGAAGAAGCTTTTTGGCTCTGGGAATGTGGCTGCAACCCTTCAGCGCCCCAGGCCTGCGGACCCGATGCCTCTTCCTGCTGAGATCCCTCCAGTTTTTCCCAGTGGGAAGTTGGGACCAAGCACAAACTCCATGTCTGGTGGGGTACAGACTCCAAGGGAAGACTGGGCTCCAAAGCCACATGCCTTTGTTGGCAGCGTCAAGAATGAGAAGACTTTTGTGGGGGGTCCTCTTAAGGCAAATGCCGAGAACAGGAAAGCTACTGGGCATAGTCCCCTGGAACTGGTGGGTCACTTGGAAGGGATGCCCTTTGTCATGGACTTGCCCTTCTGGAAATTACCCCGAGAGCCAGGGAAGGGGCTCAGTGAGCCTCTGGAGCCTTCTTCTCTCCCCTCCCAACTCAGCATCAAGCAGGCATTTTATGGGAAGCTTTCTAAACTCCAACTGAGTTCCACCAGCTTTAATTATTCCTCTAGCTCTCCCACCTTTCCCAAAGGCCTTGCTGGAAGTGTGGTGCAGCTGAGCCACAAAGCAAACTTTGGTGCGAGCCACAGTGCATCACTTTCCTTGCAAATGTTCACTGACAGCAGCACGGTGGAAAGCATCTCGCTCCAGTGTGCGTGCAGCCTGAAAGCCATGATCATGTGCCAAGGCTGCGGTGCGTTCTGTCACGATGACTGTATTGGACCCTCAAAGCTCTGTGTATTGTGCCTTGTGGTGAGATAAT'\n",
    "  \n",
    "    #List of all amino acid positions (across whole coding region) covered by (amino acid position minus 1 (as python starts at 0))\n",
    "    ASXL1_nonsense_codons_DICT={}\n",
    "    for i in range(lower_range, upper_range):\n",
    "        ASXL1_nonsense_codons_DICT[i+1]= (ASXL1_codons[i])\n",
    "\n",
    "    calls_ASXL1_nonsense = {}\n",
    "    for k,v in ASXL1_nonsense_codons_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_ASXL1_nonsense[k]=(v, changes)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_ASXL1_nonsense_extra = {}\n",
    "\n",
    "    for k, v in calls_ASXL1_nonsense.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_ASXL1_nonsense_extra[k] = (Amino_acid_change, v[0], v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_ASXL1_nonsense_extra2 = {}\n",
    "\n",
    "    for k, v in calls_ASXL1_nonsense_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_ASXL1_nonsense_extra2[k] = (Amino_acid_abbrev, v[1], v[2])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary that contains details of the positions of nonsense variants at that codon\n",
    "    ASXL1_position_in_list_dict_nonsense = {}\n",
    "\n",
    "    for k, v in calls_ASXL1_nonsense_extra2.items():\n",
    "        original_AA = v[0]\n",
    "        possible_AA_changes = v[2]\n",
    "        possible_AA_changes_string = ''.join(v[2])\n",
    "        positions_in_list = []\n",
    "        for i in possible_AA_changes:\n",
    "            if i == '*': #nonsense\n",
    "                position_in_list=-1\n",
    "                nonsense_change = i    \n",
    "                while True:\n",
    "                    position_in_list = possible_AA_changes_string.find(nonsense_change, position_in_list+1)\n",
    "    #               print(position_in_list)\n",
    "                    if position_in_list == -1: break\n",
    "                    positions_in_list.append(position_in_list)\n",
    "                    positions_in_list = set(positions_in_list) #convert into a 'set' to remove duplicate positions\n",
    "                    positions_in_list = list(positions_in_list) #convert back to a list so can be added to dictionary\n",
    "                ASXL1_position_in_list_dict_nonsense[k]=(positions_in_list, v)\n",
    "\n",
    "        aa_positions_variants_called = []\n",
    "        for k, v in ASXL1_position_in_list_dict_nonsense.items():\n",
    "            aa_positions_variants_called.append(k)\n",
    "    # print('positions at which nonsense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    ASXL1_bases_nonsense = {}\n",
    "    for k, v in ASXL1_position_in_list_dict_nonsense.items():\n",
    "        original_a_a = v[1][0]\n",
    "        codon_seq = v[1][1]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        ASXL1_bases_nonsense[k]=(v, base_changes, original_a_a)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    ASXL1_bases_nonsense2 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_nonsense.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        ASXL1_bases_nonsense2[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    ASXL1_bases_nonsense3 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_nonsense2.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_a_a = v[0][2]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = k\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(k)-1)*3)+codon_pos #codon number minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][1][1][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        ASXL1_bases_nonsense3[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, original_a_a)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    ASXL1_bases_nonsense4 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_nonsense3.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_nonsense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check(i)\n",
    "                sites_trio_nonsense.append([adapted_trio])\n",
    "                ASXL1_bases_nonsense4[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_nonsense = [ASXL1_exons_extra_bases[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_nonsense.append(site_trio_nonsense)\n",
    "                    ASXL1_bases_nonsense4[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    ASXL1_sites_nonsense5 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in ASXL1_bases_nonsense4.items():\n",
    "        original_a_a = v[0][0][0]\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, original_a_a])\n",
    "            change = change+1\n",
    "        ASXL1_sites_nonsense5[a]=info\n",
    "        a=a+1\n",
    "    #     print(ASXL1_sites_nonsense5)\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    ASXL1_sites_nonsense6={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in ASXL1_sites_nonsense5.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        original_a_a = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        ASXL1_sites_nonsense6[k] = (v[0],changes3, codon_n_912, original_a_a) #site_trio = the affected base and its neighbours\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    ASXL1_mutation_rates_nonsense = {}\n",
    "    for k, v in ASXL1_sites_nonsense6.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            original_a_a = v[3]\n",
    "            aa_positioning = v[2]\n",
    "            variant_type = '*'\n",
    "            variant_id = str(original_a_a)+str(aa_positioning)+variant_type\n",
    "            ASXL1_mutation_rates_nonsense[k] = (v[0],v[1], mutation_rate, variant_id)\n",
    "\n",
    "    ASXL1_mutation_rate_nonsense_sum = []\n",
    "\n",
    "    mutation_rates_list = []\n",
    "    for k, v in ASXL1_mutation_rates_nonsense.items():\n",
    "        codon_position_in_gene=(v[0])\n",
    "        variant = v[3]\n",
    "        site = v[1]\n",
    "        mutation_rate_value = v[2] \n",
    "        mutation_rates_list.append((v[0], variant, site, mutation_rate_value))\n",
    "        ASXL1_mutation_rate_nonsense_sum.append(mutation_rate_value)\n",
    "\n",
    "    #Get the average of all the mutation rates\n",
    "    ASXL1_mutation_rate_nonsense_sum = []\n",
    "    for k, v in ASXL1_mutation_rates_nonsense.items():\n",
    "        mut_rate = v[2]\n",
    "        ASXL1_mutation_rate_nonsense_sum.append(mut_rate)\n",
    "\n",
    "    ASXL1_Summed_mutation_rate_nonsense_added = np.sum(ASXL1_mutation_rate_nonsense_sum)\n",
    "\n",
    "    print('Total summed mutation rate across ASXL1 nonsense for this study = ', ASXL1_Summed_mutation_rate_nonsense_added)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all # e.g. lower= 0, upper = 913 for whole gene\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_missense_range_of_variants_ASXL1(lower_range, upper_range):\n",
    "\n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base = [140, 143, 252, 373, 1086, 1085]\n",
    "   \n",
    "    def intron_exon_check(position_in_gene_number):\n",
    "        if position_in_gene_number == 140:\n",
    "            gene_pos = 140\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 143:\n",
    "            gene_pos = 143\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 252:\n",
    "            gene_pos = 252\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 373:\n",
    "            gene_pos = 373\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1086:\n",
    "            gene_pos = 1086\n",
    "            first_base = 'G'\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = ASXL1_exons_extra_bases[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1085:\n",
    "            gene_pos = 1085\n",
    "            first_base = ASXL1_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = ASXL1_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #ASXL1 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    ASXL1_exons = 'ATGAAGGACAAACAGAAGAAGAAGAAGGAGCGCACGTGGGCCGAGGCCGCGCGCCTGGTATTAGAAAACTACTCGGATGCTCCAATGACACCAAAACAGATTCTGCAGGTCATAGAGGCAGAAGGACTAAAGGAAATGAGAAGTGGGACTTCCCCTCTCGCATGCCTCAATGCTATGCTACATTCCAATTCAAGAGGAGGAGAGGGGTTGTTTTATAAACTGCCTGGCCGAATCAGCCTTTTCACGCTCAAGAAGGATGCCCTGCAGTGGTCTCGCCATCCAGCTACAGTGGAGGGAGAGGAGCCAGAGGACACGGCTGATGTGGAGAGCTGTGGGTCTAATGAAGCCAGCACTGTGAGTGGTGAAAACGATGTATCTCTTGATGAAACATCTTCGAACGCATCCTGTTCTACAGAATCTCAGAGTCGACCTCTTTCCAATCCCAGGGACAGCTACAGAGCTTCCTCACAGGCGAACAAACAAAAGAAAAAGACTGGGGTGATGCTGCCTCGAGTTGTCCTGACTCCTCTGAAGGTAAACGGGGCCCACGTGGAATCTGCATCAGGGTTCTCGGGCTGCCACGCCGATGGCGAGAGCGGCAGCCCGTCCAGCAGCAGCAGCGGCTCTCTGGCCCTGGGCAGCGCTGCTATTCGTGGCCAGGCCGAGGTCACCCAGGACCCTGCCCCGCTCCTGAGAGGCTTCCGGAAGCCAGCCACAGGTCAAATGAAGCGCAACAGAGGGGAAGAAATAGATTTTGAGACACCTGGGTCCATTCTTGTCAACACCAACCTCCGTGCCCTGATCAACTCTCGGACCTTCCATGCCTTACCATCACACTTCCAGCAGCAGCTCCTCTTCCTCCTGCCTGAAGTAGACAGACAGGTGGGGACGGATGGCCTGTTGCGTCTCAGCAGCAGTGCACTAAATAACGAGTTTTTTACCCATGCGGCTCAGAGCTGGCGGGAGCGCCTGGCTGATGGTGAATTTACTCATGAGATGCAAGTCAGGATACGACAGGAAATGGAGAAGGAAAAGAAGGTGGAACAATGGAAAGAAAAGTTCTTTGAAGACTACTATGGACAGAAGCTGGGTTTGACCAAAGAAGAGTCATTGCAGCAGAACGTGGGCCAGGAGGAGGCTGAAATCAAAAGTGGCTTGTGTGTCCCAGGAGAATCAGTGCGTATACAGCGTGGTCCAGCCACCCGACAGCGAGATGGGCATTTTAAGAAACGCTCTCGGCCAGATCTCCGAACCAGAGCCAGAAGGAATCTGTACAAAAAACAGGAGTCAGAACAAGCAGGGGTTGCTAAGGATGCAAAATCTGTGGCCTCAGATGTTCCCCTCTACAAGGATGGGGAGGCTAAGACTGACCCAGCAGGGCTGAGCAGTCCCCATCTGCCAGGCACATCCTCTGCAGCACCCGACCTGGAGGGTCCCGAATTCCCAGTTGAGTCTGTGGCTTCTCGGATCCAGGCTGAGCCAGACAACTTGGCACGTGCCTCTGCATCTCCAGACAGAATTCCTAGCCTGCCTCAGGAAACTGTGGATCAGGAACCCAAGGATCAGAAGAGGAAATCCTTTGAGCAGGCGGCCTCTGCATCCTTTCCCGAAAAGAAGCCCCGGCTTGAAGATCGTCAGTCCTTTCGTAACACAATTGAAAGTGTTCACACCGAAAAGCCACAGCCCACTAAAGAGGAGCCCAAAGTCCCGCCCATCCGGATTCAACTTTCACGTATCAAACCACCCTGGGTGGTTAAAGGTCAGCCCACTTACCAGATATGCCCCCGGATCATCCCCACCACGGAGTCCTCCTGCCGGGGTTGGACTGGCGCCAGGACCCTCGCAGACATTAAAGCCCGTGCTCTGCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCCACCACTGCCATCGGAGGGGGGGGTGGCCCGGGTGGAGGTGGCGGCGGGGCCACCGATGAGGGAGGTGGCAGAGGCAGCAGCAGTGGTGATGGTGGTGAGGCCTGTGGCCACCCTGAGCCCAGGGGAGGCCCGAGCACCCCTGGAAAGTGTACGTCAGATCTACAGCGAACACAACTACTGCCGCCTTATCCTCTAAATGGGGAGCATACCCAGGCCGGAACTGCCATGTCCAGAGCTAGGAGAGAGGACCTGCCTTCTCTGAGAAAGGAGGAAAGCTGCCTACTACAGAGGGCTACAGTTGGACTCACAGATGGGCTAGGAGATGCCTCCCAACTCCCCGTTGCTCCCACTGGGGACCAGCCATGCCAGGCCTTGCCCCTACTGTCCTCCCAAACCTCAGTAGCTGAGAGATTAGTGGAGCAGCCTCAGTTGCATCCGGATGTTAGAACTGAATGTGAGTCTGGCACCACTTCCTGGGAAAGTGATGATGAGGAGCAAGGACCCACCGTTCCTGCAGACAATGGTCCCATTCTGTCTCTAGTGGGAGATGATACATTAGAGAAAGGAACTGGCCAAGCTCTTGACAGTCATCCCACTATGAAGGATCCTGTAAATGTGACCCCCAGTTCCACACCTGAATCCTCACCGACTGATTGCCTGCAGAACAGAGCATTTGATGACGAATTAGGGCTTGGTGGCTCATGCCCTCCTATGAGGGAAAGTGATACTAGACAAGAAAACTTGAAAACCAAGGCTCTCGTTTCTAACAGTTCTTTGCATTGGATACCCATCCCATCGAATGATGAGGTAGTGAAACAGCCCAAACCAGAATCCAGAGAACACATACCATCTGTTGAGCCCCAGGTTGGAGAGGAGTGGGAGAAAGCTGCTCCCACCCCTCCTGCATTGCCTGGGGATTTGACAGCTGAGGAGGGTCTAGATCCTCTTGACAGCCTTACTTCACTCTGGACTGTGCCATCTCGAGGAGGCAGTGACAGCAATGGCAGTTACTGTCAACAGGTGGACATTGAAAAGCTGAAAATCAACGGAGACTCTGAAGCACTGAGTCCTCACGGTGAGTCCACGGATACAGCCTCTGACTTTGAAGGTCACCTCACGGAGGACAGCAGTGAGGCTGACACTAGAGAAGCTGCAGTGACAAAGGGATCTTCGGTGGACAAGGATGAGAAACCCAATTGGAACCAATCTGCCCCACTGTCCAAGGTGAATGGTGACATGCGTCTGGTTACAAGGACAGATGGGATGGTTGCTCCTCAGAGCTGGGTGTCTCGAGTATGTGCGGTCCGCCAAAAGATCCCAGATTCCCTACTGCTGGCCAGTACTGAGTACCAGCCAAGAGCCGTGTGCCTGTCCATGCCTGGGTCCTCAGTGGAGGCCACTAACCCACTTGTGATGCAGTTGCTGCAGGGTAGCTTGCCCCTAGAGAAGGTTCTTCCACCAGCCCACGATGACAGCATGTCAGAATCCCCACAAGTACCACTTACAAAAGACCAGAGCCATGGCTCGCTACGCATGGGATCTTTACATGGTCTTGGAAAAAACAGTGGCATGGTTGATGGAAGCAGCCCCAGTTCTTTAAGGGCTTTGAAGGAGCCTCTTCTGCCAGATAGCTGTGAAACAGGCACTGGTCTTGCCAGGATTGAGGCCACCCAGGCTCCTGGAGCACCCCAAAAGAATTGCAAGGCAGTCCCAAGTTTTGACTCCCTCCATCCAGTGACAAATCCCATTACATCCTCTAGGAAACTGGAAGAAATGGATTCCAAAGAGCAGTTCTCTTCCTTTAGTTGTGAAGATCAGAAGGAAGTCCGTGCTATGTCACAGGACAGTAATTCAAATGCTGCTCCAGGAAAGAGCCCAGGAGATCTTACTACCTCGAGAACACCTCGTTTCTCATCTCCAAATGTGATCTCCTTTGGTCCAGAGCAGACAGGTCGGGCCCTGGGTGATCAGAGCAATGTTACAGGCCAAGGGAAGAAGCTTTTTGGCTCTGGGAATGTGGCTGCAACCCTTCAGCGCCCCAGGCCTGCGGACCCGATGCCTCTTCCTGCTGAGATCCCTCCAGTTTTTCCCAGTGGGAAGTTGGGACCAAGCACAAACTCCATGTCTGGTGGGGTACAGACTCCAAGGGAAGACTGGGCTCCAAAGCCACATGCCTTTGTTGGCAGCGTCAAGAATGAGAAGACTTTTGTGGGGGGTCCTCTTAAGGCAAATGCCGAGAACAGGAAAGCTACTGGGCATAGTCCCCTGGAACTGGTGGGTCACTTGGAAGGGATGCCCTTTGTCATGGACTTGCCCTTCTGGAAATTACCCCGAGAGCCAGGGAAGGGGCTCAGTGAGCCTCTGGAGCCTTCTTCTCTCCCCTCCCAACTCAGCATCAAGCAGGCATTTTATGGGAAGCTTTCTAAACTCCAACTGAGTTCCACCAGCTTTAATTATTCCTCTAGCTCTCCCACCTTTCCCAAAGGCCTTGCTGGAAGTGTGGTGCAGCTGAGCCACAAAGCAAACTTTGGTGCGAGCCACAGTGCATCACTTTCCTTGCAAATGTTCACTGACAGCAGCACGGTGGAAAGCATCTCGCTCCAGTGTGCGTGCAGCCTGAAAGCCATGATCATGTGCCAAGGCTGCGGTGCGTTCTGTCACGATGACTGTATTGGACCCTCAAAGCTCTGTGTATTGTGCCTTGTGGTGAGATAA'\n",
    "    \n",
    "    #Split the sequence into codons \n",
    "    ASXL1_codons = [ASXL1_exons[i:i+3] for i in range(0, len(ASXL1_exons), 3)]\n",
    "\n",
    "    #ASXL1 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    ASXL1_exons_extra_bases = 'GATGAAGGACAAACAGAAGAAGAAGAAGGAGCGCACGTGGGCCGAGGCCGCGCGCCTGGTATTAGAAAACTACTCGGATGCTCCAATGACACCAAAACAGATTCTGCAGGTCATAGAGGCAGAAGGACTAAAGGAAATGAGAAGTGGGACTTCCCCTCTCGCATGCCTCAATGCTATGCTACATTCCAATTCAAGAGGAGGAGAGGGGTTGTTTTATAAACTGCCTGGCCGAATCAGCCTTTTCACGCTCAAGAAGGATGCCCTGCAGTGGTCTCGCCATCCAGCTACAGTGGAGGGAGAGGAGCCAGAGGACACGGCTGATGTGGAGAGCTGTGGGTCTAATGAAGCCAGCACTGTGAGTGGTGAAAACGATGTATCTCTTGATGAAACATCTTCGAACGCATCCTGTTCTACAGAATCTCAGAGTCGACCTCTTTCCAATCCCAGGGACAGCTACAGAGCTTCCTCACAGGCGAACAAACAAAAGAAAAAGACTGGGGTGATGCTGCCTCGAGTTGTCCTGACTCCTCTGAAGGTAAACGGGGCCCACGTGGAATCTGCATCAGGGTTCTCGGGCTGCCACGCCGATGGCGAGAGCGGCAGCCCGTCCAGCAGCAGCAGCGGCTCTCTGGCCCTGGGCAGCGCTGCTATTCGTGGCCAGGCCGAGGTCACCCAGGACCCTGCCCCGCTCCTGAGAGGCTTCCGGAAGCCAGCCACAGGTCAAATGAAGCGCAACAGAGGGGAAGAAATAGATTTTGAGACACCTGGGTCCATTCTTGTCAACACCAACCTCCGTGCCCTGATCAACTCTCGGACCTTCCATGCCTTACCATCACACTTCCAGCAGCAGCTCCTCTTCCTCCTGCCTGAAGTAGACAGACAGGTGGGGACGGATGGCCTGTTGCGTCTCAGCAGCAGTGCACTAAATAACGAGTTTTTTACCCATGCGGCTCAGAGCTGGCGGGAGCGCCTGGCTGATGGTGAATTTACTCATGAGATGCAAGTCAGGATACGACAGGAAATGGAGAAGGAAAAGAAGGTGGAACAATGGAAAGAAAAGTTCTTTGAAGACTACTATGGACAGAAGCTGGGTTTGACCAAAGAAGAGTCATTGCAGCAGAACGTGGGCCAGGAGGAGGCTGAAATCAAAAGTGGCTTGTGTGTCCCAGGAGAATCAGTGCGTATACAGCGTGGTCCAGCCACCCGACAGCGAGATGGGCATTTTAAGAAACGCTCTCGGCCAGATCTCCGAACCAGAGCCAGAAGGAATCTGTACAAAAAACAGGAGTCAGAACAAGCAGGGGTTGCTAAGGATGCAAAATCTGTGGCCTCAGATGTTCCCCTCTACAAGGATGGGGAGGCTAAGACTGACCCAGCAGGGCTGAGCAGTCCCCATCTGCCAGGCACATCCTCTGCAGCACCCGACCTGGAGGGTCCCGAATTCCCAGTTGAGTCTGTGGCTTCTCGGATCCAGGCTGAGCCAGACAACTTGGCACGTGCCTCTGCATCTCCAGACAGAATTCCTAGCCTGCCTCAGGAAACTGTGGATCAGGAACCCAAGGATCAGAAGAGGAAATCCTTTGAGCAGGCGGCCTCTGCATCCTTTCCCGAAAAGAAGCCCCGGCTTGAAGATCGTCAGTCCTTTCGTAACACAATTGAAAGTGTTCACACCGAAAAGCCACAGCCCACTAAAGAGGAGCCCAAAGTCCCGCCCATCCGGATTCAACTTTCACGTATCAAACCACCCTGGGTGGTTAAAGGTCAGCCCACTTACCAGATATGCCCCCGGATCATCCCCACCACGGAGTCCTCCTGCCGGGGTTGGACTGGCGCCAGGACCCTCGCAGACATTAAAGCCCGTGCTCTGCAGGTCCGAGGGGCGAGAGGTCACCACTGCCATAGAGAGGCGGCCACCACTGCCATCGGAGGGGGGGGTGGCCCGGGTGGAGGTGGCGGCGGGGCCACCGATGAGGGAGGTGGCAGAGGCAGCAGCAGTGGTGATGGTGGTGAGGCCTGTGGCCACCCTGAGCCCAGGGGAGGCCCGAGCACCCCTGGAAAGTGTACGTCAGATCTACAGCGAACACAACTACTGCCGCCTTATCCTCTAAATGGGGAGCATACCCAGGCCGGAACTGCCATGTCCAGAGCTAGGAGAGAGGACCTGCCTTCTCTGAGAAAGGAGGAAAGCTGCCTACTACAGAGGGCTACAGTTGGACTCACAGATGGGCTAGGAGATGCCTCCCAACTCCCCGTTGCTCCCACTGGGGACCAGCCATGCCAGGCCTTGCCCCTACTGTCCTCCCAAACCTCAGTAGCTGAGAGATTAGTGGAGCAGCCTCAGTTGCATCCGGATGTTAGAACTGAATGTGAGTCTGGCACCACTTCCTGGGAAAGTGATGATGAGGAGCAAGGACCCACCGTTCCTGCAGACAATGGTCCCATTCTGTCTCTAGTGGGAGATGATACATTAGAGAAAGGAACTGGCCAAGCTCTTGACAGTCATCCCACTATGAAGGATCCTGTAAATGTGACCCCCAGTTCCACACCTGAATCCTCACCGACTGATTGCCTGCAGAACAGAGCATTTGATGACGAATTAGGGCTTGGTGGCTCATGCCCTCCTATGAGGGAAAGTGATACTAGACAAGAAAACTTGAAAACCAAGGCTCTCGTTTCTAACAGTTCTTTGCATTGGATACCCATCCCATCGAATGATGAGGTAGTGAAACAGCCCAAACCAGAATCCAGAGAACACATACCATCTGTTGAGCCCCAGGTTGGAGAGGAGTGGGAGAAAGCTGCTCCCACCCCTCCTGCATTGCCTGGGGATTTGACAGCTGAGGAGGGTCTAGATCCTCTTGACAGCCTTACTTCACTCTGGACTGTGCCATCTCGAGGAGGCAGTGACAGCAATGGCAGTTACTGTCAACAGGTGGACATTGAAAAGCTGAAAATCAACGGAGACTCTGAAGCACTGAGTCCTCACGGTGAGTCCACGGATACAGCCTCTGACTTTGAAGGTCACCTCACGGAGGACAGCAGTGAGGCTGACACTAGAGAAGCTGCAGTGACAAAGGGATCTTCGGTGGACAAGGATGAGAAACCCAATTGGAACCAATCTGCCCCACTGTCCAAGGTGAATGGTGACATGCGTCTGGTTACAAGGACAGATGGGATGGTTGCTCCTCAGAGCTGGGTGTCTCGAGTATGTGCGGTCCGCCAAAAGATCCCAGATTCCCTACTGCTGGCCAGTACTGAGTACCAGCCAAGAGCCGTGTGCCTGTCCATGCCTGGGTCCTCAGTGGAGGCCACTAACCCACTTGTGATGCAGTTGCTGCAGGGTAGCTTGCCCCTAGAGAAGGTTCTTCCACCAGCCCACGATGACAGCATGTCAGAATCCCCACAAGTACCACTTACAAAAGACCAGAGCCATGGCTCGCTACGCATGGGATCTTTACATGGTCTTGGAAAAAACAGTGGCATGGTTGATGGAAGCAGCCCCAGTTCTTTAAGGGCTTTGAAGGAGCCTCTTCTGCCAGATAGCTGTGAAACAGGCACTGGTCTTGCCAGGATTGAGGCCACCCAGGCTCCTGGAGCACCCCAAAAGAATTGCAAGGCAGTCCCAAGTTTTGACTCCCTCCATCCAGTGACAAATCCCATTACATCCTCTAGGAAACTGGAAGAAATGGATTCCAAAGAGCAGTTCTCTTCCTTTAGTTGTGAAGATCAGAAGGAAGTCCGTGCTATGTCACAGGACAGTAATTCAAATGCTGCTCCAGGAAAGAGCCCAGGAGATCTTACTACCTCGAGAACACCTCGTTTCTCATCTCCAAATGTGATCTCCTTTGGTCCAGAGCAGACAGGTCGGGCCCTGGGTGATCAGAGCAATGTTACAGGCCAAGGGAAGAAGCTTTTTGGCTCTGGGAATGTGGCTGCAACCCTTCAGCGCCCCAGGCCTGCGGACCCGATGCCTCTTCCTGCTGAGATCCCTCCAGTTTTTCCCAGTGGGAAGTTGGGACCAAGCACAAACTCCATGTCTGGTGGGGTACAGACTCCAAGGGAAGACTGGGCTCCAAAGCCACATGCCTTTGTTGGCAGCGTCAAGAATGAGAAGACTTTTGTGGGGGGTCCTCTTAAGGCAAATGCCGAGAACAGGAAAGCTACTGGGCATAGTCCCCTGGAACTGGTGGGTCACTTGGAAGGGATGCCCTTTGTCATGGACTTGCCCTTCTGGAAATTACCCCGAGAGCCAGGGAAGGGGCTCAGTGAGCCTCTGGAGCCTTCTTCTCTCCCCTCCCAACTCAGCATCAAGCAGGCATTTTATGGGAAGCTTTCTAAACTCCAACTGAGTTCCACCAGCTTTAATTATTCCTCTAGCTCTCCCACCTTTCCCAAAGGCCTTGCTGGAAGTGTGGTGCAGCTGAGCCACAAAGCAAACTTTGGTGCGAGCCACAGTGCATCACTTTCCTTGCAAATGTTCACTGACAGCAGCACGGTGGAAAGCATCTCGCTCCAGTGTGCGTGCAGCCTGAAAGCCATGATCATGTGCCAAGGCTGCGGTGCGTTCTGTCACGATGACTGTATTGGACCCTCAAAGCTCTGTGTATTGTGCCTTGTGGTGAGATAAT'\n",
    "  \n",
    "    #List of all amino acid positions (across whole coding region) covered (amino acid position minus 1 (as python starts at 0))\n",
    "    ASXL1_missense_codons_DICT={}\n",
    "    for i in range(lower_range, upper_range):\n",
    "        ASXL1_missense_codons_DICT[i+1]= (ASXL1_codons[i])\n",
    "\n",
    "    calls_ASXL1_missense = {}\n",
    "    for k,v in ASXL1_missense_codons_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_ASXL1_missense[k]=(v, changes)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_ASXL1_missense_extra = {}\n",
    "\n",
    "    for k, v in calls_ASXL1_missense.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_ASXL1_missense_extra[k] = (Amino_acid_change, v[0], v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_ASXL1_missense_extra2 = {}\n",
    "\n",
    "    for k, v in calls_ASXL1_missense_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_ASXL1_missense_extra2[k] = (Amino_acid_abbrev, v[1], v[2])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary that contains details of the positions of missense variants at that codon\n",
    "    ASXL1_position_in_list_dict_missense = {}\n",
    "\n",
    "    for k, v in calls_ASXL1_missense_extra2.items():\n",
    "        original_AA = v[0]\n",
    "        possible_AA_changes = v[2]\n",
    "        possible_AA_changes_string = ''.join(v[2])\n",
    "        positions_in_list = []\n",
    "        for i in possible_AA_changes:\n",
    "            if i != original_AA: #not synonymous\n",
    "                if i != '*': #not missense\n",
    "                    position_in_list=-1\n",
    "                    missense_change = i\n",
    "                    new_aa_list = []\n",
    "                    while True:\n",
    "                        position_in_list = possible_AA_changes_string.find(missense_change, position_in_list+1)\n",
    "        #               print(position_in_list)\n",
    "                        if position_in_list == -1: break\n",
    "                        positions_in_list.append(position_in_list)\n",
    "                        positions_in_list = set(positions_in_list) #convert into a 'set' to remove duplicate positions\n",
    "                        positions_in_list = list(positions_in_list) #convert back to a list so can be added to dictionary\n",
    "                    for pos in positions_in_list:\n",
    "                        new_aa = possible_AA_changes_string[pos]\n",
    "                        new_aa_list.append(new_aa)\n",
    "                    ASXL1_position_in_list_dict_missense[k]=(positions_in_list, v, new_aa_list)\n",
    "\n",
    "        aa_positions_variants_called = []\n",
    "        for k, v in ASXL1_position_in_list_dict_missense.items():\n",
    "            aa_positions_variants_called.append(k)\n",
    "    # print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    ASXL1_bases_missense = {}\n",
    "    for k, v in ASXL1_position_in_list_dict_missense.items():\n",
    "        original_a_a = v[1][0]\n",
    "        codon_seq = v[1][1]\n",
    "        base_changes=[]\n",
    "        new_a_a_list = v[2]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        ASXL1_bases_missense[k]=(v, base_changes, original_a_a, new_a_a_list)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    ASXL1_bases_missense2 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_missense.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        new_a_a_list = v[3]\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        ASXL1_bases_missense2[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    ASXL1_bases_missense3 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_missense2.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_a_a = v[0][2]\n",
    "        new_aa_list = v[0][0][2]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = k\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(k)-1)*3)+codon_pos #codon number minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][1][1][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        ASXL1_bases_missense3[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, original_a_a, new_aa_list)\n",
    "\n",
    "    # Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    ASXL1_bases_missense4 = {}\n",
    "\n",
    "    for k, v in ASXL1_bases_missense3.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "        new_a_a_list = v[6]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check(i)\n",
    "                sites_trio_missense.append([adapted_trio])\n",
    "                ASXL1_bases_missense4[k] = (v, sites_trio_missense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_missense = [ASXL1_exons_extra_bases[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_missense.append(site_trio_missense)\n",
    "                    ASXL1_bases_missense4[k] = (v, sites_trio_missense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    ASXL1_sites_missense5 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in ASXL1_bases_missense4.items():\n",
    "        original_a_a = v[0][0][0]\n",
    "        new_a_a = v[0][6]\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            new_amino_acid = new_a_a[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, original_a_a, new_amino_acid])\n",
    "            change = change+1\n",
    "        ASXL1_sites_missense5[a]=info\n",
    "        a=a+1\n",
    "    #     print(ASXL1_sites_missense5)\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    ASXL1_sites_missense6={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in ASXL1_sites_missense5.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        original_a_a = v[5]\n",
    "        new_a_a = v[6]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        ASXL1_sites_missense6[k] = (v[0],changes3, codon_n_912, original_a_a, new_a_a) #site_trio = the affected base and its neighbours\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    ASXL1_mutation_rates_missense = {}\n",
    "    for k, v in ASXL1_sites_missense6.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            original_a_a = v[3]\n",
    "            new_a_a = v[4]\n",
    "            aa_positioning = v[2]\n",
    "            variant_id = str(original_a_a)+str(aa_positioning)+str(new_a_a)\n",
    "            ASXL1_mutation_rates_missense[k] = (v[0],v[1], mutation_rate, variant_id)\n",
    "\n",
    "    ASXL1_mutation_rate_missense_sum = []\n",
    "\n",
    "    mutation_rates_list = []\n",
    "    for k, v in ASXL1_mutation_rates_missense.items():\n",
    "        codon_position_in_gene=(v[0])\n",
    "        variant = v[3]\n",
    "        site = v[1]\n",
    "        mutation_rate_value = v[2] \n",
    "        mutation_rates_list.append((v[0], variant, site, mutation_rate_value))\n",
    "        ASXL1_mutation_rate_missense_sum.append(mutation_rate_value)\n",
    "\n",
    "    #Get the average of all the mutation rates\n",
    "    ASXL1_mutation_rate_missense_sum = []\n",
    "    for k, v in ASXL1_mutation_rates_missense.items():\n",
    "        mut_rate = v[2]\n",
    "        ASXL1_mutation_rate_missense_sum.append(mut_rate)\n",
    "\n",
    "    ASXL1_Summed_mutation_rate_missense_added = np.sum(ASXL1_mutation_rate_missense_sum)\n",
    "\n",
    "    print('Total summed mutation rate across ASXL1 missense for this study = ', ASXL1_Summed_mutation_rate_missense_added)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all # e.g. lower= 0, upper = 913 for whole gene\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TP53"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_nonsense_range_of_variants_TP53(lower_range, upper_range):\n",
    "    \n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base_TP53 = [74, 75, 96, 375, 672, 782, 919, 993, 1100]\n",
    "   \n",
    "    def intron_exon_check(position_in_gene_number):\n",
    "        if position_in_gene_number == 74:\n",
    "            gene_pos = 74\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 75:\n",
    "            gene_pos = 75\n",
    "            first_base = 'G'\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = TP53_exons_extra_bases[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 96:\n",
    "            gene_pos = 96\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 375:\n",
    "            gene_pos = 375\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 672:\n",
    "            gene_pos = 672\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 782:\n",
    "            gene_pos = 782\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 919:\n",
    "            gene_pos = 919\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 993:\n",
    "            gene_pos = 993\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1100:\n",
    "            gene_pos = 1100\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #TP53 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    TP53_exons = 'ATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGACATAGTGTGGTGGTGCCCTATGAGCCGCCTGAGGTTGGCTCTGACTGTACCACCATCCACTACAACTACATGTGTAACAGTTCCTGCATGGGCGGCATGAACCGGAGGCCCATCCTCACCATCATCACACTGGAAGACTCCAGTGGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGAGGAAGAGAATCTCCGCAAGAAAGGGGAGCCTCACCACGAGCTGCCCCCAGGGAGCACTAAGCGAGCACTGCCCAACAACACCAGCTCCTCTCCCCAGCCAAAGAAGAAACCACTGGATGGAGAATATTTCACCCTTCAGATCCGTGGGCGTGAGCGCTTCGAGATGTTCCGAGAGCTGAATGAGGCCTTGGAACTCAAGGATGCCCAGGCTGGGAAGGAGCCAGGGGGGAGCAGGGCTCACTCCAGCCACCTGAAGTCCAAAAAGGGTCAGTCTACCTCCCGCCATAAAAAACTCATGTTCAAGACAGAAGGGCCTGACTCAGACTGA'\n",
    "    \n",
    "    #Split the sequence into codons \n",
    "    TP53_codons = [TP53_exons[i:i+3] for i in range(0, len(TP53_exons), 3)]\n",
    "\n",
    "    #TP53 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    TP53_exons_extra_bases = 'CATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGACATAGTGTGGTGGTGCCCTATGAGCCGCCTGAGGTTGGCTCTGACTGTACCACCATCCACTACAACTACATGTGTAACAGTTCCTGCATGGGCGGCATGAACCGGAGGCCCATCCTCACCATCATCACACTGGAAGACTCCAGTGGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGAGGAAGAGAATCTCCGCAAGAAAGGGGAGCCTCACCACGAGCTGCCCCCAGGGAGCACTAAGCGAGCACTGCCCAACAACACCAGCTCCTCTCCCCAGCCAAAGAAGAAACCACTGGATGGAGAATATTTCACCCTTCAGATCCGTGGGCGTGAGCGCTTCGAGATGTTCCGAGAGCTGAATGAGGCCTTGGAACTCAAGGATGCCCAGGCTGGGAAGGAGCCAGGGGGGAGCAGGGCTCACTCCAGCCACCTGAAGTCCAAAAAGGGTCAGTCTACCTCCCGCCATAAAAAACTCATGTTCAAGACAGAAGGGCCTGACTCAGACTGAC'\n",
    "    \n",
    "    #List of all amino acid positions (across whole coding region) covered (amino acid position minus 1 (as python starts at 0))\n",
    "    TP53_nonsense_codons_DICT={}\n",
    "    for i in range(lower_range, upper_range):\n",
    "        TP53_nonsense_codons_DICT[i+1]= (TP53_codons[i])\n",
    "\n",
    "    calls_TP53_nonsense = {}\n",
    "    for k,v in TP53_nonsense_codons_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_TP53_nonsense[k]=(v, changes)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_TP53_nonsense_extra = {}\n",
    "\n",
    "    for k, v in calls_TP53_nonsense.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_TP53_nonsense_extra[k] = (Amino_acid_change, v[0], v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_TP53_nonsense_extra2 = {}\n",
    "\n",
    "    for k, v in calls_TP53_nonsense_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_TP53_nonsense_extra2[k] = (Amino_acid_abbrev, v[1], v[2])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary that contains details of the positions of nonsense variants at that codon\n",
    "    TP53_position_in_list_dict_nonsense = {}\n",
    "\n",
    "    for k, v in calls_TP53_nonsense_extra2.items():\n",
    "        original_AA = v[0]\n",
    "        possible_AA_changes = v[2]\n",
    "        possible_AA_changes_string = ''.join(v[2])\n",
    "        positions_in_list = []\n",
    "        for i in possible_AA_changes:\n",
    "            if i == '*': #nonsense\n",
    "                position_in_list=-1\n",
    "                nonsense_change = i    \n",
    "                while True:\n",
    "                    position_in_list = possible_AA_changes_string.find(nonsense_change, position_in_list+1)\n",
    "    #               print(position_in_list)\n",
    "                    if position_in_list == -1: break\n",
    "                    positions_in_list.append(position_in_list)\n",
    "                    positions_in_list = set(positions_in_list) #convert into a 'set' to remove duplicate positions\n",
    "                    positions_in_list = list(positions_in_list) #convert back to a list so can be added to dictionary\n",
    "                TP53_position_in_list_dict_nonsense[k]=(positions_in_list, v)\n",
    "\n",
    "        aa_positions_variants_called = []\n",
    "        for k, v in TP53_position_in_list_dict_nonsense.items():\n",
    "            aa_positions_variants_called.append(k)\n",
    "    # print('positions at which nonsense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TP53_bases_nonsense = {}\n",
    "    for k, v in TP53_position_in_list_dict_nonsense.items():\n",
    "        original_a_a = v[1][0]\n",
    "        codon_seq = v[1][1]\n",
    "        base_changes=[]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TP53_bases_nonsense[k]=(v, base_changes, original_a_a)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TP53_bases_nonsense2 = {}\n",
    "\n",
    "    for k, v in TP53_bases_nonsense.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TP53_bases_nonsense2[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TP53_bases_nonsense3 = {}\n",
    "\n",
    "    for k, v in TP53_bases_nonsense2.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_a_a = v[0][2]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = k\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(k)-1)*3)+codon_pos #codon number minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][1][1][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TP53_bases_nonsense3[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, original_a_a)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TP53_bases_nonsense4 = {}\n",
    "\n",
    "    for k, v in TP53_bases_nonsense3.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_nonsense=[]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_TP53:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check_TP53(i)\n",
    "                sites_trio_nonsense.append([adapted_trio])\n",
    "                TP53_bases_nonsense4[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_nonsense = [TP53_exons_extra_bases[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_nonsense.append(site_trio_nonsense)\n",
    "                    TP53_bases_nonsense4[k] = (v, sites_trio_nonsense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TP53_sites_nonsense5 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TP53_bases_nonsense4.items():\n",
    "        original_a_a = v[0][0][0]\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, original_a_a])\n",
    "            change = change+1\n",
    "        TP53_sites_nonsense5[a]=info\n",
    "        a=a+1\n",
    "    #     print(TP53_sites_nonsense5)\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TP53_sites_nonsense6={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TP53_sites_nonsense5.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        original_a_a = v[5]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TP53_sites_nonsense6[k] = (v[0],changes3, codon_n_912, original_a_a) #site_trio = the affected base and its neighbours\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    TP53_mutation_rates_nonsense = {}\n",
    "    for k, v in TP53_sites_nonsense6.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            original_a_a = v[3]\n",
    "            aa_positioning = v[2]\n",
    "            variant_type = '*'\n",
    "            variant_id = str(original_a_a)+str(aa_positioning)+variant_type\n",
    "            TP53_mutation_rates_nonsense[k] = (v[0],v[1], mutation_rate, variant_id)\n",
    "\n",
    "    TP53_mutation_rate_nonsense_sum = []\n",
    "\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TP53_mutation_rates_nonsense.items():\n",
    "        codon_position_in_gene=(v[0])\n",
    "        variant = v[3]\n",
    "        site = v[1]\n",
    "        mutation_rate_value = v[2] \n",
    "        mutation_rates_list.append((v[0], variant, site, mutation_rate_value))\n",
    "        TP53_mutation_rate_nonsense_sum.append(mutation_rate_value)\n",
    "\n",
    "    #Get the average of all the mutation rates\n",
    "    TP53_mutation_rate_nonsense_sum = []\n",
    "    for k, v in TP53_mutation_rates_nonsense.items():\n",
    "        mut_rate = v[2]\n",
    "        TP53_mutation_rate_nonsense_sum.append(mut_rate)\n",
    "\n",
    "    TP53_Summed_mutation_rate_nonsense_added = np.sum(TP53_mutation_rate_nonsense_sum)\n",
    "\n",
    "    print('Total summed mutation rate across TP53 nonsense for this study = ', TP53_Summed_mutation_rate_nonsense_added)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all # e.g. lower= 0, upper = 913 for whole gene\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "def mutation_rates_missense_range_of_variants_TP53(lower_range, upper_range):\n",
    "\n",
    "    #The mutation rate at a base is calculated based on its trinucleotide context.  To obtain the trinucleotide context, the base either side of the mutated base needs to be ascertained.\n",
    "    #The coding sequence (+ an extra base at each end) of each gene is used when determining the trinucleotide context of the base change.  The problem with using the coding sequence is that\n",
    "    #if a base change falls at the end of an exon, its neighbouring base will actuall be an intronic base, which may be different to the 'neighbouring' base in the coding sequence.\n",
    "    #To account for this, there is an \"intron_exon_check\" function that is called at the intron/exon boundary if the neighbouring base in the intron is different to the neighbouring base in the coding sequence.\n",
    "    \n",
    "    intron_exon_boundary_caution_base_TP53 = [74, 75, 96, 375, 672, 782, 919, 993, 1100]\n",
    "   \n",
    "    def intron_exon_check(position_in_gene_number):\n",
    "        if position_in_gene_number == 74:\n",
    "            gene_pos = 74\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 75:\n",
    "            gene_pos = 75\n",
    "            first_base = 'G'\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = TP53_exons_extra_bases[gene_pos+1]\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 96:\n",
    "            gene_pos = 96\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 375:\n",
    "            gene_pos = 375\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 672:\n",
    "            gene_pos = 672\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 782:\n",
    "            gene_pos = 782\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 919:\n",
    "            gene_pos = 919\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 993:\n",
    "            gene_pos = 993\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        if position_in_gene_number == 1100:\n",
    "            gene_pos = 1100\n",
    "            first_base = TP53_exons_extra_bases[gene_pos-1]\n",
    "            middle_base = TP53_exons_extra_bases[gene_pos]\n",
    "            third_base = 'G'\n",
    "            triplet = first_base+middle_base+third_base\n",
    "#             print(triplet)\n",
    "\n",
    "        return triplet\n",
    "    \n",
    "    #TP53 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    TP53_exons = 'ATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGACATAGTGTGGTGGTGCCCTATGAGCCGCCTGAGGTTGGCTCTGACTGTACCACCATCCACTACAACTACATGTGTAACAGTTCCTGCATGGGCGGCATGAACCGGAGGCCCATCCTCACCATCATCACACTGGAAGACTCCAGTGGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGAGGAAGAGAATCTCCGCAAGAAAGGGGAGCCTCACCACGAGCTGCCCCCAGGGAGCACTAAGCGAGCACTGCCCAACAACACCAGCTCCTCTCCCCAGCCAAAGAAGAAACCACTGGATGGAGAATATTTCACCCTTCAGATCCGTGGGCGTGAGCGCTTCGAGATGTTCCGAGAGCTGAATGAGGCCTTGGAACTCAAGGATGCCCAGGCTGGGAAGGAGCCAGGGGGGAGCAGGGCTCACTCCAGCCACCTGAAGTCCAAAAAGGGTCAGTCTACCTCCCGCCATAAAAAACTCATGTTCAAGACAGAAGGGCCTGACTCAGACTGA'\n",
    "    \n",
    "    #Split the sequence into codons \n",
    "    TP53_codons = [TP53_exons[i:i+3] for i in range(0, len(TP53_exons), 3)]\n",
    "\n",
    "    #TP53 transcript:\n",
    "    #Downloaded from http://grch37.ensembl.org\n",
    "    TP53_exons_extra_bases = 'CATGGAGGAGCCGCAGTCAGATCCTAGCGTCGAGCCCCCTCTGAGTCAGGAAACATTTTCAGACCTATGGAAACTACTTCCTGAAAACAACGTTCTGTCCCCCTTGCCGTCCCAAGCAATGGATGATTTGATGCTGTCCCCGGACGATATTGAACAATGGTTCACTGAAGACCCAGGTCCAGATGAAGCTCCCAGAATGCCAGAGGCTGCTCCCCCCGTGGCCCCTGCACCAGCAGCTCCTACACCGGCGGCCCCTGCACCAGCCCCCTCCTGGCCCCTGTCATCTTCTGTCCCTTCCCAGAAAACCTACCAGGGCAGCTACGGTTTCCGTCTGGGCTTCTTGCATTCTGGGACAGCCAAGTCTGTGACTTGCACGTACTCCCCTGCCCTCAACAAGATGTTTTGCCAACTGGCCAAGACCTGCCCTGTGCAGCTGTGGGTTGATTCCACACCCCCGCCCGGCACCCGCGTCCGCGCCATGGCCATCTACAAGCAGTCACAGCACATGACGGAGGTTGTGAGGCGCTGCCCCCACCATGAGCGCTGCTCAGATAGCGATGGTCTGGCCCCTCCTCAGCATCTTATCCGAGTGGAAGGAAATTTGCGTGTGGAGTATTTGGATGACAGAAACACTTTTCGACATAGTGTGGTGGTGCCCTATGAGCCGCCTGAGGTTGGCTCTGACTGTACCACCATCCACTACAACTACATGTGTAACAGTTCCTGCATGGGCGGCATGAACCGGAGGCCCATCCTCACCATCATCACACTGGAAGACTCCAGTGGTAATCTACTGGGACGGAACAGCTTTGAGGTGCGTGTTTGTGCCTGTCCTGGGAGAGACCGGCGCACAGAGGAAGAGAATCTCCGCAAGAAAGGGGAGCCTCACCACGAGCTGCCCCCAGGGAGCACTAAGCGAGCACTGCCCAACAACACCAGCTCCTCTCCCCAGCCAAAGAAGAAACCACTGGATGGAGAATATTTCACCCTTCAGATCCGTGGGCGTGAGCGCTTCGAGATGTTCCGAGAGCTGAATGAGGCCTTGGAACTCAAGGATGCCCAGGCTGGGAAGGAGCCAGGGGGGAGCAGGGCTCACTCCAGCCACCTGAAGTCCAAAAAGGGTCAGTCTACCTCCCGCCATAAAAAACTCATGTTCAAGACAGAAGGGCCTGACTCAGACTGAC'\n",
    "    \n",
    "    #List of all amino acid positions (across whole coding region) covered (amino acid position minus 1 (as python starts at 0))\n",
    "    TP53_missense_codons_DICT={}\n",
    "    for i in range(lower_range, upper_range):\n",
    "        TP53_missense_codons_DICT[i+1]= (TP53_codons[i])\n",
    "\n",
    "    calls_TP53_missense = {}\n",
    "    for k,v in TP53_missense_codons_DICT.items(): #iterating over the list of codon numbers (and codon sequences)\n",
    "        if v in Possible_Amino_Acid_Changes.keys(): #assign the list of all the possible amino acid changes to each codon they called\n",
    "            changes = Possible_Amino_Acid_Changes_letters[v]\n",
    "    #         print(Possible_Amino_Acid_Changes.values())\n",
    "            calls_TP53_missense[k]=(v, changes)\n",
    "\n",
    "    #Changing the original amino acid to a Word (e.g. Phe) rather than a codon\n",
    "    calls_TP53_missense_extra = {}\n",
    "\n",
    "    for k, v in calls_TP53_missense.items():\n",
    "        if v[0] in Genetic_Code:\n",
    "            Amino_acid_change = Genetic_Code.get(v[0])\n",
    "            calls_TP53_missense_extra[k] = (Amino_acid_change, v[0], v[1])\n",
    "            #e.g. 12: (Phe, 'K', '*', 'T' etc..)\n",
    "\n",
    "    #Changing the original amino acid to a LETTER (rather than the codon)\n",
    "    calls_TP53_missense_extra2 = {}\n",
    "\n",
    "    for k, v in calls_TP53_missense_extra.items():\n",
    "        if v[0] in Amino_acid_abbreviations:\n",
    "            Amino_acid_abbrev = Amino_acid_abbreviations.get(v[0])\n",
    "            calls_TP53_missense_extra2[k] = (Amino_acid_abbrev, v[1], v[2])\n",
    "            #e.g. 12: ((F), ('K', '*', 'T' etc..))\n",
    "\n",
    "    #Create a dictionary that contains details of the positions of missense variants at that codon\n",
    "    TP53_position_in_list_dict_missense = {}\n",
    "\n",
    "    for k, v in calls_TP53_missense_extra2.items():\n",
    "        original_AA = v[0]\n",
    "        possible_AA_changes = v[2]\n",
    "        possible_AA_changes_string = ''.join(v[2])\n",
    "        positions_in_list = []\n",
    "        for i in possible_AA_changes:\n",
    "            if i != original_AA: #not synonymous\n",
    "                if i != '*': #not missense\n",
    "                    position_in_list=-1\n",
    "                    missense_change = i\n",
    "                    new_aa_list = []\n",
    "                    while True:\n",
    "                        position_in_list = possible_AA_changes_string.find(missense_change, position_in_list+1)\n",
    "        #               print(position_in_list)\n",
    "                        if position_in_list == -1: break\n",
    "                        positions_in_list.append(position_in_list)\n",
    "                        positions_in_list = set(positions_in_list) #convert into a 'set' to remove duplicate positions\n",
    "                        positions_in_list = list(positions_in_list) #convert back to a list so can be added to dictionary\n",
    "                    for pos in positions_in_list:\n",
    "                        new_aa = possible_AA_changes_string[pos]\n",
    "                        new_aa_list.append(new_aa)\n",
    "                    TP53_position_in_list_dict_missense[k]=(positions_in_list, v, new_aa_list)\n",
    "\n",
    "        aa_positions_variants_called = []\n",
    "        for k, v in TP53_position_in_list_dict_missense.items():\n",
    "            aa_positions_variants_called.append(k)\n",
    "    # print('positions at which missense variants called = ', aa_positions_variants_called)\n",
    "\n",
    "    #Create a dictionary that contains the position in the list of possible AA where the variant is and what it changes to\n",
    "    TP53_bases_missense = {}\n",
    "    for k, v in TP53_position_in_list_dict_missense.items():\n",
    "        original_a_a = v[1][0]\n",
    "        codon_seq = v[1][1]\n",
    "        base_changes=[]\n",
    "        new_a_a_list = v[2]\n",
    "        if codon_seq in Possible_Base_Changes.keys():\n",
    "            list_of_possible_bases = Possible_Base_Changes.get(codon_seq)\n",
    "        for i in v[0]:\n",
    "            base_change_position = i\n",
    "            base_change = list_of_possible_bases[base_change_position]\n",
    "            base_changes.append(base_change)\n",
    "        TP53_bases_missense[k]=(v, base_changes, original_a_a, new_a_a_list)\n",
    "\n",
    "    #Create dictionary which contains original codon seq and details on which base is changed\n",
    "    TP53_bases_missense2 = {}\n",
    "\n",
    "    for k, v in TP53_bases_missense.items():\n",
    "        base_change_positions = v[0][0]\n",
    "        codon_positions = []\n",
    "        positions_in_gene = []\n",
    "        new_a_a_list = v[3]\n",
    "        for i in base_change_positions:\n",
    "            base_change_position = i\n",
    "            if base_change_position <= 2:\n",
    "                codon_pos = 1\n",
    "            if 2 < base_change_position <= 5:\n",
    "                codon_pos = 2\n",
    "            if 5 < base_change_position <= 8:\n",
    "                codon_pos = 3\n",
    "            codon_positions.append(codon_pos)\n",
    "        TP53_bases_missense2[k] = (v, codon_positions)\n",
    "\n",
    "    #Create dictionary which contains original codon seq, details on which base is changed and it's position in the coding region\n",
    "    TP53_bases_missense3 = {}\n",
    "\n",
    "    for k, v in TP53_bases_missense2.items():\n",
    "        codon_positions = v[1]\n",
    "        base_changes = v[0][1]\n",
    "        original_a_a = v[0][2]\n",
    "        new_aa_list = v[0][0][2]\n",
    "        original_bases = []\n",
    "        positions_in_gene = []\n",
    "        codon_number = k\n",
    "        for i in codon_positions:\n",
    "            codon_pos = i\n",
    "            position_in_gene = ((int(k)-1)*3)+codon_pos #codon number minus 1 (e.g. codon 307 -1 = 306, then times by 3 takes you to beginning of codon 307 (then add the position))  \n",
    "            positions_in_gene.append(position_in_gene)\n",
    "            original_base = (v[0][0][1][1][codon_pos-1])\n",
    "            original_bases.append(original_base)\n",
    "        TP53_bases_missense3[k] = (v[0][0][1], positions_in_gene, original_bases, base_changes, codon_number, original_a_a, new_aa_list)\n",
    "\n",
    "    # Create dictionary which contains original codon seq, details on which base is changed and its neighbouring bases\n",
    "    TP53_bases_missense4 = {}\n",
    "\n",
    "    for k, v in TP53_bases_missense3.items():\n",
    "        position_in_gene = v[1]\n",
    "        sites_trio_missense=[]\n",
    "        new_a_a_list = v[6]\n",
    "\n",
    "        for i in position_in_gene:\n",
    "    #         print(i)\n",
    "            if i in intron_exon_boundary_caution_base_TP53:\n",
    "#                 print('positions at intron_exon_boundary =',i)\n",
    "                adapted_trio = intron_exon_check(i)\n",
    "                sites_trio_missense.append([adapted_trio])\n",
    "                TP53_bases_missense4[k] = (v, sites_trio_missense)\n",
    "\n",
    "            else:\n",
    "                for i in position_in_gene:\n",
    "                    gene_pos = i\n",
    "                    site_trio_missense = [TP53_exons_extra_bases[gene_pos:gene_pos+3] for gene_pos in range (gene_pos-1, gene_pos, 3)]\n",
    "                    sites_trio_missense.append(site_trio_missense)\n",
    "                    TP53_bases_missense4[k] = (v, sites_trio_missense)\n",
    "\n",
    "    #Create a dictionary containing pos in coding region, original base, new base, base affected with its neighbours\n",
    "    #e.g. 938, G, A, TGG = base 938 on coding region, original base = G, changes to base A, base with neighbours either side = TGG\n",
    "    TP53_sites_missense5 = {}\n",
    "\n",
    "    a=0\n",
    "    for k, v in TP53_bases_missense4.items():\n",
    "        original_a_a = v[0][0][0]\n",
    "        new_a_a = v[0][6]\n",
    "        codon_n = v[0][1] #base number in codon (out of 3 x 912)\n",
    "        codon_n_912 = v[0][4]\n",
    "        original_base = v[0][2]\n",
    "        new_base = v[0][3]\n",
    "        site_trio = v[1]\n",
    "        info=[]\n",
    "\n",
    "        change = 0\n",
    "        for i in codon_n:\n",
    "            codon_number = i\n",
    "            orig_base = original_base[change]\n",
    "            new_amino_acid = new_a_a[change]\n",
    "            n_base = new_base[change]\n",
    "            site_tri = site_trio[change]\n",
    "            info.append([codon_number, orig_base, n_base, site_tri, codon_n_912, original_a_a, new_amino_acid])\n",
    "            change = change+1\n",
    "        TP53_sites_missense5[a]=info\n",
    "        a=a+1\n",
    "    #     print(TP53_sites_missense5)\n",
    "\n",
    "    # Write in a way that can iterate over mutation rates (i.e. put in to form C[C>A]G etc)   \n",
    "    TP53_sites_missense6={}\n",
    "\n",
    "    separating_positions = {} # separate out all the possible changes to just 1 effective item in the list\n",
    "    a=0\n",
    "    for k, v in TP53_sites_missense5.items():\n",
    "        for i in v: #because some contain more than 1 change from that codon:\n",
    "            separating_positions[a]=i\n",
    "            a=a+1\n",
    "\n",
    "    for k, v in separating_positions.items():\n",
    "        base_pos = v[0]\n",
    "        original_base = v[1]\n",
    "        new_base = v[2]\n",
    "        site_trio = v[3]\n",
    "        codon_n_912 = v[4]\n",
    "        original_a_a = v[5]\n",
    "        new_a_a = v[6]\n",
    "        sequence = (str(original_base),str(new_base))\n",
    "        changes = '>'.join(sequence)\n",
    "        left_base = site_trio[0][0]\n",
    "        right_base = site_trio[0][2]\n",
    "        sequence2 = (changes, right_base)\n",
    "        changes2 = ']'.join(sequence2)\n",
    "        sequence3 = (left_base, changes2)\n",
    "        changes3 = '['.join(sequence3)\n",
    "        TP53_sites_missense6[k] = (v[0],changes3, codon_n_912, original_a_a, new_a_a) #site_trio = the affected base and its neighbours\n",
    "\n",
    "    #Assign the mutation rates to each site...\n",
    "    TP53_mutation_rates_missense = {}\n",
    "    for k, v in TP53_sites_missense6.items():\n",
    "        base_change_info = v[1]\n",
    "        if base_change_info in Mutation_rates.keys():\n",
    "            mutation_rate = Mutation_rates.get(base_change_info)\n",
    "            original_a_a = v[3]\n",
    "            new_a_a = v[4]\n",
    "            aa_positioning = v[2]\n",
    "            variant_id = str(original_a_a)+str(aa_positioning)+str(new_a_a)\n",
    "            TP53_mutation_rates_missense[k] = (v[0],v[1], mutation_rate, variant_id)\n",
    "\n",
    "    TP53_mutation_rate_missense_sum = []\n",
    "\n",
    "    mutation_rates_list = []\n",
    "    for k, v in TP53_mutation_rates_missense.items():\n",
    "        codon_position_in_gene=(v[0])\n",
    "        variant = v[3]\n",
    "        site = v[1]\n",
    "        mutation_rate_value = v[2] \n",
    "        mutation_rates_list.append((v[0], variant, site, mutation_rate_value))\n",
    "        TP53_mutation_rate_missense_sum.append(mutation_rate_value)\n",
    "\n",
    "    #Get the average of all the mutation rates\n",
    "    TP53_mutation_rate_missense_sum = []\n",
    "    for k, v in TP53_mutation_rates_missense.items():\n",
    "        mut_rate = v[2]\n",
    "        TP53_mutation_rate_missense_sum.append(mut_rate)\n",
    "\n",
    "    TP53_Summed_mutation_rate_missense_added = np.sum(TP53_mutation_rate_missense_sum)\n",
    "\n",
    "    print('Total summed mutation rate across TP53 missense for this study = ', TP53_Summed_mutation_rate_missense_added)\n",
    "\n",
    "    mutation_rates_list_all = mutation_rates_list\n",
    "    return mutation_rates_list_all # e.g. lower= 0, upper = 913 for whole gene\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Lists of variants targeted by each study"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- In order to calculate the number of times you expect to see a variant across all studies, you need to know which variants were call-able by which studies.\n",
    "- This can be calculated by using the mutation rate functions (above) as they will also produce a list of the variants within e.g. range of amino acid positions of interest or from a CSV file of variants."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Jaiswal 2014"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "code_folding": [],
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  3.8678204400000003e-07\n",
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  5.08517098e-07\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Jaiswal's DNMT3A variants\n",
    "Jaiswal_non_DNMT3A = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(0, 913)\n",
    "Jaiswal_miss_DNMT3A = mutation_rates_list_missense_variants_DNMT3A_NP072046('Data files/Jaiswal DNMT3A missense variants called.csv')\n",
    "\n",
    "#Lists of DNMT3A variants targetable in Jaiswal 2014 (i.e. included in their panel)\n",
    "Jaiswal_nonsense_DNMT3A = []\n",
    "for i in Jaiswal_non_DNMT3A:\n",
    "    Jaiswal_nonsense_DNMT3A.append(str('p.')+i[1])\n",
    "    \n",
    "Jaiswal_missense_DNMT3A = []\n",
    "for i in Jaiswal_miss_DNMT3A:\n",
    "    Jaiswal_missense_DNMT3A.append(i[3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "code_folding": [],
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TET2 nonsense_NP_01120680 for this study =  5.10993418e-07\n",
      "Total summed mutation rate across TET2 missense_NP_01120680 for this study =  1.1931589999999999e-08\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Jaiswal's TET2 variants\n",
    "Jaiswal_non_TET2 = mutation_rates_nonsense_range_of_variants_TET2(0, 1137)\n",
    "Jaiswal_miss_TET2 = mutation_rates_list_missense_variants_TET2('Data files/Jaiswal TET2 missense variants called.csv')\n",
    "\n",
    "#Lists of TET2 variants targetable in Jaiswal 2014 (i.e. included in their panel)\n",
    "Jaiswal_nonsense_TET2 = []\n",
    "for i in Jaiswal_non_TET2:\n",
    "    Jaiswal_nonsense_TET2.append(str('p.')+i[1])\n",
    "    \n",
    "Jaiswal_missense_TET2 = []\n",
    "for i in Jaiswal_miss_TET2:\n",
    "    Jaiswal_missense_TET2.append(i[3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "code_folding": [],
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across ASXL1 nonsense for this study =  6.254075530000001e-07\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Jaiswal's ASXL1 variants\n",
    "Jaiswal_non_ASXL1 = mutation_rates_nonsense_range_of_variants_ASXL1(0, 1542)\n",
    "\n",
    "#Lists of ASXL1 variants targetable in Jaiswal 2014 (i.e. included in their panel)\n",
    "Jaiswal_nonsense_ASXL1 = []\n",
    "for i in Jaiswal_non_ASXL1:\n",
    "    Jaiswal_nonsense_ASXL1.append(str('p.')+i[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "code_folding": [],
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 nonsense for this study =  1.6147847e-07\n",
      "Total summed mutation rate across TP53 missense for this study =  6.26412261e-07\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Jaiswal's TP53 variants\n",
    "Jaiswal_non_TP53 = mutation_rates_nonsense_range_of_variants_TP53(0, 394)\n",
    "Jaiswal_miss_TP53 = mutation_rates_list_missense_variants_TP53('Data files/Jaiswal TP53 missense variants called.csv')\n",
    "\n",
    "#Lists of TP53 variants targetable in Jaiswal 2014 (i.e. included in their panel)\n",
    "Jaiswal_nonsense_TP53 = []\n",
    "for i in Jaiswal_non_TP53:\n",
    "    Jaiswal_nonsense_TP53.append(str('p.')+i[1])\n",
    "    \n",
    "Jaiswal_missense_TP53 = []\n",
    "for i in Jaiswal_miss_TP53:\n",
    "    Jaiswal_missense_TP53.append(i[3])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Genovese 2014"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  2.7073501399999997e-07\n",
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  1.0921795000000001e-08\n",
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  5.383253473e-06\n",
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  4.68086602e-07\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Genovese's DNMT3A variants\n",
    "Genovese_nonsense_DNMT3A_1 = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(213, 841)\n",
    "Genovese_nonsense_DNMT3A_2 = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(866, 913)\n",
    "Genovese_missense_DNMT3A_1 = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(213, 841)\n",
    "Genovese_missense_DNMT3A_2 = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(866, 913)\n",
    "\n",
    "#Lists of DNMT3A variants targetable in Genovese 2014 (i.e. included in their panel)\n",
    "Genovese_nonsense_DNMT3A = []\n",
    "Genovese_missense_DNMT3A = []\n",
    "for i in Genovese_nonsense_DNMT3A_1:\n",
    "    Genovese_nonsense_DNMT3A.append(str('p.')+i[1])\n",
    "for i in Genovese_nonsense_DNMT3A_2:\n",
    "    Genovese_nonsense_DNMT3A.append(str('p.')+i[1])\n",
    "for i in Genovese_missense_DNMT3A_1:\n",
    "    Genovese_missense_DNMT3A.append(str('p.')+i[1])\n",
    "for i in Genovese_missense_DNMT3A_1:\n",
    "    Genovese_missense_DNMT3A.append(str('p.')+i[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "code_folding": [
     0,
     6
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  9.167888e-09\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Genovese's DNMT3A variants\n",
    "Genovese_missense_DNMT3A_partial_codons = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(846, 848)\n",
    "base_position_list = [2540, 2541, 2542, 2543]\n",
    "Genovese_partial_missense = []\n",
    "Genovese_partial_missense_sum_values = []\n",
    "for i in Genovese_missense_DNMT3A_partial_codons:\n",
    "    if i[0] in base_position_list:\n",
    "        Genovese_partial_missense.append(i)\n",
    "        Genovese_partial_missense_sum_values.append(i[3])\n",
    "        \n",
    "Genovese_partial_missense_added = np.sum(Genovese_partial_missense_sum_values)\n",
    "\n",
    "#Lists of DNMT3A variants targetable in Genovese 2014 (i.e. included in their panel)\n",
    "for i in Genovese_partial_missense:\n",
    "    Genovese_missense_DNMT3A.append(str('p.'+i[1]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "code_folding": [],
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TET2 nonsense_NP_060098 for this study =  5.192273050000001e-07\n",
      "Total summed mutation rate across TET2 missense_NP_060098 for this study =  3.8538210999999995e-08\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Genovese's TET2 variants\n",
    "Genovese_non_TET2 = mutation_rates_nonsense_range_of_variants_TET2_shorter(0, 1166)\n",
    "Genovese_miss_TET2 = mutation_rates_list_missense_variants_TET2_shorter('Data files/Genovese TET2 missense variants called.csv')\n",
    "\n",
    "#Lists of TET2 variants targetable in Genovese 2014 (i.e. included in their panel)\n",
    "Genovese_nonsense_TET2 = []\n",
    "for i in Genovese_non_TET2:\n",
    "    Genovese_nonsense_TET2.append(str('p.')+i[1])\n",
    "    \n",
    "Genovese_missense_TET2 = []\n",
    "for i in Genovese_miss_TET2:\n",
    "    Genovese_missense_TET2.append(i[3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across ASXL1 missense for this study =  4.871308e-09\n",
      "Total summed mutation rate across ASXL1 nonsense for this study =  4.4622556000000006e-08\n",
      "Total summed mutation rate across ASXL1 nonsense for this study =  2.6285938e-08\n",
      "Total summed mutation rate across ASXL1 nonsense for this study =  2.2726005200000001e-07\n",
      "Total summed mutation rate across ASXL1 nonsense for this study =  2.93543906e-07\n",
      "Total summed mutation rate across ASXL1 nonsense for this study =  1.20044e-08\n",
      "Total summed mutation rate across ASXL1 nonsense for this study =  3.78353e-10\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Genovese's ASXL1 variants\n",
    "Genovese_miss_ASXL1 = mutation_rates_list_missense_variants_ASXL1('Data files/Genovese ASXL1 missense variants called.csv')\n",
    "Genovese_non_ASXL1_1 = mutation_rates_nonsense_range_of_variants_ASXL1(19, 128)\n",
    "Genovese_non_ASXL1_2 = mutation_rates_nonsense_range_of_variants_ASXL1(143, 196)\n",
    "Genovese_non_ASXL1_3 = mutation_rates_nonsense_range_of_variants_ASXL1(200, 627)\n",
    "Genovese_non_ASXL1_4 = mutation_rates_nonsense_range_of_variants_ASXL1(655, 1541)\n",
    "Genovese_non_ASXL1_5 = mutation_rates_nonsense_range_of_variants_ASXL1(141, 143)\n",
    "Genovese_non_ASXL1_6 = mutation_rates_nonsense_range_of_variants_ASXL1(627, 629)\n",
    "\n",
    "#Lists of ASXL1 variants targetable in Genovese 2014 (i.e. included in their panel)\n",
    "Genovese_nonsense_ASXL1 = []\n",
    "for i in Genovese_non_ASXL1_1:\n",
    "    Genovese_nonsense_ASXL1.append(str('p.')+i[1])\n",
    "for i in Genovese_non_ASXL1_2:\n",
    "    Genovese_nonsense_ASXL1.append(str('p.')+i[1])\n",
    "for i in Genovese_non_ASXL1_3:\n",
    "    Genovese_nonsense_ASXL1.append(str('p.')+i[1])\n",
    "for i in Genovese_non_ASXL1_4:\n",
    "    Genovese_nonsense_ASXL1.append(str('p.')+i[1])\n",
    "for i in Genovese_non_ASXL1_5:\n",
    "    Genovese_nonsense_ASXL1.append(str('p.')+i[1])\n",
    "for i in Genovese_non_ASXL1_6:\n",
    "    Genovese_nonsense_ASXL1.append(str('p.')+i[1])\n",
    "    \n",
    "Genovese_missense_ASXL1 = []\n",
    "for i in Genovese_miss_ASXL1:\n",
    "    Genovese_missense_ASXL1.append(i[3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 nonsense for this study =  7.915323800000001e-08\n",
      "Total summed mutation rate across TP53 missense for this study =  4.1547220999999997e-07\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Genovese's TP53 variants\n",
    "Genovese_non_TP53 = mutation_rates_list_nonsense_variants_TP53('Data files/Genovese TP53 nonsense variants called.csv')\n",
    "Genovese_miss_TP53 = mutation_rates_list_missense_variants_TP53('Data files/Genovese TP53 missense variants called.csv')\n",
    "\n",
    "#Lists of TP53 variants targetable in Genovese 2014 (i.e. included in their panel)\n",
    "Genovese_nonsense_TP53 = []\n",
    "for i in Genovese_non_TP53:\n",
    "    Genovese_nonsense_TP53.append(i[3])\n",
    "    \n",
    "Genovese_missense_TP53 = []\n",
    "for i in Genovese_miss_TP53:\n",
    "    Genovese_missense_TP53.append(i[3])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## McKerrel 2015"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "#List of missense variants in DNMT3A called by McKerrel\n",
    "McKerrel_missense_DNMT3A = ['p.R882H', 'p.R882C', 'p.R882P', 'p.R882S', 'p.R882G', 'p.R882L']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Zink 2017"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "code_folding": [],
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  3.8678204400000003e-07\n",
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  1.211073846e-06\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Zink's DNMT3A variants\n",
    "Zink_non_DNMT3A = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(0, 913)\n",
    "Zink_miss_DNMT3A = mutation_rates_list_missense_variants_DNMT3A_NP072046('Data files/Zink DNMT3A missense variants in COSMIC.csv')\n",
    "\n",
    "#Lists of DNMT3A variants targetable in Zink 2017 (i.e. included in their panel)\n",
    "Zink_nonsense_DNMT3A = []\n",
    "for i in Zink_non_DNMT3A:\n",
    "    Zink_nonsense_DNMT3A.append(str('p.')+i[1])\n",
    "\n",
    "Zink_missense_DNMT3A = []\n",
    "for i in Zink_miss_DNMT3A:\n",
    "    Zink_missense_DNMT3A.append(i[3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "code_folding": [],
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TET2 nonsense_NP_01120680 for this study =  8.664021850000001e-07\n",
      "Total summed mutation rate across TET2 missense_NP_01120680 for this study =  1.090214402e-06\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Zink's TET2 variants\n",
    "Zink_non_TET2 = mutation_rates_nonsense_range_of_variants_TET2(0, 2003)\n",
    "Zink_miss_TET2 = mutation_rates_list_missense_variants_TET2('Data files/Zink TET2 missense variants in COSMIC.csv')\n",
    "\n",
    "#Lists of TET2 variants targetable in Zink 2017 (i.e. included in their panel)\n",
    "Zink_nonsense_TET2 = []\n",
    "for i in Zink_non_TET2:\n",
    "    Zink_nonsense_TET2.append(str('p.')+i[1])\n",
    "    \n",
    "Zink_missense_TET2 = []\n",
    "for i in Zink_miss_TET2:\n",
    "    Zink_missense_TET2.append(i[3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "code_folding": [],
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across ASXL1 nonsense for this study =  6.254075530000001e-07\n",
      "Total summed mutation rate across ASXL1 missense for this study =  5.189833720000001e-07\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Zink's ASXL1 variants\n",
    "Zink_non_ASXL1 = mutation_rates_nonsense_range_of_variants_ASXL1(0, 1542)\n",
    "Zink_miss_ASXL1 = mutation_rates_list_missense_variants_ASXL1('Data files/Zink ASXL1 missense variants called.csv')\n",
    "\n",
    "#Lists of ASXL1 variants targetable in Zink 2017 (i.e. included in their panel)\n",
    "Zink_nonsense_ASXL1 = []\n",
    "for i in Zink_non_ASXL1:\n",
    "    Zink_nonsense_ASXL1.append(str('p.')+i[1])\n",
    "    \n",
    "Zink_missense_ASXL1 = []\n",
    "for i in Zink_miss_ASXL1:\n",
    "    Zink_missense_ASXL1.append(i[3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "code_folding": [],
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 nonsense for this study =  1.6147847e-07\n",
      "Total summed mutation rate across TP53 missense for this study =  1.464249575e-06\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Zink's TP53 variants\n",
    "Zink_non_TP53 = mutation_rates_nonsense_range_of_variants_TP53(0, 394)\n",
    "Zink_miss_TP53 = mutation_rates_list_missense_variants_TP53('Data files/Zink TP53 missense variants called.csv')\n",
    "\n",
    "#Lists of TP53 variants targetable in Zink 2017 (i.e. included in their panel)\n",
    "Zink_nonsense_TP53 = []\n",
    "for i in Zink_non_TP53:\n",
    "    Zink_nonsense_TP53.append(str('p.')+i[1])\n",
    "    \n",
    "Zink_missense_TP53 = []\n",
    "for i in Zink_miss_TP53:\n",
    "    Zink_missense_TP53.append(i[3])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Coombs 2017"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  3.8678204400000003e-07\n",
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  8.083470887e-06\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Coombs' DNMT3A variants\n",
    "Coombs_non_DNMT3A = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(0, 913)\n",
    "Coombs_miss_DNMT3A = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(0, 913)\n",
    "\n",
    "#Lists of DNMT3A variants targetable in Coombs 2017 (i.e. included in their panel)\n",
    "Coombs_nonsense_DNMT3A = []\n",
    "for i in Coombs_non_DNMT3A:\n",
    "    Coombs_nonsense_DNMT3A.append(str('p.')+i[1])\n",
    "    \n",
    "Coombs_missense_DNMT3A = []\n",
    "for i in Coombs_miss_DNMT3A:\n",
    "    Coombs_missense_DNMT3A.append(str('p.')+i[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TET2 nonsense_NP_01120680 for this study =  8.664021850000001e-07\n",
      "Total summed mutation rate across TET2 missense_NP_01120680 for this study =  1.2216578715e-05\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Coombs' TET2 variants\n",
    "Coombs_non_TET2 = mutation_rates_nonsense_range_of_variants_TET2(0, 2003)\n",
    "Coombs_miss_TET2 = mutation_rates_missense_range_of_variants_TET2(0, 2003)\n",
    "\n",
    "#Lists of TET2 variants targetable in Coombs 2017 (i.e. included in their panel)\n",
    "Coombs_nonsense_TET2 = []\n",
    "for i in Coombs_non_TET2:\n",
    "    Coombs_nonsense_TET2.append(str('p.')+i[1])\n",
    "    \n",
    "Coombs_missense_TET2 = []\n",
    "for i in Coombs_miss_TET2:\n",
    "    Coombs_missense_TET2.append(str('p.')+i[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across ASXL1 nonsense for this study =  6.254075530000001e-07\n",
      "Total summed mutation rate across ASXL1 missense for this study =  1.1360431228e-05\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Coombs' ASXL1 variants\n",
    "Coombs_non_ASXL1 = mutation_rates_nonsense_range_of_variants_ASXL1(0, 1542)\n",
    "Coombs_miss_ASXL1 = mutation_rates_missense_range_of_variants_ASXL1(0, 1542)\n",
    "\n",
    "#Lists of ASXL1 variants targetable in Coombs 2017 (i.e. included in their panel)\n",
    "Coombs_nonsense_ASXL1 = []\n",
    "for i in Coombs_non_ASXL1:\n",
    "    Coombs_nonsense_ASXL1.append(str('p.')+i[1])\n",
    "    \n",
    "Coombs_missense_ASXL1 = []\n",
    "for i in Coombs_miss_ASXL1:\n",
    "    Coombs_missense_ASXL1.append(str('p.')+i[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 nonsense for this study =  1.6147847e-07\n",
      "Total summed mutation rate across TP53 missense for this study =  3.273721457e-06\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Coombs' TP53 variants\n",
    "Coombs_non_TP53 = mutation_rates_nonsense_range_of_variants_TP53(0, 394)\n",
    "Coombs_miss_TP53 = mutation_rates_missense_range_of_variants_TP53(0, 394)\n",
    "\n",
    "#Lists of TP53 variants targetable in Coombs 2017 (i.e. included in their panel)\n",
    "Coombs_nonsense_TP53 = []\n",
    "for i in Coombs_non_TP53:\n",
    "    Coombs_nonsense_TP53.append(str('p.')+i[1])\n",
    "    \n",
    "Coombs_missense_TP53 = []\n",
    "for i in Coombs_miss_TP53:\n",
    "    Coombs_missense_TP53.append(str('p.')+i[1])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Young 2016 & 2019"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  3.8678204400000003e-07\n",
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  8.083470887e-06\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Young DNMT3A variants\n",
    "Young_non_DNMT3A = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(0, 913)\n",
    "Young_miss_DNMT3A = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(0, 913)\n",
    "\n",
    "#Lists of DNMT3A variants targetable in Young (i.e. included in their panel)\n",
    "Young_nonsense_DNMT3A = []\n",
    "for i in Young_non_DNMT3A:\n",
    "    Young_nonsense_DNMT3A.append(str('p.')+i[1])\n",
    "    \n",
    "Young_missense_DNMT3A = []\n",
    "for i in Young_miss_DNMT3A:\n",
    "    Young_missense_DNMT3A.append(str('p.')+i[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TET2 nonsense_NP_01120680 for this study =  8.664021850000001e-07\n",
      "Total summed mutation rate across TET2 missense_NP_01120680 for this study =  1.2216578715e-05\n",
      "Total summed mutation rate across TET2 missense_NP_01120680 for this study =  5.206542600000001e-08\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Young TET2 variants\n",
    "Young_non_TET2 = mutation_rates_nonsense_range_of_variants_TET2(0, 2003)\n",
    "Young_miss_TET2 = mutation_rates_missense_range_of_variants_TET2(0, 2003)\n",
    "\n",
    "#Lists of TET2 variants targetable in Young (i.e. included in their panel)\n",
    "Young_nonsense_TET2 = []\n",
    "for i in Young_non_TET2:\n",
    "    Young_nonsense_TET2.append(str('p.')+i[1])\n",
    "    \n",
    "Young_missense_TET2_all = []\n",
    "for i in Young_miss_TET2:\n",
    "    Young_missense_TET2_all.append(str('p.')+i[1])\n",
    "    \n",
    "#Excluded regions:\n",
    "excluded_TET2_Young=[]\n",
    "Young_TET2_excluded = mutation_rates_list_missense_variants_TET2('Data files/excluded_variants_Young_TET2.csv')\n",
    "for i in Young_TET2_excluded:\n",
    "    excluded_TET2_Young.append(i[3])\n",
    "\n",
    "Young_missense_TET2=[]\n",
    "for i in Young_missense_TET2_all:\n",
    "    if i not in Young_TET2_excluded:\n",
    "        Young_missense_TET2.append(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across ASXL1 nonsense for this study =  3.3358587600000003e-07\n",
      "Total summed mutation rate across ASXL1 missense for this study =  6.820950358e-06\n",
      "Total summed mutation rate across ASXL1 missense for this study =  5.2494673999999997e-08\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Young ASXL1 variants\n",
    "Young_non_ASXL1 = mutation_rates_nonsense_range_of_variants_ASXL1(573, 1542)\n",
    "Young_miss_ASXL1 = mutation_rates_missense_range_of_variants_ASXL1(573, 1542)\n",
    "\n",
    "#Lists of ASXL1 variants targetable in Young (i.e. included in their panel)\n",
    "Young_nonsense_ASXL1 = []\n",
    "for i in Young_non_ASXL1:\n",
    "    Young_nonsense_ASXL1.append(str('p.')+i[1])\n",
    "    \n",
    "Young_missense_ASXL1_all = []\n",
    "for i in Young_miss_ASXL1:\n",
    "    Young_missense_ASXL1_all.append(str('p.')+i[1])\n",
    "    \n",
    "#Excluded regions:\n",
    "excluded_ASXL1_Young=[]\n",
    "Young_ASXL1_excluded = mutation_rates_list_missense_variants_ASXL1('Data files/excluded_variants_Young_ASXL1.csv')\n",
    "for i in Young_ASXL1_excluded:\n",
    "    excluded_ASXL1_Young.append(i[3])\n",
    "\n",
    "Young_missense_ASXL1=[]\n",
    "for i in Young_missense_ASXL1_all:\n",
    "    if i not in Young_ASXL1_excluded:\n",
    "        Young_missense_ASXL1.append(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 nonsense for this study =  1.6147847e-07\n",
      "Total summed mutation rate across TP53 missense for this study =  3.273721457e-06\n",
      "Total summed mutation rate across TP53 missense for this study =  1.27013e-09\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Young TP53 variants\n",
    "Young_non_TP53 = mutation_rates_nonsense_range_of_variants_TP53(0, 394)\n",
    "Young_miss_TP53 = mutation_rates_missense_range_of_variants_TP53(0, 394)\n",
    "\n",
    "#Lists of TP53 variants targetable in Young (i.e. included in their panel)\n",
    "Young_nonsense_TP53 = []\n",
    "for i in Young_non_TP53:\n",
    "    Young_nonsense_TP53.append(str('p.')+i[1])\n",
    "    \n",
    "Young_missense_TP53_all = []\n",
    "for i in Young_miss_TP53:\n",
    "    Young_missense_TP53_all.append(str('p.')+i[1])\n",
    "    \n",
    "#Excluded regions:\n",
    "excluded_TP53_Young=[]\n",
    "Young_TP53_excluded = mutation_rates_list_missense_variants_TP53('Data files/excluded_variants_Young_TP53.csv')\n",
    "for i in Young_TP53_excluded:\n",
    "    excluded_TP53_Young.append(i[3])\n",
    "\n",
    "Young_missense_TP53=[]\n",
    "for i in Young_missense_TP53_all:\n",
    "    if i not in Young_TP53_excluded:\n",
    "        Young_missense_TP53.append(i)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Desai 2018"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  3.8678204400000003e-07\n",
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  8.083470887e-06\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Desai DNMT3A variants\n",
    "Desai_non_DNMT3A = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(0, 913)\n",
    "Desai_miss_DNMT3A = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(0, 913)\n",
    "\n",
    "#Lists of DNMT3A variants targetable in Desai (i.e. included in their panel)\n",
    "Desai_nonsense_DNMT3A = []\n",
    "for i in Desai_non_DNMT3A:\n",
    "    Desai_nonsense_DNMT3A.append(str('p.')+i[1])\n",
    "    \n",
    "Desai_missense_DNMT3A = []\n",
    "for i in Desai_miss_DNMT3A:\n",
    "    Desai_missense_DNMT3A.append(str('p.')+i[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TET2 nonsense_NP_01120680 for this study =  8.664021850000001e-07\n",
      "Total summed mutation rate across TET2 missense_NP_01120680 for this study =  1.2216578715e-05\n",
      "Total summed mutation rate across TET2 missense_NP_01120680 for this study =  8.3116988e-08\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Desai TET2 variants\n",
    "Desai_non_TET2 = mutation_rates_nonsense_range_of_variants_TET2(0, 2003)\n",
    "Desai_miss_TET2 = mutation_rates_missense_range_of_variants_TET2(0, 2003)\n",
    "\n",
    "#Lists of TET2 variants targetable in Desai (i.e. included in their panel)\n",
    "Desai_nonsense_TET2 = []\n",
    "for i in Desai_non_TET2:\n",
    "    Desai_nonsense_TET2.append(str('p.')+i[1])\n",
    "    \n",
    "Desai_missense_TET2_all = []\n",
    "for i in Desai_miss_TET2:\n",
    "    Desai_missense_TET2_all.append(str('p.')+i[1])\n",
    "    \n",
    "#Excluded regions:\n",
    "excluded_TET2_Desai=[]\n",
    "Desai_TET2_excluded = mutation_rates_list_missense_variants_TET2('Data files/excluded_variants_Desai_2018_TET2.csv')\n",
    "for i in Desai_TET2_excluded:\n",
    "    excluded_TET2_Desai.append(i[3])\n",
    "\n",
    "Desai_missense_TET2=[]\n",
    "for i in Desai_missense_TET2_all:\n",
    "    if i not in Desai_TET2_excluded:\n",
    "        Desai_missense_TET2.append(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across ASXL1 nonsense for this study =  6.254075530000001e-07\n",
      "Total summed mutation rate across ASXL1 missense for this study =  1.1360431228e-05\n",
      "Total summed mutation rate across ASXL1 missense for this study =  5.8171873e-08\n",
      "Total summed mutation rate across ASXL1 nonsense for this study =  5.39508e-10\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Desai ASXL1 variants\n",
    "Desai_non_ASXL1 = mutation_rates_nonsense_range_of_variants_ASXL1(0, 1542)\n",
    "Desai_miss_ASXL1 = mutation_rates_missense_range_of_variants_ASXL1(0, 1542)\n",
    "\n",
    "#Lists of ASXL1 variants targetable in Desai (i.e. included in their panel)\n",
    "Desai_nonsense_ASXL1_all = []\n",
    "for i in Desai_non_ASXL1:\n",
    "    Desai_nonsense_ASXL1_all.append(str('p.')+i[1])\n",
    "    \n",
    "Desai_missense_ASXL1_all = []\n",
    "for i in Desai_miss_ASXL1:\n",
    "    Desai_missense_ASXL1_all.append(str('p.')+i[1])\n",
    "    \n",
    "#Excluded regions:\n",
    "excluded_ASXL1_Desai_missense=[]\n",
    "Desai_ASXL1_excluded_missense = mutation_rates_list_missense_variants_ASXL1('Data files/excluded_variants_Desai_2018_ASXL1_missense.csv')\n",
    "for i in Desai_ASXL1_excluded_missense:\n",
    "    excluded_ASXL1_Desai_missense.append(i[3])\n",
    "\n",
    "Desai_missense_ASXL1=[]\n",
    "for i in Desai_missense_ASXL1_all:\n",
    "    if i not in Desai_ASXL1_excluded_missense:\n",
    "        Desai_missense_ASXL1.append(i)\n",
    "        \n",
    "excluded_ASXL1_Desai_nonsense=[]\n",
    "Desai_ASXL1_excluded_nonsense = mutation_rates_list_nonsense_variants_ASXL1('Data files/excluded_variants_Desai_2018_ASXL1_nonsense.csv')\n",
    "for i in Desai_ASXL1_excluded_nonsense:\n",
    "    excluded_ASXL1_Desai_nonsense.append(i[3])\n",
    "\n",
    "Desai_nonsense_ASXL1=[]\n",
    "for i in Desai_nonsense_ASXL1_all:\n",
    "    if i not in Desai_ASXL1_excluded_nonsense:\n",
    "        Desai_nonsense_ASXL1.append(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 nonsense for this study =  1.6147847e-07\n",
      "Total summed mutation rate across TP53 missense for this study =  3.273721457e-06\n",
      "Total summed mutation rate across TP53 missense for this study =  4.730426e-09\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Desai TP53 variants\n",
    "Desai_non_TP53 = mutation_rates_nonsense_range_of_variants_TP53(0, 394)\n",
    "Desai_miss_TP53 = mutation_rates_missense_range_of_variants_TP53(0, 394)\n",
    "\n",
    "#Lists of TP53 variants targetable in Desai (i.e. included in their panel)\n",
    "Desai_nonsense_TP53 = []\n",
    "for i in Desai_non_TP53:\n",
    "    Desai_nonsense_TP53.append(str('p.')+i[1])\n",
    "    \n",
    "Desai_missense_TP53_all = []\n",
    "for i in Desai_miss_TP53:\n",
    "    Desai_missense_TP53_all.append(str('p.')+i[1])\n",
    "    \n",
    "#Excluded regions:\n",
    "excluded_TP53_Desai=[]\n",
    "Desai_TP53_excluded = mutation_rates_list_missense_variants_TP53('Data files/excluded_variants_Desai_2018_TP53.csv')\n",
    "for i in Desai_TP53_excluded:\n",
    "    excluded_TP53_Desai.append(i[3])\n",
    "\n",
    "Desai_missense_TP53=[]\n",
    "for i in Desai_missense_TP53_all:\n",
    "    if i not in Desai_TP53_excluded:\n",
    "        Desai_missense_TP53.append(i)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Acuna-Hidalgo 2017"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  3.741791e-09\n",
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  9.093981e-09\n",
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  1.3048360000000002e-09\n",
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  2.0864156e-08\n",
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  1.4313799999999999e-09\n",
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  2.5605291000000003e-08\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Acuna-Hidalgo DNMT3A variants\n",
    "Acuna_complete_nonsense_DNMT3A_1 = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(874,892)\n",
    "Acuna_complete_nonsense_DNMT3A_2 = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(851,866)\n",
    "Acuna_complete_nonsense_DNMT3A_3 = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(849,857)\n",
    "Acuna_complete_nonsense_DNMT3A_4 = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(758,774)\n",
    "Acuna_complete_nonsense_DNMT3A_5 = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(733,740)\n",
    "Acuna_complete_nonsense_DNMT3A_6 = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(317,334)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  1.987322e-07\n",
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  7.3492762e-08\n",
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  3.745106700000001e-08\n",
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  1.535384e-07\n",
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  6.461601599999999e-08\n",
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  2.0140387e-07\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Acuna-Hidalgo DNMT3A variants\n",
    "Acuna_complete_missense_DNMT3A_1 = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(874,892)\n",
    "Acuna_complete_missense_DNMT3A_2 = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(851,866)\n",
    "Acuna_complete_missense_DNMT3A_3 = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(849,857)\n",
    "Acuna_complete_missense_DNMT3A_4 = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(758,774)\n",
    "Acuna_complete_missense_DNMT3A_5 = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(733,740)\n",
    "Acuna_complete_missense_DNMT3A_6 = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(317,334)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "#Lists of DNMT3A variants targetable in Acunda-Hidalgo (i.e. included in their panel)\n",
    "Acuna_nonsense_DNMT3A = []\n",
    "for i in Acuna_complete_nonsense_DNMT3A_1:\n",
    "    Acuna_nonsense_DNMT3A.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_DNMT3A_2:\n",
    "    Acuna_nonsense_DNMT3A.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_DNMT3A_3:\n",
    "    Acuna_nonsense_DNMT3A.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_DNMT3A_4:\n",
    "    Acuna_nonsense_DNMT3A.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_DNMT3A_5:\n",
    "    Acuna_nonsense_DNMT3A.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_DNMT3A_6:\n",
    "    Acuna_nonsense_DNMT3A.append(str('p.')+i[1])\n",
    "    \n",
    "Acuna_missense_DNMT3A = []\n",
    "for i in Acuna_complete_missense_DNMT3A_1:\n",
    "    Acuna_missense_DNMT3A.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_DNMT3A_2:\n",
    "    Acuna_missense_DNMT3A.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_DNMT3A_3:\n",
    "    Acuna_missense_DNMT3A.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_DNMT3A_4:\n",
    "    Acuna_missense_DNMT3A.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_DNMT3A_5:\n",
    "    Acuna_missense_DNMT3A.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_DNMT3A_6:\n",
    "    Acuna_missense_DNMT3A.append(str('p.')+i[1])\n",
    "    \n",
    "for i in ['p.Y874S', 'p.Y874F', 'p.Y874C', 'p.F851Y', 'p.F851S', 'p.F851C', 'p.F851L', 'p.I858L', 'p.I858F', 'p.I858V', 'p.I858N', 'p.I858T', 'p.I858S',\\\n",
    "'p.A741T', 'p.A741P', 'p.A741S', 'p.A741E', 'p.A741V', 'p.A741G', 'p.E733A', 'p.E733V', 'p.E733G', 'p.E733D']:\n",
    "    Acuna_missense_DNMT3A.append(i)\n",
    "    \n",
    "for i in ['p.Y874*', 'p.E733*']:\n",
    "    Acuna_nonsense_DNMT3A.append(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TET2 nonsense_NP_01120680 for this study =  5.5247350000000006e-09\n",
      "Total summed mutation rate across TET2 nonsense_NP_01120680 for this study =  1.249072e-08\n",
      "Total summed mutation rate across TET2 nonsense_NP_01120680 for this study =  3.501569e-08\n",
      "Total summed mutation rate across TET2 nonsense_NP_01120680 for this study =  9.645627e-09\n",
      "Total summed mutation rate across TET2 nonsense_NP_01120680 for this study =  3.998035e-09\n",
      "Total summed mutation rate across TET2 nonsense_NP_01120680 for this study =  7.165286e-09\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Acuna-Hidalgo TET2 variants\n",
    "Acuna_complete_nonsense_TET2_1 = mutation_rates_nonsense_range_of_variants_TET2(527, 534)\n",
    "Acuna_complete_nonsense_TET2_2 = mutation_rates_nonsense_range_of_variants_TET2(541, 546)\n",
    "Acuna_complete_nonsense_TET2_3 = mutation_rates_nonsense_range_of_variants_TET2(547, 552)\n",
    "Acuna_complete_nonsense_TET2_4 = mutation_rates_nonsense_range_of_variants_TET2(754, 773)\n",
    "Acuna_complete_nonsense_TET2_5 = mutation_rates_nonsense_range_of_variants_TET2(844, 849)\n",
    "Acuna_complete_nonsense_TET2_6 = mutation_rates_nonsense_range_of_variants_TET2(885, 890)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TET2 missense_NP_01120680 for this study =  3.135999e-08\n",
      "Total summed mutation rate across TET2 missense_NP_01120680 for this study =  3.5936637999999997e-08\n",
      "Total summed mutation rate across TET2 missense_NP_01120680 for this study =  3.3724941999999994e-08\n",
      "Total summed mutation rate across TET2 missense_NP_01120680 for this study =  9.4240322e-08\n",
      "Total summed mutation rate across TET2 missense_NP_01120680 for this study =  2.3836145e-08\n",
      "Total summed mutation rate across TET2 missense_NP_01120680 for this study =  2.016166e-08\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Acuna-Hidalgo TET2 variants\n",
    "Acuna_complete_missense_TET2_1 = mutation_rates_missense_range_of_variants_TET2(527, 534)\n",
    "Acuna_complete_missense_TET2_2 = mutation_rates_missense_range_of_variants_TET2(541, 546)\n",
    "Acuna_complete_missense_TET2_3 = mutation_rates_missense_range_of_variants_TET2(547, 552)\n",
    "Acuna_complete_missense_TET2_4 = mutation_rates_missense_range_of_variants_TET2(754, 773)\n",
    "Acuna_complete_missense_TET2_5 = mutation_rates_missense_range_of_variants_TET2(844, 849)\n",
    "Acuna_complete_missense_TET2_6 = mutation_rates_missense_range_of_variants_TET2(885, 890)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {
    "code_folding": []
   },
   "outputs": [],
   "source": [
    "#Lists of TET2 variants targetable in Acunda-Hidalgo (i.e. included in their panel)\n",
    "Acuna_nonsense_TET2 = []\n",
    "for i in Acuna_complete_nonsense_TET2_1:\n",
    "    Acuna_nonsense_TET2.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_TET2_2:\n",
    "    Acuna_nonsense_TET2.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_TET2_3:\n",
    "    Acuna_nonsense_TET2.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_TET2_4:\n",
    "    Acuna_nonsense_TET2.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_TET2_5:\n",
    "    Acuna_nonsense_TET2.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_TET2_6:\n",
    "    Acuna_nonsense_TET2.append(str('p.')+i[1])\n",
    "    \n",
    "Acuna_missense_TET2 = []\n",
    "for i in Acuna_complete_missense_TET2_1:\n",
    "    Acuna_missense_TET2.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_TET2_2:\n",
    "    Acuna_missense_TET2.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_TET2_3:\n",
    "    Acuna_missense_TET2.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_TET2_4:\n",
    "    Acuna_missense_TET2.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_TET2_5:\n",
    "    Acuna_missense_TET2.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_TET2_6:\n",
    "    Acuna_missense_TET2.append(str('p.')+i[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across ASXL1 nonsense for this study =  3.618524e-08\n",
      "Total summed mutation rate across ASXL1 nonsense for this study =  5.3296666999999994e-08\n",
      "Total summed mutation rate across ASXL1 nonsense for this study =  1.1963488e-08\n",
      "Total summed mutation rate across ASXL1 nonsense for this study =  2.2857128e-08\n",
      "Total summed mutation rate across ASXL1 nonsense for this study =  4.384787e-09\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Acuna-Hidalgo ASXL1 variants\n",
    "Acuna_complete_nonsense_ASXL1_1 = mutation_rates_nonsense_range_of_variants_ASXL1(400, 407)\n",
    "Acuna_complete_nonsense_ASXL1_2 = mutation_rates_nonsense_range_of_variants_ASXL1(413, 573)\n",
    "Acuna_complete_nonsense_ASXL1_3 = mutation_rates_nonsense_range_of_variants_ASXL1(572, 596)\n",
    "Acuna_complete_nonsense_ASXL1_4 = mutation_rates_nonsense_range_of_variants_ASXL1(688, 694)\n",
    "Acuna_complete_nonsense_ASXL1_5 = mutation_rates_nonsense_range_of_variants_ASXL1(799, 804)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across ASXL1 missense for this study =  5.904737100000001e-08\n",
      "Total summed mutation rate across ASXL1 missense for this study =  1.1949677830000001e-06\n",
      "Total summed mutation rate across ASXL1 missense for this study =  2.43229732e-07\n",
      "Total summed mutation rate across ASXL1 missense for this study =  3.6595521e-08\n",
      "Total summed mutation rate across ASXL1 missense for this study =  2.3812267999999998e-08\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Acuna-Hidalgo ASXL1 variants\n",
    "Acuna_complete_missense_ASXL1_1 = mutation_rates_missense_range_of_variants_ASXL1(400, 407)\n",
    "Acuna_complete_missense_ASXL1_2 = mutation_rates_missense_range_of_variants_ASXL1(413, 573)\n",
    "Acuna_complete_missense_ASXL1_3 = mutation_rates_missense_range_of_variants_ASXL1(572, 596)\n",
    "Acuna_complete_missense_ASXL1_4 = mutation_rates_missense_range_of_variants_ASXL1(688, 694)\n",
    "Acuna_complete_missense_ASXL1_5 = mutation_rates_missense_range_of_variants_ASXL1(799, 804)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#Lists of ASXL1 variants targetable in Acunda-Hidalgo (i.e. included in their panel)\n",
    "Acuna_nonsense_ASXL1 = []\n",
    "for i in Acuna_complete_nonsense_ASXL1_1:\n",
    "    Acuna_nonsense_ASXL1.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_ASXL1_2:\n",
    "    Acuna_nonsense_ASXL1.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_ASXL1_3:\n",
    "    Acuna_nonsense_ASXL1.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_ASXL1_4:\n",
    "    Acuna_nonsense_ASXL1.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_ASXL1_5:\n",
    "    Acuna_nonsense_ASXL1.append(str('p.')+i[1])\n",
    "    \n",
    "Acuna_missense_ASXL1 = []\n",
    "for i in Acuna_complete_missense_ASXL1_1:\n",
    "    Acuna_missense_ASXL1.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_ASXL1_2:\n",
    "    Acuna_missense_ASXL1.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_ASXL1_3:\n",
    "    Acuna_missense_ASXL1.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_ASXL1_4:\n",
    "    Acuna_missense_ASXL1.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_ASXL1_5:\n",
    "    Acuna_missense_ASXL1.append(str('p.')+i[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 nonsense for this study =  3.2044719e-08\n",
      "Total summed mutation rate across TP53 nonsense for this study =  3.3097040000000003e-09\n",
      "Total summed mutation rate across TP53 nonsense for this study =  1.3440674e-08\n",
      "Total summed mutation rate across TP53 nonsense for this study =  1.485146e-09\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Acuna-Hidalgo TP53 variants\n",
    "Acuna_complete_nonsense_TP53_1 = mutation_rates_nonsense_range_of_variants_TP53(185, 221)\n",
    "Acuna_complete_nonsense_TP53_2 = mutation_rates_nonsense_range_of_variants_TP53(168, 185)\n",
    "Acuna_complete_nonsense_TP53_3 = mutation_rates_nonsense_range_of_variants_TP53(117, 146)\n",
    "Acuna_complete_nonsense_TP53_4 = mutation_rates_nonsense_range_of_variants_TP53(59, 81)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 missense for this study =  2.75197872e-07\n",
      "Total summed mutation rate across TP53 missense for this study =  2.0262653100000002e-07\n",
      "Total summed mutation rate across TP53 missense for this study =  1.9660014099999998e-07\n",
      "Total summed mutation rate across TP53 missense for this study =  1.9621064499999997e-07\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across Acuna-Hidalgo TP53 variants\n",
    "Acuna_complete_missense_TP53_1 = mutation_rates_missense_range_of_variants_TP53(185, 221)\n",
    "Acuna_complete_missense_TP53_2 = mutation_rates_missense_range_of_variants_TP53(168, 185)\n",
    "Acuna_complete_missense_TP53_3 = mutation_rates_missense_range_of_variants_TP53(117, 146)\n",
    "Acuna_complete_missense_TP53_4 = mutation_rates_missense_range_of_variants_TP53(59, 81)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#Lists of TP53 variants targetable in Acunda-Hidalgo (i.e. included in their panel)\n",
    "Acuna_nonsense_TP53 = []\n",
    "for i in Acuna_complete_nonsense_TP53_1:\n",
    "    Acuna_nonsense_TP53.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_TP53_2:\n",
    "    Acuna_nonsense_TP53.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_TP53_3:\n",
    "    Acuna_nonsense_TP53.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_nonsense_TP53_4:\n",
    "    Acuna_nonsense_TP53.append(str('p.')+i[1])\n",
    "    \n",
    "Acuna_missense_TP53_all = []\n",
    "for i in Acuna_complete_missense_TP53_1:\n",
    "    Acuna_missense_TP53_all.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_TP53_2:\n",
    "    Acuna_missense_TP53_all.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_TP53_3:\n",
    "    Acuna_missense_TP53_all.append(str('p.')+i[1])\n",
    "for i in Acuna_complete_missense_TP53_4:\n",
    "    Acuna_missense_TP53_all.append(str('p.')+i[1])\n",
    "    \n",
    "#Excluded regions:\n",
    "Acuna_TP53_excluded=['p.P72R']\n",
    "\n",
    "Acuna_missense_TP53=[]\n",
    "for i in Acuna_missense_TP53_all:\n",
    "    if i not in Acuna_TP53_excluded:\n",
    "        Acuna_missense_TP53.append(i)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Lists of all possible variants in DNMT3A, TET2, ASXL1, TP53"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Next, create lists of ALL possible variants in DNMT3A, TET2, ASXL1, TP53 so can look at each of them in turn and see how often they are observed and then use this to calculate what fitness effect would explain this number of observations.\n",
    "- The mutation rate calculator functions also produce a list of all the possible variants so use this function to generate a list of all the possible variants within each gene."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DNMT3A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across DNMT3A nonsense_NP_072046 for this study =  3.8678204400000003e-07\n",
      "Total summed mutation rate across DNMT3A missense_NP_072046 for this study =  8.083470887e-06\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across DNMT3A\n",
    "all_nonsense_DNMT3A = mutation_rates_nonsense_range_of_variants_DNMT3A_NP072046(0, 913)\n",
    "all_missense_DNMT3A = mutation_rates_missense_range_of_variants_DNMT3A_NP072046(0, 913)\n",
    "\n",
    "#List of possible variants\n",
    "all_nonsense_variants_DNMT3A = []\n",
    "for i in all_nonsense_DNMT3A:\n",
    "    all_nonsense_variants_DNMT3A.append(str('p.')+i[1])\n",
    "    \n",
    "all_missense_variants_DNMT3A = []\n",
    "for i in all_missense_DNMT3A:\n",
    "    all_missense_variants_DNMT3A.append(str('p.')+i[1])\n",
    "\n",
    "all_missense_variants_DNMT3A = list(dict.fromkeys(all_missense_variants_DNMT3A))\n",
    "all_nonsense_variants_DNMT3A = list(dict.fromkeys(all_nonsense_variants_DNMT3A))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TET2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TET2 nonsense_NP_01120680 for this study =  8.664021850000001e-07\n",
      "Total summed mutation rate across TET2 missense_NP_01120680 for this study =  1.2216578715e-05\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across TET2\n",
    "all_nonsense_TET2 = mutation_rates_nonsense_range_of_variants_TET2(0, 2003)\n",
    "all_missense_TET2 = mutation_rates_missense_range_of_variants_TET2(0, 2003)\n",
    "\n",
    "#List of possible variants\n",
    "all_nonsense_variants_TET2 = []\n",
    "for i in all_nonsense_TET2:\n",
    "    all_nonsense_variants_TET2.append(str('p.')+i[1])\n",
    "    \n",
    "all_missense_variants_TET2 = []\n",
    "for i in all_missense_TET2:\n",
    "    all_missense_variants_TET2.append(str('p.')+i[1])\n",
    "\n",
    "all_missense_variants_TET2 = list(dict.fromkeys(all_missense_variants_TET2))\n",
    "all_nonsense_variants_TET2 = list(dict.fromkeys(all_nonsense_variants_TET2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ASXL1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across ASXL1 nonsense for this study =  6.254075530000001e-07\n",
      "Total summed mutation rate across ASXL1 missense for this study =  1.1360431228e-05\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across TET2\n",
    "all_nonsense_ASXL1 = mutation_rates_nonsense_range_of_variants_ASXL1(0, 1542)\n",
    "all_missense_ASXL1 = mutation_rates_missense_range_of_variants_ASXL1(0, 1542)\n",
    "\n",
    "#List of possible variants\n",
    "all_nonsense_variants_ASXL1 = []\n",
    "for i in all_nonsense_ASXL1:\n",
    "    all_nonsense_variants_ASXL1.append(str('p.')+i[1])\n",
    "    \n",
    "all_missense_variants_ASXL1 = []\n",
    "for i in all_missense_ASXL1:\n",
    "    all_missense_variants_ASXL1.append(str('p.')+i[1])\n",
    "\n",
    "all_missense_variants_ASXL1 = list(dict.fromkeys(all_missense_variants_ASXL1))\n",
    "all_nonsense_variants_ASXL1 = list(dict.fromkeys(all_nonsense_variants_ASXL1))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TP53"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total summed mutation rate across TP53 nonsense for this study =  1.6147847e-07\n",
      "Total summed mutation rate across TP53 missense for this study =  3.273721457e-06\n"
     ]
    }
   ],
   "source": [
    "#Mutation rate across TP53\n",
    "all_nonsense_TP53 = mutation_rates_nonsense_range_of_variants_TP53(0, 394)\n",
    "all_missense_TP53 = mutation_rates_missense_range_of_variants_TP53(0, 394)\n",
    "\n",
    "#List of possible variants\n",
    "all_nonsense_variants_TP53 = []\n",
    "for i in all_nonsense_TP53:\n",
    "    all_nonsense_variants_TP53.append(str('p.')+i[1])\n",
    "    \n",
    "all_missense_variants_TP53 = []\n",
    "for i in all_missense_TP53:\n",
    "    all_missense_variants_TP53.append(str('p.')+i[1])\n",
    "\n",
    "all_missense_variants_TP53 = list(dict.fromkeys(all_missense_variants_TP53))\n",
    "all_nonsense_variants_TP53 = list(dict.fromkeys(all_nonsense_variants_TP53))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Actual number of observations of each variant"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- Look at each variant seen across the studies and count how many times it was observed."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DNMT3A"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#Create lists of variants by studies and counts of each variant\n",
    "import csv\n",
    "filename=\"Data files/all_studies_trimmed_all_genes.csv\"\n",
    "\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    DNMT3A_results={} #dictionary of DNMTA results (key = ID (row[13]), value = VAF (row[10]))\n",
    "\n",
    "    for row in read_reader:\n",
    "        if row_count>0:\n",
    "#             print(row)\n",
    "            if row[3] == 'DNMT3A':\n",
    "#                 print(row)\n",
    "                VAF = row[0]\n",
    "                age = row[1]\n",
    "                variant = str('p.')+row[2]\n",
    "                study = row[4]\n",
    "                DNMT3A_results[row_count]=(variant, study)\n",
    "                \n",
    "        row_count=row_count+1\n",
    "        \n",
    "all_DNMT3A_results_list = []\n",
    "for k, v in DNMT3A_results.items():\n",
    "    all_DNMT3A_results_list.append(v[0])\n",
    "        \n",
    "all_DNMT3A_df = pd.DataFrame({'variant':all_DNMT3A_results_list})\n",
    "all_grouped_df=all_DNMT3A_df.groupby(['variant']).size().reset_index(name='counts')\n",
    "all_grouped_df.to_csv('Data files/all_DNMT3A_variant_counts.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#All_DNMT3A_counts\n",
    "filename=\"Data files/all_DNMT3A_variant_counts.csv\"\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    all_studies_DNMT3A_counts={} #dictionary of DNMTA results (key = ID (row[13]), value = VAF (row[10]))\n",
    "\n",
    "    for row in read_reader:\n",
    "#         print(row)\n",
    "        if row_count>0:\n",
    "            variant = row[1]\n",
    "            counts = row[2]\n",
    "            all_studies_DNMT3A_counts[variant]=counts\n",
    "                \n",
    "        row_count=row_count+1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TET2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#Create lists of variants by studies and counts of each variant\n",
    "import csv\n",
    "filename=\"Data files/all_studies_trimmed_all_genes.csv\"\n",
    "\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    TET2_results={} #dictionary of DNMTA results (key = ID (row[13]), value = VAF (row[10]))\n",
    "\n",
    "    for row in read_reader:\n",
    "        if row_count>0:\n",
    "#             print(row)\n",
    "            if row[3] == 'TET2':\n",
    "#                 print(row)\n",
    "                VAF = row[0]\n",
    "                age = row[1]\n",
    "                variant = str('p.')+row[2]\n",
    "                study = row[4]\n",
    "                TET2_results[row_count]=(variant, study)\n",
    "                \n",
    "        row_count=row_count+1\n",
    "        \n",
    "all_TET2_results_list = []\n",
    "for k, v in TET2_results.items():\n",
    "    all_TET2_results_list.append(v[0])\n",
    "\n",
    "all_TET2_df = pd.DataFrame({'variant':all_TET2_results_list})\n",
    "all_grouped_df=all_TET2_df.groupby(['variant']).size().reset_index(name='counts')\n",
    "all_grouped_df.to_csv('Data files/all_TET2_variant_counts.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#All_TET2_counts\n",
    "filename=\"Data files/all_TET2_variant_counts.csv\"\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    all_studies_TET2_counts={} #dictionary of DNMTA results (key = ID (row[13]), value = VAF (row[10]))\n",
    "\n",
    "    for row in read_reader:\n",
    "#         print(row)\n",
    "        if row_count>0:\n",
    "            variant = row[1]\n",
    "            counts = row[2]\n",
    "            all_studies_TET2_counts[variant]=counts\n",
    "                \n",
    "        row_count=row_count+1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### ASXL1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#Create lists of variants by studies and counts of each variant\n",
    "import csv\n",
    "filename=\"Data files/all_studies_trimmed_all_genes.csv\"\n",
    "\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    ASXL1_results={} #dictionary of DNMTA results (key = ID (row[13]), value = VAF (row[10]))\n",
    "\n",
    "    for row in read_reader:\n",
    "        if row_count>0:\n",
    "#             print(row)\n",
    "            if row[3] == 'ASXL1':\n",
    "#                 print(row)\n",
    "                VAF = row[0]\n",
    "                age = row[1]\n",
    "                variant = str('p.')+row[2]\n",
    "                study = row[4]\n",
    "                ASXL1_results[row_count]=(variant, study)\n",
    "                \n",
    "        row_count=row_count+1\n",
    "\n",
    "all_ASXL1_results_list = []\n",
    "\n",
    "for k, v in ASXL1_results.items():\n",
    "    all_ASXL1_results_list.append(v[0])\n",
    "        \n",
    "all_ASXL1_df = pd.DataFrame({'variant':all_ASXL1_results_list})\n",
    "all_grouped_df=all_ASXL1_df.groupby(['variant']).size().reset_index(name='counts')\n",
    "all_grouped_df.to_csv('Data files/all_ASXL1_variant_counts.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#All_ASXL1_counts\n",
    "filename=\"Data files/all_ASXL1_variant_counts.csv\"\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    all_studies_ASXL1_counts={} #dictionary of DNMTA results (key = ID (row[13]), value = VAF (row[10]))\n",
    "\n",
    "    for row in read_reader:\n",
    "#         print(row)\n",
    "        if row_count>0:\n",
    "            variant = row[1]\n",
    "            counts = row[2]\n",
    "            all_studies_ASXL1_counts[variant]=counts\n",
    "                \n",
    "        row_count=row_count+1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### TP53"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#Create lists of variants by studies and counts of each variant\n",
    "import csv\n",
    "filename=\"Data files/all_studies_trimmed_all_genes.csv\"\n",
    "\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    TP53_results={} #dictionary of DNMTA results (key = ID (row[13]), value = VAF (row[10]))\n",
    "\n",
    "    for row in read_reader:\n",
    "        if row_count>0:\n",
    "#             print(row)\n",
    "            if row[3] == 'TP53':\n",
    "#                 print(row)\n",
    "                VAF = row[0]\n",
    "                age = row[1]\n",
    "                variant = str('p.')+row[2]\n",
    "                study = row[4]\n",
    "                TP53_results[row_count]=(variant, study)\n",
    "                \n",
    "        row_count=row_count+1\n",
    "\n",
    "all_TP53_results_list = []\n",
    "for k, v in TP53_results.items():\n",
    "    all_TP53_results_list.append(v[0])\n",
    "        \n",
    "all_TP53_df = pd.DataFrame({'variant':all_TP53_results_list})\n",
    "all_grouped_df=all_TP53_df.groupby(['variant']).size().reset_index(name='counts')\n",
    "all_grouped_df.to_csv('Data files/all_TP53_variant_counts.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#All_TP53_counts\n",
    "filename=\"Data files/all_TP53_variant_counts.csv\"\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    all_studies_TP53_counts={} #dictionary of DNMTA results (key = ID (row[13]), value = VAF (row[10]))\n",
    "\n",
    "    for row in read_reader:\n",
    "#         print(row)\n",
    "        if row_count>0:\n",
    "            variant = row[1]\n",
    "            counts = row[2]\n",
    "            all_studies_TP53_counts[variant]=counts\n",
    "                \n",
    "        row_count=row_count+1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Functions for calculating the expected number of observations of a variant"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- The number of times you expect to see a variant depends on it's fitness effect (s), the size of the study (i.e. number of participants), the ages of the individuals in the study and the study's sequencing limit of detection.\n",
    "- The distribution of ages in the studys is presumed to be ~normally distributed, except for Acuna-Hidalgo, which has a ~uniform distribution of ages."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "def density_of_clones(f, s, t, N, u): \n",
    "    return ((N*u)/(f*(1-2*f)))*np.exp(-f/((1-2*f)*((np.exp(s*t)-1)/(2*N*s))))\n",
    "\n",
    "def prevalence_single_s(fmin, s, t, N, u): #expected prevalence of variant of a particular fitness effect\n",
    "    expected_number_clones=integrate.quad(lambda f: density_of_clones(f, s, t, N, u), fmin, 0.4999)[0]\n",
    "    return 1.0-np.exp(-expected_number_clones)\n",
    "\n",
    "def prevalence_averaged_over_ages(fmin, s, N, u, mean, sigma, study_size): #expected prevalence across range of ages\n",
    "    expected_prev_clones=integrate.quad(lambda t: (prevalence_single_s(fmin, s, t, N, u))*\\\n",
    "                                 ((1/((2*np.pi*(sigma)**2)**0.5))*(np.exp(-(t-mean)**2/(2*(sigma**2))))), mean-2*sigma, mean+2*sigma)\n",
    "    prev = expected_prev_clones[0]\n",
    "    expect_number_clones = prev*study_size\n",
    "    return expect_number_clones\n",
    "\n",
    "def prevalence_averaged_over_uniform_ages(fmin, s, N, u, mean, study_size): #expected prevalence across uniform distribution of ages\n",
    "    expected_prevalence_clones=integrate.quad(lambda t: (prevalence_single_s(fmin, s, t, N, u))*(1/49), 20, 69)\n",
    "    prevalence = expected_prevalence_clones[0]\n",
    "    expected_number_clones = prevalence*study_size\n",
    "    return expected_number_clones"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Maximum Likelihood Estimation for s"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- This involves calculating the most likely fitness effect (s) that would explain the number of observations of the variant."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " **Calculating the expected number of observations of each variant**\n",
    "- To calculate the expected number of observations of each variant, need to look at each variant in turn and for each variant, need to then look at each study in turn and see if that variant is targeted by that study.  If it is, need to work out how many times you expect to see that variant in that study, taking in to account the study's limit of detection, study size and ages of the individuals in the study. Then, once looked at all studies, sum up the number of observations across all the studies to get the overall expected number of observations for that variant.\n",
    "- The expected number also depends on the fitness effect of the variant - the value of s that means the observed number of variants matches as closely as possible the expected number of variants is calculated."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## DNMT3A variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {
    "code_folding": [
     0,
     96,
     107,
     119
    ]
   },
   "outputs": [],
   "source": [
    "def expected_number_variants_missense_DNMT3A(data, params):\n",
    "    s_value = params[0]\n",
    "    \n",
    "    a = 0\n",
    "    expected_number_by_variant_missense = {}\n",
    "    \n",
    "    Jaiswal_fmin = 0.056\n",
    "    Jaiswal_mean = 58\n",
    "    Jaiswal_sigma = 11.5\n",
    "    Jaiswal_total = 17182\n",
    "    \n",
    "    Genovese_fmin = 0.1194\n",
    "    Genovese_mean = 55.1\n",
    "    Genovese_total = 12380\n",
    "    Genovese_sigma = 11.95\n",
    "    \n",
    "    McKerrel_fmin = 0.008\n",
    "    McKerrel_mean = 51.88\n",
    "    McKerrel_total = 4219\n",
    "    McKerrel_sigma = 18.36\n",
    "    \n",
    "    Zink_fmin = 0.2363\n",
    "    Zink_mean = 60\n",
    "    Zink_total = 11262\n",
    "    Zink_sigma = 1.18907720e+01\n",
    "    \n",
    "    Coombs_fmin = 0.019\n",
    "    Coombs_mean = 59.82\n",
    "    Coombs_total = 5649\n",
    "    Coombs_sigma = 15.14\n",
    "    \n",
    "    Young_fmin = 0.0009\n",
    "    Young_mean = 63\n",
    "    Young_total = 89\n",
    "    Young_sigma = 6.8\n",
    "    \n",
    "    Desai_fmin = 0.0169\n",
    "    Desai_mean = 66.1\n",
    "    Desai_total = 424\n",
    "    Desai_sigma = 6 #estimated\n",
    "    \n",
    "    Acuna_fmin = 0.0015\n",
    "    Acuna_mean = 4.5\n",
    "    Acuna_total = 2014\n",
    "    \n",
    "    N = 9.40166610e+04\n",
    "\n",
    "    for i in data:\n",
    "        variant_mutation_rate = mutation_rates_variants_DNMT3A(i)\n",
    "\n",
    "        if i in Jaiswal_missense_DNMT3A:\n",
    "            Jaiswal_number = prevalence_averaged_over_ages(Jaiswal_fmin, s_value, N, variant_mutation_rate, Jaiswal_mean, Jaiswal_sigma, Jaiswal_total)\n",
    "        else:\n",
    "            Jaiswal_number = 0\n",
    "\n",
    "        if i in Genovese_missense_DNMT3A:\n",
    "            Genovese_number = prevalence_averaged_over_ages(Genovese_fmin, s_value, N, variant_mutation_rate, Genovese_mean, Genovese_sigma, Genovese_total)\n",
    "        else:\n",
    "            Genovese_number = 0 \n",
    "\n",
    "        if i in Zink_missense_DNMT3A:\n",
    "            Zink_number = prevalence_averaged_over_ages(Zink_fmin, s_value, N, variant_mutation_rate, Zink_mean, Zink_sigma, Zink_total)\n",
    "        else:\n",
    "            Zink_number = 0\n",
    "            \n",
    "        if i in McKerrel_missense_DNMT3A:\n",
    "            McKerrel_number = prevalence_averaged_over_ages(McKerrel_fmin, s_value, N, variant_mutation_rate, McKerrel_mean, McKerrel_sigma, McKerrel_total)\n",
    "        else:\n",
    "            McKerrel_number = 0\n",
    "\n",
    "        if i in Coombs_missense_DNMT3A:\n",
    "            Coombs_number = prevalence_averaged_over_ages(Coombs_fmin, s_value, N, variant_mutation_rate, Coombs_mean, Coombs_sigma, Coombs_total)\n",
    "        else:\n",
    "            Coombs_number = 0\n",
    "\n",
    "        if i in Young_missense_DNMT3A:\n",
    "            Young_number = prevalence_averaged_over_ages(Young_fmin, s_value, N, variant_mutation_rate, Young_mean, Young_sigma, Young_total)\n",
    "        else:\n",
    "            Young_number = 0\n",
    "\n",
    "        if i in Desai_missense_DNMT3A:\n",
    "            Desai_number = prevalence_averaged_over_ages(Desai_fmin, s_value, N, variant_mutation_rate, Desai_mean, Desai_sigma, Desai_total)\n",
    "        else:\n",
    "            Desai_number = 0\n",
    "\n",
    "        if i in Acuna_missense_DNMT3A:\n",
    "            Acuna_number = prevalence_averaged_over_uniform_ages(Acuna_fmin, s_value, N, variant_mutation_rate, Acuna_mean, Acuna_total)\n",
    "        else:\n",
    "            Acuna_number = 0\n",
    "\n",
    "        expected_total = Jaiswal_number + Genovese_number + Zink_number + McKerrel_number + Coombs_number + Young_number + Desai_number + Acuna_number\n",
    "\n",
    "        expected_number_by_variant_missense[i] = expected_total\n",
    "    \n",
    "    return expected_number_by_variant_missense\n",
    "\n",
    "def prob_observed_given_expected_missense_DNMT3A(params, data): #data = list of variant (e.g. p.R882H) (or all_missense_variants_DNMT3A)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_DNMT3A_counts[variant])\n",
    "        expected_number = float(expected_number_variants_missense_DNMT3A(data, params)[variant])\n",
    "        square_distance = ((observed_number - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   \n",
    "\n",
    "def prob_observed_given_expected_missense_DNMT3A_upper_error(params, data): #data = list of variant (e.g. p.R882H) (or all_missense_variants_DNMT3A)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_DNMT3A_counts[variant])\n",
    "        observed_number_upper_error = observed_number+np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_missense_DNMT3A(data, params)[variant])\n",
    "        square_distance = ((observed_number_upper_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   \n",
    "\n",
    "def prob_observed_given_expected_missense_DNMT3A_lower_error(params, data): #data = list of variant (e.g. p.R882H) (or all_missense_variants_DNMT3A)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_DNMT3A_counts[variant])\n",
    "        observed_number_lower_error = observed_number-np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_missense_DNMT3A(data, params)[variant])\n",
    "        square_distance = ((observed_number_lower_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "metadata": {
    "code_folding": [
     0,
     91,
     102,
     114
    ]
   },
   "outputs": [],
   "source": [
    "def expected_number_variants_nonsense_DNMT3A(data, params):\n",
    "    s_value = params[0]\n",
    "    \n",
    "    a = 0\n",
    "    expected_number_by_variant_nonsense = {}\n",
    "    \n",
    "    Jaiswal_fmin = 0.056\n",
    "    Jaiswal_mean = 58\n",
    "    Jaiswal_sigma = 11.5\n",
    "    Jaiswal_total = 17182\n",
    "    \n",
    "    Genovese_fmin = 0.1194\n",
    "    Genovese_mean = 55.1\n",
    "    Genovese_total = 12380\n",
    "    Genovese_sigma = 11.95\n",
    "    \n",
    "    McKerrel_fmin = 0.008\n",
    "    McKerrel_mean = 51.88\n",
    "    McKerrel_total = 4219\n",
    "    McKerrel_sigma = 18.36\n",
    "    \n",
    "    Zink_fmin = 0.2363\n",
    "    Zink_mean = 60\n",
    "    Zink_total = 11262\n",
    "    Zink_sigma = 1.18907720e+01\n",
    "    \n",
    "    Coombs_fmin = 0.019\n",
    "    Coombs_mean = 59.82\n",
    "    Coombs_total = 5649\n",
    "    Coombs_sigma = 15.14\n",
    "    \n",
    "    Young_fmin = 0.0009\n",
    "    Young_mean = 63\n",
    "    Young_total = 89\n",
    "    Young_sigma = 6.8\n",
    "    \n",
    "    Desai_fmin = 0.0169\n",
    "    Desai_mean = 66.1\n",
    "    Desai_total = 424\n",
    "    Desai_sigma = 6 #estimated\n",
    "    \n",
    "    Acuna_fmin = 0.0015\n",
    "    Acuna_mean = 4.5\n",
    "    Acuna_total = 2014\n",
    "    \n",
    "    N = 9.40166610e+04\n",
    "\n",
    "    for i in data:\n",
    "        variant_mutation_rate = mutation_rates_variants_DNMT3A(i)\n",
    "\n",
    "        if i in Jaiswal_nonsense_DNMT3A:\n",
    "            Jaiswal_number = prevalence_averaged_over_ages(Jaiswal_fmin, s_value, N, variant_mutation_rate, Jaiswal_mean, Jaiswal_sigma, Jaiswal_total)\n",
    "        else:\n",
    "            Jaiswal_number = 0\n",
    "\n",
    "        if i in Genovese_nonsense_DNMT3A:\n",
    "            Genovese_number = prevalence_averaged_over_ages(Genovese_fmin, s_value, N, variant_mutation_rate, Genovese_mean, Genovese_sigma, Genovese_total)\n",
    "        else:\n",
    "            Genovese_number = 0 \n",
    "\n",
    "        if i in Zink_nonsense_DNMT3A:\n",
    "            Zink_number = prevalence_averaged_over_ages(Zink_fmin, s_value, N, variant_mutation_rate, Zink_mean, Zink_sigma, Zink_total)\n",
    "        else:\n",
    "            Zink_number = 0\n",
    "\n",
    "        if i in Coombs_nonsense_DNMT3A:\n",
    "            Coombs_number = prevalence_averaged_over_ages(Coombs_fmin, s_value, N, variant_mutation_rate, Coombs_mean, Coombs_sigma, Coombs_total)\n",
    "        else:\n",
    "            Coombs_number = 0\n",
    "\n",
    "        if i in Young_nonsense_DNMT3A:\n",
    "            Young_number = prevalence_averaged_over_ages(Young_fmin, s_value, N, variant_mutation_rate, Young_mean, Young_sigma, Young_total)\n",
    "        else:\n",
    "            Young_number = 0\n",
    "\n",
    "        if i in Desai_nonsense_DNMT3A:\n",
    "            Desai_number = prevalence_averaged_over_ages(Desai_fmin, s_value, N, variant_mutation_rate, Desai_mean, Desai_sigma, Desai_total)\n",
    "        else:\n",
    "            Desai_number = 0\n",
    "\n",
    "        if i in Acuna_nonsense_DNMT3A:\n",
    "            Acuna_number = prevalence_averaged_over_uniform_ages(Acuna_fmin, s_value, N, variant_mutation_rate, Acuna_mean, Acuna_total)\n",
    "        else:\n",
    "            Acuna_number = 0\n",
    "\n",
    "        expected_total = Jaiswal_number + Genovese_number + Zink_number + Coombs_number + Young_number + Desai_number + Acuna_number\n",
    "\n",
    "        expected_number_by_variant_nonsense[i] = expected_total\n",
    "    \n",
    "    return expected_number_by_variant_nonsense\n",
    "\n",
    "def prob_observed_given_expected_nonsense_DNMT3A(params, data): #data = list of variant (e.g. p.R882H) (or all_nonsense_variants_DNMT3A)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_DNMT3A_counts[variant])\n",
    "        expected_number = float(expected_number_variants_nonsense_DNMT3A(data, params)[variant])\n",
    "        square_distance = ((observed_number - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance\n",
    "\n",
    "def prob_observed_given_expected_nonsense_DNMT3A_upper_error(params, data): #data = list of variant (e.g. p.R882H) (or all_nonsense_variants_DNMT3A)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_DNMT3A_counts[variant])\n",
    "        observed_number_upper_error = observed_number+np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_nonsense_DNMT3A(data, params)[variant])\n",
    "        square_distance = ((observed_number_upper_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   \n",
    "\n",
    "def prob_observed_given_expected_nonsense_DNMT3A_lower_error(params, data): #data = list of variant (e.g. p.R882H) (or all_nonsense_variants_DNMT3A)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_DNMT3A_counts[variant])\n",
    "        observed_number_lower_error = observed_number-np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_nonsense_DNMT3A(data, params)[variant])\n",
    "        square_distance = ((observed_number_lower_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/anaconda3/lib/python3.7/site-packages/scipy/integrate/quadpack.py:385: IntegrationWarning: The integral is probably divergent, or slowly convergent.\n",
      "  warnings.warn(msg, IntegrationWarning)\n"
     ]
    }
   ],
   "source": [
    "#CALCULATING S FOR MISSENSE VARIANTS\n",
    "DNMT3A_missense_s = {}\n",
    "a = 0\n",
    "for var in all_missense_variants_DNMT3A: #iterate over all the possible missense variants in DNMT3A\n",
    "    initial_guess=[0.05] #initial guess for s = 0.05 (i.e. 5%)\n",
    "    if var in all_studies_DNMT3A_counts.keys(): #look to see if the variant is seen across any of the studies\n",
    "        observed_number = float(all_studies_DNMT3A_counts[var]) #if it is, retrieve how many times it was observed\n",
    "#         print(observed_number)\n",
    "        if observed_number >1: #only calculate s for variants seen more than 1 time\n",
    "            variant_mutation_rate = mutation_rates_variants_DNMT3A(var) #trinucleotide-context dependent mutation rate for the variant\n",
    "            \n",
    "            MLE = scipy.optimize.minimize(prob_observed_given_expected_missense_DNMT3A, initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value = MLE.x[0]\n",
    "\n",
    "            upper_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_missense_DNMT3A_upper_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_upper = upper_error_MLE.x[0]\n",
    "\n",
    "            lower_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_missense_DNMT3A_lower_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_lower = lower_error_MLE.x[0]\n",
    "\n",
    "#             print('a = ', a)\n",
    "            #Create a dictionary where key = variant\n",
    "            DNMT3A_missense_s[var]=(s_value, observed_number, variant_mutation_rate, s_value_upper, s_value_lower) \n",
    "    a = a+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#CALCULATING S FOR NONSENSE VARIANTS\n",
    "DNMT3A_nonsense_s = {}\n",
    "a = 0\n",
    "for var in all_nonsense_variants_DNMT3A: #iterate over all the possible nonsense variants in DNMT3A\n",
    "    initial_guess=[0.05] #initial guess for s = 0.05 (i.e. 5%)\n",
    "    if var in all_studies_DNMT3A_counts.keys(): #look to see if the variant is seen across any of the studies\n",
    "        observed_number = float(all_studies_DNMT3A_counts[var]) #if it is, retrieve how many times it was observed\n",
    "#         print(observed_number)\n",
    "        if observed_number >1: #only calculate s for variants seen more than 1 time\n",
    "            variant_mutation_rate = mutation_rates_variants_DNMT3A(var) #trinucleotide-context dependent mutation rate for the variant\n",
    "            \n",
    "            MLE = scipy.optimize.minimize(prob_observed_given_expected_nonsense_DNMT3A, initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value = MLE.x[0]\n",
    "\n",
    "            upper_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_nonsense_DNMT3A_upper_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_upper = upper_error_MLE.x[0]\n",
    "\n",
    "            lower_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_nonsense_DNMT3A_lower_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_lower = lower_error_MLE.x[0]\n",
    "\n",
    "#             print('a = ', a)\n",
    "            DNMT3A_nonsense_s[var]=(s_value, observed_number, variant_mutation_rate, s_value_upper, s_value_lower) \n",
    "    a = a+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#Exac - retrieve the frequencies in Exac for each of the variants\n",
    "import csv\n",
    "filename=\"Data files/DNMT3A_Exac.csv\"\n",
    "\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    Exac_DNMT3A_results={} \n",
    "    \n",
    "    for row in read_reader:\n",
    "        if row_count>1:\n",
    "#             print(row)\n",
    "            variant = row[6]\n",
    "            allele_freq = float(row[14])\n",
    "            african_freq = float(row[15])/float(row[16])\n",
    "            east_asian_freq = float(row[18])/float(row[19])\n",
    "            european_freq = float(row[21])/float(row[22])\n",
    "            finnish_freq = float(row[24])/float(row[25])\n",
    "            latino_freq = float(row[27])/float(row[28])\n",
    "            south_asian_freq = float(row[33])/float(row[34])\n",
    "            Exac_DNMT3A_results[variant]=(allele_freq, african_freq, east_asian_freq, european_freq, finnish_freq, latino_freq, south_asian_freq)\n",
    "\n",
    "        row_count=row_count+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#COSMIC - retrieve the frequencies in COSMIC for each of the variants\n",
    "import csv\n",
    "filename=\"Data files/DNMT3A_COSMIC_variant_counts.csv\"\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    DNMT3A_COSMIC_results={} #dictionary of DNMTA results (key = ID (row[13]), value = VAF (row[10]))\n",
    "\n",
    "    for row in read_reader:\n",
    "        if row_count>1:\n",
    "            variant = row[1]\n",
    "            counts = int(row[2])\n",
    "            DNMT3A_COSMIC_results[variant]=counts\n",
    "\n",
    "        row_count=row_count+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Create dictionary of results and save to .csv file\n",
    "DNMT3A_s_dictionary={}\n",
    "for k, v in DNMT3A_missense_s.items():\n",
    "#     print(k)\n",
    "    if k in Exac_DNMT3A_results.keys():\n",
    "        exac_freq = round((Exac_DNMT3A_results[k][0])*100, 4)\n",
    "        african_freq = round((Exac_DNMT3A_results[k][1])*100, 4)\n",
    "        east_asian_freq = round((Exac_DNMT3A_results[k][2])*100, 4)\n",
    "        european_freq = round((Exac_DNMT3A_results[k][3])*100, 4)\n",
    "        finnish_freq = round((Exac_DNMT3A_results[k][4])*100, 4)\n",
    "        latino_freq = round((Exac_DNMT3A_results[k][5])*100, 4)\n",
    "        south_asian_freq = round((Exac_DNMT3A_results[k][6])*100, 4)\n",
    "    if k not in Exac_DNMT3A_results.keys():\n",
    "        exac_freq = '-'\n",
    "        african_freq = '-'\n",
    "        east_asian_freq = '-'\n",
    "        european_freq = '-'\n",
    "        finnish_freq = '-'\n",
    "        latino_freq = '-'\n",
    "        south_asian_freq = '-'\n",
    "    if k in DNMT3A_COSMIC_results.keys():\n",
    "        COSMIC_number = DNMT3A_COSMIC_results[k]\n",
    "    if k not in DNMT3A_COSMIC_results.keys():\n",
    "        COSMIC_number = '-'\n",
    "    DNMT3A_s_dictionary[k]=(round(v[0]*100, 2),round(v[4]*100, 2),round(v[3]*100, 2), v[1], v[2], COSMIC_number, exac_freq, \\\n",
    "                            african_freq, east_asian_freq, european_freq, finnish_freq, latino_freq, south_asian_freq)\n",
    "\n",
    "for k, v in DNMT3A_nonsense_s.items():\n",
    "    if k in Exac_DNMT3A_results.keys():\n",
    "        exac_freq = round((Exac_DNMT3A_results[k][0])*100, 4)\n",
    "        african_freq = round((Exac_DNMT3A_results[k][1])*100, 4)\n",
    "        east_asian_freq = round((Exac_DNMT3A_results[k][2])*100, 4)\n",
    "        european_freq = round((Exac_DNMT3A_results[k][3])*100, 4)\n",
    "        finnish_freq = round((Exac_DNMT3A_results[k][4])*100, 4)\n",
    "        latino_freq = round((Exac_DNMT3A_results[k][5])*100, 4)\n",
    "        south_asian_freq = round((Exac_DNMT3A_results[k][6])*100, 4)\n",
    "    if k not in Exac_DNMT3A_results.keys():\n",
    "        exac_freq = '-'\n",
    "        african_freq = '-'\n",
    "        east_asian_freq = '-'\n",
    "        european_freq = '-'\n",
    "        finnish_freq = '-'\n",
    "        latino_freq = '-'\n",
    "        south_asian_freq = '-'\n",
    "    if k in DNMT3A_COSMIC_results.keys():\n",
    "        COSMIC_number = DNMT3A_COSMIC_results[k]\n",
    "    if k not in DNMT3A_COSMIC_results.keys():\n",
    "        COSMIC_number = '-'\n",
    "    DNMT3A_s_dictionary[k]=(round(v[0]*100, 2),round(v[4]*100, 2),round(v[3]*100, 2), v[1], v[2], COSMIC_number, exac_freq, \\\n",
    "                            african_freq, east_asian_freq, european_freq, finnish_freq, latino_freq, south_asian_freq)\n",
    "\n",
    "DNMT3A_df = pd.DataFrame.from_dict(DNMT3A_s_dictionary, orient='index', columns=['s (%)', 's lower value (%)', 's upper value (%)','observed number', 'site-specific mutation rate', 'Number of times in COSMIC',\\\n",
    "                                                                                'Exac Overall freq (%)', 'Exac African freq (%)', 'Exac East Asian freq (%)',\\\n",
    "                                                                                'Exac European freq (%)', 'Exac Finnish freq (%)', 'Exac Latino freq (%)', 'Exac South Asian freq (%)'])\n",
    "DNMT3A_df_sorted = DNMT3A_df.sort_values(by = ['s (%)'], ascending = False)\n",
    "\n",
    "DNMT3A_df_sorted.to_csv('Data files/DNMT3A_s_across_whole_gene.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s (%)</th>\n",
       "      <th>s lower value (%)</th>\n",
       "      <th>s upper value (%)</th>\n",
       "      <th>observed number</th>\n",
       "      <th>site-specific mutation rate</th>\n",
       "      <th>Number of times in COSMIC</th>\n",
       "      <th>Exac Overall freq (%)</th>\n",
       "      <th>Exac African freq (%)</th>\n",
       "      <th>Exac East Asian freq (%)</th>\n",
       "      <th>Exac European freq (%)</th>\n",
       "      <th>Exac Finnish freq (%)</th>\n",
       "      <th>Exac Latino freq (%)</th>\n",
       "      <th>Exac South Asian freq (%)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>p.E30A</th>\n",
       "      <td>196.75</td>\n",
       "      <td>28.05</td>\n",
       "      <td>196.75</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.863150e-10</td>\n",
       "      <td>3</td>\n",
       "      <td>0.2959</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0</td>\n",
       "      <td>0.3033</td>\n",
       "      <td>2.1424</td>\n",
       "      <td>0.1413</td>\n",
       "      <td>0.0247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.F751C</th>\n",
       "      <td>30.46</td>\n",
       "      <td>18.06</td>\n",
       "      <td>196.75</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.574550e-10</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.Y735C</th>\n",
       "      <td>26.10</td>\n",
       "      <td>23.52</td>\n",
       "      <td>29.07</td>\n",
       "      <td>23.0</td>\n",
       "      <td>8.813490e-10</td>\n",
       "      <td>12</td>\n",
       "      <td>0.0118</td>\n",
       "      <td>0.0097</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0182</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0087</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.F732C</th>\n",
       "      <td>22.84</td>\n",
       "      <td>16.11</td>\n",
       "      <td>29.32</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.857950e-10</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.S714C</th>\n",
       "      <td>21.81</td>\n",
       "      <td>18.73</td>\n",
       "      <td>24.73</td>\n",
       "      <td>7.0</td>\n",
       "      <td>4.368770e-10</td>\n",
       "      <td>25</td>\n",
       "      <td>0.0041</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0075</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.F414C</th>\n",
       "      <td>21.57</td>\n",
       "      <td>15.02</td>\n",
       "      <td>27.43</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.574550e-10</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.Y908C</th>\n",
       "      <td>19.88</td>\n",
       "      <td>15.03</td>\n",
       "      <td>24.63</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.472060e-09</td>\n",
       "      <td>-</td>\n",
       "      <td>0.0033</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0045</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0086</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.P507R</th>\n",
       "      <td>19.02</td>\n",
       "      <td>13.63</td>\n",
       "      <td>23.05</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.559440e-10</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.W795S</th>\n",
       "      <td>18.91</td>\n",
       "      <td>13.57</td>\n",
       "      <td>22.89</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.643030e-10</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.W860R</th>\n",
       "      <td>18.84</td>\n",
       "      <td>17.35</td>\n",
       "      <td>20.21</td>\n",
       "      <td>18.0</td>\n",
       "      <td>1.990571e-09</td>\n",
       "      <td>7</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.L737R</th>\n",
       "      <td>18.75</td>\n",
       "      <td>13.80</td>\n",
       "      <td>22.44</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.863150e-10</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R882P</th>\n",
       "      <td>18.26</td>\n",
       "      <td>15.62</td>\n",
       "      <td>20.55</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.296770e-10</td>\n",
       "      <td>36</td>\n",
       "      <td>0.0058</td>\n",
       "      <td>0.029</td>\n",
       "      <td>0.0116</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0087</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.F755S</th>\n",
       "      <td>17.66</td>\n",
       "      <td>14.96</td>\n",
       "      <td>19.87</td>\n",
       "      <td>4.0</td>\n",
       "      <td>4.958240e-10</td>\n",
       "      <td>3</td>\n",
       "      <td>0.0041</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0086</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.N838D</th>\n",
       "      <td>17.09</td>\n",
       "      <td>13.07</td>\n",
       "      <td>19.74</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.958240e-10</td>\n",
       "      <td>5</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.E599D</th>\n",
       "      <td>16.66</td>\n",
       "      <td>13.93</td>\n",
       "      <td>18.80</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.419513e-09</td>\n",
       "      <td>-</td>\n",
       "      <td>0.0034</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0061</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.E733G</th>\n",
       "      <td>16.09</td>\n",
       "      <td>13.09</td>\n",
       "      <td>18.17</td>\n",
       "      <td>3.0</td>\n",
       "      <td>8.126630e-10</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.L547H</th>\n",
       "      <td>15.90</td>\n",
       "      <td>12.41</td>\n",
       "      <td>18.22</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.783530e-10</td>\n",
       "      <td>5</td>\n",
       "      <td>0.0017</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0087</td>\n",
       "      <td>0.0061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R729G</th>\n",
       "      <td>15.80</td>\n",
       "      <td>13.05</td>\n",
       "      <td>17.88</td>\n",
       "      <td>3.0</td>\n",
       "      <td>6.552200e-10</td>\n",
       "      <td>-</td>\n",
       "      <td>0.0009</td>\n",
       "      <td>0.0097</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.G543A</th>\n",
       "      <td>15.65</td>\n",
       "      <td>12.27</td>\n",
       "      <td>17.88</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.048440e-10</td>\n",
       "      <td>5</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R882H</th>\n",
       "      <td>15.58</td>\n",
       "      <td>15.15</td>\n",
       "      <td>15.99</td>\n",
       "      <td>105.0</td>\n",
       "      <td>1.882290e-08</td>\n",
       "      <td>1069</td>\n",
       "      <td>0.0545</td>\n",
       "      <td>0.0869</td>\n",
       "      <td>0.0694</td>\n",
       "      <td>0.0571</td>\n",
       "      <td>0.0758</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.S770W</th>\n",
       "      <td>15.49</td>\n",
       "      <td>12.84</td>\n",
       "      <td>17.40</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.723300e-10</td>\n",
       "      <td>5</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.E756D</th>\n",
       "      <td>15.33</td>\n",
       "      <td>11.52</td>\n",
       "      <td>17.79</td>\n",
       "      <td>2.0</td>\n",
       "      <td>9.655700e-10</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.P307R</th>\n",
       "      <td>15.18</td>\n",
       "      <td>11.98</td>\n",
       "      <td>17.24</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.643030e-10</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.M880V</th>\n",
       "      <td>15.02</td>\n",
       "      <td>13.27</td>\n",
       "      <td>16.42</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.486370e-09</td>\n",
       "      <td>7</td>\n",
       "      <td>0.0017</td>\n",
       "      <td>0.0097</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.W753C</th>\n",
       "      <td>15.02</td>\n",
       "      <td>11.32</td>\n",
       "      <td>17.37</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.046296e-09</td>\n",
       "      <td>-</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.F732S</th>\n",
       "      <td>14.96</td>\n",
       "      <td>11.84</td>\n",
       "      <td>16.95</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.958240e-10</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.I780T</th>\n",
       "      <td>14.91</td>\n",
       "      <td>12.49</td>\n",
       "      <td>16.53</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.346250e-09</td>\n",
       "      <td>7</td>\n",
       "      <td>0.0051</td>\n",
       "      <td>0.0201</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0061</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.G543C</th>\n",
       "      <td>14.67</td>\n",
       "      <td>12.43</td>\n",
       "      <td>16.26</td>\n",
       "      <td>3.0</td>\n",
       "      <td>8.141860e-10</td>\n",
       "      <td>25</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R882C</th>\n",
       "      <td>14.55</td>\n",
       "      <td>14.05</td>\n",
       "      <td>15.01</td>\n",
       "      <td>61.0</td>\n",
       "      <td>1.415120e-08</td>\n",
       "      <td>460</td>\n",
       "      <td>0.0355</td>\n",
       "      <td>0.0677</td>\n",
       "      <td>0.0347</td>\n",
       "      <td>0.0391</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0087</td>\n",
       "      <td>0.0364</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.E733*</th>\n",
       "      <td>14.30</td>\n",
       "      <td>11.15</td>\n",
       "      <td>16.22</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.399810e-10</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R736C</th>\n",
       "      <td>10.79</td>\n",
       "      <td>10.13</td>\n",
       "      <td>11.32</td>\n",
       "      <td>13.0</td>\n",
       "      <td>1.415120e-08</td>\n",
       "      <td>21</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.G890D</th>\n",
       "      <td>10.73</td>\n",
       "      <td>9.01</td>\n",
       "      <td>11.79</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.445670e-09</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.W297*</th>\n",
       "      <td>10.72</td>\n",
       "      <td>9.36</td>\n",
       "      <td>11.56</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.610690e-09</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.C497Y</th>\n",
       "      <td>10.65</td>\n",
       "      <td>8.79</td>\n",
       "      <td>11.63</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.211140e-09</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R771*</th>\n",
       "      <td>10.63</td>\n",
       "      <td>10.05</td>\n",
       "      <td>11.10</td>\n",
       "      <td>16.0</td>\n",
       "      <td>1.882290e-08</td>\n",
       "      <td>12</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R598*</th>\n",
       "      <td>10.54</td>\n",
       "      <td>9.95</td>\n",
       "      <td>11.00</td>\n",
       "      <td>11.0</td>\n",
       "      <td>1.882290e-08</td>\n",
       "      <td>9</td>\n",
       "      <td>0.0059</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0061</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0121</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.W581*</th>\n",
       "      <td>10.48</td>\n",
       "      <td>8.66</td>\n",
       "      <td>11.45</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.524850e-09</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0027</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0016</td>\n",
       "      <td>0.0177</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0063</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.W893*</th>\n",
       "      <td>10.48</td>\n",
       "      <td>8.66</td>\n",
       "      <td>11.45</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.524850e-09</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0016</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.W601*</th>\n",
       "      <td>10.45</td>\n",
       "      <td>9.13</td>\n",
       "      <td>11.26</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.410250e-09</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R749C</th>\n",
       "      <td>10.28</td>\n",
       "      <td>9.46</td>\n",
       "      <td>10.86</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.200440e-08</td>\n",
       "      <td>10</td>\n",
       "      <td>0.0025</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0045</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R635W</th>\n",
       "      <td>10.27</td>\n",
       "      <td>9.16</td>\n",
       "      <td>11.05</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.415120e-08</td>\n",
       "      <td>18</td>\n",
       "      <td>0.0095</td>\n",
       "      <td>0.0353</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0128</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.L547F</th>\n",
       "      <td>10.24</td>\n",
       "      <td>8.47</td>\n",
       "      <td>11.17</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.087170e-09</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.W440*</th>\n",
       "      <td>10.05</td>\n",
       "      <td>8.32</td>\n",
       "      <td>10.95</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.610690e-09</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.W313*</th>\n",
       "      <td>9.88</td>\n",
       "      <td>8.18</td>\n",
       "      <td>10.76</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.140310e-09</td>\n",
       "      <td>3</td>\n",
       "      <td>0.0049</td>\n",
       "      <td>0.0096</td>\n",
       "      <td>0.0116</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.W306*</th>\n",
       "      <td>9.80</td>\n",
       "      <td>8.12</td>\n",
       "      <td>10.66</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.410250e-09</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.G550R</th>\n",
       "      <td>9.66</td>\n",
       "      <td>8.82</td>\n",
       "      <td>10.24</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.480642e-08</td>\n",
       "      <td>8</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.S770L</th>\n",
       "      <td>9.52</td>\n",
       "      <td>8.23</td>\n",
       "      <td>10.43</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.200440e-08</td>\n",
       "      <td>6</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.0096</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0086</td>\n",
       "      <td>0.0182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.P385L</th>\n",
       "      <td>9.30</td>\n",
       "      <td>7.98</td>\n",
       "      <td>10.17</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.415120e-08</td>\n",
       "      <td>-</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0092</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.W327*</th>\n",
       "      <td>9.30</td>\n",
       "      <td>7.11</td>\n",
       "      <td>10.35</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.610690e-09</td>\n",
       "      <td>2</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.W860*</th>\n",
       "      <td>9.11</td>\n",
       "      <td>6.92</td>\n",
       "      <td>10.16</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.140310e-09</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0.0096</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R326H</th>\n",
       "      <td>9.02</td>\n",
       "      <td>8.16</td>\n",
       "      <td>9.63</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.882290e-08</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0025</td>\n",
       "      <td>0.0096</td>\n",
       "      <td>0</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.D702N</th>\n",
       "      <td>8.66</td>\n",
       "      <td>7.20</td>\n",
       "      <td>9.41</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.200440e-08</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R899C</th>\n",
       "      <td>8.63</td>\n",
       "      <td>7.04</td>\n",
       "      <td>9.55</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.415120e-08</td>\n",
       "      <td>-</td>\n",
       "      <td>0.0016</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.G332R</th>\n",
       "      <td>8.55</td>\n",
       "      <td>7.23</td>\n",
       "      <td>9.45</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.480642e-08</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0041</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0075</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.G685R</th>\n",
       "      <td>8.40</td>\n",
       "      <td>6.98</td>\n",
       "      <td>9.11</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.480642e-08</td>\n",
       "      <td>4</td>\n",
       "      <td>0.0017</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R899H</th>\n",
       "      <td>8.21</td>\n",
       "      <td>6.73</td>\n",
       "      <td>9.04</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.882290e-08</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R688H</th>\n",
       "      <td>8.21</td>\n",
       "      <td>6.73</td>\n",
       "      <td>9.03</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.882290e-08</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0033</td>\n",
       "      <td>0.0096</td>\n",
       "      <td>0</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R659H</th>\n",
       "      <td>8.10</td>\n",
       "      <td>6.72</td>\n",
       "      <td>8.78</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.882290e-08</td>\n",
       "      <td>-</td>\n",
       "      <td>0.0017</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R366H</th>\n",
       "      <td>8.10</td>\n",
       "      <td>6.72</td>\n",
       "      <td>8.78</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.882290e-08</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R771Q</th>\n",
       "      <td>7.58</td>\n",
       "      <td>5.48</td>\n",
       "      <td>8.53</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.200440e-08</td>\n",
       "      <td>4</td>\n",
       "      <td>0.0025</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>102 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          s (%)  s lower value (%)  s upper value (%)  observed number  \\\n",
       "p.E30A   196.75              28.05             196.75              4.0   \n",
       "p.F751C   30.46              18.06             196.75              2.0   \n",
       "p.Y735C   26.10              23.52              29.07             23.0   \n",
       "p.F732C   22.84              16.11              29.32              2.0   \n",
       "p.S714C   21.81              18.73              24.73              7.0   \n",
       "p.F414C   21.57              15.02              27.43              2.0   \n",
       "p.Y908C   19.88              15.03              24.63              4.0   \n",
       "p.P507R   19.02              13.63              23.05              2.0   \n",
       "p.W795S   18.91              13.57              22.89              2.0   \n",
       "p.W860R   18.84              17.35              20.21             18.0   \n",
       "p.L737R   18.75              13.80              22.44              2.0   \n",
       "p.R882P   18.26              15.62              20.55              6.0   \n",
       "p.F755S   17.66              14.96              19.87              4.0   \n",
       "p.N838D   17.09              13.07              19.74              2.0   \n",
       "p.E599D   16.66              13.93              18.80              4.0   \n",
       "p.E733G   16.09              13.09              18.17              3.0   \n",
       "p.L547H   15.90              12.41              18.22              2.0   \n",
       "p.R729G   15.80              13.05              17.88              3.0   \n",
       "p.G543A   15.65              12.27              17.88              2.0   \n",
       "p.R882H   15.58              15.15              15.99            105.0   \n",
       "p.S770W   15.49              12.84              17.40              3.0   \n",
       "p.E756D   15.33              11.52              17.79              2.0   \n",
       "p.P307R   15.18              11.98              17.24              2.0   \n",
       "p.M880V   15.02              13.27              16.42              6.0   \n",
       "p.W753C   15.02              11.32              17.37              2.0   \n",
       "p.F732S   14.96              11.84              16.95              2.0   \n",
       "p.I780T   14.91              12.49              16.53              3.0   \n",
       "p.G543C   14.67              12.43              16.26              3.0   \n",
       "p.R882C   14.55              14.05              15.01             61.0   \n",
       "p.E733*   14.30              11.15              16.22              2.0   \n",
       "...         ...                ...                ...              ...   \n",
       "p.R736C   10.79              10.13              11.32             13.0   \n",
       "p.G890D   10.73               9.01              11.79              3.0   \n",
       "p.W297*   10.72               9.36              11.56              3.0   \n",
       "p.C497Y   10.65               8.79              11.63              2.0   \n",
       "p.R771*   10.63              10.05              11.10             16.0   \n",
       "p.R598*   10.54               9.95              11.00             11.0   \n",
       "p.W581*   10.48               8.66              11.45              2.0   \n",
       "p.W893*   10.48               8.66              11.45              2.0   \n",
       "p.W601*   10.45               9.13              11.26              3.0   \n",
       "p.R749C   10.28               9.46              10.86              6.0   \n",
       "p.R635W   10.27               9.16              11.05              5.0   \n",
       "p.L547F   10.24               8.47              11.17              2.0   \n",
       "p.W440*   10.05               8.32              10.95              2.0   \n",
       "p.W313*    9.88               8.18              10.76              2.0   \n",
       "p.W306*    9.80               8.12              10.66              2.0   \n",
       "p.G550R    9.66               8.82              10.24              5.0   \n",
       "p.S770L    9.52               8.23              10.43              5.0   \n",
       "p.P385L    9.30               7.98              10.17              3.0   \n",
       "p.W327*    9.30               7.11              10.35              2.0   \n",
       "p.W860*    9.11               6.92              10.16              2.0   \n",
       "p.R326H    9.02               8.16               9.63              7.0   \n",
       "p.D702N    8.66               7.20               9.41              2.0   \n",
       "p.R899C    8.63               7.04               9.55              2.0   \n",
       "p.G332R    8.55               7.23               9.45              4.0   \n",
       "p.G685R    8.40               6.98               9.11              2.0   \n",
       "p.R899H    8.21               6.73               9.04              2.0   \n",
       "p.R688H    8.21               6.73               9.03              2.0   \n",
       "p.R659H    8.10               6.72               8.78              2.0   \n",
       "p.R366H    8.10               6.72               8.78              2.0   \n",
       "p.R771Q    7.58               5.48               8.53              2.0   \n",
       "\n",
       "         site-specific mutation rate Number of times in COSMIC  \\\n",
       "p.E30A                  1.863150e-10                         3   \n",
       "p.F751C                 1.574550e-10                         -   \n",
       "p.Y735C                 8.813490e-10                        12   \n",
       "p.F732C                 1.857950e-10                         1   \n",
       "p.S714C                 4.368770e-10                        25   \n",
       "p.F414C                 1.574550e-10                         -   \n",
       "p.Y908C                 1.472060e-09                         -   \n",
       "p.P507R                 4.559440e-10                         -   \n",
       "p.W795S                 4.643030e-10                         -   \n",
       "p.W860R                 1.990571e-09                         7   \n",
       "p.L737R                 1.863150e-10                         1   \n",
       "p.R882P                 6.296770e-10                        36   \n",
       "p.F755S                 4.958240e-10                         3   \n",
       "p.N838D                 4.958240e-10                         5   \n",
       "p.E599D                 1.419513e-09                         -   \n",
       "p.E733G                 8.126630e-10                         1   \n",
       "p.L547H                 3.783530e-10                         5   \n",
       "p.R729G                 6.552200e-10                         -   \n",
       "p.G543A                 4.048440e-10                         5   \n",
       "p.R882H                 1.882290e-08                      1069   \n",
       "p.S770W                 5.723300e-10                         5   \n",
       "p.E756D                 9.655700e-10                         -   \n",
       "p.P307R                 4.643030e-10                         2   \n",
       "p.M880V                 1.486370e-09                         7   \n",
       "p.W753C                 1.046296e-09                         -   \n",
       "p.F732S                 4.958240e-10                         3   \n",
       "p.I780T                 1.346250e-09                         7   \n",
       "p.G543C                 8.141860e-10                        25   \n",
       "p.R882C                 1.415120e-08                       460   \n",
       "p.E733*                 5.399810e-10                         -   \n",
       "...                              ...                       ...   \n",
       "p.R736C                 1.415120e-08                        21   \n",
       "p.G890D                 3.445670e-09                         1   \n",
       "p.W297*                 4.610690e-09                         1   \n",
       "p.C497Y                 3.211140e-09                         3   \n",
       "p.R771*                 1.882290e-08                        12   \n",
       "p.R598*                 1.882290e-08                         9   \n",
       "p.W581*                 3.524850e-09                         2   \n",
       "p.W893*                 3.524850e-09                         2   \n",
       "p.W601*                 5.410250e-09                         1   \n",
       "p.R749C                 1.200440e-08                        10   \n",
       "p.R635W                 1.415120e-08                        18   \n",
       "p.L547F                 4.087170e-09                         1   \n",
       "p.W440*                 4.610690e-09                         2   \n",
       "p.W313*                 5.140310e-09                         3   \n",
       "p.W306*                 5.410250e-09                         1   \n",
       "p.G550R                 1.480642e-08                         8   \n",
       "p.S770L                 1.200440e-08                         6   \n",
       "p.P385L                 1.415120e-08                         -   \n",
       "p.W327*                 4.610690e-09                         2   \n",
       "p.W860*                 5.140310e-09                         1   \n",
       "p.R326H                 1.882290e-08                         2   \n",
       "p.D702N                 1.200440e-08                         -   \n",
       "p.R899C                 1.415120e-08                         -   \n",
       "p.G332R                 1.480642e-08                         2   \n",
       "p.G685R                 1.480642e-08                         4   \n",
       "p.R899H                 1.882290e-08                         3   \n",
       "p.R688H                 1.882290e-08                         1   \n",
       "p.R659H                 1.882290e-08                         -   \n",
       "p.R366H                 1.882290e-08                         -   \n",
       "p.R771Q                 1.200440e-08                         4   \n",
       "\n",
       "        Exac Overall freq (%) Exac African freq (%) Exac East Asian freq (%)  \\\n",
       "p.E30A                 0.2959                  0.02                        0   \n",
       "p.F751C                     -                     -                        -   \n",
       "p.Y735C                0.0118                0.0097                        0   \n",
       "p.F732C                     -                     -                        -   \n",
       "p.S714C                0.0041                     0                        0   \n",
       "p.F414C                     -                     -                        -   \n",
       "p.Y908C                0.0033                     0                        0   \n",
       "p.P507R                     -                     -                        -   \n",
       "p.W795S                     -                     -                        -   \n",
       "p.W860R                     -                     -                        -   \n",
       "p.L737R                     -                     -                        -   \n",
       "p.R882P                0.0058                 0.029                   0.0116   \n",
       "p.F755S                0.0041                     0                        0   \n",
       "p.N838D                     -                     -                        -   \n",
       "p.E599D                0.0034                     0                        0   \n",
       "p.E733G                     -                     -                        -   \n",
       "p.L547H                0.0017                     0                        0   \n",
       "p.R729G                0.0009                0.0097                        0   \n",
       "p.G543A                     -                     -                        -   \n",
       "p.R882H                0.0545                0.0869                   0.0694   \n",
       "p.S770W                     -                     -                        -   \n",
       "p.E756D                     -                     -                        -   \n",
       "p.P307R                     -                     -                        -   \n",
       "p.M880V                0.0017                0.0097                        0   \n",
       "p.W753C                0.0008                     0                        0   \n",
       "p.F732S                     -                     -                        -   \n",
       "p.I780T                0.0051                0.0201                        0   \n",
       "p.G543C                     -                     -                        -   \n",
       "p.R882C                0.0355                0.0677                   0.0347   \n",
       "p.E733*                     -                     -                        -   \n",
       "...                       ...                   ...                      ...   \n",
       "p.R736C                     -                     -                        -   \n",
       "p.G890D                0.0008                     0                        0   \n",
       "p.W297*                     -                     -                        -   \n",
       "p.C497Y                     -                     -                        -   \n",
       "p.R771*                     -                     -                        -   \n",
       "p.R598*                0.0059                  0.01                        0   \n",
       "p.W581*                0.0027                     0                        0   \n",
       "p.W893*                0.0016                     0                        0   \n",
       "p.W601*                     -                     -                        -   \n",
       "p.R749C                0.0025                     0                        0   \n",
       "p.R635W                0.0095                0.0353                        0   \n",
       "p.L547F                     -                     -                        -   \n",
       "p.W440*                0.0008                     0                        0   \n",
       "p.W313*                0.0049                0.0096                   0.0116   \n",
       "p.W306*                0.0008                     0                        0   \n",
       "p.G550R                     -                     -                        -   \n",
       "p.S770L                 0.005                0.0096                        0   \n",
       "p.P385L                 0.005                     0                        0   \n",
       "p.W327*                     -                     -                        -   \n",
       "p.W860*                0.0008                0.0096                        0   \n",
       "p.R326H                0.0025                0.0096                        0   \n",
       "p.D702N                     -                     -                        -   \n",
       "p.R899C                0.0016                     0                        0   \n",
       "p.G332R                0.0041                     0                        0   \n",
       "p.G685R                0.0017                     0                        0   \n",
       "p.R899H                     -                     -                        -   \n",
       "p.R688H                0.0033                0.0096                        0   \n",
       "p.R659H                0.0017                     0                        0   \n",
       "p.R366H                     -                     -                        -   \n",
       "p.R771Q                0.0025                     0                        0   \n",
       "\n",
       "        Exac European freq (%) Exac Finnish freq (%) Exac Latino freq (%)  \\\n",
       "p.E30A                  0.3033                2.1424               0.1413   \n",
       "p.F751C                      -                     -                    -   \n",
       "p.Y735C                 0.0182                     0               0.0087   \n",
       "p.F732C                      -                     -                    -   \n",
       "p.S714C                 0.0075                     0                    0   \n",
       "p.F414C                      -                     -                    -   \n",
       "p.Y908C                 0.0045                     0               0.0086   \n",
       "p.P507R                      -                     -                    -   \n",
       "p.W795S                      -                     -                    -   \n",
       "p.W860R                      -                     -                    -   \n",
       "p.L737R                      -                     -                    -   \n",
       "p.R882P                  0.003                     0               0.0087   \n",
       "p.F755S                  0.006                     0               0.0086   \n",
       "p.N838D                      -                     -                    -   \n",
       "p.E599D                 0.0061                     0                    0   \n",
       "p.E733G                      -                     -                    -   \n",
       "p.L547H                      0                     0               0.0087   \n",
       "p.R729G                      0                     0                    0   \n",
       "p.G543A                      -                     -                    -   \n",
       "p.R882H                 0.0571                0.0758                    0   \n",
       "p.S770W                      -                     -                    -   \n",
       "p.E756D                      -                     -                    -   \n",
       "p.P307R                      -                     -                    -   \n",
       "p.M880V                 0.0015                     0                    0   \n",
       "p.W753C                 0.0015                     0                    0   \n",
       "p.F732S                      -                     -                    -   \n",
       "p.I780T                 0.0061                     0                    0   \n",
       "p.G543C                      -                     -                    -   \n",
       "p.R882C                 0.0391                     0               0.0087   \n",
       "p.E733*                      -                     -                    -   \n",
       "...                        ...                   ...                  ...   \n",
       "p.R736C                      -                     -                    -   \n",
       "p.G890D                 0.0015                     0                    0   \n",
       "p.W297*                      -                     -                    -   \n",
       "p.C497Y                      -                     -                    -   \n",
       "p.R771*                      -                     -                    -   \n",
       "p.R598*                 0.0061                     0                    0   \n",
       "p.W581*                 0.0016                0.0177                    0   \n",
       "p.W893*                  0.003                     0                    0   \n",
       "p.W601*                      -                     -                    -   \n",
       "p.R749C                 0.0045                     0                    0   \n",
       "p.R635W                      0                     0                    0   \n",
       "p.L547F                      -                     -                    -   \n",
       "p.W440*                 0.0015                     0                    0   \n",
       "p.W313*                  0.006                     0                    0   \n",
       "p.W306*                 0.0015                     0                    0   \n",
       "p.G550R                      -                     -                    -   \n",
       "p.S770L                 0.0015                     0               0.0086   \n",
       "p.P385L                 0.0092                     0                    0   \n",
       "p.W327*                      -                     -                    -   \n",
       "p.W860*                      0                     0                    0   \n",
       "p.R326H                  0.003                     0                    0   \n",
       "p.D702N                      -                     -                    -   \n",
       "p.R899C                  0.003                     0                    0   \n",
       "p.G332R                 0.0075                     0                    0   \n",
       "p.G685R                  0.003                     0                    0   \n",
       "p.R899H                      -                     -                    -   \n",
       "p.R688H                  0.003                     0                    0   \n",
       "p.R659H                 0.0015                     0                    0   \n",
       "p.R366H                      -                     -                    -   \n",
       "p.R771Q                  0.003                     0                    0   \n",
       "\n",
       "        Exac South Asian freq (%)  \n",
       "p.E30A                     0.0247  \n",
       "p.F751C                         -  \n",
       "p.Y735C                         0  \n",
       "p.F732C                         -  \n",
       "p.S714C                         0  \n",
       "p.F414C                         -  \n",
       "p.Y908C                         0  \n",
       "p.P507R                         -  \n",
       "p.W795S                         -  \n",
       "p.W860R                         -  \n",
       "p.L737R                         -  \n",
       "p.R882P                         0  \n",
       "p.F755S                         0  \n",
       "p.N838D                         -  \n",
       "p.E599D                         0  \n",
       "p.E733G                         -  \n",
       "p.L547H                    0.0061  \n",
       "p.R729G                         0  \n",
       "p.G543A                         -  \n",
       "p.R882H                    0.0485  \n",
       "p.S770W                         -  \n",
       "p.E756D                         -  \n",
       "p.P307R                         -  \n",
       "p.M880V                         0  \n",
       "p.W753C                         0  \n",
       "p.F732S                         -  \n",
       "p.I780T                         0  \n",
       "p.G543C                         -  \n",
       "p.R882C                    0.0364  \n",
       "p.E733*                         -  \n",
       "...                           ...  \n",
       "p.R736C                         -  \n",
       "p.G890D                         0  \n",
       "p.W297*                         -  \n",
       "p.C497Y                         -  \n",
       "p.R771*                         -  \n",
       "p.R598*                    0.0121  \n",
       "p.W581*                    0.0063  \n",
       "p.W893*                         0  \n",
       "p.W601*                         -  \n",
       "p.R749C                         0  \n",
       "p.R635W                    0.0128  \n",
       "p.L547F                         -  \n",
       "p.W440*                         0  \n",
       "p.W313*                         0  \n",
       "p.W306*                         0  \n",
       "p.G550R                         -  \n",
       "p.S770L                    0.0182  \n",
       "p.P385L                         0  \n",
       "p.W327*                         -  \n",
       "p.W860*                         0  \n",
       "p.R326H                         0  \n",
       "p.D702N                         -  \n",
       "p.R899C                         0  \n",
       "p.G332R                         0  \n",
       "p.G685R                         0  \n",
       "p.R899H                         -  \n",
       "p.R688H                    0.0061  \n",
       "p.R659H                    0.0062  \n",
       "p.R366H                         -  \n",
       "p.R771Q                         0  \n",
       "\n",
       "[102 rows x 13 columns]"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "DNMT3A_df_sorted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "metadata": {
    "code_folding": []
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABZoAAAHvCAYAAADdMGqTAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAIABJREFUeJzs3Xl4FEX+x/F3hZBAOBIgBOQMiIAnCios6IJcCioIHiCsgHjsioKsiquuGoL6ExYPQAVdFAHFWw5FuZQbAbnPcJNAOAIhJBwhkKN+f/TMOJOZhEA4wvp5PU8/M1NdXV3dNd3idyrfNtZaRERERERERERERETOVtDF7oCIiIiIiIiIiIiIXNoUaBYRERERERERERGRQlGgWUREREREREREREQKRYFmERERERERERERESkUBZpFREREREREREREpFAUaBYRERERERERERGRQlGgWUREROQiMcYEG2OsMebji92XosgY09cYs9kYc8p1nqq5ylsZY5YaY465yv92sfsqIiIiIvJnp0CziIiIyDlmjGnhCoDmtTTJZ9sgY8xAY0yHC9nnosYY0wYYAawHHgceAlKMMZHARKAE8IyrfNF56kND11jUOB/t/1kZY+rkuh6yjTGprh8VvjLG3GuMCQ6wXWuvbXoFWO/+4WZyrvKFrvJTxpiKefTpA6+2b/Fqq6BLNWNMZWPMp8aYdcaYw8aYE8aYrcaY0caY2qc5J/1c7Rw2xpQ8w1MqIiIiUiT4/QNORERERM6ZL4GfA5RvA7DWZrmCSlle64KAGOAT4Ifz3sOiq43r9WFr7RF3oTGmBVAWeMVae77PT0OcsfgF2HWe9/VnNAP43PW+NFAHuBPoAiw3xnSy1ibmse0gY8xX1tqMAu4rE+fa+hvwrvcKY0wJ4EEgA+cHDIBsnB8xvLUAHgFGAb/lWpcC1AKuAKbjfF9OAPWA3sADxpibrbWb8+hfb2A7cDnQGZhQwOMSERERKTIUaBYRERE5f1Zaaz/Pr8IZBMr+bCoD2d5BZq9ycAJ7kg9jjAFKWWuPXey+5GFT7uvDGDMAeBYYCkw1xjSy1mbn2m45cCPQ11WvINJxgsMPkyvQDHQCygFfAN0ArLWWP4Lg7r6VwAk0/5bHdb0BuCV3oTFmomvfTwL9AqxvBDRw7fsFnKCzAs0iIiJyyVHqDBEREZGLJHeOZmNMHZyZlwCPeP1Zflbu+q4/719gjEk3xiQbY/5rjCkVYB9VjTEfGmN2u1IH7HF9jsxVr4IxZrgxZocxJsMYc8gYs9wY80yuer2MMctcqQ6OG2O2G2M+N8ZUKOAx326MmWWMSXPtZ40x5nGv9XWMMRZnNmkxr3PwizEmEWemN8AC73Pj2raEMeZlY8xGV9uHjTE/GGMaBOiHMcb83Rjzu3FyPR81xqw1xsS41r8OjM61L++xKmmMGWScdA/prn2tM8YMLsA5CHb1c4ExJsk1Lgmu9A3l89jmfmPMPNd5SzfGbHKNV7BrvTutxEPGyW0dB5wE+nu10cJ1Ho+42lhhAqeguNYY873ru3LSGLPPGDPbGNPOq85ZH39+rOMt4Guc4Ov9Aap9AawBXjTGlDuD5j8FrnUFdr09DKwA1p1FlwsiwfWaV18fAY4Ak4FxwG3GmFrnqS8iIiIi541mNIuIiIicP2G5A7rASWvt0Tzq7wd64gSb5vJHUDUnV71GOLMwP8GZddkSeAwnBUcfdyVXsOo3oJir7g6cP+1/AieYdZPXjOGJwF+AD4G1QCngSpx0Ae+42nsYGAPMA17FSQ1QA2gPRAKH8jsZxpgngA9cfXoNZ5bp7cBHxpha1toXXefgIVcfm7jOB8A+nJQZd+IE5l4DtrjPjTEmBJgJNAbG4+R3Luc6L78ZY26x1q5y1TU4aU26AIuBN4BU1/HeC8QC3+LMnvbeF7jSnrjOUw9grKuN4q5z2zK/c+BSAmfW7vc4wcXjwM04uaibucbF/YMDxpghwPM4M2bfBpJwUizcC/wb8J6x/KzruD9xncsEVxv3AN+5zuNQ1z4fBD51nXt3gL0iMBsndcSHOCkgIoGbXH2cdg6OvyA+xhmfO4Gvcq2zODN/pwEv4pybgpiC8x3tjRNYxjj5t1vhzI4uW+he4/kulgVCcNKBxLpW+aXRMX+k7fjGWnvCGPM5MAQn+P3queiPiIiIyAVjrdWiRYsWLVq0aNFyDhec4KzNY/nKq16wq+zj/MoCrMsGbsy1bgbODNaSXmU/4QQbq+Sq29jVxsuuz+Vd7Y44zXH9gJOyothZnJNqrv6ND7DuA5wgeU2vss+BrAB1H3X19ZZc5QNwgs6tc5VHAInAL15l3VxtjAWCctUPOt2+XOuOAD+c5fcjyHucvMr/7tpfZ6+ypq6yWUBorvrG631rV71kIDJXveKuc5ACVPYqDwWWuL4LtV1lnXP3IY9jKMzx13HtY1g+daJcdZYGOMb+rs+/4vzYUS3X9TE5V1sLgVTX++Gu8xDq+uz+wSQCJ3gdcLxzfR/+dprjuwffa34f8HQedbvn3ifwI06APyi//WjRokWLFi1atBS1RakzRERERM6f/+I81M57ef0ctLvQWrs8V9lsnBmUNQFcKRja4cyYPWWMiXQvOA8d2wm0dW2bjpOyo4kxpmY++00DygDtXbOCz8T9rv6N8e6Lqz8/4sy6bnWGbXr7G86M39W52g7GCUg2N8aEuuq6g3vPWWt9Zovn/pyPNJw0DFefaUettTnW2hMAxphixpgIV19nu6o09qre3fX6grX2ZK52bIDmx1prk3OV3QRUxfnxYr/X9ieBt3AC3x28jgucMS6Tz2Gc9fEXkHumfX6zjP+FEywfdAbtjsGZ8X2P6zvcE5hkrU09q14GtgjnWu+AM+N6P1DeneYkl0eA7dbahV5lY4Hq/PFATBEREZFLglJniIiIiJw/W621v5yHdncEKHOnrXDnSq4PGJxZsn/Po51scB5IaJxczO8A8caYDThBz0nW2jle9V/HedjZD0CyMWYeTvqCr+3pHzh3pet1Tj51Kp2mjfzUxwlkH8ynTnmc2aVXAIkBArJn4mmcFCfrjTHbcY7rB2BqHgFgH8aYrjhpLhrgzDj25p3L9wqccVpbwH5tCVDmzve7IcC69a7X2q7X2Tg5kB8BehhjluGkJPnaWrvJa7tCHX8BuAPMuR8G6WGtXW6M+RboaYx5G9h8ukattWuMMatwUlMcwDnuf5yD/nrv4yDgvu5/dKXDWIuTguRJdz1jTG2cv34YZZz87G5xOKlNHsH5SwURERGRS4ICzSIiIiKXnux81plcr+Nw0lAEku5+Y6193xgzCScn7l+BB4C+xpgJ1tq/uepsNsbUx0lh0ApojpNLd6Ax5q/W2p0F6Fd3nABfINvyKC+IIGA1TgqNvKR49aVQwVBr7URXoL09znlog5NaYa4xpq31yrGcmzHmAZwc0UuAfjhpLTJwAuU/4fvA7jOdOZ4eoKzAbbiCxN1dD/VrB9yKkwP5FWNMX2vtKFe9sz7+ArrO9Xq64PG/cfKVD3a9FsQYnBQaxXFSVPx6Nh0sKGttojFmNvCYMaa/17npjTM2ffDKre6lozGmgrU239znIiIiIkWFAs0iIiIiRcu5mA0KTtDWAsULOqvaWrsHJ93Hf11/5j8BJ+j4tnU9SM+VbuEn14IxpgPOQ9b648xyzctW1+vB8zTLeytQEfi1ADNqN+Okhog8zazmfNtxBQA/Az5zpWEYijNL+S5gUj6bPoQTEL7NWpvhLjTGXJNHX1sD1wIr8+tPPra7XgOlubjK9eozS95auw5YB/zHlYbld5xg7iivOmd7/AXxqOv1p/wqWWu3GWP+izNT+NYCtj0BJ2VIS2DQGaRLKYySOIHt0sBhY0wQTtqOFTjnNbeqwDCclDDDL0D/RERERApNOZpFREREihBrbTbO7NbyhWwnCSflwf3GmJtyrzeOiq73YcaYkrm2z8IJNOLuiyuPcG4rvevk42vgFDDIGFMiQH8ijDEhp2kjP+NxgnMBg93GGO+0HBNwZpL+J3eu6Vyf3elAyueqE2yMCfcucwW3VweqH0A2zoMLPf8Wd+335QB1v3C9vpn7/JxBnuxlwB7gEWNMlNf2IcBzrr784Corn7tda20KEA+UNsaEnIPjz5cx5lmgC7AK+K4AmwzCGashBWnfWnsYJ11GLDD6LLvpJ9d3zLv8WpwUGVtc+wa4HecBmeOttd8FWIYDu3FmPYuIiIhcEjSjWURERKToWQLcbox5HifYlG2t/eYs2vk7sBBYaIwZjxO4C8bJ2XsP8AlO3uWrgF9cqTM2AIddZU/gzIZd5GpvtjHmILDA1a/yOLluc8g7PQcA1toEY8xTwIfARlfe2l04s5CvAzoCdXHSSJyNd3Bm/r5rjGkNzAWOAjVw0nwc5Y+Hq30FdHb1va4x5kech9vVddVt4Kr3O86s5pddQfnj/PEgxV3GmB9wgqvuXL9P4KTnyHcWLk7wtCPO+fwM54F2nQC/ALy19jdX/uFnAXdO4v04Y3i/MeaG0+XHttZmGWP6At8Cy4wxo13H0hW4GWdWr3tGc2/gKdd3YRuQBdzmOi9fWGtPuX5wKMzxu9U3xvzN9b4UcDlwN06+7d+BTq4fXvJlrT3gOkcxBdwv1tqxBa17Bl4xxjQHfsYJzBfDmYn+N9f7J73qPoLz3cpv5vdE4GljzI0BHv4pIiIiUuQo0CwiIiJS9PwDeB9nhmsZnBmwZxxodgV3GwIvAB2AHjgpG3YDk/ljtmgCTi7nFjgBz1CcgO9HwBCv9A4fAPe7+lce5wGEK4E+1tq5BejPaGNMHM4s2ieAcCAZ2ISTaze/B/mdru1Txpg7gKdwAnuxrlV7cQL347zqWmNMF5zAX2+cAGU2TvqIb73q7TTGPIqTo3gUTuqDT1zbjcAJvrbBCZLuwwkavmmt3X+avn5ujCmNM/v6bZzg7A/AKwTIX22tfc71ALsngX/hzMbeDUzFmf1ekPMzyRjTBuc79S+c/w+IA3pbaz/1qjobJ/B/N3AZTqB5J06g+31XnaOFOX4vt7sWizMjeT/O9+kVnAdRnjbI7OUtnO9lYR4oWVg/AlVwZmNH4QSXE3F+2HjLWhsHnr8MuBtYZq3dnU973+N8R3oDCjSLiIhIkWfOzUOhzw9jzBDgRpzZJZHACZz/EZoMvO/9YAxjTDTOP4Lz8rW1tut566yIiIiIiIiIiIjIn1RRDzSfwpnVsBFndkcpoAlO8Hkv0MQ9C8Ar0LwGJxCd23prbUFyvImIiIiIiIiIiIjIGSjqqTPKej+J280Y8wbwEvAi0CfX6tXW2oEXoG8iIiIiIiIiIiIigteTrouiQEFmF3eOwisuVF9EREREREREREREJLCiPqM5L3e7XtcGWFfFGPN3oALOA2oWW2sD1RMRERERERERERGRc6BI52h2M8Y8B5TGeTL5jcAtOEHm1tbag6460eT9MMC5QE9r7a7T7WvUqFEW4IknnjCF7beIiIiIiIiIiIjIn0GRTp3h5TkgBuiPE2SeDrR1B5ld0oHXgEZAOdfSHJgDtAB+NcaUymsHxpjHjTHL3Z/T0tLsqlWrbGZmpl2+fLkF7PLly+3JkyftmjVr7OHDh+22bdtsYmKi3bdvn928ebM9cuSIXbVqlc3KyvLbZt26dTYlJcVu377d7t692+7fv99u2rTJHj161K5YscJmZ2f7bHPixAm7fv16e+jQIbtjxw67a9cum5SUZOPi4uyxY8fs8uXLbU5Ojh027Xc7dFOWHTbtd5uenm43bNhgDx48aHfu3GkTEhLsgQMH7MaNG+3x48c97S9btsyzn+PHj9uNGzfaAwcO2ISEBLtz50578OBBu2HDBpuenu7Tp5ycHLt8+XJ77NgxGxcXZ5OSkuyuXbvsjh077KFDh+z69evtiRMnfLbJzs62K1assEePHrWbNm2y+/fvt7t377bbt2+3KSkpdt26dfbkyZM+22RlZdlVq1bZI0eO2M2bN9t9+/bZxMREu23bNnv48GG7Zs0av20yMzPtqlWrbFpamt2yZYvdu3ev3bNnj926datNTU21q1evtpmZmT7HfurUKc9Ybt261e7Zs8fu3bvXbtmyxXqPv3ubZcuW2ZMnT9q1a9d6xjL3+K9cudJvLDMyMuy6det8xjL3+LvPbe7xT05O9oxlUlKS3bhxo2f8rbU+23iPf3x8vI2Pj7cHDx70G3/v12PHjvmNf3Jysl2/fr1NT0/3OV85OTl2xYoV+Y5/RkaGz/csOzvbrly50h45ciTg+K9duzbP8U9LS7ObN2+2e/fu9Rv/U6dO+Y3/6tWrbWpqqt2yZYvds2eP3bNnj92yZYvP+Htv4z3+7mt57969dvPmzZ7xD3Qte4+/97XsPf7e5y0jIyPfazn3+C9btsymp6f7jb/7WnaPf+6xdF/Lucc/0LXs/u64xz8pKclv/HNfy+7x976Wvcd/3bp1PuPvvv7d4+99Lecef+9rLL9rOdD4u7dxj39+13JRvpfndS3rXq57ue7lupfrXq57ue7lupfrXq57eVG8lw+budIOXZ9uh037Xfdy3ct1L78E7uVcIJfEjGY3Y0wloCkwGCgD3GWtXXmabYKBhUBjoL+1dnh+9b1mNJ+TPp9vQzdled4PqH+pZkIRERERERERkUuFYhEil5wLkrnhUpnRDIC1NslaOwloi5ODeXwBtskCPnZ9/Ot57J6IiIiIiIiIiIjIn9IlFWh2s9YmABuBq40xkQXYxJ1iI8/UGSIiIiIiIiIiIiJydi7JQLNLFddrdgHqNnG97jhPfRERERERERERERH50yqygWZjTH1jTOUA5UHGmDeAKOA3a+1hV3ljY0xIgPotgX+6Pn5+PvssIiIiIiIiIiIi8mdUlDO23wEMNcbMB7YDh4BKQHOgNrAfeMyr/hCcVBpzgURX2XVAS9f7V6y1v12AfouIiIiIiIiIiIj8qRTlQPMvwH+BZkADIAI4DmwBPgNGWGtTvOp/BnQCbgLaAcWBJOAb4H1r7YIL13URERERERERERGRP48iG2i21q4HnjyD+p8An5y/HomIiIiIiIiIiIhIIEU2R7OIiIiIiIiIiIiIXBoUaBYRERERERERERGRQlGgWUREREREREREREQKRYFmERERERERERERESkUBZpFREREREREREREpFAUaBYRERERERERERGRQlGgWUREREREREREREQKRYFmERERERERERERESkUBZpFREREREREREREpFAUaJYiZeDAgRhjmDt37sXuioiIiIiIiIiIiBSQAs0iIiIiIiIiIiIiUigKNEuR8tRTTxEXF8fNN998sbsiIiIiIiIiIiIiBRR8sTsg4i0yMpLIyMiL3Q0RERERERERERE5A5rRLIUWHx+PMYZevXqxfft27rvvPipUqECZMmVo27Yt69evB+DgwYM8/vjjXHbZZZQoUYKbbrqJOXPm+LSVV47mBQsWcPfdd1OtWjVCQ0OpXLkyTZo0ITY21qdeUlISzz33HPXq1aNUqVJERERQr149evXqxY4dO/z6PmPGDNq3b09kZCShoaFcfvnlDBgwgNTUVL+60dHRREdHk56ezoABA6hRowahoaHUqVOHIUOGYK312+aHH36gVatWXHbZZYSGhlKlShWaN2/OyJEj/eqmpKTw4osvcuWVV1KyZEnCw8Np1aoVM2fOPO0YiIiIiIiIiIiIXEya0SznTHx8PI0bN+bKK6+kV69exMfHM2nSJFq0aMHixYu54447KFu2LF26dCElJYWvvvqKdu3asWXLFmrUqJFnu9OnT+fOO++kbNmydOjQgapVq5KSkkJcXBwjR44kJiYGgPT0dJo1a8b27dtp06YNd999N9ZaEhISmDJlCvfddx+1a9f2tDto0CBiYmIoX748d911F1FRUaxdu5a33nqLn3/+mcWLF1O2bFmfvmRmZtK2bVv27t1Lu3btCA4OZvLkybzwwgtkZGR4+gLw3//+l7///e9UrlyZu+++m8jISA4cOMDatWv59NNP6dOnj6duQkICLVq0ID4+nltvvZU77riD48ePM3XqVO644w4++ugjHnvssXM1VCIiIiIiIiIiIueUAs1yzsybN4/XX3+df//7356y1157jVdffZXGjRvzwAMPMHLkSIKCnIn0bdq0oUePHrz77ru8++67ebY7evRocnJymDt3Lg0aNPBZl5yc7Hn/66+/sn37dvr37+/X3qlTpzh58qTn85w5c4iJieEvf/kLP//8MxEREZ51Y8eO5eGHHyYmJsavnb1799KgQQNmzZpFyZIlAYiJiaFu3bq8++67vPTSSxQvXhyAjz76iJCQENasWUNUVFSe/Qbo2bMnCQkJfPnll3Tt2tVTnpqaSosWLejXrx8dOnSgUqVKeZ4nERERERERERGRi0WB5vOk25LMS2afXzQpfk72Hx0dzQsvvOBT1rNnT1599VVOnjzJ0KFDPUFmgG7dutG7d29Wr15doPbdgV1vgfI5B6oXEhJCSEiI5/OIESMAJ4jtHWQG6NWrF8OHD2fChAkBA+AjRozw2UdUVBQdO3Zk/PjxbN68mWuuucazLjg42BN4zqvfa9asYd68edx3330+QWaAiIgIYmNjueeee/j+++99ZkGLiIiIiIiIiIgUFQo0yzlz/fXXU6xYMZ+yKlWqAFC3bl3KlCnjs65YsWJUqlSJxMTEfNvt3r07EydOpHHjxnTp0oXbbruNZs2aUa1aNZ96zZs3p2rVqgwePJiVK1fSvn17mjVrFrBfixcvpnjx4nz77bd8++23fvs8deoUBw8e5NChQ1SoUMFTHh4eTp06dfzqV69eHYDDhw/79PvZZ5/l6quvpkuXLjRv3pxmzZpRsWJFv74ApKWlMXDgQL+2Dx48CEBcXFx+p0lEREREREREROSiUaBZzpnw8HC/suDg4DzXuddnZuY/E7tz585MnTqVt99+mzFjxvDRRx8B0KhRI958803atGkDQNmyZVmyZAkxMTH88MMPzJgxA3BmD/fp04eXX37ZM7v40KFDZGVl+T1MMLdjx475BJpzz37OfZzZ2dmesmeeeYbIyEhGjhzJiBEjGDZsGMYYmjdvztChQ7nxxhs9fQGYNWsWs2bNyrcvIiIiIiIiIiIiRZECzefJuUpHcTpDN2V53g+o/787nHfeeSd33nknx48fZ+nSpUydOpVRo0Zx1113sWrVKq666ioAqlWrxieffIK1lo0bNzJ79mw++OADBg0aRE5ODq+99hrgBL5zcnJISUk5r/3u0aMHPXr0IDU1ld9++41JkyYxZswYbr/9duLi4oiKivIE4YcPH06/fv3Oa39ERERERERERETOh6DTVxEpOkqVKkXLli155513eOmllzh16hTTpk3zq2eM4eqrr6Zv376eWcKTJ0/2rG/SpAmHDx9mw4YNF6TfERERtG/fntGjR9OrVy9SUlJYsGCBpy+A57OIiIiIiIiIiMilRoFmKfJ+/fVXTpw44VeelJQEQFhYGADr168nPj7+tPUA/vnPfwLw2GOPsXfvXr9tjh8/zpIlSwrV7+nTp5OVleVXfuDAAZ/+3Hjjjdx6661MnDiRMWPGBGxr3bp1nu1ERERERERERESKmv/dXAvyP+PZZ58lPj6eFi1aEB0dTUhICCtWrGD27NnUrFmTrl27AvDLL7/wzDPP0LRpU+rXr09UVBSJiYlMmTKFoKAgBgwY4GmzVatWDB48mBdffJErrriC9u3bU6tWLY4dO0ZCQgLz5s3jlltuYfr06Wfd765du1KiRAluueUWoqOjsdayYMECli1bRqNGjWjdurWn7hdffEHLli155JFHGDFiBI0bNyYiIoLExETWrl3L+vXrWbx4MVFRUWd/IkVERERERERERM4TBZqlyHvppZeYNGkSy5cv55dffiEoKIgaNWrw0ksv0b9/f8qVKwfA7bffTv/+/Zk/fz5TpkzhyJEjXHbZZbRp08YTgPb2r3/9i2bNmjFixAgWLlzIlClTCA8Pp2rVqjz++ON069atUP0ePHgwM2bMYOXKlfz888+UKFGCmjVrMmTIEJ544gnPgwnByS29YsUK3nvvPb7//nsmTJhAdnY2lStX5qqrrqJv375ce+21heqPiIiIiIiIiIjI+WKstRe7D0XKqFGjLMATTzxxsbtSIH+WhwGKiIiIiIiISNGgWITIJcdciJ0oR7OIiIiIiIiIiIiIFIoCzSIiIiIiIiIiIiJSKAo0i4iIiIiIiIiIiEihKNAsIiIiIiIiIiIiIoWiQLOIiIiIiIiIiIiIFIoCzSIiIiIiIiIiIiJSKAo0i4iIiIiIiIiIiEihKNAsIiIiIiIiIiIiIoWiQLOIiIiIiIiIiIiIFErwxe6AiIiIiIiIiIhcGhLTLQdOWnIsBBnnc7Uwc7G7JSJFgALNIiIiIiIiIiKSr/VpOUxMzGHTUetT/vzaLOqXMXSuFsQ14frDeZE/M90BREREREREREQkT3MO5PBmXLZfkNlt01HLm3HZzD2Qc4F7JiJFiQLNIiIiIiIiIiIS0Pq0HD7ekU3gEPMfLDB6Rzbr0xRsFvmzUqBZREREREREREQCmpiYc9ogs5sFJiUq0CzyZ1WkA83GmCHGmF+NMbuNMSeMMSnGmFXGmBhjTIU8tmlqjPnZVTfdGLPWGNPfGFPsQvdfRERERERERORSlZhu80yXkZe4o5bE9DPbRkT+NxTpQDPwT6AUMAsYDkwAsoCBwFpjTHXvysaYjsB84K/AJOADIAR4F/jqgvVaREREREREROQSd7ZpMJQ+Q+TPKfhid+A0ylprM3IXGmPeAF4CXgT6uMrKAqOBbKCFtXa5q/wVYDZwnzGmq7VWAWcRERERERERkdM4kX1htxORS1uRntEcKMjs8o3r9QqvsvuAisBX7iCzVxsvuz4+cc47KcTHx2OMoVevXsTHx9O1a1ciIyMpUaIEN954I1OnTvXb5uTJkwwePJjrrruOsLAwypYty6233so333xzTto/deoUI0aMoGHDhpQrV46wsDCio6Pp2LEjv/zyi1/9TZs20atXL6pXr05oaCiVKlWiW7dubN682a9ur169MMZNkyY5AAAgAElEQVQQHx/PRx99xLXXXkuJEiWoVKkSjz/+OGlpaX7brF27lgcffJDo6GhCQ0OpWLEiDRs2pH///mRmZvrUzcrKYuTIkTRp0oSyZcsSFhbGDTfcwPvvv09Ojn4VFhERERERkQuj5FkmIT3b7UTk0lbUZzTn5W7X61qvspau1+kB6s8H0oGmxphQa+3J89m5P6uEhARuvvlmateuzUMPPURKSgpff/21J7h72223AU4Q+Pbbb2fevHnUr1+fJ598kvT0dL777ju6dOnC6tWr+b//+7+zbh+cYPCXX37JNddcQ48ePShZsiR79+5l4cKFTJ8+ndatW3vqTp8+nc6dO5OZmcndd99NnTp1SExMZOLEifz000/MmTOHhg0b+vXn+eefZ8aMGdx99920bduWOXPmMHr0aLZt28bs2bM99dauXUvjxo0xxtChQwdq1arFkSNH2LZtGyNHjuT111+nePHiAJ4+zJgxg3r16tGtWzdKlCjBnDlz6Nu3L0uXLuWzzz47Z2MmIiIiIiIikpdrwoOAM5/w5GwnIn82l0Sg2RjzHFAaCAduBG7BCTIP9qpWz/W6Jff21tosY8xO4GqgNhB3Xjv8JzV37lwGDhxITEyMp6xbt27ccccdDB061BMIfvvtt5k3bx7t2rXjhx9+IDjY+RrGxMRw88038+abb3LXXXfRtGnTs2o/LS2Nr776ikaNGrF06VKKFfP9KfXQoUOe94cPH+bBBx8kLCyM+fPnc9VVV3nWbdiwgcaNG/Poo4+ycuVKv+NdsmQJ69ato0aNGoAzE7lly5bMmTOH33//nZtvvhmAcePGkZGRweTJk+nYsaNPG4cPHyYsLMzz+Y033mDGjBk89dRTDBs2zNP37OxsHn/8ccaMGcN9993n146IiIiIiIjIuVYtzFC/jDmjBwJeWcZQLcycx16JSFF1SQSageeASl6fpwO9rLUHvcrCXa/+eQt8yyPOcd8CMveMuBC78fH8WW5nJ/c7J/uvWbMmL7/8sk/Z7bffTo0aNfj99989ZWPGjMEYwzvvvOMJMgNERUXxyiuv8Oijj/Lxxx/7BZoL2r4xBmstoaGhBAX5/4paoUIFz/vx48eTmprK+++/7xNkBrj66qt57LHHGDZsGBs3bvRb/+qrr3qCzADBwcE8/PDDLFiwwCfQ7FayZEm/vpQrV87zPicnh/fff5/KlSvz7rvv+gTIixUrxttvv82nn37KhAkTFGgWERERERGRC6JztSDejMumIKFmA3SqptnMIn9WxQYOHHix+3BaAwcOfGvgwIGxsbGxHwK/AfcD/WJjY+cMHDhwH0BsbGxfoALw3sCBA1NytxEbG/soUB34eODAgYm51xtjHo+Njf3ozjvvrAJQt25d4uLiiIyMZNWqVVSpUoXly5dToUIFNmzYQGhoKHv27CE9PZ3jx4+zd+9eQkJC2LhxIxUrVuS1b5adxzNybj1wXRmstSQnJ3P48GGysrKIj4+ndOnSrFmzxnPslStXZsWKFZQtW5bt27eTk5NDSkoKCQkJjB07loYNG9K1a1efbSpVqsT48eNJSEigS5cupKamEhsbS6VKlXj++efZtm0bERERrF69mipVqrB7926++uorTpw4QY8ePYiPjyclJYWPP/6Ym266iQ4dOrBlyxbKlSvn2WbcuHEkJCTQoUMHwsLCSEpKYvPmzcybN4+vvvqKvXv3smPHDurUqcO6deuoWrUqy5cvJzIyktdee43t27dTunRpFi9ezMyZM5k6dSrz58/n22+/ZceOHWzbto2WLVty/PhxIiMj+eyzz9i8eTOPPfYYpUqV8hn/VatW8dNPP1GjRg3uuusuli9fTp06dRg/fjzffvsty5cvZ9++fRhjOHbsGKGhoaxfv57KlSszceJExo0bR/ny5UlJSeH7779n3rx5TJ06lRkzZrBw4UJWrVpFZmYmjRs35rLLLmPFihWecx0REcGWLVswxnDw4EFSU1PJzMwkISGBsLAw1q5d66nrfi1btizbtm3zGf/s7Gx27NhB6dKlWb16ted8Va5cmZUrV/qN/6FDh8jJyWHHjh2ULVvWMy7Lli2jcuXKrF69mlKlSrFz506ysrI4fPgwycnJAGzdutVn/JcvX05UVBRr1qwhLCyMhIQETp06RVpaGgcOHMAYw5YtWyhfvrzPdVmxYkXWrVtHiRIl2LVrFxkZGRw9epT9+/dTrFgxNm3aRIUKFXy2iYyMZP369T7X8rFjx9i7dy/FixcnLi6OihUrsnLlSp/rf+PGjRQvXpy9e/dy/PhxTpw4QWJiIiEhIWzYsIGoqChWrFjhOW/ly5dn06ZNFCtWjP3793P06FEyMjLYvXs3JUuWZN26dZ5ry33eypUrx+bNmwkKCiIpKYkjR45w6tQpEhISKFmyZMCxDA8PZ9u2bQA+4x/oWnZ/d9zjn5OTw6FDh3zGv0yZMn7X/8qVKylTpoxnLL3Hf/v27YSHh/uMZaVKlTzjHx8fT2ZmJmlpaRw8eNBn/FetWkXVqlVZtmwZUVFRrF27lrCwMHbt2sXJkyc5cuQISUlJBAUF+Y2/exv3+O/evZuMjAyOHTvGvn37CA4OLtS9PPf4x8XFBRx/97VcqVIln23cY5nX+K9du9Zn/HNfywcOHPAZf+9redmyZZ7vmXv8C3Mv976Wc49/pUqVWLVqlc/4e1/Lue/l3tdy7vH3vpa97+Xuazmv8fe+lleuXOlzL3dfy4mJiXmOv3ubZcuWERkZ6XMt5x5/97XsPZbly5cnLi6O4OBgz1jmHv/cYxnoWj558iS7du3yjL/u5bqX616ue7nu5bqX616ue3lRv5dnJO2mavApdh86Qpn0A1gTRKWUraSXKEf1pDWklbmMGvtWciwskk45cVwfFaZ7ue7lupcXsXv5ZZddFnseQ4sextqC//lDUWGMqYmTImOrtfYaV9kynLQaN1prVwTYZj1O6oyrrLV5ps4YNWqUBXjiicI9N/BizGg+W4Wd0RwfH0+tWrXo2bMnY8eO9VvfokUL5s2bh7WWxMREqlevzo033siyZf7B+IyMDEqWLEl0dDQ7d+484/bdTpw4wZAhQ/jiiy/YunUrACVKlOC+++7jrbfeolIlZ4J8mzZtAj4cMLexY8fSs2dPwMn/PG7cOHbu3El0dLRPvblz53LbbbcRExOD9484ixcv5o033mD27NmcOHECgHr16hETE8ODDz4IwKJFi7jllltO2xfvcyMiIiIiIiJyIaxPy2FSYg5xAdJoXFnG0KlakHIzixRdFySfzaWSOsOHtTbBGLMRuN4YE2mtTQY24wSa6wI+gWZjTDBQC8gCdlyQPp6jdBSnM3RTluf9gPpFfzjDw50MJ/v37w+4ft++fT71zlbJkiUZOHAgAwcOZPfu3cyfP5+xY8fy+eefEx8fz4IFC3z2s2bNGq677rpC7TM/f/nLX5g6dSonT55kxYoVTJ8+nffee49u3bpRsWJFWrdu7elLp06dmDhx4nnri4iIiIiIiMiZuibcCSQnpluGb80ix0KQgaevCFZOZhEB4FL+qamK6zXb9Trb9XpHgLp/BcKA36y1J893xyRvZcqU4fLLL2fPnj2emcbe5syZA0DDhg3P2T6rV69O9+7dmTFjBldccQULFy70PBCwSZMmAJ7A8/kWGhpK06ZNGTRoECNGOLPep0yZAkD9+vWJiIhgyZIlZGZmXpD+iIiIiIiIiJyJamGGqFBD5RLOq4LMIuJWZAPNxpj6xpjKAcqDjDFvAFE4gePDrlXfAclAV2PMjV71SwCvuz6OOs/dlgLo3bs31loGDBhAdna2pzw5OZnXXnvNU+dsHTx4kKVLl/qVHz9+nKNHjxIcHExISAgADz/8MBEREcTGxvo8UNAtJyeHuXPnnnVfwAlip6X5P6MyKSkJgLCwMMB5mGDfvn3Zt28f/fr186TY8LZv3z42btxYqP6IiIiIiIiIiIica0U518IdwFBjzHxgO3AIqAQ0B2oD+4HH3JWttUeMMY/hBJznGmO+AlKADkA9V/nXF/QIJKDnnnuOadOmMWXKFBo0aED79u1JT0/n22+/5cCBAzz//PMFylWclz179tCkSROuvPJKGjZsSPXq1Tly5AhTp05l//799OvXjzJlygBQoUIFvvvuOzp16kSTJk1o1aoVV199NUFBQezatYvFixdz6NAhMjIyzro/b7/9NjNnzqRFixbUrl2b0qVLs2HDBqZNm0a5cuV4/PHHPXVfeeUV1qxZw4cffsiPP/5Iy5YtqVq1KgcOHGDr1q0sWrSIN954g6uuuuqs+yMiIiIiIiIiInKuFeVA8y/Af4FmQAMgAjiO8xDAz4AR1toU7w2stZONMc2BfwP3AiWAbcAzrvqX3pMP/weFhIQwa9Ys3nnnHb744gvee+89goODadCgAcOGDfM8HO9sRUdHExsby9y5c5kzZw7JycmUL1+eevXqMXjwYLp27epTv1WrVqxdu5a33nqLGTNmsGDBAkJCQqhSpQotW7bk3nvvLVR/+vTpQ7ly5Vi6dCmLFi0iKyuLatWq0adPH5599llq1qzpqVu8eHEmT57M559/ztixY5k6dSrHjh2jYsWK1KpVi9dee43u3bsXqj8iIiIiIiJy6fN+ZlMgl8JznETkf4tR7NXXqFGjLMATTzxxsbtSIJfawwBFREREREREpPC84wGHT/0R2ykX4uRMPp8xAsUiRC45FySZuu4GIiIiIiIiIiKXGO8ArwK/IlIU6O4jIiIiIiIiInKJ6zRkqPOmdB4TF9974cJ1RkT+lBRoFhERERERERH5X3Lg0B/voypcvH6IyJ+KAs0iIiIiIiIiIpe4Sf8aALhSZ/Qd/McKzWQWkQsk6GJ3QEREREREREREREQubQo0i4iIiIiIiIiIiEihKNAsIiIiIiIiIiIiIoWiQLOIiIiIiIiIiIiIFIoCzSIiIiIiIiIiIiJSKAo0i4iIiIiIiIiIiEihKNAsIiIiIiIiIiIiIoWiQLOIiIiIiIiIiIiIFErwxe6AnLmhm7I877cfs37lA+prWEVEREREREREROTC0YzmS1z5kD8WufT16tULYwzx8fEXuytF3ty5czHGMHDgwIvdlQJr0aIFxpiL3Q0RERERERERkXNOgeZL0ID6wZ7l/64r7lncZRdSfHw8xhh69ep1QfdbEJmZmUyaNIlHHnmEa665hrJlyxIWFsa1117Lq6++ytGjR/PcNjExkd69e1OlShVCQ0OJjo6mf//+HD58+AIegci5UZSvUxERERERERH536AcC/I/a/v27XTu3JlSpUpx2223ceedd3Ls2DFmzJjBa6+9xtdff82iRYuIjIz0265p06YcOHCAjh07Ur9+fX7//XeGDx/O9OnTWbRoERUqVLhIRyWXsvHjx5Oenn6xuyEiIiIiIiIics4p0Cz/s8qUKcMHH3xAz549KVWqlKf81KlTdO7cmZ9++onY2Fjee+89n+369OnDgQMHGDFiBH379vWUP/PMM7z77rv8+9//5sMPP7xgxyH/O2rUqHGxuyAiIiIiIiIicl4odYactYEDB1KrVi0Axo0bhzHGs4wdO9ZTLycnhw8//JCbbrqJ0qVLU6pUKW666SZGjRpFTk6OX7vGGFq0aMHevXt56KGHiIqKomTJkjRq1IgvvviiwP2rWrUqffr08QkyA4SEhPDSSy8BTp5fbzt27GDmzJlER0fz5JNP+qyLjY2lVKlSfPbZZxw/frzA/VixYgX33nsvUVFRhIaGUrNmTfr06cO+ffvy3CYnJ4d33nmH+vXrU6JECapVq8Y///lPjhw54ld37dq1PPjgg0RHRxMaGkrFihVp2LAh/fv3JzMz06duVlYWI0eOpEmTJp5UIjfccAPvv/++31h4p1vYsmULXbp0ISoqiqCgIObOnUv9+vUJCQkhOTk54DEMHjwYYwwffPCBT3liYiJPPfUUtWvXJjQ0lAoVKtChQweWLVsWsJ2kpCQeeeQRKlWqRMmSJbn++usZN25cnucukDfffBNjDCNGjAi4fu/evRQrVoybbrrJp2zQoEE0a9aMypUrExISQpUqVejWrRtxcXF+bZzufEHgHM2nTp3i/fffp3379tSsWZPQ0FDKly9P69atmTZtWsD+RkdHEx0dTXp6OgMGDKBGjRqEhoZSp04dhgwZgrV/PCS0oNepiIiIiIiIiEhhaEaznLUWLVqQmprK8OHDadCgAffcc49n3fXXX+95/9BDD/HFF19QvXp1Hn30UYwxTJo0iT59+rBw4UImTJjg1/bhw4dp2rQpERERPPzww6SmpvLNN9/QvXt39uzZw4ABAwrV9+LFiwMQHOx7CcyePRuAtm3bEhTk+ztMmTJlaNasGTNnzmTJkiW0atXqtPuZOnUq9957L9Za7rvvPmrWrMmKFSsYNWoUU6ZMYdGiRURHR/tt989//pP58+fzwAMP0LFjR2bMmMGwYcNYsGABCxcupESJEoATZG7cuDHGGDp06ECtWrU4cuQI27ZtY+TIkbz++uueY83MzOTuu+9mxowZ1KtXj27dulGiRAnmzJlD3759Wbp0KZ999plfX7Zv307jxo2pW7cu3bt358SJE5QtW5aePXvy0ksv8eWXX/rM/HYbP348ISEhdO3a1VO2cuVK2rZtS0pKCrfffjudO3cmOTmZyZMnc8sttzBp0iTat2/vqX/o0CGaNm3Kjh07uOWWW7jlllvYt28f//jHP2jbtu1pz79bjx49ePnllxk3bhz9+vXzW//555+Tk5NDz549PWXz589n8ODB3Hbbbdx7772ULl2arVu38t133/HDDz+waNEiGjRoUODzlZeUlBSefvppmjZtSps2bahYsSL79u3jxx9/pH379owePZpHH33Ub7vMzEzatm3L3r17adeuHcHBwUyePJkXXniBjIwMYmJigIJfpyIiIiIiIiIihWKt1eK1jBw50o4cOdJKwezcudMCtmfPngHXf/HFFxawN9xwgz169Kin/NixY7ZRo0YWsBMmTPDZBrCAvf/++212dranfMeOHbZcuXK2ePHidvv27YXq9z/+8Q8L2BdeeMGn/LnnnrOAfeuttwJu9+STT1qgQN+Ro0eP2goVKtigoCA7f/58n3WDBw+2gG3Tpo1Pec+ePS1gK1SoYOPj4z3l2dnZtnPnzhawgwYN8pQ/88wzFrCTJ0/2239KSorP+YuJibGAfeqpp2xWVpanPCsry/bu3duvHffYAvbFF1/0az8xMdEGBQXZRo0a+a37/fffLWA7d+7sKcvMzLSXX365DQ0NtXPnzvWpv2fPHlulShVbuXJlm5GR4Sl/7LHHLGD79+/vU3/ZsmU2ODjYAjYmJsZv/4G0bdvWAnbdunV+66666ipbvHhxm5yc7ClLSkqyR44c8au7evVqW6pUKXvHHXf4lJ/ufFlrbfPmza1z2/1DRkaG3b17t1/d1NRUe/XVV9ty5crZ9PR0n3U1a9a0gG3Xrp3PuqSkJBseHm7Dw8PtqVOn/PqW13UqInIx/CcuM99FRERECs7vv6FPvfnHciH2JyJF3QWJq2pG8/nS4P6L3YOCW/PteWt6zJgxgJNGoXTp0p7yUqVKMWTIEFq3bs3HH39Mt27dfLYrVqwYQ4YM8ZlVXKtWLfr160dsbCyfffaZZ8bmmfrhhx/46KOPqFatGs8//7zPurS0NADCw8MDbusuT01NPe1+pkyZwqFDh3jwwQe59dZbfdY9++yzfPjhh8yaNYtdu3b55e59+umnqVmzpudzUFAQQ4cOZfLkyYwZM4ZXXnnFp37JkiX99l+uXDnP+5ycHN5//30qV67Mu+++S7FixTzrihUrxttvv82nn37KhAkT6Nixo087lSpVCniuq1atSqtWrZg1axYbNmzg6quv9qxzp7bwniH8008/sX37dp577jmaN2/u01aVKlV4/vnn6d+/P7/++ivt27cnMzOTCRMmUKZMGQYOHOhT/8Ybb6R79+5nlEKjZ8+ezJw5k3HjxjF06FBP+fLly9m4cSOdOnXyechjVFRUwHYaNGhAy5YtmTlzJpmZmZ4Z4255na+8hIaGUq1aNb/y8PBwevfuzbPPPsuyZcv461//6ldnxIgRPmMfFRVFx44dGT9+PJs3b+aaa64pcD9ERC6mw6f+SPlTLsTkU1NERERERIoqBZrlvFq5ciVBQUG0aNHCb13z5s0pVqwYq1at8ltXo0YNT15Zby1atCA2NjbgNgXx22+/0a1bN0qVKsX333/vE4wtCOvKfZs7z24gK1euBKBly5Z+64KDg/nrX/9KfHw8q1at8gs05w7EAtSuXZvq1asTHx9PamoqERERdOnSheHDh3PPPfdw33330bp1a5o1a8bll1/us+2WLVs4dOgQV1xxBa+//nrA/pYsWTJg7uEGDRoQGhoacJtevXoxa9Ysxo0bx3/+8x/AyTn81VdfUbFiRZ80GIsXLwYgISHBL3AMsHXrVgDi4uJo3749mzZtIj09nVtvvTVg4L9FixZnFGju1KkT4eHhfP755wwePNgTbHe30atXL79tfvrpJz788EOWL19OcnIyWVlZPuuTk5O57LLLfMryO1952bBhA0OHDmX+/Pns27ePjIwMn/V79uzx2yY8PJw6der4lVevXh1w0s+IiBRlA+r/8c/QoZuyApaLiIiIiMilQ/+Sl/MqLS2N8uXLExIS4rcuODiYyMhIDhw44LeuUqVKAdurXLmyp90ztXjxYtq1a0dQUBDTpk3j5ptv9qvjDmjm1b77YXx5zXj25m4jdyDSzV0eaHZ0fsefkJBAWloaERER3HzzzSxYsIA33niD7777zpNjuV69esTExPDggw8CTq5jcIK5sbGxefb52LFjAfeZl06dOlG2bFk+//xz3nzzTYoVK8bUqVM5dOgQ/fv398mB7e7Dt9/mP4Pe3Qf3+Tvdd6GgSpYsyQMPPMDo0aOZOXMm7dq1IzMzky+//JKKFSvSrl07n/ojRozg6aefply5crRp04YaNWoQFhaGMYbJkyezZs0aTp48Weh+LVmyhJYtW5KVlUWrVq3o0KEDZcuWJSgoiNWrVzNlypSA+4mIiAjYnvucZ2dnn1E/REREREREREQKQ4Hm8+U8pqO4lISHh5OSkhIwxUBWVhbJyckBH5SWlJQUsL39+/d72j0TCxYs4M477yQoKIgZM2bQpEmTgPXq1asHODOAA3HPuq1bt+5p9+nuo7vPue3bt8+nnrekpCRPX7wFOv6//OUvTJ06lZMnT7JixQqmT5/Oe++9R7du3ahYsSKtW7f21O/UqRMTJ048bd+95Td72x28/fjjj5k1axZ33HFHwLQZ3n2eMmUKHTp0OO1+3fVP9104Ez179mT06NGMGzeOdu3aeYLiTz/9tM/3Mysri5iYGCpXrszKlSv9fixwz84OpCCz3b29/vrrnDhxgjlz5vjN/H/zzTeZMmXKGbUnIiIiIiIiInIxBJ2+ikje3OkH8po9ecMNN5CTk8P8+fP91s2fP5/s7GwaNmzot27Xrl3Ex8f7lc+dO9fTbkHNnj2bdu3aERwczKxZs/IMMgPcdtttAMycOZOcnByfdUePHmXRokWULFky3zbc3H1099lbVlYWCxcuBAh4/PPmzfMr27FjB7t37yY6OjrgbNbQ0FCaNm3KoEGDGDFiBIAnSFm/fn0iIiJYsmQJmZmZp+37mXCnnBg3bhzJyclMmzaN6667juuvv96nnvucLViwoEDt1q9fn7CwMFavXh1whnmg83o6zZo144orrmDKlCmkpaXlGRRPTk4mNTWVpk2b+gWZjx075kmLci5s27aN8uXLB0wvE+h7cDZOd52KiIiIiIiIiBSWAs1SKOXKlcMYw65duwKu7927NwAvvvgi6enpnvL09HReeOEFAB555BG/7bKzs/nXv/7lE+zduXMnI0aMIDg4mL/97W8F6t/MmTO56667KFGiBL/++is33XRTvvUvv/xy2rZtS3x8PB988IHPupiYGI4fP06PHj0oVarUafd9zz33UL58eb788kuWLFnis27YsGHs2LGD1q1b++VnBhg+fDgJCQmezzk5OQwYMICcnBwefvhhT/mCBQsCBmHds4DDwsIAJ51C37592bdvH/369ePEiRN+2+zbt4+NGzee9rhy8w7ejho1iszMzID5jjt27Mjll1/OBx98wM8//xywrcWLF3u+J8WLF6d79+4cPXrUL6fz8uXLmTBhwhn3FZygckZGBiNHjuTnn3/muuuu8/vhIioqirCwMFasWOGTTiQzM5Onn36a5OTks9p3INHR0aSkpLB27Vqf8k8++YQZM2ack32c7joVERERERERESkspc6QQildujSNGzdmwYIFdO/enbp161KsWDE6dOjAddddR7du3ZgyZQrffPMNV199Nffcc48nx+3OnTt54IEH6N69u1+71113HUuXLqVRo0a0bduWtLQ0vv76a1JTU/nPf/7j97C7QDZv3kzHjh3JyMigffv2TJkyJWAagtxBzJEjR9K0aVP69evHr7/+ypVXXsnSpUuZM2cOdevW5Y033ijwuRkzZgz3338/zZs35/7776dGjRqsWLGCmTNnUrlyZT766KOA2zZr1ozrr7+eLl26EB4ezowZM1izZg2NGjXi+eef99R7++23mTlzJi1atKB27dqULl2aDRs2MG3aNMqVK8fjjz/uqfvKK6+wZs0aPvzwQ3788UdatmxJ1apVOXDgAFu3bmXRokW88cYbXHXVVQU6Pm89evw/e3ceHtdV33/8fUaStUuWLMmyLSfBWbAThwSSkgIpjhP2sMUk/PorlISShOTpjyTwlABtk9hOIUBaaIASstBQKGULTgplbYMJBEqQvMmSpdiSF0m2ZHmVtW/z/f0xGnm0ju5oNs18Xs8zz2ju3DtzNGfunZnv/Z7v+bTxGe8AACAASURBVAD33XcfDz74IJmZmfzFX/zFlHWysrLYsmULb37zm7n++ut57Wtfy+WXX05eXh6tra1UV1ezf/9+2tvbxwPkn/nMZ3juuef453/+Z2pqarj66qtpb2/ne9/7Hm9729v40Y9+FFFb77//fh544AGGh4enZDMD+Hw+7rrrLj772c9y6aWX8q53vYuhoSG2bt3KyZMnWb9+PVu3bvX83NO55557+MUvfsHVV1/Ne9/7XoqLi6mpqeGFF17gxhtv5Omnn573c4TbT0VERERERERE5s3MdAm5fPWrX7WvfvWrJnO3b98+e/vb326lpaXmnDPAnnrqqfH7R0dH7V/+5V/siiuusNzcXMvNzbVXvepV9pWvfMVGR0enPB5g69ats8OHD9v73vc+Ky8vt+zsbHvlK19p3/72t+fcrq1btxoQ9jKdlpYWu+WWW6yystKysrLsnHPOsbvuustOnDjh+fX54x//aO9+97utrKzMsrKybOXKlXbHHXfY4cOHp6x78803G2DNzc32j//4j/byl7/csrOzbfny5Xb33XdbV1fXhPV/8Ytf2C233GJr1qyxoqIiy8vLs4suusg+8pGP2MGDB6c8vt/vt29+85t27bXXWklJiWVlZdny5cvtda97nX3605+2lpaW8XUPHDhggN18881h/8dDhw6Zz+czwN7+9rfPuu7Ro0ftE5/4hF1yySWWm5tr+fn5dsEFF9h73vMe+9a3vmXDw8MT1m9vb7cPfvCDVlZWZjk5OXbZZZfZU089Nd6/DzzwQNj2TXbdddcZYJmZmdbR0THtOsPDw/ZP//RPtmbNGsvJybGlS5fa+9//fjt48OB4Px04cGB8/bm8XuvWrZv2PffjH//YrrrqKisoKLDi4mJ74xvfaM8//7w99dRTU/YnM7Nzzz3Xzj333Gmf44EHHjDAtm7dOmF5uP1URCSRPt8wPH4RERER7yZ8lu5rMbvho2bvuCtwva8l/APM5/lEZCGIS1zVmVk849pJ79FHHzWAO++8M9FNSVvOOdatWxdRDV4RERFZeB5uHBn/++OrNeBORETEq4cbRzhndx2v/eEPWdnQOHWFK9bAh2+Cqy6N2vMF6bNbZEFw8XgS1WgWEREREREREVnALv3VVm769GemDzIDbGuAOx6EZ34V34aJSFpRoFlEREREREREZKF6cTdvevwJfOFGrPsNNn8NXtwdn3aJSNpRoFlEREREREREZKF67Afhg8xBfoPH5z/huIjIdFRIR5KO6oaLiIiIiIiIzEFTK2xrwPBQgLVmT2C7C1bGsGEiko6U0SwiIiIiIiIishCNlcHwPMuXymeISAwo0CwiIiIiIiIishD19sd3OxGRWSjQLCIiIiIiIiKyEOXnxnc7EZFZKNAsIiIiIiIiIrIQXXUpAJ5nOhrbTkQkmhRoFhERERERERFZiC5YCVes8Vaj+cqLNRGgiMSEAs0iIiIiIiIiIgvVh2/C7+YYavY5uP3G2LZHRNKWAs0iIiIiIiIiIgvVVZfyy9tvCx9s9jm4/w6VzRCRmEnaQLNzbolz7lbn3DPOuSbnXL9zrss594Jz7kPOOd+k9c9zztksl+8m6n8REREREREREYmV3deu5wd/97e0XLxm+hWuvBi+dh/ccG18GyYiaSUz0Q2YxU3Ao0A7sBVoAZYCG4Angbc6524ys8k173cBz07zeHUxbKuIiIiIiIiISMK0XLqWlkvX8vHMdrj3izAyCpkZ8PmPqiaziMRFMgea9wLvBH5iZv7gQufc3wJ/BN5DIOj8w0nb7TSzjfFqpIiIiIiIiIhI0rhgJayomHhbRCQOkrZ0hpn9ysx+HBpkHlveAXxt7OY1cW+YiIiIiIiIiIiIiEyQzBnNsxkeux6Z5r7lzrkPA0uAE8D/mllt3FomIiIiIiIiIiIikmYWXKDZOZcJfGDs5s+nWeWNY5fQbX4N3GxmLbFtnYiIiIiIiIiIiEj6ydi4cWOi2+DJpk2bPg+8G/ipmW0OWb4IGAA+BnwS+BzwM+A8YD3wjk2bNv3rxo0bh6c8KOCcu33Tpk2PXX/99csBLrroIhoaGigrK2PHjh0sX76cmpoalixZQn19PdnZ2Rw+fJi+vj56e3s5cuQIixYtYs+ePZSXl7N9+/YJ2zQ0NJCVlcWRI0fo7e2lv7+ftrY2srOzqaurY+nSpRO2KSkp4aWXXiIjI4OOjg66u7sZGBigtbWV3NxcamtrqaysZNu2bePbLF68mL179+Kco7OzkzNnzjA0NMShQ4fIy8ujtraW5cuXU11dzYoVK6ipqaG4uJimpibMjOPHj3Pq1ClGRkY4ePAgBQUF7Nq1a/zxg89XVFREc3Mzfr+fkydPcuLECfx+P/v376ewsHDCNkuXLmXHjh0UFhZy4MABRkZGOHXqFMePHwegqamJxYsXs3PnzvFtKioq2LVrF/n5+Rw8eJDh4WG6urro7OzEOcfevXspKSmZsE15eTm1tbXk5eXR0tLC4OAgZ86c4ejRo2RkZNDY2MiSJUvYvn37+P9eVlZGXV0d2dnZtLW1MTAwQE9PD+3t7WRmZo73f3Cb6upqysrK2LNnz3hfTu7/+vp6KioqJvRlaWkpDQ0NZGZmjvfl5P6f3JfB/vf5fBw9epQzZ84wODhIS0vLeP8vW7Zsxv4/duwYp0+fZnh4eEr/19TUjF8XFRVN6f/R0VH2799PQUEBO3fuHH+9Kisr2b59+6z9X1RUNN4v1dXVVFZWsnPnTvLz86ft/3379s3Y/3l5eRw6dIihoaEp/V9aWjphvywvL2f37t3k5OTQ0tLCwMAA3d3ddHR0TOj/0G1C+z+4L/f09HDkyBGysrJoaGiYdl8O7f/QfTm0/7dt2zb+upWWltLY2Djjvrx79+4J/V9dXT1t/wf35WD/T+7L4L4MTOj/6fbl4Hsn2P9+v58TJ05M6P/J+3Kw/0P35dD+b25upri4eEJfLl26dLz/Q/flY8eOTej/HTt2jO9jFRUVM+7LPp9vSv8Htwn2f2tr64z7so7lOpbrWJ68x/Lq6hr8RWVktdWzekmujuU6lutYrmO5juUL8Fiu7+WJPZY3HjmOMyOzs5nLly+m5+mfk5OTw+mu02S9c33Uj+W7G17Cn19C1qFd/OmFK3Qs17Fcx/IkP5YvW7ZsUwRhWM+cmcXjeaLCOXcX8AjQCLzOzE7OYZtM4AXgKuAeM3tktvUfffRRA7jzzjvn32ARERERCevhxrPV0D6+esENuBMREUm4KZ+lH/ns2Tu//MnYP5+IJDsXjydZMEcD59xfEwgy7wGum0uQGcDMRpxzTxIINL9+7DFERERERERERKIqNAA7HQVlRSSVRXSEc87lAyuAMqAf6DSzw9Fs2KTnuwf4IlBHIMjc6fEhjo1d50e1YSIiIiIiIiIi0zg1dHYEecmiuCQTikiaSbaTW3N+Nufc+cBfEZho75WAb9L9J4CtwA+BLWY2+3869+f9BPBZYCfwRjM7HsHD/OnY9f5otElEREREREREZLLQoI7KS4hIPCXDya2wRzrn3JXAPwBvIBBcHgZ2Ax3ASSAXWAK8HLgJuBE45px7BPiCmQ1G2jjn3H3AZmAb8KbZymU4564CdpjZ0KTl1wIfHbv575G2RURERERERERERCRZJNvJrVmf1Tn3b8D7gC7gceC7wB/NbGCG9c8D3gzcDHwauMM595dm9huvDXPO3UwgyDwK/Ba4y7kp0fiDZvaNsb8/B1zinPs10Da27BXAtWN/32dmv/faDhERERERERERERGZXbjw9puAjwGPzSUz2cwOAo8BjznnXgFsAtYBngPNwMvGrjOAe2ZY53ngG2N/fwu4AfgT4K1AFnAU+D7wFTP7bQRtEBEREREREREREZEwwgWaV5lZfyQPbGa1wA3OudwIt98IbPSw/teBr0fyXCIiIiIiIiIiIiISOd9sd0YaZI72Y4iIiIiIiIiIiIhI8po10DxXzrks59wrnXMvj8bjiYiIiIiIiIiIiMjC4SnQ7Jx7r3Pu+8650pBl5wP1QA2wxzm3xTmXmKkNRURERERERERERCTuvGY0/xWw2sxOhiz7J+ACYCtQC7wL+GB0miciIiIiIiIiIiIiyc5roPlioDp4wzlXBLwN+L6ZvQF4NdCIAs0iIiIiIiIiIiIiacNroLkcaA+5/RogE/gugJkNA/8NnB+V1omIiIiIiIiIiIhI0vNaS7kbKA65vQ4w4IWQZQNA4TzbJSIiIiIiIiIic3TD5x4O/FHgoK7p7B0f+Wzg+sufjH+jRCSteA007wPe6pzLJhBgvgmoNbPjIeucC3RGqX0iIiIiIiIiIuJFRUmiWyAiachroPlx4CkCAedh4DzgnknrXAXUz7tlIiIiIiIiIiIyq7Y+o3PQePTuv8Hn4O4LM6nKc4luloikIU+BZjP7N+fcy4HbxxZ9ZewCgHPuWgLB569Gq4EiIiIiIiIiIjJRXZefLW1+GrttwvJ7a0dYXejYUOVjbbHXqblERCLnNaMZM/tb4G9nuPsFoATonU+jRERERERERERkels7/Ty5fxSb4f7GbuOhhlFuWwXXVCjYLCLx4TnQPBszGwKGovmYIiIiIiKSXh5uHJn1/o+vjurPGBGRBaWua/Ygc5ABT+wfpSwbZTaLSFzMeqRxzuXO9wmi8RgiIiIiIpKeTg3Z+EVERGBLmz9skDnIgGfa/LFsjojIuHCpAAeccw8BXzOzQS8P7Jy7DNgM1AAPRtg+ERERERFJM6EZy6HZzcpkFpF019ZnU2oyh9PQbbT1mSYInCeNthEJL9zYiV8CXwDanXOPOufWz5ah7Jxb5Zy70zn3v8B24DJga/SaKyIiIiIiIiKSnuq6IstOjnQ7mZ5G24hMb9bTLWb2Aefcl4DPALePXUadcw1AO3AKyAGWAC8HygAHHAX+Dvii10xoERERERERERGZqn80vtvJWRptIxJe2L3BzGqANznnLgQ+BFwHXA5cOmnVY8AW4IfAD81sOMptFRERERERERFJW7kZ8d1ORMSLOZ92MbN9wCcBnHN5wAoCmcz9QKeZtcekhSIiIiIiIiIiwtpiH+C9DEZgOxGR2Ioov9/M+oB9YxcREREREREREYmxqjzH6kLnaULAggz4TsvE2hkq9yAisaBTWiIiIiIiIiIiUdTWZ3QOGh0Dgeu2vuhNGrehyofzsH5+Jpq4TkTiQqewRERERERERESioK7Lz5Y2/5SM43trR1hd6NhQ5Zt3GYu1xT5uXQVP7h9lttCxA25blUH1ybOlNpTJLCKxpIxmEREREREREZF52trp56GG0RnLWjR2Gw81jPLrTu81lidbX+HjU2syWFM4fW7zmkLHp9ZkcE2Fwj4iEj86lSUiIiIiCRMcWuw38LnA7ao8LwOCRUREEq+uyx82wxjAgCf2j1KWPf8J+tYWB7Kj2/qMR/aNjH+W3n1hpj5LRSQhFGgWERERkbiLx9BiERGReNnS5g8bZA4y4Jk2f9Q+56ryHBXZbsJtEZFE0Ld3EREREYmreA4tFhERibW2PpvxM20mDd3RnSBQRCQZeAo0O+fOcc4VhVmn0Dl3zvyaJSIiIiKpyOvQ4rouBZtFRCS5RfpZpc84EUk1XjOaDwB3h1nnrrH1REREREQmiGRosYiISDLrH43vdiIiycproNmNXUREREREPNHQYhERSUW5GfHdTkQkWcWiRvNSoDcGjysiIiIiC5iGFouISCqKdFI/TXorIqkmM9wKzrkPTFp0+TTLADKAc4C/BHZHoW0iIiIikkI0tFhERFJRVZ5jdaHzNGpnTaGjKk8DxkUktYQNNAPfgPFSega8a+wyWfAI2QdsmnfLRERERCSlaGixiIikqg1VPh5qCD/ZLQSCJzdUKZtZRFLPXALNHxy7dsC/As8C/znNeqPACeB/zex0dJonIiIiIqkiMETYexkMDS0WEZFkt7bYx62r4Mn9swebHXDbqgx9tolISgobaDazfwv+7Zy7GXjWzL4Z01aJiIiISMrR0GIREUll6yt8lGfDM21+Gqb5rFtT6Lihyqcgs4ikrLlkNI8zs/WxaoiIiIiIpD4NLRYRkVS2tjgQSG7rMx7ZN4LfwOfg7gszdeJURFKep2/uzrlXOefud84tneH+yrH7L49O80REREQklQSGFmcQ7qe2hhaLiMhCVpXnqMh2VOYErhVkFpF04CmjGfgb4GrgwRnuPwp8CLgA+MA82iUiIiIiKUpDi0VEREREUo/XQPNrgK1mNu1oRzMz59yvgNfPu2UiIiIikrI0tFhEREREJLV4DTRXAm1h1jkCLIusOSIiIiKSToJDi0Nvi4iIiIjIwuN1PGIfUB5mnXJgMLLmiIiIiIiIiIiIiMhC4zWjeSfwLufcx8ysZ/Kdzrki4F1j64mIiIiIiIiIiIgsaA83jsx6/8dXew2xpiavGc2PE8hY/m/n3CtC73DOXQb8EigbW29enHNLnHO3Oueecc41Oef6nXNdzrkXnHMfcs5N23bn3Gudcz91zp10zvU552qdc/c45zLm2yYRERERERERERFJX6eGbPySLNr6jM5Bo2MgcN3Wl5i2eQq3m9n3nHNvBT4A7HDOHQUOAyuApYAD/s3MvhOFtt0EPAq0A1uBlrHn2AA8CbzVOXdT6MSEzrl3AT8EBoDvASeBdwBfBF439pgiIiIiIiIiIiIicxKasRya3ZzoTOa6Lj9b2vw0dk8MLN9bO8LqQseGqsAE3PHi+dUws1ucc78HPgJcQmCCQIA64Etm9mSU2rYXeCfwEzPzBxc65/4W+CPwHgJB5x+OLS8CngBGgWvMrGZs+X3Ar4AbnXN/bmbfjVL7REREREREREREROJua6efJ/ePMlPucmO38VDDKLetgmsq4hNsjijsbmaPA4875/KAxcBpM+uLZsPM7FczLO9wzn0N+DRwDWOBZuBGAmU9vhkMMo+tP+Cc+3vgOeBOQIFmERERERERERGRheIjn539/i9/Mj7tSBJ1XbMHmYMMeGL/KI/vH73uP/4067lYt2te4Wwz6zOzI9EOMs/B8Nh1aCXua8eufz7N+r8B+oDXOueyY9kwERERERERERERiZHOE2cvaWpLmz9skDlobL37YtaYEBFlNDvnygmUrlgD5JvZrSHLXwbsNrP+qLVy4nNnEqgRDRODyi8fu947eRszG3HOHSBQ6mMV0BCLtomIiIiIiIiIiEiUhWYsh2Y3p1kmMwQm/ptck3kO1sWiLZNlbNy40dMGzrkPAb8A3gVcBVy+cePGzQCbNm1aBdQABzdu3Lgjuk0N2LRp0+eBdwM/NbPNIcs/DiwBHt64cePJabZ7H7AS+PbGjRvbJt/vnLt906ZNj11//fXLAS666CIaGhooKytjx44dLF++nJqaGpYsWUJ9fT3Z2dkcPnyYvr4+ent7OXLkCIsWLWLPnj2Ul5ezffv2Cds0NDSQlZXFkSNH6O3tpb+/n7a2NrKzs6mrq2Pp0qUTtikpKeGll14iIyODjo4Ouru7GRgYoLW1ldzcXGpra6msrGTbtm3j2yxevJi9e/finKOzs5MzZ84wNDTEoUOHyMvLo7a2luXLl1NdXc2KFSuoqamhuLiYpqYmzIzjx49z6tQpRkZGOHjwIAUFBezatWv88YPPV1RURHNzM36/n5MnT3LixAn8fj/79++nsLBwwjZLly5lx44dFBYWcuDAAUZGRjh16hTHjx8HoKmpicWLF7Nz587xbSoqKti1axf5+fkcPHiQ4eFhurq66OzsxDnH3r17KSkpmbBNeXk5tbW15OXl0dLSwuDgIGfOnOHo0aNkZGTQ2NjIkiVL2L59+/j/XlZWRl1dHdnZ2bS1tTEwMEBPTw/t7e1kZmaO939wm+rqasrKytizZ894X07u//r6eioqKib0ZWlpKQ0NDWRmZo735eT+n9yXwf73+XwcPXqUM2fOMDg4SEtLy3j/L1u2bMb+P3bsGKdPn2Z4eHhK/9fU1IxfFxUVTen/0dFR9u/fT0FBATt37hx/vSorK9m+ffus/V9UVDTeL9XV1VRWVrJz507y8/On7f99+/bN2P95eXkcOnSIoaGhKf1fWlo6Yb8sLy9n9+7d5OTk0NLSwsDAAN3d3XR0dEzo/9BtQvs/uC/39PRw5MgRsrKyaGhomHZfDu3/0H05tP+3bds2/rqVlpbS2Ng44768e/fuCf1fXV09bf8H9+Vg/0/uy+C+DEzo/+n25eB7J9j/fr+fEydOTOj/yftysP9D9+XQ/m9ubqa4uHhCXy5dunS8/0P35WPHjk3o/x07dozvYxUVFTPuyz6fb0r/B7cJ9n9ra+uM+7KO5TqW61ievMfy6uoa/EVlZLXVs3pJro7lOpZP2Jf39vrIan8Jf0EZ2S07dSzXsVzH8iQ9lut7+dRj+e7WTnx9XbjREYpPH4rpsbzxyHGcGZmdzVy+fGJfHswsJ6ulFsvOY8mZlqgcy3c3vIQ/v4SefTupoZLq6mp+O7iE6p27eaErixf3tvK/7X2cm9Gbkt/Lm4ZyyGqtw1+yjEUHtulYHudjed+WX+KcY2BwkI5Xnh/3Y3nTUC5Zrbvxlywj++D2uB/L9w7l0dl6EJ9/hLyB0xT0BTK7y081059dTNXRWroKl3FO+3a688tZ0bmbMwWVvKcqY5OnIGwEnNncI+DOuTcSyCKuBR4A3gzcYWYZIevUAofM7B1RbivOubuAR4BG4HVmdjLkvr3AhcCFZtY0zba/B14DvMbM/jDTczz66KMGcOedd0a59SIiIiIynWSauVuSj94fIrJQxfP4NdtzxaIdwcds7jHOL3AAnBo6G18qWeSi+nzJRp9NCZbgjOZE9/8zbaP8oM3vebv/+NMsF4PmTOD11fgE0A6sM7MzzrlXTrNOLYGAblQ55/6aQJB5D3BdaJB5TNfYdfEMD1E0aT0REREREREREYnQ+QVuPNCW6OCbxE5o305H/R1fuRnh10kUr5MBXgn8l5mdmWWdNqAy8iZN5Zy7B/gKUAesN7OOaVZ7aez6omm2zyRQO3oE2B/NtomIiIiIiIiIiKSDU0M2fpHEWFvsNZwbP15POSwCesOssxgYjaw5UznnPgF8FtgJvNHMjs+w6q+A9wFvAb4z6b7XA3nAb8xsMFptExERERGR2GnrMzoHDb+BzwVuV+XFfNSniIiIhAjNWFbmeuJV5TlWFzqvEwI+H6v2hPIaAj8IXBFmnas4m108L865+wgEmbcRKJcxU5AZ4GngOPDnzrkrQx4jB/iHsZuPRqNdIiIiIiISO3VdfjbXj3Bv7QiH+6F9AA73w721I2yuH6Guy3tdQhERkWgIngTtGAhct/Ups1fib0OVj7meeh9b78GYNSaE11MP/wnc65y7ycx+MPlO59wHgVcAfzffhjnnbgY2E8iO/i1wl3NTXsKDZvYNgLGa0bcRCDj/2jn3XeAk8E7g5WPLvzffdomIiIiISOxs7fTz5P5RZvrZ3thtPNQwym2r4JqK5B06KiIiqaWuy8+WNv+ULNJ7a0dYXejYUOVL6pIGKaOpFQ53wsgoZGYEbl+wMtGtiru1xT5uXcWs35kgEGS+bVUG11T4notHu7wGmj8P/DnwHefcjYxNvOec+3/AnwEbgH3Al6PQtpeNXWcA98ywzvPAN4I3zOxZ59w6AoHu9wA5QBPwMeBLZqbTTCIiIiIiSaqua/Ygc5ABT+wfpSw7uesUiogkmkoQRYdOgiaBF3fDYz+AbQ0Tl7/nY3DFGvjwTXDVpYlpW4Ksr/BRng3PtPlpmKaMxppCxw1xPgHiKdBsZqfGArnfBG4KuetLY9e/Bf7CzMLVcZ7Lc20ENkaw3e+At833+UVEREREJL62tPnDBpmDjMAPKwWaRUSmUvZt9OgkaBLY8hw8+Bj4Z+iFbQ1wx4Nw/x1ww7XxbVuCrS0O7MttfcYj+0bGTyrdfWFmQk4qea7abWYtwDXOuVcArwGWAF3AH8xsW5TbJyIiIiIiaaCtz7xOakNDtyk7T0RkEmXfRpdOgibYi7tnDzIH+Q02fw2Wl6ddZjMEJgisyHYTbifCrIFm59wW4Ltm9v2x268nUBe5xcxqgdo4tFFERERERFJcpBP81XX5qcrLiHJrREQWpu5hU/ZtFOkkaBJ47Afhg8xBfoPHn07LQHOyCHc0eTewOuT2VuCWmLVGRERERETSUv9ofLcTEUlFHQN4zr6Vmc3nJKhEQVPr1JrM4dTsCWwnCRGudMZpoCjktk7HiIiIiIhI1OVGmJQc6XayMD3cODLr/R9f7bk6pEjK6B81ejyefFP27ex0EjTBXtwd+XYXrIxuW8Zogs3ZhfsUbgD+r3OuGmgfW3beWAmNWZnZb+bbOBERERERSQ+BodveM8A05Dt9nRo6m7dZskg/8iU5hJ4Mae6xKctjeTKke/bzMDNSCaKZ6SRogvX2x3e7WWiCzbkJd4TbBDwL/EfIspvHLuFotxIRERERkTmpynOsLnSeamGuKXTKIkozoUG60ICeMpklGZUuiu/zzbWM7WTKvp2ZToImWH5ufLebgSbYnLtZP43N7JfOuTXAG4AVwEbg+bGLiIiIiIhI1Gyo8vFQQ/hJrCBQ0++GqvT+MSciySeRJz18EZ53U/btzHQSNMEindQvipMB1nXNHmQO0gSbAWGPgGZ2CPg6gHNuI/BrM9sc43aJiIiIiEiaWVvs49ZVhP1B54DbVmWk9Q85EZHJCiOMcetYOjudBE2gC1bCFWu8TQh45cVRrc+8pc3veYLNdN6nZv3PnXNfcM69KWTRBwmU0hAREREREYm69RU+PrUmgzWF02eDrSl0fGpNRtoPTRURmSw3w1HgMTtZ2bfhBU6CZhDuVdJJ0Bj58E1zT9f3Obj9xqg9dVufecpmh7MTbKarcOe77gFOA78cu/2vBOo218ayUSIiIiIikr7WFgcm1GnrNACFIwAAIABJREFUMx7ZNzI+s/vdF2amXEAktM7wdFR7WES8qMyB5l7inn3b1md0Dtr48bqtz1LqeL2+wkd5diBbtWGawOOaQscNmgwuNq66FO77MDz42OyFyH0O7r8j6mUzIt0uXSfYDPetpQfIC7mdOkcJERERERFJalV5jopsN+F2Kjs1dPYHdMmi1P5fRSQ6giermnvOHj9W5kJL/+zbRSv7tnvY2Fw/MiXr897aEVYXOjakUPA1nU6CJp0N18GKCnj8aajZM/X+Ky8OZDJHMcgMkU+Umc4TbIYLNDcBG5xzzwDtY8sWO+fOCffAZtYy38aJiIiIiIikstCM5dDsZmUyi4hXw2PJlz4HF+RDxwD0TBPwijT7NvQY1dxjDPlhwA8z5U83dhsPNYxy2ypSqtxRup0ETRpXXRq4NLXCvV+EkVHIzIDPfzSqNZlDRTpRZjpPsBnu28vDwL8Dvw9ZdvfYZTY2h8cWEREREREREZEIBU9KPdw4Mp7VXLLIjS+PVfZtbgacmb3yDxAIDj2xf5SybE06KFFywcpAdnPo7RgJvGe9l89I5/f6rMFgM/uOc+4AcD2wAriFQH3mnbFvmoiIiIiISGpQLWYRmY9wx5CZRDP7NvQ4tbl+hKODc5vwzAjUNk7n4JssTFV5jtWFztOEgOk+wWbYbzNm9gfgDwDOuVuAZ8xsc4zbJSIiIiIikpJUi1lE5iPRx5C2PvMUeANo6LaUmyBQ0sOGKh8PNYzGfYLNhcrrafMPAjti0RAREREREZFUpVrMIjIf4Y4hkWY8R6Kuy3spgeB2VXlpXLxWFqS1xT5uXQVP7p892BytCTYXOk/faszs32LVEBERERERERERSW7900wwGMvtJHm09RmdgzZe8ztdstTXV/gozw6UgGmYJps/0gk2U5Hn0+fOOR/w18D7gDVAvplljt33SuA24J/NbG80GyoiIiIiIiIiIomVG2FScqTbSeLVdfnZ0uafUjLl3toRVhc6NqRBkHVtceB/jNUEm6nCU6DZObcI+BlwDXAS6AYKQlY5APwVcAx4IDpNFBERERERERGRZBAIKHovn+E1EKlJVJPD1k7/rGUjGruNhxpGuW0VXFOR2sFmiO4Em6nI6175cWA9sBH4B+B+4L7gnWZ22jn3G+DNKNAsIiIiIiIiIpJSqvIcBRnQ46EUxppCN6+AXKInQExXdV2zB5mDDHhi/yhl2d5PKMzZRz4b0rCmqcu//MnYPK944jXQ/D7gd2a2GcA5N9177QDwjvk2TEREROJHGSMiIrKQpGudUJFkUZkDTb1zW9cBN1R5Dz5qEtXE29LmDxtkDjICNYzjUkKjoiT2zyER8bp3vgz4SZh1TgKlkTVHREREEk0ZIyIikqxUJ1QkOWy+NCtsSQUIBJlvW5Wh/XIBauuzKcfacBq6LXYn/pSxvCB4DTT3A4vDrHMOcDqy5oiIiEgiKGNERESSneqEiiSX9RU+yrPhS3tHpy2jUZABd12kIPNCVdflvQ53cLuqPM38mK68/nrcCbzJObfIzIYm3+mcKyZQn/n30WiciIiIiIiISFLVCRVJEs09Np4g0Nxj9I1OXB6PhIG1xT4uLPTTP2oc6IXSRXByCF6WD7kZTvvhAtbvoQZ3NLaT1OB1j38CWAl82zlXFHqHc24x8A2gBPhaVFonIiIiIiIiaS+SOqEi6WTYD34LXBIhN8OR7YPKnMB1bobKry10uREmJUe6naQGT6e3zOw7zrk3AB8E3gmcAnDO1QCXANnAv5jZT6PdUBEREREREUk/SVcnVCRJnF/gxrOW76gZZtimLheJVCAb3ftJO2WxpzfPvW9mHwL+CtgDlBOo7f4qoAn4kJl9JKotFBERERERkbQ1nzqhIiISmao8x+pCbyfr1hQ6neBLcxGdZjCzb5jZK4ECoAooNLNLzeypqLZORERERERE0prqhIqEN2pnS2d0DgYy+kXma0OVj7mGjR1wQ5WymdPdvMZSmFk/0B+ltoiIiIiIiIhMoDqhIjOr6/Kzpc1Pb8iJlcP9cG/tCKsLHRvSMPAXnCBxJiorMndri33cuoqwk7E64LZVGSqbIfMLNIuIiIiIiMj8tfUZnYOG38DnUH3hEKoTKjK9E4PGQw0zBwAbuwP3r8yFJdnpeTw5NXT21SlZlJ6vwXytr/BRnh2YZLVhmnr5awodN1T5UvKYO9NJi+aewOtwfoHeU5Mp0CwiIiIiIpIgwWzEyZPdhWYjpuKPdy+CdUK9TAioOqGSqoInpfpH4ORw+PUNaOmHRT6jMCs99onQjOXQQKEymSO3tjjwWdTWZzyyb2T8pOjdF2amzbE29KSFzEx7mYiIiIiISIwFgx39o8aB3sCyUYMdp2cuJBzMRrxtFVxTkd7B5g1VvlkzN0OpTqikoplOSj3/g5/Out26m94GQMcAFGbFrHmSJqryHBUh2fGpHmSe6aRFyaJEtGZhUKBZREREREQkxrqHjY4B6PE4QZ0BT+wfpSw7vUtBqE6opLOtnf6w732A/p7e8b9zC/In3NczGjjRJSLzs6S1lXPr6lnU18dQXh5kXgYXrEx0s5KGAs0iIiIiIiIxtLXTT3Mvc8rGnY4RqI2Z7sHTdKsTqgnNBAKZzLMFmYMZyzAxuzl0eVD37G+ppKUa9pIMztldx2t/+ENWNjROvOMbwBVr4MM3wVWXJqJpSUWfTCIiIiIiIjESLkg0Vw3dpuAK6VsnVBOapa8tbf55Hz+C/AssoVk17CVpbHmO9/7D13AETv5OOQpva4A7HoT774Abro1/+5JIVALNzrky4M+APuB/zMzjgDAREREREZHUE80gUV2Xn6q8jCg92sKWDnVCNaGZtPWZp0kww/HFYTfpHzUG/dAxYPPKPg5XLkQ17CVuXtwNDz42Hlye8d3sN9j8NVhentaZzZ4+oZxzdwK3AG81s5Njy64Afg6Ujq1W45y71sx6p38UEREREZHkoiHqEgvRDhL1K51HJK3Udfmj+niFMfwo6x42ekegsTtwu30gcB1J9vFcR4Kohr3ExWM/mPtwAL/B40+ndaDZ6574fwALBpnHPAyUAE8BPwX+BLgjOs0TERGReArWwOsYCFy39S2wMZYiUXBqyMYvIvMR7SBRrpKZRdJKNE8uFWRAbkZsUppPDBpNvTBTc4PZx7/unNsx0ctIkGANe5GYaGoNlMXwomZPYLs05fV81oXAT4I3xkpmrAOeNLMPjy17EfgL4J+i1UgRERGJLdXAk3SnIeoSC9HOQNZxWCS9RPPkUmVO9B4r1Najo7T0h19vrtnHkYwEUQ17iZkXd0e+3QUro9uWBcLrN5UlQGfI7deNXT8Tsuy3wLnzaVSQc+5G59yXnXO/dc6dcc6Zc+7fZ1j3vLH7Z7p8NxptEhERSTVbO/081DA645d6r1koIiISEM0g0ZpCpyCKSJqJxsklB5yTC4VZ0T1+1HX52Vw/whMH5v79cC7Zx5GOBIn2CBIRgKMnIqwK3DuHsy8pymuKxkmgLOT2OsAP/D5kmQHROlf298BlQA/QBqyewza7gGenWV4XpTaJiIikDNXAExGJncDxcv7BDwfcUKVjr0i6qcpzrC50Edd6PycP3n9uBj9rj24QNtxEfbMJl30c6UiQ6bYLloTzG/OamFDS09ZOP609OXwgko3zc6PdnAXDa6C5AXiHc+7vCJTf+T9AtZmdCVnnPKAjOs3jowQCzE0Egtpb57DNTjPbGKXnFxERSSrRnrAskhp4CjSLyHyly+SL8w0SQSDIfNuqDB17RRIgGY5VG6p8PNQQWVC3tQ+OD0a3PXNNUgj3GFV50w/5iHQkSOh2KgmX+mK9bwbf58tXrwUCv4M8naLQZIBz9giwjEDwtxWoBL4avNM5lwFcTSCreN7MbKuZ7TMzzcQiIiIyyXwnLJtPDTwRkWhJ9ckXN1T5vP04DbGm0PGpNRlcU6GAiEiiJepYtbbYx62rMiI6jgRHpHUPR6/NXpIUZjJb1nKkAeDgdioJl35isW8G3+eHl1fRcOFqb/vflRenbX1m8JjRbGY/cs7dAdw+tujbZhZaM/kNBMpm/CJK7YvEcufchwnUkz4B/K+Z1SawPSIiIlETzQnL5lMDb6YsFBGRuUinyRcDQSLmlAFYkgU5GYEh3ndfmKkh3pKUkiHLN16S5Vi1vsJHeXZgZFmDxyQBAzoGoDBr/u2IJElhOrNlLUcyEiRYw14l4dJHLPfNye/zLddv4FOPPIRvLjmwPge33zjvNixknnvAzB4HHp/hvl8AJfNt1Dy9cewyzjn3a+BmM2tJSItERESSUDRr4Il4FfqjoLnHpixPpUCFCDCnDMDCTFiSHQguK8gsC0FoBmHJIr1nY2ltcaDcQ81JP1/Y6+3LWM8o9I8auRnz66NoTbgXLrjrpVxIaA17lYSTaJj8Pq9fs5Yn338rt33riVkzm8053P13xL1sRrJ9p87YuHFjXJ8wUps2bToPuAXYvXHjxi3T3L8IGAA+BnwS+BzwMwI1o9cD79i0adO/bty4cXi6x3fO3b5p06bHrr/++uUAF110EQ0NDZSVlbFjxw6WL19OTU0NS5Ysob6+nuzsbA4fPkxfXx+9vb0cOXKERYsWsWfPHsrLy9m+ffuEbRoaGsjKyuLIkSP09vbS399PW1sb2dnZ1NXVsXTp0gnblJSU8NJLL5GRkUFHRwfd3d0MDAzQ2tpKbm4utbW1VFZWsm3btvFtFi9ezN69e3HO0dnZyZkzZxgaGuLQoUPk5eVRW1vL8uXLqa6uZsWKFdTU1FBcXExTUxNmxvHjxzl16hQjIyMcPHiQgoICdu3aNf74wecrKiqiubkZv9/PyZMnOXHiBH6/n/3791NYWDhhm6VLl7Jjxw4KCws5cOAAIyMjnDp1iuPHjwPQ1NTE4sWL2blz5/g2FRUV7Nq1i/z8fA4ePMjw8DBdXV10dnbinGPv3r2UlJRM2Ka8vJza2lry8vJoaWlhcHCQM2fOcPToUTIyMmhsbGTJkiVs3759/H8vKyujrq6O7Oxs2traGBgYoKenh/b2djIzM8f7P7hNdXU1ZWVl7NmzZ7wvJ/d/fX09FRUVE/qytLSUhoYGMjMzx/tycv9P7stg//t8Po4ePcqZM2cYHBykpaVlvP+XLVs2Y/8fO3aM06dPMzw8PKX/a2pqxq+Lioqm9P/o6Cj79++noKCAnTt3jr9elZWVbN++fdb+LyoqGu+X6upqKisr2blzJ/n5+dP2/759+2bs/7y8PA4dOsTQ0NCU/i8tLZ2wX5aXl7N7925ycnJoaWlhYGCA7u5uOjo6JvR/6Dah/R/cl3t6ejhy5AhZWVk0NDRMuy+H9n/ovhza/9u2bRt/3UpLS2lsbJxxX969e/eE/q+urp62/4P7crD/J/dlcF8GJvT/dPty8L0T7H+/38+JEycm9P/kfTnY/6H7cmj/Nzc3U1xcPKEvly5dOt7/ofvysWPHJvT/jh07xvexioqKGfdln883pf+D2wT7v7W1dcZ9ORWP5Qezl5F1YDuWV8Tik/s9H8v37mvmcN8o+QMnye8/hbNRyk4fZCC7kKqjtXQVLuOc9u2cKVjKyqO7GFyUz5LTh3hZ7ij5g6d1LNexfF7H8mYWk3VoF/6SZRS3bCejuIyKznoysnPJPNnGOVmDCTuWH8goI+vgDvwly8k+uF3H8jgcy/2F5WS11bN6Se6UY/neXh9Z7S/hLygju2XngjuW1xw8ytN7T5HhH6K0q5WhrFxWdNaNH2OD1/05RSzq3E8WRnbvCSpHT+t7eZhj+fbde7DMLDJOt7Myoz/lj+X+kuVkHdjOq89fltBj+YX+41xdnkHZySaOZCymqK2WRWXLWd+zk3evWZqy38t3H+rAN3AGNzLE4q7WhB3Lj+VUcGrvDgYXFVJ2+iA+GyVv4BT5/SfY29RO9nAfo74s/qy05+x3ufwKKjp2k5Wbx5IzLREfy1/qPMPJY0cx52PpyX305ZSw8uiu8efpzitj+bF6RjJzKOk+TNbIANnDPRT1HMXvy2Dpib1UVZSxrGPXrMfy/NE+invaaR7IpPJEI915ZZzTsXP8eXpzS6k88RJ+XxYbCjo5L6OPQ6f6+P2+NkYyF7H8WAPd+RWc07EjZJsSlp7Yi9/no6i3k5yhbk73DbCku42SAu/fy5uGcshqrcNfsoxFB7alVYzlpV5HVvte/PmlLDq0M+Hfy/d2nsEND5Bx6jBry3LnfSw/kb+Mnpe20ZdTTMWpJgzHRfsbuHRPQ9gSGoOvuZS9vuG4fi+vP3ISX/dxHEbB8WYyC4pZ2l6LrzTwmXHly5aya9culi1btilM86PCeSl/7Jx7JfAaAiUzusaW5ROo0/wuoA/4nJk9EvWGOncNgckAv21m7/ewXSbwAnAVcE+4tj366KMGcOedd0beWBERkTiY7zCxtj7j3trZh79O5/Ov0HBuSW2pXsoh2YR7vRd6f2yuH/E0BLwgAy4sdAvyf423hf7e8CJZ/9dkbVcsJMv/+kzbKD9om5pZ3HXsJC88+0vM/Djn4+p3v4ni8tLx+5flQGXO/I4tP28f5ZuH5pfVfEE+FGZN/B45U5vquvwzlgtZU+i4IWRSv0jb9oFzfbxlmfeScMnyfkiEZPvfo92eye+lSxrqvJXO+Np9yToZYFx+wHntgU8Af2ZmXw1Z9hDwl0APgbrIX3DONZjZL6PUxnkxsxHn3JMEAs2vJzChoYiISNqbTw08EREJL5J6psEh7iIi05lc37jjYBt1v6vhWGv7hOU//fr3KF+5jLWvu5LK86rwReHrWyCoG3mg+ZxcGLGzJVfClVsJlgtp6zMe2TeC32auYa+ScBItk9/nG36yZW5BZgC/weNPJ2ugOS68BpqvBH4dvOGcywJuBv4IXAOUAjuAu4CkCDSPOTZ2nZ/QVoiIiCSZSGvgxVo6TTQ0G70OC5v6TyKtZ9rtfbCJiKSJ0CBY864G/vjz52GGINix1na2fu+/ePVb1rH66jXzfu5IkhRgYvZxJNmnVXmOimw34fZks00wOJtIt5PUFfo+X3GkjTX7GjE8pAPX7IGmVrhgZQxbmby8/lqsAFpDbl8JFAKPmdmAmR0B/hN4RZTaFy1/Ona9P6GtEBERSTJri33cuioj7BcnB9y2KiMhE6acGrLxSzrT67Cwqf/SU6SZcn69TUSSSluf0TlodAwErtv6EreTVuU5CjICmcyzBZnHmVH98+dpO3A4Ks+/ocrnafz9bat83HdJZsy/Q0b6+JFs93DjCM09Nn55uHFkwkUWvuD7fG1jHRBBzYkXd0e7SQuG1zQKm7TN1WPLng9Zdgwon2e7PHPOXQXsMLOhScuvBT46dvPf490uERGRZLe+wkd5NnOugRcPoRkuyVYHLp70Oixs6j+JNFMuGkPcJXUEg5zBsgFtfaYyVnFS1+VnS5t/SgbvvbUjFGRAZc7UesNBsTzWV+bAs7+rCR9kHmNmPPdcDRdeUDXv5w4kKcCT+8OPiDsnF9ZXxCdlOFEl4Yb9cy8FIgtH8H1+/Cf9kT1Ab4TbpQCvR74WzmYHQ2ACwDYzC80UXg6cmm/DAJxz7wbePXazcuz6Nc65b4z9fdzM/mbs788Blzjnfg20jS17BXDt2N/3mdnvo9EuERGRVOOlBp6IiMxNpPVMC3UuQpg9yLm60LEhzieB083WTv+swdSeUWjqhXNybcLJoXgEG3tPnppSkzmc/QeO0HH0JKyumPfzh0tSCBeEj5V4lYT7+OrM8RPIzT023uc6kZxa1lf4OHJehBV483Oj25gFxOte8H1gk3PuaWAAeA3wz5PWWQs0R6FtAJcTqAEdatXYBeAQEAw0fwu4AfgT4K1AFnB0rM1fMbPfRqlNIiIiKWsuNfBEUpkyByWaIsmwK8iA3Ay959JduCBnY7fxUMMot62CayrmH2xWTfmJ6rpmf/1DtfbD+flng6rxeK2amtrCrzTTduvmH2iG2ZMUvtMSnRn23vEPPxr/u7nnbG/8piDwWv/47985pU1zybZOZEk4WViWr3sFfAlvNZpBkwF68EXgLcCGsds7gc3BO51zFwNXAJ+JRuPMbCOwcY7rfh34ejSeV0REJNkpGCYSXcocTIx0OJZ5ybCDQBagpLe5BjkNeGL/KGXZkdennU5oLfl0LQWwpc0/533WgI4BKMyKZYsmGhwYCr9SFLebTbySFLrO9J69UVAw43rJWBJOFrALVsIVa3DbGua+zZUXp+1EgOAx0GxmPcDrnHNrxxbtMbPQsWB9BLKKa6LUPhEREQkRSa3AuWTWhGYyhWaMBJenWyaTpJd4Zw5K6gX2Q7PupvOx298+pwy7lbnxH2ouycdrkPOZNv+89xfVlD+rrc88jUKAQBmN/lGL22iE7JxFcd0uUUIzli/+5H9Ou3w6KgknUfXhm+COB+c2U6/Pwe03xr5NSSyiTw0zq5th+UHg4DzaIyIiIjOYa63AJYuMwszIs5CGvZcTFVmwEp05mI5SPbB/+MTZrLsVSwK1HeeaYfezdh2A010kQc6GbovpaICZymr0jxoHeqF0UWqNSKjrimw/7B6JfAJQry6IcFK/SLeLlniPYlFJuNQz3Ynd0CSZj3/2XdF/0qsuhfs+jG36Go5Zymj4HNx/R1qXzYAIA83OuSzgOmANUGBmD44tzwGKCEzSp29JIiIiUeKlVuCJISjJ8paFFLruHTXD0y4XSUWJyBxMZ5EE9heC0Oy60B/BoctDM+zuqzsbuHtw7dkMOwWaw0v1ETiRBjnruvxU5c0/yjldIDDUqSGjfxROD8PgWFPbBwLXC3VEwmT9EZYXnkuyY7RULi3lZS9bzoEDR+a8zaqXLadyaWkMWzWzVBvFIskheGK3a9AoLopw0r652nAdR779P5R2dJDb0zP1/uICePhjaR9khggCzc65txCohVwJ48H8B8fuvhz4HfB+4DtRaqOIiEja8xIMg0CtQBGZXTJmDqa6SAL7OXHKEIyXqjxHtm/ibYlM6cKqAjAnkQY5I90uaK6BwE/uGubo4MyPs9BHJEDkWcm+OO/Kb7juSp78+o8xC39Udc5x3XVXRuV5p8twDz3pM1mqj2KR+JruxO5Aj/HBW66P+XP/x6cDoc8lra2cW1fPor4+hvLyuO4dl6V1TebJPAWanXNXAs8Cx4GPAq8G/m/wfjP7g3PuAIE6zQo0i4iIREGktQKTORim2e0lGSQ6czDdRBrYX11I3OqeSvJL9c+HSIOc8ynZMNdA4NuWGS394R9voZcaCrTZ++dDYZzfmhdeUMXr37aO3/z0+VmDzc45Xn/9OnyVK2jusahm/wcnjhz2Q9Y0Xd09bCpPJTE1MBoYXdExYHEr4XNi5UpOrDwbWL7ugtT+XPLK66txH4EJ/640sw7n3APTrFMNvGreLRMREREg9YNhmt1eEiVRmYPpaiHUPRVJtEiDnJEG57yUs/mJh9IuC7nUUFWeY3Wh83RirCAjMSfErv+zi7nv1SU8+P0/8nz94Sn3r3rZcq677krKzlkRteecbuLIk0PG+QVT//+OAVK6PFX/qMU9yCkBdV1+9nYbPWO5K6lWwmch8xpofh3wrJl1zLJOKxD7nHUREZE0kYrBMM1uL8kgEZmD6SzSY9LxwUBWHKROHV6RmUQS5FxT6CIObnktzeXFQi41tKHKx0MNc5sbAyAnI3HBxusuW8l1l63kvT87yi9/+Ev8fj95mT6e+/u3cck5S+LWjsn6R40ej8f9ZH/PBD+DuocDE2EG/z0FOeMrOAqjZ4YBkirHklheX/ECAmUzZpMXweOKiIjIDBQMk+DkTB0DgevJkzNJZCL9Eeh1O/VfQKTHpPyMQC3eVKzHKzKdDVU+5hpmc8ANVZEdyyIpZ+NVpCMZEm1tsY9bV2XMuR+ODwWCjYf7A8HGzfUjcf/fS8tLKVhcRFHpYkpLixIaZIbAaJRIJPt75sSg0RQSZJ4sGOT8dWdy/x8LlddJhZP9/ZSKvKYCHAYuCbPO5cD+yJojIiIik8V7GK0kD83SHluxzhxU/00U6bHsk2sykza7Dc5ORhT0x31HJ9wXOnGRyFwEgpyEDaY44LZVGfMqmxFryTy6Kpz1FT7KswPlHBo8BuSVUQn+CM9hTPeeqW85wcmTZ/D7/fh8PupbTiQkkP7WZYFM93BUczp2IplUWH0QX14DzT8D7nDOXW1mL0y+0zn3VuC1wGej0TgRERGJvFZgMgdmJDzN0h4fXoZHe8kcVP9NFe+SAInQ1N5FV98QGGRl+ujuH0p0k2SBChfkXFPouGGeJ6viEQRO5tFVc5mYeG1x4DVu6zMe2TeC32DIDyeHwtceTvVgY3C0jt8Ck7H1j9qEOtW+CA/doe+Z53a1svn7L/Kb+iMT1ll717d5/SXLuf+9V3HdZSuJFwU5EyvSSYXnU44l9DjR3DP1uaerTZ7uvL7jHwJOA790zn0OuBjAOXf92O0fAO3AF6LaShERkTTnZRgtBGoFRjI8v60vMKlJ8JLoIf7pWnJAwwLjZ67Do71kDqr/ZhavkgDx9OO/fyf3vOOVnOkfovHwKQaHRxkcGaVnYJjn64+w7u+e5rldrYlupixAa4t93HdJJp9/RSYrcmFZDqzIhc+/IpP7LsmcdwArHkHghRJkOzVk45fpVOU5KrIdlTmOwVHvE9ylkrouP5vrR7i3doTD/YGSIYN+aOyGfd02/plWGGEp/eB75uv/Xc+bNj47Jcgc9Jv6I7xp47P86//UR/ZEHs0nyCnRMZ8J0qMhWMarIPPsb6V0+n0yV552fTM77Jx7E/B94OMhd/2IwHfBZmCDmYWr4ywiIiIezHUYbdDxocDw/IIMqMyBwqxAaGemCbSSbYh/srUn3pQxE1/RzhxU/80sXiUB4unr/13P7V/9FX6b/j8KBkOe+Otr+as3hKtCKDJVMMgZejsaIi1nM1fJPiIhkolFmzURAAAgAElEQVSJk3GCu2Db+0YheBjqHY3N5KnhRuv0jMJDDaOszIUl2Y6CDG+vV/A989yu1lmPq0F+M277l19xbnlRzDOb5xPkrMpL4tT+BSQRE6SH7j/B3ycHQwLLwbrs6fD7ZK48H3HMbLtz7uXA9cBrgCVAF/AH4D/NLMKS7yIiIjKbSGoF9oxCUy+ck2ssyZ7+B06yDfFPtvbEWyKGBQrTDo/2Obj7Qm/1gdV/4cWjJEC8JGMwRGSuIilnM1cLZUSCV/OZ4G6+wcbQYPjurol9ZjZxcrrMGHyceBmt09IPi3xGZQ40984tAzz4nnm4cYRHv/Fi2ONqkN+MB7//x5gfWxMR5EwmM5WQiMUJjZmEjsLoOnaSntNnMPPjnI+uYycpLi8Nu12k0v33iRcRvRPMbJRAFvOPwq0rIiIi0RMaDPtc4wgn5lj+s7Uf3n/e1C89Xof4x7rOYLK1JxGUMZNY880cVP/NTbQC+4m2+fvJFwwR8cJLnXoIzAERLkN1IY1IgIm1hn2OWU/8RXOCu/kYG6jGqMGITc1LHxgNPOeGKAb7vYzWAegYgAsLHbeu8nkaxfLUzuMcODB9uYyZPF9/OOYTBEYarEzmOuWRKl2UmOddW+yj42ALdb+r4Vhr+4T7fvr171G+chlrX3clledVTdluPvT7xJvYn3IQERGRqKvKC9QInKuZhucneoj/5B9332mZ+4/dVC05kO4ZMwud+s+bWJUEiIf6lhMz1g6dSTAYAkWxaZSIR15Lc4ULMi+kEQmRlOmKxgR3kZqcMRrPDMtIRuv0jAZKjXgdxXJut7fjatBzta0xDTRHWmpmIewLcxGPjOVwfvG7Pfz6e7/CZjjBe6y1na3f+y9e/ZZ1nH/ZGiA6JXwS/XtpoYnoneL+P3v3HR9Hfed//DWrVa9Wb+422Mi0QAyECzZ26N0BEgJpRyAEknBJjuSSYCJjLuSOSy9cKAe/JITgEIdAQjfGhObeJMtFbpLcZEu2et/5/bFaeVVW2j67q/fz8YDV7uxov9bszHznM5/v52sY5wJzgQnASIdM0zTNpYE0TERERDyraw+8RqCVQ/w9Xdz5KhZLDihjJrpp+40fK7b4N8Hfii21UKhazRI5/CnNNdSEePju7OgZkeBvkDbQCe6CJdwZlv6O1nGVGvFlFEtzu5fD9Ybwdz1v+VNqJtLrlEcTV6kqT0HmAabJmldXkZqZTtGU0oBL+Kgkmu98OkwahpEBLAcuhlEnjDYBBZpFRERCJBjD860a4j/WxV242+ONax4avVrYS/dfG7TPGu8ZM9FO22/8iNRgiIg/RgoEdjugsdu7+rrHe+BEj0npqGGCyBBIkDY5zv8J7oIp3BmW/o66GVpqxJtRLBkp/tVl8Hc9X/hSaiZW65RbxZdSVZgmle+t54EFk4NSNsPf9cZTSTR3vt6PewRYAPwTeAqoBTT5n4iISJgFY3i+FUP8vb2480W4Sw4caGgb+LkkJzXov18ZM9FN22/8iORgiIi/3AOBu1pMn87X0TJcPNAgrT8T3AWTFRmW/o668afUyMIz/Ktj7+96vvC21Ey01SmPdP6UqqqvPUhe53EgsHIq3lxnrPrzy8Ne25Vs8Osk58/BTEiJBr4Gmq8DNgAXm6bpX1hfREREAhaM4flWDPH3dSIZb4Sj5IB7B9E9uzlUHUdlzEQ3bb/xIZBgyLb6IDdGJMg6+gIv0RWJAgnSuqTH+z7BXTBZkWHp72gdf0qNlE3K4aKyYp8Ci/PKSgbVZ35k+8l8yN2t5rDXA6k37GvNaQlcIKWqAq3b7et1RkerMyHleDcUJqUF9NnRyte9KxP4vYLMIiIi1grG8PxwD/H35+LOG7HYkVfGTHTT9hsfDh9v93mdaVOLebk9k/a+k9+Mu9b1ADA9zQh6QETEXy1+jluO9OHigQRp3VkZbLRiRJo/o3XS4pylRvzxwM3ncWn5C16VSrAZBotvnutxeXYIBpH4UnNaAmdlqSpvrpfm3XTlwM+u7OapGca4y2R28bXXsgsoCEVDRERExHulKQZpcWPP/u5u6PD8cA/x9/fibjSxVnJgaC3oll441GHS2n/B796RVcZMZFPGU+x7cNlqn9eZ/bFz2N1qDqpb2u52HA9FQETEH0Nr63or3OWsfOVv+9444qDNLfjuuhG0uMwe9mCjVZPO+jJaB5wlRvy18MyJPHb3Au4YY/I3m2Hw+D0LWHjm4BEm4bpB503NaQmclaWq/LrJYoekyL3fFnK+7n2/Bn5kGEaJaZoHQtEgERER8U5hElS3jf0+l5GG54dziH+wLz5jveSAqxZ0KjApK5U9bSZFScqYiSbKeIpd/tSLBDg1P5XCNIPj3SePuhMSnN8FZS6LlYaWGujy895wOMpZBSKQWsOebgSFO9jY5md/KtAbm76M1pmY7CwxEojbLyljdVcqK1asY8/e4cfbeWUlLL557rAgs8Qeq+t2+3qTpSh5fPfxfO3NvIJzMsD3DMNYAqwHToz0RtM0awJsm4iIiIwiPd5gUrJJTYd3719e57xqHFo+I1xD/IN58WlVyYHKmgb2Hmmmt8+BPc5GZU1DwLXf3I1WC9o9CKAgZXRRxlPs8bde5OSWg3x9Xn6QWyMSXNkJ0O2Ag52+rxvK87L7eXAk3tys8bdsWKTcIKxocgz053wRrBFgY43WSYuDr58SxyuHgjOKbeaMUmbOKOXwkUZee/71gf7Xc/ddEdT+l0S2YNTtDoS310tNRxtJ6GjhcIeDYyG4TogWvgaa9+GcXNUAnhjlfaYfv1tERER8lJNokGAzibcZI3b43W1vMXm4qo87psH8/JMXguEa4u/vxV2o2uOLFZtreXDZ6mEd3Dlff4aLyop54ObzlFEzxNAyIEON17p1EhusrBcpEgojBWkfrOwNW3ktX400KsAb4S4bFmz+TKoc7BFgI43WaeyGqanOmsxzMm1BCzS7FBZkM7UgY+D5eAzejXfBrNvtj9Gulw7vq2PPh+vZvy9M1wnVtbB6K7R1QGoynHc6zIic6xBfg8G/g6BPFi8iIiIBSI83uKLIxnYvhnSZwON7+shNHJx19MohB0lxMCsddracfP8p6c4aY8EI6vpzcTctBbpMLC058OQbldz5m7c8dmzfqTzIpeUv8Pg9C/jXT5SFtW3RwlUGBKAkJ9XClogEh5X1IkXCxdfh4qEuZ+UeDHfPbva17Ew4y4YFk7+TKt8Yopvz7qN12npNvyf+E/GGq273aH1y8Fy3OxhGuslSuaGKVS+tCs91wm3fg5pD0NQ6fFlmGjzyTWfQ2WI+HZFN0/xCiNohIiIiAfAlw8XEeTf+u7/8+8Bru1tPru2aeO6Sm68M+kWDrxd3n548ePhluIPMKzbXjtmhBXCYJnf8+i0m52Uos7nfaGVARKKd1fUiRcLB2+HiAJOSQ1s2I5jCWTYsmPydVDnS62Z7UtduUt9lDiQbdPaN7wnWxFm3e0p+BkuXrWFV5fBp46ZNLeaxL4R+lKHrJsuu6jpef2nVqBNWQpCuE5avgK27PC9vaoW7lsIDd8ENC/z7jCBReQsREZEo19Fnsr1l7Pe5q2oxB3XYk7raTy5MTAFgVobh3+RUX/uR87GtA060QG8f2OMgK505qcl8afG3R7y4W/Xnlwc9n5xq8N03TwbBv/iFq3xvS4AeXLbaqyF64OxELl22Jiid25FqQUNmwL/XCqGuay1iBavrRYqEizc1eQuTAp/4LdzCVTYsmFyTKpccrGPO9gqSOzroSE6mYtYcDhSXjrletHDVoR6avb2t2STNronWxruFZ05k4ZkTqaxp4FOPvEJvn4MTvQa33noZhQXZLAzjxLpvrlg3ZpDZxWGa3Pn0au76ctHAa15fZ63eCkt/682HwIP/C8V5lmY2+7QFDMPoA8pN01w6ynu+DywxTVNBbBERkTBoGX1uHI++dedVXF7kjDS7Z5xedOOVgTXoePOow7ou3ltJ3uwyzxetdrB1ttHZYnAABgLf4VZZ0+BTEAlgVeWBgAKpo9WCnjq1mE8sPJeZMzxfTEYS1bUWCM7kXZHK6nqRIuEy0nBxVzmrZ2uiLIo5hKe9NxLrhc7/1v18ovYw6W3D+1ctqWn84o6vs2qE9SIho9n9XOA+is71uutcsLLeMWqmeWsv7GoxebveMWi+ERl/yiblDNTt3t1qUliQHdbPP3ykkb17fbtO2LP3IDvrGjil1MfrhN/+2RlE9obDhMeej55AM87RI97cPtItJhERkTBx9TuGZgQPNe+mwQFkV4bL0IzTU440+t9ZW74CKqs9d4b6h3XNeeAu5tywYNBF62c+d+VADeagBr79tGJLrd/r+RNoHqsW9N69B3niyZe4cdF8mGV9/bXRqK61jMTfybsiVSTUixQJJ/eavK7nVhhaUqGu3fS5LWMFND1NoGyZ5SvI2l6NgTMI7v6vNYH0tla++/OHaZh1Hs+VzBi0aiRlZQNkeyhVX9E0+jZxN9J8IyLhVF1d59d69fsP8PgnCnz4oFpYX+Xbh6zb5lzPogkCQ5FCMAHoDMHvFRERkRHYRri26mg9OflactrIk69t31XLvEfXDcs4rfzpnwayZ5k1xfuGuIZ1jXXH3W1YV+l5pw+7aA1q4DsAze3dYVvP21rQpmny/PK3WXFGVsQGrVTXWtwFa/KuSDVWvch5ZSUsvnnuuPiOu98gHInqs0ugPJVU+PaWXmalGyzystSFtwFNTxMoh11//8rVWxra7XM9t5kmj1R9yIHkVN7Ndg7Pn51uhO2GgCtT+ZHtvaNmLXviz3wjCjSLVbo6/btO+OCDrVxzaOQg9YjnydVb/focVm+N3ECzYRgXDXlpygivAcQBk4BbgR1BaJuIiIh4Ib3/bO6eseye3Tw0kxlg9+YqnnvV8wzJruzZ85N8yDj1Z1hXThYfeWUzCe3tHGjr5vOtXfyupm3QW/0OfAcoI8VDyk0I1vOlFrQZxFrQoWBVXWsRqwytF7mvvhnDMDhneh5v/+cnrW6eJQ40nDyOl+SMfLNTxBfBzECOuoCmD/2rOODePVt5N7sIA/jc3vXw5D+htR3SUuDqebDgoyFtLnjOWvakrt0cdgNhLFUtpl/Z7CLBkJjk33WCzeY8lnh9nmzr8Otz/F4vCLxJJXibkyWKTODz/f+NxAAcwLcCbpmIiIh4JTnOYFY6Ax30pqONtJ5oxjQdGIaNpqONZOadzAg+vK+Ota+OPUOy6UvGqb/Duj75TRa6vfRp4Cc2Ozfnz+Kt5AkDr/sV+A7QwjP8C376up4VtaBDJZb+LSK+ctWLPNrsvLhLT/bvIjRauWdiuWc3hzKTWVnU40MwM5CjLqDpY//KBD52op6dK/5Iypt/GF7TdMVq5+M3boMvXBdQ09wzl9v7IMWtFvSEBN8mlK5ocvjVhoomB6UpEVCEWsadGX7OmXLFTZey9MJ8vvKTlwZeG/VclZrs1+f4vV4QeLPnP8jJMkAP4Aw8j1Rjvg9oAFaaprk9WA0UERGRsS0qtXHvK/vZ+t46jtYeGrTs5SefI29iEXMuPJfCKaVUvufbDMleZZz6O6xrCBPIcfTy+uEK7sidyVPphSeXhbnUQtmkHC4qK/YpcDqvrMTngGm4a0GHUiz9WyS4glFTNVbF8oSJ4aQs6tgVzAzkqAto+ti/ch1VU0zH6BNn/fQPsLsOlt7jb8sGiTd8z2J21+HFnJKntJ7gXxoPk97bQ8FbNipmzaGjdJL/HyoSgMKCbKZOLfZpQsD8icV0pWfz7S297GwxKUo2BkameuTvpH6RPBmgaZrlrp8Nw/g88IJpmr8IZaNERETEN6vXVPH2c57r4h6tPcTK5/7O6ReeS/2QQPRYvMo4DdLwLNdFURzw+LFd7LcnDspsDnephQduPo9Ly1/wqhSEzTBYfPNcnz8jnLWgQy2W/i0SHN7UVJWTYm3CxJCproXVW3kpv8eZtXXe6Vzzp40Di5XJHDuCnYHsTUBzJP6uFzA/+1deHT1efBuml/qd2XzfLHvQau8njxLDX/KjByisP0x6W+vJF/dsBqA7PQ1+/E1Lg2oyfn1i4bk88eRL3iXwGAZlF54z8LS1F3a1mExOHWNvnTERzpnt28jRc0+zrD4z+DgZoGmaU0PVEBEREfHPruo6nnxy7MnXME0q3l3r12eMmXEaguFZccDiE7WDAs0Q3lILC8+cyGN3LxhzcjubYfD4PQv8CoCHsxZ0qMXSv0UC521N1YnJkJM4foOqsTZh4tAJXYN6vL7te1BzCJpahy36vT2efzv1HNblFERdOR5ltXsW7Azk0QKao/F3vYCFevj748sDLqERDM4M9OHbev67K5mxtxqDk8PsXUwgoaUV7loKD9wFNywIT2NF+s2cUconF81j+XLP894AYBjMvXwehVOc5TbcyxxWnrDxQuVRri/L87z+l29yfs+9qdVuM+DOG338lwTX+D1jiYiIxIg3V6zzfiI5Pz9jzIzTEGSSmMD8ziZO625jW8LgYdDhLLVw+yVlTMnPYOmyNayqPDBs+byyEhbfPNfvLOtw1YIOh1j4t1zz0IsDNScB3kkbHABVpqR3fKmpWtMBCTaT9PjxG2yOBSs21/LgstXDyg3N+fozXFRWzAM3nxfYaJTlK2DrrhEXmUBWbw9PVn7IHbkzmfP1xuB8pgWU1T5YsDOQPQU0x2LZZIChztRtbYe31vo0QaD7jRH386XrdX9ujJSmGMxKNwZlr5dVVfClPzwxEFweujcMPHeY8OD/QnGeMpsl7OZ+9DTuOmOCx+uE/InFlF14DoVTSjm8r46KEcoc3vD9Z0c/Z513Oiz+Miz97ejBZpvhvOli8X4w6hHAMIy36J8A0DTNuv7n3jBN01w49ttEREQkEIePNPpUG8xfY2ac+jOsawyuC4iFHSeGBZrDXWph4ZkTWXjmRCprGvjUI68MZOo9d98VAQe8w1ULOhxi6d8C0NTcxoEu5zdRNV9940tNVYDDnZAeH7LmRIVormP95BuVo478eKfyIJeWv8Dj9/g5oevqrc4LbA9GLLtUSWCfGUauwFxdu8nPd/UOfAdumRQXNd+BUAl2BvJIAc2xzE43rNsOIehfDfP3VaMGml0B5I4+k5ZeONblfN1uOGNeBUnBacaiUhsPV528QbnoH8uxeZlIgcOEx563PMAm45P7dcKi/3qFQx19GIaNf7n+0oEJ2XdvrmLNq6vA3/PkooXOG64eRvWQmQaPREYZmbFuNc3vf0wZ8nws/iZMiYiIyCiGZpFsrfRv8jVfLf+gmjc21Xhc/tL91/o2rMsHGY7haUlWlVoom5TD1IKMQc+DwZda0IaftaDDJRx1rUPppfuvHdjPnnr6H5T0ZzQrk9l7/tRUbe2DnS1mQBlx0cqbOtaWZVN6YcXm2jHLC4Gzxr7fE7r+9s9en1vcyy4F9JlhpFrmnoUiA3loQHM0BnCD1X//EPWvBrS2j7q4pcfkcKfzOO2uC0iLg89MjgvKMWpOpo0vTYMn9vRRfLCO2bu2DyuXMap125z12y2sTSvh88j23oGM+ra+k9n1VvYjyiblkJWdQWu7sy2uIPPhfXWjBpldxjxnTchw/tfWAbWHweEAmw0mFg7MVxAJRv3Lm6ZpG+25iIiIWCc7Aey9PSH/nHllJaQnn0w1PNDQNvDzoCxP17CuJf8b1M9vtg1PS4qkUgvB4G0taMMwuHHR/IgOmISjrrVENn9rqiaOwysNb+tY3zEN5udH5h/owWWrvS7f5NeErtW1PmVzDi27FO5JZH3lzXfgh1V9JNkgof8rMJ5uyIQiA9k9oDnabzWAO6YFJ4gakDGGzbuCsQ7Ar5ampXhctLLewe42z5mErX0E9Rh1cb6NvESoe68S8CHI7LJ6a1gCzaEoHyL+SzCc10WRYKT7QRXvrRszyHxy/VHOWb/8jwBbFx5jlc74OvChaZprwtQeERERGcXQjusvapJ4x9vCVm4Mw7v+jivj1L2zc81DLw78PCzLc6xhXT5wXTitSM4a9Hokl1oIxFi1oKdNLWbhwnOZOaPUgtb5JtR1rUNh6EVj49FGDjc0c+yYA5styBOaxTh/a6ouLLBxQ6lVM26Fny91rB/f00duooV1Yj2orGnwqVQO+DGh6+qtPv3+kcouhXMSWV94+x0A6HRAZryFk9K5cc8cDEeQLRQZyK6A5l/rHFSNEMSenW5wwxijCcIabBylf2UAHWlpvL7gMq578S++ZQEDXD1vxJetOkbNybQxJ73Lv5XbOgL+fF9FSoBzvLlvln3QPBo/PCMy6m/Zhux8TUcbh9VkHkuknrO8NdaR72dAObAGwDCMPqDcNM2lIW4XhmHcCMwDzgLOBNKBZ0zTvG2UdT4G3A+cDyQB1cD/Ab80TdPPLq+IiEjk8jezt/zT57HkT2vGzJ4dmnFaWdPA3iPNAzWKR+wE/eGHzsfqWmeAoK3j5HCuHz7udWaaAbydlDmoPnMklloIptFqQb/cnml183wSyrrWobSruo5X31hL7f7BFwVBm9BsHAh2TdVY5UsdaxNnQCzSAs0rtvhXvsmnCV39DBwNLbsUzklkveVrLfPsBIPFZZGVKRmOIFuoMpDnZDoDyUNrY9870+5zTeawBBs9DJuvySmgOzmZV677JJe9+TJJ7T7sM2kpHuszW3qMSk0O73o+UsayeJIx5KtxeH+dX78nEs9Z3hpr7+gEEt2eG/gxesFP9+MMMLcCdcCs0d5sGMZ1wF9wtvk5oBG4BvgpcCFwUygbKyIiYgV/J1974FPnceGsYo8Zp6UleVxxxQX86yemAM4anA8uWz3sc0YNvs2YOHz4og91BvuApVkn1x8p8B2rRqoF/bJb5lQ0CVVd62C7b5adJ9+o5MknQzih2TgRipqqkc410mPNriOcaOsa+Nn1+tDRH/7Usa5qMSNugkB/J2b1aT0/A0dDyy6FexLZsUTzd2B6fxZhOINtwchA9qQ0xSA/0Rj03BthDzZ6GDb/nKt/0GrywQ03MP+ZP3oftLlj0YgvW/797K8163N2doTUqJXxKykO0uzQ2r9b9nT5V+Yw0s5ZvhjrKLwXuMwwjAK318I10d83gFOADOAro73RMIwM4HGc16TzTdO83TTN+3BmQ38A3GgYxqdD3F4RERFLPHDzedgM77rh7hnBC8+cyOKb53Lu9Pxh76s7cJQ3V6xjxeZannyjkkvLX/AYzHYF3/7vzcqxG+CqM9g/rmxop8L1vA+4I3cmbyVPAJxlI750+zUK7knI+Dqh2YrN4ZmIMxq5aqr6YqyaqtEkKT6OpPg4SrLTPL7H3zrW/q4XKv5OzOrTej4Gjlx78NCyS1ZNIutJrHwHwmlOpo3FZXb++ww7JclQlAQlyfDfZ9hZXGaP6ptVwbLu2muomHeRd2++dj584boRF1n+/ZwxEc6Z7VuQ+dzTNBGgRISi5JPf3PhE/0p6RNo5yxdj3YL7Lc7yGQeNkxew5YZhlI+xnmmaZkC390zTXOn62Rj74vlGIA/4nWma69x+R6dhGPcDK3AGq/8USJtEREQikb+Trz35RuWo6+zde5BLfvBX8KKe85izJLtbtBBK8uGx5zHWbRu0yADa5szk60Ye72fkUBZn47IbL6WwIHv03ynA4HqRI9FQT89CPqHZOBOKmqqRzJWxPGoNezf+1rF+44iDy4sip8aIv+WbfFqvP+AUSNklnz8zDPz9Dvi7XizxNwM5FrnXyW7vc9aK/vXn7uKhXdVMOHyYOMcIgd84G3z9Mx6DzBAh388v34Tjy0uxeTepCNx5YxA/XCKR+zkWnCOH3JeNdt4Np3Q7TEo1MIDCyf7NrRJp5yxfjNqjM03zF8BngD8Cb/e/XAO8M8Z//wxNcz1a0P/46gjL3gHagY8ZhpE4wnIREZGod/slZbxefj3zykpGXD6vrITXy68fyAj2NnvTxOtJkgeCb175wz8gJRnOOY2uxETa4uy0xtnZmptPanYG9bOmM7M4i6kFGQoy++l4tznwn4wukAnNZGTOmqpxY2aj+VpTNVb4W4+6oTuyMlpd5Zt84deErl++afgMSx4MLbvk92eGmGqZS7DFG85a0dkJcKKwkL1nnQWnTYN4O9jjnI+nTYMLzx41yAwR8v0873QeKDsPV+za0yg4bAY8cJfKZowzBxra6Ozuo7M7Mu++5SbAd2fHcf60HPImFvm0biSes3wxZlqLaZp/oj8T2DAMB/CUaZoPhrphPjq1/3Hn0AWmafYahrEXKAOmAd7dChcREYkyQydfq+/ow2azceutl/HjeYPLY/iSvekLn2dJTk3mRFERDY1NAHTER8aM0dHKPWPZPbvZikxm96yToRknMHq2Z7iFZUKzcSiUNVWjnb91rLsd8HBVH3dMg/n5kfF3e+Dm87i0/AWvzil+T+jqKru09LfgMIfVbXU9H1p2KaDPDLHxWMtcgu++WfaRz/f9dbRJm+CcNNAldwLeiJTv57LSGdxQs4up7S1k9w6uWWsAvRlp2P/nmwoyjxPufcdrHnqRA42tAMydWRBR/UoX14SjMz97Hp95+G+YoTxPRpC48vJyr9+8ZMkSgLfLy8v3h6pBHj53CvAFYGt5efnyEZbfB+QAj5SXlzeOsPxWYCLwTHl5+YhTPhqGceeSJUt+e9VVVxUDnHLKKVRVVZGbm8vGjRspLi5m3bp15OTkUFlZSWJiIgcOHKC9vZ22tjYOHjxIQkIC27ZtIy8vjw0bNgxap6qqivj4eA4ePEhbWxsdHR3U1dWRmJhIRUUFBQUFg9aZMGECO3bsIC4ujsOHD9PS0kJnZye1tbUkJyezZcsWCgsLWb9+/cA6WVlZ7Ny5E8MwqK+vp7m5me7ubvbv309KSgpbtmyhuLiYtWvXUlJSwrp168jMzKS6uhrTNDl27BjHjx+nt7eXffv2kZaWxubNmwd+v+vzMjIy2L17Nw6Hg8bGRkLB19wAACAASURBVBoaGnA4HOzZs4f09PRB6xQUFLBx40bS09PZu3cvvb29HD9+nGPHjgFQXV1NVlYWmzZtGlgnPz+fzZs3k5qayr59++jp6aGpqYn6+noMw2Dnzp1MmDBh0Dp5eXls2bKFlJQUampq6Orqorm5mSNHjhAXF8f27dvJyclhw4YNA//23NxcKioqSExMpK6ujs7OTlpbWzl06BB2u31g+7vWWbt2Lbm5uWzbtm1gWw7d/pWVleTn5w/altnZ2VRVVWG32we25dDtP3Rbura/zWbjyJEjNDc309XVRU1NzcD2Lyoq8rj9jx49yokTJ+jp6Rm2/detWzfwmJGRMWz79/X1sWfPHtLS0ti0adPA36uwsJANGzaMuv0zMjIGtsvatWspLCxk06ZNpKamjrj9d+3a5XH7p6SksH//frq7u4dt/+zs7EH7ZV5eHlu3biUpKYmamho6OztpaWnh8OHDg7a/+zru29+1L7e2tnLw4EHi4+OpqqoacV923/7u+7L79l+/fv3A3y07O5vt27d73Je3bt06aPuvXbt2xO3v2pdd23/otnTty8Cg7T/Svuz67ri2v8PhoKGhYdD2H7ovu7a/+77svv13795NZmbmoG1ZUFAwsP3d9+WjR48O2v4bN24c2Mfy8/M97ss2m23Y9net49r+tbW1Hvfl8XQstzs6eWf7UfpMk+TUFBaUOjh/5sl9edmr/+THr+wK6Nw8muS+VhacPX30Y/n0fCZ85ho2FaXxt6tu4/t1TSybNod3C3JZVH4rv1+5DYfDpLe3l1MLEjBtNuIP7eScSbmWH8v/+M4Oent7cTgcfLQ4LmTH8mff3UVtSw+NXQ667Q729qRgr9lKbVIR1ZvX88++AhL3bfB4LK/YtQczKY34mi2cP7M07Mfyx15eT1xcHB0dHRiGQUpCHCnxNtKS7LS1tXHrvFkRcyxfsfUA7+2o9/m7fkqOnY+fVuzxWP6rd6p4vwE+3HOYI8ebaWrvovdoDRvbk1izccvAfhnLx/LOI7Wcn9nDWQmtHD16hNR4GyUndvGNj+RRcmQLZ00rifpj+cLvP8fyNft59KU1bKltYl99E4eOt/OPD3fw7D938C9Tkocdy/dVbeF4ahEp+zfQlF7EpEMb6EjKIO/4HsAktb2BlK4T2By95DTtpys+ldL6rQPvXdmTT/L+jRRlZ1reL//IrMmY7cd5t/rEqKVSbAb8/IsXMCuzy79++cc/Ssfr75NggtE1POB0zGbnuoLT+GtqnttnGnz/qmncfuW5EdcvnzW5hL1b1nMouX/7J2aQe3wPhmmS2tFISucJbI4+ck/soyshjdL6rRQXF1N0aKOl/fJth08Q13wU07CRenSXZf3yfcYE4g9thzg7BR2HA+6X7+pKJL6uAkdmAYn7Nw46lkd6v7y6qQdb23FsbY2ckobzWP7pq9hUnEbxl25m3dRsCr76WTbNyCP1psvYd1rpmMfyHZvWcSI5j8S6Cnrik5jQXIe9r4ukrhYy2usxDRsFjbtoT5rAxCObaUov4iONG1g4syBox/IV2w/QMWU287Zt4FhhKe3JdkzDTq/dRmtyGoeK8jiQW8iaiz5CSm/k9ssVYwlNv/zJ1zZR39KNw+EgN9XOpWW5YT2WV3cnE1+7FceEIhL3nWzbih0naGtvB+CiaakcP36c2YVp7GvtZtu+YeHKYefJH336bD42NTUkx/KioqIlo/dsg8PwJqJuNcMw5gMrgWdM07xthOU7gZnATNM0q0dY/j5wAXCBaZofjvZZjz76qAnwla+MOv+giIhIRLvmoRcHavZ98QtXDcpo/cXfN3HvE++E7LOXfuZ87vfhTvy3VtXzzDOv4XA4yE+O47n7ruA/fvfewPKLbrxy4OdIqDHsbQ3WYHyO+zZ0L4ExIcGZqeTp71HXbvLzXb04TGen9d6Z9nFdw3Is/u4TP//SRXz96rM8LnfPMvNl+8Uqq7PsQ8X9mHCgoW3g55IcZ41gT8eJiiaH13WsRzI73WBxWeT8HVdsrmXpsjWsqjwwbNm8shIW3zw38LrmX/uR87GtA2oP09nVQ3NnL5XE02qzc23hyQljg/aZIeTLd8DAOQzb6ozmSNmPg9EO99/hOt8CTE+LrmN0qLaJld/PaNw3JLyueejFgRFzVmQ0e9rvPF0nPLK9l13VdaxYsY49e4eXawvTOSssFwPRceQcW1P/Y6aH5RlD3iciIjJuNbd3j/2mAHg7S/KKzbU8uGz1oNq4R4E5X3+G+DgbqUl2Euxx7H76H4Az2Dre+HqxW9HkYHmdg+1DyhR8e0svs9INFo3TMgVjCdWEZtESpJDA+Htx66xjDU/s8S/YXNViUtduRsxNpKHlm3r7HNjjbDx33xXBKzHzy/8Y9DSp/7+jNQ187ZFXODUUnxlC3n4Hxmst83DK9q7rMq5Y+f1cXufw+rho4izRpP1DrOJtubiZM0qZOaOUK1OaQneejACx0vvdAZwLnAKsd19gGIYdmAr0AnvC3zQREZHI4m0g2F/eBO1u+v0W/rJ8lcdaZT19Dk60dTMpNy3YzYtZK+sdo14Mbm8xI662a6RwTWjmy4SA0T5Ri0QGVx3rJ/f0caTL9/UrmhyUpkTW7HBlk3KYWpAx6HksfmawREMtc0+Zv67Xo/WmWrS2G8K3Taz4fta1m8NumI8l0m68Sehc89CLtHR0s373UTq6nd/397cfipg5QEqyx752ieZzljei98g62FvArcDlwLNDll0EpADvmKbpR/dNREQktvibvekNb4JvKzbXjhpkdlfX0Mbt1y9g5ozSYDUxJlU0jR5kdjGBx/f0kZuoCaWGCsuEZiIjmJNp46I8kz/X+T7xVkdfCBokYeeaMCoayh4p8zfyhHqbhPv7WdHk+7HQtV6k3XiT4FqxuZb3tx+isXVwaK+xtYv3tx/ilOIsS9oVzOC2+02kkUTDDbLIb6F3ngf+C/i0YRi/NE1zHYBhGEnAQ/3vedSqxomIiEQSf7I3veFt8O3BZau9CjIDOEyTv7++jmsLSwDrsqe8HRJnFQ0xDdzCMyfy2N0LuPM3b40abLYZBo/fsyCi675K9En2Mzbi73oSmUpTDPITjUHPI0E0BDbGm3BuE/fAV0vPydefrekLSVv8vYGmG2+x7ck3KkftozW2drFmVz3/92Yl//qJshHfE21Gmt8jGkTsGcMwjOuB6/ufFvY/XmAYxtP9Px8zTfPfAUzTbDYM4w6cAee3DcP4E9AIXAuc2v/6c+Fqu4iIiBWGBkPb+68Lnnr6H7yTZgwKhvqSvWn0/2+0t3obfKusafA5wH2o5iD1RxqZkJft03qh4s2QuHDSENPguf2SMqbkZ4R+QjORIZw3fnzP4tMNI5HgioVswlAKR0a7brzJUCs2146ZCADOBJU7fv0Wk/Myorav5n6MiZTJV30VyS09C/j8kNem9f8HsB/4d9cC0zRfMAxjHvB94JM454WoBr4J/ML0NnVKREQkBpRkp3Gky/Opz9fszcl5wQm+/WF1jXf/gCFq99aRlmtdoDlcGcv+ZE5riGlwhWVCM5EhSlMMZqUbPt00mp1u6GaRSAhFazZhsIU7wKUbbzLUg8tWe5UcA85g89Jla6I20BwLIjbQbJpmOVDu4zrvAVeGoj0iIiKRbmgQcqy74L5mbwYj+LbGn9muAEd3D/Hj7BrP28xpDTENjVifqEUiz6JSGz+s8m7HNIAbShVUiRWxOtleNHL9rYfWI75lUpxu7ISJbryJO39GQ66qPEBlTUNU993q2k3qu8yBY1A0jUTUGUtERCSG+Hqx6mv2ZiDBt7p2kxPEe/1+dxkp8WQnxv6Fnj+Z0xpiKhIb5mTamJTcR03H6O8zgDumxSl7j8ivn+8PTbbn2V3rThYIbu8b/vr/nutfH8NdRZOD5XWOYUHOb2/pZVa6waJSm/a9MFhUauPhqrEnOQbdeIt1K7bU+r1esAPN4bgpGAvHIAWaRUREYpQvF6vhyN6saHJQOLnUr3VTSko50BFdnaxw0RBTkdiRk2iQYDM53AmtIyQ3z043uEHHvxFFWv18Xyhj2XcG0Nsfh3EFnYfWV/b177qy3sETezwHN7e3mDxc1ccd02B+vvbBUJqTaeNL0+DxPaOP8tCNt9jX3N4d1vW85c111miB6d2tJtPTBifPrKx3jPqdj5ZjkM5oIiIiMSSSL1Y7+iAzL5u8iUUcrT3k9Xr5E4vJdJsIMFo6WeGiIaYisSU93iA93jmCw33o/r0z7dpvh4jGjGXxnytj2VXW4lh/Na6pqdDZd7Kmsj/1lCuaRg8yu5g4g5+5ibphG2oX59tYeaRPN97GuYwU/4Z5+LveaAK5zhorMO06Bo0lGo5BkXs1KiIiIjHFVaphzoXnsvK5v4M3k3oYBmUXnjPs5WjoZEH4Zq/XEFOJVKo96zvX3+nZmj5aTlYK4Nka5wWo/mYSauE6d/nK05Dy7S2QFgeFSc4bNf60b3mdw6tzKDj7IH+tc0R0/yNWuG68dfSZtPSiG2/j0MIz/JvUz9/1gmm0Y9E7Q7KZY+kYFJmtEhERkZjj6gwVTill7uXzwBjjAsEwmHv5PAqnjFxuw9XJihbHu82B/4LNOcQ0jrEuuTTEVKyUnXDyP/GO/mZitVCeu3yxst7Bw1V9HkfvtPZBdRvsazOpa/etrXXtpk+jggCqWnz/HPFfcpxBfqJBYZLzUUHm8aNsUg4XlRX7tM68spKomggw1o5Buh0uIiIiYeFe4mH6mbNJzUyn8r311NcOn0k6f2IxZRee4zHI7OLqZEXqBUc4Z6+/ON9GXqIz+F41QmdVQ0zFCsq+9Z2rZmO0/e1icWK+8Sqc5y5veFvWAuB4j+/zOVQ0+XfTuqLJQWmKZtYVCbUHbj6PS8tfwOHFaEibYbD45rlhaJXvPJ0nP/nwS9S1m8y76Uqffl+kHoOiq/ciIiIiUc29xEPhlFIKp5TSdLSRd194HdN0YBg2/uX6SwfVZB5LpHayIPwzR8/JdP6+ocEBDTEVkXCK5on5JPznrrH4MqTcxZf5HDrGLosa1PXEN+5ll1zcy7tE20058d3CMyfy2N0LuPM3b40abLYZBo/fs4CFZ1pfNmMs7udJh5+JyZF6DNIeKSIiImHjmkXcPTMpMy+btKyMgff4EmSGyO1kWTl7fWmKc2ip+3MRiUyxUsdaGcuxwcpz10j8GVLu4u18Dsl+3qv2dz3xX09/8vnxbtOvSR8let1+SRlT8jNYumwNqyoPDFs+r6yExTfPjeggs6fz5KuH+vjdft9HVkTqMSg6ei0iIiISM8Yq8eCrSOxkafb66KXh/2Il1WIWK0XiucvfshYu3kya5Vzm++fovB0anm6+AfSYkBIHExL8m/RRotvCMyey8MyJVNY08KlHXqG3z4E9zsZz910RVTWZh4q1Y5D2TBEREQm7oSUekuKcF4NdfmQnR2InK5Zmjh7PNPxfwkHBEokUkXjuCsaopbHmc3CfQ8Jbs9M1IV04uN98a+y2rh0SWcom5TC1IGPQc4kc6tWIiIjImIKd5emerdLSA73913Z2H6/ZIvFCL5CZoyPt3zIeKWNZRMajSD13BWvU0ljzObjPITEWA7ihVDeHQ8X95pt7f7Gx27l12vucmc6PbO/VjTqJCbE2Ian2ShERkXHMnwBysLM8sxPAVWbP8OG6LVIv9GKtsygiIrEvUs9d/g4pH2qszOiR5pAYiQHcMS1Oo5AskGZ3bkcT6HJAR1/g5ddEIkGsTUiqQLOIiIgAoweQg53lOTQDZWvGyWyoO6bFRfWFnpWdxViZVExERMIrUgMd/pS1GIk3mdFjzSExO93ghlJbRPY9YtV9s+xUNDlYXufgYOfJ13tN2N4CD1b2skjbRKJcrE1IqqsNERGRcSwSywRE+4VepHQWNamYiIh4K1LOXSPxpayFJ972GYbOIeEwwWbAvTPtKm9lgZX1o09Qub3F5OGqPu6YBvPzI7NfKDIWTQYoIiIiEmLRfKFnZWdRGcsiIuKPSA50eFvWwhN/5nMoTTHITzQGPZfwqmgaPcjsYgKP7+kjNzFyA28io4m1CUm1F4qIiEjEcl3oFSY5HyO1Q+XO1Vn0RSR3FkVEJPZF+rnr4nwb350dx2wf2xip8znI2JbXOby+sWDiHAknEq0Wldrw9ugW6ce1yG2ZiIiISJSKpc6iiIiMD5F+7pqTaWNxmZ3/PsNOVvzY74/k+RxkdHXtps91uataTOraNUGgRCfnyI24MY/B0XBci9yWiYiIiESpWOosiojI+BAt567SFIOpqQYzUiHNQ43o2ekG350dp7q9Uaqiyb/sZH/XE4kEY43ciJbjmgr5iYiIiIRAtE9qKCIi4080nbvS4w3S4+GWSXFRN5+DjK6jL7zriUQK93lqFlf0Dry+dE70HNcUaBYREREJkWie1FBERManaDt3aeK+2JPsIVM9VOuJRJrSFINE2+Dn0UKBZhEREYk4j2w/eQd/d6s57PX7ZkVXF0YXwSIiEm107hKrODPmfS+DEQmZ9iLjXXRdpYmIiMi4k51gdQtERERkPIi1G93RqjTFYFa64dOEgLPTDd0MEYkAOkqKiIiIJa556MWBn9fsOjLs9ZfuvzbsbRIREREB3ei22qJSGw9X9eFNqNkAbihVNrNEP/ebXe19w1+Phptdkd9CERERiXkl2WlWN0FEREQi2GjZxrtbTaanBZ7NGg1BnPFiTqaNL02DJ/aMHmw2gDumxalshsSc+ChN0NdRVERERCyhjGURERHxh7KNx4eL823kJcJf6xxUjVBGY3a6wQ2lNgWZx4nxMBrS/WaX+821aLoJFj0tFREREREREZFxabRAi3tARmLLnExnILmu3eTnu3pxmGAz4N6ZdtVkHsc0GjJyKdAsIiIiEiKaVEhERKJNtJy7oqWdEhylKQb5icag5zK+xELG8lhi4bgW+S0UERERiQEa5isiItEmWs5d0dJOERFvRetxTYFmERERkRCJhqwDERERd9Fy7oqWdoqIeCsWjmuqmC4iIiIiIiIiIiIiAVGgWUREREREREREREQCokCziIiIiIiIiIiIiAREgWYRERERERERERERCYgCzSIiIiIiIiIiIiISEAWaRURERERERERERCQgCjSLiIiIiIiIiIiISEAUaBYRERERERERERGRgCjQLCIiIiIiIiIiIiIBUaBZRERERERERERERAJit7oBIiIiIiIiIiIiQz2yvXfg592t5rDX75ulsJZIJIm5PdIwjH3AZA+Lj5imWRjG5oiIiIiIiIiISICyE6xugYiMJeYCzf2agJ+N8HpruBsiIiIiIiIiIiK+U8aySHSJ1T32hGma5VY3QkRERERERERERGQ80GSAIiIiIiIiIiIiIhKQWM1oTjQM4zZgEtAGbAHeMU2zz9pmiYiIiIiIiIiIiMSeWA00FwK/H/LaXsMwvmia5iorGiQiIiIiIiIiIiISq+LKy8utbkNQLVmyJBNYCnwH+AHwJyAOWAh8esmSJX8vLy8/MnQ9wzDuXLJkyW+vuuqqYoBTTjmFqqoqcnNz2bhxI8XFxaxbt46cnBwqKytJTEzkwIEDtLe309bWxsGDB0lISGDbtm3k5eWxYcOGQetUVVURHx/PwYMHaWtro6Ojg7q6OhITE6moqKCgoGDQOhMmTGDHjh3ExcVx+PBhWlpa6OzspLa2luTkZLZs2UJhYSHr168fWCcrK4udO3diGAb19fU0NzfT3d3N/v37SUlJYcuWLRQXF7N27VpKSkpYt24dmZmZVFdXY5omx44d4/jx4/T29rJv3z7S0tLYvHnzwO93fV5GRga7d+/G4XDQ2NhIQ0MDDoeDPXv2kJ6ePmidgoICNm7cSHp6Onv37qW3t5fjx49z7NgxAKqrq8nKymLTpk0D6+Tn57N582ZSU1PZt28fPT09NDU1UV9fj2EY7Ny5kwkTJgxaJy8vjy1btpCSkkJNTQ1dXV00Nzdz5MgR4uLi2L59Ozk5OWzYsGHg356bm0tFRQWJiYnU1dXR2dlJa2srhw4dwm63D2x/1zpr164lNzeXbdu2DWzLodu/srKS/Pz8QdsyOzubqqoq7Hb7wLYcuv2HbkvX9rfZbBw5coTm5ma6urqoqakZ2P5FRUUet//Ro0c5ceIEPT09w7b/unXrBh4zMjKGbf++vj727NlDWloamzZtGvh7FRYWsmHDhlG3f0ZGxsB2Wbt2LYWFhWzatInU1NQRt/+uXbs8bv+UlBT2799Pd3f3sO2fnZ09aL/My8tj69atJCUlUVNTQ2dnJy0tLRw+fHjQ9ndfx337u/bl1tZWDh48SHx8PFVVVSPuy+7b331fdt/+69evH/i7ZWdns337do/78tatWwdt/7Vr1464/V37smv7D92Wrn0ZGLT9R9qXXd8d1/Z3OBw0NDQM2v5D92XX9nffl923/+7du8nMzBy0LQsKCga2v/u+fPTo0UHbf+PGjQP7WH5+vsd92WazDdv+rnVc27+2ttbjvqxjuY7lOpbrWK5juY7lOpbrWK5juY7lOpbrWK5juY7l4/FYXlRUtCTgoKsXDNM0w/E5ljMM43+AbwEvmKZ5g6f3PfrooybAV77ylXA1TURERERERERERCRUjHB8yHiaDPB/+x8vsrQVIiIiIiIiIiIiIjFmPAWa6/sfUy1thYiIiIiIiIiIiEiMGU+B5gv6H/dY2goRERERERERERGRGBNTgWbDMMoMw8ge4fXJwK/6n/4hvK0SERERERERERERiW12qxsQZDcB/2EYxkpgL9ACTAeuApKAl4H/sa55IiIiIiIiIiIiIrEn1gLNK4FTgbNxlspIBU4A7wK/B35vmqZpXfNEREREREREREREYk9MBZpN01wFrLK6HSIiIiIiIiIiIiLjSUzVaBYRERERERERERGR8FOgWUREREREREREREQCokCziIiIiIiIiIiIiAREgWYRERERERERERERCYgCzSIiIiIiIiIiIiISEAWaRURERERERERERCQgCjSLiIiIiIiIiIiISEAUaBYRERERERERERGRgCjQLCIiIiIiIiIiIiIBUaBZRERERERERERERAKiQLOIiIiIiIiIiIiIBESBZhEREREREREREREJiALNIiIiIiIiIiIiIhIQBZpFREREREREREREJCAKNIuIiIiIiIiIiIhIQBRoFhEREREREREREZGAKNAsIiIiIiIiIiIiIgFRoFlEREREREREREREAqJAs4iIiIiIiIiIiIgERIFmEREREREREREREQmIAs0iIiIiIiIiIiIiEhAFmkVEREREREREREQkIAo0i4iIiIiIiIiIiEhAFGgWERERERERERERkYAo0CwiIiIiIiIiIiIiAVGgWUREREREREREREQCokCziIiIiIiIiIiIiAREgWYRERERERERERERCYgCzSIiIiIiIiIiIiISEAWaRURERERERERERCQgCjSLiIiIiIiIiIiISEAUaBYRERERERERERGRgCjQLCIiIiIiIiIiIiIBUaBZRERERERERERERAKiQLOIiIiIiIiIiIiIBESBZhEREREREREREREJiALNIiIiIiIiIiIiIhIQBZpFREREREREREREJCAKNIuIiIiIiIiIiIhIQGIu0GwYRqlhGP9nGMZBwzC6DMPYZxjGzwzDmGB120RERERERERERERikd3qBgSTYRjTgfeBfOBvwHZgLnAvcLlhGBeaptlgYRNFREREREREREREYk6sZTT/BmeQ+eumaV5vmuZ/mKa5APgpcCrwn5a2TkRERERERERERCQGxUyg2TCMacClwD7g10MW/wBoAz5rGEZqmJsmIiIiIiIiIiIiEtNiJtAMLOh/fN00TYf7AtM0W4D3gBTg/HA3TERERERERERERCSWxVKg+dT+x50elu/qfzwlDG0RERERERERERERGTdiaTLAzP7HJg/LXa9njbTQMIw7gTt/85vfAPDoo48GtXEiIiIiIiIiIiIiFjC/8pWvGKH+kFjKaB6L649pjrTQNM3HTNM8N4ztEREREREREREREYkJsZTR7MpYzvSwPGPI+0YUjui+iIiIiIiIiIiISCyJpYzmHf2Pnmowz+x/9FTDWURERERERERERET8YJjmiJUkoo5hGNOBamAfMN00TYfbsnTgEM7Aep5pmm2WNFJEREREREREREQkBsVMRrNpmruB14EpwD1DFi8BUoHfKcgsIiIiIiIiIiIiElwxk9EMA1nN7wP5wN+AKuA84GKcJTM+Zppmg3UtFBEREREREREREYk9MRVoBjAMYyLwIHA5kIOzZMYLwBLTNButbJuIiIiIiIiIiIhILIq5QHMwxMXZTIcjMv4uhmEQK9vIZkCE/FklgtgAx5jv8k0s7TehYBg23MrYSyQyDNB32HeGDWLtu63vgp8MQH+30YTi/AuR29+L9b5BpP7dY02o9ptgi/XveyCish+svsDooqn/FzHbcnz0k6LlmB1Gx0zTzAv1h9hD/QHRyOEwMddZ3Qon41yTd955x+pmBMVFF12E+SurWyGRxvgqmGfcGNzfueV5rtlxe1B/Zyx56dQnueWDbqubIaN49oIEzvnzLqubEXXW3zQTvhdjneYfGtzyqW1WtyLqPPvcaXDdz61uRkRz/O1ezAvPCvrvNd7bFDH9aHex1KceifrZ4RGKfmsoqC/sWTT2g9UvHF1U9f8ipF83XvpJoerrRCvjvU37w/E5MTMZoIiIiIiIiIiIiIhYQ4FmEREREREREREREQmISmcEWVc33P0jeHMNNDbDjFL44T1wxYVWt2xkS5cuZf369XR2dpKdnc1nPvMZrr76aqubFfN21cPpP4Qbz4Y/fN7q1kS2+bvf5sP2RuyGAUCJPZkdsy63uFXeO/CP3ez81UY6DrWRmJvMWT+6iJxzC61uVkz484IJg573JaJoiQAAIABJREFUdXUwY9FdnPutn1nUotjWVV9HzRPltO3ciGFPYML5lzPxi9/HiBtnXYljVfDaPXB4PaTkwYJH4NQbrG5VSO3c9Qx7977AiaadTJ50Feef98OBZYePfMC69Q/R3n6InJwzOH/uf5KaWmJhay225x2oWQMtB6HkHPjIrc7XHb2w7ndwohY6GuHCr0LuTGvbOoJfHTrK00ca2dreyS15WTw9czIAH7a0sXj/Ida3dRAHzM9M4xfTSilKiA/O5z4HT/8dtlbDLZfB0+XO17ftgc/9AHbXOZ+fMxt+8e9w2rSgfGxQ/OUvf+HVV19lz549LFy4kO9973vD3vPUU0/x1FNP8ZOf/IRzzz3XglZa61er4OkPYeshuOUcePqzztf3NcDUH0Bqwsn3fucSWHyFNe0MtV8dq+bp4/vZ2tnELVkTeXriRweWtTt6+fdDW1h2oo4e08GZyVm8M32+dY31w94/bKN2+S5adjZSfPV0zv7RRQDUvVjNlh+8N/A+02Hi6Ozj43+5jqw5uVY11287//wb9r78O07srmDyJZ/i/MVPDiyrefPPbH3iQdqPHiAlv5Qz71pK6bzrLGxtYOpf+T0Nby+no2YH2RdezZSv/vfAsmMrlnH4r7+l58Qx0madw+S7HyYhu8DC1vqptwteuxv2vgmdjTBhBsz/IUzvPxDtXeHs9zXXQPF5cM3TkDk55M3q6+tm3foHOXzkA7q7m0hLm8SZZ/wbxUUX0dfXzQcffpvGxgra2g+y4OKnKcifG/I2eaWvF7Ysg6M7obsd0nJh9tVQcBo07oPt/4ATdc5a1bkz4fRFkJRpdauDpsvh4O7ddbzZ1EJjTx8zkhP54eQirpiQwbb2Tj63cz+7O52le85JS+YX00o5LSXJ4lYH3zi7Ogy93j6YWACrHoNJhfDye3Dzd2Hrn2BKsdWtG+62227jO9/5DgkJCezfv597772XmTNncuqpp1rdtJh2zzL4aOjPTzHjV8Vn86WcqVY3w2dH3ztA1f+s5ZyfLiDrjDw6j7Zb3aSYctNbxwd+7u1o469XlTJpwSctbFFsq3minPiMbM547H362prZufQLHH3tGfKvHEd3yxy98Px1cPZdcMsbULMK/nwN/OtGyDnF6taFTHJyPmWnfZlDh9+jr69r4PWuruO8+969zP3og5QUX8yWrb/gvfe/xaWX/MnC1losKRNOvRTqt0Nfz+BlOdNg+jxY+7QlTfNGcUI8908s4LUTLXQ4Tk6fc7y3jzsLc7ksKx27YfDVPXV8cVcNr5ZND87n5sH9t8NrH0BH1+DXn/8vmFwEDgf8+s/w6e/Blgj6iuXm5vK5z32ONWvW0NXVNWz5gQMHePvtt8nJybGgdZGhOBPuvxxeq4KOnuHLTzwC9rjwtyvciuOTuT9/Fq+1HKHD7Bu07M66DfSaDqpOvYzsuAQ2dZywqJX+S8pPYebdZ3L0nwfo6zr57yu9dgal184YeF67fCc7f7OJzLLo3CeS84oo+8J3ObT6Dfq6OgZeb68/wAdLvsDH//svFJ1/GQfff4X3vn8L1y7fRVJ2voUt9l98dj5Fn7ybpk3/xOzuHHi9pXI1B/74Y04p/wOJhZOpfeoh9v7sG5z64B8tbK2fHL2QPhFuWwWZk6D6ZfjrzfClrZCQBssXwZVPwMxrYNVi+Oun4AsfhrxZptlLSkohCxf8jtSUIg4eeof33v8mV1z+N5KT8sjN/QinnvJZ3n3/GyFvi0/MPkieAP/yNefjkW3Ofs+C70BPO0z+GHx0tnMCx63Pw8Y/wgVfsbrVQdNrmkxMjGfVnBlMSkzg5ePN3Lx9H1vPPpXiBDvPz5rC5MQEHMCvDx3j0zv2seXsWVY3O+hUOiPIUpOh/MvOoLLNBld/HKYWw/oqq1s2sqlTp5KQ4EwjMPozRg8cOGBlk2Len9ZBVjIsjN2YhPTb8csNnHL32Uw4Kx/DZpBckEpyQarVzYpJNW/9hcQJ+eSd9S9WNyVmddfXMuFjV2JLSCR+Qh6ZZ32cjtpxNjHNse3OTNW53wBbHExZAKUXQsXvrW5ZSE0svYTS0k+QmJg16PXaujfIzJjBpImXExeXyOlz7uFE0w6am/dY1NIIUHwmFJ0BCUOO9TY7TJ8POdOdWTwRalFOFtfnZJFjH5yLcsWEDG7KzSLDHkdKnI2vFuXyXnNb8D53AVw/H3KGJDVlpTv71IYBpglxNqiuDdrHBsW8efP4+Mc/TkZGxojLf/azn3HXXXcRHx+c7O9otOgsuP5MyBnnXaBFmSVcn1lCjj1h0Os7Olt4sfkgj5WeQ549kTjD4JyUCR5+S+QqunQKRZ+YQkJW4qjvq/1rNaXXzxi49ow2E+ffQOm860jMzB70evvRA8SnZ1F8weUYhkHJhVdiT06l9UD0nhMnnHcZWXMvwZ4++PzftH4lEy64guSJM7HFJ1B04z20Vq2l63BY5hkLroRUuKgcsqY4g58zr4bMqc6RazuWQ24ZzL4J7Enw8XKo3+zsD4aY3Z7C6XO+SlpqCYZho6R4PqmppRxvrCQuLoFZp36OvLxzMIwIu0tnT4RZV0BKjvPvWTgHUrOdI7oKToOSsyE+CewJMPXj0LDX6hYHVWpcHOWTipiSlIjNMLg6O5OpSQmsb+0gy25nSlIihmFgAnEGVHcOv0EdCxRoDrEjDbCzBoKU8BESP/nJT7jkkku47bbbyMnJ4fzzz7e6STGruQMe+Af8eJHVLYku3z28ldzKF7mweiVvt9Zb3RyvmH0OTlQco+t4JysuWcYbFz3L1gffp6+z1+qmxaS9r/yBqVfcGrUXLdEg/8rP0/jeP3B0ddDdcJimTe+QcdZFVjcrzEaY0dw04WhF+JsSAZqaqsnKOjkCym5PIS11Ik1N1Ra2SsLhneZWysI41DNrPiRdCF97BL73xbB9bMBWrlyJ3W7nggsusLopEW3yA1B6P3zx93Cs1erWhN/qjkYmJ6TwgyOV5Fa+yOk7X+cvTXVWNysk2g+00LDuMBOvi7zSQYHKnnUOGZNnUffPl3D09VG36m/Y4hPJmnG61U0LOtM0nf2fky8AxEYCQusRaNzpDDAfrYT8M08uS0iFrOlwrDLszeroPEZLyz4yM2eM/eZI0tkMrUchvWj4sobdkBHbJSWPdPews6NrUJ8p68MtJL2/ma/tOcD3SqOw3IwXVDojhHp64dbF8PmrYNYUq1vj2Te/+U3uvfdeKisr2bhx40CGswTf4n/A7R+DidGXpGCZ/yo6ndMSM0gwbPzpRC3X7HufTTM/wfTENKubNqquYx2YPQ4OvbqPC5+5GsNuY+3db7Dz0U3M/sb4q88YSm2Hazi68R3O+95vrW5KTEs7bS7H3lzGxs+dDY4+cubdQNbcS6xuVnjlzILUfPjwEWdW8/6VzvIZky+2umWW6O1tJzFxcEZXfHw6Pb3By3SVyLOlrYMHa4/wt1nhK2l14m1o64D/93dnGY1o0N7ezmOPPcaPf/xjq5sSsXLTYO19cFYpNLQ5S8vd+jS89lWrWxZedT3tVHQ288mMEg7OvpoP2hu4at+7nJaYweykkTPlo1XdC9XknFtAysR0q5sSdLa4OKZecRsf/OBz9HV3YrMncOF/Pos9OfZS+TM/Mo89P/03ci+9haTCKRx6/ldgGDjcSolEpb4eePFWOP3zkDsLelqd83G4S8qE7pawNsvh6OGDD7/N1CnXkZERQZMUjMXRB+t/DxPnQvqQgGrTAdjxGpz3JWvaFgY9DpNbd+7n8/nZzHILNJ84/wza+vr4f/XHmZwYm6OdlNEcIg4HfHYxJNjhV9+xujVji/v/7d13nFTV3cfxz2+XLWyhs1QVFESKWMASFEtijZLHgpoYUYMmahLU5DHREPMoeYJ5UnjUiCaaYIjhicZgxFBEDQYVrDRRUAPCUpa21G1smZnz/HHuLsMys4Uts7t836/XeV3uuefMPcP89s6dc+89JzmZ4cOHk5+fz6xZsxLdnDZpxWb456fwvSOzP+KwnZHRlezkFNKSkrmpSz/OyujKvMJtiW5WrZLS/XW8/uOGkJ6TQVqXdI79xjB2vNHCnvltA9a/PINuw88iq3frG8e7tXCRCGt+Np5OZ1zEKTM+5KSn3ydUXEDejF/WXrktSU6BsbNg7Vz4TU94bwoMvhY69E10yxKiXbsMKioOvv2wIlRESru296NavLX7y7h09Toe7d+H0R2b94JvZnu4/Wo/OeCO3c2668Py9NNPc/HFF9O7dwucpKWFyEqDkcf48Zl7dICp18Krn/onAI8k7S2ZFDPu7zGY1KQkzs3qzvmZObxatD3RTWt0m19aS98r2t7dzADb3l/Aisd/xBcf/yfXvVnMl55YwPs/v509/16R6KY1ug4njqL3tXey7tff5aNvn0tqTl+S0jNJ6dqK7051EfjHOEhOhYun+ryULCgrOLhcWQGkNt+FEucivPPufSQlpTByxP3Ntt8GcxFY9mc/1NzwsQdvK8qHd5/0EwF2bcGP/jdAxDnGrdlAapIx9dhDfydkJidze8+u3LhmIzvKY0xe0MrpjuYm4Bzc8t+wfTfMexRSWtH/cjgc1hjNTWThGsjdDUf/xK8XlUHYweqtsOy+xLatNTEDF+vx9RYmtWMa6T0zQSM5NLncl2cweNwPE92MNi1ctJeKXVvJuWQcSSlpJKWk0e38q8l77mH6jmsFV1MbU85wGPfGgfU/jfJ3vhyBOnYcwPr1L1Wth0IlFBVtan2PdUqdbCgt54JVa/lJ3x6My+lSe4UmEIlASSnk7YAENaHOli1bdtANHHv37uWBBx7g+uuv5+tf/3qCW9cyVZ4ytfyzvMY1PL1j7YXagN1Lt1O6o4TeF/dLdFOaxJ41H5Jz8tl0HTwCgK5DRtJ16Gls++B1Oh9/coJb1/hyLrmBnEtuAKB0y3q2vfAE7Y9qpZMQOQdzb4Hi7XDdPH9jAUD3ofDRnw6UKy+GPZ/7YTWapVmO996/n9KyXZw7+nckJbWSu1+dg+XPQlkhnHmb72yuVLIb3n4Cjr8IjjotcW1sQs45blm7ie3lIeYNOZaUpNgdAhGgJBIhr7yCnNRW8tnWke5obgJ3/Bw+WQ+zH4b2zTd8Xb3t2bOHBQsWUFJSQjgc5v3332fBggWceuqpiW5am/Sts+DzB2HFj3y6/Wy4bCi88p1Et6zl2hsu55XCbZRGwoRchP/bs5E3i3ZycVbruFp+1FUDWf/n1ZTt2k/5vjLW/2kVPc47OtHNalPyV75DSf4Wjv7i1YluSpvWrkMXUnP6kv/qX3DhEKHiAnYtfJGMY9reLMm12rESQqV+5ux3fw1FW2H4zYluVZOKREKEw2VEXBjnwv7fkRB9+1zAvoI1bNr0KuFwGR+v+i2dOh7fuh7rbGyRsH/01kV8Clf4PIBwyK8fVK5ldamFnKM0EiHsHGEHpZEIIefIKyvni6vW8p2e3bm9V7fG328ISssgHIFw2P87FILX3oXln/q8giL4/sPQORsGt6AHWEKhEGVlZUQiESKRCGVlZYRCIR5++GGmT5/OtGnTmDZtGl27duWee+7hyiuvTHSTm10oDKUVwecb8f8OheG9XPhsu7+AsKsI7pwJ5w2Eju0T3eKmEXIRSiPh4O/LVZ3fnpPVnaNTMvj5jk8JuQiLi3eysDifi7Na19idkVCEcFkIF3G4sP93JBSp2r5p1hp6XdSPdlmte5jGSChEuKyUSDiMi4T9v0Mhug4eyY4PF1fdwbz7s+Xkr1jcqsdoduEQkfIyiERwkQiR8rKqvP0b/41zjvL8LWx48n5yvnwT7bJa6UWT+XfAzk/g2tmQEnUAOv5KPw/Hpy/4c79FP/U3HHRrnvPfJUsnUVCwjnPOfpx27Q7uWAqHywmH/URykUgF4XCZHzu7JVj5vB/r+oxv+TvEK+3fC4unQv+zfWqj7vh8M5+UlDJ7SH/aJx/ocn1tbyHLi0oIO0dBKMz31+fROTmZwc0450VzaUX32rYOG7bCk3+HtFToefGB/CcnwtcvTVy7YjEzZs2axZQpU4hEIvTo0YMJEyYwevToRDetTcpI9alSVhqkp0D3tjdEWaOpcI77t63i07JCkjFOSM9mVr9RDEpvHf9px3/7FMr3lPL6xTNJTkum96X9GXjHSbVXlDpbP+/PHHXeFaRkto6YaM2Ou+dxNk2fzLaXnoKkZLKHnkHfm3+c6GY1v4/+DB/+wXcSHjUavvaan2G7DVu1+nd8vOqJqvXcDbMZNvTbnDjsu5w96hGWLJvMO+/dS9cuwxk16ggfj/bfr8Jn8w+sb14Cgy7xM7AvmAz7gzEf3vmtX174X35m9hbiZ5u2MWnTgcf1Z+Tv4YGjemAY60rLmbRpG5M2HRi+qugLwxtnv9Ng0u8PrM94GR74pp9Me8KvYPMOaJ8Gpw2B+Y9Begv6k3vmmWeYPn161fqrr77KzTffzPjx4w8ql5ycTHZ2NhkZGc3cwsT72XyY9PKB9RkfwAOXwqAeMHE27CiEDulw4QnwbCua7LG+frb9Eybt+KRqfcbejTyQM5gHew7lpX6juHXzUv5nx2cck5rBM0edxgmtbHzmNb9dwb+nLq9az/vH5xz/3VMYNOFUwmUhtry8npGPfSmBLWwcq6Y/xMfTfla1njv/Lwy75X5OvPW/OPGWn7Bo4tco3bOdtE7dGXLTvfQ6o/XOZ7H1hSfY+rfHqtZ3v/USva6ZQM5lN7P+0e9Ttn0jSemZdDv/anpfd3cCW9oA+zbA8ichOQ0ejbqZ6dInYdjX4aoX/MDx/7gBep8BVzzXLM0qLs5j7efPk5SUyqx/HJh8+7QRD9Kv3xjmzvsyxSVbAFj4xjcBGHP5a2Rl9mmW9sVVshty34akdjA/aqiPk66D4nwo2eXPk6LPlS7/VfO3s4lsKC3nye27SDOj5/sHJo188ri+pCYlMWHdZjaXVdA+yTgtO4P5Q48jPant3f9rLeaqRwtiZs4tSXQrPBsJb775ZqKb0SjOOecc3NREt0JaGvsuuOrjNjX0NVfOZMxntzTqa7YlswdN42vvlCe6GVKDZ7+Qyoi/tYGZu5vZ0msGwsQ2dl7zkPG161YnuhWtzrN/HQL/8Wiim9GyvXQX7qzGf5zbFq+gpZxHR2tL59Sx6Dy7eTTFeWtT0LlwfK3xPFjnhTVrVed/LeS87og5T2qic53WyhavWOqcG9nU+2l7XeciIiIiIiIiIiIi0qzU0SwiIiIiIiIiIiIiDaKOZhERERERERERERFpEHU0i4iIiIiIiIiIiEiDqKNZRERERERERERERBrEnGsls3M2IzPLBzYkuh0iIiIiIiIiIiIiDXSMc657U+9EHc0iIiIiIiIiIiIi0iAaOkNEREREREREREREGkQdzSIiIiIiIiIiIiLSIOpoFhEREREREREREZEGUUeziIiIiIiIiIiIiDSIOppFREREREREREREpEHaJboBLVG7du1cUpL64EVEpOWKhMKYvquahMOh8wA5UkVCEcws0c0QSThnzftdEAqFSUrS356IiDSNcDi80znXvan3o47mGJKSkigoKEh0M0REROLqmtmJNwf/MtHNaJPOy/0x+fn5iW6GSEJ0zujKEz2WJroZIgl3Z8Fp5O9svu+CzMyOnHjifc22PxERObKsWPHTDc2xH92uIyIiIiIiIiIiIiINoo5mEREREREREREREWkQDZ0hIiIicQ2bfzMpXTNwkQiR/SH2vZVL4Xub6XnrSFZfMaOq3MCnriAlJ+uQvIL3NmEYWSN6s/bb/6jaNnTOjZRt3HtI3pap77Bn/hpGfHQn4ZIKwBEuLGfPK2vYPGURRFyzvG8RkaZwycyj6TIkjedOXkuk3B/Pzn64F8de2YFIeQSAos0hNr1WxMqpu6go9HkDru3IWVN6Ei71dUp3hdj2TgkrH9tFwbqKxLwZEZHDMHfu3XTvns1FF01h796SqvznnrudQYN6sXjxGk455WgAUlLaAY6KijAA8+atZPLkOYlotojUkTqaRUREpEZrJ8ym8N1NpORkMvB3V+BCEdof14V2ndsT2rMfko32x3cjUho6KC/zpJ5sefxdMKPnLSMgySDiaNc1A2uXRMbg7gflpR/TiaKlW6r2+8nYv1C2aR9p/Tsz6OmrKM3dw86/fZzA/wkRkcOX1TeFHme0p6IgwtEXZZE7p7Bq28dP7GLZL3eSnGZ0PiGNkfd357KXjmHOZbmE9vvO5fyl+5l3xUYsCbKOSmHobV34yvx+zBmzgb2flSfqbYmI1Fte3l4uueREnnvuPQAGDMghLS0FgMmT57B1614AJk26gu3bC3jiidcT1lYRqR8NnSEiIiJ1UrGjmH2LcknpnknZpn1kjegNQMbgHEo/303RkryD8syMklU7KPl4u+9YPqEbANkj+1D4wWZKc/celFe6cS8V+cWH7Lds/R6Klm2h/cCuzfRORUQa33HXdCB/2X7WPL+PAdd0jFkmXObY+WEp/7w5j7TOyQz86qHlXAQKN1Tw7sTtbHtnP6f8Z7embrqISKOaO/dDLr/8pKr1MWNOZs6cD+tUt3PnDKZOvYHXX/8hCxfey8MPf41u3bKqtl955Qjmzr2bRYsmMnv2XVxwwZCqbVdfPZIXX5zAwoX38pvffJ3u3bMb702JCKCOZhEREamjlB5ZdBzdj/2f5lO4NI+sEX0AyB7Rm8JleRQt33JQXtHKbbhQBBeKUPzR9qptWSN6U7R0C0XLtlTLy4u53/Rju5B1am9KPslvhncpItI0BoztyLq/F7Du7/voc14m6d2S45YNFUfY8mYxPc7IqPE1N7xcWGsZEZGW5qOPNpOZmUb//t1ISjIuumgY8+bVraPZzJg58wMuvfR/ufzyRwC4555LAMjOTud737uQ2277E2ef/RDjxz/N55/vAODii4dx/fVnctddf+GCC37Fp59uZfLkq5vmDYocwdTRLCIiIjU67tHLOGnxbQx6ZiyFS/LY+vsPKFqSR3ZVJ3EfipZtoXDZloPzlhzoOC5ccqBjOuvU3hQtyzu4o/nU3hQuObijefDzX+Okxd9iwNQx7Pz7KnbNWt0cb1dEpNHlnN6erL4prJ9dyK6PyijYUM6xV3aosU7J9hCpneJ3RgOUbAuRVksZEZGWyN/VfDJnnnkcubk72bGjsPZKwO7dxSxc+BllZSGKisp4+um3OPXUfgeVGTAgh9TUduTnF7J+/U7A3838hz+8wcaNuwiFIjz11EJOPvlounTJbOy3JnJE0xjNIiIiUqPP75pL4bubDsorXJrHMZO+RHKHNDKH96T43leI7K8gpXsGyR3SyDqlF9v/vKKqfNHSPLp/dTjJHdJo17k9ZRv3UbFrP/0mX0hyhzTaD+h60PjMAJ9c+yxlm/Y1y3sUEWlKA67pSN4bxZTt9hNarXuxgAHXdmT17/fErZPRsx3le8M1vm5Gr3aU1VJGRKQlmjv3Q6ZNG0+fPp2YM2dF7RUCGRmp/OAHl3LGGceSnZ0OUDW+c2FhKRMnvsANN4zipz+9kmXLNjBlyits2rSbXr068uMfj+FHP7qs6rXC4Qg9enRg9+5Dh24TkcOjjmYRERGpt/LNBVTkF9Nt7DDKtxYS2V8BQNGH2+g2dhhJGakUr9xaVb7ow60kZ6fSbewwilf4/EhxORU7/GtU5BdTnleQkPciItKUktON/mOysWTjuhUDfF6qkdYpmc5D0mLWaZdh9B6dycrf7KzxtY+5JJvt75U0eptFRJra1q37yMvbw1lnDeTBB1+qc71vfGM0PXp04IYbnmL37mJOPLEvTz89vmr7okVrWLRoDenpKdx114VMnHg5d9zxDNu3F/DII6+yYMEnTfF2RCSgjmYRERE5LEXLttDjxlPYPfezg/J6jh9ByartuLIDd9m5sjAlq3bQ48ZT2PrUBwfKL/evUfDuxmZtu4hIczn6kixcGGZ9aR2RcleVf96TfRgw9uDJ/pJSjc6DUhl5fw7l+8Ksee7QpzosCTL7pDD0ti70HJXB3DEbmvw9iIg0hUmTXqJDh/aUllaQnFy3kV0zM1MpLa2gsLCUTp0yuPXWc6q2de+ezaBBPVmyJJfy8hD795cTDkcAmDnzA2699VzWrt3Bhg27yM5O5/TT+6vjWaSRqaNZREREDkvhkjy6fHkQRcsODHlRtGwLKT8Yzc4XDx1PuXBJHlkn96Jo+cHlc64/6ZBhM0RE2ooB13RkzV/3UZwXOij/kz/u4Yz/7sHWt4oZ9u2uDLm1MyQZRZsq2PzPIv712C5C+w90THcf0Z4b1hwPBmW7w2x7u4TZl+ayb215c78lEZFGsXnzHiD+EEKxPPPM2zz00NUsXHgv27cX8Oyz7zFqVPC0SHIS48eP5uc/H4tzjtWrt/LQQ7MBmD//Y9LTU5ky5av06NGBwsJSFi9eo45mkUZmzrnaSx1hUlNTXUGBHt8VEZGWq2tmJ94c8stEN6NNOi/3x+Tn5ye6GSIJ0TmjK0/0XJroZogk3J0Fp5G/s/m+CzIzOzJ8+H3Ntj8RETmyrFjx06XOuZFNvZ+6PZsgIiIiIiIiIiIiIhKHOppFREREREREREREpEHU0SwiIiIiIiIiIiIiDaKOZhERERERERERERFpEHU0i4iIiIiIiIiIiEiDmHMu0W1occysEPgs0e2QNqEbsDPRjZA2QbEkjUFxJI1FsSSNRbEkjUWxJI1FsSSNRbEkLckg51x2U++kXVPvoJX6zDk3MtGNkNbPzJYolqQxKJakMSiOpLEolqSxKJaksSiWpLEolqSxKJakJTGzJc2xHw2dISIiIiIiIiIiIiINoo5mEREREREREREREWkQdTTH9lSiGyBthmJJGotiSRqD4kgai2JJGotiSRqLYkkai2JJGotiSVqSZolHTQYoIiIiIiIiIiIiIg2iO5pFREREREREREREpEHU0SwiIiLr3WYIAAANyUlEQVQiIiIiIiIiDaKO5oCZ9TWzp81si5mVmVmumT1iZp0T3TZJDDMba2aPmdlbZlZgZs7MZtRSZ5SZzTOz3WZWYmYrzexuM0uuoc7lZrbQzPaZWZGZvWdmNzX+O5JEMLOuZnarmb1oZmvNbH/wWS8ys1vMLOZxWLEksZjZL8xsgZltCmJpt5ktN7MHzKxrnDqKJamVmY0Lvuecmd0ap0y948LMbjKz94Py+4L6lzfNu5DmFpwvuzhpW5w6OiZJXGY22sxeMLOtwW+yrWb2qpl9OUZZxZIcxMxuruGYVJnCMeopliQmM7ssOAZtDs6915nZ38zsC3HKK5akSZg33szeNbPCIL6Wm9md1ePLzFLM7C4z+6OZrTCz8prO8YM6Z5nZL83sAzPLD76D15vZH8xsQL3aqjGawcyOA94GcoCXgE+B04Hzgc+As5xzuxLXQkkEM1sBnAQUAZuBE4D/c87dEKf8fwAvAKXAX4HdwBhgEDDTOXdNjDrfBR4DdgV1yoGxQF9ginPunkZ+W9LMzOx24LfAVuBfwEagB3AV0BEfM9e4qIOxYkniMbNyYBmwGtgBZAJnAiOBLcCZzrlNUeUVS1IrMzsK+AhIBrKAbzrn/lCtTL3jwsx+Dfwn/jt0JpAKfBXoAkxwzk1tqvckzcPMcoFOwCMxNhc5535drbyOSRKXmd0P/DewE5iDP3fqBpwC/Ms598OosoolOYSZnQxcEWfzaOCLwFzn3OVRdRRLEpOZ/QL4If5znoU/Ng0AvgK0A250zs2IKq9YkiZjZs8A4/C/AWcDxcAFwBCq9SmYWSdgT1B1Oz6ujiLGOX7U628DuuP7RpcCIeALwKhgXxc6596pU2Odc0d8Al4BHP5HT3T+/wb5v0t0G5USEhfnAwMBA84LYmFGnLIdgj/4MmBkVH568IfqgK9Wq9MP/yW0C+gXld8ZWBvU+UKi/x+UGhxHX8SfYCRVy++J73R2wNWKJaU6xlN6nPzJwef8hGJJqT4p+I77J/A58KvgM761oXGBPyl1wfbO1V5rV/B6/ZrqfSk1W/zkArl1LKtjklJN8XFN8Hm+BmTH2J6iWFJqSALeCT7nryiWlOoQLz2BMLANyKm27fzgc16nWFJqjoS/gOaAdUC3qPwU4MVg281R+anApUCvYP1BYpzjV9vHvUDvGPkTg7of1bW9R/zQGWZ2LHAR/kT58WqbH8D33I8zs8xmbpokmHPuX865NS7466rFWPzVn+ecc0uiXqMUuD9YvaNanfFAGjDVOZcbVWcP8FCwevthNl9aCOfc68652c65SLX8bcDvgtXzojYpliSuIA5ieT5YDozKUyxJXdyJvyD2Dfw5TyyHExeV65ODcpV1cvHnW2nBPuXIoWOSxGR+GLFfACXA9c65wuplnHMVUauKJakXMxuGfwIsD5gbtUmxJPEcgx9q9j3n3I7oDc65fwGF+NippFiSpnRVsJzinNtZmRl8N/4kWJ0QlV/unHvZObe1rjtwzv3CObclxqZfAPuBYRZnqMbqjviOZvyPK4BXY3QEFQKLgQz8F5NIPJVxND/GtjfxJ86jzCytjnVerlZG2qbKH02hqDzFkhyOMcFyZVSeYklqZGaDgf8BHnXOvVlD0cOJC8XSkSPNzG4ws4nBeIDnxxmLUsckiWcU0B+YB+wJxkS9N4inWOOgKpakvm4LltOcc9FjNCuWJJ41+OEGTjezbtEbzOwcIBv/RFglxZI0pZ7Bcl2MbZV5pwZDZjQ2x4H+ikPGuI9FHc1+vByAf8fZviZYHt8MbZHWK24cOedCwHr8OE7H1rHOVvydZX3NLKNxmyotgZm1A24MVqNPLhRLUiszu8fMHjSzh83sLfyYlivxnYaVFEsSV3AM+jN+CJ+JtRSvV1wET4H1wY/RG+tOCp1btS098bE0GT9W8+vAGjM7t1o5HZMkntOC5Xb8PARz8N9njwBvm9kbZhZ956BiSerMzNoDNwARoPrYpIolick5txs/lEAPYLWZPWVmPzez54FX8cP83BZVRbEkTanyLub+MbZFx9QJTbDva/AXVt51zu2tSwV1NPvJuAD2xdlemd8UVwak7TicOKprnY5xtkvr9j/AMGCec+6VqHzFktTFPfjhne4GzsZfrLjIOZcfVUaxJDX5L/wEWzc75/bXUra+caFzqyPHH4Ev4TubM4ETgSfx406+bGYnRZXVMUniyQmWtwPt8ZMbZePPk14BzgH+FlVesST1cS0+Fl52URMmBxRLEpdz7hH8kAXtgG8C9+E73TYB06sNqaFYkqY0J1h+38y6VGYGN45MiirXuTF3amb98ZNVhvATfNeJOpprZ8GyLuP0isRzOHGk2GujzOxO/IH6U/zMsfWqHiwVS0cw51xP55zhO3euwl/JXm5mp9bjZRRLRygzOx1/F/MUV9fZo2t5yWBZ37hQHLVyzrlJwVwE251zJc65j51zt+Mn1G6Pn3ymrnRMOnJVDrViwFjn3ALnXJFzbhVwJbAZODfOMBqxKJYk2reC5ZOHUVexdAQzsx8CM4HpwHH4C6oj8EMV/J+Z/bI+LxcsFUtyOJ7DD6dyHAfusH8EWAF8mQNPC9ZpaIu6MLOcYJ/dgbucc2/Xta46mmu/StShWjmRWA4njupap6AB7ZIWxsy+AzwKrAbODx7LiqZYkjoLOndexE9q2xV4JmqzYkkOETVkxr85MHlIbeobF7WVr+0OHmn9Kie7PScqT8ckiadywtB1zrkPozcET1xUPvl1erBULEmdmNkQ/Bjgm/FjgFenWJKYzOw8/CRo/3DOfd85ty64oLoMfwEsD/hPM6sctkCxJE0mmE/uK/inWrfhb1Ybjz+2nQ3sCoruiPkC9RR0Mr+OH97lLufcE/Wpr45m+CxYxhsncGCwjDeGswjUEEfBj/r++McN1tWxTi/8FdPNzrmSxm2qJIqZ3Q1MBT7GdzJvi1FMsST15pzbgL94MTRqwhLFksSShf98BwOlZuYqE344FoDfB3mPBOv1igvnXDH+B1hWsL06nVu1fZU/dDKj8nRMkngqP+d4Yz9WdkS3r1ZesSS1iTcJYCXFksRzebD8V/UNwWf7Pr4/7ZQgW7EkTco5F3LOTXHOneyca++c6+CcuwT/G/BkYD+wqqH7CWJvITAE+I5z7jf1fQ11NB84cFxkZgf9f5hZNnAW/gN7t7kbJq3K68HykhjbzgEygLedc2V1rHNptTLSypnZvcDD+Mdbzq82plc0xZIcrt7BsvKHlGJJYikDpsVJy4Myi4L1ymE1DicuFEtHtsohDqJ/UOuYJPG8ie+AGWhmqTG2DwuWucFSsSS1MrN0/F1/Efx3WiyKJYknLVh2j7O9Mr88WCqWJFHGAenA8865ioa8kJn1Bd7ATyp4e33vZK7inDviE/5xLAdMqJb/v0H+7xLdRqXEJuC8IBZmxNneAcjH/4AfGZWfDrwd1P1qtTr9gVL8Yw79ovI7A2uDOl9I9HtXapT4+UnweS4ButRSVrGkFC82TgB6xshPAiYHn/NixZLS4Sb8eLoOuLWhcYF/VNkF2ztH5fcLXqc0+rWUWl8Chsb6TgOOwY8V6ICJUfk6JinVFE8zgs/zZ9XyL8R3FO4FOimWlOoRU+OCz3V2DWUUS0rxYuPa4LPcBvSptu3S4Li0H+iqWFJqjgR0iJF3GrAbKASOraHug8Q4x69W5mjgc/xNS99oSFsteMEjmpkdh//jzwFeAj4BzgDOxz/WOco5tyv+K0hbZGZXAFcEqz2Bi/F35rwV5O10zt1TrfxM/JfFc/g/+K/gx7WZCVzrqv3BmdkE4Df4L5e/4q+IjgX64idpugdp1czsJvwEEmH8jK2xxiTNdc5Nj6qjWJJDBEOv/Ap/59fn+M+6B3AufjLAbcCXnHOro+oolqTOzOxB/PAZ33TO/aHatnrHhZlNAb6PHz9uJpAKXIcfT3yCc25qk70ZaXJBvNyHfzpwPf5HznHAZfgf1vOAK51z5VF1dEySmILxIBcDA/Dn2u/jL1pcif9xfL1z7m9R5RVLUiMzews/dulXnHOzayinWJJDBE+7vwJcgP9+exF/rj0YP6yGAXc75x6NqqNYkiZjZu/hL258jI/JofiJAMuAq5xzr1Qrfx/+RiXwQ2uchO/3rJw4cFH0+b6ZrcffELIUmBOnGdOdc7m1NjbRvfItJQFHAX8EtuL/uDfgJ+yq8e5DpbabOHDVJ17KjVHnLPwPqz3BQeAj4HtAcg37GYN/PKEQKAY+AG5K9PtXarY4csBCxZJSHWJpGPA4fviVnfjHjPcFn/OD8b6vFEtKdU3UcrfD4cQFcFNQrjio9wZweaLfq1KjxMu5wLPAp/i7TSvwd3O9BtwI/oaWGPV0TFKK9zl3wT9Ruh7/e2wX/iagMxVLSvWMpcHB99mmmuIhqrxiSSnWZ5wC3I0fRrUgOPfege+Eu0ixpNScCfgBvhN4L75zeT1+8uV+ccovpOY+iOnVytfWZ+GA8+rSVt3RLCIiIiIiIiIiIiINoskARURERERERERERKRB1NEsIiIiIiIiIiIiIg2ijmYRERERERERERERaRB1NIuIiIiIiIiIiIhIg6ijWUREREREREREREQaRB3NIiIiIiIiIiIiItIg6mgWERERERERERERkQZRR7OIiIiIiIiIiIiINIg6mkVERERERERERESkQdTRLCIiIiIiIiIiIiIN8v/7rs/xrqt+vwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 1728x576 with 3 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Plotting the predicted fitness effect for the variants seen more than once\n",
    "plt.close('all')\n",
    "f, (ax1, ax2) = plt.subplots(2, 1, sharey=True, sharex = True, figsize=(24, 8))\n",
    "gs = matplotlib.gridspec.GridSpec(3, 1, width_ratios=[1], height_ratios=[10,0.5, 0.5])\n",
    "ax1 = plt.subplot(gs[0])\n",
    "ax2 = plt.subplot(gs[1])\n",
    "ax3 = plt.subplot(gs[2])\n",
    "gs.update(hspace=0.05)\n",
    "\n",
    "nonsense_list = []\n",
    "y_list = []\n",
    "x_list=[]\n",
    "y_list_top20 = []\n",
    "x_list_top20=[]\n",
    "y_list_nonsense = []\n",
    "x_list_nonsense = []\n",
    "x=[]\n",
    "y=[]\n",
    "OBS_EXP_DICT={}\n",
    "top20_lower_err = []\n",
    "top20_upper_err = []\n",
    "missense_lower_err = []\n",
    "missense_upper_err = []\n",
    "nonsense_lower_err = []\n",
    "nonsense_upper_err = []\n",
    "\n",
    "top20 = c1\n",
    "missense = c6\n",
    "nonsense = c10\n",
    "synonymous = c3\n",
    "\n",
    "axisfont=20\n",
    "domainfont = 12\n",
    "titlefont=12\n",
    "axislabelfont=12\n",
    "\n",
    "top_20_DNMT3A = ['p.R882C', 'p.R729W', 'p.R326C', 'p.R320*', 'p.R882H', 'p.R736H', 'p.Y735C',\\\n",
    "                'p.R736C', 'p.W860R', 'p.R771*', 'p.R598*', 'p.P904L']\n",
    "\n",
    "PWWP_col = 'mediumvioletred'\n",
    "ADD_col = 'darkviolet'\n",
    "MTase_col = 'midnightblue'\n",
    "\n",
    "for k, v in DNMT3A_missense_s.items():\n",
    "    if k in top_20_DNMT3A:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_top20.append(codon_number)\n",
    "        y_list_top20.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        top20_lower_err.append(error_lower)\n",
    "        top20_upper_err.append(error_upper)\n",
    "\n",
    "    if k not in top_20_DNMT3A:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list.append(codon_number)\n",
    "        y_list.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        missense_lower_err.append(error_lower)\n",
    "        missense_upper_err.append(error_upper)\n",
    "\n",
    "    missense_err = [tuple(missense_lower_err), tuple(missense_upper_err)]   \n",
    "    \n",
    "for k, v in DNMT3A_nonsense_s.items():\n",
    "    if k in top_20_DNMT3A:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_top20.append(codon_number)\n",
    "        y_list_top20.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        top20_lower_err.append(error_lower)\n",
    "        top20_upper_err.append(error_upper)\n",
    "    \n",
    "    top20_err = [tuple(top20_lower_err), tuple(top20_upper_err)]\n",
    "        \n",
    "    if k not in top_20_DNMT3A:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_nonsense.append(codon_number)\n",
    "        y_list_nonsense.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        nonsense_lower_err.append(error_lower)\n",
    "        nonsense_upper_err.append(error_upper)\n",
    "\n",
    "    nonsense_err = [tuple(nonsense_lower_err), tuple(nonsense_upper_err)]        \n",
    "    \n",
    "msize = 12\n",
    "medgewidth = 2\n",
    "linew = 3\n",
    "csize = 4\n",
    "cthick = 3\n",
    "\n",
    "markers1, caps1, bars1 = ax1.errorbar(x_list, y_list, yerr = missense_err, fmt = 'o', ecolor = missense, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = missense, markerfacecolor = missense, zorder = 1)    \n",
    "\n",
    "markers2, caps2, bars2 = ax1.errorbar(x_list_nonsense, y_list_nonsense, yerr = nonsense_err, fmt = 'o', ecolor = nonsense, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = nonsense, markerfacecolor = nonsense, zorder = 2)    \n",
    "\n",
    "markers3, caps3, bars3 = ax1.errorbar(x_list_top20, y_list_top20, yerr = top20_err, fmt = 'o', ecolor = top20, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = top20, markerfacecolor = top20, zorder = 3)    \n",
    "\n",
    "# loop through bars and caps and set the alpha value\n",
    "a = 0.7\n",
    "[bar.set_alpha(a) for bar in bars1]\n",
    "[cap.set_alpha(a) for cap in caps1]\n",
    "[bar.set_alpha(a) for bar in bars2]\n",
    "[cap.set_alpha(a) for cap in caps2]\n",
    "[bar.set_alpha(a) for bar in bars3]\n",
    "[cap.set_alpha(a) for cap in caps3]\n",
    "\n",
    "#Only show the required axis lines\n",
    "ax1.spines['top'].set_visible(False)\n",
    "ax1.spines['right'].set_visible(False)\n",
    "\n",
    "for axis in ['bottom','left']:\n",
    "    ax1.spines[axis].set_linewidth(2)\n",
    "\n",
    "for axis in ['bottom','left']:\n",
    "    ax1.spines[axis].set_color(grey3)\n",
    "    \n",
    "ax1.set_ylabel('fitness effect (s)', fontsize = axisfont, labelpad = 10)\n",
    "ax1.set_title('Fitness effects across DNMT3A', fontsize = 18)\n",
    "\n",
    "# ax1.set_yscale('log')\n",
    "\n",
    "# STACKED BAR CHART\n",
    "# define the exons + create data\n",
    "exons = [{'2':(24-1)+1, '3': (59-25)+1, '4': (150-60)+1, '5':(164-151)+1, '6':(213-165)+1, '7': (285-214)+1, '8': (338-286)+1,\\\n",
    "          '9': (374-339)+1, '10': (427-375)+1, '11': (477-428)+1, '12': (492-478)+1, '13': (518-493)+1, '14':(556-519)+1, \\\n",
    "          '15': (617-557)+1, '16': (646-618)+1, '17': (694-647)+1, '18': (725-695)+1, '19': (774-726)+1, '20': (803-775)+1, \\\n",
    "          '21': (826-804)+1, '22': (866-827)+1, '23': (912-867)+1}]\n",
    "exon_orders = [['2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23']]\n",
    "colors2 = [c4, c0, c3, c1, c5, c6,c7, c8, c9, c10, c2,c4, c0, c3, c1, c5, c6,c7, c8, c9, c10, c2]\n",
    "names = exons[0].keys()\n",
    "values = np.array([[data[name] for name in order] for data,order in zip(exons, exon_orders)])\n",
    "lefts = np.insert(np.cumsum(values, axis=1),0,0, axis=1)[:, :-1]\n",
    "orders = np.array(exon_orders)\n",
    "\n",
    "#plot the bar chart\n",
    "for name, color in zip(names, colors2):\n",
    "    idx = np.where(orders == name)\n",
    "    value = values[idx]\n",
    "    left = lefts[idx]\n",
    "    ax2.bar(x=left, height=-10, width=value, bottom=-5, color=color, orientation = 'horizontal', label = name, edgecolor='k', linewidth=1, tick_label = name)\n",
    "    ax2.text(left + (value/2.), (-0.26), name, ha='center', va='center', fontsize = axislabelfont)\n",
    "    \n",
    "\n",
    "# STACKED BAR CHART of domains    \n",
    "domains = [{' ': (291-1)+1, 'PWWP' :(350-292)+1, '  ':(481-351)+1, 'ADD': (614-482)+1, '   ': (633-615)+1, 'MTase': (912-634)+1}]\n",
    "domain_orders = [[' ','PWWP','  ','ADD','   ','MTase']]\n",
    "# blank_col = c13\n",
    "blank_col = grey1\n",
    "colors3 = [blank_col, PWWP_col, blank_col, ADD_col, blank_col, MTase_col]\n",
    "names_domains = domains[0].keys()\n",
    "values = np.array([[data[name] for name in order] for data,order in zip(domains, domain_orders)])\n",
    "lefts = np.insert(np.cumsum(values, axis=1),0,0, axis=1)[:, :-1]\n",
    "orders = np.array(domain_orders)\n",
    "\n",
    "#plot the bar chart\n",
    "for name, color in zip(names_domains, colors3):\n",
    "    idx = np.where(orders == name)\n",
    "    value = values[idx]\n",
    "    left = lefts[idx]\n",
    "    ax3.bar(x=left, height=-10, width=value, bottom=-5, color=color, orientation = 'horizontal', alpha = 0.9, label = name, edgecolor='k', linewidth=1)\n",
    "    ax3.text(left + (value/2.), (-0.25), name, ha='center', va='center', fontsize = domainfont, color = 'white', fontweight = 530)\n",
    "\n",
    "    \n",
    "\n",
    "# CONFIGURING THE GRAPH APPEARANCE\n",
    "#Set the x and y axis limits\n",
    "ax1.set_xlim(0, 912)\n",
    "ax1.set_ylim(-0.01, 30)\n",
    "ax2.set_xlim(0, 912)\n",
    "ax2.set_ylim(-0.5, 0)\n",
    "ax3.set_xlim(0, 912)\n",
    "ax3.set_ylim(-0.5, 0)\n",
    "\n",
    "#x-axis ticks\n",
    "x1_major_ticks = []\n",
    "x1_major_tick_labels = []\n",
    "x2_major_ticks = [0, 100, 200, 300, 400, 500, 600, 700, 800, 912]\n",
    "x2_major_tick_labels = [\"0\", \"100\", \"200\", \"300\", \"400\", \"500\", \"600\", \"700\", \"800\", \"912\"]\n",
    "ax3.set_xticks(x2_major_ticks)\n",
    "ax3.set_xticklabels(x2_major_tick_labels, fontsize = axisfont)\n",
    "ax1.set_yticklabels([])\n",
    "ax1.set_yticks([])\n",
    "ax1.set_xticklabels([])\n",
    "ax1.set_xticks([])\n",
    "ax2.set_yticklabels([])\n",
    "ax2.set_yticks([])\n",
    "ax2.set_xticklabels([])\n",
    "ax2.set_xticks([])\n",
    "ax3.set_yticklabels([])\n",
    "ax3.set_yticks([])\n",
    "\n",
    "#y-axis ticks\n",
    "y1_major_ticks = [0, 5, 10, 15, 20, 25, 30, 35]\n",
    "y1_major_tick_labels = [\"0\", \"5\", \"10\", \"15\", \"20\", \"25\", \"30\", \"35\"]\n",
    "ax1.set_yticks(y1_major_ticks)\n",
    "ax1.set_yticklabels(y1_major_tick_labels, fontsize = axisfont)\n",
    "\n",
    "#set panel legends\n",
    "legend_marker_size = 2\n",
    "legend_elements = [Line2D([0], [0], marker = '_', color = missense, alpha = 1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='missense'),\n",
    "                   Line2D([0], [0], marker = '_', color=nonsense, alpha=1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='nonsense'),\n",
    "                  Line2D([0], [0], marker = '_', color=top20, alpha=1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='top 20 observed variant')]\n",
    "    \n",
    "ax1.legend(ncol=1, handles=legend_elements, loc='upper left', frameon=0, fontsize = 20)\n",
    "\n",
    "ax1.grid(which='major', linestyle=':', linewidth='0.7', color=grey3, zorder = 0)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABacAAAHvCAYAAABAPWMbAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAIABJREFUeJzs3Xd4VFX+x/H3SUJCCJDQAlKjKGAFQYWlLEgVkCKiKKyAWH6KgqyKq64aQF1xrSAEXBRFBRtKEaUpIEVAeg2dUEMghNBCIOX8/rgzszOZSQjFENbP63nmmZlzzzn33HvmzgPfOfleY61FRERERERERERERKQgBV3qAYiIiIiIiIiIiIjIn4+C0yIiIiIiIiIiIiJS4BScFhEREREREREREZECp+C0iIiIiIiIiIiIiBQ4BadFREREREREREREpMApOC0iIiIiIiIiIiIiBU7BaREREZHLiDEmxBhjjTEfXeqxFEbGmH7GmM3GmDOu81TZVd7CGLPUGHPCVf63Sz1WEREREZE/OwWnRURERAoBY0wzV9A0t0eDPNoGGWMGGWM6FuSYCxtjTCtgOLAeeBR4AEgxxpQFvgeKAk+7yhf9QWOo65qLqn9E/39Wxpirc1wPWcaYVNcPEV8ZY+42xoQEaNfSq03vANvdP/ZMzlG+0FV+xhhTLpcxjfTqu7FXX/l9VDbGVDDGfGKMWWeMOWKMOWWM2WqMGWOMueos56S/q58jxpjwczylIiIiIoWC3z/gREREROSS+hL4KUD5NgBrbaYrEJXptS0IiAU+Bqb+4SMsvFq5nh+01h5zFxpjmgElgZettX/0+amLMxc/A7v/4H39Gc0EvnC9Lg5cDbQHugHLjTF3WWv35tJ2iDHmK2ttej73lYFzbf0NeM97gzGmKHA/kI7zowdAFs4PH96aAQ8Bo4DfcmxLAa4ErgFm4HxeTgE1gT7AvcaY26y1m3MZXx9gO1Ad6AKMz+dxiYiIiBQaCk6LiIiIFC4rrbVf5FXhHIJrfzYVgCzvwLRXOTjBQMmDMcYAEdbaE5d6LLnYlPP6MMYMBJ4B3gKmGWPqWWuzcrRbDtwC9HPVy480nIDyg+QITgN3AaWACUB3AGut5b+Bc/fYiuIEp3/L5breADTOWWiM+d617yeA/gG21wNqu/b9PE6gWsFpERERueworYeIiIjIZSRnzmljzNU4KzwBHvJKGZCZs74r9cACY0yaMSbZGPMfY0xEgH1UMsaMNsbscaU12Od6XzZHvTLGmGHGmB3GmHRjzGFjzHJjzNM56vU2xixzpWE4aYzZboz5whhTJp/H3MYYM9sYc9S1nzXGmEe9tl9tjLE4q1aDvc7Bz8aYvTgrygEWeJ8bV9uixpiXjDEbXX0fMcZMNcbUDjAOY4z5P2PM78bJXX3cGLPWGBPr2v4aMCbHvrznKtwYM8Q4qSjSXPtaZ4wZmo9zEOIa5wJjTJJrXna5UkuUzqXNPcaYX13nLc0Ys8k1XyGu7e6UFw8YJ1d3PHAaGODVRzPXeTzm6mOFCZwe40ZjzHeuz8ppY0yiMWaOMaatV53zPv68WMfbwNc4Adt7AlSbAKwBXjDGlDqH7j8BbnQFg709CKwA1p3HkPNjl+s5t7E+BBwDJgPjgNuNMVf+QWMRERER+cNo5bSIiIhI4VIsZxAYOG2tPZ5L/QNAL5wA1Tz+G4jNzlGvHs5qz49xVnc2Bx7BSQ/S113JFeD6DQh21d2Bk3bgcZwA2K1eK5O/B/4CjAbWAhHAtTipDN519fcgMBb4FXgFJ21BVaAdUBY4nNfJMMY8Dox0jelVnNWsbYAPjTFXWmtfcJ2DB1xjbOA6HwCJOOk82uME814FtrjPjTEmFJgF1Ac+w8lXXcp1Xn4zxjS21q5y1TU4KVe6AYuB14FU1/HeDQwGvsVZpe29L3ClZHGdp57Ap64+irjObfO8zoFLUZzVwd/hBCRPArfh5NZu5JoX948UGGPeBJ7DWZn7DpCEk/7hbuCfgPfK6Gdcx/2x61zucvXRGZjoOo9vufZ5P/CJ69y7g/LlgDk4aS1G46SnKAvc6hrj9Itw/PnxEc78tAe+yrHN4qwwng68gHNu8mMKzme0D04wGuPkE2+Bswq75AWPGs9nsSQQipOqZLBrk1+KH/PflCLfWGtPGWO+AN7ECZi/cjHGIyIiIlJgrLV66KGHHnrooYceelziB05A1+by+MqrXoir7KO8ygJsywJuybFtJs5K2XCvsh9xApQVc9St7+rjJdf70q5+h5/luKbipNMIPo9zUtk1vs8CbBuJE1iv5lX2BZAZoO7DrrE2zlE+ECdQ3TJHeRSwF/jZq6y7q49PgaAc9YPOti/XtmPA1PP8fAR5z5NX+f+59tfFq6yhq2w2EJajvvF63dJVLxkom6NeEdc5SAEqeJWHAUtcn4WrXGVdco4hl2O4kOO/2rWP9/OoE+2qszTAMQ5wvf8F5weSyjmuj8k5+loIpLpeD3OdhzDXe/ePLFE4Ae+A853j8/C3sxxfZ3yv+UTgqVzq9si5T+AHnB8FgvLajx566KGHHnrooUdheyith4iIiEjh8h+cG/t5P167CP0utNYuz1E2B2elZjUAV3qItjgrc88YY8q6Hzg3XtsJtHa1TcNJJ9LAGFMtj/0eBUoA7Vyrj8/FPa7xjfUei2s8P+Cs7m5xjn16+xvOyuLVOfoOwQliNjXGhLnqugOCz1prfVal53yfh6M4KSKuP9eBWmuzrbWnAIwxwcaYKNdY57iq1Peq3sP1/Ly19nSOfmyA7j+11ibnKLsVqITzg8cBr/angbdxguUdvY4LnDkukcdhnPfx55N7RX9eq5n/gRNgH3IO/Y7FWVne2fUZ7gVMstamntcoA1uEc613xFnZfQAo7U7BksNDwHZr7UKvsk+BKvz3pqAiIiIilwWl9RAREREpXLZaa3/+A/rdEaDMnVLDnfu5FmBwVuP+Xy79ZIFzU0bj5JZ+F0gwxmzACZROstbO9ar/Gs4N36YCycaYX3FSK3xtz37TvWtdz3PzqFP+LH3kpRZO8PtQHnVK46xivQbYGyCIey6ewkm/st4Ysx3nuKYC03IJGvswxtyHk4KjNs7KZm/euYmvwZmntfkc15YAZe78xRsCbFvver7K9TwHJ6fzQ0BPY8wynHQpX1trN3m1u6Djzwd3UDrnDTE9rLXLjTHfAr2MMe8Am8/WqbV2jTFmFU7ajIM4x/3YRRiv9z4OAe7r/gdXqo61OOlRnnDXM8ZchfNXFqOMk2/eLR4n7cpDOH8RISIiInJZUHBaRERE5M8hK49tJsfzOJwUGYGkuV9Ya0cYYybh5Pj9K3Av0M8YM95a+zdXnc3GmFo46RVaAE1xcgMPMsb81Vq7Mx/j6oETFAxkWy7l+REErMZJ75GbFK+xXFAA1Vr7vSs43w7nPLTCSfswzxjT2nrljM7JGHMvTs7rJUB/nJQb6TjB9R/xvdH5ua5QTwtQlu8+XIHlHq4bG7YFmuDkdH7ZGNPPWjvKVe+8jz+fbnI9ny3g/E+c/OtDXc/5MRYnvUcRnPQZv5zPAPPLWrvXGDMHeMQYM8Dr3PTBmZu+eOWK99LJGFPGWptnLncRERGRwkLBaREREZHL38VYdQpOoNcCRfK7ettauw8nFcl/XCkIxuMEKt+xrpsJulJB/Oh6YIzpiHOjuQE4q2lzs9X1fOgPWk2+FSgH/JKPlbubcdJWlD3L6uk8+3EFDT8HPneliHgLZzX0ncCkPJo+gBNEvt1am+4uNMbckMtYWwI3AivzGk8etrueA6XguM717LMa31q7DlgH/NuVIuZ3nADwKK8653v8+fGw6/nHvCpZa7cZY/6DsyK5ST77Ho+TzqQ5MOQcUrlciHCcYHhx4IgxJggnpcgKnPOaUyXgfZx0NcMKYHwiIiIiF0w5p0VEREQuc9baLJxVtKUvsJ8knHQM9xhjbs253TjKuV4XM8aE52ifiROcxD0WV17knFZ618nD18AZYIgxpmiA8UQZY0LP0kdePsMJ6AUMkBtjvFOGjMdZsfrvnLmzc7x3pyopnaNOiDEm0rvMFRBfHah+AFk4N2/0/Pvdtd+XAtSd4Hp+I+f5OYe838uAfcBDxphor/ahwLOusUx1lZXO2a+1NgVIAIobY0IvwvHnyRjzDNANWAVMzEeTIThz9WZ++rfWHsFJ5TEYGHOew/ST4zPmXX4jTvqOLa59A7TBuUnoZ9baiQEew4A9OKurRURERC4LWjktIiIi8r9hCdDGGPMcToAqy1r7zXn083/AQmChMeYznGBfCE4O4s7Axzh5pK8Dfnal9dgAHHGVPY6z6naRq785xphDwALXuErj5O7NJvfUIQBYa3cZY54ERgMbXXl4d+Osdr4J6ATUwElxcT7exVlh/J4xpiUwDzgOVMVJQXKc/95g7iugi2vsNYwxP+Dc4K+Gq25tV73fcVZPv+QK5J/kvzeT3G2MmYoTkHXnLn4cJ3VInqt9cQKunXDO5+c4N/W7C/AL2ltrf3PlU34GcOdYPoAzh/cYY24+W75va22mMaYf8C2wzBgzxnUs9wG34awedq+c7gM86fosbAMygdtd52WCtfaM60eKCzl+t1rGmL+5XkcA1YEOOPnDfwfucv1Ykydr7UHXOYrN536x1n6a37rn4GVjTFPgJ5xgfjDOive/uV4/4VX3IZzPVl4rzL8HnjLG3BLgBqgiIiIihY6C0yIiIiL/Gx4DRuCspC2Bs9L2nIPTroBwXeB5oCPQEyedxB5gMv9dlboLJzd1M5wgaRhOkPhD4E2v1BMjgXtc4yuNcxPGlUBfa+28fIxnjDEmHme17uNAJJAMbMLJHZzXzQzP1vcZY8wdwJM4wcDBrk37cYL947zqWmNMN5xgYR+coGYWTmqLb73q7TTGPIyTc3kUTlqGj13thuMEbFvhBFYTcQKNb1hrD5xlrF8YY4rjrPJ+ByegOxV4mQD5uK21z7pu4vcE8A+cVd97gGk4q+zzc34mGWNa4Xym/oHzf4d4oI+19hOvqnNwfizoAFyBE5zeiRMcH+Gqc/xCjt9LG9fD4qx8PoDzeXoZ52acZw1Me3kb53N5ITfVvFA/ABVxVn1H4wSk9+L8GPK2tTYePH+B0AFYZq3dk0d/3+F8RvoACk6LiIhIoWcuzo2xLz5jTBmc/+i0x1k9UAnnzzrXAZ8An3jnejPGXIOzmqUNzh3Ky+Os4FkCvJ/jrvE599UL5x/u1+H8J2MVzj8Gp138IxMRERERERERERGRwhycfgxntUkiMBfnTzjL4wSgI3FWBdzjvnmNMeYrnBUHG3H+FDUFqImz4icYeMpaOzzAft7GWdWxF2clUCjOnyuWBvpZa0fkbCMiIiIiIiIiIiIiF6YwB6eb4/y53485VkhXwMknVwXoaq39zlXeG1jjviu8V/2mwGycP/2LsdYmem1riJMPcTtwq/tmI8aYGJy7YEcAtay1CX/IQYqIiIiIiIiIiIj8SQWdvcqlYa2dY639wTsw7So/gHNTHHByHLrLP80ZmHaV/4pzc5tQoGGOzY+5nl/3ugs2rmD0SJzciQ9e0IGIiIiIiIiIiIiIiJ9CG5w+iwzXc+YF1m/uep4RoM30HHVERERERERERERE5CK57ILTxpgQnLvGQ+Cgcs761XDuCp4GzPcqj8C5yeIJ71QfXra6nmucbR+jRo2yo0aNKpz5UUREREREREREREQKocsuOA0MBW4AfrLWzsyrojEmDBiPk55jkHfqDpybKgIczaW5uzwqj/4fNcYs9zQ4etSuWrXKZmRk2OXLl1vALl++3J4+fdquWbPGHjlyxG7bts3u3bvXJiYm2s2bN9tjx47ZVatW2czMTL8269atsykpKXb79u12z5499sCBA3bTpk32+PHjdsWKFTYrK8unzalTp+z69evt4cOH7Y4dO+zu3bttUlKSjY+PtydOnLDLly+32dnZPm3S0tLshg0b7KFDh+zOnTvtrl277MGDB+3GjRvtyZMnPXWXLVvmaXPy5Em7ceNGe/DgQbtr1y67c+dOe+jQIbthwwablpbm0797fydOnLDx8fE2KSnJ7t692+7YscMePnzYrl+/3p46dcqnTVZWll2xYoU9fvy43bRpkz1w4IDds2eP3b59u01JSbHr1q2zp0+f9mmTmZlpV61aZY8dO2Y3b95sExMT7d69e+22bdvskSNH7Jo1a/zaZGRk2FWrVtmjR4/aLVu22P3799t9+/bZrVu32tTUVLt69WqbkZHhc+xnzpzxzOXWrVvtvn377P79++2WLVus9/y72yxbtsyePn3arl271jOXOed/5cqVfnOZnp5u161b5zOXOec/51y65z85Odkzl0lJSXbjxo2e+bfW5jr/CQkJNiEhwR46dMhv/r2fT5w44Tf/ycnJdv369TYtLc3nfGVnZ9sVK1bkOf/p6ek+n7OsrCy7cuVKe+zYsYDzv3bt2lzn/+jRo3bz5s12//79fvN/5swZv/lfvXq1TU1NtVu2bLH79u2z+/bts1u2bPGZf+823vPvvpb3799vN2/e7Jn/QNey9/x7X8ve8+993tLT0/O8lnPO/7Jly2xaWprf/LuvZff855xL97Wcc/4DXcvuz457/pOSkvzmP+e17J5/72vZe/7XrVvnM//u6989/97Xcs75977G8rqWA82/u417/vO6lvVdru9yfZfru1zf5fou13e5vsv1Xa7vcn2X67tc3+X6Lv8zfpdTQArtDREDMcb0B4YBm4BG1tqUPOoGA18C9wBfA/dbr4M1xlQE9gH7rLWVA7QvApwBTltri+Y1Lveq6ccff/ycj0lERERERERERESkkDEFsZPLZuW0MeYJnMD0RuD2fASmv8AJTH8D/M36R+HdK6MjCexsK6tFRERERERERERE5DxdFsFpY8wAYASwHicwfSCPuiE4K6bvAyYA3a21fjdOtNaexFk5XdwYc0WArq5xPW+5wOGLiIiIiIiIiIiISA6FPjhtjPkH8B6wGicwfTCPuqHARJwV058BD1hrs/Lofo7r+Y4A29rmqCMiIiIiIiIiIiIiF0mhDk4bY17GuQHiCqCFtTY5j7phwCSgE/Ax8KC1Nvssuxjtev6nMaaUV18xwBPAaeCT8x2/iIiIiIiIiIiIiAQWcqkHkBtjTC9gCJAFLAD6G+OXhzvBWvup6/VooB2QjJOu45UA9edZa+e531hrfzPGvAs8Daw1xkwEQoFuQGmgn7U24eIdlYiIiIiIiIiIiIhAIQ5OA1e6noOBAbnU+RX4NEf9ssArefQ7z/uNtfYZY8xa4EngUSAbWAm8Za2dds6jFhEREREREREREZGzKrTBaWvtIGDQOdRvdgH7GgeMO9/2IiIiIiIiIiIiInJuCnXOaRERERERERERERH536TgtIiIiIiIiIiIiIgUOAWnRURERERERERERKTAKTgtIiIiIiIiIiIiIgVOwWkRERERERERERERKXAKTouIiIiIiIiIiIhIgVNwWkREREREREREREQKnILTIiIiIiIiIiIiIlLgFJwWERERERERERERkQKn4LRc9gYNGoQxhnnz5l3qoYiIiIiIiIiIiEg+KTgtIiIiIiIiIiIiIgVOwWm57D355JPEx8dz2223XeqhiIiIiIiIiIiISD6FXOoBiFyosmXLUrZs2Us9DBERERERERERETkHWjktl0RCQgLGGHr37s327dvp2rUrZcqUoUSJErRu3Zr169cDcOjQIR599FGuuOIKihYtyq233srcuXN9+sot5/SCBQvo0KEDlStXJiwsjAoVKtCgQQMGDx7sUy8pKYlnn32WmjVrEhERQVRUFDVr1qR3797s2LHDb+wzZ86kXbt2lC1blrCwMKpXr87AgQNJTU31qxsTE0NMTAxpaWkMHDiQqlWrEhYWxtVXX82bb76JtdavzdSpU2nRogVXXHEFYWFhVKxYkaZNmxIXF+dXNyUlhRdeeIFrr72W8PBwIiMjadGiBbNmzTrrHIiIiIiIiIiIiFxKWjktl1RCQgL169fn2muvpXfv3iQkJDBp0iSaNWvG4sWLueOOOyhZsiTdunUjJSWFr776irZt27JlyxaqVq2aa78zZsygffv2lCxZko4dO1KpUiVSUlKIj48nLi6O2NhYANLS0mjUqBHbt2+nVatWdOjQAWstu3btYsqUKXTt2pWrrrrK0++QIUOIjY2ldOnS3HnnnURHR7N27VrefvttfvrpJxYvXkzJkiV9xpKRkUHr1q3Zv38/bdu2JSQkhMmTJ/P888+Tnp7uGQvAf/7zH/7v//6PChUq0KFDB8qWLcvBgwdZu3Ytn3zyCX379vXU3bVrF82aNSMhIYEmTZpwxx13cPLkSaZNm8Ydd9zBhx9+yCOPPHKxpkpEREREREREROSiUnBaLqlff/2V1157jX/+85+esldffZVXXnmF+vXrc++99xIXF0dQkLPIv1WrVvTs2ZP33nuP9957L9d+x4wZQ3Z2NvPmzaN27do+25KTkz2vf/nlF7Zv386AAQP8+jtz5gynT5/2vJ87dy6xsbH85S9/4aeffiIqKsqz7dNPP+XBBx8kNjbWr5/9+/dTu3ZtZs+eTXh4OACxsbHUqFGD9957jxdffJEiRYoA8OGHHxIaGsqaNWuIjo7OddwAvXr1YteuXXz55Zfcd999nvLU1FSaNWtG//796dixI+XLl8/1PImIiIiIiIiIiFwqCk4XIt2XZFzqIeTbhAZFLko/MTExPP/88z5lvXr14pVXXuH06dO89dZbnsA0QPfu3enTpw+rV6/OV//uYLC3QPmpA9ULDQ0lNDTU83748OGAE/j2DkwD9O7dm2HDhjF+/PiAQfPhw4f77CM6OppOnTrx2WefsXnzZm644QbPtpCQEE+wOrdxr1mzhl9//ZWuXbv6BKYBoqKiGDx4MJ07d+a7777zWW0tIiIiIiIiIiJSWCg4LZdUnTp1CA4O9imrWLEiADVq1KBEiRI+24KDgylfvjx79+7Ns98ePXrw/fffU79+fbp168btt99Oo0aNqFy5sk+9pk2bUqlSJYYOHcrKlStp164djRo1CjiuxYsXU6RIEb799lu+/fZbv32eOXOGQ4cOcfjwYcqUKeMpj4yM5Oqrr/arX6VKFQCOHDniM+5nnnmG66+/nm7dutG0aVMaNWpEuXLl/MYCcPToUQYNGuTX96FDhwCIj4/P6zSJiIiIiIiIiIhcMgpOyyUVGRnpVxYSEpLrNvf2jIy8V5l36dKFadOm8c477zB27Fg+/PBDAOrVq8cbb7xBq1atAChZsiRLliwhNjaWqVOnMnPmTMBZpdy3b19eeuklzyrmw4cPk5mZ6XdDxZxOnDjhE5zOuco653FmZWV5yp5++mnKli1LXFwcw4cP5/3338cYQ9OmTXnrrbe45ZZbPGMBmD17NrNnz85zLCIiIiIiIiIiIoWRgtOFyMVKlSGO9u3b0759e06ePMnSpUuZNm0ao0aN4s4772TVqlVcd911AFSuXJmPP/4Yay0bN25kzpw5jBw5kiFDhpCdnc2rr74KOMHy7OxsUlJS/tBx9+zZk549e5Kamspvv/3GpEmTGDt2LG3atCE+Pp7o6GhP4H7YsGH079//Dx2PiIiIiIiIiIjIHyHo7FVELm8RERE0b96cd999lxdffJEzZ84wffp0v3rGGK6//nr69evnWY08efJkz/YGDRpw5MgRNmzYUCDjjoqKol27dowZM4bevXuTkpLCggULPGMBPO9FREREREREREQuNwpOy/+kX375hVOnTvmVJyUlAVCsWDEA1q9fT0JCwlnrAfz9738H4JFHHmH//v1+bU6ePMmSJUsuaNwzZswgMzPTr/zgwYM+47nlllto0qQJ33//PWPHjg3Y17p16zztREREREREREREChul9ZD/Sc888wwJCQk0a9aMmJgYQkNDWbFiBXPmzKFatWrcd999APz88888/fTTNGzYkFq1ahEdHc3evXuZMmUKQUFBDBw40NNnixYtGDp0KC+88ALXXHMN7dq148orr+TEiRPs2rWLX3/9lcaNGzNjxozzHvd9991H0aJFady4MTExMVhrWbBgAcuWLaNevXq0bNnSU3fChAk0b96chx56iOHDh1O/fn2ioqLYu3cva9euZf369SxevJjo6OjzP5EiIiIiIiIiIiJ/EAWn5X/Siy++yKRJk1i+fDk///wzQUFBVK1alRdffJEBAwZQqlQpANq0acOAAQOYP38+U6ZM4dixY1xxxRW0atXKE7T29o9//INGjRoxfPhwFi5cyJQpU4iMjKRSpUo8+uijdO/e/YLGPXToUGbOnMnKlSv56aefKFq0KNWqVePNN9/k8ccf99ycEZxc2StWrOCDDz7gu+++Y/z48WRlZVGhQgWuu+46+vXrx4033nhB4xEREREREREREfmjGGvtpR7DZW/UqFEW4PHHH7/UQxERERERERERERG5UKYgdqKc0yIiIiIiIiIiIiJS4BScFhEREREREREREZECp+C0iIiIiIiIiIiIiBQ4BadFREREREREREREpMApOC0iIiIiIiIiIiIiBU7BaREREREREREREREpcApOi4iIiIiIiIiIiEiBU3BaRERERERERERERApcoQ1OG2PKGGMeNsZMMsZsM8acMsYcNcYsNMY8ZIwJylG/iDHmKWPMJ8aY1caYM8YYa4x5OI999HbVye3x2B9/pCIiIiIiIiIiIiJ/PiGXegB5uAcYBSQCc4HdQHmgC/AR0NYYc4+11rrqRwDvu14nAQeAKvnc1xRgdYDy5ec3dBERERERERERERHJS2EOTm8BOgI/Wmuz3YXGmBeB34G7cQLV37k2pQHtgNXW2kRjzCAgNp/7mmyt/fQijVtEREREREREREREzqLQpvWw1s6x1v7gHZh2lR8ARrveNvMqP2OtnW6tTSzAYYqIiIiIiIiIiIjIeSjMK6fzkuF6zrxI/dUxxgwAigL7gLnW2r0XqW8RERERERERERERyeGyC04bY0KAnq63My5St0/leJ9ljPkIGGCtTb9I+xARERERERERERERl+BBgwZd6jGck8GDB/8b6Az8ZK0T5uPvAAAgAElEQVQdkke9ZjhpP34YNGjQylzqRAIrgL8DzwMjgaXALUAroPqgQYO+C9QWwBjz6ODBgz9s3759RYAaNWoQHx9P2bJlWbVqFRUrVmT58uWUKVOGDRs2EBYWxr59+0hLS+PkyZPs37+f0NBQNm7cSLly5Vi5cqVPm/j4eIoUKcL+/fs5efIkp06dYu/evYSFhbF+/XrKly/v06ZUqVJs3ryZ4OBgDhw4wPHjx0lPT2fPnj2Eh4ezdu1aKlSowIoVKzxtoqKi2LJlC8YYDh48yLFjxzhz5gy7du2iWLFirF27looVK7Js2TIqVarE8uXLiYyMZNu2bVhrSU5O5siRI2RmZpKQkEDx4sVZs2aNp3/3/kqWLMn27dvJzs4mJSWFw4cPk52dzY4dOyhRooRPm/Lly7Nq1SpKlCjBzp07yczM5MiRIyQnJwOwbds2oqKiWL16tadNdHQ0a9asISIigoSEBDIyMjh69CgHDx7EGMOWLVsoVaqUT5ty5cqxdu1aihUrxu7duzl9+jTHjh0jKSmJ4OBgNm3aRJkyZVi5cqXn2MuWLcv69esJCwtj7969pKenc+LECRITEwkJCfHMv7vNsmXLKFu2LBs3bvTMZc7537BhA9HR0T5zWbp0aeLj4wkJCfHMZc75zzmX7vkPCgoiKSmJY8eOcfr0aXbv3u2Z/yuuuCLX+T906BCpqalkZGT4zf/y5cs9zyVLlvSb/6ysLHbs2EHx4sVZvXq153xVqFCBlStX5jn/JUuW9MzLsmXLqFChAqtXryYiIiLg/G/dujXX+S9WrBi7du3izJkzfvNfunRpn+uyXLlyrFu3jqJFi7J7927S09M5fvw4Bw4c8Jl/7zbe8+++lk+cOMH+/fspUqQI8fHxAa9l7/n3vpa953/FihWe81a6dGk2bdqU67W8bt06n/lftmxZwPl3X8vu+c85l+5rGfCZ/0DXsvuz457/7OxsDh8+7DP/Oa9l9/x7X8ve8799+3YiIyN95rJ8+fKe+fe+lg8dOuQz/6tWrfJcY9HR0bley0FBQX7z727jnv89e/bkei3ru1zf5fou13e5vsv1Xa7vcn2X67tc3+X6Ltd3ub7L9V3+Z/wuv+KKKwbnFhO9mIy1tiD2c1EYY/oDw4BNQCNrbUoedQfh3BDxEWvtR+e4nyrAGqAUUMdauyav+qNGjbIAjz/++LnsRkRERERERERERKQwMgWxk0J7Q8ScjDFP4ASmNwK35xWYvlDW2j3AT663f/2j9iMiIiIiIiIiIiLyZ3VZBKddNyscAazHCUwfKIDdHnI9RxTAvv50EhISMMbQu3dvEhISuO+++yhbtixFixbllltuYdq0aX5tTp8+zdChQ7npppsoVqwYJUuWpEmTJnzzzTcXpf8zZ84wfPhw6tatS6lSpShWrBgxMTF06tSJn3/+2a/+pk2b6N27N1WqVCEsLIzy5cvTvXt3Nm/e7Fe3d+/eGGNISEjgww8/5MYbb6Ro0aKUL1+eRx99lKNHj/q1Wbt2Lffffz8xMTGEhYVRrlw56taty4ABA8jIyPCpm5mZSVxcHA0aNKBkyZIUK1aMm2++mREjRpCdnZ3nXIiIiIiIiIiIiFwKhf6GiMaYfwBDgdVAK2ttcgHtur7reUcB7e9PadeuXdx2221cddVVPPDAA6SkpPD11197AsK333474ASO27Rpw6+//kqtWrV44oknSEtLY+LEiXTr1o3Vq1fzr3/967z7ByeA/OWXX3LDDTfQs2dPwsPD2b9/PwsXLmTGjBm0bNnSU3fGjBl06dKFjIwMOnTowNVXX83evXv5/vvv+fHHH5k7dy5169b1G89zzz3HzJkz6dChA61bt2bu3LmMGTOGbdu2MWfOHE+9tWvXUr9+fYwxdOzYkSuvvJJjx46xbds24uLieO211yhSpAiAZwwzZ86kZs2adO/enaJFizJ37lz69evH0qVL+fzzzy/anImIiIiIiIiIiFwU1tpC+wBeBiywHCh9jm0Hudo+nEedJgHKDPCCq+0hoOTZ9hUXF2fj4uKs5N/OnTut6xzbQYMG+WybMWOGBWzbtm09Zf/61788ZRkZGZ7ypKQkW61aNQvYRYsWnXf/qamp1hhj69WrZzMzM/3Gm5yc7HmdkpJio6KibJkyZeyGDRt86q1fv95GRETYm2++2ae8V69eFrBVqlSxu3bt8pRnZGTYJk2aWMAuXbrUU/70009bwE6ePNlvLCkpKTYrK8vzPjY21gL2ySef9Bl7Zmam7dOnT679iIiIiIiIiIiI5KJA4r+FduW0MaYXMATIAhYA/Y3xy8OdYK391KvN80At19s6rucHjTGNXa8XWt+bI843xmwBlgH7gEigEXADkAb0sNYeu2gHdRam8/CC2tUFs5P7X5R+qlWrxksvveRT1qZNG6pWrcrvv//uKRs7dizGGN59911CQv77sY2Ojubll1/m4Ycf5qOPPqJhw4bn1b8xBmstYWFhBAX5Z7spU6aM5/Vnn31GamoqI0aM4LrrrvOpd/311/PII4/w/vvvs3HjRr/tr7zyClWrVvW8DwkJ4cEHH2TBggX8/vvv3HbbbT71w8PD/cZSqlQpz+vs7GxGjBhBhQoVeO+99wgODvZsCw4O5p133uGTTz5h/PjxdOrUya8vERERERERERGRS6XQBqeBK13PwcCAXOr8Cnzq9f4OoGmOOg1dDzfv4PTbwG1Ac6A0kA3sBkYC71prldLjD1anTh2fgKpblSpVWLx4MQDHjx9n27ZtVKpUiVq1avnVbd68OQCrVq06r/4BSpYsSYcOHfjhhx+oU6cOd999N02aNKF+/foUK1bMp6273Zo1axg0aJBf31u2bAEgPj7eLzh9yy23BBwLwJEjRzxl3bp1Y9iwYXTu3JmuXbvSsmVLGjVqRPXq1f32dfjwYa655hpee+01v77BCXDHx8cH3CYiIiIiIiIiInKpFNrgtLV2EE5qjnNp0+wc6w88l/py8UVFRQUsDwkJ8dzIz32zwCuuuCJgXXd5amrqefXv9vXXX/Pmm28yYcIEYmNjAShatChdu3bl7bffpnz58gAcPnwYgDFjxuR5bCdOnMjXeNwrwbOysjxlt912GwsWLOD1119n4sSJnpzRNWvWJDY2lvvvv99nLFu3bmXw4MHnNBYREREREREREZFLqdAGp/+MLlaqjP81kZGRABw4cCDg9sTERJ965ys8PJxBgwYxaNAg9uzZw/z58/n000/54osvSEhIYMGCBT77WbNmDTfddNMF7TMvf/nLX5g2bRqnT59mxYoVzJgxgw8++IDu3btTrlw5WrZs6RnLXXfdxffff/+HjUVERERERERERORi80+uK1LIlChRgurVq7Nv3z62bt3qt33u3LkA1K1b96Lts0qVKvTo0YOZM2dyzTXXsHDhQs8q5QYNGgB4gtV/tLCwMBo2bMiQIUMYPtzJSz5lyhQAatWqRVRUFEuWLCEjI6NAxiMiIiIiIiIiInIxKDgtl4U+ffpgrWXgwIE+6S+Sk5N59dVXPXXO16FDh1i6dKlf+cmTJzl+/DghISGEhoYC8OCDDxIVFcXgwYN9bqrolp2dzbx58857LOAEvt3pTLwlJSUBePJgh4SE0K9fPxITE+nfvz+nTp3ya5OYmMjGjRsvaDwiIiIiIiIiIiIXm9J6yGXh2WefZfr06UyZMoXatWvTrl070tLS+Pbbbzl48CDPPfccjRs3Pu/+9+3bR4MGDbj22mupW7cuVapU4dixY0ybNo0DBw7Qv39/SpQoAUCZMmWYOHEid911Fw0aNKBFixZcf/31BAUFsXv3bhYvXszhw4dJT08/7/G88847zJo1i2bNmnHVVVdRvHhxNmzYwPTp0ylVqhSPPvqop+7LL7/MmjVrGD16ND/88APNmzenUqVKHDx4kK1bt7Jo0SJef/11v5szioiIiIiIiIiIXEoKTstlITQ0lNmzZ/Puu+8yYcIEPvjgA0JCQqhduzbvv/++5waB5ysmJobBgwczb9485s6dS3JyMqVLl6ZmzZoMHTqU++67z6d+ixYtWLt2LW+//TYzZ85kwYIFhIaGUrFiRZo3b87dd999QePp27cvpUqVYunSpSxatIjMzEwqV65M3759eeaZZ6hWrZqnbpEiRZg8eTJffPEFn376KdOmTePEiROUK1eOK6+8kldffZUePXpc0HhEREREREREREQuNmOtvdRjuOyNGjXKAjz++OOXeigiIiIiIiIiIiIiF8oUxE6Uc1pERERERERERERECpyC0yIiIiIiIiIiIiJS4BScFhEREREREREREZECp+C0iIiIiIiIiIiIiBQ4BadFREREREREREREpMApOC0iIiIiIiIiIiIiBU7BaREREREREREREREpcApOi4iIiIiIiIiIiEiBU3BaRERERERERERERAqcgtMiIiIiIiIiIiIiUuAUnBYRERERERERERGRAqfgtIiIiIiIiIiIiIgUOAWnRURERERERERERKTAKTgtcpnr3bs3xhgSEhIu9VAKvXnz5mGMYdCgQZd6KPnWrFkzjDGXehgiIiIiIiIiIhedgtNS4BISEjDG0Lt370s9FD8ZGRlMmjSJhx56iBtuuIGSJUtSrFgxbrzxRl555RWOHz+ea9u9e/fSp08fKlasSFhYGDExMQwYMIAjR44U4BGIXByF+ToVERERERERkf8NIZd6ACKFyfbt2+nSpQsRERHcfvvttG/fnhMnTjBz5kxeffVVvv76axYtWkTZsmX92jVs2JCDBw/SqVMnatWqxe+//86wYcOYMWMGixYtokyZMpfoqORy9tlnn5GWlnaphyEiIiIiIiIictEpOC3ipUSJEowcOZJevXoRERHhKT9z5gxdunThxx9/ZPDgwXzwwQc+7fr27cvBgwcZPnw4/fr185Q//fTTvPfee/zzn/9k9OjRBXYc8r+jatWql3oIIiIiIiIiIiJ/CKX1kAI1aNAgrrzySgDGjRuHMcbz+PTTTz31srOzGT16NLfeeivFixcnIiKCW2+9lVGjRpGdne3XrzGGZs2asX//fh544AGio6MJDw+nXr16TJgwId/jq1SpEn379vUJTAOEhoby4osvAk7eYm87duxg1qxZxMTE8MQTT/hsGzx4MBEREXz++eecPHky3+NYsWIFd999N9HR0YSFhVGtWjX69u1LYmJirm2ys7N59913qVWrFkWLFqVy5cr8/e9/59ixY351165dy/33309MTAxhYWGUK1eOunXrMmDAADIyMnzqZmZmEhcXR4MGDTxpTm6++WZGjBjhNxfeqSC2bNlCt27diI6OJigoiHnz5lGrVi1CQ0NJTk4OeAxDhw7FGMPIkSN9yvfu3cuTTz7JVVddRVhYGGXKlKFjx44sW7YsYD9JSUk89NBDlC9fnvDwcOrUqcO4ceNyPXeBvPHGGxhjGD58eMDt+/fvJzg4mFtvvdWnbMiQITRq1IgKFSoQGhpKxYoV6d69O/Hx8X59nO18QeCc02fOnGHEiBG0a9eOatWqERYWRunSpWnZsiXTp08PON6YmBhiYmJIS0tj4MCBVK1albCwMK6++mrefPNNrLWeuvm9TkVERERERERELoRWTkuBatasGampqQwbNozatWvTuXNnz7Y6dep4Xj/wwANMmDCBKlWq8PDDD2OMYdKkSfTt25eFCxcyfvx4v76PHDlCw4YNiYqK4sEHHyQ1NZVvvvmGHj16sG/fPgYOHHhBYy9SpAgAISG+l82cOXMAaN26NUFBvr/3lChRgkaNGjFr1iyWLFlCixYtzrqfadOmcffdd2OtpWvXrlSrVo0VK1YwatQopkyZwqJFi4iJifFr9/e//5358+dz77330qlTJ2bOnMn777/PggULWLhwIUWLFgWcwHT9+vUxxtCxY0euvPJKjh07xrZt24iLi+O1117zHGtGRgYdOnRg5syZ1KxZk+7du1O0aFHmzp1Lv379WLp0KZ9//rnfWLZv3079+vWpUaMGPXr04NSpU5QsWZJevXrx4osv8uWXX/qsMHf77LPPCA0N5b777vOUrVy5ktatW5OSkkKbNm3o0qULycnJTJ48mcaNGzNp0iTatWvnqX/48GEaNmzIjh07aNy4MY0bNyYxMZHHHnuM1q1bn/X8u/Xs2ZOXXnqJcePG0b9/f7/tX3zxBdnZ2fTq1ctTNn/+fIYOHcrtt9/O3XffTfHixdm6dSsTJ05k6tSpLFq0iNq1a+f7fOUmJSWFp556ioYNG9KqVSvKlStHYmIiP/zwA+3atWPMmDE8/PDDfu0yMjJo3bo1+/fvp23btoSEhDB58mSef/550tPTiY2NBfJ/nYqIiIiIiIiIXBBrrR4X+IiLi7NxcXFW8mfnzp0WsL169Qq4fcKECRawN998sz1+/Lin/MSJE7ZevXoWsOPHj/dpA1jA3nPPPTYrK8tTvmPHDluqVClbpEgRu3379gsa92OPPWYB+/zzz/uUP/vssxawb7/9dsB2TzzxhAXy9Rk5fvy4LVOmjA0KCrLz58/32TZ06FAL2FatWvmU9+rVywK2TJkyNiEhwVOelZVlu3TpYgE7ZMgQT/nTTz9tATt58mS//aekpPicv9jYWAvYJ5980mZmZnrKMzMzbZ8+ffz6cc8tYF944QW//vfu3WuDgoJsvXr1/Lb9/vvvFrBdunTxlGVkZNjq1avbsLAwO2/ePJ/6+/btsxUrVrQVKlSw6enpnvJHHnnEAnbAgAE+9ZctW2ZDQkIsYGNjY/32H0jr1q0tYNetW+e37brrrrNFihSxycnJnrKkpCR77Ngxv7qrV6+2ERER9o477vApP9v5stbapk2bWuer+r/S09Ptnj17/Oqmpqba66+/3pYqVcqmpaX5bKtWrZoFbNu2bX22JSUl2cjISBsZGWnPnDnjN7bcrlMRERERERER+Z9WIHFVrZwuTGrfc6lHkH9rvv3Duh47dizgpHgoXry4pzwiIoI333yTli1b8tFHH9G9e3efdsHBwbz55ps+q5evvPJK+vfvz+DBg/n88889K0PP1dSpU/nwww+pXLkyzz33nM+2o0ePAhAZGRmwrbs8NTX1rPuZMmUKhw8f5v7776dJkyY+25555hlGjx7N7Nmz2b17t18u4qeeeopq1ap53gcFBfHWW28xefJkxo4dy8svv+xTPzw83G//pUqV8rzOzs5mxIgRVKhQgffee4/g4GDPtuDgYN555x0++eQTxo8fT6dOnXz6KV++fMBzXalSJVq0aMHs2bPZsGED119/vWebO+2G90rkH3/8ke3bt/Pss8/StGlTn74qVqzIc889x4ABA/jll19o164dGRkZjB8/nhIlSjBo0CCf+rfccgs9evQ4p/QevXr1YtasWYwbN4633nrLU758+XI2btzIXXfd5XOjy+jo6ID91K5dm+bNmzNr1iwyMjI8K9PdcjtfuQkLC6Ny5cp+5ZGRkfTp04dnnnmGZcuW8de//tWvzvDhw33mPjo6mk6dOvHZZ5+xefNmbrjhhnyPQ0RERERERETkQig4LYXOypUrCQoKolmzZn7bmjZtSnBwMKtWrfLbVrVqVU+eXG/NmjVj8ODBAdvkx2+//Ub37t2JiIjgu+++8wng5od15fLNmTc4kJUrVwLQvHlzv20hISH89a9/JSEhgVWrVvkFp3MGbwGuuuoqqlSpQkJCAqmpqURFRdGtWzeGDRtG586d6dq1Ky1btqRRo0ZUr17dp+2WLVs4fPgw11xzDa+99lrA8YaHhwfMpVy7dm3CwsICtunduzezZ89m3Lhx/Pvf/wacHMpfffUV5cqV80nRsXjxYgB27drlF2wG2Lp1KwDx8fG0a9eOTZs2kZaWRpMmTQL+WNCsWbNzCk7fddddREZG8sUXXzB06FBPgN7dR+/evf3a/Pjjj4wePZrly5eTnJxMZmamz/bk5GSuuOIKn7K8zlduNmzYwFtvvcX8+fNJTEwkPT3dZ/u+ffv82kRGRnL11Vf7lVepUgVwUuOIiIiIiIiIiBQUBael0Dl69CilS5cmNDTUb1tISAhly5bl4MGDftvKly8fsL8KFSp4+j1Xixcvpm3btgQFBTF9+nRuu+02vzruIGhu/btvSJjbympv7j5yBi/d3OWBVmHndfy7du3i6NGjREVFcdttt7FgwQJef/11Jk6c6MkZXbNmTWJjY7n//vsBJ3czOAHgwYMH5zrmEydOBNxnbu666y5KlizJF198wRtvvEFwcDDTpk3j8OHDDBgwwCent3sM336b90p99xjc5+9sn4X8Cg8P595772XMmDHMmjWLtm3bkpGRwZdffkm5cuVo27atT/3hw4fz1FNPUapUKVq1akXVqlUpVqwYxhgmT57MmjVrOH369AWPa8mSJTRv3pzMzExatGhBx44dKVmyJEFBQaxevZopU6YE3E9UVFTA/tznPCsr65zGISIiIiIiIiJyIRScLkz+wFQZl5PIyEhSUlICpj/IzMwkOTk54M3ikpKSAvZ34MABT7/nYsGCBbRv356goCBmzpxJgwYNAtarWbMm4Kw0DsS9urdGjRpn3ad7jO4x55SYmOhTz1tSUpJnLN4CHf9f/vIXpk2bxunTp1mxYgUzZszggw8+oHv37pQrV46WLVt66t911118//33Zx27t7xWibsDvh999BGzZ8/mjjvuCJjSw3vMU6ZMoWPHjmfdr7v+2T4L56JXr16MGTOGcePG0bZtW08g/amnnvL5fGZmZhIbG0uFChVYuXKl3w8M7lXggeRnVb231157jVOnTjF37ly/vzB44403mDJlyjn1JyIiIiIiIiJyKQSdvYrIxeVOjZDbKs2bb76Z7Oxs5s+f77dt/vz5ZGVlUbduXb9tu3fvJiEhwa983rx5nn7za86cObRt25aQkBBmz56da2Aa4Pbbbwdg1qxZZGdn+2w7fvw4ixYtIjw8PM8+3NxjdI/ZW2ZmJgsXLgQIePy//vqrX9mOHTvYs2cPMTExAVfNhoWF0bBhQ4YMGcLw4cMBPIHNWrVqERUVxZIlS8jIyDjr2M+FOx3GuHHjSE5OZvr06dx0003UqVPHp577nC1YsCBf/daqVYtixYqxevXqgCvZA53Xs2nUqBHXXHMNU6ZM4ejRo7kG0pOTk0lNTaVhw4Z+gekTJ054UrZcDNu2baN06dIBU98E+hycj7NdpyIiIiIiIiIiF0rBaSlwpUqVwhjD7t27A27v06cPAC+88AJpaWme8rS0NJ5//nkAHnroIb92WVlZ/OMf//AJEO/cuZPhw4cTEhLC3/72t3yNb9asWdx5550ULVqUX375hVtvvTXP+tWrV6d169YkJCQwcuRIn22xsbGcPHmSnj17EhERcdZ9d+7cmdKlS/Pll1+yZMkSn23vv/8+O3bsoGXLln75pgGGDRvGrl27PO+zs7MZOHAg2dnZPPjgg57yBQsWBAzculcbFytWDHBSPfTr14/ExET69+/PqVOn/NokJiaycePGsx5XTt4B31GjRpGRkREwf3OnTp2oXr06I0eO5KeffgrY1+LFiz2fkyJFitCjRw+OHz/ul6N6+fLljB8//pzHCk4gOj09nbi4OH766Sduuukmvx87oqOjKVasGCtWrPBJdZKRkcFTTz1FcnLyee07kJiYGFJSUli7dq1P+ccff8zMmTMvyj7Odp2KiIiIiIiIiFyoQpvWwxhTBrgLaA/cCFQCzgDrgE+AT6y12V71iwB9gTrAzcB1QBHgEWvtR2fZVy/gCVebLGAV8La1dtpFPiwBihcvTv369VmwYAE9evSgRo0aBAcH07FjR2666Sa6d+/OlClT+Oabb7j++uvp3LmzJ2fvzp07uffee+nRo4dfvzfddBNLly6lXr16tG7dmqNHj/L111+TmprKv//9b78b/gWyefNmOnXqRHp6Ou3atWPKlCkBUyTkDHzGxcXRsGFD+vfvzy+//MK1117L0qVLmTt3LjVq1OD111/P97kZO3Ys99xzD02bNuWee+6hatWqrFixglmzZlGhQgU+/PDDgG0bNWpEnTp16NatG5GRkcycOZM1a9ZQr149nnvuOU+9d955h1mzZtGsWTOuuuoqihcvzoYNG5g+fTqlSpXi0Ucf9dR9+eWXWbNmDaNHj+aHH36gefPmVKpUiYMHD7J161YWLVrE66+/znXXXZev4/PWs2dPXn75ZV599VVCQkLo3r27X50iRYrw/fff06ZNG9q3b0/Dhg2pU6cOxYoVY8+ePSxbtowdO3aQmJjoCar/61//4pdffuH9999n+fLlNG7cmMTERL7++mvatWvH1KlTz2usr7zyCrGxsWRkZPitmgYICgqif//+DB06lBtvvJFOnTpx5swZ5s6dS0pKCrfffjtz5849530HMmDAAGbOnEnjxo259957iYyMZPny5SxcuJCuXbsyceLEC97H2a5TEREREREREZELZq0tlA/gMcAC+4HxwBvAWCDVVT4RMF71o1zlFjgA7Ha9fvgs+3nbVW8P8B4wEjjsKnsyP2ONi4uzcXFxVvJv69at9s4777SlS5e2xhgL2E8++cSzPSsry44cOdLWq1fPhoeH2/DwcFu3bl07YsQIm5WV5dff/7N3p9F1peWB7/+PJZVsyZZsWZNdDEUlFNChGRoyQBGqKIZAVxKK6SZpaKqqGS4OYwYWvUIIRxC6c8PUgXAdyFBFkkoqWVAhDYGQ3FAMCdBIlmV5kMujLM9zWZ7kSc/9cI6UI1mDjyW7XPL/t9ZZW3vv9937Oefx+354vNe7gbztttty9+7d+cY3vjFbWlqytrY2n/vc5+YDDzxwyXE9/PDDWfbvaNLPRAYGBvKee+7J9vb2rKmpySc96Un5nve8Jw8fPlzx7/OjH/0o77rrrmxubs6ampp84hOfmO94xzty9+7dF7W9++67E8itW+ftf3YAACAASURBVLfmJz7xiXza056WtbW1uXz58nzve9+bx44dG9P+m9/8Zt5zzz35jGc8IxsaGrKuri5vueWWfPe73539/f0XXX94eDj//M//PO+4445csmRJ1tTU5PLly/PWW2/Nj33sYzkwMDDadvv27Qnk3XffPe133LFjR86bNy+B/Pmf//kp2+7fvz8/8IEP5E/8xE/kggULsr6+Pn/8x388X/e61+Vf/MVf5Llz58a037t3b957773Z3Nyc8+fPz2c/+9l53333jeb3wx/+8LTxjffSl740gayurs59+/ZN2ObcuXP5yU9+Mp/xjGfk/Pnzs62tLd/0pjdlf3//aJ62b98+2v5Sfq/bbrttwn9zX/3qV/Onf/qnc+HChdnY2Jgvf/nL8zvf+U7ed999F42nzMwnP/nJ+eQnP3nCe3z4wx9OIB9++OExx6cbp5IkSZIkac66KjXgyGKB9poTEXcA9cA/5NgnpNuBHwFPBF6fmV8uHb8BeCnQk5l7I6IAfJgpnpyOiBcC/wZsBX4yM4+Wjt8ErCrd/+mZ2T9VrCtXrkyAFStWXOa31UxFBLfddttlrSksSZIkSZIkaYy4Gje5ZteczsxvZeZXywvTpeP7gD8q7d5edvxsZn4jM/dWcJt3lLYfGylMl67VT/EJ6lrg3gn6SZIkSZIkSZJm4JotTk/jXGl7fobXuaO0/ccJzn1jXBtJkiRJkiRJ0ix53BWnI6IaeHNpd6Ki8qVep57iSxZPTPK09ebS9pbLvYckSZIkSZIkaWLVj3UAl+H3gGcCX8/Mb87gOo2l7bFJzo8cXzyDe+gquVbXTpckSZIkSZI0sapCofBYx3DJIuI9wEeAjcDPFwqF05O17ejouJ3imtRfLRQK3ROcXwT8BnC8UCh8aoLzVcCHgPOFQuF3J4nn7R0dHZ+/8847lwPccsst9PX10dzczOrVq1m+fDldXV0sXbqU9evXU1tby+7duzl16hQnT55kz5493HDDDWzYsIGWlha6u7vH9Onr66OmpoY9e/Zw8uRJTp8+za5du6itrWXdunW0tbWN6bNkyRIeeeQRqqqq2LdvH8ePH2doaIidO3eyYMECent7aW9vZ9WqVaN9Fi9ezKZNm4gIDhw4wODgIGfPnmXHjh3U1dXR29vL8uXL6ezs5MYbb6Srq4vGxka2bNlCZnLo0CGOHj3K+fPn6e/vZ+HChaxZs2b0+iP3a2hoYOvWrQwPD3PkyBEOHz7M8PAw27ZtY9GiRWP6tLW1sXr1ahYtWsT27ds5f/48R48e5dChQwBs2bKFxYsX09PTM9qntbWVNWvWUF9fT39/P+fOnePYsWMcOHCAiGDTpk0sWbJkTJ+WlhZ6e3upq6tjYGCAM2fOMDg4yP79+6mqqmLjxo0sXbqU7u7u0e/e3NzMunXrqK2tZdeuXQwNDXHixAn27t1LdXX1aP5H+nR2dtLc3MyGDRtGczk+/+vXr6e1tXVMLpuamujr66O6uno0l+PzPz6XI/mfN28e+/fvZ3BwkDNnzjAwMDCa/2XLlk2a/4MHD/Loo49y7ty5i/Lf1dU1um1oaLgo/xcuXGDbtm0sXLiQnp6e0d+rvb2d7u7uKfPf0NAwmpfOzk7a29vp6emhvr5+wvxv3rx50vzX1dWxY8cOzp49e1H+m5qaxozLlpYW1q5dy/z58xkYGGBoaIjjx4+zb9++Mfkv71Oe/5GxfOLECfbs2UNNTQ19fX0TjuXy/JeP5fL8r1q1avR3a2pqYuPGjZOO5bVr147Jf2dn54T5HxnLI/kfn8uRsQyMyf9EY3nk385I/oeHhzl8+PCY/I8fyyP5Lx/L5fnfunUrjY2NY3LZ1tY2mv/ysXzw4MEx+V+9evXoGGttbZ10LM+bN++i/I/0Gcn/zp07Jx3LzuXO5c7lzuXO5c7lzuXO5c7lzuXO5c7lzuXO5dfjXL5s2bKOyequsykeL0+cRsQ7gT8ENgAvLb0Ycar2BeDDwNsy808mOF8PnKC4rMeiCc43AweBA5nZNtW9Vq5cmQArVqy4tC8jSZIkSZIkSdeuuBo3eVysOR0R76NYmF4HvGS6wvSlyMyTwG5gYUQsm6DJU0vbTTO9lyRJkiRJkiRprGu+OB0RHwA+DfRQLEwfmMXLf6u0feUE5141ro0kSZIkSZIkaZZc08XpiPgQxRcgrqK4lMehWb7FH5W2H4yIJWX3vQl4J3AGuG+W7ylJkiRJkiRJ173qxzqAyUTE3RRffngB+B7wnoiLljrpz8z7y/r8d+Dppd3nlLb3RsSLSn//a/n605n5/Yj4FPDrQG9EfAm4AfgloAl4d2b2z+b3kiRJkiRJkiRdw8Vp4CmlbRXwvknafAe4v2z/lcBt49q8sPQZMebliJn5GxHRC7wLeDswDHQDH8/Mr11W5JIkSZIkSZKkKV2zxenMLACFCvvcfpn3+iLwxcvpK0mSJEmSJEmq3DW95rQkSZIkSZIkaW6yOC1JkiRJkiRJuuosTkuSJEmSJEmSrjqL05IkSZIkSZKkq87itCRJkiRJkiTpqrM4LUmSJEmSJEm66ixOS5IkSZIkSZKuOovTkiRJkiRJkqSrrvpyOkVEPXAj0AycBg5k5u7ZDEySJEmSJEmSNHddcnE6In4M+G/Ay4HnMu6p64g4DDwMfBl4KDPPz2KckiRJkiRJkqQ5ZNridEQ8H/hd4GUUC9LngLXAPuAIsABYCjwNeAPweuBgRPwB8KnMPHNlQpckSZIkSZIkPV5NWZyOiC8CbwSOAV8AHgR+lJlDk7S/Cfg54G7gY8A7IuK/ZuZ3ZzFmSZIkSZIkSdLj3HQvRHwF8OvA8sz81cz87mSFaYDM7M/Mz2fmC4HnAN3AbbMXriRJkiRJkiRpLphuWY+bM/P05Vw4M3uB10TEgsvpL0mSJEmSJEmau6Z8cvpyC9OzfQ1JkiRJkiRJ0twy3bIelyQiaiLiuRHxtNm4niRJkiRJkiRpbquoOB0R/1dE/G1ENJUd+zFgPdAFbIiIhyJiuuVCJEmSJEmSJEnXsUqfnP5vwNMz80jZsU8CPw48DPQCrwbunZ3wJEmSJEmSJElzUaXF6f8AdI7sREQD8J+Bv83MlwE/BWzE4rQkSZIkSZIkaQqVFqdbgL1l+y8AqoEHATLzHPDPwI/NSnSSJEmSJEmSpDmp0uL0caCxbP82IIF/LTs2BCyaYVySJEmSJEmSpDms0hcXbgZeFRG1FIvSbwB6M/NQWZsnAwdmKT5JkiRJkiRJ0hxU6ZPTXwBuplik7iv9/Wfj2vw0sH7moUmSJEmSJEmS5qqKitOZ+UXg94A6ist7/GHpA0BE3AHcBDw8eyFKkiRJkiRJkuaaSpf1IDN/C/itSU7/K7AEODmToCRJkiRJkiRJc1vFxempZOZZ4OxsXlOSJEmSJEmSNPdMuaxHRCyY6Q1m4xqSJEmSJEmSpLllujWnt0fEeyOittILR8SzI+Lvgd+8vNAkSZIkSZIkSXPVdMXpfwI+BeyNiJUR8ZKpnoSOiJsjYkVE/ADoBp6NL0eUJEmSJEmSJI0z5ZrTmfnmiPgM8D+At5c+FyKiD9gLHAXmA0uBpwHNQAD7gQ8Cn87MM1cufEmSJEmSJEnS49G0L0TMzC7gFRHxVOAtwEuB5wD/cVzTg8BDwJeBL2fmuVmOVZIkSZIkSZI0R0xbnB6RmZuB/w4QEXXAjRSfmD4NHMjMvVckQkmSJEmSJEnSnHPJxelymXkK2Fz6SJIkSZIkSZJUkeleiPiYiYilEfHWiPi7iNgSEacj4lhE/GtEvCUiJow9Il4YEV+PiCMRcSoieiPifRFRNUHb2yMip/j83pX/ppIkSZIkSZJ0/bmsJ6evkjcAKym+ePFhYABoA14L/Anwqoh4Q2bmSIeIeDXFNa+HgL8BjgC/AHwauLV0zYl8B/j2BMf/dTa+iCRJkiRJkiRprGu5OL0J+EXgHzJzeORgRPwW8CPgdRQL1V8uHW8A/hi4ANxeepEjEfEh4FvA6yPilzPzwQnu9e3MLFzB7yJJkiRJkiRJKnPNLuuRmd/KzK+WF6ZLx/cBf1Tavb3s1OuBFuDBkcJ0qf0Q8Nul3RVXLmJJkiRJkiRJ0qW6lp+cnsq50vZ82bE7Stt/nKD9d4FTwAsjojYzz4w7/+MR8S6gAdgHfC8zfdmjJEmSJEmSJF0hj7vidERUA28u7ZYXop9W2m4a3yczz0fEduAngJuBvnFN3lj6lN/ny8DbMvPobMQtSZIkSZIkSfp3VYVC4ZIbR8STOjo6slAojH/yuLzNoo6OjvZCoXBsNgIcr6Oj4/eBu4CvZ+ZHyo6/H1gKfLxQKByZoN8bgScCDxQKhV2lY/UUX7T4a8AHKL448WGKRew7gBd1dHTcXygUcvz1ACLi7R0dHZ+/8847lwPccsst9PX10dzczOrVq1m+fDldXV0sXbqU9evXU1tby+7duzl16hQnT55kz5493HDDDWzYsIGWlha6u7vH9Onr66OmpoY9e/Zw8uRJTp8+za5du6itrWXdunW0tbWN6bNkyRIeeeQRqqqq2LdvH8ePH2doaIidO3eyYMECent7aW9vZ9WqVaN9Fi9ezKZNm4gIDhw4wODgIGfPnmXHjh3U1dXR29vL8uXL6ezs5MYbb6Srq4vGxka2bNlCZnLo0CGOHj3K+fPn6e/vZ+HChaxZs2b0+iP3a2hoYOvWrQwPD3PkyBEOHz7M8PAw27ZtY9GiRWP6tLW1sXr1ahYtWsT27ds5f/48R48e5dChQwBs2bKFxYsX09PTM9qntbWVNWvWUF9fT39/P+fOnePYsWMcOHCAiGDTpk0sWbJkTJ+WlhZ6e3upq6tjYGCAM2fOMDg4yP79+6mqqmLjxo0sXbqU7u7u0e/e3NzMunXrqK2tZdeuXQwNDXHixAn27t1LdXX1aP5H+nR2dtLc3MyGDRtGczk+/+vXr6e1tXVMLpuamujr66O6uno0l+PzPz6XI/mfN28e+/fvZ3BwkDNnzjAwMDCa/2XLlk2a/4MHD/Loo49y7ty5i/Lf1dU1um1oaLgo/xcuXGDbtm0sXLiQnp6e0d+rvb2d7u7uKfPf0NAwmpfOzk7a29vp6emhvr5+wvxv3rx50vzX1dWxY8cOzp49e1H+m5qaxozLlpYW1q5dy/z58xkYGGBoaIjjx4+zb9++Mfkv71Oe/5GxfOLECfbs2UNNTQ19fX0TjuXy/JeP5fL8r1q1avR3a2pqYuPGjZOO5bVr147Jf2dn54T5HxnLI/kfn8uRsQyMyf9EY3nk385I/oeHhzl8+PCY/I8fyyP5Lx/L5fnfunUrjY2NY3LZ1tY2mv/ysXzw4MEx+V+9evXoGGttbZ10LM+bN++i/I/0Gcn/zp07Jx3LzuXO5c7lzuXO5c7lzuXO5c7lzuXO5c7lzuXO5dfjXL5s2bKO6eq0syEyJ6y7Ttw44gJQyMyPTtHmg8BHMrNqFuIbf+33AH8AbARuzcwjZec2AU8FnpqZWybo+33gBcALMvOH09ynAegBngLclZl/P1X7lStXJsCKFS5pLUmSJEmSJOlxL67GTSp9IWJwlQK76MYR76RYmN4AvKS8MF0y8qR24ySXaBjXblKZOQj8VWn3xRWGKkmSJEmSJEmaRqXF6UvRBpyczQtGxPuAPwTWUSxM75ug2SOl7S0T9K+m+BT0eWDbJd72YGlbX1m0kiRJkiRJkqTpTPtCxIh487hDz5ngGEAV8CTgvwJrZyG2kft/APg9istsvDwzD03S9FsUX2r4SuCvx517MVAHfDczJ10ve5yfKW0vtZgtSZIkSZIkSbpE0xangfuBkYWpE3h16TPeyHIfp4BZWTA7Ij4EfARYBbxigqU8yn0J+H+AX46Iz2ZmV+ka84HfLbVZOe76twI/yMzhccffBPwScBb429n4LpIkSZIkSZKkf3cpxel7S9sA/gz4CjDRCwIvAIcpFnsfnWlgEXE3xcL0BeB7wHsiLlruuj8z74fiOtER8TaKRepvR8SDwBHgF4GnlY7/zbj+DwDzSi9L3AXMB34S+CmKS4D835nZP9PvIkmSJEmSJEkaa9ridGZ+ceTvUsH4K5n551c0qqKnlLZVwPsmafMdik92A5CZX4mI24APAq+jWGzeAvw68JnMzHH9VwIvA24FmikW4HeXrvm/MnPNbHwRSZIkSZIkSdJYcXG9VpVauXJlAqxYseKxDkWSJEmSJEmSZuqiJSyuhHmVNI6I/xQRvxMRbZOcby+df87shCdJkiRJkiRJmosqKk4Dvwm8FTgwyfn9wFsoLqMhSZIkSZIkSdKEKi1OvwB4eIK1mwEoHf8WxTWcJUmSJEmSJEmaUKXF6XZg1zRt9gDLLi8cSZIkSZIkSdL1oNLi9CmgZZo2LcCZywtHkiRJkiRJknQ9qLQ43QO8OiIWTnQyIhqAV5faSZIkSZIkSZI0oUqL01+g+GT0P0fEs8pPRMSzgX8CmkvtJEmSJEmSJEmaUHUljTPzbyLiVcCbgdURsR/YDdwItAEBfDEz/3rWI5UkSZIkSZIkzRmVPjlNZt4DvAPYQPEFic8rbdcDb8/Me2czQEmSJEmSJEnS3FPRk9MjMvMLwBciog5YDDyamadmNTJJkiRJkiRJ0px1WcXpEaWCtEVpSZIkSZIkSVJFLqs4HREtwOuAZwD1mfnWsuNPAdZm5ulZi1KSJEmSJEmSNKdUXJyOiLcAnwHmU3wBYgJvLZ1uA34AvB3401mKUZIkSZIkSZI0x1T0QsSIeDnwBWAT8BpgZfn5zFxH8cWId81WgJIkSZIkSZKkuafSJ6c/AOwFbsvMwYh47gRteoEXzDgySZIkSZIkSdKcVdGT08Dzga9l5uAUbXYB7ZcfkiRJkiRJkiRprqu0OH0DcHKaNouBC5cXjiRJkiRJkiTpelBpcbofeN40bX4aeOSyopEkSZIkSZIkXRcqLU7/PfCzEfGGiU5GxL3As4AvzzQwSZIkSZIkSdLcVekLEX8f+GXgryPi9UAjQES8C/hZ4LXAZuCzsxmkJEmSJEmSJGluqag4nZlHI+I24M+B8qenP1Pafg/4L5k53brUkiRJkiRJkqTrWKVPTpOZA8DtEfEs4AXAUuAY8MPMXDXL8UmSJEmSJEmS5qApi9MR8RDwYGb+bWn/xUB/Zg5kZi/QexVilCRJkiRJkiTNMdO9EPEu4Oll+w8D91yxaCRJkiRJkiRJ14XpitOPAg1l+3EFY5EkSZIkSZIkXSemW3O6D/iViOgE9paO3VRa3mNKmfndmQYnSZIkSZIkSZqbpitOdwBfAf6q7Njdpc90qi43KEmSJEmSJEnS3DZlcToz/ykingG8DLgRKADfKX0kSZIkSZIkSbos0z05TWbuAP4UICIKwLcz8yNXOC5JkiRJkiRJ0hw25QsRI+JTEfGKskP3UlzmQ5IkSZIkSZKkyzZlcRp4H/AzZft/Btx15cKRJEmSJEmSJF0PpitOnwDqyvbjCsYiSZIkSZIkSbpOTLfm9BbgtRHxd8De0rHFEfGk6S6cmQMzDU6SJEmSJEmSNDdN9+T0x4EfA74PbC8de2/p76k+22YaWEQsjYi3RsTfRcSWiDgdEcci4l8j4i0RMWHsEfHCiPh6RByJiFMR0RsR74uIqinu9fMR8e3S9U9ExP+JiLtn+h0kSZIkSZIkSROb8snpzPzriNgO3AncCNwD9AI9Vz403gCspPjE9sPAANAGvBb4E+BVEfGGzMyRDhHxauDLwBDwN8AR4BeATwO3lq45RkS8C/gscBj4S+As8Hrg/oj4j5n5m1fqC0qSJEmSJEnS9SrKarvTN44YBgqZ+ZErF9Love4A6oF/yMzhsuPtwI+AJwKvz8wvl443UFyGpBG4NTO7SsfnA98CXgD8SmY+WHatm4CNwEngeZnZXzq+BOik+NT4CzPzB1PFunLlygRYsWLFTL+2JEmSJEmSJD3Wrsq7B6db1mO8e4GvXIlAxsvMb2XmV8sL06Xj+4A/Ku3eXnbq9UAL8OBIYbrUfgj47dLu+OrxfwNqgT8cKUyX+hwF/kdp9x0z+yaSJEmSJEmSpPGmeyHiGJn5xSsVSIXOlbbny47dUdr+4wTtvwucAl4YEbWZeeYS+nxjXBtJkiRJkiRJ0iyp9MlpImJeRLw7In5YeoHg+bJzz42I/zcibpndMMfcvxp4c2m3vKj8tNJ20/g+mXme4osaq4GbL7HPXorLfTwhIupmGLYkSZIkSZIkqUxVoVC45MYRcQPwTYpLXSyg+DTyokKh8BGAjo6OIeABYLBQKHx7toMt3eP3gbuAr5evfd3R0fF+YCnw8UKhcGSCfm+kuE71A4VCYVfp2MeAKuCDhUJheII+7wQWAZ8rFAonxp+PiLd3dHR8/s4771wOcMstt9DX10dzczOrV69m+fLldHV1sXTpUtavX09tbS27d+/m1KlTnDx5kj179nDDDTewYcMGWlpa6O7uHtOnr6+Pmpoa9uzZw8mTJzl9+jS7du2itraWdevW0dbWNqbPkiVLeOSRR6iqqmLfvn0cP36coaEhdu7cyYIFC+jt7aW9vZ1Vq1aN9lm8eDGbNm0iIjhw4ACDg4OcPXuWHTt2UFdXR29vL8uXL6ezs5Mbb7yRrq4uGhsb2bJlC5nJoUOHOHr0KOfPn6e/v5+FCxeyZs2a0euP3K+hoYGtW7cyPDzMkSNHOHz4MMPDw2zbto1FixaN6dPW1sbq1atZtGgR27dv5/z58xw9epRDhw4BsGXLFhYvXkxPT89on9bWVtasWUN9fT39/f2cO3eOY8eOceDAASKCTZs2sWTJkjF9Wlpa6O3tpa6ujoGBAc6cOcPg4CD79++nqqqKjRs3snTpUrq7u0e/e3NzM+vWraO2tpZdu3YxNDTEiRMn2Lt3L9XV1aP5H+nT2dlJc3MzGzZsGM3l+PyvX7+e1tbWMblsamqir6+P6urq0VyOz//4XI7kf968eezfv5/BwUHOnDnDwMDAaP6XLVs2af4PHjzIo48+yrlz5y7Kf1dX1+i2oaHhovxfuHCBbdu2sXDhQnp6ekZ/r/b2drq7u6fMf0NDw2heOjs7aW9vp6enh/r6+gnzv3nz5knzX1dXx44dOzh79uxF+W9qahozLltaWli7di3z589nYGCAoaEhjh8/zr59+8bkv7xPef5HxvKJEyfYs2cPNTU19PX1TTiWy/NfPpbL879q1arR362pqYmNGzdOOpbXrl07Jv+dnZ0T5n9kLI/kf3wuR8YyMCb/E43lkX87I/kfHh7m8OHDY/I/fiyP5L98LJfnf+vWrTQ2No7JZVtb22j+y8fywYMHx+R/9erVo2OstbV10rE8b968i/I/0mck/zt37px0LDuXO5c7lzuXO5c7lzuXO5c7lzuXO5c7lzuXO5dfj3P5smXLOiav0M6eSl+I+EHgo0AB+F3gd4APZWZVWZt/Ahoy82dmN1SIiPcAf0DxJYa3ZuaRsnObgKcCT83MLRP0/T7FlyK+IDN/WDp2FqgBakpPV4/vswdYBiwrrXU9IV+IKEmSJEmSJGkOuSZfiPhG4N8y8yOlFxVOVNneDjxpxpGNExHvpFiY3gC8pLwwXXKstG2c5BIN49pV0mewglAlSZIkSZIkSdOotDj9FOCH07Q5AjRdXjgTi4j3AX8IrKNYmJ7oKeZHStuL1rsurVP9FIovUNx2iX2WAfXArsw8dfnRS5IkSZIkSZLGq7Q4fRpYPE2bJwGPXl44F4uIDwCfBnooFqYPTNL0W6XtKyc492KgDvh+Zp65xD6vGtdGkiRJkiRJkjRLKi1O9wCvKL0Y8SIR0Qj8HPCjmQZWut6HgN8DVgEvzcxDUzT/EnAI+OWIeH7ZNeZTXB8bYOW4PvcBZ4B3RcRNZX2WAL9V2v2jGXwFSZIkSZIkSdIEqits/8fAA8ADEfGW8hMRsZhisXcJs1DQjYi7gY8AF4DvAe+JuGgd7v7MvB8gMwcj4m0Ui9TfjogHKS4x8ovA00rH/6a8c2Zuj4j3A58BuiLib4CzwOuBJwCfzMwfzPS7SJIkSZIkSZLGqqg4nZl/HREvA+6lWPQ9ChARXcBPALXA5zLz67MQ21NK2yrgfZO0+Q5wf1l8X4mI24APAq8D5gNbgF8HPpOZF73AMTM/GxH9wG8Cb6b4NPkG4Lcz84uz8D0kSZIkSZIkSePEBPXa6TtF3AO8F3gWMPI483rgU5l536xF9zixcuXKBFixYsVjHYokSZIkSZIkzdRFS1hcCZUu6wFAaSmN+yNiAcVlPI5l5snZDEySJEmSJEmSNHddVnF6RGaeBk7PUiySJEmSJEmSpOvEvMc6AEmSJEmSJEnS9cfitCRJkiRJkiTpqrM4LUmSJEmSJEm66ixOS5IkSZIkSZKuOovTkiRJkiRJkqSrzuK0JEmSJEmSJOmqq56Ni0REM/CzwCng/8vMC7NxXUmSJEmSJEnS3FTRk9MRsSIi/k9ENJUdex7QB3wJ+Drw/Yion90wJUmSJEmSJElzSaXLevwSkJl5pOzYx4ElwH0Ui9M/CbxjdsKTJEmSJEmSJM1FlRannwr0juyUlvO4DfjTzHxrZv4C0An8l9kLUZIkSZIkSZI011RanF4KHCjbv7W0/buyY98DnjyToCRJkiRJkiRJc1ulxekjQHPZ/m3AMPD9smMJzJ9hXJIkSZIkSZKkOazS4nQf8AsRsTQiFlNcg7ozMwfL2twE7Jul+CRJkiRJkiRJc1B1he3/APgKsAs4D9QBHxg5GRFVwIsY+yS1riG7TiXrjg1z+gIsqIJnNs7jCXXxWIclSZIkSZIk6TpTUXE6M/93RLwDeHvp0AOZ+ZdlTV5GcUmPb85SfJol644N89CuYTYez3Fnhnn6ouC1T5jHMxsrfZBekiRJkiRJki5PpU9Ok5lfAL4wyblvAktmGpRm18MHhvmTbRcYX5YeZpOdPgAAIABJREFUsfF48j/7LvC2m+H2VgvUkiRJkiRJkq48K5Fz3LpjUxemRyTwx9susO7Y8NUIS5IkSZIkSdJ1rqLidEQ8NyJ+NSIay47VR8QXI+LRiNgTEe+d/TB1uR7aNTxtYXpEAn+3y+K0JEmSJEmSpCuv0ienPwB8MDOPlR37n8B/LV1rKfCpiHjFLMWnGdh1KidYY3pqfceTXacq6yNJkiRJkiRJlap0zennA98e2YmIGuBu4EfA7UATsBp4D/BPsxKhLtvlLtGx7tgwT6irmuVoJEmSJEnS9eLjG89Pef79T6/4NWiS5qBKn5xuBXaW7T8fWAR8PjOHMnMP8PfAs2YpPs3A6QtXt58kSZIkSdJ4R8/m6EeSylX631Q5rs+LSse+U3bsINAyw7g0CxZc5sPPl9tPkiRJkiQJxj4ZXf4UtU9MSypX6ZPTA8DPlO2/GtiVmdvKji0Hjs40MM3cMxsrTe/M+kmSJEmSJEnSpaq0Cvm3wAsj4ksR8ZfAC4AvjWvzTGDrbASnmXlCXfD0RVFRn2csCp5QV1kfSZIkSZIkSapUpcXpTwM/AF4L/BdgDfCRkZMR8R+A5zF2mQ89hl77hHlcaqk5gNc8waemJUmSJEmSJF15FVUiM/NEZt5K8YWHzwKen5nHypqcAl4DrJy9EDUTz2ycx1tvrpq2QB3A226uckkPSZIkSZIkSVfFZa1Cn5nrJjneD/TPIB5dAS9pnUdLLfzdrmH6jl/8ZtxnLApe84R5FqYlSZIkSZIkXTWXVZyOiBrgpcAzgIWZ+dHS8flAA3AoM4dnLUrN2DMbi8XnXaeSP9h8nuGEeQHvfWq1a0xLkiRJkiRJuuoqLk5HxCuBPwXaKa4GkcBHS6efA/wb8Cbgr2cpRs2iJ9QFrbUxZl+SJEmSJEmSrraK1nGIiOcDX6FYkP414K/Kz2fmD4HtFNedliRJkiRJkiRpQpUuMvwhii89fH5mfgbYPEGbTuDZMw1MkiRJkiRJkjR3VVqcvhX4Smbum6LNTmDZ5Yf07yLi9RHx2Yj4XkQMRkRGxF9O0X5hRHw0IvoiYigiHo2If4mI/zxJ+3tK15zs847Z+B6SJEmSJEmSpLEqXXN6IXBomjZ1VF70nsxvU3wK+wSwC3j6ZA0jYjHwPeCZwHrg80A98IvAP0TEe0tPe0/k74GeCY53XX7okiRJkiRJkqTJVFqc3g38xDRtngNsu7xwLvJrFIvSW4DbgIenaFugWJh+CPilzDwPEBEtwI+AT0TENzJzoqVIvpKZ989SzJIkSZIkSZKkaVT6hPM3gJ+LiBdNdDIiXgW8EPjaTAMDyMyHM3NzZuYlNH9tafs7I4Xp0jUOAp8EagCX6ZAkSZIkSZKka0ClT07/T+CXgX+KiM8CNwFExJ3Ai4F3AnuBT81ijJeqvbSd6KntkWMvnaTvcyLifcB8ik+HP5yZu2Y5PkmSJEmSJElSSUXF6czcHRGvAP4WeH/Zqf8NBLAVeG1mTrcu9ZVwiOKLGJ8CbBh37ubSdrI1q987bv9CRPwJ8L7MHJq9ECVJkiRJkiRJAFWFQqGiDoVCYW9HR8fngG5gB7AR+DbwaeDdmblvlmMEoKOj4ybgHmBtoVB4aILzTwOeB7R2dHQ8VCgUEiAilgL3AUuA6o6Ojt8vFArnS30agVUU17b+78DngP8DPB94OfBjhULhy5PFFBFv7+jo+Pydd965HOCWW26hr6+P5uZmVq9ezfLly+nq6mLp0qWsX7+e2tpadu/ezalTpzh58iR79uzhhhtuYMOGDbS0tNDd3T2mT19fHzU1NezZs4eTJ09y+vRpdu3aRW1tLevWraOtrW1MnyVLlvDII49QVVXFvn37OH78OENDQ+zcuZMFCxbQ29tLe3s7XV2rGF6yjJrt3TzrxiVs2rSJiODAgQMMDg5y9uxZduzYQV1dHb29vSxfvpzOzk5uvPFGurq6aGxsZMuWLWQmhw4d4ujRo5w/f57+/n4WLlzImjVrRmNqb29n1apVNDQ0sHXrVoaHhzly5AiHDx9meHiYbdu2sWjRojF92traWL16NYsWLWL79u2cP3+eo0ePcuhQ8f88tmzZwuLFi+np6Rnt09raypo1a6ivr6e/v59z585x7NgxDhw4QESwadMmlixZMqZPS0sLvb291NXVMTAwwJkzZxgcHGT//v1UVVWxceNGli5dSnd39+h3b25uZt26ddTW1rJr1y6GhoY4ceIEe/fupbq6ejT/I306Oztpbm5mw4YNo7kcn//169fT2to6JpdNTU309fVRXV09msvx+R/5bcfnf968eezfv5/BwUHOnDnDwMDAaP6XLVs2ps/ixYtH83/w4EEeffRRzp07d1H+u7q6RrcNDQ0X5f/ChQts27aNhQsX0tPTM/p7tbe3093dPWX+GxoaRvPS2dlJe3s7PT091NfXT5j/zZs3T5r/uro6duzYwdmzZy/Kf1NT05hx2dLSwtq1a5k/fz4DAwMMDQ1x/Phx9u3bNyb/5X3K8z8ylk+cOMGePXuoqamhr69vwrFcnv/ysVye/1WrVo3+bk1NTWzcuHHSsbx27dox+e/s7Jww/yNjeST/43M5MpaBMfmfaCyP/NsZyf/w8DCHDx8ek//xY3kk/+VjuTz/W7dupbGxcUwu29raRvNfPpYPHjw4Jv+rV68eHWOtra2TjuV58+ZdlP+RPiP537lz56Rj+Vqeyycby87lzuXO5c7lzuXO5c7lzuXO5c7l19pc/sjJoGbvJobrl1A7sMa53LncufxxMJcvW7asY6pa7WyJS1vO+bEXEbdTfCHiA5n5pgnOtwM/BJ4MrAP+BagDXg0cp/hUdR1Qm5lnp7nXE4E1FAvaz8nMNVO1X7lyZQKsWLGisi/1GPn4xtEluXn/0ytd2UWSJEmSJOnSWYeQHpfiatyk0hciXrNKT2z/JPAZoB74VYqF6a8BLwMWAMemK0yXrrUT+Hpp98VXJGBJkiRJkiRJuo5d1n9XRcTzgZ+i+GRx1QRNMjM/OpPALkdmHqS4fvSYNaQj4iUUq/2dFVzuYGlbPzvRSZIkSZIkSZJGVFScjogG4CFgpNg7mQSuenF6Cm8rbR+ooM9Pl7bbZjkWSZIkSZIkSbruVfrk9MeBO4DvUXzJ4E7g/JQ9rpKImAfUZeaJccffCvwK0MO44nRE/Gxmfm/csaD4csQXAIeAf7yScUuSJEmSJEnS9ajS4vSrgW7gJZk5fAXiGSMi7gLuKu22l7YviIj7S38fyszfLP1dB+yPiH8GtpSO/SzF5Ue2Aq/JzHPjbvHdiNhEcbmP3UAjcCvwTOAU8MbMHJzdbyVJkiRJkiRJqrQ43Qj8xdUoTJc8B7h73LGbSx+AHcBIcfoM8CDwIuDlpWNbgQ8Dnxr/RHXJJygWr+8AmoBhYAD4XKmPS3pIkiRJkiRJ0hVQaXF6M9B2JQKZSGYWgMIltj0HvKXC67+/8qgkSZIkSZIkSTM1r8L2nwN+ISJuvBLBSJIkSZIkSZKuD5U+Of0Niktg/FtEdACrgEcnapiZAzOMTZIkSZIkSZI0R1VanO4HEgjgT6Zol5dxbUmSJEmSJEnSdaLSAvKfUyw8S5IkSZIkSZJ02SoqTmfmPVcoDkmSJEmSJEnSdaTSFyJKkiRJkiRJkjRjFRWnI+JCRHxomjYfjIjzMwtLkiRJkiRJkjSXVfrkdJQ+l9JOkiRJkiRJkqQJXYllPZYAQ1fgupIkSZIkSZKkOWLaFyJGxIvHHbppgmMAVcCTgDcCj8xCbJIkSZIkSZKkOWra4jTwbSBLfydwd+kzkQCGgd+YcWSSJEmSJEmSpDnrUorTH6FYlA7gdygWq78zQbsLwGHg4czcOFsBSpIkSZIkSZLmnmmL05lZGPk7Iu4GvpKZn7mSQUmSJEmSJEmS5rZLeXJ6VGY+5UoFIkmSJEmSJEm6fsx7rAOQJEmSJEmSJF1/pnxyOiK+RekliJm5q7R/KTIzXzrj6CRJkiRJkiRJc9J0y3rcXtrWjdufTl5OMJIkSZIkSZKk68OUxenMnDfVviRJkiRJkiRJl2PKYnNEvCcifupqBSNJkiRJkiRJuj5M9yT0/wJeObITERci4kNXNiRJkiRJkiRJ0lw3XXF6CKgt24/SR5IkSZIkSZKkyzZdcXo78HMR0VZ2zJcdSpIkSZIkSZJmZLri9OeB/wTsiYgLpWOF0vIeU33OX9mwJUmSJEmSJEmPZ9VTnczMz0TEAeBOYDnwEmAA6L/yoUmSJEmSJEmS5qopi9MAmfkg8CBARAwD92XmR650YJIkSZIkSZKkuWu6ZT3G6wC+fQXikCRJkiRJkiRdR6Z9crpcZnZcqUAkSZIkSZIkSdePSp+cliRJkiRJkiRpxixOS5IkSZIkSZKuOovTkiRJkiRJkqSrzuK0JEmSJEmSJOmqszgtSZIkSZIkSbrqLE5LkiRJkiRJkq66a7o4HRGvj4jPRsT3ImIwIjIi/nKK9gsj4qMR0RcRQxHxaET8S0T852nuc3dE/CgiTkTEsYj4dkT8/Ox/I0mSJEmSJEkSXOPFaeC3gXcBzwF2T9UwIhYDPyj1uQB8HvgS8B+Bf4iI90zS7xPA/cAy4I+Bvyz1+WpEvGtWvoUkSZIkSZIkaYzqxzqAafwasAvYAtwGPDxF2wLwTOAh4Jcy8zxARLQAPwI+ERHfyMzNIx0i4oXAbwBbgZ/MzKOl4x8HVpX6fC0z+2f5e0mSJEmSJEnSde2afnI6Mx/OzM2ZmZfQ/LWl7e+MFKZL1zgIfBKoAd4xrs/I/sdGCtOlPv3A54Ba4N7LDF+SJEmSJEmSNIlrujhdofbSdtsE50aOvXTc8TtK23+coM83xrWRJEmSJEmSJM2SuVScPlTaPmWCczeXtk8fORAR9cCNwInM3DtBn5HlP26ZtQglSZIkSZIkSQBUFQqFxzqGS9LR0XETcA+wtlAoPDTB+acBzwNaOzo6HioUCgkQEUuB+4AlQHVHR8fvFwqF8x0dHS0U15s+WCgUPjXB9c4BHwJqCoXC704UU0S8vaOj4/N33nnncoBbbrmFvr4+mpubWb16NcuXL6erq4ulS5eyfv16amtr2b17N6dOneLkyZPs2bOHG264gQ0bNtDS0kJ3d/eYPn19fdTU1LBnzx5OnjzJ6dOn2bVrF7W1taxbt462trYxfZYsWcIjjzxCVVUV+/bt4/jx4wwNDbFz504WLFhAb28v7e3tdHWtYnjJMmq2d/OsG5ewadMmIoIDBw4wODjI2bNn2bFjB3V1dfT29rJ8+XI6Ozu58cYb6erqorGxkS1btpCZHDp0iKNHj3L+/Hn6+/tZuHAha9asGY2pvb2dVatW0dDQwNatWxkeHubIkSMcPnyY4eFhtm3bxqJFi8b0aWtrY/Xq1SxatIjt27dz/vx5jh49yqFDxf9/2LJlC4sXL6anp2e0T2trK2vWrKG+vp7+/n7OnTvHsWPHOHDgABHBpk2bWLJkyZg+LS0t9Pb2UldXx8DAAGfOnGFwcJD9+/dTVVXFxo0bWbp0Kd3d3aPfvbm5mXXr1lFbW8uuXbsYGhrixIkT7N27l+rq6tH8j/Tp7OykubmZDRs2jOZyfP7Xr19Pa2vrmFw2NTXR19dHdXX1aC7H53/ktx2f/3nz5rF//34GBwc5c+YMAwMDo/lftmzZmD6LFy8ezf/Bgwd59NFHOXfu3EX57+rqGt02NDRclP8LFy6wbds2Fi5cSE9Pz+jv1d7eTnd395T5b2hoGM1LZ2cn7e3t9PT0UF9fP2H+N2/ePGn+6+rq2LFjB2fPnr0o/01NTWPGZUtLC2vXrmX+/PkMDAwwNDTE8ePH2bdv35j8l/cpz//IWD5x4gR79uyhpqaGvr6+Ccdyef7Lx3J5/letWjX6uzU1NbFx48ZJx/LatWvH5L+zs3PC/I+M5ZH8j8/lyFgGxuR/orE88m9nJP/Dw8McPnx4TP7Hj+WR/JeP5fL8b926lcbGxjG5bGtrG81/+Vg+ePDgmPyvXr16dIy1trZOOpbnzZt3Uf5H+ozkf+fOnZOO5Wt5Lp9sLDuXO5c7lzuXO5c7lzuXO5c7lzuXX2tz+SMng5q9mxiuX0LtwBrncudy5/LHwVy+bNmyjsnqtLMpLm0558deRNxO8YWID2TmmyY43w78EHgysA74F6AOeDVwHFhW2q/NzLMRsRzYDezOzCdMcL0a4CxwJjPnTxXbypUrE2DFihWX/f2upo9vHF2Sm/c//Vp/J6YkSZIkSXo8sw4hPS7F1bjJnFnWIzP3AT8JfAaoB36VYmH6a8DLgAXAscw8W+pyrLRtnOSSjePaSZIkSZIkSZJmyZwpTgNk5sHMfG9m3pyZN2RmW2a+heI61AF0lrU9SfHJ6YURsWyCyz21tN10xQOXJEmSJEmSpOvMnCpOT+Ftpe0D445/q7R95QR9XjWujSRJkiRJkiRplsyZ4nREzIuIhRMcfyvwK0APFxen/6i0/WBELCnrcxPwTuAMxZcpSpIkSZIkSZJm0TW9Cn1E3AXcVdptL21fEBH3l/4+lJm/Wfq7DtgfEf8MbCkd+1ngp4CtwGsy81z59TPz+xHxKeDXgd6I+BJwA/BLQBPw7szsn/UvJkmSJEmSJEnXuWu6OA08B7h73LGbSx+AHcBIcfoM8CDwIuDlpWNbgQ8Dn8rMExPdIDN/IyJ6gXcBbweGgW7g45n5tVn6HpIkSZIkSZKkMtd0cTozC0DhEtueA95ymff5IvDFy+krSZIkSZIkSarcnFlzWpIkSZIkSZL0+GFxWpIkSZIkSZJ01VmcliRJkiRJkiRddRanJUmSJEmSJElXncVpSZIkSZIkSdJVZ3FakiRJkiRJknTVWZyWJEmSJEmSJF11FqclSZIkSZIkSVedxWlJkiRJkiRJ0lVncVqSJOn/b+++w+woy8aPf5/t2U3fdBKSkNBC770qTUABUV+k2lFB1JfXgiCJvaGiKOJPXvIqigoiqHQQBClCGgkppFfS2ybZlj3n+f0xZzebZTfZhN1zdjffz3XNNXtmnpm5z8595sy5z5xnJEmSJElZZ3FakiRJkiRJkpR1FqclSZIkSZIkSVlncVqSJEmSJEmSlHUWpyVJkiRJkiRJWWdxWpIkSZIkSZKUdRanJUmSJEmSJElZZ3FakiRJkiRJkpR1FqclSZIkSZIkSVlncVqSJEmSJEmSlHUWpyVJkiRJkiRJWWdxWpIkSZIkSZKUdRanJUmSJEmSJElZZ3FakiRJkiRJkpR1FqclSZIkSZIkSVlncVqSJEmSJEmSlHUWpyVJkiRJkiRJWWdxWpIkSZIkSZKUdRanJUmSJEmSJElZZ3FakiRJkiRJkpR1FqclSZIkSZIkSVlncVqSJEmSJEmSlHUWpyVJkiRJkiRJWdehi9MhhEtDCD8PIbwQQqgIIcQQwr07aF8cQvhsCOHVEMKaEMLmEMLMEMLPQgjDm2l/TWadLQ3Xtu8zlCRJkiRJkqQ9U0GuA9iJm4HDgM3AUuCAlhqGEAqAZ4CTgFnAfUANcAxwPXBVCOHEGOOMZhZ/GJjSzPQJ7yh6SZIkSZIkSVKzOnpx+gskRem5wGnAsztoezFJYfoZ4OwYY7p+RghhHPB14Ebgo80s+1CMcXwbxSxJkiRJkiRJ2okO3a1HjPHZGOOcGGNsRfN9MuNHGhemMx7OjPu3XXSSJEmSJEmSpN3V0a+c3hXTM+PzQgi3NylQX5AZP93CsoeHED4PlADLgGdjjEvbKU5JkiRJkiRJ2uN1peL0I8CDwCXAtBDC00AtcBRwMvBz4I4Wlr2hyeNUCOE3wOdjjNXtFK8kSZIkSZIk7bG6THE6xhhDCJeS9C19CzCm0exngD/EGFNNFltAcrPEJ0n6tu5FUsj+LvApoCfw4XYOXZIkSZIkSZL2OPljx47NdQytMm7cuBHANcC0sWPHPth0fgihBPgDcDXweeATwDeBF4DLgC+MGzdu6tixY9+sX2bs2LGLxo4d++rYsWPXjh07duvYsWM3jx07dsa4ceP+mtnWMePGjXto7NixK5uLKYTwyXHjxt11/vnnDwHYb7/9mDlzJv369WPy5MkMGTKECRMmUF5ezvTp0ykuLmbZsmVUVlayZcsW3nrrLYqKipgxYwb9+/dn0qRJ2y0zc+ZMCgsLeeutt9iyZQtVVVUsXbqU4uJi3njjDQYOHLjdMn369OHNN98kPz+fFStWsGnTJqqrq1myZAndunVj6tSpDBo0iAkTJpLuM5jCBZM4dK8+zJ49mxACq1atoqKigtraWhYtWkRpaSlTp05lyJAhvPbaa+y1115MmDCBXr16MXfuXGKMrFmzhvXr11NXV8fChQvp3r07r7/+ekNMgwYNYuLEifTs2ZN58+aRTqdZt24da9euJZ1OM3/+fHr06LHdMgMHDmTy5Mn06NGDBQsWUFdXx/r161mzZg0Ac+fOpXfv3kyZMqVhmQEDBvD6669TVlbGwoUL2bp1Kxs3bmTVqlWEEJg9ezZ9+vTZbpn+/fszdepUSktLWbx4MTU1NVRUVLBy5Ury8/OZNWsW5eXlTJo0qeG59+vXjzfeeIPi4mKWLl1KdXU1mzdvZvny5RQUFDTs//plXnvtNfr168eMGTMa9mXT/T99+nQGDBiw3b7s27cvM2fOpKCgoGFfNt3/9f/bpvs/Ly+PlStXUlFRQU1NDYsXL27Y/4MHD95umd69ezfs/9WrV7Nhwwa2bt36tv0/YcKEhnHPnj3ftv9TqRTz58+ne/fuTJkypeH/NWjQICZNmrTD/d+zZ8+G/fLaa68xaNAgpkyZQllZWbP7f86cOS3u/9LSUhYtWkRtbe3b9n/fvn23e13279+fadOmUVJSwuLFi6murmbTpk2sWLFiu/3feJnG+7/+tbx582beeustCgsLmTlzZrOv5cb7v/FrufH+nzhxYsP/rW/fvsyaNavF1/K0adO22/+vvfZas/u//rVcv/+b7sv61zKw3f5v7rVcnzv1+z+dTrN27drt9n/T13L9/m/8Wm68/+fNm0evXr2225cDBw5s2P+NX8urV6/ebv9Pnjy54TU2YMCAFl/LeXl5b9v/9cvU7/8lS5a0+FruyMfyll7LHss9lnss91jusdxjucdyj+Ueyz2Wd7Rj+ZtbAoXLZ5Mu60Px4tc9lnss91jeCY7lgwcPHrezem1bCK2712DuhRBOB54Ffh9jvKKZ+WOBW4EbYow/azLvMGAKsCjGOKKV27sXuBz4XIzx5ztqe+edd0aAT3/6061Zdc79cFZdw9//c0CXuXhekiRJkiR1QNYhpE4pZGMjednYSJbU3/Tw2aYzYoyvA+uA4SGE8laub3VmXNYGsUmSJEmSJEmSGulKxenizLh/0xkhhGKS/qMhuUliaxyXGc9/h3FJkiRJkiRJkproSsXpFzLjmzLF6MbGktz88bUY46b6iSGEU5quJCS+CpwArAEeb59wJUmSJEmSJGnP1aE7+gkhXARclHk4KDM+IYQwPvP3mhjjjZm/vw1cCLwLmBVCeByoAk4Cjs38fUOTTTwfQpgNvAYsA3pl2h8MVAKXxxgr2vp5SZIkSZIkSdKerkMXp4HDgaubTNsnMwAsAm4EiDEuCyEcCXwZOB/4CMmV4cuB8cD3Y4yzmqzrRySF6zOBvkAaWAz8AvhxjNEuPSRJkiRJkiSpHXTo4nSMcSxJlxytbb+apFh9487aZtr/z24FJkmSJEmSJEl6R7pSn9OSJEmSJEmSpE7C4rQkSZIkSZIkKessTkuSJEmSJEmSss7itCRJkiRJkiQp6yxOS5IkSZIkSZKyzuK0JEmSJEmSJCnrLE5LkiRJkiRJkrLO4rQkSZIkSZIkKessTkuSJEmSJEmSss7itCRJkiRJkiQp6yxOS5IkSZIkSZKyzuK0JEmSJEmSJCnrLE5LkiRJkiRJkrLO4rQkSZIkSZIkKessTkuSJEmSJEmSss7itCRJkiRJkiQp6yxOS5IkSZIkSZKyzuK0JEmSJEmSJCnrLE5LkiRJkiRJkrLO4rQkSZIkSZIkKessTkuSJEmSJEmSss7itCRJkiRJkiQp6yxOS5IkSZIkSZKyzuK0JEmSJEmSJCnrLE5LkiRJkiRJkrLO4rQkSZIkSZIkKessTkuSJEmSJEmSss7itCRJkiRJkiQp6yxOS5IkSZIkSZKyzuK0JEmSJEmSJCnrLE5LkiRJkiRJkrLO4rQkSZIkSZIkKessTkuSJEmSJEmSsq5DF6dDCJeGEH4eQnghhFARQoghhHt30L44hPDZEMKrIYQ1IYTNIYSZIYSfhRCG72C5qzPLbA4hbAwhPBdCuKB9npUkSZIkSZIkqUMXp4GbgeuAw4FlO2oYQigAngHuAHoA9wG/AlYB1wOvhxDGNLPcj4DxwGDg/wH3AocAfw8hXNdWT0SSJEmSJEmStE1BrgPYiS8AS4G5wGnAsztoezFwEkmB+uwYY7p+RghhHPB14Ebgo42mnwj8NzAPOCbGuD4z/YfAROBHIYR/xBgXtuFzkiRJkiRJkqQ9Xoe+cjrG+GyMcU6MMbai+T6Z8SONC9MZD2fG/ZtMvzYz/nZ9YTqz3YXAL4Bi4CO7FrUkSZIkSZIkaWc6dHF6F03PjM8LITR9XvX9Rz/dZPqZmfHjzazvsSZtJEmSJEmSJEltpKN367ErHgEeBC4BpoUQngZqgaOAk4Gfk/RHDUAIoQzYC9gcY1zezPrmZMb7tWfQkiRJkiRJkrQn6jLF6RhjDCFcStK39C1A45sfPgP8IcaYajStV2a8sYVV1k/v3doY7rzzztY2zanujf6+c0e9eEuSJEmSJL1D1iGkTil++tOfDu29kS7TrUcIoQT4E8lNDz8LDCYpQL8HGA48H0J4326susX+rkMInwwhTNiNdUqSJEmSJEnSHq3LXDkNfAX4AHBDjPGuRtMfy1xRPQW4nW3YVUA1AAAgAElEQVQ3R6y/MroXzdvZldXEGH8N/Hq3I5YkSZIkSZKkPVSXuXKabTc9fNsPRGKMrwPrgOEhhPLMtC3AMqB7CGFwM+vbNzOe3Q6xSpIkSZIkSdIerSsVp4sz4/5NZ4QQioGemYe1jWb9MzM+t5n1ndekjSRJkiRJkiSpjXSl4vQLmfFNmWJ0Y2NJujB5Lca4qdH0X2XGXwsh9KmfGEIYQdJvdQ1wT3sEK0mSJEmSJEl7shBji/f7y7kQwkXARZmHg4BzgPlsK0SviTHemGm7F/AKMBRYCDwOVAEnAcdm/n5XjPHlJtu4DfgisBR4ACgCPgSUA9fHGO9op6cnSZIkSZIkSXusjl6cHgvcuoMmi2KMIxq17w98GTgfGElyZfhykq45vh9jnNXCdq4GrgPGAGlgEvDDGOM/3vmzkCRJkiRJkiQ11aGL051Jfn5eTKc7xv8yhEBX2a95ATrIv1UdSB7Jt0htqSu9btpaCHnE2Nb/cbW5EMAc3nUhD7pafpsLuykA/t92pD3ef6Hjnu919XODjvp/72ra63XT1rp6vr8TnfJc2HOBHetM538dZl/uGedJneWYnUVrYoxvu7dfWyto7w3sKdLpSJyQ6ygS4ejI888/n+sw2sSpp56KHauoqXAdxEMvbdt1Tn2AC9/8WJuus6v4+/53c9nLtTtvqJy674Qijrp/Tq7D6HQmfmBfuKmLnWh/J3DZh2bkOopO574/jYH33Z7rMDq09MM3EE86vM3XG16c0mHOoxvrSufUzfE8Ozva47y1PXgu3LLOeC7seeGOdarzvw5yXrennCe117lOZxVenLIoG9vpSjdElCRJkiRJkiR1EhanJUmSJEmSJElZZ7ceOVZTC5/5Hjz9KqyrgNFD4TufhfNOynVkLfvmN7/JxIkTqa6upm/fvnz4wx/mggsuyHVYXd6cVXDId+DSI+Deq3MdTcd2+rzneKVyHQUhALBXQTfePODcHEfVessemcfsOyZTtXwLxf26cfj3TqX86EG5DqtLuP/MPts9TtVUMfqSazn6v3+ao4i6tppVS1n8m7FsmT2ZUFBEn+PPZdhHvkbI34NOP9bMhCc+CysmQml/OPOHsP/FuY6q3c2e83sWLHiIDRtnM3zv8zn+uO80zFux8mUmTPwWlZXLKS8/lOOP/TZlZXvlMNocm/88LH4VNr0Fex0FR16eTE/XwYTfwoYlULUOTroO+u2b21ibccfy1YxfuY5pldVc1r834/cdDsArm7Zwy6LlTNxSRT5weq/u/GyfoQwuKmyb7f4Jxv8Dps2Fy86B8WOT6TPmw1W3wrylyeOjDoSf3Qhj9mmTzbaJv/zlLzz++OPMnz+fd73rXdx0001va3PPPfdwzz338OMf/5ijjz46B1Hm1h3/gvGvwLTlcNlRMP7KZPrCtTDyVigr2tb2y2fBLeflJs72dseauYxfv4hp1Ru5rPcwxg87pmFeZbqOG5dP5c8blrI1pjmsW2+eH3V67oLdDQvuncGSB+ewafY6hlwwiiO+dyoAS/82l6m3vtjQLqYj6eoUp/zlffQ+uF+uwt1ts+//JQse/S0b5r3B8LM+xPG33N0wb/HT9zPtN9+gcvUySgcM5bBrv8nQ096Xw2jfmVWP/Y61zz1I1eI36XvSBYy47gcN89Y882dW/PUutm5YQ/cDjmL4Z75LUd+BOYx2N9XVwBOfgQVPQ/U66DMaTv8OjMociBY8k5z7VSyGIcfBheOh1/B2DyuVqmXCxG+wYuXL1NZupHv3vTns0M8zZPCppFK1vPzKl1i37g22VL7FmWeMZ+CAY9s9plZJ1cHUP8Pq2VBbCd37wYEXwMAxsG4hzHoENixN+t7uty8ccgmU9Mp11G2iJp3mM/OW8vTGTazbmmJ0t2K+M3ww5/XpyYzKaq6avYh51UmXQkd178bP9hnKmNKSHEfdPvagT4cdU10Khg2Ef/0a9h4Ej74IH/wqTPsjjBiS6+iad8UVV/DlL3+ZoqIiFi1axA033MC+++7L/vvvn+vQurTP/hmOaf/3tC7jjiFH8PHykbkOY5etfnEZM3/0Gkf95Ex6H9qf6tWVuQ6pS/nAP9c3/F1XtYW/nj+Uvc98fw4j6toW/2YshT37cuivXyK1pYLZ37yG1U/8ngHv2UO+YUvXwQPvgyOuhcuegsX/gvsvhI9OhvL9ch1du+rWbQAHjfkUy1e8SCpV0zC9pmY9/37xBo495hvsNeQMpk77GS++9N+cfdYfcxhtjpX0gv3PhlWzILV1+3nl+8Co0+C18TkJrTWGFBVy87CBPLFhE1XpbbcQWl+X4pOD+nFO7x4UhMB185fykTmLefygUW2z3f5w88fgiZehqmb76Q98H4YPhnQafnE//NdNMLUDpVi/fv246qqrePXVV6mpqXnb/GXLlvHcc89RXl6eg+g6hiG94OZz4YmZULX17fM3/BAK8rMfV7YNKezGzQMO4IlNK6mKqe3mfXLpJOpimpn7n0Pf/CKmVG3IUZS7r2RAKft+5jBWv7CMVM225zf0vaMZ+t7RDY+XPDib2b+cQq+DOudrolv/wRx0zVdZ/p+nSNVUNUyvXLWMl8ddwyk/+AuDjz+Ht156jBe/dhnvfXAOJX0H5DDi3VfYdwCD3/8ZNk55gVhb3TB90/T/sOwPt7Hf2HspHjScJfd8iwU//QL7f+MPOYx2N6XroMcwuOJf0GtvmPso/PWD8PFpUNQdHrwE3vMb2PdC+Nct8NcPwTWvtHtYMdZRWjqId535W8pKB/PW8ud58aUvct65D9OtpD/9+h3J/vtdyb9f+kK7x7JLYgq69YGTr0/GK2ck5z1nfhm2VsLwE+GYA5ObWE57ACb/AU74dK6jbhN1MTKsuJB/HTyavYuLeHR9BR+ctZBpR+zPkKICHjhgBMOLi0gDv1i+hv96cyFTjzgg12G3C7v1yLGybjD2U0khOi8PLjgFRg6BiTNzHVnLRo4cSVFRcrlCyFyZumzZslyG1OX9cQL07gbv6tq1DAFv/nwS+33mCPocPoCQF+g2sIxuA8tyHVaXtPiff6G4zwD6H35yrkPpsmpXLaHPie8hr6iYwj796XX4KVQt2YNuzrNmVnI17LFfgLx8GHEmDD0J3vhdriNrd8OGnsXQoe+muLj3dtOXLH2KXj1Hs/ewc8nPL+aQgz/Lho1vUlExP0eRdgBDDoPBh0JRk2N9XgGMOh3KRyVXC3VQl5T35qLy3pQXbH/Ny3l9evKBfr3pWZBPaX4e1w3ux4sVW9puu2fCRadDeZOLp3r3SM6rQ4AYIT8P5i5ps822idNOO41TTjmFnj17Njv/pz/9Kddeey2FhW1zlXlndMnhcNFhUL6HnwJd0msvLuq1F+UFRdtNf7N6E3+reItfDz2K/gXF5IfAUaV9WlhLxzX47BEMfvcIinoX77Ddkr/OZehFoxs+e3Y2w06/mKGnvY/iXn23m165ehmFPXoz5IRzCSGw10nvoaBbGZuXdd73xD7HnUPvY8+ioMf27/8bJz5LnxPOo9uwfckrLGLwpZ9l88zXqFmRlXutta2iMjh1LPQekRRM970Aeo1MfiX35oPQ7yA48ANQUAKnjIVVryfnhO2soKCUQw6+ju5lexFCHnsNOZ2ysqGsXzed/PwiDtj/Kvr3P4oQOtg3ewXFcMB5UFqe/D8HHQxlfZNfjg0cA3sdAYUlUFAEI0+BtQtyHXGbKcvPZ+zegxlRUkxeCFzQtxcjS4qYuLmK3gUFjCgpJoRABPIDzK1++xfaXYXF6Q5m5VqYvRja6KKSdvPjH/+Ys846iyuuuILy8nKOP/74XIfUZVVUwdcfgdsuyXUknctXV0yj3/S/cdLcZ3lu86pch9MqMZVmwxtrqFlfzTNn/ZmnTr2Pad94iVR1Xa5D65IWPHYvI8+7vNN+0OkMBrznata9+Ajpmipq165g45Tn6Xn4qbkOK4uauQt8jLD6jeyH0kFs3DiX3r23/dKqoKCU7mXD2Lhxbg6jUjY8X7GZg7L4U9Tep0PJSXD9D+Gmj2Rts+/Ys88+S0FBASeccEKuQ+nQhn8dht4MH/kdrNmc62iy7z9V6xheVMqtK6fTb/rfOGT2k/xl49Jch9UuKpdtYu2EFQx7X8fr1uid6nvAUfQcfgBLX/g76VSKpf96mLzCYnqPPiTXobW5GGNyDrRtAkDXuGhh80pYNzspSq+eDgMO2zavqAx6j4I107MeVlX1GjZtWkivXqN33rgjqa6Azauhx+C3z1s7D3p23e4uV9ZuZXZVzXbnS71fmUrJS69z/fxl3DS0E3aD00p269GBbK2Dy2+Bq8+HA0bkOpod++IXv8gNN9zA9OnTmTx5csOV1Gp7tzwCHzsRhnW+iyFy5vuDD2FMcU+KQh5/3LCECxe+xJR9382o4u65Dm2HatZUEbemWf74Qk76/QWEgjxe+8xTzL5zCgd+Yc/rb7I9bVmxmNWTn+e4m+7KdShdWvcxx7Lm6T8z+aojIJ2i/LSL6X3sWbkOK3vKD4CyAfDKD5Orpxc9m3TtMfyMXEeWM3V1lRQXb3/lWGFhD7bWtd0Vtep4pm6p4htLVvLwAdnrbmvDc7ClCv7vH0kXH51BZWUlv/71r7nttttyHUqH1a87vPY/cPhQWLsl6fbu8vHwxHW5jiy7lm6t5I3qCt7fcy/eOvACXq5cy/kL/82Y4p4cWNL8Ffmd1dKH5lJ+9EBKh/XIdShtLi8/n5HnXcHLt15FqraavIIiTvr2fRR063o/Geh15GnM/8nn6Xf2ZZQMGsHyB+6AEEg36uakU0pthb9dDodcDf0OgK2bk3uMNFbSC2o3ZTWsdHorL7/yJUaOeB89e3agmy7sTDoFE38Hw46FHk0KsRuXwZtPwHEfz01s7WxrOnL57EVcPaAvBzQqTm84/lC2pFL836r1DC/uur+o8srpDiKdhitvgaICuOPLuY6mdfLz8zn00ENZvXo1Dz30UK7D6ZKmLIWnZ8EX9tw6xm45rrScHvmFFOflc3XfEZxUWs6jm1bkOqydyitJvi8ceeUYSgaUUty3hH0+cjCr/tXBfo/cBSx47F76HXoS3Yd0vn7JO4uYTjPnWx+l93Fnc8S9r3PY/75K3ZYKlt37g50v3FXkF8KlD8HcR+Bng+A/t8GBH4SeQ3MdWc4UFJSydev2lzlurdtMYUHX+yCuxNyqGs6bMZ/bR+7FKb2y+yVxWTe49v3JDRJXrcvqpnfL//7v/3LOOecwZEgHvfFMB9C9GI4envQ3PbAn3PFBeHJW8kvDPUm3kE9hCNw88ECK8vI4rXt/zigbwJObV+Y6tDa39OG5DL2o6101DbDi1WeY8ouvcuYvnuZDz2/hXb98hle/ey3rZ0/JdWhtruchJzLkg59j/o+uY9pnTqNowFDySsooLO/EV8HGNPztSsgvgnPuSKYVdoeaiu3b1VRAUfa+XIkxzcuvfIW8vEKOPurmrG33HYtpmPS7pCu8Qy/dft7m1fDKXcnNEMs7eDcDuyEdI1fOWURRXuCOfd7+OaEsP59rB5Vz1ZzFrKpt5kYMXYBXTncAMcLHvgkr18Gjt0NhJ9srqVTKPqfbyXNzYOE62PuW5PHmGkhFmLEcJn0lt7F1JiFAbO7n9R1MUa9iSgaVgb1MtLuFj93LgVd+KddhdGmpzRvYunY5A869krzCYvIKi+l3xvtZ9sefMPTKTvItbFsYcChc+a9tj//vxOTqmj1Ur16jWbDg4YbHdXWVbN68pPP95FStsqi6lndPn8stQwdy5YC+O1+gHaTTUFkNy1ZBjkJotUmTJm130ceGDRu49dZb+fCHP8zll1+e4+g6pvpTpo5/lte2Di3ptfNGXcC6iSupXlXJkHNG5DqUdrF+zusMOPxkyg88CoDyMUdTftAxrHjtn/TZ7/AcR9f2Bpx7BQPOvQKA6rcWsOIvv6TbsE56U6UY4ZGPwZaV8KFHkwsSAPofBNP+b1u72i2wfl7S5UdWwor859Wbqa5Zy2mn/Iq8vE5ypW2MMPk+qNkEx38qKVDXq1wHL/0S9jsbhh2TuxjbSYyRj81dwsraOh4dsw+Fec0XA9JAZTrNstqtDCjqJPt1F3jldAfw6e/CzAXw959At+x1xbdb1q9fzzPPPENlZSWpVIpXX32VZ555hiOPPDLXoXVJnzwJ5o2FKV9NhmtPhvMPgic+m+vIOq4NqVqe2LSC6nSKupjm9+sX8/zmNZzTvXN8Kz/skn1Z8LsZ1KytonZjDQv+bzoDT98712F1Kaunvkzl6rfY+8z35zqULq2gZ1+KBgxl9ZN/IKbqqNtSwdrn/krp8K55h+kWrZoKddXJ3cZf+RFsXg6HXpPrqNpdOl1HKlVDOqaIMZX8na5j6F7vZmPFHJYseZJUqoY3pt9J7177da6fnLa1dCr5WXBMJ0NqazINIFWXPN6uXccqw9XFSHU6TSpGUhGq02nqYmRZTS1nTp/LZwf159rB/dp+u3VQXQOpNKRSyd91dfDUKzB5VjKtYjN88SfQpwcc2IF+KFNXV0dNTQ3pdJp0Ok1NTQ11dXX85Cc/Yfz48dx9993cfffdlJeXc+ONN3LxxRfnOuSsq0tB9dbM/k0nf9el4D8L4c2VyZcOazfD5x6A0/eFXt1yHXH7qItpqtOpzOsrNpzfntq9P3sXlvLdVbOoi2le3LKG57as5pzunas/0nRdmlRNHTEdiank73RdumH+kofmMPjsERR079xdSKbr6kjVVJNOpYjpVPJ3XR3lBx7NqtdfbLhSet2bk1k95cVO3ed0TNWRrq2BdJqYTpOurWmYVrV4NjFGale/xaK7bmbAe66moHsn/aLl8U/Dmpnwwb9DYaMD0H4XJ/cWmfWX5Pzv399ILlTol53z3wkTx1FRMZ9TT/4FBQXbF5dSqVpSqeSGeun0VlKpmqQv8I5g6p+TvruP+2RyJXq9qg3w4h0w8uRk6II+PW8pMyur+fuYkXTL31aifWrDJiZvriQVIxV1Kb64YBl98vM5MIv378imTnaNbtezaDnc9SAUF8Ggc7ZNv+smuPy83MXVkhACDz30ELfddhvpdJqBAwdy/fXXc8opp+Q6tC6ptCgZ6nUvhpJC6N/1ulxrM1tj5OYV05lVs4l8AgeU9OChESeyf0nn+Kft95kjqF1fzT/PeYD84nyGnDeSfT992M4XVKstePR3DDv9IgrLOkdOdGajbvwFS8Z/mxUP/xry8ulx0HEMveZruQ4ru6b9Dl7/TVJUHHYKXPZUclfyLm76jF/xxvRfNjxeuOjvHHzQZzjk4Os4+cSfMmHSt3n5P1+mvO+hnHjiHt6/7uwn4c3Htz1eOgH2Pze5c/0z34aqTH8UL9+ZjM/6enJH+w7iW0tWMG7Jtq4E7l29nluHDSQQmF9dy7glKxi3ZFvXWptPOLRttns3jPt/2x7f+xjc+onkpuLX/xCWroJuxXDMGHj851DSgV52v/3tbxk/fnzD4yeffJJrrrmGj370o9u1y8/Pp0ePHpSWlmY5wtz71uMw7rFtj+99DW49D/YfCDf9HVZtgp4lcNYBcF8nuuHlrvrWypmMWzWz4fG9GxZz64ADGTvoIB4ecSIfXzqR7616k+FFpfx22DEc0Mn6m55z5xRm3zG54fGyv81jv+uOYP/rjyRVU8dbjy3g6J+/K4cRto3p47/DG3d/q+Hxwsf/wMEfu5lDPv51DvnYLfz7psuoXr+S4t79GXP1lxl8XOe9P8fyv/yS5ff/vOHxuhceZvAHrmfA+dew4PYvUrNyMXklZfQ74/0M+dDncxjpO7BxEUy+C/KL4fZGF0CddxccfDlc8pekI/y/XQFDjoOL/piVsLZsWcbceX8mL6+Ih/627Qbkxxw1lhEjLuSRR9/Dlsq3AHjuX58A4MILnqJ72V5Zia9Fletg4UuQVwCPN+qG5LAPwZbVULk2OU9qfK50wQ+zH2c7WFRdy10r11IcAoNe3XbTzLtGDaUoL4/r5y9lac1WuuUFjulRyuMHjaIkr2teYxw6zDclnVwIIcYJuY4iEY6G559/PtdhtIlTTz2VeEeuo1BHE66D2LQfqne6zqkPcOGbH2vTdXYVf9//bi57uTbXYWgn7juhiKPu7wJ3PM+yiR/YF27qYudC3wlc9qEZuY6i07nvT2PgfbfnOoyO7eEbiCe1/U/Nw4tT6Cjn0Y11pXPq5nienR3tcd7aHjwXbllnPBf2vHDHOtX5Xwc5r9tjzpPa6VynswovTpkYYzy6vbfTNUvukiRJkiRJkqQOzeK0JEmSJEmSJCnrLE5LkiRJkiRJkrLO4rQkSZIkSZIkKessTkuSJEmSJEmSsi7E2EnuUNrBhRBWA4tyHYckSZIkSZIkvUPDY4z923sjFqclSZIkSZIkSVlntx6SJEmSJEmSpKyzOC1JkiRJkiRJyjqL05IkSZIkSZKkrLM4LUmSJEmSJEnKOovTkiRJkiRJkqSsK8h1AF1FQUFBzMuz1i9J6rjSdSmC71XtIhLxPEB7qnRdmhBCrsOQci6G7L4X1NWlyMvztSdJah+pVGpNjLF/e2/H4nQbycvLo6KiItdhSJLUovKy3jx/4A9yHUaXdPrCr7F69epchyHlRJ/Scn45cGKuw5By7nMVx7B6TfbeC8rKenHIIV/J2vYkSXuWKVO+sSgb2/ESH0mSJEmSJElS1lmcliRJkiRJkiRlnd16SJKkNnPhm+NYV7eJvBDoFoo4qccYji7bj/FrnuL+fW9qaPeZBb9gdd3Gt007tvt+RGDylnn8bMS1DfMunv1NhhX1f9u0aweczzm9j+ToN26gJBQRAnTP68ZZvY7ghkHvIz/4Pbykzuv7a69iydY3+cnAFygMRQDcveGrvFL1CAWhEIB++UM4rPgM3tP9E5Tm9QDg35V/5Z6NN1MUSgDokdeH/YuO5fzun2BQwcjcPBlJ2k3Tp99OXd0mDjroixQUlDZMnzXrLqqrV9Kjxyi2bFkMQDqdIgQIIR+APn0OZdiw83MSt6TWsTgtSZLa1I+Hf4Ljuu/Pqq0buG7hnRSEfBbUrGR93Wb6FHSnLqaYU/0WJXlF202bVrWQTw18DzFGxq9+mlRMkx/yWLO1grqY5s3qpdtNW1K7hiPLRjVs977RX2JYcX8W1qzkUwt+zt7F/bm078k5/E9I0u5bU7eM2bUTKQ09mFL9T47pdm7DvPO6f5RLenyerbGGpVtnc/+mH/HdtR/m5vI/UZyXFG5GFR7OTf1+TzqmWJNaxhNb7mHcmkv5Wvl9DC3cL1dPS5J2S1FRH9avf4P+/Y8FoKpqJTHWATB06PkUF/cGYNGihykq6sHgwWfmLFZJu8bLiSRJUrsYUNibk3qMYXXdRvYqLGfSlrkAzKpayj4lgziybNR209IxMqZkGAd125u6mGJ29TIAJlfO4+iy0QwvGrDdtKFF/ehf2Ott2x1RPJDDS0cxr3pFlp6pJLW9l6oeYlThYZxUehEvVj3UbJvCUMzIokO4vs8v2ZzewL+r/vq2NnkhnwEFe3Nlr1vZv+gYHt78i/YOXZLaXJ8+h7Bu3dSGx+vWvU6fPoe2atm6ui3Mm/d7pk37EdOm/YD58//I1q2bGuavXTuJ6dNvZ+rU7zFjxs/YsGFGw7w1ayYyc+YvMsv9gdrairZ7UpIAi9OSJKmdrKhdz4ubZrB/yVCOLBvF5Mp5QFJYPqJ0FIeX7rPdtENKR1CYV0BhXgEHlw5vKFxP2jKPw8tGcVjZPttNO6J0VLPbnV+9gimV89i/ZK8sPEtJah8vVf2N47tdwPHdLmR6zYtsTK1psW23vDLGFJ/I7NoJO1znUSVnMad2YluHKkntrqxsKOl0DdXVq4kxzYYNM+jb95BWLRtjpF+/oznooBsYM+ZzACxb9iQAdXVVLFv2FKNHX8mhh36F0aOvoaSkPwDr17/B6tX/YeTI/+Lgg/+bbt0GsXjx278ElPTOWJyWJElt6sbFv+H0GV/h4wtu58iyUXyk/1kcWTaaSVsyhegt8zi8bB+OKBu13bTGXXQcWTqayZl5UyrncUTpPhxROmq7aUeVjd5uu1fM+yFnzPgKX1j0a97X5wTe2+e4bDxdSWpzs2snsjb1Fsd0O48RhQfRP38Y/6n6xw6X6Z03gC3pje+4jSR1VPVXT2/aNJ/i4nIKC3u2arnCwu706rU/eXmF5OeXMHDgyWzevGi7NtXVq0mn6ygq6tlQnF67diIDB55CSUk5IeQzaNBpbN68hK1bN7f5c5P2ZPY5LUmS2tSP9v44x3Xff7tpR5SN4pvL7qMiVckblQv51tCrKM0vZm1dBRWpSl6vnM+Hy09vaH9k2SjuX/cCFalK1tdtZu/iAfQt6MnYpfdSkapkXvVyjijb/srpe0f9D8OK+2fjKUpSu3qp8iEOKj6RHnl9ADi+2/m8WPUwZ3e/psVlNqRXUpb39q6OGlvfijaS1FH17Xsoc+b8H7W1G+jb97BWL5dK1bBs2RNs2jSfVKoagHQ66a+6oKAbw4dfzOrVr7Bo0UN07z6cvfY6m+LivtTWbmTp0kdYuvTRhnWFkMfWrRUUFnZv2ycn7cEsTkuSpHY3tKgf/Qt68eC6lxhU1IfS/GIADuk2ggfXvURluoaDS4c3tD+kdASb09U8uO4lDivdB4Du+SX0K0zW0b+gF3sVlefkuUhSe6qN1bxW/Thp0nx+5SkA1MVaKmMFi7fOanaZ6vQWZtS8zAXdr93huidVP82+RUe1ecySlA1FRb0pKupNRcUchg27sNXLrVr1Ilu3VrDffh+nsLA7W7YsZc6cexrm9+q1H7167Uc6vZW33nqaJUseYfToKyks7MmQIWfRu/eB7fF0JGVYnJYkSVlxeNk+/GHNs5zb++jtpv129TMc2G1vSvKKGqaX5BUxpmQYf1jzLB8dcPa29qXJOo5tcmW2JHUVk6qfIY98vtH/YfBy4yQAABDdSURBVAoobJh+54Yv8FLVw9u13RprWVY3h/srfkRpXi9O7nbx29aXjinWppbz5JbxvFn7Gl8rv6/dn4MktZe9976QVKqa/PwiYky3aplUqpYQki496uoqWbnyhYZ5tbUVVFWtoEePkYSQT15eISEkPeD263c0K1e+QElJf0pK+lFXV8XmzQstVkttzOK0JEnKiiPLRvPExkkcnrkSGuDw0lH8NPUw7y09vtn2U6sWbtf+iLJR/HndC2/r0kOSuoqXqh7ipNKLKc8fst30M0sv5w8V32FM0Qk8tvl/eWrL74hE+uUP4bDi0zm/+ycpzittaD9v6xQ+veIoIpEeeb3Zv+hYbin/M0MKPX5K6ryKi/vu8jIDBpzAokUP8sYbP6SwsCf9+x9LRcXczNzIqlUvsmjRg0CgtHQww4adD0CfPgeTTm9l4cL7qa3dSH5+N3r2HG1xWmpjIcaY6xi6hKKiolhRUZHrMCRJalF5WW+eH/ODXIfRJZ2+8GusXr0612FIOdGntJxfDpqY6zCknPtcxTGsXpO994Kysl4ceuhXsrY9SdKeZcqUb0yMMR6985bvTF57b0CSJEmSJEmSpKYsTkuSJEmSJEmSss7itCRJkiRJkiQp6yxOS5IkSZIkSZKyzuK0JEmSJEmSJCnrQowx1zF0CSGETcCbuY5DXUI/YE2ug1CnZx6prZhLaivmktqKuaS2Yi6prZhLagvmkTqa/WOMPdp7IwXtvYE9yJsxxqNzHYQ6vxDCBHNJ75R5pLZiLqmtmEtqK+aS2oq5pLZiLqktmEfqaEIIE7KxHbv1kCRJkiRJkiRlncVpSZIkSZIkSVLWWZxuO7/OdQDqMswltQXzSG3FXFJbMZfUVswltRVzSW3FXFJbMI/U0WQlJ70hoiRJkiRJkiQp67xyWpIkSZIkSZKUdRanJUmSJEmSJElZZ3H6HQghDA0h/G8I4a0QQk0IYWEI4achhD65jk3ZF0K4NITw8xDCCyGEihBCDCHcu5NlTgwhPBpCWBdCqAwhTA0hfD6EkL+DZS4IITwXQtgYQtgcQvhPCOHqtn9GypUQQnkI4eMhhL+GEOaGEKoy+/vfIYSPhRCaPXabT2pOCOH7IYRnQghLMrm0LoQwOYRwawihvIVlzCXtVAjhysx7XQwhfLyFNrucFyGEq0MIr2bab8wsf0H7PAtlW+Z8ObYwrGhhGY9JalEI4ZQQwl9CCMszn8mWhxCeDCG8p5m25pK2E0K4ZgfHpPoh1cxy5pKaFUI4P3MMWpo5954fQrg/hHBCC+3NJbWLkPhoCOGVEMKmTH5NDiF8rml+hRAKQwg3hBDuCSFMCSHU7ugcP7PMSSGEH4QQXgshrM68By8IIfwmhDB6l2K1z+ndE0IYBbwEDAAeBmYBxwJnAG8CJ8UY1+YuQmVbCGEKcBiwGVgKHAD8PsZ4RQvt3wf8BagG/gSsAy4E9gceiDF+oJllrgN+DqzNLFMLXAoMBW6LMd7Yxk9LORBCuBa4E1gOPAssBgYClwC9SPLmA7HRAdx8UktCCLXAJGAGsAooA44HjgbeAo6PMS5p1N5c0k6FEIYB04B8oDvwiRjjb5q02eW8CCH8CPhvkvfRB4Ai4L+AvsD1McY72us5KTtCCAuB3sBPm5m9Ocb4oybtPSapRSGEm4FvAmuAf5CcO/UDjgCejTF+qVFbc0lvE0I4HLiohdmnAGcCj8QYL2i0jLmkZoUQvg98iWQ/P0RybBoNvBcoAK6KMd7bqL25pHYTQvgtcCXJZ8C/A1uAdwNjaFJTCCH0BtZnFl1JklfDaOYcv9H6VwD9SWqjE4E64ATgxMy2zooxvtyqYGOMDrsxAE8AkeSDUuPpP85M/1WuY3TIek6cAewLBOD0TB7c20LbnpkDRA1wdKPpJZkXdgT+q8kyI0jetNYCIxpN7wPMzSxzQq7/Dw5tkktnkpyU5DWZPoikUB2B95tPDq3Mp5IWpn87s59/aS457MqQeZ97GpgH/DCzjz/+TvOC5EQ2Zub3abKutZn1jWiv5+WQtfxZCCxsZVuPSQ47yo8PZPbnU0CPZuYXmksO72QAXs7s5/eaSw6tyJdBQApYAQxoMu+MzH6eby45ZGMg+dItAvOBfo2mFwJ/zcy7ptH0IuA8YHDm8ViaOcdvso0vA0OamX5TZtlprY3Xbj12QwhhH+BskpPrXzSZfSvJNwRXhhDKshyacijG+GyMcU7MvBp34lKSb5j+GGOc0Ggd1cDNmYefbrLMR4Fi4I4Y48JGy6wHvpN5eO1uhq8OJMb4zxjj32OM6SbTVwC/yjw8vdEs80ktyuRBc/6cGe/baJq5pNb4HMmXaB8hOedpzu7kRf3jb2fa1S+zkOR8qzizTe05PCapWSHp4uz7QCXw4RjjpqZtYoxbGz00l7RLQggHk/zSbBnwSKNZ5pJaMpyk69z/xBhXNZ4RY3wW2ESSO/XMJbWnSzLj22KMa+onZt4bb8k8vL7R9NoY42MxxuWt3UCM8fsxxreamfV9oAo4OLTQjWRTFqd3z5mZ8ZPNFI82AS8CpSRvZlJz6nPo8WbmPU9yon1iCKG4lcs81qSNuq76D1p1jaaZT9odF2bGUxtNM5e0QyGEA4HvAbfHGJ/fQdPdyQtzac9RHEK4IoRwU6Z/wzNa6FvTY5JaciIwEngUWJ/p4/XLmXxqrl9Xc0m76lOZ8d0xxsZ9TptLaskckq4Qjg0h9Gs8I4RwKtCD5Jdn9cwltadBmfH8ZubVTzsy051HW4tsq1e8rc/+5lic3j37Z8azW5g/JzPeLwuxqHNqMYdijHXAApI+qfZp5TLLSa5eGxpCKG3bUNVRhBAKgKsyDxufkJhP2qkQwo0hhLEhhJ+EEF4g6aNzKkmhsZ65pBZljkG/I+le6KadNN+lvMj82mwvkj6Hm7tiw3OrrmUQSS59m6Tv6X8Cc0IIpzVp5zFJLTkmM15Jcl+Ff5C8n/0UeCmE8K8QQuMrFM0ltVoIoRtwBZAGmva1ai6pWTHGdSTdHAwEZoQQfh1C+G4I4c/AkyRdEH2q0SLmktpT/dXSI5uZ1zinDmiHbX+A5MuYV2KMG1qzgMXp3dMrM97Ywvz66e3xDYS6ht3JodYu06uF+er8vgccDDwaY3yi0XTzSa1xI0nXU58HTib5guPsGOPqRm3MJe3I10luMnZNjLFqJ213NS88t9pz3AO8i6RAXQYcAtxF0o/mYyGEwxq19ZiklgzIjK8FupHc4KkHyXnSE8CpwP2N2ptL2hUfJMmFx2Kjm0ZnmEtqUYzxpyTdKRQAnwC+QlKoWwKMb9Ldh7mk9vSPzPiLIYS+9RMzF5uMa9SuT1tuNIQwkuSGnXUkNzlvFYvT7SNkxq3pe1hqzu7kkHnXhYUQPkdycJ9FcsfdXVo8Mzaf9mAxxkExxkBSELqE5BvzySGEI3dhNebSHiqEcCzJ1dK3xdbedXsnq8yMdzUvzKNOLsY4LnNvhZUxxsoY4xsxxmtJbirejeQGPK3lMWnPVd8NTAAujTE+E2PcHGOcDlwMLAVOa6GLj+aYS2rsk5nxXbuxrLm0BwshfAl4ABgPjCL5EvYokm4Ufh9C+MGurC4zNpe0O/5I0tXLKLZdyf9TYArwHrb9KrFV3W60RghhQGab/YEbYowvtXZZi9O7Z2ffRvVs0k5qandyqLXLVLyDuNQBhRA+C9wOzADOyPxkrDHzSa2WKQj9leTGvuXAbxvNNpf0No2685jNthuo7Myu5sXO2u/sSiF1fvU3/D210TSPSWpJ/U1T58cYX288I/PLjvpfmB2bGZtLapUQwhiSPs2XkvRp3pS5pGaFEE4nuRHc32KMX4wxzs98CTuJ5EuzZcB/hxDqu1Qwl9RuMvfHey/Jr2dXkFzg9lGSY9vJwNpM01XNrmAXZQrT/yTpeuaGGOMvd2V5i9O7583MuKV+D/fNjFvqk1pqMYcyRYCRJD+DmN/KZQaTfCu7NMZY2bahKpdCCJ8H7gDeIClMr2immfmkXRZjXETyhcdBjW7aYi6pOd1J9u+BQHUIIdYPJF3FAPy/zLSfZh7vUl7EGLeQfGjrnpnflOdWXV/9h6OyRtM8Jqkl9fu5pb4s64vX3Zq0N5e0My3dCLGeuaSWXJAZP9t0RmbfvkpSgzsiM9lcUruKMdbFGG+LMR4eY+wWY+wZYzyX5DPg4UAVMP2dbieTe88BY4DPxhh/tqvrsDi9e+oPNmeHELb7H4YQegAnkezkV7IdmDqNf2bG5zYz71SgFHgpxljTymXOa9JGXUAI4cvAT0h+enNGkz7KGjOftLuGZMb1H77MJTWnBri7hWFyps2/M4/ru/zYnbwwl/Zs9d0vNP4Q7jFJLXmepGizbwihqJn5B2fGCzNjc0k7FUIoIbm6ME3yntYcc0ktKc6M+7cwv356bWZsLilXrgRKgD/HGLe+kxWFEIYC/yK5seK1u3rFdIMYo8NuDCQ/FYvA9U2m/zgz/Ve5jtEhp/lxeiYP7m1hfk9gNckH/qMbTS8BXsos+19NlhkJVJP8/GJEo+l9gLmZZU7I9XN3aLMcuiWzTycAfXfS1nxyaCk3DgAGNTM9D/h2Zj+/aC457O5A0j9wBD7+TvOC5GfUMTO/T6PpIzLrqW68LofONwAHNfeeBgwn6fswAjc1mu4xyWFH+XRvZn9+q8n0s0iKixuA3uaSwy7k1JWZ/fr3HbQxlxxayo0PZvblCmCvJvPOyxyXqoByc8khGwPQs5lpxwDrgE3APjtYdizNnOM3abM3MI/kQqePvJNYQ2aF2kUhhFEkB4wBwMPATOA44AySn5yeGGNc2/Ia1NWEEC4CLso8HAScQ3L1zwuZaWtijDc2af8AyZvLH0kOEO8l6aPnAeCDsckLNIRwPfAzkjejP5F863opMJTkJlU3ok4vhHA1yU00UiR3um2uj9WFMcbxjZYxn/Q2mW5hfkhyhdk8kn09EDiN5IaIK4B3xRhnNFrGXFKrhRDGknTt8YkY42+azNvlvAgh3AZ8kaQ/vAeAIuBDJP2jXx9jvKPdnozaXSZfvkLyK8QFJB+MRgHnk3wYfxS4OMZY22gZj0lqVqZ/yxeB0STn26+SfNFxMckH6g/HGO9v1N5c0g6FEF4g6Yv1vTHGv++gnbmkt8n8qv4J4N0k729/JTnXPpCky48AfD7GeHujZcwltZsQwn9IvhB5gyQnDyK5GWINcEmM8Ykm7b9CcnETJN1+HEZS96y/eeK/G5/vhxAWkFxEMhH4RwthjI8xLtxpsLmu5HfmARgG3AMsJzkgLCK5adkOr3J06JoD275ZamlY2MwyJ5F8EFufOWhMA74A5O9gOxeS/GxiE7AFeA24OtfP3yGruRSB58wnh1bk0sHAL0i6hllD8hPojZn9PLal9ytzyaG1Azu5qmJ38gK4OtNuS2a5fwEX5Pq5OrRJvpwG3AfMIrmqdSvJVWNPAVdBcuFMM8t5THJoaT/3Jfnl6gKSz2NrSS4cOt5cctjFXDow8362ZEf50Ki9ueTQ3D4uBD5P0sVrRebcexVJ4e5sc8khmwPwPySF4w0kBekFJDegHtFC++fYcQ1ifJP2O6tZROD01sTqldOSJEmSJEmSpKzzhoiSJEmSJEmSpKyzOC1JkiRJkiRJyjqL05IkSZIkSZKkrLM4LUmSJEmSJEnKOovTkiRJkiRJkqSsszgtSZIkSZIkSco6i9OSJEmSJEmSpKyzOC1JkiRJkiRJyjqL05IkSZIkSZKkrLM4LUmSJEmSJEnKuv8PWhcdnN0uu2sAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 1728x576 with 3 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Plotting the predicted fitness effect for the variants seen more than once\n",
    "plt.close('all')\n",
    "f, (ax1, ax2) = plt.subplots(2, 1, sharey=True, sharex = True, figsize=(24, 8))\n",
    "gs = matplotlib.gridspec.GridSpec(3, 1, width_ratios=[1], height_ratios=[10,0.5, 0.5])\n",
    "ax1 = plt.subplot(gs[0])\n",
    "ax2 = plt.subplot(gs[1])\n",
    "ax3 = plt.subplot(gs[2])\n",
    "gs.update(hspace=0.05)\n",
    "\n",
    "nonsense_list = []\n",
    "y_list = []\n",
    "x_list=[]\n",
    "y_list_top20 = []\n",
    "x_list_top20=[]\n",
    "y_list_nonsense = []\n",
    "x_list_nonsense = []\n",
    "x=[]\n",
    "y=[]\n",
    "OBS_EXP_DICT={}\n",
    "top20_lower_err = []\n",
    "top20_upper_err = []\n",
    "missense_lower_err = []\n",
    "missense_upper_err = []\n",
    "nonsense_lower_err = []\n",
    "nonsense_upper_err = []\n",
    "\n",
    "top20 = c1\n",
    "missense = c6\n",
    "nonsense = c10\n",
    "synonymous = c3\n",
    "\n",
    "axisfont=20\n",
    "domainfont = 12\n",
    "titlefont=12\n",
    "axislabelfont=12\n",
    "\n",
    "top_20_DNMT3A = ['p.R882C', 'p.R729W', 'p.R326C', 'p.R320*', 'p.R882H', 'p.R736H', 'p.Y735C',\\\n",
    "                'p.R736C', 'p.W860R', 'p.R771*', 'p.R598*', 'p.P904L']\n",
    "\n",
    "for k, v in DNMT3A_missense_s.items():\n",
    "    if k in top_20_DNMT3A:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_top20.append(codon_number)\n",
    "        y_list_top20.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        top20_lower_err.append(error_lower)\n",
    "        top20_upper_err.append(error_upper)\n",
    "\n",
    "    if k not in top_20_DNMT3A:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list.append(codon_number)\n",
    "        y_list.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        missense_lower_err.append(error_lower)\n",
    "        missense_upper_err.append(error_upper)\n",
    "\n",
    "    missense_err = [tuple(missense_lower_err), tuple(missense_upper_err)]   \n",
    "    \n",
    "for k, v in DNMT3A_nonsense_s.items():\n",
    "    if k in top_20_DNMT3A:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_top20.append(codon_number)\n",
    "        y_list_top20.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        top20_lower_err.append(error_lower)\n",
    "        top20_upper_err.append(error_upper)\n",
    "    \n",
    "    top20_err = [tuple(top20_lower_err), tuple(top20_upper_err)]\n",
    "        \n",
    "    if k not in top_20_DNMT3A:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_nonsense.append(codon_number)\n",
    "        y_list_nonsense.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        nonsense_lower_err.append(error_lower)\n",
    "        nonsense_upper_err.append(error_upper)\n",
    "\n",
    "    nonsense_err = [tuple(nonsense_lower_err), tuple(nonsense_upper_err)]        \n",
    "    \n",
    "msize = 12\n",
    "medgewidth = 2\n",
    "linew = 3\n",
    "csize = 4\n",
    "cthick = 3\n",
    "\n",
    "markers1, caps1, bars1 = ax1.errorbar(x_list, y_list, yerr = missense_err, fmt = 'o', ecolor = missense, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = missense, markerfacecolor = missense, zorder = 1)    \n",
    "\n",
    "markers2, caps2, bars2 = ax1.errorbar(x_list_nonsense, y_list_nonsense, yerr = nonsense_err, fmt = 'o', ecolor = nonsense, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = nonsense, markerfacecolor = nonsense, zorder = 2)    \n",
    "\n",
    "markers3, caps3, bars3 = ax1.errorbar(x_list_top20, y_list_top20, yerr = top20_err, fmt = 'o', ecolor = top20, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = top20, markerfacecolor = top20, zorder = 3)    \n",
    "\n",
    "# loop through bars and caps and set the alpha value\n",
    "a = 0.7\n",
    "[bar.set_alpha(a) for bar in bars1]\n",
    "[cap.set_alpha(a) for cap in caps1]\n",
    "[bar.set_alpha(a) for bar in bars2]\n",
    "[cap.set_alpha(a) for cap in caps2]\n",
    "[bar.set_alpha(a) for bar in bars3]\n",
    "[cap.set_alpha(a) for cap in caps3]\n",
    "\n",
    "#Only show the required axis lines\n",
    "ax1.spines['top'].set_visible(False)\n",
    "ax1.spines['right'].set_visible(False)\n",
    "\n",
    "for axis in ['bottom','left']:\n",
    "    ax1.spines[axis].set_linewidth(2)\n",
    "\n",
    "for axis in ['bottom','left']:\n",
    "    ax1.spines[axis].set_color(grey3)\n",
    "    \n",
    "ax1.set_ylabel('fitness effect (s)', fontsize = axisfont, labelpad = 10)\n",
    "ax1.set_title('Fitness effects across DNMT3A', fontsize = 18)\n",
    "\n",
    "# ax1.set_yscale('log')\n",
    "\n",
    "# STACKED BAR CHART\n",
    "# define the exons + create data\n",
    "exons = [{'2':(24-1)+1, '3': (59-25)+1, '4': (150-60)+1, '5':(164-151)+1, '6':(213-165)+1, '7': (285-214)+1, '8': (338-286)+1,\\\n",
    "          '9': (374-339)+1, '10': (427-375)+1, '11': (477-428)+1, '12': (492-478)+1, '13': (518-493)+1, '14':(556-519)+1, \\\n",
    "          '15': (617-557)+1, '16': (646-618)+1, '17': (694-647)+1, '18': (725-695)+1, '19': (774-726)+1, '20': (803-775)+1, \\\n",
    "          '21': (826-804)+1, '22': (866-827)+1, '23': (912-867)+1}]\n",
    "exon_orders = [['2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12', '13', '14', '15', '16', '17', '18', '19', '20', '21', '22', '23']]\n",
    "colors2 = [c4, c0, c3, c1, c5, c6,c7, c8, c9, c10, c2,c4, c0, c3, c1, c5, c6,c7, c8, c9, c10, c2]\n",
    "names = exons[0].keys()\n",
    "values = np.array([[data[name] for name in order] for data,order in zip(exons, exon_orders)])\n",
    "lefts = np.insert(np.cumsum(values, axis=1),0,0, axis=1)[:, :-1]\n",
    "orders = np.array(exon_orders)\n",
    "\n",
    "#plot the bar chart\n",
    "for name, color in zip(names, colors2):\n",
    "    idx = np.where(orders == name)\n",
    "    value = values[idx]\n",
    "    left = lefts[idx]\n",
    "    ax2.bar(x=left, height=-10, width=value, bottom=-5, color=color, orientation = 'horizontal', label = name, edgecolor='k', linewidth=1, tick_label = name)\n",
    "    ax2.text(left + (value/2.), (-0.26), name, ha='center', va='center', fontsize = axislabelfont)\n",
    "    \n",
    "\n",
    "# STACKED BAR CHART of domains    \n",
    "domains = [{' ': (291-1)+1, 'PWWP' :(350-292)+1, '  ':(481-351)+1, 'ADD': (614-482)+1, '   ': (633-615)+1, 'MTase': (912-634)+1}]\n",
    "domain_orders = [[' ','PWWP','  ','ADD','   ','MTase']]\n",
    "# blank_col = c13\n",
    "blank_col = grey1\n",
    "colors3 = [blank_col, PWWP_col, blank_col, ADD_col, blank_col, MTase_col]\n",
    "names_domains = domains[0].keys()\n",
    "values = np.array([[data[name] for name in order] for data,order in zip(domains, domain_orders)])\n",
    "lefts = np.insert(np.cumsum(values, axis=1),0,0, axis=1)[:, :-1]\n",
    "orders = np.array(domain_orders)\n",
    "\n",
    "#plot the bar chart\n",
    "for name, color in zip(names_domains, colors3):\n",
    "    idx = np.where(orders == name)\n",
    "    value = values[idx]\n",
    "    left = lefts[idx]\n",
    "    ax3.bar(x=left, height=-10, width=value, bottom=-5, color=color, orientation = 'horizontal', alpha = 0.9, label = name, edgecolor='k', linewidth=1)\n",
    "    ax3.text(left + (value/2.), (-0.25), name, ha='center', va='center', fontsize = domainfont, color = 'k', fontweight = 530)\n",
    "\n",
    "    \n",
    "\n",
    "# CONFIGURING THE GRAPH APPEARANCE\n",
    "#Set the x and y axis limits\n",
    "ax1.set_xlim(0, 912)\n",
    "ax1.set_ylim(180, 220)\n",
    "ax2.set_xlim(0, 912)\n",
    "ax2.set_ylim(-0.5, 0)\n",
    "ax3.set_xlim(0, 912)\n",
    "ax3.set_ylim(-0.5, 0)\n",
    "\n",
    "#x-axis ticks\n",
    "x1_major_ticks = []\n",
    "x1_major_tick_labels = []\n",
    "x2_major_ticks = [0, 100, 200, 300, 400, 500, 600, 700, 800, 912]\n",
    "x2_major_tick_labels = [\"0\", \"100\", \"200\", \"300\", \"400\", \"500\", \"600\", \"700\", \"800\", \"912\"]\n",
    "ax3.set_xticks(x2_major_ticks)\n",
    "ax3.set_xticklabels(x2_major_tick_labels, fontsize = axisfont)\n",
    "ax1.set_yticklabels([])\n",
    "ax1.set_yticks([])\n",
    "ax1.set_xticklabels([])\n",
    "ax1.set_xticks([])\n",
    "ax2.set_yticklabels([])\n",
    "ax2.set_yticks([])\n",
    "ax2.set_xticklabels([])\n",
    "ax2.set_xticks([])\n",
    "ax3.set_yticklabels([])\n",
    "ax3.set_yticks([])\n",
    "\n",
    "#y-axis ticks\n",
    "y1_major_ticks = [180, 185, 190, 195, 200, 205, 210, 215, 220]\n",
    "y1_major_tick_labels = [\"180\", \"185\", \"190\", \"195\", \"200\", \"205\", \"210\", \"215\", \"220\"]\n",
    "ax1.set_yticks(y1_major_ticks)\n",
    "ax1.set_yticklabels(y1_major_tick_labels, fontsize = axisfont)\n",
    "\n",
    "#set panel legends\n",
    "legend_marker_size = 2\n",
    "legend_elements = [Line2D([0], [0], marker = '_', color = missense, alpha = 1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='missense'),\n",
    "                   Line2D([0], [0], marker = '_', color=nonsense, alpha=1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='nonsense'),\n",
    "                  Line2D([0], [0], marker = '_', color=top20, alpha=1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='top 20 observed variant')]\n",
    "    \n",
    "ax1.legend(ncol=1, handles=legend_elements, loc='upper left', frameon=0, fontsize = 20)\n",
    "\n",
    "ax1.grid(which='major', linestyle=':', linewidth='0.7', color=grey3, zorder = 0)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TET2 variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {
    "code_folding": [
     0,
     91,
     102,
     114
    ]
   },
   "outputs": [],
   "source": [
    "def expected_number_variants_missense_TET2(data, params):\n",
    "    s_value = params[0]\n",
    "    \n",
    "    a = 0\n",
    "    expected_number_by_variant_missense = {}\n",
    "    \n",
    "    Jaiswal_fmin = 0.056\n",
    "    Jaiswal_mean = 58\n",
    "    Jaiswal_sigma = 11.5\n",
    "    Jaiswal_total = 17182\n",
    "    \n",
    "    Genovese_fmin = 0.1194\n",
    "    Genovese_mean = 55.1\n",
    "    Genovese_total = 12380\n",
    "    Genovese_sigma = 11.95\n",
    "    \n",
    "    McKerrel_fmin = 0.008\n",
    "    McKerrel_mean = 51.88\n",
    "    McKerrel_total = 4219\n",
    "    McKerrel_sigma = 18.36\n",
    "    \n",
    "    Zink_fmin = 0.2363\n",
    "    Zink_mean = 60\n",
    "    Zink_total = 11262\n",
    "    Zink_sigma = 1.18907720e+01\n",
    "    \n",
    "    Coombs_fmin = 0.019\n",
    "    Coombs_mean = 59.82\n",
    "    Coombs_total = 5649\n",
    "    Coombs_sigma = 15.14\n",
    "    \n",
    "    Young_fmin = 0.0009\n",
    "    Young_mean = 63\n",
    "    Young_total = 89\n",
    "    Young_sigma = 6.8\n",
    "    \n",
    "    Desai_fmin = 0.0169\n",
    "    Desai_mean = 66.1\n",
    "    Desai_total = 424\n",
    "    Desai_sigma = 6 #estimated\n",
    "    \n",
    "    Acuna_fmin = 0.0015\n",
    "    Acuna_mean = 4.5\n",
    "    Acuna_total = 2014\n",
    "    \n",
    "    N = 9.40166610e+04\n",
    "\n",
    "    for i in data:\n",
    "        variant_mutation_rate = mutation_rates_variants_TET2(i)\n",
    "\n",
    "        if i in Jaiswal_missense_TET2:\n",
    "            Jaiswal_number = prevalence_averaged_over_ages(Jaiswal_fmin, s_value, N, variant_mutation_rate, Jaiswal_mean, Jaiswal_sigma, Jaiswal_total)\n",
    "        else:\n",
    "            Jaiswal_number = 0\n",
    "\n",
    "        if i in Genovese_missense_TET2:\n",
    "            Genovese_number = prevalence_averaged_over_ages(Genovese_fmin, s_value, N, variant_mutation_rate, Genovese_mean, Genovese_sigma, Genovese_total)\n",
    "        else:\n",
    "            Genovese_number = 0 \n",
    "\n",
    "        if i in Zink_missense_TET2:\n",
    "            Zink_number = prevalence_averaged_over_ages(Zink_fmin, s_value, N, variant_mutation_rate, Zink_mean, Zink_sigma, Zink_total)\n",
    "        else:\n",
    "            Zink_number = 0\n",
    "\n",
    "        if i in Coombs_missense_TET2:\n",
    "            Coombs_number = prevalence_averaged_over_ages(Coombs_fmin, s_value, N, variant_mutation_rate, Coombs_mean, Coombs_sigma, Coombs_total)\n",
    "        else:\n",
    "            Coombs_number = 0\n",
    "\n",
    "        if i in Young_missense_TET2:\n",
    "            Young_number = prevalence_averaged_over_ages(Young_fmin, s_value, N, variant_mutation_rate, Young_mean, Young_sigma, Young_total)\n",
    "        else:\n",
    "            Young_number = 0\n",
    "\n",
    "        if i in Desai_missense_TET2:\n",
    "            Desai_number = prevalence_averaged_over_ages(Desai_fmin, s_value, N, variant_mutation_rate, Desai_mean, Desai_sigma, Desai_total)\n",
    "        else:\n",
    "            Desai_number = 0\n",
    "\n",
    "        if i in Acuna_missense_TET2:\n",
    "            Acuna_number = prevalence_averaged_over_uniform_ages(Acuna_fmin, s_value, N, variant_mutation_rate, Acuna_mean, Acuna_total)\n",
    "        else:\n",
    "            Acuna_number = 0\n",
    "\n",
    "        expected_total = Jaiswal_number + Genovese_number + Zink_number + Coombs_number + Young_number + Desai_number + Acuna_number\n",
    "\n",
    "        expected_number_by_variant_missense[i] = expected_total\n",
    "    \n",
    "    return expected_number_by_variant_missense\n",
    "\n",
    "def prob_observed_given_expected_missense_TET2(params, data): #data = list of variant (e.g. p.R882H) (or all_missense_variants_TET2)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_TET2_counts[variant])\n",
    "        expected_number = float(expected_number_variants_missense_TET2(data, params)[variant])\n",
    "        square_distance = ((observed_number - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance\n",
    "\n",
    "def prob_observed_given_expected_missense_TET2_upper_error(params, data): #data = list of variant (e.g. p.R882H) (or all_missense_variants_TET2)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_TET2_counts[variant])\n",
    "        observed_number_upper_error = observed_number+np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_missense_TET2(data, params)[variant])\n",
    "        square_distance = ((observed_number_upper_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   \n",
    "\n",
    "def prob_observed_given_expected_missense_TET2_lower_error(params, data): #data = list of variant (e.g. p.R882H) (or all_missense_variants_TET2)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_TET2_counts[variant])\n",
    "        observed_number_lower_error = observed_number-np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_missense_TET2(data, params)[variant])\n",
    "        square_distance = ((observed_number_lower_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {
    "code_folding": [
     0,
     91,
     102,
     114
    ]
   },
   "outputs": [],
   "source": [
    "def expected_number_variants_nonsense_TET2(data, params):\n",
    "    s_value = params[0]\n",
    "    \n",
    "    a = 0\n",
    "    expected_number_by_variant_nonsense = {}\n",
    "    \n",
    "    Jaiswal_fmin = 0.056\n",
    "    Jaiswal_mean = 58\n",
    "    Jaiswal_sigma = 11.5\n",
    "    Jaiswal_total = 17182\n",
    "    \n",
    "    Genovese_fmin = 0.1194\n",
    "    Genovese_mean = 55.1\n",
    "    Genovese_total = 12380\n",
    "    Genovese_sigma = 11.95\n",
    "    \n",
    "    McKerrel_fmin = 0.008\n",
    "    McKerrel_mean = 51.88\n",
    "    McKerrel_total = 4219\n",
    "    McKerrel_sigma = 18.36\n",
    "    \n",
    "    Zink_fmin = 0.2363\n",
    "    Zink_mean = 60\n",
    "    Zink_total = 11262\n",
    "    Zink_sigma = 1.18907720e+01\n",
    "    \n",
    "    Coombs_fmin = 0.019\n",
    "    Coombs_mean = 59.82\n",
    "    Coombs_total = 5649\n",
    "    Coombs_sigma = 15.14\n",
    "    \n",
    "    Young_fmin = 0.0009\n",
    "    Young_mean = 63\n",
    "    Young_total = 89\n",
    "    Young_sigma = 6.8\n",
    "    \n",
    "    Desai_fmin = 0.0169\n",
    "    Desai_mean = 66.1\n",
    "    Desai_total = 424\n",
    "    Desai_sigma = 6 #estimated\n",
    "    \n",
    "    Acuna_fmin = 0.0015\n",
    "    Acuna_mean = 4.5\n",
    "    Acuna_total = 2014\n",
    "    \n",
    "    N = 9.40166610e+04\n",
    "\n",
    "    for i in data:\n",
    "        variant_mutation_rate = mutation_rates_variants_TET2(i)\n",
    "\n",
    "        if i in Jaiswal_nonsense_TET2:\n",
    "            Jaiswal_number = prevalence_averaged_over_ages(Jaiswal_fmin, s_value, N, variant_mutation_rate, Jaiswal_mean, Jaiswal_sigma, Jaiswal_total)\n",
    "        else:\n",
    "            Jaiswal_number = 0\n",
    "\n",
    "        if i in Genovese_nonsense_TET2:\n",
    "            Genovese_number = prevalence_averaged_over_ages(Genovese_fmin, s_value, N, variant_mutation_rate, Genovese_mean, Genovese_sigma, Genovese_total)\n",
    "        else:\n",
    "            Genovese_number = 0 \n",
    "\n",
    "        if i in Zink_nonsense_TET2:\n",
    "            Zink_number = prevalence_averaged_over_ages(Zink_fmin, s_value, N, variant_mutation_rate, Zink_mean, Zink_sigma, Zink_total)\n",
    "        else:\n",
    "            Zink_number = 0\n",
    "\n",
    "        if i in Coombs_nonsense_TET2:\n",
    "            Coombs_number = prevalence_averaged_over_ages(Coombs_fmin, s_value, N, variant_mutation_rate, Coombs_mean, Coombs_sigma, Coombs_total)\n",
    "        else:\n",
    "            Coombs_number = 0\n",
    "\n",
    "        if i in Young_nonsense_TET2:\n",
    "            Young_number = prevalence_averaged_over_ages(Young_fmin, s_value, N, variant_mutation_rate, Young_mean, Young_sigma, Young_total)\n",
    "        else:\n",
    "            Young_number = 0\n",
    "\n",
    "        if i in Desai_nonsense_TET2:\n",
    "            Desai_number = prevalence_averaged_over_ages(Desai_fmin, s_value, N, variant_mutation_rate, Desai_mean, Desai_sigma, Desai_total)\n",
    "        else:\n",
    "            Desai_number = 0\n",
    "\n",
    "        if i in Acuna_nonsense_TET2:\n",
    "            Acuna_number = prevalence_averaged_over_uniform_ages(Acuna_fmin, s_value, N, variant_mutation_rate, Acuna_mean, Acuna_total)\n",
    "        else:\n",
    "            Acuna_number = 0\n",
    "\n",
    "        expected_total = Jaiswal_number + Genovese_number + Zink_number + Coombs_number + Young_number + Desai_number + Acuna_number\n",
    "\n",
    "        expected_number_by_variant_nonsense[i] = expected_total\n",
    "    \n",
    "    return expected_number_by_variant_nonsense\n",
    "\n",
    "def prob_observed_given_expected_nonsense_TET2(params, data): #data = list of variant (e.g. p.R882H) (or all_nonsense_variants_TET2)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_TET2_counts[variant])\n",
    "        expected_number = float(expected_number_variants_nonsense_TET2(data, params)[variant])\n",
    "        square_distance = ((observed_number - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance\n",
    "\n",
    "def prob_observed_given_expected_nonsense_TET2_upper_error(params, data): #data = list of variant (e.g. p.R882H) (or all_nonsense_variants_TET2)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_TET2_counts[variant])\n",
    "        observed_number_upper_error = observed_number+np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_nonsense_TET2(data, params)[variant])\n",
    "        square_distance = ((observed_number_upper_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   \n",
    "\n",
    "def prob_observed_given_expected_nonsense_TET2_lower_error(params, data): #data = list of variant (e.g. p.R882H) (or all_nonsense_variants_TET2)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_TET2_counts[variant])\n",
    "        observed_number_lower_error = observed_number-np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_nonsense_TET2(data, params)[variant])\n",
    "        square_distance = ((observed_number_lower_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#CALCULATING S FOR MISSENSE VARIANTS\n",
    "TET2_missense_s = {}\n",
    "a = 0\n",
    "for var in all_missense_variants_TET2:\n",
    "    initial_guess=[0.05]\n",
    "    if var in all_studies_TET2_counts.keys():\n",
    "        observed_number = float(all_studies_TET2_counts[var])\n",
    "#         print(observed_number)\n",
    "        if observed_number >1:\n",
    "            variant_mutation_rate = mutation_rates_variants_TET2(var)\n",
    "            \n",
    "            MLE = scipy.optimize.minimize(prob_observed_given_expected_missense_TET2, initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value = MLE.x[0]\n",
    "\n",
    "            upper_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_missense_TET2_upper_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_upper = upper_error_MLE.x[0]\n",
    "\n",
    "            lower_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_missense_TET2_lower_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_lower = lower_error_MLE.x[0]\n",
    "\n",
    "#             print('a = ', a)\n",
    "            TET2_missense_s[var]=(s_value, observed_number, variant_mutation_rate, s_value_upper, s_value_lower) \n",
    "    a = a+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/anaconda3/lib/python3.7/site-packages/scipy/integrate/quadpack.py:385: IntegrationWarning: The integral is probably divergent, or slowly convergent.\n",
      "  warnings.warn(msg, IntegrationWarning)\n"
     ]
    }
   ],
   "source": [
    "#CALCULATING S FOR NONSENSE VARIANTS\n",
    "TET2_nonsense_s = {}\n",
    "a = 0\n",
    "for var in all_nonsense_variants_TET2:\n",
    "    initial_guess=[0.05]\n",
    "    if var in all_studies_TET2_counts.keys():\n",
    "        observed_number = float(all_studies_TET2_counts[var])\n",
    "#         print(observed_number)\n",
    "        if observed_number >1:\n",
    "            variant_mutation_rate = mutation_rates_variants_TET2(var)\n",
    "            \n",
    "            MLE = scipy.optimize.minimize(prob_observed_given_expected_nonsense_TET2, initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value = MLE.x[0]\n",
    "\n",
    "            upper_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_nonsense_TET2_upper_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_upper = upper_error_MLE.x[0]\n",
    "\n",
    "            lower_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_nonsense_TET2_lower_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_lower = lower_error_MLE.x[0]\n",
    "\n",
    "#             print('a = ', a)\n",
    "            TET2_nonsense_s[var]=(s_value, observed_number, variant_mutation_rate, s_value_upper, s_value_lower) \n",
    "    a = a+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#Exac - retrieve the frequencies in Exac for each of the variants\n",
    "import csv\n",
    "filename=\"Data files/TET2_Exac.csv\"\n",
    "\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    Exac_TET2_results={} \n",
    "    \n",
    "    for row in read_reader:\n",
    "#         print(row)\n",
    "        if row_count>1:\n",
    "            variant = row[6]\n",
    "            allele_freq = float(row[14])\n",
    "            african_freq = float(row[15])/float(row[16])\n",
    "            east_asian_freq = float(row[18])/float(row[19])\n",
    "            european_freq = float(row[21])/float(row[22])\n",
    "            finnish_freq = float(row[24])/float(row[25])\n",
    "            latino_freq = float(row[27])/float(row[28])\n",
    "            south_asian_freq = float(row[33])/float(row[34])\n",
    "            Exac_TET2_results[variant]=(allele_freq, african_freq, east_asian_freq, european_freq, finnish_freq, latino_freq, south_asian_freq)\n",
    "\n",
    "        row_count=row_count+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#COSMIC - retrieve the frequencies in COSMIC for each of the variants\n",
    "import csv\n",
    "filename=\"Data files/TET2_COSMIC_variant_counts.csv\"\n",
    "\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    TET2_COSMIC_results={} \n",
    "    \n",
    "    for row in read_reader:\n",
    "#         print(row)\n",
    "        if row_count>1:\n",
    "            variant = row[1]\n",
    "            counts = int(row[2])\n",
    "            TET2_COSMIC_results[variant]=counts\n",
    "\n",
    "        row_count=row_count+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Create dictionary of results and save to .csv file\n",
    "TET2_s_dictionary={}\n",
    "for k, v in TET2_missense_s.items():\n",
    "#     print(k)\n",
    "    if k in Exac_TET2_results.keys():\n",
    "        exac_freq = round((Exac_TET2_results[k][0])*100, 4)\n",
    "        african_freq = round((Exac_TET2_results[k][1])*100, 4)\n",
    "        east_asian_freq = round((Exac_TET2_results[k][2])*100, 4)\n",
    "        european_freq = round((Exac_TET2_results[k][3])*100, 4)\n",
    "        finnish_freq = round((Exac_TET2_results[k][4])*100, 4)\n",
    "        latino_freq = round((Exac_TET2_results[k][5])*100, 4)\n",
    "        south_asian_freq = round((Exac_TET2_results[k][6])*100, 4)\n",
    "    if k not in Exac_TET2_results.keys():\n",
    "        exac_freq = '-'\n",
    "        african_freq = '-'\n",
    "        east_asian_freq = '-'\n",
    "        european_freq = '-'\n",
    "        finnish_freq = '-'\n",
    "        latino_freq = '-'\n",
    "        south_asian_freq = '-'\n",
    "    if k in TET2_COSMIC_results.keys():\n",
    "        COSMIC_number = TET2_COSMIC_results[k]\n",
    "    if k not in TET2_COSMIC_results.keys():\n",
    "        COSMIC_number = '-'\n",
    "    TET2_s_dictionary[k]=(round(v[0]*100, 2),round(v[4]*100, 2),round(v[3]*100, 2), v[1], v[2], COSMIC_number, exac_freq, \\\n",
    "                            african_freq, east_asian_freq, european_freq, finnish_freq, latino_freq, south_asian_freq)\n",
    "\n",
    "for k, v in TET2_nonsense_s.items():\n",
    "    if k in Exac_TET2_results.keys():\n",
    "        exac_freq = round((Exac_TET2_results[k][0])*100, 4)\n",
    "        african_freq = round((Exac_TET2_results[k][1])*100, 4)\n",
    "        east_asian_freq = round((Exac_TET2_results[k][2])*100, 4)\n",
    "        european_freq = round((Exac_TET2_results[k][3])*100, 4)\n",
    "        finnish_freq = round((Exac_TET2_results[k][4])*100, 4)\n",
    "        latino_freq = round((Exac_TET2_results[k][5])*100, 4)\n",
    "        south_asian_freq = round((Exac_TET2_results[k][6])*100, 4)\n",
    "    if k not in Exac_TET2_results.keys():\n",
    "        exac_freq = '-'\n",
    "        african_freq = '-'\n",
    "        east_asian_freq = '-'\n",
    "        european_freq = '-'\n",
    "        finnish_freq = '-'\n",
    "        latino_freq = '-'\n",
    "        south_asian_freq = '-'\n",
    "    if k in TET2_COSMIC_results.keys():\n",
    "        COSMIC_number = TET2_COSMIC_results[k]\n",
    "    if k not in TET2_COSMIC_results.keys():\n",
    "        COSMIC_number = '-'\n",
    "    TET2_s_dictionary[k]=(round(v[0]*100, 2),round(v[4]*100, 2),round(v[3]*100, 2), v[1], v[2], COSMIC_number, exac_freq, \\\n",
    "                            african_freq, east_asian_freq, european_freq, finnish_freq, latino_freq, south_asian_freq)\n",
    "\n",
    "TET2_df = pd.DataFrame.from_dict(TET2_s_dictionary, orient='index', columns=['s (%)', 's lower value (%)', 's upper value (%)','observed number', 'site-specific mutation rate', 'Number of times in COSMIC',\\\n",
    "                                                                                'Exac Overall freq (%)', 'Exac African freq (%)', 'Exac East Asian freq (%)',\\\n",
    "                                                                                'Exac European freq (%)', 'Exac Finnish freq (%)', 'Exac Latino freq (%)', 'Exac South Asian freq (%)'])\n",
    "TET2_df_sorted = TET2_df.sort_values(by = ['s (%)'], ascending = False)\n",
    "\n",
    "TET2_df_sorted.to_csv('Data files/TET2_s_across_whole_gene.csv')\n",
    "# TET2_s_dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s (%)</th>\n",
       "      <th>s lower value (%)</th>\n",
       "      <th>s upper value (%)</th>\n",
       "      <th>observed number</th>\n",
       "      <th>site-specific mutation rate</th>\n",
       "      <th>Number of times in COSMIC</th>\n",
       "      <th>Exac Overall freq (%)</th>\n",
       "      <th>Exac African freq (%)</th>\n",
       "      <th>Exac East Asian freq (%)</th>\n",
       "      <th>Exac European freq (%)</th>\n",
       "      <th>Exac Finnish freq (%)</th>\n",
       "      <th>Exac Latino freq (%)</th>\n",
       "      <th>Exac South Asian freq (%)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>p.L532*</th>\n",
       "      <td>17.92</td>\n",
       "      <td>13.36</td>\n",
       "      <td>21.20</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.175600e-10</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.Q916*</th>\n",
       "      <td>15.73</td>\n",
       "      <td>13.86</td>\n",
       "      <td>17.20</td>\n",
       "      <td>5.0</td>\n",
       "      <td>9.919690e-10</td>\n",
       "      <td>43</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.S588*</th>\n",
       "      <td>13.49</td>\n",
       "      <td>10.88</td>\n",
       "      <td>15.05</td>\n",
       "      <td>2.0</td>\n",
       "      <td>8.230880e-10</td>\n",
       "      <td>4</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.Q530*</th>\n",
       "      <td>11.01</td>\n",
       "      <td>8.62</td>\n",
       "      <td>12.23</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.964580e-09</td>\n",
       "      <td>4</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.C1135Y</th>\n",
       "      <td>10.70</td>\n",
       "      <td>8.79</td>\n",
       "      <td>11.75</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.211140e-09</td>\n",
       "      <td>3</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0087</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.Q3*</th>\n",
       "      <td>10.65</td>\n",
       "      <td>8.79</td>\n",
       "      <td>11.63</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.211140e-09</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R1516*</th>\n",
       "      <td>8.21</td>\n",
       "      <td>6.73</td>\n",
       "      <td>9.04</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.882290e-08</td>\n",
       "      <td>26</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R550*</th>\n",
       "      <td>8.02</td>\n",
       "      <td>7.19</td>\n",
       "      <td>8.60</td>\n",
       "      <td>7.0</td>\n",
       "      <td>3.269290e-08</td>\n",
       "      <td>61</td>\n",
       "      <td>0.0058</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0116</td>\n",
       "      <td>0.0075</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0087</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R1359H</th>\n",
       "      <td>7.49</td>\n",
       "      <td>6.14</td>\n",
       "      <td>8.19</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.269290e-08</td>\n",
       "      <td>6</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          s (%)  s lower value (%)  s upper value (%)  observed number  \\\n",
       "p.L532*   17.92              13.36              21.20              2.0   \n",
       "p.Q916*   15.73              13.86              17.20              5.0   \n",
       "p.S588*   13.49              10.88              15.05              2.0   \n",
       "p.Q530*   11.01               8.62              12.23              2.0   \n",
       "p.C1135Y  10.70               8.79              11.75              2.0   \n",
       "p.Q3*     10.65               8.79              11.63              2.0   \n",
       "p.R1516*   8.21               6.73               9.04              2.0   \n",
       "p.R550*    8.02               7.19               8.60              7.0   \n",
       "p.R1359H   7.49               6.14               8.19              2.0   \n",
       "\n",
       "          site-specific mutation rate Number of times in COSMIC  \\\n",
       "p.L532*                  2.175600e-10                         -   \n",
       "p.Q916*                  9.919690e-10                        43   \n",
       "p.S588*                  8.230880e-10                         4   \n",
       "p.Q530*                  1.964580e-09                         4   \n",
       "p.C1135Y                 3.211140e-09                         3   \n",
       "p.Q3*                    3.211140e-09                         -   \n",
       "p.R1516*                 1.882290e-08                        26   \n",
       "p.R550*                  3.269290e-08                        61   \n",
       "p.R1359H                 3.269290e-08                         6   \n",
       "\n",
       "         Exac Overall freq (%) Exac African freq (%) Exac East Asian freq (%)  \\\n",
       "p.L532*                      -                     -                        -   \n",
       "p.Q916*                 0.0008                     0                        0   \n",
       "p.S588*                      -                     -                        -   \n",
       "p.Q530*                 0.0008                     0                        0   \n",
       "p.C1135Y                0.0008                     0                        0   \n",
       "p.Q3*                        -                     -                        -   \n",
       "p.R1516*                     -                     -                        -   \n",
       "p.R550*                 0.0058                     0                   0.0116   \n",
       "p.R1359H                     -                     -                        -   \n",
       "\n",
       "         Exac European freq (%) Exac Finnish freq (%) Exac Latino freq (%)  \\\n",
       "p.L532*                       -                     -                    -   \n",
       "p.Q916*                       0                     0                    0   \n",
       "p.S588*                       -                     -                    -   \n",
       "p.Q530*                  0.0015                     0                    0   \n",
       "p.C1135Y                      0                     0               0.0087   \n",
       "p.Q3*                         -                     -                    -   \n",
       "p.R1516*                      -                     -                    -   \n",
       "p.R550*                  0.0075                     0               0.0087   \n",
       "p.R1359H                      -                     -                    -   \n",
       "\n",
       "         Exac South Asian freq (%)  \n",
       "p.L532*                          -  \n",
       "p.Q916*                     0.0061  \n",
       "p.S588*                          -  \n",
       "p.Q530*                          0  \n",
       "p.C1135Y                         0  \n",
       "p.Q3*                            -  \n",
       "p.R1516*                         -  \n",
       "p.R550*                          0  \n",
       "p.R1359H                         -  "
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TET2_df_sorted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABZ8AAAHvCAYAAAA7GaHXAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAIABJREFUeJzs3Xd4VFX+x/H3CSEVktACUoMFUFQQVBBwCVVBqaIguBKx7MIKsnZdNWBZYa0gAq6KoqKrohRRCChBiqD00HuoIZCE0EIg5fz+mPKbyUxCCyTq5/U8eSZz7rnnfu89c+eRb47fa6y1iIiIiIiIiIiIiIgUp4CSDkBERERERERERERE/niUfBYRERERERERERGRYqfks4iIiIiIiIiIiIgUOyWfRURERERERERERKTYKfksIiIiIiIiIiIiIsVOyWcRERERERERERERKXZKPouIiIj8jhljAo0x1hjzQUnHUhoZYwYbYzYZY045r1NNZ3s7Y8yvxphjzvZ7SjpWEREREZE/GiWfRUREREohY0ysMyla2E/zIvYNMMYMM8Z0vZgxlzbGmA7AaGAt8BDwVyDDGFMZ+BYIAR51ti+6QDE0cc5F7Qsx/p+RMWbPae4Nz59WHn+gKfJ+Msa8fBbjfuCMpbsx5mNjzEZjzHFjzD5jzBxjTMeSvk4iIiIipUFgSQcgIiIiIkX6AvjBT/tWAGttrjEmFMj12BYAxAMfAtMveISlVwfn633W2iOuRmNMLBABPG+tvdDXpwmOufgR2HWBj/VnMRgI93jfEHgamAxMK9B3k8fvy4G3CxlzK/A1sLFA+2jgFPC4n/4AHwDpzuNuBioC9wMJxpinrbUjT3cyIiIiIn9kSj6LiIiIlG4rrLWfFdXBWpt9sYL5nakG5Hkmnj3aATIucjy/O8YYA4Rba4+VdCwu1topnu+NMe1xJJ9X+7tXjDGuf/PsOc29lAasLrDvCCC7iP3ustbOLbDPu85xhhtjxltrDxd5QiIiIiJ/YCq7ISIiIvI7VrDmszHmciDHufl+jzIBuQX7O0sSLDDGZBlj0owx/zXGhPs5Rg1jzHhjzG5n7eS9zveVC/SrZIwZZYzZbozJNsakG2OWGWMeLdAvzhiz1BiT6SxVsM0Y85kxptIZnvMtztIGh53HWW2Mechj++XGGIujnEYZj2vwozFmD44V4QALPK+Nc98QY8xzxpj1zrEPGWOmG2Ma+YnDGGP+Zoz5zThqRx81xiQZY+Kd218G3i9wLM+5CjXGvGgcNamznMda40x4nu4aBDrjXGCMSXXOy05jzLvGmIqF7HOnMeZn53XLcpaKGOVKzhpj2jvj+6tx1MreAJwEhnqMEeu8jkecYyw3xsT5OdY1xphvnJ+Vk8aYFGPMXGNMJ48+53z+pUXBxLOz7TjwPRAMXHHRgxIREREpRbTyWURERKR0CyuY5AVOWmuPFtJ/P9AfmAjM4/8TrfkF+jUFeji3fwa0BR7EUb5jkKuTMaYu8AtQxtl3O46E2kCgjTHmBo+Vxd8CNwHjgSQcpRGuBGKBN53j3QdMAH4GXgBOALWBzkBlHCUMCmWMGQi864zpJSALuAV4zxhT11r7jPMa/NUZY3Pn9QBIwVFu4zYcpRFewlEqId85dhAwG2gGfIKj5EIF53X5xRjTylq70tnX4CiJ0htYDLwCZDrP9w5gOI4yDtUKHAv+v2TDeOBe4GPnGGWd17ZtUdfAKQR4DPgGmAocB27EUdu6pXNeXH+EwBgzEngSWAe8AaQClzlj/RfgubL5Med5f+i8ljudY3THUdoiBXjNecy7gY+c196VdK8CzAXynOe4C8fc3uCMcWYxnP+5CPJzLwHkW2uLexV8TefrgWIeV0REROR3xVhrSzoGERERESnAOOoSJxay+UtrbR9nv0AcK50/tNY+UFibx7iubflAM2vtMo9tCTgSxVHW2hPOtu9xJKqbWGv3efRthiMBHG+tfdm52jYdeMdaO6SI85oOtAKqWGvzzvByuPatCWxznv+9Bba9C/wNuMxa60qWfgb0sdYGFuj7AI4VyTdbaxd6tD8BjAQ6Wmt/9GiPwvHQwo3W2vbOtr7AJBxJ/gHW2nyP/gGu94Udy7ntCDDPWnvWD4Y0xgQAwa558mj/G46k7h3W2m+dbS1wPFDxR+B2a+1Jj/7GOv9BYBzlK+bgmMcG1to0j35lgR1AGHCVtXa/sz0Yxx8SbgCusNZuN8b0xJEUd8dQyDmc8/kXGMcV9/PW2pf9bHd95gtz2FobVcjYe3CU3bj8LOJpAvwGzLfWXqhEuoiIiMjvgspuiIiIiJRu/8Xx4DzPH58E2zlY6Jl4dpoLBAF1AJwJ5U44VtaeMsZUdv3gSALvADo6983CkeBrboypU8RxDwPlgc7O1cNn405nfBM8Y3HG8x2O1dntznJMT/fgWBm8qsDYgcBPQGtnshWgH2CBxz0TzwAF3xfhMHCNMabh2QZqrc33+ANBGWNMlDNWVxmIZh7d+zlfn/ZMPDvH8bcS5WPPxLPTDUAN4ANX4tm5/0ngdRz/rnAlkV01jjsbY8oXcRrnfP7n6Bd876UOQLfiOoAxpiqOxPtxHKvQRURERP7UVHZDREREpHTb4rkKtxht99PmKnnhqr3cADA4VhT/rZBx8sDx0EPjqO38JpBsjFmHIxE6xVrruYL7ZRwrn6cDacaYn3GUYfjyDB5qd6XztbAV4QBVTzNGURrgSG4fLKJPRRxlJ67A8QC7gknas/EIjpXTa40x23Cc13RgRiFJYS/GmD44SmQ0wlGywlMFj9+vwDFPSWcY12Y/bXWdr+v8bFvrfL3U+ToX+BxHuZF7jTFLcZQz+dJau9Fjv/M6/3Nw8ALdSwA4k/8/4vgMdrbWbj3NLiIiIiJ/eEo+i4iIiPw5FVXywhR4nYijLrQ/Wa5frLVjjDFTcNRU/gtwFzDYGDPJWnuPs88mY0wDoD2OVcqtgQ+AYcaYv1hrd5xBXP0ovJbu+ST8AoBVwBNF9HHVBjY4Vj6fM2vtt87ke2cc16ED8AAwzxjT0bNmc0HGmLtw1JxeAgwB9gDZOJLn3+P9fzie7QrzLD9tZzyGM3Hcz/ngwE7AzTjqTT9vjBlsrR3n7HfO51/aGMfDMn8ELge6WmvnlWxEIiIiIqWDks8iIiIifzzFtWp0q3Ossme6YtRauxdHqZD/OmvtTsKRiHzD9bA+Z6mG750/GGO6AtOAoThWwxZmi/P1Qq1g3QJUAX46g5W3m3CUlah8mtXPRY5jrU0HPgU+dZYheQ3HaubbgSlF7PpXHEniNtbabFejMebqQmJtD1wDrCgqniJsc776K5FxlfPVazW9tXYNsAb4j7OEy2/ACGCcR59zPf9Sw5l4/gnHyvmu1to5JRySiIiISKmhms8iIiIifzDOB/ll4ygRcT7jpOIol3CnMeaGgtuNQxXn72HGmNAC++fiSD7iisVZmqCgFZ59ivAlcAp40RgT4ieeKGNM0GnGKMonOOoa+02AO+v5ukzCsRr4PwVrVxd47yolUrFAn0BjTKRnmzPhvcpffz/ycDw00v3f887jPuen7+fO11cLXp+zqLu9FNgL3G+MifbYPwh43BnLdGdbxYLjWmszgGSgnDEmqBjOv1RwJtV/xJF47mGtnV3CIYmIiIiUKlr5LCIiIvLHtAS4xRjzJLAbyLPWfnUO4/wNWAgsNMZ8AqzE8d+QdYHuwIc46jhfBfzoLLuxDjjkbBuIY9XsIud4c40xB4EFzrgqAvfhSF4WVtoDAGvtTmPMw8B4YL0x5jNgF47VytfieHBcPRwlKM7FmzhWCL9ljGkPzAOOArVxlAg5iqM0BMD/gJ7O2OsZY77D8QC9es6+jZz9fsOx+vk5Z6L+OP//sMZdxpjpOBKuB3DUTB6Io7TH96eJdbLzfOcaYz4FgoEegE9S3lr7izHmDRwripcZY74G9uOYwzuNMdedrt62tTbXGDMY+BpYaox533kufYAbgRetta6VzwOAh52fha1ALtDGeV0+t9aecv4R4nzO/1zUNMbcU8i2XzziPyPOBPuPQGMcf4yo5Gf8RacpJSMiIiLyh6bks4iIiMgf09+BMThWwpbHsVL2rJPPzoRvE+BpoCtwL45yD7uBqTiSoAA7cdSGjsWRBA3GkQR+DxjpURriXeBOZ3wVcTzkcAUw6Ezq5Fpr3zfGbMCx2nYgEAmkARuBf1H0wwJPN/YpY8ytwMPAPcBw56Z9OJL5Ez36WmNMb+AfOJKt8Tiu8XYcCVpXvx3GmAdw1Dweh+PBgB869xuNIyHbAQjH8SDDKcCr1tr9p4n1M2NMORyrtN/AkbCdDjyPn3rY1trHjTErncd9Cseq7d3ADByr5M/k+kwxxnTA8Zl6Cse/JTYAA6y1H3l0nYvjjwFdgEtwJJ934Eh+j3H2OXo+53+OmuIo8eHPffh/CGdRygDXOX/v5/wp6K84zl1ERETkT8lcmAdJFw9jzEjgehwrSCoDJ3D8w2YqMMZZI67gPi1w/AdxcxwrP7YCE4B3nP8LqoiIiIiIiIiIiIhcYKU9+XwKx0qY9ThWcITjSCpfj2MFSnNr7W6P/t2Ab3Cs3vgSxwqQLkB9YLK19s6LegIiIiIiIiIiIiIif1KlPfkc4vn0bo/2V4BngXHW2kHOtggcq5wjgZbW2mWuMXD8r383AXdba/93seIXERERERERERER+bMKOH2XkuMv8ezkqld4hUdbLxwPm/mfK/HsMYbrqd8Diz1IEREREREREREREfFRqpPPRejifE3yaGvrfJ3lp/98HA/GaWGMCb6QgYmIiIiIiIiIiIiI4wnVpZ4x5nGgHI6SGtcDrXAknkd4dKvvfN1ccH9rba4xZgfQELgUx1O5/Ro3bpwFGDhwoCmW4EVERERERERERET+hH4vK58fB+KBoTgSz7OAjtbagx59Ip2vhwsZw9Ue5W+jMeYhY4y7XMfhw4ftypUrbU5Ojl22bJkF7LJly+zJkyft6tWr7aFDh+zWrVvtnj17bEpKit20aZM9cuSIXblypc3NzfXZZ82aNTYjI8Nu27bN7t692+7fv99u3LjRHj161C5fvtzm5eV57XPixAm7du1am56ebrdv32537dplU1NT7YYNG+yxY8fssmXLbH5+vtc+WVlZdt26dfbgwYN2x44ddufOnfbAgQN2/fr19vjx4+6+S5cude9z/Phxu379envgwAG7c+dOu2PHDnvw4EG7bt06m5WV5TW+63jHjh2zGzZssKmpqXbXrl12+/btNj093a5du9aeOHHCa5+8vDy7fPlye/ToUbtx40a7f/9+u3v3brtt2zabkZFh16xZY0+ePOm1T25url25cqU9cuSI3bRpk01JSbF79uyxW7dutYcOHbKrV6/22ScnJ8euXLnSHj582G7evNnu27fP7t27127ZssVmZmbaVatW2ZycHK9zP3XqlHsut2zZYvfu3Wv37dtnN2/ebD3n37XP0qVL7cmTJ21SUpJ7LgvO/4oVK3zmMjs7265Zs8ZrLgvOf8G5dM1/Wlqaey5TU1Pt+vXr3fNvrS10/pOTk21ycrI9ePCgz/x7vh47dsxn/tPS0uzatWttVlaW1/XKz8+3y5cvL3L+s7OzvT5neXl5dsWKFfbIkSN+5z8pKanQ+T98+LDdtGmT3bdvn8/8nzp1ymf+V61aZTMzM+3mzZvt3r177d69e+3mzZu95t9zH8/5d93L+/bts5s2bXLPv7972XP+Pe9lz/n3vG7Z2dlF3ssF53/p0qU2KyvLZ/5d97Jr/gvOpeteLjj//u5l12fHNf+pqak+81/wXnbNv+e97Dn/a9as8Zp/1/3vmn/Pe7ng/HveY0Xdy/7m37WPa/6Lupf1Xa7vcn2X67tc3+X6Ltd3ub7L9V2u73J9l+u7XN/l+i7/M36XU4JK9QMHCzLGVAVa4FjxXB643Vq7wrltM44a0FdYa7f62fcXHA8dvMlau6SwY3isfC7+ExARERERERERERG5uEqswsPvZeUzANbaVGvtFKAjUAn4xGOza2VzpM+ODhEF+omIiIiIiIiIiIjIBfK7Sj67WGt3AuuBhsaYys7mTc7XegX7G2MCgbpALrD9ogQpIiIiIiIiIiIi8if2u0w+O1V3vuY5X+c6X2/10/cvQBjwi7X25IUOTEREREREREREROTPrtQmn40xDYwx1fy0BxhjXgGicSSTDzk3TQbSgD7GmOs9+ocALzvfjrvAYYuIiIiIiIiIiIgIEFjSARThVuA1Y8x8YBuQDlQFWgOXAvuBB12drbVHjDEP4khCzzPG/A/IALoC9Z3tX17UMxARERERERERERH5kyrNyecfgf8CLYFGQBRwHNgMfAqMttZmeO5grZ1qjGkN/Au4AwgBtgKPOvvbixe+iIiIiIiIiIiIyJ9XqU0+W2vXAv84h/0WAZ2LPyIREREREREREREROVOltuaziIiIiIiIiIiIiPx+KfksIiIiIiIiIiIiIsVOyWcRERERERERERERKXZKPouIiIiIiIiIiIhIsVPyWURERERERERERESKnZLPIiIiIiIiIiIiIlLslHwWERERERERERERkWKn5LOIiIiIiIiIiIiIFDsln0VERERERERERESk2Cn5LKXKsGHDMMYwb968kg5FREREREREREREzoOSzyIiIiIiIiIiIiJS7JR8llLl4YcfZsOGDdx4440lHYqIiIiIiIiIiIich8CSDkDEU+XKlalcuXJJhyEiIiIiIiIiIiLnSSuf5bwlJydjjCEuLo5t27bRq1cvKlWqRPny5enYsSNr164F4ODBgzz00ENccsklhISEcMMNN5CYmOg1VmE1nxcsWECXLl2oWbMmwcHBVKtWjebNmzN8+HCvfqmpqTz++OPUr1+f8PBwoqKiqF+/PnFxcWzfvt0n9oSEBDp37kzlypUJDg7msssu44knniAzM9Onb0xMDDExMWRlZfHEE09Qu3ZtgoODufzyyxk5ciTWWp99pk+fTrt27bjkkksIDg6mevXqtG7dmrFjx/r0zcjI4JlnnuHKK68kNDSUyMhI2rVrx+zZs087ByIiIiIiIiIiIqWNVj5LsUlOTqZZs2ZceeWVxMXFkZyczJQpU4iNjWXx4sXceuutRERE0Lt3bzIyMvjf//5Hp06d2Lx5M7Vr1y503FmzZnHbbbcRERFB165dqVGjBhkZGWzYsIGxY8cSHx8PQFZWFi1btmTbtm106NCBLl26YK1l586dTJs2jV69enHppZe6x33xxReJj4+nYsWK3H777URHR5OUlMTrr7/ODz/8wOLFi4mIiPCKJScnh44dO7Jv3z46depEYGAgU6dO5emnnyY7O9sdC8B///tf/va3v1GtWjW6dOlC5cqVOXDgAElJSXz00UcMGjTI3Xfnzp3ExsaSnJzMzTffzK233srx48eZMWMGt956K++99x4PPvhgcU2ViIiIiIiIiIjIBafksxSbn3/+mZdffpl//etf7raXXnqJF154gWbNmnHXXXcxduxYAgIcC+47dOjAvffey1tvvcVbb71V6Ljvv/8++fn5zJs3j0aNGnltS0tLc//+008/sW3bNoYOHeoz3qlTpzh58qT7fWJiIvHx8dx000388MMPREVFubd9/PHH3HfffcTHx/uMs2/fPho1asScOXMIDQ0FID4+nnr16vHWW2/x7LPPUrZsWQDee+89goKCWL16NdHR0YXGDdC/f3927tzJF198QZ8+fdztmZmZxMbGMmTIELp27UrVqlULvU4iIiIiIiIiIiKliZLPF0jfJTklHcIZ+7x52WIZJyYmhqefftqrrX///rzwwgucPHmS1157zZ14Bujbty8DBgxg1apVZzS+K9nryV99aH/9goKCCAoKcr8fPXo04EhseyaeAeLi4hg1ahSTJk3ymxQfPXq01zGio6Pp1q0bn3zyCZs2beLqq692bwsMDHQnowuLe/Xq1fz888/06tXLK/EMEBUVxfDhw+nevTvffPON12ppERERERERERGR0kzJZyk2jRs3pkyZMl5t1atXB6BevXqUL1/ea1uZMmWoWrUqe/bsKXLcfv368e2339KsWTN69+5NmzZtaNmyJTVr1vTq17p1a2rUqMGIESNYsWIFnTt3pmXLln7jWrx4MWXLluXrr7/m66+/9jnmqVOnOHjwIOnp6VSqVMndHhkZyeWXX+7Tv1atWgAcOnTIK+7HHnuMhg0b0rt3b1q3bk3Lli2pUqWKTywAhw8fZtiwYT5jHzx4EIANGzYUdZlERERERERERERKFSWfpdhERkb6tAUGBha6zbU9J6foVeI9e/ZkxowZvPHGG0yYMIH33nsPgKZNm/Lqq6/SoUMHACIiIliyZAnx8fFMnz6dhIQEwLHKeNCgQTz33HPuVcjp6enk5ub6PLCwoGPHjnklnwuuki54nnl5ee62Rx99lMqVKzN27FhGjx7N22+/jTGG1q1b89prr3H99de7YwGYM2cOc+bMKTIWERERERERERGR3wslny+Q4iplIQ633XYbt912G8ePH+fXX39lxowZjBs3jttvv52VK1dy1VVXAVCzZk0+/PBDrLWsX7+euXPn8u677/Liiy+Sn5/PSy+9BDiS4fn5+WRkZFzQuO+9917uvfdeMjMz+eWXX5gyZQoTJkzglltuYcOGDURHR7sT86NGjWLIkCEXNB4REREREREREZGLJeD0XURKj/DwcNq2bcubb77Js88+y6lTp5g5c6ZPP2MMDRs2ZPDgwe7VxFOnTnVvb968OYcOHWLdunUXJe6oqCg6d+7M+++/T1xcHBkZGSxYsMAdC+B+LyIiIiIiIiIi8keg5LOUej/99BMnTpzwaU9NTQUgLCwMgLVr15KcnHzafgD//Oc/AXjwwQfZt2+fzz7Hjx9nyZIl5xX3rFmzyM3N9Wk/cOCAVzzXX389N998M99++y0TJkzwO9aaNWvc+4mIiIiIiIiIiPweqOyGlHqPPfYYycnJxMbGEhMTQ1BQEMuXL2fu3LnUqVOHPn36APDjjz/y6KOP0qJFCxo0aEB0dDR79uxh2rRpBAQE8MQTT7jHbNeuHSNGjOCZZ57hiiuuoHPnztStW5djx46xc+dOfv75Z1q1asWsWbPOOe4+ffoQEhJCq1atiImJwVrLggULWLp0KU2bNqV9+/buvp9//jlt27bl/vvvZ/To0TRr1oyoqCj27NlDUlISa9euZfHixURHR5/7hRQREREREREREbmIlHyWUu/ZZ59lypQpLFu2jB9//JGAgABq167Ns88+y9ChQ6lQoQIAt9xyC0OHDmX+/PlMmzaNI0eOcMkll9ChQwd3UtrTU089RcuWLRk9ejQLFy5k2rRpREZGUqNGDR566CH69u17XnGPGDGChIQEVqxYwQ8//EBISAh16tRh5MiRDBw40P3wQ3DUql6+fDnvvPMO33zzDZMmTSIvL49q1apx1VVXMXjwYK655przikdERERERERERORiMtbako6hVBk3bpwFGDhwYEmHIiIiIiIiIiIiInK+TEkdWDWfRURERERERERERKTYKfksIiIiIiIiIiIiIsVOyWcRERERERERERERKXZKPouIiIiIiIiIiIhIsVPyWURERERERERERESKnZLPIiIiIiIiIiIiIlLslHwWERERERERERERkWKn5LOIiIiIiIiIiIiIFDsln0VERERERERERESk2Cn5LCIiIiIiIiIiIiLFTslnERERERERERERESl2Sj6LiIiIiIiIiIiISLFT8llEREREREREREREip2SzyIiIiIiIiIiIiJS7JR8FhEREREREREREZFip+SziIiIiIiIiIiIiBQ7JZ9FREREREREREREpNgp+SznLTk5GWMMcXFxJCcn06dPHypXrkxISAjXX389M2bM8Nnn5MmTjBgxgmuvvZawsDAiIiK4+eab+eqrr4pl/FOnTjF69GiaNGlChQoVCAsLIyYmhm7duvHjjz/69N+4cSNxcXHUqlWL4OBgqlatSt++fdm0aZNP37i4OIwxJCcn895773HNNdcQEhJC1apVeeihhzh8+LDPPklJSdx9993ExMQQHBxMlSpVaNKkCUOHDiUnJ8erb25uLmPHjqV58+ZEREQQFhbGddddx5gxY8jPzy9yLkREREREREREREoLY60t6RhKlXHjxlmAgQMHlnQovxvJycnUrVuX2NhY1q1bx6WXXspNN91ERkYGX375JTk5Ofz444+0adMGcCSGO3bsyM8//0yDBg24/fbbycrKYvLkyRw4cIBnnnmGf//73+c8PkDfvn354osvuPrqq2nbti2hoaHs27ePhQsX0rNnT15//XV331mzZtGzZ09ycnLo0qULl19+OXv27OHbb78lODiYxMREmjRp4u4fFxfHxIkTufPOO0lISKBLly5UrVqVxMREVq5cSZs2bZg7d667f1JSEs2aNcMYQ9euXalbty5Hjhxh69atJCYmkpGRQbly5QDcMSQkJFC/fn1iY2MJCQkhMTGRpKQk7rnnHj799NMLNpciIiIiIiIiIvKHY0rsyNZa/Xj8jB071o4dO9bKmduxY4cFLGCHDRvmtW3WrFkWsJ06dXK3/fvf/3a35eTkuNtTU1NtnTp1LGAXLVp0zuNnZmZaY4xt2rSpzc3N9Yk3LS3N/XtGRoaNioqylSpVsuvWrfPqt3btWhseHm6vu+46r/b+/ftbwNaqVcvu3LnT3Z6Tk2NvvvlmC9hff/3V3f7oo49awE6dOtUnloyMDJuXl+d+Hx8fbwH78MMPe8Wem5trBwwYUOg4IiIiIiIiIiIihSixXGtgiWS8/wRM99ElHcIZs1OHFMs4derU4bnnnvNqu+WWW6hduza//fabu23ChAkYY3jzzTcJDPz/j2B0dDTPP/88DzzwAB988AEtWrQ4p/GNMVhrCQ4OJiDAt7JMpUqV3L9/8sknZGZmMmbMGK666iqvfg0bNuTBBx/k7bffZv369T7bX3jhBWrXru1+HxgYyH333ceCBQv47bffuPHGG736h4aG+sRSoUIF9+/5+fmMGTOGatWq8dZbb1GmTBn3tjJlyvDGG2/w0UcfMWmwSU9OAAAgAElEQVTSJLp16+YzloiIiIiIiIiISGmi5LMUm8aNG3slTF1q1arF4sWLATh69Chbt26lRo0aNGjQwKdv27ZtAVi5cuU5jQ8QERFBly5d+O6772jcuDF33HEHN998M82aNSMsLMxrX9d+q1evZtiwYT5jb968GYANGzb4JJ+vv/56v7EAHDp0yN3Wu3dvRo0aRffu3enVqxft27enZcuWXHbZZT7HSk9P54orruDll1/2GRscCewNGzb43SYiIiIiIiIiIlKaKPksxSYqKspve2BgoPtBea6H8V1yySV++7raMzMzz2l8ly+//JKRI0fy+eefEx8fD0BISAi9evXi9ddfp2rVqgCkp6cD8P777xd5bseOHTujeFwrufPy8txtN954IwsWLOCVV15h8uTJ7prN9evXJz4+nrvvvtsrli1btjB8+PCzikVERERERERERKS0UfL5AimuUhZ/NJGRkQDs37/f7/aUlBSvfucqNDSUYcOGMWzYMHbv3s38+fP5+OOP+eyzz0hOTmbBggVex1m9ejXXXnvteR2zKDfddBMzZszg5MmTLF++nFmzZvHOO+/Qt29fqlSpQvv27d2x9OjRg2+//faCxSIiIiIiIiIiInIx+BbEFbmAypcvz2WXXcbevXvZsmWLz/bExEQAmjRpUmzHrFWrFv369SMhIYErrriChQsXulcZN2/eHMCdjL7QgoODadGiBS+++CKjRzvqgk+bNg2ABg0aEBUVxZIlS8jJybko8YiIiIiIiIiIiFwoSj7LRTdgwACstTzxxBNe5SnS0tJ46aWX3H3O1cGDB/n111992o8fP87Ro0cJDAwkKCgIgPvuu4+oqCiGDx/u9dBCl/z8fObNm3fOsYAjse0qN+IpNTUVwF2HOjAwkMGDB5OSksKQIUM4ceKEzz4pKSmsX7/+vOIRERERERERERG5GFR2Qy66xx9/nJkzZzJt2jQaNWpE586dycrK4uuvv+bAgQM8+eSTtGrV6pzH37t3L82bN+fKK6+kSZMm1KpViyNHjjBjxgz279/PkCFDKF++PACVKlVi8uTJ9OjRg+bNm9OuXTsaNmxIQEAAu3btYvHixaSnp5OdnX3O8bzxxhvMnj2b2NhYLr30UsqVK8e6deuYOXMmFSpU4KGHHnL3ff7551m9ejXjx4/nu+++o23bttSoUYMDBw6wZcsWFi1axCuvvOLz8EMREREREREREZHSptQmn40xlYAewG3ANUAN4BSwBvgI+Mham+/RPwbYUcSQX1pr+1yoeOXMBQUFMWfOHN58800+//xz3nnnHQIDA2nUqBFvv/22+wF85yomJobhw4czb948EhMTSUtLo2LFitSvX58RI0bQp4/3x6Bdu3YkJSXx+uuvk5CQwIIFCwgKCqJ69eq0bduWO+6447ziGTRoEBUqVODXX39l0aJF5ObmUrNmTQYNGsRjjz1GnTp13H3Lli3L1KlT+eyzz/j444+ZMWMGx44do0qVKtStW5eXXnqJfv36nVc8IiIiIiIiIiIiF4Ox1pZ0DH4ZY/4OjANSgERgF1AV6AlEAt8Ad1rnCXgkn1cDU/0MudZaO/l0xx03bpwFGDhw4Hmfg4iIiIiIiIiIiEgJMyV14FK78hnYDHQFvi+wwvlZ4DfgDhyJ6G8K7LfKWjvsYgUpIiIiIiIiIiIiIr5K7QMHrbVzrbXfeSaene37gfHOt7EXPTAREREREREREREROa3SvPK5KDnO11w/26obY/4GVALSgcXW2qSLFpmIiIiIiIiIiIiI/P6Sz8aYQOBe59tZfrp0cP547jMP6G+t3XVhoxMRERERERERERERgDLDhg0r6RjOyvDhw/8DdAd+sNa+6NEeBGQDjwJPAyOBmUAM0AboMnz48AnDhg3L8RkUMMY8NHz48Pduu+226gD16tVjw4YNVK5cmZUrV1K9enWWLVtGpUqVWLduHcHBwezdu5esrCyOHz/Ovn37CAoKYv369VSpUoUVK1Z47bNhwwbKli3Lvn37OH78OCdOnGDPnj0EBwezdu1aqlat6rVPhQoV2LRpE2XKlGH//v0cPXqU7Oxsdu/eTWhoKElJSVSrVo3ly5e794mKimLz5s0YYzhw4ABHjhzh1KlT7Ny5k7CwMJKSkqhevTpLly6lRo0aLFu2jMjISLZu3Yq1lrS0NA4dOkRubi7JycmUK1eO1atXu8d3HS8iIoJt27aRn59PRkYG6enp5Ofns337dsqXL++1T9WqVVm5ciXly5dnx44d5ObmcujQIdLS0gDYunUrUVFRrFq1yr1PdHQ0q1evJjw8nOTkZHJycjh8+DAHDhzAGMPmzZupUKGC1z5VqlQhKSmJsLAwdu3axcmTJzly5AipqamUKVOGjRs3UqlSJVasWOE+98qVK7N27VqCg4PZs2cP2dnZHDt2jJSUFAIDA93z79pn6dKlVK5cmfXr17vnsuD8r1u3jujoaK+5rFixIhs2bCAwMNA9lwXnv+BcuuY/ICCA1NRUjhw5wsmTJ9m1a5d7/i+55JJC5//gwYNkZmaSk5PjM//Lli1zv0ZERPjMf15eHtu3b6dcuXKsWrXKfb2qVavGihUripz/iIgI97wsXbqUatWqsWrVKsLDw/3O/5YtWwqd/7CwMHbu3MmpU6d85r9ixYpe92WVKlVYs2YNISEh7Nq1i+zsbI4ePcr+/fu95t9zH8/5d93Lx44dY9++fZQtW5YNGzb4vZc959/zXvac/+XLl7uvW8WKFdm4cWOh9/KaNWu85n/p0qV+5991L7vmv+Bcuu5lwGv+/d3Lrs+Oa/7z8/NJT0/3mv+C97Jr/j3vZc/537ZtG5GRkV5zWbVqVff8e97LBw8e9Jr/lStXuu+x6OjoQu/lgIAAn/l37eOa/927dxd6L+u7XN/l+i7Xd7m+y/Vdru9yfZfru1zf5fou13e5vsv1Xf5n/C6/5JJLhp9DGrZYGGttSR37rBljhgCjgI1AS2ttxhnsEwgsBJoBQ621o4rqP27cOAswcODA8w9YREREREREREREpGSZkjpwqX3gYEHGmH/gSDyvB9qcSeIZwFqbC3zgfPuXCxSeiIiIiIiIiIiIiHj4XSSfjTFDgTHAWhyJ5/1nOcRB52t4sQYmIiIiIiIiIiIiIn6V+uSzMeYp4C1gFY7E84FzGKa583V7sQUmIiIiIiIiIiIiIoUq1clnY8zzwAhgOdDOWptWRN9mxpggP+1tgX863352QQIVERERERERERERES+BJR1AYYwx/YEXgTxgATDEGJ/a2MnW2o+dv48EGhpj5gF7nG3XAm2dvz9vrf3lQsYsIiIiIiIiIiIiIg6lNvkM1HW+lgGGFtLnZ+Bj5++fAj2AG4BOQFkgFfgKGGOtXXDBIhURERERERERERERL6U2+WytHQYMO4v+HwIfXqh4REREREREREREROTMleqazyJ/NnFxcRhjSE5OLulQSr158+ZhjGHYsGElHcoZi42NxU/5IBERERERERGRPyQln+W8JCcnY4whLi6upEPxkZOTw5QpU7j//vu5+uqriYiIICwsjGuuuYYXXniBo0ePFrrvnj17GDBgANWrVyc4OJiYmBiGDh3KoUOHLuIZiBSP0nyfioiIiIiIiMgfV6ktuyFyvrZt20bPnj0JDw+nTZs23HbbbRw7doyEhAReeuklvvzySxYtWkTlypV99mvRogUHDhygW7duNGjQgN9++41Ro0Yxa9YsFi1aRKVKlUrorOT37JNPPiErK6ukwxARERERERERuSiUfJY/rPLly/Puu+/Sv39/wsPD3e2nTp2iZ8+efP/99wwfPpx33nnHa79BgwZx4MABRo8ezeDBg93tjz76KG+99Rb/+te/GD9+/EU7D/njqF27dkmHICIiIiIiIiJy0ajshpyzYcOGUbduXQAmTpyIMcb98/HHH7v75efnM378eG644QbKlStHeHg4N9xwA+PGjSM/P99nXGMMsbGx7Nu3j7/+9a9ER0cTGhpK06ZN+fzzz884vho1ajBo0CCvxDNAUFAQzz77LOCoG+xp+/btzJ49m5iYGP7xj394bRs+fDjh4eF8+umnHD9+/IzjWL58OXfccQfR0dEEBwdTp04dBg0aREpKSqH75Ofn8+abb9KgQQNCQkKoWbMm//znPzly5IhP36SkJO6++25iYmIIDg6mSpUqNGnShKFDh5KTk+PVNzc3l7Fjx9K8eXN3GZLrrruOMWPG+MyFZ6mGzZs307t3b6KjowkICGDevHk0aNCAoKAg0tLS/J7DiBEjMMbw7rvverXv2bOHhx9+mEsvvZTg4GAqVapE165dWbp0qd9xUlNTuf/++6latSqhoaE0btyYiRMnFnrt/Hn11VcxxjB69Gi/2/ft20eZMmW44YYbvNpefPFFWrZsSbVq1QgKCqJ69er07duXDRs2+IxxuusF/ms+nzp1ijFjxtC5c2fq1KlDcHAwFStWpH379sycOdNvvDExMcTExJCVlcUTTzxB7dq1CQ4O5vLLL2fkyJFYa919z/Q+FREREREREREpblr5LOcsNjaWzMxMRo0aRaNGjejevbt7W+PGjd2///Wvf+Xzzz+nVq1aPPDAAxhjmDJlCoMGDWLhwoVMmjTJZ+xDhw7RokULoqKiuO+++8jMzOSrr76iX79+7N27lyeeeOK8Yi9btiwAgYHet8DcuXMB6NixIwEB3n+bKV++PC1btmT27NksWbKEdu3anfY4M2bM4I477sBaS69evahTpw7Lly9n3LhxTJs2jUWLFhETE+Oz3z//+U/mz5/PXXfdRbdu3UhISODtt99mwYIFLFy4kJCQEMCReG7WrBnGGLp27UrdunU5cuQIW7duZezYsbz88svuc83JyaFLly4kJCRQv359+vbtS0hICImJiQwePJhff/2VTz/91CeWbdu20axZM+rVq0e/fv04ceIEERER9O/fn2effZYvvvjCa4W4yyeffEJQUBB9+vRxt61YsYKOHTuSkZHBLbfcQs+ePUlLS2Pq1Km0atWKKVOm0LlzZ3f/9PR0WrRowfbt22nVqhWtWrUiJSWFv//973Ts2PG019/l3nvv5bnnnmPixIkMGTLEZ/tnn31Gfn4+/fv3d7fNnz+fESNG0KZNG+644w7KlSvHli1bmDx5MtOnT2fRokU0atTojK9XYTIyMnjkkUdo0aIFHTp0oEqVKqSkpPDdd9/RuXNn3n//fR544AGf/XJycujYsSP79u2jU6dOBAYGMnXqVJ5++mmys7OJj48Hzvw+FREREREREREpdtZa/Xj8jB071o4dO9bKmdmxY4cFbP/+/f1u//zzzy1gr7vuOnv06FF3+7Fjx2zTpk0tYCdNmuS1D2ABe+edd9q8vDx3+/bt222FChVs2bJl7bZt284r7r///e8WsE8//bRX++OPP24B+/rrr/vd7x//+IcFzugzcvToUVupUiUbEBBg58+f77VtxIgRFrAdOnTwau/fv78FbKVKlWxycrK7PS8vz/bs2dMC9sUXX3S3P/rooxawU6dO9Tl+RkaG1/WLj4+3gH344Ydtbm6uuz03N9cOGDDAZxzX3AL2mWee8Rl/z549NiAgwDZt2tRn22+//WYB27NnT3dbTk6Oveyyy2xwcLCdN2+eV/+9e/fa6tWr22rVqtns7Gx3+4MPPmgBO3ToUK/+S5cutYGBgRaw8fHxPsf3p2PHjhawa9as8dl21VVX2bJly9q0tDR3W2pqqj1y5IhP31WrVtnw8HB76623erWf7npZa23r1q2t42v3/2VnZ9vdu3f79M3MzLQNGza0FSpUsFlZWV7b6tSpYwHbqVMnr22pqak2MjLSRkZG2lOnTvnEVth9KiIiIiIiIiJ/aCWWa9XK5wul0Z0lHcGZW/31BRt6woQJgKMEQ7ly5dzt4eHhjBw5kvbt2/PBBx/Qt29fr/3KlCnDyJEjvVYf161blyFDhjB8+HA+/fRT98rOszV9+nTee+89atasyZNPPum17fDhwwBERkb63dfVnpmZedrjTJs2jfT0dO6++25uvvlmr22PPfYY48ePZ86cOezatcunFvAjjzxCnTp13O8DAgJ47bXXmDp1KhMmTOD555/36h8aGupz/AoVKrh/z8/PZ8yYMVSrVo233nqLMmXKuLeVKVOGN954g48++ohJkybRrVs3r3GqVq3q91rXqFGDdu3aMWfOHNatW0fDhg3d21xlMTxXEn///fds27aNxx9/nNatW3uNVb16dZ588kmGDh3KTz/9ROfOncnJyWHSpEmUL1+eYcOGefW//vrr6dev31mV3+jfvz+zZ89m4sSJvPbaa+72ZcuWsX79enr06OH1IMno6Gi/4zRq1Ii2bdsye/ZscnJy3CvLXQq7XoUJDg6mZs2aPu2RkZEMGDCAxx57jKVLl/KXv/zFp8/o0aO95j46Oppu3brxySefsGnTJq6++uozjkNEREREREREpLgp+SwX1IoVKwgICCA2NtZnW+vWrSlTpgwrV6702Va7dm13nVpPsbGxDB8+3O8+Z+KXX36hb9++hIeH880333glaM+EddbSLVi3158VK1YA0LZtW59tgYGB/OUvfyE5OZmVK1f6JJ8LJmcBLr30UmrVqkVycjKZmZlERUXRu3dvRo0aRffu3enVqxft27enZcuWXHbZZV77bt68mfT0dK644gpefvllv/GGhob6rWXcqFEjgoOD/e4TFxfHnDlzmDhxIv/5z38ARw3j//3vf1SpUsWrhMbixYsB2Llzp08yGWDLli0AbNiwgc6dO7Nx40aysrK4+eab/f4xIDY29qySzz169CAyMpLPPvuMESNGuBPwrjHi4uJ89vn+++8ZP348y5YtIy0tjdzcXK/taWlpXHLJJV5tRV2vwqxbt47XXnuN+fPnk5KSQnZ2ttf2vXv3+uwTGRnJ5Zdf7tNeq1YtwFG6RkRERERERESkJCn5LBfU4cOHqVixIkFBQT7bAgMDqVy5MgcOHPDZVrVqVb/jVatWzT3u2Vq8eDGdOnUiICCAmTNncuONN/r0cSU5Cxvf9cC/wlZGe3KNUTA56eJq97eKuqjz37lzJ4cPHyYqKoobb7yRBQsW8MorrzB58mR3zeb69esTHx/P3XffDThqJ4MjwTt8+PBCYz527JjfYxamR48eRERE8Nlnn/Hqq69SpkwZZsyYQXp6OkOHDvWqqe2K4euvi15p74rBdf1O91k4U6Ghodx11128//77zJ49m06dOpGTk8MXX3xBlSpV6NSpk1f/0aNH88gjj1ChQgU6dOhA7dq1CQsLwxjD1KlTWb16NSdPnjzvuJYsWULbtm3Jzc2lXbt2dO3alYiICAICAli1ahXTpk3ze5yoqCi/47mueV5e3lnFISIiIiIiIiJS3JR8vlAuYCmL35PIyEgyMjL8lifIzc0lLS3N78PYUlNT/Y63f/9+97hnY8GCBdx2220EBASQkJBA8+bN/farX78+4Fgp7I9rdW69evVOe0xXjK6YC0pJSfHq5yk1NdUdiyd/53/TTTcxY8YMTp48yfLly5k1axbvvPMOffv2pUqVKrRv397dv0ePHnz77benjd1TUau8XQndDz74gDlz5nDrrbf6LbnhGfO0adPo2rXraY/r6n+6z8LZ6N+/P++//z4TJ06kU6dO7kT5I4884vX5zM3NJT4+nmrVqrFixQqfPyC4VnH7cyar4j29/PLLnDhxgsTERJ//Q+DVV19l2rRpZzWeiIiIiIiIiEhpEXD6LiKFc5UuKGyV5XXXXUd+fj7z58/32TZ//nzy8vJo0qSJz7Zdu3aRnJzs0z5v3jz3uGdq7ty5dOrUicDAQObMmVNo4hmgTZs2AMyePZv8/HyvbUePHmXRokWEhoYWOYaLK0ZXzJ5yc3NZuHAhgN/z//nnn33atm/fzu7du4mJifG76jU4OJgWLVrw4osvMnr0aAB34rJBgwZERUWxZMkScnJyThv72XCVq5g4cSJpaWnMnDmTa6+9lsaNG3v1c12zBQsWnNG4DRo0ICwsjFWrVvldie7vup5Oy5YtueKKK5g2bRqHDx8uNFGelpZGZmYmLVq08Ek8Hzt2zF1SpThs3bqVihUr+i1N4+9zcC5Od5+KiIiIiIiIiFwISj7LealQoQLGGHbt2uV3+4ABAwB45plnyMrKcrdnZWXx9NNPA3D//ff77JeXl8dTTz3llQDesWMHo0ePJjAwkHvuueeM4ps9eza33347ISEh/PTTT9xwww1F9r/sssvo2LEjycnJvPvuu17b4uPjOX78OPfeey/h4eGnPXb37t2pWLEiX3zxBUuWLPHa9vbbb7N9+3bat2/vU+8ZYNSoUezcudP9Pj8/nyeeeIL8/Hzuu+8+d/uCBQv8JmZdq4XDwsIARymGwYMHk5KSwpAhQzhx4oTPPikpKaxfv/6051WQZ0J33Lhx5OTk+K2f3K1bNy677DLeffddfvjhB79jLV682P05KVu2LP369ePo0aM+NaKXLVvGpEmTzjpWcCSas7OzGTt2LD/88APXXnutzx8zoqOjCQsLY/ny5V6lSHJycnjkkUdIS0s7p2P7ExMTQ0ZGBklJSV7tH374IQkJCcVyjNPdpyIiIiIiIiIiF4LKbsh5KVeuHM2aNWPBggX069ePevXqUaZMGbp27cq1115L3759mTZtGl999RUNGzake/fu7pq5O3bs4K677qJfv34+41577bX8+uuvNG3alI4dO3L48GG+/PJLMjMz+c9//uPzQD1/Nm3aRLdu3cjOzqZz585MmzbNbwmDgonNsWPH0qJFC4YMGcJPP/3ElVdeya+//kpiYiL16tXjlVdeOeNrM2HCBO68805at27NnXfeSe3atVm+fDmzZ8+mWrVqvPfee373bdmyJY0bN6Z3795ERkaSkJDA6tWradq0KU8++aS73xtvvMHs2bOJjY3l0ksvpVy5cqxbt46ZM2dSoUIFHnroIXff559/ntWrVzN+/Hi+++472rZtS40aNThw4ABbtmxh0aJFvPLKK1x11VVndH6e7r33Xp5//nleeuklAgMD6du3r0+fsmXL8u2333LLLbdw22230aJFCxo3bkxYWBi7d+9m6dKlbN++nZSUFHfS/N///jc//fQTb7/9NsuWLaNVq1akpKTw5Zdf0rlzZ6ZPn35Osb7wwgvEx8eTk5Pjs+oZICAggCFDhjBixAiuueYaunXrxqlTp0hMTCQjI4M2bdqQmJh41sf2Z+jQoSQkJNCqVSvuuusuIiMjWbZsGQsXLqRXr15Mnjz5vI9xuvtUREREREREROSCsNbqx+Nn7NixduzYsVbO3JYtW+ztt99uK1asaI0xFrAfffSRe3teXp599913bdOmTW1oaKgNDQ21TZo0sWPGjLF5eXk+4wG2devWdu/evbZfv362SpUqNjg42F533XV20qRJZxxXYmKiBU7748+uXbtsXFycrVatmi1btqytXbu2HTJkiE1PTz/r6/Pbb7/Z7t2728qVK9uyZcvaWrVq2b///e927969Pn379+9vAbtt2zb7+uuv2/r169vg4GBbvXp1+8gjj9jDhw979U9ISLBxcXH2yiuvtBERETYsLMzWq1fPDh482CYnJ/uMn5+fbz/55BPbtm1bW6FCBVu2bFlbvXp127JlS/vKK6/YXbt2ufvu2LHDArZ///6nPcedO3fagIAAC9jbb7+9yL6pqan2qaeesg0bNrShoaE2PDzcXn755faOO+6wn376qc3JyfHqn5KSYu+77z5buXJlGxISYhs1amQ/+ugj9/zGx8efNr6C2rVrZwEbGBho9+/f77dPTk6OfeONN+yVV15pQ0JCbNWqVe0999xjk5OT3fO0Y8cOd/8zuV6tW7f2+5n77rvvbLNmzWy5cuVsZGSk7dChg/3555/tRx995HM/WWttnTp1bJ06dfweIz4+3gI2MTHRq/1096mIiIiIiIiI/GGVWK7VWGsvZq671Bs3bpwFGDhwYEmH8qdljKF169bnVNNXREREREREREREvJiSOrBqPouIiIiIiIiIiIhIsVPyWURERERERERERESKnZLPIiIiIiIiIiIiIlLsAks6AJGCVIdcRERERERERETk908rn0VERERERERERESk2Cn5LCIiIiIiIiIiIiLFTslnERERERERERERESl2Sj6LiIiIiIiIiIiISLFT8llEREREREREREREip2SzyIiIiIiIiIi8n/s3Xt0ndV94P3vlmTLyLZs624JO0AnhFQUJ1MKIelAghPalBcCaULyTtuQhMBbwlvSdoVp1wRYR8BM2uGdpE2aMIGQSxsm4JUQCjNlSgoE0qYB+SLLlmV8tyQLy7ZsLN9kWzr7/UPnqEd3H+lIsqzvZ62zznmeZ+9zfvjnZyP/tM/ekpRzFp8lSZIkSZIkSTln8VmSJEmSJEmSlHMWnyVJkiRJkiRJOWfxWZIkSZIkSZKUcxafJUmSJEmSJEk5Z/FZkiRJkiRJkpRzFp8lSZIkSZIkSTln8VmSJEmSJEmSlHMWnyVJkiRJkiRJOWfxWZIkSZIkSZKUcxafJUmSJEmSJEk5Z/FZkiRJkiRJkpRzFp8lSZIkSZIkSTln8VmSJEmSJEmSlHMWnyVJkiRJkiRJOVcw3QFIkgRww0PPjnr9uXtvnKJIJEmSJElSLlh8liSddfZ0Hut/XVM6fxojkSRJkiRJ42XxWZJ0Vsic2Zw5C9oZz5IkSZIkzUyu+SxJkiRJkiRJyjmLz5IkSZIkSZKknLP4LEmSJEmSJEnKuXGt+RxCmA/UAGXACWBfjHFPLgOTJEmSJEmSJM1cZ1x8DiH8CvBZ4EPAuxk0azqE0Am8DPwYeDrG2JPDOCVJkiRJkiRJM8iYxecQwuXAQ8AH6Ss4nwY2AHuBg8B5QCnwDuDjwMeA/SGEvwa+EmM8OZg2zpEAACAASURBVDmhS5IkSZIkSZLOVqMWn0MI3wd+DzgMPAo8CbweY+weof0FwG8BtwL/BfjDEMIfxBhfzWHMkiRJkiRJkqSz3FgbDl4H/ClQHWP8fIzx1ZEKzwAxxl0xxm/FGN8LvAtYC1yTu3AlSZIkSZIkSTPBWMtuXBRjPDGeN44xNgI3hxDOG0//EEIpcDNwPfBr9G1weIq+JT++C3w3xpgcpt97gXuB9wDzgG3Ad4Cvxxh7xxOLJEmSJEmSJCk7oxafx1t4ztF7fBx4BHiTvo0MW4BK4KPAt4EPhxA+HmOM6Q4hhI/Qt+FhN/AUfWtS3wB8FXhf6j0lSZIkSZIkSZNszA0Hz0QIYQ5wKXA8xvhGLt4T2ALcCPzvzBnOIYT/DLwO/C59hegfp84XA48BvcD7Y4yrU+fvA14CPhZC+GSM8ckcxSdJkiRJkiRJGsFYaz4PEEK4JYSwKoRQknHuV4AmYDWwKYTwdAhhwkXtGONLMcbnBi+tEWPcC/yP1OH7My59DCgHnkwXnlPtu+lbhgPgzonGJUmSJEmSJEkaW1bFZ+CzwCUxxoMZ5/478O/oWxqjEfgI8JnchDei06nnnoxz16ae/88w7V8FjgPvDSEUTmZgkiRJkiRJkqTsi8+/CtSnD1JLXfwOsCrG+EHgCmAzk1h8Ts2q/lTqMLPQ/I7U85bBfWKMPcBO+pYZuWiyYpMkSZIkSZIk9cm2+FxO3waAaVfRV9B9EiDGeBr4KfArOYlueH9B3/rS/xBj/MeM84tSz4dH6Jc+v3iyApMkSZIkSZIk9clPJBJn3Liuru4LwLZEIvF86vg24H3A3YlE4njq3PuBqxOJxH/JdbAhhLuBB+ibXf1/JRKJExmx/RFQCnw9kUgcHNy3rq7uc8Ay4NuJRKJtmPe+o66u7lvXX399NcDFF19Mc3MzZWVlrFu3jurqalavXk1paSlNTU0UFhayZ88ejh8/zrFjx2hvb2fu3Lls2rSJ8vJy1q5dO6BPc3Mzc+bMob29nWPHjnHixAna2tooLCxk48aNVFZWDuizZMkS3njjDfLz89m7dy9Hjhyhu7ub1tZWzjvvPBobG6mqqmLNmjX9fRYvXsyWLVsIIbBv3z66uro4deoUu3fvpqioiMbGRqqrq6mvr6empobVq1ezaNEitm3bRoyRAwcOcOjQIXp6eti1axcLFixg/fr1/e+f/rzi4mK2b99OMpnk4MGDdHZ2kkwm2bFjBwsXLhzQp7KyknXr1rFw4UJ27txJT08Phw4d4sCBAwBs27aNxYsX09DQ0N+noqKC9evXM3/+fHbt2sXp06c5fPgw+/btI4TAli1bWLJkyYA+5eXlNDY2UlRUREtLCydPnqSrq4uOjg7y8/PZvHkzpaWlrF27tv+/vaysjI0bN1JYWEhbWxvd3d0cPXqUN998k4KCgv78p/vU19dTVlbGpk2b+nM5OP9NTU1UVFQMyGVJSQnNzc0UFBT053Jw/gfnMp3/vLw8Ojo66Orq4uTJk7S0tPTnf+nSpSPmf//+/bz11lucPn16SP5Xr17d/1xcXDwk/729vezYsYMFCxbQ0NDQ/+dVVVXF2rVrR81/cXFxf17q6+upqqqioaGB+fPnD5v/rVu3jpj/oqIidu/ezalTp4bkv6SkZMB9WV5ezoYNG5g3bx4tLS10d3dz5MgR9u7dOyD/mX0y85++l48ePUp7eztz5syhubl52Hs5M/+Z93Jm/tesWdP/51ZSUsLmzZtHvJc3bNgwIP/19fXD5j99L6fzPziX6XsZGJD/4e7l9N+ddP6TySSdnZ0D8j/4Xk7nP/Nezsz/9u3bWbRo0YBcVlZW9uc/817ev3//gPyvW7eu/x6rqKigsbGRF5oP0n3iBMlkZOU7FtPR0UFeXt6Q/Kf7pPPf2to64r3sWO5Y7ljuWO5YPrVj+XD3smO5Y7ljuWO5Y7ljuWO5Y7lj+dSP5UuXLq3LqgibQyHGeOaNQ/gXoBKoBSJ9Gw0ejTG+O6PND4GrYowX5DTQEO4C/gbYBKxMbTyYeb0euBy4PMa4Zpj+G1Nx/2qMsXmkz3nkkUciwJ13ujehJE2XGx56tv/1c/feOI2RSJIkSZI044Xp+uBsl914lL41k7cCzanX3xnU5kr6itI5E0L4Y/oKzxuBDwwuPKe8kXq+eJj+BcCF9G1QuCOXsUmSJEmSJEmShsqq+Bxj/D59ay4X0bfG8t+kHgCEEK4FLgBezlWAIYQ/A74KNNBXeN43QtOXUs+/Pcy1q+mL+RcxxpO5ik2SJEmSJEmSNLxsZz4TY/zPMcay1OMLceC6Hf8MLAH+KhfBhRDuo6/YvYa+pTYOjNL8R8AB4JMhhMsz3mMe8FDq8JFcxCVJkiRJkiRJGl1BLt8sxngKOJWL9woh3Erf5oK9wM+Bu0MYsjzJrhjj91Kf3RVCuJ2+IvTPQghPAgeBG4F3pM4/lYvYJEmSJEmSJEmjG7X4HEI4L8Z4YiIfMIH3uDD1nA/88QhtXgG+lz6IMT4TQrgG+BLwu8A8YBvwp8DXYja7K0qSJEmSJEmSxm2sZTd2hhC+EEIozPaNQwgrQgh/D3xxPIHFGBMxxjDG4/3D9PuXGOPvxBiXxBjPizH+WozxqzHG3vHEIUmSJEmSJEnK3ljF5xeArwBvhhAeCSF8IIRw3kiNQwgXhRDuDCH8K7AWWEEONx+UJEmSJEmSJM0Moy67EWP8VAjha8B/Be5IPXpDCM3Am8Ah+pa2KKVvXeUyIAAd9C198dUY48nJC1+SJEmSJEmSdDYac8PBGONq4LoQwtuB24CVwLuAXxvUdD/wNPBj4McxxtM5jlWSJEmSJEmSNEOMWXxOizFuBf4cIIRQBNTQN+P5BLAvxvjmpEQoSZIkSZIkSZpxzrj4nCnGeBzYmnpIkiRJkiRJkjTAWBsOSpIkSZIkSZKUNYvPkiRJkiRJkqScs/gsSZIkSZIkSco5i8+SJEmSJEmSpJyz+CxJkiRJkiRJyjmLz5IkSZIkSZKknMuq+BxCWB5CKB6jzcIQwvKJhSVJkiRJkiRJmsmynfm8E/jCGG3uTrWTJEmSJEmSJM1SBVm2D6mHJEmToqmlk50dXfT0JinIz6OppZPa5aXTHZYkSZIkScpStsXnM1EJHJuE95UkncNeXN/KA6te49Wm9gHnL737Ca6ureb+W65k5Ypl0xSdJEmSJEnK1pjF5xDCpwadetcw5wDygeXAHwAbchCbJGmWePynTdzxzZdIxjjs9Veb2rku8QyP3XUtn/1g7RRHJ0mSJEmSxuNMZj5/D0hXAyLwkdRjsPRyHMeBuglHJkmaFV5c3zpq4TktGSO3f+Ml3lZe7AxoSZIkSZJmgDMpPn8m9RyA7wDPAH8/TLteoBP41xjjW7kJT5J0rntg1WtjFp7TkjHy4KrXLT5LkiRJkjQDjFl8jjF+P/06hHAr8EyM8W8nNSpJ0qzQ1NI5ZI3nsbzStMdNCCVJkiRJmgGy2nAwxviByQpEkjT7vNjYOu5+Fp8lSZIkSTq75WXTOITw70MI94cQKke4XpW6/q7chCdJOpd1HT81pf0kSZIkSdLUyar4DHwR+Bywb4TrHcBtwJ9OJChJ0uxQXDR3SvtJkiRJkqSpk23x+Srg5RiH3xkqdf4l4H0TDUySdO5bedn4Ng4cbz9JkiRJkjR1si0+VwFtY7RpB5aOLxxJ0mxSu7yUq2urs+pzTW2N6z1LkiRJkjQDZFt8Pg6Uj9GmHDg5vnAkSbPN/bdcSV4IZ9Q2LwTuu+WKSY5IkiRJkiTlQrbF5wbgIyGEBcNdDCEUAx9JtZMkaUwrVyzj0c9fO2YBOi8EHrvrWlaucMkNSZIkSZJmgmyLz4/SN7P5pyGEyzIvhBBWAC8AZal2kiSdkds+VMsLiZu4prZm2OvX1NbwQuImPvvB2imOTJIkSZIkjVdBNo1jjE+FED4MfApYF0LoAPYANUAlEIDvxxh/mPNIJUnntJUrlrFyxTKaWjr5xMPP09ObpCA/j6fu+bBrPEuSJEmSNANlVXwGiDF+OoTwC+CPgFr6NiEE2Ah8Lcb47RzGJ0maZWqXl3JhZfGAY0nnvhseenbU68/de+MURSJJkiQpV7IuPgPEGB8FHg0hFAGLgbdijMdzGpkkSZJmpT2dx/pf15TOn8ZIJEmSJE3EuIrPaamCs0VnSZIkTUjmzObMWdDOeJYkSZJmrnEVn0MI5cDvAu8E5scYP5dx/kJgQ4zxRM6ilCRJkiRJkiTNKFkXn0MItwFfA+bRt8FgBD6XulwJ/CtwB/B4jmKUJEmSJEmSJM0wedk0DiF8CHgU2ALcDDySeT3GuBFoAm7KVYCSJEmSJEmSpJkn25nPfwa8CVwTY+wKIbx7mDaNwFUTjkySJEmSJEmSNGNlNfMZuBz4XzHGrlHatAFV4w9JkiRJkiRJkjTTZVt8ngscG6PNYqB3fOFIkiRJkiRJks4F2RafdwG/PkabK4E3xhWNJEmSJEmSJOmckG3x+e+B/xBC+PhwF0MInwEuA3480cAkSZIkSZIkSTNXthsO/jfgk8APQwgfAxYBhBD+X+A/AB8FtgJfz2WQkiRJkiRJkqSZJavic4zxUAjhGuBvgczZz19LPf8c+I8xxrHWhZYkSZIkSZIkncOynflMjLEFeH8I4TLgKqAUOAz8Msa4JsfxSZIkSZIkSZJmoFGLzyGEp4EnY4yrUsdXA7tijC0xxkagcQpilCRJ0izR1NLJzo4uenqTFOTn0dTSSe3y0ukOS5IkSdI4jDXz+SagIeP4ZaAOeGDSIpIkSdKs8+L6Vh5Y9RqvNrUPOH/p3U9wdW01999yJStXLJum6CRJkiSNR94Y198CijOOwyTGIkmSpFno8Z82cV3imSGF57RXm9q5LvEM3/mnpimOTJIkSdJEjDXzuRn4v0MI9cCbqXMXpJbfGFWM8dWJBidJkqRz24vrW7njmy+RjHHUdskYuf0bL/G28mJnQEuSJEkzxFjF5zrgGeB/Zpy7NfUYS/54g5IkSdLs8MCq18YsPKclY+TBVa9bfJYkSZJmiFGLzzHGF0II7wQ+CNQACeCV1EOSJEkat6aWzhGX2hjJK0173IRQkiRJmiHGmvlMjHE38DhACCEB/CzG6IaDkiRJmpAXG1vH3c/isyRJknT2G3XDwRDCV0II12Wc+gx9y3BIkiRJE9J1/NSU9pMkSZI0tUYtPgN/DLwn4/g7wE2TF44kSZJmi+KiuVPaT5IkSdLUGqv4fBQoyjgOkxiLJEmSZpGVl41v48Dx9pMkSZI0tcZa83kb8NEQwk+AN1PnFocQlo/1xjHGlokGJ0mSpHNX7fJSrq6tzmrTwWtqa1zvWZIkSZohxpr5/DDwK8AvgJ2pc19IvR7tsSMXwYUQPhZC+HoI4echhK4QQgwh/GCEthekro/0eDIXMUmSJCl37r/lSvLCmX25Li8E7rvlikmOSJIkSVKujDrzOcb4wxDCTuB6oAb4NNAINEx+aADcC6ygb/mPNuCSM+iznuE3RdyYw7gkSZKUAytXLOPRz1/LHd98iWSMI7bLC4HH7rqWlStcckOSJEmaKcZadoMY4y+BXwKEED4N/CTG+MAkx5X2J/QVnbcB1wAvn0GfhhhjYjKDkiRJUu7c9qFaLqgo5sFVr/NK054h16+preG+W66w8CxJkiTNMGMWnwf5DLBuMgIZToyxv9gczvDrmJIkSZp5Vq5YxsoVy2hq6eQTDz9PT2+Sgvw8nrrnw67xLEmSJM1QWRWfY4zfn6xAcqg6hPD/AKVAJ/CvMcbGaY5JkjRD3fDQs6Nef+7eG6coEml2qF1eyoWVxQOOJUmSJM1M2c58JoSQB9wF/B7wTmB+jLEgde3dwO3AX8UYt+Qy0Cx8KPXoF0L4GXBrjLFlWiKSJJ0T9nQe639dUzp/GiORJEmSJOnsl59IJM64cQhhLvCPwB8C5wHHgYWJROIBgLq6um7gCaArkUj8LJeB1tXVXUDfhocbEonE08Ncnwt0A38K/Dnwl8DzwAXAB4Ab6urqvpNIJE4P9/4hhDvq6uq+df3111cDXHzxxTQ3N1NWVsa6deuorq5m9erVlJaW0tTURGFhIXv27OH48eMcO3aM9vZ25s6dy6ZNmygvL2ft2rUD+jQ3NzNnzhza29s5duwYJ06coK2tjcLCQjZu3EhlZeWAPkuWLOGNN94gPz+fvXv3cuTIEbq7u2ltbeW8886jsbGRqqoq1qxZ099n8eLFbNmyhRAC+/bto6uri1OnTrF7926KiopobGykurqa+vp6ampqWL16NYsWLWLbtm3EGDlw4ACHDh2ip6eHXbt2sWDBAtavX9///unPKy4uZvv27SSTSQ4ePEhnZyfJZJIdO3awcOHCAX0qKytZt24dCxcuZOfOnfT09HDo0CEOHDgAwLZt21i8eDENDQ39fSoqKli/fj3z589n165dnD59msOHD7Nv3z5CCGzZsoUlS5YM6FNeXk5jYyNFRUW0tLRw8uRJurq66OjoID8/n82bN1NaWsratWv7/9vLysrYuHEjhYWFtLW10d3dzdGjR3nzzTcpKCjoz3+6T319PWVlZWzatKk/l4Pz39TUREVFxYBclpSU0NzcTEFBQX8uB+d/cC7T+c/Ly6Ojo4Ouri5OnjxJS0tLf/6XLl06Yv7379/PW2+9xenTp4fkf/Xq1f3PxcXFQ/Lf29vLjh07WLBgAQ0NDf1/XlVVVaxdu3bU/BcXF/fnpb6+nqqqKhoaGpg/f/6w+d+6deuI+S8qKmL37t2cOnVqSP5LSkoG3Jfl5eVs2LCBefPm0dLSQnd3N0eOHGHv3r0D8p/ZJzP/6Xv56NGjtLe3M2fOHJqbm4e9lzPzn3kvZ+Z/zZo1/X9uJSUlbN68ecR7ecOGDQPyX19fP2z+0/dyOv+Dc5m+l4EB+R/uXk7/3UnnP5lM0tnZOSD/g+/ldP4z7+XM/G/fvp1FixYNyGVlZWV//jPv5f379w/I/7p16/rvsYqKChobG3mh+SDdJ06QTEZWvmMxHR0d5OXlDcl/uk86/62trSPey+Mdy7/4yQ9wcdER/ujm97Lq1SYWzy9kTl6Sb3/uCm76jeWO5Y7ljuWTMJb/4OVNJJNJenp6eM/5cxzLZ+hYPty9PF1juT+XO5Y7lvtzuWO5Y7ljuWP5bB7Lly5dWnemNdhcC3GUXcWHNA7hS8CDQAJ4CLgfuC/GmJ/R5gWgOMb4npwGGsL76dtw8IkY4+9n0a8A+GfgSuCPY4x/PVr7Rx55JALceeed4w9WkjQhmUtdnE3LWpytcUnnGu81SZIkKaembTO9vCzb/x7wLzHGB2KMSWC4yvVOYPmEI8uRGGMP8O3U4dXTGYskSZIkSZIkzRbZFp8vBH45RpuDQMn4wpk0+1PPLtApSZIkSZIkSVMg2+LzCWDxGG2WA2+NL5xJk14CZMe0RiFJkiRJkiRJs0S2xecG4LrUxoNDhBAWAb8FvD7RwLIVQrhyuLhCCNcCf5I6/MHURiVJkiRJkiRJs1NBlu0fA54Anggh3JZ5IYSwGPgusAT4H7kILoRwE3BT6rAq9XxVCOF7qdcHYoxfTL3+S6A2hPAzoC117jLg2tTr+2KMv8hFXJIkSZIkSZKk0WVVfI4x/jCE8EHgM8CNwCGAEMJqoBYoBL4RY/yHHMX3LuDWQecuSj0AdgPp4vPfATcDvwF8GJgDdACrgL+JMf48RzFJkiRJkiRJksaQ7cxnYoy3hRB+DnyBvpnFAfj3QBPwlRjjd3MVXIwxASTOsO3jwOO5+mxJkiRJkiRJ0vhlXXwGiDF+D/heCOE8+pbZOBxjPJbLwCRJkiRJkiRJM9e4is9pMcYTwIkcxSJJkiRJkiRJOkfkTXcAkiRJkiRJkqRzj8VnSZIkSZIkSVLOTWjZDUmSZoumlk52dnTR05ukID+PppZOapeXTndYkiRJkiSdtSw+S5I0ihfXt/LAqtd4tal9wPlL736Cq2uruf+WK1m5Ytk0RSedO2546Nn+169v7Rhy/rl7b5zymCRJkiRNjMtuSJI0gsd/2sR1iWeGFJ7TXm1q57rEM3znn5qmODLp3FZTsqD/IUmSJGnmcuazJEnDeHF9K3d88yWSMY7aLhkjt3/jJd5WXuwMaGkCnNksSZIknXtyMvM5hFAWQrg5hPBbIYT8XLynJEnT6YFVr41ZeE5LxsiDq16f5IgkSZIkSZpZsio+hxDuDCG8FkIoyTj360Az8CPgH4BfhBDm5zZMSZKmTlNL54hLbYzklaY9NLV0TlJEkiRJkiTNPNnOfP4EEGOMBzPOPQwsAb5LX/H5N4A/zE14kiRNvRcbW6e0nyRJkiRJ56Jsi89vBxrTByGEMuAa4PEY4+dijDcA9cB/zF2IkiRNra7jp6a0nyRJkiRJ56JsNxwsBfZlHL8v9fyTjHM/Bz49gZgkSbPQDQ892//69a0dQ85P5WZkxUVzp7SfJEmSJEnnomyLzweBsozja4Ak8IuMcxGYN8G4JEmzWE3Jgmn9/JWXLZvSfpIkSZIknYuyLT43AzeEEL4E9NK3BnR9jLEro80FwN7chCdJmi2mcmbzWGqXl3J1bXVWmw5eU1tD7fLSSYxKkiRJkqSZJds1n/8aWAq0Aa1AFfDN9MUQQj7wm8D6XAUoSdJ0uP+WK8kL4Yza5oXAfbdcMckRSZIkSZI0s2RVfI4xPgv8IdAEvAF8Mcb4g4wmH6RvyY1/zFmEkiRNg5UrlvHo568dswCdFwKP3XUtK1e45IYkSZIkSZmyXXaDGOOjwKMjXPtHYMlEg5Ik6Wxw24dquaCimAdXvc4rTXuGXL+mtob7brnCwrMkSZIkScPIuvgsSdJssnLFMlauWEZTSyefePh5enqTFOTn8dQ9H3aNZ0mSJEmSRpHVshshhHeHED4fQliUcW5+COH7IYS3QgjtIYQv5D5MSZKmV+3yUi6sLObt1Yu5sLLYwrMkSZIkSWPIdsPBPwO+FGM8nHHuy8AfpN6rFPhKCOG6HMUnSZIkSZIkSZqBsi0+Xw78LH0QQpgD3Aq8DlQAFwIHgLtzFJ8kSZIkSZIkaQbKtvhcAbRmHF8OLAS+FWPsjjG2A38PXJaj+CRJkiRJkiRJM1C2xefIwE0KfzN17pWMc/uB8gnGJUmSJEmSJEmawbItPrcA78k4/gjQFmPckXGuGjg00cAkSZIkSZIkSTNXtsXnVcB7Qwg/CiH8ALgK+NGgNpcC23MRnCRJkiRJkiRpZioYu8kAXwV+G/ho6rgBeCB9MYTwq8CvA/81J9FJkiRJkiRJkmakrIrPMcajwPtCCJemTm2KMSYzmhwHbgZW5yg+SdIscMNDz456/bl7b5yiSCRJkiRJUq5kO/MZgBjjxhHO7wJ2TSAeSdIst6fzWP/rmtL50xiJJEmSJEmaiHEVn0MIc4CVwDuBBTHGB1Pn5wHFwIFBM6IlSRpR5szmzFnQzniWJEmSJGnmynbDQUIIv03f7Ob/Dfx3IJFx+V3Am8AnchCbJEmSJEmSJGmGyqr4HEK4HHgGiMCfAP8z83qM8ZfATvrWfZYkSZIkSZIkzVLZzny+j75NBS+PMX4N2DpMm3pgxUQDkyRJkiRJkiTNXNkWn98HPBNj3DtKm1Zg6fhDkiRJkiRJkiTNdNluOLgAODBGmyLGsZa0JEmSJE3Ew5t7Rr1+zyXj2m9dkiRJ45TtT197gNox2rwL2DG+cM4eNzz07JBzz9174zREIkmSJClbh07F/tdL5oZpjESSJGn2yrb4/DzwhyGE34wx/vPgiyGEDwPvBf4iF8FNpz2dx/pf15TOn8ZIJEmSJJ2JzJnNmbOgnfEsSZI0PbL9KezLwCeBF0IIXwcuAAghXA9cDdwFvAl8JYcxTovMgrMzniVJkiRJkiQpO1kVn2OMe0II1wGrgHsyLj0LBGA78NEY41jrQkuSNERTSyc7O7ro6U1SkJ9HU0sntctLpzssSZIkSZI0Dll//yzGuDaE8A7geuAqoBQ4DPwS+PsY4+i7fEiSNMiL61t5YNVrvNrUPuD8pXc/wdW11dx/y5WsXLFsmqKTJEmSJEnjMa7Fz2KMvfTNdh66K58kSVl4/KdN3PHNl0jGOOz1V5vauS7xDI/ddS2f/eBYe95KkiRJkqSzRd50ByBJmr1eXN86auE5LRkjt3/jJV5c3zpFkUmSJEmSpIka18znEMLlwBXAEiB/mCYxxvjgRAKbbq45KkmT74FVr41ZeE5LxsiDq153+Q1J0pjajkf2nYwkI+SFvuPzi8J0hyVJkjTrZFV8DiEUA08DH6Bvg8GRRGBGF5+bWg/2v3bNUUnKvaaWziFrPI/llaY9/kJQkjSijYeTPN2WZPORgb/Y/E+NPVyyMPDR8/O4dJFf/pQkSZoq2c58fhi4Fvg58F2gFZgVGwy65qgk5daLjeNbQuPFxlaLz5KkIV7el+TbO3oZ6fs0m49Evtzcy+0XwfsrLEBLkiRNhWyLzx8B1gIfiDEmJyGes1p6zdG3lRc7A1qSJqjr+Kkp7TdeNzz0b3vrvr61Y8j55+69cUrjkSQNtfHw6IXntAg8tqOXskKcAS1JkjQFsv2JaxHw8mwsPKel1xyVJE1McdHcKe2XCzUlC/ofkqSzx9NtyTELz2kR+EnbrP3njCRJ0pTKdubzVqByMgKZSVxzVJImbuVl4/sGyXj7jZczmyXp7NZ2PA5Z43kszUeimxBKkiRNgWxnPn8DuCGEUDMZwcwk412rVJLUp3Z5KVfXVmfV55raGn/xJ0kaYOPh8c1iHm8/SZIknblsi8/PAy8A/xJC+EwI4bIQwvLhHpMQ61llqtcclaRz0f23XEleOLNZZ3khcN8tV0xyRJKkmeZE79T2kyRJ0pnLdtmNXfQtkxaAb4/SLo7jvWeU6VxzVJLOFStXLOPRz1/LHd98iWQc+SvTeSHw2F3XZx311gAAIABJREFUutmrJGmI8/Kntp8kSZLOXLYF4r+FM97L45w21WuOStK56rYP1XJBRTEPrnqdV5r2DLl+TW0N991yhYVnSdKwLl2UB2S/hEZfP0mSJE2mrIrPMcZPT1IcQ4QQPgZcA7wLWAEsBJ6IMf7+KH3eC9wLvAeYB2wDvgN8PcaYsy/WueaoJOXWyhXLWLliGU0tnXzi4efp6U1SkJ/HU/d82PFWkjSq84sClywMWW06+M6Fwc0GJUmSpsDZvDTGvfQVnY8CbcAlozUOIXwE+DHQDTwFHARuAL4KvA/4eC6Ccs1RSZo8tctLubCyeMCxJElj+ej5eXy5ufeMvqIZgJvPd9azJEnSVMjqp64QQm8I4b4x2nwphNAzsbAA+BPgYqAYuHOMzywGHgN6gffHGG+LMd5D36zpfwU+FkL45EQDcs1RSZIk6exz6aI8PndRPmPNZQ7A7Rflu+SGJEnSFMl25nNIPc6k3YTEGF/uf7Mw5tt9DCgH/jbGuDrjPbpDCPcCL9JXwH5yvPG45qgkSZJ09vpARR7lhfCTtiTNwyzB8c6FgZvPz7PwLEmSNIUmY9mNJfQtfTGVrk09/59hrr0KHAfeG0IojDGePJM3rF1W4pqjkiRJ0gxy6aK+4nLb8chfb+0hGSEvwBfeXuAaz5IkSdNgzOJzCOHqQacuGOYcQD6wHPg94I0cxJaNd6Setwy+EGPsCSHsBGqBi4DmM3lD1xyVJEmSZqbziwIVhWHAsSRJkqZefiKRGLVBXV3dTuBW4NOpU+9KvR78+BRwE1AE3JNIJDbkKsi6uroLUp+xIZFIPD3M9XuAUuDhRCJxcJjrvwcsA55IJBJtw31GCOGOurq6b11//fXVAJuOFHH06FHmzp3LO+Yfpbq6mtWrV1NaWkpTUxOFhYXs2bOH48ePc+zYMdrb25k7dy6bNm2ivLyctWvXDujT3NzMnDlzaG9v59ixY5w4cYK2tjYKCwvZuHEjlZWVA/osWbKEN954g/z8fPbu3cuRI0fo7u6mtbWV8847j8bGRqqqqlizZk1/n8WLF7NlyxZCCOzbt4+uri5OnTrF7t27KSoqorGxkerqaurr66mpqWH16tUsWrSIbdu2EWPkwIEDHDp0iJ6eHnbt2sWCBQtYv359//unP6+4uJjt27eTTCY5ePAgnZ2dJJNJduzYwcKFCwf0qaysZN26dSxcuJCdO3fS09PDoUOHOHDgAADbtm1j8eLFNDQ09PepqKhg/fr1zJ8/n127dnH69GkOHz7Mvn37CCGwZcsWlixZMqBPeXk5jY2NFBUV0dLSwsmTJ+nq6qKjo4P8/Hw2b95MaWkpa9eu7f9vLysrY+PGjRQWFtLW1kZ3dzdHjx7lzTffpKCggObmZsrKyvr71NfXU1ZWxqZNm/pzOTj/TU1NVFRUDMhlSUkJzc3NFBQU9OdycP4H5zKd/7y8PDo6Oujq6uLkyZO0tLT053/p0qUj5n///v289dZbnD59ekj+V69e3f9cXFw8JP+9vb3s2LGDBQsW0NDQ0P/nVVVVxdq1a0fNf3FxcX9e6uvrqaqqoqGhgfnz5w+b/61bt46Y/6KiInbv3s2pU6eG5L+kpIR169YNyP+GDRuYN28eLS0tdHd3c+TIEfbu3Tsg/5l9MvOfvpePHj1Ke3s7c+bMobm5edh7OTP/mfdyZv7XrFnT/+dWUlLC5s2bR7yXN2zYMCD/9fX1w+Y/fS+n8z84l+l7GRiQ/+Hu5fTfnXT+k8kknZ2dA/I/+F5O5z/zXs7M//bt21m0aNGAXFZWVvbnP/Ne3r9//4D8r1u3rv8eq6io4NHn15Kfn0/3iRN86FdL++/lvLy8IflP90nnv7W1dcR7OTP/juWO5Y7ljuWO5ZM7lo90L0/lWN64sYnkwjLm7G7gPW+vcSx3LHcsdyx3LJ+BY7k/lzuWO5bnZixfunRp3Rjl10kTYhx9T+gQQgKI9K3jfD/wM+CVYZr2Ap3AyzHGzTkNMoT3Ay8DT8QYf3+Y61uAtwNvjzFuG+b6L4CrgKtijL8c7bMeeeSRCPAPnTX9556798aJhC9JysINDz3b/9rxV5I0Xg9v/rc90O+5ZDJWG5QkSZoxpu1rYGP+FBZjTKRfhxBuBZ6JMX5tMoMah8Op50UjXC8e1E6SJEmSJEmSNImymgIQY7xwsgKZoDeAy4GLgTWZF0IIBcCFQA+wY+pDkyRJkiRJkqTZJ2+6A8iRl1LPvz3MtavpW4f6FzHGk1MXkiRJkiRJkiTNXqPOfA4hvETfes+3xhjbUsdnIsYYV044ujP3I+AvgU+GEL4eY1wNEEKYBzyUavPIFMYjSZIkSZIkSbPaWMtuvD/1XDToeCyj72J4BkIINwE3pQ6rUs9XhRC+l3p9IMb4RYAYY1cI4Xb6itA/CyE8CRwEbgTekTr/1ERjkiRJkiRJkiSdmVGLzzHGvNGOJ9m7gFsHnbso9QDYDXwxfSHG+EwI4RrgS8DvAvOAbcCfAl+LMU64IC5JkiRJkiRJOjNjLbtxN/DLGOPrUxRPvxhjAkhk2edfgN+ZjHgkSZIkSZIkSWdurJnMf0XGJn4hhN4Qwn2TG5IkSZIkSZIkaaYbq/jcDRRmHIfUQ5IkSZIkSZKkEY1VfN4J/FYIoTLjnGsnS5IkSZIkSZJGNeqaz8C36Ft6oz2E/gnPiRBCYox+McY41ntLkiRJkiRNyA0PPTvq9efuvXGKIpEkDTZqgTjG+LUQwj7geqAa+ADQAuya/NAkSZIkSZLO3J7OY/2va0rnT2MkkiQYe+YzMcYngScBQghJ4LsxxgcmOzBJkiRJkqSxZM5szpwF7YxnSZp+Y635PFgd8LNJiEOSJEmSJEmSdA7Jal3mGGPdZAVytnl9a0f/6/RvTv2tqSRJknT2enhzT//r7UfjkPP3XOK2NJIkSVPJn75GUFOyYLpDkCRJkjROJXOnOwJJU62ppZOdHV309CYpyM+jqaWT2uWl0x2WJM1qIcY4dqtZ5JFHHokAd95553SHIkmzRubafJnfPLni7ZXA8N88cVdzSZIkAby4vpUHVr3Gq03tQ65dXVvN/bdcycoVy6YhMkk6a4Tp+mBnPkuSzirj+eaJu5pLkiTNTo//tIk7vvkSyREm1r3a1M51iWd47K5r+ewHa6c4OkmSxWdJ0rQbzyzldJ+mlk4+8fDz/V+v/ItPvc+vV0qSJM0CL65vHbXwnJaMkdu/8RJvKy92BrQkTTGLz5KkGWmkr1deevcTfr1SkiRpFnhg1WtjFp7TkjHy4KrX/flQkqZY3nQHIElSth7/aRPXJZ4Zdl0/+LevV37nn5qmODJJkiRNhaaWzhF/FhzJK017aGrpnKSIJEnDsfgsSZpRsv165YvrW6coMkmSJE2VFxvH9zPeePtJksbHZTckSTOKX6+UJElT6eHNPaNev+cS/1k9HbqOn5rSfpKk8XHmsyRpxvDrlZIkaTodOhX7H5pexUVzp7SfJGl8/BWtJGnGmMjXK2uXl+Y4GkmSNBtkzmzOnAXtjOfptfKy8X2zbbz9JEnj48xnSdKM4dcrJUnSdGk7Htl3MrK3u++57bizn6dT7fJSrq6tzqrPNbU1TkiQpCnmr2olSTOGX6+UJElTbePhJE+3Jdl8ZGCx+T819nDJwsBHz8/j0kXO65oO999yJdclnjmj/UDyQuC+W66YgqgkSZn8P6Qkacbw65WSJGkqvbwvyZebe4cUntM2H4l8ubmXn+1LTnFkAli5YhmPfv5a8kIYtV1eCDx217VuQi1J08DisyRpxvDrlZIkaapsPJzk2zt6GWtObQQe29HLxsMWoKfDbR+q5YXETVxTWzPs9Wtqa3ghcROf/WDtFEcmSQKX3ZAkzTB+vVKSJE2Fp9uSYxae0yLwk7aky29Mk5UrlrFyxTKaWjr5xMPP09ObpCA/j6fu+bCTECRpmvl/RknSjOLXKyVJ0mRrOx5HXGpjJM1H3IRwutUuL+XCymLeXr2YCyuLLTxL0lnA4rMkacbx65WSJGkyjXcJDZfekCRpIJfdkCTNSH69UpIkTZYTvVPbT5Kkc5XFZ0nSjJb+emXmsSRJ0kSclz+1/SRJOle57IYkSZIkSRnGu3GgGw5KkjSQM58lSTPSDQ892//69a0dQ84/d++NUx6TJEk6N5xfFLhkYchq08F3LgycXzT6hsiSJM02/lpWkjTj1ZQs6H9IkiTlwkfPz+NMS8kBuPl8/3ktSdJgznyWJM1IzmyWJEmT6dJFeXzuIvj2jl5Gm/8cgNsvynfJDUmShmHxWZIkSZKkYXygIo/yQvhJW5LmYZbgeOfCwM3n51l4liRpBBafJUmSJEkawaWL+orLbccjf721h2SEvABfeHuBazxLkjQGi8+SJEmSJI3h/KJARWEYcCxJkkZn8VmSJEmSJM1YNzz0bP/r17d2DDnvXiGSNH0sPkuSJEmSpHNCTcmC6Q5BkpTB4rMkSZIkSZqxnNksSWcvt+SVJEmSJEmSJOWcxWdJkiRJkiRJUs5ZfJYkSZIkSZIk5ZzFZ0mSJEmSJElSzll8liRJkiRJkiTlXMF0ByBJkiRJ0tnq4c09/a+3H41Dzt9zif+sliRpJP5fUpIkSZKkM1Ayd7ojkCRpZrH4LEmSJEnSCJzZLEnS+LnmsyRJkiRJkiQp5yw+S5IkSZIkSZJyzuKzJEmSJEmSJCnnLD5LkiRJkiRJknLO4rMkSZIkSZIkKecsPkuSJEmSJEmScs7isyRJkiRJkiQp58654nMIYVcIIY7w2Dvd8UmSJEmSJEnSbFAw3QFMksPAXw1z/uhUByJJkiRJkiRJs9G5Wnx+K8aYmO4gJEmSJEmSJGm2OueW3ZAkSZIkSZIkTb9zdeZzYQjh94HlwDGgEXg1xtg7vWFJkiRJkiRJ0uxwrhafq4C/G3RuZwjhMzHGV6YjIEmSJEmSJEmaTc7FZTe+C6ykrwA9H/g14FvABcDzIYQV0xeaJEmSJEmSJM0O+YlEYrpjyKlEIvFKIpHYmUgkjiUSidOJRGJfIpH4X3V1dQuBq4GqRCLx1OB+IYQ76urqvnX99ddXA1x88cU0NzdTVlbGunXrqK6uZvXq1ZSWltLU1ERhYSF79uzh+PHjHDt2jPb2dubOncumTZsoLy9n7dq1A/o0NzczZ84c2tvbOXbsGCdOnKCtrY3CwkI2btxIZWXlgD5LlizhjTfeID8/n71793LkyBG6u7tpbW3lvPPOo7GxkaqqKtasWdPfZ/HixWzZsoUQAvv27aOrq4tTp06xe/duioqKaGxspLq6mvr6empqali9ejWLFi1i27ZtxBg5cOAAhw4doqenh127drFgwQLWr1/f//7pzysuLmb79u0kk0kOHjxIZ2cnyWSSHTt2sHDhwgF9KisrWbduHQsXLmTnzp309PRw6NAhDhw4AMC2bdtYvHgxDQ0N/X0qKipYv3498+fPZ9euXZw+fZrDhw+zb98+Qghs2bKFJUuWDOhTXl5OY2MjRUVFtLS0cPLkSbq6uujo6CA/P5/NmzdTWlrK2rVr+//by8rK2LhxI4WFhbS1tdHd3c3Ro0d58803KSgo6M9/uk99fT1lZWVs2rSpP5eD89/U1ERFRcWAXJaUlNDc3ExBQUF/Lgfnf3Au0/nPy8ujo6ODrq4uTp48SUtLS3/+ly5dOmL+9+/fz1tvvcXp06eH5H/16tX9z8XFxUPy39vby44dO1iwYAENDQ39f15VVVWsXbt21PwXFxf356W+vp6qqioaGhqYP3/+sPnfunXriPkvKipi9+7dnDp1akj+S0pKBtyX5eXlbNiwgXnz5tHS0kJ3dzdHjhxh7969A/Kf2Scz/+l7+ejRo7S3tzNnzhyam5uHvZcz8595L2fmf82aNf1/biUlJWzevHnEe3nDhg0D8l9fXz9s/tP3cjr/g3OZvpeBAfkf7l5O/91J5z+ZTNLZ2Tkg/4Pv5XT+M+/lzPxv376dRYsWDchlZWVlf/4z7+X9+/cPyP+6dev677GKiooR7+W8vLwh+U/3See/tbV1xHvZsdyx3LHcsdyx3LHcsdyx3LHcsdyx3LHcsdyxfDaO5UuXLq3LTeU1eyHGOF2fPaVCCP8O2AocjDGWjtTukUceiQB33nnnVIUmSZIkSZIkSZMlTNcHn4vLboxkX+p5/rRGIUmSJEmSJEmzwGwqPl+Vet4xrVFIkiRJkiRJ0ixwThWfQwi1IYSSYc6/Dfib1OEPpjYqSZIkSZIkSZp9CqY7gBz7OPDnIYSXgZ3AEeBXgOuBecA/AP/f9IUnSZIkSZIkSbPDuVZ8fhl4B/Bu+pbZmA+8Bfwz8HfA38XZssOiJEmSJEmSJE2jc6r4HGN8BXhluuOQJEmSJEmSpNnunFrzWZIkSZIkSZJ0drD4LEmSJEmSJEnKOYvPkiRJkiRJkqScs/gsSZIkSZIkSco5i8+SJEmSJEmSpJyz+CxJkiRJkiRJyjmLz5IkSZIkSZKknLP4LEmSJEmSJEnKOYvPkiRJkiRJkqScs/gsSZIkSZIkSco5i8+SJEmSJEmSpJyz+CxJkiRJkiRJyjmLz5IkSZIkSZKknLP4LEmSJEmSJEnKOYvPkiRJkiRJkqScs/gsSZIkSZIkSco5i8+SJEmSJEmSpJyz+CxJkiRJkiRJyjmLz5IkSZIkSZKknLP4LEmSJEmSJEnKOYvPkiRJkiRJkqScs/gsSZIkSZIkSco5i8+SJEmSJEmSpJyz+CxJkiRJkiRJyjmLz5IkSZIkSZKknLP4LEmSJEmSJEnKOYvPkiRJkiRJkqScs/gsSfr/27vzKDnKcvHj3yd72CEkbAECISQsen8ssgSBgDu73IDiEYEruAAiVz0iKhfwuKEgqCjqJYr3ooCgglzRoGAIiyhBlCUsCUk0QFbCkj2Zmff3R9WQpunJLKmZmun+fs6pU9O1vj3v01XVT7/1liRJkiRJUuFMPkuSJEmSJEmSCmfyWZIkSZIkSZJUOJPPkiRJkiRJkqTCmXyWJEmSJEmSJBXO5LMkSZIkSZIkqXAmnyVJkiRJkiRJhTP5LEmSJEmSJEkqnMlnSZIkSZIkSVLhTD5LkiRJkiRJkgpn8lmSJEmSJEmSVDiTz5IkSZIkSZKkwpl8liRJkiRJkiQVzuSzJEmSJEmSJKlwJp8lSZIkSZIkSYUz+SxJkiRJkiRJKpzJZ0mSJEmSJElS4Uw+S5IkSZIkSZIKZ/JZkiRJkiRJklQ4k8+SJEmSJEmSpMKZfJYkSZIkSZIkFc7ksyRJkiRJkiSpcCafJUmSJEmSJEmFM/ksSZIkSZIkSSqcyWdJkiRJkiRJUuFMPkuSJEmSJEmSCld3yeeIGBkRP46IFyJidUTMiYirImLLsssmSZIkSZIkSY1iQNkFKFJEjAYeAEYAtwFPAQcAnwTeHRGHpJReLLGIkiRJkiRJktQQ6q3l8/fJEs/npZROSCl9LqV0JHAlMBb4SqmlkyRJkiRJkqQGUTfJ54jYFXgnMAf4XtXsi4HlwKkRsXEPF02SJEmSJEmSGk7dJJ+BI/PxnSmllsoZKaWlwP3ARsBBPV0wSZIkSZIkSWo09dTn89h8/Ewb82eQtYzeHbirvY1dc801BRVLkiRJkiRJkkqTPv7xj0cZO66nls+b5+NX2pjfOn2LWjMj4iMRMa3wUkmSJEmSJElSA6qnls/tac3up1ozU0o/An7Uc8WRJEmSJEmSpPpVTy2fW1s2b97G/M2qlpMkSZIkSZIkdZN6Sj4/nY93b2P+mHzcVp/QkiRJkiRJkqSCREo1e6HocyJiNDATmAOMTim1VMzbFJhHlmwfnlJaXkohJUmSJEmSJKlB1E3L55TSs8CdwCjgnKrZlwIbA/9j4lmSJEmSJEmSul/dtHyG11o/PwCMAG4DngQOBI4g625jfErpxfJKKEmSJEmSJEmNoa6SzwARsSPwJeDdwDCy7jZuBS5NKS0ps2ySJEmSJEmS1CjqLvksSZIkSZIkSSpf3fT5LEmSJEmSJEnqPUw+S5IkSZIkSZIKN6DsAvRG/fv3Sy0tdkciSZIk1YOIoMjuBvsFFPl1oR/QUtzmelTR/9uul6MfKRX4X4yAMt5X9IMi30ejKaveJKkXGzhwEGvXrlmcUhpexv5NPtfQ0pJI08ouhSRJkqQixP6JqVOnFra9ww47jHR1YZsjzoX05onFbbAHxaO3cOzTHy67GNw+dhKn/HlNYdu74eBB7HfzjMK211EPnzQGPm/ytMu+Gpzyvulll0KSepUbbtoT4J9l7d9uNyRJkiRJkiRJhTP5LEmSJEmSJEkqnMlnSZIkSZIkSVLh7PNZktRwPngR3PVXWL4Kth0Gn/0QnHlC2aWSJJVp7ty5nHHGGRx++OFcdNFFG7StCVfBg3NgQN7UZ4ct4On/2vAydsWNL8/l0gXT+deaFWw7cAjX7bg/h25cyvOGXueOfX76utfNq5oZ9YE9eNNFB3fL/pbNm8O0b57H4scfpP/Awex4xInse/4V9BtQ3Ffilc/NZO61l7J81uMM3Gwrdjj1ArY88J2Fbb9HTbsaHr0OFj0Ge54Cx163bt7su2DyOfDqv2D7A7N5m+9cUkGlcj0z42fMnn0rL7/yDDvvdDQHHfhVAJqb1/DnBz/LkiWPs3zFCxx5xHVsM+KAkksrlcPksySp4Vx4Oky6CAYPgqfmwISPwj5jYb89yi6ZJKksV155JePGjStse1efDGeOL2xzXfKHpQu4YN5j3LTTgRyw0VbMa1pVboEqHPXIaa/93bRiLXce8nO2f/eobtvftG+ex5Ath/Pe2//FmmUv86fz3sOMX/2AsSefW8j2U3MTz37j4wx/xymMueg6lk7/K89e9lGG7ngbQ7bfpZB99KhNtodDvgizJkPTynXTVyyGX50IR10LY46Fey6CX78PTn+wvLJKJRo6dAR77flR5s2/n+bm1a+bt/XW+zJ291O574H/LKl0Uu9gtxuSpIaz1+gs8QwQ+fDsc2WWSJJUprvuuotNN92Ufffdt+yiFOriBU/wX9vswUEbD6NfBDsMHMoOA4eWXaw3mDd5DoO3GspW+2/bbftY/sJsdnrbRPoPHsLQYduy3UHv4pVZ0wvb/qrnZ7F2yUJGHHMG0b8/m73pYDYeuy8vTr21sH30qHEnwtgTYOiw109/+lew9V6wx0kwYAgcegks/AcsfqqUYkpl23HkOxg58u0MHrzF66b37z+IcWM/xPDh+xHRv6TSSb2DyWdJUkM6++uw0SEwbiJstzUcdUjZJZIklWH58uVMmjSJs88+u9DtXvgb2PoCOORbMOWZQjfdIc0pMW3lSyxqWs1uT/2OkU/+lnOff4SVLc09X5h2zP31DEaesBsR0W372P3kT/DPP/yCplUrWLHweeY9+Hu2O6jALjFSqjlt1b9mFLeP3mDREzDi39a9HrQxbDEaFj9RXpkkSb2ayWdJUkP6/udg6VS491o48Yh1LaElSY3l2muv5eijj2abbbYpbJuXnQCzLoHnvwwfGQ/H/hCeXVTY5jtkQdMq1qbELa88z72jJ/D3MW/nkZUv8+WFT/ZsQdqx4oVlvPjQfHY8YUy37mfEvofyyuzp3PL2Ydx2/C5sNW4/Rh5+fGHbH7LDrgzYfCsW/Oa/SU1refUf97Js+kO0rFnZ/sp9ydplMGTz108bsjmsWVpOeSRJvZ7JZ0lSw+rfH976/+C5hXDNLWWXRpLU02bMmMHDDz/MySefXOh2DxwFmw6BwQPhtIPgkF3hjh5uGDo0v837E8NGs93AoWw9YDCfGj6GO16d37MFacdzt85gq/22YaMdN+22faSWFqacfww7TjiBk+5+mRN/P481S1/i79+7sLB9xICBjP7sNbzy8BT+cdZ4Ftz+Y7Yc/x4GbtV9XYmUYuAmsPrV109b/SoM6r76kyT1bT5wUJLU8Jqa7fNZkhrRI488wvz58znppJMAWLlyJc3NzcyZM4dJkyYVtp8IqNEpQ7facsAgRg4c2q1dWRThudtmsttZ/9b+ghtgzatLWLFgLmMmnk3/QYPpP2gwux59Go/+6GL2Offrhe1no53HMfZLP3/t9VNfOJlhE95b2PZ7heF7wWM/Xfd6zXJ46dmsH2hJkmqw5bMkqaEsXAI3ToZlK6C5GSb/GW6YDEfuX3bJJEk97bjjjuOGG25g0qRJTJo0ieOOO46DDz6YK664osvbfHkFTJ4Oq9ZmP27+7CGYOhPetUeBBe+gM7YcxXcXz2Rh0ypealrDVYtmcMxmvacl7pK/LWDVghVs/+5R3bqfwVtszcbb78LMX/2QlqYm1ix9mdl3/C9b7PbmQvez4p9P0bJmNS2rVzL/N9ey9qVFDJtwYqH76DEtTdC0ClJzNjStyqbt/l5Y9Dg89cts2n1fghFvhq3HlV1iqRQtLU00N6+mJTWTUnP2d0sTAM3Na2huXp0vt5bm5tWkWv3DS3XOls+SpIYSAdf8Ej72NWhJsPO2cNWn4fgJZZdMktTThgwZwpAhQ157PXToUAYNGsQWW2zR5W2ubYYv/h88tQD694Nx28CtZ8HY4rqU7rCLttmDxc2r2f2pyQzp14+TN9+RL4woIQvehrm3zmDbd+zMgE26/8ELh37tJv521WeYfv3lRL/+bLPv4ez7ycsL3ceSe25l8d03k5qa2GSP/Rlz0XX0Gzi40H30mPu+DPdduu7149fDWy+Gwy6BE38Jk8+F33wQtj8QTrixtGJKZXti+g94/Invv/Z6zj9vZ++9zuZNe5/Lb+84iuUrXgBgyj1nAXDsMX9gk413KKWsUlnCX13eKCJSmlZ2KSRJkiQVIfaHqVOnFra9ww47jHR1YZsjzoX05onFbbAHxaO3cOzTHy67GNw+dhKn/HlNYdu74eBB7HfzjMK211EPnzQGPu939C77anDK+6aXXQpJ6lVuuGlPgIdTSqXc72u3G5KmamrMAAAQyUlEQVQkSZIkSZKkwpl8liRJkiRJkiQVzuSzJEmSJEmSJKlwJp8lSZIkSZIkSYUz+SxJkiRJkiRJKlyk5JN0q0XEUuDpssuhXmVrYHHZhVCvY1yoFuNC1YwJ1WJcqBbjQtWMCdViXKgW40Jt2RrYOKU0vIydDyhjp33A0yml/csuhHqPiJhmTKiacaFajAtVMyZUi3GhWowLVTMmVItxoVqMC7Ulj41RZe3fbjckSZIkSZIkSYUz+SxJkiRJkiRJKpzJ59p+VHYB1OsYE6rFuFAtxoWqGROqxbhQLcaFqhkTqsW4UC3GhdpSamz4wEFJkiRJkiRJUuFs+SxJkiRJkiRJKpzJZ0mSJEmSJElS4Uw+5yJiZET8OCJeiIjVETEnIq6KiC3LLps2TEQMi4gzI+LXETEzIlZGxCsRcV9EfDgian4OImJ8RNwREUsiYkVEPBoR50dE//Xs65iImJJvf1lE/CUiTuu+d6ciRcSpEZHy4cw2lul0HUfEaRHx13z5V/L1j+med6EiRMShEfHLiJiXnxPmRcSdEXFUjWU9VjSAiDg6j4Hn8vPIrIi4OSIObmN546IORMTEiPhuRNwbEa/m54fr21mnR+rec0s5OhMTETEmIi6IiLsjYm5ErImIBRFxW0Qc0c5+OlW/EdE/j7NH82PUkjwOx2/oe1b7unKsqFp/UsU16G5tLNPpOo6IoRFxaUQ8HRGrImJhRPwiIvboyvtU53TxHBL5539KXscrI2J2Xm+7t7GOx4s+orMxERGDI+KcvH4X53X8ZER8JyJ2Xs96xkQfE700bxURh0TENyLioYhYFNl349kRcW1b56uaUkoNPwCjgQVAAm4Fvg7cnb9+ChhWdhkdNqh+P5bX5QvAz4CvAT8GXs6n30Le/3nFOscDTcAyYBLwzTwWEnBzG/s5N5+/GPgecCUwN592edn/B4d242THPCaW5nV2ZhF1DFyez5+bL/894MV82rllv2+HmnX2xbx+FgE/Ab5K9oCGh4BvVC3rsaIBBuCyijq7Nr9OuAVYA7QAHzQu6nMA/p7XwVLgyfzv69ezfI/UveeWvhETwI35/CeAH5Jdg/4qj5EEnFdE/QIB3My67y7fzONvWb6v48v+v9X70NljRdW6x1asm4DdiqhjYDBwX77OQ/m57OfAWmA5cGDZ/7d6H7pwDhkC3F5Rz1fnx42fArOAY2qs4/GiDw2dPIcMqPgMPwl8N6/ve/JpLwN7GhP1MdBL81bAfKAZuBe4Ko+v+/PllwEHd+j9lf0P7g0DMDn/x32iavq38uk/KLuMDhtUv0eSXdT1q5q+LfCvvI7/vWL6ZsBCYDWwf8X0IcAD+fLvr9rWKGBVflAfVTF9S2Bmvk6HPpQOpcRIAH8Ens0P2G9IPneljoHx+fSZwJZV23ox396o7npfDl2KhZPyOvsDsGmN+QMr/vZY0QBDfq5ozi+8RlTNOyKvs1nGRX0OeR2Pyc8TE1j/l8QeqXvPLX0qJk4H9qkx/XCyH69WA9ttaP0Cp+Tr3A8MqZj+lnwfC6lxTnMoJy6q1huen19uBKbQdvK503UMXJivczMV34PIkhWtP4r068r7deieuCBLBCWyhg9vqBsqrkPz1x4v+tjQyXNI6/eSP1bHA3BpPu/HxkR9DPTSvBVwAbB9jfJ+Pl/+sY68v4bvdiMidgXeCcwhO9hXupjsV+FTI2LjHi6aCpJSujuldHtKqaVq+nzgB/nLCRWzJpJdCN6YUppWsfwqshaRAB+v2s1/kLUuuDqlNKdinZfILh4g+yVLvdN5ZAf7M8g+87V0pY5bX38lX651nTlkx5vB+T7VC+S3Ml0GrAA+kFJaWr1MSmltxUuPFY1hZ7Juyv6SUlpYOSOl9CeylivDKyYbF3UkpfSnlNKMlF9lt6On6t5zS4k6ExMppetSSo/UmH4PWaJxEFmioFJX6rc1rr6Yx1vrOg8BN5HF5cT2yquu6+SxotKP8vE57SzXqTqOiGBdLH228ntQSuk2shZse5L9EKJu0pm4iIjRZHX2EPCF6u+u+fbWVk3yeNHHdPJYsWs+/m2NeLgtHw+vmm5M9FG9NW+VUrospfRCjSJfBqwE9o6IYe29v4ZPPpMlnADurFHJS8l+/dkIOKinC6Ye0XoCb6qY1hoTv6+x/FSyxNT4iBjcwXV+V7WMepG8z7uvA99OKU1dz6JdqWPjom8ZD+wC3AG8FFkfvxdExCejdr++HisawwyyFooHRMTWlTMi4jBgU7IWKa2Mi8bVU3VvvNSHWteg0Mn6zeNpPFl83duRddQ7RMTpwAnAx1JKL65nua7U8WhgJ+CZlNLsDq6jcp1Clp/5KbBZRHwwIi6MiI+sp19Vjxf17Yl8/J4a/f229t/8x6rpxkR96o15q1RRnub2Fh7QwY3Ws7H5+Jk25s8gaxm9O3BXj5RIPSIiBgAfyl9WfvjajImUUlNEzAb2Ivsl8skOrDMvIpYDIyNio5TSiiLKrw2Xx8D/kt3G8vl2Fu9UHed3S+wALEspzauxvRn5uOaDQ1SKt+TjBcDfgDdVzoyIqcDElNKifJLHigaQUloSEReQdcU1PSJuJbtVbTRwHFkXLR+tWMW4aFzdXveeW+pD/pCot5F9MZxaMb0r9bsb0J+s+5/qRHZb66hkeQx8m+x2+1vbWbwrddyR77jV66hcrdehm5N1BVjZkjBFxDVk/cQ3g8eLBvFbsucEnAg8FhF/JGsQsR/wVrI+oK9uXdiYqE+9OG91ElkjnAdTSi+39z5s+Zwd3AFeaWN+6/QteqAs6llfB/YG7kgpTa6Y3pWY6Og6m7cxX+X4L2Af4PSU0sp2lu1sHXts6XtG5OOPAUOBt5OdUPcmezbAYWT9JrbyWNEgUkpXkV34DwDOAj5HdsE1F7iuqjsO46Jx9UTde27p4/IWSD8ju+31ksrbouneGDImeom8BeNPyR7UdF4HVjEuGkPrdeiXgGlkjSA2Jfuh6lngbOCiiuWNizqXd80xEbiELGl4HvAZsn6jpwI/b/0xImdM1Kdel7eKiF3IfvxoAj69vmVbmXxuX+TjzvbfpV4sIs4j+5A8BZza2dXzcWdiwjjqZSLiALLWzleklP5cxCbzcWfr2JjoPfrn4yBr4XxXSmlZSukJ4L3Ac8DhbXTBUYvHijoREZ8le8L0dWQtnjcma3UyC/hZRHyjM5vLx8ZF4+nJujdWeqGI6E92x9UhZP1oXt7FTXn86Nv+k6yv5bOqfnzoKs8r9aH1OnQe8N6U0uP5dejdZAnIFuBTETGok9s1LvqoiBhCdq74DFm/8NuRJQWPInsmydSIOL4LmzYm+ojemLeKiBFkXXQMBz6ZUnqgIxs2+dx+Zn+zquXUx0XEOWS3uU0HjkgpLalapCsx0dF1Xu1EUdVNKrrbeIbXtyBYn87WcXvLt/ero3pe6xfAWSmlf1TOyFvGt/7SfEA+9ljRACJiAtkDNX6TUvpUSmlWSmlFSulvZD9KPA98On+AMRgXjawn6t5zSx+VJ56vJ7tr4hfAB2s8cKor9et3mT4kIsYAXwF+klK6o4OrdeexxbjoPVqvQ39ffUdmfl06m6wl9B75ZI8X9a/1TrsvpJR+mFKan1J6NaX0O7IfJAaS5TVaGRN1pDfmrfLE891kLfE/mVL6fhvbeQOTz/B0Pm6rD5sx+bit/rLUh0TE+WT9Ij1O9gGeX2OxNmMiT1ruQnZ7wawOrrMdWSu55+yrs9fYhKyu9gBWRURqHYCL82X+O592Vf66U3WcUlpOlpTaJJ9fzWNL79Nax231WdX6pWBo1fIeK+pb6wNd/lQ9I6+nv5JdT+2TTzYuGle3173nlr4pr/8bgPcDPwc+UKtvzS7W70yyB/3smu+nI+uoPHuRdblyRuX1Z34Neni+zIx82gn5667Usd9x+55OXYd6vGgI67sG/QewBNg5Iobl04yJOtEb81b5/CnAnsA5KaXvdPT9gMlnWPdBfmf1E0QjYlOy2+JWAg/2dMFUrPyBUVcCfyf7AC9sY9G78/G7a8w7DNgIeCCltLqD67ynahmVbzUwqY3hkXyZ+/LXrV1ydKWOjYu+ZSrZCXpMG7c07p2P5+RjjxWNofUJ0cPbmN86fU0+Ni4aV0/VvfHSh+Tnk1vIWq/9D3BqVR+d1TpVv3k8PUAWX4d2ZB2Vag5tX4O2Jhduzl/PgS7X8bNkD9TePe+bsyPrqFx35eO9q2fkfcW3JgHnVMzyeFHf2rwGzWOitZXqmopZxkQf1xvzVhExErgHGAd8rDMtnl+TUmr4gexW6gR8omr6t/LpPyi7jA4bXMcX5XU5DdiqnWU3AxaRJSj3r5g+hOzAnID3V62zC7AKeBEYVTF9S7JfExNwcNn/B4cOxcoleX2duaF1DIzPp88EtqyYPirfzqrKbTmUP5DdEp2AL1dNfwdZX3svA1vk0zxWNMAAnJzXy3xgh6p578njYiUwzLio7wGYkNfF9W3M75G699zSe4YOxMRg4Lf5MtcC/TqwzU7XL3BKvs79wJCK6W/J43EhsFnZ/69GGdqLi/WsNyVfb7ca8zpdx8CF+To3V8YecHw+/YmOxKRDz8QFMIjsR4MW4B1V876crzularrHiz48dCAmvp/P/yMwuGre1/J5fzUm6megF+atgJ3yY1MzcEZX31vkG2toETGarHJGALcBTwIHkj1F9BlgfErpxfJKqA0REaeRPSSqmeyJnLX6K5qTUrquYp0TyFqprAJuJLul5Tiyvm1uAU5OVR+eiPgE8B2yD/JNZL9ATgRGkj3U7jNFvi91j4i4hKzrjbNSStdWzet0HUfEFcCnyB5WdwvZheX7gGFkP3hd3W1vRp2W92N1P7AbcC9Zlwo7k/Xtm8hulb65YnmPFXUuvytqMvB2YCnwa7JE9B5kt0MGcH5K6dsV6xgXdSKvy9Zb37cF3kV2++K9+bTFlXXTU3XvuaU8nYmJiPgJcDqwmHVJhGpTUkpTqvbRqfqNiCDrR3oi2UOJbs+XfR/Zl9B/Tynd1tX3rPZ19ljRxjamkHW9MSalNLNqXqfrOG8ZeTdZMmoaWcvancha4a8Bjkwp/aULb1cd1IVzyFuBO8k+878G/kmWADyMLMn01pTS67o/8HjRt3TyHLID2R34I8lavP+erMHDIWTPoFkJvC2l1Hqnbus+jIk+qLfmrSJiNtmPFw8D/9dG8a9LKc1Z7xssO7PfWwZgR+AnZE+XXUN2oP827fza4ND7B9a1ZF3fMKXGeocAd5D1r7USeIzs6dT917OvY8luR1gKLAceAk4r+3/g0KV4ObON+Z2uY+C0fLnl+Xr3AMeU/V4d2qyvrcjufJmdnw9eJPth8qA2lvdYUecD2QNdzif7AvAqWfcsC8kuwN5pXNTv0IFriDll1b3nlt4fE6xrybq+4ZIi6hcYkMfZY3ncvZTH4fiy/2eNMHTlWFFjG63x8oaWz12tY7L+gS8FZpC1jltE1hJ6z7L/Z40wdPEcsidZQmgh2XXoXOCHwMj17MfjRR8ZOhsTZF1uXE7WQHIV63JVPwHGGRP1M3QgNkrJW3WgTAmY0N77s+WzJEmSJEmSJKlwPnBQkiRJkiRJklQ4k8+SJEmSJEmSpMKZfJYkSZIkSZIkFc7ksyRJkiRJkiSpcCafJUmSJEmSJEmFM/ksSZIkSZIkSSqcyWdJkiRJkiRJUuFMPkuSJEmSJEmSCmfyWZIkSZIkSZJUOJPPkiRJkiRJkqTC/X/ZmxXeTDLuKwAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 1728x576 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Plotting the predicted fitness effect for the variants seen more than once\n",
    "plt.close('all')\n",
    "f, (ax1, ax2) = plt.subplots(2, 1, sharey=True, sharex = True, figsize=(24, 8))\n",
    "gs = matplotlib.gridspec.GridSpec(2, 1, width_ratios=[1], height_ratios=[10,0.5])\n",
    "ax1 = plt.subplot(gs[0])\n",
    "ax2 = plt.subplot(gs[1])\n",
    "gs.update(hspace=0.05)\n",
    "\n",
    "nonsense_list = []\n",
    "y_list = []\n",
    "x_list=[]\n",
    "y_list_top20 = []\n",
    "x_list_top20=[]\n",
    "y_list_nonsense = []\n",
    "x_list_nonsense = []\n",
    "x=[]\n",
    "y=[]\n",
    "OBS_EXP_DICT={}\n",
    "top20_lower_err = []\n",
    "top20_upper_err = []\n",
    "missense_lower_err = []\n",
    "missense_upper_err = []\n",
    "nonsense_lower_err = []\n",
    "nonsense_upper_err = []\n",
    "\n",
    "top20 = c1\n",
    "missense = c6\n",
    "nonsense = c10\n",
    "synonymous = c3\n",
    "\n",
    "axisfont=20\n",
    "domainfont = 12\n",
    "titlefont=12\n",
    "axislabelfont=12\n",
    "\n",
    "top_20_TET2 = []\n",
    "\n",
    "#ERROR NEEDS TO BE IN THE FORMAT: [tuple(lower_err),tuple(upper_err)]\n",
    "\n",
    "for k, v in TET2_missense_s.items():\n",
    "    if k in top_20_TET2:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_top20.append(codon_number)\n",
    "        y_list_top20.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        top20_lower_err.append(error_lower)\n",
    "        top20_upper_err.append(error_upper)\n",
    "\n",
    "    if k not in top_20_TET2:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list.append(codon_number)\n",
    "        y_list.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        missense_lower_err.append(error_lower)\n",
    "        missense_upper_err.append(error_upper)\n",
    "\n",
    "    missense_err = [tuple(missense_lower_err), tuple(missense_upper_err)]   \n",
    "    \n",
    "for k, v in TET2_nonsense_s.items():\n",
    "    if k in top_20_TET2:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_top20.append(codon_number)\n",
    "        y_list_top20.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        top20_lower_err.append(error_lower)\n",
    "        top20_upper_err.append(error_upper)\n",
    "    \n",
    "    top20_err = [tuple(top20_lower_err), tuple(top20_upper_err)]\n",
    "        \n",
    "    if k not in top_20_TET2:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_nonsense.append(codon_number)\n",
    "        y_list_nonsense.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        nonsense_lower_err.append(error_lower)\n",
    "        nonsense_upper_err.append(error_upper)\n",
    "\n",
    "    nonsense_err = [tuple(nonsense_lower_err), tuple(nonsense_upper_err)]        \n",
    "    \n",
    "msize = 12\n",
    "medgewidth = 2\n",
    "linew = 3\n",
    "csize = 4\n",
    "cthick = 3\n",
    "\n",
    "markers1, caps1, bars1 = ax1.errorbar(x_list, y_list, yerr = missense_err, fmt = 'o', ecolor = missense, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = missense, markerfacecolor = missense, zorder = 1)    \n",
    "\n",
    "markers2, caps2, bars2 = ax1.errorbar(x_list_nonsense, y_list_nonsense, yerr = nonsense_err, fmt = 'o', ecolor = nonsense, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = nonsense, markerfacecolor = nonsense, zorder = 2)    \n",
    "\n",
    "markers3, caps3, bars3 = ax1.errorbar(x_list_top20, y_list_top20, yerr = top20_err, fmt = 'o', ecolor = top20, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = top20, markerfacecolor = top20, zorder = 3)    \n",
    "\n",
    "# loop through bars and caps and set the alpha value\n",
    "a = 0.7\n",
    "[bar.set_alpha(a) for bar in bars1]\n",
    "[cap.set_alpha(a) for cap in caps1]\n",
    "[bar.set_alpha(a) for bar in bars2]\n",
    "[cap.set_alpha(a) for cap in caps2]\n",
    "[bar.set_alpha(a) for bar in bars3]\n",
    "[cap.set_alpha(a) for cap in caps3]\n",
    "\n",
    "#Only show the required axis lines\n",
    "ax1.spines['top'].set_visible(False)\n",
    "ax1.spines['right'].set_visible(False)\n",
    "\n",
    "for axis in ['bottom','left']:\n",
    "    ax1.spines[axis].set_linewidth(2)\n",
    "\n",
    "for axis in ['bottom','left']:\n",
    "    ax1.spines[axis].set_color(grey3)\n",
    "    \n",
    "ax1.set_ylabel('fitness effect (s)', fontsize = axisfont, labelpad = 10)\n",
    "ax1.set_title('Fitness effects across TET2', fontsize = 18)\n",
    "\n",
    "# ax1.set_yscale('log')\n",
    "\n",
    "# STACKED BAR CHART\n",
    "# define the exons + create data\n",
    "exons = [{'3':(1137-1)+1, '4': (1167-1138)+1, '5': (1198-1168)+1, '6':(1268-1199)+1, '7':(1318-1269)+1, '8': (1348-1319)+1, '9': (1394-1349)+1,'10': (1513-1395)+1, '11': (2002-1514)+1}]\n",
    "exon_orders = [['3', '4', '5', '6', '7', '8', '9', '10', '11']]\n",
    "colors2 = [c4, c0, c3, c1, c5, c6,c7, c8, c9]\n",
    "names = exons[0].keys()\n",
    "values = np.array([[data[name] for name in order] for data,order in zip(exons, exon_orders)])\n",
    "lefts = np.insert(np.cumsum(values, axis=1),0,0, axis=1)[:, :-1]\n",
    "orders = np.array(exon_orders)\n",
    "\n",
    "#plot the bar chart\n",
    "for name, color in zip(names, colors2):\n",
    "    idx = np.where(orders == name)\n",
    "    value = values[idx]\n",
    "    left = lefts[idx]\n",
    "    ax2.bar(x=left, height=-10, width=value, bottom=-5, color=color, orientation = 'horizontal', label = name, edgecolor='k', linewidth=1, tick_label = name)\n",
    "    ax2.text(left + (value/2.), (-0.26), name, ha='center', va='center', fontsize = 12)\n",
    "\n",
    "# CONFIGURING THE GRAPH APPEARANCE\n",
    "#Set the x and y axis limits\n",
    "ax1.set_xlim(0, 2004)\n",
    "ax1.set_ylim(-0.01, 30)\n",
    "ax2.set_xlim(0, 2004)\n",
    "ax2.set_ylim(-0.5, 0)\n",
    "\n",
    "#x-axis ticks\n",
    "x1_major_ticks = []\n",
    "x1_major_tick_labels = []\n",
    "x2_major_ticks = [0, 200, 400, 600, 800, 1000, 1200, 1400, 1600, 1800, 2002]\n",
    "x2_major_tick_labels = [\"0\", \"200\", \"400\", \"600\", \"800\", \"1000\", \"1200\", \"1400\", \"1600\", \"1800\", \"2002\"]\n",
    "ax1.set_xticks(x1_major_ticks)\n",
    "ax1.set_xticklabels(x1_major_tick_labels, fontsize = axisfont)\n",
    "ax2.set_xticks(x2_major_ticks)\n",
    "ax2.set_xticklabels(x2_major_tick_labels, fontsize = axisfont)\n",
    "\n",
    "#y-axis ticks\n",
    "y1_major_ticks = [0, 5, 10, 15, 20, 25, 30]\n",
    "y1_major_tick_labels = [\"0\", \"5\", \"10\", \"15\", \"20\", \"25\", \"30\"]\n",
    "ax1.set_yticks(y1_major_ticks)\n",
    "ax1.set_yticklabels(y1_major_tick_labels, fontsize = axisfont)\n",
    "\n",
    "#set panel legends\n",
    "legend_marker_size = 2\n",
    "legend_elements = [Line2D([0], [0], marker = '_', color = missense, alpha = 1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='missense'),\n",
    "                   Line2D([0], [0], marker = '_', color=nonsense, alpha=1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='nonsense'),\n",
    "                  Line2D([0], [0], marker = '_', color=top20, alpha=1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='top 20 observed variant')]\n",
    "    \n",
    "ax1.legend(ncol=1, handles=legend_elements, loc='upper left', frameon=0, fontsize = 20)\n",
    "\n",
    "ax1.grid(which='major', linestyle=':', linewidth='0.7', color=grey3, zorder = 0)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ASXL1 variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {
    "code_folding": [
     0,
     86,
     97,
     109
    ]
   },
   "outputs": [],
   "source": [
    "def expected_number_variants_missense_ASXL1(data, params):\n",
    "    s_value = params[0]\n",
    "    \n",
    "    a = 0\n",
    "    expected_number_by_variant_missense = {}\n",
    "    \n",
    "    Jaiswal_fmin = 0.056\n",
    "    Jaiswal_mean = 58\n",
    "    Jaiswal_sigma = 11.5\n",
    "    Jaiswal_total = 17182\n",
    "    \n",
    "    Genovese_fmin = 0.1194\n",
    "    Genovese_mean = 55.1\n",
    "    Genovese_total = 12380\n",
    "    Genovese_sigma = 11.95\n",
    "    \n",
    "    McKerrel_fmin = 0.008\n",
    "    McKerrel_mean = 51.88\n",
    "    McKerrel_total = 4219\n",
    "    McKerrel_sigma = 18.36\n",
    "    \n",
    "    Zink_fmin = 0.2363\n",
    "    Zink_mean = 60\n",
    "    Zink_total = 11262\n",
    "    Zink_sigma = 1.18907720e+01\n",
    "    \n",
    "    Coombs_fmin = 0.019\n",
    "    Coombs_mean = 59.82\n",
    "    Coombs_total = 5649\n",
    "    Coombs_sigma = 15.14\n",
    "    \n",
    "    Young_fmin = 0.0009\n",
    "    Young_mean = 63\n",
    "    Young_total = 89\n",
    "    Young_sigma = 6.8\n",
    "    \n",
    "    Desai_fmin = 0.0169\n",
    "    Desai_mean = 66.1\n",
    "    Desai_total = 424\n",
    "    Desai_sigma = 6 #estimated\n",
    "    \n",
    "    Acuna_fmin = 0.0015\n",
    "    Acuna_mean = 4.5\n",
    "    Acuna_total = 2014\n",
    "    \n",
    "    N = 9.40166610e+04\n",
    "\n",
    "    for i in data:\n",
    "        variant_mutation_rate = mutation_rates_variants_ASXL1(i)\n",
    "\n",
    "        if i in Genovese_missense_ASXL1:\n",
    "            Genovese_number = prevalence_averaged_over_ages(Genovese_fmin, s_value, N, variant_mutation_rate, Genovese_mean, Genovese_sigma, Genovese_total)\n",
    "        else:\n",
    "            Genovese_number = 0 \n",
    "\n",
    "        if i in Zink_missense_ASXL1:\n",
    "            Zink_number = prevalence_averaged_over_ages(Zink_fmin, s_value, N, variant_mutation_rate, Zink_mean, Zink_sigma, Zink_total)\n",
    "        else:\n",
    "            Zink_number = 0\n",
    "\n",
    "        if i in Coombs_missense_ASXL1:\n",
    "            Coombs_number = prevalence_averaged_over_ages(Coombs_fmin, s_value, N, variant_mutation_rate, Coombs_mean, Coombs_sigma, Coombs_total)\n",
    "        else:\n",
    "            Coombs_number = 0\n",
    "\n",
    "        if i in Young_missense_ASXL1:\n",
    "            Young_number = prevalence_averaged_over_ages(Young_fmin, s_value, N, variant_mutation_rate, Young_mean, Young_sigma, Young_total)\n",
    "        else:\n",
    "            Young_number = 0\n",
    "\n",
    "        if i in Desai_missense_ASXL1:\n",
    "            Desai_number = prevalence_averaged_over_ages(Desai_fmin, s_value, N, variant_mutation_rate, Desai_mean, Desai_sigma, Desai_total)\n",
    "        else:\n",
    "            Desai_number = 0\n",
    "\n",
    "        if i in Acuna_missense_ASXL1:\n",
    "            Acuna_number = prevalence_averaged_over_uniform_ages(Acuna_fmin, s_value, N, variant_mutation_rate, Acuna_mean, Acuna_total)\n",
    "        else:\n",
    "            Acuna_number = 0\n",
    "\n",
    "        expected_total = Genovese_number + Zink_number + Coombs_number + Young_number + Desai_number + Acuna_number\n",
    "\n",
    "        expected_number_by_variant_missense[i] = expected_total\n",
    "    \n",
    "    return expected_number_by_variant_missense\n",
    "\n",
    "def prob_observed_given_expected_missense_ASXL1(params, data): #data = list of variant (e.g. p.R882H) (or all_missense_variants_ASXL1)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_ASXL1_counts[variant])\n",
    "        expected_number = float(expected_number_variants_missense_ASXL1(data, params)[variant])\n",
    "        square_distance = ((observed_number - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance\n",
    "\n",
    "def prob_observed_given_expected_missense_ASXL1_upper_error(params, data): #data = list of variant (e.g. p.R882H) (or all_missense_variants_ASXL1)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_ASXL1_counts[variant])\n",
    "        observed_number_upper_error = observed_number+np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_missense_ASXL1(data, params)[variant])\n",
    "        square_distance = ((observed_number_upper_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   \n",
    "\n",
    "def prob_observed_given_expected_missense_ASXL1_lower_error(params, data): #data = list of variant (e.g. p.R882H) (or all_missense_variants_ASXL1)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_ASXL1_counts[variant])\n",
    "        observed_number_lower_error = observed_number-np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_missense_ASXL1(data, params)[variant])\n",
    "        square_distance = ((observed_number_lower_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {
    "code_folding": [
     0,
     91,
     102,
     114
    ]
   },
   "outputs": [],
   "source": [
    "def expected_number_variants_nonsense_ASXL1(data, params):\n",
    "    s_value = params[0]\n",
    "    \n",
    "    a = 0\n",
    "    expected_number_by_variant_nonsense = {}\n",
    "    \n",
    "    Jaiswal_fmin = 0.056\n",
    "    Jaiswal_mean = 58\n",
    "    Jaiswal_sigma = 11.5\n",
    "    Jaiswal_total = 17182\n",
    "    \n",
    "    Genovese_fmin = 0.1194\n",
    "    Genovese_mean = 55.1\n",
    "    Genovese_total = 12380\n",
    "    Genovese_sigma = 11.95\n",
    "    \n",
    "    McKerrel_fmin = 0.008\n",
    "    McKerrel_mean = 51.88\n",
    "    McKerrel_total = 4219\n",
    "    McKerrel_sigma = 18.36\n",
    "    \n",
    "    Zink_fmin = 0.2363\n",
    "    Zink_mean = 60\n",
    "    Zink_total = 11262\n",
    "    Zink_sigma = 1.18907720e+01\n",
    "    \n",
    "    Coombs_fmin = 0.019\n",
    "    Coombs_mean = 59.82\n",
    "    Coombs_total = 5649\n",
    "    Coombs_sigma = 15.14\n",
    "    \n",
    "    Young_fmin = 0.0009\n",
    "    Young_mean = 63\n",
    "    Young_total = 89\n",
    "    Young_sigma = 6.8\n",
    "    \n",
    "    Desai_fmin = 0.0169\n",
    "    Desai_mean = 66.1\n",
    "    Desai_total = 424\n",
    "    Desai_sigma = 6 #estimated\n",
    "    \n",
    "    Acuna_fmin = 0.0015\n",
    "    Acuna_mean = 4.5\n",
    "    Acuna_total = 2014\n",
    "    \n",
    "    N = 9.40166610e+04\n",
    "\n",
    "    for i in data:\n",
    "        variant_mutation_rate = mutation_rates_variants_ASXL1(i)\n",
    "\n",
    "        if i in Jaiswal_nonsense_ASXL1:\n",
    "            Jaiswal_number = prevalence_averaged_over_ages(Jaiswal_fmin, s_value, N, variant_mutation_rate, Jaiswal_mean, Jaiswal_sigma, Jaiswal_total)\n",
    "        else:\n",
    "            Jaiswal_number = 0\n",
    "\n",
    "        if i in Genovese_nonsense_ASXL1:\n",
    "            Genovese_number = prevalence_averaged_over_ages(Genovese_fmin, s_value, N, variant_mutation_rate, Genovese_mean, Genovese_sigma, Genovese_total)\n",
    "        else:\n",
    "            Genovese_number = 0 \n",
    "\n",
    "        if i in Zink_nonsense_ASXL1:\n",
    "            Zink_number = prevalence_averaged_over_ages(Zink_fmin, s_value, N, variant_mutation_rate, Zink_mean, Zink_sigma, Zink_total)\n",
    "        else:\n",
    "            Zink_number = 0\n",
    "\n",
    "        if i in Coombs_nonsense_ASXL1:\n",
    "            Coombs_number = prevalence_averaged_over_ages(Coombs_fmin, s_value, N, variant_mutation_rate, Coombs_mean, Coombs_sigma, Coombs_total)\n",
    "        else:\n",
    "            Coombs_number = 0\n",
    "\n",
    "        if i in Young_nonsense_ASXL1:\n",
    "            Young_number = prevalence_averaged_over_ages(Young_fmin, s_value, N, variant_mutation_rate, Young_mean, Young_sigma, Young_total)\n",
    "        else:\n",
    "            Young_number = 0\n",
    "\n",
    "        if i in Desai_nonsense_ASXL1:\n",
    "            Desai_number = prevalence_averaged_over_ages(Desai_fmin, s_value, N, variant_mutation_rate, Desai_mean, Desai_sigma, Desai_total)\n",
    "        else:\n",
    "            Desai_number = 0\n",
    "\n",
    "        if i in Acuna_nonsense_ASXL1:\n",
    "            Acuna_number = prevalence_averaged_over_uniform_ages(Acuna_fmin, s_value, N, variant_mutation_rate, Acuna_mean, Acuna_total)\n",
    "        else:\n",
    "            Acuna_number = 0\n",
    "\n",
    "        expected_total = Jaiswal_number + Genovese_number + Zink_number + Coombs_number + Young_number + Desai_number + Acuna_number\n",
    "\n",
    "        expected_number_by_variant_nonsense[i] = expected_total\n",
    "    \n",
    "    return expected_number_by_variant_nonsense\n",
    "\n",
    "def prob_observed_given_expected_nonsense_ASXL1(params, data): #data = list of variant (e.g. p.R882H) (or all_nonsense_variants_ASXL1)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_ASXL1_counts[variant])\n",
    "        expected_number = float(expected_number_variants_nonsense_ASXL1(data, params)[variant])\n",
    "        square_distance = ((observed_number - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   \n",
    "\n",
    "def prob_observed_given_expected_nonsense_ASXL1_upper_error(params, data): #data = list of variant (e.g. p.R882H) (or all_nonsense_variants_ASXL1)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_ASXL1_counts[variant])\n",
    "        observed_number_upper_error = observed_number+np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_nonsense_ASXL1(data, params)[variant])\n",
    "        square_distance = ((observed_number_upper_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   \n",
    "\n",
    "def prob_observed_given_expected_nonsense_ASXL1_lower_error(params, data): #data = list of variant (e.g. p.R882H) (or all_nonsense_variants_ASXL1)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_ASXL1_counts[variant])\n",
    "        observed_number_lower_error = observed_number-np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_nonsense_ASXL1(data, params)[variant])\n",
    "        square_distance = ((observed_number_lower_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#CALCULATING S FOR MISSENSE VARIANTS\n",
    "ASXL1_missense_s = {}\n",
    "a = 0\n",
    "for var in all_missense_variants_ASXL1:\n",
    "    initial_guess=[0.05]\n",
    "    if var in all_studies_ASXL1_counts.keys():\n",
    "        observed_number = float(all_studies_ASXL1_counts[var])\n",
    "#         print(observed_number)\n",
    "        if observed_number >1:\n",
    "            variant_mutation_rate = mutation_rates_variants_ASXL1(var)\n",
    "            \n",
    "            MLE = scipy.optimize.minimize(prob_observed_given_expected_missense_ASXL1, initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value = MLE.x[0]\n",
    "\n",
    "            upper_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_missense_ASXL1_upper_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_upper = upper_error_MLE.x[0]\n",
    "\n",
    "            lower_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_missense_ASXL1_lower_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_lower = lower_error_MLE.x[0]\n",
    "\n",
    "#             print('a = ', a)\n",
    "            ASXL1_missense_s[var]=(s_value, observed_number, variant_mutation_rate, s_value_upper, s_value_lower) \n",
    "    a = a+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#CALCULATING S FOR NONSENSE VARIANTS\n",
    "ASXL1_nonsense_s = {}\n",
    "a = 0\n",
    "for var in all_nonsense_variants_ASXL1:\n",
    "    initial_guess=[0.05]\n",
    "    if var in all_studies_ASXL1_counts.keys():\n",
    "        observed_number = float(all_studies_ASXL1_counts[var])\n",
    "#         print(observed_number)\n",
    "        if observed_number >1:\n",
    "            variant_mutation_rate = mutation_rates_variants_ASXL1(var)\n",
    "            \n",
    "            MLE = scipy.optimize.minimize(prob_observed_given_expected_nonsense_ASXL1, initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value = MLE.x[0]\n",
    "\n",
    "            upper_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_nonsense_ASXL1_upper_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_upper = upper_error_MLE.x[0]\n",
    "\n",
    "            lower_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_nonsense_ASXL1_lower_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_lower = lower_error_MLE.x[0]\n",
    "\n",
    "#             print('a = ', a)\n",
    "            ASXL1_nonsense_s[var]=(s_value, observed_number, variant_mutation_rate, s_value_upper, s_value_lower) \n",
    "    a = a+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#Exac - retrieve the frequencies in Exac for each of the variants\n",
    "import csv\n",
    "filename=\"Data files/ASXL1_Exac.csv\"\n",
    "\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    Exac_ASXL1_results={} \n",
    "    \n",
    "    for row in read_reader:\n",
    "#         print(row)\n",
    "        if row_count>1:\n",
    "            variant = row[6]\n",
    "            allele_freq = float(row[14])\n",
    "            african_freq = float(row[15])/float(row[16])\n",
    "            east_asian_freq = float(row[18])/float(row[19])\n",
    "            european_freq = float(row[21])/float(row[22])\n",
    "            finnish_freq = float(row[24])/float(row[25])\n",
    "            latino_freq = float(row[27])/float(row[28])\n",
    "            south_asian_freq = float(row[33])/float(row[34])\n",
    "            Exac_ASXL1_results[variant]=(allele_freq, african_freq, east_asian_freq, european_freq, finnish_freq, latino_freq, south_asian_freq)\n",
    "\n",
    "        row_count=row_count+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#COSMIC - retrieve the frequencies in COSMIC for each of the variants\n",
    "import csv\n",
    "filename=\"Data files/ASXL1_COSMIC_variant_counts.csv\"\n",
    "\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    ASXL1_COSMIC_results={} \n",
    "    \n",
    "    for row in read_reader:\n",
    "#         print(row)\n",
    "        if row_count>1:\n",
    "            variant = row[1]\n",
    "            counts = int(row[2])\n",
    "            ASXL1_COSMIC_results[variant]=counts\n",
    "\n",
    "        row_count=row_count+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Create dictionary of results and save to .csv file\n",
    "ASXL1_s_dictionary={}\n",
    "for k, v in ASXL1_missense_s.items():\n",
    "#     print(k)\n",
    "    if k in Exac_ASXL1_results.keys():\n",
    "        exac_freq = round((Exac_ASXL1_results[k][0])*100, 4)\n",
    "        african_freq = round((Exac_ASXL1_results[k][1])*100, 4)\n",
    "        east_asian_freq = round((Exac_ASXL1_results[k][2])*100, 4)\n",
    "        european_freq = round((Exac_ASXL1_results[k][3])*100, 4)\n",
    "        finnish_freq = round((Exac_ASXL1_results[k][4])*100, 4)\n",
    "        latino_freq = round((Exac_ASXL1_results[k][5])*100, 4)\n",
    "        south_asian_freq = round((Exac_ASXL1_results[k][6])*100, 4)\n",
    "    if k not in Exac_ASXL1_results.keys():\n",
    "        exac_freq = '-'\n",
    "        african_freq = '-'\n",
    "        east_asian_freq = '-'\n",
    "        european_freq = '-'\n",
    "        finnish_freq = '-'\n",
    "        latino_freq = '-'\n",
    "        south_asian_freq = '-'\n",
    "    if k in ASXL1_COSMIC_results.keys():\n",
    "        COSMIC_number = ASXL1_COSMIC_results[k]\n",
    "    if k not in ASXL1_COSMIC_results.keys():\n",
    "        COSMIC_number = '-'\n",
    "    ASXL1_s_dictionary[k]=(round(v[0]*100, 2),round(v[4]*100, 2),round(v[3]*100, 2), v[1], v[2], COSMIC_number, exac_freq, \\\n",
    "                            african_freq, east_asian_freq, european_freq, finnish_freq, latino_freq, south_asian_freq)\n",
    "\n",
    "for k, v in ASXL1_nonsense_s.items():\n",
    "    if k in Exac_ASXL1_results.keys():\n",
    "        exac_freq = round((Exac_ASXL1_results[k][0])*100, 4)\n",
    "        african_freq = round((Exac_ASXL1_results[k][1])*100, 4)\n",
    "        east_asian_freq = round((Exac_ASXL1_results[k][2])*100, 4)\n",
    "        european_freq = round((Exac_ASXL1_results[k][3])*100, 4)\n",
    "        finnish_freq = round((Exac_ASXL1_results[k][4])*100, 4)\n",
    "        latino_freq = round((Exac_ASXL1_results[k][5])*100, 4)\n",
    "        south_asian_freq = round((Exac_ASXL1_results[k][6])*100, 4)\n",
    "    if k not in Exac_ASXL1_results.keys():\n",
    "        exac_freq = '-'\n",
    "        african_freq = '-'\n",
    "        east_asian_freq = '-'\n",
    "        european_freq = '-'\n",
    "        finnish_freq = '-'\n",
    "        latino_freq = '-'\n",
    "        south_asian_freq = '-'\n",
    "    if k in ASXL1_COSMIC_results.keys():\n",
    "        COSMIC_number = ASXL1_COSMIC_results[k]\n",
    "    if k not in ASXL1_COSMIC_results.keys():\n",
    "        COSMIC_number = '-'\n",
    "    ASXL1_s_dictionary[k]=(round(v[0]*100, 2),round(v[4]*100, 2),round(v[3]*100, 2), v[1], v[2], COSMIC_number, exac_freq, \\\n",
    "                            african_freq, east_asian_freq, european_freq, finnish_freq, latino_freq, south_asian_freq)\n",
    "\n",
    "ASXL1_df = pd.DataFrame.from_dict(ASXL1_s_dictionary, orient='index', columns=['s (%)', 's lower value (%)', 's upper value (%)','observed number', 'site-specific mutation rate', 'Number of times in COSMIC',\\\n",
    "                                                                                'Exac Overall freq (%)', 'Exac African freq (%)', 'Exac East Asian freq (%)',\\\n",
    "                                                                                'Exac European freq (%)', 'Exac Finnish freq (%)', 'Exac Latino freq (%)', 'Exac South Asian freq (%)'])\n",
    "ASXL1_df_sorted = ASXL1_df.sort_values(by = ['s (%)'], ascending = False)\n",
    "\n",
    "ASXL1_df_sorted.to_csv('Data files/ASXL1_s_across_whole_gene.csv')\n",
    "# ASXL1_s_dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s (%)</th>\n",
       "      <th>s lower value (%)</th>\n",
       "      <th>s upper value (%)</th>\n",
       "      <th>observed number</th>\n",
       "      <th>site-specific mutation rate</th>\n",
       "      <th>Number of times in COSMIC</th>\n",
       "      <th>Exac Overall freq (%)</th>\n",
       "      <th>Exac African freq (%)</th>\n",
       "      <th>Exac East Asian freq (%)</th>\n",
       "      <th>Exac European freq (%)</th>\n",
       "      <th>Exac Finnish freq (%)</th>\n",
       "      <th>Exac Latino freq (%)</th>\n",
       "      <th>Exac South Asian freq (%)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>p.I1220F</th>\n",
       "      <td>21.46</td>\n",
       "      <td>16.50</td>\n",
       "      <td>25.98</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.042010e-10</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0099</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0181</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.C856*</th>\n",
       "      <td>14.17</td>\n",
       "      <td>11.34</td>\n",
       "      <td>15.92</td>\n",
       "      <td>2.0</td>\n",
       "      <td>6.414520e-10</td>\n",
       "      <td>1</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.S392*</th>\n",
       "      <td>13.53</td>\n",
       "      <td>10.95</td>\n",
       "      <td>15.09</td>\n",
       "      <td>2.0</td>\n",
       "      <td>8.230880e-10</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.Y591*</th>\n",
       "      <td>12.69</td>\n",
       "      <td>10.72</td>\n",
       "      <td>13.95</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.431380e-09</td>\n",
       "      <td>23</td>\n",
       "      <td>0.0025</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0046</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.Q588*</th>\n",
       "      <td>12.59</td>\n",
       "      <td>9.90</td>\n",
       "      <td>14.08</td>\n",
       "      <td>2.0</td>\n",
       "      <td>9.919690e-10</td>\n",
       "      <td>11</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.Q708*</th>\n",
       "      <td>11.55</td>\n",
       "      <td>9.48</td>\n",
       "      <td>12.67</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.964580e-09</td>\n",
       "      <td>8</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.Q768*</th>\n",
       "      <td>11.55</td>\n",
       "      <td>9.48</td>\n",
       "      <td>12.67</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.964580e-09</td>\n",
       "      <td>3</td>\n",
       "      <td>0.0017</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.Q803*</th>\n",
       "      <td>11.53</td>\n",
       "      <td>9.74</td>\n",
       "      <td>12.62</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.322790e-09</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0025</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0174</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.W1037*</th>\n",
       "      <td>10.48</td>\n",
       "      <td>8.66</td>\n",
       "      <td>11.45</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.524850e-09</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.Q432*</th>\n",
       "      <td>10.08</td>\n",
       "      <td>7.88</td>\n",
       "      <td>11.16</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.211140e-09</td>\n",
       "      <td>-</td>\n",
       "      <td>0.0025</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0045</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.W583*</th>\n",
       "      <td>9.30</td>\n",
       "      <td>7.11</td>\n",
       "      <td>10.35</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.610690e-09</td>\n",
       "      <td>3</td>\n",
       "      <td>0.0017</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R404*</th>\n",
       "      <td>9.12</td>\n",
       "      <td>8.28</td>\n",
       "      <td>9.71</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.882290e-08</td>\n",
       "      <td>6</td>\n",
       "      <td>0.0058</td>\n",
       "      <td>0.0096</td>\n",
       "      <td>0.0116</td>\n",
       "      <td>0.006</td>\n",
       "      <td>0.0151</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R693*</th>\n",
       "      <td>9.02</td>\n",
       "      <td>8.16</td>\n",
       "      <td>9.63</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.882290e-08</td>\n",
       "      <td>75</td>\n",
       "      <td>0.005</td>\n",
       "      <td>0.0098</td>\n",
       "      <td>0.0233</td>\n",
       "      <td>0.0046</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R417*</th>\n",
       "      <td>8.13</td>\n",
       "      <td>6.70</td>\n",
       "      <td>8.93</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.415120e-08</td>\n",
       "      <td>8</td>\n",
       "      <td>0.0074</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0116</td>\n",
       "      <td>0.009</td>\n",
       "      <td>0.0302</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          s (%)  s lower value (%)  s upper value (%)  observed number  \\\n",
       "p.I1220F  21.46              16.50              25.98              3.0   \n",
       "p.C856*   14.17              11.34              15.92              2.0   \n",
       "p.S392*   13.53              10.95              15.09              2.0   \n",
       "p.Y591*   12.69              10.72              13.95              3.0   \n",
       "p.Q588*   12.59               9.90              14.08              2.0   \n",
       "p.Q708*   11.55               9.48              12.67              2.0   \n",
       "p.Q768*   11.55               9.48              12.67              2.0   \n",
       "p.Q803*   11.53               9.74              12.62              3.0   \n",
       "p.W1037*  10.48               8.66              11.45              2.0   \n",
       "p.Q432*   10.08               7.88              11.16              2.0   \n",
       "p.W583*    9.30               7.11              10.35              2.0   \n",
       "p.R404*    9.12               8.28               9.71              7.0   \n",
       "p.R693*    9.02               8.16               9.63              7.0   \n",
       "p.R417*    8.13               6.70               8.93              3.0   \n",
       "\n",
       "          site-specific mutation rate Number of times in COSMIC  \\\n",
       "p.I1220F                 5.042010e-10                         1   \n",
       "p.C856*                  6.414520e-10                         1   \n",
       "p.S392*                  8.230880e-10                         -   \n",
       "p.Y591*                  1.431380e-09                        23   \n",
       "p.Q588*                  9.919690e-10                        11   \n",
       "p.Q708*                  1.964580e-09                         8   \n",
       "p.Q768*                  1.964580e-09                         3   \n",
       "p.Q803*                  2.322790e-09                         2   \n",
       "p.W1037*                 3.524850e-09                         1   \n",
       "p.Q432*                  3.211140e-09                         -   \n",
       "p.W583*                  4.610690e-09                         3   \n",
       "p.R404*                  1.882290e-08                         6   \n",
       "p.R693*                  1.882290e-08                        75   \n",
       "p.R417*                  1.415120e-08                         8   \n",
       "\n",
       "         Exac Overall freq (%) Exac African freq (%) Exac East Asian freq (%)  \\\n",
       "p.I1220F                0.0099                     0                        0   \n",
       "p.C856*                 0.0008                     0                        0   \n",
       "p.S392*                      -                     -                        -   \n",
       "p.Y591*                 0.0025                     0                        0   \n",
       "p.Q588*                      -                     -                        -   \n",
       "p.Q708*                      -                     -                        -   \n",
       "p.Q768*                 0.0017                     0                        0   \n",
       "p.Q803*                 0.0025                     0                        0   \n",
       "p.W1037*                     -                     -                        -   \n",
       "p.Q432*                 0.0025                     0                        0   \n",
       "p.W583*                 0.0017                     0                        0   \n",
       "p.R404*                 0.0058                0.0096                   0.0116   \n",
       "p.R693*                  0.005                0.0098                   0.0233   \n",
       "p.R417*                 0.0074                     0                   0.0116   \n",
       "\n",
       "         Exac European freq (%) Exac Finnish freq (%) Exac Latino freq (%)  \\\n",
       "p.I1220F                 0.0181                     0                    0   \n",
       "p.C856*                  0.0015                     0                    0   \n",
       "p.S392*                       -                     -                    -   \n",
       "p.Y591*                  0.0046                     0                    0   \n",
       "p.Q588*                       -                     -                    -   \n",
       "p.Q708*                       -                     -                    -   \n",
       "p.Q768*                   0.003                     0                    0   \n",
       "p.Q803*                  0.0015                     0               0.0174   \n",
       "p.W1037*                      -                     -                    -   \n",
       "p.Q432*                  0.0045                     0                    0   \n",
       "p.W583*                   0.003                     0                    0   \n",
       "p.R404*                   0.006                0.0151                    0   \n",
       "p.R693*                  0.0046                     0                    0   \n",
       "p.R417*                   0.009                0.0302                    0   \n",
       "\n",
       "         Exac South Asian freq (%)  \n",
       "p.I1220F                         0  \n",
       "p.C856*                          0  \n",
       "p.S392*                          -  \n",
       "p.Y591*                          0  \n",
       "p.Q588*                          -  \n",
       "p.Q708*                          -  \n",
       "p.Q768*                          0  \n",
       "p.Q803*                          0  \n",
       "p.W1037*                         -  \n",
       "p.Q432*                          0  \n",
       "p.W583*                          0  \n",
       "p.R404*                          0  \n",
       "p.R693*                          0  \n",
       "p.R417*                          0  "
      ]
     },
     "execution_count": 120,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ASXL1_df_sorted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABZ8AAAHvCAYAAAA7GaHXAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAIABJREFUeJzs3XucTfX+x/HXd4wZBjPjNkQYKZSKqAgd41rklhRxTqRy4kS66HaqofKLdCOhU0epdFVRiqGMXKLc73fGdQwzY9zGMJfv74+19z57z94zboNR7+fjsR975ru+67s+a3332g995ttnGWstIiIiIiIiIiIiIiIFKehiByAiIiIiIiIiIiIifz5KPouIiIiIiIiIiIhIgVPyWUREREREREREREQKnJLPIiIiIiIiIiIiIlLglHwWERERERERERERkQKn5LOIiIiIiIiIiIiIFDgln0VEREQuAcaYYGOMNcZ8cLFjKYyMMQOMMRuNMSdd1+lyV3tLY8zvxpijrva/X+xYRURERET+KpR8FhEREbmIjDExrqRoXq9G+ewbZIwZYozpeCFjLmyMMa2B0cAaoC/wDyDVGFMO+BYoBjzual9wnmKo75qLqudjfPH8AWaf67545hT9ehljFhhjkowxGcaY3caY2caYocaYEK++XxljcowxzQKMU8IYs8UYc8AYE+U1tjXGTDmNeLsbYz40xqwyxmR5/1FERERE5K8i+GIHICIiIiIAfA78FKB9C4C1NssYUxzI8toWBMQC/wW+P+8RFl6tXe/3W2sPuxuNMTFAOPCCtfZ8X5/6OHPxM7DzPB/rr6o9UAHYCvQBhufR70ugCzAPGAmkAVWBhsC/gbeAk66+/YFmwIfGmOuttUe9xnkNqAHcba3dfxbxPoLzuVgJbAOuOosxRERERC5pSj6LiIiIFA7LrLWf5tfBWptxoYK5xFQEsr0Tz17tAKkXOJ5LjjHGACVyJV8LmweATcCzwDfGmFuttfO8OxhjGuIknidba+/OPYAxpiLgOUdrbbIxph/wDU6yub+rX0ugH/CFtXbyWcbbE9jj+sPReJR8FhERkb8gld0QERERuQTkrvlsjLkSyHRtfsCrTEdW7v7GmKbGmHnGmHRjTLIx5j/GmBIBjlHZGDPeGLPLVTt5j+v3crn6lTXGjDLGbHOVNEgxxiwxxjyeq19vY8xiY0yaMeaYMWarMeZTY0zZ0zzn24wxs4wxh1zHWWmM6eu1/UpjjMUpp1HE6xr8bIzZjbMiHGCe97Vx7VvMGPO8MWada+yDxpjvjTF1A8RhjDH/NMb84aodfcRVSiHWtf0V4P1cx/Keq+LGmJeMU5M63XWs1caYvFbueh872BXnPFcJiZPGmB3GmHeNMWXy2OduY8yvruuWbozZ4JqvYNf2Vq74/mGcWtnrgRPAIK8xYlzX8bBrjKXGmN4BjnWdMeYb12flhDEm0TjlLdp69Tnr8/ca4zKgLTARmAak4CSjc3MneH8JNI61dp+1NitX27c4/+fBw8apER4OTACScFYvnxVr7Y7cxxIRERH5q9HKZxEREZHCISx3khc4Ya09kkf/fUAvnGTcHP6XaM3J1a8BcKdr+6dAC+AhnPId/d2djDHVgd+AIq6+7jIB/YDmxpibvFYWfwvcAowHVgElgKuBGOBN13j34yTwfgVeBI7jlD5oB5TDSR7myTirUd91xfQykA7cBrxnjKlurX3WdQ3+4Yqxket6ACTilNu4AydB+TLOitkc19ghwEycMgwf49SLLu26Lr8ZY5paa5e7+hqcxGQ3YCEwDKeMw9XAXcBQ4GucVdbexwJXyRTXdboP+Mg1RlHXtW2R3zVwKQY8gbMydwpwDLgZp7Z1E9e8uP8IgTFmBPAUsBZ4AyeBWsMV67/xWvXrGrc0znzvA3a4xugMTHZdx5GuY96LU5qiurXWnXQvD8wGsl3nuBNnbm9yxTi9AM7frTdggE+stSeNMZ8BfYwxA3LdI1td7/cYY76w1qad5viP4Hx+J+B85qoCnay1+X5ORURERCR/xlp7sWMQERER+csyTl3i+Dw2f2mt7e7qF4yz0vm/1toH82rzGte9LQdoaK1d4rUtDifRFmmtPe5q+xEnUV3fWrvXq29DnGRcrLX2Fddq2xTgHWvtwHzO63ugKVDeWpt9mpfDve/lOEnEL6219+Xa9i7wT6CGtdadLP0U6G6tDc7V90GcFcm3Wmvne7UPBkYAbay1P3u1R+I8tHCDtbaVq60HMAknyd/HWpvj1T/I/Xtex3JtOwzMsdae8YMhjTFBQKh7nrza/4mT1L3LtXIXY0xjnAcq/gy0t9ae8OpvrOsf/saYVsAsnHmsba1N9upXFNgOhAHXWGv3udpDcf6QcBNwlbV2mzGmC05S3BNDHudw1ufvNcYmYKfXvDQAlgB9rbXv5+r7E84q6WM4n93fXa/Z1tr0fI7Rgf/VTv/YWtsrQB/3fTXVWtv5DOIfj/O5rWKt3X26+4mIiIhc6lR2Q0RERKRw+A/Og/O8X68UwLjzvRPPLrOBEKAagCuh3BZnZe1JY0w59wsnCbwdaOPaNx0n+dbIGFMtn+MeAkoB7Vyrh8/E3a74JnjH4ornB5zV2S3PcExvf8dZGbwi19jBOOUamrmSreDU7bXAk96JZ4Dcv+fjEHCdMabOmQZqrc3x+gNBEWNMpCvW2a4uDb2693S9P+OdeHaNE2jFyUfeiWeXm4DKwAfuxLNr/xPA6zj//eBOIh9yvbczxpTK5zTO+vwBjDF/w1kpPdErnqXAapwHD+bWGXgUWIezuvp5nM9NojFmUID+bin87/8cmHY2sYqIiIiILyWfRURERAqHzdban3O91hTAuNsCtLlLCbhrL9fGKWnwT+BAgFcNoAJ4Hnr4OFAPSDDGrDHGjDbGNM91jFeA3TgrSfcbYyYbYx4wxpQ8jZivdr3HB4jFXcqhwmmMk5fawLUBxj6AUx4iGHDXU74K2B0gSXsmHsUpR7HGGLPFGPO+MabD6SbljTHdjTGLcUqXHHTF6S7tUdqr61U4JTBWnWZcmwK0VXe9rw2wzf15vML1Phv4DKfcSIoxZoExJtYYUzvXfud0/q7xTwIrjVPn+0rj1DyfifNHkGu8O1trT1prR1trb8Ypv/I3nJXuRYC3jDGBHkQYhlMWJBXY4+oXeZrxiYiIiEgeVPNZRERE5M8tv5IXJtf7RJy60IF4yhVYa8cYY77Dqan8N+AeYIAxZpK19u+uPhtdSchWOKuUmwEfAEOMMX+z1m4/jbh6Avvz6LMlj/bTEQSsAAbn0yfVK5ZzqlNnrf3WGPMrTr3rZjir2h8E5hhj2njXbM7NGHMPTs3pRcBAnIR+Bs7K8B/xXUxypivMA5WgOO0xXKupe7oeHNgWuBWn3vQLrlrM41z9zuX8w4GuOOe7Mo9ufYAn84gxHZiH8yDIX4GfcJLZX+fqOhwned8NOOLq9xZwf74XQURERETypeSziIiIyKWroB7escU1VlHvGsj5HtjaPTilQv7jqoM7CScR+Yb7YX2uUg0/ul4YYzoCU4FBOKth87LZ9X7gdOM5Q5uB8sAveZSj8LYRp6xEuVOsfs53HNeD6z4BPnGt+B2J88C/9sB3+ez6D5wkcXPXqnMAjDHX5hFrK+A6YFl+8eTD/cC+QCUy3CuMfVbTW2tX45TAeM1VwuUPnGTuOK8+Z3v+3XHqTz+d+7gujwH/MMY8m18S22WR672yd6Or7vojwNfW2q9cbR8C9xtjvrbW/nSKcUVEREQkDyq7ISIiInKJcj3IL4P/lYg423GScEoY3G2MuSn3duMo7/o5zBhTPNf+WTjJR9yxuOoS57bMu08+vsQps/CSMaZYgHgijTEhpxgjPx/jJCADJsCNMd4lPSbhrAZ+LXeZiFy/H3W9l8nVJ9gYE+Hd5kp4rwjUP4BsnDrEnn+3u477fIC+n7neX819fc6gxMVinLITDxhjorz2D8FZXZyD66F8xpgyuce11qYCCUBJY0xIAZz/A0Ay8Ia1dnLuFzABiMJJYmOMqWWMuSKPsdwPCFzndV4lgQ9xSpn09+r7GM4q8/+o/IaIiIjI2dPKZxEREZFL2yLgNmPMU8AuINu9evMM/ROYD8w3xnwMLMf5t2J1nKTdf3HqOF8D/Owqu7EWpwbxNUA/nFWzC1zjzTbGHMApebALJ8l4P07yMq/SHgBYa3cYYx4BxgPrjDGfAjtxVitfD3QCauIkB8/GmzgrhN8yxrQC5uCUWqiKUyLkCE5pCIAvgC6u2GsaY37AeYBeTVffuq5+f+Csfn7elag/xv8e1rjTGPM9TsJ1P07N5H44pT1+PEWsk13nO9sY8wkQCtwJ+CXlrbW/GWPewFlRvMQY8zWwD2cO7zbG3GCtPZp7v1xjZBljBuCUpVhsjHnfdS7dgZuBl6y17hXIfYBHXJ+FLUAW0Nx1XT6z1p50/RHirM7ftbr7ZpyHH+ZVPmYq8B5Okvo74AZgkjEmHpiL8xkpCTTCKd9xCHjZa/83gGjgLu+V7dbaQ8aYvuRdfqOmMSbQHwDASZQfN8Y0wylFAlDf9T7QGHMYyLHW/l9e5y4iIiLyZ6Hks4iIiMil7WFgDM5K2FI4K2XPOPnsSvjWB54BOuI8eC8dJ3E8BScJCrADpzZ0DE4SNBQnwfceMMKrNMS7wN2u+MrgPORwGdDfWjvnNOJ53xizHme1bT8gAmcF7Abg3zgrVc+KKyl6O06phb8DQ12b9uIk8yd69bXGmG7Av3CSrbE413gbXnWDrbXbjTEP4tQ8HgcUxUnY/wsYjZOQbQ2UABJxEqWvWmv3nSLWT12rcx/FSZSm4qw8foEA9bCttU8aY5a7jvs0zqrtXcA0nFXyp3N9vjPGtMb5TD2N898M64E+1toPvbrOxvljQAfgMpzk83ac5PcYV58j53D+D7jev80n1mRXLefbjTGVcB5S+TTOHxcewFkVDc5ndALwmjt5boxpA/QFPrfW+h3DWjvdGDMB6BOg/MbV+CaxvY3HeThka5zPqjd3nfFsQMlnERER+dMzpy5zd/EYY0YAN+KsLCmH84+4HTj/ATTGVTsu9z6Ncf6h3AhnRcgWnH9ovpPPigkRERERERERERERKUCFPfl8EmeFzDqclR0lcJLKN+KsTGlkrd3l1b8T8A3Oqo4vcVaGdABqAZOttXdf0BMQERERERERERER+Ysq7MnnYt5P9fZqHwY8B4yz1vZ3tYXjrHKOAJpYa5e4x8D5XwJvAe611n5xoeIXERERERERERER+asKOnWXiydQ4tnFXcfwKq+2rjgPofnCnXj2GsP9MJB+BR6kiIiIiIiIiIiIiPgp1MnnfHRwva/yamvhep8RoP9cnAfmNDbGhJ7PwERERERERERERETEeXJ1oWeMeRIoiVNS40agKU7iebhXt1qu902597fWZhljtgN1gCtwntYd0Lhx4yxAv379TIEELyIiIiIiIiIiIvIXdKmsfH4SiAUG4SSeZwBtrLUHvPpEuN4P5TGGuz0y0EZjTF9jjKdcx6FDh+zy5cttZmamXbJkiQXskiVL7IkTJ+zKlSvtwYMH7ZYtW+zu3bttYmKi3bhxoz18+LBdvny5zcrK8ttn9erVNjU11W7dutXu2rXL7tu3z27YsMEeOXLELl261GZnZ/vsc/z4cbtmzRqbkpJit23bZnfu3GmTkpLs+vXr7dGjR+2SJUtsTk6Ozz7p6el27dq19sCBA3b79u12x44ddv/+/XbdunX22LFjnr6LFy/27HPs2DG7bt06u3//frtjxw67fft2e+DAAbt27Vqbnp7uM777eEePHrXr16+3SUlJdufOnXbbtm02JSXFrlmzxh4/ftxnn+zsbLt06VJ75MgRu2HDBrtv3z67a9cuu3XrVpuammpXr15tT5w44bNPVlaWXb58uT18+LDduHGjTUxMtLt377ZbtmyxBw8etCtXrvTbJzMz0y5fvtweOnTIbtq0ye7du9fu2bPHbt682aalpdkVK1bYzMxMn3M/efKkZy43b95s9+zZY/fu3Ws3bdpkveffvc/ixYvtiRMn7KpVqzxzmXv+ly1b5jeXGRkZdvXq1T5zmXv+c8+le/6Tk5M9c5mUlGTXrVvnmX9rbZ7zn5CQYBMSEuyBAwf85t/7/ejRo37zn5ycbNesWWPT09N9rldOTo5dunRpvvOfkZHh8znLzs62y5Yts4cPHw44/6tWrcpz/g8dOmQ3btxo9+7d6zf/J0+e9Jv/FStW2LS0NLtp0ya7Z88eu2fPHrtp0yaf+ffex3v+3ffy3r177caNGz3zH+he9p5/73vZe/69r1tGRka+93Lu+V+8eLFNT0/3m3/3veye/9xz6b6Xc89/oHvZ/dlxz39SUpLf/Oe+l93z730ve8//6tWrfebfff+759/7Xs49/973WH73cqD5d+/jnv/87mV9l+u7XN/l+i7Xd7m+y/Vdru9yfZfru1zf5fou13e5vsv/it/lXESF+oGDuRljKgCNcVY8lwLaW2uXubZtwqkBfZW1dkuAfX/DeejgLdbaRXkdw2vlc8GfgIiIiIiIiIiIiMiFddEqPFwqK58BsNYmWWu/A9oAZYGPvTa7VzZH+O3oCM/VT0RERERERERERETOk0sq+exmrd0BrAPqGGPKuZo3ut5r5u5vjAkGqgNZwLYLEqSIiIiIiIiIiIjIX9glmXx2qeR6z3a9z3a93x6g79+AMOA3a+2J8x2YiIiIiIiIiIiIyF9doU0+G2NqG2MqBmgPMsYMA6JwkskHXZsmA8lAd2PMjV79iwGvuH4dd57DFhEREREREREREREg+GIHkI/bgZHGmLnAViAFqAA0A64A9gEPuTtbaw8bYx7CSULPMcZ8AaQCHYFarvYvL+gZiIiIiIiIiIiIiPxFFebk88/Af4AmQF0gEjgGbAI+AUZba1O9d7DWTjHGNAP+DdwFFAO2AI+7+tsLF76IiIiIiIiIiIjIX1ehTT5ba9cA/zqL/RYA7Qo+IhERERERERERERE5XYW25rOIiIiIiIiIiIiIXLqUfBYRERERERERERGRAqfks4iIiIiIiIiIiIgUOCWfRURERERERERERKTAKfksIiIiIiIiIiIiIgVOyWcRERERERERERERKXBKPouIiIiIiIiIiIhIgVPyWUREREREREREREQKnJLPIiIiIiIiIiIiIlLglHyWQmXIkCEYY5gzZ87FDkVERERERERERETOgZLPIiIiIiIiIiIiIlLglHyWQuWRRx5h/fr13HzzzRc7FBERERERERERETkHwRc7ABFv5cqVo1y5chc7DBERERERERERETlHWvks5ywhIQFjDL1792br1q107dqVsmXLUqpUKdq0acOaNWsAOHDgAH379uWyyy6jWLFi3HTTTcTHx/uMlVfN53nz5tGhQwcuv/xyQkNDqVixIo0aNWLo0KE+/ZKSknjyySepVasWJUqUIDIyklq1atG7d2+2bdvmF3tcXBzt2rWjXLlyhIaGUqNGDQYPHkxaWppf3+joaKKjo0lPT2fw4MFUrVqV0NBQrrzySkaMGIG11m+f77//npYtW3LZZZcRGhpKpUqVaNasGWPHjvXrm5qayrPPPsvVV19N8eLFiYiIoGXLlsycOfOUcyAiIiIiIiIiIlLYaOWzFJiEhAQaNmzI1VdfTe/evUlISOC7774jJiaGhQsXcvvttxMeHk63bt1ITU3liy++oG3btmzatImqVavmOe6MGTO44447CA8Pp2PHjlSuXJnU1FTWr1/P2LFjiY2NBSA9PZ0mTZqwdetWWrduTYcOHbDWsmPHDqZOnUrXrl254oorPOO+9NJLxMbGUqZMGdq3b09UVBSrVq3i9ddf56effmLhwoWEh4f7xJKZmUmbNm3Yu3cvbdu2JTg4mClTpvDMM8+QkZHhiQXgP//5D//85z+pWLEiHTp0oFy5cuzfv59Vq1bx4Ycf0r9/f0/fHTt2EBMTQ0JCArfeeiu33347x44dY9q0adx+++289957PPTQQwU1VSIiIiIiIiIiIuedks9SYH799VdeeeUV/v3vf3vaXn75ZV588UUaNmzIPffcw9ixYwkKchbct27dmvvuu4+33nqLt956K89x33//fXJycpgzZw5169b12ZacnOz5+ZdffmHr1q0MGjTIb7yTJ09y4sQJz+/x8fHExsZyyy238NNPPxEZGenZ9tFHH3H//fcTGxvrN87evXupW7cus2bNonjx4gDExsZSs2ZN3nrrLZ577jmKFi0KwHvvvUdISAgrV64kKioqz7gBevXqxY4dO/j888/p3r27pz0tLY2YmBgGDhxIx44dqVChQp7XSUREREREREREpDBR8vk86bEo82KHcNo+a1S0QMaJjo7mmWee8Wnr1asXL774IidOnGDkyJGexDNAjx496NOnDytWrDit8d3JXm+B6kMH6hcSEkJISIjn99GjRwNOYts78QzQu3dvRo0axaRJkwImxUePHu1zjKioKDp16sTHH3/Mxo0bufbaaz3bgoODPcnovOJeuXIlv/76K127dvVJPANERkYydOhQOnfuzDfffOOzWlpERERERERERKQwU/JZCky9evUoUqSIT1ulSpUAqFmzJqVKlfLZVqRIESpUqMDu3bvzHbdnz558++23NGzYkG7dutG8eXOaNGnC5Zdf7tOvWbNmVK5cmeHDh7Ns2TLatWtHkyZNAsa1cOFCihYtytdff83XX3/td8yTJ09y4MABUlJSKFu2rKc9IiKCK6+80q9/lSpVADh48KBP3E888QR16tShW7duNGvWjCZNmlC+fHm/WAAOHTrEkCFD/MY+cOAAAOvXr8/vMomIiIiIiIiIiBQqSj5LgYmIiPBrCw4OznObe3tmZv6rxLt06cK0adN44403mDBhAu+99x4ADRo04NVXX6V169YAhIeHs2jRImJjY/n++++Ji4sDnFXG/fv35/nnn/esQk5JSSErK8vvgYW5HT161Cf5nHuVdO7zzM7O9rQ9/vjjlCtXjrFjxzJ69GjefvttjDE0a9aMkSNHcuONN3piAZg1axazZs3KNxYREREREREREZFLhZLP50lBlbIQxx133MEdd9zBsWPH+P3335k2bRrjxo2jffv2LF++nGuuuQaAyy+/nP/+979Ya1m3bh2zZ8/m3Xff5aWXXiInJ4eXX34ZcJLhOTk5pKamnte477vvPu677z7S0tL47bff+O6775gwYQK33XYb69evJyoqypOYHzVqFAMHDjyv8YiIiIiIiIiIiFwoQafuIlJ4lChRghYtWvDmm2/y3HPPcfLkSaZPn+7XzxhDnTp1GDBggGc18ZQpUzzbGzVqxMGDB1m7du0FiTsyMpJ27drx/vvv07t3b1JTU5k3b54nFsDzu4iIiIiIiIiIyJ+Bks9S6P3yyy8cP37crz0pKQmAsLAwANasWUNCQsIp+wE89thjADz00EPs3bvXb59jx46xaNGic4p7xowZZGVl+bXv37/fJ54bb7yRW2+9lW+//ZYJEyYEHGv16tWe/URERERERERERC4FKrshhd4TTzxBQkICMTExREdHExISwtKlS5k9ezbVqlWje/fuAPz88888/vjjNG7cmNq1axMVFcXu3buZOnUqQUFBDB482DNmy5YtGT58OM8++yxXXXUV7dq1o3r16hw9epQdO3bw66+/0rRpU2bMmHHWcXfv3p1ixYrRtGlToqOjsdYyb948Fi9eTIMGDWjVqpWn72effUaLFi144IEHGD16NA0bNiQyMpLdu3ezatUq1qxZw8KFC4mKijr7CykiIiIiIiIiInIBKfkshd5zzz3Hd999x5IlS/j5558JCgqiatWqPPfccwwaNIjSpUsDcNtttzFo0CDmzp3L1KlTOXz4MJdddhmtW7f2JKW9Pf300zRp0oTRo0czf/58pk6dSkREBJUrV6Zv37706NHjnOIePnw4cXFxLFu2jJ9++olixYpRrVo1RowYQb9+/TwPPwSnVvXSpUt55513+Oabb5g0aRLZ2dlUrFiRa665hgEDBnDdddedUzwiIiIiIiIiIiIXkrHWXuwYCpVx48ZZgH79+l3sUERERERERERERETOlblYB1bNZxEREREREREREREpcEo+i4iIiIiIiIiIiEiBU/JZRERERERERERERAqcks8iIiIiIiIiIiIiUuCUfBYRERERERERERGRAqfks4iIiIiIiIiIiIgUOCWfRURERERERERERKTAKfksIiIiIiIiIiIiIgVOyWcRERERERERERERKXBKPouIiIiIiIiIiIhIgVPyWUREREREREREREQKnJLPIiIiIiIiIiIiIlLglHwWERERERERERERkQKn5LOIiIiIiIiIiIiIFDgln0VERERERERERESkwCn5LCIiIiIiIiIiIiIFTslnERERERERERERESlwSj7LOUtISMAYQ+/evUlISKB79+6UK1eOYsWKceONNzJt2jS/fU6cOMHw4cO5/vrrCQsLIzw8nFtvvZWvvvqqQMY/efIko0ePpn79+pQuXZqwsDCio6Pp1KkTP//8s1//DRs20Lt3b6pUqUJoaCgVKlSgR48ebNy40a9v7969McaQkJDAe++9x3XXXUexYsWoUKECffv25dChQ377rFq1invvvZfo6GhCQ0MpX7489evXZ9CgQWRmZvr0zcrKYuzYsTRq1Ijw8HDCwsK44YYbGDNmDDk5OfnOhYiIiIiIiIiISGFhrLUXO4ZCZdy4cRagX79+FzuUS0ZCQgLVq1cnJiaGtWvXcsUVV3DLLbeQmprKl19+SWZmJj///DPNmzcHnMRwmzZt+PXXX6lduzbt27cnPT2dyZMns3//fp599ln+7//+76zHB+jRoweff/451157LS1atKB48eLs3buX+fPn06VLF15//XVP3xkzZtClSxcyMzPp0KEDV155Jbt37+bbb78lNDSU+Ph46tev7+nfu3dvJk6cyN13301cXBwdOnSgQoUKxMfHs3z5cpo3b87s2bM9/VetWkXDhg0xxtCxY0eqV6/O4cOH2bJlC/Hx8aSmplKyZEkATwxxcXHUqlWLmJjip1ojAAAgAElEQVQYihUrRnx8PKtWreLvf/87n3zyyXmbSxERERERERER+dMxF+vAwRfrwPLnM2fOHIYMGUJsbKynrUePHtx+++2MHDnSkxx+4403+PXXX2nbti3ff/89wcHOxzA2Npabb76ZV199lfbt29O4ceOzGv/QoUN88cUXNGjQgN9//50iRYr4jJOSkuL5+eDBg9x7772EhYUxd+5crrnmGs+2tWvX0rBhQx588EGWLVvmd76LFi1i9erVVK1aFXBWLLdo0YL4+Hj++OMPbr75ZgAmTpxIRkYGU6ZMoVOnTj5jHDx4kLCwMM/vw4YNIy4ujkceeYS3337bE3t2djZ9+/ZlwoQJdO3a1W8cERERERERubSM3JCV7/bBtZWyEZFLn77JzhPTefTFDuG02SkDC2ScatWq8fzzz/u03XbbbVStWpU//vjD0zZhwgSMMbz55puexDNAVFQUL7zwAg8++CAffPCBX/L5dMc3xmCtJTQ0lKAg/8oyZcuW9fz88ccfk5aWxpgxY3wSzwB16tThoYce4u2332bdunV+21988UVP4hkgODiY+++/n3nz5vkkn92KFy/uF0vp0qU9P+fk5DBmzBgqVqzIW2+95ZM0L1KkCG+88QYffvghkyZNUvJZRERERETkT+Tgyf/9X+mlQy7aAkURkQKn5LMUmHr16vmtMgaoUqUKCxcuBODIkSNs2bKFypUrU7t2bb++LVq0AGD58uVnNT5AeHg4HTp04IcffqBevXrcdddd3HrrrTRs2NBnlTHg2W/lypUMGTLEb+xNmzYBsH79er/k84033hgwFnBWNLt169aNUaNG0blzZ7p27UqrVq1o0qQJNWrU8DtWSkoKV111Fa+88orf2OAksNevXx9wm4iIiIiIiFw6vFc2e6+C1opnEfkz0TeaFJjIyMiA7cHBwZ4H5bkfxnfZZZcF7OtuT0tLO6vx3b788ktGjBjBZ5995inTUaxYMbp27crrr79OhQoVgP+V4Hj//ffzPbejR4+eVjzuldzZ2dmetptvvpl58+YxbNgwJk+e7KnZXKtWLWJjY7n33nt9Ytm8eTNDhw49o1hEREREREREREQKGyWfz5OCKmXxZxMREQHAvn37Am5PTEz06Xe2ihcvzpAhQxgyZAi7du1i7ty5fPTRR3z66ackJCQwb948n+OsXLmS66+//pyOmZ9bbrmFadOmceLECZYuXcqMGTN455136NGjB+XLl6dVq1aeWO68806+/fbb8xaLiIiIiIiIiIjIheBfEFfkPCpVqhQ1atRgz549bN682W97fHw8APXr1y+wY1apUoWePXsSFxfHVVddxfz58z2rjBs1agTgSUafb6GhoTRu3JiXXnqJ0aOduuBTp04FoHbt2kRGRrJo0SIyMzMvSDwiIiIiIiIiIiLni5LPcsH16dMHay2DBw/2KU+RnJzMyy+/7Olztg4cOMDvv//u137s2DGOHDlCcHAwISEhANx///1ERkYydOhQn4cWuuXk5DBnzpyzjgWcxLa73Ii3pKQkAE8d6uDgYAYMGEBiYiIDBw7k+PHjfvskJiaybt26c4pHRERERERERETkQlDZDbngnnzySaZPn87UqVOpW7cu7dq1Iz09na+//pr9+/fz1FNP0bRp07Mef8+ePTRq1Iirr76a+vXrU6VKFQ4fPsy0adPYt28fAwcOpFSpUgCULVuWyZMnc+edd9KoUSNatmxJnTp1CAoKYufOnSxcuJCUlBQyMjLOOp433niDmTNnEhMTwxVXXEHJkiVZu3Yt06dPp3Tp0vTt29fT94UXXmDlypWMHz+eH374gRYtWlC5cmX279/P5s2bWbBgAcOGDfN7+KGIiIiIiIiIiEhhU2iTz8aYssCdwB3AdUBl4CSwGvgQ+NBam+PVPxrYns+QX1pru5+veOX0hYSEMGvWLN58800+++wz3nnnHYKDg6lbty5vv/225wF8Zys6OpqhQ4cyZ84c4uPjSU5OpkyZMtSqVYvhw4fTvbvvx6Bly5asWrWK119/nbi4OObNm0dISAiVKlWiRYsW3HXXXecUT//+/SldujS///47CxYsICsri8svv5z+/fvzxBNPUK1aNU/fokWLMmXKFD799FM++ugjpk2bxtGjRylfvjzVq1fn5ZdfpmfPnucUj4iIiIiIiIiIyIVgrLUXO4aAjDEPA+OARCAe2AlUALoAEcA3wN3WdQJeyeeVwJQAQ66x1k4+1XHHjRtnAfr163fO5yAiIiIiIiIiciojN2R5fh5cu9CuExSRS5e5WAcuzN9om4COwI+5Vjg/B/wB3IWTiP4m134rrLVDLlSQIiIiIiIiIiIiIuKv0D5w0Fo721r7g3fi2dW+Dxjv+jXmggcmIiIiIiIiIiIiIqdUmFc+5yfT9Z4VYFslY8w/gbJACrDQWrvqgkUmIiIiIiIiIiIiIpde8tkYEwzc5/p1RoAurV0v733mAL2stTvPb3QiIiIiIiIiIiIiAlBkyJAhFzuGMzJ06NDXgM7AT9bal7zaQ4AM4HHgGWAEMB2IBpoDHYYOHTphyJAhmX6DAsaYvkOHDn3vjjvuqARQs2ZN1q9fT7ly5Vi+fDmVKlViyZIllC1blrVr1xIaGsqePXtIT0/n2LFj7N27l5CQENatW0f58uVZtmyZzz7r16+naNGi7N27l2PHjnH8+HF2795NaGgoa9asoUKFCj77lC5dmo0bN1KkSBH27dvHkSNHyMjIYNeuXRQvXpxVq1ZRsWJFli5d6tknMjKSTZs2YYxh//79HD58mJMnT7Jjxw7CwsJYtWoVlSpVYvHixVSuXJklS5YQERHBli1bsNaSnJzMwYMHycrKIiEhgZIlS7Jy5UrP+O7jhYeHs3XrVnJyckhNTSUlJYWcnBy2bdtGqVKlfPapUKECy5cvp1SpUmzfvp2srCwOHjxIcnIyAFu2bCEyMpIVK1Z49omKimLlypWUKFGChIQEMjMzOXToEPv378cYw6ZNmyhdurTPPuXLl2fVqlWEhYWxc+dOTpw4weHDh0lKSqJIkSJs2LCBsmXLsmzZMs+5lytXjjVr1hAaGsru3bvJyMjg6NGjJCYmEhwc7Jl/9z6LFy+mXLlyrFu3zjOXued/7dq1REVF+cxlmTJlWL9+PcHBwZ65zD3/uefSPf9BQUEkJSVx+PBhTpw4wc6dOz3zf9lll+U5/wcOHCAtLY3MzEy/+V+yZInnPTw83G/+s7Oz2bZtGyVLlmTFihWe61WxYkWWLVuW7/yHh4d75mXx4sVUrFiRFStWUKJEiYDzv3nz5jznPywsjB07dnDy5Em/+S9TpozPfVm+fHlWr15NsWLF2LlzJxkZGRw5coR9+/b5zL/3Pt7z776Xjx49yt69eylatCjr168PeC97z7/3vew9/0uXLvVctzJlyrBhw4Y87+XVq1f7zP/ixYsDzr/7XnbPf+65dN/LgM/8B7qX3Z8d9/zn5OSQkpLiM/+572X3/Hvfy97zv3XrViIiInzmskKFCp75976XDxw44DP/y5cv99xjUVFRed7LQUFBfvPv3sc9/7t27crzXtZ3ub7L9V2u73J9l+u7XN/l+i7Xd7m+ywvrd/mKDVvA5hB0NIWK2Wn6Ltd3ub7L9V1eoN/ll1122dCzSMMWCGOtvVjHPmPGmIHAKGAD0MRam3oa+wQD84GGwCBr7aj8+o8bN84C9OvX79wDFhERERERERE5hZEb/ldVdHDtS+5/UheRws9crAMX2gcO5maM+RdO4nkd0Px0Es8A1tos4APXr387T+GJiIiIiIiIiIiIiJdLIvlsjBkEjAHW4CSe953hEAdc7yUKNDARERERERERERERCajQJ5+NMU8DbwErcBLP+89imEau920FFpiIiIiIiIiIiIiI5KlQJ5+NMS8Aw4GlQEtrbXI+fRsaY0ICtLcAHnP9+ul5CVREREREREREREREfBTaKvbGmF7AS0A2MA8YaIxfbewEa+1Hrp9HAHWMMXOA3a6264EWrp9fsNb+dj5jFhERERERERERERFHoU0+A9Vd70WAQXn0+RX4yPXzJ8CdwE1AW6AokAR8BYyx1s47b5GKiIiIiIiIiIiIiI9Cm3y21g4BhpxB//8C/z1f8YiIiIiIiIiIiIjI6SvUNZ9F/mp69+6NMYaEhISLHUqhN2fOHIwxDBky5GKHctpiYmIIUD5IRERERERERORPSclnOScJCQkYY+jdu/fFDsVPZmYm3333HQ888ADXXnst4eHhhIWFcd111/Hiiy9y5MiRPPfdvXs3ffr0oVKlSoSGhhIdHc2gQYM4ePDgBTwDkYJRmO9TEREREREREfnzKrRlN0TO1datW+nSpQslSpSgefPm3HHHHRw9epS4uDhefvllvvzySxYsWEC5cuX89mvcuDH79++nU6dO1K5dmz/++INRo0YxY8YMFixYQNmyZS/SWcml7OOPPyY9Pf1ihyEiIiIiIiIickEo+Sx/WqVKleLdd9+lV69elChRwtN+8uRJunTpwo8//sjQoUN55513fPbr378/+/fvZ/To0QwYMMDT/vjjj/PWW2/x73//m/Hjx1+w85A/j6pVq17sEERERERERERELhiV3ZCzNmTIEKpXrw7AxIkTMcZ4Xh999JGnX05ODuPHj+emm26iZMmSlChRgptuuolx48aRk5PjN64xhpiYGPbu3cs//vEPoqKiKF68OA0aNOCzzz477fgqV65M//79fRLPACEhITz33HOAUzfY27Zt25g5cybR0dH861//8tk2dOhQSpQowSeffMKxY8dOO46lS5dy1113ERUVRWhoKNWqVaN///4kJibmuU9OTg5vvvkmtWvXplixYlx++eU89thjHD582K/vqlWruPfee4mOjiY0NJTy5ctTv359Bg0aRGZmpk/frKwsxo4dS6NGjTxlSG644QbGjBnjNxfepRo2bdpEt27diIqKIigoiDlz5lC7dm1CQkJITk4OeA7Dhw/HGMO7777r0757924eeeQRrrjiCkJDQylbtiwdO3Zk8eLFAcdJSkrigQceoEKFChQvXpx69eoxceLEPK9dIK+++irGGEaPHh1w+969eylSpAg33XSTT9tLL71EkyZNqFixIiEhIVSqVIkePXqwfv16vzFOdb0gcM3nkydPMmbMGNq1a0e1atUIDQ2lTJkytGrViunTpweMNzo6mujoaNLT0xk8eDBVq1YlNDSUK6+8khEjRmCt9fQ93ftURERERERERKSgaeWznLWYmBjS0tIYNWoUdevWpXPnzp5t9erV8/z8j3/8g88++4wqVarw4IMPYozhu+++o3///syfP59Jkyb5jX3w4EEaN25MZGQk999/P2lpaXz11Vf07NmTPXv2MHjw4HOKvWjRogAEB/veArNnzwagTZs2BAX5/m2mVKlSNGnShJkzZ7Jo0SJatmx5yuNMmzaNu+66C2stXbt2pVq1aixdupRx48YxdepUFixYQHR0tN9+jz32GHPnzuWee+6hU6dOxMXF8fbbbzNv3jzmz59PsWLFACfx3LBhQ4wxdOzYkerVq3P48GG2bNnC2LFjeeWVVzznmpmZSYcOHYiLi6NWrVr06NGDYsWKER8fz4ABA/j999/55JNP/GLZunUrDRs2pGbNmvTs2ZPjx48THh5Or169eO655/j88899Voi7ffzxx4SEhNC9e3dP27Jly2jTpg2pqancdtttdOnSheTkZKZMmULTpk357rvvaNeunad/SkoKjRs3Ztu2bTRt2pSmTZuSmJjIww8/TJs2bU55/d3uu+8+nn/+eSZOnMjAgQP9tn/66afk5OTQq1cvT9vcuXMZPnw4zZs356677qJkyZJs3ryZyZMn8/3337NgwQLq1q172tcrL6mpqTz66KM0btyY1q1bU758eRITE/nhhx9o164d77//Pg8++KDffpmZmbRp04a9e/fStm1bgoODmTJlCs888wwZGRnExsYCp3+fioiIiIiIiIgUOGutXl6vsWPH2rFjx1o5Pdu3b7eA7dWrV8Dtn332mQXsDTfcYI8cOeJpP3r0qG3QoIEF7KRJk3z2ASxg7777bpudne1p37Ztmy1durQtWrSo3bp16znF/fDDD1vAPvPMMz7tTz75pAXs66+/HnC/f/3rXxY4rc/IkSNHbNmyZW1QUJCdO3euz7bhw4dbwLZu3dqnvVevXhawZcuWtQkJCZ727Oxs26VLFwvYl156ydP++OOPW8BOmTLF7/ipqak+1y82NtYC9pFHHrFZWVme9qysLNunTx+/cdxzC9hnn33Wb/zdu3fboKAg26BBA79tf/zxhwVsly5dPG2ZmZm2Ro0aNjQ01M6ZM8en/549e2ylSpVsxYoVbUZGhqf9oYcesoAdNGiQT//Fixfb4OBgC9jY2Fi/4wfSpk0bC9jVq1f7bbvmmmts0aJFbXJysqctKSnJHj582K/vihUrbIkSJeztt9/u036q62Wttc2aNbPO1+7/ZGRk2F27dvn1TUtLs3Xq1LGlS5e26enpPtuqVatmAdu2bVufbUlJSTYiIsJGRETYkydP+sWW130qIiIiIiIX12vrMz0vEZHz4KLlWrXy+Xype/fFjuD0rfz6vA09YcIEwCnBULJkSU97iRIlGDFiBK1ateKDDz6gR48ePvsVKVKEESNG+Kw+rl69OgMHDmTo0KF88sknnpWdZ+r777/nvffe4/LLL+epp57y2Xbo0CEAIiIiAu7rbk9LSzvlcaZOnUpKSgr33nsvt956q8+2J554gvHjxzNr1ix27tzpVwv40UcfpVq1ap7fg4KCGDlyJFOmTGHChAm88MILPv2LFy/ud/zSpUt7fs7JyWHMmDFUrFiRt956iyJFini2FSlShDfeeIMPP/yQSZMm0alTJ59xKlSoEPBaV65cmZYtWzJr1izWrl1LnTp1PNvcZTG8VxL/+OOPbN26lSeffJJmzZr5jFWpUiWeeuopBg0axC+//EK7du3IzMxk0qRJlCpViiFDhvj0v/HGG+nZs+cZld/o1asXM2fOZOLEiYwcOdLTvmTJEtatW8edd97p8yDJqKiogOPUrVuXFi1aMHPmTDIzMz0ry93yul55CQ0N5fLLL/drj4iIoE+fPjzxxBMsXryYv/3tb359Ro8e7TP3UVFRdOrUiY8//piNGzdy7bXXnnYcIiIiIiIiIiIFTclnOa+WLVtGUFAQMTExftuaNWtGkSJFWL58ud+2qlWreurUeouJiWHo0KEB9zkdv/32Gz169KBEiRJ88803Pgna02FdtXRz1+0NZNmyZQC0aNHCb1twcDB/+9vfSEhIYPny5X7J59zJWYArrriCKlWqkJCQQFpaGpGRkXTr1o1Ro0bRuXNnunbtSqtWrWjSpAk1atTw2XfTpk2kpKRw1VVX8corrwSMt3jx4gFrGdetW5fQ0NCA+/Tu3ZtZs2YxceJEXnvtNcCpYfzFF19Qvnx5nxIaCxcuBGDHjh1+yWSAzZs3A7B+/XratWvHhg0bSE9P59Zbbw34x4CYmJgzSj7feeedRERE8OmnnzJ8+HBPAt49Ru/evf32+fHHHxk/fjxLliwhOTmZrKwsn+3JyclcdtllPm35Xa+8rF27lpEjRzJ37lwSExPJyMjw2b5nzx6/fSIiIrjyyiv92qtUqQI4pWtERERERERERC4mJZ/lvDp06BBlypQhJCTEb1twcDDlypVj//79ftsqVKgQcLyKFSt6xj1TCxcupG3btgQFBTF9+nRuvvlmvz7uJGde47sf+JfXymhv7jFyJyfd3O2BVlHnd/47duzg0KFDREZGcvPNNzNv3jyGDRvG5MmTPTWba9WqRWxsLPfeey/g1E4GJ8E7dOjQPGM+evRowGPm5c477yQ8PJxPP/2UV199lSJFijBt2jRSUlIYNGiQT01tdwxff53/Snt3DO7rd6rPwukqXrw499xzD++//z4zZ86kbdu2ZGZm8vnnn1O+fHnatm3r03/06NE8+uijlC5dmtatW1O1alXCwsIwxjBlyhRWrlzJiRMnzjmuRYsW0aJFC7KysmjZsiUdO3YkPDycoKAgVqxYwdSpUwMeJzIyMuB47muenZ19RnGIiIiIiIiIiBQ0JZ/Pl/NYyuJSEhERQWpqasDyBFlZWSQnJwd8GFtSUlLA8fbt2+cZ90zMmzePO+64g6CgIOLi4mjUqFHAfrVq1QKclcKBuFfn1qxZ85THdMfojjm3xMREn37ekpKSPLF4C3T+t9xyC9OmTePEiRMsXbqUGTNm8M4779CjRw/Kly9Pq1atPP3vvPNOvv3221PG7i2/Vd7uhO4HH3zArFmzuP322wOW3PCOeerUqXTs2PGUx3X3P9Vn4Uz06tWL999/n4kTJ9K2bVtPovzRRx/1+XxmZWURGxtLxYoVWbZsmd8fENyruAM5nVXx3l555RWOHz9OfHy83/8h8OqrrzJ16tQzGk9EREREREREpLAIOnUXkby5SxfktcryhhtuICcnh7lz5/ptmzt3LtnZ2dSvX99v286dO0lISPBrnzNnjmfc0zV79mzatm1LcHAws2bNyjPxDNC8eXMAZs6cSU5Ojs+2I0eOsGDBAooXL57vGG7uGN0xe8vKymL+/PkAAc//119/9Wvbtm0bu3btIjo6OuCq19DQUBo3bsxLL73E6NGjATyJy9q1axMZGcmiRYvIzMw8Zexnwl2uYuLEiSQnJzN9+nSuv/566tWr59PPfc3mzZt3WuPWrl2bsLAwVqxYEXAleqDreipNmjThqquuYurUqRw6dCjPRHlycjJpaWk0btzYL/F89OhRT0mVgrBlyxbKlCkTsDRNoM/B2TjVfSoiIiIiIiIicj4o+SznpHTp0hhj2LlzZ8Dtffr0AeDZZ58lPT3d056ens4zzzwDwAMPPOC3X3Z2Nk8//bRPAnj79u2MHj2a4OBg/v73v59WfDNnzqR9+/YUK1aMX375hZtuuinf/jVq1KBNmzYkJCTw7rvv+myLjY3l2LFj3HfffZQoUeKUx+7cuTNlypTh888/Z9GiRT7b3n77bbZt20arVq386j0DjBo1ih07dnh+z8nJYfDgweTk5HD//fd72ufNmxcwMeteLRwWFgY4pRgGDBhAYmIiAwcO5Pjx4377JCYmsm7dulOeV27eCd1x48aRmZkZsH5yp06dqFGjBu+++y4//fRTwLEWLlzo+ZwULVqUnj17cuTIEb8a0UuWLGHSpElnHCs4ieaMjAzGjh3LTz/9xPXXX+/3x4yoqCjCwsJYunSpTymSzMxMHn30UZKTk8/q2IFER0eTmprKqlWrfNr/+9//EhcXVyDHONV9KiIiIiIiIiJyPqjshpyTkiVL0rBhQ+bNm0fPnj2pWbMmRYoUoWPHjlx//fX06NGDqVOn8tVXX1GnTh06d+7sqZm7fft27rnnHnr27Ok37vXXX8/vv/9OgwYNaNOmDYcOHeLLL78kLS2N1157ze+BeoFs3LiRTp06kZGRQbt27Zg6dWrAEga5E5tjx46lcePGDBw4kF9++YWrr76a33//nfj4eGrWrMmwYcNO+9pMmDCBu+++m2bNmnH33XdTtWpVli5dysyZM6lYsSLvvfdewH2bNGlCvXr16NatGxEREcTFxbFy5UoaNGjAU0895en3xhtvMHPmTGJiYrjiiisoWbIka9euZfr06ZQuXZq+fft6+r7wwgusXLmS8ePH88MPP9CiRQsqV67M/v372bx5MwsWLGDYsGFcc801p3V+3u677z5eeOEFXn75ZYKDg+nRo4dfn6JFi/Ltt99y2223cccdd9C4cWPq1atHWFgYu3btYvHixWzbto3ExERP0vz//u//+OWXX3j77bdZsmQJTZs2JTExkS+//JJ27drx/fffn1WsL774IrGxsWRmZvqtegYICgpi4MCBDB8+nOuuu45OnTpx8uRJ4uPjSU1NpXnz5sTHx5/xsQMZNGgQcXFxNG3alHvuuYeIiAiWLFnC/Pnz6dq1K5MnTz7nY5zqPhUREREREREROS+stXp5vcaOHWvHjh1r5fRt3rzZtm/f3pYpU8YaYyxgP/zwQ8/27Oxs++6779oGDRrY4sWL2+LFi9v69evbMWPG2OzsbL/xANusWTO7Z88e27NnT1u+fHkbGhpqb7jhBjtp0qTTjis+Pt4Cp3wFsnPnTtu7d29bsWJFW7RoUVu1alU7cOBAm5KScsbX548//rCdO3e25cqVs0WLFrVVqlSxDz/8sN2zZ49f3169elnAbt261b7++uu2Vq1aNjQ01FaqVMk++uij9tChQz794+LibO/eve3VV19tw8PDbVhYmK1Zs6YdMGCATUhI8Bs/JyfHfvzxx7ZFixa2dOnStmjRorZSpUq2SZMmdtiwYXbnzp2evtu3b7eA7dWr1ynPcceOHTYoKMgCtn379vn2TUpKsk8//bStU6eO/X/27j087uo+8P/7SLJly7Zs624pNpA2JEQUpy2LS9LFxAbSlELIzck2aUhCwq+E35K2T9nuswEyBnbbLk+TNm3iDZSQNOWX4E1TAt3SJmuISdMEyzfJlmV8t2TLV9lYvsm2pPP7QxpldPdIo/v79TzzzMz3e87MZ/Tx9yA+c3TOzJkz46xZs+Iv//Ivxw9+8IPx29/+drx06VK39ocOHYqf+tSnYlFRUZwxY0ZcvHhxfOaZZ7ry+8UvfnHQ+Hpavnx5BGJOTk48fPhwn20uXboU/+Iv/iJec801ccaMGbG0tDR+/OMfj/v27evK0969e7vaX87Pa+nSpX3+m3vxxRfjkiVL4uzZs+PcuXPjrbfeGteuXRufeeaZXtdTjDFeccUV8YorrujzPb74xS9GIL7yyivdjg92nUqSJEkaO/+z7lLXTZJGwJjVWkOMcTRr3ePeqlWrIsB999031qFMWSEEli5dOqQ1fSVJklw9rVIAACAASURBVCRJmmie2N7a9fjBt/lH6pIyLozVG7vmsyRJkiRJkiQp4yw+S5IkSZIkSZIyzuKzJEmSJEmSJCnjXEhI447rkEuSJEmSJEkTnzOfJUmSJEmSJEkZZ/FZkiRJkiRJkpRxFp8lSZIkSZIkSRln8VmSJEmSJEmSlHEWnyVJkiRJkiRJGWfxWZIkSZIkSZKUcRafJUmSJEmSJEkZZ/FZkiRJkiRJkpRxFp8lSZIkSZIkSRln8VmSJEmSJEmSlHEWnyVJkiRJkiRJGWfxWZIkSZIkSZKUcTljHYAkSZIkSdJUdeBc5OiFSHuErNDx/E15YazDkqSMsPgsSZIkSZI0yraeauf7B9rZfjp2O/5falp525zAB96UxbVz/YN1SRObo5gkSZIkSdIoeuVoO39a19ar8Jy0/XTkT+va+PHR9lGOTJIyy+KzJEmSJEnSKNl6qp2/3dNG32XnX4jAU3va2HrKArSkicvisyRJkiRJ0ij5/oH2QQvPSRH4xwMWnyVNXBafJUmSJEmSRsGBc7HfpTb6U3c6cuBcen0kabyw+CxJkiRJkjQKhrqEhktvSJqoLD5LkiRJkiSNgvNto9tPksaaxWdJkiRJkqRRMDN7dPtJ0liz+CxJkiRJkjQKrp07tDLMUPtJ0lhz9JIkSZIkSRoFb8oLvG1OSKvPNXMCb8pLr48kjRcWnyVJkiRJkkbJB96UxeWWkgPw/jdZupE0cTmCSZIkSZIkjZJr52bxmTdnD1qADsBn35ztkhuSJrScsQ5AkiRJkiRpKnl3SRbFufCPB9qpOx17nb9mTuD9b8qy8CxpwrP4LEmSJEmSNMqundtRXD5wLvJXO1tpj5AV4PNvyXGNZ0mTxpCKzyGEWUAFUAScB47GGA9mMjBJkiRJkqTJ7k15gZLc0O25JE0Wl118DiH8EvBp4FbgV+mxXnQIoQl4BfgH4PsxxtYMxilJkiRJkiRJmkAGLT6HEK4HHgduoaPgfAnYAhwGTgAzgULgrcCHgQ8Bx0IIfwV8KcZ4YWRClyRJkiRJkiSNVwMWn0MI3wI+BpwCngS+C6yLMbb00/5K4D3A3cB/B34/hPB7McZXMxizJEmSJEmSJGmcG2zb1NuAPwLKY4yfizG+2l/hGSDGuC/G+PUY4zuBdwAbgaWZC1eSJEmSJEmSNBEMtuzGm2OM54fywjHGGuD9IYSZQ+kfQigE3g/cDvwKHRscXqRjyY9ngGdijO199Hsn8BDwG8AMYBfwDeCvY4xtQ4lFkiRJkiRJkpSeAYvPQy08Z+g1PgysAg7RsZFhPVAKfAD4W+C9IYQPxxhjskMI4X10bHjYAjxHx5rUdwBfBt7V+ZqSJEmSJEmSpBE26IaDlyOEMA24FjgXY3w9E68J7ADuBP5P6gznEMJ/A9YBH6SjEP0PncfzgaeANuDmGOP6zuMPAy8DHwohfDTG+N0MxSdJkiRJkiRJ6sdgaz53E0JYEUJYHUIoSDn2S0AtsB7YFkL4fghh2EXtGOPLMcYXey6tEWM8DPyvzqc3p5z6EFAMfDdZeO5s30LHMhwA9w03LkmSJEmSJEnS4NIqPgOfBt4WYzyRcuwvgF+mY2mMGuB9wKcyE16/LnXet6YcW9Z5/y99tH8VOAe8M4SQO5KBSZIkSZIkSZLSLz6/HahKPulc6uK3gdUxxluAG4DtjGDxuXNW9Sc6n6YWmt/aeb+jZ58YYyuwl45lRt48UrFJkiRJkiRJkjqkW3wupmMDwKQb6SjofhcgxngJ+BHwSxmJrm9/Rsf60v8cY/zXlONzO+9P9dMveXzeSAUmSZIkSZIkSeqQnUgkLrvxypUrPw/sSiQSL3U+vwd4F/BAIpE413nsZuCmRCLx3zMdbAjhAeBROmZX/04ikTifEtt/BgqBv04kEid69l25cuVngIXA3yYSiQN9vPa9K1eu/Prtt99eDnD11VdTV1dHUVERmzZtory8nPXr11NYWEhtbS25ubkcPHiQc+fOcfbsWRobG5k+fTrbtm2juLiYjRs3dutTV1fHtGnTaGxs5OzZs5w/f54DBw6Qm5vL1q1bKS0t7dZn/vz5vP7662RnZ3P48GFOnz5NS0sLDQ0NzJw5k5qaGsrKytiwYUNXn3nz5rFjxw5CCBw9epTm5mYuXrzI/v37ycvLo6amhvLycqqqqqioqGD9+vXMnTuXXbt2EWPk+PHjnDx5ktbWVvbt28fs2bOprq7uev3k++Xn57N7927a29s5ceIETU1NtLe3s2fPHubMmdOtT2lpKZs2bWLOnDns3buX1tZWTp48yfHjxwHYtWsX8+bNY/PmzV19SkpKqK6uZtasWezbt49Lly5x6tQpjh49SgiBHTt2MH/+/G59iouLqampIS8vj/r6ei5cuEBzczNHjhwhOzub7du3U1hYyMaNG7s+e1FREVu3biU3N5cDBw7Q0tLCmTNnOHToEDk5OV35T/apqqqiqKiIbdu2deWyZ/5ra2spKSnplsuCggLq6urIycnpymXP/PfMZTL/WVlZHDlyhObmZi5cuEB9fX1X/hcsWNBv/o8dO8Ybb7zBpUuXeuV//fr1Xff5+fm98t/W1saePXuYPXs2mzdv7vp5lZWVsXHjxgHzn5+f35WXqqoqysrK2Lx5M7Nmzeoz/zt37uw3/3l5eezfv5+LFy/2yn9BQUG367K4uJgtW7YwY8YM6uvraWlp4fTp0xw+fLhb/lP7pOY/eS2fOXOGxsZGpk2bRl1dXZ/Xcmr+U6/l1Pxv2LCh6+dWUFDA9u3b+72Wt2zZ0i3/VVVVfeY/eS0n898zl8lrGeiW/76u5eS/nWT+29vbaWpq6pb/ntdyMv+p13Jq/nfv3s3cuXO75bK0tLQr/6nX8rFjx7rlf9OmTV3XWElJSb/XclZWVq/8J/sk89/Q0NDvtexY7ljuWO5Y7ljuWO5Y7ljuWO5YPl7H8s3bd0FsJ+tME2VtbziWO5Y7ljuWZ3QsX7Bgwcq0irAZFGKMl984hJ8CpUAlEOnYaPBMjPFXU9p8B7gxxnhlRgMN4X7gb4BtwPLOjQdTz1cB1wPXxxg39NF/a2fcb48x1vX3PqtWrYoA993n3oSSJEmSJGnkPbH9F1taPfi2nDGMRNIkFcbqjdNdduNJOtZM3gnUdT7+Ro82S+goSmdMCOEP6Cg8bwXe3bPw3On1zvur++ifA1xFxwaFezIZmyRJkiRJkiSpt7SKzzHGb9Gx5nIeHWss/03nDYAQwjLgSuCVTAUYQvgT4MvAZjoKz0f7afpy5/1v9XHuJjpi/vcY44VMxSZJkiRJkiRJ6lu6M5+JMf63GGNR5+3zsfu6Hf8GzAf+MhPBhRAepqPYvYGOpTaOD9D8e8Bx4KMhhOtTXmMG8Hjn01WZiEuSJEmSJEmSNLCMLiQUY7wIXMzEa4UQ7qZjc8E24CfAAyH0Wp5kX4zxm53v3RxC+CwdRegfhxC+C5wA7gTe2nn8uUzEJkmSJEmSJEka2IDF5xDCzBjj+eG8wTBe46rO+2zgD/ppsxb4ZvJJjPH5EMJS4AvAB4EZwC7gj4CvxHR2V5QkSZIkSZIkDdlgy27sDSF8PoSQm+4LhxAWhxB+APzxUAKLMSZijGGQ28199PtpjPG3Y4zzY4wzY4y/EmP8coyxbShxSJIkSZIkSZLSN1jx+YfAl4BDIYRVIYR3hxBm9tc4hPDmEMJ9IYSfARuBxWRw80FJkiRJkiRJ0sQw4LIbMcZPhBC+AvwP4N7OW1sIoQ44BJykY2mLQjrWVS4CAnCEjqUvvhxjvDBy4UuSJEmSJEmSxqNBNxyMMa4HbgshvAW4B1gOvAP4lR5NjwHfB/4B+IcY46UMxypJkiRJkiRJmiAGLT4nxRh3Av8VIISQB1TQMeP5PHA0xnhoRCKUJEmSJEmSJE04l118ThVjPAfs7LxJkiRJkiRJktTNYBsOSpIkSZIkSZKUNovPkiRJkiRJkqSMs/gsSZIkSZIkSco4i8+SJEmSJEmSpIyz+CxJkiRJkiRJyjiLz5IkSZIkSZKkjEur+BxCWBRCyB+kzZwQwqLhhSVJkiRJkiRJmsjSnfm8F/j8IG0e6GwnSZIkSZIkSZqi0i0+h86bJEmSJEmSJEn9Gok1n0uBsyPwupIkSZIkSZKkCSJnsAYhhE/0OPSOPo4BZAOLgN8DtmQgNkmSJEmSJEnSBDVo8Rn4JhA7H0fgfZ23npLLcZwDVg47MkmSJEmSJEnShHU5xedPdd4H4BvA88AP+mjXBjQBP4sxvpGZ8CRJkiRJkiRJE9GgxecY47eSj0MIdwPPxxj/bkSjkiRJkiRJkiRNaJcz87lLjPHdIxWIJEmSJEmSJGnyyEqncQjh10IIj4QQSvs5X9Z5/h2ZCU+SJEmSJEmSNBGlVXwG/hj4DHC0n/NHgHuAPxpOUJIkSZIkSZKkiS3d4vONwCsxxtjXyc7jLwPvGm5gkiRJkiRJkqSJK93icxlwYJA2jcCCoYUjSZIkSZIkSZoM0i0+nwOKB2lTDFwYWjiSJEmSJEmSpMkg3eLzZuB9IYTZfZ0MIeQD7+tsJ0mSJEmSJEmaotItPj9Jx8zmH4UQrks9EUJYDPwQKOpsJ0mSJEmSJEmaonLSaRxjfC6E8F7gE8CmEMIR4CBQAZQCAfhWjPE7GY9UkiRJkiRJkjRhpDvzmRjjJ4HfB7bRsQHhr3fe1wL3xhg/lckAJUmSJEmSJEkTT1ozn5NijE8CT4YQ8oB5wBsxxnMZjUySJEmSJEmSNGENqfic1FlwtugsSZIkSZIkSepmSMXnEEIx8EHgGmBWjPEzKcevArbEGM9nLEpJkiRJkiRJ0oSSdvE5hHAP8BVgBh0bDEbgM52nS4GfAfcCT2coRkmSJEmSJEnSBJPWhoMhhFuBJ4EdwPuBVannY4xb6dh48K5MBShJkiRJkiRJmnjSnfn8J8AhYGmMsTmE8Kt9tKkBbhx2ZJIkSZIkSZKkCSutmc/A9cA/xRibB2hzACgbekiSJEmSJEmSpIku3eLzdODsIG3mAW1DC0eSJEmSJEmSNBmkW3zeB/z6IG2WAK8PKRpJkiRJkiRJ0qSQ7prPPwD+SwjhwzHG/93zZAjhU8B1wBcyEZwkSZIkSdJk9MT21q7Hu8/EXscffFu6JRtJGn/SHcn+J/BR4DshhA8BcwFCCP8v8B+BDwA7gb/OZJCSJEmSJEmTVcH0sY5AkkZGWsXnGOPJEMJS4O+AD6ec+krn/U+A340xDrYutCRJkiRJ0pTlzGZJU0HaI12MsR64OYRwHXAjUAicAn4eY9yQ4fgkSZIkSZIkSRPQgMXnEML3ge/GGFd3Pr8J2BdjrI8x1gA1oxCjJEmSJEmSJGmCyRrk/F3A21KevwJ8csSikSRJkiRJkiRNCoMVn98A8lOehxGMRZIkSZIkSZI0SQy25nMd8J9CCFXAoc5jV3YuvzGgGOOrww1OkiRJ0tRyx+MvDHj+xYfuHKVIJEmSNFyDFZ9XAs8D/1/Ksbs7b4PJHmpQkiRJknSw6WzX44rCWWMYiSRJkoZiwOJzjPGHIYRrgFuACiABrO28SZIkSVJGpc5sTp0F7YxnSZKkiWewmc/EGPcDTwOEEBLAj2OMj45wXJIkSZIkSZKkCWzADQdDCF8KIdyWcuhTdCzDIUmSJEmSJElSvwYsPgN/APxGyvNvAHeNXDiSJEmSJEmSpMlgsOLzGSAv5XkYwVgkSZIkSZIkSZPEYGs+7wI+EEL4R+BQ57F5IYRFg71wjLF+uMFJkiRJkiRJkiamwYrPTwB/D/x7yrHPd94GEi/jtQcVQvgQsBR4B7AYmAM8G2P8eB9trwT2DvByz8UYPzrcmCRJkiRJkiRJgxuwQBxj/E4IYS9wO1ABfBKoATaPfGgAPERH0fkMcAB422X0qabvTRG3ZjAuSZIkSZIkSdIABp2dHGP8OfBzgBDCJ4F/jDE+OsJxJf0hHUXnXXTMgH7lMvpsjjEmRjIoSZIkSZIkSdLA0l0a41PAppEIpC8xxq5icwjudShJkiRJkiRJE0VaxecY47dGKpAMKg8h/D9AIdAE/CzGWDPGMUnSlHDH4y8MeP7Fh+4cpUgkSZIkSdJYS3tTwBBCFnA/8DHgGmBWjDGn89yvAp8F/jLGuCOTgabh1s5blxDCj4G7Y4z1YxKRJE1BB5vOdj2uKJw1hpFIkiRJkqSxkJ1IJC67cQhhOvCvwO8DM4FzwJxEIvEowMqVK1uAZ4HmRCLx40wGunLlyivp2PBwSyKR+H4f56cDLcAfAf8V+HPgJeBK4N3AHStXrvxGIpG41NfrhxDuXbly5ddvv/32coCrr76auro6ioqK2LRpE+Xl5axfv57CwkJqa2vJzc3l4MGDnDt3jrNnz9LY2Mj06dPZtm0bxcXFbNy4sVufuro6pk2bRmNjI2fPnuX8+fMcOHCA3Nxctm7dSmlpabc+8+fP5/XXXyc7O5vDhw9z+vRpWlpaaGhoYObMmdTU1FBWVsaGDRu6+sybN48dO3YQQuDo0aM0Nzdz8eJF9u/fT15eHjU1NZSXl1NVVUVFRQXr169n7ty57Nq1ixgjx48f5+TJk7S2trJv3z5mz55NdXV11+sn3y8/P5/du3fT3t7OiRMnaGpqor29nT179jBnzpxufUpLS9m0aRNz5sxh7969tLa2cvLkSY4fPw7Arl27mDdvHps3b+7qU1JSQnV1NbNmzWLfvn1cunSJU6dOcfToUUII7Nixg/nz53frU1xcTE1NDXl5edTX13PhwgWam5s5cuQI2dnZbN++ncLCQjZu3Nj12YuKiti6dSu5ubkcOHCAlpYWzpw5w6FDh8jJyenKf7JPVVUVRUVFbNu2rSuXPfNfW1tLSUlJt1wWFBRQV1dHTk5OVy575r9nLpP5z8rK4siRIzQ3N3PhwgXq6+u78r9gwYJ+83/s2DHeeOMNLl261Cv/69ev77rPz8/vlf+2tjb27NnD7Nmz2bx5c9fPq6ysjI0bNw6Y//z8/K68VFVVUVZWxubNm5k1a1af+d+5c2e/+c/Ly2P//v1cvHixV/4LCgq6XZfFxcVs2bKFGTNmUF9fT0tLC6dPn+bw4cPd8p/aJzX/yWv5zJkzNDY2Mm3aNOrq6vq8llPzn3otp+Z/w4YNXT+3goICtm/f3u+1vGXLlm75r6qq6jP/yWs5mf+euUxey0C3/Pd1LSf/7STz397eTlNTU7f897yWk/lPvZZT8797927mzp3L2/PP88cffTdX551mw8ELxEvnmT9nJl/6yDV86MY3c+rUKY4dO9Yt/5s2beq6xkpKSvq9lrOysnrlP9knmf+GhoZ+r2XHcsdyx3LHcsfyyxvLU3NZWlralf/Ua3m0xvLn/m0nISuLM2fO8PGbr3Esdyx3LHcsdyyfgGO5v5c7ljuWj/1YvmDBgpWXW4PNtBBjvPzGIXwBeAxIAI8DjwAPxxizU9r8EMiPMf5GRgMN4WY6Nhx8Nsb48TT65QD/BiwB/iDG+FcDtV+1alUEuO+++4YerCSp2xIcLrchSRoK/1siSZKUEWO2mV5Wmu0/Bvw0xvhojLEd6KtyvRdYNOzIMiTG2Ar8befTm8YyFkmSJEmSJEmaKtItPl8F/HyQNieAgqGFM2KOdd676KgkSZIkSZIkjYJ0Nxw8D8wbpM0i4I2hhTNikkuA7BnTKCRJkiRdltr6JvYeaaa1rZ2c7Cxq65uoXFQ41mFJkiQpDekWnzcDt4UQpscYL/Y8GUKYC7wH+PdMBJeOEMISYFPPuEIIy4A/7Hz696MdlyRJkqTLt6a6gUdXv8artY3djl/7wLPcVFnOIyuWsHzxwjGKTpIkSelIt/j8FPAs8GwI4Z7UEyGEecAzwHzgf2UiuBDCXcBdnU/LOu9vDCF8s/Px8RjjH3c+/nOgMoTwY+BA57HrgGWdjx+OMY56UVySJEnS5Xn6R7Xc+7WXae9nU/RXaxu5LfE8T92/jE/fUjnK0UmSJCldaRWfY4zfCSHcAnwKuBM4CRBCWA9UArnAV2OM/5yh+N4B3N3j2Js7bwD7gWTx+dvA+4H/ALwXmAYcAVYDfxNj/EmGYpIkSZKUYWuqGwYsPCe1x8hnv/oyVxTnOwNakiRpnEt3w0FijPcAnwa2AcVAAH4N2AXcE2P8z5kKLsaYiDGGAW5XprR9Osb4OzHGK2OMs2OMuTHGRTHGj1h4liRJksa3R1e/NmjhOak9Rh5bvW6EI5IkSdJwpbvsBgAxxm8C3wwhzKRjmY1TMcazmQxMkiRJ0tRQW9/Ua43nwaytPegmhJIkSeNc2jOfU8UYz8cYGy08S5IkSRqqNTUNo9pPkiRJo2NIM58lSZImqjsef2HA8y8+dOcoRSIpqfncxVHtJ0mSpNFh8VmSJE1ZB5t+8cdbFYWzxjASaWrLz5s+qv0kSZI0Oiw+S5Iyrra+ib1HmmltaycnO8s1OTWupM5sLr37qa7HyeJz6sxoZ0FLo2P5dQtHtZ8kSZJGh8VnSVLGrKlu4NHVr/XaNOraB57lpspyHlmxhOWLLRRofHIWtDR2KhcVclNleVqbDi6trPCLTUmSpHHO4rMkKSOe/lEt937tZdpj7PP8q7WN3JZ4nqfuX8anb6kc5eikvt3wltI+jzvjWRp9j6xYwm2J5/v970iqrBB4eMUNoxCVJEmShiNrrAOQJE18a6obBiw8J7XHyGe/+jJrqhtGKTJJ0kSxfPFCnvzcMrJCGLBdVgg8df8y/5JGkiRpAshI8TmEUBRCeH8I4T0hhOxMvKYkaeJ4dPVrlzVTDToK0I+tXjfCEUmSJqJ7bq3kh4m7WFpZ0ef5pZUV/DBxl39BI0mSNEGktexGCOE+4JPAe2OMJzqP/TrwL0BBZ7P1IYRlMcazfb+KJGkyqa1vSmuNToC1tQfdhFCS1KflixeyfPFCauub+MgTL3VtXvvcg+/1vxuSJEkTTLoznz8CxGThudMTwHzgGeCfgf8A/H5mwpMkjXdraoa2hMZQ+0mSpobKRYVcVZrPW8rncVVpvoVnSZKkCSjd4vNbgJrkkxBCEbAUeDrG+JkY4x1AFfC7mQtRkjSeNZ+7OKr9JEmSJEnSxJBu8bkQOJry/F2d9/+YcuwnwBXDCUqSNHHk500f1X6SJEmSJGliSLf4fAIoSnm+FGgH/j3lWARmDDMuSdIEsfy6haPaT5IkSZIkTQzpFp/rgDtCCIUhhHl0rAFdFWNsTmlzJXA4Q/FJksa5ykWF3FRZnlafpZUVrt0pSZIkSdIkl27x+a+ABcABoAEoA76WPBlCyAZ+E6jOVICSpPHvkRVLyArhstpmhcDDK24Y4Yik9Jw+f5G9R5rZ2fgGe480U1vfNNYhSZIkSdKEl5NO4xjjCyGE3wfu7Tz0bIzx71Oa3ELHkhv/mqH4JEkTwPLFC3nyc8u492sv0x5jv+2yQuCp+5exfLFLbmjs1dY3sWV/E0dPneP8xbZu56594FluqiznkRVL/PcqSZIkSUOUVvEZIMb4JPBkP+f+FZg/3KAkSRPPPbdWcmVJPo+tXsfa2oO9zi+trODhFTdYyNOYW1PdwKOrX+PV2sYB271a28htied56v5lfPqWylGKTpIkSZImj7SLz5Ik9Wf54oUsX7yQ2vomPvLES7S2tZOTncVzD77XNZ41Ljz9o9pBZ+inao+Rz371Za4ozveLE0mSJElKU1prPocQfjWE8LkQwtyUY7NCCN8KIbwRQmgMIXw+82FKkiaSykWFXFWaz1vK53FVab6FZ40La6ob0io8J7XHyGOr141QVJIkSZI0eaW74eCfAF+IMZ5KOfanwO91vlYh8KUQwm0Zik+SJCkjHl39WtqF56S1tQfdhFCSJEmS0pTushvXAz9OPgkhTAPuBtYBNwMFwCbgAeCHGYlQkiRpmGrrmwZd43kwa2oanMU/Su54/IUBz7/40J2jFElmTLbPI0mSJF2udIvPJUBDyvPrgTnA12OMLUBjCOEHwG9lKD5JkqRhW1PTMHijQTSfu5iBSJSug01nux5XFM4aw0gyYzQ/j0VvSZIkjbV0i8+xR5/f7Dy2NuXYMaB4mHFJkiRlTCYKx/l50zMQiS5HalE0tYA6UYul4+HzTLYiviRJkiaGdIvP9cBvpDx/H3Agxrgn5Vg5cHK4gUmSJGVKJgrHy69bmIFIpNEzHorekiRJmtrS3XBwNfDOEML3Qgh/D9wIfK9Hm2uB3ZkITpIkKROGWzheWlnhes+SJEmSlKZ0i89fBn4GfAD4XaAaeDR5MoTwduDX6b4MhyRJ0piqXFTITZXlQ+qbFQIPr7ghwxFJkiRJ0uSXVvE5xngmxvgu4LrO2/UxxlMpTc4B7wdWZS5ESZKk4XtkxRKyQkirT1YIPHX/MpYvdskNSZIkSUpXujOfAYgxbu28tfc4vi/G+IMY48HMhCdJkpQZyxcv5MnPLbvsAvTSygp+mLiLT99SOcKRSZIkSdLklO6GgwCEEKYBy4FrgNkxxsc6j88A8oHjPQvTkiRJY+2eWyu5siSfx1avY21t7+/KZ07PpmRuHv/n4Ttd41mSJEmShint4nMI4beAp4EyIAAReKzz9DuAnwIfB76ToRglSZIyZvnihSxfvJDa+iY+8sRLtLa1k5OdRVH+DObMnA5g4XmcqK1vYu+R5q4c1dY3TejcTLbPI0mSJA0mreJzCOF64HngOPCHwA3Af0qejzH+PISwl451ny0+S5KkcatyUSFXleaPdRjqw5rqBh5d/Rqv1jZ2O37tA89yU2U5j6xYMqHW4R7rz2PRW5IkSWMl3TWfH6ZjU8HrY4xfAXb20aYKWDzcwCRJkjT1PP2jWm5LPN+rUJv0am0jtyWe5xv/t3aUIxuasfw8a6obWPqF73HtA89S23CC1xvfoLbhBNc+8CxLv/A91lQ3ZPw9JUmSpFTpFp/fBTwfYzw8QJsGYMHQQ5IkSdJUNhFtjgAAIABJREFUtKa6gXu/9jLtMQ7Yrj1GPvvVl8d98XQsP89kK+JLkiRpYkq3+DybjiU3BpI3hNeVJEnSFPfo6tcGLdQmtcfIY6vXjXBEwzNWn2eyFfElSZI0caVbJD4IVA7S5h3AnqGFI0mSpKmotr6p31m6/Vlbe5Da+qYRimh4xvLzTLYiviRJkiaudIvPLwHvCSH8Zl8nQwjvBd4J/NNwA5MkSdLUsaZmaLNvh9pvpI3V55lsRXxJkiRNbOkWn/8UeAP4YQjhz4G3A4QQbu98/r+BQ8CXMhqlJEmSJrXmcxdHtd9IG6vPM9mK+JIkSZrYctJpHGM8GEK4DVgNPJhy6gUgALuBD8QYB1sXWpIkSeqSnzd9VPuNtLH6PJOtiC+Nhjsef2HA8y8+dOcoRSJJ0uSTVvEZIMa4MYTwVuB24EagEDgF/Bz4QYyxNbMhSpIkabJbft3CUe030sbq80yGIn5qIXDdziO9jlsI1Eg62HS263FF4awxjESSpMkh7eIzQIyxjY7ZzgN/RSxJkiRdhspFhdxUWZ7WesVLKyuoXFQ4glEN3Vh9nslWxK8omD3WIWgKSP1CI/XLD7/okCRp+IZUfJYkSZIy7ZEVS7gt8TztMQ7aNisEHl5xwyhENXRj8XkmQxHfgp8kSdLkMaTicwjheuAGYD6Q3UeTGGN8bDiBSZI0HK7fKE08yxcv5MnPLePer708YME2KwSeun8ZyxePz9m6SWP1eSZbEV+SJEkTV1rF5xBCPvB94N10bDDYnwhYfJYkjQuu3yhNHPfcWsmVJfk8tnoda2sP9jq/tLKCh1fcMO4Lz0lj8XkmWxFfkiRJE1e6M5+fAJYBPwGeARoANxiUJAHja5Mo12+UJq7lixeyfPFCauub+MgTL9Ha1k5OdhbPPfjecbU8xOUai88z2Yr4kiRJmpjSLT6/D9gIvDvG2D4C8UiSJgk3iZI0XJWLCrmqNL/b84lstD/PZCviSyOttr6JvUeau66V2vomrxVJkoYp3eLzXODbFp4lSX1xVrEkjT+TrYgvZdqa6gYeXf1ar406r33gWW6qLOeRFUv8KwFJkoYoK832O4HSkQhEkiRJkqTR9PSParkt8XyvwnPSq7WN3JZ4nm/839pRjkySpMkh3ZnPXwX+LIRQEWPsvXicJEnSBJK6NnmqsVinXJI0utZUNwy6MSdAe4x89qsvc0VxvjOgJUlKU7rF55fo2HDwpyGElcAG4I2+GsYY64cZmyRJ0qhxnXJJmloeXf3aoIXnpPYYeWz1OovPkiSlKd3i8z4gAgH42wHaxSG8tiRJGefmQRrIDW/5xWpiznKWpKmjtr6p36U2+rO29qC/R0iSlKZ0C8R/R0dhWZKkcc3NgyRJUn/W1DQMuZ/FZ0mSLl9axecY4ydHKI5eQggfApYC7wAWA3OAZ2OMHx+gzzuBh4DfAGYAu4BvAH8dY2wb8aAlSePC0z+qHXANx+TmQU/dv4xP31I5ytFptCXXb+7Lup1Hus1+liRNDc3nLo5qP0mSpqrxvDTGQ3QUnc8AB4C3DdQ4hPA+4B+AFuA54ARwB/Bl4F3Ah0cyWEnS+ODmQRrIwaazXY8rCmeNYSSSpLGUnzd9VPtJkjRVZaXTOITQFkJ4eJA2XwghtA4vLAD+ELgayAfuG+Q984GngDbg5hjjPTHGB+mYNf0z4EMhhI9mICZJ0jg3lM2DNLm9+NCdXbeKwlldtxcfutNZz5I0RS2/bmhfPA+1nyRJU1VaxWc6NhoMl9luWGKMr8QYd8Z4WRWEDwHFwHdjjOtTXqOFjhnUMEgBW5I08Q1n8yBJkjR1VC4q5KbK8rT6LK2scL1nSZLSlG7x+XLMp2Ppi9G0rPP+X/o49ypwDnhnCCF39EKSJI224WweJEmSppZHViwhK1zevKmsEHh4xQ0jHJEkSZPPoGs+hxBu6nHoyj6OAWQDi4CPAa9nILZ0vLXzfkfPEzHG1hDCXqASeDNQN5qBSZJGj5sHSZKky7V88UKe/NyyQfeKyAqBp+5f5h4RkiQNQXYikRiwwcqVK/cCdwOf7Dz0js7HPW+fAO4C8oAHE4nElkwFuXLlyis732NLIpH4fh/nHwQKgScSicSJPs5/DFgIPJtIJA709R4hhHtXrlz59dtvv70c4Oqrr6auro6ioiI2bdpEeXk569evp7CwkNraWnJzczl48CDnzp3j7NmzNDY2Mn36dLZt20ZxcTEbN27s1qeuro5p06bR2NjI2bNnOX/+PAcOHCA3N5etW7dSWlrarc/8+fN5/fXXyc7O5vDhw5w+fZqWlhYaGhqYOXMmNTU1lJWVsWHDhq4+8+bNY8eOHYQQOHr0KM3NzVy8eJH9+/eTl5dHTU0N5eXlVFVVUVFRwfr165k7dy67du0ixsjx48c5efIkra2t7Nu3j9mzZ1NdXd31+sn3y8/PZ/fu3bS3t3PixAmamppob29nz549zJkzp1uf0tJSNm3axJw5c9i7dy+tra2cPHmS48ePA7Br1y7mzZvH5s2bu/qUlJRQXV3NrFmz2LdvH5cuXeLUqVMcPXqUEAI7duxg/vz53foUFxdTU1NDXl4e9fX1XLhwgebmZo4cOUJ2djbbt2+nsLCQjRs3dn32oqIitm7dSm5uLgcOHKClpYUzZ85w6NAhcnJyuvKf7FNVVUVRURHbtm3rymXP/NfW1lJSUtItlwUFBdTV1ZGTk9OVy57575nLZP6zsrI4cuQIzc3NXLhwgfr6+q78L1iwoN/8Hzt2jDfeeINLly71yv/69eu77vPz83vlv62tjT179jB79mw2b97c9fMqKytj48aNA+Y/Pz+/Ky9VVVWUlZWxefNmZs2a1Wf+d+7c2W/+8/Ly2L9/PxcvXuyV/4KCgm7XZXFxMVu2bGHGjBnU19fT0tLC6dOnOXz4cLf8p/ZJzX/yWj5z5gyNjY1MmzaNurq6Pq/l1PynXsup+d+wYUPXz62goIDt27f3ey1v2bKlW/6rqqr6zH/yWk7mv2cuk9cy0C3/fV3LyX87yfy3t7fT1NTULf89r+Vk/lOv5dT87969m7lz53bLZWlpaVf+U6/lY8eOdcv/pk2buq6xkpKSfq/lrKysXvlP9knm/yc1e3j19eNp/3fm5rcWcHVxrmP5FBjLV72wjunTp3P69Bne9+tv4tm122lva6O1rY0lFTmO5eNwLH/up7sIIYuzZ8/w8ZuvmfBj+bNrO16n+fRpPn3b4lEby1e9uI7cGTNoPnWKD9345nE9ljc0NPR7Lft7uWN5pn8vv6Z8Dm8vm0HTuUj98TO9fkf4j9cs4KHf+SXed8NVjuX+Xu7v5Y7ljuXjdCz39/LBx/IFCxasTPt/lDMkDLakcgghAUQ61nF+BPgxsLaPpm1AE/BKjHF7RoMM4WbgFeDZGOPH+zi/A3gL8JYY464+zv87cCNwY4zx5wO916pVqyLAffe5RLQkTTS19U1c+8Czaffb+pWPuYbjFHHH4y+wbucRAG54S2nX4+Rz6NigUOPHHY+/0PV4MuRmrD7PZPs5SiOhtr6JjzzxEq1t7eRkZ/Hcg+/19wNJ0mQx7P35hmrQZTdijInk4xDC3cDzMcavjGRQQ3Cq835uP+fze7STJE1Cyc2D0tl00M2DpraKgtljHYIkaZyoXFTIVaX53Z5LkqThGbT4nCrGeNVIBTJMrwPXA1cDG1JPhBBygKuAVmDP6IcmSepL6iy8vgx1Zt4jK5ZwW+L5AdduTHLzoKmltr6JvUeamT8rl5zsLP7sE++ysKBJLXWcTZ3lnzzuDGhJkiSNtLSKz+PYy3RsdPhbwHd6nLuJjnWoX40xXhjtwCRJgzvYdLbrcUXhrGG9lpsHqac11Q08uvq1XjPir33gWW6qLOeRFUv8dzCCRuqLJqXHWf6SJEkaCwMWn0MIL9Ox3vPdMcYDnc8vR4wxLh92dJfve8CfAx8NIfx1jHE9QAhhBvB4Z5tVoxiPJGkQyYJTz/UVMzEb9Z5bK7myJJ/HVq9jbe3BXueXVlbw8IobLDhOAU//qHbALyJerW3ktsTzPHX/Mj59S+UoRzf1ZPKLJg3Owr4kSZLG2oAbDoYQ2jsfvi3GuCPl+WBijDF7WIGFcBdwV+fTMuA9dCyb8ZPOY8djjH/co/33gBbgu8AJ4E7grZ3HV8TBdlfEDQclabT0NxsVyOhsVDcPmrrWVDektQTLDxN3+YXECErnWuxvuYiJuinkZPs80mTm5pySpElqfG44GGPMGuj5CHsHcHePY2/uvAHsB7qKzzHG50MIS4EvAB8EZgC7gD8CvnI5hWdJ0ugYzdmobh40dT26+rXLKjwDtMfIY6vXWXweAcNd9mSyLRcx2T6PJEmSNJDBlt14APh5jHHdKMXTJcaYABJp9vkp8NsjEY8kKTPWVDcMuh4zdBQDP/vVl7miON+CoNJWW9/U56z6gaytPUhtfZNfUGTQUL9ommyzDSfb55EkSZIu12Azmf+Sjk38AAghtIUQHh7ZkCRJk9lQZqNK6VpT0zCq/dRbul80ran2Zy9JkiRNNoMVn1uA3JTngTFcI0SSNLENZzaqlI7mcxdHtZ9684smSZIkSYMVn/cC7wkhlKYcc+1kSdKQOBtVoyU/b/qo9lN3ftEkSZIkCQYvPn8d+DWgMYTQ1nks0bn8xkC31pENW5I0ETkbVaNl+XVDWyd8qP3UnV80SZIkSYJBNhyMMX4lhHAUuB0oB94N1AP7Rj40SdJk42xUjZbKRYXcVFme1uzbpZUVbjaYIX7RJEmSJAkGKT4DxBi/C3wXIITQDjwTY3x0pAOTJE0+Q51V+v2f7eJHm+v7Pf/iQ3cONSRNYo+sWMJtiecva93hrBB4eMUNoxDV1OAXTZIkSZJg8GU3eloJ/HgE4pAkTQHJ2ajpWFpZwZyZvyhIHWw623WTBrJ88UKe/NwyssLAeyVnhcBT9y9j+WKX3MgUlz2RJEmSBJcx8zlVjHHlSAUiSZoahjIbNbUoeMfjL3Q9dsazBnPPrZVcWZLPY6vXsbb2YK/zSysrev0b0/C57IkkSZIkSLP4LEnScCVno977tZcHLEA7G1WZsnzxQpYvXkhtfRMfeeIlWtvaycnO4rkH32uxcwS57ImkiSL1i+11O4/0Ou6X3ZIkDZ3FZ0nSqBvqbNTa+ib2HmnuKh7W1jdZPNRlq1xUyFWl+d2ea+T4RZOkiaiiYPZYhyBJ0qRi8VmSNCbSmY26prqBR1e/1utP+K994FluqiznkRVLLFxJ45DLnkiaCJzZLEnSyLH4LEkaU4PNRn36R7UDzpx8tbaR2xLP89T9y/j0LZUjGquk9LnsiSRJkjR1WXyWJI1ba6obBv2TfYD2GPnsV1/miuJ8Z1BK45TLnkiSJElTj8VnSdKYuJzNfR5d/dplbVYGHQXox1avs/gsSZIkSdI4YfFZkjTm+trcp7a+qdcaz4NZW3vQTQglSZIkSRonLD5LksbEYJv7fOWfNg/pddfUNFh8lsaRy/krB0mSJEmTk8VnSdK41Hzu4qj2kzTy+vorB0mSJEmTl8VnSdK4lJ83fVT7SRoZzmyWJEmSpq6ssQ5AkqS+LL9uaBsHDrWfJEmSJEnKLGc+S5LGpcpFhdxUWZ7WpoNLKyu61nt2nVlJkiRJksaWM58lSePWIyuWkBXCZbXNCoGHV9zQ57mKgtldN0mSJEmSNDqc+SxJGreWL17Ik59bxr1fe5n2GPttlxUCT92/jOWLf7HkhjObJUmSJEkaWxafJUnj2j23VnJlST6PrV7H2tqDvc4vrazg4RU3dCs8S5IkSZKksWfxWZI07i1fvJDlixdSW9/ER554ida2dnKys3juwfd2rfEsSZIkSZLGF4vPkqQJo3JRIVeV5nd7LkmSJEmSxic3HJQkSZIkSZIkZZzFZ0mSJEmSJElSxll8liRJkiRJkiRlnMVnSZIkSZIkSVLGueGgJEma1O54/IWux+t2Hul1/MWH7hz1mCRJkiRpKrD4LEmSpoyKgtljHYIkSZIkTRkWnyVJ0qTmzGZJkiRJGhuu+SxJkiRJkiRJyjiLz5IkSZIkSZKkjLP4LEmSJEmSJEnKONd8liSNe3c8/kLX43U7j/Q67pq+kiRJkiSNPxafJUkTSkXB7LEOQZIkSZIkXQaLz5Kkcc+ZzZIkSZIkTTyu+SxJkiRJkiRJyjiLz5IkSZIkSZKkjLP4LEmSJEmSJEnKOIvPkiRJkiRJkqSMs/gsSZIkSZIkSco4i8+SJEmSJEmSpIyz+CxJkiRJkiRJyjiLz5IkSZIkSZKkjLP4LEmSJEmSJEnKOIvPkiRJkiRJkqSMs/gsSZIkSZIkSco4i8+SJEmSJEmSpIyz+CxJkiRJkiRJyjiLz5IkSZIkSZKkjJt0xecQwr4QQuzndnis45MkSZIkSZKkqSBnrAMYIaeAv+zj+JnRDkSSJEmSJEmSpqLJWnx+I8aYGOsgJEmSJEmSJGmqmnTLbkiSJEmSJEmSxt5knfmcG0L4OLAIOAvUAK/GGNvGNixJkiRJkiRJmhoma/G5DPh2j2N7QwifijGuHYuAJEmSJEmSJGkqmYzLbjwDLKejAD0L+BXg68CVwEshhMVjF5okSZIkSZIkTQ3ZiURirGPIqEQisTaRSOxNJBJnE4nEpUQicTSRSPzTypUr5wA3AWWJROK5nv1CCPeuXLny67fffns5wNVXX01dXR1FRUVs2rSJ8vJy1q9fT2FhIbW1teTm5nLw4EHOnTvH2bNnaWxsZPr06Wzbto3i4mI2btzYrU9dXR3Tpk2jsbGRs2fPcv78eQ4cOEBubi5bt26ltLS0W5/58+fz+uuvk52dzeHDhzl9+jQtLS00NDQwc+ZMampqKCsrY8OGDV195s2bx44dOwghcPToUZqbm7l48SL79+8nLy+PmpoaysvLqaqqoqKigvXr1zN37lx27dpFjJHjx49z8uRJWltb2bdvH7Nnz6a6urrr9ZPvl5+fz+7du2lvb+fEiRM0NTXR3t7Onj17mDNnTrc+paWlbNq0iTlz5rB3715aW1s5efIkx48fB2DXrl3MmzePzZs3d/UpKSmhurqaWbNmsW/fPi5dusSpU6c4evQoIQR27NjB/Pnzu/UpLi6mpqaGvLw86uvruXDhAs3NzRw5coTs7Gy2b99OYWEhGzdu7PrsRUVFbN26ldzcXA4cOEBLSwtnzpzh0KFD5OTkdOU/2aeqqoqioiK2bdvWlcue+a+traWkpKRbLgsKCqirqyMnJ6crlz3z3zOXyfxnZWVx5MgRmpubuXDhAvX19V35X7BgQb/5P3bsGG+88QaXLl3qlf/169d33efn5/fKf1tbG3v27GH27Nls3ry56+dVVlbGxo0bB8x/fn5+V16qqqooKytj8+bNzJo1q8/879y5s9/85+XlsX//fi5evNgr/wUFBd2uy+LiYrZs2cKMGTOor6+npaWF06dPc/jw4W75T+2Tmv/ktXzmzBkaGxuZNm0adXV1fV7LqflPvZZT879hw4aun1tBQQHbt2/v91resmVLt/xXVVX1mf/ktZzMf89cJq9loFv++7qWk/92kvlvb2+nqampW/57XsvJ/Kdey6n53717N3Pnzu2Wy9LS0q78p17Lx44d65b/TZs2dV1jJSUl/V7LWVlZvfKf7JPMf0NDQ7/XsmO5Y7ljuWO5Y7ljuWO5Y7ljuWO5Y7ljuWO5Y/lUHMsXLFiwMjOV1/SFGONYvfeoCiH8MrATOBFjLOyv3apVqyLAfffdN1qhSZIkSZIkSdJICWP1xpNx2Y3+HO28nzWmUUiSJEmSJEnSFDCVis83dt7vGdMoJEmS/v/27jtOrrre//jrk957QigJJUTQAIIgJSAEVK6CCPZGiQ27KOpPBZFw70XxCkguRVARBKyggnhFQSBEykXqDUUglJBQAiSQkEbq9/fH9yxMltlkZnfKsvt6Ph7zmN1T5nzPe87MnPOZM98jSZIkSd1Alyo+R8SkiBhRZviWwFnFv5c0tlWSJEmSJEmS1P30anYDauwDwLci4nrgMWAJMAE4GOgH/AU4tXnNkyRJkiRJkqTuoasVn68HtgN2IXezMRBYBNwIXAxcnLrLFRYlSZIkSZIkqYm6VPE5pXQDcEOz2yFJkiRJkiRJ3V2X6vNZkiRJkiRJktQ5WHyWJEmSJEmSJNWcxWdJkiRJkiRJUs1ZfJYkSZIkSZIk1ZzFZ0mSJEmSJElSzVl8liRJkiRJkiTVnMVnSZIkSZIkSVLNWXyWJEmSJEmSJNWcxWdJkiRJkiRJUs1ZfJYkSZIkSZIk1ZzFZ0mSJEmSJElSzVl8liRJkiRJkiTVnMVnSZIkSZIkSVLNWXyWJEmSJEmSJNWcxWdJkiRJkiRJUs1ZfJYkSZIkSZIk1ZzFZ0mSJEmSJElSzVl8liRJkiRJkiTVnMVnSZIkSZIkSVLNWXyWJEmSJEmSJNWcxWdJkiRJkiRJUs1ZfJYkSZIkSZIk1ZzFZ0mSJEmSJElSzVl8liRJkiRJkiTVnMVnSZIkSZIkSVLNWXyWJEmSJEmSJNWcxWdJkiRJkiRJUs1ZfJYkSZIkSZIk1ZzFZ0mSJEmSJElSzVl8liRJkiRJkiTVnMVnSZIkSZIkSVLNWXyWJEmSJEmSJNWcxWdJkiRJkiRJUs1ZfJYkSZIkSZIk1ZzFZ0mSJEmSJElSzVl8liRJkiRJkiTVnMVnSZIkSZIkSVLNWXyWJEmSJEmSJNWcxWdJkiRJkiRJUs1ZfJYkSZIkSZIk1ZzFZ0mSJEmSJElSzVl8liRJkiRJkiTVnMVnSZIkSZIkSVLNWXyWJEmSJEmSJNWcxWdJkiRJkiRJUs1ZfJYkSZIkSZIk1ZzFZ0mSJEmSJElSzVl8liRJkiRJkiTVnMVnSZIkSZIkSVLNWXyWJEmSJEmSJNWcxWdJkiRJkiRJUs1ZfJYkSZIkSZIk1ZzFZ0mSJEmSJElSzXW54nNEbBERP4+IpyJiZUTMiYgzImJ4s9smSZIkSZIkSd1Fr2Y3oJYiYgJwMzAGuAJ4ANgdOAZ4R0TsnVJa2MQmSpIkSZIkSVK30NXOfD6HXHj+ckrpsJTSt1JKBwA/ArYDTm5q6yRJkiRJkiSpm+gyxeeI2AY4EJgDnN1q9InAMuCIiBjY4KZJkiRJkiRJUrfTZYrPwAHF/dUppXWlI1JKS4CbgAHAno1umCRJkiRJkiR1N12pz+ftivuH2hg/m3xm9OuAazf2YD/+8Y9r1CxJkiRJkiRJapr0uc99Lpqx4K505vPQ4n5xG+Nbhg8rNzIijo6I22veKkmSJEmSJEnqhrrSmc8b01LdT+VGppR+Avykcc2RJEmSJEmSpK6rK5353HJm89A2xg9pNZ0kSZIkSZIkqU66UvH5weL+dW2Mn1jct9UntCRJkiRJkiSpRiKlsr1QvOZExATgYWAOMCGltK5k3GDgaXKxfXRKaVlTGilJkiRJkiRJ3USXOfM5pfQIcDWwFfCFVqNPAgYCF1l4liRJkiRJkqT66zJnPsPLZz/fDIwBrgD+BewB7E/ubmNySmlh81ooSZIkSZIkSd1Dlyo+A0TEOODfgXcAI8ndbVwOnJRSer6ZbZMkSZIkSZKk7qLLFZ8lSZIkSZIkSc3XZfp8liRJkiRJkiR1HhafJUmSJEmSJEk116vZDeiMevbskdata393JBFBvbsz6RHQgSZ2fPnAuuYtfj2NyHv95fUgpQaufQS8lrrHiR7QyHxqqSlZB9D457czvYalWmr256MkSZIkdSZ9+vRi1ao1C1JKo5uxfIvPZaxbl0i3t3/+2C0xc+bM2jWojH333Zd0Vl0XsUHxRUg7vb95DSgRsy7jkAc/2bDlXbnd+XzkllUNW96v9+rDrpfObtjyOuqOD0yE416jlZ/vBR/50P0NXeSvf/sGOHR6Q5cJsO6KY0h779zw5Ur1Fjfd3aHPcEmSJEnqSmK3NQCPN2v5drshSZIkSZIkSao5i8+SJEmSJEmSpJqz240KzbwTTr0Y7ngAnnoOLjgRph7S/se75JJLmDlzJnPnzqV3795MmjSJo48+mm222aZ2jS5x9g1w3k0w5/n8/6Sx8J13wME71GVxFfves//i+Pn38YWREzhr813qvrwHz7yTh866a71hfUf158CbPlq3Za5Y8DR3n3M8T9/yV1YvX8Kgzbbmzd84izFv2rduy2zLPZ+fwqrnnnzV8CG77MfE437W8PZUZN1a+Mc0uPcSWPo0DNoUJn0M9p0GPZr7FrZ69TJm3fPfPPHk31m58nmGD3s9b9rl24wcuWPtFrLgYXjkelg0D15aDLt8FMbv8cr4lODBv8Kcm2H1Chi+Jez0fhiyae3aoIrMXLyUU598ljuWreCpVau5YNtxTN1k5Mvj/7BwEefNX8idS5ezYM1art9hAlOGDm5iiyVJUkds6Bhx9Rr4zjlw1c3wyBMwZCDsvxuc8iUYP7a57ZYkqZEsPldo6XLYYVs48mA48sSOP95dd93FYYcdxvbbb09KiZ///Occe+yxXHTRRQwZMqTjC2hli+Hwg0Nh4hhYtw5+cSsc9hO445uw0+Y1X1xF/nfZQn668DF26je0ocsduPVQJl980Mv/R8+o27JWLVnENZ+ZwuidJrPfqVfQd/golj75GH1HNKWPd7b//u/zBlBYvehZ/vXN9zBi8kEbmKvJbvkB3HE2HPILGL0jPDsL/nwU9OoL+5zQ1Kb987YTWLToIfbc4/sM6L8Jcx6/kutv+CQHveNKBgzYpDYLWbsSBo+FcW+GOy959fiHr4WHr4c3fRQGjYEH/wY3nwNvPR5696tNG1SRpWvXscPAfhw5ZgRHzn51d1rL1q5j8uCBHD56OEfOntuEFkqSpFra0DHi8pfgzgfg+E/Azq+DxUvha2fAO74Es34NvTwSlyR1E37kVeigffINYOpJHX+80047bb3/jz/+eA466CAi5Em1AAAcPklEQVTuuece9t57744voJVDd1r//5PfDT++EW55rDnF58VrV/Oxef/k/HG78e/PNPYCbz16Bf1GD2jIsv51yan0HzmWvU684OVhgzbbuiHLLqf30JHr/b/gukvp2X8Qw/d6Z5NaVIEnb4aJh+QbwLCtYOK74albm9qsNWteYt4T17DP3mewyZjdAdhxhy/y5FMzePiR37DTjsfUZkGbTMo3gDt/uf64lOCRG2Di22Cz4uKBb/oYXPUdePIO2Kr27yVq20EjhnDQiPzl4dQyxeUjxowAYMHqNQ1tlyRJqo8NHSMOHQTXnLP+sPOOg0kfhH/NgR23bUgTJUlqOvt87iSWL1/OunXrGDy4/j/BXrsOfnM7LF0Jk5tUBz36iTt4/9DNOWDQmIYve9m8JVzzll/z9wN+yx1fvY5l816s27KemPknRk7anZu+81H+cNDmXHXkbjx06TmklOq2zEqllFhw3aWMeMuh9Ojbv9nNadsW+8Dj18OCB/L/z90Pc66DCc09WzultaS0lp49+q43vGfPfjz33J2NacTyhbDyRRizXUkD+sCoCfD8Y41pgyRJkiry4rJ8P9xetyRJ3YhnPncSZ555JhMnTmTSpEl1W8Y9T8Jep8FLa2BQX/jjp2HHJpz1/NOFj/LwqqVcPG73hi972E6j2eX7+zJom2GsfH4Fs398Nzd9+M9M+fN76TO89l0ULH3qMWb/4Vy2+9CXmXLEN1g0exZ3nP4VAF73gc/XfHnVWDLrRlY9+wSj3vqBprZjo/b6JqxaAj95A/ToCevWwOTjYdfm5te790BGjdyZ++4/l6FDJ9Kv3ygen/s/LFx4N4MGjW9MI1Yuyfd9Wx3B9B0MKxY3pg2SJEnaqFWr4Ws/gkPeAlvUqHc2SZJeCyw+dwJnnXUWs2bN4uyzz6Znz551W852m8Dd34ZFy+H3d8NRF8OMY2CHzeq2yFd58KUlHDf/Xv4xYQp9ejT+xPtN9hu33v/D3ziGa9/2O+ZdPpsJH6/hReJarFvHiO13ZefPnwzAiO12Ycm82cz+/Y+bXnx+7u+/Y8CEnRiw9Rua2o6Nuv+3cM9FcOivYPQkeOZuuOYYGLY17PzJpjZtzz1P4dZ/focrrtyfiJ4MH/4Gxo8/iBdeaGxXMkSrfstTgvp1ZS5JkqQqrFkDh58Ai5bCn05vdmskSWosi89NduaZZ3Lttdcyffp0NtusvlXgPr1g2+I6d7ttCbfNhR9dD+d/rK6LXc8tyxeyYO0qdnjompeHrSUxc9kCzl34KMt2OIy+PepXgG+t18DeDN52OMvm1KfrjX6jNmXI1q9fb9iQrbZn2e/OqsvyKrV68UIW33Yt4z5Vg6tn1tt134A9vg6TPpz/H7MjLH4cbvl+04vPgweN520HXMSaNctZvXoZ/fuP5qabj2XQwC0a04CWM55fehH6D39l+Kqlrz4bWpIkSQ23Zg185Hi452GYcR6MHNbsFkmS1FgWn5to+vTpXHfddUyfPp0tt9yy4ctfl2Blg697ddjQzdhtwNvXG/bxebczse8gjhuzPX2isWdDr125hqWPLWLUHpvW5fFH77gXS+Y+tN6wJXNnM3Bsg7plaMPCGb8nevdhxN4HN7UdFVmzPHe3UapHT0jrmtOeMnr1GkCvXgNYtWoxT8+/iZ3f+LXGLHjASOg7BJ57EIYX7yFrV8PCR2DSoY1pgyRJkspavQY+/G249xGY8RMYO6rZLZIkqfEsPldo6XJ4eF7+e906mDsf7n4QRgyF8WOrf7zTTz+dq6++mpNPPpnBgwezcOFCAPr378+AAQNq2PLsW1fAwZNg3HBY8hL86naYMRv+57M1X9QGDevZh2E9+6w3bGCPnozo2Ycd+g2t+/Lv+8GtbLL/eAZsOoiVz6/goXPuZu3yNWzxnvpcbnq7Dx/DNUfvy30Xfp/xb/0ALzx0Nw9dejY7ffY/6rK8SqSUWHDtpYzY+2B69h/UtHZUbNtD4JZTcjcboybBM3fBrafDjkc2u2U8/fSNJNYxZPA2LFk6l7v/74cMGbwV22z9ntotZM1KWPZc8U+CFS/A4ieg9wAYMAIm7AcPXQ2DNoFBo/PfPfvC5rvWrg2qyNK1a3l4xUoA1pGYu3I1dy9dzojevRjftw/Pr17D3JWrWLR2LQAPr1jFsJ7LGdunN2P79G5m0yVJUjts6Bhxs1HwgW/CbffDlT/KPaLNX5CnHToI+tf+cjOSJHVKFp8rdPv9sH9JofbE8/LtqHfBhdOqf7zLL78cgK9+9avrDZ86dSqf+MQnOtDS8ua/CIf/AuYvgaH9YKfN4arPwb918u5+a+2l+cu489gZrFr0En2H92PYzmPY53eHMGDz+nRRMPINu/GWH1zGrHNP4N4LvsfATcax49HTmPi+Blf9Syy971ZWPj2Hrb90atPaUJUDz4SZJ8BfPw/Ln4VBm8Iun4Z9vtvslrF69RL+b9YZLF8xnz59hjJuiwPZacdj6NGjhoXERXPhppJuWh64Kt/G7Q5v+hhs+9Z8tvOsy2D18nwG9OTPQW+PaBrt9qXL2f/eR17+/8R58zlx3nyOGjOcCyduyZ+eX8zHW45QgU8/kv8+cdwmTBtfn19fSJKk+tnQMeK0o+GKG/LwXQ9ff74LToSphzSunZIkNZPF5wpN2Q3S7bV7vJkzZ9buwSpw4RENXVxVZkyY0rBl7fqjAxq2rBab730Qm+99UMOX25bBO+zJrpfObnYzKtd3MLz9jHzrZMaPfyfjx7+zvgsZNREOnd72+AjY/p35pqaaMnQwae+d2xw/dZORTN1kZANbJEmS6mljx4i1PH6UJOm1qrEd7EqSJEmSJEmSugWLz5IkSZIkSZKkmrP4LEmSJEmSJEmqOYvPkiRJkiRJkqSas/gsSZIkSZIkSaq5SCk1uw2dTkQsAR5sdju6qFHAgmY3oosy2/ow1/ox2/ox2/ox2/ox2/ox2/ow1/ox2/ox2/ox2/ox2/ox2+5hFDAwpTS6GQvv1YyFvgY8mFLardmN6Ioi4nazrQ+zrQ9zrR+zrR+zrR+zrR+zrR+zrQ9zrR+zrR+zrR+zrR+zrR+z7R6K53mrZi3fbjckSZIkSZIkSTVn8VmSJEmSJEmSVHMWn8v7SbMb0IWZbf2YbX2Ya/2Ybf2Ybf2Ybf2Ybf2YbX2Ya/2Ybf2Ybf2Ybf2Ybf2YbffQ1OfZCw5KkiRJkiRJkmrOM58lSZIkSZIkSTVn8VmSJEmSJEmSVHMWnwsRsUVE/DwinoqIlRExJyLOiIjhzW5bZxER74+IMyPiHxHxYkSkiLhkI/NMjoi/RMTzEbE8ImZFxFcioucG5nlXRMyIiMURsTQibo2Io2q/Rp1DRIyMiE9FxB8j4uGIWFGs+40R8cmIKPs6NdvKRMQPIuLaiJhXZPt8RNwVESdGxMg25jHbdoiII4r3hRQRn2pjmqpzioijIuKfxfSLi/nfVZ+1aL7i8ye1cZvfxjxus1WIiLdExO8j4uniM//piLg6Ig4qM63ZbkRETN3ANttyW1tmPrOtUEQcXGyjTxSfZY9GxKURsVcb05vtRkT2iYj434hYUuR0V0R8ua2czHV90YmPDeI1vu9QTbYR0TsijomICyLi7ohYFRvYF2s1b1U5RUTP4vmaFa/sV/8lIiZ3ZH0bqcpsJ0bENyPiusjHEqsi4pmIuCIi9t/IcrpVtlXmOi4iziley/Mj74s9Vcz78YjovYHldKtcoX3vta3mPz9e2R/bto1pqs4pIvpHxEkR8WBEvBQRz0bE7yLi9e1ZT1X9OtoqNrzv/ZsKltc3Iu4tpn+ijWneHhGnRa6pPF9Me2PVK5dS6vY3YALwDJCAy4FTgOuK/x8ARja7jZ3hBtxdZLIE+Ffx9yUbmP5QYA2wFDgf+GGRZwIubWOeLxbjFwBnAz8C5hXDTm12BnXK9bPF+j0F/BL4PvBzYFEx/DKK/tnNtl35rgL+t8j0FOBM4LZivZ8ExpltTXIeV2yzS4r1/lQtcgJOLcbPK6Y/G1hYDPtis9e7TlnOKbKcVub29TLTu81Wl+93ivV8DrgA+B75Ahy3Af9ltu3KdOc2ttdpwLXFuv/ZbNud7w9K1vtn5M+yy8ifb+uAw822XbleVKzfM0Wu04H7cN+rmgw75bEBXWDfoZpsgWHF+ATMB+bSxr5YR3ICAriUV46Rf1g8j0uL5/XQZudWh2x/U4y/DziPfJz2h2J9E/Bls21XrlOAxcDVwLnkfbHzSrbd64Fe5lp9tmXmPaRk3gRsW4ucgL7AjcU8t5H3VX4FrAaWAXs0O7fX4q3K19FWxfi7Kb8P/v4KlndaybbxRBvTXF6MXwHcU/x9Y9Xr1uxwO8MN+FsR4JdaDT+9GH5us9vYGW7A/sDE4s1pykZeCEOAZ4GVwG4lw/sBNxfzfrjVPFsBLxUfIFuVDB8OPFzMs1ezc6hDrgcUHwo9Wg0fyysfwO8z23bn26+N4ScX632O2XY44wD+DjxC3ll51QFPe3ICJhfDHwaGt3qshcXjbVWv9WpinnOAORVO6zZbXbYfKNbvGmBwmfG9zbbmmd9SrPe7zbZd+Y0F1pILSmNajdu/WO9HzbbqXA9ryQ4YVTK8N/DHYtxUc91ojp3u2IAusu9QZbZ9gHcCmxb/T2Mjxef25AR8pJjnJkr2r4E3F8/rs5T5bO1styqznQrsUmb4fuQvAFe25N7ds23HNtujzPDe5MJzAj5ortVn22q+0eT9h98AM2i7+Fx1TsC3i3kuLX0uyV8yJvIXNq96jr3V7rnmleLzhe1c1hTySQwtJ0O2VXzeC5gE9CxZpsXndgS+TRHeY61fHMBg8rc9y4CBzW5rZ7pV8EL4RDH+F2XGHVCMu6HV8H8vhp9UzeN15RtwXLHeZ5ptzbN9Y7He15hth7M8pvjg2pc2DnjakxOvnJX28TLztPl4r/Ub1RWf3WYrz7UHudC0DBhttg3JfIdinZ8AepptuzLco1i3K9oY/yKwxGyrzrXl8+ULZca1bLd3mGtVmU6hExwb0AX3HTaWbZnpp7Hx4nPVOQEzi+H7V/N4nflWbbat5r2aVicJmW1Ncj2mmPd4c+1YtuQvU+cDI9lw8bmqnMiF0ceL4VtX83jeavdc04HiM/kL4TkUtRA2UHxuY5lVF5/t8znv7ABcnVJaVzoipbSE/O3PAGDPRjfsNa4l17+WGTcTWA5Mjoi+Fc5zVatpuovVxf2akmFmWxuHFPezSoaZbZWKPr1OAaanlGZuYNL25NSds+0bEYdHxHGR+3HcP8r3h+k2W7nJwNbAX4AXIveh+80i33L95pptx32muD8/pVTa57PZVm42+ey63SNiVOmIiNiXfKLE30sGm21lxhb3j5YZ1zLsTRExrPjbXDuuURmae2Wqyql4XiaTn6d/VDJPN1DuOA3Mtt2Kfd2W62/MajXaXKsQEVPJv/L5bEpp4Qama09OE4DxwEMppccqnEf1s1lEfKY4bvxMROxUwTz/Tf4V0Sfr3LaX9WrUgjqx7Yr7h9oYPxs4EHgdud9CVabNXFNKayLiMfKp+9uQ+7LZ2DxPR8QyYIuIGJBSWl6HNncqEdELOLL4t/RD1mzbISK+DgwChgK7AfuQd2pOKZnMbKtQbKMXk7uHOW4jk1eVU0QMBDYHlqaUni7zeLOL+9e1r/Wd3lhytqUei4iPp5RuKBnmNlu5Nxf3zwB3AjuWjoyImeS+0Z4rBpltB0REf+Bw8q8iftZqtNlWKKX0fER8k9wV3P0RcTn558UTgHeTu5D5TMksZluZBcX91mXGbVPy9/bk60aYa8fVPUP3HSrTzpy2Jf/k+tGUUutia1vzdFkRsSXwVnLBbmbJcLOtQvGl6hfJZ9KOBt5OzuNXwJ9LpjPXKhTb53TyGbOXb2Ty9uRUSQ2t9Tyqn7cXt5dFxAzgqJTS3NYTR8R7gKPIv4551fh68cznXIiC3OF9OS3Dh7UxXuW1J9dK5xnaxviu5hTyTz//klL6W8lws22frwMnAl8hF57/ChxYUmgCs63Wd4FdyP1irtjItNXm1J3fmy8gH9CMBQaSi6TnkX/mdFVEvLFkWrfZyo0p7j8L9AfeRj5rdAfytR/2Jfdb18JsO+aD5GyuSinNazXObKuQUjoDeC/5pJFPA98i918+j/xTy2dLJjfbyrQUNY6NiBEtA4svVU8qmW54cW+uHdeIDLvzvkM16vlcdPlsizNFf0m+4Nq0lNILJaPNtjqjyMdn3wU+R/5i9VTysUUqmc5cKxQRPYBfkLuP/XIFs5jta9dy4D+AXcn7K8PJ/dFfT+6y49rii5uXRcQm5OPKq1JK5zeysRafNy6K+7TBqVSt9uTabZ6LiPgy8DXy1WaPqHb24t5sS6SUxqaUglzQey/5zJq7IuJNVTyM2RYiYnfy2c6npZRuqcVDFvfV5tSlcgVIKZ2UUroupfRMSml5SunelNJnyWc+9if35Vgpt9lXtHRbEuQznK9NKS1NKd0HvIfcL/F+bXTBUY7ZbtjRxf157ZjXbEtExP8DLgMuJB+YDyQfaDwK/DIi/quahyvuu3u2vyH/LHgC+Yzyn0TEGeQrxh/EK2dtrW1j/tbMteMamaGZV8btuZWiW4iLgb2B35ILpe1htkBK6YHi+KwXsCXwVfL+w8zSLwarecgqpu2quX6VXID8dKsvRtrLz7dOKqX0bErpuymlO1NKi4rbTHLPDbeSz2r/VKvZfkq+sOenG9xci89s/KyDIa2mU2Xak2ul87zYgXZ1ehHxBfLPZO4nd9L/fKtJzLYDioLeH8lvyiPJF1FoYbYVKOlu4yHghApnqzanjU2/sW/cu6Jzi/t9S4a5zVauZQf80ZTS/5WOKM7cb/mFye7Fvdm2U0S8gdx/4BPkPrZbM9sKRcQU4AfAn1JKx6aUHi2+lLqT/KXJk8DXIqKlqwizrUBxnZd3k38VNZ/8Rf8nyNvsPuSuTQBazio3145rRIbuO1SmPTl1+2PmovB8CfmXJ78DDm91di6YbbuklNamlOamlKaTu5Lak3wRwRbmWoGImAicDFyQUiq3/1VOPd+bu0y2ryVF9yktXd69fNwYEUeSr3t1TErpyUa3y+IzPFjct9UfzcTivq3+bFRem7kWhautyRdneLTCeTYln+nzRFfuFy8ivgKcBdxLLjzPLzOZ2dZASulxcoF/UslFnMy2MoPI6/t64KWISC038k/nAH5aDDuj+L+qnFJKy8hFlUHF+Na643tzSxGk9OdTbrOVa1nvRW2MbylO9281vdlWr60LDbYw28q9q7i/vvWIYl3/Sd6f36UYbLYVSimtSSmdllLaOaXUP6U0JKX0DvK+wc7ACuC+YnJz7bi6Z+i+Q2XamdPD5F8CbFM8X5XM02UU6/xr4MPk/og/Wq5/XLOtiZaL1U1pGWCuFZtE7g7m46XHZ8Ux2n7FNLOLYYcV/7cnJ2tonV9L16Klx40tv/j+RZntA2DzkmE17zLF4vMrO/MHFv3jvCwiBpN/UrOCfLERVe664v4dZcbtCwwAbk4praxwnne2mqbLKS4o9CPyTz73b9WHYymzrZ3NivuW4ojZVmYlcH4bt7uKaW4s/m/pkqM9OXXHbDekpTuI0gNzt9nKzSQXNiZGRJ8y43co7ucU92bbDhHRj3wW6Trye0A5Zlu5vsX96DbGtwxfVdybbccdAfQDfpdSWl0MM9eOa1SG5l6ZqnIqnpebyc/TWyqZp6so9hkuI5/xfBFwRBtfrLYw247ZvLhvXdw3142bQ9vHaC0ntV1a/D8H2p3TI+QLzr8uIspduLcrZvtas2dxX3rceAttbx+Q+5Bu+b/0s7g2Ukrd/kb+qW0CvtRq+OnF8HOb3cbOdiN/E5nIV1AtN34I+duWlcBuJcP7kd/cEvDhVvNsDbxE/qnjViXDh5O/kUvAXs1e9zrleUKxfrcDIzYyrdlWnuv2wNgyw3uQf5KUgJvMtqaZTyvW+VMdzYn8s/1UjB9eMnyr4nFeKn2srnAjn7HwqvcAcj94s4s8jisZ7jZbXb6XFOv3n62Gv51cLF0EDDPbDmV8RLGeV25gGrOtPM8PFus2H9i81bh3FtvtCmCk2Vad7ZAyw94MPA8sAbZxm60qzyl0gmMDuuC+w8ayLTP9NMrsi3U0J+AjxTw3Af1Khr+5eF6fLfe66sy3CrbbvsD/FNP8DOhRwWN2+2wryHUPYECZ4YOAa4p5TzbX6rPdwHwzivm2LTOu6pyAbxfzXFr6ugAOLYbfV8nrxVv7n+viddSnzPADitdDAiZXuKxE/iXRxqbbqpj2xmrXJ4oH6NYiYgJ5p2cMcAXwL/ITuT/5pwKTU0oL236E7qH4aUbLzzPGAv9G/iblH8WwBSmlr7ea/jLyhv8b8s78u4HtiuEfTK02wIj4EvDf5A+Q35LP5Hk/sAX5wmZfp4uJiKPIFxFaC5xJ+b6R5qSULiyZx2wrUHRj8kPyGY+PkNd9E/LPjrYhH8y/NaV0f8k8ZtsBETGN3PXGp1NKP2s1ruqcIuI04FhyP5yXAX2AD5H76/5SSumsuq1MExT5fYv8q5zHyAWQCcDB5AP0vwDvSSmtKpnHbbZCETGGvGO9Lfmz65/kwv57yDtSH00pXVoyvdlWKSL+Qe4z990ppSs3MJ3ZVqD4Vd7fgLeR3w/+SP7sej25S44AvpJyP5kt85htBSLiVnLh/l5ytpPIFxtcCbw3pfS3VtObayud9digK+w7tCPbb5FPuoDcbcwbyce3LRfPvLHMfllVOUVEkPs6fj/5ouhXFtN+iLyP8r6U0hUdWvEGqCbbiLgAmAosAM6h/AXUZqSUZrRaRrfLtspcLycX1m4gn0G7HBhH/lJ1GHnb/beU0tJWy+h2uUL17wdtPMYM8jHwxJTSw63GVZ1TRPQln9k8mXwC3bXAePIvBFYBB6SUbm3H6nZrVb6OZpD3XWaQXxMAO5GLzwAnpJT+s8LlJuDJlNIWZcbtwysXLhwEvI/8hURLFzmklKZudCGNrt531hv5ze4C4Gnyi+Vx8kXfNngWane68cq36G3d5pSZZ29yseQF8g7+PeQrsPbcwHIOIX8QLQGWAbcBRzV7/ZuYayLv1Jht9dnuAJxN7spkAfnnW4uL9Z7W1uvbbDuUecv2XPZsm/bkBBxVTLesmO8G4F3NXtc65bcfuU/BB8hn4a4mnyl2DXAk5C+Ny8znNlt5xiPIv2x6jPx5v5D8xfOeZtvhbF9fvP7nbSgfs606197AV8hdwL1YfJY9C/wZONBs253rN4A7ivfalcV7wrls4KxYc33Vuk2jkx4b8Brfd6g2W145q7Gt24W1yAnoVTxf9xTP3wvF81nR2XWd4VZNthXkmoBpZlt1rgcDvySf6LeYvL/7LPB34Gig1waW061yrTbbDTxGy7b8qjOf25sT+TopJ5G/5FpJPma5FHhDszN7rd6qfB19krwvOAdYWjwHc8lf1L6lyuUm2jjzmfwF3AbfBytZhmc+S5IkSZIkSZJqzgsOSpIkSZIkSZJqzuKzJEmSJEmSJKnmLD5LkiRJkiRJkmrO4rMkSZIkSZIkqeYsPkuSJEmSJEmSas7isyRJkiRJkiSp5iw+S5IkSZIkSZJqzuKzJEmSJEmSJKnmLD5LkiRJkiRJkmrO4rMkSZIkSZIkqeb+P3grnEhb0HDbAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 1728x576 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Plotting the predicted fitness effect for the variants seen more than once\n",
    "plt.close('all')\n",
    "f, (ax1, ax2) = plt.subplots(2, 1, sharey=True, sharex = True, figsize=(24, 8))\n",
    "gs = matplotlib.gridspec.GridSpec(2, 1, width_ratios=[1], height_ratios=[10,0.5])\n",
    "ax1 = plt.subplot(gs[0])\n",
    "ax2 = plt.subplot(gs[1])\n",
    "gs.update(hspace=0.05)\n",
    "\n",
    "nonsense_list = []\n",
    "y_list = []\n",
    "x_list=[]\n",
    "y_list_top20 = []\n",
    "x_list_top20=[]\n",
    "y_list_nonsense = []\n",
    "x_list_nonsense = []\n",
    "x=[]\n",
    "y=[]\n",
    "OBS_EXP_DICT={}\n",
    "top20_lower_err = []\n",
    "top20_upper_err = []\n",
    "missense_lower_err = []\n",
    "missense_upper_err = []\n",
    "nonsense_lower_err = []\n",
    "nonsense_upper_err = []\n",
    "\n",
    "top20 = c1\n",
    "missense = c6\n",
    "nonsense = c10\n",
    "synonymous = c3\n",
    "\n",
    "axisfont=20\n",
    "domainfont = 12\n",
    "titlefont=12\n",
    "axislabelfont=12\n",
    "\n",
    "top_20_ASXL1 = []\n",
    "\n",
    "#ERROR NEEDS TO BE IN THE FORMAT: [tuple(lower_err),tuple(upper_err)]\n",
    "\n",
    "for k, v in ASXL1_missense_s.items():\n",
    "    if k in top_20_ASXL1:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_top20.append(codon_number)\n",
    "        y_list_top20.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        top20_lower_err.append(error_lower)\n",
    "        top20_upper_err.append(error_upper)\n",
    "\n",
    "    if k not in top_20_ASXL1:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list.append(codon_number)\n",
    "        y_list.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        missense_lower_err.append(error_lower)\n",
    "        missense_upper_err.append(error_upper)\n",
    "\n",
    "    missense_err = [tuple(missense_lower_err), tuple(missense_upper_err)]   \n",
    "    \n",
    "for k, v in ASXL1_nonsense_s.items():\n",
    "    if k in top_20_ASXL1:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_top20.append(codon_number)\n",
    "        y_list_top20.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        top20_lower_err.append(error_lower)\n",
    "        top20_upper_err.append(error_upper)\n",
    "    \n",
    "    top20_err = [tuple(top20_lower_err), tuple(top20_upper_err)]\n",
    "        \n",
    "    if k not in top_20_ASXL1:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_nonsense.append(codon_number)\n",
    "        y_list_nonsense.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        nonsense_lower_err.append(error_lower)\n",
    "        nonsense_upper_err.append(error_upper)\n",
    "\n",
    "    nonsense_err = [tuple(nonsense_lower_err), tuple(nonsense_upper_err)]        \n",
    "    \n",
    "msize = 12\n",
    "medgewidth = 2\n",
    "linew = 3\n",
    "csize = 4\n",
    "cthick = 3\n",
    "\n",
    "markers1, caps1, bars1 = ax1.errorbar(x_list, y_list, yerr = missense_err, fmt = 'o', ecolor = missense, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = missense, markerfacecolor = missense, zorder = 1)    \n",
    "\n",
    "markers2, caps2, bars2 = ax1.errorbar(x_list_nonsense, y_list_nonsense, yerr = nonsense_err, fmt = 'o', ecolor = nonsense, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = nonsense, markerfacecolor = nonsense, zorder = 2)    \n",
    "\n",
    "markers3, caps3, bars3 = ax1.errorbar(x_list_top20, y_list_top20, yerr = top20_err, fmt = 'o', ecolor = top20, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = top20, markerfacecolor = top20, zorder = 3)    \n",
    "\n",
    "# loop through bars and caps and set the alpha value\n",
    "a = 0.7\n",
    "[bar.set_alpha(a) for bar in bars1]\n",
    "[cap.set_alpha(a) for cap in caps1]\n",
    "[bar.set_alpha(a) for bar in bars2]\n",
    "[cap.set_alpha(a) for cap in caps2]\n",
    "[bar.set_alpha(a) for bar in bars3]\n",
    "[cap.set_alpha(a) for cap in caps3]\n",
    "\n",
    "#Only show the required axis lines\n",
    "ax1.spines['top'].set_visible(False)\n",
    "ax1.spines['right'].set_visible(False)\n",
    "\n",
    "for axis in ['bottom','left']:\n",
    "    ax1.spines[axis].set_linewidth(2)\n",
    "\n",
    "for axis in ['bottom','left']:\n",
    "    ax1.spines[axis].set_color(grey3)\n",
    "    \n",
    "ax1.set_ylabel('fitness effect (s)', fontsize = axisfont, labelpad = 10)\n",
    "ax1.set_title('Fitness effects across ASXL1', fontsize = 18)\n",
    "\n",
    "# ax1.set_yscale('log')\n",
    "\n",
    "# # STACKED BAR CHART\n",
    "# define the exons + create data\n",
    "exons = [{'1':(19-1)+1, '2': (47-20)+1, '3': (84-48)+1, '4':(125-85)+1, '5':(157-126)+1, '6': (189-158)+1,\\\n",
    "          '7': (240-190)+1,'8': (294-241)+1, '9': (327-295)+1, '10': (362-328)+1, '11': (573-363)+1, \\\n",
    "          '12': (1541-574)+1}]\n",
    "exon_orders = [['1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11', '12']]\n",
    "colors2 = [c4, c0, c3, c1, c5, c6,c7, c8, c9, c10, c2,c4]\n",
    "names = exons[0].keys()\n",
    "values = np.array([[data[name] for name in order] for data,order in zip(exons, exon_orders)])\n",
    "lefts = np.insert(np.cumsum(values, axis=1),0,0, axis=1)[:, :-1]\n",
    "orders = np.array(exon_orders)\n",
    "\n",
    "# #plot the bar chart\n",
    "for name, color in zip(names, colors2):\n",
    "    idx = np.where(orders == name)\n",
    "    value = values[idx]\n",
    "    left = lefts[idx]\n",
    "    ax2.bar(x=left, height=-10, width=value, bottom=-5, color=color, orientation = 'horizontal', label = name, edgecolor='k', linewidth=1, tick_label = name)\n",
    "    ax2.text(left + (value/2.), (-0.26), name, ha='center', va='center', fontsize = 14)\n",
    "    \n",
    "# # CONFIGURING THE GRAPH APPEARANCE\n",
    "# #Set the x and y axis limits\n",
    "ax1.set_xlim(0, 1542)\n",
    "ax1.set_ylim(-0.01, 30)\n",
    "ax2.set_xlim(0, 1542)\n",
    "ax2.set_ylim(-0.5, 0)\n",
    "\n",
    "#x-axis ticks\n",
    "x1_major_ticks = []\n",
    "x1_major_tick_labels = []\n",
    "x2_major_ticks = [0, 100, 200, 300, 400, 500, 600, 700, 800, 900, 1000, 1100, 1200, 1300, 1400, 1541]\n",
    "x2_major_tick_labels = [\"0\", \"100\", \"200\", \"300\", \"400\", \"500\", \"600\", \"700\", \"800\", \"900\", \"1000\", \"1100\", \"1200\", \"1300\", \"1400\", \"1541\"]\n",
    "ax1.set_xticks(x1_major_ticks)\n",
    "ax1.set_xticklabels(x1_major_tick_labels, fontsize = axisfont)\n",
    "ax2.set_xticks(x2_major_ticks)\n",
    "ax2.set_xticklabels(x2_major_tick_labels, fontsize = axisfont)\n",
    "\n",
    "#y-axis ticks\n",
    "y1_major_ticks = [0, 5, 10, 15, 20, 25, 30]\n",
    "y1_major_tick_labels = [\"0\", \"5\", \"10\", \"15\", \"20\", \"25\", \"30\"]\n",
    "ax1.set_yticks(y1_major_ticks)\n",
    "ax1.set_yticklabels(y1_major_tick_labels, fontsize = axisfont)\n",
    "\n",
    "#set panel legends\n",
    "legend_marker_size = 2\n",
    "legend_elements = [Line2D([0], [0], marker = '_', color = missense, alpha = 1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='missense'),\n",
    "                   Line2D([0], [0], marker = '_', color=nonsense, alpha=1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='nonsense'),\n",
    "                  Line2D([0], [0], marker = '_', color=top20, alpha=1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='top 20 observed variant')]\n",
    "    \n",
    "ax1.legend(ncol=1, handles=legend_elements, loc='upper left', frameon=0, fontsize = 20)\n",
    "\n",
    "ax1.grid(which='major', linestyle=':', linewidth='0.7', color=grey3, zorder = 0)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## TP53 variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {
    "code_folding": [
     0,
     86,
     97,
     109
    ]
   },
   "outputs": [],
   "source": [
    "def expected_number_variants_missense_TP53(data, params):\n",
    "    s_value = params[0]\n",
    "    \n",
    "    a = 0\n",
    "    expected_number_by_variant_missense = {}\n",
    "    \n",
    "    Jaiswal_fmin = 0.056\n",
    "    Jaiswal_mean = 58\n",
    "    Jaiswal_sigma = 11.5\n",
    "    Jaiswal_total = 17182\n",
    "    \n",
    "    Genovese_fmin = 0.1194\n",
    "    Genovese_mean = 55.1\n",
    "    Genovese_total = 12380\n",
    "    Genovese_sigma = 11.95\n",
    "    \n",
    "    McKerrel_fmin = 0.008\n",
    "    McKerrel_mean = 51.88\n",
    "    McKerrel_total = 4219\n",
    "    McKerrel_sigma = 18.36\n",
    "    \n",
    "    Zink_fmin = 0.2363\n",
    "    Zink_mean = 60\n",
    "    Zink_total = 11262\n",
    "    Zink_sigma = 1.18907720e+01\n",
    "    \n",
    "    Coombs_fmin = 0.019\n",
    "    Coombs_mean = 59.82\n",
    "    Coombs_total = 5649\n",
    "    Coombs_sigma = 15.14\n",
    "    \n",
    "    Young_fmin = 0.0009\n",
    "    Young_mean = 63\n",
    "    Young_total = 89\n",
    "    Young_sigma = 6.8\n",
    "    \n",
    "    Desai_fmin = 0.0169\n",
    "    Desai_mean = 66.1\n",
    "    Desai_total = 424\n",
    "    Desai_sigma = 6 #estimated\n",
    "    \n",
    "    Acuna_fmin = 0.0015\n",
    "    Acuna_mean = 4.5\n",
    "    Acuna_total = 2014\n",
    "    \n",
    "    N = 9.40166610e+04\n",
    "\n",
    "    for i in data:\n",
    "        variant_mutation_rate = mutation_rates_variants_TP53(i)\n",
    "\n",
    "        if i in Genovese_missense_TP53:\n",
    "            Genovese_number = prevalence_averaged_over_ages(Genovese_fmin, s_value, N, variant_mutation_rate, Genovese_mean, Genovese_sigma, Genovese_total)\n",
    "        else:\n",
    "            Genovese_number = 0 \n",
    "\n",
    "        if i in Zink_missense_TP53:\n",
    "            Zink_number = prevalence_averaged_over_ages(Zink_fmin, s_value, N, variant_mutation_rate, Zink_mean, Zink_sigma, Zink_total)\n",
    "        else:\n",
    "            Zink_number = 0\n",
    "\n",
    "        if i in Coombs_missense_TP53:\n",
    "            Coombs_number = prevalence_averaged_over_ages(Coombs_fmin, s_value, N, variant_mutation_rate, Coombs_mean, Coombs_sigma, Coombs_total)\n",
    "        else:\n",
    "            Coombs_number = 0\n",
    "\n",
    "        if i in Young_missense_TP53:\n",
    "            Young_number = prevalence_averaged_over_ages(Young_fmin, s_value, N, variant_mutation_rate, Young_mean, Young_sigma, Young_total)\n",
    "        else:\n",
    "            Young_number = 0\n",
    "\n",
    "        if i in Desai_missense_TP53:\n",
    "            Desai_number = prevalence_averaged_over_ages(Desai_fmin, s_value, N, variant_mutation_rate, Desai_mean, Desai_sigma, Desai_total)\n",
    "        else:\n",
    "            Desai_number = 0\n",
    "\n",
    "        if i in Acuna_missense_TP53:\n",
    "            Acuna_number = prevalence_averaged_over_uniform_ages(Acuna_fmin, s_value, N, variant_mutation_rate, Acuna_mean, Acuna_total)\n",
    "        else:\n",
    "            Acuna_number = 0\n",
    "\n",
    "        expected_total = Genovese_number + Zink_number + Coombs_number + Young_number + Desai_number + Acuna_number\n",
    "\n",
    "        expected_number_by_variant_missense[i] = expected_total\n",
    "    \n",
    "    return expected_number_by_variant_missense\n",
    "\n",
    "def prob_observed_given_expected_missense_TP53(params, data): #data = list of variant (e.g. p.R882H) (or all_missense_variants_TP53)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_TP53_counts[variant])\n",
    "        expected_number = float(expected_number_variants_missense_TP53(data, params)[variant])\n",
    "        square_distance = ((observed_number - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance\n",
    "\n",
    "def prob_observed_given_expected_missense_TP53_upper_error(params, data): #data = list of variant (e.g. p.R882H) (or all_missense_variants_TP53)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_TP53_counts[variant])\n",
    "        observed_number_upper_error = observed_number+np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_missense_TP53(data, params)[variant])\n",
    "        square_distance = ((observed_number_upper_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   \n",
    "\n",
    "def prob_observed_given_expected_missense_TP53_lower_error(params, data): #data = list of variant (e.g. p.R882H) (or all_missense_variants_TP53)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_TP53_counts[variant])\n",
    "        observed_number_lower_error = observed_number-np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_missense_TP53(data, params)[variant])\n",
    "        square_distance = ((observed_number_lower_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {
    "code_folding": [
     0,
     91,
     102,
     114
    ]
   },
   "outputs": [],
   "source": [
    "def expected_number_variants_nonsense_TP53(data, params):\n",
    "    s_value = params[0]\n",
    "    \n",
    "    a = 0\n",
    "    expected_number_by_variant_nonsense = {}\n",
    "    \n",
    "    Jaiswal_fmin = 0.056\n",
    "    Jaiswal_mean = 58\n",
    "    Jaiswal_sigma = 11.5\n",
    "    Jaiswal_total = 17182\n",
    "    \n",
    "    Genovese_fmin = 0.1194\n",
    "    Genovese_mean = 55.1\n",
    "    Genovese_total = 12380\n",
    "    Genovese_sigma = 11.95\n",
    "    \n",
    "    McKerrel_fmin = 0.008\n",
    "    McKerrel_mean = 51.88\n",
    "    McKerrel_total = 4219\n",
    "    McKerrel_sigma = 18.36\n",
    "    \n",
    "    Zink_fmin = 0.2363\n",
    "    Zink_mean = 60\n",
    "    Zink_total = 11262\n",
    "    Zink_sigma = 1.18907720e+01\n",
    "    \n",
    "    Coombs_fmin = 0.019\n",
    "    Coombs_mean = 59.82\n",
    "    Coombs_total = 5649\n",
    "    Coombs_sigma = 15.14\n",
    "    \n",
    "    Young_fmin = 0.0009\n",
    "    Young_mean = 63\n",
    "    Young_total = 89\n",
    "    Young_sigma = 6.8\n",
    "    \n",
    "    Desai_fmin = 0.0169\n",
    "    Desai_mean = 66.1\n",
    "    Desai_total = 424\n",
    "    Desai_sigma = 6 #estimated\n",
    "    \n",
    "    Acuna_fmin = 0.0015\n",
    "    Acuna_mean = 4.5\n",
    "    Acuna_total = 2014\n",
    "\n",
    "    N = 9.40166610e+04\n",
    "    \n",
    "    for i in data:\n",
    "        variant_mutation_rate = mutation_rates_variants_TP53(i)\n",
    "\n",
    "        if i in Jaiswal_nonsense_TP53:\n",
    "            Jaiswal_number = prevalence_averaged_over_ages(Jaiswal_fmin, s_value, N, variant_mutation_rate, Jaiswal_mean, Jaiswal_sigma, Jaiswal_total)\n",
    "        else:\n",
    "            Jaiswal_number = 0\n",
    "\n",
    "        if i in Genovese_nonsense_TP53:\n",
    "            Genovese_number = prevalence_averaged_over_ages(Genovese_fmin, s_value, N, variant_mutation_rate, Genovese_mean, Genovese_sigma, Genovese_total)\n",
    "        else:\n",
    "            Genovese_number = 0 \n",
    "\n",
    "        if i in Zink_nonsense_TP53:\n",
    "            Zink_number = prevalence_averaged_over_ages(Zink_fmin, s_value, N, variant_mutation_rate, Zink_mean, Zink_sigma, Zink_total)\n",
    "        else:\n",
    "            Zink_number = 0\n",
    "\n",
    "        if i in Coombs_nonsense_TP53:\n",
    "            Coombs_number = prevalence_averaged_over_ages(Coombs_fmin, s_value, N, variant_mutation_rate, Coombs_mean, Coombs_sigma, Coombs_total)\n",
    "        else:\n",
    "            Coombs_number = 0\n",
    "\n",
    "        if i in Young_nonsense_TP53:\n",
    "            Young_number = prevalence_averaged_over_ages(Young_fmin, s_value, N, variant_mutation_rate, Young_mean, Young_sigma, Young_total)\n",
    "        else:\n",
    "            Young_number = 0\n",
    "\n",
    "        if i in Desai_nonsense_TP53:\n",
    "            Desai_number = prevalence_averaged_over_ages(Desai_fmin, s_value, N, variant_mutation_rate, Desai_mean, Desai_sigma, Desai_total)\n",
    "        else:\n",
    "            Desai_number = 0\n",
    "\n",
    "        if i in Acuna_nonsense_TP53:\n",
    "            Acuna_number = prevalence_averaged_over_uniform_ages(Acuna_fmin, s_value, N, variant_mutation_rate, Acuna_mean, Acuna_total)\n",
    "        else:\n",
    "            Acuna_number = 0\n",
    "\n",
    "        expected_total = Jaiswal_number + Genovese_number + Zink_number + Coombs_number + Young_number + Desai_number + Acuna_number\n",
    "\n",
    "        expected_number_by_variant_nonsense[i] = expected_total\n",
    "    \n",
    "    return expected_number_by_variant_nonsense\n",
    "\n",
    "def prob_observed_given_expected_nonsense_TP53(params, data): #data = list of variant (e.g. p.R882H) (or all_nonsense_variants_TP53)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_TP53_counts[variant])\n",
    "        expected_number = float(expected_number_variants_nonsense_TP53(data, params)[variant])\n",
    "        square_distance = ((observed_number - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance\n",
    "\n",
    "def prob_observed_given_expected_nonsense_TP53_upper_error(params, data): #data = list of variant (e.g. p.R882H) (or all_nonsense_variants_TP53)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_TP53_counts[variant])\n",
    "        observed_number_upper_error = observed_number+np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_nonsense_TP53(data, params)[variant])\n",
    "        square_distance = ((observed_number_upper_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   \n",
    "\n",
    "def prob_observed_given_expected_nonsense_TP53_lower_error(params, data): #data = list of variant (e.g. p.R882H) (or all_nonsense_variants_TP53)\n",
    "    total_square_distance = 0\n",
    "    a = 0\n",
    "    for variant in data:\n",
    "        observed_number = float(all_studies_TP53_counts[variant])\n",
    "        observed_number_lower_error = observed_number-np.sqrt(observed_number)\n",
    "        expected_number = float(expected_number_variants_nonsense_TP53(data, params)[variant])\n",
    "        square_distance = ((observed_number_lower_error - expected_number)**2)\n",
    "        total_square_distance = total_square_distance + square_distance\n",
    "\n",
    "    return total_square_distance   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#CALCULATING S FOR MISSENSE VARIANTS\n",
    "TP53_missense_s = {}\n",
    "a = 0\n",
    "for var in all_missense_variants_TP53:\n",
    "    initial_guess=[0.05]\n",
    "    if var in all_studies_TP53_counts.keys():\n",
    "        observed_number = float(all_studies_TP53_counts[var])\n",
    "#         print(observed_number)\n",
    "        if observed_number >1:\n",
    "            variant_mutation_rate = mutation_rates_variants_TP53(var)\n",
    "            \n",
    "            MLE = scipy.optimize.minimize(prob_observed_given_expected_missense_TP53, initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value = MLE.x[0]\n",
    "\n",
    "            upper_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_missense_TP53_upper_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_upper = upper_error_MLE.x[0]\n",
    "\n",
    "            lower_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_missense_TP53_lower_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_lower = lower_error_MLE.x[0]\n",
    "\n",
    "#             print('a = ', a)\n",
    "            TP53_missense_s[var]=(s_value, observed_number, variant_mutation_rate, s_value_upper, s_value_lower) \n",
    "    a = a+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#CALCULATING S FOR NONSENSE VARIANTS\n",
    "TP53_nonsense_s = {}\n",
    "a = 0\n",
    "for var in all_nonsense_variants_TP53:\n",
    "    initial_guess=[0.05]\n",
    "    if var in all_studies_TP53_counts.keys():\n",
    "        observed_number = float(all_studies_TP53_counts[var])\n",
    "#         print(observed_number)\n",
    "        if observed_number >1:\n",
    "            variant_mutation_rate = mutation_rates_variants_TP53(var)\n",
    "            \n",
    "            MLE = scipy.optimize.minimize(prob_observed_given_expected_nonsense_TP53, initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value = MLE.x[0]\n",
    "\n",
    "            upper_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_nonsense_TP53_upper_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_upper = upper_error_MLE.x[0]\n",
    "\n",
    "            lower_error_MLE = scipy.optimize.minimize(prob_observed_given_expected_nonsense_TP53_lower_error, \\\n",
    "                                                      initial_guess, args=([var],), method='Nelder-Mead')\n",
    "            s_value_lower = lower_error_MLE.x[0]\n",
    "\n",
    "#             print('a = ', a)\n",
    "            TP53_nonsense_s[var]=(s_value, observed_number, variant_mutation_rate, s_value_upper, s_value_lower) \n",
    "    a = a+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#Exac - retrieve the frequencies in Exac for each of the variants\n",
    "import csv\n",
    "filename=\"Data files/TP53_Exac.csv\"\n",
    "\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    Exac_TP53_results={} \n",
    "    \n",
    "    for row in read_reader:\n",
    "#         print(row)\n",
    "        if row_count>1:\n",
    "            variant = row[6]\n",
    "            allele_freq = float(row[14])\n",
    "            african_freq = float(row[15])/float(row[16])\n",
    "            east_asian_freq = float(row[18])/float(row[19])\n",
    "            european_freq = float(row[21])/float(row[22])\n",
    "#             print(row[24])\n",
    "            if row[25]=='0':\n",
    "                finnish_freq = '-'\n",
    "            if row[25]!='0':\n",
    "#                 print(row)\n",
    "                finnish_freq = float(row[24])/float(row[25])\n",
    "            latino_freq = float(row[27])/float(row[28])\n",
    "            south_asian_freq = float(row[33])/float(row[34])\n",
    "            Exac_TP53_results[variant]=(allele_freq, african_freq, east_asian_freq, european_freq, finnish_freq, latino_freq, south_asian_freq)\n",
    "\n",
    "        row_count=row_count+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [],
   "source": [
    "#COSMIC - retrieve the frequencies in COSMIC for each of the variants\n",
    "import csv\n",
    "filename=\"Data files/TP53_COSMIC_variant_counts.csv\"\n",
    "with open(filename, 'r') as csvfile:\n",
    "    read_reader = csv.reader(csvfile)  #csv.reader returns a reader object which will iterate over lines in the csvfile\n",
    "    row_count=0\n",
    "    TP53_COSMIC_results={} \n",
    "    \n",
    "    for row in read_reader:\n",
    "#         print(row)\n",
    "        if row_count>1:\n",
    "            variant = row[1]\n",
    "            counts = int(row[2])\n",
    "            TP53_COSMIC_results[variant]=counts\n",
    "\n",
    "        row_count=row_count+1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {
    "code_folding": [
     0
    ],
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "#Create dictionary of results and save to .csv file\n",
    "TP53_s_dictionary={}\n",
    "for k, v in TP53_missense_s.items():\n",
    "#     print(k)\n",
    "    if k in Exac_TP53_results.keys():\n",
    "        exac_freq = round((Exac_TP53_results[k][0])*100, 4)\n",
    "        african_freq = round((Exac_TP53_results[k][1])*100, 4)\n",
    "        east_asian_freq = round((Exac_TP53_results[k][2])*100, 4)\n",
    "        european_freq = round((Exac_TP53_results[k][3])*100, 4)\n",
    "        finnish_freq = round((Exac_TP53_results[k][4])*100, 4)\n",
    "        latino_freq = round((Exac_TP53_results[k][5])*100, 4)\n",
    "        south_asian_freq = round((Exac_TP53_results[k][6])*100, 4)\n",
    "    if k not in Exac_TP53_results.keys():\n",
    "        exac_freq = '-'\n",
    "        african_freq = '-'\n",
    "        east_asian_freq = '-'\n",
    "        european_freq = '-'\n",
    "        finnish_freq = '-'\n",
    "        latino_freq = '-'\n",
    "        south_asian_freq = '-'\n",
    "    if k in TP53_COSMIC_results.keys():\n",
    "        COSMIC_number = TP53_COSMIC_results[k]\n",
    "    if k not in TP53_COSMIC_results.keys():\n",
    "        COSMIC_number = '-'\n",
    "    TP53_s_dictionary[k]=(round(v[0]*100, 2),round(v[4]*100, 2),round(v[3]*100, 2), v[1], v[2], COSMIC_number, exac_freq, \\\n",
    "                            african_freq, east_asian_freq, european_freq, finnish_freq, latino_freq, south_asian_freq)\n",
    "\n",
    "for k, v in TP53_nonsense_s.items():\n",
    "    if k in Exac_TP53_results.keys():\n",
    "        exac_freq = round((Exac_TP53_results[k][0])*100, 4)\n",
    "        african_freq = round((Exac_TP53_results[k][1])*100, 4)\n",
    "        east_asian_freq = round((Exac_TP53_results[k][2])*100, 4)\n",
    "        european_freq = round((Exac_TP53_results[k][3])*100, 4)\n",
    "        finnish_freq = round((Exac_TP53_results[k][4])*100, 4)\n",
    "        latino_freq = round((Exac_TP53_results[k][5])*100, 4)\n",
    "        south_asian_freq = round((Exac_TP53_results[k][6])*100, 4)\n",
    "    if k not in Exac_TP53_results.keys():\n",
    "        exac_freq = '-'\n",
    "        african_freq = '-'\n",
    "        east_asian_freq = '-'\n",
    "        european_freq = '-'\n",
    "        finnish_freq = '-'\n",
    "        latino_freq = '-'\n",
    "        south_asian_freq = '-'\n",
    "    if k in TP53_COSMIC_results.keys():\n",
    "        COSMIC_number = TP53_COSMIC_results[k]\n",
    "    if k not in TP53_COSMIC_results.keys():\n",
    "        COSMIC_number = '-'\n",
    "    TP53_s_dictionary[k]=(round(v[0]*100, 2),round(v[4]*100, 2),round(v[3]*100, 2), v[1], v[2], COSMIC_number, exac_freq, \\\n",
    "                            african_freq, east_asian_freq, european_freq, finnish_freq, latino_freq, south_asian_freq)\n",
    "\n",
    "TP53_df = pd.DataFrame.from_dict(TP53_s_dictionary, orient='index', columns=['s (%)', 's lower value (%)', 's upper value (%)','observed number', 'site-specific mutation rate', 'Number of times in COSMIC',\\\n",
    "                                                                                'Exac Overall freq (%)', 'Exac African freq (%)', 'Exac East Asian freq (%)',\\\n",
    "                                                                                'Exac European freq (%)', 'Exac Finnish freq (%)', 'Exac Latino freq (%)', 'Exac South Asian freq (%)'])\n",
    "TP53_df_sorted = TP53_df.sort_values(by = ['s (%)'], ascending = False)\n",
    "\n",
    "TP53_df_sorted.to_csv('Data files/TP53_s_across_whole_gene.csv')\n",
    "# TP53_s_dictionary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>s (%)</th>\n",
       "      <th>s lower value (%)</th>\n",
       "      <th>s upper value (%)</th>\n",
       "      <th>observed number</th>\n",
       "      <th>site-specific mutation rate</th>\n",
       "      <th>Number of times in COSMIC</th>\n",
       "      <th>Exac Overall freq (%)</th>\n",
       "      <th>Exac African freq (%)</th>\n",
       "      <th>Exac East Asian freq (%)</th>\n",
       "      <th>Exac European freq (%)</th>\n",
       "      <th>Exac Finnish freq (%)</th>\n",
       "      <th>Exac Latino freq (%)</th>\n",
       "      <th>Exac South Asian freq (%)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>p.N235S</th>\n",
       "      <td>19.63</td>\n",
       "      <td>16.52</td>\n",
       "      <td>22.31</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.094130e-09</td>\n",
       "      <td>6</td>\n",
       "      <td>0.0239</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0345</td>\n",
       "      <td>0.0913</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.S241C</th>\n",
       "      <td>17.65</td>\n",
       "      <td>13.41</td>\n",
       "      <td>20.52</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.368770e-10</td>\n",
       "      <td>9</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0152</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.K132E</th>\n",
       "      <td>16.07</td>\n",
       "      <td>12.00</td>\n",
       "      <td>18.56</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.441820e-10</td>\n",
       "      <td>9</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.Y220C</th>\n",
       "      <td>14.92</td>\n",
       "      <td>12.67</td>\n",
       "      <td>16.50</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.472060e-09</td>\n",
       "      <td>145</td>\n",
       "      <td>0.0025</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0045</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R110L</th>\n",
       "      <td>14.38</td>\n",
       "      <td>11.22</td>\n",
       "      <td>16.17</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.064910e-09</td>\n",
       "      <td>7</td>\n",
       "      <td>0.0017</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.G245C</th>\n",
       "      <td>14.11</td>\n",
       "      <td>11.02</td>\n",
       "      <td>15.85</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.163110e-09</td>\n",
       "      <td>10</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.V216M</th>\n",
       "      <td>10.51</td>\n",
       "      <td>7.78</td>\n",
       "      <td>11.96</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.211140e-09</td>\n",
       "      <td>60</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R248Q</th>\n",
       "      <td>9.83</td>\n",
       "      <td>8.63</td>\n",
       "      <td>10.65</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.415120e-08</td>\n",
       "      <td>189</td>\n",
       "      <td>0.0058</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0231</td>\n",
       "      <td>0.0075</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.W146*</th>\n",
       "      <td>9.30</td>\n",
       "      <td>7.11</td>\n",
       "      <td>10.35</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.610690e-09</td>\n",
       "      <td>9</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R273H</th>\n",
       "      <td>8.75</td>\n",
       "      <td>7.90</td>\n",
       "      <td>9.38</td>\n",
       "      <td>5.0</td>\n",
       "      <td>3.269290e-08</td>\n",
       "      <td>106</td>\n",
       "      <td>0.0026</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0047</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R110C</th>\n",
       "      <td>8.63</td>\n",
       "      <td>7.04</td>\n",
       "      <td>9.55</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.415120e-08</td>\n",
       "      <td>-</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0151</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R248W</th>\n",
       "      <td>8.63</td>\n",
       "      <td>7.04</td>\n",
       "      <td>9.53</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.415120e-08</td>\n",
       "      <td>129</td>\n",
       "      <td>0.0008</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0015</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>p.R283H</th>\n",
       "      <td>8.21</td>\n",
       "      <td>6.73</td>\n",
       "      <td>9.04</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.882290e-08</td>\n",
       "      <td>2</td>\n",
       "      <td>0.0066</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.003</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0262</td>\n",
       "      <td>0.0183</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         s (%)  s lower value (%)  s upper value (%)  observed number  \\\n",
       "p.N235S  19.63              16.52              22.31              5.0   \n",
       "p.S241C  17.65              13.41              20.52              2.0   \n",
       "p.K132E  16.07              12.00              18.56              2.0   \n",
       "p.Y220C  14.92              12.67              16.50              4.0   \n",
       "p.R110L  14.38              11.22              16.17              2.0   \n",
       "p.G245C  14.11              11.02              15.85              2.0   \n",
       "p.V216M  10.51               7.78              11.96              2.0   \n",
       "p.R248Q   9.83               8.63              10.65              4.0   \n",
       "p.W146*   9.30               7.11              10.35              2.0   \n",
       "p.R273H   8.75               7.90               9.38              5.0   \n",
       "p.R110C   8.63               7.04               9.55              2.0   \n",
       "p.R248W   8.63               7.04               9.53              2.0   \n",
       "p.R283H   8.21               6.73               9.04              2.0   \n",
       "\n",
       "         site-specific mutation rate Number of times in COSMIC  \\\n",
       "p.N235S                 1.094130e-09                         6   \n",
       "p.S241C                 4.368770e-10                         9   \n",
       "p.K132E                 5.441820e-10                         9   \n",
       "p.Y220C                 1.472060e-09                       145   \n",
       "p.R110L                 1.064910e-09                         7   \n",
       "p.G245C                 1.163110e-09                        10   \n",
       "p.V216M                 3.211140e-09                        60   \n",
       "p.R248Q                 1.415120e-08                       189   \n",
       "p.W146*                 4.610690e-09                         9   \n",
       "p.R273H                 3.269290e-08                       106   \n",
       "p.R110C                 1.415120e-08                         -   \n",
       "p.R248W                 1.415120e-08                       129   \n",
       "p.R283H                 1.882290e-08                         2   \n",
       "\n",
       "        Exac Overall freq (%) Exac African freq (%) Exac East Asian freq (%)  \\\n",
       "p.N235S                0.0239                     0                        0   \n",
       "p.S241C                0.0008                     0                        0   \n",
       "p.K132E                0.0008                     0                        0   \n",
       "p.Y220C                0.0025                     0                        0   \n",
       "p.R110L                0.0017                     0                        0   \n",
       "p.G245C                     -                     -                        -   \n",
       "p.V216M                     -                     -                        -   \n",
       "p.R248Q                0.0058                     0                   0.0231   \n",
       "p.W146*                     -                     -                        -   \n",
       "p.R273H                0.0026                     0                        0   \n",
       "p.R110C                0.0008                     0                        0   \n",
       "p.R248W                0.0008                     0                        0   \n",
       "p.R283H                0.0066                     0                        0   \n",
       "\n",
       "        Exac European freq (%) Exac Finnish freq (%) Exac Latino freq (%)  \\\n",
       "p.N235S                 0.0345                0.0913                    0   \n",
       "p.S241C                      0                0.0152                    0   \n",
       "p.K132E                 0.0015                     0                    0   \n",
       "p.Y220C                 0.0045                     0                    0   \n",
       "p.R110L                 0.0015                     0                    0   \n",
       "p.G245C                      -                     -                    -   \n",
       "p.V216M                      -                     -                    -   \n",
       "p.R248Q                 0.0075                     0                    0   \n",
       "p.W146*                      -                     -                    -   \n",
       "p.R273H                 0.0047                     0                    0   \n",
       "p.R110C                      0                0.0151                    0   \n",
       "p.R248W                 0.0015                     0                    0   \n",
       "p.R283H                  0.003                     0               0.0262   \n",
       "\n",
       "        Exac South Asian freq (%)  \n",
       "p.N235S                         0  \n",
       "p.S241C                         0  \n",
       "p.K132E                         0  \n",
       "p.Y220C                         0  \n",
       "p.R110L                    0.0061  \n",
       "p.G245C                         -  \n",
       "p.V216M                         -  \n",
       "p.R248Q                         0  \n",
       "p.W146*                         -  \n",
       "p.R273H                         0  \n",
       "p.R110C                         0  \n",
       "p.R248W                         0  \n",
       "p.R283H                    0.0183  "
      ]
     },
     "execution_count": 130,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "TP53_df_sorted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {
    "code_folding": [
     0
    ]
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABZoAAAHvCAYAAADdMGqTAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAIABJREFUeJzs3Xd4VFX+x/H3CYHQEzqCQLAAigqCClKWUBWUIqIiuBLrCivI2lZdNaCywmJFDPqzICi6KgooSlOCBASl9w4JkISEJIQQQiDl/P64M7MzmUkooern9Tx5JnPuued+7z1z55Fvjt9rrLWIiIiIiIiIiIiIiJyqoHMdgIiIiIiIiIiIiIhc2JRoFhEREREREREREZESUaJZREREREREREREREpEiWYRERERERERERERKRElmkVERERERERERESkRJRoFhEREREREREREZESUaJZRERE5AJmjAk2xlhjzIfnOpbzkTFmqDFmizHmmOs6Xexq72yM+c0Yk+Vqv+dcxyoiIiIiciFTollERETkPGSMiXAlQIv6aV3MvkHGmBHGmF5nM+bzjTGmKzAOWA88DPwVSDfGVAe+BcoCj7vaF5+hGFq45qL+mRj/z8gYs/c494b3TzuvP8Z4/+QYY7YaY14zxlQ5ifHDCvW9zxgz17VPjjFmvzHmV2PMvcYY/VtLRERE/lSCz3UAIiIiIlKsL4AfA7RvB7DW5hljygF5XtuCgCjgI+C7Mx7h+aur6/U+a22mu9EYEwFUBl6w1p7p69MCZy5+Anaf4WP9WQwFKni9bwo8A0wFZhTqu8Xr9xXAW67fqwK3Ak8AXYwx11trc736bgBGBzj24ULvWwJpwHhgP1DJNe4koC3wtxM7JREREZELnxLNIiIiIue3ldbaz4rrYK3NOVvBXGBqA/neSWavdoD0sxzPBccYY4AK1tqscx2Lm7V2mvd7Y0wXnETzmkD3ijHG/W+evYW2jzPG/Ah0x0kOe4+773j3nSuWRwM0v2WMmQ08aIz5l7U29XjjiIiIiPwR6H/nEhEREbmAFa7RbIy5DHCvzHzA63/5zyvc31VWINYYk22MSTXG/J8xpkKAY9Q1xrxnjNnjqnWc4HpfvVC/asaYt40xO11lBNKMMcuNMY8X6hdpjFlmjMkwxhw2xuwwxnxmjKl2gud8kzFmnjHmoOs4a4wxD3ttv8wYY3FKYpTyugY/GWP24qz0Boj1vjaufcsaY543xmx0jX3AGPOdMaZZgDiMMeZvxpjfjVPr+ZAxZq0xJsq1/RXgg0LH8p6rcsaYl4xTQzrbdax1xphAK2kLHzvYFWesMSbZNS/xxph3jTFVi9jnDmPML67rlm2M2eyar2DX9i6u+P5qnNrWm4CjwHCvMSJc1zHTNcYKY0xkgGNdbYz5xvVZOWqMSTLGzDfGdPfqc8rnfxrNdb1eVniD6xpXOsVxd+P8Wyv0VAMTERERudBoRbOIiIjI+a184YQucNRae6iI/vuAQTj/6/4C/pdULSjUryVwm2v7Z0An4CGcEhxD3J2MMQ2BX4FSrr47gcuBwUBH45QccK8Y/ha4EXgPWItT3uAKIAJ4wzXefcDHwC/Ai8ARoD7QA6iOU4agSMaYwcC7rpheBrKBm4D3jTENrbXPuq7BX10xtnZdD4AknJIZtwAPuPbf6r42xpgyOInHVsBknPrOVVzX5VdjTDtr7SpXX4NT1uQuYAkwCshwne/twEjga5zV097HAlfZE9d1uhf4xDVGade17VTcNXApi1P24RtgOk5JhxtwalG3NYVKQRhjxgBP45SEeB1IBi51xfovwHvF8hOu8/7IdS3jXWP0wSlPkQSMdR3zbmCi69q7E+w1gPlAvuscd+PM7fWuGGedhvM/XS53vRZeddwW57NV2hiTgXONn7PWJgUaxDi1m4Nxrlt3nPPaBOw6E0GLiIiInI+MtfZcxyAiIiIihRinjnBMEZu/tNb2d/ULxlnB/JG19sGi2rzGdW8rAFpZa5d7bZuDkxQOs9YecbX9gJOUbmGtTfTq2won2RtlrX3FtYo2DXjHWjusmPP6DmgH1LDW5p/g5XDvezGww3X+9xba9i5OPdxLrbXuxOhnQH9rbXChvg/irDRub61d5NX+FDAG6Gat/cmrPQzngYKbrbVdXG0DgCk4Cf37rbUFXv2D3O+LOpZrWyawwFp70g9tNM6D5kLc8+TV/jecBO7t1tpvXW1tcB52+BNwq7X2qFd/Y13/IDBOCYp5OPPYxLvkgzGmNE7StDxwpbV2n6s9BOePBtcDl1trdxpj+uIkwD0xFHEOp3z+hcZxx/2CtfaVANvdn/lZOAlgcBLCvXDqMGe5Yk919Z8FLAI2A2WAjjh/LEgAbnCfe6FjrAbcq94LXPE8Yq2NK8m5iYiIiFxIVDpDRERE5Pz2fzgPtfP+8UumnYJF3klml/k4ibUGAK7kcXec1ZzHjDHV3T84Cd9dQDfXvtk4ybzWxpgGxRz3IM4D03q4VgWfjDtc8X3sHYsrnu9xVl13Pskxvd2Ds+J3daGxg4GfgQ6uxCrAQMACT3onmQEKvy/GQeBqY0zTkw3UWlvg9ceAUsaYMFes811dWnl1H+h6fcY7yewaJ9Cqk08C1BW+HqgLfOidaHWN9xrOvyvcCeODrtcexyk9ccrnf4q64zywbz/O6vLXgHVAV+/ztdZ2t9aOstZ+Y639wlr7MHAfUA/nwY6B/A3nXhiEs+q7DBB2xs5ERERE5Dyk0hkiIiIi57dt3qtrT6OdAdrcZSvctZKbAAYnifa3IsbJB+eBhMapxfwGEGeM2YCT9JxmrfVemf0Kzorm74BUY8wvOCtNvzyBB85d4XotaqU3QK3jjFGcJjgJwv3F9KmKUzricpyHy5XkQW+P4ayIXm+M2YFzXt8BM4tIAPswxvTHKXPRDKfshLcqXr9fjjNPa08wrq0B2hq6XjcE2Lbe9XqJ63U+8DnOKuB7jTHLcEqSfGmt3ey1X4nO/xT8yv8SxUeBOGvtnhPZ0Vo72Tg1t28pYvtvXm8nG2PGAouMMVdba1U+Q0RERP4UlGgWERER+XMqrmyFKfQ6CaeOcyDZ7l+steONMdNwknF/Ae4Ehhpjplhr73H12WKMaQJ0wVl93AH4EBhhjPnLcZJy7ngGAilF9NleRPuJCAJWA08V0yfdK5YSJUOttd+6Eu09cK5DV+BBYIExppt3jeXCjDF34tSIXgoMA/YCOTiJ8h/w/T8XT3bleHaAthMew5UkHuh6qF93oD1OfegXjDFDrbUTXP1O+fxP0f4S/tEmHqeMzImYBDyJU6pjZAmOKSIiInLBUKJZRERE5I/ndK0G3e4aq/SJJuistQk45T7+z1UbdwpO0vF194P0XOUWfnD9YIzpBcwAhuOsci3KNtdrSROGxY1fA/j5BFbUbsEpDVH9OKuaix3HWpsGfAp86iolMhZnlfKtwLRidv0rTkK4o7U2x91ojLmqiFi7AFcDK4uLpxg7XK+Bylxc6Xr1WSVvrV2HU5riP64yLL/j1ESe4NXnVM//rHLFdinOQxRPRDnXa9UzE5GIiIjI+Uc1mkVERET+YFwP2cuhhEkua20yTsmDO4wx1xfebhw1XL+XN8aUK7R/Hk6iEXcsrjrCha307lOML4FjwEvGmLIB4gkzxpQ5zhjFmYxThzhgstsY412WYwrOKt//FK41Xei9uxxI1UJ9go0xod5truT26kD9A8jHeeic57/nXcd9PkDfz12vrxa+PidRJ3sZzsPwHjDG1PTavwzOyt0CnLIXGGOqFh7XWpsOxAEVjTFlTsP5nxHGmGpFbBoGXIRTC9zdt7QrgR7IUNfr0tMYnoiIiMh5TSuaRURERP6YlgI3GWOeBvYA+dbar05hnL8Bi3DqzU4GVuH8N2RDoA/wEU7d5SuBn1ylMzYAB1xtg3FWwy52jTffGLMfiHXFVRXnQWsFFF2eAwBrbbwx5lHgPWCjMeYzYDfOKuRrgN5AI5wyEqfiDZyVv28aY7oAC4BDQH2cMh+HcMo7APwX6OuKvZEx5nuch9s1cvVt5ur3O86q5uddSfnD/O9BiruNMd/hJFdTcGocD8Ypz/HDcWKd6jrf+caYT4EQ4DbALwFvrf3VGPM6zkrh5caYr4F9OHN4hzHm2uPVx7bW5hljhgJfA8uMMR+4zqU/cAPwkrXWvaL5fuBR12dhO5AHdHRdl8+ttcdcf3AoyfmfKfcZY+4FZuMkxssAETjXehvwklffUJxzmIZTpzoFqI0zDy1x/khzKveciIiIyAVJiWYRERGRP6ZHgPE4K1wr4ayAPemklyu52wJ4BuiFU3M2GydJPB0n4QlO/dpJOEm523ASn3uB94ExXuUd3gXucMVXFecBhCuBIdbaBScQzwfGmE04q2gH4yT7UoHNwL8o/kF+xxv7mDHmZuBR4B7+V1s3ESdxP8mrrzXG3AX8HSexGoVzjXfiJGPd/XYZYx7EqVE8AeehfR+59huHk3ztClTAecjgNOBVa+2+48T6mTGmIs7q69dxkrPfAS8QoH61tfZJY8wq13H/ibMaew8wE2f1+4lcn2nGmK44n6l/4vxbYhNwv7V2olfX+TiJ/544q4DzcBLrT+B8JsFJ2p/y+Z9Bv+PUi+6P8wcMcGJ/FedzfNCrbxbOHz3aAzcBYTjntR4YAnzg+r8LRERERP4UzJl5oPPpYYwZA1yHszKkOnAE5x8x04Hxrppu7r7hOP8RWJQvrbX9z1iwIiIiIiIiIiIiIn9S53ui+RjOCpeNOCszKgCtcZLPiUBra+0eV99wnETzGpxEdGHrrbVTA7SLiIiIiIiIiIiISAmc76UzKns/RdvNGDMKeA54Fud/S/O22lo74izEJiIiIiIiIiIiIiJ4PaX6fBQoyeziri94+dmKRUREREREREREREQCO99XNBelp+t1bYBtdYwxfwOq4TxcZom1NlA/ERERERERERERETkNzusazW7GmCeBijhPFb8OaIeTZO5ird3v6hNO0Q8DXAAMstbuPt6xJkyYYAEGDx5sShq3iIiIiIiIiIiIyJ/BeV06w8uTQBQwHCfJPBvo5k4yu2QDLwMtgSqunw5ADBAB/GyMqVDUAYwxDxtjlrvfHzx40K5atcrm5uba5cuXW8AuX77cHj161K5Zs8YeOHDAbt++3e7du9cmJSXZLVu22MzMTLtq1Sqbl5fnt8+6detsenq63bFjh92zZ4/dt2+f3bx5sz106JBdsWKFzc/P99nnyJEjdv369TYtLc3u3LnT7t692yYnJ9tNmzbZrKwsu3z5cltQUOCzT3Z2tt2wYYPdv3+/3bVrl42Pj7cpKSl248aN9vDhw56+y5Yt8+xz+PBhu3HjRpuSkmLj4+Ptrl277P79++2GDRtsdna2z/ju42VlZdlNmzbZ5ORku3v3brtz506blpZm169fb48cOeKzT35+vl2xYoU9dOiQ3bx5s923b5/ds2eP3bFjh01PT7fr1q2zR48e9dknLy/Prlq1ymZmZtotW7bYpKQku3fvXrt9+3Z74MABu2bNGr99cnNz7apVq+zBgwft1q1bbWJiok1ISLDbtm2zGRkZdvXq1TY3N9fn3I8dO+aZy23bttmEhASbmJhot27dar3n373PsmXL7NGjR+3atWs9c1l4/leuXOk3lzk5OXbdunU+c1l4/gvPpXv+U1NTPXOZnJxsN27c6Jl/a22R8x8XF2fj4uLs/v37/ebf+zUrK8tv/lNTU+369ettdna2z/UqKCiwK1asKHb+c3JyfD5n+fn5duXKlTYzMzPg/K9du7bI+T948KDdsmWLTUxM9Jv/Y8eO+c3/6tWrbUZGht26datNSEiwCQkJduvWrT7z772P9/y77+XExES7ZcsWz/wHupe959/7Xvaef+/rlpOTU+y9XHj+ly1bZrOzs/3m330vu+e/8Fy67+XC8x/oXnZ/dtzzn5yc7Df/he9l9/x738ve879u3Tqf+Xff/+75976XC8+/9z1W3L0caP7d+7jnv7h7Wd/l+i7Xd7m+y/Vdru9yfZfru1zf5fou13e5vsv1Xa7v8j/jdzlnyQWxotnNGFMLaAOMBioBt1prVx5nn2BgEdAKGG6tfbu4/l4rmk9LzCIiIiIiIiIiIiLn0Fmp3HChrGgGwFqbbK2dBnTDqcE8+QT2yQM+dL39yxkMT0RERERERERERORP6YJKNLtZa+OBjUBTY0z1E9jFXWKjyNIZIiIiIiIiIiIiInJqLshEs0sd12v+CfRt7XrdeYZiEREREREREREREfnTOm8TzcaYJsaY2gHag4wxo4CawK/W2gOu9lbGmDIB+ncC/uF6+9mZjFlERERERERERETkzyj4XAdQjJuBscaYhcAOIA2oBXQALgH2AQ959R+DU0pjAbDX1XYN0Mn1+wvW2l/PQtwiIiIiIiIiIiIifyrnc6L5J+D/gLZAMyAMOAxsBT4Fxllr0736fwrcBlwPdAdKA8nAV8B4a23s2QtdRERERERERERE5M/jvE00W2vXA38/if4fAR+duYhEREREREREREREJJDztkaziIiIiIiIiIiIiFwYlGgWERERERERERERkRJRollERERERERERERESkSJZhEREREREREREREpESWaRURERERERERERKRElGgWERERERERERERkRJRollERERERERERERESkSJZhEREREREREREREpESWaRURERERERERERKRElGiW88qIESMwxrBgwYJzHYqIiIiIiIiIiIicICWaRURERERERERERKRElGiW88qjjz7Kpk2buOGGG851KCIiIiIiIiIiInKCgs91ACLeqlevTvXq1c91GCIiIiIiIiIiInIStKJZSiwuLg5jDJGRkezYsYN+/fpRrVo1KlWqRLdu3Vi/fj0A+/fv5+GHH+aiiy6ibNmyXH/99cTExPiMVVSN5tjYWHr27MnFF19MSEgItWvXpnXr1owcOdKnX3JyMk8++SSNGzemQoUKhIWF0bhxYyIjI9m5c6df7HPmzKFHjx5Ur16dkJAQLr30Up566ikyMjL8+oaHhxMeHk52djZPPfUU9evXJyQkhMsuu4wxY8ZgrfXb57vvvqNz585cdNFFhISEUKdOHTp06EB0dLRf3/T0dJ599lmuuOIKypUrR2hoKJ07d2bu3LnHnQMREREREREREZFzSSua5bSJi4ujVatWXHHFFURGRhIXF8e0adOIiIhgyZIl3HzzzVSuXJm77rqL9PR0/vvf/9K9e3e2bt1K/fr1ixx39uzZ3HLLLVSuXJlevXpRt25d0tPT2bRpE9HR0URFRQGQnZ1N27Zt2bFjB127dqVnz55Ya4mPj2fGjBn069ePSy65xDPuSy+9RFRUFFWrVuXWW2+lZs2arF27ltdee40ff/yRJUuWULlyZZ9YcnNz6datG4mJiXTv3p3g4GCmT5/OM888Q05OjicWgP/7v//jb3/7G7Vr16Znz55Ur16dlJQU1q5dy8SJExkyZIinb3x8PBEREcTFxdG+fXtuvvlmDh8+zMyZM7n55pt5//33eeihh07XVImIiIiIiIiIiJxWSjTLafPLL7/wyiuv8K9//cvT9vLLL/Piiy/SqlUr7rzzTqKjowkKchbSd+3alXvvvZc333yTN998s8hxP/jgAwoKCliwYAHNmjXz2Zaamur5/eeff2bHjh0MHz7cb7xjx45x9OhRz/uYmBiioqK48cYb+fHHHwkLC/Ns++STT7jvvvuIioryGycxMZFmzZoxb948ypUrB0BUVBSNGjXizTff5LnnnqN06dIAvP/++5QpU4Y1a9ZQs2bNIuMGGDRoEPHx8XzxxRf079/f056RkUFERATDhg2jV69e1KpVq8jrJCIiIiIiIiIicq4o0XyGDFiae65DOGGfty59WsYJDw/nmWee8WkbNGgQL774IkePHmXs2LGeJDPAgAEDuP/++1m9evUJje9O7HoLVM85UL8yZcpQpkwZz/tx48YBThLbO8kMEBkZydtvv82UKVMCJsDHjRvnc4yaNWvSu3dvJk+ezJYtW7jqqqs824KDgz2J56LiXrNmDb/88gv9+vXzSTIDhIWFMXLkSPr06cM333zjswpaRERERERERETkfKFEs5w2zZs3p1SpUj5tderUAaBRo0ZUqlTJZ1upUqWoVasWe/fuLXbcgQMH8u2339KqVSvuuusuOnbsSNu2bbn44ot9+nXo0IG6desyevRoVq5cSY8ePWjbtm3AuJYsWULp0qX5+uuv+frrr/2OeezYMfbv309aWhrVqlXztIeGhnLZZZf59a9Xrx4ABw4c8In7iSeeoGnTptx111106NCBtm3bUqNGDb9YAA4ePMiIESP8xt6/fz8AmzZtKu4yiYiIiIiIiIiInDNKNMtpExoa6tcWHBxc5Db39tzc4ld/9+3bl5kzZ/L666/z8ccf8/777wPQsmVLXn31Vbp27QpA5cqVWbp0KVFRUXz33XfMmTMHcFYPDxkyhOeff96zujgtLY28vDy/hwkWlpWV5ZNoLrz6ufB55ufne9oef/xxqlevTnR0NOPGjeOtt97CGEOHDh0YO3Ys1113nScWgHnz5jFv3rxiYxERERERERERETkfKdF8hpyuchTiuOWWW7jllls4fPgwv/32GzNnzmTChAnceuutrFq1iiuvvBKAiy++mI8++ghrLRs3bmT+/Pm8++67vPTSSxQUFPDyyy8DTuK7oKCA9PT0Mxr3vffey7333ktGRga//vor06ZN4+OPP+amm25i06ZN1KxZ05OEf/vttxk2bNgZjUdERERERERERORMCDp+F5HzR4UKFejUqRNvvPEGzz33HMeOHWPWrFl+/YwxNG3alKFDh3pWCU+fPt2zvXXr1hw4cIANGzaclbjDwsLo0aMHH3zwAZGRkaSnpxMbG+uJBfC8FxERERERERERudAo0SznvZ9//pkjR474tScnJwNQvnx5ANavX09cXNxx+wH84x//AOChhx4iMTHRb5/Dhw+zdOnSEsU9e/Zs8vLy/NpTUlJ84rnuuuto37493377LR9//HHAsdatW+fZT0RERERERERE5Hyj0hly3nviiSeIi4sjIiKC8PBwypQpw4oVK5g/fz4NGjSgf//+APz00088/vjjtGnThiZNmlCzZk327t3LjBkzCAoK4qmnnvKM2blzZ0aPHs2zzz7L5ZdfTo8ePWjYsCFZWVnEx8fzyy+/0K5dO2bPnn3Kcffv35+yZcvSrl07wsPDsdYSGxvLsmXLaNmyJV26dPH0/fzzz+nUqRMPPPAA48aNo1WrVoSFhbF3717Wrl3L+vXrWbJkCTVr1jz1CykiIiIiIiIiInKGKNEs573nnnuOadOmsXz5cn766SeCgoKoX78+zz33HMOHD6dKlSoA3HTTTQwfPpyFCxcyY8YMMjMzueiii+jatasnAe3tn//8J23btmXcuHEsWrSIGTNmEBoaSt26dXn44YcZMGBAieIePXo0c+bMYeXKlfz444+ULVuWBg0aMGbMGAYPHux5MCE4taVXrFjBO++8wzfffMOUKVPIz8+ndu3aXHnllQwdOpSrr766RPGIiIiIiIiIiIicKcZae65jOK9MmDDBAgwePPhchyIiIiIiIiIiIiJSUuZsHEQ1mkVERERERERERESkRJRoFhEREREREREREZESUaJZREREREREREREREpEiWYRERERERERERERKRElmkVERERERERERESkRJRoFhEREREREREREZESUaJZREREREREREREREpEiWYRERERERERERERKRElmkVERERERERERESkRJRoFhEREREREREREZESUaJZREREREREREREREpEiWYRERERERERERERKRElmkVERERERERERESkRJRoFhEREREREREREZESUaJZREREREREREREREpEiWYRERERERERERERKRElmkVERERERERERESkRM7rRLMxZowx5mdjzB5jzBFjTLoxZpUxJsoYU62IfdoYY3509c02xqw1xgw3xpQ62/H/WcTFxWGMITIykri4OPr370/16tUpW7Ys1113HTNnzvTb5+jRo4wePZprrrmG8uXLU7lyZdq3b89XX311WsY/duwY48aNo0WLFlSpUoXy5csTHh5O7969+emnn/z6b968mcjISOrVq0dISAi1atViwIABbNmyxa9vZGQkxhji4uJ4//33ufrqqylbtiy1atXi4Ycf5uDBg377rF27lrvvvpvw8HBCQkKoUaMGLVq0YPjw4eTm5vr0zcvLIzo6mtatW1O5cmXKly/Ptddey/jx4ykoKCh2LkRERERERERERM4FY6091zEUyRhzDFgJbARSgApAa+A6IBFoba3d49W/N/ANkAN8CaQDPYHGwFRr7R3HO+aECRMswODBg0/rufyRxcXF0bBhQyIiItiwYQOXXHIJN954I+np6Xz55Zfk5uby008/0bFjR8BJAnfr1o1ffvmFJk2acOutt5Kdnc3UqVNJSUnh2Wef5d///vcpjw8wYMAAvvjiC6666io6depEuXLlSExMZNGiRfTt25fXXnvN03f27Nn07duX3NxcevbsyWWXXcbevXv59ttvCQkJISYmhhYtWnj6R0ZGMmnSJO644w7mzJlDz549qVWrFjExMaxatYqOHTsyf/58T/+1a9fSqlUrjDH06tWLhg0bkpmZyfbt24mJiSE9PZ2KFSsCeGKYM2cOjRs3JiIigrJlyxITE8PatWu55557+PTTT8/YXIqIiIiIiIiIyB+OOStHsdaetz9A2SLaRwEWiPZqq4yTjD4KXOc9BvCrq3//4x0zOjraRkdHWzlxu3btsq7ra0eMGOGzbfbs2Raw3bt397T9+9//9rTl5uZ62pOTk22DBg0sYBcvXnzK42dkZFhjjG3ZsqXNy8vzizc1NdXze3p6ug0LC7PVqlWzGzZs8Om3fv16W6FCBXvttdf6tA8aNMgCtl69ejY+Pt7Tnpuba9u3b28B+9tvv3naH3/8cQvY6dOn+8WSnp5u8/PzPe+joqIsYB999FGf2PPy8uz9999f5DgiIiIiIiIiIiJFOCu53OCzks2llPccAAAgAElEQVQ+RdbanCI2fQU8B1zu1dYPqAFMttYu9x7DGPM88DMwGPjvGQrXh+kz7mwc5rSw04edlnEaNGjA888/79N20003Ub9+fX7//XdP28cff4wxhjfeeIPg4P99BGvWrMkLL7zAgw8+yIcffkibNm1OaXxjDNZaQkJCCAryrw5Trdr/qq5MnjyZjIwMxo8fz5VXXunTr2nTpjz00EO89dZbbNy40W/7iy++SP369T3vg4ODue+++4iNjeX333/nhhtu8Olfrlw5v1iqVKni+b2goIDx48dTu3Zt3nzzTUqV+l+1l1KlSvH6668zceJEpkyZQu/evf3GEhEREREREREROVfO60RzMXq6Xtd6tXVyvc4O0H8hkA20McaEWGuPnsng/qyaN2/ukxx1q1evHkuWLAHg0KFDbN++nbp169KkSRO/vp06OdO4atWqUxofoHLlyvTs2ZPvv/+e5s2bc/vtt9O+fXtatWpF+fLlffZ177dmzRpGjBjhN/bWrVsB2LRpk1+i+brrrgsYC8CBAwc8bXfddRdvv/02ffr0oV+/fnTp0oW2bdty6aWX+h0rLS2Nyy+/nFdeecVvbHCS1Zs2bQq4TURERERERERE5Fy5IBLNxpgngYpAKE595nY4SebRXt0au163Ft7fWptnjNkFNAUuAZSpOwPCwsICtgcHB3seYud+UN5FF10UsK+7PSMj45TGd/vyyy8ZM2YMn3/+OVFRUQCULVuWfv368dprr1GrVi0A0tLSAPjggw+KPbesrKwTise9Qjs/P9/TdsMNNxAbG8uoUaOYOnWqp8Zy48aNiYqK4u677/aJZdu2bYwcOfKkYhERERERERERETmXLohEM/AkUMvr/Wwg0lq736st1PV6sIgx3O2Bs5Wn2ekqR/FHExrqTNO+ffsCbk9KSvLpd6rKlSvHiBEjGDFiBHv27GHhwoV88sknfPbZZ8TFxREbG+tznDVr1nDNNdeU6JjFufHGG5k5cyZHjx5lxYoVzJ49m3feeYcBAwZQo0YNunTp4onltttu49tvvz1jsYiIiIiIiIiIiJxu/gVsz0PW2trWWgPUBvrirEpeZYxpcRLDuJ+uaANuNOZhY4yntvPBgwdZtWoVubm5LF/uNC9fvpyjR4+yZs0aDhw4wPbt29m7dy9JSUls2bKFzMxMVq1aRV5ent8+69atIz09nR07drBnzx727dvH5s2bOXToECtWrCA/P99nnyNHjrB+/XrS0tLYuXMnu3fvJjk5mU2bNpGVlcXy5cspKCjw2Sc7O5sNGzawf/9+du3aRXx8PCkpKWzcuJHDhw97+i5btsyzz+HDh9m4cSMpKSnEx8eza9cu9u/fz4YNG8jOzvYZ3328rKwsNm3aRHJyMrt372b37t2AUy7iyJEjPvvk5+dz6NAhADZv3szhw4dp0KABCQkJLFu2jHXr1nH06FHPPhMnTgScesyZmZls2bKFlJQUADIzMzlw4ABr1qzx2cc9/qpVqzh48CBbt24lMTGRhIQEtm3bRqVKlWjatCkzZ86kXr16LFq0iJ9++oljx455aix/8803JCQkkJiYyNatW33m3329li1bxtGjRz1lMeLj4/3m311qIzEx0XMNcnJyWLduHWlpaSQkJHDxxRczZMgQnnnmGcBZTV1QUEBWVhZhYWHExsaSmZnJ+vXrSU1N9cxlcnIyGzdu9My/tbbI+Y+LiyMuLo79+/f7zb/3a1ZWlt/8p6amsn79erKzs30+KwUFBaxYscJv/nfu3ElaWhrr168nJyfH53OWn5/PypUryczMZPPmzezbt489e/awY8cO0tPTWbt2rc9cLl++nLy8PM9cbtmyhcTERPbu3cv27ds983/s2DGffXJzc1m9ejUZGRls3bqVhIQEEhIS2Lp1KxkZGaxevdrvXj527JjfvZyYmMiWLVs88x/oXl67dm3AezkzM5OVK1eSn5/vc91ycnKKvZdXrFjhcy8vW7aM7Oxsv/l338vu+S88l+57ufD8B7qX3Z8d9/wnJyf7zX/he9k9/4cOHfLMpff8r1u3zmf+3fe/e/63bNlCUlISe/fu9Zt/73ssNze3yHs50Py793HP/7Zt24q8l/VdXvx3ufe9HOi7vPD8e9/Lhb/Lve/lwvPvfS8X3qe4+fe+l73P3fteLm7+C3+Xe9/LhefffS97x+b9Xe6ey8LzX3gu3fOv73J9l+u7XN/l+i7Xd7m+y/Vdru9yfZfru1zf5efyu/ysOVtPHTydP0AD4Ciw3qttGU4SuWUR+6x3bb+iuLGjo6NtdHR08c9pFB+7du2ygB00aFDA7R06dLDOR80xatQoC9jevXvbvLw8T/v+/ftteHi4BWxsbOwpj5+SkmKXLl3q1y8zM9PWrl3bBgcH28zMTGuttampqTYsLMzWqFHD/vbbb3775Ofn25iYGJ+2QYMGWcDu2rXLr39MTIwFbFRUlKdt4cKFNiMjw6/v2LFjLWCffvppT9sLL7xgAfvII4/Y7Oxsv30SExPthg0b/NpFRERERERERESKcFZythdK6Qwf1tp4Y8xGoLkxprq1NhXYglO/uRGwwru/MSYYaAjkATvPdrzi68knn2TWrFnMmDGDZs2a0aNHD7Kzs/n6669JSUnh6aefpl27dqc8fkJCAq1bt+aKK66gRYsW1KtXj8zMTGbOnMm+ffsYNmwYlSpVAqBatWpMnTqV2267jdatW9O5c2eaNm1KUFAQu3fvZsmSJaSlpZGTk3PK8bz++uvMnTuXiIgILrnkEipWrMiGDRuYNWsWVapU4eGHH/b0feGFF1izZg3vvfce33//PZ06daJu3bqkpKSwbds2Fi9ezKhRo/weTCgiIiIiIiIiInIuXZCJZpc6rlf3U9fmAwOBm4EvCvX9C1AeWGitPXp2wpOilClThnnz5vHGG2/w+eef88477xAcHEyzZs146623PA/HO1Xh4eGMHDmSBQsWEBMTQ2pqKlWrVqVx48aMHj2a/v37+/Tv3Lkza9eu5bXXXmPOnDnExsZSpkwZ6tSpQ6dOnbj99ttLFM+QIUOoUqUKv/32G4sXLyYvL89TPuOJJ56gQYMGnr6lS5dm+vTpfPbZZ3zyySfMnDmTrKwsatSoQcOGDXn55ZcZOHBgieIRERERERERERE53Yy1AUsWn3PGmCZAhrV2X6H2IOBl4DngV2ttW1d7ZWAHUBloa61d7movi5OEvhG421r73+KOO2HCBAswePDg03tCIiIiIiIiIiIiImefOX6XkjufVzTfDIw1xizESSCnAbWADjgPA9wHPOTubK3NNMY8BEwFFhhj/gukA72Axq72L8/qGYiIiIiIiIiIiIj8CZzPieafgP8D2gLNgDDgMLAV+BQYZ61N997BWjvdGNMB+BdwO1AW2A487up/fi7fFhEREREREREREbmAnbeJZmvteuDvp7DfYqDH6Y9IRERERERERERERAIJOtcBiIiIiIiIiIiIiMiFTYlmERERERERERERESkRJZpFREREREREREREpESUaBYRERERERERERGRElGiWURERERERERERERKRIlmERERERERERERESkRJZpFREREREREREREpESUaBYRERERERERERGRElGiWURERERERERERERKRIlmERERERERERERESkRJZpFziORkZEYY4iLizvXoZz3FixYgDGGESNGnOtQTlhERATGmHMdhoiIiIiIiIjIaadEs5RIXFwcxhgiIyPPdSh+cnNzmTZtGg888ABXXXUVlStXpnz58lx99dW8+OKLHDp0qMh99+7dy/3330+dOnUICQkhPDyc4cOHc+DAgbN4BiKnx/l8n4qIiIiIiIjIH0PwuQ5A5EzZsWMHffv2pUKFCnTs2JFbbrmFrKws5syZw8svv8yXX37J4sWLqV69ut9+bdq0ISUlhd69e9OkSRN+//133n77bWbPns3ixYupVq3aOToruZBNnjyZ7Ozscx2GiIiIiIiIiMhpp0Sz/GFVqlSJd999l0GDBlGhQgVP+7Fjx+jbty8//PADI0eO5J133vHZb8iQIaSkpDBu3DiGDh3qaX/88cd58803+de//sV777131s5D/jjq169/rkMQERERERERETkjVDpDTtmIESNo2LAhAJMmTcIY4/n55JNPPP0KCgp47733uP7666lYsSIVKlTg+uuvZ8KECRQUFPiNa4whIiKCxMRE/vrXv1KzZk3KlStHy5Yt+fzzz084vrp16zJkyBCfJDNAmTJleO655wCnzq+3nTt3MnfuXMLDw/n73//us23kyJFUqFCBTz/9lMOHD59wHCtWrOD222+nZs2ahISE0KBBA4YMGUJSUlKR+xQUFPDGG2/QpEkTypYty8UXX8w//vEPMjMz/fquXbuWu+++m/DwcEJCQqhRowYtWrRg+PDh5Obm+vTNy8sjOjqa1q1be0qJXHvttYwfP95vLrzLLWzdupW77rqLmjVrEhQUxIIFC2jSpAllypQhNTU14DmMHj0aYwzvvvuuT/vevXt59NFHueSSSwgJCaFatWr06tWLZcuWBRwnOTmZBx54gFq1alGuXDmaN2/OpEmTirx2gbz66qsYYxg3blzA7YmJiZQqVYrrr7/ep+2ll16ibdu21K5dmzJlylCnTh0GDBjApk2b/MY43vWCwDWajx07xvjx4+nRowcNGjQgJCSEqlWr0qVLF2bNmhUw3vDwcMLDw8nOzuapp56ifv36hISEcNlllzFmzBistZ6+J3qfioiIiIiIiIiUhFY0yymLiIggIyODt99+m2bNmtGnTx/PtubNm3t+/+tf/8rnn39OvXr1ePDBBzHGMG3aNIYMGcKiRYuYMmWK39gHDhygTZs2hIWFcd9995GRkcFXX33FwIEDSUhI4KmnnipR7KVLlwYgONj3Fpg/fz4A3bp1IyjI9+8wlSpVom3btsydO5elS5fSuXPn4x5n5syZ3H777Vhr6devHw0aNGDFihVMmDCBGTNmsHjxYsLDw/32+8c//sHChQu588476d27N3PmzOGtt94iNjaWRYsWUbZsWcBJMrdq1QpjDL169aJhw4ZkZmayfft2oqOjeeWVVzznmpubS8+ePZkzZw6NGzdmwIABlC1blpiYGIYOHcpvv/3Gp59+6hfLjh07aNWqFY0aNWLgwIEcOXKEypUrM2jQIJ577jm++OILn5XfbpMnT6ZMmTL079/f07Zy5Uq6detGeno6N910E3379iU1NZXp06fTrl07pk2bRo8ePTz909LSaNOmDTt37qRdu3a0a9eOpKQkHnnkEbp163bc6+9277338vzzzzNp0iSGDRvmt/2zzz6joKCAQYMGedoWLlzI6NGj6dixI7fffjsVK1Zk27ZtTJ06le+++47FixfTrFmzE75eRUlPT+exxx6jTZs2dO3alRo1apCUlMT3339Pjx49+OCDD3jwwQf99svNzaVbt24kJibSvXt3goODmT59Os888ww5OTlERUUBJ36fioiIiIiIiIiUiLVWP14/0dHRNjo62sqJ2bVrlwXsoEGDAm7//PPPLWCvvfZae+jQIU97VlaWbdmypQXslClTfPYBLGDvuOMOm5+f72nfuXOnrVKlii1durTdsWNHieJ+5JFHLGCfeeYZn/Ynn3zSAva1114LuN/f//53C5zQZ+TQoUO2WrVqNigoyC5cuNBn2+jRoy1gu3bt6tM+aNAgC9hq1arZuLg4T3t+fr7t27evBexLL73kaX/88cctYKdPn+53/PT0dJ/rFxUVZQH76KOP2ry8PE97Xl6evf/++/3Gcc8tYJ999lm/8ffu3WuDgoJsy5Yt/bb9/vvvFrB9+/b1tOXm5tpLL73UhoSE2AULFvj0T0hIsHXq1LG1a9e2OTk5nvaHHnrIAnb48OE+/ZctW2aDg4MtYKOiovyOH0i3bt0sYNetW+e37corr7SlS5e2qampnrbk5GSbmZnp13f16tW2QoUK9uabb/ZpP971stbaDh06WOdr939ycnLsnj17/PpmZGTYpk2b2ipVqtjs7GyfbQ0aNLCA7d69u8+25ORkGxoaakNDQ+2xY8f8YivqPhURERERERGRP7SzklfViuYzpdkd5zqCE7fm6zM29Mcffww4ZRQqVqzoaa9QoQJjxoyhS5cufPjhhwwYMMBnv1KlSjFmzBifVcUNGzZk2LBhjBw5kk8//dSzYvNkfffdd7z//vtcfPHFPP300z7bDh48CEBoaGjAfd3tGRkZxz3OjBkzSEtL4+6776Z9+/Y+25544gnee+895s2bx+7du/1q9z722GM0aNDA8z4oKIixY8cyffp0Pv74Y1544QWf/uXKlfM7fpUqVTy/FxQUMH78eGrXrs2bb75JqVKlPNtKlSrF66+/zsSJE5kyZQq9e/f2GadWrVoBr3XdunXp3Lkz8+bNY8OGDTRt2tSzzV3awnuF8A8//MCOHTt48skn6dChg89YderU4emnn2b48OH8/PPP9OjRg9zcXKZMmUKlSpUYMWKET//rrruOgQMHnlQJjUGDBjF37lwmTZrE2LFjPe3Lly9n48aN3HbbbT4PeaxZs2bAcZo1a0anTp2YO3cuubm5nhXjbkVdr6KEhIRw8cUX+7WHhoZy//3388QTT7Bs2TL+8pe/+PUZN26cz9zXrFmT3r17M3nyZLZs2cJVV111wnGIiIiIiIiIiJSEEs1yRq1cuZKgoCAiIiL8tnXo0IFSpUqxatUqv23169f31JX1FhERwciRIwPucyJ+/fVXBgwYQIUKFfjmm298krEnwrpq3xausxvIypUrAejUqZPftuDgYP7yl78QFxfHqlWr/BLNhROxAJdccgn16tUjLi6OjIwMwsLCuOuuu3j77bfp06cP/fr1o0uXLrRt25ZLL73UZ9+tW7eSlpbG5ZdfziuvvBIw3nLlygWsPdysWTNCQkIC7hMZGcm8efOYNGkS//nPfwCn5vB///tfatSo4VMGY8mSJQDEx8f7JY4Btm3bBsCmTZvo0aMHmzdvJjs7m/bt2wdM/EdERJxUovm2224jNDSUzz77jNGjR3uS7e4xIiMj/fb54YcfeO+991i+fDmpqank5eX5bE9NTeWiiy7yaSvuehVlw4YNjB07loULF5KUlEROTo7P9oSEBL99QkNDueyyy/za69WrBzjlZ0REREREREREzhYlmuWMOnjwIFWrVqVMmTJ+24KDg6levTopKSl+22rVqhVwvNq1a3vGPVlLliyhe/fuBAUFMWvWLG644Qa/Pu6EZlHjux/GV9SKZ2/uMQonIt3c7YFWRxd3/vHx8Rw8eJCwsDBuuOEGYmNjGTVqFFOnTvXUWG7cuDFRUVHcfffdgFPrGJxk7siRI4uMOSsrK+Axi3LbbbdRuXJlPvvsM1599VVKlSrFzJkzSUtLY/jw4T41sN0xfP118Svo3TG4r9/xPgsnqly5ctx555188MEHzJ07l+7du5Obm8sXX3xBjRo16N69u0//cePG8dhjj1GlShW6du1K/fr1KV++PMYYpk+fzpo1azh69GiJ41q6dCmdOnUiLy+Pzp0706tXLypXrkxQUBCrV69mxowZAY8TFhYWcDz3Nc/Pzz+pOERERERERERESkKJ5jPlDJajuJCEhoaSnp4esMRAXl4eqampAR+UlpycHHC8ffv2ecY9GbGxsdxyyy0EBQUxZ84cWrduHbBf48aNAWcFcCDuVbeNGjU67jHdMbpjLiwpKcmnn7fk5GRPLN4Cnf+NN97IzJkzOXr0KCtWrGD27Nm88847DBgwgBo1atClSxdP/9tuu41vv/32uLF7K271tjt5++GHHzJv3jxuvvnmgGUzvGOeMWMGvXr1Ou5x3f2P91k4GYMGDeKDDz5g0qRJdO/e3ZMUf+yxx3w+n3l5eURFRVG7dm1Wrlzp98cC9+rsQE5ktbu3V155hSNHjhATE+O38v/VV19lxowZJzWeiIiIiIiIiMi5EHT8LiJFc5cfKGr15LXXXktBQQELFy7027Zw4ULy8/Np0aKF37bdu3cTFxfn175gwQLPuCdq/vz5dO/eneDgYObNm1dkkhmgY8eOAMydO5eCggKfbYcOHWLx4sWUK1eu2DHc3DG6Y/aWl5fHokWLAAKe/y+//OLXtnPnTvbs2UN4eHjA1awhISG0adOGl156iXHjxgF4kpRNmjQhLCyMpUuXkpube9zYT4a75MSkSZNITU1l1qxZXHPNNTRv3tynn/uaxcbGntC4TZo0oXz58qxevTrgCvNA1/V42rZty+WXX86MGTM4ePBgkUnx1NRUMjIyaNOmjV+SOSsry1MW5XTYvn07VatWDVheJtDn4FQc7z4VERERERERESkpJZqlRKpUqYIxht27dwfcfv/99wPw7LPPkp2d7WnPzs7mmWeeAeCBBx7w2y8/P59//vOfPsneXbt2MW7cOIKDg7nnnntOKL65c+dy6623UrZsWX7++Weuv/76YvtfeumldOvWjbi4ON59912fbVFRURw+fJh7772XChUqHPfYffr0oWrVqnzxxRcsXbrUZ9tbb73Fzp076dKli199ZoC3336b+Ph4z/uCggKeeuopCgoKuO+++zztsbGxAZOw7lXA5cuXB5xyCkOHDiUpKYlhw4Zx5MgRv32SkpLYuHHjcc+rMO/k7YQJE8jNzQ1Y77h3795ceumlvPvuu/z4448Bx1qyZInnc1K6dGkGDhzIoUOH/Go6L1++nClTppx0rOAklXNycoiOjubHH3/kmmuu8fvDRc2aNSlfvjwrVqzwKSeSm5vLY489Rmpq6ikdO5Dw8HDS09NZu3atT/tHH33EnDlzTssxjnefioiIiIiIiIiUlEpnSIlUrFiRVq1aERsby8CBA2nUqBGlSpWiV69eXHPNNQwYMIAZM2bw1Vdf0bRpU/r06eOpcbtr1y7uvPNOBg4c6DfuNddcw2+//UbLli3p1q0bBw8e5MsvvyQjI4P//Oc/fg+7C2TLli307t2bnJwcevTowYwZMwKWISicxIyOjqZNmzYMGzaMn3/+mSuuuILffvuNmJgYGjVqxKhRo0742nz88cfccccddOjQgTvuuIP69euzYsUK5s6dS+3atXn//fcD7tu2bVuaN2/OXXfdRWhoKHPmzGHNmjW0bNmSp59+2tPv9ddfZ+7cuURERHDJJZdQsWJFNmzYwKxZs6hSpQoPP/ywp+8LL7zAmjVreO+99/j+++/p1KkTdevWJSUlhW3btrF48WJGjRrFlVdeeULn5+3ee+/lhRde4OWXXyY4OJgBAwb49SldujTffvstN910E7fccgtt2rShefPmlC9fnj179rBs2TJ27txJUlKSJ0H+73//m59//pm33nqL5cuX065dO5KSkvjyyy/p0aMH33333SnF+uKLLxIVFUVubq7famaAoKAghg0bxujRo7n66qvp3bs3x44dIyYmhvT0dDp27EhMTMxJHzuQ4cOHM2fOHNq1a8edd95JaGgoy5cvZ9GiRfTr14+pU6eW+BjHu09FRERERERERErMWqsfr5/o6GgbHR1t5cRt27bN3nrrrbZq1arWGGMBO3HiRM/2/Px8++6779qWLVvacuXK2XLlytkWLVrY8ePH2/z8fL/xANuhQwebkJBgBw4caGvUqGFDQkLstddea6dMmXLCccXExFjguD+B7N6920ZGRtratWvb0qVL2/r169thw4bZtLS0k74+v//+u+3Tp4+tXr26LV26tK1Xr5595JFHbEJCgl/fQYMGWcDu2LHDvvbaa7Zx48Y2JCTE1qlTxz722GP24MGDPv3nzJljIyMj7RVXXGErV65sy5cvbxs1amSHDh1q4+Li/MYvKCiwkydPtp06dbJVqlSxpUuXtnXq1LFt27a1o0aNsrt37/b03bVrlwXsoEGDjnuO8fHxNigoyAL21ltvLbZvcnKy/ec//2mbNm1qy5UrZytUqGAvu+wye/vtt9tPP/3U5ubm+vRPSkqy9913n61evbotW7asbdasmZ04caJnfqOioo4bX2GdO3e2gA0ODrb79u0L2Cc3N9e+/vrr9oorrrBly5a1tWrVsvfcc4+Ni4vzzNOuXbs8/U/kenXo0CHgZ+7777+3rVq1shUrVrShoaG2a9eu9pdffrETJ070u5+stbZBgwa2QYMGAY8RFRVlARsTE+PTfrz7VERERERERET+sM5KXtVYa89mXvu8N2HCBAswePDgcx3Kn5Yxhg4dOpxSDV4RERERERERERHxYc7GQVSjWURERERERERERERKRIlmERERERERERERESkRJZpFREREREREREREpESCz3UAIoWpbriIiIiIiIiIiMiFRSuaRURERERERERERKRElGgWERERERERERERkRJRollEROT/2bvz8LjPs9D731uSLVu2JFuW5CVqSM1JapPQFCh047RJCoWSt0tMC30ph7Y0KQRe2tKXlK3lklMOAXJBSYET2qallPbQ9qVJKAfKgUOTLrSl8iLLsqV4t6x4kbfIsmXZluZ5/5iROrK1jTRa/f1c11wzv+WZua3bv9/M755nnkeSJEmSJE2JhWZJkiRJkiRJ0pRYaJYkSZIkSZIkTYmFZkmSJEmSJEnSlFholiRJkiRJkiRNiYVmSZIkSZIkSdKUWGiWJEmSJEmSJE2JhWZJkiRJkiRJ0pRYaJYkSZIkSZIkTYmFZkmSJEmSJEnSlFholiRJkiRJkiRNyZwtNEfEqoi4NyKeiIh9EXExIroj4hsR8c6IKLlq/5siIo1x+9xs/VskSZIkSZIkaSErm+0AxvBm4FHgGPAU0AGsBjYBjwGvjYg3p5TSVe12AE+O8Hyt0xirJEmSJEmSJF235nKheQ/weuCfUkqZwZUR8TvAd4CfJlt0/uJV7ZpTSo0zFaQkSZIkSZIkXe/m7NAZKaWvpHlye9QAACAASURBVJT+Mb/InFt/HPir3OIdMx6YJEmSJEmSJGmYudyjeSxXcvf9I2xbFxG/BKwCTgPfSim1zFhkkiRJkiRJknSdmXeF5ogoA34ht/gvI+zy47lbfpungbellDqmNzpJkiRJkiRJuv6UNjY2znYMBdm8efMfA28E/jml9GDe+sVAH/A+4LeAPwK+DNwE3Am8bvPmzZ9sbGy8cs2TAhHxrs2bN3/07rvvXgdwyy230NbWRm1tLdu3b2fdunVs2bKFVatWsWvXLsrLy3n22Wfp7e3lwoULHD16lMWLF7N7927q6urYtm3bsDZtbW0sWrSIo0ePcuHCBS5evEhnZyfl5eW0trayevXqYW1WrlzJM888Q2lpKcePH6enp4e+vj6OHDnC0qVLaWlpYc2aNWzdunWozYoVK9izZw8RQVdXF+fOnePy5cscPnyYiooKWlpaWLduHU1NTdxwww1s2bKF6upq9u3bR0qJU6dOcfbsWfr7+zl06BDLly9nx44dQ88/+HpVVVXs37+fTCbDmTNnOH36NJlMhgMHDlBZWTmszerVq9m+fTuVlZUcPHiQ/v5+zp49y6lTpwDYt28fK1asoLm5eahNfX09O3bsYNmyZRw6dIgrV67Q3d1NV1cXEcGePXtYuXLlsDZ1dXW0tLRQUVFBR0cHly5d4ty5c5w4cYLS0lLa29tZtWoV27ZtG/q319bW0traSnl5OZ2dnfT19XH+/HmOHTtGWVnZUP4H2zQ1NVFbW8vu3buHcnl1/nft2kV9ff2wXNbU1NDW1kZZWdlQLq/O/9W5HMx/SUkJJ06c4Ny5c1y6dImOjo6h/K9du3bU/J88eZLnnnuOK1euXJP/LVu2DN1XVVVdk/+BgQEOHDjA8uXLaW5uHvp7rVmzhm3bto2Z/6qqqqG8NDU1sWbNGpqbm1m2bNmI+d+7d++o+a+oqODw4cNcvnz5mvzX1NQMOy7r6urYuXMnS5YsoaOjg76+Pnp6ejh+/Piw/Oe3yc//4LF8/vx5jh49yqJFi2hraxvxWM7Pf/6xnJ//rVu3Dv3dampqaG9vH/VY3rlz57D8NzU1jZj/wWN5MP9X53LwWAaG5X+kY3nw/85g/jOZDKdPnx6W/6uP5cH85x/L+fnfv38/1dXVw3K5evXqofznH8snT54clv/t27cPHWP19fWjHsslJSXX5H+wzWD+jxw5Muqx7Lncc7nncs/lnss9l3su91zuudxzuedyz+Weyz2XX4/n8rVr126eRBm2YJFSmonXKYqIeDfwCNAOvCKldGYCbcqAbwAvAd6bUnpkrP0fffTRBHD//fdPPWBJkiRJkiRJml0xEy8yZycDvFpE/CrZIvNu4M6JFJkBUkr9wGO5xVdOU3iSJEmSJEmSdN2aF4XmiHgv8BdAK9ki8/ECn+Jk7n5ZUQOTJEmSJEmSJM39QnNE/CbwYaCZbJG5axJP89Lc/YGiBSZJkiRJkiRJAuZ4oTkiPgj8IbAVeHVK6dQY+74kIhaPsP4u4Ndzi5+ZlkAlSZIkSZIk6TpWNtsBjCYi3gY8CAwAXwfeHXHNuNWHUkqfyj3+I+DWiHga6MyteyFwV+7xB1NK35zOmCVJkiRJkiTpejRnC83A83P3pcB7R9nnq8Cnco//FrgH+GHgtcAi4ATwBeAvUkpfn7ZIJUmSJEmSJOk6NqlCc0QsA24AaoGLQFdK6dliBpZSagQaC9j/E8AnihmDJEmSJEmSJGl8Ey40R8T3Ar8I/DjwA1w1vnNEnAaeAr4IPJ5S6i9inJIkSZIkSZKkOWrcQnNEvBj4feDHyBaXrwA7gePAGWApsAp4AfBm4E3AyYh4BPjTlNKl6QldkiRJkiRJkjQXjFlojoi/Ad4KdAMfAz4HfCel1DfK/jcBPwG8DfjvwC9HxH9LKX2tiDFLkiRJkiRJkuaQknG2vwZ4H7AupfQrKaWvjVZkBkgpHUopfTSl9HLgRcA24FXFC1eSJEmSJEmSNNeMN3TG+pTSxck8cUqpBbgnIpZOpr0kSZIkSZIkaX4Ys0fzZIvMxX4OSZIkSZIkSdLcNd7QGRMSEYsi4gci4gXFeD5JkiRJkiRJ0vxRUKE5In4mIr4QETV5674X2AVsAXZHxOMRMd6QHJIkSZIkSZKkBaLQHs2/CGxIKZ3JW/cnwH8BngJagDcA7yhOeJIkSZIkSZKkua7QQvP3AU2DCxFRBfwU8IWU0o8BPwK0Y6FZkiRJkiRJkq4bhRaa64BjecsvA8qAzwGklK4A/wZ8b1GikyRJkiRJkiTNeYUWmnuA6rzlVwEJ+Ebeuj6gcopxSZIkSZIkSZLmiUIn7dsLvDYiyskWmN8MtKSUTuXt8z1AV5HikyRJkiRJkiTNcYX2aP4YsJ5swbkt9/iTV+3zEmDX1EOTJEmSJEmSJM0HBRWaU0p/A/whUEF2CI2/yN0AiIi7gJuAp4oXoiRJkiRJkiRpLit06AxSSr8D/M4om78BrAQuTCUoSZIkSZIkSdL8UXCheSwppcvA5WI+pyRJkiRJkiRpbhtz6IyIWDrVFyjGc0iSJEmSJEmS5q7xxmg+GBHviYjyQp84Im6PiH8AfmNyoUmSJEmSJEmS5oPxCs3/CvwpcCwiHo2IO8fqoRwR6yPi/oj4FrANuB0nBpQkSZIkSZKkBW3MMZpTSr8QER8B/gB4V+42EBFtwDHgLLAEWAW8AKgFAjgB/C7w4ZTSpekLX5IkSZIkSZI028adDDCltAV4TUTcDLwTeDXwIuD7r9r1JPA48EXgiymlK0WOVZIkSZIWhIfb+8fc/sCGos7bLkmSNO0m/OklpbQX+C2AiKgAbiDbk/ki0JVSOjYtEUqSJEnSAnb2chp6vHJxzGIkkiRJkzepr8lTSr3A3txNkiRJklSA/B7L+b2b7cksSZLmq/EmA5QkSZIkSZIkaUwWmiVJkiRJkiRJU2KhWZIkSZIkSZI0JRaaJUmSJEmSJElTYqFZkiRJkiRJkjQlFpolSZIkSZIkSVNSUKE5Im6MiKpx9qmMiBunFpYkSZIkSZIkab4otEfzQeA94+zz7tx+kiRJkiRJkqTrQKGF5sjdJEmSJEmSJEkCpmeM5tXAhWl4XkmSJEmSJEnSHFQ23g4R8QtXrXrRCOsASoEbgf8G7CxCbJIkSZIkSZKkeWDcQjPwKSDlHifgDbnb1QaH1OgFNk85MkmSJEmSJEnSvDCRQvM7cvcBfBJ4EviHEfYbAE4D30opPVec8CRJkiRJkiRJc924heaU0t8MPo6ItwFPppQ+Pa1RSZIkSZIkSZLmjYn0aB6SUrpzugKRJEmSJEmSJM1PJYXsHBE/GBG/FxGrR9m+Jrf9RcUJT5IkSZIkSZI01xVUaAZ+A7gX6Bpl+wngncD7phKUJEmSJEmSJGn+KLTQ/DLgqZRSGmljbv1XgFdMNTBJkiRJkiRJ0vxQ0BjNwBqgc5x9jgJrJxeOJEmSJKlQD7f3j7n9gQ2FXvpJkiQVptAezb1A3Tj71AGXJhfOd0XEqoi4NyKeiIh9EXExIroj4hsR8c6IGDH2iHh5RPxzRJyJiN6IaImI90ZE6VRjkiRJkqS57uzlNHSTJEmaKYV+rd0MvCEi3pdSOn/1xoioAt6Q22+q3gw8ChwDngI6gNXAJuAx4LUR8eb8YTwi4g3AF4E+4PPAGeB1wIfJDufx5iLEJUmSJElzSn6P5fzezfZkliRJM6XQHs0fI9tj+d8i4oX5GyLiduBfgdrcflO1B3g90JBSemtK6bdTSr8IbACOAD9Ntug8+PpVwMeBAeCOlNI7U0oPAC8CvgW8KSLeUoS4JEmSJEmSJEl5Cio0p5Q+D3waeAmwPSKORkRTRBwFtgE/Anw6pfR3Uw0spfSVlNI/ppQyV60/DvxVbvGOvE1vIlsE/1xKaUve/n3AB3KL9081LkmSJEkqls7eRNelxPG+7H1nr8NdSJKk+ang31GllN4eEd8Efg24lewEgQCtwEdSSo8VMb7RXMnd5894cVfu/l9G2P9rZMeXfnlElKeUpjyGtCRJkiRNVmt3hsc7M7T3DC8sv7+lnw2VwaaGEm6rLvQHqJIkSbNnUp9cUkofSyl9P7AcaACWp5ReOBNF5ogoA34ht5hfVH5B7n7P1W1SSv3AQbKF9fXTGqAkSZIkjeGprgwPtQ1cU2Qe1N6TeKhtgKe7MiNulyRJmotKGxsbJ924sbHxSmNjY09jY+OV8fcujs2bN/8x8Ebgn1NKD+atfwBYBTzc2Nh4ZoR2bwWeB3y2sbGx8+rtEfGuzZs3f/Tuu+9eB3DLLbfQ1tZGbW0t27dvZ926dWzZsoVVq1axa9cuysvLefbZZ+nt7eXChQscPXqUxYsXs3v3burq6ti2bduwNm1tbSxatIijR49y4cIFLl68SGdnJ+Xl5bS2trJ69ephbVauXMkzzzxDaWkpx48fp6enh76+Po4cOcLSpUtpaWlhzZo1bN26dajNihUr2LNnDxFBV1cX586d4/Llyxw+fJiKigpaWlpYt24dTU1N3HDDDWzZsoXq6mr27dtHSolTp05x9uxZ+vv7OXToEMuXL2fHjh1Dzz/4elVVVezfv59MJsOZM2c4ffo0mUyGAwcOUFlZOazN6tWr2b59O5WVlRw8eJD+/n7Onj3LqVOnANi3bx8rVqygubl5qE19fT07duxg2bJlHDp0iCtXrtDd3U1XVxcRwZ49e1i5cuWwNnV1dbS0tFBRUUFHRweXLl3i3LlznDhxgtLSUtrb21m1ahXbtm0b+rfX1tbS2tpKeXk5nZ2d9PX1cf78eY4dO0ZZWdlQ/gfbNDU1UVtby+7du4dyeXX+d+3aRX19/bBc1tTU0NbWRllZ2VAur87/1bkczH9JSQknTpzg3LlzXLp0iY6OjqH8r127dtT8nzx5kueee44rV65ck/8tW7YM3VdVVV2T/4GBAQ4cOMDy5ctpbm4e+nutWbOGbdu2jZn/qqqqobw0NTWxZs0ampubWbZs2Yj537t376j5r6io4PDhw1y+fPma/NfU1Aw7Luvq6ti5cydLliyho6ODvr4+enp6OH78+LD857fJz//gsXz+/HmOHj3KokWLaGtrG/FYzs9//rGcn/+tW7cO/d1qampob28f9VjeuXPnsPw3NTWNmP/BY3kw/1fncvBYBoblf6RjefD/zmD+M5kMp0+fHpb/q4/lwfznH8v5+d+/fz/V1dXDcrl69eqh/OcfyydPnhyW/+3btw8dY/X19aMeyyUlJdfkf7DNYP6PHDky6rHsudxzuedyz+Weyz2Xz4VzeUf/Er70rR10V67lxmPbhu4vLqmi7uwBILGs9zRLLz1H8+nLLDl9mLUrCjuXt+7ZT2ZJFYs6WnjpzTd4Lvdc7rncc7nncj+Xey6/zs/la9eu3VxQAXaSIqXCxwCLiDqyk/FtBJallO7NW/98YGdK6WIxA809/7uBR4B24BUppTN52/YANwM3p5T2jdD2m8DLgJellL492ms8+uijCeD++x3OWZIkSVJxPbirf9SezCPZWBl88NbCRjx8uP27Iww+sKHg0RIlSdLCEzPxIgUPnRER7wQOAX9Jdpzmd+RtXg18C/i5YgR31ev+Ktki827gzvwic0537r56lKeoumo/SZIkSZoxnb2poCIzQFuPEwRKkqT5oaBCc0T8OPAxsuMg3wM8mr89pdQK7CI7tEXRRMR7gb8gO+HgnSml4yPs9kzu/pYR2peR7WndDxwoZmySJEmSNBGt3ZMbc3my7SRJkmZSoT2afxM4BrwqpfQloGuEfVqA75tqYIMi4jeBDwPNZIvMI70mwFdy9z85wrZXAhXAN1NKl4oVmyRJkiRN1MWBmW0nSZI0kwotNL8Y+F8ppXNj7NMJrJl8SN8VER8E/hDYCrw6pXRqjN3/HjgFvCUiXpz3HEuA388tPjpSQ0mSJEmabktLZ7adJEnSTCp0ZojFwIVx9lkBTPk794h4G/Bg7rm+Drw74ppxqw+llD4FkFI6FxH3kS04Px0RnwPOAK8HXpBb//mpxiVJkiRJk3FbdQlQ+DAY2XaSJElzW6GF5kPAD42zz0v47njJU/H83H0p8N5R9vkq8KnBhZTSkxHxKuB3gZ8GlgD7gPcBH0kpOYuGJEmSpFnRUBFsqIyCJgTcWBk0VMzIRPGSJElTUuhX4/8A/NeIePNIGyPiHcALgS9ONbCUUmNKKca53TFCu/9IKf1USmllSmlpSun7U0ofTik5spkkSZKkWbWpoYSJlo0DuKfB3sySJGl+KPRTyx8DHcDfRcTngZcBRMT/k1v+GLAX+POiRilJkiRJC8Bt1SXcu7503GJzAPetL3XYDEmSNG8UNHRGSulsbmiKTwP5vZo/krv/OvBzKaXxxnGWJEmSpOvSnfUl1JXDE50Z2kYYRmNjZXBPQ4lFZkmSNK8UOkYzKaUO4I6IeCHZHs2rgG7g2ymlrUWOT5IkSZIWnNuqs4Xkzt7EI3v7ySQoCXjPzWWOySxJkualMQvNEfE48LmU0hdyy68EDqWUOlJKLUDLDMQoSZIkSQtSQ0VQXx7DliVJkuaj8X6L9UZgQ97yU8Dbpy0aSZIkSZIkSdK8M16h+TmgKm/Zr9clSZIkSZIkScOMN0ZzG/B/R0QTcCy37qbcEBpjSil9barBSZIkSZIkSZLmvvEKzZuBJ4H/mbfubbnbeEonG5QkSZo5D7f3j7n9gQ0Fzx0sSZIkSbrOjHnlmFL614jYCPwYcAPQCHw1d5MkSQvM2ctp6PHKxY6YJUmSJEmamHG7KKWUDgOfAIiIRuDplNKD0xyXJEmaIfk9lvN7N9uTWZIkSZI0UWNOBhgRfxoRr8lb9Q6yQ2lIkiRJkiRJkgSMU2gG3gu8NG/5k8Abpy8cSZIkSdJkdfYmui4ljvdl7zt70/iNJEmSimC838SeByrylh2sUZIkSZLmmNbuDI93ZmjvGV5Yfn9LPxsqg00NJdxWPV4/I0mSpMkbr9C8D9gUEU8Ax3LrVkTEjeM9cUqpY6rBSZIkSZLG9lRXhscODDBa3+X2nsRDbQPctx7uqLfYLEmSpsd4heaHgc8A38xb957cbSxpAs8tSZIkSZqC1u6xi8yDEvDxAwPUlmPPZkmSNC3GLAanlP4uIg4CdwM3AG8HWoDm6Q9NkiRJkjSWxzsz4xaZByXgic6MhWZJkjQtxu11nFL6NvBtgIh4O/BESunBaY5LkiRJkjSGzt50zZjM42nryU4Q2FDh9DuSJKm4Cv0q+x3Ak9MRiCRJkiRp4lq7MzPaTpIkaSwFjaOcUvqb6QpEkiRJkjRxFwdmtp0kSdJYCh6cKyJKIuLXIuLbEdEdEf15234gIv5HRNxS3DAlSZIkSfmWls5sO0mSpLEUVGiOiMXAvwF/Bnwv0APkD+51EPhF4K3FClCSJEmSdK3JTurnZICSJGk6FPoJ4wHgTmAzsBp4LH9jSuk54GvATxQlOkmSJEnSiBoqgg2VhU3qt7EynAhQkiRNi0ILzW8F/iOl9GBKKQOMNMXxQeDGKUcmSZIkSRrTpoYSJlo2DuCeBnszS5Kk6VHop4znA98eZ58zQM3kwpEkSZIkTdRt1SXcu7503GJzAPetL3XYDEmSNG3KCtz/IrBinH1uBJ6bXDiSJEmSpELcWV9CXTk80ZmhrefaH51urAzuaSixyCxJkqZVoYXmZuA1EbE4pXT56o0RUU12fOZvFiM4SZIkSdL4bqvOFpI7exOP7O0nk6Ak4D03lzkmsyRJmhGFfqX9ceB5wGcjoip/Q0SsAD4FrAT+qijRSZIkSZImrKEiqC8P1izJ3ltkliRJM6WgHs0ppb+LiB8D3gG8HjgLEBFbgFuBcuAvU0r/XOxAJUmSJEmSJElzU8GDdKWU3gn8IrAbqCM7r8QPAvuAd6aUfq2oEUqSJEmSJEmS5rRCx2gGIKX0KeBTEbGU7FAZ3SmlC8UMTJIkzazO3kTXpTQ0rmdnb/In15IkSZKkCZlUoXlQSukicLFIsUiSpFnQ2p3h8c4M7T1p2Pr3t/SzoTLY1JCdYEqSZtvD7f1jbn9gw5Qub2ZU/r9l//l0zfr59G+RJEmCKRaaJUnS/PZUV4bHDgyQRtne3pN4qG2A+9bDHfUWmyXNHWcvf/fMtXLx/P71Rc3i2Y5AkiRp6iw0S5J0nWrtHrvIPCgBHz8wQG059myWNKvye/nm9wiej71/52PMkiRJY/FqUZKk69TjnZlxi8yDEvBEZ2Y6w5EkSZIkzWN+jS5JWpAW0jie06GzN10zJvN42nqSEwRKkiRJkkZkj2ZJ0oJ39nIauimrtXtyvZMn206SJEmStLBd3925JEkL1kIax3M6XByY2XaSJEmSpIWtKFfbEVEL/FegF/g/KSUvQyVJmsOWls5sO0mSJEnSwlZQoTki7gfeDrw2pXQmt+6HgH8BanK7bYmIu1JKF4oZqCRJKp7bqkuAwofByLaTJC0UzmkgSZKKpdCrxZ8F0mCROedhYCXw18A/Az8M/HJxwpMkSdOhoSLYUFnYpH4bK8OJACVpAXNOA0mSNBWFfj19M/BPgwu5ITNeBTyWUvql3Lr/BH4O+JNiBSlJkopvU0MJD7UNMJFyQgD3NNibWZIWGuc0kCRJxVLoFeMqoCtv+RW5+yfy1n0d+J6pBCVJkqbfbdUl3Lu+lPH6KAdw3/pSh82QJEmSJI2q0K+pzwC1ecuvIjvA4zfz1iVgyRTjkiRJM+DO+hLqyuGJzgxtPdf2bd5YGdzTUGKRWZIkSZI0pkILzW3A6yLid4EBsmM2N6WUzuXtcxNwvBjBRcSbyBazXwTcDlQCn00p/fwI+94EHBzj6T6fUnpLMeKSJGkhua06W0ju7E08srefTIKSgPfcXOaYzJLmpM7eRNelNHS+6uxNnq8kSZJmWaGF5keAJ4FOoB+oAH5zcGNElAI/yvAezlPxAbIF5vO519wwgTY7cjFerbVIMUmStCA1VAT15TFsWZLmktbuDI93Zmi/6hcY72/pZ0NlsMlfYEiSJM2aggrNKaUvRcQvA+/KrfpsSukzebv8GNlhM/53keL7dbIF5n1kezY/NYE2zSmlxiK9viRJkqQ54KmuDI8dGH0C0/aexENtA9y3Hu6ot9gsSZI00wqeSjil9DHgY6Ns+9/AyqkGlfd8Q4XlCHtVSZIkSdej1u6xi8yDEvDxAwPUlmPPZkmSpBlWcKF5HlgXEb8ErAJOA99KKbXMckySJEmSJunxzsy4ReZBiewEpxaaJUmSZlZBn74i4gci4lciojpv3bKI+JuIeC4ijkbEe4ofZkF+HPgr4L/n7ndExFMRcePshiVJkiSpUJ296ZoxmcfT1pPo7C2sjSRJkqamtLGxccI7b968+c+An08p/UHeuj8B7gOuAJXAazdv3vztxsbG/cUMdPPmzTcBbwd2NjY2Pj7C9sVAH/A+4LeAPwK+DNwE3Am8bvPmzZ9sbGy8MtLzR8S7Nm/e/NG77757HcAtt9xCW1sbtbW1bN++nXXr1rFlyxZWrVrFrl27KC8v59lnn6W3t5cLFy5w9OhRFi9ezO7du6mrq2Pbtm3D2rS1tbFo0SKOHj3KhQsXuHjxIp2dnZSXl9Pa2srq1auHtVm5ciXPPPMMpaWlHD9+nJ6eHvr6+jhy5AhLly6lpaWFNWvWsHXr1qE2K1asYM+ePUQEXV1dnDt3jsuXL3P48GEqKipoaWlh3bp1NDU1ccMNN7Blyxaqq6vZt28fKSVOnTrF2bNn6e/v59ChQyxfvpwdO3YMPf/g61VVVbF//34ymQxnzpzh9OnTZDIZDhw4QGVl5bA2q1evZvv27VRWVnLw4EH6+/s5e/Ysp06dAmDfvn2sWLGC5ubmoTb19fXs2LGDZcuWcejQIa5cuUJ3dzddXV1EBHv27GHlypXD2tTV1dHS0kJFRQUdHR1cunSJc+fOceLECUpLS2lvb2fVqlVs27Zt6N9eW1tLa2sr5eXldHZ20tfXx/nz5zl27BhlZWVD+R9s09TURG1tLbt37x7K5dX537VrF/X19cNyWVNTQ1tbG2VlZUO5vDr/V+dyMP8lJSWcOHGCc+fOcenSJTo6Oobyv3bt2lHzf/LkSZ577jmuXLlyTf63bNkydF9VVXVN/gcGBjhw4ADLly+nubl56O+1Zs0atm3bNmb+q6qqhvLS1NTEmjVraG5uZtmyZSPmf+/evaPmv6KigsOHD3P58uVr8l9TUzPsuKyrq2Pnzp0sWbKEjo4O+vr66Onp4fjx48Pyn98mP/+Dx/L58+c5evQoixYtoq2tbcRjOT//+cdyfv63bt069Herqamhvb191GN5586dw/Lf1NQ0Yv4Hj+XB/F+dy8FjGRiW/5GO5cH/O4P5z2QynD59elj+rz6WB/Offyzn53///v1UV1cPy+Xq1auH8p9/LJ88eXJY/rdv3z50jNXX1496LJeUlFyT/8E2g/k/cuTINcfyt1va2DFQQ9mhZs4sX0vZwW2sq6/1XD7Ksdy6Zz+ZJVUs6mjhpTff4Lncc7nncs/ls34ub+0p4eyhZ7hQsYrnHW+mu3ItNx7bxvmKVaw91UZ/6SJW9BxjcX8vi6/0Un3+GAMli1h0tI0NDfWeyws4l+/LZNtkVq6l/NA2z+Weyz2Xey4v6udyayzWWDyXz965fO3atZsnWIKdkkhp4t/0R8Q+4D9TSm/NLS8CTgFtwB1ADbAdaEop/V9FDTTiDrKTAX42pfTzBbQrA74BvAR4b0rpkbH2f/TRRxPA/fffP/lgJUmzrrU7w+OdmRF7wW2oDDY1lPiz6hE83N4/9PiBDQtxhC1J880TnQP8f52Zgtu9uaGEexpKpyGiuW+y53LfAyRJWrBmZPK7Qq+w64EjecsvJtuL+aMppb6U0lHgH4AXFim+KUsp9QOP5RZfOZuxSJJmxlNdGR5qGxj1p9btPYmH2gZ4uqvwYZ+tDQAAIABJREFUwoUkaWYtnWSteLLtJEmSNDmFFpoTwycQ/NHcuq/mrTsJ1E0xrmI7mbtfNqtRSJKmXWt3hscODIw7aVQCPn5ggNZui82SNJdN9tcn/mqlMJ29ia5LieN92XvHuJYkSYUq9PdQHcBL85bfAHSmlA7krVsHnJ1qYEU2GPOBMfeSJM17j3dmxi0yD0rAE50ZixGSNIc1VAQbKqOgCQE3VgYNFTPyC9F5b7Shpt7f0u9QU5IkqSCFfmL4AvDyiPj7iPgM8DLg76/a5zagqBMBTkREvCQiFo+w/i7g13OLn5nZqCRJM6mzNxVUiABo67HXliTNdZsaSiY8sGAA9zRYGJ0Ih5qSJEnFVGiP5g8DPwlsyi03Aw8OboyI7wN+CPiDYgQXEW8E3phbXJO7f1lEfCr3+FRK6Tdyj/8IuDUingY6c+teCNyVe/zBlNI3ixGXJGlumuwwGK3dGRoqHMxTkuaq26pLuHc94w6NFMB960vtgTsBhQ41VVvucCSSJGlsBRWaU0rngVdExG25VbtTSvlX9b3APcCWIsX3IuBtV61bn7sBHAYGC81/m3vtHwZeCywCTpDthf0XKaWvFykmSdIcdXFgZttJkmbOnfUl1JVnhzxqG6EH7sbK4B6HeZgwh5qSJEnFVmiPZgBSSq2jrD8EHJpCPFc/XyPQOMF9PwF8olivLUmaf5ZOslPyZNtJkmbWbdXZQnJnb+KRvf1kEpQEvOfmMsdkLsBUhpry7yxJkkYzqUJzRCwCXg1sBJanlD6UW78EqCI7pIUDeUmSZlS2p1Xhbz/20JKk+aWhIqgvj2HLmjiHmpIkSdOh4CvriPhJsr2W/wn4E4b3OH4RcAz42SLEJklSQRoqgg2VhRUbNlaGBQpJ0nXFoaYkSdJ0KKjQHBEvBp4kO0zXrwP/M397SunbwEGyYyVLkjTjNjWUMNGycQD3NNibWZJ0fXGoKUmSNB0Kvbr+INkJ/16cUvoIsHeEfZqA26camCRJk3FbdQn3ri8dt9gcwH3rSx02Q5J03Znse5/vmZIkaSyFjtH8CuDJlNLxMfY5Atw9+ZAkSZqaO+tLqCuHJzoztI0w2dHGyuCehhIvmCVJ16XBoaYKmRDQoaYkSdJ4Ci00LwdOjbNPBZMY+1mSpGK6rTpbSO7sTTyyt59MgpKA99xc5oWyJOm6t6mhhIfaBphIqdmhpiRJ0kQU+mnhWeDWcfZ5EXBgcuFIklRcDRVBfXmwZkn23iKzJEkONSVJkoqv0E8LXwZ+IiJ+dKSNEfFa4OXA/5pqYJIkSZKk6XNnfQm/vbGUjZUjl5s3Vga/vbGUO+otMkuSpPEVOnTGQ8BbgH+NiD8HbgKIiLuBVwK/ChwD/rSIMUqSJEmSxvBwe//Q4/3n0zXrH9gw8qWfQ01JkqRiKajQnFJ6NiJeA3wBeCBv05fI/qpqP7AppTTeOM6SJEmSpGlQs7jwNoNDTeUvS5IkFaLQHs2klLZFxAuAu4GXAauAbuDbwD+klPrHai9JkiRJKq7ReixLkiTNlEl9GkkpDZDtxfyl4oYjSZIkSZIkSZpvnNVBkiRJkiRJkjQlk+rRHBEvBn4EWAmUjrBLSil9aCqBSZIkSZIkSZLmh4IKzRFRBTwO3El28r/RJMBCsyRJkiRJkiRdBwrt0fwwcBfwdeCvgSOAk/9JkiRJkiRJ0nWs0ELzG4BtwJ0ppcw0xCNJkiRJkiRJmmcKLTRXA39rkVmSJEnSTHu4/bs/ptx/Pl2z/oENk5qCRpIkSUVQ6CexvcDq6QhEkiRJkiaqZvFsRyBJkqR8hRaa/xL4w4i4IaX07HQEJEmSJEkjsceyJEnS3FXoJ7Uvk50M8D8iYjOwFXhupB1TSh1TjE2SJEmSJEmSNA8UWmg+BCQggMfG2C9N4rklSZIkSZIkSfNQocXgT5MtIkuSJEmSJEmSBBRYaE4pvX2a4pAkSZIkSZIkzVMObyFJ0nXu4fb+ocf7z6dr1jv5liRJkiRpPAVdOUbEANCYUvrQGPv8LrA5peRVqSRJ80zN4tmOQNL1IP8LrpH4BZckSdL8U+gnuMjdJrKfJEmaByzoSJpNZy9/95cUKxd7GSFJkjRfTceV5UqgbxqeV5IkSdICkP8FV37vZr/4kiRJmr/G/SQXEa+8atVNI6wDKAVuBN4KPFOE2CRJkiRJkiRJ88BEugw8DQz+ni0Bb8vdRhJABvh/pxyZJEmSJEmSJGlemEih+UGyBeYAfo9s4fmrI+w3AJwGnkoptRcrQEmSJEmSJEnS3DZuoTml1Dj4OCLeBjyZUvrIdAYlSZIkSZIkSZo/CpptI6X0/OkKRJIkSZIkSZI0PzmtsyRpQXq4vX/o8f7z6Zr1D2zwLVCSJEmSpGIZ8yo7Ir5CbgLAlFJnbnkiUkrp1VOOTpKkIqhZPNsRSJIkSZK0sI3XneuO3H3FVcvjSePvIknS9LHHsiRJkiRJM2fMq/CUUslYy5IkSZI0WZ29ia5LiUyCksguN1TEbId1XXGoKUmSVCzjDZ3xbuDbKaXvzFA8kiRJkha41u4Mj3dmaO8Z/kPI97f0s6Ey2NRQwm3V9nGZaQ41JUmSpmK8r6f/DGgEvgMQEQNAY0rpQ9MclyRJkqQF6KmuDI8dGBh1rL32nsRDbQPctx7uqLfYPN3ssSxJkoplvE9ufUB53nLkbpIkSZJUkNbusYvMgxLw8QMDtHZnZiIsSZIkFcF4X18fBH4iIj6SUjqRW+dEf5IkSZIK9nhnZsIXEwl4ojPjEBpzRP5YziOxZ7QkSRrvU9tHgR8EjuaGzQBojIiBcW5jfwqRJEmSdF3p7E3XjMk8nraeRGev/VzmmrOX09BNkiRp0JhfO6eUPhIRXcDdwDrgTqADODTdgUXEm4BXAS8Cbgcqgc+mlH5+jDYvBz4AvBRYAuwDPgn8eUppYLR2kiRJkqbXZIfBaO3O0FBRWuRoVKj8Hsv5vZvtySxJkgaN+6kgpfQ54HMAEZEB/jql9OB0B0a2YHw7cB7oBDaMtXNEvAH4ItlxpT8PnAFeB3wYeAXw5ukMVpIkzbzX/f6Xxtz+jx94/QxFImk8FyfZ7WOy7SRJkjSzCv36eTPw9DTEMZJfJ1tg3ke2Z/NTo+0YEVXAx4EB4I6U0pbc+g8CXwHeFBFvyRXNJUnSAvTs6QtDj29YtWwWI5E0kqWT7JQ82XaaHp29ia5LiUyCksguN1Q4X7wkSSqw0JxS2jxdgYzwWkOF5YhxP7i8CagDPj1YZM49R19EfAD4d+B+cj2zJUnSwpDfYzm/d7M9maW5JzupX+HDZzgZ4NzQ2p3h8c7MNeNsv7+lnw2VwaaGEnMlSdJ1bqF8Ergrd/8vI2z7GtALvDwiymcuJEmSJEmDGiqCDZWF9XzdWBn2lp0DnurK8FDbwKiTObb3JB5qG+DprsmNwy1JkhaGhVJofkHufs/VG1JK/cBBsr23189kUJIkSZK+a1NDCRMtGwdwT8NCuVyZv1q7Mzx2YICRS8zflYCPHxiY9KSPkiRp/lson9yqc/fdo2wfXL9iBmKRJEmSNILbqku4d33puMXmAO5bX+pQDHPA452ZcYvMgxLwRKeFZkmSrleljY2Nsx3DuDZv3nwT8HZgZ2Nj4+MjbP81YBXw542NjWdG2H4v8DzgscbGxs6RXiMi3rV58+aP3n333esAbrnlFtra2qitrWX79u2sW7eOLVu2sGrVKnbt2kV5eTnPPvssvb29XLhwgaNHj7J48WJ2795NXV0d27ZtG9amra2NRYsWcfToUS5cuMDFixfp7OykvLyc1tZWVq9ePazNypUreeaZZygtLeX48eP09PTQ19fHkSNHWLp0KS0tLaxZs4atW7cOtVmxYgV79uwhIujq6uLcuXNcvnyZw4cPU1FRQUtLC+vWraOpqYkbbriBLVu2UF1dzb59+0gpcerUKc6ePUt/fz+HDh1i+fLl7NixY+j5B1+vqqqK/fv3k8lkOHPmDKdPnyaTyXDgwAEqKyuHtVm9ejXbt2+nsrKSgwcP0t/fz9mzZzl16hQA+/btY8WKFTQ3Nw+1qa+vZ8eOHSxbtoxDhw5x5coVuru76erqIiLYs2cPK1euHNamrq6OlpYWKioq6Ojo4NKlS5w7d44TJ05QWlpKe3s7q1atYtu2bUP/9traWlpbWykvL6ezs5O+vj7Onz/PsWPHKCsrG8r/YJumpiZqa2vZvXv3UC6vzv+uXbuor68flsuamhra2tooKysbyuXV+b86l4P5Lykp4cSJE5w7d45Lly7R0dExlP+1a9eOmv+TJ0/y3HPPceXKlWvyv2XLlqH7qqqqa/I/MDDAgQMHWL58Oc3NzUN/rzVr1rBt27Yx819VVTWUl6amJtasWUNzczPLli0bMf979+4dNf8VFRUcPnyYy5cvX5P/mpqaYcdlXV0dO3fuZMmSJXR0dNDX10dPTw/Hjx8flv/8Nvn5HzyWz58/z9GjR1m0aBFtbW0jHsv5+c8/lvPzv3Xr1qG/W01NDe3t7aMeyzt37hyW/6amphHzP3gsD+b/6lwOHsvAsPyPdCwP/t8ZzH8mk+H06dPD8n/1sTyY//xjOT//+/fvp7q6elguV69ePZT//GP55MmTw/K/ffv2oWOsvr5+1GO5pKTkmvwPthnM/5EjR0Y9lj2XL/xz+WeebiOTydDfP8BLGxZ5Lvdc7rl8jp7LLz+7l5ffVMPFfTs4tnQtNx7bRk9FLetO7qK/bAm3XjnKptWX+S9lFzyXz/K5/EplHf+xdQeXF1VQc66D0oErLL3UzfLeU6SA+jP76V2ygued2EF3ZTaXB8vqqDnWysrlfi73XO65fCGfy/1cPn/O5dZYPJfv2LGDtWvXzsi8e5HSRL+fnj0RcQfwFPDZlNLPj7C9CXgx8OKU0tYRtrcCtwLfl1JqG+u1Hn300QRw//33FyFySZI0E3Z1nOZnH/4y/QMZykpL+PwDr+XWG1fNdliSxtHZm3hkbz+ZBCUB77m5zDGZ55B/OTbApw8X3kP5F76nhJ9cWzoNEUmSpEmakQ9YZTPxIjPgGbKF5luAYYXmiCgDng/0AwdmPjRJkjRd/n3HER78wn/ytV1Hh62/7d2f5ZW3ruP3fuYlvPr2581SdJLG01AR1JfHsGXNHRcHZradJEma3xbKoGdfyd3/5AjbXglUAN9MKV2auZAkSdJ0+sS/7eI1jU9eU2Qe9LVdR3lN45N88v/smuHIJGlhWDrJTsmTbSdJkua3hdKj+e+BPwLeEhF/nlLaAhARS4Dfz+3z6GwFJ0maeQ+394+5/YENC+Ut8Pr07zuO8K7/8RUy4wwBlkmJ+/7yK3xPXZU9myWpQNnJGAsfOsNJHCVJuj7N2avsiHgj8Mbc4prc/csi4lO5x6dSSr8BkFI6FxH3kS04Px0RnwPOAK8HXpBb//mZil2SNLecvfzdYuTKxf4seyF48Av/OW6ReVAmJT70he9YaJakAjVUBBsqg/aeic/rs7EyHAJFkqTr1JwtNAMvAt521br1uRvAYeA3BjeklJ6MiFcBvwv8NLAE2Ae8D/hImg+zHkqSiia/x3J+72Z7Ms9/uzpOjzpcxmi+uutZdnWcdoJASSrQpoYSHmobYCIXUwHc02Bv5ony11eSpIVmzn4KSCk1ppRijNtNI7T5j5TST6WUVqaUlqaUvj+l9OGUktNRSJK0QPx7y5EZbSdJ17Pbqku4d33puFPVB3Df+lKHzZiEiwOJwxcSBy9k7y8O2EdKkjQ/+RWpJGlB6+xNdF1KZBKURHbZn/TOb+d6L89oO0m63t1ZX0JdOTzRmaFthGE0NlYG9zSUWGQu0GvXlvB4Z4b2nuHrz1yBDZXQ2p3xbypJmlcsNEuSFqTW7kzu4m34BfH7W/rZUBls8oJ43qqqWDyj7SRJWaP1s7X/beGe6srw2IHRhyRp70k81DbAfevhjno/r0iS5gcLzZKkBceLt4Xt1S+c3KR+k20nSdc731eLq7V77L/noAR8/MAAteVM65fjjhUtSSoWPwVIkhaUQi/eWrszMxGWiujWG1fxylvXFdTmVbfe4ESAkjQJvq8W3+OdmQn3Ak9khyyZKWcvp6GbJEmFstAsSVpQ5vLFm4rn937mJZTExMbaLonggz/zI9MckSQtTL6vFldnb7pmWK/xtPUkOnunr/D7wIayodvKxTF0G1wnSdJEWWiWJC0Yc/HiTdPj1bc/j4/9yl3jFptLIvj4r97Fq2932AxJKpTvq8U32R7fM9FTfHAC5eN92XvzKEkqlF9PSpIWjKlcvDVUlBY5Gk23d/74rdxUX8WHvvAdvrrr2Wu2v+rWG/jgz/yIRWZJmiTfV4vv4sDMtpsIJ1CWJBWLhWZJ0oIxFy/eNL1effvzePXtz2NXx2l+9uEv0z+Qoay0hM8/8FrHZJakKfJ9tfiWTrL+Ptl243GiR0lSMVloliQtGHPt4k0z59YbV/H81VXDliVJU+P7avFlewYX3lN8OnoUFzrRY2359MQhSVo4fJeQJC0Yk7348aJJkqRr+b5afA0VwYbKiU1mO2hjZdBQUVibiXCiR0lSsfkJQJK0YMylizdJkuY731enx6aGEib6FwrgnobiX7Y70aMkaTpYaJYkLShz4eJNkqSFwvfV4rutuoR715eO+3cN4L71pdM2bMZMtpMkXR/8FCBJWlDmwsWbJEkLhe+r0+PO+hJ+e2MpG0fpMb6xMvjtjaXTNgGfEz1KkqaDkwFKkhacO+tLqCvPjiXYNsLPQjdWBvc0lHgxLEnSBPi+Oj1uq87+zTp7E4/s7SeToCTgPTeXTfvwI070KEmaDhaaJUkL0mxevEmStND4vjp9GiqC+vIYtjzdsl8KFD4Mhl8mSJLGYqFZkrSgzcbFmyRJC5Xvq8XzcHv/0OP959M16x/YMH2X64MTPRYyIaATPUqSxmOhWZIkSdKMms0CmzQX1Sye+dfc1FDCQ20DTKTU7ESPkqSJ8BOcJEmSpFkzGwU2aS6Y7S9UshM9wmMHxi42O9GjJGmiLDRLkiRJmlGzXWCTlOVEj5KkYvITniRJkiRJ1yknepQkFYuFZkmSJEmSrnNO9ChJmip//yJJkiRJkiRJmhILzZIkSZIkSZKkKbHQLEmSJEmSJEmaEsdoliRJ89brfv9LQ4+/s/fENev/8QOvn/GYJEmSJOl6ZKFZkrQgPdzeP/R4//l0zfoHNvgWuNDcULN8tkOQpAXL91VJkjQePw1Ikha8msWzHYGmiz2WJWnm+b4qSZJGYqFZkrQg2bNKkqTi8X1VkiSNx08LkiRJkiRdpxwWRZJULL5jSJIkSZIkh0WRJE2JhWZJkiRJkq5T9liWJBVLyWwHIEmSJEmSJEma3yw0S5IkSZIkSZKmxEKzJEmSJEmSJGlKLDRLkiRJkiRJkqbEQrMkSZIkSZIkaUosNEuSJEmSJEmSpsRCsyRJkiRJkiRpSiw0S5IkSZIkSZKmxEKzJEmSJEmSJGlKLDRLkiRJkiRJkqbEQrMkSZIkSZIkaUosNEuSJEmSJEmSpmTBFZoj4lBEpFFux2c7PkmSJEmSJElaaMpmO4Bp0g382Qjrz890IJIkSZIkSZK00C3UQvNzKaXG2Q5CkiRJkiRJkq4HC27oDOn/b+/O4+uq64SPf75N2nTf0o1SsIBQFKyiDKtUQFEQGDeYkREGHRwFwWFGfT04Omgr4uM8iqLWQXk5jiPwiMo8lnFAwEGhDqB0pXRf6L7ve9Mm+T1/nJuQhqRNepOc5Obzfr3O6+Se7X6T873f3n7vub8jSZIkSZIkqWOV6hXNFRFxPXAisBeYA0xNKdXkG5YkSZIkSZIklZ5SbTSPAh5otGx5RHwspfRsHgFJkiRJkiRJUqkqmzhxYt4xtKlJkyYNAu4C7gC+DDwMlAHvBD48adKk/5o4ceLGxvtFxCcmTZr0wyuvvHI0wGmnncaCBQsYNmwYs2bNYvTo0UyfPp3KykrmzZtHRUUFa9euZd++fezdu5d169bRq1cv5s+fz/Dhw5k5c+Zh+yxYsICePXuybt069u7dy/79+1mzZg0VFRXMnTuXkSNHHrbPkCFDWLRoEWVlZWzYsIHdu3dz4MABVq9eTZ8+fZgzZw6jRo1ixowZ9fsMHjyYxYsXExFs2rSJXbt2cfDgQVauXEnfvn2ZM2cOo0ePZtq0aRx//PFMnz6dQYMGsXTpUlJKbNmyhe3bt1NdXc2KFSvo378/L730Uv3x655v4MCBLFu2jNraWrZt28bWrVupra3llVdeYcCAAYftM3LkSGbNmsWAAQNYvnw51dXVbN++nS1btgCwdOlSBg8ezOzZs+v3GTFiBC+99BL9+vVjxYoVHDp0iJ07d7Jp0yYigsWLFzNkyJDD9hk+fDhz5syhb9++rFq1iqqqKnbt2sXGjRspKytj4cKFVFZWMnPmzPrffdiwYcydO5eKigrWrFnDgQMH2LNnD+vXr6e8vLz+/NftM23aNIYNG8b8+fPrz2Xj8z9v3jxGjBhx2LkcOnQoCxYsoLy8vP5cNj7/jc9l3fnv0aMHGzduZNeuXVRVVbFq1ar683/cccc1e/43b97Mjh07OHTo0GvO//Tp0+vnAwcOfM35r6mp4ZVXXqF///7Mnj27/u81atQoZs6cecTzP3DgwPrzMm3aNEaNGsXs2bPp169fk+d/yZIlzZ7/vn37snLlSg4ePPia8z906NDDXpfDhw/n5Zdfpnfv3qxatYoDBw6we/duNmzYcNj5b7hPw/Nf91res2cP69ato2fPnixYsKDJ13LD89/wtdzw/M+YMaP+7zZ06FAWLlzY7Gv55ZdfPuz8T5s2rcnzX/darjv/jc9l3WsZOOz8N/VarsuduvNfW1vL1q1bDzv/jV/Ldee/4Wu54flftmwZgwYNOuxcjhw5sv78N3wtb968+bDzP2vWrPrX2IgRI5p9Lffo0eM1579un7rzv3r16mZfy9Zya7m13FpuLbeWW8ut5dZya7m13FpuLbeWd8daftxxx00quunaApFS6ojnyV1EfBP4LDAlpfSB5ra77777EsAtt9zSUaFJkiRJkiRJUnuJjniS7nQzwB8U5hNyjUKSJEmSJEmSSkx3ajRvKsz75RqFJEmSJEmSJJWY7tRoPr8wfyXXKCRJkiRJkiSpxJRUozkizoiIoU0sfx0wufDwwY6NSpIkSZIkSZJKW3neAbSxa4HPR8TvgeXAbuAU4EqgN/A48M38wpMkSZIkSZKk0lNqjebfA+OAs8iGyugH7AD+B3gAeCCllPILT5IkSZIkSZJKT0k1mlNKzwLP5h2HJEmSJEmSJHUnJTVGsyRJkiRJkiSp49loliRJkiRJkiQVxUazJEmSJEmSJKkoNpolSZIkSZIkSUWx0SxJkiRJkiRJKoqNZkmSJEmSJElSUWw0S5IkSZIkSZKKYqNZkiRJkiRJklQUG82SJEmSJEmSpKLYaJYkSZIkSZIkFcVGsyRJkiRJkiSpKDaaJUmSJEmSJElFsdEsSZIkSZIkSSqKjWZJkiRJkiRJUlFsNEuSJEmSJEmSimKjWZIkSZIkSZJUFBvNkiRJkiRJkqSi2GiWJEmSJEmSJBXFRrMkSZIkSZIkqSg2miVJkiRJkiRJRbHRLEmSJEmSJEkqio1mSZIkSZIkSVJRbDRLkiRJkiRJkopio1mSJEmSJEmSVBQbzZIkSZIkSZKkotholiRJkiRJkiQVxUazJEmSJEmSJKkoNpolSZIkSZIkSUWx0SxJkiRJkiRJKoqNZkmSJEmSJElSUWw0S5IkSZIkSZKKYqNZkiRJkiRJklQUG82SJEmSJEmSpKLYaJYkSZIkSZIkFcVGsyRJkiRJkiSpKDaaJUmSJEmSJElFsdEsSZIkSZIkSSqKjWZJkiRJkiRJUlFsNEuSJEmSJEmSimKjWZIkSZIkSZJUFBvNkiRJkiRJkqSi2GiWJEmSJEmSJBXFRrMkSZIkSZIkqSgl12iOiDER8eOIWBcRVRGxIiLujYgheccmSZIkSZIkSaWoPO8A2lJEnAI8D4wAHgUWAucAtwOXR8SFKaWtOYYoSZIkSZIkSSWn1K5o/heyJvPfpZTen1L6fErpUuDbwDjg7lyjkyRJkiRJkqQSVDKN5og4GXg3sAL4fqPVXwb2AjdERL8ODk2SJEmSJEmSSlrJNJqBSwvzp1JKtQ1XpJR2A88BfYHzOjowSZIkSZIkSSplpTRG87jCfHEz65eQXfF8GvD00Q523333tVFYkiRJkiRJkpSbdMstt0R7P0kpXdE8qDDf2cz6uuWDm1oZEZ+IiOltHpUkSZIkSZIklbhSuqL5aOq69qmplSml+4H7Oy4cSZIkSZIkSSoNpXRFc90Vy4OaWT+w0XaSJEmSJEmSpDZQSo3mRYX5ac2sP7Uwb24MZ0mSJEmSJEnSMYiUmhxJosuJiFOApcAK4JSUUm2DdQOA9WSN9eEppb25BClJkiRJkiRJJahkrmhOKS0DngLGArc2Wj0J6Af81CazJEmSJEmSJLWtkrmiGeqvan4eGAE8CiwAzgUuIRsy44KU0tb8IpQkSZIkSZKk0lNSjWaAiDgB+ApwOVBJNmTGFGBSSmlbnrFJkiRJkiRJUikquUazJEmSJEmSJKljlcwYzZIkSZIkSZKkfNholiRJkiRJkiQVpTzvADqjsrIeqbbWIUXaUkTQnsO09AjwlEnqLnoAtXkHIZWA9n5/os4pogcpWUW7nQjw9a6OFD3AWlM8X7vdRACe53a2JaU0vL2fxEZzE2prE2l63lGUljg7MXXq1HY7/oQJE0iT2+3wktSpxG2Qxl+TdxhSlxdzHuHqRTflHYY62K/H/SvXvXAw7zDUwX52fi/e9ssleYehbmTGtafCF2ycFe1rwXV/OT/vKNTOfvbzN8L7vpN3GKXt0dtXdsTTOHSGJEmSJEmSJKkoNpolSZIkSZKfGf4aAAAVDklEQVQkSUWx0SxJkiRJkiRJKopjNHcxVQfhU1+H/34Rtu2C14+Br90KV1yYd2Rt46677mLGjBkcOHCAoUOH8ld/9VdcddVVeYclNWnJJnjT1+Cas+DBG/OORlJncfGyZ/jjvm2URwBwfHkfFp1+ec5RSa9a+9gyFk+exf71e6kY1oe3fH0ClWePyjsstZNfXjrksMc1Vft5/Qdv5uzP3ptTROoqqjatYdWPJrJ38SyivBdDzrucEz72RaLMNkKXt2UBPHkrbJgBfYfDpd+AcR/IO6puZ/GSh1i+fAo7di7mdSdeyXnnfq1+3YaNLzB9xlfZt289lZXjOe+cu+nX7/gco+2mXpkKq16E3evg+LfBWz+SLa+thuk/hR2rYf82uPA2GHZqvrF2Ev4L0cVU18AJI+HZ++HEUfD4c/AX/wgvPwxjR+cdXfGuv/567rjjDnr16sXKlSu5/fbbOfXUUxk3blzeoUmvcesv4M9el3cUkjqjyaPP4uOVJ+UdhvQam59by4JvTuNt376UweOHc2DzvrxDUju79nfb63+u3r+XX105hhMv/VCOEamrWPWjifQcOJTx9z9Pzd5dLL7ro2x+8iFGvNcrLLq02mp45H1w1s1w3W9h1bPwy6vhb2ZB5Wl5R9et9OkzgjPe+EnWb3iOmpqq+uVVVdv5n+du55w/+wrHj76EOS9/l+ee/yzvvuzhHKPtpnoPgnHvhk0LoebQ4esqT4ZT3gHTfpJLaJ2VQ2d0Mf36wMRPZk3lHj3gqovgpNEwY0HekbWNk046iV69egEQhSvB1q5dm2dIUpMeng6D+8A7fS8mSepCFn1vJqd96iyGvGUE0SPoM7IffUb2yzssdZBVv/sPKoaMYPhb3p53KOoCDm5azZAL3kuPXhX0HDKcQW+5iP2rl+Qdloq1ZWF2deY5/wA9ymDspTDmQpj7QN6RdTsnjLmMMWPeRUXF4MOWr17zWwYNfD0nnnA5ZWUVvOnMW9mxcxG7dr2SU6Td2Og3w3HjoVej90o9yuGUi6HyFCj0rpSx0dzFbdwKi1fBGafkHUnb+da3vsVll13G9ddfT2VlJeedd17eIUmH2bUfvvQY3PPBvCOR1Fn944aXGTbvP7lw6e95Zs+mvMORAEg1teyYu4Wq7Qd4+rJf8NsJP+PlrzxPzYHqvENTB1n+mwc56YqP1F/QIR3JiPfeyLbnHqO2aj8Ht25g5+ypDHzLhLzDUtFSE4sSbJ7b8aGoSTt3LmXw4Fe/1V1e3pf+/U5g586lOUYltYyN5i7sUDV85E648Uo4fWze0bSdz3zmMzzxxBNMnjyZCRMm1F/hLHUWdz4GN10AJww5+raSup9/Pu5NvHL6Fax9w5V8YuhJXL3ieZZV7ck7LImqLftJh2pZ/8QKLnzoKiZM+QA7529l8X2z8w5NHWDvhlVsnjWVk957Q96hqIvo/8ZzOLB6CbP++ixevvki+p18JoPPuSzvsFSsytOh3wj44zeyoQBeeSobPuOQQyl1FtXV++jZc8Bhy3r2HMCh6r05RSS1nI3mLqq2Fm64E3qVw+Q78o6m7ZWVlTF+/Hg2b97MlClT8g5Hqjd7Dfz3QviHS/KORFJndW7fSgaU9aSiRxk3Dh3LhX0reXz3hrzDkujRO7s9y0k3vJHeI/pSMbQ3J3/sTDY9uzrnyNQRlv/mQYaNv5D+ox0/XkeXamtZ8tW/YfC57+asB1/izT9+keq9u1j74P/JOzQVq6wnXDMFlj4G3x0Ff7oH3vAXMHBM3pGpoLy8L4cOHX6RwqHqPfQsd6grdX7eDLALSgluugs2boPHvwM9S/gs1tTUOEazOpVnlsCKbXDindnjPVVQk2D+epj5+Xxjk9Q5RUBq6muqUgfrNaiC3qP6gaMmdEsrfvMgb7jhf+UdhrqImj07OLR1PSMuv4EePSvo0bOCYZd8iLUPf5sxN5TglU7dzYjxcMOzrz7+9wvgTd7ksbMYNOj1LF/+aP3j6up97NmzmkGDXp9jVFLLeEVzF3TL/4YFy+HX34Y+vfOOpu1s376dp59+mn379lFTU8OLL77I008/zVvf+ta8Q5PqfeJCWDYRZv9jNt38drjyDHjy1rwjk9QZ7Kg5yJO7N3CgtobqVMtD21cxdc8W3tN/VN6hSQCc8MFTWf7AfKq27ufgziqW//s8Rl58Yt5hqZ1tnvMC+zav48RLP5R3KOoiygcOpdeIMWx+6v+Saqqp3ruLrc/8ir6vOz3v0NQWNs2B6gPZcBl//CbsWQ/jP5p3VN1ObW01NTVV1KYaUqrJfq6tZszx72LnriWsXv0UNTVVzJ13H4MHncbAgSfnHXL3U1uTDTGTarOp5lC2DKCmOnt82HZeXFLC18KWppXr4Yf/Dyp6waj3vLr8h1+Aj1yRX1xtISKYMmUK99xzD7W1tYwcOZJPf/rTXHTRRXmHJtXr2yub6vSvgN49YfiA5veR1H0cSol/2jCPhVW7KSM4vfcApoy9gHG9LRLqHE771Fkc3H6A373nEcoqyhh9xUmcesub8w5L7Wz54w9wwsXvp2c/a5Fa7pTPfZ/VP7mbDY/eDz3KGHDGuYz56BfzDktt4eUH4KUfZY2xEy6C634L5RV5R9XtzJv/A+bO+5f6xytW/pozz/gUbzrzNt5+wb1Mn3k3L/zpDiqHjueCC+7JMdJubPFTsOiJVx+vmQ7jLofTr4Cn74b927LlL9yXzS/7EvSt7Pg4O5FIdttfIyJSmp53FKUlzoapU6e22/EnTJhAmtxuh5ekTiVugzT+mrzDkLq8mPMIVy+6Ke8w1MF+Pe5fue6Fg3mHoQ72s/N78bZfLsk7DHUjM649Fb5gv6VoXwuu+8v5eUehdvazn78R3vedvMMobY/ePiOldHZ7P41DZ0iSJEmSJEmSimKjWZIkSZIkSZJUFBvNkiRJkiRJkqSi2GiWJEmSJEmSJBXFRrMkSZIkSZIkqSiRkndBbSwidgOL8o5DOophwJa8g5BawFxVV2CeqiswT9VVmKvqCsxTdQXmqdrKuJTSgPZ+kvL2foIualFK6ey8g5COJCKmm6fqCsxVdQXmqboC81RdhbmqrsA8VVdgnqqtRMT0jngeh86QJEmSJEmSJBXFRrMkSZIkSZIkqSg2mpt2f94BSC1gnqqrMFfVFZin6grMU3UV5qq6AvNUXYF5qrbSIbnkzQAlSZIkSZIkSUXximZJkiRJkiRJUlFsNEuSJEmSJEmSimKjuSAixkTEjyNiXURURcSKiLg3IobkHZu6l0LupWamDc3sc0FEPB4R2yJiX0TMiYi/j4iyjo5fpSUiromI70XEHyJiVyEPHzzKPq3Ox4i4KiKeiYidEbEnIv4UETe2/W+kUtSaPI2IsUeosSkiHj7C89wYES8WcnRnIWevar/fTKUkIioj4uMR8auIWBoR+wt59D8RcVNENPm+3JqqjtTaPLWmKi8R8c8R8XRErC7k6baImBURX46Iymb2sZ6qw7UmV62pak5ra15E9I+IuyJiQUQciIgdhf3f28zxJ0TEAxExNyK2FvZZHhH/GRHvbFWsjtEMEXEK8DwwAngUWAicA1wCLAIuTCltzS9CdScRsQIYDNzbxOo9KaVvNtr+fcB/AAeAnwPbgKuBccAjKaVr2zVglbSImA28GdgDrAFOBx5KKV3fzPatzseIuA34HrC1sM9B4BpgDHBPSulzbfxrqcS0Jk8jYiywHHgJmNLE4eamlB5pYr9vAp8tHP8RoBfwYWAo8OmU0uS2+F1UuiLiZuA+YD3we2AVMBL4IDCIrHZemxq8ObemqqO1Nk+tqcpLRBwEZgLzgU1AP+A84GxgHXBeSml1g+2tp8pFa3LVmqrmtDKPBgN/AM4E5gFPF7b/c2A4cHtK6buNjj8R+FvgT2R5tBc4sbBPf+CrKaU7WxRsSqnbT8CTQCJ7ATZc/q3C8h/kHaNT95mAFcCKFm47sFBkqoCzGyzvTfbhSQI+nPfv5NR1J7IP3E4FAri4kFMPNrNtq/MRGEv2hn8rMLbB8iHA0sI+5+f9d3Dq3FMr83RsYf1PWnH8Cwr7LAWGNDrW1kIOj8377+DUuSfgUrKmRo9Gy0eRNfMS8KEGy62pTh0+HUOeWlOdcpmA3s0sv7uQX//SYJn11Cm3qZW5ak11aos8urew7D+A8gbLh5N9kHEQOLWFxz8e2AjUAMe1JNZuP3RGRJwMvJusuff9Rqu/TNbFvyEi+nVwaFJLXENWLB5OKU2vW5hSOgD8U+HhLXkEptKQUvp9SmlJKvwrcxTHko9/A1QAk1NKKxrssx34WuHhzccYvrqJVubpsajLwbsLuVn3vCvI3jtUAB9rp+dWiUgp/S6l9OuUUm2j5RuAHxQeXtxglTVVHe4Y8vRYWFNVtEItbMovCvNTGyyznio3rczVY2FN7QZamUcfLMy/lFKqbnCMzcA9QE8a1a/mjp9SWkv2gVwP4OSWxNrtG81kn9oDPNXEG6rdwHNAX7JL0qWOUhER10fEFyLi9oi4pJmxw+ry94km1k0F9gEXRERFu0UqvepY8vFI+/ym0TZSWxodEZ8s1NlPRsT4I2xrnqq9HSrMqxsss6aqs2kqT+tYU9VZXF2Yz2mwzHqqzqipXK1jTVVLNZVHowrzV5rYvm5Zi8ZdjogRwLlk3whZ1JJ9yluyUYkbV5gvbmb9ErIrnk8jG9dE6gijgAcaLVseER9LKT3bYFmz+ZtSqo6I5cAZZJ88LWiXSKVXHUs+Hmmf9RGxFxgTEX1TSvvaIWZ1X5cVpnoR8QxwY0ppVYNl/ci+MrYnpbS+ieMsKcxPa6c4VeIiohz468LDhv9JtKaq0zhCntaxpioXEfE5svFDB5GNVfp2sobL1xtsZj1V7lqYq3WsqWpSC/NoC3AccBLZmM4N1V2VfHozxz8buIqsXzyGbIzmgWRDDW9pSYw2mrOTA7CzmfV1ywd3QCwSwL+RDdw+D9hNVghuAz4B/CYizk8pvVTY1vxVZ3Is+diSffoVtvNNvNrCPuAushus1H2iPx6YSDbW89MR8ZaU0t7COuus2tvXyW7W8nhK6ckGy62p6kyay1NrqvL2ObIbVtZ5Avho4Svidayn6gxakqvWVB1NS/Lov8hu7DcxIq5LKdUAREQl8JnCNhUR0SeltL/R8c8mG0a4zm7gYymlxhdCNsuhM44uCvP2GvdROkxKaVJhfLyNKaV9KaW5KaWbyW5O2YfsH5mWMn/VmRxLPprDalMppU0ppS+llGamlHYUpqlk3176E/B64OPHcug2DVTdQkT8Hdld4hcCN7R298Lcmqp2daQ8taYqbymlUSmlIPtG6AfJLtKZFRFvbcVhrKdqdy3JVWuqjqaFNe9LwErgWmB2RNwbEfeTXd1cy6sfjtU0cfwfFI7fB3gj2YWQP42IHzTetjk2ml/9hGdQM+sHNtpOykvdC3tCg2XmrzqTY8nHlu6zq4i4pKMq3CjjR4WHramzR7uSRGpSRNwKfIfsTf8lKaVtjTaxpip3LcjTJllT1dEKF+n8iqwhVwn8tMFq66k6jaPkanP7WFN1mCPlUeEGvn8GfJfsmxefAt5HdqXzu8iayDtTSgePcPwDKaUFKaXbgR8Cn4yIa1oSm43mVwezbm7Mmro7NzY3hrPUUTYV5v0aLGs2fwtj6Z1EdsOWpgaBl9raseTjkfY5jizf1zj2nTpI3VfO6uts4auJa4H+hZxszPcJarWI+HtgMjCXrHm3oYnNrKnKVQvz9EisqepwKaWVZB+MnBERwwqLrafqdJrJ1SOxpuo1msujlNLmlNLtKaWTU0q9UkojU0o3kdW7AKa14mnqbip5cUs2ttEMvy/M3x0Rh/09ImIAcCGwH/hjRwcmNXJ+Yd7wDdDvCvPLm9h+AtAXeD6lVNWegUkFx5KPR9rnikbbSO3tvMK88Ydz5qnaTETcAXwbmE3WvNvUzKbWVOWmFXl6JNZU5WV0YV73tXDrqTqrxrl6JNZUNac1efS3hflDrTj+8YV5dUs27vaN5pTSMuApYCxwa6PVk8g+Lfppg8HWpXYTEWdExNAmlr+O7IoSgAcbrHqE7I6iHy7cHbRu+97AVwsP72uncKXGjiUf/w2oAm6LiLEN9hkCfKHwsMXjQUlHExHnRkSvJpZfCvxD4eGDjVbX5eAXC7lZt89YsvcOVWS5LB1RRNxJdlO1GcA7j3L3bmuqctGaPLWmKg8RcXpEjGpieY+IuBsYQdY43l5YZT1VLlqbq9ZUNeUY8qhHRPRvYvuPA9eRfYj8UKN172h88W1h+SnAFwsPH2tRvCk5JnjhD/c82cl5FFgAnEt2V8/FwAUppa35RajuIiImAp8nu9J+OdkdPk8BrgR6A48DH2g4lk5EvJ/szdMB4GFgG/DnwLjC8r9IvtB1jAr59f7Cw1HAe8g+Rf9DYdmWlNLnGm3fqnyMiE+TjR+1Ffg5cBC4BhgD3NPw+FJTWpOnEfEMcAbwDLCmsH48cGnh5ztTSnX/6Wz4HPeQ3aV5DVku9wL+kmxMtE+nlCY33kdqKCJuBH5CdrXJ92h6vMQVKaWfNNjHmqoO1do8taYqD4VhXb4BTAWWkdW7kcA7yG6MtYHsQ5L5DfaxnqrDtTZXralqyjHkUX9gI/BbYGnhMBcB5xT2f1dKaUWj59gB7CC76eRqoJysF3V54efvpZT+rkXx2n/KRMQJwFfI/oiVwHpgCjCppTe9kIoVEe8AbgbOImuW9CN7sc8GHgAeaKppHBEXkn3KdD5ZQ3op8GPguymllnx9QmpS4cOPLx9hk5UppbGN9ml1PkbE1cDngLeSfdtmPjA5pfTvRf4K6gZak6cRcRPwAeBMYBjQk+yN2AtkOfeH5g5SaMDcRnYH5lpgJvCNlNJ/Ff9bqNS1IE8Bnk0pXdxoP2uqOkxr89SaqjxExJnALWTDXI4BBgN7yS4Se4ysPr7m//DWU3W01uaqNVVNOYY86kl2pfvbC9tD1mB+BPhWSmlPE89xO9mNBc8EhgNlZLn3J+BHKaUnWxyvjWZJkiRJkiRJUjG6/RjNkiRJkiRJkqTi2GiWJEmSJEmSJBXFRrMkSZIkSZIkqSg2miVJkiRJkiRJRbHRLEmSJEmSJEkqio1mSZIkSZIkSVJRbDRLkiRJkiRJkopio1mSJEmSJEmSVBQbzZIkSZIkSZKkotholiRJkiRJkiQV5f8D5uJ+8XcjTh0AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 1728x576 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Plotting the predicted fitness effect for the variants seen more than once\n",
    "plt.close('all')\n",
    "f, (ax1, ax2) = plt.subplots(2, 1, sharey=True, sharex = True, figsize=(24, 8))\n",
    "gs = matplotlib.gridspec.GridSpec(2, 1, width_ratios=[1], height_ratios=[10,0.5])\n",
    "ax1 = plt.subplot(gs[0])\n",
    "ax2 = plt.subplot(gs[1])\n",
    "gs.update(hspace=0.05)\n",
    "\n",
    "nonsense_list = []\n",
    "y_list = []\n",
    "x_list=[]\n",
    "y_list_top20 = []\n",
    "x_list_top20=[]\n",
    "y_list_nonsense = []\n",
    "x_list_nonsense = []\n",
    "x=[]\n",
    "y=[]\n",
    "OBS_EXP_DICT={}\n",
    "top20_lower_err = []\n",
    "top20_upper_err = []\n",
    "missense_lower_err = []\n",
    "missense_upper_err = []\n",
    "nonsense_lower_err = []\n",
    "nonsense_upper_err = []\n",
    "\n",
    "top20 = c1\n",
    "missense = c6\n",
    "nonsense = c10\n",
    "synonymous = c3\n",
    "\n",
    "axisfont=20\n",
    "domainfont = 12\n",
    "titlefont=12\n",
    "axislabelfont=12\n",
    "\n",
    "top_20_TP53 = []\n",
    "\n",
    "for k, v in TP53_missense_s.items():\n",
    "    if k in top_20_TP53:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_top20.append(codon_number)\n",
    "        y_list_top20.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        top20_lower_err.append(error_lower)\n",
    "        top20_upper_err.append(error_upper)\n",
    "\n",
    "    if k not in top_20_TP53:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list.append(codon_number)\n",
    "        y_list.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        missense_lower_err.append(error_lower)\n",
    "        missense_upper_err.append(error_upper)\n",
    "\n",
    "    missense_err = [tuple(missense_lower_err), tuple(missense_upper_err)]   \n",
    "    \n",
    "for k, v in TP53_nonsense_s.items():\n",
    "    if k in top_20_TP53:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_top20.append(codon_number)\n",
    "        y_list_top20.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        top20_lower_err.append(error_lower)\n",
    "        top20_upper_err.append(error_upper)\n",
    "    \n",
    "    top20_err = [tuple(top20_lower_err), tuple(top20_upper_err)]\n",
    "        \n",
    "    if k not in top_20_TP53:\n",
    "        codon_number = int(((re.findall('\\d+', k)))[0])\n",
    "        s_value = float(v[0])*100\n",
    "        x_list_nonsense.append(codon_number)\n",
    "        y_list_nonsense.append(s_value)\n",
    "        \n",
    "        s_upper = float(v[3])*100\n",
    "        s_lower = float(v[4])*100\n",
    "        error_upper = s_upper-s_value\n",
    "        error_lower = s_value-s_lower\n",
    "        nonsense_lower_err.append(error_lower)\n",
    "        nonsense_upper_err.append(error_upper)\n",
    "\n",
    "    nonsense_err = [tuple(nonsense_lower_err), tuple(nonsense_upper_err)]        \n",
    "    \n",
    "msize = 12\n",
    "medgewidth = 2\n",
    "linew = 3\n",
    "csize = 4\n",
    "cthick = 3\n",
    "\n",
    "markers1, caps1, bars1 = ax1.errorbar(x_list, y_list, yerr = missense_err, fmt = 'o', ecolor = missense, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = missense, markerfacecolor = missense, zorder = 1)    \n",
    "\n",
    "markers2, caps2, bars2 = ax1.errorbar(x_list_nonsense, y_list_nonsense, yerr = nonsense_err, fmt = 'o', ecolor = nonsense, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = nonsense, markerfacecolor = nonsense, zorder = 2)    \n",
    "\n",
    "markers3, caps3, bars3 = ax1.errorbar(x_list_top20, y_list_top20, yerr = top20_err, fmt = 'o', ecolor = top20, elinewidth = linew, \\\n",
    "             capsize = csize, capthick = cthick, markersize = msize, markeredgewidth = medgewidth, \\\n",
    "             markeredgecolor = top20, markerfacecolor = top20, zorder = 3)    \n",
    "\n",
    "# loop through bars and caps and set the alpha value\n",
    "a = 0.7\n",
    "[bar.set_alpha(a) for bar in bars1]\n",
    "[cap.set_alpha(a) for cap in caps1]\n",
    "[bar.set_alpha(a) for bar in bars2]\n",
    "[cap.set_alpha(a) for cap in caps2]\n",
    "[bar.set_alpha(a) for bar in bars3]\n",
    "[cap.set_alpha(a) for cap in caps3]\n",
    "\n",
    "#Only show the required axis lines\n",
    "ax1.spines['top'].set_visible(False)\n",
    "ax1.spines['right'].set_visible(False)\n",
    "\n",
    "for axis in ['bottom','left']:\n",
    "    ax1.spines[axis].set_linewidth(2)\n",
    "\n",
    "for axis in ['bottom','left']:\n",
    "    ax1.spines[axis].set_color(grey3)\n",
    "    \n",
    "ax1.set_ylabel('fitness effect (s)', fontsize = axisfont, labelpad = 10)\n",
    "ax1.set_title('Fitness effects across TP53', fontsize = 18)\n",
    "\n",
    "# ax1.set_yscale('log')\n",
    "\n",
    "# STACKED BAR CHART\n",
    "# define the exons + create data\n",
    "exons = [{'2':(25-1)+1, '3': (32-26)+1, '4': (125-33)+1, '5':(187-126)+1, '6':(224-188)+1, '7': (261-225)+1, '8': (307-262)+1,'9': (331-308)+1, '10': (367-332)+1, '11': (393-368)+1}]\n",
    "exon_orders = [['2', '3', '4', '5', '6', '7', '8', '9', '10', '11']]\n",
    "colors2 = [c4, c0, c3, c1, c5, c6,c7, c8, c9, c10]\n",
    "names = exons[0].keys()\n",
    "values = np.array([[data[name] for name in order] for data,order in zip(exons, exon_orders)])\n",
    "lefts = np.insert(np.cumsum(values, axis=1),0,0, axis=1)[:, :-1]\n",
    "orders = np.array(exon_orders)\n",
    "\n",
    "#plot the bar chart\n",
    "for name, color in zip(names, colors2):\n",
    "    idx = np.where(orders == name)\n",
    "    value = values[idx]\n",
    "    left = lefts[idx]\n",
    "    ax2.bar(x=left, height=-10, width=value, bottom=-5, color=color, orientation = 'horizontal', label = name, edgecolor='k', linewidth=1, tick_label = name)\n",
    "    ax2.text(left + (value/2.), (-0.26), name, ha='center', va='center', fontsize = 12)\n",
    "    \n",
    "# CONFIGURING THE GRAPH APPEARANCE\n",
    "#Set the x and y axis limits\n",
    "ax1.set_xlim(0, 393)\n",
    "ax1.set_ylim(0, 35)\n",
    "# ax1.set_ylim(10**1, 10**5)\n",
    "ax2.set_xlim(0, 393)\n",
    "ax2.set_ylim(-0.5, 0)\n",
    "\n",
    "#x-axis ticks\n",
    "x1_major_ticks = []\n",
    "x1_major_tick_labels = []\n",
    "x2_major_ticks = [0, 50, 100, 150, 200, 250, 300, 350, 393]\n",
    "x2_major_tick_labels = [\"0\", \"50\", \"100\", \"150\", \"200\", \"250\", \"300\", \"350\", \"393\"]\n",
    "ax1.set_xticks(x1_major_ticks)\n",
    "ax1.set_xticklabels(x1_major_tick_labels, fontsize = axisfont)\n",
    "ax2.set_xticks(x2_major_ticks)\n",
    "ax2.set_xticklabels(x2_major_tick_labels, fontsize = axisfont)\n",
    "\n",
    "#y-axis ticks\n",
    "y1_major_ticks = [0, 5, 10, 15, 20, 25, 30, 35]\n",
    "y1_major_tick_labels = [\"0\", \"5\", \"10\", \"15\", \"20\", \"25\", \"30\", \"35\"]\n",
    "ax1.set_yticks(y1_major_ticks)\n",
    "ax1.set_yticklabels(y1_major_tick_labels, fontsize = axisfont)\n",
    "\n",
    "#set panel legends\n",
    "legend_marker_size = 2\n",
    "legend_elements = [Line2D([0], [0], marker = '_', color = missense, alpha = 1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='missense'),\n",
    "                   Line2D([0], [0], marker = '_', color=nonsense, alpha=1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='nonsense'),\n",
    "                  Line2D([0], [0], marker = '_', color=top20, alpha=1.0, markersize = legend_marker_size, \\\n",
    "                          lw=3, label='top 20 observed variant')]\n",
    "    \n",
    "ax1.legend(ncol=1, handles=legend_elements, loc='upper left', frameon=0, fontsize = 20)\n",
    "\n",
    "ax1.grid(which='major', linestyle=':', linewidth='0.7', color=grey3, zorder = 0)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.0"
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": true,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": true
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
