This repository contains the code and necessary files to accompany the manuscript **"The Evolutionary Dynamics and Fitness Landscape of Clonal Hematopoiesis"**.  Code was written in Jupyter notebook using Python. 
<br>
<br>
If there are problems with the code (.ipynb) pages rendering, please try viewing them here: https://nbviewer.jupyter.org/github/blundelllab/ClonalHematopoiesis/tree/master/The%20Evolutionary%20Dynamics%20and%20Fitness%20Landscape%20of%20Clonal%20Hematopoiesis%20%282020%29/
<br>
<br>
## Navigation: 
### Code & data for generation of figures/ analyses in main text:
#### Figure 1:
- **Figure 1b (distribution of variants across DNMT3A)**:
  - Github folder: *Distribution of variants across DNMT3A*.
- **Figure 1d (simulated data)**:
  - Github folder: *Simulations*.
- **Figure 1e (VAF-density histrogram)**: 
  - Github folder: *Maximum likelihood estimations/ DNMT3A R882H, DNMT3A nonsynonymous, all synonymous*.
#### Figure 2:
- **Figure 2a (fitness effects & mutation rates for top 20 most commonly observed CH variants)**:
  - Github folder: *Maximum likelihood estimations/ Fitness landscape of top 20 most commonly observed variants in CH*.
- **Figure 2b (distribution of fitness effects in key CH driver genes)**:
  - Github folder: *Maximum likelihood estimations/ Fitness landscape of key CH driver genes (DNMT3A, TET2, ASXL1, TP53)*.
#### Figure 3: predicted prevalence of CH as a function of age):
  - Github folder: *Predicted prevalence of CH and double mutations*.

### Code & data for generation of figures/ analyses in supplement:
#### Supplementary Methods 1 (data trimming):
- Github folder: *Trimming the data*.
#### Supplementary Methods 5 (study-specific mutation rates):
- Github folder: *Mutation rate calculations*.
#### Supplementary Methods 6 (emergence of clones with multiple mutations):
- Github folder: *Predicted prevalence of CH and double mutations*.
#### Supplementary Methods 7 (maximum likelihood parameter estimation):
- **Figure S9 (DNMT3A R882H)**:
  - Github folder: *Maximum likelihood estimations/ DNMT3A R882H, DNMT3A nonsynonymous, all synonymous*.
- **Figure S10 (top 20 most commonly observed variants)**:
  - Github folder: *Maximum likelihood estimations/ Fitness landscape of top 20 most commonly observed variants*.
- **Figure S11 (R882H vs R882HC)**:
  - Github folder: *Maximum likelihood estimations/ Fitness of DNMT3A R882H vs R882C*.
- **Figure S12 (nonsynonymous variants parameter estimation)**:
  - DNMT3A:
    - Github folder: *Maximum likelihood estimations/ DNMT3A R882H, DNMT3A nonsynonymous, all synonymous*.
  - TET2, ASXL1, TP53:
    - Github folder: *Maximum likelihood estimations/ Fitness landscape of key CH driver genes (DNMT3A, TET2, ASXL1, TP53)*.
- **Figure S13 (synoymous variants parameter estimation)**:
   - Github folder: *Maximum likelihood estimations/ DNMT3A R882H, DNMT3A nonsynonymous, all synonymous*.
#### Supplementary Methods 8 (developmental mutations and neutral 'hitchhikers'):
- **Figure S16 (hitchhiker variants parameter estimation)**:
  - Github folder: *Maximum likelihood estimations/ DNMT3A R882H, DNMT3A nonsynonymous, all synonymous*.
#### Supplemetary Methods 9 (odds ratio of AML stratified by variant fitness):
- Github folder: *Highly fit variants are enriched in pre-AML blood samples*.
#### Supplementary Methods 10 (age prevalence of R882H and R882C mutations):
- Github folder: *Age prevalence of DNMT3A R882H and R882C variants*.
#### Supplementary Methods 11 (parameter estimation for 10 commonly mutated CH genes):
- Github folder: *Predicted prevalence of CH and double mutants*.
#### Supplemenetary Methods 12: (Estimating fitness effects of infrequently mutated sites)
- Github folder: *Estimating fitness effects of infrequently mutated sites*.
#### Supplemenetary Methods 13: (Limitations of study size and sequencing limit)
- Github folder: *Limitations of study size and sequencing limit*.

## LICENSE
[GNU General Public License v3.0](https://choosealicense.com/licenses/gpl-3.0/)
